home.38_06_06_130655 <- path.expand("~")
hs.home.expand <- function(fp1, q = 0) {
    for (i in seq_along(fp1)) {
        a <- fp1[i]
        if ((a == "~") || (substr(a, 1, 2) == "~/")) 
            fp1[i] <- paste0(home.38_06_06_130655, substring(a, 2))
    }
    if (q) 
        fp1 <- shQuote(fp1)
    fp1
}
hs.machine.isLocal <- function() {
    file.exists("~/wk/db/这里是本地")
}
wk.dm.wjj <- switch(Sys.info()["nodename"], node9 = "/dev/shm/wk/dm/r", hs.home.expand("~/org/dm/r"))
hs.getTextFile <- function(wj1) {
    wj1 <- hs.home.expand(wj1)
    if (file.exists(wj1)) 
        return(wj1)
    wj2 <- paste0(wj1, ".xz")
    if (file.exists(wj2)) 
        return(wj2)
    wj2 <- paste0(wj1, ".gz")
    if (file.exists(wj2)) 
        return(wj2)
}
hs.readLines.robust <- function(wj) {
    con1 <- gzfile(hs.getTextFile(wj))
    on.exit(close(con1))
    unlist(strsplit(readChar(con1, 1e+07), "[\r\n]+"))
}
hs.source.robust <- function(wj, env = .GlobalEnv) {
    eval(expr = parse(text = hs.readLines.robust(wj = hs.getTextFile(wj1 = wj))), envir = env)
}
hs.gre <-
structure(function (s1, p1, r1 = {
}, match = 0, perl = 1, ignore.case = 1, useBytes = 0, fixed = 0, no_match = {
}, ...) 
{
    m1 <- gregexpr(p1, s1, perl = as.logical(perl), ignore.case = as.logical(ignore.case), fixed = as.logical(fixed), useBytes = as.logical(useBytes))
    if (match) 
        return(m1)
    0L
    if (length(r1)) {
        hs.regmatches(s1, m1, no_match = no_match, replacement = r1, ...)
        s1
    }
    else hs.regmatches(s1, m1, no_match = no_match, ...)
}, usage = "【功能】提取或替换模式. s1: string. p1: pattern. r1: replacement. match: if true, return regmatches.")
hs.grep <-
function (x, pattern, fixed = FALSE, perl = !fixed, ...) 
grep(pattern, x, ..., perl = perl, fixed = fixed)
hs.grepl <-
function (x, pattern, fixed = FALSE, perl = !fixed, ...) 
grepl(pattern, x, ..., perl = perl, fixed = fixed)
hs.grepV <-
function (..., hs0 = {
}) 
{
    if (length(hs0)) {
        xi <- match("x", ...names(), nomatch = 1)
        x <- hs0(...elt(xi))
        L <- list(...)
        L[[xi]] <- x
        i <- do.call(hs.grep, L)
        ...elt(xi)[i]
    }
    else hs.grep(..., v = 1)
}
hs.gsub <-
function (x, pattern, replacement = "", fixed = FALSE, perl = !fixed, ...) 
gsub(pattern, x, replacement = replacement, ..., perl = perl, fixed = fixed)
hs.re <-
structure(function (s1, p1, r1 = {
}, match = 0, perl = 1, ignore.case = 1, useBytes = 0, fixed = 0, no_match = {
}, ..., after = FALSE) 
{
    m1 <- regexpr(p1, s1, perl = as.logical(perl), ignore.case = as.logical(ignore.case), fixed = as.logical(fixed), useBytes = as.logical(useBytes))
    if (match) 
        return(m1)
    if (after) 
        return(substring(s1, m1 + attr(m1, "match.length")))
    if (length(r1)) {
        hs.regmatches(s1, m1, no_match = no_match, replacement = r1, ...)
        s1
    }
    else hs.regmatches(s1, m1, no_match = no_match, ...)
}, usage = "【功能】提取或替换模式. s1: string. p1: pattern. r1: replacement. match: if true, return regmatches.")
hs.sub <-
function (x, pattern, replacement = "", fixed = FALSE, perl = !fixed, ...) 
sub(pattern, x, replacement = replacement, ..., perl = perl, fixed = fixed)
hs.colAlls <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colAlls"(x, na.rm = na.rm, ...)
}
hs.colAnys <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colAnys"(x, na.rm = na.rm, ...)
}
hs.colCumsums <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colCumsums"(x, na.rm = na.rm, ...)
}
hs.colCvs <-
function (x) 
{
    Mean <- hs.colMeans(x)
    Sd <- hs.colSds(x)
    Sd/Mean
}
hs.colMads <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colMads"(x, na.rm = na.rm, ...)
}
hs.colMaxs <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colMaxs"(x, na.rm = na.rm, ...)
}
hs.colMeans <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colMeans2"(x, na.rm = na.rm, ...)
}
hs.colMedians <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colMedians"(x, na.rm = na.rm, ...)
}
hs.colMins <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colMins"(x, na.rm = na.rm, ...)
}
hs.colProds <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colProds"(x, na.rm = na.rm, ...)
}
hs.colSds <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colSds"(x, na.rm = na.rm, ...)
}
hs.colSums <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colSums2"(x, na.rm = na.rm, ...)
}
hs.colVars <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"colVars"(x, na.rm = na.rm, ...)
}
hs.rowAlls <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowAlls"(x, na.rm = na.rm, ...)
}
hs.rowAnys <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowAnys"(x, na.rm = na.rm, ...)
}
hs.rowCumsums <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowCumsums"(x, na.rm = na.rm, ...)
}
hs.rowCvs <-
function (x) 
{
    Mean <- hs.rowMeans(x)
    Sd <- hs.rowSds(x)
    Sd/Mean
}
hs.rowMads <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowMads"(x, na.rm = na.rm, ...)
}
hs.rowMaxs <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowMaxs"(x, na.rm = na.rm, ...)
}
hs.rowMeans <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowMeans2"(x, na.rm = na.rm, ...)
}
hs.rowMedians <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowMedians"(x, na.rm = na.rm, ...)
}
hs.rowMins <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowMins"(x, na.rm = na.rm, ...)
}
hs.rowProds <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowProds"(x, na.rm = na.rm, ...)
}
hs.rowSds <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowSds"(x, na.rm = na.rm, ...)
}
hs.rowSums <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowSums2"(x, na.rm = na.rm, ...)
}
hs.rowVars <-
function (x, na.rm = TRUE, ...) 
{
    if (!inherits(x, c("matrix", "Matrix"))) 
        x <- as.matrix(x)
    MatrixGenerics::"rowVars"(x, na.rm = na.rm, ...)
}
hs.grepV <- function(..., hs0 = {
}) {
    if (length(hs0)) {
        xi <- match("x", ...names(), nomatch = 1)
        x <- hs0(...elt(xi))
        L <- list(...)
        L[[xi]] <- x
        i <- do.call(hs.grep, L)
        ...elt(xi)[i]
    }
    else hs.grep(..., v = 1)
}
hs.grep_sub <- function(v, p, r = "", ...) {
    a <- hs.grepV(v, p, ...)
    hs.sub(a, p, r, ...)
}
hs.re.capture <- local({
    match.hs.capture <- function(m1, c1) {
        m2 <- attr(m1, "capture.start")[, c1]
        attr(m2, "match.length") <- attr(m1, "capture.length")[, c1]
        m2
    }
    function(s1, p1, c1 = {
    }, all = 0) {
        m1 <- (if (all) 
            gregexpr
        else regexpr)(p1, s1, perl = TRUE)
        if (is.list(m1)) 
            "capture.start" %in% names(attributes(m1[[1]]))
        cn.db <- attr(if (all) m1[[1]] else m1, "capture.names")
        c1 <- if (length(c1)) {
            if (is.character(c1)) {
                if (any(c1 %not.in% cn.db)) 
                  hs.browser("37.04.15.182833")
                match(c1, cn.db)
            }
            else if (is.numeric(c1)) {
                if (any(c1 %not.in% seq_along(cn.db))) 
                  hs.browser("37.04.15.184825")
                c1
            }
            else hs.browser("37.04.15.182809")
        }
        else seq_along(cn.db)
        jg <- lapply(c1, function(c1) {
            cm <- if (all) 
                lapply(m1, match.hs.capture, c1)
            else match.hs.capture(m1, c1)
            regmatches(s1, cm)
        })
        names(jg) <- cn.db
        jg
    }
})
hs.has_x <- function(E) {
    if (length(E) > 1) {
        any(sapply(E, hs.has_x))
    }
    else identical(E, quote(x))
}
"%=>%" <- local({
    hs_blank <- function(x, ...) {
    }
    add_x <- function(E) {
        if (is.symbol(E)) 
            return(as.call(list(E, quote(x))))
        if (E[[1]] == quote(`[`)) {
            return(as.call(append(as.list(E), quote(x), 1)))
        }
        else as.call(append(as.list(E), quote(x), after = 1))
    }
    function(x, e, env = parent.frame(), assign = 0, substitute = !assign, out = 0) {
        if (substitute) {
            e <- substitute(e)
            x <- substitute(x)
        }
        if ((length(e) == 1) && is.character(e)) {
            e <- as.symbol(e)
        }
        assign_back <- local({
            right <- list()
            while (is.call(x)) {
                if (identical(x[[1]], quote(`%=>%`))) {
                  right <- list(x[[3]], right)
                  x <- x[[2]]
                }
                else if (identical(x[[1]], quote(`%<=>%`))) {
                  right <- list(x[[3]], right)
                  return(list(x[[2]], right))
                }
                else return()
            }
        })
        if (length(assign_back)) {
            right <- assign_back[[2]]
            call_1 <- assign_back[[1]]
            while (length(right)) {
                call_1 <- call("%=>%", call_1, right[[1]])
                right <- right[[2]]
            }
            call_1 <- call("%=>%", call_1, e)
            y <- eval(call_1, envir = env)
            return(eval(as.call(list(quote(`<-`), assign_back[[1]], y)), envir = env))
        }
        if (is.symbol(x) && (x == quote(.))) {
            if (is.symbol(e) && is.function(e)) {
                return(e)
            }
            else {
                if (!hs.has_x(e)) 
                  e <- add_x(e)
                body(hs_blank) <- if (e[[1]] == quote(`{`)) 
                  bquote(.(e))
                else bquote({
                  .(e)
                })
                environment(hs_blank) <- env
                return(hs_blank)
            }
        }
        start_with_. <- local({
            right <- list()
            while (is.call(x)) {
                if (identical(x[[1]], quote(`%=>%`))) {
                  right <- list(x[[3]], right)
                  if (x[[2]] == quote(.)) {
                    return(right)
                  }
                  x <- x[[2]]
                }
                else return()
            }
        })
        if (length(start_with_.)) {
            call_1 <- quote(x)
            while (length(start_with_.)) {
                call_1 <- call("%=>%", call_1, start_with_.[[1]])
                start_with_. <- start_with_.[[2]]
            }
            call_1 <- call("%=>%", call_1, e)
            body(hs_blank) <- call_1
            environment(hs_blank) <- env
            return(hs_blank)
        }
        if (is.symbol(e) && do.call(is.function, list(e), envir = env) || is.call(e) && {
            e1 <- paste(e[[1]])
            (length(e1) == 1) && (e1 %in% c("hs.hsgly", "::"))
        }) {
            e <- bquote(.(e)(x))
        }
        else if (inherits(e, c("{", "("))) {
        }
        else if (!hs.has_x(e)) 
            e <- add_x(e)
        if (assign) {
            hs1 <- if (out) 
                `<<-`
            else `<-`
            eval(bquote(.(hs1)(.(x), with(list(x = .(x)), .(e)))), envir = env)
        }
        else {
            eval(bquote(with(list(x = .(x)), .(e))), envir = env)
        }
    }
})
"%<=>%" <- function(x, e, env = parent.frame()) {
    `%=>%`(x = substitute(x), e = substitute(e), env = env, assign = 1)
}
hs.with <- function(..., env.2023_07_21_174408 = parent.frame()) {
    A <- substitute(list(...))
    data <- head(A, -1)
    if (is.null(names(data))) {
        if (length(data) == 2) {
            names(data)[2] <- "x"
        }
        else if (length(data) > 2) {
            names(data)[2:length(data)] <- paste0("x", seq(length(data) - 1))
        }
    }
    e1 <- A[[length(A)]]
    if ((length(data) == 2) && (length(e1) == 1) && is.symbol(e1) && eval(bquote(is.function(.(e1))), envir = env.2023_07_21_174408)) 
        e1 <- bquote(.(e1)(x))
    e1 <- bquote(with(data = .(data), expr = .(e1)))
    eval(e1, envir = env.2023_07_21_174408)
}
hs.pipe1 <- function(..., env.2023_07_24_205312 = parent.frame()) {
    A <- substitute(list(...))
    hs.switch(...length(), 1, ..1, 0, {
    }, {
        x <- ..1
        for (i1 in seq(3, length(A))) x <- do.call("%=>%", list(x, A[[i1]]), envir = env.2023_07_24_205312)
        x
    })
}
hs.update <- function(x, e, hs.if = function(x) TRUE, env = parent.frame(), n = (length(x) == 1) && inherits(x, "character"), out = FALSE) {
    hs1 <- substitute(hs.if)
    target <- substitute(x)
    env
    do <- if ((is.name(hs1) && is.function(hs.if)) || (paste(as.list(hs1)[[1]]) == "function")) {
        eval(bquote(.(hs.if)(.(if (is.symbol(target)) 
            target
        else x))), envir = env)
    }
    else {
        eval(bquote(with(list(x = .(x)), .(hs1))), envir = env)
    }
    if (do) {
        exp <- substitute(e)
        if (is.name(exp) && is.function(e)) {
            exp <- bquote(.(exp)(x))
        }
        else if (hs.has_x(exp)) 
            exp <- bquote(with(list(x = .(target)), .(exp)))
        e1 <- bquote(.(if (out) 
            `<<-`
        else `<-`)(.(target), .(exp)))
        eval(e1, envir = env)
        return(TRUE)
    }
    return(FALSE)
    eval(bquote(.(hs1)(.(x))), envir = env)
    target <- substitute(x)
    eval(bquote(.(hs1)(.(target))), envir = env)
    if (identical(hs.if, missing)) {
        target <- substitute(x)
        eval(bquote(missing(.(target))), envir = env)
    }
    else {
        if (inherits(x, "character")) 
            x <- as.symbol(x)
        target <- substitute(x)
        if (!hs1(eval(target, envir = env))) 
            return("没更新")
    }
    exp <- substitute(e)
    if (is.name(exp) && is.function(e)) 
        exp <- bquote(.(exp)(x))
    e1 <- bquote(.(if (out) 
        `<<-`
    else `<-`)(.(target), .(if (is.symbol(target) && (!exists(paste(target), envir = env))) {
        substitute(exp)
    }
    else substitute(with(list(x = target), exp)))))
    invisible(eval(e1, envir = env))
}
"%<<=>%" <- function(x, e, env = parent.frame()) `%=>%`(x = substitute(x), e = substitute(e), env = env, assign = 1, out = 1)
..x <- function(up = 1) {
    env1 <- parent.frame()
    while (up) {
        env1 <- parent.env(env1)
        up <- up - 1
    }
    get("x", envir = env1)
}
hs.libPaths <- function(fp) {
    .libPaths(unique(c(fp, .libPaths())))
}
wk.true <- !0
wk.false <- !1
wk.null <- {
}
wk.t <- "\t"
wk.n <- "\n"
hs.havePackages <- function(...) {
    all(c(...) %in% .packages(1, .libPaths()))
}
hs.browser <- wk.browser <- function(text = "", condition = NULL, expr = TRUE, skipCalls = 0L, env = parent.frame(), debug = 0) {
    eval(bquote({
        if (.(if (debug) 
            debug4browse.38_09_01_174554
        else 1)) {
            message(rep("\n", 3), "browse at ", .(text), "\n")
            browser(.(text), .(condition), .(expr), .(skipCalls))
        }
    }), env)
}
mirror_db <- c(同济 = "https://mirrors.tongji.edu.cn/CRAN/", 上海交大 = "https://mirrors.sjtug.sjtu.edu.cn/cran/", 南大 = "https://mirrors.nju.edu.cn/CRAN/", 清华 = "https://mirrors.tuna.tsinghua.edu.cn/CRAN/", 兰大 = "https://mirror.lzu.edu.cn/CRAN/", `0-Cloud-East-Asia` = "https://cran.asia/")
wk.lib <- file.path(R.home(), "library")
wk.home <- path.expand("~")
wk.lib.mine <- file.path(wk.home, ifelse(length(grep("(?i)virtualbox", Sys.info()["nodename"])), "rj2/lib/R", "rj/lib/R"))
wk_node_local <- c("wk-VirtualBox", "wk-mbp-2017-13.local", "kai-PC", "wk-NBLK-WAX9X", "calvin_mpb", "wk-ThinkPad-X220", "wk-MS-7A21", "wk-HP-EliteBook-Revolve-810-G2")
hs.wj.is.Symlink <- function(wj1) {
    wj1 %=>% sub("/*$", "", x) %=>% {
        file.exists(x) && (path.expand(x) != normalizePath(x))
    }
}
hs.with_wd <- function(wd = ".", exp1, skip_if_wd_not_exist = 0, local = 1, env1 = parent.frame(), temp = 0, temp.rm = 1) {
    if (skip_if_wd_not_exist && (!file.exists(wd))) 
        return()
    if (!dir.exists(wd)) 
        wd <- dirname(wd)
    if (temp) {
        wd %<=>% hs.wjj(.muid4machine())
        if (temp.rm) 
            on.exit(hs.wj.rm(wd, 1), add = TRUE)
    }
    if (!identical(normalizePath("."), normalizePath(wd))) {
        wjj.37_11_13_163320 <- getwd()
        dir.create(wd, recursive = TRUE)
        setwd(wd)
        on.exit(setwd(wjj.37_11_13_163320), add = TRUE)
    }
    eval(substitute(exp1), envir = if (local) 
        new.env(parent = env1)
    else env1)
}
hs.with_wd_temp <- function(wjj0 = ".", ...) {
    if (skip_if_wd_not_exist && (!file.exists(wd))) 
        return()
    if (!dir.exists(wd)) 
        wd <- dirname(wd)
    if (!identical(normalizePath("."), normalizePath(wd))) {
        wjj.37_11_13_163320 <- getwd()
        dir.create(wd, recursive = TRUE)
        setwd(wd)
        on.exit(setwd(wjj.37_11_13_163320))
    }
    eval(substitute(exp1), envir = if (local) 
        new.env(parent = env1)
    else env1)
}
hs.c <- function(name, value) {
    if (missing(value)) {
        value <- name
        a <- substitute(name)
        name <- as.character(a)
        if (class(a) == "call") 
            name <- name[-1]
    }
    names(value) <- name
    value
}
hs.c1 <- function(..., recursive = FALSE, use.names = FALSE) {
    a <- list(...)
    hs.c(names(a), hs.unlist(a, recursive = recursive, use.names = use.names))
}
hs.db <- function(x, inv = FALSE, ot = "nv") {
    if (ot != "nv") 
        hs.browser("2023.07.29.113830", debug = 0)
    x_ <- substitute(x)
    if (inherits(x_, "call")) 
        x <- as.list(substitute(x)[-1])
    if (!length(names(x))) 
        x <- if (inherits(x, "list")) 
            hs.c(unlist(x[seq(1, length(x), by = 2)]), unlist(x[seq(2, length(x), by = 2)]))
        else hs.c(x[seq(1, length(x), by = 2)], x[seq(2, length(x), by = 2)])
    if (inv) 
        x <- hs.c(x, names(x))
    x
}
hs.normalizePath <- function(fp) {
    suppressWarnings(normalizePath(fp))
}
hs.wjm.home <- local({
    h1 <- sprintf("^(%s$|%s(?=/))", home.38_06_06_130655, home.38_06_06_130655)
    Map <- c("~/org", "~/xt", "~/swxx", "~/1", "~") %=>% hs.c(normalizePath(x), x)
    need_abbreviation <- function(x, before) {
        y <- hs.normalizePath(x)
        L <- nchar(before)
        if ((nchar(y) == L) && (y == before)) 
            return(TRUE)
        if ((nchar(y) > L) && (substr(y, 1, L + 1) == paste0(before, "/"))) 
            return(TRUE)
        return(FALSE)
    }
    abbreviate <- function(x, before, after) paste0(after, substring(hs.normalizePath(x), nchar(before) + 1))
    function(wj1) {
        for (h in seq_along(wj1)) {
            x <- wj1[h]
            for (i in seq_along(Map)) {
                before <- names(Map)[i]
                if (need_abbreviation(x, before = before)) {
                  wj1[h] <- abbreviate(x, before = before, after = Map[i])
                  break
                }
            }
        }
        wj1
    }
})
hs.fp.components <- function(fp1) {
    jg <- rep(NA, length(gregexpr("/", fp1, fixed = TRUE)[[1]]) + 1)
    i1 <- length(jg)
    hs.browser("38.04.28.225512")
    repeat {
        if (grepl("^~/*$", fp1, perl = TRUE)) {
            jg[i1] <- "~"
            break
        }
        if (fp1 == "/") {
            jg[i1] <- ""
            break
        }
        bn1 <- basename(fp1)
        if (nzchar(bn1)) {
            jg[i1] <- bn1
            fp1 <- substr(fp1, 1, nchar(fp1) - nchar(bn1))
            basename("~/a/\\/")
        }
        else break
        i1 <- i1 - 1
        if (i1 < 0) 
            hs.browser("38.04.28.221919")
    }
    if (i1 > 1) 
        tail(jg, 1 - i1)
    else jg
}
hs.wj.symlink <- function(wj1, wjj1 = hs.home.expand("~/1/常用"), name = basename(wj1), R = 0, F = 0, ot = "") {
    hs.with_wd(wjj1, {
        wj2 <- file.path(wjj1, name)
        if (F) {
            unlink(wj2)
        }
        if (R) {
            a1 <- strsplit(c(wj1, wjj1), "/")
            depth.min <- min(sapply(a1, length))
            i1 <- 0
            repeat {
                i1 <- i1 + 1
                if (hs.any(i1 > depth.min, a1[[1]][i1] != a1[[2]][i1])) {
                  i1 <- i1 - 1
                  break
                }
            }
            wj1 <- paste(c(rep("..", length(a1[[2]]) - i1), tail(a1[[1]], -i1)), collapse = "/")
        }
        switch(ot, wj1 = wj1, file.symlink(wj1, wj2))
    })
}
file.is_symlink <- function(wj1) {
    nzchar(Sys.readlink(wj1), keepNA = TRUE)
}
hs.symlink.target <- function(wj1) {
}
hs.wj.symlink.force <- function(wj1, wj2) {
    if (file.exists(wj2)) {
        if (hs.normalizePath(wj1) != hs.normalizePath(wj2)) {
            unlink(wj2)
        }
    }
    if (file.exists(wj2)) {
    }
}
hs.wjj.create <- function(wj, force = 0) {
    if (!dir.exists(wj)) {
        if (file.exists(wj)) {
            file.rename(wj, paste0(wj, ".bf"))
        }
        dir.create(wj, recursive = TRUE)
    }
    wj
}
hs.seqForward <- function(from, to, ...) {
    if (to >= from) 
        seq.int(from = from, to = to, ...)
}
hs.all <- function(...) {
    for (i in hs.seqForward(1, ...length())) if (!...elt(i)) 
        return(FALSE)
    return(TRUE)
}
hs.any <- function(...) {
    for (i in hs.seqForward(1, ...length())) if (...elt(i)) 
        return(TRUE)
    return(FALSE)
}
hs.cat <- function(..., file = "", sep = if (less) "\n" else "", end = if (sep == "\n") "" else "\n", append = file.exists(hs.home.expand(file)), intern = FALSE, less = 0) {
    file <- hs.home.expand(file)
    if (intern) {
        capture.output(cat(..., sep = sep))
    }
    else {
        arg <- list(..., sep = sep, append = as.logical(append), file = file)
        if (less) 
            arg[["file"]] <- "| less -Ni"
        do.call("cat", arg)
        if (less) 
            return()
        if (nzchar(end)) 
            cat(end, append = TRUE, file = file)
        if (nzchar(file)) 
            return(file)
    }
}
hs.catn <- function(...) hs.cat(..., sep = "\n")
hs.say <- function(..., sep = " ") {
    hs.cat(..., sep = sep)
}
hs.info <- function(..., call. = TRUE, domain = NULL, appendLF = TRUE, stop = 0) {
    A <- list("\n\t", ..., "\n\n")
    A[["domain"]] <- domain
    if (stop) {
        hs <- "stop"
        A[["call."]] <- call.
    }
    else {
        hs <- "message"
        A[["appendLF"]] <- appendLF
    }
    do.call(hs, A)
}
.try <- function(expr, silent = TRUE) {
    op <- options(warn = -1)
    on.exit(options(op))
    try(expr, silent)
}
.try1 <- function(expr, value.on.error = NA, silent = TRUE) {
    on.exit(options(op))
    op <- options(warn = -1)
    r <- try(expr, silent)
    if (class(r) == "try-error") 
        value.on.error
    else r
}
hs.is_try_error <- function(x1) {
    is(x1, "try-error")
}
hs.load <- function(file, keep.name = FALSE, env = new.env(), mc = 0) {
    if (inherits(file, "connection")) {
        jg <- .try(load(file, envir = env))
        if (!hs.is_try_error(jg)) {
            if (keep.name || (length(jg) > 1)) {
                return(mget(jg, envir = env))
            }
            else return(env[[jg]])
        }
        hs.browser("2023.07.03.150920", debug = 0)
    }
    if (is.character(file)) 
        file %<=>% hs.home.expand
    if (hs.is_try_error(.try(infoRDS(file)))) {
        if (mc) {
            con1 <- .sys("pigz", "-cd", file)
            con1 %<=>% pipe(x, "rb")
            on.exit(close(con1), add = TRUE)
            b <- load(con1, envir = env)
        }
        else b <- load(file, envir = env)
        if (keep.name || (length(b) > 1)) {
            mget(b, envir = env)
        }
        else env[[b]]
    }
    else readRDS(file)
}
hs.source.part <- function(wj, env = if (local) parent.frame() else .GlobalEnv, local = 0, sep = "## ka 34.09.27.190116", nr = {
}) {
    if (!grepl("^\\s*#", sep)) 
        stop("需用注释行作为分割行")
    if (!missing(wj)) {
        wj %<=>% hs.home.expand
        nr <- hs.readLines.robust(wj)
    }
    if (0) {
        i1 <- 15692
        eval(parse(text = append(head(nr, match(sep, nr, length(nr))), deparse(quote(message("37_01_25_105517"))), after = i1), encoding = "UTF-8"), env)
        cat(append(head(nr, match(sep, nr, length(nr))), deparse(quote(message("37_01_25_105517"))), after = 1103)[seq.int(1100, 1200)], sep = "\n")
        cat(head(nr, match(sep, nr, length(nr)))[seq.int(1110, 1130)], sep = "\n")
        wk.less(head(nr, match(sep, nr, length(nr))), start = i1)
    }
    if (length(nr)) 
        eval(parse(text = head(nr, match(sep, nr, length(nr))), encoding = "UTF-8"), env)
}
hs.source.rd <- function(wj, env = .GlobalEnv, text) {
    if (missing(text)) 
        text <- hs.load(hs.home.expand(wj))
    hs.source.part(nr = text, env = env)
}
hs.source.wj <- function(wj, local = 0, up = parent.frame(), env = {
}) {
    wj %<=>% hs.home.expand
    rd <- grepl("(?i)\\.rd$", wj, perl = !0)
    hs.update(env, if (local) 
        up
    else .GlobalEnv, hs.if = is.null)
    if (rd) {
        hs.source.rd(wj, env = env)
    }
    else hs.source.part(wj, env = env)
}
hs.source.mtime <- local({
    mt <- list()
    function(wj, smart = 1) {
        for (wj1 in wj) {
            wj1 %<=>% normalizePath(hs.home.expand(x))
            if (smart && length(mt[[wj1]]) && (wj1 %=>% (mt[[x]] >= file.mtime(x)))) {
            }
            else {
                mt[[wj1]] <<- file.mtime(wj1)
                hs.source.wj(wj1)
            }
        }
    }
})
hs.r.2rd <- function(wj, wj.rd) {
    wj %<=>% hs.home.expand
    if (missing(wj.rd)) {
        wj.rd <- sub("[^\\.]+$", "rd", wj)
        for (i1 in seq_along(wj)) {
            line <- hs.readLines.robust(wj)
            save(line, file = wj.rd[i1])
        }
        wj.rd
    }
    else {
        wj.rd <- hs.home.expand(wj.rd)
        line <- unlist(lapply(wj, hs.readLines.robust))
        save(line, file = wj.rd)
        line
    }
}
hs.install.packages <- function(bao, reinstall = 0, lib = wk.lib[1], destdir = hs.wjj("~/rj/src/r包"), ...) {
    for (b1 in bao) {
        e1 <- 1
        if (reinstall) 
            unlink(system.file(package = b1), recursive = TRUE)
        if (file.exists(b1)) {
            description.tar_fp <- .wk("cat_wk", .sys.fpGood(b1), "2>/dev/null | head -n 20 | tar -tf - 2>/dev/null | grep -m1 ", .q("DESCRIPTION$"), intern = 1)
            description <- .wk("cat_wk", .sys.fpGood(b1), "| tar -xvf -", .q(description.tar_fp), "--to-stdout 2>/dev/null", intern = 1)
            b1.name <- hs.re(hs.grepV(description, "^(?i)package:"), "[^ ]*$")
            b1.version <- hs.re(hs.grepV(description, "^(?i)version:"), "[^ ]*$")
            b1.fp_lib <- hs.fp(.libPaths()[1], b1.name)
            if (file.exists(b1.fp_lib)) {
                b1.version_lib <- hs.re(hs.grepV(readLines(hs.fp(b1.fp_lib, "DESCRIPTION")), "^(?i)version:"), "[^ ]*$")
                if (package_version(b1.version) > package_version(b1.version_lib)) {
                }
                else {
                  message("[message@37_12_17_134422]", .q(b1.name), " installed already for ", .sys.fpGood(b1))
                  next
                }
            }
            install.packages(b1, lib = lib, destdir = destdir, ...)
        }
        if (hs.any(!hs.havePackages(b1), reinstall)) {
            if (hs.grepl(b1, "/")) {
                devtools::install_github(b1, lib = lib, destdir = destdir, ...)
                if (0) {
                  devtools::install_github
                  install_github <- function(repo, ref = "HEAD", subdir = NULL, auth_token = remotes:::github_pat(quiet), host = "api.github.com", dependencies = NA, upgrade = c("default", "ask", "always", "never"), force = FALSE, quiet = FALSE, build = TRUE, build_opts = c("--no-resave-data", "--no-manual", "--no-build-vignettes"), build_manual = FALSE, build_vignettes = FALSE, repos = getOption("repos"), type = getOption("pkgType"), ...) {
                    pkgbuild::with_build_tools({
                      ellipsis::check_dots_used(action = getOption("devtools.ellipsis_action", rlang::warn))
                      {
                        remotes <- lapply(repo, remotes::github_remote, ref = ref, subdir = subdir, auth_token = auth_token, host = host)
                        install_remotes(remotes, auth_token = auth_token, host = host, dependencies = dependencies, upgrade = upgrade, force = force, quiet = quiet, build = build, build_opts = build_opts, build_manual = build_manual, build_vignettes = build_vignettes, repos = repos, type = type, ...)
                      }
                    }, required = FALSE)
                  }
                  install_github("MRCIEU/TwoSampleMR")
                  .libPaths()[1] %=>% .bash
                  remotes:::install_remotes
                  install_remotes <- function(remotes, ...) {
                    res <- character(length(remotes))
                    for (i in seq_along(remotes)) {
                      tryCatch(res[[i]] <- install_remote(remotes[[i]], ...), error = function(e) {
                        stop(remotes:::remote_install_error(remotes[[i]], e))
                      })
                    }
                    invisible(res)
                  }
                  remotes::install_remote
                  install_remote <- function(remote, dependencies, upgrade, force, quiet, build, build_opts, build_manual, build_vignettes, repos, type, ...) {
                    stopifnot(remotes:::is.remote(remote))
                    package_name <- remotes::remote_package_name(remote)
                    local_sha <- remotes:::local_sha(package_name)
                    remote_sha <- remotes::remote_sha(remote, local_sha)
                    if (!isTRUE(force) && !remotes:::different_sha(remote_sha = remote_sha, local_sha = local_sha)) {
                      if (!quiet) {
                        message("Skipping install of '", package_name, "' from a ", sub("_remote", "", class(remote)[1L]), " remote,", " the SHA1 (", substr(remote_sha, 1L, 8L), ") has not changed since last install.\n", "  Use `force = TRUE` to force installation")
                      }
                      return(invisible(package_name))
                    }
                    if (inherits(remote, "cran_remote")) {
                      hs.browser("2023.07.22.172435", debug = 0)
                      install_packages(package_name, repos = remote$repos, type = remote$pkg_type, dependencies = dependencies, quiet = quiet, destdir = "~/rj/src/r包", ...)
                      return(invisible(package_name))
                    }
                    res <- try(bundle <- remotes::remote_download(remote, quiet = quiet), silent = quiet)
                    remotes::remote_download
                    UseMethod("remote_download")
                    remotes::remote_download
                    getS3method("remote_download", class = "github_remote")
                    getAnywhere("remote_download")
                    findMethods("remote_download", class = class(remote))
                    findMethods("remote_download", class = "github_remote")
                    remotes:::remote_download.github_remote
                    remotes:::download
                    remotes:::base_download
                    remotes:::download_method()
                    remotes:::base_download_headers
                    utils::download.file
                    if (inherits(res, "try-error")) {
                      return(NA_character_)
                    }
                    hs.browser("2023.07.22.172347", debug = 0)
                    on.exit(unlink(bundle), add = TRUE)
                    source <- remotes:::source_pkg(bundle, subdir = remote$subdir)
                    on.exit(unlink(source, recursive = TRUE), add = TRUE)
                    update_submodules(source, remote$subdir, quiet)
                    add_metadata(source, remotes::remote_metadata(remote, bundle, source, remote_sha))
                    remotes:::clear_description_md5(source)
                    remotes:::install(source, dependencies = dependencies, upgrade = upgrade, force = force, quiet = quiet, build = build, build_opts = build_opts, build_manual = build_manual, build_vignettes = build_vignettes, repos = repos, type = type, ...)
                  }
                }
            }
            else {
                (if ("repos" %in% ...names()) 
                  install.packages
                else BiocManager::install)(b1, lib = lib, destdir = destdir, update = FALSE, ...)
            }
            e1 <- 1
        }
    }
}
hs.require <- function(bao, install = reinstall, reinstall = 0, lib = .libPaths()[1], load = 0, ...) {
    bao %<=>% unique
    r <- data.frame(installed = rep(TRUE, length(bao)))
    rownames(r) <- bao
    for (b1 in bao) {
        if (grepl("@", b1)) 
            hs.browser("39.05.27.221613", debug = 0)
        b2 <- basename(b1)
        b2 %<=>% hs.sub("-.*")
        e1 <- 1
        if (reinstall) 
            unlink(system.file(package = b2), recursive = TRUE)
        if (hs.any(!hs.havePackages(b2), reinstall)) {
            e1 <- 0
            r[b1, "installed"] <- FALSE
            if (install) {
                hs.switch(b2, "Rfast", if (system("type gsl-config")) 
                  .wk("sudo apt install libgsl-dev"), "data.table", {
                  if (system("type pkgconf")) {
                    system("sudo apt install pkgconf")
                  }
                  if (system("pkg-config --libs zlib")) {
                    system("sudo apt install zlib1g-dev")
                  }
                }, "scRNAtoolVis", hs.require(c("junjunlab/jjAnno", "sajuukLyu/ggunchull", "ComplexHeatmap", "junjunlab/scRNAtoolVis"), 1), "Rediscover", {
                  if (system("pkg-config --libs fftw3")) 
                    system("sudo apt install libfftw3-dev")
                }, "car", {
                  if (system("type gfortran")) 
                    system("sudo apt install gfortran")
                })
                wj1 <- "~/.Rprofile"
                wj2 <- paste0(wj1, ".备份")
                if (file.exists(wj1)) 
                  file.rename(wj1, wj2)
                A <- list(b1, lib = lib, ...)
                if ((length(b1) > 1) && ("stringi" %in% b1)) 
                  hs.browser("2023.08.28.214104", debug = 0)
                if (b1 == "stringi") 
                  A[["configure.vars"]] <- paste0("ICUDT_DIR=", "~/db/软件/src/R包/stringi")
                do.call(hs.install.packages, A)
                if (!file.exists(wj1)) 
                  file.rename(wj2, wj1)
                e1 <- 1
                r[b1, "installed"] <- 1
            }
        }
        if (e1 && load) 
            suppressMessages(do.call("require", list(b2, quietly = 1)))
    }
    if (!all(r)) 
        stop("[38_01_08_114620]缺包 ", paste(rownames(r)[!r], collapse = ", "))
}
hs.source2wj <- function(hs1, wjj2 = "~/tmp") {
    on <- paste(substitute(hs1))
    if ((length(on) > 1) && hs.grepl(on[1], "::+")) 
        on %<=>% paste(x[-1], collapse = "_")
    wj2 <- hs.fp(wjj2, paste0(on, ".r"))
    if (!file.exists(wj2)) {
        code <- capture.output(hs1)
        hs.update(code[length(code)], paste("##", x), hs.grepl(x, "^\\<environment: .*\\>$"))
        hs.update(code[length(code) - 1], paste("##", x), hs.grepl(x, "^\\<bytecode: .*\\>$"))
        wk.dt.fwrite(code, wj2)
    }
    wj2
}
hs.init <- function(wjj = wk.dm.wjj, smart = 1) {
    wjj %<=>% hs.home.expand
    ..debug <- 0
    wj <- sapply(file.path(wjj, paste0(c("internal.r", "string.r", "wj.r", "data.table.r", "xm.r", "function.common.r", "代码分析.r", "swxx.r", "fig.r", "math.r", if (interactive()) c("zim.r", "org1.r"), if (hs.machine.isLocal()) c("my.r")), "")), hs.getTextFile) %=>% unlist
    sapply(wj, . %=>% {
        if (file.exists(x)) {
            if (hs.machine.isLocal()) 
                message("加载", x)
            hs.source.robust(x)
        }
    })
    return()
    identical(Sys.getenv()[["USERID"]], "38_06_28_092642")
    if (!interactive()) {
        sapply(wj, function(wj1) {
            source(wj1)
        })
        return()
    }
    wj.rd <- file.path(wjj, "all.r.rd")
    if (1) {
        convert <- (!file.exists(wj.rd)) || (length(wj) && (max(file.mtime(c(wj, hs.getTextFile(file.path(wjj, "init.r"))))) > file.mtime(wj.rd)))
        if (..debug) 
            convert <- 1
        if (convert) {
            line <- hs.r.2rd(wj, wj.rd)
            hs.source.rd(text = line)
        }
    }
    if (..debug) 
        hs.browser("36.02.07.203804")
    hs.source.mtime(wj.rd, smart = smart)
    if (0) 
        if (hs.machine.isLocal()) 
            sapply(hs.wj(wk.dm.wjj, paste0(c("my.r"), "")), function(wj1) {
                hs.source.mtime(wj1, smart = smart)
            })
}
hs.init1 <- function() {
    .s()
    if (!exists("wk.ssh", envir = .GlobalEnv)) 
        assign("wk.ssh", {
        }, envir = .GlobalEnv)
    if (!exists("wk.rsync.wj.wj", envir = .GlobalEnv)) 
        assign("wk.rsync.wj.wj", hs.wj.portable(hs.wj(.mfp("36.02.15.203034.txt.gz"))), envir = .GlobalEnv)
    {
        evalq({
            wj.wj <- hs.home.expand("~/org") %=>% .wk("ag -g 文件/找/find.39_01_08_150201", x, intern = 1) %=>% parse(x)[[1]][[2]][["wj2"]] %=>% eval
            uid.wj <- uid.wj.wj.hs()
            if (!exists(env.lt, envir = .GlobalEnv)) 
                assign(env.lt, new.env(), envir = .GlobalEnv)
        }, envir = environment(hs.hsgly))
    }
    for (wj1 in c("org1.r", "my.r")) hs.fp(wk.dm.wjj, wj1) %=>% hs.getTextFile %=>% hs.source.robust
    {
        pager_fp <- "~/org/dm/sh/rj/pager"
        if (!file.exists(pager_fp)) {
            cat("#!/bin/bash", "less -i", sep = "\n", file = pager_fp %=>% hs.wj)
            Sys.chmod(pager_fp, "0700")
        }
        options(pager = normalizePath(pager_fp))
    }
    "~/.db/2do.org.gz" %=>% hs.wj %=>% if (!file.exists(x)) 
        cat(letters, file = .sys("| gzip >", hs.fpGood(x)))
}
my.r <- function() {
    wj1 <- hs.getTextFile(hs.wj(wk.dm.wjj, "my.r"))
    hs.source.robust(wj1)
}
hs.gc_here <- function(env = parent.frame()) {
    evalq(rm(list = ls()), envir = env)
}
.First <- function() if (interactive()) loadhistory(hs.home.expand("~/.Rhistory"))
hs.save.history <- function(keep.incomplete_line = 1, size = Sys.getenv("HISTSIZE")) {
    Sys.setenv(R_HISTSIZE = size)
    try({
        wj1 <- hs.home.expand("~/.Rhistory")
        wj2 <- paste0(wj1, ".tmp")
        on.exit(hs.wj.rm(wj2), add = TRUE)
        savehistory(wj2)
        h0 <- rev(readLines(wj1))
        h1 <- rev(readLines(wj2))
        if (!keep.incomplete_line) 
            h1 <- unlist(sapply(h1, function(h2) {
                r1 <- .try(parse(text = h2))
                if (!input.is.incomplete(r1)) 
                  h2
            }))
        cat(rev(h1), sep = "\n", file = wj1)
        wj1
    })
}
.Last <- function() {
    if (hs.machine.isLocal()) 
        local({
            if (0) {
                wj1 <- "~/db/软件库/.config/vivaldi/Default"
                wj2 <- "~/.config/vivaldi/Default"
                if (!dir.exists(wj2)) {
                  dir.create(dirname(wj2), recursive = TRUE)
                  file.symlink(wj1, wj2)
                }
                wjj1 <- "~/db/软件库/.config/vivaldi/"
                wjj2 <- "~/.config/vivaldi/"
                if (length(.wk("find", hs.fpGood(wjj2), "-maxdepth 1", "-iname", .q("singleton*"), intern = 1))) 
                  hs.browser("2023.07.11.105504|vivaldi在运行", debug = 0)
                hs1 <- function(wjj) hs.hsgly("列出部分文件.2023_07_11_102948")(wjj, exclude = "Default") %=>% file.mtime %=>% max(na.rm = 1)
                time1 <- hs1(wjj1)
                time2 <- hs1(wjj2)
                if (time1 != time2) {
                  if (time1 < time2) 
                    hs.swap(wjj1, wjj2)
                  .wk("rsync -ahPvkK --delete", "-f", .q("- /Default/"), hs.fpGood(wjj1), hs.fpGood(wjj2))
                }
            }
            if (0) {
                ln_wjj2_db <- hs.home.expand("~/db/软件库")
                wjj_pair <- list(c("~/rjk/ln-s/.ssh/", "~/.ssh/"), c(paste0(ln_wjj2_db, "/.config/opera/"), "~/.config/opera/"))
                for (i1 in seq_along(wjj_pair)) {
                  p1 <- wjj_pair[[i1]]
                  wjj1 <- hs.home.expand(p1[1])
                  wjj2 <- hs.home.expand(p1[2])
                  if (hs.wj.is.Symlink(wjj2)) 
                    wjj2 %=>% hs.sub("/*$") %=>% file.rename(x, paste0(x, ".bf"))
                  t2 <- hs.wjj.ls(wjj2, R = 1) %=>% file.mtime %=>% max(na.rm = 1)
                  t1 <- hs.wjj.ls(wjj1, R = 1) %=>% file.mtime %=>% max(na.rm = 1)
                  if ((paste(t2) != "NA") && (t1 != t2)) {
                    if (as.difftime(t2 - t1, units = "secs") > 0.1) 
                      hs.swap(wjj1, wjj2)
                    .wk(wj_rsync, "-ahPvz", "--delete", "-f", .q(paste("-", .q("X[0-9]"))), wjj1, wjj2)
                  }
                  "~/.ssh" %=>% c(x, hs.wjj.ls(x)) %=>% hs.fp.make_personal
                }
            }
        })
    if (interactive()) {
        hs.save.history()
        for (o1 in c("hs.org.db.tbl")) if (exists(o1)) 
            get(o1)(save.and.exit = 1)
    }
}
hs.ls <- function(..., all.names = TRUE, sorted = FALSE) {
    if (inherits(..1, "character") && (length(..1) == 1)) {
        return(getNamespaceExports(..1) %=>% x[order(tolower(x))])
    }
    ls(..., all.names = all.names, sorted = sorted)
}
hs.str <- function(x) {
    wj_temp <- "/dev/shm/2023_08_15_182030.txt"
    on.exit(hs.wj.rm(wj_temp), add = TRUE)
    sink(wj_temp)
    str(x)
    sink()
    return(.wk("less -Ni", wj_temp))
}
hs.pkg.depend <- function(pkg) {
    system.file("DESCRIPTION", package = pkg) %=>% read.dcf(fields = "Depends") %=>% hs.gsub("\\([^\\(\\)]*\\)") %=>% strsplit(",") %=>% unlist %=>% trim
}
hs.pkg.exported <- function(pkg) {
    getNamespaceExports(pkg) %=>% x[order(tolower(x))]
}
hs.pkg.get <- function(pkg, what, include_depend = 1, re = FALSE) {
    e <- getNamespaceExports(pkg)
    if (re) {
        alike <- hs.grepV(e, what)
        if (length(alike)) 
            return(hs.lapply(alike, . %=>% do.call("::", list(pkg, x))))
    }
    else if (what %in% e) {
        return(do.call("::", list(pkg, what)))
    }
    if (include_depend) {
        depend <- hs.pkg.depend(pkg) %rm% c("R", NA)
        for (d1 in depend) {
            a <- hs.pkg.get(d1, what, include_depend = 0)
            if (length(a)) 
                return(a)
        }
    }
}
hs.lvl <- base::levels
hs.nc <- base::nchar
hs.p <- base::paste
hs.p0 <- base::paste0
hs.nm <- base::names
`hs.nm<-` <- base::`names<-`
hs.ev <- base::eval
hs.bq <- base::bquote
hs.env <- base::environment
`hs.env<-` <- base::`environment<-`
hs.ss <- function(..., perl = TRUE) {
    strsplit(..., perl = perl)
}
hs.df <- function(..., rn = {
}, check.names = FALSE, stringsAsFactors = FALSE) {
    .. <- list(..., check.names = check.names, stringsAsFactors = stringsAsFactors)
    if (length(rn)) 
        ..[["row.names"]] <- rn
    hs.call...(E = data.frame(...), .. = ..)
}
hs.na <- function(x) {
    a <- x[1]
    a[1] <- NA
    a
}
hs.sample <- function(x, size = hs.switch(length(dim(x)), 0, length(x), 2, dim(x)[1], hs.browser("2023.08.02.142428", debug = 0)), replace = hs.switch(length(dim(x)), 0, if ((length(x) == 1) && inherits(x, "numeric") && (x >= 1)) size > x else size > length(x), 2, size > nrow(x), hs.browser("2023.08.02.142705", debug = 0)), prob = {
}) {
    hs.switch(length(dim(x)), 0, sample(x = x, size = size, replace = replace, prob = prob), 2, hs.ij(x, sample(x = nrow(x), size = size, replace = replace, prob = prob)))
}
hs.list <- function(..., env.36_11_30_195548 = parent.frame()) {
    a1 <- substitute(alist(...))
    name.L <- lapply(a1[seq(2, length(a1), by = 2)], eval, env.36_11_30_195548)
    i1 <- sapply(name.L, length) == 1
    if (0) 
        hs.c(lapply(a1[seq(2, length(a1), by = 2)], eval, env.36_11_30_195548), lapply(a1[seq(3, length(a1), by = 2)], eval, env.36_11_30_195548))
    if (any(i1)) {
        hs.c(unlist(name.L[i1]), lapply(a1[seq(3, length(a1), by = 2)[i1]], function(x1) {
            "language" %in% class(x1)
            if (is.language(x1)) {
                eval(x1, env.36_11_30_195548)
            }
            else x1
        }))
    }
}
hs.runCmd <- function(..., intern = FALSE) {
    wj_cmd <- tempfile()
    on.exit(unlink(wj_cmd), add = TRUE)
    cmd <- paste(c(...), collapse = " ")
    cat(cmd, file = wj_cmd)
    system(paste("bash", wj_cmd), intern = intern)
}
hs.methods <- function(O) {
    for (c1 in base::class(O)) {
        hs.say(base::sprintf("Methods for %s:", c1))
    }
}
hs.filter <- function(..., out = "i") {
    of <- list(...)
    if (length(of)%%2) 
        stop("37_02_01_215621 the argument number should be even")
    oi <- seq(1, length(of), by = 2)
    n0 <- unique(sapply(oi, function(oi) length(of[[oi]])))
    if (length(n0) > 1) 
        stop("37_01_12_102345")
    i1 <- seq_len(n0)
    for (o1i in oi) {
        if (length(i1) < 1) 
            break
        hs1 <- of[[o1i + 1]]
        i2 <- hs1(of[[o1i]][i1])
        if (is.logical(i2)) 
            i2 <- which(i2)
        i1 <- i1[i2]
    }
    if (length(of) > 2) 
        out <- "i"
    switch(out, i = i1)
}
hs.filter.1 <- function(X, ..., out = "v", x1 = 1) {
    i1 <- seq_along(X)
    hs <- list(...)
    for (hs1.i in seq_along(hs)) {
        if (length(i1) < 1) {
            break
        }
        hs1 <- hs[[hs1.i]]
        i1 <- i1[which(hs1(X[i1]))]
    }
    switch(out, i = i1, v = X[i1])
}
hs.attr.slice <- function(a1, i1, null = {
}) {
    if (!length(i1)) 
        return(null)
    r1 <- a1[i1]
    for (n1 in names(attributes(a1))) {
        attr(r1, n1) <- if (length(attr(a1, n1)) == length(a1)) {
            attr(a1, n1)[i1]
        }
        else attr(a1, n1)
    }
    r1
}
hs.match <- function(..., fail = FALSE) {
    r1 <- match(...)
    if (is.na(r1)) {
    }
    else r1
}
hs.skip <- function(v, v1) {
    i <- match(v1, v, nomatch = 0)
    if (i) 
        tail(v1, -i)
    else v
}
v2.hs_chain.v1 <- function(x1, ...) {
    chain <- list(...)
    for (c1 in chain) x1 %<=>% c1[[2]][match(x, c1[[1]])]
    x1
}
hs.eval <- function(E, env1 = parent.frame(), local = 0) {
    if (local) 
        env1 <- new.env(parent = env1)
    eval(E, env1)
}
hs.local <- function(E, restoreWd = 1, env1 = parent.frame()) {
    if (restoreWd) {
        wjj0 <- getwd()
        on.exit(setwd(wjj0), add = TRUE)
    }
    hs.eval(E, env1, local = 1)
}
hs01.expression2language <- function(e1) switch(paste(length(e1)), `0` = {
}, `1` = e1[[1]], as.call(c(as.symbol("{"), as.list(e1))))
hs.evalLazy <- function(exp1, uid, root = formals(.mfp.swxx)[["dataDir"]], fp = hs.home.expand(hs.wjm(.mfp(uid, dtt = dtt), add = ext)), ext = if ("TxDb" %in% class) "sql" else "rd", lazy = 1, ow = 1, env = parent.frame(), fp.only = 0, dtt = "handy", save.jg = 1, want = "jg", hs4save = if (hs.grepl(fp, "(?i)\\.[ct]sv(\\.gz)?$")) {
    function(x, fp) wk.dt.fwrite(x, fp)
} else if (hs.grepl(fp, "(?i)\\.xlsx?$")) {
    function(x, fp) wk.xls.write(x, fp)
} else if (hs.grepl(fp, "(?i)\\.rd$")) {
    function(x, fp) hs.save(x, file = fp)
} else {
    function(x, fp) fp
}, hs4load = if (hs.grepl(fp, "(?i)\\.[ct]sv(\\.gz)?$")) {
    wk.fread
} else if (hs.grepl(fp, "(?i)\\.xlsx?$")) {
    function(fp) wk.xls.read(fp, if (xls.all.sheet) 
        ""
    else 1)
} else if (hs.grepl(fp, "(?i)\\.rd$")) {
    hs.load
} else identity, load = 1, xls.all.sheet = 0, lib = {
}, class = {
}, local = 1) {
    if (fp.only) 
        return(fp)
    if (file.exists(fp) && lazy) 
        return(if (load) hs4load(fp))
    if (length(lib)) 
        hs.require(lib)
    R <- eval(substitute(exp1), envir = if (local) 
        new.env(parent = env)
    else env)
    hs4save(R, hs.wj(fp))
    if (load) 
        R
}
hs.nonempty <- function(L, length = 1, nrow = 0, ot = "v") {
    hs <- eval({
        if (length && nrow) {
            bquote(function(a) (length(a) >= .(length)) && (nrow(a) >= .(nrow)))
        }
        else if (length) {
            bquote(function(a) length(a) >= .(length))
        }
        else if (nrow) {
            bquote(function(a) nrow(a) >= .(nrow))
        }
    })
    if (length(L)) {
        i1 <- sapply(L, hs)
        switch(ot, v = L[i1], i = which(i1), l = i1)
    }
}
hs.is_empty <- function(x1, co1 = 0) {
    if (is(x1, "data.table")) 
        return(x1[, .N <= co1])
    if (is(x1, "character")) 
        return((length(x1) == 0) || ((length(x1) == 1) && (nchar(x1) <= co1)))
    length(x1) <= co1
}
hs.graph.node.path <- function(N, Z, P = {
}) {
    P <- c(P, N)
    p1 <- Z[N, ]
    down <- names(p1)[p1 > 0] %not% P
    self <- sys.function()
    if (length(down)) {
        lapply(down, function(n2) {
            self(n2, Z, P)
        })
    }
    else P
}
hs0.seq.eql <- function(s1, s2) {
    (length(s1) == length(s2)) && all(s1 == s2)
}
hs.exists <- wk.exists <- ieiwk.exists <- local.exists <- function(x, env1 = parent.frame(), inherits = FALSE, ...) {
    if (length(x)) {
        sapply(x, function(x) {
            exists(x, inherits = inherits, envir = env1, ...)
        })
    }
    else FALSE
}
hs.make.connection <- function(wj, open = {
}) {
    if (hs.len1(wj)) {
        if ("connection" %not.in% class(wj)) {
            if (is.character(wj)) {
                wj <- {
                  if (file.exists(hs.home.expand(wj))) {
                    gz <- hs.grepl(wj, "\\.gz$")
                    args <- list(description = wj)
                    if (length(open)) {
                      args[["open"]] <- open
                    }
                    do.call(if (gz) 
                      gzfile
                    else file, args)
                  }
                  else {
                    pipe(wj)
                  }
                }
            }
            else {
                stop("what is the connection? 34.10.05.070316 @ ~/org/dm/r/function.common.R")
            }
        }
        if (!isOpen(wj, "r")) {
            if (isOpen(wj, "w")) {
                stop("hs.make.connection: use only for reading. 34.10.05.071940 @ ~/org/dm/r/internal.r")
            }
            open(wj)
        }
        wj
    }
    else {
        hs.browser("38.06.06.145143")
    }
}
hs.readLines <- function(con, iv = {
}, toEnd = 0, n = -1, skip = {
}, ok = TRUE, warn = TRUE, encoding = "UTF-8", close = n < 1) {
    con %<=>% hs.make.connection
    if (close) 
        on.exit(close(con), add = TRUE)
    skip.chunk.size <- 1000L
    skip %=>% if (length(x)) {
        j <- x
        while (j > 0) {
            readLines(con, n = min(skip.chunk.size, j), ok = ok, warn = warn)
            j <- j - skip.chunk.size
        }
    }
    if (length(iv)) {
        iv_2 <- unique(sort(iv))
        skip <- diff(c(0, iv_2)) - 1
        r <- vector("character", length(skip))
        for (i in seq_along(skip)) {
            j <- skip[i]
            while (j > 0) {
                readLines(con, n = min(skip.chunk.size, j), ok = ok, warn = warn)
                j <- j - skip.chunk.size
            }
            r[i] <- readLines(con, n = 1, ok = ok, warn = warn)
        }
        if (toEnd) {
            c(r[rank(iv)], readLines(con, ok = ok, warn = warn))
        }
        else r[rank(iv)]
    }
    else readLines(con, n = n, ok = ok, warn = warn, encoding = encoding)
}
hs.source <- ieiwk.source <- function(f.in, pattern, ignore.case = TRUE, fixed = FALSE, perl = FALSE, local = FALSE, encoding = "UTF-8", ...) {
    f.in %<=>% hs.home.expand
    flag <- 0
    uid <- {
        a <- readLines(f.in, 10)
        hs.trim(re("[^ ]*", grep(" # file uid$", a, v = 1)), "\"")
    }
    if (!length(apropos(uid))) 
        flag <- 1
    if (!flag) {
        mtime <- file.info(f.in)[["mtime"]]
        if (Sys.time() - mtime < 24) 
            flag <- 1
    }
    if (flag) {
        if (!missing(pattern)) {
            e <- parse(f.in)
            env <- {
                if (is.logical(local)) {
                  if (local) {
                    .GlobalEnv
                  }
                  else parent.frame()
                }
                else local
            }
            e.s <- sapply(seq_along(e), function(a.i) {
                capture.output(e[a.i])
            })
            e.i <- unlist(sapply(pattern, function(p1) {
                for (a.i in seq_along(e)) if (any(grepl(p1, e.s[[a.i]], fixed = fixed))) 
                  return(a.i)
            }))
            sapply(seq_along(e.i), function(i) {
                j <- e.i[i]
                .ss(sprintf("%s # %s sourced.", f.in, names(e.i)[i]))
                eval(e[[j]], env)
            })
            invisible()
        }
        else {
            .ss(sprintf("%s sourced.", f.in))
            source(f.in, encoding = encoding, local = local, ...)
            source(f.in, encoding = encoding, local = FALSE, ...)
        }
    }
}
hs.isOdd <- function(x) as.logical(x%%2)
hs.isEven <- function(x) !(x%%2)
hs.recycle <- function(i, l) {
    if (is.logical(i)) 
        i <- which(i)
    j <- i%%l
    if (0 %in% j) 
        j[j %in% 0] <- l
    j
}
hs.ifelse <- function(test, yes, no, envUp = parent.frame()) {
    if (!(identical(environment(yes), envUp) && identical(environment(no), envUp))) 
        stop("[38_01_01_122923]")
    jg <- rep(NA, length(test))
    i1 <- test
    if (!is.logical(i1)) 
        i1 <- as.logical(i1)
    if (any(i1)) 
        jg[i1] <- yes(i1)
    if (FALSE %in% i1) {
        i1 <- !i1
        jg[i1] <- no(i1)
    }
    jg
}
hs.attr <- function(o1, nm, vl, simplify = 1) {
    if (missing(nm)) {
        names(attributes(o1))
    }
    else {
        if (missing(vl)) {
            r <- wk.sapply(nm, function(n1) {
                attr(o1, n1)
            }, nc = 1)
            if (hs.len1(r) && simplify) {
                r[[1]]
            }
            else r
        }
        else {
            if (length(nm) > 1) {
                for (i in seq_along(nm)) attr(o1, nm[i]) <- vl[[hs.recycle(i, length(vl))]]
            }
            else attr(o1, nm) <- vl
            o1
        }
    }
}
expandingSequence <- function(capacity = 10, named = 0, type = "list") {
    buffer <- vector(switch(type, list = "list", vector = "logical", type), capacity)
    if (named) 
        names <- character(capacity)
    length <- 0
    methods <- list()
    {
        a1 <- function() {
        }
        on.exit(rm(list = "a1"), add = TRUE)
        body(a1) <- bquote({
            buffer <<- c(buffer, vector(.(switch(type, list = "list", vector = "logical", type)), capacity))
            capacity <<- capacity * 2
            .(if (named) 
                quote(names <<- c(names, character(capacity))))
        })
        methods[["double.size"]] <- a1
    }
    {
        a1 <- if (named) 
            function(name, val = NULL) {
            }
        else function(val = NULL) {
        }
        body(a1) <- bquote({
            if (length == capacity) 
                methods$double.size()
            length <<- length + 1
            buffer[length] <<- .(if (type == "list") {
                quote(list(val))
            }
            else quote(val))
            .(if (named) 
                quote(names[length] <<- name))
        })
        methods[["add"]] <- a1
    }
    {
        a1 <- function(i = {
        }, ignore_out_of_range = 1) {
        }
        body(a1) <- bquote({
            if (length(i)) {
                if (max(i) > length) {
                  if (ignore_out_of_range) {
                    warning("out of range @33.07.17.204126@")
                    i <- i[i <= length]
                  }
                  else stop("out of range @33.07.17.203755@")
                }
                .(if (named) {
                  quote(hs.c(names[i], buffer[i]))
                }
                else quote(buffer[i]))
            }
            else .(if (named) {
                quote(hs.c(head(names, length), head(buffer, length)))
            }
            else quote(head(buffer, length)))
        })
        methods[["get"]] <- a1
    }
    methods
}
linkedList <- function() {
    head <- list(0)
    length <- 0
    methods <- list()
    methods$add <- function(val) {
        length <<- length + 1
        head <<- list(head, val)
    }
    methods$get <- function() {
        b <- vector("list", length)
        if (length) 
            for (i in length:1) {
                b[[i]] <- head[[2]]
                head <- head[[1]]
            }
        b
    }
    methods
}
hs.linkedList <- function() {
    head <- list(0)
    lengthTotal <- length <- 0
    methods <- list()
    methods$add <- function(val) {
        length <<- length + 1
        head <<- list(head, val)
    }
    R <- expandingList()
    methods$get <- function() {
        if (length) {
            if (get("length", environment(R$get)) == 0) 
                R <<- expandingList(length)
            for (i in length:1) {
                R$add(head[[2]])
                head <- head[[1]]
            }
            length <<- 0
        }
        R$get()
    }
    methods
}
expandingList <- function(...) {
    expandingSequence(..., named = 0, type = "list")
}
namedExpandingList <- function(...) {
    expandingSequence(..., named = 1, type = "list")
}
expandingVector <- function(..., type = "vector") expandingSequence(..., named = 0, type = type)
namedExpandingVector <- function(capacity = 10) {
    buffer <- vector(length = capacity)
    length <- 0
    methods <- list()
    methods$double.size <- function() {
        buffer <<- c(buffer, vector(length = capacity))
        capacity <<- capacity * 2
    }
    methods$add <- function(val) {
        end <- length + length(val)
        while (end >= capacity) methods$double.size()
        buffer[seq(length + 1, end)] <<- val
        length <<- end
    }
    methods$get <- function(i, rm = 0) {
        if (missing(i)) {
            R <- head(buffer, length)
            if (rm) {
                buffer <<- vector(length = capacity)
                length <<- 0
            }
        }
        else {
            R <- buffer[i]
            if (rm) {
                i <- sort(unique(i))
                buffer <<- buffer[-i]
                length <<- length - length(i)
            }
        }
        R
    }
    methods
}
hs.acc <- wk.acc <- function(once = 0, type = "list") {
    env1 <- new.env()
    nL <- switch(type, vector = expandingVector, list = expandingList)()
    N <- 0
    methods <- list()
    methods$add <- {
        hs1 <- function(n, v) {
        }
        b1 <- quote({
            if (missing(v)) {
                v <- n
                n <- N + 1
                N <<- n
            }
            n <- as.character(n)
        })
        b2 <- quote({
            if (!methods$has(n)) {
                assign(n, v, envir = env1)
                nL$add(n)
            }
        })
        if (!once) {
            b2 <- b2[[2]][[3]]
        }
        b1 <- c(`{`, append(as.list(b1[-1]), as.list(b2[-1])))
        body(hs1) <- as.call(b1)
        hs1
    }
    methods$get <- function() {
        a <- nL$get() %and% ls(envir = env1)
        if (length(a)) {
            mget(a, envir = env1)
        }
    }
    methods$has <- function(n) {
        exists(as.character(n), envir = env1, inherits = FALSE)
    }
    methods$rm <- function(n) {
        a <- as.character(n)
        if (methods$has(a)) {
            rm(a, envir = env1)
        }
    }
    methods$n <- function() {
        nL$get() %and% ls(envir = env1)
    }
    methods
}
wk.flat <- function(x, flag = "wk.flat.me") {
}
hs.list.flatten <- function(L, hs1 = identity, D = 2, hs2 = rbindlist) {
    a1 <- expandingList()
    do.call("lapply", c(list(L), unname(sapply(rep("lapply", D - 1), as.symbol)), function(.) {
        a1[["add"]](hs1(.))
    }))
    rbindlist(a1[["get"]]())
}
hs.list.merge <- function(L, name = 1) {
    a1 <- tapply(seq_along(L), names(L), list)
    i1 <- which(sapply(a1, length) > 1)
    if (length(i1)) {
        for (i1.1 in i1) {
            i2 <- a1[[i1.1]]
            L[[i2[1]]] %<=>% (x %or% unlist(L[i2[-1]]))
        }
        i2rm <- lapply(i1, . %=>% a1[[x]][-1]) %=>% unlist
        L[i2rm] <- {
        }
    }
    if (!name) {
        for (i1 in seq_along(L)) {
            n1 <- names(L)[i1]
            if (n1 %in% L[[i1]]) 
                L[[i1]] %<=>% (x %not% n1)
        }
    }
    L
}
hs.abline <- function() abline(0, 1, v = 0, h = 0, col = Mycol2rgb("black", 0.25))
hs.check.list <- function(L) {
    lapply(L, function(.) {
        r1 <- table(.)
        a1 <- length(r1)
        if (a1 < (length(.)/2)) 
            r1
    })
}
hs.unique <- function(a1) {
    i1 <- which(duplicated(a1))
    if (length(i1)) {
        a1[-i1]
    }
    else a1
}
hs.duplicated <- function(x, cn = {
}, inv = FALSE, ot = "v") {
    hs.cond(is.atomic(x), {
        i <- x %in% x[duplicated(x)]
        if (inv) 
            i <- !i
        switch(ot, v = x[i], i = which(i))
    }, inherits(x, "data.table"), {
        if (inv) 
            hs.browser("39.03.15.234353", debug = 0)
        i <- if (length(cn)) 
            x[, duplicated(.SD), .SDcols = cn]
        else duplicated(x)
        x[i]
    })
}
hs.v.start_with <- function(v1, e1, ot = "v") {
    i1 <- match(e1, v1)
    switch(ot, v = v1[i1:length(v1)], i = i1:length(v1))
}
hs.v.within <- function(v1, e1, e2, ot = "v") {
    i1 <- match(c(e1, e2), v1)
    i1 <- seq(i1[1], i1[2])
    switch(ot, v = v1[i1], i = i1)
}
hs.unlist <- function(x, recursive = TRUE, use.names = FALSE) unlist(x = x, recursive = recursive, use.names = use.names)
intersection_string.hs.list <- function(L) {
    names(L) <- seq_along(L)
    L1 <- inverseList(L)
    names(L1)[sapply(L1, length) == length(L)]
}
hs.table <- function(..., order = "smart", M = 0, prop = type == "p", type = "c", percentage = 0, margin = NULL) {
    sj <- list(...)
    jg <- table(..., useNA = "ifany")
    if (M) {
        jg <- rbind(jg, colSums(jg))
        jg <- cbind(jg, rowSums(jg))
    }
    if (order == "smart") {
        dimname <- lapply(seq_along(sj), function(d1) {
            a1 <- rle(as.vector(sj[[d1]]))[["values"]]
            a2 <- attr(jg, "dimnames")[[d1]]
            if (anyDuplicated(a1)) {
                seq_along(a2)
            }
            else match(a1, a2)
        })
        jg <- do.call("[", c(list(jg), dimname, drop = FALSE))
    }
    switch(type, c = jg, p = prop.table(jg, margin = margin) * ifelse(percentage, 100, 1), b = jg %=>% list(x, prop.table(x, margin = margin) * ifelse(percentage, 100, 1)) %=>% hs.c(c("count", ifelse(percentage, "pct", "prop.")), x) %=>% do.call(cbind, x))
}
hs.tapply <- function(..., simplify = 0) tapply(..., simplify = as.logical(simplify))
hs.timestamp <- function(quiet = FALSE) {
    timestamp(stamp = Sys.time(), quiet = quiet)
}
hs.cat.percentage <- local({
    def_in <- 1
    db1 <- seq(0, 100, def_in)[-1]
    function(n1, def = 1, new = 0) {
        if (new) {
            db1 <<- seq(0, 100, def_in)
            return()
        }
        if (def != def_in) {
            def_in <<- def
            db1 <<- seq(0, 100, by = def)
        }
        n2 <- ceiling(n1 * 100)
        if (n2 %in% db1) {
            cat("**** ", n2, "%", " ****", "\n", sep = "")
            db1 <<- db1 %not% n2
        }
    }
})
hs.not <- `!`
hs.some_false <- hs.not_all <- function(..., na.rm = FALSE) !all(..., na.rm = na.rm)
hs.none <- function(..., na.rm = FALSE) {
    !any(..., na.rm = na.rm)
}
hs.which <- function(x, na = 1, inv = 0) {
    a <- which(if (inv) {
        !x
    }
    else as.logical(x))
    if (na) {
        i1 <- which(is.na(x))
        if (length(i1)) 
            a <- sort(a %or% i1)
    }
    a
}
hs.updatePackages <- ieiwk.update.packages <- function(lib.loc = wk.path[["lib"]][1], ask = 0) {
    update.packages(lib.loc, ask = ask)
}
hs.removePackages <- ieiwk.remove.packages <- function(pkgs, lib = wk.path[["lib"]][1]) {
    remove.packages(pkgs, lib)
}
hs.lines_comment <- function(wj1, opt = if (plus > 0) "l" else "i", comment = "#", skip = 1, plus = 0) {
    wj1 <- hs.home.expand(wj1)
    con1 <- hs.file(wj1)
    on.exit(close(con1), add = TRUE)
    stepSize <- 10
    comments <- {
    }
    while (length(L <- readLines(con1, stepSize))) {
        i1 <- hs.which(substr(L, 1, nchar(comment)) %not.in% c(comment, ""))
        if (length(i1)) {
            if (i1[1] > 1) 
                comments <- c(comments, head(L, i1[1] - 1))
            break
        }
        else comments <- c(comments, L)
    }
    if (plus > 0) {
        comments <- c(comments, readLines(con1, plus))
    }
    switch(opt, i = length(comments), l = comments, h = unlist(strsplit(comments[length(comments) - plus], "\t")), c = {
        con2 <- hs.file(wj1)
        if (length(comments)) readLines(con2, length(comments) - 1)
        con2
    })
}
hs.makeBatch <- function(V, N = wk.xcs, K = {
}, name = 1, qtl = 1) {
    if (length(V) == 0) 
        return()
    if (length(V) == 1) 
        return(list(V))
    if (!length(K)) {
        K <- length(V)/N
    }
    f1 <- if (length(V)%%K) {
        breaks <- seq(1, length(V), by = K) %or% length(V)
        cut(V, breaks, include.lowest = TRUE)
    }
    else {
        as.integer(base::gl(length(V)/K + K, K, length(V)))
    }
    R <- tapply(V, f1, list)
    if (name) {
        R
    }
    else unname(R)
}
hs.doBatch <- function(V, hs = identity, n = wk.xcs, k = {
}, name = 1, reduce = 0, hs1 = hs) {
    H <- hs.makeBatch(V, N = n, K = k)
    R <- wk.sapply(H, hs, nc = n)
    if (reduce) 
        R <- do.call(hs1, unname(R))
    R
}
hs.is_empty_arg <- function(x) is(x, "name") && identical(paste(x), "")
hs.switch <- function(o1, ..., env1 = parent.frame()) {
    n1 <- ...length()
    lang <- substitute(c(...))
    matched <- FALSE
    hs.eq <- if (is.list(o1)) {
        function(x1, x2) {
            (length(x1) == length(x2)) && {
                i <- seq_along(x1)
                all(sapply(i, function(i1) {
                  a1 <- x1[[i1]]
                  a2 <- x2[[i1]]
                  (length(a1) == length(a2)) && all(a1 == a2)
                }))
            }
        }
    }
    else function(x1, x2) {
        (length(x1) == length(x2)) && all(x1 == x2)
    }
    for (i in hs.seqForward(1, n1 - 1, by = 2)) {
        if (matched) 
            if (!hs.is_empty_arg(lang[[i + 2]])) 
                return(...elt(i + 1))
            else next
        if (hs.eq(o1, ...elt(i))) {
            matched <- TRUE
            if (!hs.is_empty_arg(lang[[i + 2]])) 
                return(...elt(i + 1))
        }
    }
    if (n1%%2) 
        ...elt(n1)
}
hs.switch.hs <- function(AL, ..., no_match = NULL, env1 = parent.frame()) {
    if (!is.list(AL)) 
        AL <- list(AL)
    n1 <- ...length()
    lang <- substitute(c(...))
    matched <- FALSE
    for (i in seq(1, n1 - 1, by = 2)) {
        if (do.call(...elt(i), AL)) 
            matched <- TRUE
        if (hs.all(matched, !hs.is_empty_arg(lang[[i + 2]]))) 
            return(...elt(i + 1))
    }
    if (n1%%2) 
        ...elt(n1)
}
hs.cond <- function(...) {
    n1 <- ...length()
    lang <- substitute(c(...))
    matched <- FALSE
    for (i in seq(1, n1 - 1, by = 2)) {
        if (...elt(i)) 
            return(...elt(i + 1))
    }
    if (n1%%2) 
        ...elt(n1)
}
hs.case <- hs.cond
hs.if <- `if`
hs.repeat <- function(exp1, env1 = parent.frame()) {
    eval(substitute(repeat exp1), envir = env1)
}
hs.on.exit <- function(..., env1 = parent.frame()) {
    evalq(substitute(on.exit(...)), envir = env1)
    eval(bquote(.(substitute(on.exit(...)))), envir = env1)
    evalq(bquote(.(substitute(on.exit(...)))), envir = env1)
    do.call(evalq, list(bquote(.(substitute(on.exit(...))))))
}
hs.case.over <- function(over, hs, R = {
}, no_match = NULL, env1 = parent.frame(), env2 = if (local) new.env(parent = env1) else env1, local = 1) {
    if (!is.list(hs)) 
        hs <- list(hs)
    N <- max(length(over), length(hs), length(R))
    for (i1 in seq_len(N)) {
        r1 <- hs[[hs.recycle(i1, length(hs))]](over[[hs.recycle(i1, length(over))]])
        if (r1) 
            return(if (length(R)) R[[hs.recycle(i1, length(R))]] else r1)
    }
    no_match
}
hs.for0 <- function(..., .env = parent.frame()) {
    lang <- substitute(hs.for(...))
    for (i in hs.seqForward(2, length(lang))) {
        if (is(lang[[i]], "name")) {
            do.call("for", list(lang[[i]], lang[[i + 1]], lang[-seq(2, i + 1)]), envir = .env)
            break
        }
        else ...elt(i - 1)
    }
}
hs.for <- function(..., .env = parent.frame(), .named = {
}, .local = 0) {
    if (.local) 
        .env <- new.env(parent = .env)
    if (length(.named)) {
        e1 <- substitute(hs.catch(hs.for0(...), named = .named))
        eval(e1, envir = .env)
    }
    else hs.for0(..., .env = .env)
}
hs.screen.clear <- function(n1 = 9) {
    cat(rep("", n1), sep = "\n")
}
hs.deInf <- function(v1) {
    na.i <- is.na(v1)
    for (fx1 in c("+", "-")) {
        switch(fx1, `+` = {
            i1 <- which(is.infinite(v1) & (v1 > 0))
            if (length(i1)) {
                a1 <- max(v1[-i1], na.rm = 1)
                v1[i1] <- a1
            }
        }, `-` = {
            i1 <- which(is.infinite(v1) & (v1 < 0))
            if (length(i1)) {
                a1 <- min(v1[-i1], na.rm = 1)
                v1[i1] <- a1
            }
        })
    }
    v1
}
order.hs <- function(..., order = -1, na.last = 1) {
    tbl <- as.data.table(list(...))
    tbl[, `:=`("i.36_03_11_203708", 1:.N)]
    setorderv(tbl, names(tbl) %rm% "i.36_03_11_203708", order = order, na.last = as.logical(na.last))
    tbl[, i.36_03_11_203708]
}
set.clean.hs <- function(V) V %not% c("", NA)
hs.clean <- function(V, U = 0, dirty = c("", NA, NaN), dirty.extra = {
}, dirty.pattern = {
}, ot = if (is.list(V)) "l" else "v", inv = 0, how_many = "any") {
    if (length(dirty.extra)) 
        dirty <- dirty %or% dirty.extra
    if (is.list(V)) {
        jg <- lapply(V, hs.clean, dirty = dirty, dirty.extra = dirty.extra, ot = "l", inv = inv)
        cleanN <- hs.rowSums(jg)
        jg <- switch(how_many, any = cleanN > 0, all = cleanN == length(V))
        return(if (inv) !jg else jg)
    }
    if (U) 
        V <- unique(V)
    is_clean <- V %not.in% dirty
    for (p1 in dirty.pattern) {
        i <- which(is_clean)
        if (!length(i)) 
            break
        i1 <- hs.grep(V[i], p1)
        is_clean[i[i1]] <- FALSE
    }
    switch(ot, v = if (inv) V[!is_clean] else V[is_clean], i = which(if (inv) !is_clean else is_clean), l = if (inv) !is_clean else is_clean, n = sum(if (inv) !is_clean else is_clean), t = hs.table(if (inv) !is_clean else is_clean))
}
hs.dirty <- function(...) {
    hs.clean(..., inv = 1)
}
hs.qp <- function() cat(rep("", 30), sep = "\n")
hs.matrix_to_list <- function(m1, side = "column", name = 0) {
    jg <- switch(side, column = lapply(1:ncol(m1), function(i1) m1[, i1]), row = lapply(1:nrow(m1), function(i1) m1[i1, ]))
    if (name) {
        switch(side, column = {
            if (length(colnames(m1))) names(jg) <- colnames(m1)
        }, row = {
            if (length(rownames(m1))) names(jg) <- rownames(m1)
        })
    }
    else {
        names(jg) <- {
        }
    }
    jg
}
hs.ij <- hs.matrix_ij <- local({
    hs.flag <- function(i, m, side = 1) {
        (!length(i)) || isTRUE(i) || {
            if (is.logical(i)) {
                all(i)
            }
            else {
                N <- if (side == 1) {
                  nrow(m)
                }
                else if (side == 2) {
                  ncol(m)
                }
                if (length(i) != N) 
                  return(FALSE)
                if (is.numeric(i)) {
                  !is.unsorted(i)
                }
                else if (inherits(i, "character")) {
                  all(i == hs.switch(side, 1, rownames(m), 2, colnames(m)))
                }
            }
        }
    }
    function(m, i = {
    }, j = {
    }, drop = FALSE, ...) {
        flag_i <- hs.flag(i = i, m = m, side = 1)
        flag_j <- hs.flag(i = j, m = m, side = 2)
        if (inherits(m, "data.table")) {
            if (...length()) {
                hs.browser("39.01.17.131158", debug = 0)
                A <- list(...)
                exp_l
                wk.dt.ij(m, exp_l = exp_l)
                hs.switch(c(flag_i, flag_j), c(0, 1), m[i], c(1, 0), m[, .SD, .SDcols = c(j)], c(0, 0), m[i, .SD, .SDcols = c(j)], m)
                wk.dt.ij(m, ...)
            }
            else hs.switch(c(flag_i, flag_j), c(0, 1), m[i], c(1, 0), m[, .SD, .SDcols = c(j)], c(0, 0), m[i, .SD, .SDcols = c(j)], m)
        }
        else if (inherits(m, c("matrix", "data.frame"))) {
            drop <- as.logical(drop)
            hs.switch(c(flag_i, flag_j), c(0, 1), m[i, , drop = drop], c(1, 0), m[, j, drop = drop], c(0, 0), m[i, j, drop = drop], m)
        }
        else hs.switch(c(flag_i, flag_j), c(0, 1), m[i, ], c(1, 0), m[, j], c(0, 0), m[i, j], m)
    }
})
hs.i <- function(v, i, vt = if (inherits(v, "character") && (length(v) == 1) && (!length(names(v))) && inherits(i, "character") && (length(i) == 1) && (!length(names(i)))) "pkg") {
    if (missing(i) && inherits(v, "character") && (length(v) == 1) && (!length(names(v)))) {
        return(packageDescription(v))
    }
    if (length(vt) && (vt == "pkg")) 
        return(do.call(`::`, list(v, i)))
    v_len <- if (length(dim(v)) == 2) 
        dim(v)[2]
    else length(v)
    if (inherits(i, "character")) {
        i <- match(i, names(v))
    }
    else {
        ii <- which(i < 0)
        if (length(ii)) 
            i[ii] <- v_len + (i[ii] + 1)
        ii <- which(i > v_len)
        if (length(ii)) {
            r <- i[ii]%%v_len
            i[ii] <- ifelse(r > 0, r, v_len)
        }
    }
    hs.switch(length(dim(v)), 0, , 1, v[i], 2, hs.ij(v, , i), hs.browser("2023.08.02.141804", debug = 0))
}
hs.put2end <- function(x, i) {
    db <- seq_along(x)
    x[c(db[-i], db[i])]
}
popular.hs <- function(L, N = length(L) - D, D = 0, exclude = c("", NA)) {
    if (hs.len1(L)) 
        if (N == 1) 
            return(unlist(L))
    L2 <- inverseList(L)
    names(L2)[sapply(L2, length) >= N] %not% exclude
}
hs.getter <- function(o1, name = {
}, name.pattern = {
}, always.list = 0) {
    if (inherits(o1, "function")) 
        o1 <- environment(o1)
    if (inherits(o1, "environment")) {
        if (length(name.pattern)) 
            name <- ls(envir = o1, pattern = name.pattern)
        if (length(name)) {
            if ((length(name) == 1) && inherits(name, "character")) 
                return(o1[[name]])
            if (name == -1) {
                mget(ls(o1), o1)
            }
            else mget(name, o1)
        }
        else ls(o1)
    }
    else {
        if (length(name.pattern)) 
            name <- hs.grepV(names(o1), name.pattern)
        if (length(name)) {
            if (identical(name, -1)) {
                o1
            }
            else o1[name]
        }
        else names(o1)
    }
}
hs.matrix.merge <- function(m1, rng = {
}, cng = {
}) {
    if (length(rng)) {
        a1 <- lapply(rng, function(.) {
            colSums(m1[., , drop = 0])
        })
        jg <- do.call("rbind", a1)
        rownames(jg) <- local({
            . <- rng
            if (length(names(.))) {
                names(.)
            }
            else {
                sapply(., function(.) {
                  a1 <- if (is.numeric(.)) {
                    rownames(m1)[.]
                  }
                  else .
                  paste(a1, collapse = "+")
                })
            }
        })
        m1 <- jg
    }
    if (length(cng)) {
        m1 <- t(hs.matrix.merge(t(m1), cng))
    }
    m1
}
hs.cut <- function(x1, ..., lo = include.lowest, include.lowest = TRUE, hi = right, right = !include.lowest) {
    i1 <- hs.clean(x1, ot = "i")
    if (length(i1)) {
        if (length(i1) < length(x1)) {
            x1[i1] <- cut(x1[-i1], ..., include.lowest = lo, right = hi)
            return(x1)
        }
        else return(cut(x1, ..., include.lowest = lo, right = hi))
    }
    else return(x1)
}
hs.cut_fine <- function(x1, breaks) {
    if (any(hs.grepl(breaks, "Inf"))) 
        hs.browser("38.01.02.091035")
    breaks <- unique(breaks)
    N <- as.numeric(hs.sub(breaks, "[\\(\\)\\[\\]]"))
    jg <- rep(NA, length(x1))
    i1 <- order(N)
    if (is.unsorted(i1)) {
        breaks <- breaks[i1]
        N <- N[i1]
    }
    boundary <- hs.re(breaks, "[\\(\\)\\[\\]]")
    L <- rep("", length(breaks) + 1)
    b2 <- boundary[1]
    n2 <- N[1]
    if (b2 %in% c("(", "]")) {
        L[1] <- paste0("<=", n2)
        i1 <- x1 <= n2
    }
    else {
        L[1] <- paste0("<", n2)
        i1 <- x1 < n2
    }
    if (any(i1)) 
        jg[i1] <- L[1]
    for (i1 in hs.seqForward(2, length(breaks))) {
        n1 <- n2
        b1 <- b2
        b2 <- boundary[i1]
        n2 <- N[i1]
        bb <- paste0(b1, b2)
        if (0) {
            hs1 <- function(p1) {
                strsplit(p1, "")[[1]] %=>% lapply(. %=>% {
                  c(x, switch(x, `(` = "]", `[` = ")", `)` = "[", `]` = "("))
                }) %=>% {
                  do.call(expand.grid, x)
                } %=>% apply(1, paste, collapse = "") %=>% .q %=>% paste(collapse = ", ") %=>% {
                  paste0("c( ", x, ")")
                } %=>% .ss
            }
            null <- sapply(c("()", "[)", "[]", "(]"), hs1)
        }
        nn <- paste(n1 %or% n2, collapse = ", ")
        if (bb %in% c("()", "])", "([", "][")) {
            i2 <- (x1 > n1) & (x1 < n2)
            L[i1] <- paste0("(", nn, ")")
        }
        else if (bb %in% c("[)", "))", "[[", ")[")) {
            i2 <- (x1 >= n1) & (x1 < n2)
            L[i1] <- paste0("[", nn, ")")
        }
        else if (bb %in% c("[]", ")]", "[(", ")(")) {
            i2 <- (x1 >= n1) & (x1 <= n2)
            L[i1] <- paste0("[", nn, "]")
        }
        else if (bb %in% c("(]", "]]", "((", "](")) {
            i2 <- (x1 > n1) & (x1 <= n2)
            L[i1] <- paste0("(", nn, "]")
        }
        if (any(i2)) 
            jg[i2] <- L[i1]
    }
    b1 <- boundary[length(breaks)]
    n1 <- N[length(N)]
    if (b1 %in% c("[", ")")) {
        L[length(L)] <- paste0(">=", n1)
        i1 <- x1 >= n1
    }
    else {
        L[length(L)] <- paste0(">", n1)
        i1 <- x1 > n2
    }
    if (any(i1)) 
        jg[i1] <- L[length(L)]
    hs.factor(jg, L %and% jg, N = 0)
}
hs.top <- function(x1, N, end = "left", ot = "v", sort = 1) {
    l1 <- length(x1)
    if (N < 1) 
        N <- floor(l1 * N)
    N <- as.integer(N)
    if (sort) {
        i1 <- head(order(x1, decreasing = switch(end, left = FALSE, right = TRUE)), N)
    }
    else {
        i1 <- switch(end, left = seq.int(N), right = {
            seq.int(l1 - N + 1, l1)
        })
    }
    switch(ot, v = x1[i1], i = i1)
}
hs.hs <- function(x1, ...) {
    lapply(list(...), function(hs1) hs1(x1))
}
hs.factor <- function(v1, levels = as.character(unique(v1)), N = 1, nl = 1, ref = {
}, sort = 0, ot = "xxx", ...) {
    if (ot == "str") 
        return(levels(v1)[v1])
    if (length(ref)) 
        levels <- ref %or% levels
    if (is.factor(v1)) {
        levels <- levels %and% levels(v1)
        tbl <- hs.table(v1)
        if (0 %in% tbl) {
            level2drop <- names(tbl)[tbl == 0]
            levels <- levels[levels %not.in% level2drop]
        }
    }
    else {
        levels <- levels[levels %in% v1]
    }
    if (!(is.factor(v1) && all(levels(v1) == levels))) {
        v1 <- factor(v1, levels, ...)
    }
    if (N) {
        v.pct.hs(v1, nl = nl)
    }
    else v1
}
eval. <- function(o1, ..., env1 = parent.frame()) {
    o1.exp <- substitute(o1)
    eval(bquote({
        eval(substitute(.(substitute(...)), list(. = quote(.(o1.exp)))))
    }), env1)
}
isTrue.hs.on <- function(on1, ..., env1 = parent.frame()) {
    L <- list(...)
    if (on1 %in% names(L)) {
        as.logical(L[[on1]])
    }
    else exists(on1, envir = env1, inherits = FALSE) && get(on1, envir = env1)
}
hs.intersection <- function(...) {
    sj <- list(...)
    jg <- sj[[1]]
    if (length(sj) > 1) {
        for (sj1.i in 2:length(sj)) {
            if (!length(jg)) 
                break
            jg <- jg %and% sj[[sj1.i]]
        }
    }
    jg
}
hs.list.common <- function(L) {
    names(L) <- seq(length(L))
    L2 <- inverseList(L)
    names(L2)[sapply(L2, length) == length(L)]
}
hs.as.numeric <- function(x1) {
    switch(as.character(length(dim(x1))), `0` = {
        jg <- as.numeric(x1, ...)
        if (length(names(x1))) names(jg) <- names(x1)
        jg
    }, `2` = {
        matrix(as.numeric(x1), ncol = ncol(x1), dimnames = dimnames(x1))
    })
}
isList <- function(x) {
    (typeof(x) == "list") && (length(dim(x)) < 2)
}
chk <- local({
    chk.nonList <- function(x, howMany = 3) {
        a <- dim(x)
        if (!length(a)) 
            a <- length(x)
        do.call("[", append(list(x), append(lapply(a, function(n) {
            1:min(n, howMany)
        }), list(drop = FALSE))))
    }
    function(x, nlines = 9, width = 125, dim.n = 3) {
        if (substitute(x) %=>% {
            inherits(x, "character") && (length(x) == 1)
        }) {
            return(help(package = (x)))
        }
        if (dim.n != 3) {
            chk.nonList <<- function(x, howMany = dim.n) {
                a <- dim(x)
                do.call("[", c(list(x), sapply(a, function(n) {
                  1:min(n, howMany)
                }), list(drop = FALSE)))
            }
        }
        f <- c("length", "dim", "names", "typeof", "class")
        sapply(f, function(f) {
            .ss(sprintf("\t**** %s ****", f))
            y <- f %(% list(x)
            .s(y, max.print = 9)
        })
        flag <- 0
        if (isList(x)) {
            y <- sapply(x, length)
            .ss("length of element: ")
            .s(y, max.print = 9)
            .ss("sum of numbers of element: ")
            .s(sum(y), max.print = 9)
            flag <- listPrintLinesCounts(x)
        }
        if (flag < 999) {
            .ss("\t**** contents ****")
            if (is.list(x) & !is.data.frame(x)) {
                x <- rapply(x, chk.nonList, how = "list")
            }
            else x <- chk.nonList(x, dim.n)
            zz <- textConnection(capture.output(.s(head(x, nlines))))
            cat(sapply(readLines(zz, nlines), toString, width), sep = "\n")
            close(zz)
        }
    }
})
chk2 <- function(M, Nline = 20, width = 128) {
    dm <- dim(M)
    if (is.null(dm)) 
        ieiwk.cat.line(c("length: ", length(M)))
    else ieiwk.cat.line(c("dim: ", dim(M)))
    ieiwk.cat.line(c("type: ", typeof(M)))
    ieiwk.cat.line(c("class: ", class(M)))
    if (length(M) > 9 | (class(M) %in% c("matrix", "data.frame"))) {
        a <- capture.output(.s(head(M, max(10, Nline))))
    }
    else {
        a <- capture.output(.str(M))
    }
    i <- as.integer({
        width - 3
    }/2)
    a <- sapply(a, function(x) {
        if (nchar(x) > width) {
            i <- min(as.integer({
                nchar(x) - 3
            }/2), i)
            c(substr(x, 1, i), substring(x, nchar(x) - i)) %,% "..."
        }
        else x
    })
    .ss(head(a, Nline))
}
replicate.hs.group <- function(lv, sep = "") {
    . <- lv
    . <- tapply(., ., function(.) {
        paste0(., sep, seq_along(.))
    })
    unlist(unname(.))
}
file.maker <- function(target, dependency, hs1) {
    if (FALSE %in% file.exists(target)) {
        hs1(target, dependency)
    }
    target
}
obj.maker <- function(..., env1 = parent.frame(), env2 = new.env(parent = env1)) {
    oh <- list(...)
    if (length(oh)) {
        if (FALSE %in% sapply(oh[[1]], hs.exists, env1)) {
            if (length(oh) > 2) 
                do.call(obj.maker, append(oh[-c(1, 2)], env1 = env1))
            jg <- oh[[2]]()
            hs.browser("37.01.18.234903")
        }
    }
}
hs.order <- order10 <- function(..., decreasing = fx < 0, fx = -1, use = {
}) {
    A <- list(...)
    if (inherits(A[[1]], "matrix")) 
        A[[1]] <- hs.df(A[[1]])
    if (inherits(A[[1]], c("list", "data.frame"))) 
        A <- c(as.list(unname(hs.ij(A[[1]], , j = use))), A[-1])
    do.call(order, c(A, list(decreasing = as.logical(decreasing))))
}
hs.sort <- sort10 <- function(..., decreasing = fx < 0, fx = -1, use = {
}) {
    if (inherits(..1, "data.table")) {
        use <- if (length(use)) {
            if (is.numeric(use)) {
                names(..1)[use]
            }
        }
        else names(..1)
        return(wk.dt.setorderv(..1, cols = use, order = fx))
    }
    i <- hs.order(..., decreasing = decreasing, use = use)
    xi <- match("x", ...names(), nomatch = 1)
    x <- ...elt(xi)
    if (length(dim(x)) == 2) 
        return(hs.ij(x, i))
    return(x[i])
}
hs.apply <- function(m1, side, hs1, ...) {
    hs.switch(side, 1, lapply(seq_len(nrow(m1)), function(i1) hs1(m1[i1, ], ...)), 2, lapply(seq_len(ncol(m1)), function(i1) hs1(m1[, i1], ...)), no_match = lapply(asplit(m1, side), hs1, ...))
}
hs.match.from_tail <- function(v1, t1, n1 = 10000L) {
    t2 <- rev(tail(t1, n1))
    m1 <- length(t2) - match(v1, t2) + 1L
    if (length(t2) < length(t1)) 
        m1 <- length(t1) - length(t2) + m1
    m1
}
hs.rename <- function(x1, n1, n2, hs1 = names) {
    i1 <- match(n1, hs1(x1))
    eval(bquote(.(substitute(hs1))(x1)[i1] <- n2))
    x1
}
hs.version_newest <- function(...) {
    ver0 <- unique(c(...))
    max_i <- 1
    ver <- hs.re(ver0, "\\d.*")
    for (i1 in hs.seqForward(2, length(ver0))) if (utils::compareVersion(ver[max_i], ver[i1]) < 0) 
        max_i <- i1
    ver0[max_i]
}
hs.catch <- local({
    env.38_01_04_223523 <- new.env()
    evalq(hs.throw <- function(...) {
        acc.38_01_04_221205[["add"]](...)
        return()
    }, env.38_01_04_223523)
    function(exp1, named = 0, envUp = parent.frame()) {
        on.exit(evalq(rm(list = ls(all.names = TRUE, sorted = FALSE) %not% "hs.throw"), env.38_01_04_223523), add = TRUE)
        if (!identical(parent.env(env.38_01_04_223523), envUp)) 
            parent.env(env.38_01_04_223523) <- envUp
        env.38_01_04_223523[["acc.38_01_04_221205"]] <- if (named) 
            namedExpandingList()
        else expandingList()
        eval(substitute(exp1), envir = env.38_01_04_223523)
        env.38_01_04_223523[["acc.38_01_04_221205"]][["get"]]()
    }
})
hs.replace <- function(x, old, new = {
}) {
    if (!length(new)) {
        i <- seq(2, length(old), by = 2)
        new <- old[i]
        old <- old[-i]
    }
    if (inherits(x, "factor")) {
        i <- which(old %in% levels(x))
        if (length(i)) {
            if (length(i) < length(old)) {
                old <- old[i]
                new <- new[i]
            }
            old <- match(old, levels(x))
            levels(x) <- replace(levels(x), old, new)
        }
        x
    }
    else {
        paste(hs.replace(factor(x), old, new))
    }
}
hs.merge <- function(x, y = {
}, sort = FALSE, all.x = TRUE, by = intersect(names(x), names(y)), by.x = by, by.y = by, ...) {
    if (length(y)) {
        if (length(names(by)) && missing(by.x) && missing(by.y)) {
            by.x <- names(by)
            by.y <- by
        }
        merge(x = x, y = y, sort = sort, all.x = all.x, by.x = by.x, by.y = by.y, by = by, ...)
    }
    else {
        if (inherits(x, "list") && (!inherits(x, "data.frame"))) {
            return(hs.list.merge(x))
        }
        else hs.browser("2023.07.25.162653", debug = 0)
    }
}
"%which.not.in%" <- function(x1, x2) {
    which(x1 %not.in% x2)
}
"%which.in%" <- function(x1, x2) {
    which(x1 %in% x2)
}
"%rm%" <- function(x, y) {
    if (any(typeof(y) %in% c("language", "expression", "symbol"))) {
        x[eval(as.cmd(sprintf(ifelse(length(grep("%s", y[[1]])), y[[1]], paste("%s", y[[1]])), "x")))]
        x[!tfs(c(deparse(substitute(x)), y[[1]]) %,% "")]
    }
    else {
        x[x %not.in% y]
    }
}
"%not.in%" <- function(x, y) !(x %in% y)
"%all.in%" <- function(x, y) all(x %in% y)
"%not.all.in%" <- function(x, y) !all(x %in% y)
"%none.in%" <- function(x, y) !any(x %in% y)
"%any.in%" <- function(x, y) any(x %in% y)
"%keep%" <- function(x, y) {
    if (any(typeof(y) %in% c("language", "expression"))) {
        x[eval(as.cmd(sprintf(ifelse(length(grep("%s", y[[1]])), y[[1]], paste("%s", y[[1]])), "x")))]
    }
    else {
        x[x %in% y]
    }
}
hs.len1 <- function(x, y = 1L) {
    length(x) == y
}
hs.of_length <- hs.len1
merge_list <- function(old, new = {
}, update = 1) {
    nm <- names(old) %and% names(new)
    if (length(nm)) {
        if (anyDuplicated(names(old)[names(old) %in% nm])) 
            stop("[37_04_25_165259] the \"old\" list has duplicated names")
        if (anyDuplicated(names(new)[names(new) %in% nm])) 
            stop("[37_04_25_165504] the \"new\" list has duplicated names")
    }
    if (length(nm) && update) {
        old[nm] <- new[nm]
    }
    c(old, if (length(nm)) new[names(new) %not.in% nm] else new)
}
"%set==%" <- function(v1, v2) {
    (v1 %all.in% v2) && (v2 %all.in% v1)
}
"%not%" <- setdiff
"%and%" <- intersect
"%or%" <- union
hs.xor <- function(v1, v2, ot = "l") {
    switch(ot, l = {
        a <- list(v1 %not% v2, v2 %not% v1)
        if (all(sapply(a, length) == 0)) a <- {
        }
        a
    }, n = c(sum(v1 %not.in% v2), sum(v2 %not.in% v1)))
}
hs.msg <- function(msg = "", ln = 0, lw = 10, ot = {
}) {
    msg %<=>% paste(sprintf(paste0("%-", lw, "s"), if (ln) 
        paste(rep("*", ln), collapse = "")
    else ""), "[", Sys.time(), "] ", x)
    hs.switch(ot, "s", msg, message(msg))
}
.muid4machine <- .m4m <- function(names, n = 1, interval = 0.01, precision = 1e+09, sep = if (precision > 0) "." else "_", sep1 = "_", hz = {
}, time_pattern = paste(c("%Y", "%m", "%d", "%H%M%S"), collapse = sep)) {
    birth.time <- "2010-09-15 22:26:51 CST"
    if (!missing(names)) 
        n <- length(names)
    r <- {
    }
    while ({
        i <- length(r)
    } < n) {
        n1 <- n - i + 1
        r[seq(i + 1, n)] <- sapply(seq(i + 1, n), function(i) {
            a <- Sys.time()
            y1 <- as.integer(re("\\d+", a)) - 1984
            paste(c(sub("\\d+", y1, format(a, time_pattern)), if (precision > 0) c(sep1, as.numeric(difftime(Sys.time(), birth.time, units = "secs") * precision))), collapse = "")
        })
        r <- unique(r)
    }
    if (!missing(names)) 
        names(r) <- names
    r
}
hs.uid.fp <- function(base_nyr = 1) {
    a <- Sys.time()
    y1 <- as.integer(re("\\d+", a)) - 1984
    f1 <- if (base_nyr) 
        "aaaa/%m/%d/aaaa_%m_%d_%H%M%S"
    else "aaaa/%m/%d/%H%M%S"
    hs.gsub(format(a, f1), "aaaa", y1)
}
hs.seed_generate <- function() {
    a1 <- hs.sub(.muid4machine(), "_\\d+$")
    a1 <- strsplit(a1, "\\.")[[1]]
    as.integer(hs.pastec0(a1[-1]))
}
today_file_path <- function(ext1 = "org", root = "") {
    jg <- hs.fp(as.integer(format(Sys.Date(), "%Y")) - 1984L, format(Sys.Date(), paste0("%m/%d.", ext1)))
    if (nzchar(root)) 
        hs.fp(root, jg)
    else jg
}
osType <- function() {
    Sys.info()["sysname"]
}
wj_rsync <- switch(osType(), Linux = "rsync", Darwin = "/usr/local/bin/rsync")
.sys <- function(..., sep = " ", wait = 1, intern = 1, chk = 1, dz = {
}, pipe = {
}, fifo = 0, report = 0, file, file.append = TRUE, use.tmpfile = 0) {
    cmd <- {
        A <- c(...)
        if (inherits(A, "list")) {
            A <- unlist(A)
        }
        a <- paste(A, collapse = sep)
        a <- hs.gsub(a, sprintf("%s\n%s", sep, sep), "\n")
        if (hs.grepl(a, "^cmd")) 
            a <- hs.gsub(a, "\\s*&&\\s*", "&&")
        a
    }
    if (!missing(file)) {
        .cat(cmd, file = file, append = file.append)
        return(invisible())
    }
    if (length(pipe)) {
        con <- pipe(cmd, pipe)
        return(con)
    }
    if (fifo) 
        return(c("<(", cmd, ")"))
    if (chk) {
        if (intern) {
            cmd
        }
        else message(cmd)
    }
    else {
        if (report) {
            r <- system(cmd, wait = TRUE, intern = as.logical(intern))
            .r()
            if (intern) 
                r
        }
        if (use.tmpfile) {
            f1 <- hs.wj.temp()
            on.exit(hs.rm.wj(f1), add = TRUE)
            cat(cmd, file = f1, sep = "\n")
            aaa <- system2("/bin/bash", f1, wait = as.logical(wait), stdout = if (intern) 
                TRUE
            else "")
        }
        else {
            aaa <- .try(system(cmd, wait = as.logical(wait), intern = as.logical(intern)))
        }
        aaa
    }
}
.sys.io.2wj <- function(wj1, cmd = "", Z = grepl("\\.gz$", wj1), append = 0) {
    pipe(.sys(cmd, if (Z) 
        c(if (cmd != "") "|", "gzip"), ifelse(append, ">>", ">"), hs.fpGood(wj1)), "w")
}
wk.io.lines.2wj <- function(L, wj1, ...) {
    io1 <- .sys.io.2wj(wj1, ...)
    on.exit(close(io1))
    cat(L, file = io1, sep = "\n", ...)
    wj1
}
.r <- ieiwk.email.job.is.done <- function(title = Sys.time(), title.postfix = if ("difftime" %in% class(title)) "in total" else ", done at", ps.pattern, nodes = cluster.all(), chk.ps = 0, update.time = 0, time.diff = {
}, time.diff.2 = {
}, content = "bbb") {
    if (.hostname() %in% c("wk-NBLK-WAX9X", "wk-MS-7A21")) {
        .wk("wps", wk.wjlj(paste0(title, title.postfix)), "&")
        return()
    }
    if (0) 
        if (!length(time.diff)) 
            time.diff <- local({
                wj1 <- .mfp("33.06.06.230513.rd")
                if (file.exists(wj1) && !update.time) 
                  hs.load(wj1)
                else {
                  n1 <- "192.115.110.254"
                  x0 <- ieiwk.eval(quote(list(readLines("https://time.is"), Sys.time())), intern = 1, n1 = n1)
                  x <- x0[[1]]
                  x1 <- x[wk.re(x, "Exact time now:", "which") + 1]
                  a <- wk.re(x1, "twd", "sub")
                  i <- wk.re(a, ">", "locate") + 1
                  t1 <- substr(a, i, i + 7)
                  d1 <- do.call(".j", c(wk.re(wk.re(x1, "title", "sub", sub.len = 50), "\\d+", all = 1), collapse = "-"))
                  dt1 <- .jc(d1, t1, sep = " ")
                  dt2 <- x0[[2]]
                  r <- difftime(dt1, dt2)
                  ieiwk.save("r", wj1)
                  r
                }
            })
    if (0) 
        if (!length(time.diff.2)) 
            time.diff.2 <- local({
                wj1 <- .mfp("33.06.06.231803.rd")
                if (file.exists(wj1) && !update.time) 
                  hs.load(wj1)
                else {
                  n1 <- "log01"
                  t1 <- ieiwk.eval(quote(Sys.time()), n1 = n1, intern = 1)
                  t2 <- Sys.time()
                  r <- difftime(t1, t2)
                  ieiwk.save("r", wj1)
                  r
                }
            })
    if (chk.ps) {
        while (wdjc.search(ps.pattern, nodes)) {
        }
    }
    title <- paste0(title, title.postfix)
    hn <- Sys.info()["nodename"]
    n1 <- ieiwk("host.with.internet")
    cmd1 <- bquote({
        .wk("echo", .q(.(content)), "| mutt -s", .q(.(title)), "wengkai_report@163.com")
    })
    if (hn == n1) {
        eval(cmd1)
        return()
    }
    ieiwk.eval(cmd1, bg = 0, n1 = n1, via.file.result = 0, email = 0)
}
.wk <- .sd <- .sys.do <- function(..., intern = 0, wait = 1, log = FALSE, log.append = TRUE, log.cmd = TRUE, target, ssh = Sys.info()["nodename"], use.tmpfile = 1, fifo = 0, fifo.4r = 0, report.via.email = 0, email = report.via.email, redo = 0) {
    if (is.logical(log) && log) 
        log <- hs.wjm(dn = hs.home.expand("~/log"), .muid4machine(), ext = "log")
    if (is.character(log) && file.exists(log) && {
        !redo
    }) 
        return()
    args <- {
        a <- lapply(substitute(list(...)), function(x) c(x))[-1]
        if (is.null(names(a))) {
            a1 <- a
            a2 <- list()
        }
        else {
            a1 <- a[names(a) == ""]
            a2 <- a[names(a) != ""]
        }
        env1 <- parent.frame()
        append(lapply(unlist(a1), eval, env1), a2)
    }
    args[["intern"]] <- intern
    if (osType() == "Windows") 
        args <- c("cmd /C", args)
    if (is.character(log)) {
        a <- ieiwk.file.path(log)
        .cat.time(file = log, append = log.append)
        if (log.cmd) {
            cmd <- local({
                args["intern"] <- 1
                do.call(.sys, args)
            })
            hs.say(cmd, file = log)
            hs.cat("run on ", if (nchar(ssh)) 
                ssh
            else Sys.info()["nodename"], file = log)
        }
    }
    if (fifo) {
        f.in <- ieiwk.tempfile()
        .sys.do("mkfifo", f.in)
        args <- append(args, list(">", unname(.sys.fpGood(f.in)), "&"))
        do.call(.sys, c(args, chk = 0, use.tmpfile = use.tmpfile))
        if (fifo.4r) 
            return(.sys("cat", unname(.sys.fpGood(f.in)), pipe = 1))
        else return(f.in)
    }
    r <- do.call(.sys, c(args, chk = 0, use.tmpfile = unname(ifelse(osType() == "Windows", 0, use.tmpfile)), wait = wait))
    if (is.character(log)) {
        if (intern || {
            !intern && !r
        }) 
            .cat.time(file = log)
    }
    if (!missing(target)) 
        cat("done", file = target)
    if (email && !wait) 
        ieiwk.email.job.is.done()
    r
}
.sys.less <- function(...) {
    args <- .sys(...)
    args <- {
        if (file.exists(args)) {
            c("less -N", args)
        }
        else c(args, "| less -N")
    }
    .sys.do(args)
}
.sys.junction <- function(s, t, p = 1, b = "D:/wk/my/rj/Junction/junction.exe") {
    if (p) 
        tp <- t
    tp <- ifelse(p, t, dirname(t))
    t <- file.path(tp, basename(s))
    .sys.do(path.expand(b), .sys.fpGood(t), .sys.fpGood(s))
}
.sys.str <- function(...) {
    .sys(..., intern = 1)
}
.sys.bsub <- function(...) {
    .sys("bsub", shQuote(.sys.str(...)))
}
.wk.bsub.2mem <- function(e1, J = paste0("wk.job", ".", wk.re(wk.re(as.character(Sys.time()), "\\s+", "replace", replacement = ",", all = 1), ":", "replace", replacement = "-", all = 1)), n = 8, R = shQuote(paste0("span[ptile=", min(n, 10), "]"), "cmd"), o = paste0(J, "_output"), e = paste0(J, "_errput"), q = "mem", m = "fat01") {
    wj1 <- ieiwk.tempfile()
    .ss(capture.output(e1), file = wj1)
    wj2 <- ieiwk.tempfile()
    .ss("#!/bin/bash", .sys("Rscript", wj1), .sys("unlink", wj1), .sys("unlink", wj2), file = wj2)
    .wk("bsub", sapply(sort(unlist(strsplit("JnoemqR", ""))), function(x) {
        paste0("-", x, " ", get(x))
    }), "<", wj2)
    c(J, o, e)
}
.sys.bsub.do <- function(...) {
    .sys.do("bsub", shQuote(.sys.str(...)))
}
.sys.man <- function(bin) {
    .ss(.sys("man", bin, intern = 1, chk = 0))
}
.sys.fpGood <- function(wj, home.expand = 1, qn = 1) {
    if (home.expand) 
        wj <- hs.home.expand(wj)
    a1 <- hs.str_split(wj, sep = "/+")
    jg <- sapply(a1, function(a1) {
        a1 <- sapply(a1, function(a1) {
            if (a1 %in% c(".", "..", "~", "")) 
                a1
            else .q(a1, n = qn)
        })
        paste(a1, collapse = "/")
    })
    return(jg)
    if (0) {
        if (hs.grepl(wj, "[ <>]")) 
            wj <- hs.gsub(wj, "([ <>]+)", "'\\1'")
    }
    wj %=>% sapply(. %=>% {
        if (hs.grepl(x, "^((\\.*|~)?/)")) {
            a1 <- strsplit(c("../../boot", "~"), "/")
            a <- sapply(a1, function(a1) {
                a1 <- sapply(a1, function(a1) {
                  if (a1 %in% c(".", "..", "~")) 
                    a1
                  else .q(a1)
                })
                paste(a1, collapse = "/")
            })
        }
        if (substr(x, 1, 2) == "~/") {
            a <- paste0("~/", .q(substring(x, 3)))
            x <- if (qn > 1) 
                .q(a)
            else a
        }
        else if (substr(x, 1, 1) == "/") {
            a <- .q(x)
            x <- if (qn > 1) 
                .q(a)
            else a
        }
        else if (hs.grepl(x, " ")) {
            x <- .q(x)
        }
        x
    })
}
.sys.fpGood.file <- function(f1) {
    if (!length(system(paste("file", f1), intern = TRUE))) 
        f1 <- shQuote(f1)
    f1
}
.sys.view <- function(f) {
    f2 <- .sys.fpGood(f)
    if (file.info(f)$isdir) {
        .sys.do("nautilus --browser", f2, "&")
        return()
    }
    else {
        if (length(grep("\\.pdf$", f))) {
            .sys.do("evince", f2, "&")
        }
    }
}
.sys.l <- function(fp = "") {
    .sys.do("ls -ahl", fp)
}
.sys.nl <- function(fp) {
    a <- sapply(fp, function(fp) .sys("wc -l", fp, intern = 1))
    r <- as.numeric(re("\\d+", a))
    names(r) <- fp
    r
}
.sys.cp <- function(f1, f2, d = 0) {
    .sys.do("cp", f1, f2)
    if (d) 
        .sys.do("rm -r", f1)
    f2
}
.q <- ieiwk.shQuote <- function(s1, n = 1, type = "sh") {
    r <- s1
    while (n) {
        r <- shQuote(r, type = type)
        n <- n - 1
    }
    r
}
.qq <- function(...) {
    .q(..., type = "cmd")
}
ieiwk.fp.cygwin2win <- function(f) {
    a <- tail(strsplit(f, "/")[[1]], -2)
    a[1] <- sprintf("%s:", a[1])
    a %,% "/"
}
ieiwk.fp.win2cygwin <- function(f) {
    p <- re("^[^:]+", f)
    sub("^[^:]+:", sprintf("/cygdrive/%s", p), f)
}
Ncpu <- function(hostname) {
    a <- ieiwk.mpstat()
    c(ceiling({
        as.numeric(a["all", "%idle"])/100
    } * {
        nrow(a) - 1
    } - 0.5), {
        nrow(a) - 1
    })
}
hs.NC <- function() {
    max(parallel::detectCores()/2 - 1, 1)
}
ieiwk.rm <- function(fp, r = 0) {
    system(sprintf("rm %s %s", ifelse(r, "-r", ""), fp))
}
wk.find.wj.2tmp <- function(wjj1 = ".", wj = 1, wjj = 0, wjj.empty = 1, wj2exclude = {
}, follow.symlink = 0, recursive = 0, maxdepth = 1, find.extra = list(), wj.tmp = {
}) {
    if (!length(wj.tmp)) 
        stop("must supply wj.tmp")
    find.par <- c(if (follow.symlink) "-L", .q(wjj1), if (length(maxdepth)) c("-maxdepth", maxdepth))
    r <- {
    }
    if (wj) 
        r <- c(r, .wk("find", find.par, "-type f", intern = 1))
    if (wjj.empty) 
        r <- c(r, .wk("find", find.par, "-type d", "-empty", intern = 1))
    if (wjj) 
        r <- c(r, wjj)
    .ss(r %not% wj2exclude, file = wj.tmp, append = 1)
    if (recursive) {
        folders <- .wk("find", find.par, "-type d", intern = 1)[-1] %not% wj2exclude
        for (folder1 in folders) wk.find.wj.2tmp(folder1, wj2exclude = wj2exclude, follow.symlink = follow.symlink, recursive = recursive, maxdepth = maxdepth, wj.tmp = wj.tmp)
    }
    {
    }
}
wk.list.wj <- function(wjj1 = ".", ms = {
}, ms.bn = {
}, invert = 0, all = 1, wj = 1, wjj = !wj, size.min, qm = 1, recursive = 0, ignore.case = 1, no.. = 1) {
    if (hs.len1(wjj1)) {
        if (0) 
            for (o1 in c("invert", "all", "wj", "wjj", "qm", "recursive", "ignore.case", "no..")) assign(o1, as.logical(get(o1)))
        r <- {
            if (!wj) 
                list.dirs(wjj1, full.names = qm, recursive = recursive)
            else list.files(wjj1, all.files = all, full.names = qm, recursive = recursive, ignore.case = ignore.case, include.dirs = wjj, no.. = no..)
        }
        if (!wjj && !recursive) 
            r <- r[!file.info(r)[, "isdir"]]
        {
            if (length(ms.bn) && length(r)) {
                i <- wk.re(basename(r), ms.bn, "detect")
                r <- r[if (invert) 
                  !i
                else i]
            }
            if (length(ms) && length(r)) {
                i <- wk.re(r, ms, "detect")
                r <- r[if (invert) 
                  !i
                else i]
            }
        }
        r
    }
    else sapply(wjj1, wk.list.wj, ms = ms, ms.bn = ms.bn, invert = invert, all = all, wj = wj, wjj = wjj, size.min = size.min, qm = qm, recursive = recursive, ignore.case = ignore.case, no.. = no..)
}
ieiwk.list.files <- function(path = ".", pattern = NULL, invert = FALSE, all.files = TRUE, only.file = TRUE, only.dir = !only.file, size.min, full.names = TRUE, recursive = FALSE, ignore.case = TRUE, include.dirs = only.dir, no.. = TRUE) {
    a <- list.files(path, pattern, all.files, full.names, recursive, ignore.case, include.dirs, no..)
    b <- list.files(path, pattern, all.files, full.names = TRUE, recursive, ignore.case, include.dirs, no..)
    names(a) <- basename(a)
    if (only.dir) 
        return(a[hs.isDir(b)])
    if (only.file) 
        a <- a[!file.info(b)[, "isdir"]]
    if (!missing(size.min) && is.numeric(size.min)) 
        a <- a[file.info(a)[["size"]] > size.min]
    a
}
ieiwk.list.dirs <- function(wjj = ".", ...) {
    ieiwk.list.files(wjj, ..., only.dir = 1)
}
ieiwk.file.rename <- function(from, to.dir) {
    sapply(from, function(x) file.rename(x, file.path(to.dir, basename(x))))
}
wk.cp.wj2wjj <- function(wj, wjj) {
    for (i1 in seq_along(wj)) {
        file.copy(wj[i1], wjj, )
    }
}
ieiwk.file.copy <- function(from, to.dir, to, copy.time = TRUE, copy.mode = TRUE, ...) {
    sapply(from, function(x) {
        from <- ""
        fp <- list(from = from, to = to, copy.mode = copy.mode, copy.date = copy.time)
        switch(osType(), Linux = {
            file.copy %(% fp[c("from", "to")]
            if (copy.time) {
                Sys.setFileTime(to, ieiwk.mtime(from))
            }
        }, Windows = {
            file.copy %(% fp[c("from", "to", "copy.date")]
        })
    })
}
ieiwk.list.files.bak <- function(path = ".", pattern = NULL, all.files = FALSE, only.file = 0, only.dir = 0, full.names = 1, recursive = 0, ignore.case = 1) {
    browser()
    full.names <- as.logical(full.names)
    recursive <- as.logical(recursive)
    ignore.case <- as.logical(ignore.case)
    r <- list.files(path, pattern, all.files, full.names, recursive, ignore.case, include.dirs = 1)
    if (full.names) 
        a <- ieiwk.fp.norm(a)
    r <- file.path(path, a)
    if (full.names) 
        r <- ieiwk.fp.norm(r)
    names(r) <- a
    if (only.dir) 
        a <- a[file.info(r)[, "isdir"]]
    if (only.file) 
        a <- a[!sapply(r, function(x) {
            file.info(x)$isdir
        })]
    a
}
.sys.ls <- function(d = ".", full = 1, dz = {
}) {
    a <- .sys.do("ls", d, intern = 1, dz = dz)
    r <- file.path(d, a)
    names(r) <- a
    r
}
.sys.scp <- function(fp1, fp2 = fp1, fp1.prefix = .sys.do("echo ~", intern = 1), fp2.prefix = .sys.do("echo ~", intern = 1), l, r = "hh", mb = "r", type.fp = "file", ow = 0) {
    .ss(fp1)
    l.info <- ieiwk.ssh.db(l)
    r.info <- ieiwk.ssh.db(r)
    d1 <- ieiwk.dirname.4home(fp1)
    d2 <- ieiwk.dirname.4home(fp2)
    switch(mb, l = {
        .sys.mkdir(d1)
        if (.sys.file.exists(ieiwk.fp.4sys(fp1))) {
            if (!ow) {
                stop("File exists!")
            }
        }
        .sys.do("scp -r -C -p -P", r.info["port"], sprintf("%s@%s:%s", r.info["user"], r.info["dz"], fp2), fp1)
    }, r = {
        .sys.mkdir(d2, dz = r)
        if (.sys.file.exists(ieiwk.fp.4sys(fp2), dz = r)) {
            if (!ow) {
                stop("File exists!")
            }
        }
        .sys.do("scp -r -C -p -P", l.info["port"], fp1, sprintf("%s@%s:%s", r.info["user"], r.info["dz"], fp2))
    })
}
.sys.scp.2newer <- function(fp1, fp2 = fp1, l, r, type.fp = "file", ow = 0) {
    r.info <- ieiwk.ssh.db(r)
    fn <- switch(judge, modify = .sys.stat.modify, size = .sys.stat.size)
    l.fp.modify <- do.call(fn, list(fp = fp1))
    r.fp.modify <- do.call(fn, list(fp = fp2, dz = dz))
    if (l.fp.modify == r.fp.modify) {
        ieiwk.cat.line("Of the same modification time; no need of transfer.")
        return()
    }
    if (l.fp.modify > r.fp.modify) 
        d <- "r"
    if (l.fp.modify < r.fp.modify) 
        d <- "l"
    .sys.scp(fp1, l = l, r = r, mb = d, ow = 1)
}
.sys.scp.2larger <- function(fp1, fp2 = fp1, l, r, type.fp = "file", ow = 0) {
    r.info <- ieiwk.ssh.db(r)
    fn <- .sys.stat.size
    l.fp.stat <- do.call(fn, list(fp = fp1))
    r.fp.stat <- do.call(fn, list(fp = fp2, dz = r))
    if (l.fp.stat == r.fp.stat) {
        ieiwk.cat.line("Of the same state; no need of transfer.")
        return()
    }
    d <- ifelse(l.fp.stat > r.fp.stat, "r", "l")
    .sys.scp(fp1, fp2, l = l, r = r, mb = d, ow = 1)
}
ieiwk.find.file <- function(pat.txt = c(), dir = ieiwk("meta.dir.code"), ext = "[r|R]", pat.fn = sprintf("\\.%s$", ext), arg.ls = "", recursive = 1, ignore.case = 1, fixed = 0) {
    a <- unlist(sapply(dir, ieiwk.list.files, pat.fn, recursive = 1))
    a <- grep(sprintf("\\.%s$", ext), a, val = 1)
    a <- grep(pat.fn, a, val = 1)
    if (!is.null(pat.txt)) {
        sapply(a, function(x) {
            b <- readLines(x)
            line <- grep(pat.txt, b, fixed = fixed)
            if (length(line)) {
                cat(x, "\n", sep = "")
                apply(cbind(line, ": ", b[line], "\n"), 1, cat, sep = "")
            }
        })
        cat(c())
    }
    else a
}
ieiwk.cpDirTree <- function(from, to) {
    from <- "~/apps/src"
    to <- "~/test"
    if (!file.exists(to)) 
        dir.create(to)
    a <- list.dirs(from, recursive = 1)
    b <- sprintf("%s%s", to, substring(a[-1], nchar(a[1]) + 1))
    sapply(b, dir.create)
    {
    }
}
.sys.backup <- function(from, to, way = "1to1", db = "") {
    from <- "/home/wengkai/F_DRIVE/doc.mobile/code"
    to <- "/home/wengkai/Z_DRIVE/wengkai/backup/f_pan"
    if (file.info(from)$isdir) {
        to2 <- file.path(to, basename(from))
        dir.create(to2, recursive = 1)
    }
    switch(way, `1to1` = {
    })
    file.info(to2)
    .sys.do("file", to2)
    .get(to2, , .sys.do, list("ls"), 2)
    fl <- list.files(from, recursive = 1)
    fl.full <- file.path(from, fl)
    fm.full <- file.path(to2, fl)
    file.copy(from, to, recursive = 1, copy.date = 1)
    null <- sapply(seq(length(fl)), function(i) {
        f1 <- fl.full[i]
        f2 <- fm.full[i]
        file.exists(f2)
        file.copy(f1, f2, copy.date = 1, copy.mode = 1, recursive = 1)
    })
}
alarm <- function(time = 25, time2 = time * 60, text = sprintf("%s minutes past", time)) {
    .sys.do("sleep", time2, "; gedit", text)
}
.cut <- function(f1, col = 1, sep = "\t", sep.out = "\t", o, intern = 0, chk = !intern) {
    .wk("cut", "-d", shQuote(sep), "-f", paste(col, collapse = ","), "--output-delimiter", shQuote(sep.out), if (!missing(o)) 
        o, f1, if (chk) 
        "| head")
}
hs.uniq <- function(v1, which = {
}) {
    switch(as.character(length(v1)), `0` = v1, `1` = {
        if (length(which)) {
            1
        } else v1
    }, {
        a1 <- rle(v1)
        a2 <- a1[["lengths"]]
        if (length(which)) {
            unlist(tapply(seq_along(v1), rep(seq_along(a2), a2), switch(which, first = "head", last = "tail"), 1))
        } else {
            a1[["values"]]
        }
    })
}
wk.wj.read.some.columns <- function(wj1, cn1 = {
}, cn2 = {
}, header = 1) {
    cn <- names(fread(cmd = .sys.cat.skip(wj1, N = 1), header = TRUE))
    if (hs.len1(cn1, 0) && hs.len1(cn2, 0)) 
        return(cn)
    ci <- sort(c(if (length(cn1)) {
        match(cn1, cn)
    }, cn2))
    fread(cmd = .sys(.sys.cat.skip(wj1), "| cut", paste0("-f", wk.integerRange.4cut(ci))), header = as.logical(header))
}
.grep <- function(p1, f1, fix = 0, max, line.number = 0, o = "", intern = 0, head = !intern, head.n = 6) {
    .sys.do("grep", if (fix) 
        "-F", if (!missing(max)) 
        c("-m", max), if (line.number) 
        "-n", o, shQuote(p1), .sys.fpGood(f1), if (head) 
        c("|head -n", head.n), intern = intern)
}
.pwd <- function(intern = 1) .wk("pwd", intern = intern)
wk.ls.path <- switch(Sys.info()["sysname"], Linux = "ls", Darwin = "gls")
.ls <- function(fp = ".", p = "AhlF", p1 = "", less = 0) {
    fp <- unlist(fp)
    .wk(wk.ls.path, sprintf("-%s", p), "--color", "--group-directories-first", "--time-style=long-iso", "-I", .q("*[#~]"), "--full-time", "--time-style=long-iso", if (nchar(p1)) 
        sprintf("-%s", p1), .sys.fpGood(fp), if (less) 
        "| less -N")
}
.ls1 <- function(wj = ".", p = "Ahlpg", less = 0) {
    .ls(dirname(wj[1]))
}
.cp <- function(f1, f2, ...) .wk("cp", hs.fpGood(f1), hs.fpGood(f2), ...)
.wc <- function(f, p = "l", intern = 1, z = grepl("\\.gz$", f)) {
    R <- .wk(.sys.cat.text(f), hs.paste("| wc -", p), intern = intern)
    if (intern) 
        as.numeric(R)
    else R
}
.cd <- function(d = "~") {
    if (!file.info(d)[["isdir"]]) 
        d <- dirname(d)
    setwd(d)
}
.less <- function(f, r = 1, skip = 0) {
    less <- osType() %=>% hs.switch("Darwin", "/usr/local/bin/less", "less")
    .wk(.sys.cat.skip(f, skip = 0), if (skip) 
        c("| perl -e", .q(paste0("foreach ( 1..", skip, ") { <>}; while( <>) { print}"))), "|", less, " -NfiF", if (r) 
        "-R")
}
.head <- function(wj1, n = 10, intern = 0, z = hs.file.extension(wj1) %in% c("gz")) {
    .wk("cat", .sys.fpGood(wj1), if (z) 
        c(.pipe(), "gzip -cd"), .pipe(), "head -n", format(n, scientific = FALSE), intern = intern)
}
.cat <- function(wj, gz = hs.grepl(wj, "\\.gz$")) {
    .wk("cat", wj, if (gz) 
        "| gzip -cd")
}
.pipe <- function(no.err = 1) {
    c(if (no.err) "2>/dev/null", "|")
}
.sys.pipe <- function(x, ..., no.err = 0) {
    .sys(x, .pipe(no.err), ...)
}
.perl <- function(..., perl = "perl", w = 1) {
    .sys(perl, if (w) 
        "-w", ...)
}
.perl.e <- function(..., l = 1, a = 0, n = 1, p = 0, F = "\\t", c = 0, args = {
}, intern = 0, sys = 0, pipe = 0, begin = {
}) {
    cmd <- .sys("perl", if (c) 
        "-c", if (n) 
        "-n", if (p) 
        "-p", if (l) 
        "-l", if (a) 
        "-a", if (length(F)) 
        paste0("-F", .q(F)), paste0("-e ", c(if (length(begin)) .q(paste0(c("BEGIN{", paste0(begin, ";"), "}"), collapse = "")), .q(paste0(c(...), ";")))), if (length(args)) 
        args)
    if (pipe) {
        pipe(cmd, "rt")
    }
    else if (sys) {
        cmd
    }
    else {
        .wk(cmd, intern = intern)
    }
}
wj_row.hs.v_wj__perl <- function(v1, wj1, wj1.cn2use, mn = -1, ot = "dt", cn = mn != 1) {
    wj_temp <- hs.wj.temp()
    on.exit(hs.wj.rm(wj_temp), add = TRUE)
    cat(v1, sep = "\n", file = wj_temp)
    pattern1 <- paste0("^", paste(rep("[^\\t]*\\t", wj1.cn2use - 1), collapse = ""), "([^\\t\\n]*)")
    cmd1 <- .sys("perl -e", .q(hs.pastec0("\nopen( in1, \"cat ", wj_temp, " |\");\nwhile( <in1>) {\n  chomp;\n  $dict1{ $_} = 1;\n}\nclose( in1);\nopen( in1, \"", .sys.cat.text(wj1), " |\");\nwhile( <in1>) {\n  /", pattern1, "/;\n  if( exists( $dict1{ $1})) {\n    print;\n", if (mn > 0) {
        c("\n    $found = $found + 1;\n    if( $found == ", mn, ") {\n      exit;\n    }\n")
    }, " }\n}\nclose( in1);\n")))
    jg <- .wk(cmd1, intern = 1)
    switch(ot, row = jg, dt = {
        jg <- fread(text = jg)
        setnames(jg, cn.hs.wj(wj1))
    })
}
.tail <- function(f, n = 10) {
    .wk("cat_wk", hs.fpGood(f), "| perl -e", .q(paste0("\n$n = ", n, ";\n$i = 0;\nwhile( <>) {\n  $i = $. % 10;\n  $a[ $i] = $_;\n}\nif( $i < $n) {\n  foreach $j ( ( $i + 1) .. ( $n - 1)) {\n    print( $a[ $j]);\n  }\n}\nforeach $j ( 0 .. $i) {\n  print( $a[ $j]);\n}\n")))
}
.man <- function(x) .wk("man", x)
.top <- function() .wk("top")
.htop <- function(u = .whoami()) .wk("htop -u", u)
.df <- function() .wk("df -h")
.rm <- function(f) .wk("rm", .sys.fpGood(f))
.hostname <- function() .wk("hostname", intern = 1)
.sleep <- function(t = 3) .wk("sleep", t)
.uuidgen <- function(append) paste(c(.sys.do("uuidgen", intern = 1), if (!missing(append)) append), collapse = "_")
.whoami <- function() Sys.getenv("USER")
.Rscript <- function(..., ssh, nohup = FALSE, slave = TRUE, quiet = TRUE, no.save = TRUE) {
    dian <- c(...)
    .sys.do(if (!missing(ssh)) 
        c("ssh", ssh), if (nohup) 
        "nohup", "Rscript", if (quiet) 
        "--quiet", if (no.save) 
        "--no-save", dian, if (nohup && !length(grep("&", dian))) 
        "&")
    .sys(if (!missing(ssh) && ssh != .hostname()) 
        c("ssh", ssh), if (nohup) 
        "nohup", "Rscript", if (quiet) 
        "--quiet", if (no.save) 
        "--no-save", dian, if (nohup && !length(grep("&", dian))) 
        "&")
}
wk.zip2gz <- function(wj1) {
    wj2 <- hs.wjm(wj1, ext = "gz")
    if (!file.exists(wj2)) 
        .wk("unzip", "-p", wj1, "|", "gzip", "-c", ">", wj2)
}
hs.wj.rm <- hs.rm.wj <- wk.rm.wj <- function(wj, recursive = 0) {
    unlink(wj, recursive)
}
ieiwk.dirname.1 <- function(f1) {
    repeat {
        d1 <- dirname(f1)
        if (d1 == ".") 
            return(basename(f1))
        f1 <- d1
    }
}
ieiwk.split.file <- function(file, parts = 5, size, mb = dirname(file)) {
    c1 <- file(file, "rb")
    if (missing(size)) {
        s0 <- file.info(file)[["size"]]
        size <- ceiling(s0/parts)
    }
    part <- 1
    while (length(chunk <- readBin(c1, "raw", size))) {
        f.out <- file(ieiwk.filename(file, add = sprintf("part%s", part), dn = mb), "")
        `?`(file)
        writeBin(chunk, f.out)
        part <- part + 1
    }
    close(c1)
    part - 1
}
ieiwk.comine.files <- function(file1) {
    files <- ieiwk.list.files(dirname(file1), sub("\\d+$", "", basename(file1)))
    f.out <- ieiwk.filename(file1)
    r <- unlist(lapply(files, function(file) {
        c1 <- file(file, "rb")
        chunk <- readBin(c1, "raw", file.info(files)[["size"]])
        close(c1)
        chunk
    }), use.names = FALSE)
    writeBin(r, f.out)
    f.out
}
.sys.ssh.file.path <- function(..., fsep = .Platform$file.sep, recursive = TRUE, type = "file", sys = FALSE, expand = 0, normalize = 1, v.only = 0) {
    a <- file.path(...)
    if (v.only) 
        return(a)
    switch(type, file = {
        a.p <- dirname(a)
        if (!file.exists(a.p)) dir.create(a.p, recursive = TRUE, showWarnings = FALSE)
    }, dir = {
        a.len <- nchar(a)
        if (substr(a, a.len, a.len) != "/") a <- sprintf("%s/", a)
        if (!file.exists(a)) dir.create(a, recursive = recursive)
    })
    r <- {
        if (sys) {
            r <- gsub("\\\\ ", " ", a)
            gsub(" ", "\\\\ ", a)
        }
        else {
            a
        }
    }
    if (expand) 
        r <- path.expand(r)
    if (normalize) 
        r <- ieiwk.fp.norm(r)
    r
}
.sys.file.exists <- function(fp, dz = {
}) {
    ifelse(length(grep("No such file or directory)$", .sys.do("file", .q(ieiwk.fp.4sys(fp)), dz = dz, intern = 1))), FALSE, TRUE)
}
.sys.ssh <- function(..., dz = "sg", p = 22, sep = " ", wait = 1, intern = 0, chk = 0) {
    dz.db <- c(sg = "172.18.202.8", hp = "172.18.202.7", g1 = "gate1.picb.ac.cn", g2 = "gate2.picb.ac.cn")
    p.db <- c(sg = 53641, hp = 53641, g1 = 22, g2 = 22)
    if (dz %in% names(dz.db)) {
        p <- p.db[dz]
        dz <- dz.db[dz]
    }
    cmd <- .sys(..., intern = 1)
    if (chk) {
        ieiwk.cat.line(cmd)
    }
    else {
        .sys("ssh -p", p, dz, shQuote(cmd), sep = sep, wait = wait, intern = intern, chk = 0)
    }
}
.sys.ssh.stat <- function(fp, dz, p = 22) {
    .sys.do("ssh -p", p, dz, shQuote(.sys("stat", fp, intern = 1)), intern = 1)
}
.sys.ssh.stat.modify <- function(fp, dz, p = 22) {
    a <- .sys.ssh.stat(fp, dz, p)
    ifelse(length(a), sapply(strsplit(grep("Modify:", a, v = 1), "[ ]+")[[1]][2:3], function(x) head(strsplit(x, "[-\\:\\.]")[[1]], 3)) %,% "", 0)
}
.sys.stat <- function(fp, dz = {
}) {
    .sys.do("stat", fp, intern = 1, dz = dz)
}
.sys.stat.modify <- function(fp, dz = {
}) {
    a <- .sys.stat(fp, dz = dz)
    ifelse(length(a), sapply(strsplit(grep("Modify:", a, v = 1), "[ ]+")[[1]][2:3], function(x) head(strsplit(x, "[-\\:\\.]")[[1]], 3)) %,% "", 0)
}
.sys.stat.size <- function(fp, dz = {
}) {
    a <- .sys.stat(fp, dz = dz)
    ifelse(length(a), as.numeric(re("\\d+", grep("Size:", a, v = 1))), 0)
}
.sys.mkdir <- function(d, dz = {
}) {
    .sys.do("mkdir -p", d, dz = dz)
}
.wget <- function(dz, P = ".", T = 1, t = 9999, nc = 0, d = 0) {
    wjj0 <- getwd()
    on.exit(setwd(wjj0), add = TRUE)
    setwd(P)
    repeat {
        a1 <- .wk("wget -c", if (d) 
            "-d", "-t", t, "-T", T, dz, "2>&1", intern = 1)
        if (length(hs.grep(a1, "\\s+The file is already fully retrieved; nothing to do."))) 
            break
    }
    a <- .wk("wget -c", if (d) 
        "-d", "-nc", "-t", t, "-T", T, dz, "2>&1", intern = 1)
    a <- hs.grep(a, "^File .(.*). already there; not retrieving\\.$", v = 1)
    normalizePath(hs.re(a, "(?<=^File .).*(?=. already there; not retrieving\\.$)"))
}
ieiwk.cp.dir_str <- function(d1, d2) {
    fp.v <- ieiwk.list.files(d1, only.dir = 1, recursive = 1)
    sapply(file.path(d2, gsub(path.expand(d1), "", fp.v)), dir.create, recursive = 1)
}
ieiwk.fp.norm <- function(fp) {
    for (i in seq_along(fp)) {
        a <- fp[i]
        if (!file.exists(a)) 
            next
        if (file.info(a)$isdir && substring(a, nchar(a)) != .Platform$file.sep) {
            fp[i] <- file.path(a, "")
        }
    }
    fp
}
ieiwk.dirname.4home <- function(fp) {
    a <- head(strsplit(fp, .Platform$file.sep)[[1]], -1) %,% .Platform$file.sep
    sub("^~", "$HOME", a)
}
ieiwk.fp.4sys <- function(fp) {
    if (length(grep("/~", fp))) 
        stop(sprintf("Please check the file path: %s, in which ~ is preceeded by /.", fp))
    sub("~", "$HOME", fp)
}
ieiwk.fp.4r <- function(fp, dz = {
}) {
    sub(.sys.do("echo ~", dz = dz, intern = 1), "~", fp)
}
assign.zjjg <- function(..., extN, env = parent.frame()) {
    vn <- paste(substitute(c(...))[-1])
    for (v1 in paste(substitute(c(...))[-1])) {
        assign(paste(v1, "zjjg", sep = "."), ieiwk.filename(get(v1, env = env), zjjg = TRUE), envir = env)
    }
}
ieiwk.filename.append <- function(f1, append = "", dn = TRUE) {
    paste(ieiwk.filename(f1, dn = dn), append, ".", ieiwk.file.extension(f1), sep = "")
}
ieiwk.file.extension.change <- function(f, alt, bn = 0) {
    if (bn) 
        f <- basename(f)
    sub("([^.]+)$", alt, f)
}
ieiwk.file.size <- function(f1, co = 1024^2, bj = 0) {
    s1 <- file.info(f1)[["size"]]
    if (bj) {
        s1 >= co
    }
    else s1
}
.sys.file.fifo <- function(f1, act = c("read", "write"), gzip = "gzip", compression.level = 6, named.file, append = 0) {
    act <- match.arg(act)
    f2 <- sprintf("%s %s )", switch(act, read = paste("<(", sapply(ieiwk.file.extension(f1), switch, gz = "pigz -dc", "cat")), write = paste(">(", if (length(gzip)) sprintf("%s -c -%d >%s", gzip, compression.level, ifelse(append, ">", "")))), f1)
    if (missing(named.file)) {
        if (length(f1) > 1) 
            c("<( cat", f2, ")")
        else f2
    }
    else {
        if (is.logical(named.file) && named.file) {
            named.file <- hs.wjm(f1[1], append = .uuidgen("fifo"))
        }
        .wk("mkfifo", named.file)
        switch(act, read = .wk("cat", f2, ">", named.file, "&"), write = .wk("cat", named.file, ">", f2, "&"))
        named.file
    }
}
.sys.grep <- local({
    bin <- switch(osType(), Darwin = "ggrep", "grep")
    function(p1, inv = 0, perl = 1, m = {
    }, n = {
    }, ic = 1, start = 0, end = 0) {
        if (start) 
            p1 <- paste0("^", p1)
        if (end) 
            p1 <- paste0(p1, "$")
        .sys("LC_ALL=C", bin, if (ic) 
            "-i", if (inv) 
            "-v", if (perl) 
            "-P", if (length(m)) 
            c("-m", m), if (length(n)) 
            "-n", .q(p1))
    }
})
.sys.cat.skip <- function(wj1, comment = "#", cn = 1, header = cn, only = {
}, inv = 0, pipe = {
}, N = 0, skip = 1, plus = 0) {
    wj2 <- tolower(hs.re(wj1, "[^\\.]*$")) %=>% {
        if (x %in% c("7z")) 
            hs.browser("2023.07.04.162016", debug = 0)
        db <- environment(.sys.cat.text)[["ext.cat"]]
        if (x %in% names(db)) 
            db[x]
        else "cat"
    } %=>% .sys(.sys.fpGood(wj1))
    if (skip) {
        i1 <- hs.lines_comment(wj1, opt = "i", comment = comment) + 1
        i1 <- i1 + plus
        if (length(only)) 
            return(switch(only, i = i1, "pipe"))
        if (inv) {
            .sys(wj2, "|", .sys.head(i1 - 1, pipe = 0), if (N) 
                c("| head -n", N), pipe = pipe)
        }
        else {
            i2 <- i1 + ifelse(header, -1, 0)
            .sys(wj2, if (i2 > 1) 
                .sys("| perl -e", .q(paste0("$n = ", i2, "; while( $n) { <>; $n--;}; while( <>) { print;}"))), if (N) 
                c("| head -n", N), pipe = pipe)
        }
    }
    else wj2
}
.sys.uniq <- function(sort = 1) {
    c("| uniq", if (sort) "| sort | uniq")
}
.sys.wget <- function(wj1, q = 1, bg = 1, v = !q, c = 1, root = getwd(), o = hs.wjm(wj1, add = "wget.log", dn = root)) {
    c("wget", if (c) "-c", if (q) "-q", if (v) "-v", str_c("--directory-prefix=", wk.wjlj(root)), wj1, if (length(o)) c("-a", o), if (bg) "&")
}
.tabix <- function(w1, r1 = {
}, tabix = hs.home.expand("~/rj/tabix-0.2.6/tabix"), p = "vcf", cn = 1, start = 2, end = 3, name = 1) {
    if (file.size(w1) < 10000) 
        stop("【34.01.19.193142】file size is too small")
    .tabix.index(w1, tabix, start = start, end = end, name = name)
    if (length(r1)) {
        wj.tmp <- tempfile(tmpdir = "/dev/shm")
        on.exit(hs.rm.wj(wj.tmp), add = TRUE)
        if (hs.len1(dim(r1), 0)) {
            r2 <- do.call("rbind", strsplit(r1, "[:-]"))
            r1 <- as.data.table(r2)
        }
        if ((hs.len1(dim(r1), 2)) && (ncol(r1) >= 3)) {
            if (!is.data.table(r1)) 
                r1 <- as.data.table(r1)[, .SD, .SDcols = 1:3]
            if (length(r1) > 3) 
                r1[, `:=`(c(tail(names(r1), -3)), {
                })]
            r1[[2]] <- as.numeric(r1[[2]]) - 1
            wk.fwrite(r1, wj.tmp, append = 0, col.names = 0)
            r1 <- wj.tmp
        }
        R <- wk.fread(cmd = .sys(tabix, "-B", w1, r1), cn = 0)
        if (nrow(R) && cn) 
            setnames(R, names(wk.fread(w1, N = 1)))
        R
    }
}
.tabix.sort <- function(w1, name = 1, start = 2, end = 3) {
    w1_tmp <- hs.wjm(w1, append = "33.12.29.180245")
    comments.i <- hs.lines_comment(w1, opt = "i")
    cmd.cat.sort <- if (comments.i) {
        .sys("(", "zcat", w1, "|", "head -n", comments.i, ";", "zcat", w1, "|", "tail", "-n", paste0("+", comments.i + 1), "|", "sort", paste0("-k", name, ",", name), paste0("-k", start, ",", start, "n"), paste0("-k", end, ",", end, "n"), ")")
    }
    else .sys("zcat", w1, "|", "sort", paste0("-k", name, ",", name), paste0("-k", start, ",", start, "n"), paste0("-k", end, ",", end, "n"), ")")
    .wk(cmd.cat.sort, "|", "pbgzip", "-c", ">", w1_tmp)
    if (0) {
        if (diff(file.size(c(w1, w1_tmp))) < 0) 
            warning("size check failed; check here: 34.01.19.164703")
    }
    local({
        a <- file.size(c(w1_tmp, w1))
        a <- (a[1]/a[2]) < 0.9
        if (a) 
            stop("new file's size too small; check 34.01.30.201935")
    })
    file.rename(w1_tmp, w1)
    w1
}
.tabix.index <- function(w1, tabix = hs.home.expand("~/rj/tabix-0.2.6/tabix"), co.ratio = 0.01, start = 2, end = 3, name = 1) {
    flag <- 0
    flag.sort <- 0
    wj_tbi <- hs.wjm(w1, add = "tbi")
    if (file.exists(wj_tbi)) {
        fi <- file.info(c(wj_tbi, w1))
        if (fi[, "mtime"] %=>% diff %=>% {
            x >= 1
        }) {
            flag <- 1
        }
        if (0) {
            if (fi[, "size"] %=>% {
                x[1]/x[2] < co.ratio
            }) 
                flag.sort <- 1
        }
        if ((fi[, "size"] %=>% {
            x[1]/x[2]
        } < co.ratio) && (fi[1, "size"] < 1000)) {
            flag.sort <- 1
        }
    }
    else {
        flag.sort <- 1
    }
    if (flag.sort) {
        .tabix.sort(w1, start = start, end = end, name = name)
        flag <- 1
    }
    a <- 0
    if (flag) {
        a <- .wk(tabix, "-s", name, "-b", start, "-e", end, "-f", .q(normalizePath(w1)))
        if (a) {
            if (flag.sort) {
                hs.browser("34.07.04.112349")
            }
            else {
                .tabix.sort(w1)
                .tabix.index(w1, p, tabix, co.ratio)
            }
        }
    }
}
.sys.tabix.index <- function(wj, bin = hs.home.expand("~/rj/tabix-0.2.6/tabix"), s = 1, b = 2, e = 3, S = 0, sort = 1) {
    wj <- wj2
    if (sort) {
        if (str_sub(wj, -3, -1) == ".gz") {
            wjj1 <- hs.wjj(tempfile(tmpdir = "/dev/shm"))
            on.exit(unlink(wjj1, recursive = TRUE), add = TRUE)
        }
        else {
            wjj1 <- dirname(wj)
        }
        args <- list(wj, dn = wjj1)
        if (str_sub(wj, -3, -1) != "gz") 
            args[["add"]] <- "gz"
        wj2 <- hs.wjm %(% args
        .sys.do("sort", str_interp("-k$${ s},${ s}"), str_interp("-k${ b},${ b}n"), .sys.file.fifo(wj2), "|", "pbgzip", ">", wj2)
        .less(.sys.file.fifo(wj2))
    }
    else {
        if (str_sub(wj, -3, -1) != ".gz") {
            .sys.do("pbgzip", wj)
            wj <- hs.wjm(wj, add = "gz")
        }
    }
    for (o1 in unlist(str_split("s,b,e,S", ","))) {
        eval(bquote({
            .(as.symbol(o1)) <- sprintf("-%s%s", .(o1), formatSafe(.(as.symbol(o1))))
        }))
    }
    .sys.do(bin, s, b, e, S, wj)
}
.sys.sort <- function(k = {
}) {
    .sys("sort", if (length(k)) 
        str_c("-k", k))
}
.sys.find <- function(wjj = ".", iregex = "", regex = "", name = "", path = "", iname = "", ipath = "", wholename = "", iwholename = "", maxdepth = 99, regextype = "posix-extended", ...) {
    L <- list(...)
    e1 <- unlist(sapply(unlist(strsplit("iregex, regex, name, path, iname, ipath, wholename, iwholename", ",\\s+")), function(p1) {
        v1 <- get(p1)
        inv <- L[[paste0(p1, ".inv")]]
        if (!length(inv)) 
            inv <- rep(0, )
        if (nzchar(v1[1])) 
            sprintf("%s -%s %s", ifelse(inv == 1, "!", ""), p1, .q(get(p1)))
    }))
    .sys("find", wjj, "-maxdepth", maxdepth, "-regextype", regextype, e1)
}
.sys.gawk <- function(col2print = 0, sep.in = "\t", ofs = "\t") {
    .sys("gawk", "-F", .q(sep.in), "-v", sprintf("OFS=\"%s\"", ofs))
}
.sys.gawk.print <- function(col2print = 0, sep.in = "\t", sep.out = sep.in) {
    cmd <- paste("BEGIN{ OFS = \"\\t\"} { print(", paste(paste("$", col2print, sep = ""), collapse = ","), ")}", sep = "")
    .sys("gawk", "-F", .q(sep.in), .q(cmd))
}
.sys.gawk.selectLines <- function(i) {
    i <- wk.integerRange.4cut(i)
    .sys(.sys.gawk(), .q(str_c("BEGIN{ split( ", .q(i, type = "cmd"), ", i, \",\"); for( ii = 1; ii <= length( i); ii++) { split( i[ ii], i1, \"-\"); if( ii == 1) { i2 = 1} else { split( i1[ ii - 1], i0, \"-\"); i2 = i0[ length( i0)] + 1}; for( i_skip = i2; i_skip < i1[ 1]; i_skip++) getline; for( i_keep = i1[ 1]; i_keep <= i1[ length( i1)]; i_keep++) { getline a; print a}}}")))
}
.sys.cut <- function(col2print = "1-", sep.out = "\t", d = {
}, m = "f") {
    .sys("cut", sprintf("-%s", m), col2print, sprintf("--output-delimiter=\"%s\"", sep.out), if (length(d)) 
        c("-d", d))
}
.sys.less <- function(N = 1) .sys("less", if (N) "-N")
.sys.head <- function(n = 10, pipe = 0) {
    if (length(n)) 
        .sys("head", "-n", n, if (pipe) 
            "|")
    else c("", if (pipe) "|")
}
.sys.tail <- function(n = 10, from.beginning = 0) {
    n <- formatSafe(n)
    .sys("tail", "-n", if (from.beginning) 
        sprintf("+%s", n)
    else n)
}
.sys.pigz.c <- function(cl = 6, tn = 8, pipe = 0) {
    r <- sprintf("pigz -p%s -%s", tn, cl)
    if (pipe) 
        r <- c("|", r)
    r
}
.sys.tr <- function(set2 = ";", set1 = "\t") .sys("tr", .q(set1), .q(set2))
.sys.parallel <- function() hs.home.expand("~/rj/speedseq_2016.05.15/bin/parallel")
.sys.gzip <- function(args, l = 6) {
    .sys("gzip", args, if (length(l)) 
        str_c("-", l))
}
.sys.gzip.out <- function(...) .sys.gzip("-cd", ..., l = {
})
.sys.gzip.in <- function(...) .sys.gzip("-c", ...)
.sys.pigz <- function(act = "cd", N = n, n = {
}) {
    .sys("pigz", if (nchar(act)) 
        sprintf("-%s", act), if (length(N)) 
        c("-p", as.integer(N)))
}
.sys.pigz.out <- function(N = n, n = {
}) {
    .sys.pigz(N = N)
}
.sys.pigz.in <- function(N = n, n = {
}) {
    .sys.pigz("c", N = N)
}
.sys.pbgzip <- function(N = n, n = {
}, c = 1, d = 0) {
    .sys("pbgzip", if (length(N)) 
        c("-n", N), if (c) 
        "-c", if (d) 
        "-d", "2>&1")
}
.sys.pbgzip.out <- function(...) .sys.pbgzip(..., d = 1)
.sys.pbgzip.in <- function(...) .sys.pbgzip(..., d = 0)
.sys.genomes.human <- function(which = 1) c(hs.home.expand("~/swxx/sj/jyz/human/1kg/37/human_g1k_v37.fasta"), .sys.speedseq("wj.ref"))[which]
.sys.java <- function(mem.max = 750, jar = {
}) {
    .sys("java", str_interp("-Xmx${ formatSafe( mem.max)}m"), if (length(jar)) 
        c("-jar", jar))
}
ieiwk.copy.wjj.jiegou <- function(wjj1, wjj2 = dirname(wjj1[1]), qianzhui = "speedseq") {
    ieiwk.file.path(wjj2, qianzhui, type = "wjj")
}
hs.with_sink <- function(wj2, ...) {
    sink.con <- if (hs.grepl(wj2, "(?i)\\.gz$")) {
        on.exit({
            sink()
            close(sink.con)
        }, add = TRUE)
        gzfile(wj2, "w")
    }
    else {
        on.exit(sink(), add = TRUE)
        wj2
    }
    sink(sink.con)
    list(...)
    wj2
}
uid_fp_source <- function(uid1 = c(), fp1 = c(), less = 0, ...) {
    if (length(uid1)) {
        uid1 <- hs.re(uid1, "\\d+[_\\.]\\d{2}[\\._]\\d{2}[\\._]\\d{6}")
        fp1 <- .wk("ag -g", uid1, hs.home.expand("~/org"), intern = 1)
    }
    if (!((hs.len1(fp1)) && nzchar(fp1))) 
        hs.browser("37.05.07.204631")
    if (less) {
        .less(fp1)
        return()
    }
    source(fp1, ...)[["value"]]
}
hs.hs <- function(hs1, env1 = new.env(parent = environment(hs1)), onExit = {
}) {
    hs.browser("34.07.05.104924")
    hs2 <- eval(bquote(local({
        env1.33.06.24.181214 <- environment()
        env2.33.06.24.181217 <- new.env()
        .(hs1)
    })))
    a <- hs.localize(hs1)
    ls(environment(a))
    ls(environment(hs1))
    typeof(hs1)
    class(function(x) {
    })
    ls(environment(hs2))
    ls(env1)
    env1 <- environment(hs1)
    apropos("parent")
    parent.env(env1)
    self <- sys.function()
    hs1 <- function() {
    }
    AN <- 0
    list(...)
    formalArgs()
    args <- head(formalArgs(hs.parent), -3)
}
wk.get <- ieiwk.get <- local.get <- function(...) get(..., inherits = FALSE)
.getl <- .geta <- function(uid, env1 = .GlobalEnv) {
    o1 <- ls(pattern = uid, all.names = TRUE, envir = env1)
    if (length(o1)) 
        eval(bquote(get(.(o1))), env1)
    else stop("No such object.")
}
hs.dmgly <- ieiwk.dmgly <- ieiwk.sinkORget.code <- function(wjm, bds, local = 0, ext = "R", wjj0 = ieiwk("wjj.dm"), env1 = .GlobalEnv, PO = 0, path.only = PO, blm.only = 0, load = 0, reload = 0, call = 0, argList = list(), rename = 0, bl = 1, parse.network = 0, less = 0, trim = 0) {
    {
        wj <- .mfp(wjm, wjj0, act = "find", uid.pos = "tail")
        wj2 <- hs.wjm(.mfp(wjm, wjj0, uid.pos = "tail"), add = ext)
        if (length(wj)) {
            if (wj != wj2) 
                if (!missing(bds)) {
                  file.rename(wj, wj2)
                  wj <- wj2
                }
        }
        else wj <- wj2
    }
    if (path.only) 
        return(wj)
    if ({
        !file.exists(wj)
    } && missing(bds)) 
        stop(paste("####", wj, "is not found. 34.07.04.195822 ####"))
    if (less) 
        return(.less(wj))
    if (!missing(bds)) {
        bds <- substitute(bds)
        bds.nr <- format(bds)
        if (trim) 
            bds.nr <- trim(bds.nr)
        cat2wj <- 1
        if (file.exists(wj)) {
            wj.nr <- readLines(wj, encoding = "UTF-8")
            if ({
                length(wj.nr) == length(bds.nr)
            } && identical(bds.nr, wj.nr)) 
                cat2wj <- 0
        }
        if (cat2wj) 
            cat(bds.nr, file = wj, sep = "\n")
    }
    {
        uid <- hs.file.extension(wjm, extN = 4)
        wj.blm <- hs.wjm(wj, dn = 0, ext = 0)
        if (wj.blm == uid) 
            stop(paste("####", wjm, "is not meaningful, 34.07.04.195811@function.common.R ####"))
        blm <- ls(env1, pattern = uid, all.names = TRUE)
        if (!length(blm)) 
            blm <- wj.blm
    }
    if (blm.only) 
        return(blm)
    {
        env.lt <- "env.loadtime.33.04.27.080101"
        if (!exists(env.lt, env1, inherits = FALSE)) 
            env1[[env.lt]] <- new.env(parent = env1)
        blm.loadtime <- env1[[env.lt]][[blm]]
    }
    if (missing(bds)) {
        if (exists(blm, env1, inherits = FALSE)) {
            if ((!length(blm.loadtime)) || (difftime(file.info(wj)[, "mtime"], blm.loadtime, units = "mins") > 1)) {
                load <- 1
            }
            else return(env1[[blm]])
        }
        else load <- 1
    }
    if (load || reload) {
        file.exists(wj) || stop("no file found # 34.07.13.054404@function.common.R")
        bds <- parse(wj, encoding = "UTF-8", keep.source = FALSE)
    }
    if (length(bds)) {
        if (local) 
            bds <- bquote(local(.(bds)))
        R <- eval(bds, env1)
        if (bl) {
            env1[[env.lt]][[blm]] <- Sys.time()
            env1[[blm]] <- R
        }
        else R
    }
}
a.33.03.17.191105 <- function(wj1) {
    a <- ieiwk.readLines(wj1, n = -1)
    .ss(sort(grep("ieiwk.dmgly\\(\\s*'.*\\.\\d+.\\d{2}.\\d{2}.\\d{6}',", a, v = 1)))
}
.getls <- function(uid, env1 = .GlobalEnv, update = 0) {
    uid1 <- ieiwk.file.extension(uid, extN = 4)
    if (update) 
        ieiwk.sinkORget.code(uid, reload = 1)
    o1 <- ls(pattern = uid1, all.names = TRUE, envir = env1)
    if (!length(o1)) {
        eval(bquote(ieiwk.dmgly(.(uid), bl = 1)), env1)
        o1 <- ls(pattern = uid1, all.names = TRUE, envir = env1)
    }
    eval(bquote(get(.(o1))), env1)
}
uid.filePath.code <- function(uid) {
    ieiwk.sinkORget.code(uid, path.only = 1)
}
ieiwk.sjgly <- function(wjm, bds, ext = "RData", wjj0 = ieiwk("wjj.sj"), env1 = .GlobalEnv, path.only = 0, blm.only = 0, load = 0, reload = 1, call = 0, argList = list(), rename = 0, bl = 0, parse.network = 0) {
    wj <- hs.wjm(.mfp(wjm, wjj0, uid.pos = "tail"), add = ext)
    if (path.only) 
        return(wj)
    if ({
        !file.exists(wj)
    } && missing(bds)) 
        stop(paste("####", wj, "is not found. ####"))
    blm <- wjm
    uid <- ieiwk.file.extension(wjm, extN = 4)
    if (blm == uid) 
        stop(paste("####", wjm, "is not meaningful ####"))
    if (blm.only) 
        return(blm)
    if (missing(bds)) {
        if (load || reload) {
            e1 <- bquote(hs.load(.(wj)))
            if (bl) {
                eval(bquote({
                  .(as.symbol(blm)) <- .(e1)
                }), env1)
            }
            else {
                eval(e1)
            }
        }
    }
    else {
        if (file.exists(wj) && reload) 
            ieiwk.sjgly(wjm)
        else {
            e1 <- bquote({
                local({
                  eval(.(bds))
                })
            })
            if (bl) {
                eval(bquote({
                  .(as.symbol(blm)) <- .(e1)
                }), env1)
            }
            blm
        }
    }
}
hs.xmgly <- function(wjm, wjj0 = hs.home.expand("~/wk/gz/db"), go = 1, gj = 1) {
    uid <- ieiwk.file.extension(wjm, extN = 4)
    wjj <- .mfp(uid, wjj0, uid.pos = "tail", fp.type = "dir")
    if (gj) 
        sapply(c("sj", "jg"), function(x) hs.wjj(wjj, x))
    if (go) 
        setwd(wjj)
    wjj
}
ieiwk.xmbm <- function(uid.xm, ms1, ms2) {
    if (missing(ms2)) 
        ms2 <- ieiwk.file.extension(uid.xm, extN = 4)
    xm.wjj <- ieiwk.xmgly(uid.xm)
    hs.wjj(xm.wjj, "sj", go = 1)
    wj1 <- ieiwk.list.files(, recursive = 1, full.names = 0)
    wj2 <- gsub(ms1, ms2, wj1)
    for (i1 in seq_along(wj1)) {
        if (wj1[i1] != wj2[i1]) {
            ieiwk.file.path(wj2[i1])
            file.rename(wj1[i1], wj2[i1])
        }
    }
    getwd()
    for (wjj1 in ieiwk.list.dirs()) {
        if (!length(ieiwk.list.files(wjj1, recursive = 1))) 
            unlink(wjj1, recursive = TRUE)
    }
}
ieiwk.xmbmhy <- function(wjm, ms21) {
    ms2 <- names(ms21)[2]
    for (ms2 in names(ms21)) {
        wj2 <- grep(ms2, wjm, v = 1)
        if (!length(wj2)) 
            next
        ms1 <- ms21[[ms2]]
        wj1 <- sub(ms2, ms1, wj2, fixed = TRUE)
        file.rename(wj2, wj1)
    }
}
ieiwk.function <- function(..., body) {
    f1 <- function(...) {
    }
    cmpfun(f1)
}
evalbq <- function(e1, env1 = parent.frame(), env2 = env1) {
    eval(substitute(eval(bquote(e1))), envir = env1)
}
evalsub <- function(exp1, map1, env1 = parent.frame()) {
    e <- evalbq(substitute(.(substitute(exp1)), .(substitute(map1))))
    eval(e, env1)
}
ieiwk.eval.save.load <- function(e1, f1, rerun = 0, lib = {
}, c1 = Sys.info()["nodename"]) {
    if (file.exists(f1) && !rerun) {
        hs.load(f1)
    }
    else {
        r <- ieiwk.eval(e1, lib = lib, c1 = c1)
        save(r, file = f1)
        r
    }
}
input.is.incomplete <- function(r) {
    (class(r) == "try-error") && (!is.null(attr(r, "condition"))) && (re("(?<=: )[^:]*$", readLines(textConnection(attr(r, "condition")[["message"]]), 1)) %in% c("unexpected end of input", "unexpected INCOMPLETE_STRING", "unexpected end of line", "意外的INCOMPLETE_STRING"))
}
wk.e1 <- function(e1, env1 = parent.frame(), td.co = 10, bg = 0) {
    if (missing(e1)) 
        e2 <- quote({
        })
    e2 <- as.call(c(as.symbol("{"), substitute(e1)))
    e3 <- bquote({
        wk.cat(Sys.info()["user"], "@", Sys.info()["nodename"], getwd())
        t1 <- Sys.time()
        r <- eval(.(e2), .(if (!bg) 
            env1))
        t2 <- Sys.time()
        td <- difftime(t2, t1, units = "secs")
        msg1 <- capture.output(as.difftime(td))
        .ss(msg1)
        if (td > .(td.co)) {
            if (wk.is.a.local.machine()) {
                .wk("kate", .q(msg1), "> /dev/null", "2> /dev/null", "&")
            }
            else {
                .r(msg1, "")
            }
        }
        .cat.time()
        r
    })
    wk.eval(e3, bg = bg, email = 0, report.if.done.cutoff.time = 99999)
}
wk. <- wk.time <- function(expr, gcFirst = 0, td.co = 10) {
    ppt <- function(y) {
        if (!is.na(y[4L])) 
            y[1L] <- y[1L] + y[4L]
        if (!is.na(y[5L])) 
            y[2L] <- y[2L] + y[5L]
        y[1L:3L]
    }
    if (!exists("proc.time")) 
        return(rep(NA_real_, 5L))
    if (gcFirst) 
        gc(FALSE)
    time <- proc.time()
    on.exit(cat("Timing stopped at:", ppt(proc.time() - time), "\n"))
    expr
    new.time <- proc.time()
    on.exit()
    .ss(paste(Sys.info()["user"], "@", Sys.info()["nodename"], getwd(), "\n", format(Sys.time(), "%Y %b %d %a %X"), " ? ", sep = ""))
    td <- new.time - time
    if (td["elapsed"] > td.co) 
        .r(.sys(td["elapsed"], "elapsed"), content = paste(capture.output(substitute(expr)), collapse = "\n"))
    structure(new.time - time, class = "proc_time")
}
fp_temp.38_05_01_195807 <- file.path(if (dir.exists("/dev/shm")) "/dev/shm" else hs.home.expand("~"), Sys.info()["user"], "38_05_01_171338.txt.gz")
ieiwk.repl <- function(env = parent.frame(), env1 = new.env(parent = env), init = {
}, ...) {
    A <- list(...)
    repeat {
        E.38_05_01_195517 <- do.call(ieiwk.parse, A)
        if (!length(E.38_05_01_195517)) 
            next
        if (hs.all(hs.len1(E.38_05_01_195517), is(E.38_05_01_195517[[1]], "name"), paste(E.38_05_01_195517[[1]]) == ".qq")) 
            return()
        loadhistory(hs.home.expand("~/.Rhistory"))
        time1.38_05_02_230459 <- Sys.time()
        cat("\033[31m****************", paste(time1.38_05_02_230459), "[表达式] ****************\033[0m\n", file = paste("| gzip >>", fp_temp.38_05_01_195807))
        cat(paste(E.38_05_01_195517), sep = "\n", file = paste("| gzip >>", fp_temp.38_05_01_195807))
        a.38_05_01_195727 <- .try(eval(E.38_05_01_195517, env))
        time_diff.38_05_03_000503 <- Sys.time() - time1.38_05_02_230459
        jg_tag.38_05_03_230943 <- paste(paste(Sys.time()), "[结果]", capture.output(time_diff.38_05_03_000503))
        jg_line.38_05_04_111217 <- paste("\033[31m****************", jg_tag.38_05_03_230943, "****************\033[0m\n")
        cat(jg_line.38_05_04_111217, file = paste("| gzip >>", fp_temp.38_05_01_195807))
        cat(jg_line.38_05_04_111217)
        units(time_diff.38_05_03_000503) <- "secs"
        if (time_diff.38_05_03_000503 > 30) {
            hs.hsgly("report.job.is.done.36.03.02.104436")(jg_tag.38_05_03_230943, body = paste(E.38_05_01_195517))
        }
        if (file.size(fp_temp.38_05_01_195807) > 1e+07) {
            L.38_05_01_195750 <- readLines(fp_temp.38_05_01_195807)
            cat(tail(L.38_05_01_195750, max(1, floor(length(L.38_05_01_195750) * 0.7))), sep = "\n", file = paste("| gzip >", fp_temp.38_05_01_195807))
        }
        local({
            con1.38_05_01_195851 <- pipe(paste("head -n 20 | gzip >>", fp_temp.38_05_01_195807), open = "w")
            on.exit(close(con1.38_05_01_195851), add = TRUE)
            capture.output(a.38_05_01_195727, file = con1.38_05_01_195851)
        })
        if (((class(E.38_05_01_195517[[1]]) %in% c("<-", "=")) && (!is(a.38_05_01_195727, "try-error"))) || is(a.38_05_01_195727, "help_files_with_topic")) {
        }
        else local({
            a <- capture.output(a.38_05_01_195727)
            if (length(a) > 20) 
                a[20] <- "..."
            cat(head(a, 20), sep = "\n")
        })
    }
}
ieiwk.parse <- local({
    hs.readline_saveHistory <- function(p1) {
        cmd1 <- readline(p1)
        e1 <- .try(parse(text = cmd1))
        if (!input.is.incomplete(e1)) 
            cat(cmd1, sep = "\n", append = TRUE, file = hs.home.expand("~/.Rhistory"))
        cmd1
    }
    function(prefix = "") {
        p1 <- paste0("\033[31m", Sys.info()["user"], "@", Sys.info()["nodename"], ":", getwd(), "\033[0m\n", format(Sys.time(), "%Y %b %d %a %X"), "\n", prefix, " R> ", sep = "")
        cmd <- hs.readline_saveHistory(p1)
        repeat {
            e1 <- .try(parse(text = cmd))
            if (input.is.incomplete(e1)) {
                cmd1 <- hs.readline_saveHistory("and? > ")
                cmd <- c(cmd, cmd1)
            }
            else {
                return(e1)
            }
        }
    }
})
ieiwk.repl.eval <- function(e1, env = .GlobalEnv) {
    length(e1[[1]])
    {
        if (1) 
            aaa
    }
    mode(e1[[1]])
    mode(e1[[1]][[2]])
    is.function(e1[[1]][[length(e1[[1]])]][[2]])
    mode(e1[[1]][[length(e1[[1]])]])
    mode(tail(e1[[1]], 1)[[1]])
    apropos("function")
    cmd <- {
    }
    repeat {
        cmd <- c(cmd, readline() %not% "")
        if (!length(cmd)) 
            next
        r <- .try(eval(parse(text = cmd), env = env))
        if (!input.is.incomplete) 
            return(r)
    }
}
ieiwk.get.function <- function(s, r = c("definition", "symbol", "name")) {
    r <- match.arg(r)
    a <- apropos(s, mode = "function")
    switch(r, definition = get(a), symbol = as.symbol(a), name = a)
}
ieiwk.load.or.rerun <- function(exp, fp, rerun, env = parent.frame()) {
    if ({
        !rerun
    } & {
        !file.exists(fp)
    }) 
        rerun <- 1
    if (rerun) 
        eval(exp, env)
    else load(fp, env)
}
ieiwk.get.callForm <- function(...) {
    match.call()[-1]
}
ieiwk.dynamicScope <- function(fn, env = parent.frame()) {
    environment(fn) <- env
    fn
}
defmacro <- function(..., expr) {
    expr <- substitute(expr)
    a <- substitute(list(...))[-1]
    nn <- names(a)
    if (is.null(nn)) 
        nn <- rep("", length(a))
    for (i in 1:length(a)) {
        if (nn[i] == "") {
            nn[i] <- paste(a[[i]])
            msg <- paste(a[[i]], "not supplied")
            a[[i]] <- substitute(stop(foo), list(foo = msg))
        }
        if (nn[i] == "DOTS") {
            nn[i] <- "..."
            a[[i]] <- formals(function(...) {
            })[[1]]
        }
    }
    names(a) <- nn
    a <- as.list(a)
    ff <- eval(substitute(function() {
        tmp <- substitute(body)
        eval(tmp, parent.frame())
    }, list(body = expr)))
    formals(ff) <- a
    mm <- match.call()
    mm$expr <- NULL
    mm[[1]] <- as.name("macro")
    attr(ff, "source") <- c(deparse(mm), deparse(expr))
    ff
}
strmacro <- function(..., expr, strexpr) {
    if (!missing(expr)) 
        strexpr <- deparse(substitute(expr))
    a <- substitute(list(...))[-1]
    nn <- names(a)
    if (is.null(nn)) 
        nn <- rep("", length(a))
    for (i in 1:length(a)) {
        if (nn[i] == "") {
            nn[i] <- paste(a[[i]])
            msg <- paste(a[[i]], "not supplied")
            a[[i]] <- substitute(stop(foo), list(foo = msg))
        }
        else {
            a[[i]] <- a[[i]]
        }
    }
    names(a) <- nn
    a <- as.list(a)
    ff <- function(...) {
        reptab <- a
        reptab$... <- NULL
        args <- match.call(expand.dots = TRUE)[-1]
        for (item in names(args)) reptab[[item]] <- args[[item]]
        body <- strexpr
        for (i in 1:length(reptab)) {
            pattern <- paste("\\b", names(reptab)[i], "\\b", sep = "")
            value <- reptab[[i]]
            if (missing(value)) 
                value <- ""
            body <- gsub(pattern, value, body)
        }
        fun <- parse(text = body)
        eval(fun, parent.frame())
    }
    formals(ff) <- a
    mm <- match.call()
    mm$expr <- NULL
    mm[[1]] <- as.name("macro")
    attr(ff, "source") <- c(deparse(mm), strexpr)
    ff
}
evalStr <- function(s, env = parent.frame()) {
    eval(parse(text = s), env)
}
evalStrArg <- function(s, env = parent.frame()) {
    s0 <- try(parse(text = s), silent = TRUE)
    switch(typeof(s0), character = {
        s1 <- strsplit(attr(s0, "condition")[[1]], "\n")[[1]]
        se <- as.integer(strsplit(re("\\d+:\\d+", s1[1]), ":")[[1]])
        evalStr(substr(s, se[1], se[2] - 1L), env)
        s <- substring(s, se[2] + 1)
        evalStrArg(s, env)
    }, eval(s0, env))
    invisible()
}
ieiwk.mp.code2str.r <- function(..., trim = 0) {
    a <- list(...)
    r <- sapply(a, function(x) {
        if (typeof(x) != "character") 
            x <- capture.output(x)
        x
    })
    if (is.list(r)) 
        r <- unlist(r)
    r <- c(r)
    if (trim) 
        r <- do.call("trim", list(r))
    r
}
wk.evalLazy <- ieiwk.evalLazy <- function(r, e, uid, root = formals(.mfp.swxx)[["dataDir"]], fp = {
}, ext = if ("TxDb" %in% class) "sql" else "rd", lazy = 1, ow = 1, env = parent.frame(), env2 = if (local) new.env() else parent.frame(), local = 1, fp.only = 0, dtt = "handy", save.jg = 1, want = "jg", hs1.save = {
}, reader = {
}, lib = {
}, class = {
}) {
    env
    env2
    if (substitute(e)[1] != substitute(expression())) 
        e <- substitute(e)
    if (!length(fp)) 
        fp <- hs.wjm(.mfp(uid, root, dtt = dtt), add = ext)
    if (fp.only) 
        return(fp)
    if (!length(reader)) 
        reader <- hs.case(hs.grepl(fp, "(?i)\\.(tsv|csv|txt)(\\.gz)?$"), fread, hs.grepl(fp, "(?i)\\.xlsx?$"), wk.xls.read, hs.grepl(fp, "(?)\\.rda?(ta)?$"), hs.load, no_match = readLines)
    if (!length(hs1.save)) 
        hs1.save <- switch(tolower(hs.file.extension(fp)), xlsx = , xls = wk.xls.write, gz = wk.dt.fwrite, rd = hs.save, function(x1, file) cat(x1, sep = "\n", file = file))
    e1 <- substitute(e)
    if (class(e1) == "{") 
        e <- e1
    if ("TxDb" %in% class) {
        hs1.save <- AnnotationDbi::saveDb
        reader <- AnnotationDbi::loadDb
    }
    if (save.jg) {
        if (file.exists(fp) && (lazy || (!ow))) {
        }
        else {
            for (lib1 in lib) library(lib1)
            jg <- eval(e, env2)
            hs1.save(jg, file = fp)
        }
    }
    switch(want, jg = {
        if (missing(r)) {
            reader(fp)
        } else {
            assign(r, reader(fp), envir = env)
        }
    }, fp = fp)
}
ieiwk.evalLazy.2 <- local(function(bds, bds2, update = 0, pre = if (!missing(uid)) .mfp(uid), uid, what = "jg") {
    if (length(pre)) 
        wj <- sapply(c("bds", "jg"), . %=>% hs.wjm(pre, add = x))
    env1 <- environment(sys.function())
    if (!missing(bds)) 
        bds2 <- as.expression(substitute(bds))
    if (missing(bds2)) {
        if (length(pre) && file.exists(wj["bds"])) 
            bds2 <- parse(wj["bds"], encoding = "UTF-8")
    }
    if (missing(bds2)) {
    }
    else {
        e1 <- quote(assign("bds", bds2, envir = env1))
        if (ieiwk.exists("bds", envir = env1)) {
            bds.old <- trim(capture.output(get("bds", envir = env1)))
            bds.new <- trim(capture.output(bds2))
            if ({
                length(bds.old) != length(bds.new)
            } || any(bds.old != bds.new)) {
                eval(e1)
                if (length(pre)) 
                  .ss(bds.new, file = wj["bds"])
                update <- 1
            }
        }
        else eval(e1)
    }
    {
        e2 <- quote(eval(quote(jg <- eval(bds)), env = env1))
        if (ieiwk.exists("jg", envir = env1)) {
            if (update) {
                eval(e2)
                if (length(pre)) 
                  save(jg, file = wj["jg"], envir = env1)
            }
        }
        else {
            if (length(pre) && file.exists(wj["jg"])) 
                eval(bquote(jg <- hs.load(.(wj["jg"]))), env1)
            else {
                eval(e2)
                if (length(pre) && file.exists(wj["jg"])) 
                  ieiwk.save("jg", wj["jg"], env1 = env1)
            }
        }
    }
    switch(what, jg = , bds = get(what, envir = env1), pre = pre, wj.jg = , wj.bds = hs.wjm(pre, add = re("[^\\.]+$", what)))
})
ieiwk.mp.assign <- function(..., v = {
}) {
    obj <- substitute(...)
    if (length(obj) > 1) {
        r <- sapply(seq(2, length(obj)), function(i) {
            x <- obj[[i]]
            b.29.12.07.032017 <- eval(x)
            substitute(a <- b, list(a = x, b = b.29.12.07.032017))
        })
        r <- sapply(r, function(x) {
            capture.output(x)
        })
    }
    else {
        r <- capture.output(substitute(a <- b, list(a = obj, b = eval(obj))))
    }
    r
}
bin.perl <- "perl"
ieiwk.mp.perl.cmd <- function(cmd, options = {
}, run = 0, intern = 1) {
    my.perl.lib <- switch(hostname, `kai-VirtualBox` = )
    cmd.default <- "use strict;\nuse warnings;\nuse lib \"/public/home/wengkai/calvin/code/perl\";\nuse subroutineCommon;\nuse feature qw(say);\n"
    cmd <- sprintf("%s%s", cmd.default, cmd)
    if (run) 
        .sys.do(bin.perl, options, "-e " %.% shQuote(strsplit(cmd, "\\n")[[1]]), intern = intern)
    else .sys(bin.perl, options, "-e " %.% shQuote(strsplit(cmd, "\\n")[[1]]), intern = intern)
}
ieiwk.eval.sysCmd <- function(cmd, wj.log) {
    ieiwk.eval(bquote({
        .sys.do(.(cmd), log = .(wj.log))
    }))
}
wk.e <- wk.eval <- ieiwk.eval <- function(e1, n1 = Sys.info()["nodename"], f1 = ieiwk.tempfile(), f2 = ieiwk.tempfile(), bg = 0, intern = !bg, ssh = n1 != Sys.info()["nodename"], via.file.code = bg || ssh || length(ssh.info) || length(ssh.name), via.file.result = ssh && intern && !bg, f2.remove = via.file.result, env = parent.frame(), use.wd = as.logical(one.sys), lib = {
}, job.uid, email = 0, report.if.done = email, time1 = Sys.time(), report.if.done.cutoff.time = 10, newR = 0, one.sys = 1, ssh.info = {
}, ssh.name = {
}) {
    e0 <- list()
    if (length(lib)) 
        e0 <- append(e0, unname(sapply(lib, function(l1) bquote(library(.(l1))))))
    if (use.wd) 
        e0 <- append(e0, bquote(setwd(.(getwd()))))
    e1 <- bquote({
        t1 <- Sys.time()
        "33.04.19.192621" <- .(switch(typeof(e1), expression = e1[[1]], language = e1))
        t2 <- Sys.time()
        td <- difftime(t2, t1, units = "secs")
        if (.(report.if.done) || td > .(report.if.done.cutoff.time)) {
            .ss(c("33.04.19.185828", paste(t2), paste(t1), difftime(t2, t1, units = "mins")), file = hs.home.expand("~/test.33.04.19.185834.txt"))
            .r(capture.output(td), "")
        }
    })
    if (via.file.code && !bg) 
        on.exit(unlink(f1), add = TRUE)
    if (f2.remove && !bg) 
        on.exit(unlink(f2), add = TRUE)
    if (via.file.code) {
        sapply(c(e0, e1), function(e1) {
            cat(capture.output(e1), file = f1, append = TRUE, sep = "\n")
        })
        if (0) {
            .less(f1)
        }
        cmd <- {
            acc1 <- paste(c("source(", shQuote(f1), ")", ";", "unlink(", shQuote(f1), ")", if (via.file.result) c(";", "save( ", .q("33.04.19.192621"), ", file = ", wk.wjlj(f2), ")")), collapse = " ")
            if (ssh) 
                acc1 <- shQuote(acc1)
            paste(c(if (ssh) c("ssh", n1), .q(c("source activate ER; R --no-save --slave -e", acc1)), if (bg) "&"), collapse = " ")
        }
        if (!one.sys) {
            wk.rsync.out(f1, hs.home.expand("~/tmp"), ssh = ssh.name)
            ssh.info <- ieiwk.ssh.db(ssh.name)
            .wk("ssh", "-p", ssh.info["port"], "-l", ssh.info["user"], ssh.info["dz"], shQuote(cmd))
            return()
        }
        r <- system(cmd)
        if (via.file.result) {
            r <- hs.load(f2)
        }
        r
    }
    else {
        if (typeof(e1) == "list") {
            for (e1.1 in head(e1, length(e1) - 1)) r <- eval(e1.1, env = env)
            eval(e1[[length(e1)]], env = env)
            r
        }
        else {
            eval(e1, env = env)
            get("33.04.19.192621", envir = env)
        }
    }
}
hs.substitute <- function(E) {
    E1 <- substitute(E)
    if ("name" %in% class(E1)) 
        return(E)
    if ("call" %in% class(E1)) 
        if (paste(E1[[1]]) %in% c("quote", "bquote", "expression")) 
            E <- E
    substitute(E)
}
ieiwk.eval.helper <- function(f1, f2 = ieiwk.tempfile()) {
    r <- eval(parse(text = readLines(f1)))
    save(r, file = f2)
    invisible(f2)
}
ieiwk.call0 <- function(f, ...) {
    a <- list(...)
    r <- as.call(c(as.symbol(tail(f, 1)), a))
    f <- head(f, -1)
    while (length(f)) {
        r <- call(tail(f, 1), r)
        f <- head(f, -1)
    }
    r
}
ieiwk.call <- function(f, ...) {
    a <- tail(as.list(match.call()), -2)
    r <- as.call(c(as.symbol(tail(f, 1)), a))
    f <- head(f, -1)
    while (length(f)) {
        r <- call(as.symbol(tail(f, 1)), r)
        f <- head(f, -1)
    }
    r
}
ieiwk.access.list <- function(p, last = c("element", "list")) {
    r <- as.symbol(p[1])
    if (length(p) > 2) 
        for (i in seq(2, length(p) - 1)) r <- bquote(.(r)[[.(p[i])]])
    if (length(p) > 1) 
        r <- switch(match.arg(last), list = bquote(.(r)[.(tail(p, 1))]), element = bquote(.(r)[[.(tail(p, 1))]]))
    r
}
wk.a <- ieiwk.access <- function(...) {
    p2 <- list(...)
    acc <- p2[[1]]
    for (p1 in tail(p2, -1)) {
        fn <- switch(class(acc), list = {
            ifelse(length(p1) > 1, `[`, `[[`)
        }, environment = `$`, S4 = `@`, `[`)
        acc <- fn(acc, p1)
    }
    acc
}
.get <- function(...) {
    p <- as.list(match.call())
    fn <- switch(eval(call("typeof", p[[2]]), parent.frame()), list = , environment = "[[", S4 = "@", "[")
    p[[3]] <- eval(p[[3]], parent.frame())
    fa <- call(fn, p[[2]], p[[3]])
    for (p1 in tail(p, -3)) {
        fn <- switch(eval(call("typeof", fa), parent.frame()), list = , environment = "[[", S4 = "@", "[")
        fa <- call(fn, fa, eval(p1, parent.frame()))
    }
    eval(fa, parent.frame())
}
.set <- function(x, value = "", envir = parent.frame()) {
    fa <- as.list(as.list(match.call())$x[-1])
    eval(call("<-", eval(as.call(c(as.symbol("ieiwk.access"), fa))), value), envir)
}
.bt <- browserText
.brs <- defmacro(a, expr = {
    .cat("browsing at ", a)
    browser(a)
})
ieiwk.env.here <- function(env = parent.frame()) env
wk.33.06.24.202656 <- function(env.33.06.24.205427 = parent.frame(), hs.parent = sys.function(1)) {
    evalq(local({
        hs.parent <- sys.function(1)
        formals(hs.33.06.24.203851) <- head(formals(hs.parent), -3)
        args <- head(formalArgs(hs.parent), -3)
        PL <- mget(args %rm% "...", inherits = TRUE)
        if ("..." %in% args) 
            PL <- append(PL, list(...))
        an <- names(PL)
        if (length(an)) 
            PL <- PL[order(c(an, rep("", length(PL) - length(an))), sapply(PL, object.size))]
        pc1 <- hs.try_with_time_limit(digest::digest(PL), 3)
        if (length(pc1)) {
            if (!hs.exists(pc1, env2.33.06.24.181217)) {
                assign(pc1, eval(bquote(do.call(hs.33.06.24.203851, PL))), env2.33.06.24.181217)
            }
            assign("jg", get(pc1, envir = env2.33.06.24.181217), envir = env1.33.06.24.181214)
        }
        else {
            .ss("(re)calculating 34.07.05.191435")
            assign("jg", do.call(hs.33.06.24.203851, PL), envir = env1.33.06.24.181214)
        }
        get("jg", envir = env1.33.06.24.181214)
    }), env.33.06.24.205427)
}
wk.33.06.24.202656.bak <- function(env.33.06.24.205427 = parent.frame(), hs.parent = sys.function(1)) {
    eval(bquote(local({
        args <- .(names(formals(hs.parent)))
        PL <- sapply(args %not% "...", dynGet)
        if ("..." %in% args) 
            PL <- c(PL, list(...))
        if (object.size(PL) > mnemonic.object_size.cutoff.33.06.24.180623) 
            mnemonic.33.06.24.180617 <- 0
        if (mnemonic.33.06.24.180617) {
            PL <- PL[order(names(PL))]
            pc1 <- digest::digest(PL)
            if (!wk.exists(pc1, env2.33.06.24.181217)) {
                assign(pc1, hs.33.06.24.203851(), envir = env2.33.06.24.181217)
                assign("jg", get(pc1, envir = env2.33.06.24.181217), envir = env1.33.06.24.181214)
            }
        }
        else {
            if (!wk.exists("jg", env1.33.06.24.181214) || ..new) 
                assign("jg", hs.33.06.24.203851(), envir = env1.33.06.24.181214)
        }
        jg
    })), env.33.06.24.205427)
}
hs.import <- wk.import <- function(e1, e2 = parent.frame(), only.hs = 0, exclude.hs = 0, what = {
}, exclude = "env2.33.06.24.181217", tolist = 0, ls = 0, replace = 0) {
    if (is.function(e1)) 
        e1 <- environment(e1)
    if (is.function(e2)) 
        e2 <- environment(e2)
    if (inherits(e1, "list")) {
        for (on1 in names(e1)) assign(on1, e1[[on1]], envir = e2)
        return()
    }
    a <- unlist(eapply(e1, base::is.function, all.names = TRUE))
    if (length(what)) 
        a <- a[base::match(what, base::names(a))]
    exclude <- exclude %and% base::names(a)
    if (length(exclude)) {
        a <- a[-base::match(exclude, base::names(a))]
    }
    if (NA %in% a) {
        message("####", base::names(a[base::is.na(a)]), "is not found.", sep = " ")
    }
    if (only.hs) 
        a <- a[a]
    if (exclude.hs) 
        a <- a[!a]
    if (length(a)) {
        if (ls) {
            return(names(a))
        }
        if (tolist) {
            mget(names(a), e1)
        }
        else {
            for (i in seq_along(a)) {
                on1 <- names(a[i])
                eval(hs.bq({
                  if ((!hs.exists(.(on1), inherits = FALSE)) || .(replace)) {
                    assign(.(on1), .(get(on1, envir = e1)))
                    if (.(a[i])) 
                      hs.env(.(as.symbol(on1))) <- hs.env()
                  }
                }), e2)
            }
        }
    }
}
exp_str.hs.v <- function(v1) {
    .q(v1) %=>% paste(collapse = ", ") %=>% paste0("c( ", x, ")")
}
hs.swap <- function(x, y, env = parent.frame()) {
    x.name <- paste(substitute(x))
    y.name <- paste(substitute(y))
    if (all(hs.exists(c(x.name, y.name), env))) {
        a <- x
        assign(x.name, y, envir = env)
        assign(y.name, a, envir = env)
    }
    else stop("[38_05_14_183507]", x.name, "和", y.name, "须同时存在于当前环境")
}
hs.formals.merge <- function(..., old = "replace") {
    fl <- list(...)
    f1 <- formals(fl[[1]])
    for (i in hs.seqForward(2, length(fl))) {
        f2 <- formals(fl[[i]])
        a <- names(f1) %and% names(f2)
        if (length(a)) {
            switch(old, keep = c(f1, f2[names(f2) %not.in% a]), replace = {
                f1[a] <- f2[a]
                f2[a] <- {
                }
            })
        }
        f1 <- c(f1, f2)
    }
    f1
}
hs.call <- function(hs.38_09_26_084109, ..., env1 = parent.frame()) {
    nv <- names(formals(hs.38_09_26_084109))
    a <- list(...)
    if (length(a)) {
        nv <- nv %not% c("...", names(a))
        n1 <- sum(names(a) == "")
        if (n1) 
            nv <- tail(nv, -n1)
    }
    if (length(nv)) 
        a <- c(a, mget(nv, envir = env1, ifnotfound = formals(hs.38_09_26_084109)[nv]))
    do.call(hs.38_09_26_084109, a)
}
hs.call.cmd <- function(f1) {
    a <- names(formals(f1))
    cmd <- paste0(substitute(f1), "( ", paste(paste0(a, " = ", a), collapse = ", "), ")")
    message(cmd)
}
hs.find_function <- function(hs1) {
    a <- findFunction(hs1)
    if (length(a)) {
        env1 <- a[[1]]
        env1[[hs1]]
    }
}
hs.semicolon2 <- function(p, f) {
    do.call("::", list(p, f))
}
hs.call... <- function(E, .., env1 = parent.frame()) {
    eval(bquote(local({
        hs1 <- function(...) {
            .(substitute(E))
        }
        do.call(hs1, as.list(.(..)))
    })), envir = env1)
}
hs.hs.reduce <- function(e1) {
    A <- as.list(e1[-1])
    hs.switch(paste(e1[[1]]), "hs.require", {
        e1[[1]] <- quote(library)
    }, "hs.fp", {
        hs.browser("2023.08.14.190031", debug = 0)
        i <- match("robust", names(A), 0)
        if (i && A[[i]]) {
            B <- body(hs.fp) %=>% as.list
            B[[3]] %=>% as.list
            A
            formals(hs.fp)
        }
        e1[[1]] <- quote(file.path)
    })
    e1
}
hs.arg_supplied_p <- function(name, hs2, ...) {
    i <- match(name, ...names(), nomatch = 0)
    if (i) 
        return(i)
    hs.browser("2023.08.29.192954", debug = 0)
    al <- formals(hs2)
    p1 <- match(name, names(al))
    ...names()
}
hs.pastec <- function(..., collapse = " ") .Internal(paste0(list(c(...)), collapse))
hs.pastec0 <- function(...) .Internal(paste0(list(c(...)), ""))
hs.pastec. <- function(...) .Internal(paste0(list(c(...)), "."))
hs.pastec_ <- function(...) .Internal(paste0(list(c(...)), "_"))
re.helper <- function(a, data) {
    att <- attributes(a)
    x <- a
    y <- a + att$match.length - 1
    r <- substring(data, x, y)
    if ("capture.names" %in% names(att)) {
        r <- list(match = r)
        r[["capture"]] <- {
            b <- sapply(seq_along(att$capture.names), function(i) {
                sapply(seq.int(nrow(att$capture.start)), function(j) {
                  start <- att$capture.start[j, i]
                  len <- att$capture.length[j, i]
                  substring(data[j], start, start + len - 1)
                })
            })
            if (hs.len1(dim(b), 2)) {
                colnames(b) <- att$capture.names
            }
            else {
                names(b) <- att$capture.names
            }
            b
        }
    }
    r
}
re <- function(pat, data, ignore.case = TRUE, perl = TRUE, fixed = FALSE) {
    a <- regexpr(pat, data, ignore.case, perl, fixed)
    re.helper(a, data)
}
gre.helper <- function(a, data) {
    att <- attributes(a)
    x <- a
    y <- a + att$match.length - 1
    r <- sapply(seq.int(length(x)), function(i) substr(data, x[i], y[i]))
    if ("capture.names" %in% names(att)) {
        r <- list(match = r)
        r[["capture"]] <- {
            b <- as.vector(sapply(seq.int(length(att$capture.names)), function(i) {
                sapply(seq.int(nrow(att$capture.start)), function(j) {
                  start <- att$capture.start[j, i]
                  len <- att$capture.length[j, i]
                  substr(data, start, start + len - 1)
                })
            }))
            b <- matrix(NA, nrow(att$capture.start), length(att$capture.names))
            for (i in seq.int(length(att$capture.names))) {
                for (j in seq.int(nrow(att$capture.start))) {
                  start <- att$capture.start[j, i]
                  len <- att$capture.length[j, i]
                  b[j, i] <- substr(data, start, start + len - 1)
                }
            }
            colnames(b) <- att$capture.names
            b
        }
    }
    r
}
gre <- function(pat, data, ignore.case = 0, perl = TRUE, fixed = 0, useBytes = FALSE) {
    a <- gregexpr(pat, data, ignore.case, perl, fixed, useBytes)
    lapply(seq_along(a), function(i) gre.helper(a[[i]], data[i]))
}
hs.strsplit.v <- function(split = sep, sep = "auto") {
    if (!is.character(x)) 
        x <- hs.str_as(x)
    if (split == "auto") {
        split <- hs.detect.sep(x, sep_set = split)
        if (!nzchar(split)) 
            split <- {
            }
    }
    jg <- if (length(split)) {
        strsplit(x = x, split = split, fixed = fixed, perl = perl, useBytes = useBytes)
    }
    else x
    unlist(jg)
}
formals(hs.strsplit.v) <- hs.formals.merge(strsplit, hs.strsplit.v, old = "replace")
hs.strsplit <- function(x, split, fixed = FALSE, perl = TRUE, useBytes = FALSE) {
    strsplit(x = x, split = split, fixed = fixed, perl = perl, useBytes = useBytes)
}
hs.substr.replace <- function(s1, i1, i2, s2) {
    sprintf("%s%s%s", substr(s1, 1, i1 - 1), s2, substring(s1, i2 + 1))
}
hs.regmatches <- function(s1, m1, invert = FALSE, no_match = "", replacement = {
}, env1 = parent.frame()) {
    if (length(replacement)) {
        eval(bquote(regmatches(.(substitute(s1)), .(m1), .(invert)) <- .(replacement)), env1)
    }
    else {
        a1 <- regmatches(s1, m1)
        if (length(no_match)) {
            a2 <- rep(no_match, length(s1))
            a2[m1 > 0] <- a1
            a2
        }
        else a1
    }
}
hs.detect.sep <- function(v, sep_set = "", ...) {
    if (!nzchar(sep_set)) 
        sep_set <- "[;\\|]"
    step_size <- 2^10
    lv <- length(v)
    i1 <- 1
    repeat {
        if (i1 > lv) 
            return("")
        i2 <- min(i1 + step_size, lv)
        v1 <- v[i1:i2]
        gi <- hs.grep(v1, sep_set, ...)
        jg <- if (length(gi)) 
            hs.re(v1[gi[1]], sep_set, ...)
        else ""
        if (nzchar(jg)) 
            return(jg)
        i1 <- i2 + 1
    }
}
hs.format.same_width <- function(V) {
    width1 <- max(nchar(V))
    sprintf(sprintf("%%%ds", width1), V)
}
hs.paste <- function(..., sep = "", cat = 0, end = "\n") {
    R <- paste(c(...), collapse = sep)
    if (cat) {
        cat(R, sep = end)
    }
    R
}
paste. <- function(...) {
    paste(..., sep = ".")
}
hs.paste. <- function(...) {
    paste.(c(...), collapse = ".")
}
pasteT <- function(...) {
    paste(..., sep = "\t")
}
hs.pasteT <- function(...) {
    paste.(c(...), collapse = "\t")
}
hs.file.extension <- function(fp, ext.pattern = "\\.[^.]+$", extN = 1, include. = FALSE) {
    if (extN) {
        hs.re(fp, sprintf("%s%s$", ifelse(include., "\\.", ""), paste(rep("[^\\./]*", extN), collapse = "\\.")))
    }
    else fp
}
hs.str_merge <- function(s1, s2 = {
}, sep = ";", sep2 = sep, clean = 0, unique = clean, ...) {
    a <- strsplit(c(s1, s2), sep, ...)
    a <- unlist(a)
    if (unique) {
        a <- unique(a)
    }
    if (clean) {
        a <- hs.clean(a, U = unique)
    }
    if (length(a)) 
        a <- a[order(nchar(a), a)]
    paste(a, collapse = sep2)
}
hs.str_subtract <- function(s1, s2, sep = ";", ...) {
    a <- strsplit(c(s1, s2), sep, ...)
    a <- a[[1]] %not% a[[2]]
    a <- hs.clean(a, U = 1)
    a <- a[order(nchar(a), a)]
    paste(a, collapse = sep)
}
hs.str_n <- function(s1, sep = ";") {
    sapply(hs.str_split(s1, sep), length)
}
hs.str_in <- function(s1, s2, sep = ";", sep2 = sep, all = 0, ic = 1) {
    s2 <- hs.clean(hs.strsplit.v(s2, split = sep2), U = 1)
    if (ic) {
        s1 <- tolower(s1)
        s2 <- tolower(s2)
    }
    hs1 <- if (all) 
        base::all
    else base::any
    sapply(strsplit(paste(s1), sep), function(x) if (length(x)) 
        hs1(x %in% s2)
    else FALSE)
}
hs.str_split <- function(s1, sep = ";", fixed = FALSE, ..., reversible = TRUE) {
    jg <- strsplit(s1, sep, fixed = fixed, ...)
    if (reversible) {
        f1 <- hs.grepl(s1, paste0(sep, "$"))
        nm <- names(jg)
        jg <- lapply(seq_along(jg), function(i1) {
            . <- jg[[i1]]
            if (is.null(.)) {
                . <- ""
            }
            else if (f1[i1]) {
                . <- c(., "")
            }
            .
        })
        names(jg) <- nm
    }
    jg
}
hs.str_flat <- function(s1, sep = ";", ot = "v") {
    a1 <- hs.str_split(s1, sep)
    switch(ot, v = unlist(a1), i = rep(seq_along(s1), sapply(a1, length)), all = list(v = unlist(a1), i = rep(seq_along(s1), sapply(a1, length))))
}
hs.str_pad <- function(str, width = max(nchar(str)), pad = " ", side = "left") {
    if (identical(pad, " ") && (side == "left")) {
        sprintf(paste0("%0", width, "s"), str)
    }
    else {
        jg <- paste(str)
        ns <- nchar(jg)
        i <- which(ns < width)
        if (length(i)) {
            w1 <- width - ns[i]
            a <- hs.gsub(sprintf(paste0("%", w1, "s"), ""), ".", pad)
            switch(side, left = {
                jg[i] <- paste0(a, jg[i])
            }, right = {
                jg[i] <- paste0(jg[i], a)
            })
        }
        jg
    }
}
hs.str_nonredundant <- function(x) {
    y <- toupper(x)
    if (uniqueN(y) == uniqueN(x)) 
        x
    else y
}
hs.str_list_merge <- function(..., toupper = 0) {
    a <- trim(unlist(list(...)))
    a <- if (toupper) 
        toupper(a)
    else hs.str_nonredundant(a)
    sort(unique(a))
}
hs.str_and <- function(v1, v2, inv = 0) {
    i <- match(tolower(v1), tolower(v2))
    if (inv) {
        v1[which(is.na(i))]
    }
    else {
        ii <- which(!is.na(i))
        if (length(ii)) 
            unique(v2[i[ii]])
    }
}
hs.str_not <- function(v1, v2) {
    hs.str_and(v1, v2, inv = 1)
}
hs.str_match <- function(v1, v2, ot = "i", ic = 1, ...) {
    if (ic) {
        v1 <- tolower(v1)
        v2 <- tolower(v2)
    }
    i <- match(v1, v2, ...)
    hs.switch(ot, "i", i, "l", !is.na(i), "v2", v2[i], "v1", v1[!is.na(i)])
}
hs.str_diversify <- function(v1, sep = "_", skip1 = 0) {
    a <- tapply(seq_along(v1), v1, function(i) {
        if (length(i) > 1) 
            i
    })
    a <- hs.nonempty(a)
    i <- unlist(a)
    v1[i] <- if (skip1) {
        paste0(v1[i], unlist(lapply(a, function(i) c("", paste0(sep, seq(length(i) - 1))))))
    }
    else paste0(v1[i], sep, unlist(lapply(a, seq_along)))
    v1
}
hs.str_as <- function(x, ...) {
    y <- as.character(x)
    i <- hs.clean(x, ot = "l", inv = 1, ...)
    if (any(i)) 
        y[i] <- ""
    y
}
hs.str_cut <- function(x, p1, all = 1, hs1 = if (all) gregexpr else regexpr, ic = TRUE, perl = !fixed, fixed = FALSE, at = "p1", at_once = 1, inclusive = at != "p1") {
    m <- hs1(p1, x, ignore.case = ic, perl = perl, fixed = fixed)
    if (at == "after") 
        for (i in seq_along(m)) {
            m1 <- m[[i]]
            if (identical(m1[[1]], -1L)) 
                next
            m1 <- m1 + attr(m1, "match.length")
            j <- m1 > nchar(x[i])
            if (any(j)) {
                j <- which(!j)
                if (length(j)) {
                  m[[i]] <- hs.attr.slice(m1, j)
                }
                else m[i] <- list({
                })
            }
            else m[[i]] <- m1
        }
    if (at != "p1") 
        for (i in seq_along(m)) {
            if (length(m[[i]])) 
                attr(m[[i]], "match.length") <- rep(0, length(m[[i]]))
        }
    if (!at_once) {
        a <- lapply(m, function(m1) {
            if (length(m1) > 1) 
                lapply(seq_along(m1), function(i) hs.attr.slice(m1, i))
            else list(m1)
        })
        m <- unlist(a, recursive = FALSE)
        x <- rep(x, sapply(a, length))
    }
    regmatches(x, m, invert = TRUE)
}
hs.str_sort <- function(s, ic = TRUE, ...) {
    s[order(if (ic) 
        tolower(s)
    else s, ...)]
}
hs.pv.format <- function(p1, digits = 3) {
    paste("pv:", format(p1, scientific = 1, digits = digits))
}
format.hs.size <- function(size1, unit = "auto") {
    class(size1) <- "object_size"
    format(size1, unit = unit)
}
hs.title <- function(string1) {
    substr(string1, 1, 1) <- toupper(substr(string1, 1, 1))
    substring(string1, 2) <- tolower(substring(string1, 2))
    string1
}
hs.str_split_fixed <- function(string, pattern = "\t", N = {
}, I = {
}, from = 1, ot = "dt", perl = !fixed, fixed = FALSE, ignore.case = TRUE) {
    m1 <- gregexpr(pattern, string, perl = perl, fixed = fixed)
    n1 <- length(m1[[1]])
    if (!length(N)) 
        N <- n1 + 1
    V <- if (length(I)) {
        I
    }
    else {
        switch(from, seq_len(N - 1), seq.int(n1 - N + 2, n1))
    }
    m1 <- lapply(m1, hs.attr.slice, V)
    jg <- regmatches(string, m1, invert = 1)
    switch(ot, l = jg, m = do.call(rbind, jg), dt = as.data.table(do.call(rbind, jg)))
}
hs.str_2num <- function(x, as.num = 1) {
    jg <- rep(hs.na(""), length(x))
    i <- hs.grep(x, "%$")
    if (length(i)) 
        jg[i] <- as.numeric(hs.sub(x[i], "%"))/100
    i <- hs.grep(x, ",")
    if (length(i)) 
        x[i] <- hs.gsub(x[i], ",")
    i <- which(is.na(jg))
    if (length(i)) 
        jg[i] <- x[i]
    if (as.num) 
        as.numeric(jg)
    else jg
}
hs.dna.reversComplement <- function(S) {
    paste(rev(sapply(unlist(strsplit(S, "")), switch, a = "t", t = "a", c = "g", g = "c", A = "T", T = "A", C = "G", G = "C", wk.browser("34.09.16.165416 @ hs.dna.reversComplement"))), collapse = "")
}
hs.fpGood <- function(fp, q = 1, type = "sh") {
    R <- hs.normalizePath(fp)
    i <- hs.grep(fp, "/$")
    if (length(i)) 
        R[i] <- paste0(R[i], "/")
    if (q) 
        R <- shQuote(R, type)
    R
}
hs.fp.make_read_only <- function(wj) {
    Mode <- as.character(file.info(wj)[, "mode"])
    i1 <- hs.grep(Mode, "[2367]")
    if (length(i1)) {
        wj <- wj[i1]
        Mode <- Mode[i1]
        p1 <- hs.gre(Mode, ".")[[1]]
        Mode2 <- sapply(hs.gre(Mode, "."), function(p1) {
            sapply(p1, . %=>% {
                intToBits(x) %=>% head(3) %=>% as.numeric %=>% {
                  x[2] <- 0
                  x
                } %=>% rev %=>% paste(collapse = "") %=>% strtoi(2)
            }) %=>% as.character %=>% paste(collapse = "")
        })
        base::Sys.chmod(wj, Mode2)
    }
    wj
}
hs.fp.make_personal <- function(wj) {
    Mode <- paste(file.info(wj)[, "mode"])
    i1 <- hs.grep(Mode, ".[^0]")
    if (length(i1)) 
        local({
            wj <- wj[i1]
            Mode <- Mode[i1]
            p1 <- hs.gre(Mode, ".")[[1]]
            Mode2 <- hs.sub(Mode, "(?<=.)..", "00")
            base::Sys.chmod(wj, Mode2)
        })
    wj
}
hs.wj.parent <- function(wj) basename(dirname(wj))
hs.fp <- function(..., e = FALSE, robust = FALSE, n = FALSE) {
    a <- file.path(...)
    if (robust) {
        if (length(a) > 1) 
            hs.browser("2023.08.04.190335", debug = 0)
        p1 <- "(?i)\\.gz$"
        a <- if (hs.grepl(a, p1)) 
            c(hs.sub(a, p1), a)
        else c(a, paste0(a, ".gz"))
        a <- a[file.exists(a)]
        if (length(a) > 1) 
            a <- a[1]
    }
    else if (e) {
        a <- a[file.exists(a)]
    }
    if (n) 
        names(a) <- hs.wjm(a, dn = 0, ext = 0)
    a
}
hs.fp1 <- function(...) {
    hs.pastec(..., collapse = .Platform$file.sep)
}
hs.wj <- ieiwk.file.path <- function(..., fsep = .Platform[["file.sep"]], recursive = 1, type = "file", sys = 0, expand = 0, normalize = FALSE, v.only = 0) {
    a <- file.path(...)
    if (v.only) 
        return(a)
    switch(type, file = {
        sapply(dirname(a), function(x) {
            if (!file.exists(x)) dir.create(x, recursive = TRUE, showWarnings = FALSE)
        })
    }, wjj = , dir = {
        if (!file.exists(a)) dir.create(a, recursive = recursive)
    })
    r <- {
        if (sys) {
            r <- gsub("\\\\ ", " ", a)
            gsub(" ", "\\\\ ", a)
        }
        else a
    }
    if (expand) 
        r <- path.expand(r)
    if (normalize) 
        r <- ieiwk.fp.norm(r)
    r
}
MyMakePath <- ieiwk.file.path
hs.wjj <- wk.wjj <- wk.dir.path <- ieiwk.dir.path <- function(..., fsep = .Platform$file.sep, recursive = TRUE, type = "dir", sys = FALSE, expand = FALSE, normalize = FALSE, v.only = FALSE, go = FALSE) {
    wjj <- hs.wj(..., fsep = fsep, recursive = recursive, type = type, sys = sys, expand = expand, normalize = normalize, v.only = v.only)
    if (go) 
        setwd(wjj)
    wjj
}
hs.wjj.clean <- wk.wjj.clean <- function(...) {
    wjj <- hs.wjj(...)
    unlink(hs.fp(wjj, "*"), recursive = 1)
    wjj
}
hs.wjlj <- wk.wjlj <- function(wj1, path.expand = 1, quote = 1) {
    if (path.expand) 
        wj1 <- path.expand(wj1)
    if (quote) 
        wj1 <- .q(wj1)
    wj1
}
hs.with_wd_temp <- function(wd = hs.fp("/dev/shm", .muid4machine()), exp1) {
    wjj.37_11_28_135147 <- getwd()
    setwd(hs.wjj(wd))
    on.exit(setwd(wjj.37_11_28_135147))
    exp1
    return(wd)
}
hs.fp2name <- function(fp, extN = 1 + hs.grepl(fp, "(?i)\\.gz$")) {
    hs.wjm(fp, dn = 0, ext = 0, extN = 1 + hs.grepl(fp, "(?i)\\.gz$"))
}
hs.fp_add_name <- function(fp, ...) {
    hs.c(sapply(fp, hs.fp2name, ...), fp)
}
.bash <- function(wd = ".") {
    wd <- wd[1]
    if (!dir.exists(wd)) 
        wd <- dirname(wd)
    hs.with_wd(wd, {
        .wk("bash")
    })
}
.file <- function(f1) {
    sapply(f1, function(f1) {
        r <- .sys.do("file", .sys.fpGood(f1), intern = 1)
        sep.n <- sum(nchar(unlist(gre(":", f1))) == 1)
        for (i in seq(sep.n + 1)) {
            r <- sub("[^:]*:", "", r)
        }
        trim(r)
    })
}
symlink.target <- function(wj1) {
    Sys.readlink(wj1)
}
symlink.fix <- function(wj1) {
    wj1 <- wj1[!file.exists(wj1)]
    wj2 <- symlink.target(wj1) %=>% hs.wj.portable(static = 1)
    hs.wj.rm(wj1)
    file.symlink(wj2, wj1)
}
wk.muid.sep.pattern <- "[\\.\\-_]"
wk.muid.pattern <- {
    paste0("\\b\\d+", wk.muid.sep.pattern, "\\d{2}", wk.muid.sep.pattern, "\\d{2}", wk.muid.sep.pattern, "\\d{6}")
}
uid.hs <- function() {
    uid1 <- .wk("xsel --clipboard --output", intern = 1)
    uid1 <- hs.re(uid1, wk.muid.pattern)
    l1 <- .wk("ag -G", .q("\\.txt$"), .q(paste0("{[^{]*", uid1, "}")), hs.home.expand("~/xt"), intern = 1)
    if (hs.len1(l1)) {
        wj.line <- strsplit(l1, ":\\d+:")[[1]]
        a <- hs.re(wj.line[2], paste0("{[^{]*", uid1, "}"))
        jg <- hs.sub(a, "{") %=>% hs.sub("}$")
        soi <- hs.re(wj.line[2], paste0("{\\d+}{[^{]*", uid1, "}"))
        if (length(soi)) {
            up_level <- as.integer(hs.re(soi, "\\d+"))
            up_str <- paste(tail(strsplit(hs.sub(wj.line[1], "\\.txt$"), "/")[[1]], up_level), collapse = "/")
            jg <- hs.fp(up_str, jg)
        }
        jg <- paste0("hs.hsgly( ", .q(jg), ")")
        message(jg)
        cat(jg, file = "| xsel -ib")
    }
}
.mfp <- .muid.file.path <- function(.muid, dataDir = hs.home.expand(switch(data.type.time, handy = hs.home.expand("~/wk/gz/handy"), long = hs.home.expand("~/wk/db/long"))), fp.type = "file", split = wk.muid.sep.pattern, uid.pos = c("head", "tail"), uid.pat = "\\d+.\\d+.\\d+.\\d+$", normalize = FALSE, v.only = FALSE, include.uid = TRUE, smart = TRUE, act = c("compose", "find"), find.exclude.pat = "~$", data.type.time = dtt, dtt = "handy", fp.only = FALSE) {
    lvl <- 3
    a <- as.list(strsplit(.muid, split)[[1]])
    uid <- hs.gre(.muid, wk.muid.pattern)[[1]]
    if (!hs.len1(uid)) 
        hs.browser("36_10_20_121717")
    a <- strsplit(uid, wk.muid.sep.pattern)[[1]]
    p <- do.call("file.path", as.list(c(dataDir, head(a, lvl))))
    if (fp.only) 
        return(file.path(p, .muid))
    switch(match.arg(act), compose = {
        fp <- file.path(p, .muid)
        if (hs.all(fp.type == "file", !include.uid)) {
            flag <- 1
            if (smart) if (nchar(ieiwk.filename(sub(uid.pat, "", .muid))) == 0) flag <- 0
            if (flag) fp <- hs.sub(fp, wk.muid.pattern)
        }
    }, find = {
        fp <- hs.wjj.ls(p, hs.re(.muid, wk.muid.pattern))
        for (p1 in find.exclude.pat) {
            fp <- grep(p1, fp, inv = 1, v = 1)
        }
    })
    fp
}
.mfp.swxx <- function(..., dataDir = hs.home.expand("~/swxx/xm/db")) {
    .mfp(..., dataDir = dataDir)
}
.mfp.org <- function(..., dataDir = hs.home.expand("~/org/db")) {
    .mfp(..., dataDir = dataDir)
}
wj.head.hs <- function(wj) {
    names(fread(cmd = .sys.cat.skip(wj, N = 1)))
}
hs.wj.portable <- local({
    hs.wj.abs_pattern.rel <- function(wj) {
        c(paste0("^", normalizePath(hs.home.expand(wj))), wj)
    }
    db <- matrix(hs.wj.abs_pattern.rel(c("~/swxx", "~/in2/swxx", "~/org", "~/wk", "~")), ncol = 2)
    function(wj, static = 0) {
        wj_in <- wj
        wj %<=>% {
            unique(x) %=>% hs.c(x, x)
        }
        if (static) {
            wj <- hs.sub(wj, "^/work/wengkailab/swxx", "~/swxx")
        }
        else {
            i1 <- hs.grep(wj, "^~", inv = 1)
            wj[i1] %<=>% normalizePath
            for (db.ci in seq.int(nrow(db))) {
                if (!length(i1)) 
                  break
                db1 <- db[db.ci, ]
                i2 <- regexpr(db1[1], wj[i1], perl = TRUE)
                i2.1 <- which(i2 > 0)
                if (length(i2.1)) {
                  regmatches(wj[i1], i2) <- db1[2]
                  i1 %<=>% {
                    x %not% x[i2.1]
                  }
                }
            }
        }
        if (length(wj_in) > length(wj)) {
            wj[wj_in]
        }
        else wj
    }
})
hs.basename <- function(wj, time = 1, sep2 = "/", skip = {
}, keep = {
}) {
    if (time == 1) {
        basename(wj)
    }
    else {
        if (length(skip)) {
            skip <- skip[skip <= time]
            keep <- keep[keep <= time]
        }
        sapply(strsplit(wj, "/"), function(L) {
            a1 <- tail(L, time)
            if (length(keep)) {
                keey <- keep %not% skip
                a1 <- a1[keep]
            }
            else if (length(skip)) {
                a1 <- a1[-skip]
            }
            paste(a1, collapse = sep2)
        })
    }
}
hs.filename <- ieiwk.filename <- function(fp, ext.pattern = hs.pastec0("\\.", if (..) "+", "[^.]+$"), ignore.case = TRUE, dn = 1, dn.depth, zjjg, zjjg_append = "zjjg", zjjg_rename = if (!missing(zjjg)) !zjjg, append = if (!missing(zjjg)) {
}, append_sep = "_", insert, insert_sep = ".", zjjg_pattern = sprintf("%s%s", append_sep, zjjg_append), extN = if (length(zjjg_pattern) && length(grep(zjjg_pattern, fp[1]))) {
    length(strsplit(re(sprintf("(?<=%s\\.).+$", zjjg_pattern), fp[1]), "\\.")[[1]])
} else if (length(a <- hs.re(fp, "(?i)(\\.([tc]sv|txt|tar|fa|r|py))?\\.gz$"))) {
    sapply(hs.gre(a, "\\."), length)
} else 1, pop, pop_sep = append_sep, add = {
}, add_sep = ".", ext = {
    !missing(add)
} || {
    !missing(zjjg)
}, .. = 1) {
    fp.original <- fp
    fp.dn <- hs.case(inherits(dn, "character"), {
    }, dn > 0, dirname(fp), dn < 0, hs.basename(dirname(fp), abs(dn)))
    fp.ext <- rep("", length(fp))
    fp <- basename(fp)
    for (fp_i1 in seq_along(fp)) {
        extN1 <- hs.i(extN, fp_i1)
        for (i in seq(extN1)) {
            ext.v <- hs.re(fp[fp_i1], ext.pattern)
            if (!length(ext.v)) 
                break
            if (!length(ext.v)) 
                hs.browser("36.07.16.160155")
            fp.ext[fp_i1] %<=>% paste0(ext.v, x)
            fp[fp_i1] %<=>% sub(ext.v, "", x, fixed = TRUE)
        }
    }
    if (!missing(zjjg)) {
        if (zjjg) {
            append <- paste(c(append, zjjg_append), collapse = append_sep)
        }
        else pop <- zjjg_append
    }
    if (length(append)) 
        fp %<=>% paste(x, append, sep = append_sep)
    if (!missing(pop)) 
        fp %<=>% sub(sprintf("%s%s$", pop_sep, pop), "", x)
    if (is.logical(ext)) {
        if (ext) 
            fp %<=>% paste0(x, fp.ext)
    }
    else fp %<=>% paste(x, ext, sep = ".")
    if (is.character(dn)) {
        fp %<=>% file.path(dn, x)
    }
    else {
        if (dn && {
            !all(fp.dn %in% c("", "."))
        }) 
            fp <- file.path(fp.dn, fp)
    }
    if (length(add)) 
        fp %<=>% paste(x, add, sep = add_sep)
    if (length(zjjg_rename) && {
        !zjjg
    } && zjjg_rename) 
        file.rename(fp.original, fp)
    fp
}
hs.wjm <- function(..., append = {
}, append_sep = if (length(append)) ifelse(hs.grepl(append, "^\\\\{0,1}[_\\+\\-\\.,]"), "", "_"), add = {
}, add_sep = if (length(add)) ifelse(hs.grepl(add, "^[_\\-\\.]"), "", "."), ext = 1) {
    args <- list(...)
    args["append"] <- list(append)
    args["append_sep"] <- list(append_sep)
    args["add"] <- list(add)
    args["add_sep"] <- list(add_sep)
    if (!is.character(ext)) 
        ext <- as.logical(ext)
    args[["ext"]] <- ext
    jg <- do.call(hs.filename, args)
    hs.wjm.home(jg)
}
hs.fp.ext <- function(fp, i = 1) {
    fp <- basename(fp)
    while (i > 1) {
        fp <- hs.sub(fp, "\\.[^\\.]+$")
        i <- i - 1
    }
    hs.re(fp, "(?<=\\.)[^\\.]+$", no_match = "")
}
hs.wjj.ls <- hs.listWj <- function(wjj1 = ".", ms = {
}, R = 0, F = 1, A = 0, wj = if (all) TRUE else !wjj, wjj = if (all) TRUE else FALSE, all = FALSE, no.. = TRUE, ic = FALSE, ref_wj = {
}, time = ">", nm = 0) {
    if (wj) {
        r1 <- list.files(path = wjj1, pattern = ms, recursive = R, all.files = A, include.dirs = wjj, no.. = no.., ignore.case = ic, full.names = FALSE)
        if (!wjj) 
            r1 <- r1[!dir.exists(file.path(wjj1, r1))]
    }
    else {
        r1 <- list.dirs(wjj1, full.names = 0, recursive = R)
        r1 <- r1[dir.exists(hs.fp(wjj1, r1))]
        if (length(ms)) 
            r1 <- hs.grepV(r1, ms)
    }
    i1 <- hs.grepl(basename(r1), "^~")
    if (any(i1)) {
        r1 <- r1[!i1]
    }
    if (F) 
        r1 <- hs.fp(wjj1, r1)
    if (length(ref_wj)) {
        r1 <- r1[do.call(time, list(file.mtime(r1), file.mtime(ref_wj)))]
    }
    if (nm) {
        names(r1) <- basename(r1) %=>% {
            p1 <- "[^\\._]+[\\._]"
            repeat {
                f1 <- hs.re(x, p1, no_match = "")
                if ((uniqueN(f1) == 1) && (f1[1] != "")) {
                  x <- hs.sub(x, p1)
                }
                else {
                  break
                }
            }
            p1 <- "[\\._][^\\._]+$"
            repeat {
                f1 <- hs.re(x, p1, no_match = "")
                if ((uniqueN(f1) == 1) && (f1[1] != "")) {
                  x <- hs.sub(x, p1)
                }
                else {
                  break
                }
            }
            x
        }
    }
    r1
}
wk.find.wj <- function(wjj1 = ".", ms = "", ms.bn = "", wj = 1, wjj = 0, wj2exclude = {
}, wjm2exclude.ms = "\"*~\"", follow.symlink = 0, recursive = 0, maxdepth = 0, find.extra = list(), context = 3, less = 1, wj.tmp = hs.home.expand("~/.tmp.34.04.08.150846")) {
    find.par <- c(if (follow.symlink) "-L", wjj1 %=>% path.expand %=>% wk.re.replace("[/]*$", "/"), if (maxdepth) c("-maxdepth", maxdepth), if (wj) "-type f")
    .wk("find", find.par, "!", "-iname", wjm2exclude.ms, "|", .sys.grep(ms), "--color", ">", wj.tmp)
    .ss("--", file = wj.tmp, append = 1)
    .wk("find", find.par, "!", "-iname", wjm2exclude.ms, "-print0", "|", "xargs", "-0", .sys.grep(ms), "--color=never", "-n", paste0("-", context), "-T", ">>", wj.tmp)
    if (less) {
        .less(wj.tmp)
    }
    else {
        .wk("cat", wj.tmp)
    }
    if (0) {
        if (recursive) {
            folders <- .wk("find", find.par, "-type d", intern = 1)[-1] %not% wj2exclude
            r <- {
            }
            if (length(folders)) {
                if (wj) 
                  r <- .wk("find", find.par, "-type f", intern = 1) %not% wj2exclude
                r <- c(r, unlist(lapply(folders, wk.find.wj, wj2exclude = wj2exclude, follow.symlink = follow.symlink, recursive = recursive, maxdepth = maxdepth, wj.tmp = wj.tmp)))
            }
            else {
                if (wj) 
                  r <- c(r, .wk("find", find.par, "-type f", intern = 1))
                if (wjj) 
                  r <- c(r, wjj)
            }
            if (length(wj.tmp)) 
                .ss(folders, file = wj.tmp)
            else r
        }
        else {
            r <- {
                if (wj & wjj) {
                  .wk("find", find.par, intern = 1)[-1]
                  .wk("find", find.par, intern = 1)
                }
                else if (wj) {
                  .wk("find", find.par, "-type f", intern = 1)
                }
                else .wk("find", find.par, "-type d", intern = 1)[-1]
            }
            r %not% wj2exclude
        }
    }
}
wk.ag.wj <- function(ms = "", wjj1 = hs.home.expand("~/org"), wjm2exclude.ms = "\"*~\"", follow.symlink = 0, recursive = 0, maxdepth = 0, find.extra = list(), context = 3, less = 1, wj.tmp = hs.home.expand("~/.tmp.34.04.14.104524"), C = 3) {
    wjj1 %=>% wk.re.replace("/*$", "/") %=>% {
        .wk("find", x, "!", "-iname", wjm2exclude.ms, "-print", "| ag", .q(ms), "--pager \"less -R\"")
        .wk("ag", "-i", .q(ms), x, "--pager", .q("less -R"), "-C", "C")
    }
}
hs.wj.real <- function(wj) base::normalizePath(wj)
hs.uid.wj <- function(uid) {
    .wk("ag", "-g", uid, hs.home.expand("~/org"), intern = 1)
}
hs.wj.copy <- function(from, to, ..., copy.date = TRUE, recursive = TRUE) {
    file.copy(from = from, to = to, ..., copy.date = copy.date, recursive = recursive)
}
hs.wj.move <- function(from, to = ".", ..., copy.date = TRUE, recursive = TRUE, ow = 0) {
    if ((!length(from))) 
        return()
    if (length(to) %not.in% c(1, length(from))) 
        hs.browser("39.04.05.161937", debug = 0)
    if (length(to) > 1) 
        hs.browser("39.04.08.110341", debug = 0)
    if ((length(to) == 1) && (dir.exists(to) || (length(from) > 1))) {
        to %<=>% hs.fp(basename(from))
    }
    if (any(file.exists(to))) 
        hs.browser("39.04.09.202117", debug = 0)
    for (to1 in unique(dirname(to))) hs.wjj(to1)
    file.rename(from = from, to = to)
    i <- which(file.exists(from))
    if (length(i)) {
        hs.wj.copy(from = from[i], to = to[i], ..., copy.date = copy.date, recursive = recursive)
        hs.wj.rm(from[i], 1)
    }
    to
}
hs.dirname <- local({
    home <- normalizePath(hs.home.expand("~"))
    function(fp, time = 1, tilde.expansion = 0) {
        while (time > 0) {
            fp <- dirname(fp)
            time <- time - 1
        }
        return(hs.wjm.home(fp))
        if (!tilde.expansion) {
            i1 <- which(substr(fp, 1, 1) == "~")
            if (length(i1)) {
                i2 <- which(hs.grepl(R[i1], paste0("^", home)))
                if (length(i2)) 
                  R[i1[i2]] <- hs.sub(R[i1[i2]], paste0("^", home, "(?=/)"), "~")
            }
        }
        R
    }
})
hs.file <- function(wj, open = "r", ...) {
    gz <- hs.grepl(wj, "(?i)\\.[gx]z$")
    con1 <- (if (gz) 
        gzfile
    else file)(wj, open = open, ...)
}
hs.list.write <- function(L1, wj2) {
    con1 <- hs.file(wj2, "w")
    on.exit(close(con1), add = TRUE)
    if (!is.null(names(L1))) 
        hs.cat("#named=1", file = con1)
    for (i1 in seq_along(L1)) hs.cat(names(L1)[i1], L1[[i1]], sep = "\t", append = 1, file = con1)
    wj2
}
hs.list.read <- function(wj2) {
    nr <- readLines(wj2)
    if (nr[1] == "#named=1") {
        L1 <- strsplit(nr[-1], "\t")
        names(L1) <- sapply(L1, "[", 1)
        lapply(L1, "[", -1)
    }
    else strsplit(nr, "\t")
}
hs.wj.open <- function(wj, ..., open = "r", skip.comments = 1, comment = "#", cn = 1) {
    con1 <- gzfile(wj, open = open, ...)
    if (skip.comments) {
        skip <- hs.lines_comment(wj, comment = comment, opt = "i")
        if (skip && cn) 
            skip <- skip - 1
        if (skip) 
            readLines(con1, skip)
    }
    con1
}
hs.wj.temp <- local({
    wjj <- {
        shm <- file.exists("/dev/shm")
        wjj <- if (shm) {
            hs.wjj(tempfile(tmpdir = "/dev/shm"))
        }
        else tempdir()
    }
    function(ext = "", ...) {
        if (nchar(ext) && (!hs.grepl(ext, "^\\."))) 
            ext <- paste0(".", ext)
        hs.wj(tempfile(tmpdir = wjj, fileext = ext, ...))
    }
})
hs.wj.temp <- function(pattern = "file", fileext = "", tmpdir = hs.wjj(if (file.exists("/dev/shm")) "/dev/shm" else hs.wjj(hs.home.expand("~/tmp")), "36_10_16_121431"), tmpdir.clean = 0, old = 10) {
    if (tmpdir.clean) {
        a <- list.files(tmpdir, full = 1)
        c1 <- 0
        sapply(a, function(a1) {
            if (difftime(Sys.Date(), file.info(a1)[["mtime"]], u = "d") > old) {
                file.remove(a1)
                c1 <<- c1 + 1
            }
        })
        hs.cat(c1, "old temporary files have just been deleted.", sep = " ")
        return()
    }
    pattern <- sprintf("..%s..", pattern)
    tempfile(pattern, tmpdir, fileext)
}
hs.wj.bf <- function(wj1, wj2 = hs.wjm(wj1, add = post), post = "bf", inv = 0, mv = 1, ow = 0, ...) {
    if (inv) {
        a1 <- wj1
        wj1 <- wj2
        wj2 <- a1
    }
    if (ow && any(file.exists(wj2))) {
        hs.wj.rm(wj2[file.exists(wj2)], 1)
    }
    if (mv) {
        hs.wj.move(wj1, wj2)
    }
    else {
        hs.wj.copy(wj1, wj2, ...)
    }
    wj2
}
hs.wj.exists <- function(wj, dir = 0, wj_is = size, size = 0) {
    file.exists(wj) & {
        if (wj_is) 
            !dir.exists(wj)
        else 1
    } & {
        if (dir) 
            dir.exists(wj)
        else 1
    } & {
        if (size) 
            file.size(wj) >= size
        else 1
    }
}
hs.Sys.setFileTime <- function(wj, ...) Sys.setFileTime(normalizePath(wj), ...)
hs.wj.update <- function(wj, message = 1) {
    td <- difftime(Sys.time(), file.mtime(wj), units = "days")
    if (message) {
    }
}
hs.wjm.friendly <- function(wj, rep1 = "_") {
    wj <- hs.gsub(wj, "(?<=[,\\.]) ")
    wj <- hs.gsub(wj, " ", rep1)
    wj <- hs.gsub(wj, ":", "-")
    wj
}
yb.hs.gn <- function(gn, sep1 = "-", smart = 1) {
    if (smart) {
        if (!anyDuplicated(gn)) 
            return(gn)
    }
    a1 <- tapply(seq_along(gn), gn, . %=>% {
        hs.c(seq_along(x), x)
    }, simplify = 0)
    paste0(gn, sep1, names(sort(unlist(unname(a1)))))
}
hs.readLines4gz <- wk.readLines4gz <- function(wj, n = -1L, ...) {
    con <- gzfile(wj, "r", "UTF-8")
    hs.readLines(con, n = n, ...)
}
hs.writeLines <- function(text, con, closeCon = TRUE) {
    writeLines(text, con)
    if (closeCon && ("connection" %in% class(con))) 
        on.exit(close(con), add = TRUE)
}
hs.writeLines2gz <- wk.writeLines2gz <- function(text, wj) {
    con <- gzfile(wj, "w", "UTF-8")
    hs.writeLines(text, con)
}
hs.url_good <- function(url) {
    curl::curl_unescape(url)
}
.sys.perl.print <- function(e1, sep = "\t", a = 1) {
    c("perl", if (a) c(sprintf("-F\"\\%s\"", switch(sep, `\t` = "t", sep)), "-a"), "-lne", .q(e1))
}
.sys.skip_line <- function(n = 0) {
    if (n > 0) {
        .sys("| perl -e", .q(paste0("$i = ", formatSafe(ceiling(n)), ";\nwhile( $i) {\n <>;\n $i--;\n}\nwhile( <>) {\n  print;\n}")))
    }
    else ""
}
wk.readLines.using_perl <- function(wj, lineN, intern = 1, ...) {
    lineN.wj <- paste0(hs.wj.temp(), ".tsv.gz")
    on.exit(hs.wj.rm(lineN.wj))
    if (hs.len1(dim(lineN), 2)) {
        wk.dt.fwrite(lineN, lineN.wj, cn = 0)
    }
    else cat(lineN, sep = "\n", file = lineN.wj)
    .perl.e("%line_wanted = ()", paste("open( DATA, \"", if (hs.grepl(lineN.wj, "\\.gz$")) 
        "pigz -cd"
    else "cat", .sys.fpGood(lineN.wj), "|\")"), "$toEnd = 0", "while( <DATA>) { chomp; @v = split( \"\t\"); if( @v == 1) { $v[ 1] = $v[ 0]}; if( $v[ 1] < 0) { $toEnd = $v[ 0]}; for( $i = $v[ 0]; $i <= $v[ 1]; $i += 1) { $line_wanted{ $i} = 1}}", "close( DATA)", paste("open( DATA, \"", if (hs.grepl(wj, "\\.gz$")) 
        "pigz -cd"
    else "cat", .sys.fpGood(wj), "|\")"), "while( <DATA>) { if( exists( $line_wanted{ $.}) || ( $toEnd && ( $. >= $toEnd))) { print}}", "close( DATA)", intern = intern, ...)
}
.sys.uniq <- function() {
    "uniq | sort | uniq"
}
.unzip <- function(f1, l = 1, l.head = 9, d = dirname(f1), d.use.self.name = 1, t = 0) {
    if (t) 
        return(.wk("unzip -t", hs.fpGood(f1)))
    if (hs.grepl(f1, "(?i)\\.((tar(\\.(gz|xz|bz2))?)|tgz)$")) {
        tar.p <- local({
            a <- c("t", "x")
            b <- paste0("-", a, switch(hs.file.extension(f1), tgz = , gz = "z", xz = "J", bz2 = "j"), "vf")
            names(b) <- a
            b
        })
        if (l) {
            a <- .wk("tar", tar.p["t"], hs.fpGood(f1), "|", .sys.head(1000), intern = 1)
            if (length(a) > l.head) 
                a <- c(head(a, 4), "...", tail(a, 4))
            hs.catn(a)
        }
        else {
            extN <- hs.re(f1, "(?i)\\.((tar\\.(gz|xz|bz2))|tgz)$") %=>% hs.gre("\\.") %=>% unlist %=>% length
            wjj.out <- hs.wjm(f1, dn = d, ext = 0, extN = extN)
            if (!file.exists(wjj.out)) {
                wjj.tmp <- hs.wj.temp(tmpdir = dirname(f1))
                .wk("tar", "-k", tar.p["x"], hs.fpGood(f1), "-C", hs.fpGood(hs.wjj(wjj.tmp)))
                wjj <- hs.wjj.ls(wjj.tmp, wjj = 1, F = 0)
                if (wjj %=>% {
                  hs.len1(x) && (x == basename(f1))
                }) {
                  hs.wj.move(hs.fp(wjj.tmp, wjj), d)
                  unlink(wjj.tmp, 1)
                }
                else hs.wj.move(wjj.tmp, wjj.out)
            }
            return(wjj.out)
        }
    }
    if (hs.grepl(f1, "(?i)\\.tar\\.xz$")) {
        if (l) {
            .wk("xzcat", .sys.fpGood(f1), "| tar", "-tf -", "| head")
        }
        else {
            if (d.use.self.name) 
                d <- wk.wjj(hs.wjm(f1, dn = d, ext = 0, extN = 2))
            .wk("xzcat", .sys.fpGood(f1), "| tar", "-xf -", "-C", .sys.fpGood(d))
        }
        return(d)
    }
    if (hs.grepl(f1, "\\.zip$")) {
        bin <- "unzip"
        if (!length(d.use.self.name)) {
            if (0) {
                fc.0 <- .sys.do("unzip", "-l", .sys.fpGood(f1), intern = 1)
                fc <- {
                  i <- grep("^-", fc.0)
                  substring(fc.0[seq(i[1] + 1, i[2] - 1)], regexpr("Name", fc.0[2]))
                }
            }
            fc.0 <- .wk(bin, "-l", wk.wjlj(f1), "| gawk", .q("{ print $4}"), intern = 1)
            fc <- {
                i <- grep("^-", fc.0)
                tail(fc.0, -i) %rm% ""
            }
            d.use.self.name <- as.integer(substring(fc[1], nchar(fc[1])) != "/" || length(table(sapply(fc, ieiwk.dirname.1))) > 1)
        }
        if (l) {
            .wk("unzip", "-l", .sys.fpGood(f1), "|", .sys.head(l.head))
        }
        else {
            if (d.use.self.name) 
                d <- hs.wj(hs.wjm(f1, ext = 0, dn = d))
            if (file.exists(d)) {
                message("[target folder already exists]", f1)
                return(f1)
            }
            .wk(bin, "-d", .sys.fpGood(d), .sys.fpGood(f1))
        }
        return()
    }
    if (wk.re(f1, "\\.gz$", "detect")) {
        bin <- "pigz"
        if (l) {
            .wk("gzip", "-l", wk.wjlj(f1), "|", .sys.head(l.head))
        }
        else {
            .wk(bin, "-dc", wk.wjlj(f1), ">", wk.wjlj(hs.wjm(f1, ext = 0, dn = d)))
        }
        return(ieiwk.filename(f1))
    }
    if (hs.grepl(f1, "\\.7z$")) {
        bin <- "7z"
        if (l) {
            .wk(bin, "l", f1)
        }
        else {
            if (!length(d.use.self.name)) {
                browser("33.06.15.224151")
            }
            if (d.use.self.name) 
                d <- dirname(f1)
            .wk(bin, "x", "-aos -y", paste0("-o", hs.fpGood(d)), hs.fpGood(f1))
            return(hs.wjm(f1, ext = 0))
        }
    }
    if (wk.re(f1, "\\.rar$", "detect")) {
        bin <- "unrar"
        if (l) {
            .wk(bin, "v", f1, "|", .sys.head())
        }
        else {
            wjj2 <- hs.wjm(f1, ext = 0)
            .wk(bin, "x -u", .sys.fpGood(f1), .sys.fpGood(hs.wjj(wjj2)))
            return(f1)
        }
    }
}
wk.save_to_zip <- function(x1, wj2 = "a.xls", ow = 0, wjj2_temp = hs.wj.temp(tmpdir = if (dir.exists("/dev/shm")) "/dev/shm" else dirname(wj2)), hs1 = function(x1, wj2) wk.dt.fwrite(x1, wj2)) {
    wj_temp <- hs.wjm(wj2, dn = wjj2_temp)
    hs.update(wj_temp, hs.wjm(x, ext = 0), hs.grepl(x, "(?i)\\.zip$"))
    local({
        x <- wj_temp %=>% c(x, hs.wjm(x, dn = dirname(wj2), add = "zip"))
        if (ow) 
            hs.wj.rm(x)
        if (file.exists(x[1])) {
            stop("[39.03.15.044857|中间文件已存在]", .q(x[1]))
        }
        else if (file.exists(x[2])) 
            stop("[39_03_15_045856|文件已存在]", .q(x[2]))
    })
    on.exit(hs.wj.rm(dirname(wj_temp), 1), add = TRUE)
    hs1(x1, wj_temp)
    hs.with_wd(wj_temp, .wk("zip", hs.fpGood(wj2), .q(basename(wj_temp))))
    return(wj2)
    {
        con1 <- pipe(.sys("zip", .sys.fpGood(wj2), "-"), open = "w")
        sink(con1)
        wk.dt.fwrite(x1, "")
        sink()
        close(con1)
        if (!hs.grepl(wj2, "(?i)\\.zip$")) 
            wj2 %<=>% paste0(".zip")
        return(wj2)
    }
}
wk.unrar <- function(wj1, wjj2 = hs.wjm(wj1, ext = 0)) {
    hs.with_wd(wjj2, {
        .wk("unrar x -u -r", .sys.fpGood(wj1))
    })
    hs.wj.portable(wjj2)
}
wk.compressed_file_extension_pattern <- "(?i)\\.((tar\\.|[bt]?)gz|zip|rar|xz|bz2)$"
hs.wj.is_compressed <- function(wj1) length(wj1) && hs.grepl(wj1, wk.compressed_file_extension_pattern)
wk.untar <- function(wj1 = {
}, wj2 = {
}, ..., wj2.gz = !hs.wj.is_compressed(wj2), wjj2 = hs.wjm(wj1, ext = 0, extN = gz + 1), gz = hs.grepl(wj1, "(?i)\\.t?gz$"), list = 0, Av = list, O = !length(wj1), kn = 1, ow = 0, keep_old = 1, tar = switch(osType(), Linux = "tar", Darwin = "gtar")) {
    cmd1 <- .sys(tar, if (list) 
        "-t"
    else "-x", if (!ow) 
        "--skip-old-files", if (Av) 
        "-v", if (gz) 
        "-z", if (!(list || O)) 
        c("-C", hs.wjj(wjj2) %=>% .sys.fpGood), if (O || (length(wj2) && wj2.gz)) 
        "-O", "-f", .sys.fpGood(wj1), if (length(wj2)) 
        .q(wj2), ..., if (O) {
        if (hs.wj.is_compressed(wj2)) 
            "| gzip -cd"
    }
    else {
        if (length(wj2) && wj2.gz) 
            c("| gzip >", hs.wjm(wj2, dn = wjj2, add = "gz"))
    })
    if (O) {
        cmd1
    }
    else {
        .wk(cmd1)
        normalizePath(wjj2)
    }
}
wk.decompress <- function(wj1, wjj2 = {
}, ...) {
    lapply(wj1, function(wj1) {
        if (dir.exists(wj1)) {
            wj1 <- hs.wjj.ls(wj1, wk.compressed_file_extension_pattern, F = 1)
            wk.decompress(wj1, wjj2, ...)
        }
        else {
            if (!length(wjj2)) 
                wjj2 <- hs.sub(wj1, wk.compressed_file_extension_pattern)
            hs.with_wd(hs.wjj(wjj2), {
                if (hs.grepl(wj1, "(?i)\\.zip$")) {
                  unzip(wj1, setTimes = 1, overwrite = 0, ...)
                }
                else if (hs.grepl(wj1, "(?i)\\.(tar\\.gz|tgz)$")) {
                  if (0) {
                    hs1 <- utils::untar
                    hs1(gzfile(wj1, "rb"), extras = "--skip-old-files")
                  }
                  .wk("tar -xz --skip-old-files -f", .sys.fpGood(wj1))
                }
                else hs.browser("37_12_29_152514")
            })
        }
    })
}
.sys.cat.text <- local({
    ext.cat <- c(gz = "pigz -cd", bgz = "pbgzip -cd", xz = "xz -cd", zip = "unzip -p", rar = "unrar -p")
    function(wj, bash = 0, pipe = 0) {
        if (length(wj)) {
            wj.cmd <- sapply(wj, function(wj1) {
                ext <- hs.re(wj1, "[^\\.]+$")
                wj1 <- .sys.fpGood(wj1)
                if (ext %in% names(ext.cat)) {
                  a <- paste(ext.cat[ext], wj1)
                  if (length(wj) > 1) 
                    a <- paste("<(", a, ")")
                  a
                }
                else {
                  paste("cat", wj1)
                }
            })
            cmd1 <- if (length(wj.cmd) > 1) {
                paste(c("cat", wj.cmd), collapse = " ")
            }
            else wj.cmd
            if (bash) 
                cmd1 <- .sys("bash -c", .q(cmd1))
            if (pipe) 
                cmd1 <- .sys(cmd1, "|")
            cmd1
        }
    }
})
.usu <- "uniq | sort | uniq"
hs.isDir <- function(f, a = FALSE) {
    ifelse(a, length(f) == 1, TRUE) && dir.exists(f)
}
hs.isFile <- function(f, a = FALSE) {
    ifelse(a, length(f) == 1, TRUE) && file.exists(f) && (!dir.exists(f))
}
hs.wj.mtime <- function(f) file.info(f)[["mtime"]]
wk.file.newer <- function(f1, f2, tol = 1) {
    (.sys.mtime(f1) - .sys.mtime(f2)) >= tol
}
wk.read <- function(wj) {
    lapply(wj, function(wj1) {
        hs.case.over(paste0("\\.", c("rd", "xls(x)?", "(tsv|csv|txt)(\\.gz)?"), "$"), function(p1) hs.grepl(wj1, p1), list(hs.load, wk.xls.read, fread), no_match = readLines)(wj1)
    })
}
wk.rsync <- function(wj1, wjj2, bin = wj_rsync, type = if (hs.grepl(wj1, "/$")) "wjj" else wj, del = 0, dry = 0, z = 1, p = 1, ssh = wk.ssh, ssh.side = if (length(ssh)) 2 else 0, mode = "update", recursive = 1, cat.cmd = 0, L = 0, wait = 1, sync.type = "get", fs = "linux", checksum = 0, excluded = {
}, included = {
}, U = 0, home.only = 0, files_from = {
}, size_max = {
}, size_min = {
}, ext = {
}, bf_wjj0 = {
}, show_wjj = 0, additional = {
}) {
    if (home.only) 
        wj1 %<=>% {
            hs.wj.portable(x) %=>% hs.grep("^~", v = 1)
        }
    wj1.wjj.i <- which(dir.exists(wj1))
    if (length(wj1.wjj.i)) 
        switch(type, wj = {
            hs.browser("2023.08.06.223939", debug = 0)
            hs.sub(wj1[wj1.wjj.i], "/*")
        }, wjj = if (!hs.grepl(wj1[wj1.wjj.i], "/$")) wj1[wj1.wjj.i] %<=>% paste0("/"))
    if (!hs.grepl(wjj2, "/$")) 
        wjj2 <- paste0(wjj2, "/")
    node_configured <- local({
        a <- readLines(hs.home.expand("~/.ssh/config"))
        a <- hs.sepSeq(a, sep.p = "^(?i)host")
        sapply(a, . %=>% {
            x[1] %=>% hs.re("(?<=\\s)[^\\s]*")
        })
    })
    if (hs.len1(ssh) && is.character(ssh)) {
        if (ssh %not.in% node_configured) {
            ssh <- ieiwk.ssh.db(ssh)
        }
    }
    ssh.str <- paste(c(if (hs.len1(ssh)) ssh else c(ssh["usr"], "@", ssh["dz"]), ":"), collapse = "")
    if (0) 
        switch(as.character(ssh.side), `0` = {
            hs.wjj(wjj2)
            wjj2 <- .sys.fpGood(wjj2)
        }, `1` = {
            wj1 %<=>% sapply(.sys.fpGood)
            wj1 <- paste0(ssh.str, wj1)
            hs.wjj(wjj2)
            wjj2 <- .sys.fpGood(wjj2)
        }, `2` = {
            wjj2 %<=>% sapply(.sys.fpGood)
            wjj2_kept <- wjj2
            wjj2 <- paste0(ssh.str, wjj2)
            message(wjj2)
        })
    always <- "-rtDPvh --force --progress --inplace --no-implied-dirs"
    always.linux <- "-lgo"
    if (mode == "sync") 
        del <- 1
    if (0) {
        wj1 <- hs.gsub(wj1, "(\\()", "\\\\\\1")
        wj1 <- hs.gsub(wj1, "(\\))", "\\\\\\1")
        wj1 <- hs.gsub(wj1, "(;)", "\\\\\\1")
        wjj2 <- hs.gsub(wjj2, "( )", "\\\\\\1")
        wjj2 <- hs.gsub(wjj2, "(\\()", "\\\\\\1")
        wjj2 <- hs.gsub(wjj2, "(\\))", "\\\\\\1")
        wjj2 <- hs.gsub(wjj2, "(;)", "\\\\\\1")
    }
    if (length(wj1) > 1) {
        hs.browser("38.04.11.012803")
        wj.tmp <- tempfile()
        on.exit(hs.wj.rm(wj.tmp))
        cat(wj1, sep = "\n", file = wj.tmp)
        files_from <- paste0("--files-from=", wj.tmp)
        wj1 <- hs.dirname(wj1[1])
    }
    cmd <- .sys(bin, always, if (U) 
        "-u", if (p) 
        "-p", get(paste("always.", fs, sep = "")), if (checksum) 
        "-c", if (z) 
        c("-z", "--compress-level=6"), if (recursive) 
        "-r"
    else "-d", if (del) 
        "--delete", if (L) 
        "-L", if (length(included)) 
        paste0("--include=", .q(included)), if (length(excluded)) 
        paste0("--exclude=", .q(excluded)), if (length(ssh) > 1) 
        c("-e", .q(paste("ssh -p", " ", ssh["port"], sep = ""))), if (dry) 
        "-n", if (length(files_from)) 
        files_from, if (ssh.side == 2) 
        paste0("--rsync-path=", .q(paste0("mkdir -p ", .sys.fpGood(wjj2, home.expand = 0), "; rsync"))), if (length(size_max)) 
        paste0("--max-size=", size_max), if (length(size_min)) 
        paste0("--min-size=", size_min), bf_wjj0 %=>% {
        if (length(x)) {
            c("-b", paste0("--backup-dir=", hs.switch(ssh.side, 0, , 1, bf_wjj0 %=>% hs.fp(hs.uid.fp(base_nyr = 0)) %=>% normalizePath %=>% hs.sub("/*$", "/") %=>% .q, 2, .sys.fpGood(hs.fp(bf_wjj0, hs.uid.fp(base_nyr = 0)), home.expand = 0, qn = 1))))
        }
    }, additional, if (length(ext)) 
        paste("-f", c(.q("+ */"), tolower(ext) %=>% strsplit("") %=>% sapply(x, . %=>% {
            paste0("[", x, toupper(x), "]") %=>% {
                paste(c("+ *.", x), collapse = "")
            }
        }) %=>% .q, .q("- *"))), hs.switch(ssh.side, 0, , 2, .sys.fpGood(wj1), 1, paste0(ssh.str, .sys.fpGood(wj1, home.expand = 0, qn = switch(osType(), Darwin = 1, 1)))), hs.switch(ssh.side, 0, , 1, .sys.fpGood(hs.wjj(wjj2), home.expand = 0), 2, paste0(ssh.str, .sys.fpGood(wjj2, home.expand = 0, qn = 1))), if (!show_wjj) 
        "| grep -v", .q("/$"))
    if (0) {
        hs.browser("38.07.15.092108")
        message(cmd)
        .wk(cmd, "| less -Ni")
    }
    .wk(cmd, wait = wait)
    if (cat.cmd) 
        cat("cmd is", "\n", "\t", cmd, "\n")
}
wk.rsync.in <- function(wj1 = {
}, wjj2 = c("~/jz002", "/Volumes/bf4wk@mpb", "/media/wk/sda1-usb-WD_Elements_25A1", "/media/wk/sdb1-usb-WD_Elements_25A1", "/Volumes/in2") %=>% x[file.exists(x)] %=>% {
    if (length(x)) 
        x[1]
    else hs.home.expand("~/tmp")
}, ssh = wk.ssh, ssh.side = 1, type = "wj", ...) {
    if (!length(wj1)) {
        if (0) {
            wk.rsync("/", wjj2, ssh = ssh, ssh.side = ssh.side, additional = paste0("--files-from=:", .qq(paste0("<( gzip -cd ", wk.rsync.wj.wj, ")"))), ...)
            .wk("cat", "<( gzip -cd ", wk.rsync.wj.wj, ")")
            wk.rsync.wj.wj %=>% .head
            return()
        }
        wk.rsync.in(wk.rsync.wj.wj, wjj2, ssh = ssh)
        wj.wj <- hs.wjm(wk.rsync.wj.wj, dn = wjj2)
        on.exit(hs.wj.rm(wj.wj), add = TRUE)
        wj1 <- readLines(wj.wj)
    }
    if (0) {
        bn <- 1
        while (anyDuplicated(hs.basename(wj1, bn))) {
            bn <- bn + 1
        }
    }
    a <- if (length(wj1) > 1) {
        hs.hsgly("文件/名/公私.39_05_12_131245")(wj1)[["private"]]
    }
    else list(basename(wj1))
    sapply(seq_along(wj1), function(i1) {
        wj1 <- wj1[i1]
        wjj2 <- c(wjj2, a[[i1]]) %=>% head(-1) %=>% do.call(hs.fp, as.list(x))
        wk.rsync(wj1, wjj2, type = type, ssh = ssh, ssh.side = ssh.side, ...)
        hs.wjm(wj1, dn = wjj2)
    })
}
wk.rsync.in.same_place <- function(wj = {
}, ssh = wk.ssh, recursive = 1, U = 1, wj.wj = wk.rsync.wj.wj, home.only = 1, ...) {
    if (!length(wj)) {
        wk.rsync.in.same_place(wj.wj)
        wj <- readLines(wj.wj)
    }
    for (wj1 in wj) {
        wjj2 <- if (hs.grepl(wj1, "/$")) 
            wj1
        else hs.wjj(substr(wj1, 1, nchar(wj1) - nchar(basename(wj1))))
        wk.rsync.in(wj1, wjj2, ssh = ssh, recursive = recursive, U = U, home.only = home.only, type = ifelse(hs.grepl(wj1, "/$"), "wjj", "wj"), ...)
    }
}
wk.rsync.out <- function(wj1 = {
}, wjj2 = hs.home.expand("~/tmp"), ssh = {
}, ssh.side = 2, recursive = {
}, checksum = 0, U = 0, type = {
}, wj1.type = "auto", ...) {
    if (hs.any(!length(wjj2), !length(ssh))) {
        wj1_tmp <- .Last.value
        if ((!length(wj1)) && file.exists(wj1_tmp)) 
            wj1 <- wj1_tmp
        if (identical(type, "doc")) {
            if (hs.len1(wj1) && dir.exists(wj1)) {
                wj1 <- hs.wjj.ls(wj1, "\\.(pdf|png|xlsx)", R = 1)
                if (any(duplicated(basename(wj1)))) 
                  stop("【36.04.28.122010】有重复的文件名")
            }
            else hs.browser("36.04.17.071257")
        }
        cat(wj1, sep = "\n", file = paste("| gzip >", wk.rsync.wj.wj))
        return(wk.rsync.wj.wj)
    }
    sapply(wj1, function(wj1) {
        if (!file.exists(wj1)) 
            stop("36.04.08.134538 文件\"", wj1, "\"不存在")
        if (wj1.type %in% "auto") 
            wj1.type <- ifelse(hs.grepl(wj1, "/$"), "wjj", "wj")
        hs.with_wd(switch(wj1.type, wjj = wj1, wj = dirname(wj1)), {
            if (!length(recursive)) 
                recursive <- hs.isDir(wj1)
            wk.rsync(switch(wj1.type, wjj = "./", wj = basename(wj1)), wjj2, ssh = ssh, ssh.side = ssh.side, recursive = recursive, checksum = checksum, U = U, type = wj1.type, ...)
        })
    })
}
wk.rsync.out.same_place <- function(wj, ssh = wk.ssh, ssh.side = 2, U = 1, wj.wj = wk.rsync.wj.wj, home.only = 1, ...) {
    if (home.only) 
        wj %<=>% hs.wj.portable
    for (wj1 in wj) {
        wjj2 <- if (hs.grepl(wj1, "/$")) 
            wj1
        else hs.wjj(substr(wj1, 1, nchar(wj1) - nchar(basename(wj1))))
        if (!file.exists(wj1)) 
            stop(sprintf("文件（%s）不存在", wj1))
        wk.rsync(wj1, wjj2, ssh = ssh, ssh.side = ssh.side, recursive = dir.exists(wj1), U = U, home.only = home.only, ...)
    }
}
wk.rsync.wjj.local <- function(wjj1, wjj2, wj.log = .wk("date", "+%Y.%m.%d_%H.%M.%S", intern = 1) %=>% str_c("~/rsync.", x, ".log"), linux = 1, del = 1, dry = 1, method = "", wjj1.type = "wjj", u = 1, others = "") {
    switch(wjj1.type, wjj = {
        if (basename(wjj1) != basename(wjj2)) stop("basenames of wjj1 and wjj2 should be the same")
        wjj1 %<=>% wk.re("[/\\\\]*$", "replace", replacement = "/", all = 1)
    }, wj = {
        wjj1 %<=>% wk.re.replace("[/\\\\]*$", "")
    })
    .wk("rsync", "-rtvvh", "--force", "--inplace", "--progress", if (method == "chksum") 
        "-c", if (linux) 
        "-ogplD", if (del) 
        "--del", if (u) 
        "-u", if (dry) 
        "-n", wjj1 %=>% wk.re.replace("/*$", "/"), wjj2 %=>% wk.wjj, if (nchar(wj.log)) 
        c(">", wj.log %=>% path.expand %=>% .q), others)
    wj.log
}
wk.rsync.wjj.myWay <- function(wjj1, wjj2, wj.log = .wk("date", "+%Y.%m.%d_%H.%M.%S", intern = 1) %=>% str_c("~/rsync.", x, ".log"), linux = 1, del = 1, dry = 1, method = "", wjj1.type = "wjj", u = 1, others = "", way = "out") {
    wjj2.1 <- wk.wjj(wjj2, basename(wjj1))
    wjj.current <- wjj2.1 %=>% wk.wj("current")
    cmd1 <- .sys("rsync", "-rtuvh", "--force", "--inplace", "--progress", "--del", if (method == "chksum") 
        "-c", if (linux) 
        "-ogplD", if (dry) 
        "-n")
    if (.wk("find", wjj2.1 %=>% wk.re.replace("/*$", "/") %=>% path.expand %=>% .q, "| head", intern = 1) %=>% hs.len1) {
        .wk(cmd1, wjj1 %=>% .sys.fpGood %=>% wk.re.replace("/*$", "/"), wjj.current %=>% wk.wjj %=>% .sys.fpGood)
    }
    else {
        wjj.bf <- Sys.time() %=>% format("%Y.%m.%d.%H%M%S") %=>% wk.re.replace("\\d+", as.numeric(format(Sys.time(), "%Y")) - 1984, all = 0) %=>% wk.wjj(wjj2.1, x)
        cmd2 <- str_c("--backup --backup-dir=", wjj.bf %=>% .sys.fpGood)
        {
            wj.ref <- {
                a <- list.files(wjj2.1) %=>% (`%not%`(c("current", basename(wjj.bf))))
                a[file.info(wk.wj("..", a))[, "mtime"] %=>% order %=>% tail(1)] %=>% wk.wj(wjj2.1, x)
            }
            if (length(wj.ref)) {
                1
                {
                  setwd(wjj.current)
                  wj.new <- .wk("find", "-newer", wj.ref, intern = 1) %=>% {
                    wj.new <- x
                    wj.new %=>% wk.wj(wjj1, x) %=>% file.exists %=>% wj.new[!x]
                  }
                  if (length(wj.new)) {
                    wk.browser("34.04.19.220858")
                    for (wj1 in wj.new) {
                      .wk("cp", "-a", wj1 %=>% .sys.fpGood, wj1 %=>% wk.wj(wjj1, x) %=>% dirname %=>% wk.wjj %=>% .sys.fpGood)
                    }
                  }
                }
                {
                  setwd(wjj1)
                  wj.old <- .wk("find", "! -newer", wj.ref, "-print", intern = 1) %=>% {
                    a <- x
                    a %=>% wk.wj(wjj.current, x) %=>% file.exists %=>% {
                      a[!x]
                    }
                  }
                  wj.old %x>% {
                    if (length(x)) {
                    }
                  }
                }
            }
        }
        .wk(cmd1, cmd2, wjj1 %=>% .sys.fpGood %=>% wk.re.replace("/*$", "/"), wjj.current %=>% .sys.fpGood, ">>", wj.log %=>% .sys.fpGood)
        .wk(cmd1, cmd2, wjj.current %=>% .sys.fpGood %=>% wk.re.replace("/*$", "/"), wjj1 %=>% .sys.fpGood, ">>", wj.log %=>% .sys.fpGood)
        .wk("touch", wjj.bf %=>% .sys.fpGood)
        wj.log
    }
}
wk.rsync.sync2way <- function(wjj1, wjj2, type = "wjj", mode = "sync") {
    wk.rsync(wjj1, wjj2, type = type, del = 1)
}
wk.get_open <- local({
    wj_prev <- ""
    function() {
        wj1 <- .wk("xsel --clipboard --output", intern = 1)
        if (hs.len1(wj1)) {
            wk.rsync.in(wj1, "/dev/shm")
            wj_prev <<- hs.fp("/dev/shm", basename(wj1))
        }
        .wk(hs.home.expand("~/org/dm/sh/rj/o"), .sys.fpGood(wj_prev))
    }
})
ieiwk.sync.wjj2do <- function() {
    file.path("/cygdrive", "d", c("wk", "wk.os", "rjk"))
}
ieiwk.sync.wjj.mobile <- function(fx) {
    flag <- 1
    for (wjj1 in ieiwk.sync.wjj2do()) {
        wjj2 <- sub("\\/d\\/", "/m/", wjj1)
        if (flag) {
            .cat.time()
            flag <- 0
        }
        .cat("doing ", wjj1)
        switch(fx, come = ieiwk.sync.wjj(wjj2, wjj1), go = ieiwk.sync.wjj(wjj1, wjj2))
        .cat.time()
    }
}
wk.rsync.safe <- function(dz1, wjj2 = ".", co1 = 30, update = FALSE) {
    if (normalizePath(wjj2) != getwd()) {
        wjj.36_03_04_003344 <- getwd()
        on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
        hs.wjj(wjj2, go = 1)
    }
    wj_obtained <- paste0(basename(dz1), ".obtained")
    if (file.exists(wj_obtained) && update && wj.update.reporter(basename(dz1), co1 = co1)) {
        hs.wj.rm(wj_obtained)
    }
    if (file.exists(wj_obtained)) {
    }
    else {
        if (hs.grepl(dz1, "^(ftp|https?):")) {
            dz1 <- hs.sub(dz1, "(ftp|https?)", "rsync")
        }
        msg <- 1
        msg <- .try(.wk("rsync -ahPv --no-motd", dz1, "."))
        if (!msg) {
            cat("", file = wj_obtained)
        }
    }
    normalizePath(wj_obtained)
}
.s <- function(..., max.print = 99999, width = 80, h = 100, w = 10000, file = hs.home.expand("~/tmp.r.sink.txt"), O = 0, append = TRUE, type = c("output", "message"), split = FALSE, print.result = 1) {
    op <- options(max.print = max.print, width = w)
    on.exit(options(max.print = h, width = width), add = TRUE)
    if (file != hs.home.expand("~/tmp.r.sink.txt")) 
        O <- 1
    if (O) 
        sink(file, append, type, split)
    r <- list(...)
    if (print.result) 
        invisible(sapply(r, print))
    if (O) 
        sink()
    invisible()
}
.s.no.blank.line <- function(...) {
    b <- capture.output(.s(...))
    b.1 <- grep("^(\\s+)*$", b)
    if (length(b.1)) 
        b <- b[-b.1]
    cat(b, sep = "\n")
}
.s1 <- .s.no.blank.line
.sh.no.blank.line <- function(...) {
    b <- capture.output(.sh(...))
    b.1 <- grep("^(\\s+)*$", b)
    if (length(b.1)) 
        b <- b[-b.1]
    cat(b, sep = "\n")
}
.sh1 <- .sh.no.blank.line
.s.h <- function(x, how.many = 3) {
    .s(.h(x, how.many))
}
.sh <- function(..., max.print = 99999, width = 80, h = 50, w = 120, file = NULL, append = FALSE, type = c("output", "message"), split = FALSE) {
    op <- options(max.print = max.print, width = w)
    on.exit(options(op), add = TRUE)
    if (is.null(file)) {
        sapply(list(...), print)
    }
    else {
        sink(file, append, type, split)
        sapply(list(...), print)
        sink()
    }
    invisible()
}
.sh.h <- function(x, how.many) {
    .sh(.h(x, how.many))
}
.cat.inside.function.31.10.20.111835 <- function(file.function = "", file.out = hs.home.expand("~/31.10.20.111907.txt"), append = TRUE) {
    a <- sys.calls()
    .cat(paste(a[[length(a) - 1]])[1], get(file.function, envir = .GlobalEnv), sep = "\t", file = file.out, append = append)
}
.c <- function(..., file = "", sep = " ", end = "\n", fill = FALSE, labels = NULL, append = FALSE, intern = FALSE) {
    if (intern) 
        return(capture.output(cat(..., file = "", sep = sep, fill = fill, labels = labels, append = append)))
    if (is.character(file)) 
        if (file == "") 
            file <- stdout()
        else if (substring(file, 1L, 1L) == "|") {
            file <- pipe(substring(file, 2L), "w")
            on.exit(close(file))
        }
        else {
            file <- file(file, ifelse(append, "a", "w"))
            on.exit(close(file))
        }
    .Internal(cat(list(...), file, sep, fill, labels, append))
    .Internal(cat(list(end), file, sep, fill, labels, append))
}
.jc <- function(..., sep = "") str_c(c(...), collapse = sep)
.cat.time <- function(intern = FALSE, file = "", append = 1) hs.say("    **** ", paste(Sys.time()), " ****", intern = intern, file = file, append = append)
.cat4browser <- function(a) {
    .cat("****    browsing at ", a)
    a
}
.cat.exp <- defmacro(DOTS, expr = {
    invisible(sapply(list(...), function(x) {
        .cat(x, " =>")
        .cat(eval(parse(text = x)))
    }))
})
println <- catline <- .ss <- function(..., file = "", sep = "\n", fill = FALSE, labels = NULL, append = FALSE) {
    cat(..., file = file, sep = sep, fill = fill, labels = labels, append = append)
}
report <- function(x, file = "tmp") {
    sink(file)
    .s(x)
    sink()
}
ieiwk.s <- function(..., file = hs.home.expand("~/tmp.txt"), append = FALSE, sep = "\n") {
    sink(file, append)
    sapply(list(...), try)
    on.exit(sink())
}
ieiwk.s.list.in.row <- function(a, sep = "|") {
    if (is.null(names(a))) 
        names(a) <- seq.int(length(a))
    r <- c()
    ir <- sapply(names(a), function(x) {
        x <- a[x]
        a <- capture.output(.s(x))
        for (i in seq.int(max(nchar(a)))) {
            b <- substr(a, i, i)
            r <- rbind(r, ifelse(nchar(b), b, " "))
        }
        r <<- rbind(r, rep(sep, ncol(r)))
        max(nchar(a))
    })
    if (all(r[-sapply(seq_along(ir), function(i) {
        sum(ir[1:i]) + i
    }), ncol(r)] == " ")) 
        r[sapply(seq_along(ir), function(i) {
            sum(ir[1:i]) + i
        }), ncol(r)] <- paste(rep(" ", nchar(sep)), collapse = "")
    apply(r, 2, paste, collapse = "")
}
.str <- function(object, list.len = 3, ...) {
    str(object, list.len = list.len, ...)
}
ieiwk.cat.line <- function(v, sep = "\n", ...) {
    cat(v, "", sep = sep, ...)
}
ieiwk.cat.headline <- function(x, l = "_", r = "#") cat(sprintf("%s\\ %s /%s\n", rep(l, 9) %,% "", x, rep(r, 9) %,% ""))
hs.save <- ieiwk.save <- function(..., blm, file, wj, nt = 8, bin = wk.gz.good(), cl = 6, env1 = parent.frame()) {
    if (missing(blm)) 
        blm <- as.character(substitute(list(...)))[-1L]
    if (!{
        all(is.character(blm)) && all(substr(blm, 100, 100) == "")
    }) 
        stop("need a variable name")
    if (missing(wj)) 
        wj <- file
    hs.wj(wj)
    con1 <- pipe(.sys(switch(bin, pbgzip = .sys.pbgzip.in(n = nt), pigz = c("pigz -p", nt, "-c"), gzip = "gzip", browser("33.08.05.222656")), "2>&1 >", .sys.fpGood(wj)), "wb")
    on.exit(close(con1), add = TRUE)
    if (length(blm) == 1) {
        evalbq(do.call("saveRDS", list(object = .(as.symbol(blm)), file = .(con1))))
    }
    else save(list = blm, file = con1, envir = env1)
}
ieiwk.save.image <- function(file = c("image", readline("Input an ID: "), "RData") %,% ".", version = NULL, ascii = FALSE, compress = FALSE, safe = TRUE) {
    save.image(file, version, ascii, compress, safe)
}
hs.load.image <- function(f1 = ".RData", env = .GlobalEnv) {
    load(f1, envir = env)
}
ieiwk.where2seek4parallel <- function(f, chunk.size = 1e+07, n = {
}, skip = 0, skip.pattern) {
    con <- file(f)
    open(con)
    if (skip) {
        readLines(con, skip)
    }
    i <- seek(con)
    f.s <- .sys.stat.size(f)
    if (!is.null(n)) 
        chunk.size <- (f.s - i)/n
    while ({
        j <- seek(con)
    } < f.s) {
        seek(con, where = j + chunk.size)
        readLines(con, 1)
        i <- append(i, seek(con))
    }
    i[length(i)] <- f.s
    on.exit(close(con))
    cbind(head(i, -1), diff(i))
}
wk.fread.txtFromXls <- function(wj, header.ms1 = "^[^\t]\t", n2try = 100, ...) {
    n1 <- readLines(wj, n2try) %=>% hs.grep(header.ms1)
    wk.fread(wj, skip = n1 - 1, ...)
}
wk.gz.good <- local({
    bin <- {
    }
    function() {
        if (length(bin)) {
            return(bin)
        }
        wj.tmp <- tempfile()
        bin1 <- "gzip"
        for (bin1 in (bin %or% c("pbgzip", "pigz", "bgzip", "gzip"))) {
            unlink(wj.tmp)
            .try({
                con1 <- .sys(bin1, "-c", "2>/dev/null", ">", wj.tmp) %=>% pipe("w")
                cat(letters, sep = "\n", file = con1)
                close(con1)
            })
            if (hs.wj.exists(wj.tmp, size = 9)) {
                bin <<- bin1
                break
            }
        }
        unlink(wj.tmp)
        bin
    }
})
wk.fread.gz <- ieiwk.fread.gz <- function(wj.gz, bin = wk.gz.good(), N = {
}, n = N, ...) {
    wk.fread(cmd = .sys(switch(bin, pigz = .sys.pigz(), gzip = c("gzip -cd"), pbgzip = .sys.pbgzip.out()), wj.gz, if (length(n)) 
        c("|", .sys.head(n, pipe = 0))), ...)
}
ieiwk.fread.pipe <- function(input, sep = "auto", sep2 = "auto", nrows = -1L, header = "auto", na.strings = "NA", stringsAsFactors = FALSE, verbose = getOption("datatable.verbose"), autostart = 1L, skip = 0L, select = NULL, drop = NULL, colClasses = NULL, integer64 = getOption("datatable.integer64"), dec = if (sep != ".") "." else ",", col.names, check.names = FALSE, encoding = "unknown", strip.white = TRUE, showProgress = getOption("datatable.showProgress"), data.table = getOption("datatable.fread.datatable"), 
    chunk = 1e+09) {
    h1 <- list()
    i1 <- 1
    while (length(s1 <- readChar(input, chunk))) {
        .ss(i1)
        i1 <- i1 + 1
        s2 <- readLines(input, 1)
        s <- ifelse(length(s2), sprintf("%s%s", s1, s2), s1)
        h1[[length(h1) + 1]] <- ieiwk.fread(input = s, sep = sep, sep2 = sep2, nrows = nrows, header = header, na.strings = na.strings, stringsAsFactors = stringsAsFactors, verbose = verbose, autostart = autostart, skip = skip, select = select, drop = drop, colClasses = colClasses, integer64 = integer64, dec = dec, col.names = col.names, check.names = check.names, encoding = encoding, strip.white = strip.white, showProgress = showProgress, data.table = FALSE)
    }
    rbind %(% h1
}
wk.fwrite <- function(x, wj = "", append = FALSE, quote = "auto", sep = "\t", sep2 = c("", "|", ""), eol = if (.Platform$OS.type == "windows") "\r\n" else "\n", na = "", dec = ".", row.names = 0, col.names = !append, qmethod = c("double", "escape"), logicalAsInt = 0, dateTimeAs = c("ISO", "squash", "epoch", "write.csv"), buffMB = 8L, nThread = getDTthreads(), showProgress = TRUE, verbose = getOption("datatable.verbose"), compress = grepl("\\.gz$", wj)) {
    if (!is.list(x)) {
        x <- if (is.matrix(x)) {
            wk.dt.as(x, keep.rownames = as.logical(row.names))
        }
        else if (is.vector(x)) {
            list(x)
        }
        else {
            hs.browser("33.07.03.173719")
            stop("x不是列表、矩阵或向量")
        }
    }
    if (is.numeric(quote)) 
        quote <- as.logical(quote)
    fwrite(x = x, file = wj, append = as.logical(append), quote = quote, sep = sep, sep2 = sep2, eol = eol, na = na, dec = dec, row.names = FALSE, col.names = as.logical(col.names), qmethod = match.arg(qmethod), logicalAsInt = as.logical(logicalAsInt), dateTimeAs = match.arg(dateTimeAs), buffMB = buffMB, nThread = nThread, showProgress = showProgress, verbose = verbose)
    wj
}
wk.fwrite.gz <- function(x, wj, ...) {
    wk.fwrite(x, wj, ..., compress = 1)
}
ieiwk.write.tsv <- write.tsv <- function(x, file = "", append = 0, quote = 0, sep = "\t", eol = "\n", na = "NA", dec = ".", row.names = 0, col.names = 0, qmethod = c("escape", "double"), compress = 0) {
    {
        row.names <- as.logical(row.names)
        col.names <- as.logical(col.names)
        append <- as.logical(append)
        quote <- as.logical(quote)
    }
    if (row.names && col.names) {
        x <- cbind(rownames(x), x)
        row.names = FALSE
    }
    con <- if (compress) 
        pipe(sprintf("pigz -p8 -6 >%s %s.gz", ifelse(append, ">", ""), file))
    else file
    write.table(x, con, append, quote, sep, eol, na, dec, row.names, col.names, qmethod)
    ifelse(compress, hs.wjm(file, add = "gz"), file)
}
wk.writeLines <- function(text, file, append = file.exists(file), sep = "\n", quote = FALSE, ...) {
    wk.fwrite(as.list(text), wj = file, append = append, sep = sep, quote = quote, ...)
}
ieiwk.write.tsv.gz <- function(x, ...) {
    ieiwk.write.tsv(x, ..., compress = 1)
}
ieiwk.smart.sep <- function(f1, sep2try = c(","), quote = "\"", dec = ".", fill = TRUE, comment.char = "#", as.is = TRUE, skip = FALSE, ...) {
    {
        s1 <- "\t"
        if (grepl(s1, readLines(f1, skip + 1))) 
            return(s1)
    }
    a <- ieiwk.sapply(sep2try, function(s1) {
        read.csv(f1, header = FALSE, s1, quote, dec, fill, comment.char, as.is = as.is, nrows = 99, ...)
    })
    clg(sapply(names(a), function(s1) {
        a1 <- a[[s1]]
        flag <- 1
        ncol(a1) == 1
        if (nchar(a1[1, ncol(a1)]) == 0 & any(nchar(a1[-1, ncol(a1)]) != 0)) 
            flag <- 0
        flag
    }), 1)
}
wk.do.con.. <- function(con1, e1, ..., s1, env = parent.frame()) {
    e1 <- substitute(e1)
    repeat {
        assign(s1, {
        }, envir = env)
        assign(s1, wk.read.chunk(con1, ...), envir = env)
        if (!length(get(s1, envir = env))) 
            break
        eval(e1, env)
    }
}
ieiwk.rt <- function(f1) {
}
wk.read.csv <- ieiwk.read.table <- ieiwk.read.csv <- function(file, header = TRUE, sep = ",", sheet = 1, skip = 0, smart.sep = TRUE, quote = "\"", dec = ".", fill = TRUE, comment.char = "#", as.is = TRUE, row.names = {
}, ...) {
    if (local({
        bin <- hs.home.expand("~/wk/my/rj2/cygwin64/bin/file.exe")
        if (file.exists(bin)) {
            a <- .try(.sys.do(ifelse(osType() == "Windows", hs.home.expand("~/wk/my/rj2/cygwin64/bin/file.exe"), "file"), path.expand(file), intern = 1))
            re("[^:]+$", a) %=>% trim %=>% tolower %=>% grepl("microsoft excel", x)
        }
        else {
            tolower(ieiwk.filename(file)) %in% c("xlx", "xlxs")
        }
    })) {
        require(readxl)
        read_excel(file, sheet = sheet, col_names = header, skip = skip)
    }
    else {
        if (smart.sep) {
            sep <- ieiwk.smart.sep(file)
        }
        a <- read.table(file = file, header = FALSE, sep = sep, quote = quote, dec = dec, fill = fill, comment.char = comment.char, row.names = {
        }, as.is = as.is, skip = skip + as.integer(header), ...)
        if (header) 
            colnames(a) <- as.matrix(read.table(file = file, header = FALSE, sep = sep, quote = quote, dec = dec, fill = fill, comment.char = comment.char, row.names = {
            }, skip = skip, nrow = 1, colClasses = "character", ...))
        if (hs.len1(row.names) && is.numeric(row.names) && (row.names > 0)) {
            rownames(a) <- as.matrix(a[, row.names])
            a <- a[, -row.names, drop = FALSE]
        }
        a
    }
}
wk.readTableExtra <- function(wj, sep = "\t", header = 1) {
    a <- readLines(wj)
    i1 <- wk.re(a, sep, "locate", all = 1)
    b <- wk.re(a, act = "split", info = i1 %=>% lapply(wk.attr.hs, . %=>% head(x, length(i1[[1]]))))
    if (header) {
        b[-1] %=>% {
            rbind %(% x
        } %=>% data.table %=>% setnames(b[[1]])
    }
    else b %=>% {
        rbind %(% x
    } %=>% data.table
}
ieiwk.readxl <- function(path, sheet = 1, col_names = TRUE, col_types = NULL, na = "", skip = 0) {
    require(readxl)
    read_excel(path = path, sheet = sheet, col_names = col_names, col_types = col_types, na = na, skip = skip)
}
ieiwk.xls.list.sheetNames <- function(wj1) {
    if (require(readxl)) {
        excel_sheets(path.expand(wj1))
    }
    else warning("install readxl first")
}
wk.xls.read <- ieiwk.xls.read <- function(wj1, sheet = 1, skip = 0, cn = 1, header = cn, as.dt = !rn, rn = 0, ...) {
    sn <- ieiwk.xls.list.sheetNames(wj1)
    hs.switch(sheet, 0, sn, "", {
        wk.xls.read(wj1, seq_along(sn), skip = skip, header = header, as.data.table = as.dt, ...)
    }, {
        i1 <- which(is.character(sheet))
        if (length(i1)) 
            sheet[i1] <- match(sheet[i1], sn)
        sheet %<=>% as.numeric
        r <- lapply(sheet, . %=>% {
            readxl::read_excel(wj1, x, skip = skip, col_names = as.logical(header), ...) %=>% {
                if (as.dt) {
                  wk.dt.as(x)
                }
                else {
                  if (rn) 
                    hs.df(x, rn = rn)
                  else x
                }
            }
        })
        if (hs.len1(sheet)) {
            r <- r[[1]]
        }
        else names(r) <- sn[sheet]
        r
    })
}
wk.xls.write <- function(bg, wj, sheetName = "Sheet1", cn = 1, rn = 0, append = 0, file = wj, widths = 8.43, ...) {
    wj <- path.expand(file)
    {
        hs.wj.rm(wj)
    }
    if (is.matrix(bg)) 
        bg <- wk.dt.as(bg, keep.rownames = "")
    openxlsx::write.xlsx(bg, hs.wj(file), rowNames = as.logical(rn), colNames = as.logical(cn), colWidths = widths)
    wj
}
wk.2xls <- function(w1 = ".", ow = 0) {
    if (dir.exists(w1)) 
        w1 <- hs.wjj.ls(w1, "\\.(tsv|csv)(\\.gz)?$", R = 1)
    wk.mclapply(w1, function(w1) {
        w2 <- hs.sub(w1, "\\.(tsv|csv)(\\.gz)?$", ".xlsx")
        if ((!file.exists(w2)) || ow) 
            wk.fread(w1) %=>% openxlsx::write.xlsx(w2)
        w2
    }) %=>% unlist
}
read.tsv <- function(file, header = 1, sep = "\t", quote = "\"", dec = ".", fill = TRUE, comment.char = "", as.is = TRUE, ...) {
    read.csv(file, header, sep, quote, dec, fill, comment.char, as.is = as.is, ...)
}
ieiwk.line.sep.good <- function(f1) {
    f1 <- .sys.fpGood.file(f1)
    if (.file(f1) == "ASCII text, with CRLF line terminators") 
        .sys.do("sed -i", shQuote("s/\r//"), f1)
}
wk.r2sys <- function(lines = letters, cmd = "head", cmd.init = "cat", intern = 0) {
    wj1 <- hs.wj.temp()
    zz <- fifo(wj1, "w+")
    on.exit({
        close(zz)
        unlink(wj1)
    }, add = TRUE)
    writeLines(lines, zz)
    .wk(cmd.init, wj1, "|", cmd, "&", intern = intern)
}
wk.sys2sys <- function(cmd1, cmd2 = "head", cmd.init = "cat", intern = 0) {
    wj1 <- ieiwk.tempfile()
    zz <- fifo(wj1, "w+")
    on.exit({
        close(zz)
        unlink(wj1)
    }, add = TRUE)
    browser("33.07.02.154349")
    .wk(cmd1, ">", wj1, "&")
    .wk(cmd.init, wj1, "|", cmd2, intern = intern)
}
hs.bgz <- function(wj1, wj2 = paste0(wj1, ".bgz")) {
    .wk("zcat", wj1, "| pbgzip -c >", wj2)
    if (hs.grepl(wj1, "\\.bgz$")) {
        file.rename(wj2, wj1)
    }
    else hs.wj.rm(wj1)
    wj2
}
cn.hs.wj <- function(wj1, skip.comment = "auto") {
    size1 <- 1000
    repeat {
        line <- readLines(wj1, size1)
        if (substr(line[1], 1, 1) == "#") {
            i1 <- which(substr(line, 1, 1) == "#")
            i2 <- which(diff(i1) > 1)
            if (hs.len1(i2, 0)) 
                i2 <- tail(i1, 1)
            if (i2 == length(line)) {
                size1 <- size1 + 1000
                if (size1 > 10000) {
                  stop("37_03_09_125412")
                }
                next
            }
            if (is.character(skip.comment)) {
                switch(skip.comment, auto = {
                  a1 <- readLines(wj1, i2 + 1)
                  skip.comment <- length(strsplit(a1[i2], "\t")[[1]]) < length(strsplit(a1[i2 + 1], "\t")[[1]])
                })
            }
            else {
                if (is.numeric(skip.comment)) 
                  skip.comment <- as.logical(skip.comment)
            }
            if (skip.comment) 
                i2 <- i2 + 1
            line1 <- line[i2]
            break
        }
        else {
            line1 <- line[1]
            break
        }
    }
    strsplit(line1, "\t")[[1]]
}
text.cmd.text <- function(L, cmd1, reader = function(con1) fread(text = readLines(con1), header = FALSE)) {
    fifo1 <- hs.wj.temp()
    .wk("mkfifo", fifo1)
    in1 <- pipe(.sys("cat <", fifo1, "|", cmd1, "&"), "r")
    out1 <- fifo1
    hs.browser("37.03.05.184231")
    cat(L, sep = "\n", file = out1)
    close(out1)
    jg <- reader(in1)
    close(in1)
    hs.wj.rm(fifo1)
    jg
}
wk.wget2wj <- function(dz1, wj2, redo = 0, gzip = !hs.grepl(wj2, "(?i)\\.gz$"), extra = "") {
    if (redo) 
        hs.wj.rm(wj2)
    if (!file.exists(wj2)) {
        .wk("wget -O -", extra, .q(dz1), if (gzip) 
            "| gzip", ">", .sys.fpGood(hs.wj(wj2)))
    }
    wj2
}
hs.file_system_is_linkable <- function(wjj1) {
    hs.with_wd(wjj1, {
        wj1 <- "39_01_29_102630"
        wj2 <- paste0(wj1, ".lnk")
        hs.wj.rm(wj2)
        cat("", file = wj1)
        a <- file.symlink(wj1, wj2)
        hs.wj.rm(c(wj1, wj2))
        a
    })
}
.jl <- function(wj.org = hs.home.expand("~/org/文件历史.org")) {
    wj1 <- osType() %=>% hs.switch("Darwin", "pbpaste", "xsel --clipboard --output") %=>% .wk(intern = 1) %=>% if ((length(x) == 1) && file.exists(x)) 
        hs.wjm.home(x)
    if (length(wj1)) {
        old <- if (file.exists(wj.org)) 
            wk.fread(wj.org, colClasses = "character")
        else rep("", 2) %=>% t %=>% wk.dt.as(x)[-1]
        i <- old[, match(wj1, V1, nomatch = 0)]
        current <- if (i) {
            Sys.time() %=>% paste(old[i, V2], sep = "; ") %=>% list(wj1, x) %=>% wk.dt.as %=>% rbind(old[-i])
        }
        else {
            list(wj1, Sys.time() %=>% paste) %=>% wk.dt.as %=>% rbind(old)
        }
        wk.dt.fwrite(current, wj.org, cn = 0)
    }
    .less(wj.org)
    wj.org
}
wk.dt <- function(..., check.names = FALSE, key = NULL, stringsAsFactors = FALSE, sorted = TRUE, value.name = "value", na.rm = TRUE, as = 0, keep.rownames = "row_name") {
    if (...length()) {
        i <- match("x", ...names(), nomatch = 1)
        if (is.array(...elt(i)) && is.null(rownames(...elt(i)))) 
            keep.rownames <- FALSE
    }
    if (as) {
        data.table::as.data.table(..., keep.rownames = keep.rownames, key = key, sorted = sorted, value.name = value.name, na.rm = na.rm)
    }
    else data.table::data.table(..., keep.rownames = keep.rownames, check.names = check.names, key = key, stringsAsFactors = stringsAsFactors)
}
wk.dt.as <- data.table::as.data.table
wk.dt.uniq_n <- data.table::uniqueN
wk.dt.setattr <- function(x, name = {
}, value) {
    if (length(name)) 
        data.table::setattr(x, name, value)
    else attributes(x)
}
wk.dt.copy <- data.table::copy
wk.dt.absorb <- function(dt1, ..., env1 = parent.frame()) {
    for (dt2 in list(...)) {
        if ((if (is.list(dt2)) 
            dt2[[1]]
        else nrow(dt2)) %not.in% c(dt1[, .N], 1)) {
            stop("[36.02.04.153623]")
        }
        if (!is.list(dt2)) 
            dt2 <- data.table::as.data.table(dt2)
        dt1[, `:=`(names(dt2), dt2)]
    }
    dt1
}
wk.dt.is_clean <- function(dt, by = {
}, ...) {
    a <- if (length(by)) {
        dt[, lapply(.SD, hs.clean, ot = "l", ...), .SDcols = c(by)]
    }
    else dt[, lapply(.SD, hs.clean, ot = "l", ...)]
    hs.rowAlls(a)
}
wk.dt.absorb1.core <- local({
    hs.matcher <- function(c1, c2, ic) if (ic && (is.character(c1[1]) || is.character(c2[1]))) {
        hs.str_match(c1, c2, ic = TRUE, nomatch = 0L)
    }
    else match(c1, c2, nomatch = 0L)
    hs.eq <- function(c1, c2, ic) {
        if (ic && (is.character(c1[1]) || is.character(c2[1]))) {
            c1 <- tolower(c1)
            c2 <- tolower(c2)
        }
        c1 == c1
    }
    function(dt1, dt2, by = (names(dt1) %and% names(dt2))[1], by1 = if (length(dt1) == 1) 
        names(dt1)
    else by, by2 = by, what = names(dt2) %not% by2, as = what, ic = FALSE, ri2do, old = "replace", sep = ";") {
        if (!length(ri2do)) {
            ri2do <- which(wk.dt.is_clean(dt1, by = by1))
            if (length(ri2do) == dt1[, .N]) 
                ri2do <- {
                }
        }
        rj2do <- which(wk.dt.is_clean(dt2, by = by2))
        if (!length(rj2do)) 
            return(dt1)
        hs.switch(c(length(ri2do) > 0, length(rj2do) < dt2[, .N]), c(1, 1), {
            m1.38_09_24_233212 <- hs.matcher(dt1[[by1[1]]][ri2do], dt2[[by2[1]]][rj2do], ic = ic)
            i <- which(m1.38_09_24_233212 > 0)
            dt1.matched.ri <- ri2do[i]
            m1.38_09_24_233212 <- rj2do[m1.38_09_24_233212[i]]
        }, c(1, 0), {
            m1.38_09_24_233212 <- hs.matcher(dt1[[by1[1]]][ri2do], dt2[[by2[1]]], ic = ic)
            i <- which(m1.38_09_24_233212 > 0)
            if (length(i) < length(m1.38_09_24_233212)) {
                m1.38_09_24_233212 <- m1.38_09_24_233212[i]
                dt1.matched.ri <- ri2do[i]
            }
            else dt1.matched.ri <- ri2do
        }, c(0, 1), {
            m1.38_09_24_233212 <- hs.matcher(dt1[[by1[1]]], dt2[[by2[1]]][rj2do], ic = ic)
            dt1.matched.ri <- which(m1.38_09_24_233212 > 0)
            m1.38_09_24_233212 <- rj2do[m1.38_09_24_233212[dt1.matched.ri]]
        }, c(0, 0), {
            m1.38_09_24_233212 <- hs.matcher(dt1[[by1[1]]], dt2[[by2[1]]], ic = ic)
            dt1.matched.ri <- which(m1.38_09_24_233212 > 0)
            m1.38_09_24_233212 <- m1.38_09_24_233212[dt1.matched.ri]
        })
        if (!length(dt1.matched.ri)) 
            return(dt1)
        for (i in hs.seqForward(2, length(by1))) {
            m2 <- hs.eq(dt1[[by1[i]]][dt1.matched.ri], dt2[[by2[i]]][m1.38_09_24_233212], ic = ic)
            if (!all(m2)) {
                dt1.matched.ri <- dt1.matched.ri[m2]
                m1.38_09_24_233212 <- m1.38_09_24_233212[m2]
            }
            if (!length(dt1.matched.ri)) 
                return(dt1)
        }
        if (old == "merge") {
            for (i in seq_along(as)) local({
                as1 <- as[i]
                what1 <- what[i]
                a <- dt2[[what1]][m1.38_09_24_233212]
                b <- dt1[[as1]][dt1.matched.ri]
                bi <- hs.clean(b, ot = "i")
                if (length(bi)) {
                  if (!is.character(a)) 
                    a <- hs.str_as(a)
                  if (all(substring(b[bi], nchar(b[bi]) - nchar(a[bi]) + 1) == a[bi])) {
                    if (length(bi) < length(b)) {
                      a <- a[-bi]
                      dt1.matched.ri <- dt1.matched.ri[-bi]
                    }
                    else return()
                  }
                  else a[bi] <- paste(b[bi], a[bi], sep = sep)
                }
                if (is.character(dt2[[what1]]) && (!is.character(dt1[[as1]]))) 
                  wk.dt.set(dt1, , as1, hs.str_as(dt1[[as1]]))
                wk.dt.set(dt1, dt1.matched.ri, as1, a)
            })
            return(dt1)
        }
        else {
            wk.dt.set(dt1, dt1.matched.ri, as, dt2[, .SD[m1.38_09_24_233212], .SDcols = c(what)])
        }
    }
})
wk.dt.absorb1 <- function(sep = "auto", sep2 = if (sep %in% c("auto", "")) ";" else sep, sep_set = "[;\\|]", old = "merge", na2blank = TRUE, collapse = TRUE, hs.clean = .GlobalEnv[["hs.clean"]], hs.dirty = .GlobalEnv[["hs.dirty"]]) {
    force(sep2)
    if (!length(as)) 
        as <- what
    if (hs.clean(sep, ot = "n")) {
        if (sep == "auto") 
            sep <- ""
        for (by11 in by1) {
            sep <- hs.detect.sep(dt1[[by11]], sep_set = sep)
            if (nzchar(sep)) 
                break
        }
    }
    if (nzchar(sep)) {
        cn2do <- c(by1, if (old == "keep") as)
        dt1.by <- if (cn2do %all.in% names(dt1)) {
            dt1
        }
        else {
            dt1[, .SD, .SDcols = c(cn2do)]
        }
        dt1.by[, `:=`(i.38_09_11_140638, .I)]
    }
    else dt1.by <- dt1
    ri2do <- if (old == "keep") 
        local({
            as2do <- as %and% names(dt1.by)
            if (length(as2do)) {
                ri2do <- hs.dirty(dt1.by[[as2do[1]]], ot = "i")
                for (as1 in as2do[-1]) {
                  if (!length(ri2do)) 
                    break
                  ri2do <- ri2do[hs.dirty(dt1.by[[as1]][ri2do], ot = "l")]
                }
                ri2do
            }
        })
    if (hs.clean(sep, ot = "n")) 
        for (by11 in by1) dt1.by <- wk.dt.expand(dt1.by, by11, sep = sep, ri2do = ri2do)
    if (collapse) 
        dt2 <- wk.dt.collapse(dt2, by = by2, U = 1, sep = sep2)
    wk.dt.absorb1.core(dt1 = dt1.by, dt2 = dt2, by = by, by1 = by1, by2 = by2, what = what, as = as, ic = ic, ri2do = ri2do, old = old, sep = sep2)
    if ("i.38_09_11_140638" %in% names(dt1.by)) {
        if (dt1.by[, anyDuplicated(i.38_09_11_140638)]) {
            as_not_to_collapse <- local({
                as <- as %and% names(dt1)
                if (!length(as)) 
                  return()
                a <- dt1.by[, unique(.SD), by = i.38_09_11_140638, .SDcols = c(as)]
                as[sapply(as, function(as1) {
                  i <- is.na(a[[as1]])
                  if (!all(is.na(dt1[[as1]][i]))) 
                    return(FALSE)
                  i <- !i
                  all(a[[as]][i] == dt1[[as1]][i])
                })]
            })
            dt1.by <- wk.dt.collapse(dt1.by, by = unique(c("i.38_09_11_140638", as_not_to_collapse, names(dt1.by) %not% c(by1, as))), sep = sep2)
            if (length(as_not_to_collapse)) 
                as <- as %not% as_not_to_collapse
            if (length(as)) 
                wk.dt.set(dt1, , as, dt1.by[, .SD, .SDcols = c(as)])
        }
        dt1[, `:=`("i.38_09_11_140638", {
        })]
    }
    return(dt1)
    if (ic) {
        dt1.by <- dt1[, .SD, .SDcols = c(by1)]
        dt2.by <- dt2[, .SD, .SDcols = c(by2)]
        wk.dt.replace(dt1.by, . %=>% hs.str_as %=>% tolower)
        wk.dt.replace(dt2.by, . %=>% hs.str_as %=>% tolower)
    }
    {
        switch(old, keep = {
            for (as1 in as %and% names(dt1)) {
                i1 <- hs.clean(dt1[[as1]], ot = "l")
                i.38_09_23_103500[i1] <- -1
            }
        })
        ri2do <- which(i.38_09_23_103500 >= 0)
        for (by_i in seq_along(by1)) {
            m1 <- hs.str_match(dt1[[by1[by_i]]][ri2do], dt2[[by2[by_i]]], ic = ic)
            i1 <- !is.na(m1)
            i.38_09_23_103500[ri2do[i1]] <- m1[i1]
        }
    }
    ri2do <- which(i.38_09_23_103500 > 0)
    if (length(ri2do)) {
        i.38_09_23_103500 <- i.38_09_23_103500[ri2do]
        a <- dt2[, .SD[i.38_09_23_103500], .SDcols = -c(by2)]
        if (old == "merge") {
            hs.browser("38.09.23.154900")
            hs.for(as1, as %and% names(dt1), wk.dt.set(dt1, , as1, hs.str_as(dt1[[as1]])), by1i, seq_along(by1), by1[by1i])
        }
        wk.dt.set(dt1, ri2do, as, a)
    }
    return(dt1)
    if (sep == "auto") {
        sep <- if (length(by1) == 1) 
            local({
                i <- hs.grep(dt1[[by1]], sep_set)
                if (length(i)) 
                  hs.re(dt1[[by1]][i[1]], sep_set)
            })
        if (missing(sep2)) 
            sep2 <- sep
    }
    if (dt2[, anyDuplicated(.SD), .SDcols = c(by2)]) 
        dt2 <- wk.dt.collapse(dt2, by = by2, sep = if (length(sep2)) 
            sep2
        else ";")
    if (length(by2) && length(what)) {
        if (!length(as)) 
            as <- what
        if (length(as) != length(what)) 
            hs.browser("36.12.05.215327")
        dt1.by <- dt1[, .SD, .SDcols = c(c(by1, as) %and% names(dt1))]
        if (length(sep)) {
            dt1.by[, `:=`("i.38_09_11_140638", .I)]
            dt1.by <- wk.dt.expand(dt1.by, cn1 = by1, split = sep)
        }
        if (ic) {
            for (a in by1) dt1.by[, `:=`(c(a), tolower(get(a)))]
            dt2 <- wk.dt.copy(dt2)
            for (a in by2) dt2[, `:=`(c(a), tolower(get(a)))]
        }
        dt2 <- wk.dt.subset(dt2, dt1.by, by2, ot = "in")
        if (hs.is_empty(dt2)) 
            return(invisible(dt1))
        i.38_09_23_082919 <- wk.dt.subset.na(dt2, dt1, by2, ot = "i")
        if (all(is.na(i.38_09_23_082919))) 
            return(invisible(dt1))
        common <- wk.dt.subset(dt1.by, dt2.by, mult = "1")
        common <- wk.dt.is_clean(common)
        if (hs.is_empty(common)) 
            return(invisible(dt1))
        if (old %in% c("keep", "merge") && any(as %in% names(dt1))) {
            hs.browser("38.09.23.073514")
            lapply(which(as %in% names(dt1)), function(as1.i) {
                as <- as[as1.i]
                what <- what[as1.i]
                if (old == "keep") 
                  common %<>% (`%not%`(dt1.by[hs.clean(dt1.by[[as]], ot = "l"), get(by1)]))
                i1.38_09_12_065901 <- which(dt1.by[[by1]] %in% common)
                i2.38_09_11_075403 <- which(dt2.by %in% common)
                i <- match(dt1.by[[by1]][i1.38_09_12_065901], dt2.by[i2.38_09_11_075403])
                if (length(sep)) {
                  if (!is.character(dt1[[as]])) 
                    dt1[, `:=`(c(as), hs.str_as(as))]
                  wk.dt.set(dt1.by, i1.38_09_12_065901, as, dt2[[what]][i2.38_09_11_075403[i]])
                  a <- dt1.by[i.38_09_11_140638 %in% i.38_09_11_140638[i1.38_09_12_065901]]
                  a <- a[, lapply(.SD, function(x) {
                    if (na2blank) 
                      x <- ifelse(is.na(x), "", x)
                    if (all(x == "")) 
                      x <- ""
                    paste(x, collapse = sep2)
                  }), by = i.38_09_11_140638]
                  wk.dt.set(dt1, a[, i.38_09_11_140638], as, a[[as]])
                }
                else {
                  a <- dt2[[what]][i2.38_09_11_075403[i]]
                  if (old == "merge") 
                    a <- paste(dt1[i1.38_09_12_065901, get(as)], a, sep = sep2)
                  wk.dt.set(dt1, i1.38_09_12_065901, as, a)
                }
                {
                }
            })
            i <- as %not.in% names(dt1)
            as <- as[i]
            what <- what[i]
        }
        if (length(as)) {
            i1 <- wk.dt.subset(dt1.by, common, ot = "i")
            i2.38_09_11_205626 <- wk.dt.subset(dt2.by, common, ot = "i")
            i <- match(dt1.by[[by1]][i1], dt2.by[i2.38_09_11_205626])
            dt1.by
            if (length(sep)) {
                for (as1.i in seq_along(as)) wk.dt.set(dt1.by, i1, as[as1.i], dt2[[what[as1.i]]][i2.38_09_11_205626[i]])
                dt1.by <- dt1.by[, lapply(.SD, function(x) {
                  if (na2blank) 
                    x <- ifelse(is.na(x), "", x)
                  if (all(x == "")) 
                    x <- ""
                  paste(x, collapse = sep2)
                }), by = i.38_09_11_140638]
                dt1[, `:=`(c(as), dt1.by[, .SD, .SDcols = c(as)])]
            }
            else wk.dt.set(dt1, i1, as, dt2[, .SD[i2.38_09_11_205626[i]], .SDcols = c(what)])
        }
    }
    invisible(dt1)
}
formals(wk.dt.absorb1) <- hs.formals.merge(wk.dt.absorb1.core, wk.dt.absorb1)
wk.dt.absorb2 <- function(dt1, by1, ..., sep = "auto") {
    if (sep == "auto") 
        sep <- hs.cond(any(hs.grepl(dt1[[by1]], ";")), ";", any(hs.grepl(dt1[[by1]], ",")), ",", any(hs.grepl(dt1[[by1]], "|", fixed = TRUE)), "|")
    hs.browser("38.09.11.125104")
    if (length(sep)) {
        cn2do <- c(by1, {
            i <- match("what", ...names(), nomatch = 0L)
            if (i) ...elt(i)
        }, {
            i <- match("as", ...names(), nomatch = 0L)
            if (i) ...elt(i)
        })
        cn2do <- cn2do %and% names(dt1)
        a <- dt1[, .SD, .SDcols = c(cn2do)]
        a[, `:=`(i.38_09_11_112018, seq(.N))]
        a <- wk.dt.expand(a, by1, split = sep)
        wk.dt.absorb1(dt1 = a, by1 = by1, ...)
        a <- a[, lapply(.SD, paste, collapse = sep), by = i.38_09_11_112018]
        a[, `:=`(i.38_09_11_112018, {
        })]
        list(...)
    }
    else {
        wk.dt.absorb1(dt1 = dt1, by1 = by1, ...)
    }
}
wk.dt.list.merge <- function(L, cn1 = {
}) {
    if (length(L) == 1) 
        return(L[[1]])
    jg <- data.table::copy(L[[1]])
    for (i1 in seq_along(L)[-1]) {
        hs.say(i1)
        jg <- if (length(cn1)) {
            wk.dt.absorb(jg, wk.dt.subset.na(L[[i1]], jg[[cn1]], cn1)[, .SD, .SDcols = -cn1])
        }
        else wk.dt.absorb(jg, L[[i1]])
    }
    jg
}
wk.dt.list.combine <- function(sj, cn_common = lapply(sj, names) %=>% hs.list.common) {
    cn_common %<>% (`%and%`(lapply(sj, names) %=>% hs.list.common))
    jg <- lapply(sj, . %=>% x[, .SD, .SDcols = cn_common])
    jg <- rbindlist(jg) %=>% unique
    for (i1 in if (length(names(sj))) 
        names(sj) %=>% order
    else seq_along(sj)) {
        sj1 <- sj[[i1]]
        cn2do <- names(sj1) %not% cn_common
        jg[match(sj1[, Accession], Accession), `:=`(paste0(paste0("【", names(sj)[i1], "】"), cn2do), sj1[, .SD, .SDcols = c(cn2do)])]
    }
    if (0) {
        null <- lapply(names(sj), . %=>% {
            sj1 <- sj[[x]]
            cn_specific <- names(sj1) %=>% x[x %not.in% cn_common]
            setnames(sj1, cn_specific, paste0("【", x, "】", cn_specific))
        })
        jg <- sj[[1]]
        for (i1 in 2:length(sj)) {
            jg %<>% merge(sj[[i1]], all = 1)
            names(sj[[1]]) %not% names(sj[[i1]])
        }
        cn_common <- cn_common %and% names(jg)
    }
    if (0) {
        setcolorder(jg, c(cn_common, hs.grepV(names(jg), "# Unique Peptides")))
        setcolorder(jg, c(cn_common, hs.grepV(names(jg), "# PSMs")))
        Sds <- jg[, hs.rowSds(.SD), .SDcols = patterns("# PSMs")]
        not_NA <- jg[, hs.rowSums(!is.na(.SD)), .SDcols = patterns("# PSMs")]
        jg[, `:=`("not_NA", not_NA)]
        jg[, `:=`("Sds", Sds)]
        setorderv(jg, c("not_NA", "Sds"), c(-1, -1), na.last = TRUE)
        jg[, `:=`(c("not_NA", "Sds"), {
        })]
        {
            cn1 <- "# AAs"
            a1 <- lapply(jg[, .SD, .SDcols = patterns("】# PSMs$")], "/", jg[[cn1]])
            names(a1) %<>% paste0("/", cn1)
            jg[, `:=`(names(a1), a1)]
            wk.dt.col.reorder(jg, names(a1), after = tail(hs.grep(names(jg), "】# PSMs$"), 1))
        }
    }
    jg
}
wk.dt.ij <- wk.dt.subset <- function(dt, i, on = {
}, nomatch = 0, no_na = 1, inv = 0, ot = "t", exp_l = {
}, exp_j = {
}, env1 = parent.frame(), cum = 0, mult = "all", with_i = 0) {
    env2 <- new.env(parent = env1)
    env2$i1.38_08_30_224756 <- TRUE
    env2$with_i <- with_i
    if (!missing(i)) {
        if (!length(on)) {
            on <- wk.dt.index(dt, vectors = TRUE)
            if (length(on) == 1) {
                on <- unlist(on)
            }
            else {
                if (missing(i)) 
                  hs.browser("38.07.17.220949")
                on <- if (is.list(i)) {
                  if (names(i) %all.in% names(dt)) 
                    names(i)
                  else seq_along(i)
                }
                else {
                  dim(i)[2] %=>% {
                    if (length(x)) {
                      if (colnames(i) %all.in% names(dt)) 
                        names(i)
                      else seq.int(x)
                    }
                    else 1
                  }
                }
            }
        }
        if (length(on) && is.numeric(on)) 
            on <- colnames(dt)[on]
        if (ot == "on") 
            return(on)
        if (length(on)) {
            i1 <- local({
                if (!is.list(i)) 
                  i <- list(i)
                if (is.null(names(i))) 
                  names(i) <- on
                if (on %all.in% names(i)) {
                  hs.switch(mult, 1, dt[i, on = on, nomatch = nomatch, which = TRUE, allow.cartesian = TRUE, mult = "first"], "all", dt[i, on = on, nomatch = nomatch, which = TRUE, allow.cartesian = TRUE])
                }
                else {
                  if (!is(i, "data.table")) 
                    hs.browser("38.08.31.001908")
                  dt[i[, .SD, .SDcols = seq_along(on)] %=>% wk.dt.setnames(on), on = on, nomatch = nomatch, which = TRUE, allow.cartesian = TRUE]
                }
            })
            if (inv) 
                i1 <- dt[, .I] %rm% i1
        }
        else i1 <- TRUE
        env2$i <- i
        env2$on <- on
        env2$i1.38_08_30_224756 <- i1
    }
    env2$dt <- dt
    env2$nomatch <- nomatch
    env2$no_na <- no_na
    if (!missing(exp_l)) 
        env2$exp_l.37_12_30_093732 <- substitute(exp_l)
    env2$ot <- ot
    if (!missing(exp_j)) {
        env2$exp_j.37_12_31_092013 <- substitute(exp_j)
        env2$cum.38_08_07_224434 <- cum
    }
    eval(quote({
        if (exists("exp_l.37_12_30_093732", inherits = FALSE) && (length(exp_l.37_12_30_093732) > 1)) {
            acc.38_08_07_225803 <- if (exists("cum.38_08_07_224434", inherits = FALSE) && cum.38_08_07_224434) expandingList()
            for (exp1.38_01_05_234704 in as.list(exp_l.37_12_30_093732)[-1]) {
                if (!length(i1.38_08_30_224756)) break
                if (isTRUE(exp1.38_01_05_234704)) {
                  ...a <- TRUE
                } else {
                  ...a <- dt[i1.38_08_30_224756, eval(exp1.38_01_05_234704)]
                  if (is.logical(...a)) {
                    ...a <- which(...a)
                  } else if (is.null(...a)) next
                }
                i1.38_08_30_224756 <- if (isTRUE(i1.38_08_30_224756)) {
                  ...a
                } else i1.38_08_30_224756[...a]
                if (length(acc.38_08_07_225803)) acc.38_08_07_225803[["add"]](dt[i1.38_08_30_224756, eval(exp_j.37_12_31_092013)])
            }
            if (length(acc.38_08_07_225803)) return(acc.38_08_07_225803[["get"]]())
        }
        switch(ot, t = {
            jg <- if (exists("exp_j.37_12_31_092013", inherits = FALSE)) {
                dt[i1.38_08_30_224756, eval(exp_j.37_12_31_092013)]
            } else dt[i1.38_08_30_224756]
            if (is.na(nomatch) && no_na) {
                if (length(on) > 1) hs.browser("38.06.14.231859")
                if (is(jg, "data.table") && (on %in% names(jg))) {
                  i.38_06_14_235830 <- which(is.na(jg[[on]]))
                  if (length(i.38_06_14_235830)) data.table::set(jg, i.38_06_14_235830, on, (if (is(i, "list") || is(i, "data.frame")) i[[on]] else i)[i.38_06_14_235830])
                }
            }
            if (with_i) jg[, `:=`("i1.38_08_30_224756", i1.38_08_30_224756)]
            jg
        }, i = i1.38_08_30_224756, r = local({
            a.38_07_17_123841 <- {
            }
            dt[i1.38_08_30_224756, {
                a.38_07_17_123841 <<- eval(exp_j.37_12_31_092013)
                {
                }
            }]
            a.38_07_17_123841
        }), `in` = {
            i1.38_08_30_224756 <- unique(i1.38_08_30_224756)
            if (length(i1.38_08_30_224756) < dt[, .N]) {
                dt[sort(i1.38_08_30_224756)]
            } else dt
        })
    }), envir = env2)
}
wk.dt.subset.na <- function(..., no_na = 1) {
    wk.dt.subset(..., nomatch = NA, no_na = no_na)
}
wk.dt.index <- function(x, ...) {
    data.table::indices(x, ...)
}
wk.dt.setindexv <- function(x, cols = {
}, ...) {
    if (length(cols)) {
        if (is.numeric(cols)) 
            cols <- names(x)[cols]
        data.table::setindexv(x = x, cols = cols, ...)
    }
    else {
        wk.dt.setattr(x, "index", {
        })
    }
}
wk.dt.expand <- function(dt1, cn1 = names(dt1)[1], split = ";", sep = split, sep_set = "[;\\|]", l1 = {
}, ri2ignore = if (length(ri2do)) dt1[, .I] %not% ri2do, ri2do = {
}, ...) {
    if (!is.character(dt1[[cn1]])) 
        return(dt1)
    if (is.numeric(cn1)) 
        cn1 <- names(dt1)[cn1]
    if (length(l1) != dt1[, .N]) {
        if (length(l1)) 
            hs.browser("38.09.24.072841")
        if (length(sep)) {
            sep <- if (sep == "auto") {
                hs.detect.sep(dt1[[cn1]])
            }
            else hs.detect.sep(dt1[[cn1]], sep_set = sep)
            if (!nzchar(sep)) 
                return(dt1)
        }
        i <- hs.grepl(dt1[[cn1]], sep)
        i[ri2ignore] <- FALSE
        i[ri2do] <- TRUE
        if (!any(i)) 
            return(dt1)
        l1 <- as.list(dt1[[cn1]])
        l1[i] <- hs.str_split(dt1[[cn1]][i], sep = sep, ...)
        if (0) {
            l1 <- vector("list", dt1[, .N])
            i1 <- which((dt1[[cn1]] %in% "") | (dt1[[cn1]] %in% NULL))
            if (length(i1)) {
                l1[i1] <- as.list(dt1[[cn1]][i1])
                i1 <- seq_along(l1) %not% i1
            }
            else i1 <- seq_along(l1)
            l1[i1] <- strsplit(dt1[[cn1]][i1], sep)
        }
    }
    l1.len <- sapply(l1, length)
    lt1 <- l1.len > 1
    if (any(lt1)) {
        i_expanded <- dt1[, rep(.I, ifelse(lt1, l1.len, 1))]
        jg <- dt1[i_expanded, .SD]
        wk.dt.set(jg, which(i_expanded %in% which(lt1)), cn1, unlist(l1[lt1]))
    }
    else dt1
}
wk.dt.j <- wk.dt.get.j <- wk.dtSubsetCol <- function(dt, j1 = if (length(j2rm)) seq.int(length(dt)), s = {
}, hs = {
}, S, E, grep_or = {
}, grep_and = {
}, inv = FALSE, j2rm = {
}, sort = FALSE, unique = FALSE, ot = "ci") {
    if (length(grep_or)) {
        j1 %<=>% c(x, sapply(grep_or, . %=>% {
            hs.grep(names(dt), x) %=>% unlist %=>% unique
        }))
    }
    if (length(grep_and)) {
        j1 %<=>% c(x, grep_and %=>% {
            a1 <- lapply(x, . %=>% {
                hs.grep(names(dt), x)
            }) %=>% inverseList
            hs.nonempty(a1, length(x)) %=>% names %=>% as.integer
        })
    }
    if (length(j2rm)) 
        j1 %<=>% x[x %not.in% j2rm]
    if (length(s)) 
        j1 %<=>% c(x, match(s, colnames(dt)))
    if (inv) 
        j1 %<=>% {
            seq.int(length(dt)) %not% x
        }
    if (sort) 
        j1 <- sort(j1)
    return(switch(ot, cn = names(dt)[j1], ci = j1, dt = dt[, .SD, .SDcols = j1]))
    if (!(missing(S) & missing(E))) {
        if (missing(S)) 
            S <- 1
        if (missing(E)) 
            E <- ncol(dt)
        if (is.character(S)) 
            S %<>% match(colnames(dt))
        if (is.character(E)) 
            E %<>% match(colnames(dt))
        i %<>% c(S:E)
    }
    if (length(hs)) 
        i %<=>% c(x, colnames(dt) %=>% match(hs(x), x))
    if (length(j)) 
        i %<=>% x[x %not.in% j]
    if (length(i)) {
        if (sort) 
            i %<=>% sort
        if (unique) 
            i %<=>% unique
        dt[, .SD, .SDcols = colnames(dt)[i]]
    }
}
wk.dt.colConvert <- function(dt, cols, hs) {
    cols <- unique(cols)
    if (is.character(cols)) 
        cols <- which(names(dt) %in% cols)
    for (ci1 in cols) {
        v1.37_12_29_141028 <- hs(dt[[ci1]])
        dt[, `:=`(c(ci1), v1.37_12_29_141028)]
    }
    dt
}
wk.dt.col.reorder <- local({
    hs.in <- function(dt, i1, ci, type) {
        if (is.character(i1)) 
            i1 <- which(names(dt) == i1)
        if (length(i1) > 1) 
            i1 <- switch(type, before = i1[1], after = tail(i1, 1))
        n1 <- i1
        if (type == "before") 
            n1 <- n1 - 1
        if (n1) {
            c(seq.int(n1) %not% ci, ci)
        }
        else ci
    }
    function(dt, cn, i1 = {
    }, cn.pattern = {
    }, after = {
    }, before = {
    }, replace = TRUE) {
        if (length(cn.pattern)) {
            if (length(cn.pattern) > 1) {
                for (cn.pattern1 in cn.pattern) wk.dt.col.reorder(dt, i1 = i1, cn.pattern = cn.pattern1, after = after, before = before, replace = replace)
                return()
            }
            else cn <- hs.grep(names(dt), cn.pattern)
        }
        if (is.character(cn)) 
            cn <- match(cn, names(dt))
        i2 <- if (length(after)) {
            hs.in(dt, after[1], cn, "after")
        }
        else if (length(before)) {
            hs.in(dt, before[1], cn, "before")
        }
        else {
            if (length(i1) == 0) {
                hs.in(dt, min(cn) - 1, cn, "after")
            }
            else {
                if (length(i1) == length(cn)) {
                  if (is.character(i1)) 
                    i1 <- match(i1, names(dt))
                  a1 <- rep(0, length(dt))
                  a1[i1] <- cn
                  if (replace) 
                    a1[cn] <- i1
                  if (0 %in% a1) 
                    a1[a1 == 0] <- seq.int(length(dt)) %not% a1
                  a1
                }
                else stop("【36.02.14.202704】")
            }
        }
        setcolorder(dt, i2)
    }
})
wk.dt.fwrite <- function(dt, wj, i2rm = {
}, quote = 0, add = smart && file.exists(wj), cn = !add, rn = {
}, sep = "\t", smart = FALSE, ...) {
    hs.wj(wj)
    if (length(rn) && inherits(dt, "data.frame") && (!inherits(dt, "data.table"))) 
        dt <- wk.dt.as(dt, keep.rownames = rn)
    if (length(i2rm)) {
        dt[-i2rm, data.table::fwrite(.SD, wj, sep = sep, quote = as.logical(quote), col.names = as.logical(cn), append = as.logical(add), ...)]
    }
    else {
        if ((!is.list(dt)) && (!length(dim(dt)))) 
            dt <- list(dt)
        data.table::fwrite(dt, wj, sep = sep, quote = as.logical(quote), col.names = as.logical(cn), append = as.logical(add), ...)
    }
}
wk.setindex <- function(dt, cols = 1) {
    if (length(cols)) {
        if (is.numeric(cols)) 
            cols <- colnames(dt)[cols]
    }
    else stop("need indices")
    if (indices(dt, TRUE) %=>% sapply(identical, cols) %=>% any) {
        dt
    }
    else setindexv(dt, cols)
}
wk.setkey <- function(dt, cols = {
}, physical = TRUE) {
    if (!length(cols)) 
        cols <- colnames(dt)
    if (!is.character(cols)) 
        cols <- colnames(dt)[cols]
    if (haskey(dt) && (all(key(dt) == cols))) {
        dt
    }
    else setkeyv(dt, cols)
}
wk.fread <- function(..., cn = {
}, ot = ifelse(length(rn.cn), "df", "dt"), rn.cn = {
}, rn.cn.smart = 0, skip.comments = "#", tmpdir = hs.wj(dirname(hs.wj.temp()))) {
    A <- list(...)
    if ("skip" %not.in% names(A)) 
        A[["skip"]] <- 0
    fp1 <- hs.wj.temp()
    if ("cmd" %in% names(A)) {
        cmd1 <- A[["cmd"]]
        wj_tmp <- hs.wj.temp()
        on.exit(hs.wj.rm(wj_tmp), add = TRUE)
        cat(cmd1, file = wj_tmp)
        A[["cmd"]] <- .sys("bash", wj_tmp)
        (if (.Platform$OS.type == "unix") 
            system
        else shell)(paste("(", A[["cmd"]], ") >", fp1))
        A[["cmd"]] <- {
        }
        A[["file"]] <- fp1
        on.exit(unlink(fp1), add = TRUE)
    }
    else if ("text" %in% names(A)) {
        cat(A[["text"]], sep = "\n", file = fp1)
        A[["text"]] <- {
        }
        A[["file"]] <- fp1
    }
    fp1 <- A[[match("file", names(A), nomatch = 1)]]
    if (file.exists(fp1) && nzchar(skip.comments) && (A[["skip"]] == 0)) {
        v <- readLines(fp1, 1000)
        if (length(v)) {
            i <- which(substr(v, 1, 1) != skip.comments)[1] - 1
            if (i) 
                A[["skip"]] <- i - hs.grepl(v[i], "\t")
        }
    }
    A[["tmpdir"]] <- tmpdir
    if (length(cn)) 
        A[["header"]] <- as.logical(cn)
    jg <- do.call(data.table::fread, A)
    if (hs.all(!length(rn.cn), rn.cn.smart, length(jg) > 1, identical(unname(which(sapply(jg, is.character))), 1L))) 
        rn.cn <- 1
    switch(ot, dt = jg, m = wk.dt.2matrix(jg, rn.cn = rn.cn), df = wk.dt.2df(jg, rn.cn = rn.cn))
}
wk.fread0 <- function(...) wk.fread(..., cn = FALSE)
wk.fread1 <- function(...) wk.fread(..., cn = TRUE)
wk.fread.zip <- function(zip.wj, wj = {
}) {
    wk.fread(cmd = .sys("unzip -p", .sys.fpGood(zip.wj), .q(wj)))
}
wk.dt.setorderv <- function(..., na.last = TRUE) data.table::setorderv(..., na.last = na.last)
wk.dt.setnames <- function(dt, old = {
}, new = {
}, skip_absent = FALSE, fp = {
}, fp.sep = "\t") {
    A <- list(x = dt, skip_absent = skip_absent)
    if (hs.all(length(old), !length(new))) {
        new <- old
        old <- {
        }
    }
    if (hs.all(length(old) > 1, length(old) != length(new))) 
        hs.browser("38.05.27.235812")
    i <- which(is.na(new))
    if (length(i)) {
        if (length(old) == length(new)) 
            old <- old[-i]
        new <- new[-i]
    }
    if (hs.all(length(new) < ncol(dt), length(old) < length(new))) 
        old <- hs.switch(length(old), 0, seq_along(new), 1, seq(old, old + length(new) - 1))
    if (length(old)) 
        A[["old"]] <- old
    if (length(fp)) 
        new <- head(strsplit(readLines(fp, 1), fp.sep)[[1]], length(dt))
    if (length(new)) 
        A[["new"]] <- new
    do.call(data.table::setnames, A)
}
wk.dt.set <- function(x, i = NULL, j, value) {
    if (length(j)) {
        data.table::set(x = x, i = i, j = j, value = value)
    }
    else data.table::setDT(x)
}
wk.dt.setDT <- data.table::setDT
wk.dt.setDF <- data.table::setDF
wk.dt.fromList <- function(L, subset = if (length(passenger)) {
    seq_along(L) %not% {
        if (inherits(passenger, "character")) 
            which(names(L) %in% passenger)
        else passenger
    }
} else seq_along(L), passenger = {
}, ..., nm = c("group", if ((!length(dim(L[[1]]))) || ((length(dim(L[[1]])) == 2) && (ncol(L[[1]]) == 1))) "member"), by_row = {
}, list_name = 1) {
    if (inherits(subset, "character")) 
        subset <- match(subset, names(L))
    if (missing(passenger)) 
        passenger <- seq_along(L) %not% subset
    if (inherits(passenger, "character")) 
        passenger <- match(passenger, names(L))
    if (any(nm %in% names(L)[passenger])) 
        stop("2023_07_25_182415|列名重复：", nm %and% names(L)[passenger])
    by_row %<=>% if (length(x)) {
        as.logical(x)
    }
    else hs.any(length(dim(L[[subset[1]]])) == 2, is(L[[subset[1]]], "table"))
    lh <- sapply(subset, function(i) {
        x <- L[[i]]
        hs.switch(length(dim(x)), 2, nrow(x), if (by_row) 
            length(x) > 0
        else length(x))
    })
    i1 <- which(lh > 0)
    row_n <- sum(lh[i1])
    jg <- wk.dt(rep(hs.na(""), row_n))
    a <- L[[subset[i1[1]]]]
    lw <- hs.cond(is(a, "table"), 2, length(dim(a)) == 2, ncol(a), if (by_row) 
        length(a)
    else 1)
    ci_max <- max(lw) + 1
    a <- L[[subset[i1[which.max(lw)]]]]
    a <- hs.cond(inherits(a, "data.frame"), unname(lapply(a, hs.na)), is.atomic(a), local({
        b <- hs.i(L, subset[i1]) %=>% sapply(class)
        sapply(hs.i(L, subset[i1[match(unique(b), b)]]), hs.na)[1]
    }), inherits(a, "table"), list(hs.na(""), hs.na(1L)), inherits(a, "list"), hs.browser("38.07.09.214005"), as.list(rep(hs.na(a[1]), ci_max - 1)))
    jg[, `:=`(paste0("V", seq(2, ci_max)), a)]
    wk.dt.setnames(jg, nm)
    ri1 <- 1L
    for (i in i1) {
        a <- L[[subset[i]]]
        if (is.null(a)) 
            next
        if (inherits(a, "table")) 
            a <- wk.dt.as(a)
        if (length(dim(a)) == 2) {
            ri <- seq(ri1, ri1 + nrow(a) - 1)
            if (is(a, "data.frame")) {
                wk.dt.set(jg, ri, seq(2, ncol(a) + 1), a)
            }
            else for (j in 1:ncol(a)) wk.dt.set(jg, ri, j + 1L, a[, j])
        }
        else {
            if (by_row) {
                if (length(passenger)) 
                  hs.browser("2023.07.25.182749", debug = 0)
                ri <- ri1
                ci <- seq(2, 2 + length(a) - 1)
                for (i2 in seq_along(ci)) wk.dt.set(jg, ri, ci[i2], a[i2])
            }
            else {
                ri <- seq(ri1, ri1 + length(a) - 1)
                wk.dt.set(jg, ri, 2L, a)
                if (length(passenger)) {
                  wk.dt.set(jg, ri, names(L)[passenger], if (inherits(L, c("list"))) 
                    L[passenger]
                  else if (inherits(L, "data.table")) 
                    L[, .SD, .SDcols = passenger]
                  else if (inherits(L, "data.frame")) 
                    L[, passenger])
                }
            }
        }
        if (length(names(L))) 
            wk.dt.set(jg, ri, 1L, names(L)[subset[i]])
        ri1 <- ri[length(ri)] + 1L
    }
    L2 <- list(...)
    if (length(L2)) 
        hs.browser("38.05.20.150047")
    if (length(nm) == 1) {
        a <- L[[subset[i1[1]]]]
        nm2 <- if (is(a, "table")) {
            c("V1", "N")
        }
        else if (hs.len1(names(a), ncol(jg) - 1)) 
            names(a)
        if (length(nm2)) 
            wk.dt.setnames(jg, 2, nm2)
    }
    if (length(nm)) 
        wk.dt.setnames(jg, nm)
    if (!list_name) 
        wk.dt.set(jg, , 1L, {
        })
    if (!length(names(L))) 
        jg[, `:=`(1, {
        })]
    return(jg)
    a <- lapply(i1, function(i1) {
        hs.browser("38.05.20.141127")
        a <- do.call(wk.dt, c(if (by_row) c(list(names(L)[i1]), L[[i1]]) else list(names(L)[i1], L[[i1]]), lapply(L2, "[[", i1)))
        if (length(nm)) 
            wk.dt.setnames(a, nm)
        a
    })
    jg <- rbindlist(unname(a))
    if (length(nm)) 
        setnames(jg, seq_along(nm), nm)
    jg
}
wk.dt.fromMatrix <- function(m1, rn_name = "rn") {
    dt1 <- data.table::as.data.table(m1)
    if (length(rn_name)) {
        dt1[, `:=`(c(rn_name), rownames(m1))]
        wk.dt.col.reorder(dt1, rn_name, before = 1)
    }
    dt1
}
wk.dt.fromUTF16 <- function(wj1) {
    con1 <- file(wj1, encoding = "UTF-16")
    on.exit(close(con1), add = TRUE)
    fread(text = readLines(con1))
}
wk.dt.2matrix <- function(dt, cn = seq.int(ncol(dt)), cn.ms = {
}, rn.cn = if (inherits(dt[[1]], "character")) 1, cn.ri = {
}, cn2rm = {
}, df = 0) {
    hs1 <- if (df) 
        as.data.frame
    else as.matrix
    if (length(cn.ms)) 
        cn <- hs.grep(names(dt), cn.ms)
    if (length(cn2rm)) {
        if (is.character(cn2rm)) 
            cn2rm <- match(cn2rm, names(dt))
        cn <- cn %rm% cn2rm
    }
    if (length(rn.cn)) {
        if (is.character(rn.cn)) 
            rn.cn <- match(rn.cn, names(dt))
        cn <- cn %not% rn.cn
    }
    jg <- if (length(cn.ri)) 
        dt[-cn.ri, hs1(.SD), .SDcols = c(cn)]
    else dt[, hs1(.SD), .SDcols = c(cn)]
    if (length(cn.ri)) {
        colnames(jg) <- if (length(rn.cn)) 
            dt[cn.ri, as.character(.SD), .SDcols = -rn.cn]
        else dt[cn.ri, as.character(.SD)]
    }
    if (length(rn.cn)) {
        rownames(jg) <- if (length(cn.ri)) 
            dt[[rn.cn]][-cn.ri]
        else dt[[rn.cn]]
    }
    return(jg)
    if (length(rn.cn)) {
        if (is.numeric(rn.cn)) {
            cn <- cn[-rn.cn]
            rn.cn <- names(dt)[rn.cn]
        }
        else cn <- cn %rm% rn.cn
    }
    if (length(cn2rm)) {
        hs.browser("37.12.10.210528")
        if (is.numeric(cn2rm)) 
            cn2rm <- names(dt)[cn2rm]
        cn <- cn %rm% cn2rm
    }
    hs.browser("37.12.10.210531")
    X <- (if (df) 
        as.data.frame
    else as.matrix)(dt[, .SD, .SDcols = c(cn)])
    if (length(rn.cn)) 
        rownames(X) <- dt[[rn.cn]]
    X
}
wk.dt.2df <- function(dt, ...) {
    wk.dt.2matrix(dt, ..., df = 1)
}
wk.dt.widen <- function(dt, rn = names(dt)[1], cn = names(dt)[2], vn = names(dt)[3], vn.hs = if (is.numeric(dt[[vn]])) sum else identity, rn.sort = dn.sort, cn.sort = dn.sort, dn.sort = 0) {
    for (on1 in c("rn", "cn", "vn")) hs.update(on1, names(dt)[x], is.numeric)
    if (!inherits(dt, "data.table")) 
        dt <- wk.dt.as(dt)
    dt[, {
        dn <- list(unique(get(rn)) %=>% if (rn.sort) sort(x) else x, unique(get(cn)) %=>% if (rn.sort) sort(x) else x) %=>% lapply(paste)
        df <- lapply(dn[[2]], function(c1) {
            a <- dt[list(dn[[1]], c1), .SD, .SDcols = c(rn, vn), on = c(rn, cn)]
            a[, vn.hs(get(vn)), by = c(rn)][[2]]
        }) %=>% hs.df
        dimnames(df) <- dn
        df
    }]
}
wk.dt.tapply <- function(dt, element.cn = names(dt)[1], name.cn = names(dt)[2], hs = identity, ..., hs.element = identity, hs.name = identity, hs1 = {
}, hs2 = {
}) {
    a1 <- if (length(hs1)) 
        hs1(dt)
    else hs.element(dt[[element.cn]])
    a2 <- if (length(hs2)) 
        hs2(dt)
    else hs.name(dt[[name.cn]])
    tapply(a1, a2, hs, ...)
}
wk.fst.write <- function(sj, wj, ...) {
    write_fst(x = sj, path = wj, ..., compress = 100)
    wj
}
wk.fst.write.fromWj <- function(wj1, wj2 = if (file.exists(wj1)) hs.wjm(wj1, ext = "fst", extN = ifelse(hs.grepl(wj1, "\\.(tsv|csv|txt)\\.gz$"), 2, 1)), uniform_encoding = TRUE, ow = 0, smart = 1, hs1 = identity, ...) {
    if (!length(wj2)) 
        stop("37_02_05_150526")
    if (file.exists(wj2)) {
        if (smart && (!ow)) 
            ow <- file.mtime(wj2) < file.mtime(wj1)
        if (!ow) 
            return(wj2)
    }
    sj <- if (file.exists(wj1)) {
        wk.fread(wj1, ...)
    }
    else wk.fread(cmd = wj1, ...)
    wk.fst.write(hs1(sj), wj2, uniform_encoding = uniform_encoding)
}
ranges.hs.vector <- function(v1) {
    v1 <- unique(sort(v1))
    d1 <- diff(v1)
    i1 <- which(d1 > 1)
    t(matrix(c(v1[1 %or% (i1 + 1)], v1[i1 %or% length(v1)]), ncol = 2))
}
wk.fst.read <- function(wj, ri = NULL, cn = NULL, from = 1, to = NULL, as.data.table = TRUE, rbind = 1, dt = 1, ...) {
    ft <- fst(wj)
    if (length(cn) && is.numeric(cn)) {
        cn <- colnames(ft)[cn]
    }
    jg <- hs.switch(c(length(cn), length(ri)) > 0, rep(TRUE, 2), ft[ri, cn], c(TRUE, FALSE), ft[, cn], c(FALSE, TRUE), ft[ri, cn], rep(FALSE, 2), ft[TRUE])
    if (dt && is.data.frame(jg)) 
        setDT(jg)
    return(jg)
    if (identical(cn, "")) {
        return(wk.fst.read(wj, ri = 1))
    }
    if (length(ri)) {
        ri <- ranges.hs.vector(ri)
        jg <- apply(ri, 2, function(ri) {
            wk.fst.read(wj, cn = cn, from = ri[1], to = ri[2], as.data.table = as.data.table, ...)
        })
        if (rbind) {
            rbindlist(jg)
        }
        else jg
    }
    else {
        read_fst(path = wj, columns = cn, from = from, to = to, as.data.table = as.data.table)
    }
}
wk.dt.rbind <- function(..., use.names, U = FALSE, fill = "auto") {
    L <- list(...)
    nc <- sapply(L, . %=>% {
        a <- ncol(x)
        if (is.null(a)) 
            a <- length(x)
        a
    })
    ref_i <- which.max(nc)
    ref_cn <- wk.dt.copy(names(L[[ref_i]]))
    if (inherits(L[[ref_i]], "data.table")) {
        wk.dt.reset_name(L[[ref_i]])
        on.exit(wk.dt.setnames(L[[ref_i]], ref_cn), add = TRUE)
    }
    else {
        names(L[[ref_i]]) <- paste0("V", seq_along(L[[ref_i]]))
    }
    if (missing(use.names)) 
        use.names <- all(sapply(L[-ref_i], function(x) names(x) %all.in% ref_cn))
    if (missing(fill)) 
        fill <- TRUE
    if (wk.dt.uniq_n(nc) > 1 && !fill) 
        stop("2023_08_03_023343|不等宽")
    if (use.names) {
        for (li in seq_along(L)[-ref_i]) {
            a <- names(L[[li]])
            names(L[[li]]) <- paste0("V", match(a, ref_cn))
        }
    }
    else {
        for (li in seq_along(L)[-ref_i]) {
            names(L[[li]]) <- paste0("V", seq_along(L[[li]]))
        }
    }
    jg <- do.call(data.table::rbindlist, list(L, use.names = use.names, fill = fill))
    wk.dt.setnames(jg, ref_cn)
    if (U) 
        unique(jg)
    else jg
}
wk.dt.reshape <- function(dt1, row_name_name, column_name_name, value_function) {
    cn <- dt1[, get(column_name_name) %=>% hs.clean(U = 1)]
    rn <- dt1[, get(row_name_name) %=>% hs.clean(U = 1)]
    m1 <- matrix(NA, length(rn), length(cn), dimnames = list(rn, cn))
    for (i1 in dt1[, seq.int(.N)]) {
        if (0) {
            i1 <- 1426
            message(i1)
            if (i1 == 1426) 
                hs.browser("37.05.12.144909")
        }
        rn1 <- dt1[i1, get(row_name_name)]
        if (is.na(rn1)) 
            next
        cn1 <- dt1[i1, get(column_name_name)]
        m1[rn1, cn1] <- value_function(dt1[i1])
    }
    m1
}
wk.dt.clean <- function(dt1) {
    ci2rm <- sapply(dt1, wk.dt.uniq_n) %=>% seq_along(x)[x %in% 1]
    if (length(ci2rm)) 
        dt1[, `:=`(c(ci2rm), {
        })]
    dt1
}
wk.dt.replace <- function(dt, hs1, hs2 = function(...) TRUE, ci = {
}, ci_pattern = {
}) {
    if (!length(ci)) 
        ci <- seq_along(dt)
    if (is.character(ci)) 
        ci <- match(ci, names(dt))
    if (length(ci_pattern)) 
        ci <- ci %or% hs.grep(names(dt), ci_pattern)
    for (i1 in ci) {
        if (!hs2(dt[[i1]])) 
            next
        cn1 <- names(dt)[i1]
        cn2 <- paste0(cn1, ".38_03_16_210715")
        setnames(dt, i1, cn2)
        dt[, `:=`(c(i1), hs1(get(cn2)))]
        setnames(dt, i1, cn1)
    }
}
wk.dt.reset_name <- function(dt) {
    data.table::setnames(dt, paste0("V", seq_along(dt)))
}
wk.dt.1to1 <- function(dt, c1 = 1, c2 = 2, side = 1:2) {
    c1i <- if (is.character(c1)) 
        match(c1, names(dt))
    else c1
    c2i <- if (is.character(c2)) 
        match(c2, names(dt))
    else c2
    tag1 <- "_38_03_24_111356"
    c1 <- paste0(names(dt)[c1i], tag1)
    wk.dt.setnames(dt, c1i, c1)
    c2 <- paste0(names(dt)[c2i], tag1)
    wk.dt.setnames(dt, c2i, c2)
    dt0 <- dt
    for (i1 in side) {
        cn1 <- c(c1, c2)[i1]
        cn2 <- c(c1, c2)[-i1]
        dt <- evalbq(dt[, {
            if (wk.dt.uniq_n(.(as.symbol(cn2))) == 1) 
                .SD
        }, by = c(cn1)])
    }
    wk.dt.col.reorder(dt, c(c1, c2), c(c1i, c2i))
    wk.dt.setnames(dt0, c1, hs.sub(c1, tag1))
    wk.dt.setnames(dt0, c2, hs.sub(c2, tag1))
    wk.dt.setnames(dt, c1, hs.sub(c1, tag1))
    wk.dt.setnames(dt, c2, hs.sub(c2, tag1))
}
wk.dt.collapse <- function(dt, cn2do = {
}, by = {
}, sep = ";", keep_order = 1, U = 0, expand = 0) {
    if (!length(cn2do)) 
        cn2do <- names(dt) %not% by
    if (is.numeric(cn2do)) {
        cn2do.ci <- cn2do
        cn2do <- names(dt)[cn2do.ci]
    }
    else cn2do.ci <- match(cn2do, names(dt))
    if (!length(by)) 
        by <- (names(dt) %not% cn2do)[1]
    if (expand) 
        dt <- wk.dt.expand(dt, by, split = sep)
    sep.38_09_04_232534 <- sep
    if (dt[, wk.dt.uniq_n(.SD) == .N, .SDcols = c(by)]) 
        return(dt)
    jg <- if (U) {
        dt[, lapply(.SD, function(x) {
            paste(hs.str_as(unique(x)), collapse = sep.38_09_04_232534)
        }), by = c(by)]
    }
    else dt[, lapply(.SD, function(x) {
        paste(hs.str_as(x), collapse = sep.38_09_04_232534)
    }), by = c(by)]
    wk.dt.setnames(jg, length(by) + 1, cn2do)
    if (keep_order) 
        wk.dt.col.reorder(jg, names(dt) %and% names(jg))
    jg
}
wk.dt.db <- function(x, i = {
}, inv = FALSE) {
    if (!inherits(x, "data.table")) {
        if (inherits(x, "character") && (!length(dim(x)))) {
            x <- data.table::data.table(x[seq(1, length(x), by = 2)], x[seq(2, length(x), by = 2)])
        }
        else hs.browser("2023.07.29.120726", debug = 0)
    }
    if (inv) {
        data.table::setcolorder(x, rev(names(x)))
        wk.dt.reset_name(x)
    }
    cn1 <- names(x)[1]
    if (!identical(data.table::key(x), cn1)) 
        data.table::setkeyv(x, cn1)
    if (length(i)) {
        return(x[i, .SD[[2]]])
    }
    else return(x)
}
hs.hs.key2wj <- function(uid, rm = 0) {
    hs.name <- ls(pattern = sprintf("%s$", uid), envir = .GlobalEnv)
    if (length(hs.name)) {
        hs <- get(hs.name, envir = .GlobalEnv)
        if (rm) {
            db <- environment(hs)[["db"]]
            for (uid1 in db) unlink(.mfp(uid1))
            unlink(environment(hs)[["db.wj.save2"]])
            rm(list = hs.name, envir = .GlobalEnv)
        }
        else hs
    }
    else {
        hs.name <- sprintf("hs.key2wj.%s", uid)
        hs <- local({
            db <- c()
            db.wj.save2 <- .mfp(sprintf("%s.rd", uid))
            cat(sprintf("database file: %s\n", db.wj.save2))
            if (file.exists(db.wj.save2)) 
                db <- hs.load(db.wj.save2)
            save <- 0
            function(...) {
                kw <- as.character(c(...))
                if (!length(kw)) 
                  return(db)
                for (kw1 in kw) if (!(kw1 %in% names(db))) {
                  save <<- 1
                  db[kw1] <<- .muid4machine()
                }
                if (save) {
                  save(db, file = db.wj.save2)
                  save <<- 0
                }
                db[unlist(kw)]
            }
        })
        assign(hs.name, hs, envir = .GlobalEnv)
        get(hs.name, envir = .GlobalEnv)
    }
}
wk.sync.code.r.2server <- local({
    mt <- list()
    mt.wj <- .mfp("34.10.02.060845.rd", hs.home.expand("~/org/db"))
    if (file.exists(mt.wj)) 
        mt <- hs.load(mt.wj)
    function(where = wk.ssh, only = {
    }, rp = 0) {
        hs.init()
        wj <- c(hs.wj(wk.dm.wjj, c("init.r", "all.r.rd")), if (rp) hs.home.expand("~/rjk/ln-s/.Rprofile"))
        did <- 0
        if (length(only)) 
            return(switch(only, wj = wj))
        for (i1 in seq_along(wj)) {
            wj1 <- wj[i1]
            if (length(mt[[wj1]]) && (mt[[wj1]] >= file.mtime(wj1))) {
            }
            else {
                r1 <- wk.rsync.out.same_place(wj1, L = 1, ssh = where)
                mt[[wj1]] <<- Sys.time()
                did <- 1
            }
        }
        if (did) {
            save(mt, file = mt.wj)
            if (length(only)) 
                switch(only, wj = wj2sync)
        }
        else hs.cat("wk.sync.code.r.2server: no need to sync")
    }
})
hs.r.2server.4jdi <- function(wj1 = hs.home.expand("~/tmp/jdi.r")) {
    if (nchar(wj1) == 0) 
        wj1 <- readline("请输入文件路径：\n")
    wk.rsync.out.same_place(wj1)
}
hs.dm.2svr <- function(wjj = hs.home.expand("~/org/dm"), where = wk.ssh, only = {
}) {
    wjj_sub <- hs.wjj.ls(wjj) %=>% x[dir.exists(x)]
    if (length(wjj)) {
        if (length(only)) {
            switch(only, wj = wjj_sub)
        }
        else wk.rsync.out(wjj_sub, wjj, ssh = where)
    }
}
.j <- hs.jdi <- function(local = 1, up = parent.frame()) {
    wj <- switch(tolower(Sys.info()["sysname"]), linux = "/dev/shm/wk/39_08_27_220959.txt", "~/.db/buffer4tmux.txt") %=>% hs.getTextFile
    env1 <- if (local) 
        up
    else .GlobalEnv
    hs.source.wj(wj, env = env1)
}
hs.xm.do <- local({
    hs1 <- function(mb = {
    }, ff = {
    }) {
        if (length(mb)) {
            cat("The aim is ", mb, "\n", sep = "")
            if (length(ff)) {
            }
            else hs1(mb = mb, ff = readline("What is the method?\n"))
        }
        else hs1(mb = readline("What is the aim?\n"))
    }
    function(mb = {
    }, ff = {
    }) repeat {
        hs1()
    }
})
hs.xm.template <- function(uid) {
    wjj0 <- hs.wjj(.mfp(uid))
    hs.wj(wjj0, "main.r")
    "mb,ff,sj,dm,jg,gj,wd"
}
hs.source.jdi <- function(wj1 = hs.home.expand("~/tmp/jdi.r"), ...) {
    if (!file.exists(wj1)) {
        sep <- formals(hs.source.part)[["sep"]]
        cat(sep, sep = "\n", file = hs.wj(wj1))
    }
    hs.source.part(wj1, ...)
}
hs.jdi.source <- function(wj1, i1, i2, envUp = .GlobalEnv) {
    if (file.exists(wj1)) {
        nr <- readLines(wj1)
        if (missing(i2)) {
            i2 <- i1
            if (is.character(i2)) {
                mi <- which(nr == i1)
                if (!hs.len1(mi, 2)) 
                  stop("please check the logic @ 34.11.26.005157 @ ", hs.home.expand("~/org/dm/r/xm.r"))
                i1 <- mi[1]
                i2 <- mi[2]
            }
        }
        else {
            if (is.character(i1)) {
                i1 <- match(i1, nr)
            }
            if (is.character(i2)) {
                i2 <- match(i2, nr)
            }
        }
        if (i1 > i2) 
            stop("please check the logic @ 34.11.26.005442 @ ", hs.home.expand("~/org/dm/r/xm.r"))
        nr <- head(nr, i2)
        nr <- tail(nr, -i1 + 1)
        eval(parse(text = nr), envUp)
    }
}
hs.hsgly <- local({
    uid.pattern <- wk.muid.pattern
    uid_end.pattern <- hs.p0(uid.pattern, "$")
    env.lt <- "env.loadtime.33.04.27.080101"
    eval(bquote({
        if (!exists(.(env.lt))) 
            .(env.lt) <- new.env()
    }), .GlobalEnv)
    {
        r.wj.wjj.remote <- hs.home.expand("~/tmp2/dm")
        uid.wj.wj <- .mfp("36.03.04.104650.tsv.gz", hs.home.expand("~/tmp2")) %=>% hs.wj
        if (0) {
            if (!hs.machine.isLocal()) 
                uid.wj.wj <- hs.sub(uid.wj.wj, "~", r.wj.wjj.remote)
        }
        wj.wj <- {
        }
        wj.wj.scan_time <- {
        }
        uid.wj.wj.hs <- function(uid.wj = {
        }) {
            wj <- .wk("rg --no-line-number --no-filename -z", .q(paste0("(?i)", uid.pattern, "[^/]*\\.r(\\.[gx]z)?$")), wj.wj, intern = 1)
            wj.wj.scan_time <<- Sys.time()
            if (0) {
                wj <- hs.switch(Sys.info()["nodename"], "node9", "/dev/shm/wk", "~/1") %=>% normalizePath %=>% .wk("ag -g", .q(paste0("(?i)", uid.pattern, ".*\\.r(\\.[gx]z)?$")), .q(x), intern = 1) %=>% x[hs.grep(basename(x), paste0("(?i)", uid.pattern, ".*\\.r(\\.[gx]z)?$"))]
            }
            if (file.exists(r.wj.wjj.remote)) {
                wj <- c(wj, hs.wjj.ls(r.wj.wjj.remote, hs.p0(uid.pattern, "\\.r(\\.[gx]z)?$"), 1))
            }
            uid <- hs.re(basename(wj), uid.pattern)
            uid <- hs.gsub(uid, "[^0-9]", "_")
            if (length(uid) != length(wj)) 
                hs.browser("39.05.20.113718", debug = 0)
            tbl <- wk.dt(uid = uid, wj = wj)
            tbl <- tbl[, .(wj = if (.N > 1) 
                wj[which.max(file.mtime(wj))]
            else wj), by = uid]
            if (!identical(uid.wj, tbl)) {
                wk.dt.fwrite(tbl, uid.wj.wj)
                uid.wj <<- tbl
            }
        }
        uid.wj <- {
        }
        do.find.wj <- function(uid1) {
            uid1 <- hs.gsub(uid1, "[^0-9]", "_")
            jg <- uid.wj[hs.match(uid1, uid), hs.home.expand(wj)]
            if (length(jg) && (!file.exists(jg))) {
                a <- parse(uid.wj[hs.match("39_01_08_150201", uid), wj])
                eval(a[[1]])()
                uid.wj.wj.hs()
                jg <- uid.wj[hs.match(uid1, uid), wj]
            }
            jg
        }
    }
    function(wjm = {
    }, obj = {
    }, local = 0, ext = "R", wjj0 = wk.dm.wjj, env1 = .GlobalEnv, PO = 0, path.only = PO, blm.only = 0, load = 0, reload = 0, call = 0, argList = list(), rename = 0, bl = 1, parse.network = 0, less = 0, trim = 0, keep.mtime = 0, smart = 1, update.uid = !length(wjm)) {
        if (length(obj)) 
            return(uid.obj(wjm, obj))
        if (update.uid) {
            uid.wj.wj.hs()
            return()
        }
        uid <- hs.re(wjm, uid_end.pattern)
        if ((!hs.machine.isLocal()) && ("util.390816171110" %in% .packages(1))) {
            uid %<=>% hs.gsub("[^0-9]", "_")
            return(eval(call("::", "util.390816171110", uid)))
        }
        if (!length(uid.wj)) 
            uid.wj.wj.hs()
        wj <- do.find.wj(uid)
        if (!length(wj)) 
            stop(hs.p("####", wjm, "is not found. 34.07.04.195822 @ ", hs.home.expand("~/org/dm/r/xm.r"), " ####"))
        if (path.only) 
            return(wj)
        if (less) {
            .less(wj)
            return(wj)
        }
        blm <- hs.ls(env1, pattern = uid)
        if (length(blm)) {
            if (length(blm) > 1) {
                blm.lt_max.i1 <- which.max(sapply(blm, function(x1) env1[[env.lt]][[x1]]))
                rm(list = blm[-blm.lt_max.i1], envir = env1)
                blm <- blm[blm.lt_max.i1]
            }
            if (blm.only) 
                return(blm)
            blm.loadtime <- env1[[env.lt]][[blm]]
            if ((!length(blm.loadtime)) || (difftime(file.mtime(wj), blm.loadtime, units = "secs") > 1.5)) {
                message("Need to (re)build the function ", blm, " due to its lagged mtime.")
                blm <- {
                }
            }
        }
        if ((!length(blm)) || reload) {
            nr <- unlist(strsplit(hs.readLines.robust(wj), "[\n\r]"))
            blm <- {
                blm1 <- hs.sub(nr[1], "[# ]+")
                if (blm1 == "") 
                  blm1 <- hs.sub(basename(wj), paste0("[_\\.]?", wk.muid.pattern, ".*"))
                if (blm1 == "") 
                  blm1 <- basename(dirname(wj))
                paste0(blm1, ".", uid)
            }
            if (blm.only) 
                return(blm)
            src <- parse(text = nr)
            if (!identical(env1, .GlobalEnv)) 
                hs.browser("37.12.21.173759")
            eval(bquote({
                .(blm) <- wk.33.06.24.202656.v2.auto(.(src[[1]]))
                .(as.symbol(env.lt))[[.(blm)]] <- .(if (file.exists(wj)) 
                  file.mtime(wj)
                else Sys.time())
            }), envir = env1)
        }
        R <- env1[[blm]]
        if (0) {
            hs.env(R)[["env2.33.06.24.181217"]][["body.34.07.07.065614"]]
        }
        t1 <- file.mtime(wj)
        t2 <- -Inf
        uid1 <- unique(unlist(hs.gre(hs.readLines.robust(wj), uid.pattern)))
        if (length(uid1)) {
            t2.wj <- unlist(sapply(uid1, do.find.wj))
            if (length(t2.wj)) 
                t2.wj <- t2.wj[file.exists(t2.wj)]
            if (length(t2.wj)) 
                t2 <- max(file.mtime(t2.wj))
        }
        if (t2 > t1) {
            if (smart) {
                formals(R)[["..new"]] <- 1
                hs.Sys.setFileTime(wj, t2)
            }
            else if (!keep.mtime) 
                hs.Sys.setFileTime(wj, t2)
        }
        if (0) {
            hs.browser("38.04.20.222242")
            ls(hs.env(R))
            hs.env(R)[["env2.33.06.24.181217"]][["body.34.07.07.065614"]]
        }
        R
    }
})
hs.hsgly.bak <- function(wjm, bds, local = 0, ext = "R", wjj0 = ieiwk("wjj.dm"), env1 = .GlobalEnv, PO = 0, path.only = PO, blm.only = 0, load = 0, reload = 0, call = 0, argList = list(), rename = 0, bl = 1, parse.network = 0, less = 0, trim = 0, keep.mtime = 0, smart = 1) {
    uid <- hs.re(wjm, "\\d+\\.\\d{2}\\.\\d{2}\\.\\d{6}$")
    wj <- .mfp(wjm, wjj0, act = "find", uid.pos = "tail")
    wj2 <- hs.wjm(.mfp(wjm, wjj0, uid.pos = "tail"), add = ext)
    if (length(wj) > 1) 
        wk.browser("34.10.31.172013 @ ~/org/dm/r/xm.r")
    if (missing(bds)) {
        if (!length(wj)) 
            wj <- wj2
    }
    else {
        if (!hs0.seq.eql(wj, wj2)) {
            if (length(wj)) 
                file.rename(wj, wj2)
            wj <- wj2
        }
    }
    if (path.only) 
        return(wj)
    if (((!length(wj)) || (!file.exists(wj))) && missing(bds)) {
        wj3 <- .mfp(wjm, wk.dm.wjj, act = "find", uid.pos = "tail")
        wj3 <- hs.grep(wj3, "\\.r$", v = 1)
        if (length(wj3)) {
            if (length(wj3) > 1) {
                wk.browser("34.10.13.105040")
            }
            else source(wj3)
        }
        else {
            hs.browser("35.11.26.104320")
            stop(paste("####", wj, "is not found. 34.07.04.195822 @ ~/org/dm/r/function.common.R ####"))
        }
    }
    if (less) 
        return(.less(wj))
    cat2wj <- 0
    if (!missing(bds)) {
        bds.nr <- format(substitute(bds))
        if (trim) 
            bds.nr <- trim(bds.nr)
        cat2wj <- 1
        if (file.exists(wj)) {
            wj.nr <- readLines(wj, encoding = "UTF-8")
            if (hs0.seq.eql(wj.nr, bds.nr)) 
                cat2wj <- 0
        }
        if (cat2wj) {
            if (keep.mtime) 
                mt1 <- file.mtime(wj)
            cat(bds.nr, file = wj, sep = "\n")
            if (keep.mtime) {
                hs.say("hs.hsgly: keeping the previous mtime of", wj)
                Sys.setFileTime(wj, mt1)
            }
        }
    }
    wj.blm <- hs.wjm(wj, dn = 0, ext = 0)
    if (wj.blm == uid) {
        if (wjm != uid) {
            wj.blm <- wjm
        }
        else warning(paste("####", wjm, "is not meaningful, 34.07.04.195811 @", hs.home.expand("~/org/dm/r/xm.r"), " ####"))
    }
    blm <- ls(env1, pattern = uid, all.names = TRUE, sorted = FALSE)
    if (!length(blm)) 
        blm <- wj.blm
    if (blm.only) 
        return(blm)
    env.lt <- "env.loadtime.33.04.27.080101"
    if (!exists(env.lt, env1, inherits = FALSE)) 
        env1[[env.lt]] <- new.env(parent = env1)
    blm.loadtime <- env1[[env.lt]][[blm]]
    new <- 0
    if (missing(bds)) {
        t1 <- file.mtime(wj)
        t2 <- -Inf
        uid1 <- unique(unlist(hs.gre(readLines(wj), "\\b\\d+\\.\\d{2}\\.\\d{2}\\.\\d{6}\\b")))
        if (length(uid1)) {
            t2.wj <- sapply(uid1, hs.hsgly, path.only = 1)
            t2.wj <- t2.wj[file.exists(t2.wj)]
            if (length(t2.wj)) 
                t2 <- max(file.mtime(t2.wj))
        }
        if (t2 > t1) {
            if (smart) {
                new <- load <- 1
                Sys.setFileTime(wj, t2)
            }
            else if (!keep.mtime) 
                Sys.setFileTime(wj, t2)
        }
    }
    if (exists(blm, env1, inherits = FALSE)) {
        if ((!length(blm.loadtime)) || (difftime(file.mtime(wj), blm.loadtime, units = "mins") > 1)) 
            load <- 1
    }
    else {
        if (bl) {
            load <- 1
        }
        else wk.browser("34.08.12.223657")
    }
    if (load || reload) {
        if (!file.exists(wj)) 
            stop("no file found. 34.07.13.054404 @ ", hs.home.expand("~/org/dm/r/xm.r"))
        bds <- eval(parse(wj, encoding = "UTF-8", keep.source = FALSE))
    }
    if (0) {
        if (length(bds)) {
            bds.str <- format(bds)
            .less(wj)
            if (any(grepl("hs.dmgly(\"sj.34.01.09.135117\")()", bds.str, fix = TRUE))) {
                hs.browser("34.08.12.233818")
                sys.calls()
            }
        }
    }
    if (cat2wj || new || (!missing(bds))) {
        bds <- bquote(wk.33.06.24.202656.v2.auto(.(bds), new = .(new)))
        R <- eval(bds, env1)
        if (bl) {
            env1[[env.lt]][[blm]] <- Sys.time()
            env1[[blm]] <- R
        }
        R
    }
    else {
        if (missing(bds)) {
            R <- env1[[blm]]
            if (formals(R)[["..new"]]) 
                formals(R)[["..new"]] <- 0
            R
        }
    }
}
hs.dmgly <- function(uid = {
}, wj = {
}, env1 = parent.frame()) {
    if (length(uid)) {
        wj %<=>% (`%or%`(sapply(uid, function(uid1) {
            uid1 <- hs.re(basename(uid1), environment(hs.hsgly)[["uid.pattern"]])
            hs.hsgly(uid1, PO = 1)
        })))
    }
    for (wj1 in wj) source(wj1, local = env1)
}
hs.sjgly <- function(uid1, bf = {
}, ot = "wjj2") {
    hs.hsgly()
    uid1 <- hs.re(basename(uid1), wk.muid.pattern)
    wjj0 <- hs.hsgly(uid1, PO = 1) %=>% hs.dirname
    if (ot == "wjj0") 
        return(wjj0)
    wjj0 %<=>% normalizePath
    wjj1 <- wjj0
    repeat {
        wj1 <- hs.wjj.ls(wjj1, paste0("^根?文件夹\\.", wk.muid.pattern, "\\.r(\\.gz)?$"))
        if (length(wj1) == 1) 
            break
        if (length(wj1) > 1) 
            stop("[38_11_27_001548]")
        wjj1 <- dirname(wjj1)
        if (wjj1 %=>% (x == dirname(x))) 
            stop("[39_03_02_160542|没有上层文件夹]", wjj0)
    }
    wjj2 <- wjj0_jg <- hs.hsgly(hs.re(basename(wj1), wk.muid.pattern))()
    if (nchar(wjj0) > nchar(wjj1)) 
        wjj2 <- paste0(wjj0_jg, substring(wjj0, nchar(wjj1) + 1))
    wj_flag <- hs.fp(wjj2, paste0("代码号_", uid1))
    if (0) {
        wj_flag %=>% {
            if (file.exists(x) && (!file.is_symlink(x))) 
                hs.wj.rm(x)
        }
    }
    if (length(bf)) {
        if (inherits(bf, c("numeric", "logical"))) 
            bf <- hs.uid.fp()
        wjj_bf <- hs.fp(wjj2, "备份", bf)
        wj2bf <- hs.wjj.ls(wjj2, wjj = 1, wj = 1) %=>% x[-hs.grep(basename(x), "^(代码号_|备份)")]
        hs.wj.move(wj2bf, wjj_bf)
        return()
    }
    if ((!file.exists(wj_flag))) 
        local({
            a <- hs.wjj.ls(wjj0_jg, uid1, R = 1)
            hs.switch(length(a), 1, {
                wjj_old <- dirname(a)
                if (dir.exists(wjj2)) {
                  for (wj1 in hs.wjj.ls(wjj_old, A = 1, wjj = 1, wj = 1)) hs.wj.move(wj1, wjj2)
                  hs.wj.rm(wjj_old, 1)
                }
                else {
                  wjj_temp <- paste0(wjj_old, ".39_05_15_112003")
                  hs.wj.move(wjj_old, wjj_temp)
                  hs.wj.move(wjj_temp, hs.wj(wjj2))
                }
            }, 0, {
                a <- environment(hs.hsgly)[["uid.wj"]][hs.grep(uid, hs.gsub(uid1, "_", ".")), wj]
                {
                  if (0) {
                    file.symlink(a, wj_flag)
                  }
                  cat(normalizePath(a), file = hs.wj(wj_flag))
                }
            }, hs.browser("38.11.27.083724", debug = 0))
        })
    if (file.exists(wj_flag)) 
        local({
            fp_real <- environment(hs.hsgly)[["uid.wj"]][hs.grep(uid, hs.gsub(uid1, "_", ".")), wj %=>% normalizePath]
            fp_recorded <- readLines(wj_flag)
            if (fp_recorded != fp_real) 
                cat(fp_real, file = wj_flag)
        })
    wjj2
}
uid.obj <- function(uid1, on1, ...) {
    if (0) {
        uid1 <- "36_12_21_002840"
        on1 <- "ver.hs.org1"
    }
    obj1 <- hs.getter(hs.hsgly(uid1), on1)
    if (...length()) {
        obj1(...)
    }
    else obj1
}
uid.hs.str <- function(s1) {
    hs.re(s1, wk.muid.pattern)
}
hsm.hs.uid <- function(uid) {
    wj <- hs.hsgly(uid, PO = 1)
    if (hs.len1(wj) && file.exists(wj)) {
        l1 <- readLines(wj, 1)
        paste0(hs.sub(l1, "^#+ *"), ".", uid.hs.str(uid))
    }
}
hs.hsm.updator <- function() {
    if (any(hs.grepl(hs.getter(sjk.uid.hsm, "hsm"), "[\"']"))) 
        stop("36_11_11_230733 警告：函数名带引号")
    if (min(nchar(hs.sub(hs.getter(sjk.uid.hsm, "hsm"), wk.muid.pattern))) < 3) 
        stop("36_11_11_233445 函数名太短")
    wj.V <- .wk("ag -g \"\\.(r|org)$\"", hs.home.expand("~/org"), intern = 1)
    for (wj1 in wj.V) {
        changed <- 0
        nr <- readLines(wj1)
        if (!length(nr)) 
            next
        re_match <- gregexpr("(?<=hs\\.hsgly\\() *[\"'][^\"']+", nr, perl = TRUE)
        li <- which(sapply(re_match, function(.) hs.len1(.) && (. > 0)))
        for (li1 in rev(li)) {
            matched <- regmatches(nr[li1], re_match[[li1]])
            hsm <- hs.clean(sjk.uid.hsm(matched))
            if (!length(hsm)) 
                next
            if (hs.re(matched, "[^\"']+$", no_match = "") != hsm) {
                changed <- 1
                regmatches(nr[li1], re_match[[li1]]) <- hs.sub(matched, "[^\"']+$", hsm)
            }
        }
        if (changed) {
            hs.cat(wj1)
            cat(nr, sep = "\n", file = wj1)
        }
    }
}
hs.uid_hs <- function() {
    line.V <- .wk("ag -G \"(?i)\\.(r|org)$\" --nofilename \"hs\\.hsgly\\(\"", hs.home.expand("~/org"), intern = 1)
    str_raw <- hs.gre(hs.clean(line.V), "(?<=hs\\.hsgly\\() *[\"'][^\"']+") %=>% unlist %=>% unique
    unique(uid.hs.str(str_raw))
}
sjk.uid.hsm <- local({
    hsm <- c()
    hsm.hs <- function() {
        uid.V <- hs.uid_hs() %=>% sort
        hsm <- hs.lapply(uid.V, hsm.hs.uid)
        if (max(sapply(hsm, length)) > 1) 
            stop("36_11_11_224307")
        hsm <<- unlist(hsm)
    }
    function(uid) {
        if (!length(hsm)) 
            hsm.hs()
        hsm[uid.hs.str(uid)]
    }
})
wk.33.06.24.203348 <- function(env = parent.frame()) {
    evalq({
        env1.33.06.24.181214 <- environment()
        env2.33.06.24.181217 <- new.env()
    }, env)
}
ieiwk.make.arg.list <- function(...) {
    args <- list(...)
    names(args) <- paste(substitute(list(...)))[-1]
    args
}
hs.make.arg.wj <- function(..., link = "=", sep = "-", outer = sep, before = {
}, after = {
}) {
    z <- c(...)
    m <- paste(substitute(list(...)))[-1]
    str_c(c(before, if (length(z)) str_c(c(outer, str_c(str_c(m, z, sep = link), collapse = sep), outer), collapse = ""), after), collapse = ".")
}
hs.make.arg.wj.1 <- function(x, default, before = {
}, link = "=", sep = "-", outer = sep, after = {
}, up = parent.frame()) {
    if (any(c("symbol", "character") %in% typeof(x))) {
        M <- as.character(x)
        x <- get(M, envir = up)
    }
    else M <- as.character(substitute(x))
    if (missing(default) || (default != x)) {
        paste(c(before, sprintf("%s%s%s%s%s", outer, M, link, x, outer), after), collapse = ".")
    }
    else before
}
hs.mfp.arg <- function(uid, L, ext = "rd", up = parent.frame(), ...) {
    str_arg <- paste(unlist(sapply(seq_along(L), function(i1) {
        L1 <- L[i1]
        hs.make.arg.wj.1(names(L1), L1, up = up)
    })), collapse = ".")
    .mfp(paste(c(uid, if (nchar(str_arg)) str_arg, ext), collapse = "."), ...)
}
env_name.37_12_22_104219 <- ""
hs.localize <- function(hs1, e1 = parent.frame(), full = !identical(environment(hs1), .GlobalEnv), new = 1) {
    if (full) {
        ep <- if (new) 
            new.env(parent = e1)
        else e1
        on <- hs.ls(environment(hs1))
        for (on1 in on) ep[[on1]] <- environment(hs1)[[on1]]
        environment(hs1) <- ep
    }
    else environment(hs1) <- e1
    hs1
}
wk.33.06.24.202656.v2.auto <- function(hs1, envUp = parent.frame(), new = 0) {
    e1 <- substitute(hs1)
    env1 <- if (is.call(e1) && (e1[[1]] == as.symbol("local"))) {
        environment(hs1)
    }
    else new.env(parent = envUp)
    {
        i1 <- sapply(seq_along(body(hs1)), function(i1) {
            e1 <- body(hs1)[[i1]]
            if (hs.len1(e1) && ("name" %in% class(e1)) && (as.character(e1)[1] == "..meta.wj_out")) 
                i1
        }) %=>% hs.unlist
        if (length(i1) == 1) 
            body(hs1)[[i1]] <- body(..meta.wj_out)
    }
    eval(bquote({
        env2.33.06.24.181217 <- new.env()
        env2.33.06.24.181217[["body.34.07.07.065614"]] <- body(.(hs1))
    }), env1)
    a1 <- list(..mem = 0, ..mem.object_size.cutoff = 1e+05, ..only.wj_out = 0, ..only.wj2pass = 0, ..wj_old = {
    }, ..load = local({
        e1 <- as.character(tail(body(hs1), 1)[[1]])
        ifelse(hs.len1(e1) && (e1 == "wj_out"), 0, 1)
    }), ..load.hs = "smart", ..suppress.time = 0, ..new = new, ..finally = identity, ..debug = 0, ..ssh = {
    })
    a1 <- a1[names(a1) %rm% names(formals(hs1))]
    if (length(a1)) 
        formals(hs1) <- c(formals(hs1), a1)
    body(hs1) <- bquote({
        if (..mem) {
            ac.34.07.17.183129 <- "jg.34.07.06.060444"
            ac.34.07.17.183129 <- wk.digest.args(failed = "jg.34.07.06.060444", by = "size", fa = .(formalArgs(hs1)))
            if ((ac.34.07.17.183129 == "jg.34.07.06.060444") || (!hs.exists(ac.34.07.17.183129, env2.33.06.24.181217)) || ..new) {
                env2.33.06.24.181217[[ac.34.07.17.183129]] <- eval(env2.33.06.24.181217[["body.34.07.07.065614"]])
            }
            if (length(ac.34.07.17.183129) && ("jg.34.07.06.060444" != ac.34.07.17.183129)) 
                env2.33.06.24.181217[["jg.34.07.06.060444"]] <- env2.33.06.24.181217[[ac.34.07.17.183129]]
            env2.33.06.24.181217[["jg.34.07.06.060444"]]
        }
        else eval(env2.33.06.24.181217[["body.34.07.07.065614"]], NULL)
    })
    hs.localize(hs1, env1, new = 0)
}
..meta.wj_out <- function() {
    if (exists("..wj2pass", inherits = FALSEs)) {
        if (..only.wj2pass) 
            return(..wj2pass)
    }
    if (exists("wj_out", inherits = FALSE) && (length(wj_out))) {
        if (length(..ssh)) {
            ..only.wj_out <- 1
            if (all(file.exists(wj_out))) {
                for (wj1 in wj_out) wk.rsync.out.same_place(wj1, ssh = ..ssh)
            }
            else stop("结果文件不存在：", wj_out %=>% x[!file.exists(x)] %=>% paste(collapse = "\n"))
        }
        if (..only.wj_out) 
            return(wj_out)
        if (..load == "smart") 
            ..load <- hs.wj.exists(wj_out) && (file.size(wj_out) < 1e+06)
        if (..load && identical(..load.hs, "smart") && hs.len1(wj_out)) {
            ..load.hs <- if (any(hs.grepl(wj_out, "(?i)\\.(txt|tsv|csv)(\\.gz)?$"))) {
                fread
            }
            else hs.load
        }
        if (!exists("..wj_old", inherits = FALSE)) 
            ..wj_old <- {
            }
        ..wj_old <- hs.unlist(..wj_old)
        if (exists("wj_in", inherits = FALSE)) 
            ..wj_old <- ..wj_old %or% hs.unlist(wj_in)
        if (length(..wj_old) && ..suppress.time) {
            if (hs.wj.exists(wj_out)) 
                local({
                  t2 <- min(file.mtime(hs.unlist(wj_out)))
                  for (wj1 in ..wj_old) {
                    if (file.mtime(wj1) >= file.mtime(wj_out)) {
                      hs.Sys.setFileTime(wj1, t2)
                    }
                  }
                })
        }
        if ((!..new) && all(hs.wj.exists(wj_out)) && ifelse(length(..wj_old), max(file.mtime(..wj_old)) <= min(file.mtime(hs.unlist(wj_out))), TRUE)) {
            if (..load && hs.len1(wj_out)) {
                return(..finally(..load.hs(wj_out)))
                wj_out %=>% .less
            }
            else return(wj_out)
        }
    }
}
wk.33.06.24.202656.v2 <- function(env1 = parent.frame(), fa) {
    eval(bquote({
        ac.34.07.17.183129 <- "jg.34.07.06.060444"
        if (..mem) {
            ac.34.07.17.183129 <- wk.digest.args(failed = "jg.34.07.06.060444", by = "size", fa = .(fa), env1 = .(env1))
        }
        if ((!hs.exists(ac.34.07.17.183129, env2.33.06.24.181217)) || ..new) {
            env2.33.06.24.181217[[ac.34.07.17.183129]] <- eval(env2.33.06.24.181217[["body.34.07.07.065614"]])
        }
        if (length(ac.34.07.17.183129) && ("jg.34.07.06.060444" != ac.34.07.17.183129)) {
            env2.33.06.24.181217[["jg.34.07.06.060444"]] <- env2.33.06.24.181217[[ac.34.07.17.183129]]
        }
        env2.33.06.24.181217[["jg.34.07.06.060444"]]
    }), env1)
}
hs.try_with_time_limit <- function(expr, cpu = Inf, elapsed = Inf, failed = {
}) {
    y <- try({
        expr
    }, silent = TRUE)
    if (inherits(y, "try-error")) {
        message("time out 34.07.07.061813 @ ~/org/dm/r/xm.r")
        failed
    }
    else y
}
wk.digest.args <- function(env1 = parent.frame(), failed = {
}, by = "size", fa = formalArgs(sys.function(-1))) {
    PL <- mget(fa %rm% "...", env1, inherits = TRUE)
    if ("..." %in% fa) 
        PL <- append(PL, evalq(list(...), env1))
    an <- names(PL)
    if (length(an)) 
        PL <- PL[order(c(an, rep("", length(PL) - length(an))), sapply(PL, object.size))]
    switch(by, size = {
        if (object.size(PL) < 1e+06) {
            for (i1 in seq_along(PL)) if ("function" %in% class(PL[[i1]])) PL[i1] <- list(format(PL[[i1]]))
            digest::digest(PL)
        } else failed
    }, time = {
        hs.try_with_time_limit(digest::digest(PL), 3, failed = .(failed))
    })
}
.r1 <- function() hs.say("34.11.29.071505")
.ag.less <- function(cmd1, C = 9, l = 0, H = 1, exclude_fixed = {
}, exclude = {
}, zim = 0, less = 1, include = {
}) {
    a1 <- .wk(cmd1, "| ag --passthrough", .q(s1), intern = 1)
    if (length(exclude_fixed)) 
        a1 <- a1 %rm% exclude_fixed
    for (e1 in exclude) a1 <- hs.grep(a1, e1, v = 1, inv = 1)
    for (p1 in include) 1
    if (zim) {
        i1 <- hs.grep(a1, sprintf("^\033\\[1;32m%s", hs.home.expand("~")))
        wj <- a1[i1]
        wj <- hs.sub(wj, "^\033\\[1;32m")
        wj <- hs.sub(wj, "\033\\[0m\033\\[K$")
        wj <- hs.path2zim(wj)
        a1[i1] <- sprintf("\033[1;32m%s\033[0m\033[K", wj)
    }
    if (less) {
        wj4less <- hs.wj.temp()
        on.exit(hs.rm.wj(wj4less), add = TRUE)
        cat(a1, file = wj4less, sep = "\n")
        .wk("less -R -N", wj4less)
    }
    else return(a1)
    if (0) 
        if (length(a1) && (!hs.match("", a1))) {
            wj1 <- hs.sub(a1[1], "^\033\\[1;32m")
            wj1 <- hs.sub(wj1, "\033\\[0m\033\\[K")
            .wk("~/wk/rj/zim/zim.py", "~/xt", wj1, "&")
        }
    if (0) {
        wj <- .ag(sprintf("%s", s1), "~/xt", G = "\\.txt$", l = 1, m = 1, intern = 1)
        wj <- hs.wjm(wj, ext = 0)
        wj <- hs.sub(wj, path.expand("~/xt"))
        wj <- hs.gsub(wj, "/", ":")
        if (!length(wj)) {
            hs.say("nothing")
        }
        else if (hs.len1(wj)) {
            .wk("~/wk/rj/zim/zim.py", "~/xt", wj, "&")
        }
        else {
            if (l) {
                cat(wj, sep = "\n")
            }
            else .ag(sprintf("%s", s1), "~/xt", G = "\\.txt$", H = H)
        }
    }
}
.ag1 <- function(s1, wjj1, ext1 = "txt|org|r|el", C = 9, l = 0, H = 1, exclude_fixed = {
}, exclude = {
}, exclude_wj = {
}, zim = 0, less = 1, include = {
}, m1 = 0) {
    if (0) {
        s1 <- c("@前线", "@尽快")
        wjj1 <- "~/xt"
    }
    if (0) {
        cmd1 <- .sys("ag -l", .q(s1[1]), wjj1)
        if (length(s1) > 1) 
            for (i1 in 2:length(s1)) cmd1 <- .sys(cmd1, "| xargs ag -l", .q(s1[i1]))
        .wk(cmd1, "| xargs ag --color -H -C", .q(paste(paste0("(", s1, ")"), collapse = "|")))
    }
    p1 <- .q(paste(rep(paste0("(", paste(paste0("(", s1, ")"), collapse = "|"), ")"), length(s1)), collapse = ".*"))
    cmd1 <- .sys("ag --color -H -C", C, if (length(exclude_wj)) 
        paste("--ignore", exclude_wj), "-m", m1, sprintf("-G \"(?i)\\.(%s)$\"", ext1), p1, wjj1, "2>/dev/null", "| ag --passthrough", p1)
    a1 <- suppressWarnings(.wk(cmd1, intern = 1))
    if (length(exclude_fixed)) 
        a1 <- a1 %rm% exclude_fixed
    for (e1 in exclude) a1 <- hs.grep(a1, e1, v = 1, inv = 1)
    if (zim) {
        i1 <- hs.grep(a1, sprintf("^\033\\[1;32m%s", hs.home.expand("~")))
        wj <- a1[i1]
        wj <- hs.sub(wj, "^\033\\[1;32m")
        wj <- hs.sub(wj, "\033\\[0m\033\\[K$")
        wj <- hs.path2zim(wj)
        a1[i1] <- sprintf("\033[1;32m%s\033[0m\033[K", wj)
    }
    if (less) {
        wj4less <- hs.wj.temp()
        on.exit(hs.rm.wj(wj4less), add = TRUE)
        cat(a1, file = wj4less, sep = "\n")
        .wk("less -R -N", wj4less)
    }
    else return(a1)
    if (0) 
        if (length(a1) && (!hs.match("", a1))) {
            wj1 <- hs.sub(a1[1], "^\033\\[1;32m")
            wj1 <- hs.sub(wj1, "\033\\[0m\033\\[K")
            .wk("~/wk/rj/zim/zim.py", "~/xt", wj1, "&")
        }
    if (0) {
        wj <- .ag(sprintf("%s", s1), "~/xt", G = "\\.txt$", l = 1, m = 1, intern = 1)
        wj <- hs.wjm(wj, ext = 0)
        wj <- hs.sub(wj, path.expand("~/xt"))
        wj <- hs.gsub(wj, "/", ":")
        if (!length(wj)) {
            hs.say("nothing")
        }
        else if (hs.len1(wj)) {
            .wk("~/wk/rj/zim/zim.py", "~/xt", wj, "&")
        }
        else {
            if (l) {
                cat(wj, sep = "\n")
            }
            else .ag(sprintf("%s", s1), "~/xt", G = "\\.txt$", H = H)
        }
    }
}
.ag <- function(s1, wjj2do = hs.home.expand("~/1"), wj = {
}, C = 9, less = 1, l = 0, intern = 0, G = {
}, ig = {
}, m = 0, silent = 1, H = 1) {
    if (!length(wjj2do)) 
        wjj2do <- c("~/org", "~/xt", "~/rj/sh", "~/rjk/emacs/myLisp") %=>% hs.home.expand
    if (length(wj)) {
        wjj2do <- switch(wj, r = "~/org/dm/r", org = "~/org", zim = hs.home.expand("~/xt"), sh = hs.home.expand("~/rj/sh"), el = hs.home.expand("~/rjk/emacs/myLisp"))
        G <- switch(wj, r = "\\.[rR]$", org = "\\.org$", zim = "\\.txt$", el = "\\.el$")
        ig <- switch(wj, org = c("org/2do.org", "org/dm.r.org"))
    }
    .wk("ag", if (l) 
        "-l", if (C) 
        c("-C", C), if (less) 
        c("--pager", .q("less -R"))
    else "--nocolor", "--ignore", .q("*~"), "--ignore", .q("*#"), if (length(G)) 
        c("-G", .q(G)), if (length(ig)) 
        paste("--ignore", .q(ig)), if (m) 
        c("-m", m), if (silent) 
        "--silent", ifelse(H, "--heading", "--noheading"), "-S", .q(s1), wjj2do)
    a1 <- unlist(sapply(wjj2do, function(wjj1) {
        .wk("ag", if (l) 
            "-l", if (C) 
            c("-C", C), if (less) 
            "--color"
        else "--nocolor", "--ignore", .q("*~"), "--ignore", .q("*#"), if (length(G)) 
            c("-G", .q(G)), if (length(ig)) 
            paste("--ignore", .q(ig)), if (m) 
            c("-m", m), if (silent) 
            "--silent", ifelse(H, "--heading", "--noheading"), "-S", .q(s1), wjj1, intern = 1)
    }))
    unlist(sapply(wjj2do, function(wjj1) {
        .wk("ag", if (l) 
            "-l", if (C) 
            c("-C", C), if (less) 
            c("--pager", .q("less -R"))
        else "--nocolor", "--ignore", .q("*~"), "--ignore", .q("*#"), if (length(G)) 
            c("-G", .q(G)), if (length(ig)) 
            paste("--ignore", .q(ig)), if (m) 
            c("-m", m), if (silent) 
            "--silent", ifelse(H, "--heading", "--noheading"), "-S", .q(s1), wjj1, intern = intern)
    }))
}
.orz <- function(C = 2) {
    .wk("ag -az -C", C, "-G \"$4\" --pager \"less -NiR\" -S \"$1\" $3")
}
r.str <- local({
    hs1 <- function(v1, color = "number", width.equal = 1) {
        paste0({
        }, 1)
        color1 <- if (length(color)) 
            switch(color, number = "\033[1;32m", , search = "\033[30;43m", space1 = "\033[7;96m", space2 = "\033[7;93m")
        color1.end <- if (length(color)) 
            "\033[0m\033[K"
        sprintf(paste0(color1, "%", if (width.equal) 
            max(nchar(v1)), "s", color1.end), v1)
    }
    tab.space.hs <- function(w1) sprintf(paste0("%", w1, "s"), " ")
    tab.width.in <- 2
    tab.space <- tab.space.hs(tab.width.in)
    wj1 <- ""
    function(wj.r = readline("输入文件名："), tab.width = 2, ud = 0) {
        if ((!length(wj.r)) || (!file.exists(wj.r[1]))) 
            wj.r <- wj1
        if (!file.exists(wj.r)) 
            stop("【文件不存在】", wj.r)
        if (wj1 != wj.r) 
            wj1 <<- wj.r
        orz("^(#|.*{\\s*##)", wj2do = wj.r, ud = ud)
        return()
        if (tab.width.in != tab.width) {
            tab.width.in <<- tab.width
            tab.space <<- tab.space.hs(tab.width)
        }
        a1 <- .wk("ag -az", .q("^(##+|.*{ *##)"), wj.r, intern = 1)
        rn <- hs1(hs.re(a1, "\\d+"))
        lines <- hs.re(a1, "(?<=:).*")
        spaces_leading <- {
            a1 <- hs.re(lines, "^\\s*")
            hs.gsub(a1, "\t", tab.space)
        }
        cn <- {
            cn <- 1 + nchar(spaces_leading)
            hs1(ceiling(cn/2), color = {
            })
        }
        spaces_leading_colored <- sapply(spaces_leading, function(s1) {
            l1 <- nchar(s1)
            if (l1) {
                if (l1%%tab.width) {
                  hs1(s1, color = "search", width.equal = 0)
                }
                else {
                  n1 <- l1/tab.width
                  block.color <- head(rep(c("space1", "space2"), ceiling(n1/2)), n1)
                  paste(sapply(block.color, function(c1) hs1(tab.space, color = c1)), collapse = "")
                }
            }
            else s1
        })
        wj_temp <- hs.wj.temp()
        on.exit(hs.rm.wj(wj_temp), add = TRUE)
        cat(paste0(rn, ":", cn, ":", spaces_leading_colored, trim(lines)), file = wj_temp, sep = "\n")
        .wk("less -NiR", wj_temp)
    }
})
z.str <- local({
    wj.wjj <- hs.wjj(.mfp("36.02.16.141824"))
    function(node = .wk("xsel --clipboard --output", intern = 1), tab.width = 2) {
        wj <- paste0("~/xt/", hs.gsub(node[1], ":", "/"), ".txt")
        if (file.exists(wj)) {
            wj.wjj %=>% hs.wjj.clean
            wjj <- hs.wjm(wj, ext = 0)
            file.symlink(c(wj, if (file.exists(wjj)) wjj), wj.wjj)
        }
        if (length(hs.wjj.ls(wj.wjj))) {
            wjj0 <- getwd()
            on.exit(setwd(wjj0), add = TRUE)
            setwd(wj.wjj)
            hs.wjj.ls() %=>% x[!file.exists(x)] %=>% {
                if (length(x)) 
                  hs.rm.wj(x)
            }
            .wk("ag -f", "--pager", .q("less -NiR"), .q("^(=+ .* =+|\\s*\\* .*)$"), "--color-match", .q("1;34"))
        }
    }
})
.gz <- function(s1 = readline()) {
    .wk("ag", "--pager", .q("less -Rr"), "-f", "-G", .q("dw\\.(txt|org)$"), s1, "~/wk/热点")
}
.ot <- function(s1, C = 9, l = 0, H = 1) {
    .ag1(s1 = s1, wjj1 = "~/org", ext1 = "org", C = C, l = l, H = H, less = 1, exclude_wj = c("dm.r.org", "wl.r.org", "wl.org"))
}
.ag.org <- function(s1, C = 9, l = 0, H = 1) {
    .ag1(s1 = s1, wjj1 = "~/org", C = C, l = l, H = H, less = 1, exclude_wj = c("dm.r.org", "wl.r.org", "wl.org"))
}
zim <- hs.path2zim <- function(wj1 = readline(), open = 1) {
    a <- hs.re(wj1, "[^:]*")
    if (file.exists(a)) {
        wj1 <- a
    }
    else {
        ag_jg <- .wk("ag", .q(wj1), "~/xt", intern = 1)
        wj1 <- hs.re(ag_jg, "[^:]*")
    }
    wj1 <- hs.sub(normalizePath(wj1), normalizePath(sprintf("%s/xt", path.expand("~"))))
    wj1 <- hs.gsub(wj1, "/", ":")
    wj1 <- hs.wjm(wj1, ext = 0)
    if (open) {
        .wk("~/wk/rj/zim-desktop-wiki-develop/zim.py ~/xt", .q(wj1), "&")
    }
    else wj1
}
.zt <- function(s1, C = 9, l = 0, H = 1) {
    .ag1(s1 = s1, wjj1 = "~/xt", C = C, l = l, H = H, exclude_fixed = c("\033[1;33m1\033[0m\033[K-Content-Type: text/x-zim-wiki", "\033[1;33m2\033[0m\033[K-Wiki-Format: zim 0.4"), exclude = "-Creation-Date", zim = 1, less = 1)
}
.rt <- function(s1 = readline(), C = 9, l = 0, H = 1) {
    .ag1(s1 = s1, wjj1 = "~/org/dm/r", C = C, l = l, H = H, less = 1)
}
.ohd <- function(s1, ic = 1) {
    .wk("find", "-L ~/org", "-type f", "-name", .q("*.org"), "-print0", "|", "xargs", "-0", "gawk", if (ic) 
        "-v IGNORECASE=1", .q(sprintf("{ if( FNR > 1) { nextfile}} { if( /%s/) { print FILENAME\" : \"$0; nextfile}}", s1)))
    invisible()
}
.zhd <- function(s1, ic = 1) {
    .wk("find", "-L ~/xt", "-type f", "-name", .q("*.txt"), "-print0", "|", "xargs", "-0", "gawk", if (ic) 
        "-v IGNORECASE=1", .q(sprintf("\n{ if( ! length( $0)) { nbl++}}\n{ if( FNR > 6) { nbl = 0; nextfile}}\n{ if( nbl > 1) { nbl = 0; nextfile}}\n{ if( /%s/) { sub( \"%s\", \"\", FILENAME); sub( \"\\\\.txt$\", \":\", FILENAME); gsub( \"/\", \":\", FILENAME); print FILENAME\" : \"$0; nbl = 0; nextfile}}", s1, path.expand("~/xt"))))
    invisible()
}
.oh <- function(s1) {
    .ag1(sprintf("^\\*+ +.*%s.*", s1), "~/org", "org", exclude_wj = c("dm.r.org", "wl.r.org", "wl.org"))
}
.oh1 <- function(s1) {
    .ag1(sprintf("^\\* +.*%s.*", s1), "~/org", "org", exclude_wj = c("dm.r.org", "wl.r.org", "wl.org"), m1 = 1)
}
.oht <- function(h = {
}, t = {
}, C = 9) {
    .ag.ht(h = h, t = t, type = "org")
}
.zh <- function(s1) {
    if (0) {
        s1 <- "轴"
        .ag1(s1 = sprintf("^(\\t*\\* +.*%s)|(=+ +.*%s.* =+)", s1, s1), wjj1 = "~/xt", "txt")
    }
    .ag1(s1 = sprintf("^(\\t*\\* +.*%s)|(=+ +.*%s.* =+)", s1, s1), wjj1 = "~/xt", "txt", exclude_fixed = c("\033[1;33m1\033[0m\033[K-Content-Type: text/x-zim-wiki", "\033[1;33m2\033[0m\033[K-Wiki-Format: zim 0.4"), exclude = "-Creation-Date", zim = 1)
}
.zht <- function(h = {
}, t = {
}, C = 9) {
    .ag.ht(h = h, t = t, type = "zim")
}
.zh2 <- function(s1, C = 9) {
    s1 <- unlist(strsplit(s1, ",,+"))
    cmd1 <- .sys("ag", if (length(s1) > 1) 
        c("-l", "-0"), .q(sprintf("^=+ .*%s.* =+", s1[1])), "~/xt")
    for (p1 in s1[-1]) cmd1 <- .sys(cmd1, "| xargs -0", "ag", "-l -0", .q(sprintf("^=+ .*%s.* =+$", p1)))
    if (length(s1) > 1) 
        cmd1 <- .sys(cmd1, "| xargs -0 ag -H -C", C, .q(sprintf("^=+ .*(%s).* =+$", paste(s1, collapse = "|"))))
    .wk(cmd1)
}
.ag.ht <- function(h = {
}, t = {
}, C = 9, type = "zim") {
    cmd1 <- {
    }
    wjj1 <- sprintf("~/%s", switch(type, org = "org", zim = "xt"))
    if (length(h)) {
        h <- s1 <- unlist(strsplit(h, ",{2,}"))
        ph <- switch(type, zim = "^=+ .*%s.* =+", org = "^\\*+ .*%s")
        cmd1 <- .sys("ag", "-l -0 -S", .q(sprintf(ph, s1[1])), wjj1)
        for (p1 in s1[-1]) cmd1 <- .sys(cmd1, "| xargs -0", "ag", "-l -0", .q(sprintf(ph, p1)))
    }
    if (length(t)) {
        t <- s1 <- unlist(strsplit(t, ",{2,}"))
        cmd1 <- .sys(cmd1, if (length(cmd1)) 
            "| xargs -0", "ag", "-l -0 -S", .q(s1[1]), if (!length(cmd1)) 
            wjj1)
        for (p1 in s1[-1]) cmd1 <- .sys(cmd1, "| xargs -0", "ag", "-l -0", .q(p1))
    }
    if (length(cmd1)) {
        cmd1 <- strsplit(cmd1, "-0") %=>% unlist %=>% paste0(paste0(head(x, -1), collapse = "-0"), tail(x, 1))
        wj <- .wk(cmd1, intern = 1)
        if (length(wj)) {
            wj_tmp <- hs.wj.temp()
            on.exit(hs.rm.wj(wj_tmp), add = TRUE)
            cat(wj, file = wj_tmp, sep = "\n")
            cmd1 <- .sys("cat", wj_tmp, "| tr", .q("\n"), .q("\\0"), "| xargs -0 ag -C", C, "-H", .q(paste(paste0("(", c(sprintf(ph, h), t), ")"), collapse = "|")))
            .wk(cmd1)
        }
    }
    cmd1
}
.oln <- function(s1) {
    .ag(sprintf("\\[\\[.*%s.*\\]\\]", s1), "~/org")
}
.zln <- function(s1) {
    .wk("ag", .q(sprintf("\\[\\[[^\\[\\]]*%s[^\\[\\]]*\\]\\]", s1)), "~/xt")
}
.e <- function(wj1) .wk("source activate emacs;", "e", wj1)
.dm <- hs.org.uid2wj.openInEmacs <- function(uid1 = readline()) .e(hs.org.uid2wj(uid1))
.wj <- function(s1, wjj1 = "~/org") {
    wj <- .ag(s1, wjj1, C = 0, l = 1, intern = 1)
    if (!length(wj)) {
        hs.say("nothing")
    }
    else if (hs.len1(wj)) {
        .e(wj)
        return()
    }
    if (length(wj) > 9) {
        hs.browser("35.03.28.082317")
        hs.say("inspect wj:")
    }
    else cat(wj, sep = "\n")
}
wk.bf.wk.bypy.wjj2rm <- hs.wj(path.expand("~"), c("wk/rj/exe/Firefox_56/firefox56/cache2", "wk/rjk/ff/p4current/wk/cache2", "wk/rjk/tb/p/wk/cache2", "wk/rjk/ztr/p/wk/cache2", "wk/wk.1/rj/firefox/wk/cache2", "wk/wk.1/rjk/ln-s/.wiznote", c("wk/wk.1/rjk/ln-s/.cache/chromium", "wk/wk.1/rjk/ln-s/.cache/vivaldi", "wk/wk.1/rjk/ln-s/.config/chromium", "wk/wk.1/rjk/ln-s/.config/SogouPY", "wk/wk.1/rjk/ln-s/.config/vivaldi", "wk/rjk/有道云笔记", "wk/wk.1/rj/firefox/wk/cache2", "wk/wk.1/rjk/ztr/p/wk/cache2"), 
    c("wk/wk.1/rjk/ln-s/.wiznote"), "wk/wk.1/rjk/ln-s/.julia"))
wk.bf.bypy.syncup <- function(wjj1, wjj2 = {
}, wjj2rm = wk.bf.wk.bypy.wjj2rm) {
    wjj1 <- path.expand(wjj1)
    if (hs.home.expand("~") %=>% {
        substr(wjj1, 1, nchar(x)) != x
    }) 
        stop("35.04.08.054812")
    if (!length(wjj2)) 
        wjj2 <- hs.sub(wjj1, sprintf("%s/", hs.home.expand("~")))
    hs.say("doing", wjj1)
    .wk("bypy", "-v", "--include-regex", .q(sprintf("^((?!%s).)*$", paste(wjj2rm, collapse = "|"))), "syncup", .q(wjj1), .q(wjj2), "yes")
}
wk.bypy.upload.doc <- function(local.wjj, baidu.wjj) {
    wjj.36_03_04_003344 <- getwd()
    on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
    setwd(local.wjj)
    baidu.wjj %<=>% file.path(basename(local.wjj))
    wjj2do <- c(".", hs.wjj.ls(local.wjj, wjj = 1, wj = 0, R = 1, F = 0))
    for (wjj1 in wjj2do) {
        wjj2 <- ifelse(wjj1 == ".", baidu.wjj, file.path(baidu.wjj, wjj1))
        .wk("bypy mkdir", wjj2)
        wj <- hs.wjj.ls(wjj1, "(?i)\\.(pdf|png|jpg|jpeg|xls|xlsx)$")
        for (wj1 in wj) {
            message(wj1)
            .wk("bypy upload", wj1, wjj2)
        }
    }
}
wk.bf.bypy.wk <- function() {
    if (0) {
        wjj <- .wk("find", "-L", hs.home.expand("~/wk"), "-type d", intern = 1)
        {
            wjj2rm <- {
            }
            wjj1 <- "/home/wk/wk/rj/exe/Firefox_56/firefox56/cache2"
            for (wjj1 in sprintf("%s/%s", path.expand("~"), wk.bf.wk.bypy.wjj2rm)) {
                p1 <- sprintf("^%s", wjj1)
                wjj2rm <- c(wjj2rm, hs.grep(wjj, p1, v = 1))
                wjj1.up <- wjj1
                while (wjj1.up != "/") {
                  wjj2 <- dirname(wjj1.up)
                  wjj2rm <- c(wjj2rm, wjj2)
                  wjj1.up <- wjj2
                }
            }
            wjj <- wjj %not% wjj2rm
        }
        {
            wjj <- sort(wjj)
            i1 <- 1
            while (i1 <= length(wjj)) {
                i2 <- hs.grep(wjj, sprintf("^%s", wjj[i1]))
                if (length(i2) > 1) {
                  wjj <- wjj[-i2[-1]]
                }
                i1 <- i1 + 1
            }
        }
        if (0) {
            wjj1 <- wjj[1]
            wjj1 <- hs.grep(wjj, " ", v = 1)
        }
        null <- sapply(wjj, function(wjj1) {
            hs.say("\ndoing", wjj1, "\n")
            wjj2 <- hs.sub(wjj1, sprintf("%s/", path.expand("~")))
            .wk("bypy", "-v", "syncup", .q(wjj1), .q(wjj2), "yes")
        })
    }
    wk.bf.bypy.syncup(hs.home.expand("~/wk"))
}
wk.bf.swxx.bypy <- function() {
    wjj1 <- hs.home.expand("~/e/swxx")
    if (!file.exists(wjj1)) 
        stop("35.04.08.055924")
    wk.bf.bypy.syncup(wjj1, "swxx", wjj2rm = hs.wj(path.expand(wjj1), "sj/xm"))
}
wk.bf.swxx.byRsync.2in2 <- function() {
    wjj1 <- hs.home.expand("~/e/swxx")
    wk.rsync(wjj1, hs.home.expand("~/in2/bf"), excluded = "sj/xm")
}
wk.bf.bypy.1h <- function() {
    repeat {
        for (wjj1 in hs.wj(hs.home.expand("~"), c("wk/wk.1/org", "wk/wk.1/xt", "wk/wk.1/rjk/ln-s"))) wk.bf.bypy.syncup(wjj1)
        d1 <- 60 * 60
        hs.say("sleep for", d1, "seconds")
        .sleep(d1)
    }
}
wj.zim.bin <- hs.home.expand("~/wk/rj/zim/zim.py")
wj.zim.notebook.xt <- paste0(hs.home.expand("~/.cache/zim/notebook-home_"), Sys.info()["user"], "_xt/index.db")
wk.zim.down <- function(p1, rz = 0) {
    if (0) {
        p1.txt <- paste0("~/xt", hs.gsub(l1, ":", "/"), ".txt")
        down <- readLines(p1.txt) %=>% hs.re("\\[\\[.+\\]\\]") %=>% unique %=>% {
            substr(x, 3, nchar(x) - 2) %not% p1
        }
    }
    {
        hs.require("RSQLite")
        con1 <- dbConnect(RSQLite::SQLite(), dbname = wj.zim.notebook.xt)
        on.exit(dbDisconnect(con1), add = TRUE)
        p1.id <- dbGetQuery(con1, paste0("select * from pages where name is ", "\"", substring(p1, 2), "\""))[, "id"]
        down <- paste0(":", dbGetQuery(con1, paste0("select * from links where source is ", p1.id))[, "names"])
    }
    if (0) {
        down <- paste0(":", .wk(wj.zim.bin, "--search", "~/xt", paste0("LinksFrom", p1), intern = 1)) %not% p1
    }
    if (!rz) 
        down %<=>% hs.grep(":\\d+:\\d{2,2}:\\d{2,2}:{0,1}$", inv = 1, v = 1)
    down
}
wk.zim.up <- function(p1, rz = 0, exclude = {
}) {
    hs.require("RSQLite")
    con1 <- dbConnect(RSQLite::SQLite(), dbname = wj.zim.notebook.xt)
    on.exit(dbDisconnect(con1), add = TRUE)
    p1.id <- dbGetQuery(con1, paste0("select * from pages where name is ", "\"", substring(p1, 2), "\""))[, "id"]
    up.id <- dbGetQuery(con1, paste0("select * from links where target is ", p1.id))[, "source"]
    up <- sapply(up.id, function(id1) dbGetQuery(con1, paste0("select * from pages where id is ", id1))[, "name"])
    if (!length(up)) 
        return()
    up <- paste0(":", up)
    if (!rz) {
        i1 <- hs.grep(up, ":\\d+:\\d{2,2}:\\d{2,2}:{0,1}$")
        if (length(i1)) 
            up <- up[-i1]
    }
    for (i1 in seq_along(exclude)) {
        ms1 <- exclude[i1]
        i2 <- hs.grep(up, ms1)
        if (length(i2)) 
            up <- up[-i2]
    }
    up
}
wk.zim.up.print <- function(p1, lvl = 1, seen = p1) {
    if (lvl == 1) {
        wjj0 <- hs.wjj("~/tmp")
        sep <- ",,,"
        wj_tmp <- hs.wj(wjj0, hs.gsub(p1, ":", sep)) %=>% hs.wjm(add = "org")
        p1 %<=>% paste0(":", x)
        sink(wj_tmp)
    }
    cat(rep("*", lvl), " ", p1, "\n", sep = "")
    up <- wk.zim.up(p1, rz = 0, ) %not% seen
    if (length(up)) {
        for (p1 in up) wk.zim.up.print(p1, lvl = lvl + 1, seen = c(seen, p1))
    }
    if (lvl == 1) {
        sink()
        if (file.exists(wj_tmp)) {
            .wk("e", wj_tmp)
        }
        wj_tmp
    }
}
hs.jl <- function(what = {
}) {
    wj_out <- c(午休 = "~/org/35/04/08/35.04.08.191442.org", 早睡 = "~/org/35/04/08/35.04.08.193007.org", 健身 = "~/org/35/04/08/35.04.08.193141.org")
    if (length(what)) {
        wj_out <- wj_out[what]
        hs.say(wj_out)
        jt <- paste0(as.integer(format(Sys.time(), "%Y")) - 1984, "/", format(Sys.time(), "%m/%d"))
        if (length(.wk("grep", "-m 1", .q(sprintf("^%s$", jt)), wj_out, intern = 1))) {
            hs.say("已经记录过了")
        }
        else cat(jt, file = wj_out, append = TRUE)
    }
    else hs.say(paste(names(wj_out), collapse = ", "))
}
wk.backup_file.clean <- function(wjj1) {
    if (0) {
        wjj1 <- "~/org"
        wjj1 <- "~/rjk/emacs"
    }
    wj <- .wk("ag", "-g", .q("~$"), wjj1, intern = 1)
    hs.rm.wj(wj)
}
hs.emacs.windows <- function() {
    wj1 <- hs.home.expand(c("~/.windows", "~/.emacs.d/eshell/lastdir"))
    lapply(wj1, function(wj1) {
        if (file.exists(wj1)) {
            nr1 <- readLines(wj1, encoding = "UTF8")
            nr1 <- hs.gsub(nr1, "/(home|Users)/[^/]+", path.expand("~"))
            writeLines(nr1, wj1)
        }
    })
}
wk.path_wine <- function(wj1 = readline()) {
    wj1 <- .wk("wine", .sys.fpGood(wj1), "2>&1", intern = 1)
    hs.re(wj1, "(?<=wine: Bad EXE format for ).*(?=\\.)") %=>% hs.gsub("\\\\", "\\\\\\\\") %=>% cat("\n", sep = "")
}
wk.make.home.ready <- function() {
    if (!hs.machine.isLocal()) 
        return()
    {
        wj1 <- hs.home.expand("~/.bash_history")
        if (!file.is_symlink(wj1)) {
            wj2 <- hs.home.expand("~/rjk/ln-s/.bash_history")
            cat(readLines(wj1) %or% readLines(wj2), sep = "\n", file = wj2)
            hs.rm.wj(wj1)
            file.symlink(wj2, wj1)
        }
    }
    {
        wj1 <- hs.wjj(hs.home.expand("~/.cache/mozilla/firefox/cache2"))
        wj2 <- hs.wjm(wj1, dn = hs.home.expand("~/wk/rjk/ff/p4current/wk"))
        if (!file.exists(wj2)) {
            hs.rm.wj(wj2)
            file.symlink(hs.home.expand("~/.cache/mozilla/firefox/cache2"), hs.home.expand("~/wk/rjk/ff/p4current/wk"))
        }
    }
    {
        wj1 <- hs.home.expand("~/Zotero")
        if (!file.exists(wj1)) {
            hs.rm.wj(wj1)
            file.symlink("~/wk/wk/wd/db/mf/ff/ztr", wj1)
        }
    }
}
hs.org.options.export <- function() c("#+options: ^:{} _:{} \\n:nil", "#+LANGUAGE: zh-CN")
hs.give.me <- function(n1) switch(n1, samtools = hs.hsgly("rj.find.35.08.20.070052")(n1), swxx = hs.hsgly("wjj.swxx.35.09.21.183350")())
wk.find.uid <- function(uid1, update = 0) {
    cat("【文件名】\n")
    wk.mlocate(pat = uid1, update = update)
    .wk("ag", "-t", uid1, "~/org")
    .wk("ag", "-t", uid1, "~/xt")
    .wk("ag", "-t", "-G", "\\.org$", uid1, "~")
}
wk.zip1 <- function(wjj1, wjj2 = dirname(wjj1), zip.name.pre = basename(wjj1), hs.wjj.ls.A = {
}) {
    if (!file.exists(wjj1)) 
        stop("[38_06_01_144332]", wjj1, "does not exists")
    if (!file.exists(wjj2)) 
        stop("[38_06_01_144337]", wjj2, "does not exists")
    hs.with_wd(wjj1, {
        wj1 <- tempfile()
        on.exit(hs.rm.wj(wj1), add = TRUE)
        do.call(hs.wjj.ls, c(list(wjj1, R = 1, F = 0), hs.wjj.ls.A)) %=>% cat(file = wj1, sep = "\n")
        wj2 <- paste0("../", zip.name.pre, ".zip")
        .wk("cat", wj1, "| zip -ru", "-@", wj2)
        normalizePath(wj2)
    })
}
wk.7z1 <- function(wjj1, wjj2 = dirname(wjj1), name.pre = basename(wjj1), ...) {
    if (!file.exists(wjj1)) 
        stop("[38_06_01_145352]", wjj1, "does not exists")
    hs.with_wd(wjj1, {
        wj2 <- hs.wj(wjj2, paste0(name.pre, ".7z"))
        .wk("7z u -r", hs.fpGood(wj2), "*")
        normalizePath(wj2)
    })
}
hs.36_02_03_174626 <- function(wj.wj, wjm = hs.wjm(wj.wj, dn = 0, ext = 0, extN = ifelse(hs.grepl(wj.wj, "\\.gz$"), 2, 1))) {
    wjj0 <- getwd()
    on.exit(setwd(wjj0), add = TRUE)
    wjj1 <- dirname(wj.wj)
    setwd(wjj1)
    wj_temp <- tempfile()
    on.exit(hs.rm.wj(wj_temp), add = TRUE)
    hs.browser("36.07.23.132608")
    wj1 <- readLines(wj.wj) %not% ""
    wj1 <- hs.sub(normalizePath(wj1), normalizePath(wjj1), fixed = TRUE) %=>% hs.sub("^/")
    cat(path.expand(wj1), sep = "\n", file = wj_temp)
    cat(basename(wj1), sep = "\n", file = wj_temp)
    cat(wj1, sep = "\n", file = wj_temp)
    .wk("cat", wj_temp, "| zip -@ -ruj", paste0(wjm, ".zip"))
    wj.wj %=>% .head
    .wk("unzip -l", paste0(wjm, ".zip"), "| less -Ni")
    paste0(wjm, ".zip") %=>% .ls
    paste0(wjm, ".zip") %=>% hs.wj.rm
    wj.wj %=>% hs.wjm(ext = "zip")
}
wk.bash <- function(wjj1 = ".") {
    if (!dir.exists(wjj1)) 
        wjj1 %<=>% dirname
    wjj0 <- getwd()
    on.exit(setwd(wjj0), add = TRUE)
    setwd(wjj1)
    .wk("bash")
}
note.update_time <- function(wj1, case_id = "", wj_info = {
}, ot = "message", co1 = 365/2) {
    if (!length(wj_info)) 
        wj_info <- wj1
    t1 <- difftime(Sys.time(), wj1 %=>% file.mtime %=>% min, units = "days") %=>% as.numeric
    switch(ot, message = message("\t#【", if (nzchar(case_id)) c(case_id, "|"), "提示】", wj_info, "更新于", t1, "天前. 更新频率: 每", co1, "天."), time = t1, `01` = t1 > co1)
}
hs.time.4filename <- function(t1 = Sys.time()) format(t1, "%Y-%m-%d-%H%M%S")
wj.updater <- function(wj1 = {
}, dz1 = {
}, wjj2 = if (length(wj1)) dirname(wj1) else ".", co1 = if (missing(wj1)) eval(formals(wj.update.reporter)[["co1"]]) else Inf, gz = !hs.grepl(dz1, hs.hsgly("不需要压缩的模式.2023_08_10_015813")())) {
    if (!length(dz1)) 
        stop("[37_12_19_203015]无意义！")
    if (!length(wj1)) {
        wj1 <- file.path(wjj2, basename(dz1))
        if (gz && (!hs.grepl(wj1, "(?i)\\.gz$"))) 
            wj1 <- paste0(wj1, ".gz")
    }
    if (file.exists(wj1)) {
        if (note.update_time(wj1, ot = "01", co1 = co1)) {
            wj1_bf <- paste0(wj1, ".bf_", hs.time.4filename())
            file.rename(wj1, wj1_bf)
        }
        else {
            note.update_time(normalizePath(wj1), co1 = co1)
        }
    }
    if (!file.exists(wj1)) {
        hs.msg(paste0("获取", dz1, "到", wj1))
        hs.browser("2023.06.29.214655", debug = 0)
        hs.hsgly("1.信息学/文件/获取/用aria2c.2023_06_26_120607")(dz = dz1, wjj2 = wjj2)
    }
    hs.wj.portable(wj1)
}
wj.update.reporter.co1 <- 30 * 3
wj.update.reporter <- function(wj2, wj1 = {
}, dz1 = {
}, co1 = wj.update.reporter.co1) {
    co1 <- wj.update.reporter.co1
    if (!all(file.exists(wj2))) 
        return(TRUE)
    if (length(dz1)) {
        if (!note.update_time(wj2, ot = "01", co1 = co1)) 
            return(FALSE)
        hs.msg(paste0("检查时间：", dz1))
        a1 <- .try(.wk("curl -I", dz1, intern = 1))
        if (is.character(a1) && any(hs.grepl(a1, "(?i)Last-Modified:"))) {
            time_pattern1 <- "(?i)Last-Modified:"
            t1 <- hs.grepV(a1, time_pattern1) %=>% hs.sub(time_pattern1) %=>% trim
            t1 <- as.POSIXlt(.wk("date", "-d", .q(t1), "-u", "+\"%F %T\"", intern = 1))
            a1 <- difftime(Sys.time(), t1, units = "days") >= co1
            if (a1) {
                .wk("touch", .sys.fpGood(wj2))
            }
            return(!a1)
        }
        else hs.browser(paste0("\t[38.03.23.190500]没能获取", dz1, "的修改时间"))
    }
    if (length(wj1)) {
        !all(file.exists(wj1)) || min(file.mtime(wj2)) < max(file.mtime(wj1))
    }
    else note.update_time(wj2, ot = "01", co1 = co1)
}
hs.wj.ext.sub <- function(wj1, ext) hs.wjm(wj1, ext = ext, extN = ifelse(hs.grepl(wj1, "\\.gz$"), 2, 1))
wk.2xlsx <- function(...) hs.hsgly("txt2xlsx.36.02.01.203801")(...)
wk.pack.wj <- function(wj) {
    a1 <- strsplit(wj, "/")
    n1 <- min(sapply(a1, length))
    b1 <- 1
    repeat {
        if (b1 > n1) 
            stop("【36.04.29.164929】不能得到唯一的文件名")
        a2 <- sapply(a1, . %=>% {
            tail(x, b1) %=>% paste(collapse = ".")
        })
        if (uniqueN(a2) < length(a2)) {
            b1 <- b1 + 1
            next
        }
        else break
    }
    wjj_temp <- file.path(hs.home.expand("~/tmp"), .muid4machine())
    link <- hs.wj(wjj_temp, a2)
    file.symlink(normalizePath(wj), link)
    wjj_temp
}
wk.pack <- function(wjj1 = {
}, wj.wj = {
}, name = paste(tail(unlist(strsplit(normalizePath(wjj1), "/")) %rm% "", up), collapse = "."), up = 1, pattern = "\\.(pdf|png|jpg|jpeg|xlsx)$", ext = "pdf,xlsx,xls,png,jpg,jpeg,tiff,doc,docx,ppt,pptx,html", wj = {
}, R = 1, to_xls = 1) {
    if (length(wj)) {
        wj %<=>% x[file.exists(x)]
        if (!length(wj)) 
            return()
        wjj1 <- wk.pack.wj(wj)
        name <- hs.wjm(wj[1], ext = 0, dn = 0)
    }
    if (!length(wjj1)) 
        wjj1 <- hs.home.expand("~/tmp") %=>% hs.wjj
    wjj.36_03_04_003344 <- getwd()
    on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
    setwd(wjj1)
    if (to_xls && hs.grepl(ext, "\\bxlsx?\\b")) 
        .try(wk.2xls(if (length(wj)) 
            wj
        else "."))
    {
        if (0) {
            if (!length(wj.wj)) {
                wj.wj <- hs.wj(wjj1, paste0(name, ".txt"))
                if (!length(wj)) 
                  wj <- hs.wjj.ls(wjj1, pattern, R = R)
                cat(wj, sep = "\n", file = wj.wj)
            }
            hs.36_02_03_174626(wj.wj)
        }
        if (!hs.grepl(name, "(?i)\\.zip$")) 
            name %<=>% paste0(".zip")
        .wk("zip -ru", .q(name), ". -i", .q(paste0("*.", unlist(strsplit(ext, ",")))))
    }
    normalizePath(name)
}
.l <- wk.less <- function(s1, start = {
}, max.print = 10000, wj2 = {
}) {
    if (!length(wj2)) {
        wj2 <- hs.wj.temp()
        on.exit(hs.rm.wj(wj2), add = TRUE)
    }
    op <- options(max.print = max.print)
    on.exit(options(op), add = TRUE)
    if (!is.character(s1)) 
        s1 <- capture.output(s1)
    cat(s1, sep = "\n", file = wj2)
    .wk("less -Ni", if (length(start)) 
        paste0("+", start), wj2)
    wj2
}
get.function_definition <- function(f1, wj2 = hs.home.expand("~/tmp/a.r")) {
    h1 <- head(capture.output(args(f1)), -1)
    b1 <- capture.output(body(f1))
    fn1 <- as.character(substitute(f1))
    if (length(fn1) > 1) 
        fn1 <- tail(fn1, 1)
    cat(paste(fn1, "<- "), h1, b1, sep = "\n", file = wj2)
    wj2
}
hs.wj.open.in_emacs <- local({
    bin <- switch(Sys.info()["sysname"], Darwin = "/Applications/Emacs.app/Contents/MacOS/bin/emacsclient", Linux = "emacsclient", hs.browser("36.02.29.213731"))
    function(wj) {
        .wk(bin, "-n -nw", .sys.fpGood(wj))
    }
})
hs.do.xm <- function() {
    wjj1 <- hs.home.expand("~/org/dm/r")
    hs.wjj.ls(wjj1, "(?i)\\.r$") %=>% sapply(hs.wj.open.in_emacs)
}
hs.org.jt <- function() {
    .muid4machine()
}
hs.2hotspot <- function(wj, up = 0) {
    wjj <- hs.wjj("~/org/经常访问的")
    name <- if (up) {
        paste(tail(strsplit(wj, "/")[[1]], up + 1), collapse = ".")
    }
    else basename(wj)
    hs.wj.symlink(wj, wjj, name = name)
}
rw.wjj0.hs <- function(wj, uid1 = {
}) {
    if (length(uid1)) 
        wj <- environment(hs.hsgly)[["uid.wj"]][uid == uid1][["wj"]]
    con1 <- .sys.cat.skip(wj, pipe = "r")
    .sys.cat.skip(wj)
    on.exit(close(con1), add = TRUE)
    line1 <- readLines(con1, 1)
    eval(parse(text = line1)[[1]][[3]])
}
hs.head <- function(X, N = 6, P = {
}, ...) {
    if (length(P)) 
        N <- floor(length(X) * P)
    head(X, N, ...)
}
hs.tail <- function(X, N = 6, P = {
}, ...) {
    if (length(P)) 
        N <- ceiling(length(X) * P)
    tail(X, N, ...)
}
wk.wait <- function(N = 9) {
    n1 <- sample(1:N, 1, prob = rev(seq(N))/N)
    hs.say("\twait for", n1, "seconds.")
    .sleep(n1)
}
hs.wjj.str_4parameters <- function(...) {
    jg <- expandingSequence(named = 1)
    wjj2 <- file.path(...)
    repeat {
        a1 <- list(wjj = hs.wjj.ls(wjj2, wjj = 1, F = 0), wj = hs.wjj.ls(wjj2, F = 0))
        jg[["add"]](wjj2, a1)
        if (length(a1[["wjj"]])) {
            wjj2 %<=>% file.path(a1[["wjj"]][1])
        }
        else break
    }
    jg[["get"]]()
}
hs.wjj.str_4parameters.parameters.L <- function(wjj1) {
    wjj_str1 <- hs.wjj.str_4parameters(wjj1)
    names(wjj_str1) %=>% x[seq(1, length(), 2)] %=>% lapply(x, . %=>% list(basename(x), wjj_str1[[x]][["wjj"]]))
}
hs.wjj.str_4parameters.parameters_str <- function(parameters.L, type1 = "wjm", sep1 = switch(type1, wjm = ",", wj = "/"), sep2 = switch(type1, wjm = "__", wj = "/")) {
    do.call(expand.grid, lapply(parameters.L, . %=>% do.call(paste, c(x, sep = sep1)))) %=>% apply(1, paste, collapse = sep2)
}
hs.wjj.str_4parameters.parameter_set <- function(wjj.V, p1 = wjj.V[1] %=>% hs.wjj.ls(wjj = 1, F = 0) %=>% x[1]) {
    parameters.L <- hs.wjj.str_4parameters.parameters.L(file.path(wjj.V[1], p1))
    sj.wj.part_right <- hs.wjj.str_4parameters.parameters_str(parameters.L, "wj")
    jg.wj.part <- hs.wjj.str_4parameters.parameters_str(parameters.L, "wjm")
    hs.c(jg.wj.part, seq_along(sj.wj.part_right) %=>% lapply(x, . %=>% file.path(wjj.V, sj.wj.part_right[x])))
}
pv.hs.str <- function(x1) hs.grepV(x1, "^(?i)p[-_\\.]?v(al)?")
hs.xm <- function(wd = hs.home.expand("~/org")) {
    if (dir.exists(wd)) {
        cfg_wj <- hs.fp(wd, ".windows")
    }
    else cfg_wj <- wd
    if (!file.exists(cfg_wj)) 
        cat("", file = cfg_wj)
    wj2 <- hs.home.expand("~/.windows")
    hs.wj.rm(wj2)
    hs.wj.symlink(wj1 = cfg_wj, wjj1 = dirname(wj2))
}
hs.send2wiz <- function() hs.hsgly("send2wiz.36.02.25.233554")()
.chk <- function(wj1 = ".") {
    if (dir.exists(wj1)) {
        .wk("tree", "-aCfFhi -D", paste0("--timefmt=", .q("%F %H:%M:%S")), hs.fpGood(wj1), " | less -FiNr")
    }
    else {
        hs.cond(hs.grepl(wj1, "(?i)\\.(png|pdf)$"), .wk("opera", hs.fpGood(wj1)), hs.browser("2023.08.20.212323", debug = 0))
    }
}
.h <- .hist <- function(size = 1000) {
    .wk("less -Ni +G", hs.save.history(size = size))
}
"30.10.31.202238" <- NULL
.update <- function(all = rp, where = wk.ssh, U = TRUE, rp = FALSE) {
    hs.source.robust(hs.home.expand("~/.Rprofile"))
    {
        .Last()
        .First()
    }
    if (all) {
        .Rsync(rp = rp, where = where, U = U)
    }
    else {
        hs.hsgly(update.uid = TRUE)
    }
    return()
    wj1 <- hs.home.expand("~/org/dm/r/function.common.R")
    wj2 <- sub("[^\\.]+$", "rd", wj1)
    hs.ready()
    if (file.exists(wj1) && do.call("difftime", as.list(file.info(c(wj2, wj1))[, "mtime"])) < 0) {
        a <- readLines(wj1, encoding = "UTF-8")
        save(a, file = wj2)
        if (0) 
            .try({
                wk.rsync("~/.Rprofile", "~", ssh = ieiwk.ssh.db(where), ssh.side = 2, dry = 0, L = 1)
                hs.wjm("function.common", add = c("R", "rd")) %=>% file.path("~/wk/my/wd/dm/R", x) %=>% sapply(wk.rsync.out, "~/wk/my/wd/dm/R")
            })
    }
    `33.06.02.124853` <- load(wj2)
    eval(parse(text = get(`33.06.02.124853`), encoding = "UTF-8"), envir = .GlobalEnv)
}
.Rsync <- function(where = wk.ssh, wjj2 = hs.home.expand("~/wk/"), rp = FALSE, U = TRUE) {
    hs.hsgly(update.uid = TRUE)
    if (0) {
        wj <- "~/org/dm"
        local({
            wj <- tempfile()
            on.exit(hs.wj.rm(wj), add = TRUE)
            environment(hs.hsgly) %=>% {
                c(x[["uid.wj.wj"]], x[["uid.wj"]][, wj %=>% hs.grepV("~/org/dm/", inv = 1)]) %=>% hs.sub("~/") %=>% cat(sep = "\n", file = wj)
            }
            ssh <- ieiwk.ssh.db(where)
            bin <- switch(osType(), Linux = "rsync", Darwin = "/usr/local/bin/rsync")
            .wk(bin, "-rtDPvhlpgou", "--progress", "-e", .q(paste0("ssh -p ", ssh["port"])), paste0("--files-from=", wj), "~", paste0(ssh["user"], "@", ssh["dz"], ":", environment(hs.hsgly)[["r.wj.wjj.remote"]]))
        })
    }
    if (length(where)) {
        wjj_sub <- "1"
        cmd1 <- .sys(wj_rsync, "-ahPzv --delete", "-c", "--ignore-errors", "-e ssh", "-f", .q("- .DS_Store"), "-f", .q("+ **.[rR]"), "-f", .q("+ **.[rR].[gGxX][zZ]"), "-f", .q("+ **.py"), "-f", .q("+ **.py.[gGxX][zZ]"), "-f", .q("+ **.py.[zZ][iI][pP]"), "-f", .q("+ **.json"), "-f", .q("+ */"), "-f", .q("- **"), paste0("--rsync-path=", .q(paste0("mkdir -p ", .sys.fpGood(wjj2, home.expand = 0), "; rsync"))), paste0(hs.home.expand("~/wk/"), wjj_sub, "/"), paste0(ieiwk.ssh.db(where)["user"], "@", where, 
            ":", paste0(wjj2, wjj_sub, "/")), "| grep -v", .q("^cannot delete non-empty directory: "), "| grep -v", .q("/$"))
        .wk(cmd1)
    }
    if (rp) 
        wk.rsync.out.same_place(hs.home.expand("~/.Rprofile"), ssh = where, L = 1, U = 1)
}
.r820 <- function(...) {
    .Rsync("r820", ...)
}
hostname <- Sys.getenv("HOSTNAME")
r.home <- Sys.getenv("R_HOME")
wk.path <- ieiwk.path <- local({
    a <- osType()
    list(lib = {
        if (a == "Windows") {
            file.path(dirname(hs.home.expand("~")), "wk.os/rjk/R")
        } else if (length(grep("^CYGWIN_NT", a))) {
            hs.wjj("/cygdrive/c/rjk/R")
        } else hs.home.expand(c("~/rj/lib/R", "~/rj2/lib/R"))
    })
})
hs.libPaths <- wk.libPaths <- function(...) {
    .libPaths(unique(c(..., .libPaths())))
}
wk.install_github <- function(..., lib = wk.path[["lib"]][1]) {
    hs.require("devtools", 1)
    install_github(..., lib = lib)
}
ieiwk.workflowInstall <- function(package, lib = ieiwk.path[["lib"]][1], update.Bioconductor = 0) {
    f1 <- ieiwk.file.path("~/my/wd/dm/R/workflows.R")
    if (update.Bioconductor) {
        unlink(f1)
    }
    {
        if (file.exists(f1)) {
            a <- readLines(f1)
        }
        else {
            f2 <- "http://bioconductor.org/workflows.R"
            a <- readLines(f2)
            writeLines(a, con = f1)
        }
    }
    {
        i <- grep("[^'\"]install.packages", a)
        a[i] <- sub("install.packages", "ieiwk.install.packages", a[i])
    }
    if (0) {
        options(repos = c(CRAN = "http://cran.rstudio.com"))
        options(BioC_mirror = "http://bioconductor.org")
    }
    {
        i <- grep("(?i)https", a)
        for (i1 in i) {
            a[i1] <- gsub("https", "http", a[i1])
        }
    }
    eval(parse(text = a))
    workflowInstall(package, lib = lib)
}
hs.library <- ieiwk.library <- function(package) {
    library(package, lib.loc = ieiwk.path[["lib"]])
}
pkgNSisMissing <- function() {
    pkgNS <- NULL
    for (i in seq_along(.libPaths())) {
        libLoc <- .libPaths()[i]
        pkgs <- installed.packages(lib.loc = libLoc)[, 1]
        pkgNS <- c(pkgNS, sapply(unname(pkgs), packageHasNamespace, package.lib = libLoc))
    }
    pkgNS[!pkgNS]
}
.def.1 <- function(p, to = "", env = parent.frame()) {
    eval(body(.t(p, 1, env = env)), env = env)
}
.def.4 <- function(p, v, env = parent.frame()) {
    .t(p = p, actCode = 4, v = v, env = env)
}
.def.3 <- function(p = ".", child = FALSE, level = Inf, env = parent.frame(), format = "vector", sep = formals(.t)$sep, print = TRUE) {
    if (child) 
        return(names(attributes(.t(p, actCode = 1, env = env, self = FALSE))))
    a <- .t(p, actCode = 3, env = env)
    b <- sort(unlist(a))
    if (print) 
        switch(format, vector = .ss(sapply(strsplit(b, sep), function(x) sprintf(".t( c( %s))", paste(shQuote(x), collapse = ", ")))), string = .ss(sprintf(".t( %s)", b)))
    invisible(b)
}
.def.expr <- function(...) {
    list(...)
}
.def.expr.show <- function(progn = NULL, pool = .t(c(".Expr", "all expressions"), env = def.30.04.26.152622)) {
    if (is.null(progn)) 
        progn <- names(pool)
    sapply(progn, function(b) {
        src <- capture.output(attr(pool[[b]], "srcref"))
        .ss(sprintf("\"%s\" = function() {", b))
        .ss(src[-c(1, length(src))])
        .ss("}")
    })
    invisible()
}
.def.expr.eval <- function(progn = NULL, pool = .t(c(".Expr", "all expressions"), env = def.30.04.26.152622), env) {
    if (length(progn)) {
        sapply(progn, function(b) {
            eval(body(pool[[b]]), env = env)
        })
    }
    invisible()
}
.def.save <- function(env = ef.30.04.26.152622, fp = .ma("30.04.26.152817.RData", "~/data/wengkai/handy")) {
    save(env, file = fp)
}
ieiwk <- local({
    ieiwk.dir <- hs.home.expand("~/calvin")
    meta.dir <- hs.home.expand("~/calvin/meta")
    meta.dir.data <- hs.home.expand("~/calvin/data")
    meta.dir.code <- hs.home.expand("~/calvin/code")
    wjj.dm <- switch(system("hostname", intern = TRUE), ThinkPad = , `AFAAW-704251807` = , `wk-VirtualBox` = , `wk-ThinkPad-X220-Tablet` = hs.home.expand("~/wk/my/wd/dm/R"), hs.home.expand("~/wk/my/wd/dm/R"))
    wjj.sj <- hs.wjj(hs.home.expand("~/wk/my/wd"), "sj")
    function(x) {
        switch(x, dir = file.path(hs.home.expand("~"), "calvin"), meta.dir = file.path(ieiwk("dir"), "meta"), meta.dir.data = file.path(ieiwk("dir"), "data"), meta.dir.code = file.path(ieiwk("dir"), "code"), wjj.dm = wjj.dm, wjj.sj = wjj.sj, host.with.internet = local({
            n1 <- Sys.info()["nodename"]
            switch(n1, n1)
        }))
    }
})
.muid.file.path.2 <- function(.muid, dataDir = ieiwk("meta.dir.data"), split = "\\.") {
    a <- strsplit(.muid, split)[[1]]
    p <- ieiwk.file.path(dataDir, a[1], a[2], a[3])
    r <- ieiwk.list.files(p, pattern = head(a, 6) %,% ".")
    if (length(r)) 
        gsub(sprintf("^%s", path.expand(p)), p, r)
    else ieiwk.file.path(p, .muid)
}
ieiwk.move.dirContents <- function(wjj1, wjj2) {
    wjj0 <- getwd()
    setwd(wjj1)
    for (wj1 in list.files(recursive = 1)) {
        .ss(wj1)
        wj2 <- ieiwk.file.path(wjj2, wj1)
        file.rename(wj1, wj2)
    }
    setwd(wjj0)
}
.mfp.ws <- function(.muid, dataDir = ieiwk("meta.dir.data"), split = "\\.") {
    a <- strsplit(.muid, split)[[1]]
    lvl <- 3
    p <- do.call("ieiwk.file.path", as.list(c(dataDir, head(a, lvl))))
    fp <- ieiwk.file.path(p, .muid)
    dir.create(fp)
    fp
}
.mfp.data <- function(.muid, split = "\\.") {
    .muid.file.path(.muid, ieiwk("meta.dir.data"), split)
}
.mfp.code <- function(.muid, split = "\\.") {
    .muid.file.path(.muid, ieiwk("meta.dir.code"), split)
}
.mfp2load <- function(.muid, split = "\\.") {
    hs.load(.muid.file.path(.muid))
}
.mfi <- function(.muid, type = "data") {
    switch(type, data = file.info(.mfp.data(.muid)), code = file.info(.mfp.code(.muid)))
}
.mfr <- function(.muid, type = "data") {
    file.remove(switch(type, data = .mfp.data(.muid), code = .mfp.code(.muid)))
}
ieiwk.muid.ln <- function(.muid, fp.mesa, ..., env = .GlobalEnv, compress = 1) {
    fp.meta <- .muid.file.path(.muid)
    try(if (file.info(fp.mesa)[["isdir"]]) 
        fp.mesa <- ieiwk.file.path(fp.mesa, .muid), sil = 1)
    if (...length()) {
        if (file.exists(fp.meta)) 
            stop(".muid already used")
        save(..., file = fp.mesa, envir = env, compress = compress, compression_level = 1)
        file.symlink(fp.mesa, fp.meta)
    }
}
.mln <- ieiwk.muid.ln
ieiwk.muid.rm <- function(.muid) {
    fp.meta <- .muid.file.path(.muid)
    fp.mesa <- Sys.readlink(fp.meta)
    if (!is.na(fp.mesa)) 
        file.remove(fp.mesa)
    file.remove(fp.meta)
}
.mrm <- ieiwk.muid.rm
.ma.code <- function(.muid) {
    exp <- sprintf("{\nexpression({\n\n.ma( '%s', xxx)\n})\n.ma( '%s')\n}", .muid, .muid)
    .ss(exp)
}
.ma.code.pdf <- function(.muid) {
    src <- sprintf("{\npdf( file = .mfp( '%s.pdf')) # [[%s.pdf][ ]]\nxxx\ndev.off()\n}", .muid, .mfp(.muid))
    .ss(src)
}
ieiwk.muid.access <- ieiwk.muid.save <- function(.muid, root = ieiwk("meta.dir.data"), obj, act, r = 0, d = 0, rc = 0, fp = "", v.only = 0, ow = 0, env = parent.frame(), compress = TRUE, verbose = TRUE) {
    if (fp == "") {
        fp <- .mfp(.muid, root, v.only = 1)
        if (verbose) 
            .ss(sprintf(".ma( \"%s\", \"%s\")", .muid, root))
    }
    if (v.only) 
        return(fp)
    if (d) {
        .sys.do("rm -R", fp)
        return(fp)
    }
    obj.nm <- as.character(substitute(obj))
    if (r) {
        if ({
            !file.exists(fp)
        } & {
            obj.nm != ""
        }) {
            save(list = obj.nm, file = fp, envir = env, compress = compress, compression_level = 1)
        }
        else return(hs.load(fp))
    }
    if (obj.nm != "") {
        if (file.exists(fp) & !ow) {
            .ss(fp)
            stop("file already exists")
        }
        dir.create(dirname(fp), recursive = TRUE)
        save(list = obj.nm, file = fp, envir = env, compress = compress, compression_level = 1)
    }
    else {
        hs.load(fp)
    }
}
.ms <- ieiwk.muid.save
.ma <- ieiwk.muid.access
hs.listAsVector <- function(x, hs, name = 0, flat = FALSE) {
    sapply(x, hs) %=>% unlist(use.names = name, recursive = flat)
}
ieiwk.reorder <- function(x, i, j) {
    if (length(x) == 1) 
        x <- seq(x)
    x1 <- x[-i]
    if (j == 1) {
        c(i, x1)
    }
    else if (j == length(x)) {
        c(x1, j)
    }
    else {
        c(head(x1, j - 1), i, tail(x1, -j + 1))
    }
}
clg <- ieiwk.clg <- function(cl, kc = 1, c = NULL, sep = ".", invert = 0) {
    b <- length(dim(cl))
    if (b == 0) 
        b <- 1
    sep_1 <- glob2rx(sep)
    sep_1 <- substr(sep_1, 2, nchar(sep_1) - 1)
    switch(paste(b), `1` = {
        c1 <- cl %in% kc
        if (invert) c1 <- !c1
        i1 <- which(c1)
        if (length(names(cl))) {
            names(cl)[i1]
        } else i1
    }, `2` = {
        browser("32.12.01.191540")
        if (is.null(rownames(cl))) rownames(cl) <- seq.int(nrow(cl))
        if (is.null(c)) {
            a <- strsplit(as.character(kc), sep_1)
        } else {
            c <- rep(c, len = length(kc))
            a <- lapply(1:length(kc), function(i) {
                list(kc[i], c[i])
            })
        }
        r <- lapply(a, function(a.1) {
            rownames(cl)[cl[, a.1[[1]]] %in% {
                cl[, a.1[[1]]] %keep% a.1[[2]]
            }]
        })
        names(r) <- sapply(a, function(a.1) {
            paste(unlist(a.1), collapse = sep)
        })
        if (length(r) == 1) unlist(r) else r
    })
}
ieiwk.clg2 <- function(cl, kc, c = NULL, sep = ".") {
    b <- length(dim(cl))
    if (b == 0) 
        b <- 1
    sep_1 <- glob2rx(sep)
    sep_1 <- substr(sep_1, 2, nchar(sep_1) - 1)
    if (is.list(cl)) 
        cl <- as.matrix(cl)
    switch(as.character(b), `1` = {
        r <- ifelse(kc %in% cl, 1, NA)
        i.cl <- {
            a <- which(cl %in% kc)
            names(a) <- cl %keep% kc
            a
        }
        r[!is.na(r)] <- names(cl[i.cl[kc %keep% cl]])
        r
    }, `2` = {
        if (is.null(c)) {
            a <- strsplit(as.character(kc), sep_1)
        } else {
            c <- rep(c, len = length(kc))
            a <- lapply(1:length(kc), function(i) {
                list(kc[i], c[i])
            })
        }
        r <- lapply(a, function(a.1) {
            names(cl[, a.1[[1]]] %keep% a.1[[2]])
        })
        names(r) <- sapply(a, function(a.1) {
            paste(unlist(a.1), collapse = sep)
        })
        if (length(r) == 1) unlist(r) else r
    })
}
clg2 <- ieiwk.clg2
ieiwk.in.range <- function(pool, range, inclusive = "both") {
    switch(inclusive, both = {
        fn1 <- fn2 <- "<="
    }, right = {
        fn1 <- "<"
        fn2 <- "<="
    }, left = {
        fn1 <- "<="
        fn2 <- "<"
    }, none = {
        fn1 <- fn2 <- "<"
    })
    if (!is.matrix(range)) 
        range <- t(range)
    apply(range, 1, function(range) which(do.call(fn1, list(range[1], pool)) & do.call(fn2, list(pool, range[2]))))
}
ieiwk.merge.range <- function(rng) {
    rng <- t(apply(rng, 1, sort))
    rng <- rng[order(rng[, 1], rng[, 2]), , drop = FALSE]
    row2do <- nrow(rng)
    while (row2do != 1) {
        a <- rng[row2do - 1, ]
        b <- rng[row2do, ]
        if (b[1] - a[2] < 2) {
            rng <- rng[-row2do, , drop = FALSE]
            rng[row2do - 1, ] <- c(a[1], b[2])
        }
        row2do <- row2do - 1
    }
    rng
}
ieiwk.range.change <- function(r, size) {
    c(min(r) - size, max(r) + size)
}
ieiwk.eql <- function(a, b, tolerance = .Machine$double.eps^0.5) {
    isTRUE(all.equal(a, b, tolerance = tolerance))
}
ieiwk.range.centerRadius <- function(radius, center = 0) {
    radius * c(-1, 1) + center
}
ieiwk.i <- function(x, i, is.list = 0) {
    if (is.list) 
        x[[i]]
    else x[i]
    tfs(sprintf(ifelse(is.list, "x[[ %s]]", "x[ %s]"), i))
}
.i <- ieiwk.i
ieiwk.replace <- function(x, i, y) {
    append(x[-i], y, i[1] - 1)
}
hs.sepSeq <- ieiwk.sepSeq <- function(Lines, sep = "", sep.p, sep.i = {
}, sep.i.pattern, sep.i.1is1 = 1, stepSize = {
}, iv = {
}, jv = {
}, k) {
    if (missing(k)) {
        if (is.null(sep)) {
            if (is.null(jv)) {
                jv <- c(iv[-1] - 1, length(Lines))
            }
        }
        else {
            if (is.null(sep.i)) {
                sep.i <- {
                  if (missing(sep.p)) 
                    which(Lines == sep)
                  else hs.grep(Lines, sep.p)
                }
            }
            if (sep.i.1is1 & sep.i[1] != 1) 
                sep.i <- c(1, sep.i)
            if (length(Lines) >= tail(sep.i, 1)) 
                sep.i <- append(sep.i, length(Lines) + 1)
            block.size <- diff(sep.i)
            sep.i.i <- which(block.size > 0)
            iv <- sep.i[sep.i.i]
            jv <- sep.i[sep.i.i + 1] - 1
        }
    }
    else {
        a <- sort(cbind(seq_along(Lines), seq(k))[, 2])
        iv <- match(unique(a), a)
        jv <- c(iv[-1] - 1, length(a))
    }
    lapply(seq_along(iv), function(i) {
        Lines[iv[i]:jv[i]]
    })
}
ieiwk.i.block <- function(dat, i.sep, iv) {
    L1 <- ieiwk.sepSeq(dat, i.sep)
    lapply(iv, function(i) {
        i.sep.i <- tail(which(i.sep < i), 1)
        L1[[i.sep.i]]
    })
}
hs.connection <- ieiwk.make.connection <- function(con, skip.line.n = 0, skip.char.n = 0, skip.order = c("line", "char"), skip.char.lineSep.count = FALSE) {
    if (!"connection" %in% class(con)) {
        if (is.character(con)) {
            con <- {
                if (file.exists(con)) {
                  file(con)
                }
                else {
                  pipe(con)
                }
            }
        }
        else {
            stop("what is the connection? 34.10.05.070316 @ ~/org/dm/r/function.common.R")
        }
    }
    if (!isOpen(con)) {
        open(con)
    }
    for (o1 in skip.order) {
        switch(o1, line = {
            step_size.max <- 10000
            repeat {
                step_size <- min(skip.line.n, step_size.max)
                a <- readLines(con, step_size)
                if (!length(a)) break
                skip.line.n <- skip.line.n - step_size
            }
        }, char = {
            nchars.remain <- skip.char.n
            while (skip.char.n) {
                a <- readChar(con, skip.char.n)
                if (!length(a)) break
                if (!skip.char.lineSep.count) skip.char.n <- skip.char.n - nchar(gsub("\n", "", a))
            }
        })
    }
    con
}
ieiwk.readChar <- function(con, nchars = 10, useBytes = FALSE, keep.line.sep = FALSE, count.line.sep = FALSE, skip.char.n = 0) {
    con <- ieiwk.make.connection(con)
    r <- list()
    stepSize.max <- 1e+09
    while (nchars) {
        stepSize <- min(stepSize.max, nchars)
        a <- readChar(con, stepSize)
        if (!length(a)) 
            break
        nchars <- nchars - nchar(a)
        if (!count.line.sep) 
            nchars <- nchars + str_count(a, "\n")
        r[length(r) + 1] <- a
    }
    r <- do.call(paste0, r)
    if (!keep.line.sep) 
        gsub("\n", "", r)
    else r
}
ieiwk.readChar.all <- function(con, stepSize = 1e+09, useBytes = FALSE, skip.char.n = 0) {
    con <- ieiwk.make.connection(con)
    r <- list()
    if (skip.char.n) {
        readChar(con, skip.char.n)
    }
    while (length(a <- readChar(con, stepSize))) {
        r[[length(r) + 1]] <- a
    }
    do.call(paste0, r)
}
wk.read.chunk <- local({
    db1 <- list()
    db2 <- list()
    function(wj, chunkSize = 1e+08, redo = 0, reset = 0) {
        if (is.character(wj)) {
            con_id <- wj
            if (reset) {
                if (length(db1[[wj]])) {
                  .try(close(db1[[wj]]))
                }
                db1[[wj]] <<- {
                }
                db2[[con_id]] <<- {
                }
                return()
            }
            if (redo) {
                if (length(db1[[wj]])) {
                  .try(close(db1[[wj]]))
                  db1[[wj]] <<- {
                  }
                }
                db2[[con_id]] <<- {
                }
            }
            if (length(db1[[wj]]) == 0) {
                db1[[wj]] <<- {
                  if (file.exists(wj)) {
                    file(wj)
                  }
                  else {
                    pipe(wj)
                  }
                }
            }
            con1 <- db1[[wj]]
        }
        else if ("connection" %in% class(wj)) {
            con1 <- wj
            con_id <- capture.output(attr(wj, "conn_id"))
        }
        else {
            wk.browser("34.10.05.162756 @ ~/org/dm/r/function.common.R")
        }
        if (!isOpen(con1)) 
            open(con1)
        reader1 <- {
            if (length(db2[[con_id]]) == 0) 
                db2[con_id] <<- list(chunk.reader(con1))
            db2[[con_id]]
        }
        chunk1 <- read.chunk(reader1, max.size = chunkSize)
        if (length(chunk1)) {
            rawToChar(chunk1)
        }
        else {
            db2[[con_id]] <<- {
            }
            close(con1)
            db1[[wj]] <<- {
            }
        }
    }
})
wk.fread.chunk.1 <- ieiwk.fread.chunk <- function(con, stepSize = 1e+08, useBytes = FALSE, skip.char.n = 0, gz = 0, ...) {
    chunk <- wk.read.chunk(con)
    if (length(chunk)) 
        return(wk.fread(chunk, ...))
    if (skip.char.n) {
        readChar(con, skip.char.n)
    }
    a <- readChar(con, stepSize, useBytes = TRUE)
    a1 <- readLines(con, 1)
    if (1) {
        return(ieiwk.fread(sprintf("%s%s\n", a, a1), ...))
    }
    if (0) {
        tt <- tempfile(tmpdir = "/dev/shm")
        on.exit(unlink(tt), add = TRUE)
        cat(a, a1, "\n", sep = "", file = tt)
        if (file.info(tt)[["size"]] < 2) 
            return()
        ieiwk.fread(tt, ...)
    }
}
wk.fread.chunk.2 <- function(con, stepSize = 1e+09, complete.line = 1, useBytes = FALSE, skip.char.n = 0, ...) {
    con <- hs.make.connection(con)
    if (skip.char.n) 
        readChar(con, skip.char.n)
    a <- readChar(con, stepSize)
    a1 <- readLines(con, 1)
    wk.fread(paste0(a, a1, "\n"), ...)
}
wk.fread.chunk.3 <- function(con, stepSize = 1e+08, useBytes = FALSE, skip.char.n = 0, ...) {
    con <- hs.make.connection(con)
    if (skip.char.n) 
        readChar(con, skip.char.n)
    a <- readChar(con, stepSize)
    a1 <- readLines(con, 1)
    a.len <- str_length(a)
    substr(a, a.len + 1, a.len + 2 + str_length(a1)) <- sprintf("%s\n", a1)
    wk.fread(a, ...)
}
wk.fread.chunk.4 <- function(con, nrows = 1e+05, ...) {
    con <- hs.make.connection(con)
    wj_tmp <- hs.wj.temp()
    on.exit(hs.rm.wj(wj_tmp), add = TRUE)
    a1 <- readLines(con, nrows)
    writeLines(a1, wj_tmp)
    wk.fread(wj_tmp, ...)
}
wk.fread.chunk.5 <- local({
    wj_tmp <- {
    }
    remaining <- ""
    function(con, stepSize = 1e+08, pre = {
    }, ...) {
        con <- hs.make.connection(con)
        if (length(wj_tmp) == 0) 
            wj_tmp <<- hs.wj.temp()
        if (length(pre)) {
            sink(wj_tmp, append = TRUE)
            writeChar(pre, stdout(), eos = {
            })
            sink()
        }
        a1 <- readChar(con, stepSize)
        if (length(a1)) {
            sink(wj_tmp, append = TRUE)
            writeChar(a1, stdout(), eos = {
            })
            sink()
        }
        else close(con)
        repeat {
            a2 <- readChar(con, 1000)
            if (length(a2)) {
                if (hs.grepl(a2, "\n$")) {
                  sink(wj_tmp, append = TRUE)
                  writeChar(a2, stdout())
                  sink()
                  break
                }
                else {
                  v1 <- strsplit(a2, "[\r\n]")[[1]]
                  if (length(v1) > 1) {
                    sink(wj_tmp, append = TRUE)
                    writeLines(head(v1, -1))
                    sink()
                    remaining <<- tail(v1, 1)
                    break
                  }
                  else {
                    sink(wj_tmp, append = TRUE)
                    writeChar(a2, stdout())
                    sink()
                  }
                }
            }
            else break
        }
        if (file.exists(wj_tmp) && file.size(wj_tmp)) {
            on.exit({
                sink(wj_tmp, append = FALSE)
                writeChar(remaining, stdout(), eos = {
                })
                sink()
                remaining <<- ""
            }, add = TRUE)
            wk.fread(wj_tmp, ...)
        }
        else {
            hs.rm.wj(wj_tmp)
            remaining <<- ""
            {
            }
        }
    }
})
wk.fread.chunk.6 <- local({
    remaining <- ""
    function(con, stepSize = 1e+08, ...) {
        con <- hs.make.connection(con)
        a1 <- readChar(con, stepSize)
        if (length(a1)) {
            if (nchar(remaining)) 
                a1 <- sprintf("%s%s", remaining, a1)
        }
        else {
            close(con)
            if (nchar(remaining)) {
                remaining.nc <- nchar(remaining)
                a1 <- if (substr(remaining, remaining.nc, remaining.nc) == "\n") {
                  remaining
                }
                else sprintf("%s\n", remaining)
            }
            else return()
        }
        a1.nc <- nchar(a1)
        if (substr(a1, a1.nc, a1.nc) == "\n") {
            remaining <<- ""
            suppressWarnings(wk.fread(a1, ...))
        }
        else {
            lei <- gregexpr("\n", a1, useBytes = TRUE)[[1]]
            remaining <<- substr(a1, tail(lei, 1) + 1, a1.nc)
            suppressWarnings(wk.fread(a1, nrow = length(lei), ...))
        }
    }
})
wk.fread.chunk.7 <- function(con, nrows = 1e+06, dt = 1, ...) {
    con <- hs.make.connection(con)
    a1 <- readLines(con, nrows)
    if (dt && length(a1)) 
        a1 <- fread(text = a1, header = FALSE)
    a1
}
wk.fread.chunk.8 <- local({
    left <- {
    }
    column_name <- {
    }
    function(in1, nbits = 1e+06, sep_n = charToRaw("\n"), ot = "dt", long_line = 0, cn = ot == "dt", ...) {
        in1 <- hs.make.connection(in1, "rb")
        if ("binary" %not.in% summary(in1)[["text"]]) 
            stop("37_03_05_201133")
        a1 <- readBin(in1, "raw", nbits)
        if (is.null(left) && cn) {
            ni <- match(sep_n, if (long_line) 
                a1
            else head(a1, 10000))
            column_name <<- strsplit(readBin(head(a1, ni - 1), "character"), "\t")[[1]]
            a1 <- tail(a1, -ni)
        }
        if (length(left) > 0) {
            if (length(a1) > 0) {
                ni <- hs.match.from_tail(sep_n, a1)
                if (ni < length(a1)) {
                  left2 <- tail(a1, length(a1) - ni)
                  a1 <- c(left, head(a1, ni))
                  left <<- left2
                }
                else {
                  a1 <- c(left, a1)
                  left <<- raw()
                }
            }
            else {
                a1 <- if (tail(left, 1) == sep_n) 
                  left
                else c(left, sep_n)
                left <<- NULL
            }
        }
        else {
            if (length(a1) > 0) {
                ni <- hs.match.from_tail(sep_n, a1)
                if (ni < length(a1)) {
                  left <<- tail(a1, -ni)
                  a1 <- head(a1, ni)
                }
            }
            else {
                left <<- NULL
                column_name <<- NULL
                close(in1)
                return()
            }
        }
        switch(ot, dt = {
            jg <- fread(text = readBin(a1, "character"), header = FALSE)
            if (cn) setnames(jg, column_name)
            jg
        }, row = fread(readBin(a1, "character"), sep = "\n", header = FALSE), raw = a1)
    }
})
wk.fread.chunk <- wk.fread.chunk.8
wk.dt.inList <- function() {
    methods <- list()
    dtList <- list()
    check <- 0
    methods$set <- function(dtl) {
        dtList <<- lapply(dtl, . %=>% {
            attr(x, "rows") <- x[, seq.int(.N)]
            x
        })
        check <<- 1
    }
    methods$add <- function(dt) {
        if (dt[, .N]) {
            attr(dt, "rows") <- dt[, seq.int(.N)]
            dtList[[length(dtList) + 1]] <<- dt
            check <<- 1
        }
    }
    methods$clean <- function() {
        if (check == 0) 
            return()
        l1 <- which(sapply(dtList, . %=>% attr("rows") %=>% length) == 0)
        if (length(l1)) 
            dtList[i1] <<- {
            }
        check <<- 0
    }
    methods$nrow <- function() sapply(dtList, . %=>% attr("rows") %=>% length) %=>% sum
    methods$ncol <- function() {
        methods$clean()
        length(dtList[[1]])
    }
    methods$getRows <- function(i, rm = 0) {
        if (length(i) == 0) 
            return()
        methods$clean()
        i1 <- sapply(dtList, . %=>% attr("rows") %=>% length) %=>% cumsum
        g1 <- cut(i, c(0, i1), labels = seq_along(i1))
        i2 <- tapply(i, g1, c) %=>% hs.nonempty
        for (g2 in names(i2) %not% "1") i2[[g2]] %<=>% {
            x - i1[as.integer(g2) - 1]
        }
        if (rm) {
            check <<- 1
            for (g2 in names(i2)) attr(dtList[[as.integer(g2)]], "rows") <<- attr(dtList[[as.integer(g2)]], "rows")[-sort(unique(i2[[g2]]))]
        }
        a <- names(i2) %=>% lapply(. %=>% dtList[[as.integer(x)]][i2[[x]]]) %=>% rbindlist
        if (isSorted(i)) {
            a
        }
        else a[seq.int(.N) %=>% match(order(g1))]
    }
    methods$get <- function(i = seq_along(dtList), rbind = 1) {
        methods$clean()
        R <- lapply(dtList[i], function(x) {
            i <- attr(x, "rows")
            if (length(i) == x[, .N]) {
                x
            }
            else x[i]
        })
        if (rbind) {
            rbindlist(R)
        }
        else R
    }
    methods$do <- function(hs, ex, sep = 0) {
        methods$clean()
        if (sep) {
            if (missing(hs)) {
                lapply(seq_along(dtList), function(i1) {
                  dtList[[i1]] %=>% {
                    l1 <- attr(x, "rows")
                    if (.[, .N == length(l1)]) {
                      x[, eval(ex)]
                    }
                    else x[l1, eval(ex)]
                  }
                })
            }
            else lapply(seq_along(dtList), function(i1) {
                dtList[[i1]] %=>% {
                  l1 <- attr(x, "rows")
                  if (x[, .N == length(l1)]) {
                    hs(x)
                  }
                  else hs(x[l1])
                }
            })
        }
        else if (missing(hs)) {
            methods$get()[, eval(ex)]
        }
        else hs(methods$get())
    }
    methods
}
wk.fread.chunk.do.line.1 <- local({
    cn1 <- {
    }
    function(io, hs = identity, hs.cn = function(chunk) if (sep != "\n") 
        names(fread(chunk, nrow = 0, header = TRUE, sep = sep)), all = 0, simplify = all, header = sep != "\n", sep = "\t", smart = 1, ...) {
        if (smart) 
            io <- hs.make.connection(io)
        if (header && (length(cn1) == 0)) {
            chunk1 <- readChar(io, 1e+05)
            i1 <- regexpr("\n", chunk1)
            if (length(i1)) {
                cn1 <<- hs.cn(chunk1)
                on.exit(cn1 <<- {
                }, add = TRUE)
                environment(wk.fread.chunk.6)[["remaining"]] <- substr(chunk1, i1 + 1, nchar(chunk1))
            }
            else stop("incomplete line read @34.12.06.085808")
        }
        R <- expandingList()
        Args <- c(list(io), sep = sep, header = FALSE, ...)
        if (all) 
            Args["stepSize"] <- list(rep(1e+09, 1000))
        repeat {
            L <- .try(do.call(wk.fread.chunk.6, Args))
            if ((length(L) == 0) || (class(L) == "try-error")) 
                break
            if (length(cn1)) 
                setnames(L, cn1)
            t1 <- system.time(r1 <- hs(L), gcFirst = FALSE)
            if (t1["elapsed"] < 3) 
                warning("too short running time @34.12.06.094358")
            R$add(r1)
        }
        R <- R$get()
        if (simplify) {
            R[[1]]
        }
        else R
    }
})
wk.fread.chunk.do.line.2 <- local({
    cn1 <- {
    }
    function(io = {
    }, hs = identity, hs.cn = function(io) readLines(io, 1) %=>% strsplit(sep) %=>% unlist, nrows = 1e+05, all = 0, simplify = all, header = sep != "\n", sep = "\t", smart = 1, clear.cn = length(io), co1 = Inf, co.hs = length, co.acc = 0, dt = sep != "\n", reader = wk.fread.chunk.7, ...) {
        if (clear.cn) {
            cn1 <<- {
            }
        }
        if (!length(io)) 
            return()
        if (smart) 
            io <- hs.make.connection(io)
        if (header && (length(cn1) == 0)) 
            cn1 <<- hs.cn(io)
        on.exit(cn1 <<- {
        }, add = TRUE)
        R <- expandingList()
        Args <- c(list(io), nrows = nrows, sep = sep, dt = dt, ...)
        if (all) 
            Args["stepSize"] <- list(-1L)
        repeat {
            if (co.acc >= co1) 
                break
            L <- do.call(reader, Args)
            if (length(L) == 0) {
                break
            }
            if (length(cn1)) 
                setnames(L, cn1)
            t1 <- system.time(r1 <- hs(L), gcFirst = FALSE)
            co.acc %<=>% {
                x + co.hs(r1)
            }
            if (t1["elapsed"] < 3) 
                warning("too short running time @34.12.06.094358@")
            R$add(r1)
        }
        R <- R$get()
        if (simplify) {
            R[[1]]
        }
        else R
    }
})
wk.fread.chunk.do.line.3 <- local({
    cn1 <- {
    }
    function(io = {
    }, hs = identity, hs.cn = function(io) readLines(io, 1) %=>% strsplit(sep) %=>% unlist, nrows = 1e+05, all = 0, simplify = all, header = sep != "\n", sep = "\t", smart = 1, clear.cn = length(io), co1 = Inf, co.hs = length, co.acc = 0, dt = sep != "\n", reader = wk.fread.chunk.7, ...) {
        if (clear.cn) {
            cn1 <<- {
            }
        }
        if (!length(io)) 
            return()
        if (smart) 
            io <- hs.make.connection(io)
        if (header && (length(cn1) == 0)) 
            cn1 <<- hs.cn(io)
        on.exit(cn1 <<- {
        }, add = TRUE)
        R <- expandingList()
        Args <- c(list(io), nrows = nrows, sep = sep, dt = dt, ...)
        if (all) 
            Args["stepSize"] <- list(-1L)
        repeat {
            if (co.acc >= co1) 
                break
            L <- do.call(reader, Args)
            if (length(L) == 0) {
                break
            }
            if (length(cn1)) 
                setnames(L, cn1)
            t1 <- system.time(r1 <- hs(L), gcFirst = FALSE)
            co.acc %<=>% {
                x + co.hs(r1)
            }
            if (t1["elapsed"] < 3) 
                warning("too short running time @34.12.06.094358@")
            R$add(r1)
        }
        R <- R$get()
        if (simplify) {
            R[[1]]
        }
        else R
    }
})
wk.fread.chunk.do.line <- wk.fread.chunk.do.line.2
wk.fread.chunk.do.lines <- local({
    L <- {
    }
    function(io, hs, hs.sep.i, hs.sep = hs.sepLines, hs.sep.arg = {
    }, hs.stop = {
    }, ...) {
        L <<- {
        }
        if ("connection" %not.in% class(io)) {
            if (file.exists(io)) {
                io %<=>% gzfile("r")
            }
            else io %<=>% pipe("r")
        }
        on.exit(close(io), add = TRUE)
        R <- expandingList()
        L <- c()
        miss <- 0
        repeat {
            if (miss > 3) 
                hs.browser("Too many times reading without doing anything. 34.06.29.173224")
            L1 <- wk.fread.chunk(io, dt = 0, ...)
            if (length(L1)) 
                if (length(L)) {
                  L %<=>% c(L1)
                }
                else L <- L1
            if (length(L) == 0) 
                break
            parts <- do.call(hs.sep, append(list(L, hs.i = hs.sep.i, at.last = !length(L1)), hs.sep.arg))
            if (length(parts[[1]])) {
                R$add(do.call(hs, parts[-2]))
                miss <- 0
            }
            if (length(hs.stop) && hs.stop(L)) 
                break
            if (identical(parts[[2]], L) && length(L1)) 
                miss <- miss + 1
            L <- parts[[2]]
        }
        R$get()
    }
})
hs.sepLines <- function(., hs.i, bs = 1, keep.i = bs > 1, at.last = 0) {
    parts <- vector("list", 2 + as.logical(keep.i))
    i1 <- hs.i(.)
    bn <- (length(i1) - ifelse(at.last, 0, 1))%/%bs
    if (bn) {
        i1.end <- bn * bs
        i2 <- ifelse(at.last, length(.), i1[i1.end + 1] - 1)
        parts[[1]] <- .[seq(i1[1], i2)]
        parts[[2]] <- tail(., length(.) - i2)
        if (keep.i) {
            parts[[3]] <- head(i1, i1.end) %=>% {
                if (x[1] > 1) {
                  x <- x - (x[1] - 1)
                }
                x
            }
        }
    }
    else {
        parts[[2]] <- .
    }
    parts
}
wk.fread.chunk.do <- function(io, hs, hs.sep, ...) {
    if ("connection" %not.in% class(io)) 
        io %<=>% pipe("r")
    on.exit(close(io), add = TRUE)
    R <- expandingList()
    tmp <- expandingList()
    L <- wk.dt.inList()
    it <- 0
    repeat {
        L1 <- ieiwk.fread.chunk(io, ...)
        if (missing(hs.sep)) {
            if (length(L1) == 0) 
                break
        }
        else {
            if (length(L1)) 
                L$add(L1)
            if (L$nrow() == 0) 
                break
            i1 <- L$do(hs.sep)
            if (length(i1)) 
                R$add(hs(L$getRows(i1, rm = 1)))
        }
        it %<=>% {
            x + 1
        }
        .ss(it)
        if (0) 
            if (it == 9) 
                hs.browser("34.06.28.225616")
    }
    R$get()
}
ieiwk.readChar.chunk <- function(con, stepSize = 1e+09, complete.line = 1, useBytes = FALSE, skip.char.n = 0) {
    con <- ieiwk.make.connection(con)
    if (skip.char.n) {
        readChar(con, skip.char.n)
    }
    a <- readChar(con, stepSize)
    browser("33.04.21.122438")
    tt <- tempfile(tmpdir = "/dev/shm")
    if (length(a)) {
        a.len <- str_length(a)
        if (grepl("\n$", substring(a, a.len))) {
            a1 <- paste0(readLines(con, 1), "\n")
            system.time(substr(a, a.len + 1, a.len + 1 + str_length(a1)) <- a1)
            system.time(a <- paste0(a, readLines(con, 1), "\n"))
            system.time(a <- sprintf("%s%s\n", a, readLines(con, 1)))
            system.time(cat(a, file = tt))
            cat(a1, file = tt, append = 1)
            readChar()
            .less(tt)
            unlink(tt)
        }
    }
    a
}
ieiwk.readChar.chunk.all <- function(con, stepSize = 1e+09, complete.line = 1, useBytes = FALSE, skip.char.n = FALSE) {
    con <- ieiwk.make.connection(con)
    r <- list()
    if (skip.char.n) {
        readChar(con, skip.char.n)
    }
    while (length(a <- ieiwk.readChar.chunk(con, stepSize = stepSize))) {
        r[[length(r) + 1]] <- a
    }
    r
}
ieiwk.nrc <- function(n, L) {
    a <- n%%L
    b <- n/L
    if (a) {
        c(ceiling(b), a)
    }
    else {
        c(b, L)
    }
}
ieiwk.nrc.seq <- function(i, j = NULL, S) {
    if (is.null(j)) {
        if (length(i) == 1) {
            j <- i
        }
        else {
            j <- i[2]
            i <- i[1]
        }
    }
    if (i > j) 
        stop("小的在左")
    L <- nchar(S[1])
    i.xy <- ieiwk.nrc(min(i, j), L)
    j.xy <- ieiwk.nrc(max(i, j), L)
    switch(as.character(diff(c(i.xy[1], j.xy[1]))), `0` = substr(S[i.xy[1]], i.xy[2], j.xy[2]), `1` = {
        paste(c(substr(S[i.xy[1]], i.xy[2], L), substr(S[j.xy[1]], 1, j.xy[2])), collapse = "")
    }, paste(c(substr(S[i.xy[1]], i.xy[2], L), S[seq.int(i.xy[1] + 1, j.xy[1] - 1)], substr(S[j.xy[1]], 1, j.xy[2])), collapse = ""))
}
hs.pick <- hs.select <- function(x, hs, v = 0, inv = 0) {
    if (missing(hs)) 
        return(x)
    i1 <- {
        if ("function" %not.in% class(hs)) {
            hs
        }
        else if (v) {
            hs(x)
        }
        else sapply(x, hs)
    }
    if (inv) 
        i1 <- !i1
    x[i1]
}
"%no.less.than%" <- function(x, y) as.numeric(x) >= as.numeric(y)
"%larger.than%" <- function(x, y) as.numeric(x) > as.numeric(y)
"%smaller.than%" <- function(x, y) as.numeric(x) < as.numeric(y)
"%no.larger.than%" <- function(x, y) as.numeric(x) <= as.numeric(y)
ieiwk.tvvt <- function(tv) {
    a <- names(tv)
    names(a) <- tv
    a
}
inverseList <- function(l) {
    rId <- unlist(l, use.names = FALSE)
    lId <- rep(names(l), sapply(l, length))
    return(split(lId, rId))
}
ieiwk.intersect <- function(...) {
    a <- list(...)
    repeat {
        if (length(a) < 2) 
            break
        a[[2]] <- intersect(a[[1]], a[[2]])
        a <- a[-1]
    }
    a[[1]]
}
tt <- function(x) table(table(x))
.lu <- function(x) length(unique(x))
.luu <- function(x) length(unique(unlist(x)))
nm <- names
"%|%" <- function(x, y) {
    .sys(x, "|", y)
}
"%(%" <- function(x, y, env1 = parent.frame()) {
    if (!is.function(x)) {
        x <- paste(c(x), collapse = "")
    }
    ye <- substitute(y)
    if (("call" %in% class(ye)) && (length(ye) > 1)) {
        ye <- as.call(c(list, as.list(ye)[-1]))
    }
    y <- eval(ye, env1)
    do.call(x, as.list(y))
}
"%(c%" <- function(x, y, env1 = parent.frame()) {
    if (!is.function(x)) {
        x <- paste(c(x), collapse = "")
    }
    do.call(x, as.list(y))
}
"%(unname%" <- function(x, y) {
    do.call(x, unname(as.list(y)))
}
"%)%" <- function(y, x) {
    if (!is.list(y)) 
        y <- as.list(y)
    do.call(x, y)
}
`%xor%` <- function(x, y) setdiff(union(x, y), intersect(x, y))
"%overlap%" <- function(r1, r2) {
    !{
        r1[2] < r2[1] | r1[1] > r2[2]
    }
}
"%within%" <- function(x, r) {
    r <- sort(r)
    x >= r[1] & x <= r[2]
}
"%!in%" <- function(x, y) {
    !(x %in% y)
}
wk.which <- function(l, inv = 0, na = 1) {
    if (inv) {
        if (na) {
            i <- which(l)
            if (length(i)) {
                seq_along(l)[-i]
            }
            else seq_along(l)
        }
        else which(!l)
    }
    else {
        if (na) {
            wk.which(!l, inv = 1, na = 1)
        }
        else {
            which(l)
        }
    }
}
ieiwk.within.core <- function(x, range, left = ">=", right = "<=", output = "position") {
    range <- sort(range)
    i <- do.call(left, list(x, range[1])) & do.call(right, list(x, range[2]))
    switch(output, position = i, content = x[i])
}
hs.within <- wk.within <- ieiwk.within <- function(x, ranges, left = ">=", right = "<=", output = "position") {
    if (!is.list(ranges)) {
        if (length(dim(ranges))) {
            ranges2 <- lapply(seq(nrow(ranges)), function(i) ranges[i, ])
            if (length(rownames(ranges))) 
                names(ranges2) <- rownames(ranges)
            ranges <- ranges2
        }
        else ranges <- list(ranges)
    }
    lapply(ranges, function(range) {
        ieiwk.within.core(x, range)
    }) %=>% as.data.table
}
ieiwk.sample <- function(pool, sizes) {
    r <- c()
    while (length(sizes)) {
        i <- sample(seq.int(length(pool)), sizes[1])
        r[[length(r) + 1]] <- pool[i]
        pool <- pool[-i]
        sizes <- tail(sizes, -1)
    }
    r
}
ieiwk.combn <- function(x, y, sep = "|") {
    as.vector(sapply(x, function(x) {
        sprintf("%s%s%s", x, sep, y)
    }))
}
wk.combn <- function(x, m = 2, sep, diag = 1) {
    if (!diag) 
        return(combn(x, m) %=>% t %=>% data.table)
    g <- seq_along(x) %=>% list %=>% rep(m) %=>% (`%)%`(expand.grid))
    for (ci in hs.seqForward(1, m - 1)) {
        ri <- which(g[, ci] <= g[, ci + 1])
        g <- g[ri, ]
    }
    hs.browser("2023.08.04.213453", debug = 0)
    R <- g %=>% data.table %=>% setkey %=>% {
        x[, lapply(.SD, . %>% x[.])]
    } %=>% setnames(x, paste("V", seq.int(length(x)), sep = ""))
    if (!missing(sep)) 
        R %<=>% ieiwk.sprintf(sep)
    R
}
wk.sapply <- ieiwk.sapply <- function(x, hs, ..., mc = 1, nc = ifelse(mc, hs.NC(), 1), nc.smart = 1, finally = {
}) {
    if (!length(x)) 
        return()
    if (nc == "max") 
        nc <- length(x)
    Names <- {
        if (length(names(x))) {
            names(x)
        }
        else if ("character" %in% class(x)) {
            x
        }
        else if (is.numeric(x)) {
            as.character(x)
        }
    }
    hs <- match.fun(hs)
    R <- {
        if (identical(hs, match.fun("switch"))) {
            hs2 <- substitute(list(...))
            if (nc.smart && (capture.output(hs2) %=>% wk.re("^\\s*wk\\.browser\\(", "detect") %=>% any)) 
                nc <- 1
            c1 <- hs2 %=>% as.list %=>% {
                x[[1]] <- as.symbol("switch")
                x
            }
            if (0) {
                wk.sapply(names(x), . %>% {
                  a <- .
                  hs2[[.]] %>% {
                    if (typeof(.) %in% c("language")) 
                      .
                    else match.fun(.)
                  } %>% {
                    eval(bquote({
                      hs <- .(.)
                      hs(x[[a]])
                    }))
                  }
                }, mc = mc, nc = nc)
            }
            hs.browser("2023.08.04.213640", debug = 0)
            wk.sapply(names(x), . %=>% {
                a <- x
                c1 %=>% append(a, 1) %=>% as.call %=>% eval %=>% {
                  if (length(x)) 
                    {
                      if (typeof(x) %in% c("language")) 
                        x
                      else match.fun(x)
                    } %>% {
                      .(x[[a]])
                    }
                }
            }, mc = mc, nc = nc)
        }
        else {
            if (nc.smart && (ifelse("fseq" %in% class(hs), functions, identity)(hs) %>% capture.output %>% wk.re("^\\s*(wk\\.browser\\(|<<-)", "detect", do.group = 0) %>% any)) {
                nc <- 1
            }
            a <- wk.mclapply(x, hs, ..., mc.cores = nc)
            if (length(Names)) 
                names(a) <- Names
            a
        }
    }
    if (!missing(finally)) 
        eval(finally)
    R
}
hs.lapply <- wk.lapply <- function(x1, FUN, ..., simplify = FALSE, smart = TRUE, mz = {
}, x1.sep = ",") {
    if ((length(x1) == 1) && is.character(x1)) 
        x1 <- strsplit(x1, ",")[[1]]
    mz <- if (smart) {
        if (length(names(x1))) {
            names(x1)
        }
        else if (is(x1, "character")) {
            x1
        }
        else if (is.numeric(x1)) {
            as.character(x1)
        }
    }
    if (is.symbol(substitute(FUN))) 
        FUN <- paste(substitute(FUN))
    jg <- sapply(x1, FUN, ..., simplify = simplify)
    if (length(mz)) 
        names(jg) <- mz
    jg
}
wk.apply <- function(x, m, hs, ..., mc = 1, nc = ifelse(mc, 8, 1)) {
    hs <- match.fun(hs)
    hs.switch(m, 1, wk.sapply(seq(nrow(x)), function(.) hs(x[., ], ...), mc = mc, nc = nc) %=>% {
        if (!length(names(x))) 
            names(x) <- rownames(x)
        x
    }, 2, wk.sapply(seq(ncol(x)), function(.) hs(x[, ., with = FALSE][[1]], ...), mc = mc, nc = nc) %=>% {
        names(x) %<=>% {
            if (length(x)) 
                x
            else colnames(x)
        }
        x
    })
}
wk.tapply <- function(x, i, hs, ..., mc = 1, nc = ifelse(mc, hs.NC(), 1), chk.i = 1) {
    hs <- match.fun(hs)
    if (is.factor(i) && chk.i) {
        if (levels(i) %=>% {
            (length(x) > length(i)) || length(x %not% i)
        }) 
            i %<=>% as.character
    }
    j <- tapply(seq_along(x), as.vector(i), c)
    if (identical(hs, match.fun("switch"))) {
        hs2 <- substitute(list(...))
        if (capture.output(hs2) %=>% wk.re("^[^#]*\\bbrowser\\(", "detect") %=>% any) 
            nc <- 1
        c1 <- hs2 %=>% as.list %=>% {
            x[[1]] <- as.symbol("switch")
            x
        }
        . <- "gene"
        wk.sapply(names(j), . %>% {
            a <- .
            c1 %>% append(a, 1) %>% as.call %>% eval %>% {
                if (typeof(.) %in% c("language")) 
                  .
                else match.fun(.)
            } %>% {
                .(x[j[[a]]])
            }
        }, mc = mc, nc = nc)
    }
    else {
        if (capture.output(hs) %>% wk.re("^[^#]*\\bbrowser\\(", "detect") %>% any) 
            nc <- 1
        wk.sapply(j, . %>% {
            hs(x[.], ...)
        }, mc = mc, nc = nc)
    }
}
ieiwk.rapply.get <- function(l, i) {
    r <- list()
    if (missing(i)) {
        rapply(l, function(x) {
            r <<- append(r, list(x))
        })
    }
    else {
        it <- 1
        rapply(l, function(x) {
            if (it == i) 
                r <<- append(r, list(x))
            it <<- it + 1
        })
    }
    r
}
wk.rapply.flat <- local({
    acc <- {
    }
    hs1 <- function(x, node = {
    }, sep = ":") {
        if ("list" %in% class(x)) {
            sapply(names(x), . %>% {
                hs1(x[[.]], c(node, .))
            })
        }
        else {
            acc[["add"]](node %>% paste(collapse = sep), x)
            acc %>% names
        }
    }
    function(x, node = {
    }, sep = ":") {
        acc <<- namedExpandingList()
        hs1(x = x, node = node, sep = sep)
        acc[["get"]]()
    }
})
ieiwk.replace.stepwise <- function() local({
    i.considered <- {
    }
    function(d1, i1, c1, reset = 0) {
        if (reset) 
            i.considered <<- {
            }
        i1 <- i1 %not% i.considered
        if (length(i1)) {
            d1[i1] <- c1
            i.considered <<- unique(c(i.considered, i1))
        }
        d1
    }
})
hs.once <- wk.once <- function(x, inv = 0, out = "x") {
    i <- duplicated(x)
    i <- x %in% x[i]
    if (!inv) 
        i <- !i
    switch(out, i = i, x = x[i])
}
hs.once_simple <- function(x) {
    i <- duplicated(x)
    x %not% x[i]
}
hs.twice <- wk.twice <- function(x, out = "x") {
    R <- hs.once(x[duplicated(x)])
    if (out == "i") 
        R <- x %in% R
    R
}
wk.make.hash_table <- function() {
    env1 <- new.env()
    evalq(h1 <- new.env(), env1)
    evalq(V <- expandingVector(), env1)
    evalq(hasher <- function(v1) {
        if (is.character(v1) && (length(v1) == 1)) {
            v1
        }
        else rlang::hash(v1)
    }, env1)
    evalq(set <- function(v1, k1) {
        v1 <- hasher(v1)
        V[["add"]](v1)
        h1[[v1]] <- k1
    }, env1)
    evalq(has <- function(v1) {
        v1 <- hasher(v1)
        exists(v1, envir = h1, inherits = FALSE)
    }, env1)
    evalq("+=" <- function(v1, k1, n1 = 0) {
        v1 <- hasher(v1)
        n0 <- if (has(v1)) {
            h1[[v1]]
        }
        else {
            V[["add"]](v1)
            n1
        }
        h1[[v1]] <- k1 + n0
    }, env1)
    evalq(get <- function(v1) {
        if (missing(v1)) {
            v1 <- V[["get"]]() %and% ls(envir = h1, all.names = TRUE, sorted = FALSE)
            mget(v1, h1)
        }
        else {
            v1 <- hasher(v1)
            h1[[v1]]
        }
    }, env1)
    evalq(mget <- function(v1) {
        if (missing(v1)) 
            v1 <- V[["get"]]() %and% ls(envir = h1, all.names = TRUE, sorted = FALSE)
        mget(as.character(v1), h1)
    }, env1)
    evalq(pop <- del <- function(v1) {
        v1 <- hasher(v1)
        if (has(v1)) {
            k1 <- h1[[v1]]
            rm(list = v1, envir = h1)
            k1
        }
    }, env1)
    env1
}
wk.strCutL <- function(x, L) {
    M <- cumsum(L) %=>% {
        cbind(c(1, head(x, -1) + 1), x)
    }
    substr(x, M[, 1], M[, 2])
}
wk.strCutSL <- function(x, S, L) {
    substr(x, S, S + L - 1)
}
ieiwk.string.col <- function(str, col = 1, sep = "\t") {
    if (!length(col)) 
        return()
    col.0 <- col
    col <- sort(unique(col))
    acc <- matrix(nrow = length(str), ncol = length(col))
    colnames(acc) <- col
    if (max(col) < 3) {
        for (c1 in as.character(col)) {
            acc[, c1] <- wk.re(str, switch(c1, `1` = paste0("[^", sep, "]+"), `2` = paste0("(?<=", sep, ")[^", sep, "]+")))
        }
    }
    else {
        a <- wk.re(str, "[^:]+", which = col, all = 1)
        a <- {
            if (length(col) > 1) 
                t(sapply(a, c))
            else as.matrix(unlist(a))
        }
        for (i in seq_along(col)) acc[, paste0(col[i])] <- a[, i]
        if (0) {
            sep.pos <- str_locate_all(str, sep)
            start <- matrix(nrow = length(str), ncol = length(col))
            end <- matrix(nrow = length(str), ncol = length(col))
            for (i in seq_along(sep.pos)) {
                start[i, ] <- sep.pos[[i]][col - 1, 2] + 1
                end[i, ] <- sep.pos[[i]][col, 1] - 1
            }
            for (i in seq_along(col)) acc[, paste0(col[i])] <- substr(str, start[, i], end[, i])
        }
    }
    acc[, paste0(col.0), drop = 0]
}
wk.str_split_fixed <- ieiwk.str_split_fixed <- function(string, pattern = "\t", n = str_count(string[1], pattern) + 1) data.table(str_split_fixed(string, pattern, n))
ieiwk.str_subset <- function(s, p, inv = 1) {
    i <- str_detect(s, p)
    if (inv) 
        s[!i]
    else s[i]
}
hs.trim <- trim <- function(x, trim = "\\s", l = TRUE, r = TRUE, m = FALSE) {
    if (l) 
        x <- sub(sprintf("^%s+", trim), "", x)
    if (r) 
        x <- sub(sprintf("%s+$", trim), "", x)
    if (m) 
        x <- gsub(sprintf("(?<=[^%s])%s+(?=[^%s])", trim, trim, trim), "", x, perl = TRUE)
    x
}
ieiwk.capitalize.1st.letter <- function(s) {
    sapply(s, function(x) {
        substr(x, 1, 1) <- toupper(substr(x, 1, 1))
        x
    })
}
ieiwk.strsplit.pick <- ieiwk.substr.i <- function(str, i, sep = "\\.") {
    strsplit(str, sep)[[1]][i]
}
"%,%" <- function(x, y) paste(x, collapse = y)
"%.%" <- function(x, y) sprintf("%s%s", x, y)
revStr <- function(x, sep = "") {
    sapply(strsplit(as.character(x), sep), function(a) {
        paste(rev(a), collapse = sep)
    })
}
revStr.int <- function(x) intToUtf8(rev(utf8ToInt(x)))
ieiwk.dna.nulceotide.complementary <- function(x) switch(x, A = "T", T = "A", C = "G", G = "C")
ieiwk.dna.complementaryRev.viaInt <- local({
    a <- c()
    a[utf8ToInt("ATCG")] <- utf8ToInt("TAGC")
    a[utf8ToInt("atcg")] <- utf8ToInt("tagc")
    a[utf8ToInt("N")] <- utf8ToInt("N")
    function(x) {
        b <- a[utf8ToInt(x)]
        i.na <- which(is.na(b))
        rev(seq_along(b))
        intToUtf8(rev(a[utf8ToInt(x)]))
    }
})
ieiwk.str.insert <- function(string, pos, insert) {
    paste(read.fwf(textConnection(string), c(pos, nchar(string)), as.is = TRUE), collapse = insert)
}
ieiwk.substr <- function(x, starts = 1, lengths = 1, stops = NULL) {
    if (length(lengths)) {
        stops <- starts + lengths - 1
    }
    else {
        if (!length(stops)) {
            return(substring(x, starts))
        }
    }
    substr(x, starts, stops)
}
wk.re <- function(text, pattern, act = "extract", replacement = "", which = {
}, do.group = if (missing(pattern)) {
    0
} else {
    grepl("[^\\\\]\\([^\\?]", pattern, perl = TRUE)
}, sub.len = 30, ic = if (missing(pattern)) 1 else {
    !grepl("[A-Z]", pattern, perl = TRUE)
}, ignore.case = ic, perl = 1, fixed = 0, whole = 0, useBytes = 0, invert = 0, all = 0, info, shift.start = 0, shift.length = 0) {
    invert <- as.logical(invert)
    if (missing(info)) {
        if (whole) {
            if (str_sub(pattern, 1, 1) != "^") 
                pattern <- .j("^", pattern)
            if (str_sub(pattern, -1) != "$") 
                pattern <- .j(pattern, "$")
        }
        if (act %in% c("detect", "which", "subset")) {
            hs1 <- regexpr
            i1 <- seq_along(text)
            i2 <- seq_along(pattern)
            i <- cbind(i1, i2)
            r <- rep(TRUE, nrow(i))
            for (i.p in unique(i[, 2])) {
                i1.1 <- which(i[, 2] %in% i.p)
                m <- hs1(pattern = pattern[i.p], text = text[i[i1.1, 1]], ignore.case = as.logical(ignore.case), perl = as.logical(perl), fixed = as.logical(fixed), useBytes = as.logical(useBytes))
                r[i1.1] <- m %in% -1
            }
            if (!invert) 
                r <- !r
            return(switch(act, detect = r, which = which(r), subset = text[r]))
        }
        if (act == "count") 
            all <- 1
        hs1 <- ifelse(all, gregexpr, regexpr)
        m <- hs1(pattern = pattern, text = text, ignore.case = as.logical(ignore.case), perl = as.logical(perl), fixed = fixed, useBytes = useBytes)
        if (0) {
            regexpr(".*腾讯.*", "腾讯123", perl = TRUE, ignore.case = TRUE)
            grepl("[A-Z]", pattern, perl = FALSE)
            grepl("[A-Z]", ".*腾讯A.*", perl = TRUE)
            grepl("[A-Z]", "aBc")
            hs1(pattern = pattern, text = text, perl = TRUE)
            hs1(pattern = pattern, text = text, perl = TRUE, ignore.case = as.logical(ignore.case))
            hs1(pattern = pattern, text = text, ignore.case = as.logical(ignore.case), perl = as.logical(perl), fixed = fixed, useBytes = useBytes)
        }
    }
    else m <- info
    switch(act, extract = {
        if (do.group) {
            if (all) {
                m1 <- {
                  if (length(which)) {
                    lapply(m, function(x) {
                      m1 <- attr(x, "capture.start")[, which, drop = 0]
                      attr(m1, "match.length") <- attr(x, "capture.length")[, which, drop = 0]
                      m1
                    })
                  } else {
                    lapply(m, function(x) {
                      m1 <- attr(x, "capture.start")
                      attr(m1, "match.length") <- attr(x, "capture.length")
                      m1
                    })
                  }
                }
            } else {
                if (length(which)) {
                  m1 <- attr(m, "capture.start")[, which, drop = 0]
                  attr(m1, "match.length") <- attr(m, "capture.length")[, which, drop = 0]
                } else {
                  if (0) m1 <- lapply(seq_along(m), function(i) {
                    m1 <- attr(m, "capture.start")[i, ]
                    attr(m1, "match.length") <- attr(m, "capture.length")[i, ]
                    m1
                  })
                  return(lapply(seq.int(ncol(attr(m, "capture.start"))), function(i) {
                    m1 <- attr(m, "capture.start")[, i]
                    attr(m1, "match.length") <- attr(m, "capture.length")[, i]
                    regmatches(text, m1, invert)
                  }))
                }
            }
            if (0) {
                r <- vector("character", length(m))
                r[m %not.in% -1] <- regmatches(text, m, invert)
                r
            }
            regmatches(text, m1, invert)
        } else {
            if (all && length(which)) m <- lapply(m, function(x) {
                m1 <- x[which]
                attr(m1, "match.length") <- attr(x, "match.length")[which]
                m1
            })
            r <- vector("character", length(m))
            r[m %not.in% -1] <- regmatches(text, m, invert)
            r
        }
    }, count = {
        sapply(m, . %=>% sum(x > 0))
    }, length = {
        if (all) {
            sapply(m, attr, "match.length")
        } else attr(m, "match.length")
    }, end = {
        if (all) {
            sapply(m, function(x) {
                as.vector(x + attr(x, "match.length") - 1)
            })
        } else as.vector(m + attr(m, "match.length") - 1)
    }, replace = {
        regmatches(text, m, invert) <- replacement
        text
    }, locate = m, sub = {
        str_sub(text, m, m + sub.len - 1)
    }, split = {
        regmatches(text, m, invert = 1)
    })
}
wk.re.replace <- function(text, pattern, replacement = "", all = 1, ...) {
    wk.re(text, pattern, act = "replace", replacement = replacement, all = all, ...)
}
wk.re.split <- function(text, pattern = "\t", all = 1, ...) {
    wk.re(text, pattern, act = "split", all = all, ...)
}
wk.re.ss <- function(text, pattern = "\t", ...) {
    wk.re(text, pattern, act = "subset", ...)
}
wk.striLocSub <- function(text, pattern = "\t", which = "self", ic = TRUE, last = 0, fixed = 0) {
    args <- list(str = text, mode = ifelse(last, "last", "first"), case_insensitive = ic)
    args[[ifelse(fixed, "fixed", "regex")]] <- pattern
    i <- do.call("stri_locate", args)
    switch(which, self = {
        j <- i[, "start"]
        k <- i[, "end"]
    }, before = {
        j <- 1
        k <- i[, "start"] - 1
    }, after = {
        j <- i[, "end"] + 1
        k <- -1L
    })
    stri_sub(text, j, k)
}
wk.re.ms <- function(n = 1, sep = "\t") {
    p1 <- p2 <- str_c("[^", sep, "]+")
    if (n > 1) 
        p2 <- str_c("(", p1, sep, ")")
    str_c(c(p2, if (n > 1) c("{", n - 1, "}", p1)), collapse = "")
}
wk.gre <- function(pattern, text, ignore.case = TRUE, perl = TRUE, fixed = FALSE, useBytes = FALSE, invert = FALSE) {
    wk.re(pattern = pattern, text = text, ignore.case = as.logical(ignore.case), perl = as.logical(perl), fixed = fixed, useBytes = useBytes, all = TRUE)
}
ieiwk.grep <- function(patterns, x, ignore.case = TRUE, extended = TRUE, perl = FALSE, value = TRUE, fixed = FALSE, useBytes = FALSE, invert = FALSE) {
    sapply(patterns, function(pattern) {
        grep(pattern, x, ignore.case, perl, value, fixed, useBytes, invert)
    })
}
ieiwk.grep.all <- function(patterns, x, ignore.case = FALSE, extended = TRUE, perl = FALSE, value = TRUE, fixed = FALSE, useBytes = FALSE, invert = FALSE) {
    i <- seq_along(x)
    args <- ieiwk.make.arg.list(perl, fixed, useBytes, invert)
    args["value"] <- FALSE
    if (!fixed) 
        args["ignore.case"] <- ignore.case
    for (p1 in patterns) {
        if (!length(i)) 
            break
        args[["pattern"]] <- p1
        args[["x"]] <- x[i]
        i <- i[do.call("grep", args)]
    }
    if (value) 
        x[i]
    else i
}
ieiwk.grep.any <- function(patterns, x, ignore.case = TRUE, extended = TRUE, perl = FALSE, value = TRUE, fixed = FALSE, useBytes = FALSE, invert = FALSE) {
    i <- rep(FALSE, length(x))
    j <- seq_along(x)
    args <- ieiwk.make.arg.list(perl, fixed, useBytes, invert)
    args["value"] <- FALSE
    if (!fixed) 
        args["ignore.case"] <- ignore.case
    p1 <- patterns[1]
    for (p1 in patterns) {
        args[["pattern"]] <- p1
        args[["x"]] <- x[j]
        k <- j[do.call("grep", args)]
        i[k] <- TRUE
        j <- j %not% k
    }
    i <- which(i)
    if (value) 
        x[i]
    else i
}
ieiwk.match <- function(x, y, all = 0, ...) {
    if (is.list(y)) {
        g <- rep(seq_along(y), sapply(y, length))
        y2 <- unlist(y)
    }
    else {
        g <- seq_along(y)
        y2 <- y
    }
    if (all) {
        lapply(x, function(x1) {
            g[which(y2 == x1)]
        })
    }
    else g[match(x, y2, ...)]
}
.conda <- function(act = "list") {
    .wk("conda", act)
}
wk.sprintf <- ieiwk.sprintf <- function(matrix, sep = "\t") {
    if (0) {
        matrix %<=>% as.data.table
    }
    hs.update(matrix, wk.dt.as, !inherits(x, "data.table"))
    matrix[, do.call(str_c, c(unname(.SD), sep = sep))]
}
ieiwk.as.date <- function(x, sep = c("/", "-", ".")) {
    f1 <- 1
    x <- trim(x)
    for (s1 in sep) {
        if (grepl(s1, x, fixed = 1)) {
            f1 <- 0
            break
        }
    }
    if (f1) {
        paste(apply(matrix(c(1, 4, 5, 6, 7, 8), 2), 2, function(i) {
            substr(x, i[1], i[2])
        }), collapse = "-")
    }
    else x
}
ieiwk.thisYear <- function(n = 0) {
    a <- format(Sys.time(), "%Y")
    if (n) 
        as.integer(a)
    else a
}
ieiwk.time.add <- function(t1 = 0, week = 0, day = 0, hour = 0, minute = 0, second = 0) {
    t1 + (((week * 7 + day) * 24 + hour) * 60 + minute) * 60 + second
}
.sys.mtime <- function(f) {
    strptime(paste(strsplit(.sys.do("ls -l -G -n --time-style=full-iso", .sys.fpGood(f), intern = 1), " ")[[1]][c(5, 6)], collapse = " "), "%Y-%m-%d %H:%M:%S")
}
data.check <- local({
    helper <- function(d1, d2) {
        r <- list()
        i1 <- which(sapply(d1, is.na))
        if (length(i1)) 
            r[["na"]] <- i1
        i1 <- which(sapply(d2, function(x) {
            is.na(x) || class(x) == "try-error"
        })) %not% r[["na"]]
        if (length(i1)) 
            r[["illegal"]] <- data.frame(row = i1, value = d1[i1])
        if (length(r)) 
            r
        else "nothing wrong"
    }
    function(d1, type = "age") {
        switch(type, age = helper(d1, as.numeric(d1)), date = helper(d1, lapply(d1, function(x) .try(as.Date(x)))), string = helper(d1, d1))
    }
})
wk.table.margin <- function(x, n1 = "Total") {
    r <- cbind(rbind(x, colSums(x)), c(rowSums(x), sum(x)))
    colnames(r)[ncol(r)] <- n1
    rownames(r)[nrow(r)] <- n1
    r
}
ieiwk.select.matrix <- function(m1, ic, ir, f1, rt = "v", inv = 0, sm) {
    q1 <- {
        g1 <- as.list(f1)
        if (missing(sm)) {
            cbind(seq_along(ic), seq_along(g1))
        }
        else {
            if (length(dim(sm))) {
                sm
            }
            else {
                matrix(sm, byrow = 1, ncol = 2)
            }
        }
    }
    i0 <- i <- seq.int(nrow(m1))
    for (i1 in seq.int(nrow(q1))) {
        if (!length(i)) 
            break
        ic1 <- ic[q1[i1, 1]]
        if (identical(g1[[q1[i1, 2]]], "-")) {
            if (is.character(ic1)) 
                ic1 <- match(ic1, colnames(m1))
            m1 <- m1[, -ic1, drop = 0]
        }
        else {
            x <- m1[i, ic1]
            i <- i[which({
                f2 <- g1[[q1[i1, 2]]]
                if (is.character(f2)) {
                  eval(parse(text = str_c("x", f2, sep = " ")))
                }
                else f2(x)
            })]
        }
    }
    if (inv) 
        i <- i0[-i]
    switch(rt, i = i, v = m1[i, , drop = 0])
}
wk.seq.nrow <- function(m) {
    seq.int(nrow(m))
}
wk.seq.ncol <- function(m) {
    seq.int(ncol(m))
}
ieiwk.as.matirx <- function(x, f1 = identity) {
    x <- as.matrix(x)
    r <- f1(x)
    dim(r) <- dim(x)
    r
}
ieiwk.t <- function(M) {
    sapply(if (is.null(rownames(M))) 
        seq(nrow(M))
    else rownames(M), function(c1) {
        sapply(if (is.null(colnames(M))) 
            seq(ncol(M))
        else colnames(M), function(c2) {
            M[c1, c2]
        })
    })
}
ieiwk.l2m <- function(l) {
    matrix(unlist(l), nrow = length(l[[1]]))
}
ieiwk.data.frame <- function(..., row.names = NULL, check.rows = FALSE, check.names = TRUE, stringsAsFactors = FALSE) {
    data.frame(..., row.names = row.names, check.rows = check.rows, check.names = check.names, stringsAsFactors = stringsAsFactors)
}
"%[%" <- function(...) {
    a <- as.list(substitute(...[]))
    a[[3]][[2]] <- a[[2]]
    eval(as(c(as.list(a[[3]]), drop = FALSE), "call"))
}
ieiwk.mpstat <- function(where = "", interval = 1, count = 1) {
    mpstat <- {
        a <- .sys.do(if (!missing(where)) 
            c("ssh", where), "mpstat -P ALL", interval, count, intern = 1)
        tail(a, -tail(which(a == ""), 1))
    }
    mpstat.tbl <- {
        a <- t(sapply(mpstat, function(x) {
            strsplit(x, "\\s+")[[1]]
        }))[, -1]
        rownames(a) <- c()
        b <- a[-1, -1]
        rownames(b) <- a[-1, 1]
        colnames(b) <- a[1, -1]
        b
    }
    mpstat.tbl
}
wk.xcs <- floor(parallel::detectCores()/2)
wk.mclapply <- ieiwk.mclapply <- local({
    require(parallel)
    mc.cores.max <- detectCores() - 1
    function(X, FUN, ..., mc.preschedule = TRUE, mc.set.seed = TRUE, mc.silent = FALSE, mc.cleanup = TRUE, mc.cores = {
    }, mc.cores.factor = 0.9, mc.cores.win = 1, required.packages = {
    }) {
        if (!length(X)) 
            return()
        if (!length(mc.cores)) 
            mc.cores <- wk.xcs
        if (is.na(mc.cores.max)) 
            mc.cores.max <- 4
        if (osType() == "Windows") 
            mc.cores <- mc.cores.win
        mc.cores <- min(mc.cores, mc.cores.max, length(X), na.rm = TRUE)
        windows <- Sys.info()["sysname"] == "Windows"
        a <- {
            if (windows) {
                cl <- makeCluster(mc.cores)
                f1 <- function(x, fun, required.packages = {
                }, ...) {
                  if (length(required.packages)) {
                    for (p1 in required.packages) {
                      eval(call("require", p1))
                    }
                  }
                  fun(x, ...)
                }
                a <- try(parLapply(cl, X, f1, FUN, required.packages = required.packages, ...), silent = TRUE)
                stopCluster(cl)
                a
            }
            else {
                a <- mclapply(X, FUN, ..., mc.preschedule = mc.preschedule, mc.set.seed = mc.set.seed, mc.silent = mc.silent, mc.cores = mc.cores, mc.cleanup = mc.cleanup)
            }
        }
        if (is.null(names(X))) {
            if ((class(X) %in% c("character", "numeric", "integer")) && (!anyDuplicated(X))) 
                names(a) <- X
        }
        else {
            names(a) <- names(X)
        }
        a
    }
})
ieiwk.pvec <- function(v, FUN, ..., mc.set.seed = TRUE, mc.silent = FALSE, mc.cleanup = TRUE) {
    mc.cores <- Ncpu()[1]
    if (mc.cores > length(v)) 
        mc.cores <- length(v)
    parallel::pvec(v, FUN, ..., mc.set.seed = mc.set.seed, mc.silent = mc.silent, mc.cores = mc.cores, mc.cleanup = mc.cleanup)
}
hs.methods <- function(obj) .sh(methods(class = class(obj)))
ieiwk.last.value <- function() .Last.value
ieiwk.get.singleElement <- function(o1, i = 1) {
    f1 <- if (class(o1) %in% "list") 
        "[["
    else "["
    do.call(f1, list(o1, i))
}
wk.attr.hs <- function(o, hs = head) {
    r <- hs(o)
    mz <- names(attributes(o))
    for (mz1 in mz) {
        attr(r, mz1) <- {
            if (length(attr(o, mz1)) == length(o)) {
                hs(attr(o, mz1))
            }
            else attr(o, mz1)
        }
    }
    r
}
.last.cmd <- function(n = 1, file = hs.home.expand("~/.Rhistory")) {
    .sys.do("tail -n", n, file, intern = 1)
    a <- capture.output(history())
    a <- savehistory(con)
}
ieiwk.clean.obj <- function(x) {
    lj <- c("na.action")
    for (a in lj) {
        attr(x, a) <- {
        }
    }
    x
}
os <- function(..., env = parent.frame(), units = "auto") {
    args <- list(...)
    if (!length(args)) 
        args <- sapply(ls(env, all.names = TRUE), as.symbol)
    a <- lapply(args, function(o) {
        eval(bquote({
            object.size(.(as.symbol(o)))
        }), env)
    })
    size <- sum %(% a
    class(size) <- class(a[[1]])
    format(size, units = units)
}
ll <- function(pos = 1, env = parent.frame()) {
    if (is.null(env)) 
        env <- as.environment(pos)
    a <- sapply(ls(env, all.names = TRUE), function(o) eval(call("object.size", as.symbol(o)), env = env))
    sapply(names(a)[order(a)], function(x) {
        cat(x, "\t")
        print(eval(call("object.size", as.symbol(x)), env = env), units = "auto")
    })
    a <- sum(a)
    class(a) <- "object_size"
    print(a, units = "auto")
}
rm.all <- function() {
    rm(list = ls(pos = 1), pos = 1)
}
rm.data <- function(but = "") {
    sapply(ls(pos = 1) %not% but, function(o) {
        eval(substitute(if (class(get(o)) != "function") rm(o), list(o = o)), env = .GlobalEnv)
    })
    cat(c())
}
rm.var <- function(but = c("@", "handy", "Somel2008Plos1", "Khaitovich2005Science", "Blekhman2009GenomeResearch", "Blekhman2009GenomeResearch_Khaitovich2005Science", ls(pat = "^\\d{4}\\.\\d{2}\\.\\d{2}\\.\\d{2}\\.\\d{2}\\.\\d{2}[(\\.\\w)|(_\\d+\\.\\d+)]", env = parent.frame()))) {
    browser()
    sapply(ls(envir = env) %not% but, function(o) {
        eval(substitute(if (class(get(o)) != "function") rm(o), list(o = o)), env = env)
    })
    cat(c())
}
rm.var.4def <- function(but = c("."), env = parent.frame()) {
    sapply(ls(envir = env), function(o) {
        if (!eval(call("typeof", as.symbol(o)), env) %in% c("closure")) 
            eval(call("rm", as.symbol(o)), env)
    })
}
rm.function <- function(but = "") {
    sapply(ls(pos = 1) %not% but, function(o) {
        eval(substitute(if (class(get(o)) == "function") rm(o), list(o = o)), env = .GlobalEnv)
    })
    cat(c())
}
ls.var <- function() {
    obj.v <- ls(pos = 1)
    obj.nonFn <- {
        a <- sapply(obj.v, function(o) {
            eval(substitute(class(o), list(o = as.symbol(o))), envir = .GlobalEnv)
        })
        names(a[a != "function"])
    }
    obj.nonFn
}
listElementCounts <- function(l, recursive = 1) if (isList(l)) sum(sapply(l, listElementCounts, recursive)) else 1
listPrintLinesCounts <- function(l) {
    sum(rapply(l, function(x) {
        a <- dim(x)
        if (is.null(a)) {
            1
        } else {
            if (length(a) == 2) {
                nrow(x)
            }
        }
    }))
}
.at <- function(o, path = NULL, fn = NULL, as = "", ..., leaf = 0, v = !leaf, r = 0) {
    if (leaf) 
        return(ieiwk.attr(o, path, "identity", leaf = 0))
    if (is.null(fn)) 
        fn <- switch(as.character(length(path)), `0` = "names", ".str")
    if (is.numeric(path)) 
        path[1] <- names(attributes(o))[path[1]]
    cmd <- ifelse(length(path), sprintf("attr( o, %s)", shQuote(path[1])), "attributes( o)")
    cmd <- paste(cmd, sprintf("$%s", path[-1]), sep = "")
    cmd <- sprintf("do.call( \"%s\", list( %s quote( %s), ...))", fn[1], as, cmd)
    R <- eval(parse(text = cmd))
    for (fn.1 in fn[-1]) {
        R <- do.call(fn.1, list(R))
    }
    if (r) 
        return(o)
    if (v) {
        if (tail(fn, 1) != ".str") 
            return(R)
    }
}
.t <- function(p, actCode = 1, v = NULL, v.fun.env, to = "", sub = 1, env, self = 1, sep = "////") {
    if (missing(env)) 
        env <- parent.frame()
    p <- unlist(sapply(p, function(p) if (grepl(sep, p)) 
        strsplit(p, sep)[[1]]
    else p))
    fa <- a1 <- as.symbol(p[1])
    if (self) 
        p <- c(p, ".self.30.04.27")
    switch(as.character(actCode), `1` = {
        fa <- parse(text = paste(c(rep("attr(", length(p) - 1), p[1], sprintf(",\"%s\")", p[-1])), collapse = ""))[[1]]
        o <- eval(fa, env = env)
        switch(to, eval = eval(body(o), env = env), show = {
            src <- attr(o, "srcref")
            src <- capture.output(src)
            .ss("{")
            .ss(head(src[-1], -1))
            .ss("}")
        }, {
            if (!sub) {
                for (a in attributes(o)) {
                  attr(o, a) <- NULL
                }
            }
            o
        })
    }, `4` = {
        for (p1 in tail(p, -1)) {
            if (eval(call("is.null", fa), env = env)) eval(call("<-", fa, list()), env = env)
            fa <- call("attr", fa, p1)
        }
        eval(call("<-", fa, substitute(quote(v), list(v = v))), env = env)
        invisible()
    }, `3` = {
        p <- p[!p %in% ".self.30.04.27"]
        r <- NULL
        for (p1 in tail(p, -1)) {
            fa <- call("attr", fa, p1)
        }
        leaf <- eval(call("names", call("attributes", fa)), env) %not% ".self.30.04.27"
        if (length(leaf)) {
            r <- lapply(leaf, function(leaf) eval(call(".t", c(p, leaf), 3, env = env), env))
        }
        r <- r[!sapply(r, is.null)]
        if (eval(call("is.null", call("attr", fa, ".self.30.04.27")), env)) {
            as.list(r)
        } else {
            a <- list(p, r)
            as.list(unlist(rapply(a, paste, collapse = sep, "character", how = "list")))
        }
    }, `2` = {
    })
}
.tl <- function(p, actCode = 1, v = NULL, v.fun.env, to = "", sub = 1, env, self = TRUE, sep = "////", cat.p = 1, tree = list(), tree.form = 1) {
    if (missing(env)) 
        env <- parent.frame()
    p <- unlist(strsplit(p, sep))
    if (self) 
        p <- c(p, ".self.30.04.27")
    switch(paste(actCode), `1` = {
        fa <- ieiwk.access.list(p)
        o <- eval(fa, env = env)
        switch(to, eval = eval(body(o), env = env), show = {
            src <- attr(o, "srcref")
            src <- capture.output(src)
            .ss("{")
            .ss(head(src[-1], -1))
            .ss("}")
        }, {
            o
        })
    }, `4` = {
        fa <- ieiwk.access.list(p, "element")
        eval(bquote({
            .(fa) <- list()
            .(fa) <- .(v)
        }), env)
        fa <- ieiwk.access.list(head(p, -1))
        eval(bquote({
            if (length(.(fa)) == 0) .(fa) <- {
            }
        }), env)
        {
        }
        invisible()
    }, `3` = {
        p <- p[!p %in% ".self.30.04.27"]
        fa <- ieiwk.access.list(p)
        leaf <- eval(call("names", fa), env)
        r <- sort(unlist(c(if (".self.30.04.27" %in% leaf) paste(p, collapse = sep), sapply(leaf %not% ".self.30.04.27", function(leaf) eval(call(".tl", c(p, leaf), 3, cat.p = FALSE, env = env), env))), use.names = FALSE))
        if (cat.p) .ss(r)
        invisible(r)
    }, `2` = {
        switch(to, rm = {
            browser("31.04.24.101902")
            fa <- ieiwk.access.list(p)
        })
    })
}
.te <- function(p, actCode = 1, v = NULL, v.fun.env, to = "", sub = 1, env, self = TRUE, sep = "////", cat.p = FALSE, tree = list(), tree.form = 1) {
    if (missing(env)) 
        env <- parent.frame()
    p <- unlist(strsplit(p, sep))
    if (self) 
        p <- c(p, ".self.30.04.27")
    switch(paste(actCode), `1` = {
        fa <- ieiwk.access.list(p)
        o <- eval(fa, env = env)
        switch(to, eval = eval(body(o), env = env), show = {
            src <- attr(o, "srcref")
            src <- capture.output(src)
            .ss("{")
            .ss(head(src[-1], -1))
            .ss("}")
        }, {
            o
        })
    }, `4` = {
        fa <- ieiwk.access.list(p)
        eval(call("<-", fa, list()), env = env)
        eval(call("<-", fa, substitute(quote(v), list(v = v))), env = env)
        invisible()
    }, `3` = {
        p <- p[!p %in% ".self.30.04.27"]
        fa <- ieiwk.access.list(p)
        leaf <- eval(call("names", fa), env)
        r <- sort(unlist(c(if (".self.30.04.27" %in% leaf) paste(p, collapse = sep), sapply(leaf %not% ".self.30.04.27", function(leaf) eval(call(".tl", c(p, leaf), 3, cat.p = FALSE, env = env), env))), use.names = FALSE))
        if (cat.p) .ss(r)
        invisible(r)
    }, `2` = {
    })
}
lb <- function(e1 = "", n = 0, env = new.env(parent = parent.frame())) {
    while (n && {
        length(e1) > 1
    }) {
        eval(e1[[2]], env = env)
        e1[[2]] <- {
        }
        n <- n - 1
    }
    eval(bquote({
        browser()
        .(e1)
    }), env)
}
lb.env <- local({
    fn.30.04.22.143406 <- function(where) {
        env.root <- .GlobalEnv
        .env <- try(get(where, env.root), silent = 1)
        if (identical(class(.env), "try-error")) 
            assign(.env, new.env(), env = env.root)
        eval(quote(browser()), env = .env)
    }
    function(what = "lb", where = "lb.env.env", ...) {
        switch(what, lb = fn.30.04.22.143406(where), keep = {
            o = list(...)
            names(o) <- match.call(expand.dots = 0)$...
            sapply(names(o), function(x) assign(x, o[[x]], env = .GlobalEnv))
        }, ls = ls(environment(lb.env)[["aEnv"]]))
    }
})
lb.env.update <- function(act = 1) {
    switch(act, assign("aEnv.30.04.21", environment(lb.env)[["aEnv"]], env = .GlobalEnv), assign("aEnv", aEnv.30.04.21, env = environment(lb.env)))
}
bin.Rscript <- "/public/software/R-3.0.2/bin/Rscript"
.sys.rs <- function(..., bin = bin.Rscript, myFun = 0, trim = 0, mc.cores = 1, wait = 0, qsub = 0, jn = "testing", qsub.o.d = "$HOME/test/qsub", nodes = "1", ncpu = 1, dz = {
}, chk = 0, report = 1) {
    code <- {
        a <- ieiwk.mp.code2str.r(..., trim = trim)
        a <- c(sprintf("mc.cores <- %s", mc.cores), a)
        if (myFun) {
            b <- ieiwk.mp.code2str.r({
                source("~/calvin/code/R/function.common.R")
            }, trim = trim)
            a <- c(b, a, ".r()", "cat( Sys.time(), \"\n\", sep = \"\")")
        }
        shQuote(a)
    }
    code <- gsub("\\\\", "\\\\\\\\\\\\\\\\", code)
    if (qsub) 
        .sys.qsub(bin, "-e " %.% code, qsub.o.d = qsub.o.d, jn = jn, dz = dz, nodes = nodes, ncpu = ncpu, chk = chk, report = report)
    else {
        .sys.do(bin, "-e " %.% code, wait = wait, report = report)
    }
}
.sys.qsub.rsStdin <- function(..., bin = bin.Rscript, myFun = 1, trim = 0, jn = "testing", qsub.o.d = "$HOME/test/qsub", nodes = "1", ncpu = 1, dz = {
}, chk = 0, interactive = 0, report = 1) {
    code <- {
        a <- ieiwk.mp.code2str.r(..., trim = trim)
        if (myFun) {
            b <- ieiwk.mp.code2str.r(quote({
                source("~/calvin/code/R/function.common.R")
                ieiwk.cat.line(sprintf("\n========= Start at %s", Sys.time()))
            }), trim = trim)
            a <- c(b, a)
        }
        c(a, "cat( sprintf( \"========= end at %s\n\", Sys.time()))", ".r()")
    }
    cmd.all <- c(sprintf("%s - <<RscriptEOF", bin), code, "RscriptEOF") %,% "\n"
    cmd.all <- gsub("\\\\", "\\\\\\\\\\\\\\\\", cmd.all)
    .sys.qsub(cmd.all, qsub.o.d = qsub.o.d, jn = jn, dz = dz, nodes = nodes, ncpu = ncpu, chk = chk, interactive = interactive, report = report)
}
ieiwk.rsStdin <- function(..., bin = "/home/local/bin/R-3.0.2/bin/Rscript", myFun = 1, trim = 0, report = 1) {
    code <- {
        a <- ieiwk.mp.code2str.r(..., trim = trim)
        if (myFun) {
            b <- ieiwk.mp.code2str.r(quote({
                source("~/calvin/code/R/function.common.R")
            }), trim = trim)
            a <- c(b, a)
        }
        c(a, "cat( sprintf( \"end at %s\n\", Sys.time()))", ifelse(report, ".r()", {
        }))
    }
    cmd.all <- c(sprintf("%s - <<RscriptEOF", bin), code, "RscriptEOF") %,% "\n"
    ieiwk.cat.line(cmd.all)
    ieiwk.cat.line(sprintf("\nStart at %s", Sys.time()))
    .sys.do(cmd.all)
}
.sys.qsub.i <- function(nd = 1, ncpu = 1) {
    .sys.do(sprintf("qsub -I -l nodes=%s:ppn=%s -l walltime=%s:00:00", nd, ncpu, prod(24, 365)))
}
.sys.qsub.sleep <- function(nd = 1, ncpu = 1, howLong = prod(60, 60)) {
    .sys.qsub("sleep", howLong, nodes = sprintf("node%s", nd), ncpu = ncpu)
}
.sys.qsub <- function(..., jn = "testing", qsub.o.d = "$HOME/test/qsub", nodes = "1", ncpu = 1, dz = {
}, chk = 0, interactive = 0, report = 1) {
    cmd <- .sys(..., intern = 1)
    cmd.all <- c(sprintf("qsub %s- <<EOF", ifelse(interactive, "-I ", "")), sprintf("#PBS -N %s", jn), sprintf("#PBS -l nodes=%s:ppn=%s", nodes, ncpu), "#PBS -l walltime=1000:00:00", "#PBS -j oe", sprintf("#PBS -o %s", qsub.o.d), sprintf("#PBS -e %s", qsub.o.d), cmd, "EOF") %,% "\n"
    ieiwk.cat.line(cmd.all)
    if (chk) 
        return()
    a <- .sys.do(cmd.all, dz = dz, intern = 1)
    re("\\d+", a)
}
.sys.qstat <- function(dz = {
}) {
    .sys.do("qstat -n -1 -M -R", dz = dz)
}
.sys.qsub.chk <- function(job.id, d = "~/test/qsub", n = 20, dz = "sg") {
    f.size.limit <- 1024
    f.p <- grep(job.id, .sys.ls(d, dz = dz), v = 1)
    f.size <- .sys.stat.size(f.p, dz = dz)
    ieiwk.cat.line(sprintf("file size: %s", f.size))
    .sys.do(sprintf("head -%s", n), f.p, dz = dz, intern = 1)
}
.sys.qsub.chk.r <- function(job.id, dz = "sg") {
    .sys.do("myperl -w ~/bin/report.qsub.jobs.done.pl", job.id, dz = dz, wait = 0)
}
.rls <- function(f = "~/calvin/code/2010/07/23/2010.07.23.23.08.30.Fri.R") source(f)
jdi <- function(which = "hh", env = parent.frame()) {
    wj <- switch(which, nx3200 = "/home/share/nx3200/windows/wengkai/calvin/code/R/justDoIt.R", local = "F:/data.daily/code/justDoIt.R", local.cygwin = "", hh = "/gpfsDATA/home/calvin/wk/my/wd/dm/R/justDoIt.R")
    eval(call("source", wj, local = TRUE), env)
    {
    }
}
.sys.exit <- function() {
    .sys.do("~/apps/scripts/exit.sh")
}
ieiwk.favor.node <- "fat01"
ieiwk.wait.file <- function(f, m, ..., ti = 5, tl = 15) {
    repeat {
        if (missing(m)) {
            t1 <- max(file.info(f)[["mtime"]])
            if (any(is.na(t1))) {
                td <- tl - 1
            }
            else {
                t2 <- Sys.time()
                td <- as.integer(difftime(t2, t1, units = "secs"))
            }
            if (td > tl) 
                break
        }
        else {
            if (length(.sys.do("grep", "-e", shQuote(m), ..., .sys.fpGood(f), intern = 1))) 
                break
        }
        .sys.do("sleep", ti)
    }
}
ieiwk.wait.thread <- function(thread.str, until = c("quit", "cpu0"), co.pcpu = 5, node = Sys.info()["nodename"], wait = 7) {
    u1 <- match.arg(until)
    e1 <- switch(u1, cpu0 = {
        expression({
            pid <- re("\\d+", wdjc.search(thread.str, node, jg = "ps"))
            !jc.is.sleeping(pid)
        })
    }, quit = {
        expression({
            if (wdjc.search(thread.str, node)) {
                .sleep(wait)
                1
            } else 0
        })
    })
    while (eval(e1)) {
    }
}
cluster.all <- function() .sys.do("bhosts", "|gawk", shQuote("$4>3 { print $1}"), "| tail -n +2", intern = 1)
.sys.killall <- function(c1 = .hostname(), dd = "gzfy") {
    if (any(c1 %in% cluster.all())) {
        c1 <- c1 %and% cluster.all()
    }
    else {
        c1 <- switch(c1, here = .sys.do("hostname", intern = 1), all = .sys.do("bhosts", "|gawk", shQuote("$4>3 { print $1}"), "| tail -n +2", intern = 1))
    }
    switch(dd, gzfy = {
        u <- "calvin"
        lapply(c1, function(c1) {
            .sys.do("ssh", c1, "killall -u", u)
        })
    })
}
wdjc <- function(c1 = cluster.all(), o = paste(c("pid", "cmd"), collapse = ",")) {
    a <- ieiwk.mclapply(c1, function(c1.1) {
        u <- Sys.info()["user"]
        hn <- Sys.info()["nodename"]
        a <- trim(.sys.do(if (c1.1 != .hostname()) 
            c("ssh", c1.1, "\""), "ps", "-U", u, "-u", u, "-o", o, if (c1.1 != .hostname()) 
            "\"", intern = 1)[-1])
    }, mc.cores = length(c1))
    names(a) <- c1
    a
}
jc.is.sleeping <- function(pid, co = 5, time.interval = 3) {
    as.numeric(.sys.do("pidstat", "-p", pid, "-u", time.interval, 1, "| tail -n +4", "| gawk", .q("{ print( $6)}"), intern = 1)[1]) < co
}
wdjc.search <- function(p1, nodes = "log01", times = 1, wait = 3, fix = 1, jg = c("logical", "length", "ps")) {
    ps <- wdjc(nodes)
    p1.ps <- sapply(ps, function(ps1) {
        ieiwk.grep.all(p1, ps1, fix = fix, value = 1)
    })
    jg <- match.arg(jg)
    if (jg == "ps") 
        p1.ps
    else {
        p1.ps.length <- sapply(p1.ps, length)
        switch(jg, logical = any(p1.ps.length), length = p1.ps.length)
    }
}
disk.is.busy <- function(c1 = Sys.info()["nodename"], times = 4) {
    this <- Sys.info()["nodename"]
    wk.sapply(c1, function(c1.1) {
        n1 <- times
        any(sapply(seq(times), function(i) {
            a <- table(ieiwk.eval(quote(ps.state()), n1 = c1.1, email = 0))
            n1 <<- n1 - 1
            if (n1) .sleep(2)
            "D" %in% names(a)
        }))
    })
}
ps.state <- function() {
    a <- .sys("ps", "-eo pid,ppid", "|gawk", shQuote("BEGIN{OFS=\"\\t\"}{ print( $1,$2)}")) %=>% pipe %=>% read.table(header = TRUE)
    .sys.do("ps", "-p", paste(a[a[, "PPID"] != 2, "PID"], collapse = ","), "-o state", intern = 1)
}
disk.wait.not.busy <- function(c1 = "fat01", wait = 7) {
    while (disk.is.busy(c1)[[1]]) .sleep(wait)
    1
}
mem.is.enough <- function(c1, co = 50) {
    sapply(c1, function(c1) {
        m1 <- cluster.mem(c1)
        if (sum(m1[c("free", "cached")]) > co) 
            1
        else 0
    })
}
cluster.mem <- function(c1 = {
}) {
    m1 <- .wk(if (length(c1)) 
        c("ssh", c1), "cat /proc/meminfo", intern = 1)
    hs.re(m1, "\\d+")
    strsplit(m1, ":") %=>% {
        hs.c(sapply(x, . %=>% hs.sub(x[1], "^(?i)mem")), sapply(x, . %=>% hs.re(x[2], "\\d+") %=>% as.numeric)/(1024^2))
    }
}
give.me.a.node <- function(co.mem = 50, ps.pattern, ps.pattern.max = 1, wait = 3, wait.first = 0, verbose = 1, nodes2exclude = "", target.wj) {
    if (!missing(target.wj)) 
        if (all(file.exists(target.wj))) 
            return()
    if (wait.first) 
        .sleep(wait)
    repeat {
        cluster.all.mem <- {
            nodes2do <- setdiff(cluster.all(), nodes2exclude)
            a <- t(sapply(ieiwk.mclapply(nodes2do, cluster.mem), c))
            rownames(a) <- nodes2do
            a[rev(order(apply(a[, c("free", "cached")], 1, sum))), ]
        }
        houxuan <- clg(apply(cluster.all.mem[, c("free", "cached")], 1, sum) > co.mem, 1)
        houxuan <- setdiff(houxuan, nodes2exclude)
        for (h1 in houxuan) {
            if (unlist(disk.is.busy(h1))) {
                if (verbose) 
                  cat(paste("Disk is busy for", h1), sep = "\n")
                next
            }
            if (!missing(ps.pattern) && wdjc.search(ps.pattern, h1, jg = "length") >= ps.pattern.max) 
                next
            return(h1)
        }
        if (wait > 0) {
            .sleep(wait)
        }
        else break
    }
}
node.wait.available <- function(c1 = Sys.info()[["nodename"]], wait = sample(seq(3, 100), 1), mem.co = switch(c1, fat01 = 50, 10)) {
    repeat {
        if (disk.wait.not.busy(c1, wait) && mem.is.enough(c1, mem.co)) {
            return(1)
        }
    }
}
ieiwk.vt <- function(vcf.wj) {
}
.gemini <- function(cmd = c("query", "load", "db_info"), db, q, load, v, p, t = "VEP", cores = 4, skip.gene.tables = 1, header = 1) {
    switch(match.arg(cmd), query = .sys.do("gemini", "query", "-q", .q(q), if (header) "--header", db), load = .sys.do("gemini", "load", "-v", v, "-p", p, "-t", t, "--cores", cores, if (skip.gene.tables) "--skip-gene-tables", db), db_info = .sys.do("gemini", cmd, db))
}
.samtools <- function(f.bam, cmd = "view", col = 0, less = 1, pipe = 0) {
    .sys("samtools", cmd, f.bam, "| gawk", .q(sprintf("{ print %s}", paste(sprintf("$%s", col), collapse = ","))), if (less && !pipe) 
        "| less -n", pipe = pipe, chk = 0)
}
.blat <- function(base = "~/prj/pub/genomes/hg19.fa", qry, f.out = hs.wjm(qry, ext = "blatOut"), stepSize = 5, repMatch = 2253, minScore = 0, minIdentity = 0, out = "psl", c1 = ieiwk.favor.node, bg = 0) {
    env1 <- ieiwk.env.here()
    o <- sapply(c("stepSize", "repMatch", "minScore", "minIdentity", "out"), function(o1) {
        sprintf("-%s=%s", o1, get(o1, env = env1))
    })
    ieiwk.eval(bquote({
        .sys.do("blat", .(base), .(qry), .(f.out), .(o))
    }), c1, bg = bg)
    f.out
}
.bwa <- function(qry, idxbase = "~/prj/pub/genomes/hg19", k = 48, nt = 10, d.out = dirname(qry[1]), f.out = ieiwk.filename(qry[1], dn = 0, ext = "bam"), c1 = ieiwk.favor.node) {
    f.out <- ieiwk.file.path(d.out, f.out)
    ieiwk.eval(bquote({
        .sys.do("samtools view -b", "<(", "bwa mem", "-k", .(k), "-t", nt, .(idxbase), .(qry), ")", "-o", .(f.out))
    }), c1)
    f.out
}
.bwa.index <- function(fa, p = {
}, a = c("bwtsw", "is"), n1 = ieiwk.favor.node, fa.fifo.name = {
}) {
    ieiwk.eval(bquote({
        fa <- .(fa)
        a <- .(match.arg(a))
        p <- .(p)
        if (length(p)) 
            p <- ieiwk.file.path(p)
        if (length(fa) == 1) {
            if (!length(p)) 
                p <- ieiwk.file.path(ieiwk.filename(fa))
            .sys.do("bwa index", "-a", a, "-p", p, fa)
        }
        else if (length(fa) > 1) {
            fa.fifo.name <- .(fa.fifo.name)
            if (!length(fa.fifo.name)) 
                fa.fifo.name <- ieiwk.filename(p, add = "fa")
            if (file.exists(fa.fifo.name)) 
                file.remove(fa.fifo.name)
            .sys.do("mkfifo", .sys.fpGood(fa.fifo.name))
            .sys.do("cat", fa, ">", fa.fifo.name, "&")
            .sys.do("bwa index", "-a", a, "-p", p, fa.fifo.name, "&")
            browser("32.07.21.143544")
            .sleep(11)
            jc.ms <- c("bwa", "index", p)
            ieiwk.wait.thread(jc.ms, until = "cpu0")
            wdjc.search(c("bwa", "index", p), ieiwk.favor.node, jg = "ps")
            .sys.do("cat", fa, ">", fa.fifo.name)
            ieiwk.wait.thread(jc.ms, until = "quit")
        }
    }), n1)
}
.sys.bam.chk <- function(fp, nl = 10, dz = {
}) {
    bin <- "~/apps/samtools/samtools view"
    if (length(dz)) {
        .sys.ssh(bin, fp, "| head", sprintf("-%s", nl))
    }
    else {
        .sys.do(bin, fp, "| head", sprintf("-%s", nl))
    }
}
ieiwk.index.bam <- function(f.bam, f.picard = "~/rj/picard-tools-2.1.0/picard.jar") {
    .sys.do("java -jar", f.picard, "BuildBamIndex", "I=", f.bam, log = ieiwk.filename(f.bam, ext = "bai", add = "log"))
}
ieiwk.index.fa <- function(f.fa, f.picard = "~/rj/picard-tools-2.1.0/picard.jar") {
    f.out <- ieiwk.filename(f.fa, ext = "dict")
    .sys.do("java -jar", f.picard, "CreateSequenceDictionary", "R=", f.fa, "O=", f.out, log = ieiwk.filename(f.out, add = "log"))
}
bi.qse <- function(qs) {
    as.numeric(charToRaw(qs)) - 33
}
bin.samtools <- "~/apps/samtools/samtools"
bin.bwa <- "~/apps/bwa-0.7.5a/bwa"
bin.java <- "/public/software/jdk1.7.0_45/bin/java"
bin.MarkDuplicates <- "~/apps/picard-tools-1.102/MarkDuplicates.jar"
bin.intersectBed <- "/public/software/bedtools/bin/intersectBed"
bin.ValidateSamFile <- "~/apps/picard-tools-1.102/ValidateSamFile.jar"
fn.x.left.30.03.26.162558 <- function(v, m, tolerance, plain.width.min, plain.fluc.overall, adjacent) {
    i <- seq(m)
    j <- which(diff((diff(v[i]) + tolerance) >= 0) == 1)
    if (length(j)) {
        x1 <- i[tail(j, 1) + 2]
        if (abs(v[x1 + plain.width.min] - v[x1]) < plain.fluc.overall) {
            x1 <- x1 + plain.width.min
        }
        else {
            j2 <- seq(max(1, x1 - adjacent), x1 - 1)
            m2 <- j2[which.max(v[j2])]
            if (v[m2] - v[m2 - 1] > tolerance) 
                x1 <- fn.x.left.30.03.26.162558(v, m2, tolerance, plain.width.min, plain.fluc.overall, adjacent)
        }
    }
    else x1 <- 1
    x1
}
fn.x.right.30.03.26.162646 <- function(v, m, tolerance, plain.width.min, plain.fluc.overall, adjacent) {
    i <- seq(m, length(v))
    j <- which(diff((diff(v[i]) - tolerance) <= 0) == -1)
    if (length(j)) {
        x2 <- i[j[1] + 1]
        if (abs(v[x2] - v[x2 - plain.width.min]) < plain.fluc.overall) {
            x2 - plain.width.min
        }
        else {
            j2 <- seq(x2 + 1, min(length(v), x2 + adjacent))
            m2 <- j2[which.max(v[j2])]
            if (v[m2] - v[m2 + 1] > tolerance) 
                x2 <- fn.x.right.30.03.26.162646(v, m2, tolerance, plain.width.min, plain.fluc.overall, adjacent)
        }
    }
    else x2 <- length(v)
    x2
}
ieiwk.peak2valley <- function(v, to = min(v), adjacent = 15, x1x2 = NULL) {
    tolerance <- (function(v, extra = 4) {
        fluc <- diff(v)
        i <- sapply(seq(1, length(fluc) - extra), function(i) {
            i.v <- seq(i, i + extra)
            if (length(unique(sign(fluc[i.v]))) > 1) 
                i.v
        })
        r <- abs(fluc[unique(unlist(i))])
        quantile(r, 0.95)
    })(v)
    plain.width.min <- 50
    plain.fluc.overall <- tolerance * 10
    m <- which.max(v)
    if (m < 3) 
        x1 <- 1
    else {
        x1 <- fn.x.left.30.03.26.162558(v, m, tolerance, plain.width.min, plain.fluc.overall, adjacent)
    }
    x2 <- fn.x.right.30.03.26.162646(v, m, tolerance, plain.width.min, plain.fluc.overall, adjacent)
    v[x1:x2] <- to
    v
}
fastq_EndType <- function(f1) {
    has3OrMoreReads <- {
        n <- 3
        nl <- n * 4
        as.integer(.sys.do("head -n", formatSafe(nl), f1, "| wc -l", intern = 1)) >= nl
    }
    r1 <- strsplit(readLines(f1, 1), "\\s+")[[1]]
    r2 <- strsplit(readLines(f1, 5)[5], "\\s+")[[1]]
    if ({
        r1[[1]] == r2[[1]]
    } && {
        substr(r1[[2]], 1, 2) == "1:"
    } && {
        substr(r2[[2]], 1, 2) == "2:"
    }) 
        return("interleavedPairedEnd")
    rLast <- strsplit(.sys.do("tail -n 4", f1, intern = 1)[1], "\\s+")[[1]]
    if (length(grep("^2\\:", rLast[[2]])) && has3OrMoreReads) 
        return("catPairedEnd")
}
fastq_read_HowMany <- function(f1, want = c("read", "line")) {
    switch(fastq_EndType(f1), catPairedEnd = {
        r1 <- .sys.do("bioawk", "-c fastx", shQuote("{print $1}"), f1, "| head -n 1", intern = 1)
        r <- as.integer(.sys.do("bioawk -c fastx", .q(paste(c("BEGIN{ RN =0 } { if( $1==", shQuote(r1, "cmd"), ")", "RN = RN + 1} { if( RN == 2) { print NR; exit}}"), collapse = " ")), f1, intern = 1)) - 1
        switch(match.arg(want), read = r, line = r * 8)
    }, -1)
}
ieiwk.make.hs <- function(f1, f2.base = "hs.RData", save2file = 0, load = 1, load2env = parent.frame()) {
    source(f1, local = TRUE)
    obj <- ls()
    fl <- obj[sapply(obj, function(x) is.function(get(x)))]
    if (save2file) {
        save(list = fl, file = file.path(dirname(f1), f2.base))
    }
    if (load) {
        ieiwk.sapply(fl, function(f1) {
            get(f1) %=>% assign(f1, x, load2env)
        })
    }
    invisible()
}
ieiwk.make.hs.dir <- function(d1, f2.base = "hs.RData", unique = 1) {
}
.si <- function(file = ".RData", version = NULL, ascii = FALSE, compress = !ascii, compression_level = TRUE, safe = TRUE) {
    if (!is.character(file) || file == "") 
        stop("'file' must be non-empty string")
    opts <- getOption("save.image.defaults")
    if (is.null(opts)) 
        opts <- getOption("save.defaults")
    if (missing(safe) && !is.null(opts$safe)) 
        safe <- opts$safe
    if (missing(ascii) && !is.null(opts$ascii)) 
        ascii <- opts$ascii
    if (missing(compress) && !is.null(opts$compress)) 
        compress <- opts$compress
    if (missing(version)) 
        version <- opts$version
    if (safe) {
        outfile <- paste(file, "Tmp", sep = "")
        i <- 0
        while (file.exists(outfile)) {
            i <- i + 1
            outfile <- paste(file, "Tmp", i, sep = "")
        }
    }
    else outfile <- file
    on.exit(file.remove(outfile))
    save(list = ls(envir = .GlobalEnv, all.names = TRUE), file = outfile, version = version, ascii = ascii, compress = compress, compression_level = compression_level, envir = .GlobalEnv, precheck = FALSE)
    if (safe) 
        if (!file.rename(outfile, file)) {
            on.exit()
            stop("image could not be renamed and is left in ", outfile)
        }
    on.exit()
}
ieiwk.password_random.generator <- function(n = 32, set = 1) {
    candidates <- list(letters = list(lower = letters, upper = toupper(letters)), numbers = seq(0, 9), others = list(normal = c("_", ".", "-"), wild = c("~", "`", "!", "@", "#", "$", "%", "^", "&", "*", "(", ")", "+", "=", "\\", "|", "[", "]", "{", "}", "'", "\"", ":", ";", ",", "<", ">")))
    set <- switch(paste(set), `1` = unlist(candidates), `2` = unlist(c(candidates[["letters"]], candidates[["numbers"]], candidates[["others"]][["normal"]])), `3` = unlist(c(candidates[["letters"]], candidates[["numbers"]])))
    sample(set, n) %,% ""
}
ieiwk.fp.par <- function(..., sep = ";") {
    a <- list(...)
    for (v in names(a)) {
        assign(v, a[[v]], envir = parent.frame())
    }
    paste(sprintf("%s=%s", names(a), a), collapse = sep)
}
random.password.generator <- function(n, set = 1) {
    switch(paste(set), `1` = unlist(c(letters, toupper(letters), seq(0, 9), strsplit("~!@#$%^&*()_-+-=<>[]{}|<>,./?;\\", ""))))
    paste(sample(unlist(c(letters, toupper(letters), seq(0, 9), strsplit("~!@#$%^&*()_-+-=<>[]{}|<>,./?;\\", ""))), n), collapse = "")
}
not <- get("!")
wk.zim.find <- function(s1, where = "name", nb = "系统") {
    if (where == "name") 
        s1 <- str_c("*", s1, "*")
    s1 <- if (nchar(where)) 
        str_c(where, ":", s1)
    .wk("zim", "--search", nb, .q(s1), intern = 1) %=>% {
        if (length(x)) 
            str_c(":", x) %=>% .ss
    }
}
ieiwk.zim.31.12.27.100406 <- function(what = tp, to = "ss", ss.exact = FALSE, ss.field = c("name", "tag", "LinksTo")) {
    ss.field <- match.arg(ss.field)
    switch(to, ss = {
        .sys.do("zim --search", "/home/wk/wk/my/wd/db/zim/zb", sprintf("%s:%s", ss.field, sprintf(ifelse(ss.exact, "%s", "*%s*"), what)), intern = 1)
    }, switch(what, tp = {
        t1 <- Sys.time()
        sprintf(":%s:%s", as.integer(format(t1, "%Y")) - 1984, format(t1, "%m:%d:"))
    }))
}
wk.mlocate <- function(update = 0, pat = readline(), ignore.case = 1, regex = 1, db.wjj = hs.wjj(hs.home.expand("~/.db"), "mlocate"), wjj2do = c(hs.home.expand("~/wk")) %=>% x[dir.exists(x)], intern = 0) {
    db.wj <- wk.sapply(wjj2do, function(wjj1) {
        hs.wj(db.wjj, ifelse(wjj1 == "~", "home", basename(wjj1)))
    })
    1
    {
        if (0) 
            for (wjj1 in names(db.wj)) {
                t1 <- file.mtime(db.wj[[wjj1]])
                if (is.na(t1) || (difftime(Sys.time(), t1, units = "mins") > 100)) {
                  update <- 1
                  break
                }
            }
        if (update) {
            for (wjj1 in names(db.wj)) {
                hs.msg(paste("[locate更新]", wjj1))
                .wk("updatedb -l 0 -o", .sys.fpGood(db.wj[[wjj1]]), "-U", .sys.fpGood(wjj1))
            }
            hs.msg("完成", 0)
            return()
        }
    }
    1
    {
        r1 <- .wk("locate", "-b", "-e", if (ignore.case) 
            "-i", if (regex) 
            "--regex", sapply(db.wj, function(x) c("-d", x)), .q(pat), "2>/dev/null | sort | uniq", intern = 1)
        n1 <- 20
        cat(head(r1, n1), sep = "\n")
        if (length(r1) > n1) {
            hs.browser("34.11.10.225103")
        }
    }
}
.find <- function(ext) {
    if (!missing(ext)) 
        sprintf("*%s", ext)
}
wk.head <- function(x, n = 10) {
    a <- readLines(x, n)
    cat(a, sep = "\n")
}
ieiwk.deComment.codeFile.R <- function(wj1) {
    wj1.nr <- parse(wj1)
    a <- unlist(lapply(wj1.nr, function(x) {
        capture.output(bquote(.(x)))
    }))
    .ss(a, file = wj1)
}
ieiwk.deComment.codeFile.awk <- function(wj1) {
    nr <- readLines(wj1)
    p1 <- " *#ieiwk.*"
    lines2do <- which(grepl(p1, nr))
    if (length(lines2do)) {
        nr[lines2do] <- sub(p1, "", nr[lines2do])
    }
    .ss(nr, file = wj1)
}
ieiwk.deComment.codeFiles <- function(wjj1 = "/gpfsDATA/home/calvin/wk/my/wd/dm/R") {
    {
        wj <- ieiwk.list.files(wjj1, "(?i).*\\.r$", recursive = 1)
        sapply(wj, ieiwk.deComment.codeFile.R)
    }
    {
        wj <- ieiwk.list.files(wjj1, "(?i).*\\.(awk|pl|py)$", recursive = 1)
        sapply(wj, ieiwk.deComment.codeFile.awk)
    }
}
ieiwk.append <- function(x, values, after = length(x), replace = 0, use.what = ifelse(length(names(x)), "names", "identity")) {
    y <- use.what %(% list(x)
    if (replace) {
        i1 <- na.omit(match(values, y))
        if (length(i1)) {
            y <- y[-i1]
            x <- x[-i1]
        }
    }
    if (is.character(after)) 
        after <- match(after, y)
    r <- append(x, values, after)
    r
}
ieiwk.IRanges <- function(start = NULL, end = NULL, width = NULL, names = NULL) {
    if (!require(IRanges)) 
        stop("install IRanges first")
    IRanges(start = start, end = end, width = width, names = names)
}
ieiwk.GRanges <- function(seqnames = Rle(), ranges = NULL, strand = NULL, ..., seqlengths = NULL, seqinfo = NULL, start = NULL, end = NULL, width = NULL) {
    if (!require(GenomicRanges)) 
        stop("install GenomicRanges first")
    if (is.null(ranges)) 
        ranges <- ieiwk.IRanges(start, end, width)
    GRanges(seqnames = seqnames, ranges = ranges, strand = strand, ..., seqlengths = seqlengths, seqinfo = seqinfo)
}
wk.browser <- function(text = "", condition = {
}, expr = TRUE, skipCalls = 0L, t1, env = parent.frame()) {
    eval(bquote({
        text <- .(text)
        .r(str_interp("browser at ${ text} at ${ date()} on ${ .hostname()}"), title.postfix = "")
        browser(text = .(text), condition = .(condition), expr = .(expr), skipCalls = .(skipCalls))
    }), envir = env)
}
ieiwk.pubmed <- function(kw = {
}, high.if = 1, year.start = 2006, year.end = 3000) {
    browser("33.03.03.171028")
    r <- paste(c(if (length(kw)) paste("(", paste(sprintf("\"%s\"[Title/Abstract]", kw), collapse = "OR"), ")", sep = ""), if (high.if) "(\"0007-9235\"[Journal] OR \"0028-4793\"[Journal] OR \"0732-0582\"[Journal] OR \"0034-6861\"[Journal] OR \"0009-2665\"[Journal] OR \"1471-0072\"[Journal] OR \"0140-6736\"[Journal] OR \"1471-0056\"[Journal] OR \"1474-175X\"[Journal] OR \"0001-8732\"[Journal] OR \"0028-0836\"[Journal] OR \"1061-4036\"[Journal] OR \"0066-4154\"[Journal] OR \"1474-1733\"[Journal] OR \"1476-1122\"[Journal] OR \"0092-8674\"[Journal] OR \"1301-8361\"[Journal] OR \"0036-8075\"[Journal] OR \"1471-003X\"[Journal] OR \"0098-7484\"[Journal] OR \"1749-4885\"[Journal] OR \"1474-1776\"[Journal] OR \"0306-0012\"[Journal] OR \"1748-3387\"[Journal] OR \"0031-9333\"[Journal] OR \"1535-6108\"[Journal] OR \"0066-4146\"[Journal] OR \"1529-2908\"[Journal] OR \"1543-5008\"[Journal] OR \"0147-006X\"[Journal] OR \"1934-5909\"[Journal] OR \"0079-6700\"[Journal] OR \"1474-4422\"[Journal] OR \"1087-0156\"[Journal] OR \"1470-2045\"[Journal] OR \"1078-8956\"[Journal] OR \"0066-4197\"[Journal] OR \"0001-4842\"[Journal] OR \"0362-1642\"[Journal] OR \"1074-7613\"[Journal] OR \"1740-1526\"[Journal] OR \"0066-4278\"[Journal] OR \"1755-4330\"[Journal] OR \"0370-1573\"[Journal] OR \"0031-6997\"[Journal] OR \"1553-4006\"[Journal] OR \"0027-8424\"[Journal] OR \"0006-4971\"[Journal] OR \"1063-5157\"[Journal] OR \"0021-9525\"[Journal] OR \"0002-9297\"[Journal] OR \"1544-9173\"[Journal] OR \"0890-9369\"[Journal] OR \"0962-8924\"[Journal] OR \"0955-0674\"[Journal] OR \"1088-9051\"[Journal] OR \"1534-5807\"[Journal] OR \"1097-2765\"[Journal] OR \"0896-6273\"[Journal] OR \"0009-7322\"[Journal] OR \"1097-6256\"[Journal] OR \"1549-1277\"[Journal] OR \"1465-7392\"[Journal] OR \"0168-9525\"[Journal] OR \"0960-9822\"[Journal] OR \"0261-4189\"[Journal] OR \"1474-7596\"[Journal] OR \"1553-7390\"[Journal] OR \"1744-4292\"[Journal] OR \"0028-3878\"[Journal] OR \"1001-0602\"[Journal] OR \"0305-1048\"[Journal])", 
        sprintf("(\"%s/01/01\"[Date - Publication] : \"%s\"[Date - Publication])", year.start, year.end)), collapse = " AND ")
    .ss(r)
}
.wk.baloosearch <- function(q1, d = {
}) {
    .wk("baloosearch", if (length(d)) 
        c("-d", d), q1, "| uniq", "|", .sys.less())
}
.wk.open <- function(wj1, bin = "nautilus") {
    .wk(bin, wk.wjlj(wj1), "&")
}
.wk.open.pdf <- function(wj1, bin = "PDFXEdit") {
    if (bin == "PDFXEdit") {
        wj1 <- path.expand(wj1)
        if (!wk.re(wj1, "^/", "detect")) 
            wj1 <- wk.wj(getwd(), wj1)
        wj1 <- str_c("Z:", wj1)
        wj1 <- wk.re(wj1, "/", "replace", replacement = "\\", all = 1)
    }
    bin <- switch(bin, PDFXEdit = "~/rj/exe/PDFXEdit6_Portable/PDFXEdit.exe", FoxitReader = "/home/wk/rj/foxitreader/FoxitReader.sh")
    browser("33.06.16.181001")
    .sys("wine", bin, wj1)
}
.wk.wjj.db <- function(uid1, open = 1) {
    wjj1 <- path.expand(wk.wjj(.mfp(uid1, "~/wk/my/wd/db")))
    if (open) 
        .wk.open(wjj1)
    else .ss(path.expand(wk.wjj(.mfp(uid1, "~/wk/my/wd/db"))))
}
wk.is.a.local.machine <- function(hn1 = Sys.info()["nodename"]) {
    hn1 %in% c("wk-Veriton-D630")
}
wk.remote.setup <- function() {
    wj.jg.33.06.19.154855 <- "~/R.communication.txt"
    port <- 6011L
    op <- options(timeout = 99999)
    on.exit(options(op), add = TRUE)
    repeat {
        con1 <- socketConnection(port = port, blocking = TRUE, open = "rb", server = TRUE)
        r <- .try(eval(parse(con1, encoding = "UTF-8")))
        close(con1)
        .cat.time()
        if (0) {
            .wk("sleep .5")
            con1 <- socketConnection("log01", port = port)
            writeLines(capture.output(r), con1)
            close(con1)
        }
        .ss(head(capture.output(r), 99), file = wj.jg.33.06.19.154855, append = FALSE)
    }
}
wk.remote <- function(e1) {
    e1 <- substitute(e1)
    wj.jg.33.06.19.154855 <- "~/R.communication.txt"
    wj.comm <- ieiwk.tempfile()
    on.exit(unlink(wj.comm), add = TRUE)
    .ss(head(trim(capture.output(bquote({
        con1 <- socketConnection("fat01", port = .(port))
        writeLines(capture.output(quote({
            .(e1)
        })), con1)
        close(con1)
    }))), -1)[-1], file = wj.comm)
    wk.rsync(wj.comm, wj.comm, ssh = wk.ssh, ssh.side = 2)
    .wk("ssh", "-p", "7050", "calvin@113.108.182.54", "R --no-save --slave -e", .q(.q(.sys("source(", .q(wj.comm), ")"))), wait = 1)
    wk.rsync(wj.jg.33.06.19.154855, dirname(wj.jg.33.06.19.154855), ssh = wk.ssh, ssh.side = 1)
    .ss(readLines("~/R.communication.txt"))
}
wk.filter <- function(x, f) {
    x[wk.sapply(x, f) %=>% unlist]
}
wk.axel <- function(url1, wjj2 = ".", shm = file.exists("/dev/shm") && (!hs.grepl(url1, "(?i)\\.([xg]z|zip|tgz|bgz|rar)$"))) {
    if (0) {
        wj1 <- "/home/wk/swxx/sj/xm/hy/gtf2f1/sh/rna-seq/ENCFF401ZXD.fastq.gz"
        t1 <- file.mtime(wj1)
        repeat {
            .wk("sleep 30")
            t2 <- file.mtime(wj1)
            if (difftime(t2, t1, units = "secs") > 15) 
                0
        }
    }
    wj2 <- hs.wjm(url1, dn = wjj2, add = if (shm) 
        "gz") %=>% normalizePath
    wj1.st <- paste0(wj2, ".st")
    if (file.exists(wj2) && (!file.exists(wj1.st))) 
        return(wj2)
    wjj2_tmp <- if (shm) {
        hs.wjj(hs.wj.temp())
    }
    else wjj2
    if (0) 
        .wk("df", "/dev/shm", intern = 1)[2] %=>% strsplit("\\s+") %=>% unlist %=>% x[4] %=>% as.integer
    wj2_tmp <- hs.with_wd(wjj2_tmp, {
        wj1 <- basename(url1)
        cmd1 <- .sys("axel -ca -T 2", .q(url1))
        .wk(cmd1)
        N <- 0
        while (file.exists(wj1.st) && (N < 1000)) {
            message("retrying...")
            .wk(cmd1)
            N <- N + 1
        }
        normalizePath(wj1)
    })
    if (shm) {
        .wk("cat", wj2_tmp, "|", wk.gz.good(), ">", .sys.fpGood(wj2))
        on.exit(hs.wj.rm(wj2_tmp), add = TRUE)
    }
    else wj2 <- wj2_tmp
    wj2
}
wk.wget <- function(url1, wjj2 = ".", wjm = basename(url1), timestamping = 0, touch = 0, T = 10, log = 0, gz = !hs.grepl(wjm, "(?i)\\.(t?gz|zip|xz|7z|rar|feather)$")) {
    hs.with_wd(hs.wjj(wjj2), {
        if (0) {
            wjjTemp <- tempfile("wget.", tmpdir = ".") %=>% hs.wjj
            on.exit(hs.rm.wj(wjjTemp, 1), add = TRUE)
            wjLog <- tempfile("log", tmpdir = wjjTemp)
        }
        if (log) {
            wjLog <- tempfile("log_wget.", tmpdir = ".")
        }
        if (0) 
            .wk("wget", "-c", "--random-wait", hs.url_good(url1), "-o", wjLog)
        if (gz) {
            if (!hs.grepl(wjm, "\\.gz$")) {
                wjm %<=>% paste0(".gz")
                if (file.exists(wjm)) 
                  return(normalizePath(wjm))
            }
            wjm <- c("- | gzip >", .q(wjm))
        }
        else {
            if (file.exists(wjm)) 
                return(hs.wj.portable(wjm))
        }
        cmd1 <- .sys(hs.home.expand("~/rj/miniconda3/bin/wget") %=>% {
            if (file.exists(x)) 
                x
            else "wget"
        }, if (!gz) 
            "-c", if (timestamping && (!gz)) 
            "--timestamping", "-T", T, "-t 9999", if (log) 
            c("-o", wjLog), .q(url1), if (length(wjm)) 
            c("-O", wjm))
        .wk(cmd1)
        if (touch) 
            .wk("touch", .q(wjm))
        if (log) {
            info <- tail(readLines(wjLog) %not% "", 1)
            return(hs.wj(wjj2, hs.sub.capture(info, "File ‘(.+)’ already there")))
        }
        return(hs.wj.portable(wjm))
        if (0) {
            Log <- wjLog %=>% readLines
            if (hs.grepl(tail(tail(Log) %not% "", 1), "saved")) {
                wjGot <- normalizePath(hs.listWj(wjjTemp, recursive = FALSE)) %not% normalizePath(wjLog)
                if (length(wjGot)) {
                  file.rename(wjGot, wjj2)
                }
            }
            else {
                hs.browser("35.08.24.182324")
            }
        }
    })
}
wk.curl <- function(dz1, wjj2) {
    hs.with_wd(hs.wjj(wjj2), {
        .wk("curl", "-C -", "-R", "-J", "-O", dz1)
    })
}
R_session.location.hs <- function() {
    a1 <- .wk("ifconfig", intern = 1)
    a1 <- hs.sepSeq(a1, sep.i = which(switch(Sys.info()["sysname"], Linux = {
        a1 == ""
    }, Darwin = {
        hs.grepl(a1, "^[^\\t]")
    })))
    a1 %<=>% {
        lapply(x, "%not%", "") %=>% hs.nonempty
    }
    ip <- unlist(sapply(a1, . %=>% hs.re("\\d+\\.\\d+\\.\\d+\\.\\d+")))
    if ("192.168.10.237" %in% ip) {
        "fwq"
    }
    else if ("172.16.28.105" %in% ip) {
        "hp1"
    }
    else if (length(hs.grep(ip, "^192\\.115\\."))) {
        "a409"
    }
    else if ("10.80.0.29" %in% ip) {
        "cpu14"
    }
    else if ("192.168.0.108" %in% ip) {
        "r820"
    }
    else if ("192.168.0.100" %in% ip) {
        ""
    }
    else ""
}
R_session.location <- R_session.location.hs()
wk_node_remote <- c(R_session.location.hs())
ieiwk.ssh.db <- local({
    if (0) {
        dz.db <- c(sg = "172.18.202.8", hp = "172.18.202.7", g1 = "gate1.picb.ac.cn", g2 = "gate2.picb.ac.cn", hh = "113.108.182.54", gsp = "172.21.109.131")
        p.db <- c(sg = 53641, hp = 53641, g1 = 22, gsp = 22, g2 = 22, hh = 7050)
        u.db <- c(sg = "wengkai", hp = "wengkai", g1 = "ieiwk", g2 = "ieiwk", hh = "calvin", gsp = "wk")
        h.db <- c(hh = "/gpfsDATA/home")
    }
    {
        if (!file.exists(hs.home.expand("~/.ssh/config"))) 
            return()
        ssh.config <- readLines(hs.home.expand("~/.ssh/config"))
        ssh.config <- hs.sepSeq(ssh.config, sep.p = "^Host\\s")
        names(ssh.config) <- sapply(ssh.config, function(x1) {
            hs.re(x1[1], "(?<=\\s)[^\\s]+")
        })
        ssh.config["apt"]
        ssh.ip <- function(ssh1) {
            hs.sub(hs.grepV(ssh.config[[ssh1]], "^\\s*HostName\\s*"), "^\\s*HostName\\s*")
        }
    }
    db1 <- list(jz = c("calvin", "192.115.110.254"), cpu14 = c("calvin", "cpu14"), a409 = c("wk", switch(R_session.location, jz = , cpu14 = , cpu13 = , a409 = "192.115.103.196", "172.16.3.62")), ucdom = c("ucdom", "s2.natfrp.org", "48875"), seq1 = c("wk", "172.46.114.134"), r820 = c("wk", "r820"), hp1 = c("wk", "hp1"), apt = c("root", "apt"), n9 = c("kweng", "n9"), n8 = c("kweng", "n8"), hh = c("calvin", "113.108.182.54", "7050"))
    function(dz, p = 22, u = "") {
        if (dz %in% names(db1)) {
            db1 <- db1[[dz]]
            u <- db1[1]
            dz <- db1[2]
            p <- db1[3]
            if (is.na(p)) 
                p <- "22"
        }
        a <- c(dz, p, u)
        names(a) <- c("dz", "port", "user")
        a
    }
})
wk.match_partial <- local({
    hs.require("stringdist")
    hs1 <- function(s1, dot2_ = 1) {
        s1 <- tolower(s1)
        if (dot2_) 
            s1 %<=>% hs.gsub("_ \\.", "_")
        s1
    }
    function(set1, set2, dot2_ = 1) {
        hs2 <- stringdist::stringsim
        set1_new <- hs1(set1, dot2_ = dot2_)
        set2_new <- hs1(set2, dot2_ = dot2_)
        nm <- set1_new %not% set2_new
        jg <- lapply(nm, function(nm1) {
            D <- hs2(nm1, set2_new, weight = c(0.9, 0.9, 1, 1))
            i1 <- which.max(D)
            set2_new1 <- set2_new[i1]
            data.table(similarity = D[i1], set1_new = nm1, set2_new = set2_new1, set1 = set1[match(nm1, set1_new)], set2 = set2[match(set2_new1, set2_new)])
        })
        jg %<=>% rbindlist
        setorderv(jg, "similarity", -1)
    }
})
wk.view <- function(text1, pattern1, C = 3) {
    cat(text1, sep = "\n", file = paste("| ag -C", C, .q(pattern1)))
}
wk.color <- c(blue = "navy", red = "firebrick1")
hs.snv_is <- function(ref, alt) {
    (tolower(ref) %in% c("a", "c", "g", "t")) & (tolower(alt) %in% c("a", "c", "g", "t"))
}
v.pct.hs <- function(v1, order = {
}, nl = 0, P = 1) {
    a1 <- hs.table(v1)
    a2 <- round(a1/sum(a1) * 100, 2)
    a2 <- paste0(names(a1), if (nl) 
        "\n"
    else " ", "(n=", a1, if (P) 
        paste0("; ", a2[names(a1)], "%")
    else "", ")")
    names(a2) <- names(a1)
    r1 <- factor(a2[v1], a2)
    if (length(order)) 
        r1 <- reorder(r1, X = order)
    r1
}
fq_wj.hs.wjj <- function(wjj1, R = 0, str2exclude = "unpaired", str2include = {
}, fq.pattern = "(?i)\\.f(ast)?q(\\.gz)?$", ...) {
    jg <- hs.wjj.ls(wjj1, fq.pattern, R = R, ...)
    for (str1 in str2exclude) {
        i1 <- hs.grep(basename(jg), paste0("(?i)", str1))
        if (length(i1)) 
            jg <- jg[-i1]
    }
    for (str1 in str2include) {
        i1 <- hs.grep(basename(jg), paste0("(?i)", str1))
        jg <- jg[i1]
        if (!length(jg)) 
            break
    }
    jg
}
.sys.speedseq <- function(what = "wjj.speedseq") {
    f1 <- sys.function()
    switch(what, wjj.speedseq = "~/swxx/rj/speedseq", wjj.anno = file.path(f1("wjj.speedseq"), "annotations"), f.speedseq = , wj.speedseq = file.path(f1("wjj.speedseq"), "bin/speedseq"), wj.sambamba = file.path(f1("wjj.speedseq"), "bin/sambamba"), f.ref = , wj.ref = "~/swxx/sj/jyz/human/1kg/37/human_g1k_v37.fasta", `f.ceph18.b37.include.2014-01-15.bed` = , `wj.ceph18.b37.include.2014-01-15.bed` = file.path(f1("wjj.anno"), "ceph18.b37.include.2014-01-15.bed"), `ceph18.b37.lumpy.exclude.2014-01-15.bed` = file.path(f1("wjj.anno"), 
        "ceph18.b37.lumpy.exclude.2014-01-15.bed"))
}
.sys.annovar <- function(what = "", wjj0 = "~/swxx/rj/annovar") {
    file.path(wjj0, what)
}
gene.id_symbol.hs.str <- function(str1) {
    strsplit(str1, "[,; ]+") %=>% sapply(hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937"))
}
gene_id_name.37_04_29_145914 <- function(v1) {
    hs.grepV(v1, "(?i)gene[_\\.]?id")
}
log_fold_change_name.37_04_29_150019 <- function(v1) {
    hs.grepV(v1, "(?i)l(og[0-9]*)?[_\\.]?f(old)?[_\\.]?c(hange)?")
}
qvalue_name.37_04_29_150328 <- function(v1) {
    hs.grepV(v1, "(?i)fdr|q[_\\.]?v(al(ue)?)?")
}
pvalue_name.37_04_29_150328 <- function(v1) {
    hs.grepV(v1, "(?i)p[_\\.]?v(al(ue)?)?")
}
hs.pfm2pwm <- function(pfm1, bg = {
}, pseudocounts = 0.8) {
    if (!length(bg)) 
        bg <- hs.c(c("A", "C", "G", "T"), rep(0.25, 4))
    prior <- bg/sum(bg)
    post <- (pfm1 + bg * pseudocounts)/(colSums(pfm1) + sum(bg) * pseudocounts)
    log2(post/prior)
}
wk.pfm2pwm.for_TFBSTools <- function(pfm1, bg = {
}, ...) {
    if (!length(bg)) 
        bg <- hs.c(c("A", "C", "G", "T"), rep(0.25, 4))
    pwm1 <- hs.pfm2pwm(pfm1, bg = {
    })
    PWMatrix(profileMatrix = pwm1, bg = bg, ...)
}
.sys.wjj.fa_wj <- function(wjj1, cat = ot != "jg", ot = "cmd") {
    cmd1 <- .sys.find(.sys.fpGood(wjj1), ".*\\.fn?a(\\.b?gz)?$", iname = c("*unlocalized*", "*rna*"), iname.inv = 1)
    jg <- .wk(cmd1, intern = 1)
    if ((length(jg) > 1) && any(hs.grepl(jg, "/unplaced\\."))) {
        cmd1 <- .sys(cmd1, "! -iname unplaced.*")
    }
    if (cat) 
        cmd1 <- .sys(cmd1, "| xargs -I {} ~/org/dm/sh/rj/cat_wk {}")
    switch(ot, cmd = cmd1, jg = .wk(cmd1))
}
seq_org.hs.fa <- function(fa.wj, org, fa.wjj = {
}, seq.org.wj = if (length(fa.wjj)) hs.fp(fa.wjj, "seq.org.tsv") else hs.wjm(fa.wj, ext = "seq.org.tsv")) {
    if (!file.exists(seq.org.wj)) {
        if (missing(org)) 
            stop("37_03_05_113031")
        cmd1 <- if (length(fa.wjj)) {
            .sys(.sys.wjj.fa_wj(fa.wjj), "| perl -lane", .q("if( /^>/) { print}"))
        }
        else {
            .sys(.sys.cat.text(fa.wj), "| perl -lane", .q("if( /^>/) { print}"))
        }
        a1 <- .wk(cmd1, intern = 1)
        for (a2 in hs.re(a1, "(?<=^>)[^\\s]*")) hs.cat(a2, org, sep = "\t", file = seq.org.wj)
    }
    seq.org.wj
}
blastdb.wk.fa <- function(wjj1, org) {
    makeblastdb <- "~/swxx/rj/ncbi-blast-2.11.0+-x64-linux/ncbi-blast-2.11.0+/bin/makeblastdb"
    wjj.36_03_04_003344 <- getwd()
    on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
    setwd(wjj1)
    title1 <- basename(normalizePath("."))
    seq.org.wj <- "seq.org.tsv"
    if (!file.exists(seq.org.wj)) {
        if (missing(org)) 
            stop("37_03_05_124954")
        seq_org.hs.fa(fa.wjj = ".", org = org, seq.org.wj = seq.org.wj)
    }
    if (!all(file.exists(paste0("blastdb.", c("ndb", "nhr", "nin", "nog", "nos", "not", "nsq", "ntf", "nto"))))) {
        .wk(.sys.wjj.fa_wj("."), "|", makeblastdb, "-title", title1, "-out", "blastdb", "-parse_seqids -blastdb_version 5", "-taxid_map", seq.org.wj, "-dbtype", "nucl")
    }
    hs.fp(normalizePath("."), "blastdb")
}
wk.blast <- function(fa, db, wj2 = {
}, seqid = {
}, blastn = "~/swxx/rj/ncbi-blast-2.11.0+-x64-linux/ncbi-blast-2.11.0+/bin/blastn", blastdb_aliastool = hs.fp(dirname(blastn), "blastdb_aliastool"), global_best = 1, append = hs.len1(wj2) && file.exists(wj2), nc = Ncpu()[1]) {
    if (length(seqid)) {
        hs.browser("37.03.09.184634")
        .wk(blastdb_aliastool, "-seqid_file_in")
    }
    arg <- .sys(if (global_best) 
        c("-max_hsps", global_best, "-num_alignments", global_best), "-db", .sys.fpGood(db), if (nc > 0) 
        c("-num_threads", nc), if (length(wj2) == 1) 
        c(if (hs.grepl(wj2, "(?i)\\.gz$")) "| gzip", if (append) ">>" else ">", .sys.fpGood(wj2)))
    if (hs.len1(fa) && file.exists(fa)) {
        .wk(blastn, "-query", fa, arg)
    }
    else cat(fa, sep = wk.n, file = .sys("|", blastn, arg))
    if (length(wj2)) 
        wj2
}
geneset.heatmap.37_04_29_151302 <- function(geneset1, exp1, geneset1_name, gene_name = {
}, de1 = if (hs.len1(de1_wj)) wk.fread(de1_wj), de1_wj = {
}, q.co1 = 0.05, lfc.co1 = log2(fc.co1), fc.co1 = 1.2, q.cn = qvalue_name.37_04_29_150328(names(de1)), lfc.cn = log_fold_change_name.37_04_29_150019(names(de1)), fc.cn = "fc", gene_id_cn = gene_id_name.37_04_29_145914(names(de1)), sample_info1, control1 = {
}, subject = sample_info1[[1]] %not% control1, org1, condition1 = "", to_z = 1, sample_info_for_z_i = 1, legend.name = if (to_z) "Z" else "Exp", wj2 = hs.wj(wjj2, hs.pastec0(geneset1_name, if (nzchar(condition1)) c(".", condition1), ".heatmap.pdf")), wjj2 = ".") {
    if ((!length(gene_name)) && length(de1)) 
        gene_name <- local({
            i1 <- (de1[[q.cn]] < q.co1) & (abs(de1[[lfc.cn]]) > lfc.co1)
            if (any(i1)) {
                star_n <- rep("", nrow(de1))
                i2 <- de1[[q.cn]] < 0.05
                if (any(i2)) 
                  star_n[i2] <- "*"
                i2 <- de1[[q.cn]] < 0.1
                if (any(i2)) 
                  star_n[i2] <- "**"
                de1[[gene_id_cn]][i1] %=>% hs.c(x, paste0(x, " (", star_n[i1], ")"))
            }
        })
    geneset1 <- geneset1 %and% rownames(exp1)
    if (length(geneset1) < 2) 
        stop("[37_04_28_222754|基因太少了]")
    exp1 <- exp1[geneset1, , drop = 0]
    sample_to_use <- rownames(sample_info1)[sample_info1[[1]] %in% (control1 %or% subject)]
    exp1 <- exp1[, colnames(exp1) %in% sample_to_use, drop = 0]
    sample_info1 <- sample_info1[sample_to_use, , drop = 0]
    if (to_z) {
        sample_to_use_for_z <- if (missing(control1)) {
            colnames(exp1)
        }
        else rownames(sample_info1)[sample_info1[[sample_info_for_z_i]] %in% control1]
        mean_for_z <- hs.rowMeans(exp1[, sample_to_use_for_z, drop = 0])
        sd_for_z <- hs.rowSds(exp1[, sample_to_use_for_z, drop = 0])
        m1 <- sapply(seq.int(nrow(exp1)), function(row_i) {
            (exp1[row_i, ] - mean_for_z[row_i])/sd_for_z[row_i]
        }) %=>% t
        i1 <- !hs.clean(m1, ot = "l")
        if (any(i1)) 
            m1[i1] <- NA
        rownames(m1) <- rownames(exp1)
    }
    else m1 <- exp1
    m1 <- apply(m1, 2, hs.deInf)
    gene_name %<>% {
        .[names(.) %in% rownames(m1)]
    }
    if (length(gene_name)) {
        i1 <- match(names(gene_name), rownames(m1))
        rownames(m1)[i1] <- gene_name
    }
    m1 <- m1[hs.rowMeans(is.na(m1)) < 0.5, , drop = 0]
    hs.hsgly("ComplexHeatmap/36.08.12.233720")(m1, legend.name = "Z", wj = wj2, column_split = sample_info1)
}
wk.color <- c(blue = "navy", red = "firebrick1", green = "darkolivegreen2", gray = "darkgray")
wk.plot.step <- function(x, y, ...) {
    plot(x, y, type = "n", ...)
    for (i in seq.int(length(x) - 1)) {
        lines(x[c(i, i + 1)], rep(y[i], 2))
        lines(rep(x[i + 1], 2), y[c(i, i + 1)])
    }
}
ieiwk.heatmap <- function(x, Rowv = NA, Colv = NA, distfun = dist, hclustfun = hclust, reorderfun = function(d, w) reorder(d, w), add.expr, symm = FALSE, revC = identical(Colv, "Rowv"), scale = c("none", "row", "column"), na.rm = TRUE, margins = c(5, 5), ColSideColors, RowSideColors, cexRow = 0.2 + 1/log10(nr), cexCol = 0.2 + 1/log10(nc), labRow = NULL, labCol = NULL, main = NULL, xlab = NULL, ylab = NULL, keep.dendro = FALSE, verbose = getOption("verbose"), revR = 1, di = dim(x), nr = di[1L], nc = di[2L], 
    ...) {
    if (revR) 
        x <- x[rev(seq(nrow(x))), , drop = FALSE]
    heatmap(x, Rowv, Colv, distfun, hclustfun, reorderfun, add.expr, symm, revC, match.arg(scale), na.rm, margins, ColSideColors, RowSideColors, cexRow, cexCol, labRow, labCol, main, xlab, ylab, keep.dendro, verbose, ...)
}
ieiwk.heatmap.2 <- function(m1, col = colorpanel(75, "green", "gray", "red"), key = TRUE, symkey = FALSE, density.info = "none", trace = "none", cexRow = 0.5, margins = c(5, 5)) {
    require("gplots")
    heatmap.2(m1, col = col, key = as.logical(key), symkey = as.logical(symkey), density.info = density.info, trace = trace, cexRow = cexRow, margins = margins)
}
ieiwk.density <- function(x, from = min(x), to = max(x), bw = "sj", ...) {
    density(x = x, from = from, to = to, bw = bw, ...)
}
Mycol2rgb <- function(x, alpha = 1) {
    x <- col2rgb(x)
    a1 <- as.list(x/255)
    if (length(alpha)) 
        a1[[length(a1) + 1]] <- alpha
    do.call(rgb, a1)
}
ieiwk.col2rgb <- Mycol2rgb
hs.fig_wj_wjj <- function() {
    c(if (dir.exists("/dev/shm")) hs.fp("/dev/shm", Sys.info()["user"]), "~/swxx", "~/tmp") %=>% x[dir.exists(hs.home.expand(x))][1]
}
.pdf <- function(exp1, wj = "tmp", width = height + 5, height = 15 * rh, rh = 1, wjj2 = hs.fig_wj_wjj(), family = "mono", local = 1, env1 = parent.frame(), wjm = "tmp", append = 0, el = {
}, verbose = 1) {
    if (missing(wj)) 
        wj <- hs.wj(wjj2, wj)
    if (!hs.grepl(wj, "(?i)\\.pdf$")) 
        wj %<=>% paste0(".pdf")
    if (0) 
        if (length(el)) {
            wj_temp <- paste0(sapply(el, function(i1) hs.wj.temp()), ".pdf")
            on.exit(hs.wj.rm(wj_temp), add = TRUE)
            for (i1 in seq_along(el)) {
                wj2 <- wj_temp[i1]
                .pdf(el[[i1]], wj = wj2, env1 = env1, width = width[hs.recycle(i1, length(el))], height = height[hs.recycle(i1, length(el))], verbose = 0)
            }
            wjj.36_03_04_003344 <- getwd()
            on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
            setwd(dirname(wj_temp[1]))
            .wk(hs.home.expand("~/swxx/rj/pdftk/pdftk"), .q(basename(wj_temp)), "cat", "output", .sys.fpGood(wj))
            if (verbose) 
                message(wj)
            return(wj)
        }
    wj_tmp <- if (append && file.exists(wj)) 
        hs.wj.temp(ext = "pdf")
    else wj
    if (inherits(exp1, c("grob", "ggplot")) || (inherits(exp1, "list") && inherits(exp1[[1]], c("grob", "ggplot")))) {
        if (inherits(exp1, "ggplot")) 
            exp1 %<=>% +ggplot2::theme(text = ggplot2::element_text(family = family))
        ggplot2::ggsave(wj, exp1, width = width, height = height, units = "cm", limitsize = FALSE)
        return(wj)
    }
    pdf(hs.wj(wj_tmp), height = height/2.54, width = width/2.54, family = family, fonts = family)
    on.exit(dev.off(), add = TRUE)
    {
        par.37_02_23_180042 <- par(no.readonly = TRUE)
        on.exit(par(par.37_02_23_180042), add = TRUE)
    }
    eval(substitute(exp1), if (local) 
        new.env(parent = env1)
    else env1)
    if (wj_tmp != wj) {
        hs.browser("38.03.10.093340")
        wj2_tmp <- hs.wj.temp(ext = "pdf")
        on.exit(hs.wj.rm(wj_tmp), add = TRUE)
        hs.hsgly("pdf/合并多个文件.36_11_09_121255")(c(wj, wj_tmp), wj2_tmp)
        hs.wj.move(wj2_tmp, wj)
    }
    if (0) {
        env2 <- if (local) 
            new.env(parent = env1)
        else env1
        sapply(list(...), eval, env2)
    }
    if (verbose) {
        1
        {
            message(normalizePath(wj) %=>% hs.wjm.home)
        }
    }
    wj
}
.png <- function(exp1, wj = "tmp", width = height + 5, height = 10 * rh, wjj2 = hs.fig_wj_wjj(), units = "cm", rw = 1, rh = 1, pointsize = 12, res = 100, env1 = parent.frame(), family = "mono", crop = 1) {
    off <- 0
    on.exit(if (!off) dev.off(), add = TRUE)
    if (!hs.grepl(wj, "(?i)\\.png$")) 
        wj %<=>% paste0(".png")
    if ((!hs.grepl(wj, "/")) && dir.exists(wjj2)) 
        wj <- hs.fp(wjj2, wj)
    here <- new.env(parent = env1)
    if (crop) {
        wj_real <- wj
        wj_temp <- hs.wj.temp()
        on.exit(hs.wj.rm(wj_temp), add = TRUE)
        wj <- wj_temp
    }
    png(hs.wj(wj), width, height, units = units, pointsize = pointsize, res = res, family = family)
    eval(substitute(exp1), here)
    if (crop) {
        dev.off()
        off <- 1
        hs.hsgly("切白边/2023_08_18_180631")(wj, wj2 = wj_real)
        wj <- wj_real
    }
    wj <- normalizePath(wj)
    message(hs.wjm.home(wj))
    wj
}
.figs <- function(..., type = list(.png, .pdf), nc = Inf, env1 = parent.frame()) {
    fc <- as.list(substitute(list(...)))
    i <- match("wj", names(fc), nomatch = 2 + 1)
    wj <- eval(fc[[i]], envir = env1)
    if (hs.grepl(wj, "(?i)\\.(pdf|png)$")) 
        fc[[i]] <- hs.sub(wj, "(?i)\\.[^\\.]+$")
    wk.mclapply(type, . %=>% {
        fc[[1]] <- x
        fc[["env1"]] <- env1
        eval(as.call(fc), envir = env1)
    }, mc.cores = nc)
}
.fig <- function(..., env1 = parent.frame()) {
    fc <- as.list(substitute(list(...)))
    i <- match("wj", names(fc), nomatch = 2 + 1)
    wj <- eval(fc[[i]], envir = env1)
    if (hs.grepl(wj, "(?i)\\.(pdf|png)$")) {
        hs1 <- tolower(hs.re(wj, "(?i)\\.[^\\.]+$"))
        fc[[i]] <- hs.sub(wj, "(?i)\\.[^\\.]+$")
    }
    else hs1 <- ".pdf"
    fc[[1]] <- hs.find_function(hs1)
    fc[["env1"]] <- env1
    eval(as.call(fc), envir = env1)
}
.tiff <- function(exp1, wj = "tmp", width = 480, height = 480, wjj2 = hs.fig_wj_wjj(), pointsize = 12, env1 = parent.frame(), ...) {
    on.exit(dev.off(), add = TRUE)
    if (!hs.grepl(wj, "(?i)\\.tiff$")) 
        wj %<>% hs.wjm(add = "tiff")
    if ((!hs.grepl(wj, "/")) && dir.exists(wjj2)) 
        wj <- hs.fp(wjj2, wj)
    here <- new.env(parent = env1)
    tiff(wj, width, height, pointsize = pointsize, ...)
    eval(substitute(exp1), here)
    wj
}
.jpeg <- function(..., wj = hs.home.expand("~/tmp.jpeg"), wjj2 = hs.fig_wj_wjj(), width = 480, height = 480, pointsize = 12) {
    on.exit(dev.off(), add = TRUE)
    if (tolower(hs.re(wj, "[^\\.]+$")) != "jpeg") 
        wj %<>% hs.wjm(add = "jpeg")
    jpeg(wj, width, height, pointsize = pointsize)
    sapply(list(...), function(src) {
        eval(src, env = .GlobalEnv)
    })
}
ieiwk.plot.blank <- function(main) {
    op <- par(xaxt = "n", yaxt = "n")
    plot(1:3, type = "n", xlab = "", ylab = "", main = main)
    par(op)
}
ieiwk.plot.text <- function(main = "", t1, width = 79, cex = 0.7, family = "mono", wrap = 0) {
    ieiwk.plot.blank(strwrap(main, 31))
    if (wrap) 
        t1 <- strwrap(t1, width)
    mtext(paste(t1, collapse = "\n"), line = -1, at = 1, adj = 0, padj = 1, family = family, cex = cex)
}
mergePDF <- function(files, file, gsversion = NULL, in.file = NULL) {
    if (is.null(gsversion)) {
        gsversion <- names(which(Sys.which(c("gs", "gswin32c", "gswin64c")) != ""))
        if (length(gsversion) == 0) 
            stop("Please install Ghostscript and ensure it is in your PATH")
        if (length(gsversion) > 1) 
            stop("More than one Ghostscript executable was found:", paste(gsversion, collapse = " "), ". Please specify which version should be used with the gsversion argument")
    }
    .sys.do(gsversion, "-dBATCH -dNOPAUSE -q -sDEVICE=pdfwrite", sprintf("-sOutputFile=%s", .sys.fpGood(file)), files)
    invisible(file)
}
wk.venn.diagram.2 <- local({
    hs.require("VennDiagram")
    function(x1, u1, col1 = c("cornflowerblue", "darkorchid1"), wjm = hs.gsub(paste0(main, ".", paste(names(x1), collapse = "_vs_")), " ", "_"), wjj = hs.home.expand("~/tmp"), wj.type = "pdf", side.length = 10, main = paste(names(x1), collapse = " vs "), col = "black", ...) {
        x1 <- lapply(x1, . %=>% (`%and%`(u1)))
        t1 <- fisher.test.ele(u1, x1[[1]], x1[[2]])
        sub <- paste0("Odds ratio: ", round(t1$est, 2), " (P ", t1$p.v %=>% {
            if (x < 2.2e-16) 
                "< 2.2e-16"
            else paste("=", format(t1$p.v, sci = TRUE))
        }, ")")
        inv <- sapply(x1, length) %>% {
            .[1] < .[2]
        }
        cat.pos <- 20 %=>% {
            ifelse(!inv, ., 180 - .)
        } %=>% (x * c(-1, 1))
        wj <- hs.wj(wjj, paste0(wjm, ".", wj.type))
        .pdf({
            grid.draw(venn.diagram(x = x1, filename = {
            }, height = side.length, width = side.length, units = "cm", lwd = 1, fill = col1, col = col, alpha = 0.66, label.col = "black", cex = 1, fontfamily = "serif", fontface = "bold", margin = 0.03, cat.col = col1, cat.cex = 1, cat.fontfamily = "serif", cat.fontface = "bold", cat.dist = c(0.05, 0.05), cat.just = list(c(NA, NA), c(0.3, 0)) %=>% {
                if (inv) 
                  rev(x)
                else x
            }, cat.pos = cat.pos, ext.length = 0.9, ext.pos = 180, main = main, sub = sub, sub.cex = 0.8, inverted = FALSE, ...))
        }, wj = wj)
        wj
    }
})
wk.heatmap.0orMore <- function(m1, wj_out = "heatmap.pdf", name = "Exp", color.breaks = c("white", "black"), log = 0, row_names_side = "left", column_names_side = "top", ...) {
    hs.require("ComplexHeatmap")
    m1.range <- range(as.vector(m1))
    if (log) 
        m1 <- log(m1 + 1, log)
    colors <- circlize::colorRamp2(c(0, max(as.vector(m1))), color.breaks)
    gray1 <- do.call(rgb, c(as.list(c(132, 133, 135)/255), 0.33))
    pdf.width <- 0.5 * ncol(m1)
    pdf.height <- 0.5 * nrow(m1)
    pdf(wj_out, width = pdf.width, height = pdf.height)
    on.exit(dev.off(), add = TRUE)
    ht <- Heatmap(m1, name = name, row_names_side = row_names_side, column_names_side = column_names_side, cluster_rows = FALSE, cluster_columns = FALSE, col = colors, border = ieiwk.col2rgb("black", 1), rect_gp = gpar(col = gray1, lwd = 1), cluster_column_slices = FALSE, column_title = {
    }, row_title = {
    }, column_gap = unit(0, "mm"), row_gap = unit(0, "mm"), width = grid::unit(0.5 * ncol(m1), "cm"), height = grid::unit(0.5 * nrow(m1), "cm"), cell_fun = function(j, i, x, y, width, height, fill) {
    }, show_heatmap_legend = !log)
    if (log) {
        lgd_spc <- Legend(title = name, col = colors, at = c(0, max(as.vector(m1))), labels = c("0", m1.range[2]))
        draw(ht, annotation_legend_list = list(lgd_spc))
    }
    else draw(ht)
}
text.width.hs <- function(t1, min1 = Inf, cm = 1, scale = 1, ...) {
    if (0) {
        wj <- hs.wj.temp()
        on.exit(hs.wj.rm(wj), add = TRUE)
        pdf(wj)
        on.exit(dev.off(), add = TRUE)
    }
    if (length(list(...))) {
        op <- par(...)
        on.exit(par(op), add = TRUE)
    }
    jg <- t1 %=>% strwidth("inches") %=>% max %=>% min(min1)
    if (cm) 
        jg <- jg * 2.54
    jg * scale
}
wk.gseaplot2 <- function(x, geneSetID, title = "", color = "green", base_size = 11, rel_heights = c(1.5, 0.5, 1), subplots = 1:3, pvalue_table = FALSE, ES_geom = "line", gene2label = {
}) {
    ES_geom <- match.arg(ES_geom, c("line", "dot"))
    geneList <- position <- NULL
    if (length(geneSetID) == 1) {
        gsdata <- enrichplot:::gsInfo(x, geneSetID)
    }
    else {
        gsdata <- do.call(rbind, lapply(geneSetID, enrichplot:::gsInfo, object = x))
    }
    p <- ggplot(gsdata, aes_(x = ~x)) + xlab(NULL) + theme_classic(base_size)
    if (length(gene2label)) {
        xi <- match(gene2label, names(attr(x, "geneList")))
        p <- p + geom_linerange(aes_(ymin = 0, ymax = ~runningScore), color = "black", data = gsdata[xi, ])
    }
    p <- p + theme(panel.grid.major = element_line(colour = "grey92"), panel.grid.minor = element_line(colour = "grey92"), panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank()) + scale_x_continuous(expand = c(0, 0))
    if (ES_geom == "line") {
        es_layer <- geom_line(aes_(y = ~runningScore, color = ~Description), size = 1)
    }
    else {
        es_layer <- geom_point(aes_(y = ~runningScore, color = ~Description), size = 1, data = subset(gsdata, position == 1))
    }
    p.res <- p + es_layer + theme(legend.position = c(0.8, 0.8), legend.title = element_blank(), legend.background = element_rect(fill = "transparent"))
    p.res <- p.res + ylab("Running Enrichment Score") + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.line.x = element_blank(), plot.margin = margin(t = 0.2, r = 0.2, b = 0, l = 0.2, unit = "cm"), axis.title.y = element_text(size = 8))
    i <- 0
    for (term in unique(gsdata$Description)) {
        idx <- which(gsdata$ymin != 0 & gsdata$Description == term)
        gsdata[idx, "ymin"] <- i
        gsdata[idx, "ymax"] <- i + 1
        i <- i + 1
    }
    p2 <- ggplot(gsdata, aes_(x = ~x)) + geom_linerange(aes_(ymin = ~ymin, ymax = ~ymax, color = ~Description)) + xlab(NULL) + ylab(NULL) + theme_classic(base_size) + theme(legend.position = "none", plot.margin = margin(t = -0.1, b = 0, unit = "cm"), axis.ticks = element_blank(), axis.text = element_blank(), axis.line.x = element_blank()) + scale_x_continuous(expand = c(0, 0)) + scale_y_continuous(expand = c(0, 0))
    if (length(geneSetID) == 1) {
        v <- seq(1, sum(gsdata$position), length.out = 9)
        inv <- findInterval(rev(cumsum(gsdata$position)), v)
        if (0) {
            v <- seq(1, 100, length.out = 9)
            inv <- c(rep(100, 100), rep(0, length(gsdata$position) - 100))
            inv <- inv + 1
            xmin <- which(!duplicated(inv))
            xmax <- xmin + as.numeric(table(inv)[unique(inv)])
        }
        if (min(inv) == 0) 
            inv <- inv + 1
        col = c(rev(RColorBrewer::brewer.pal(5, "Blues")), brewer.pal(5, "Reds"))
        ymin <- min(p2$data$ymin)
        yy <- max(p2$data$ymax - p2$data$ymin) * 0.3
        xmin <- which(!duplicated(inv))
        xmax <- xmin + as.numeric(table(inv)[as.character(unique(inv))])
        d <- data.frame(ymin = ymin, ymax = yy, xmin = xmin, xmax = xmax, col = col[unique(inv)])
        p2 <- p2 + geom_rect(aes_(xmin = ~xmin, xmax = ~xmax, ymin = ~ymin, ymax = ~ymax, fill = ~I(col)), data = d, alpha = 0.9, inherit.aes = FALSE)
    }
    df2 <- p$data
    df2$y <- p$data$geneList[df2$x]
    p.pos <- p + geom_segment(data = df2, aes_(x = ~x, xend = ~x, y = ~y, yend = 0), color = "grey")
    p.pos <- p.pos + ylab(quote("Ranked Log"[2] * "FoldChange")) + xlab("Rank in Ordered Dataset") + theme(plot.margin = margin(t = -0.1, r = 0.2, b = 0.2, l = 0.2, unit = "cm"), axis.title.y = element_text(size = 8))
    if (!is.null(title) && !is.na(title) && title != "") 
        p.res <- p.res + ggtitle(title)
    if (length(color) == length(geneSetID)) {
        p.res <- p.res + scale_color_manual(values = color)
        if (length(color) == 1) {
            p.res <- p.res + theme(legend.position = "none")
            p2 <- p2 + scale_color_manual(values = "black")
        }
        else {
            p2 <- p2 + scale_color_manual(values = color)
        }
    }
    if (pvalue_table) {
        pd <- x[geneSetID, c("Description", "pvalue", "p.adjust")]
        pd <- pd[order(pd[, 1], decreasing = FALSE), ]
        rownames(pd) <- pd$Description
        pd <- pd[, -1]
        pd <- round(pd, 4)
        tp <- tableGrob2(pd, p.res)
        p.res <- p.res + theme(legend.position = "none") + annotation_custom(tp, xmin = quantile(p.res$data$x, 0.5), xmax = quantile(p.res$data$x, 0.95), ymin = quantile(p.res$data$runningScore, 0.75), ymax = quantile(p.res$data$runningScore, 0.9))
    }
    plotlist <- list(p.res, p2, p.pos)[subplots]
    n <- length(plotlist)
    plotlist[[n]] <- plotlist[[n]] + theme(axis.line.x = element_line(), axis.ticks.x = element_line(), axis.text.x = element_text())
    if (length(subplots) == 1) 
        return(plotlist[[1]] + theme(plot.margin = margin(t = 0.2, r = 0.2, b = 0.2, l = 0.2, unit = "cm")))
    if (length(rel_heights) > length(subplots)) 
        rel_heights <- rel_heights[subplots]
    cowplot::plot_grid(plotlist = plotlist, ncol = 1, align = "v", rel_heights = rel_heights)
}
wk.pheatmap <- local({
    hs.callback <- function(hc, mat) {
        sv <- svd(t(mat))$v[, 1]
        dend <- reorder(as.dendrogram(hc), wts = sv)
        as.hclust(dend)
    }
    function(M, wj, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100), breaks.boundary = max(abs(M), na.rm = 1), breaks.N = 100, breaks = seq(-breaks.boundary, breaks.boundary, length.out = breaks.N), cluster = 1, cluster_rows = cluster, cluster_cols = cluster, column.anno = NA, row.anno = NA, anno.color = NA, info = FALSE, info.color = "green", fontsize_row = if (show_rownames) 
        fontsize
    else NA, fontsize_col = fontsize, fontsize = 5, show_rownames = nrow(M) < 100, show_colnames = 1, width = NA, height = if (show_rownames) 
        NA
    else 5, ...) {
        hs.require("pheatmap")
        na.i1 <- which(is.na(M))
        if (length(na.i1)) {
            M[na.i1] <- 0
        }
        if (0) {
            method1 <- "spearman"
            method1 <- "pearson"
            row_hclust <- if (cluster_rows) 
                hclust(M %=>% t %=>% cor(method = method1) %=>% {
                  1 - x
                } %=>% as.dist)
            else FALSE
            row_hclust <- 1
            row_hclust <- hclust(dist(M))
        }
        row_hclust <- cluster_rows
        column_hclust <- if (cluster_cols) 
            hclust(M %=>% cor(method = method1) %=>% {
                1 - x
            } %=>% as.dist)
        else FALSE
        if (length(na.i1)) 
            M[na.i1] <- NA
        pheatmap(M, filename = wj, cellwidth = fontsize_col, cellheight = fontsize_row, fontsize = fontsize, color = colorRampPalette(c("navy", "white", "firebrick3"))(51), breaks = breaks, na_col = "gray", border_color = NA, cluster_rows = row_hclust, cluster_cols = column_hclust, clustering_callback = hs.callback, annotation_col = column.anno, annotation_row = row.anno, annotation_colors = anno.color, display_numbers = info, number_color = info.color, show_rownames = as.logical(show_rownames), 
            show_colnames = as.logical(show_colnames), width = width, height = height, ...)
        wj
    }
})
wk.ggsave <- function(fig_list, wj2 = hs.home.expand("~/tmp.pdf"), units = "cm", family = "GB1", env1 = base::parent.frame(), env2 = if (local) base::new.env(parent = env1) else env1, local = 1, redo = 1, ...) {
    if (redo) 
        hs.wj.rm(wj2)
    if (!file.exists(wj2)) {
        if (is.language(fig_list)) 
            fig_list <- base::eval(fig_list, env2)
        if ("list" %not.in% class(fig_list)) {
            fig_list <- list(fig_list)
        }
        args <- list(..., family = family)
        height <- if ("height" %in% names(args)) 
            args[["height"]]
        else NA
        width <- if ("width" %in% names(args)) 
            args[["width"]]
        else NA
        hs.wj(wj2)
        hs.switch(length(fig_list), 1, {
            ggplot2::ggsave(wj2, fig_list[[1]], units = units, family = family, ...)
        }, 0, hs.browser("38.02.16.140027"), {
            wjj_temp <- hs.wjj(hs.wj.temp())
            on.exit(hs.wj.rm(wjj_temp, 1), add = TRUE)
            wj_temp <- sapply(seq_along(fig_list), function(i1) {
                wj2 <- hs.wj(wjj_temp, paste0(i1, ".pdf"))
                args[["width"]] <- width[hs.recycle(i1, length(width))]
                args[["height"]] <- height[hs.recycle(i1, length(height))]
                args <- c(wj2, fig_list[i1], args, units = units)
                do.call(ggplot2::ggsave, args)
                wj2
            })
            hs.hsgly("pdf/合并多个文件.36_11_09_121255")(wj_temp, wj2)
        })
    }
    wj2
}
hs.lab_log <- function(vn = "Exp", logN = 2) {
    bquote(~Log[.(logN)] * .(vn))
}
bg_white.hs <- function(go1) {
    go1 + ggplot2::theme(panel.grid.major = ggplot2::element_blank(), panel.grid.minor = ggplot2::element_blank(), panel.background = ggplot2::element_blank(), axis.line = ggplot2::element_line(colour = "black"))
}
vhline00.hs <- function(go1) {
    go1 + ggplot2::geom_hline(yintercept = 0, color = "grey", linetype = "dashed", lwd = 1) + ggplot2::geom_vline(xintercept = 0, color = "grey", linetype = "dashed", lwd = 1)
}
hline_qv.hs <- function(go) {
    go + ggplot2::geom_hline(yintercept = -log10(0.05), color = "grey", linetype = "dashed", lwd = 1)
}
annotate_text <- function(label, x, y, facets = "", hjust = 0, vjust = 0, color = "black", alpha = NA, family = thm$text$family, size = thm$text$size, fontface = 1, lineheight = 1, box_just = ifelse(c(x, y) < 0.5, 0, 1), margin = unit(size/2, "pt"), thm = ggplot2::theme_get()) {
    x <- scales::squish_infinite(x)
    y <- scales::squish_infinite(y)
    data <- if (is.null(facets)) 
        data.frame(x = NA)
    else data.frame(x = NA, facets)
    tg <- grid::textGrob(label, x = 0, y = 0, hjust = hjust, vjust = vjust, gp = grid::gpar(col = alpha(color, alpha), fontsize = size, fontfamily = family, fontface = fontface, lineheight = lineheight))
    ts <- grid::unit.c(grid::grobWidth(tg), grid::grobHeight(tg))
    vp <- grid::viewport(x = x, y = y, width = ts[1], height = ts[2], just = box_just)
    tg <- grid::editGrob(tg, x = ts[1] * hjust, y = ts[2] * vjust, vp = vp)
    inner <- grid::grobTree(tg, vp = grid::viewport(width = unit(1, "npc") - margin * 2, height = unit(1, "npc") - margin * 2))
    layer(data = NULL, stat = StatIdentity, position = PositionIdentity, geom = GeomCustomAnn, inherit.aes = TRUE, params = list(grob = grid::grobTree(inner), xmin = -Inf, xmax = Inf, ymin = -Inf, ymax = Inf))
}
hs_plot.track.add_text <- local({
    x_unit <- {
    }
    y_unit <- {
    }
    th <- {
    }
    function(text, X, y_range, cex1 = 1, th.f = 1.3, tw.f = 1.1, only = {
    }, ...) {
        x_unit <<- par("usr")[1:2] %=>% diff %=>% {
            x/par("pin")[1]
        }
        y_unit <<- par("usr")[3:4] %=>% diff %=>% {
            x/par("pin")[2]
        }
        th <<- strheight("a", "inches", cex1) * th.f * y_unit
        y1 <- y_range[1] + th/2 * th.f
        Y <- rep(0, length(X))
        occupied <- list()
        for (i1 in seq_along(text)) {
            t1 <- text[i1]
            x1 <- X[i1]
            t1w <- strwidth(t1, "inches", cex1) * tw.f * x_unit
            o1 <- cbind(x1 %=>% {
                c(x, x + t1w)
            }, y1 + th/2 * c(-1, 1))
            i2 <- sapply(occupied, function(o2) {
                is_overlapping.hs.range1(o1[, 1], o2[, 1])
            })
            if (length(i2) && any(i2)) {
                occupied_x <- occupied[i2]
                y_allowed <- sapply(occupied_x, . %=>% x[, 2]) %=>% t %=>% reduced.hs.range_1d %=>% dif.hs.range_1d(c(y1 - th/2 * th.f, y_range[2]), x)
                o1[, 2] <- y_allowed[which(apply(y_allowed, 1, diff) >= th)[1], 1] %=>% c(x, x + th)
            }
            Y[i1] <- mean(o1[, 2])
            occupied[length(occupied) + 1] <- list(o1)
        }
        if (length(only)) 
            return(switch(only, xy = list(X, Y)))
        text(X, Y, text, pos = 4, offset = 0, cex = cex1, ...)
    }
})
hs.with_par <- function(l1, exp1) {
    par.37_11_16_121131 <- do.call(par, l1)
    on.exit(par(par.37_11_16_121131))
    exp1
}
hs.grid.add_text <- function(text = "A", wh_ratio = 1) {
    grid.text(text, 0.05, unit(1, "npc") - unit(1, "lines"), just = 1, gp = gpar(fontsize = 20, fontface = "bold"))
    if (0) {
        hs.with_par(list(mar = c(0, 0, 0, 0), xaxt = "n", yaxt = "n", bty = "n"), {
            pushViewport(viewport(x = 0.01, y = 1, width = 0.05, height = 0.05 * wh_ratio, just = c("left", "top")))
            par(fig = gridFIG())
            plot(c(0, 1), c(0, 1), type = "n")
            text(0.5, 0.5, "A", cex = 2)
            popViewport()
        })
    }
}
hs.grid.print <- function(fig1, ..., pop = 1) {
    vp1 <- viewport(...)
    pushViewport(vp1)
    print(fig1, newpage = FALSE)
    if (pop) 
        popViewport()
}
hs.rowCvs <- function(x) {
    Mean <- hs.rowMeans(x)
    Sd <- hs.rowSds(x)
    Sd/Mean
}
hs.colCvs <- function(x) {
    Mean <- hs.colMeans(x)
    Sd <- hs.colSds(x)
    Sd/Mean
}
hs.num_is_equal <- function(n1, n2 = 0, co1 = 0.01) {
    n1 <- as.numeric(n1)
    n2 <- as.numeric(n2)
    abs(n1 - n2) < co1
}
hs.se <- function(x1) {
    BiocGenerics::sd(x1)/(base::length(x1)^0.5)
}
hs.sd_median <- function(v1) {
    (base::sum((v1 - stats::median(v1))^2)/(base::length(v1) - 1))^0.5
}
hs.mad <- function(v1, side = 1) {
    median1 <- stats::median(v1)
    v2 <- if (side) 
        v1[v1 >= median1]
    else v1[v1 < median1]
    stats::mad(v2, median1)
}
geometric_mean.hs <- function(v1) {
    exp(mean(log(v1)))
}
hs.prcomp <- function(..., scale = 1) {
    stats::prcomp(..., scale = as.logical(scale))
}
z_score.hs.v <- function(v1, log = 0, na.rm = TRUE, log_min = 1) {
    if (log) {
        v1 <- log(v1 + log_min, log)
    }
    a1 <- v1 - mean(v1, na.rm = na.rm)
    a1/sd(a1, na.rm = na.rm)
}
z_score.hs.m <- function(m1, side = 1) {
    if (side != 1) 
        hs.browser("38.05.04.234329")
    t((t(m1) - rep(hs.rowMeans(m1), each = ncol(m1)))/rep(hs.rowSds(m1), each = ncol(m1)))
}
hs.log_mean <- function(v1, base = {
}, base2 = base) {
    hs.exp <- if (length(base)) {
        function(x1) base^x1
    }
    else function(x1) expm1(x1) + 1
    hs.log <- if (length(base2)) {
        switch(paste(base2), `2` = base::log2, `10` = base::log10, function(x1) base::log(x1, base))
    }
    else base::log
    hs.log(base::mean(hs.exp(v1)))
}
hs.lfc <- function(v1, v2, base = 2, pseudocount = 1, loged = TRUE, base0 = {
}) {
    if (loged) {
        hs.log_mean(v1, base0, base) - hs.log_mean(v2, base0, base)
    }
    else {
        m1 <- base::mean(v1) + pseudocount
        m2 <- base::mean(v2) + pseudocount
        hs.log <- switch(paste(base), `2` = base::log2, `10` = base::log10, function(x1) base::log(x1, base))
        hs.log(m1) - hs.log(m2)
    }
}
fisher.test.ele <- function(u, s1, s2, workspace = 2e+05, hybrid = FALSE, control = list(), or = 1, alternative = "two.sided", conf.int = TRUE, conf.level = 0.95, simulate.p.value = FALSE, B = 2000, m.only = 0, verbose = TRUE, no0 = FALSE, u.n, s1.n, s2.n, m = {
}) {
    if (missing(u.n)) 
        u.n <- deparse(substitute(u), width.cutoff = 500L)
    if (missing(s1.n)) 
        s1.n <- deparse(substitute(s1), width.cutoff = 500L)
    if (missing(s2.n)) 
        s2.n <- deparse(substitute(s2), width.cutoff = 500L)
    if (length(m)) {
        if (hs.len1(dim(m), 2)) {
            if (!hs.len1(m, 4)) 
                stop("36_10_09_190618")
        }
        else m <- matrix(head(m, 4), nrow = 2)
    }
    else {
        s1 <- intersect(u, s1)
        s2 <- intersect(u, s2)
        m <- matrix(c(sum(s1 %in% s2), length(u %and% s1 %not% s2), length(u %and% s2 %not% s1), length(u %not% (s1 %or% s2))), 2)
        rownames(m) <- c(s2.n %.% "+", s2.n %.% "-")
        colnames(m) <- c(s1.n %.% "+", s1.n %.% "-")
    }
    if (no0) 
        m[m == 0] <- 1
    if (m.only) 
        return(m)
    a <- fisher.test(m, workspace = workspace, hybrid = hybrid, control = control, or = or, alternative = alternative, conf.int = conf.int, conf.level = conf.level, simulate.p.value = simulate.p.value, B = B)
    m <- cbind(m, rowSums(m))
    m <- rbind(m, colSums(m))
    if (verbose) {
        prt <- capture.output(.s(m))
        substr(prt[substr(prt, 1, 1) == " "], 1, 1) <- "."
        prt[length(prt)] <- paste(prt[length(prt)], " (", u.n, ")", sep = "")
        cat(prt, sep = "\n")
    }
    b <- list(as.vector(round(a$est, 5)), a$p.v)
    names(b) <- c("odds ratio", "p-value")
    if (verbose) 
        cat(ieiwk.s.list.in.row(b), sep = "\n")
    a
}
hs.cos_sim <- function(x, y) {
    if (length(dim(x)) == 2) {
        if ((ncol(x) != 2) && (nrow(x) == 2)) 
            x <- t(x)
        if (ncol(x) != 2) 
            hs.browser("38.08.09.224140")
        y <- x[, 2]
        x <- x[, 1]
    }
    (x %*% y)/(sum(x^2) * sum(y^2))^0.5
}
aov.pv <- function(a) {
    b <- head(summary(a)[[1]][["Pr(>F)"]], -1)
    names(b) <- gsub(" ", "", head(rownames(summary(a)[[1]]), -1))
    b
}
ieiwk.normalize.quantiles <- function(x, copy = 1, co.os = 100) {
    if (!is.matrix(x)) {
        x <- as.matrix(x)
        copy <- 0
    }
    if (as.numeric(sub("\\w+$", "", os("x", units = "Mb"))) > co.os) 
        copy <- 0
    a <- preprocessCore::normalize.quantiles(x, as.logical(copy))
    dimnames(a) <- dimnames(x)
    a
}
hs.p.adjust <- function(p, method = "fdr", n = length(p)) {
    stats::p.adjust(p, method = method, n = n)
}
reduced.hs.range_1d <- function(R) {
    if (is.vector(R)) 
        R <- t(R)
    ro <- order(R[, 1])
    wanted <- rep(TRUE, nrow(R))
    if (nrow(R) > 1) {
        for (i1 in 2:nrow(R)) {
            r1 <- R[ro[i1], ]
            i2 <- max(which(head(wanted, i1 - 1)))
            r2 <- R[ro[i2], ]
            if (r1[1] <= r2[2]) {
                if (r1[2] > r2[2]) 
                  R[ro[i2], 2] <- r1[2]
                wanted[i1] <- FALSE
            }
        }
    }
    R[ro[wanted > 0], , drop = 0]
}
is_overlapping.hs.range1_1d <- local({
    hs1 <- function(r1, r2) r2[1] <= r1[2]
    function(r1, r2) {
        if (r1[1] <= r2[1]) 
            hs1(r1, r2)
        else hs1(r2, r1)
    }
})
is_overlapping.hs.range1 <- function(r1, r2) {
    if (is.vector(r1)) 
        r1 <- matrix(r1)
    if (is.vector(r2)) 
        r2 <- matrix(r2)
    if (ncol(r1) != ncol(r2)) 
        stop("37_01_28_205438")
    for (i1 in 1:ncol(r1)) if (!is_overlapping.hs.range1_1d(r1[, i1], r2[, i1])) 
        return(FALSE)
    TRUE
}
dif.hs.range1_1d <- function(r1, r2) {
    if (length(r1)) {
        if (is_overlapping.hs.range1(as.vector(r1), as.vector(r2))) {
            rbind(if (r1[1] < r2[1]) 
                c(r1[1], r2[1]), if (r1[2] > r2[2]) 
                c(r2[2], r1[2]))
        }
        else r1
    }
}
dif.hs.range_1d <- function(r1, r2) {
    if (length(r1)) {
        if (is.vector(r1)) 
            r1 <- t(r1)
        if (length(r2)) {
            r2 <- reduced.hs.range_1d(r2)
            r2 <- r2[order(r2[, 1]), , drop = 0]
            a1 <- hs.apply(r1, 1, function(r1) {
                r2 <- r2[hs.filter(r2[, 1], function(x1) x1 >= r1[1], r2[, 2], function(x1) x1 <= r1[2]), , drop = 0]
                t(matrix(c(if (r1[1] < r2[1]) c(r1[1], r2[1]), if (nrow(r2) > 1) t(r2)[seq.int(2, length(r2) - 1)], if (r1[2] > r2[length(r2)]) c(r2[length(r2)], r1[2])), 2))
            })
            do.call("rbind", a1)
        }
        else r1
    }
}
cor.wt <- function(d, w = rep(1, nrow(d))/nrow(d), rank = 0, nt = 1) {
    w <- w/sum(w)
    if (rank) {
        d <- matrix(rank(d), nrow(d))
    }
    M <- apply(d, 2, function(d1) {
        sum(d1 * w)
    })
    N <- ncol(d)
    jg <- matrix(1, N, N)
    if (nt > 1) {
        tri <- vector("double", N * (N - 1)/2)
        combination <- combn(N, 2)
        tri <- seq.int(ncol(combination)) %=>% wk.mclapply(x, . %=>% {
            i1 <- combination[1, x]
            i2 <- combination[2, x]
            m1 <- M[i1]
            m2 <- M[i2]
            (sum(d[, i1] * d[, i2] * w) - m1 * m2)/sqrt((sum(d[, i1]^2 * w) - m1^2) * (sum(d[, i2]^2 * w) - m2^2))
        }, mc.cores = nt) %=>% unlist
        hs.browser("37.10.11.061326")
        lower
    }
    else {
        for (i1 in seq.int(N - 1)) {
            m1 <- M[i1]
            for (i2 in seq.int(i1 + 1, N)) {
                m2 <- M[i2]
                c1 <- (sum(d[, i1] * d[, i2] * w) - m1 * m2)/sqrt((sum(d[, i1]^2 * w) - m1^2) * (sum(d[, i2]^2 * w) - m2^2))
                jg[i1, i2] <- jg[i2, i1] <- c1
            }
        }
    }
    rownames(jg) <- colnames(jg) <- colnames(d)
    jg
}
wk.least_common_multiple <- function(...) {
    n1 <- unique(c(...))
    jg <- n1[1]
    if (length(n1) > 1) {
        for (i1 in 2:length(n1)) {
            jg <- pracma::Lcm(jg, n1[i1])
        }
    }
    jg
}
wk.integerRange.4cut <- function(v1, d1 = {
}, sep = "-", sep2 = ",") {
    v1 <- unique(sort(v1))
    if (length(v1) < 2) 
        return(v1)
    if (!length(d1)) 
        d1 <- diff(v1)
    tbl <- list(dif = hs.uniq(d1), first = hs.uniq(d1, "first"), last = hs.uniq(d1, "last")) %=>% wk.dt.as
    tbl <- tbl[dif == 1]
    if (tbl[, .N] == 0) 
        return(paste(v1, collapse = ","))
    tbl[, `:=`("last", last + 1)]
    c(tbl[1, first %=>% {
        if (x > 1) v1[seq.int(x - 1)]
    }], sapply(seq.int(tbl[, .N]), . %=>% {
        a1 <- tbl[x, v1[c(first, last)] %=>% unique %=>% paste(collapse = "-")]
        if (x < tbl[, .N]) {
            i1 <- tbl[x, last + 1]
            i2 <- tbl[x + 1, first - 1]
            if (i1 <= i2) a1 <- c(a1, paste(unique(v1[c(i1, i2)]), collapse = ","))
        }
        a1
    }) %=>% unlist) %=>% paste(collapse = ",")
}
formatSafe <- function(x, scientific = FALSE, trim = TRUE, ...) {
    format(x, scientific = scientific, trim = trim, ...)
}
wk.bigz <- function(n1) {
    if (inherits(n1, "character")) {
        return(gmp::as.bigz(n1))
    }
    if (is.infinite(n1)) 
        hs.browser("39.03.03.095309", debug = 0)
    if (n1 >= 100L) {
        n1 <- ceiling(log(n1, 10))
    }
    gmp::as.bigz(10)^n1
}
hs.median_i <- function(x) which.min(abs(x - median(x)))
hs.is_assignment <- function(e, assigner = c("<-", "<<-", "="), dt = {
}) {
    if (!is.call(e)) 
        return(FALSE)
    if (length(e) < 2) 
        return(FALSE)
    if ((length(e[[1]]) == 1) && (paste(e[[1]]) %in% assigner)) 
        return(TRUE)
    if ((length(e) == 4) && (e[[1]] == quote(`[`)) && (length(e[[4]]) > 1) && (e[[4]][[1]] == quote(`:=`)) && (if (length(dt)) 
        paste(e[[2]]) %in% dt
    else TRUE)) 
        return(TRUE)
    return(FALSE)
}
hs.has_assignment <- function(e1, assigner = eval(formals(hs.is_assignment)[["assigner"]])) {
    if (!is.call(e1)) 
        return(FALSE)
    if (length(e1) < 2) 
        return(FALSE)
    if (hs.is_assignment(e1, assigner = assigner)) 
        return(TRUE)
    if ((length(e1[[1]]) == 1)) {
        if (paste(e1[[1]]) %in% c("local", "apply", "sapply", "lapply", "function")) 
            return()
    }
    return(as.list(e1[-1]) %=>% lapply(hs.has_assignment) %=>% unlist %=>% any)
}
hs.get_assignment <- function(e1, assigner = eval(formals(hs.is_assignment)[["assigner"]])) {
    if (!is.call(e1)) 
        return()
    if (length(e1) < 2) 
        return()
    if (hs.is_assignment(e1, assigner = assigner)) 
        return(e1)
    if ((length(e1[[1]]) == 1)) {
        if (paste(e1[[1]]) %in% c("local", "apply", "sapply", "lapply", "function")) 
            return()
    }
    return(as.list(e1[-1]) %=>% lapply(hs.get_assignment) %=>% unlist)
}
hs.is_block <- function(e) {
    is.call(e) && ((e[[1]] == quote(`for`)) || (e[[1]] == quote(`{`)))
}
hs.get_symbol <- function(e, side = "right", fn = 0) {
    if (is.symbol(e)) 
        return(paste(e))
    if (side == "left") {
        if (e[[1]] == quote(`for`)) 
            return(hs.get_symbol(e[[4]], side = "left"))
        if ((e[[1]] == quote(`<-`)) || (e[[1]] == quote(`<<-`)) || (e[[1]] == quote("[")) || (e[[1]] == quote("[["))) 
            return(hs.get_symbol(e[[2]], side = "left"))
        if (e[[1]] == quote(`$`)) 
            return(paste(e[[2]]))
        if ((e[[1]] == quote(`[`)) && (length(e) == 4) && is.call(e[[4]]) && (e[[4]][[1]] == quote(`:=`))) 
            return(paste(e[[2]]))
        return(lapply(as.list(e[-1]), hs.get_symbol, side = "left") %=>% unlist %=>% unique)
    }
    if (is.call(e)) {
        if (length(e[[1]]) == 1) {
            if (e[[1]] == quote(`$`)) 
                return(hs.get_symbol(e[[2]], fn = fn))
            if (fn && e[[1]] == quote(`::`)) 
                return(lapply(as.list(e[-1]), hs.get_symbol, fn = fn) %=>% unlist %=>% unique)
            if ((e[[1]] == quote(`<-`)) || (e[[1]] == quote(`=`)) || (e[[1]] == quote(`:=`))) 
                return(hs.get_symbol(e[[3]], fn = fn))
        }
        lapply(as.list(if (fn) 
            e
        else e[-1]), hs.get_symbol, fn = fn) %=>% unlist %=>% unique
    }
}
hs.is_assignment_or_block <- function(e) {
    is.call(e) && ((e[[1]] == quote(`<-`)) || (e[[1]] == quote(`=`)) || (e[[1]] == quote(`for`)) || ((e[[1]] == quote(`[`)) && (e[[4]][[1]] == quote(`:=`))))
}
hs.hs_absPath <- local({
    pkg_base <- c("rlang", "magrittr", "tidyselect", "pillar")
    pkg_common <- c("ggplot2", "ggfittext", "cowplot", "RColorBrewer", "dplyr", "tidyr", "tibble", "fst", "data.table", "MASS", "car")
    pkg.fn <- pkg.fn.extra <- {
    }
    function(e, pkg = {
    }) {
        if (!length(pkg.fn)) 
            local({
                pkg.fn <- hs.lapply(pkg_common, . %=>% {
                  pkg_other <- c(pkg_common, pkg_base) %not% x
                  exported <- getNamespaceExports(x)
                  imported <- getNamespaceImports(x)
                  imported <- getNamespaceImports(x) %=>% x[names(x) %in% pkg_other] %=>% unlist
                  exported %not% imported
                }) %=>% wk.dt.fromList(nm = c("pkg", "fn"))
                dt_2keep <- c("as.data.table", "is.data.table", "data.table", "fread", "fwrite", "set", "setDT", "setDF", "setkeyv", "setkey", "setorderv", "setorder", "setindex", "setindexv", "setnames", "setDTthreads", "getDTthreads", "uniqueN", "key", "haskey", "indices", "copy", "melt")
                pkg.fn %<=>% x[!((pkg %in% "data.table") & (fn %not.in% dt_2keep))]
                u <- pkg.fn[fn %=>% {
                  x %not.in% hs.duplicated(x)
                }]
                s <- pkg.fn[fn %not.in% u[, fn]][fn %=>% match(unique(x), x)]
                pkg.fn <<- wk.dt.rbind(u, s)
            })
        pkg.fn.extra <<- if (length(pkg)) {
            hs.lapply(pkg, getNamespaceExports) %=>% wk.dt.fromList(nm = c("pkg", "fn"))
        }
        else list(pkg = "", fn = "") %=>% wk.dt.as(x)[-1]
        if (missing(e)) 
            return()
        if (is.call(e)) {
            if (length(e[[1]]) == 1) {
                if (e[[1]] == quote(`%>%`)) {
                  e[[1]] <- quote(`%=>%`)
                }
                else if ({
                  i <- pkg.fn[, match(paste(e[[1]]), fn, nomatch = 0)]
                }) {
                  e[[1]] <- pkg.fn[i, call("::", as.symbol(pkg), as.symbol(fn))]
                }
                else if ({
                  i <- pkg.fn.extra[, match(paste(e[[1]]), fn, nomatch = 0)]
                }) {
                  e[[1]] <- pkg.fn.extra[i, call("::", as.symbol(pkg), as.symbol(fn))]
                }
            }
            if (length(e) > 1) {
                for (i in 2:length(e)) {
                  e[[i]] <- hs.hs_absPath(e[[i]])
                }
            }
        }
        e
    }
})
hs.pkg_used <- function(e) {
    if (is.call(e)) {
        if ((length(e) > 1) && ((e[[1]] == quote(`::`)) || (e[[1]] == quote(`:::`)))) 
            return(paste(e[[2]]))
        lapply(as.list(e), hs.pkg_used) %=>% unlist
    }
}
"39_08_29_182030" <- function(pkg, add_a_blank_link_at_the_end = 1) {
    bquote({
        if (!("devtools" %in% .packages(TRUE))) 
            install.packages("devtools")
        if (!("util.390816171110" %in% .packages(TRUE))) 
            devtools::install_github("ieiwk/util.390816171110")
        library(util.390816171110)
        local({
            for (pkg_1 in .(pkg)) {
                if (pkg_1 %not.in% .packages(TRUE)) {
                  hs.require(pkg_1)
                }
            }
        })
    }) %=>% as.list(x)[-1] %=>% lapply(deparse) %=>% unlist %=>% 
        cat("# r包", x, if (add_a_blank_link_at_the_end) 
            "", sep = "\n")
}
"2023_08_29_191711" <- function(e, abs = 1, abs_argList = list()) {
    if (is.call(e)) {
        e <- list(e)
    }
    else e <- as.list(e)
    if (abs) {
        hs1 <- hs.hs_absPath
        if (length(abs_argList)) 
            formals(hs1)[names(abs_argList)] <- abs_argList
        hs1()
    }
    lapply(e, . %=>% {
        if (abs) 
            hs1(x)
        else x
    } %=>% hs.pkg_used) %=>% unlist %=>% hs.table %=>% hs.sort %=>% 
        names
}
"39_08_29_190716" <- local({
    hs1 <- function(left, left.right) {
        right <- left.right[left] %=>% unlist %=>% (`%or%`(left))
        if (length(right) > length(left)) {
            right %or% hs1(left %or% right, left.right)
        }
    }
    function(e, e.fp, target_pattern, include_target = 1) {
        if (!missing(e.fp)) 
            e <- parse(e.fp)
        el <- hs.catch(seq_along(e) %=>% for (i in seq_along(x)) {
            hs.msg(i)
            e1 <- e[[i]]
            a <- hs.get_assignment(e1)
            if (length(a)) {
                if (!is.list(a)) 
                  a %<=>% list
                left <- sapply(a, . %=>% hs.get_symbol(x[[2]], 
                  side = "left"))
                if (is.list(left)) 
                  hs.browser("2023.08.24.015157", debug = 0)
                right <- lapply(a, . %=>% as.list(x[-c(1, 2)]) %=>% 
                  lapply(hs.get_symbol, fn = 1) %=>% unlist %=>% 
                  hs.clean)
                hs.throw(i, hs.c(left, right) %=>% hs.list.merge(name = 0) %=>% 
                  list)
            }
            else hs.throw(i, list())
        }, named = 1)
        i <- hs.catch(seq_along(e) %=>% for (i in x) {
            s <- e[[i]] %=>% deparse(width.cutoff = 500)
            if (any(hs.grepl(s, target_pattern))) {
                hs.throw(i)
            }
        }) %=>% unlist
        if (length(i) != 1) 
            hs.browser("2023.08.29.190939", debug = 0)
        i1 <- i[1]
        node <- seq(i1) %=>% lapply(x, . %=>% el[[x]] %=>% if (length(x)) 
            names(x[[1]])) %=>% unlist %=>% unique
        l.r <- el %=>% head(i1) %=>% lapply(x, . %=>% {
            if (length(x)) {
                for (j in seq_along(x[[1]])) x[[1]][[j]] %<=>% 
                  (`%and%`(node))
            }
            x
        }) %=>% hs.catch(lapply(x, . %=>% {
            if (length(x)) {
                x <- x[[1]]
                for (i in seq_along(x)) hs.throw(names(x)[i], 
                  x[[i]] %=>% (`%and%`(node)))
            }
        }), named = 1)
        l.r %<=>% hs.list.merge(name = 0)
        left <- hs.get_symbol(e[[i1]], fn = 1) %and% node
        right <- hs1(left, l.r) %=>% unique
        e_involved <- hs.catch(seq(i1) %=>% for (i in x) {
            e1 <- e[[i]]
            a <- hs.get_assignment(e1)
            if (length(a)) {
                if (!is.list(a)) 
                  a %<=>% list
                left <- sapply(a, . %=>% hs.get_symbol(x[[2]], 
                  side = "left"))
                if (is.list(left)) 
                  hs.browser("2023.08.24.015157", debug = 0)
                if (left %=>% (`%in%`(right)) %=>% sum) {
                  hs.throw(i, e1)
                }
            }
        }, named = 1)
        if (include_target && (i1 %not.in% names(e_involved))) 
            e_involved[[paste(i1)]] <- e[[i1]]
        e_involved
    }
})
"39_08_29_205824" <- function(e.fp, wj2 = "~/a.r.gz", e_involved, 
    pkg = {
    }, ow = 0) {
    if (file.exists(wj2) && (!ow)) 
        stop("[39_08_29_210757|文件已存在]", wj2)
    if (!missing(e.fp)) 
        e <- parse(e.fp)
    if (names(e_involved) %=>% {
        length(x) && (x %all.in% seq_along(e))
    }) {
        pkg_used <- hs.hsgly("代码操作/r/包/用到的/2023_08_29_191711")(e_involved)
        {
            a <- readLines(e.fp)
            hs.hs_absPath(pkg = pkg)
            on.exit(hs.hs_absPath(), add = TRUE)
            hs.with_sink(wj2, {
                hs.hsgly("代码操作/r/包/确保已安装/39_08_29_182030")(pkg = pkg_used)
                i1 <- i2 <- 1
                ei <- 0
                ei_max <- e_involved %=>% names %=>% as.integer %=>% 
                  max
                while (i2 <= length(a)) {
                  if (ei > ei_max) 
                    break
                  code1 <- a[seq(i1, i2)]
                  e1 <- .try(parse(text = code1))
                  if (input.is.incomplete(e1)) {
                    i2 %<=>% +1
                    next
                  }
                  i1 <- i2 <- i2 + 1
                  if (length(e1)) {
                    ei %<=>% +1
                    if (!(ei %in% names(e_involved))) {
                      hs.catn(hs.msg(ei, ot = "s"))
                      next
                    }
                    code2 <- e_involved[[paste(ei)]] %=>% hs.hs_absPath %=>% 
                      deparse(control = "all")
                    comment_end <- code1 %=>% tail(1) %=>% hs.re(" *#.*$")
                    if (length(comment_end)) {
                      if (hs.sub(comment_end, "[^#]*#+") %=>% 
                        hs.grepl("#")) 
                        hs.browser("2023.08.25.205004", debug = 0)
                      code2 %<=>% paste0(comment_end)
                    }
                    hs.catn(code2)
                  }
                  else hs.cat(code1)
                }
            })
        }
        wj2
    }
    else hs.browser("2023.08.29.210057", debug = 0)
}
"37_12_08_170744" <- local({
    conda_wjj <- c("~/miniconda3", "~/rj/miniconda3", "/opt/biosoft/anaconda3") %=>% 
        x[file.exists(x)][1]
    bin_wjj <- hs.home.expand("~/.local/bin")
    in_cran <- c("factoextra", "ggpubr", "readxl", "tidyverse", 
        "survival", "gt", "broom", "gtsummary", "rsample", "labelled", 
        "ggextra", "rjson", "stringdist", "venndiagram", "pheatmap", 
        "iotools", "ade4", "ggdendro", "ggvenn", "ggVennDiagram", 
        "sf") %=>% tolower
    in_bioconductor <- c("clusterprofiler", "org.hs.eg.db", "enrichplot", 
        "ggtree")
    in_bioconda <- c("sambamba", "velocyto.r", "dropest")
    others <- c("synapseclient", "awscli")
    better_using_install_packages <- "GenomeInfoDbData"
    function(pkg = {
    }, r_pkg = {
    }, conda_env = "wk", bin = pkg, cmd = "install") {
        if (!length(conda_wjj)) 
            hs.browser("37.12.08.171804")
        bin_wjj %=>% hs.wjj
        conda <- hs.fp(conda_wjj, "bin/conda")
        if (!dir.exists(hs.fp(conda_wjj, "envs", conda_env))) 
            .wk(conda, "create -n", conda_env)
        switch(cmd, install = {
            {
                if (!length(r_pkg)) {
                  installed <- hs.fp(conda_wjj, "envs", conda_env, 
                    "lib/R/library") %=>% hs.wjj.ls(wjj = 1, 
                    F = 0) %=>% tolower
                  r_pkg <- c(in_cran, in_bioconductor) %not% 
                    installed
                }
                if (length(r_pkg)) r_pkg %<=>% tolower
                if (!length(pkg)) {
                  if (length(r_pkg)) {
                    pkg <- rep("", length(r_pkg))
                    i <- r_pkg %in% c(in_cran, in_bioconda)
                    if (any(i)) pkg <- r_pkg[i] %=>% paste0("r-", 
                      x)
                    i <- r_pkg %in% in_bioconductor
                    if (any(i)) pkg <- r_pkg[i] %=>% paste0("bioconductor-", 
                      x)
                    if ("" %in% pkg) {
                      r_pkg[pkg %in% ""]
                    }
                  }
                }
            }
            if (!length(pkg)) pkg <- in_bioconda %or% others
            .wk(conda, cmd, conda_env %=>% if (length(x) && nzchar(x)) c("-n", 
                x), "-c conda-forge", "-c bioconda", pkg)
        }, update = {
            .wk(conda, "update -n base -c conda-forge conda")
        }, hs.browser("37.12.08.172029"))
    }
})
"38_10_08_222950" <- function(..mem = 1) {
    .wk("conda info", intern = 1) %>% hs.grepV("envs directories") %>% 
        strsplit(":") %>% {
        .[[1]][2] %>% trim %>% dirname
    }
}
"2023_07_06_121233" <- function(wjj = ".", lib = .libPaths()[1]) {
    hs.with_wd(wjj, .wk("R --vanilla CMD INSTALL -l", hs.fpGood(lib), 
        "./"))
}
"38_11_30_134231" <- function(wj1) {
    wk.fread(text = .sys.cat.text(wj1) %>% .wk("| head", intern = 1))
}
"36_07_24_094159" <- function(wj, sj = {
}, ri = {
}, ci = {
}, cn = {
}, ri.cn, ri.v, name = "subset", jg_wj = hs.wjm(wj, append = paste0(".", 
    name), ext = "tsv.gz", dn = hs.wjj(dirname(wj), "jg")), extN = hs.hsgly("后缀.36.07.24.100015")(wj, 
    "n"), ow = 0) {
    if (ow) 
        hs.wj.rm(jg_wj)
    if (!file.exists(jg_wj)) {
        if (is.null(sj)) 
            sj <- wk.fread(wj)
        if (!length(ri)) {
            if (length(ri.cn) == 1) 
                ri.v <- list(ri.v)
            ri <- lapply(seq_along(ri.cn), function(ci1) {
                cn1 <- ri.cn[ci1]
                v1 <- ri.v[[ci1]]
                wk.dt.subset(sj, v1, cn1, out = "i")
            })
            ri <- unique(unlist(ri))
        }
        if (!length(ci)) {
            ci <- if (length(cn)) {
                names(sj) %>% {
                  i1 <- which(. %in% cn)
                  unlist(tapply(i1, .[i1], list)[cn])
                }
            }
            else seq.int(length(sj))
        }
        wk.dt.fwrite(sj[ri, .SD, .SDcols = c(ci)], jg_wj)
    }
    jg_wj
}
"38_11_29_171547" <- function(dt1) {
    wk.dt.setnames(dt1, unlist(dt1[1]))
    dt1[-1]
}
"39_05_14_113348" <- function(n, an = c(letters, LETTERS, seq(0, 
    9)), b = {
}, n2 = min(length(b), 5)) {
    n1 <- n - n2
    cat(paste(hs.sample(c(hs.sample(an, n1), hs.sample(b, n2))), 
        collapse = ""), sep = "\n")
}
"38_12_17_000138" <- function(n = 1, ssh = "fwq.tsj", ...) {
    hs.hsgly("rsync到台式机.38_12_16_235921")(n = n, ssh = ssh, 
        fx = "in", ...)
}
"38_12_16_235921" <- function(n = 1, ssh = "qg", fx = "out", 
    in2 = 0, wjj2do = c("~/wk", "~/db", if (fx == "out") "~/d/rjk"), 
    wjj2 = "/media/kai/adata2t/db/备份/用Rsync/wk/同步的", 
    wjj_bf = {
    }) {
    if (length(hs.clean(ssh)) && hs.grepl(ssh, "mbp\\d+")) 
        wjj2do %<=>% {
            c(x, c("~/Library/Containers/com.tencent.xinWeChat", 
                "~/Library/Containers/com.apple.Safari")) %=>% 
                x[file.exists(x)]
        }
    wjj2do %<=>% hs.sub("/*$", "/")
    if (0) {
        wjj_special <- if (Sys.info()["nodename"] == "calvin_mpb") 
            c("~/wk/db/书", "~/wk/wk/wd/小说", "~/wk/wk/wd/音乐", 
                "~/wk/wk/wd/图片", "~/wk/rjk/tb", "~/wk/wk/wd/db/mf/ff/ztr")
        wjj_special %=>% if (length(x) && any(hs.grepl(x, " "))) 
            hs.browser("39.04.20.222618", debug = 0)
    }
    additional <- c(if (n) "-n", "-f", .q("- .DS_Store"), "-kK", 
        "--safe-links")
    hs1 <- hs.cond(!length(hs.clean(ssh)), wk.rsync, fx == "out", 
        wk.rsync.out.same_place, fx == "in", wk.rsync.in.same_place)
    if (fx == "out") {
        if (!n) 
            local({
                wj1 <- "~/org/dm/r/org.r.gz"
                wj2 <- "~/org/dm/r/org1.r.gz"
                hs.wj.move(wj1, wj2)
                on.exit(hs.wj.move(wj2, wj1))
                .update()
            })
    }
    if (length(hs.clean(ssh))) {
        for (wjj1 in wjj2do) {
            wjj1.bf <- {
            }
            if (wjj1 %in% "~/wk/") {
                wjj1.bf <- wjj1 %=>% hs.sub("/*$", ".备份")
                hs1(paste0(wjj1, "rjk/"), ssh = ssh, U = 0, del = 1, 
                  additional = additional)
                hs1(wjj1, ssh = ssh, U = 0, del = 1, additional = additional, 
                  bf_wjj0 = wjj1.bf)
            }
            else hs1(wjj1, ssh = ssh, U = 0, del = 1, additional = additional, 
                bf_wjj0 = wjj1.bf)
        }
    }
    else {
        hs.browser("2023.07.14.195516", debug = 0)
        if (length(wjj2do) != 1) 
            hs.browser("39.05.17.094537", debug = 0)
        if (!dir.exists(wjj_bf)) 
            hs.browser("39.05.17.092431", debug = 0)
        wk.rsync(wjj2do, wjj2 %=>% hs.sub("/*$", "/"), ssh = ssh, 
            U = 0, del = 1, additional = additional, bf_wjj0 = wjj_bf)
    }
}
"38_01_27_154027" <- function(wjj1, wjj2 = ".", ext = "pdf,png,xls,xlsx,tsv.gz,org,htm,html", 
    del = 0, L = 1, n = 1, ssh = {
    }) {
    a <- tolower(ext) %>% strsplit(",") %>% unlist %>% unique %>% 
        strsplit("") %>% sapply(. %>% {
        a <- sapply(., . %>% {
            a <- . %or% toupper(.)
            if (length(a) > 1) 
                a <- c("[", a, "]")
            paste(a, collapse = "")
        })
        paste(c("+ *.", a), collapse = "")
    })
    cmd1 <- .sys(wj_rsync, "-ahPvz", if (n) 
        "-n", if (L) 
        "-L", if (del) 
        "--del --delete", "-e ssh", "-f", .q("- .DS_Store"), 
        paste("-f", .q(a)), "-f", .q("+ */"), "-f", .q("- **"), 
        .sys.fpGood(wjj1), wjj2, "| grep -v", .q("/$"), "| grep -v", 
        .q("^cannot delete non-empty directory"))
    message(cmd1)
    .wk(cmd1)
}
"39_05_07_212034" <- function(wjj2 = ".", remote = "r820") {
    if (0) {
        wk.fwrite(wj, wk.rsync.wj.wj)
    }
    wk.rsync.in.same_place(wk.rsync.wj.wj, ssh = remote)
    wj <- wk.rsync.wj.wj %=>% readLines
}
"38_10_20_103018" <- function(fp1, data, ...) {
    hs.switch(c(missing(data), hs.grepl(fp1, "(?i)\\.xlsx?$")), 
        c(1, 0), wk.fread(fp1, ...), c(1, 1), wk.xls.read(fp1, 
            ...), c(0, 0), wk.dt.fwrite(data, fp1, ...), c(0, 
            1), wk.xls.write(data, fp1, ...))
}
"38_12_12_173550" <- function(M, wj2) {
    hs.wj(wj2)
    gz <- hs.grepl(wj2, "(?i)\\.gz$")
    if (gz) {
        fifo <- hs.wj.temp()
        on.exit(hs.wj.rm(fifo), add = TRUE)
        .wk("mkfifo", .sys.fpGood(fifo))
        .wk("cat", .sys.fpGood(fifo), "| gzip >", .sys.fpGood(wj2), 
            "&")
        Matrix::writeMM(M, fifo)
    }
    else Matrix::writeMM(M, wj2)
    wj2
}
"38_10_09_060215" <- function(wj = {
}, wjj) {
    if ((!missing(wjj)) && dir.exists(wjj)) 
        wj <- hs.wjj.ls(wjj, "\\.tsv\\.gz$", R = 1)
    for (i1 in seq_along(wj)) {
        wj1 <- wj[i1]
        wj2 <- hs.wjm(wj1, ext = "xls", extN = 2)
        if (wj.update.reporter(wj2, wj1)) {
            a <- readLines(wj1, 10)
            wk.dt.fwrite(a, wj2)
        }
    }
}
"38_09_10_214239" <- function(v1 = {
}, db.fp, ci1, ci2 = {
}, head = 0, once = 0, consecutive = 0, ic = 0, wj2 = {
}, cn = 0, what = "", ...) {
    cmd1 <- if (length(v1)) {
        paste0("while( <>) {\n  chomp;\n  $v{ ", if (ic) 
            "lc( $_)"
        else "$_", "} = 1;\n}\nopen( F1, ", .q(.sys.cat.skip(db.fp, 
            cn = 0) %=>% paste("|")), ");", if (head) 
            "\n$a = <F1>;\nprint( $a);", "\nwhile( <F1>) {", 
            hs.switch(ci1, 1, "\n  /^[^\t]*/;\n  $f_ci1 = $&;", 
                2, "\n  /\t([^\t]*)/;\n  $f_ci1 = $1;", .sys("\n  chomp;\n  @F = split( \"\t\");\n  $f_ci1 = $F[ ", 
                  ci1 - 1, "];\n")), if (ic) 
                "\n  $f_ci1 = lc( $f_ci1);", "\n  if( exists( $v{ $f_ci1})) {", 
            if ((length(v1) == 1) && consecutive) 
                "\n    $found = 1;", if (length(ci2)) {
                paste0("\n    print( ", paste(paste0("$F[ ", 
                  ci2 - 1, "]"), collapse = ", \"\t\", "), ", \"\n\");")
            }
            else paste0("print( $_", if (ci1 > 2) 
                ", \"\n\"", ");"), if (once) 
                "\n    delete( $v{ $f_ci1});\n    if( %v == 0) {\n      close( F1);\n      exit;\n    }", 
            "\n  }", if ((length(v1) == 1) && consecutive) 
                " else {\n    if( $found) {\n      close( F1);\n      exit();\n    }\n  }", 
            "\n}\nclose( F1);")
    }
    else paste0("\nopen( F1, ", .q(.sys.cat.skip(db.fp, cn = 0) %=>% 
        paste("|")), ");", if (length(ci2)) {
        paste0("\n$\" = \"\t\";\nwhile( <F1>) {\n  chomp;\n  @F = split( \"\t\");\n  ", 
            paste0("print( \"@F[(", paste(ci2 - 1, collapse = ", "), 
                ")]\n\");"), "\n}")
    }
    else "while( <F1>) { print;}")
    if (what == "cmd") 
        return(cmd1)
    message(cmd1)
    hs.with_wd(db.fp, hs.hsgly("1.信息学/文件/读写/向量到perl到结果.38_09_26_203407")(v1, 
        cmd1, wj2 = wj2, cn = cn, ...))
}
"38_09_26_203407" <- function(x, perl.cmd, wj2 = {
}, reader = if (!length(wj2)) wk.fread, what = "", ...) {
    force(reader)
    if (length(x)) {
        if (!length(wj2)) {
            wj2 <- hs.wj.temp() %=>% paste0(".gz")
            on.exit(hs.wj.rm(wj2), add = TRUE)
        }
        hs.msg(" ")
        message(perl.cmd)
        cat(x, sep = "\n", file = .sys("| perl -e", .q(perl.cmd), 
            if (hs.grepl(wj2, "(?i)\\.gz$")) 
                "| gzip", ">", .sys.fpGood(hs.wj(wj2))))
        .sys("| perl -e", .q(perl.cmd), if (hs.grepl(wj2, "(?i)\\.gz$")) 
            "| gzip", ">", .sys.fpGood(hs.wj(wj2))) %=>% message
        hs.msg()
        if (length(reader)) 
            reader(wj2, ...)
        else wj2
    }
    else {
        cmd <- .sys("perl -e", .q(perl.cmd))
        if (length(reader)) {
            if (identical(reader, wk.fread)) {
                wk.fread(cmd = cmd, ...)
            }
            else hs.browser("39.03.02.233503", debug = 0)
        }
        else {
            .wk(cmd, if (hs.grepl(wj2, "(?i)\\.gz$")) 
                "| gzip", ">", .sys.fpGood(hs.wj(wj2)))
            wj2
        }
    }
}
"38_12_07_172444" <- function(h5fp, i = {
}) {
    if (is(h5fp, "H5D")) 
        return(if (length(i)) h5fp[i] else h5fp[])
    if (is(h5fp, "H5Group")) {
        if (!setequal(names(h5fp), c("categories", "codes"))) 
            hs.browser("38.11.18.112335")
        jg <- factor(if (length(i)) 
            h5fp[["codes"]][i]
        else h5fp[["codes"]][])
        na_i <- match("-1", levels(jg), nomatch = 0)
        if (na_i) 
            levels(jg)[na_i] <- NA
        non_na_i <- which(!is.na(levels(jg)))
        if (length(non_na_i)) 
            levels(jg)[non_na_i] <- h5fp[["categories"]][as.numeric(levels(jg)[non_na_i]) + 
                1]
        return(jg)
    }
    hs.browser("38.11.18.112139", debug = 0)
}
"36_11_09_121255" <- function(wj_in, wj_out = "~/tmp/tmp.pdf") {
    .wk("pdftk", .q(normalizePath(wj_in)), "cat output", .q(normalizePath(wj_out)))
}
"36_11_04_211749" <- function(wj) {
    tmp_wj <- hs.wj.temp()
    on.exit(hs.wj.rm(tmp_wj), add = TRUE)
    .wk("pdftk", .sys.fpGood(wj), "cat 2-end output", tmp_wj)
    hs.wj.move(tmp_wj, wj)
}
"39_01_27_222335" <- local({
    createFonts <- function(wb) {
        list(data = xlsx::Font(wb, heightInPoints = 11, name = "Arial"), 
            title = xlsx::Font(wb, heightInPoints = 16, name = "Arial", 
                isBold = TRUE), subtitle = xlsx::Font(wb, heightInPoints = 13, 
                name = "Arial", isBold = FALSE, isItalic = TRUE))
    }
    createFill <- function(color) {
        xlsx::Fill(color, color, "SOLID_FOREGROUND")
    }
    flag <- 1
    align <- dataFormatDate <- dataFormatNumberD <- {
    }
    function(tbl, cell2fill = {
    }, sheetName = "sheet 1", wj2 = "~/tmp/a.xlsx", fill = "#cc0000") {
        if (flag) {
            align <<- hs.lapply("left,right,center", . %=>% {
                xlsx::Alignment(horizontal = paste0("ALIGN_", 
                  toupper(x)), vertical = "VERTICAL_CENTER", 
                  wrapText = TRUE)
            })
            dataFormatDate <<- xlsx::DataFormat("yyyy/m/d")
            dataFormatNumberD <<- xlsx::DataFormat("0.0")
            flag <<- 0
        }
        wb <- xlsx::createWorkbook()
        sh <- xlsx::createSheet(wb, sheetName)
        f <- createFonts(wb)
        cslist <- lapply(1:ncol(tbl), function(x) {
            xlsx::CellStyle(wb) + f$data + alignRight + dataFormatNumberD
        })
        ri <- 1
        ci <- 1
        Fill1 <- createFill(fill)
        (xlsx::addDataFrame)(tbl, sh, col.names = TRUE, row.names = FALSE, 
            startRow = ri, startColumn = ci, colStyle = hs.c(1:ncol(tbl), 
                cslist))
        if (length(cell2fill)) {
            apply(cell2fill, 1, . %=>% {
                ri <- x[1] + 1
                ci <- x[2]
                (xlsx::addDataFrame)(tbl[ri, ci], sh, col.names = FALSE, 
                  row.names = FALSE, startRow = ri, startColumn = ci, 
                  colStyle = hs.c(1, lapply(cslist, . %=>% {
                    x + Fill1
                  })), )
                invisible()
            })
        }
        if (length(wj2)) {
            xlsx::saveWorkbook(wb, wj2)
            wj2
        }
        else wb
    }
})
"37_04_29_185838" <- function(dz0, wjj2 = hs.fp(wjj0, hs.re(dz0, 
    "ftp\\..*")), what = {
}, wjj0, co1 = wj.update.reporter.co1) {
    co1 <- wj.update.reporter.co1
    if (length(what)) {
        dz1 <- hs.fp(dz0, what)
        wj2 <- hs.fp(wjj2, what)
        if (!hs.wj.exists(wj2, size = 9)) 
            hs.wj.rm(wj2)
        wj.updater(wj2, dz1, co1 = co1)
    }
    else {
        wj2 <- hs.fp(wjj2, "ftp_files.37_04_29_165233.txt.gz")
        if (wj.update.reporter(wj2)) {
            a1 <- readLines(paste0("http://", dz0))
            ftp_files <- hs.re(a1, "(?<=^<a href=\")[^\"]*")
            wk.dt.fwrite(list(ftp_files), wj2)
        }
        note.update_time(wj2)
        message("[37_04_29_165055]files in ", dz0, ":")
        readLines(wj2)
    }
}
"2023_06_26_120607" <- function(dz, wjj2 = ".", wjm = {
}, j = 5, gz = 1, win = 0) {
    if (hs.grepl(dz, "^(?i)ftp://")) 
        dz %<=>% hs.re("(?i)ftp", "https")
    if (win) {
        .sys(.qq("D:\\db\\软件\\exe\\aria2-1.36.0-win-64bit-build1\\aria2c.exe"), 
            "--all-proxy=127.0.0.1:7890", "--file-allocation=none", 
            "-c -R", dz) %=>% message
        return()
    }
    if (length(wjj2) != 1) 
        hs.browser("2023.06.26.120829", debug = 0)
    if (gz) {
        wjj2_temp <- hs.wj.temp()
        on.exit(hs.wj.rm(wjj2_temp, recursive = 1), add = TRUE)
        wjj2_temp %=>% hs.wjj %=>% hs.with_wd({
            for (dz1 in dz) .wk("aria2c --file-allocation=falloc -c -R -j", 
                j, dz1)
        })
        hs.hsgly("文件/找/压缩/值得压缩的.2023_08_10_015150") %=>% 
            x(wjj2_temp) %=>% for (wj1 in x) hs.with_wd(wj1, 
            .wk("gzip", .q(basename(wj1))))
        wjj2_temp %=>% .ls
        .wk("mv", hs.fp(hs.fpGood(wjj2_temp), "*"), hs.fpGood(wjj2))
    }
    else wjj2 %=>% hs.wjj %=>% hs.with_wd({
        for (i in seq_along(dz)) .wk("aria2c --file-allocation=falloc -c -R -j", 
            j, dz[i])
    })
}
"38_10_04_161056" <- function(dz1, wj2, wjj2, gz = !hs.grepl(wj2, 
    "(?i)\\.gz$")) {
    if (missing(wj2)) 
        wj2 <- basename(dz1)
    if (gz) 
        wj2 %<=>% paste0(".gz")
    if (!missing(wjj2)) 
        wj2 <- hs.fp(wjj2, wj2)
    if (!file.exists(wj2)) 
        .wk("wget -O -", .q(dz1), if (gz) 
            "| gzip", ">", .sys.fpGood(wj2))
    normalizePath(wj2)
}
"38_07_07_093129" <- function(wjj1, wjj_all = {
}, wj_all = {
}, wj2 = {
}) {
    a <- hs.with_wd(wjj1, {
        if (length(wj2)) {
            if (inherits(wj2, "function")) {
                wj2 <- wj2()
            }
        }
        else wj2 <- hs.fp("..", .q(paste0(basename(wjj1), ".zip")))
        hs.update(wj2, paste0(x, ".zip"), !hs.grepl(x, "(?i)\\.zip$"))
        cat.file <- .sys("| zip -ru -@", wj2)
        if (inherits(wjj_all, "function")) 
            wjj_all <- wjj_all()
        for (a in wjj_all) if (file.exists(a)) 
            hs.wjj.ls(a, R = 1) %>% cat(sep = "\n", file = cat.file)
        if (length(wj_all)) {
            cat(wj_all, sep = "\n", file = cat.file)
        }
        else hs.for(a, hs.wjj.ls(wjj = 1, F = 0) %not% wjj_all, 
            hs.wjj.ls(a, "(?i)\\.(png|pdf|xls|xlsx|tsv\\.gz)$", 
                R = 1) %=>% cat(sep = "\n", file = cat.file))
        .wk("realpath", wj2, intern = 1)
    })
    message(a)
}
"2023_07_26_150436" <- function(root, sub) {
    sub %=>% hs.hsgly("文件夹/打包/38_07_07_093129")(root, 
        wj_all = hs.wjj.ls(x, wj = 1, wjj = 1), wj2 = hs.gsub(x, 
            "/", "."))
}
"38_11_27_005936" <- function(uid1) {
    hs.hsgly()
    uid1 <- hs.re(basename(uid1), wk.muid.pattern)
    wjj0 <- hs.hsgly(uid1, PO = 1) %>% dirname %>% normalizePath
    wjj1 <- wjj0
    repeat {
        wj1 <- hs.wjj.ls(wjj1, paste0("(?i)文件夹\\.", wk.muid.pattern, 
            "\\.r(.gz)?$"))
        if (length(wj1) == 1) 
            break
        if (length(wj1) > 1) 
            stop("[38_11_27_001548]")
        wjj1 <- dirname(wjj1)
    }
    wjj2 <- wjj0_jg <- hs.hsgly(hs.re(basename(wj1), wk.muid.pattern))()
    if (nchar(wjj0) > nchar(wjj1)) {
        wjj2 <- paste0(wjj0_jg, substring(wjj0, nchar(wjj1) + 
            1))
    }
    wj_flag <- hs.fp(wjj2, paste0("代码号_", uid1))
    if ((!hs.machine.isLocal()) && (!file.exists(wj_flag))) 
        local({
            a <- hs.wjj.ls(wjj0_jg, uid1, R = 1)
            hs.switch(length(a), 1, {
                wjj_old <- dirname(a)
                if (dir.exists(wjj2)) {
                  hs.browser("38.11.27.233122", debug = 0)
                  for (wj1 in hs.wjj.ls(wjj_old, A = 1, wjj = 1, 
                    wj = 1)) hs.wj.move(wj1, wjj2)
                  hs.wj.rm(wjj_old, 1)
                }
                else hs.wj.move(wjj_old, hs.wj(wjj2))
                if (dir.exists(wjj_old)) 
                  hs.browser("38.11.27.144937", debug = 0)
            }, 0, {
                cat("", file = hs.wj(wj_flag))
            }, hs.browser("38.11.27.083724", debug = 0))
        })
    wjj2
}
"2023_07_11_095839" <- function(wjj1, wjj2) {
    wj <- hs.wjj.ls(wjj1, A = 1)
    wj2 <- hs.fp(wjj2, basename(wj))
    i <- !file.exists(wj2)
    wj <- wj[i]
    wj2 <- wj2[i]
    if (!file.rename(wj, wj2)[1]) {
        file.copy(wj, wjj2, copy.date = TRUE)
    }
}
"2023_07_11_101608" <- function(wjj1, wjj2) {
    wj <- hs.wjj.ls(wjj1, wjj = 1, A = 1)
    hs.browser("2023.07.11.101800", debug = 0)
    wj2 <- hs.fp(wjj2, basename(wj))
    i <- !file.exists(wj2)
    wj <- wj[i]
    file.copy(wj, wjj2, recursive = TRUE, copy.date = TRUE)
}
"2023_07_11_102948" <- function(wjj1, exclude = {
}) {
    wj <- hs.wjj.ls(wjj1, A = 1, wj = 1, wjj = 1, F = 0)
    if (length(exclude)) 
        wj %<=>% x[x %not.in% exclude]
    a <- hs.with_wd(wjj1, {
        a <- lapply(seq_along(wj), function(i) {
            wj1 <- wj[i]
            if (dir.exists(wj1)) {
                a <- hs.wjj.ls(wj1, A = 1, R = 1)
                if (length(a)) {
                  a
                }
                else {
                  wj1
                }
            }
            else wj1
        })
    })
    hs.fp(wjj1, unlist(a))
}
"2023_07_11_101012" <- function(wjj1, wjj2) {
    hs.browser("2023.07.11.101116", debug = 0)
    wj <- hs.wjj.ls(wjj1, wjj = 1, wj = 0)
    i <- !file.exists(hs.fp(wjj2, basename(wj)))
    wj <- wj[i]
    file.symlink(wj, wjj2)
}
"2023_06_21_152354" <- function(wjj) {
    hs.with_wd(wjj, {
        while (1) {
            wj <- .wk("find -type d -empty", intern = 1)
            if (!length(wj)) 
                break
            hs.wj.rm(wj, 1)
        }
    })
}
"2023_07_11_094726" <- function(wjj1, wjj2) {
    wj <- hs.wjj.ls(wjj1, wjj = 1, wj = 0)
    for (i in seq_along(wj)) {
        wj1 <- wj[i]
        wj2 <- hs.wjm(wj1, dn = wjj2)
        if (file.exists(wj2)) 
            next
        if (!file.rename(wj1, wj2)) {
            wk.rsync(wj1, wjj2 %=>% hs.sub("/*$", "/"))
        }
        if (file.exists(wj1)) 
            hs.wj.rm(wj1, 1)
    }
}
"37_12_08_111112" <- local({
    hs1 <- function(gz1) {
        wj2 <- paste0(gz1, ".is_ok")
        if (wj.update.reporter(wj2, gz1)) {
            a1 <- .wk("gzip -t 2>&1", gz1, intern = 1)
            cat(length(a1) == 0, file = wj2)
        }
        as.logical(readLines(wj2))
    }
    function(gz) {
        wk.mclapply(gz, hs1, mc.cores = 4) %>% unlist
    }
})
"38_05_13_124417" <- function(wj) {
    if (dir.exists(wj)) 
        wj <- hs.wjj.ls(wj, "(?i)(md5(sum)?\\.txt|\\.md5)$", 
            R = 1)
    wk.mclapply(wj, . %=>% {
        hs.with_wd(x, readLines(x) %=>% cat(sep = "\n", file = .sys("| md5sum -c -")))
    }, mc.cores = 4)
}
"38_07_14_100544" <- function(wj1, ot = "t", wjj = 0, intern = ot != 
    "t", named = 1) {
    ext1 <- tolower(hs.re(wj1, "[^\\.]+$"))
    hs.with_wd(wj1, {
        wj1 <- basename(wj1)
        hs.cond(hs.grepl(wj1, "(?i)\\.(tar(\\.gz)?|tgz)$"), {
            Z <- hs.grepl(wj1, "(?i)z$")
            hs.switch(ot, "v", {
                info <- .wk("tar -tf", .q(wj1), if (Z) 
                  "-z", intern = 1)
                jg <- hs.grepV(info, "/$", inv = 1)
                if (named) 
                  jg %<=>% hs.hsgly("1.信息学/文件/名/带名称.38_12_12_185553")(x)
                jg
            }, "t", .wk("tar -tf", .q(wj1), if (Z) 
                "-z", intern = intern))
        }, hs.grepl(wj1, "(?i)\\.zip$"), hs.switch(ot, "v", {
            info <- .wk("unzip -l", .q(wj1), intern = 1)
            ci1 <- regexpr("-+$", info[3])
            info %<=>% x[seq(4, length(x) - 2)]
            jg <- substring(info, ci1)
            if (named) 
                jg %<=>% hs.hsgly("1.信息学/文件/名/带名称.38_12_12_185553")(x)
            jg
        }, "t", .wk("unzip -l", .q(wj1), intern = intern)), hs.grepl(wj1, 
            "(?i)\\.7z$"), hs.switch(ot, "v", {
            info <- .wk("7z l", .q(wj1), intern = 1)
            i <- hs.grep(info, "^[- ]*$")
            info <- info[seq(i[length(i) - 1] - 1, tail(i, 1) - 
                1)]
            ci1 <- regexpr("-+$", info[2])
            jg <- substring(tail(info, -2), ci1)
            if (!wjj) 
                jg <- jg %not% dirname(jg)
            if (named) 
                jg %<=>% hs.hsgly("1.信息学/文件/名/带名称.38_12_12_185553")(x)
            jg
        }, "t", .wk("7z l", .q(wj1), intern = intern)), hs.browser("38.07.14.101031"))
    })
}
"38_09_30_141844" <- function(f1, l = 1, l.head = 9, d = dirname(f1), 
    d.use.self.name = 1, t = 0) {
    if (t) {
        ext1 <- hs.re(f1, "(?i)(?<=.)[gx]z$")
        a <- .wk(hs.switch(ext1, "zip", "unzip -t", "gz", "gzip -t"), 
            .sys.fpGood(f1))
        return(a)
    }
    ext1 <- hs.re(f1, "(?i)(?<=.)(tar(\\.(gz|xz|bz2))?|tgz)$")
    if (length(ext1)) {
        tar.p <- local({
            a <- c("t", "x")
            b <- paste0("-", a, switch(hs.file.extension(ext1), 
                tgz = , gz = "z", xz = "J", bz2 = "j"), "vf")
            names(b) <- a
            b
        })
        if (l) {
            a <- .wk("tar", tar.p["t"], hs.fpGood(f1), "|", .sys.head(1000), 
                intern = 1)
            if (length(a) > l.head) 
                a <- c(head(a, 4), "...", tail(a, 4))
            cat(a, sep = "\n")
        }
        else {
            extN <- hs.strsplit.v(ext1, sep = "\\.") %=>% hs.clean(ot = "n")
            wjj.out <- hs.wjm(f1, dn = d, ext = 0, extN = extN)
            if (!file.exists(wjj.out)) {
                wjj.tmp <- hs.wj.temp(tmpdir = dirname(f1))
                .wk("tar", "-k", tar.p["x"], hs.fpGood(f1), "-C", 
                  hs.fpGood(hs.wjj(wjj.tmp)))
                wjj <- hs.wjj.ls(wjj.tmp, wjj = 1, F = 0)
                if (wjj %=>% {
                  (length(x) == 1) && (x == basename(f1))
                }) {
                  hs.wj.move(hs.fp(wjj.tmp, wjj), d)
                  unlink(wjj.tmp, 1)
                }
                else hs.wj.move(wjj.tmp, wjj.out)
            }
            return(wjj.out)
        }
    }
    if (hs.grepl(f1, "(?i)\\.tar\\.xz$")) {
        if (l) {
            .wk("xzcat", .sys.fpGood(f1), "| tar", "-tf -", "| head")
        }
        else {
            if (d.use.self.name) 
                d <- wk.wjj(hs.wjm(f1, dn = d, ext = 0, extN = 2))
            .wk("xzcat", .sys.fpGood(f1), "| tar", "-xf -", "-C", 
                .sys.fpGood(d))
        }
        return(d)
    }
    if (hs.grepl(f1, "(?i)\\.zip$")) {
        bin <- "unzip"
        if (!length(d.use.self.name)) {
            if (0) {
                fc.0 <- .sys.do("unzip", "-l", .sys.fpGood(f1), 
                  intern = 1)
                fc <- {
                  i <- grep("^-", fc.0)
                  substring(fc.0[seq(i[1] + 1, i[2] - 1)], regexpr("Name", 
                    fc.0[2]))
                }
            }
            fc.0 <- .wk(bin, "-l", wk.wjlj(f1), "| gawk", .q("{ print $4}"), 
                intern = 1)
            fc <- {
                i <- grep("^-", fc.0)
                tail(fc.0, -i) %rm% ""
            }
            d.use.self.name <- as.integer(substring(fc[1], nchar(fc[1])) != 
                "/" || length(table(sapply(fc, ieiwk.dirname.1))) > 
                1)
        }
        if (l) {
            .wk(bin, "-O CP936", "-l", .sys.fpGood(f1), "|", 
                .sys.head(l.head))
        }
        else {
            if (d.use.self.name) 
                d <- hs.wj(hs.wjm(f1, ext = 0, dn = d))
            if (file.exists(d)) {
                message("[target folder already exists]", f1)
                return(f1)
            }
            .wk(bin, "-O CP936", "-d", .sys.fpGood(d), .sys.fpGood(f1))
        }
        return()
    }
    if (wk.re(f1, "\\.gz$", "detect")) {
        bin <- "pigz"
        if (l) {
            .wk("gzip", "-l", wk.wjlj(f1), "|", .sys.head(l.head))
        }
        else {
            .wk(bin, "-dc", wk.wjlj(f1), ">", wk.wjlj(hs.wjm(f1, 
                ext = 0, dn = d)))
        }
        return(ieiwk.filename(f1))
    }
    if (hs.grepl(f1, "\\.7z$")) {
        bin <- "7z"
        if (l) {
            .wk(bin, "l", f1)
        }
        else {
            if (!length(d.use.self.name)) {
                browser("33.06.15.224151")
            }
            if (d.use.self.name) 
                d <- dirname(f1)
            .wk(bin, "x", "-aos -y", paste0("-o", hs.fpGood(d)), 
                hs.fpGood(f1))
            return(hs.wjm(f1, ext = 0))
        }
    }
    if (wk.re(f1, "\\.rar$", "detect")) {
        bin <- "unrar"
        if (l) {
            .wk(bin, "v", f1, "|", .sys.head())
        }
        else {
            wjj2 <- hs.wjm(f1, ext = 0)
            .wk(bin, "x -u", .sys.fpGood(f1), .sys.fpGood(hs.wjj(wjj2)))
            return(f1)
        }
    }
}
"38_07_14_102931" <- function(wj_compressed, wj_in = {
}, wj_to = if (!temp) "-", wjj_to = if (temp) hs.wj.temp() else hs.wjm(wj_compressed, 
    ext = 0), ot = if (temp) "fp" else "cmd", temp = 0) {
    if ((wj_to != "-") && (!hs.grepl(wj_compressed, "(?i)\\.zip$"))) 
        hs.browser("38.07.14.103605")
    cmd1 <- hs.cond(hs.grepl(wj_compressed, "(?i)\\.zip$"), {
        hs.switch(wj_to, "-", .sys("unzip -p", .q(normalizePath(wj_compressed)), 
            .q(wj_in)), .sys("unzip", "-d", hs.fpGood(hs.wjj(wjj_to)), 
            hs.fpGood(wj_compressed), .q(wj_in)))
    }, hs.grepl(wj_compressed, "(?i)\\.7z$"), {
        .sys("7z e", if (wj_to == "-") 
            "-so", .sys.fpGood(wj_compressed), .sys.fpGood(wj_in))
    }, hs.grepl(wj_compressed, "(?i)\\.(tar(\\.gz)?|tgz)$"), 
        {
            Z <- hs.grepl(wj_compressed, "(?i)z$")
            .sys("tar", if (wj_to == "-") 
                c("-O", if (osType() == "Darwin") "-q"), "-xvf", 
                .sys.fpGood(wj_compressed), if (Z) 
                  "-z", .sys.fpGood(wj_in))
        }, hs.browser("38_08_02_143117"))
    hs.switch(ot, "cmd", cmd1, "fp", {
        .wk(cmd1)
        hs.fp(wjj_to, wj_in)
    }, "dt", wk.fread(cmd = cmd1), "less", .wk(cmd1, "| less -Ni"))
}
"38_12_12_185553" <- function(wj) {
    hs.c(hs.wjm(wj, dn = 0, ext = 0), wj)
}
"39_05_12_131245" <- function(wj) {
    a <- strsplit(wj, "/")
    len_common <- local({
        for (i1 in hs.seqForward(1, min(sapply(a, length)))) {
            if (sapply(a, "[", i1) %=>% {
                uniqueN(x) > 1
            }) {
                return(i1 - 1)
            }
        }
        return(i1)
    })
    jg <- list(common = a[[1]] %=>% head(len_common) %=>% as.list %=>% 
        do.call(hs.fp, x), private = sapply(a, . %=>% {
        tail(x, -len_common) %=>% do.call(hs.fp, as.list(x))
    }))
    return(jg)
    common_parent <- {
    }
    repeat {
        a1 <- sapply(a, "[", 1)
        if (anyNA(a1) || (uniqueN(a1) > 1)) {
            break
        }
        else {
            common_parent %<=>% c(a[[1]][1])
            a <- lapply(a, "[", -1)
        }
    }
    list(common = common_parent %=>% do.call(hs.fp, as.list(x)), 
        private = sapply(a, . %=>% do.call(hs.fp, as.list(x))))
}
"36_07_24_100015" <- function(wj, ot = "ext") {
    ext <- hs.re(wj, "(?i)(?<=\\.)(tsv|csv|txt|xlsx?|tar|tgz|zip|rar|xlsx|gz|xz)(\\.(tar|tgz|zip|rar|xlsx|gz|xz))?$")
    hs.switch(ot, "ext", ext, "n", sapply(hs.gre(ext, "\\.+"), 
        length), "wjm", hs.sub(wj, paste0(".", ext)))
}
"2023_06_21_152044" <- function(wjj) {
    hs.with_wd(wjj, {
        .wk("chmod -R go+rx .")
        .wk("chmod -R go-w .")
    })
}
"38_01_21_131617" <- function(rp = all, bash = all, all = 0, 
    wj2exclude = c("/工作/个人/", "/工作/文献/信息", 
        "/工作/单位.其它/", "/资源", "/记忆", "/工作室", 
        "/常见任务", "/emacs", "/jdi.r", "/dm/r/my.r") %not% 
        wj_not_to_exclude, wj_not_to_exclude = hs.switch(to, 
        "9y", c("/工作/个人/")), to = wk.ssh, wjj1 = "~/wk/", 
    wjj2 = hs.fp(root, basename(wjj1)) %=>% paste0("/"), root = "~", 
    bf_wjj0 = if (root == "~") "~/wk.备份") {
    if (0) {
        hs.browser("2023.06.21.145648", debug = 0)
        wjj2 <- "/home/wk/14t1/wk/temp/code/"
        wk.rsync.out("~/wk/", wjj2, ssh = to, ext = if (!all) 
            c("r", "r.gz", "py", "py.gz", "py.zip", "sh", "json"), 
            bf_wjj0 = bf_wjj0, additional = .sys(paste("-f", 
                paste0("- ", wj2exclude, "*") %=>% .q)), included = "/1/org/dm/sh/rj/*", 
            del = 1)
        hs.hsgly("1.信息学/文件夹/删空文件夹/2023_06_21_152354")(wjj2)
        hs.hsgly("1.信息学/文件/权限/别人只读/2023_06_21_152044")(wjj2)
    }
    if (!length(to)) 
        return(hs.msg("2023_07_30_222628|没有指定目标服务器"))
    wk.rsync.out(wjj1, wjj2, ssh = to, ext = if (!all) 
        c("r", "r.gz", "py", "py.gz", "py.zip", "json", "sh", 
            "rprofile"), bf_wjj0 = bf_wjj0, additional = .sys(paste("-f", 
        paste0("- ", wj2exclude, "*") %=>% .q), "-m"), included = "/1/org/dm/sh/rj/*", 
        del = 1)
    if (rp) {
        wk.rsync.out.same_place("~/.Rprofile", ssh = to, L = 1)
    }
    if (bash) {
        wk.rsync.out(normalizePath("~/.bashrc.module"), "~/", 
            ssh = to, p = 0)
        wk.rsync.out(normalizePath("~/.bashrc.portable"), "~/", 
            ssh = to, p = 0)
        wk.rsync.out.same_place("~/.ssh/config", ssh = to, p = 0)
    }
}
"39_07_29_011356" <- function(fx = "out", dry = 1) {
    hs.msg(ln = 1)
    wjj_out_root <- "~/ssd1t"
    if (!file.exists(wjj_out_root)) 
        hs.browser("2023.07.29.013908", debug = 0)
    if (0) {
        wj_temp <- "~/.db/2023_07_29_010348" %=>% hs.wj
        if (file.exists(wj_temp)) {
            if (difftime(Sys.time(), file.mtime(wj_temp), units = "mins") > 
                10) 
                hs.browser("2023.07.29.013904", debug = 0)
            dry <- 0
            hs.wj.rm(wj_temp)
        }
        else {
            dry <- 1
            cat("", file = wj_temp)
        }
    }
    ln_wjj2_db <- hs.home.expand("~/db/软件库")
    db1 <- c("~/wk/", "~/ssd1t/备份/用rsync/wk/同步的/") %=>% 
        hs.db
    db2 <- c("~/.config/opera/", hs.fp(ln_wjj2_db, ".config/opera/"), 
        "~/.cache/opera/", hs.fp(ln_wjj2_db, ".cache/opera/"), 
        "~/.config/microsoft-edge/", hs.fp(ln_wjj2_db, ".config/microsoft-edge/"), 
        "~/.cache/microsoft-edge/", hs.fp(ln_wjj2_db, ".cache/microsoft-edge/"), 
        "~/.config/vivaldi/", hs.fp(ln_wjj2_db, ".config/vivaldi/"), 
        "~/.cache/vivaldi/", hs.fp(ln_wjj2_db, ".cache/vivaldi/"), 
        "~/.ssh/", "~/rjk/ln-s/.ssh/", "~/db/", "~/ssd1t/备份/用rsync/db/") %=>% 
        hs.db
    if (fx == "in") {
        db1 %<=>% hs.db(x, inv = 1)
        db2 %<=>% hs.db(x, inv = 1) %=>% rev
    }
    db <- c(db1, db2)
    for (i in seq_along(db)) {
        wjj1 <- names(db)[i]
        wjj2 <- db[i]
        short <- c(wjj1, wjj2) %=>% x[which.min(nchar(x))]
        hs.msg(short, ln = 2)
        wk.rsync(wjj1, wjj2, del = 1, dry = dry, ssh = {
        }, bf_wjj0 = if (!dry) 
            {
                if ((short == "~/wk/") && (fx == "out")) {
                  wjj2 %=>% hs.sub("[^/]*/$", "备份的/")
                }
                else if (short == "~/db/") 
                  wjj2 %=>% hs.sub("/*$", ".备份/")
            } %=>% {
                if (length(x)) {
                  if (!dir.exists(x)) 
                    hs.browser("2023.08.29.235237", debug = 0)
                  hs.wjj(x)
                }
            })
        if (dry && (short == "~/wk/")) 
            return()
    }
    if (fx == "in") {
        .wk("chmod 600 ~/.ssh/*")
    }
    hs.msg()
}
"2023_06_09_121941" <- function(wj) {
    wj %=>% sapply(x, . %=>% .wk("du -s", hs.fpGood(x), intern = 1)) %=>% 
        wk.fread(text = x, cn = 0)[, V1 %=>% sum] %=>% {
        class(x) <- "object_size"
        x
    } %=>% format(units = "auto")
}
"38_12_11_174401" <- function(ms, wjj, ...) {
    .wk("rg --sortr modified --color always -z", .q(ms), ..., 
        .sys.fpGood(wjj), "| less -NirF")
}
"2023_08_10_015813" <- function() {
    "(?i)\\.(zip|xz|[bt]?gz|bz2|[jr]ar|feather|epub|png|jpe?g)$"
}
"37_05_29_133955" <- function(wjj1 = ".", ot = "fp", z_pattern = hs.hsgly("不需要压缩的模式.2023_08_10_015813")()) {
    switch(ot, fp = hs.wjj.ls(wjj1, z_pattern), glob = local({
        hs.browser("2023.08.10.020846", debug = 0)
        ext <- strsplit(tolower("gz,bgz,tgz,xz,7z,bz2,zip,rar"), 
            ",")[[1]]
        ext <- sapply(ext, . %=>% {
            a1 <- strsplit(x, "")[[1]]
            a2 <- sapply(a1, . %=>% {
                if (x %in% letters) {
                  paste0("[", x, toupper(x), "]")
                } else x
            })
            paste(a2, collapse = "")
        })
        paste0("*.", ext)
    }))
}
"2023_08_10_015150" <- function(wjj1 = ".", z_pattern = hs.hsgly("不需要压缩的模式.2023_08_10_015813")(), 
    size_min = 4096) {
    wj <- hs.wjj.ls(wjj1)
    wj %=>% x[file.size(x) > size_min] %=>% hs.grepV(z_pattern, 
        inv = 1)
}
"2023_06_29_124404" <- function(wjj, pattern) {
    if (!file.exists(wjj)) 
        return()
    if (!dir.exists(wjj)) 
        wjj <- hs.dirname(wjj)
    while (1) {
        wj <- hs.wjj.ls(wjj, pattern, F = 1)
        if (length(wj)) 
            return(wj)
        a <- hs.dirname(wjj)
        if (a == wjj) 
            return()
        wjj <- a
    }
}
"39_01_08_150201" <- function(wj2 = hs.wj(hs.home.expand("~/.db/wk.txt.xz")), 
    ot = "") {
    if (ot == "wj") 
        return(wj2)
    bin1 <- hs.switch(osType(), "Darwin", "gfind", "find")
    hs.msg("hs.hsgly( '文件/找/find.39_01_08_150201')")
    hs.with_wd(hs.home.expand("~/wk"), .wk("cat <(", bin1, "-type d", 
        "-empty | perl -e", .q("while( <>) { print( substr( $_, 1, -1), \"/\\n\");}"), 
        ";", bin1, "-type f | perl -e", .q("while( <>) { print( substr( $_, 1));}"), 
        ") | sort", "| perl -e", .q("\nwhile( <>) {\n  if( /\\.DS_Store\n$/) {\n    next;\n  }\n  if( /wk\\/wd\\/图片\\/pic\\/(b|美)/) {\n    next;\n  }\n  if( /\\/rjk\\/ln-s.个人\\/\\.mozilla(\\.bf)?/) {\n    next;\n  }\n  if( /\\/rjk\\/ztr\\/p\\//) {\n    next;\n  }\n  if( m{^/1/(org|xt)/}) {\n    print( \"~/$1/\", $');\n  } elsif ( m{^/(rjk|tmp)/}) {\n    print( \"~/$1/\", $');\n  } else {\n    print( \"~/wk\", $_);\n  }\n}\n"), 
        "| xz -1 >", hs.fpGood(wj2)))
    hs.msg()
    wj2
}
"38_12_02_195318" <- function(update = 0, pat = readline(), ignore.case = 1, 
    regex = 1, db.wjj = hs.wjj("~/.db", "mlocate"), wjj2do = c("~/wk", 
        "~/in2", "~/f", "~/m") %>% {
        .[dir.exists(.)]
    }, intern = 0) {
    db.wj <- wk.sapply(wjj2do, function(wjj1) {
        hs.wj(db.wjj, ifelse(wjj1 == "~", "home", basename(wjj1)))
    })
    1
    {
        wjj1 <- names(db.wj)[1]
        if (0) 
            for (wjj1 in names(db.wj)) {
                t1 <- file.mtime(db.wj[[wjj1]])
                if (is.na(t1) || (difftime(Sys.time(), t1, units = "mins") > 
                  100)) {
                  update <- 1
                  break
                }
            }
        if (update) 
            for (wjj1 in names(db.wj)) .wk("updatedb -l 0 -o", 
                db.wj[[wjj1]], "-U", wjj1)
    }
    1
    {
        r1 <- .wk("locate", "-b", "-e", if (ignore.case) 
            "-i", if (regex) 
            "--regex", sapply(db.wj, function(x) c("-d", x)), 
            .q(pat), intern = 1)
        n1 <- 20
        cat(head(r1, n1), sep = "\n")
        if (length(r1) > n1) {
            hs.browser("34.11.10.225103")
        }
    }
}
"38_10_08_172545" <- function(max = 16) {
    nc <- Ncpu()[1]
    hs.cond(nc < 1, 1, nc > max, max)
}
"2023_07_14_192031" <- function(df, class = 1, score = 2, smooth = 0, 
    ot = "4js") {
    hs.update(df, hs.df, !is.data.frame(x))
    hs.update(class, colnames(df)[x], is.numeric)
    hs.update(score, colnames(df)[x], is.numeric)
    dca <- evalbq(dcurves::dca(.(as.symbol(class)) ~ .(as.symbol(score)), 
        data = df))
    levels(dca[["dca"]][["label"]]) %<=>% replace(x, match(c("Treat All", 
        "Treat None"), x), c("All", "None"))
    color <- hs.c(levels(dca[["dca"]][["label"]]), c("gray", 
        "black", "red"))
    hs.switch(ot, "4js", {
        dt <- dca[["dca"]] %=>% wk.dt.as
        if (smooth) {
            a <- hs.catch(dt[, {
                fit1 <- stats::loess(net_benefit ~ threshold, 
                  span = 0.2)
                l1 <- paste(wk.dt.copy(label))
                a <- list(name = l1, color = color[l1], point = fit1 %=>% 
                  cbind(x[["x"]], x[["fitted"]]))
                hs.throw(a)
            }, by = "label"])
        }
        else {
            a <- hs.catch(dt[, {
                l1 <- paste(wk.dt.copy(label))
                hs.throw(list(name = l1, color = color[l1], point = cbind(threshold, 
                  net_benefit)))
            }, by = "label"])
        }
        jsonlite::toJSON(a) %=>% cat
    }, hs.browser("2023.07.14.225831", debug = 0))
}
"37_10_12_094644" <- function(m1, wjj2 = {
}, k1 = 20, pcol = 0.8, prow = 1, most.var.n = 5000, cor.method = "spearman", 
    cluster.method = "pam", ...) {
    hs.require("ConsensusClusterPlus", 1)
    if (normalizePath(wjj2) != getwd()) {
        wjj.36_03_04_003344 <- getwd()
        on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
        setwd(dirname(wjj2))
        wjj2 <- basename(wjj2)
    }
    if (!is.matrix(m1)) 
        m1 <- as.matrix(m1)
    if (nrow(m1) > most.var.n) {
        a1 <- rowMads(m1)
        m1 <- m1[tail(order(a1), most.var.n), ]
    }
    d1 <- as.dist(1 - stats::cor(m1, method = cor.method))
    wjj2 <- "~/37_10_12_124437"
    reps <- 1000
    r1 <- ConsensusClusterPlus(d1, maxK = k1, reps = reps, pItem = pcol, 
        pFeature = prow, clusterAlg = cluster.method, seed = 123456, 
        plot = "png", writeTable = TRUE, title = wjj2)
    icl <- calcICL(r1, title = wjj2, plot = "png")
    icl[["clusterConsensus"]] %>% .s
    r1[2]
    pac <- seq.int(2, k1) %>% sapply(. %>% {
        cdf <- r1[[.]][["consensusMatrix"]] %>% {
            .[lower.tri(.)]
        } %>% ecdf
        cdf(0.9) - cdf(0.1)
    })
    which.min(pac)
}
"2023_07_12_183214" <- function(sj, response = 1, predictor = 2, 
    ci = TRUE, ci.alpha = 0.05, ci.color = "lightblue", ...) {
    response %<=>% sj[, x]
    if (uniqueN(response) != 2) 
        stop("2023_07_12_183831|response的种数不是2：", 
            sort(unique(response)))
    predictor %<=>% sj[, x]
    roc <- pROC::roc(response = response, predictor = predictor, 
        ci = ci, ci.alpha = ci.alpha, print.auc = TRUE, plot = TRUE)
    if (ci) {
        roc_ci <- pROC::ci.sp(roc, sensitivities = seq(0, 1, 
            0.001), boot.n = 1000)
        plot(roc_ci, type = "shape", alpha = 0.5, col = ci.color)
    }
    else hs.browser("2023.07.12.192814", debug = 0)
}
"36_07_30_134144" <- local({
    hgnc <- {
    }
    hs.surv_plot <- function(df, wj = "~/tmp/tmp.pdf", width1 = 25, 
        height1 = 15, status.cn, time.unit = "year", time.cn = switch(time.unit, 
            year = "survtime.year", month = "survtime.month"), 
        exp_type = {
        }, risk.table = TRUE, xlab = hs.title(time.unit), xlim2 = switch(time.unit, 
            year = 10, month = 12 * 10), break.x.by = switch(time.unit, 
            year = 1, month = 12), ylab = "Survival probability", 
        legend.title = "SAGE1\nexpression", legend.labs.pre = c("Neg", 
            "Low", "High"), legend.labs.detail = c("", "0<TPM<=1", 
            "TPM>1")) {
        if ("data.table" %in% class(df)) {
            df <- as.data.frame(df)
        }
        if (NA %in% df[, time.cn]) {
            df <- df[!is.na(df[, time.cn]), , drop = FALSE]
        }
        if ("dead" %not.in% names(df)) {
            df[, "dead"] <- 1
        }
        fit1 <- eval(bquote(survfit(Surv(.(as.symbol(time.cn)), 
            dead) ~ .(as.symbol(status.cn)), data = df)))
        fit1.diff <- eval(bquote(survdiff(Surv(.(as.symbol(time.cn)), 
            dead) ~ .(as.symbol(status.cn)), data = df)))
        p1 <- 1 - pchisq(fit1.diff$chisq, length(fit1.diff$n) - 
            1)
        if (0) {
            fit1 %>% summary %>% names
            fit1 %>% summary %>% {
                .[["table"]][, "median"]
            }
        }
        local({
            p1 %<>% {
                paste("P =", format(., scientific = 1, digits = 3))
            }
            if (0) 
                .png({
                  par(mar = c(5, 4 * 3, 4, 2) + 0.1, mfrow = c(1, 
                    2))
                  print(ggsurvplot(fit1, data = df, risk.table = TRUE, 
                    conf.int = TRUE, pval = p1))
                  print(ggsurvplot(fit1, data = df, risk.table = TRUE, 
                    conf.int = FALSE, pval = p1))
                }, wj = wjOut, width = 480 * 6, height = 480 * 
                  3)
            if (0) {
                p1 <- "P < 1e-6 (Neg. VS Low)\nP < 1e-16 (Low VS High)"
                P <- local({
                  col.db <- grDevices::colors()
                  c(0) %>% as.logical %>% lapply(function(conf.int) {
                    lapply(col.db[17], function(col1) {
                      args <- list(fit1, data = df, risk.table = risk.table, 
                        conf.int = conf.int, pval = p1, xlim = c(0, 
                          xlim2 - 0.45), xlab = xlab, break.x.by = break.x.by, 
                        ylab = ylab, surv.median.line = "hv", 
                        censor.shape = "|", censor.size = 1.5, 
                        fontsize = 3)
                      ln1 <- df[[status.cn]] %>% levels %>% length
                      if (ln1 %in% (c(2, 3))) 
                        args["palette"] <- list(switch(as.character(ln1), 
                          `2` = c("blue", "red"), `3` = c(col1, 
                            "blue", "red"))[df[[status.cn]] %>% 
                          hs.table %>% {
                          . > 0
                        }])
                      args["legend.labs"] <- local({
                        a1 <- df[, status.cn] %>% hs.table
                        if (!length(legend.labs.pre)) {
                          legend.labs.pre <- names(a1)
                        }
                        if (!length(legend.labs.detail)) 
                          legend.labs.detail <- rep("", length(a1))
                        a2 <- hs.c(names(a1), paste0(legend.labs.pre, 
                          " (", legend.labs.detail %>% {
                            ifelse(nchar(.), paste0(., "; "), 
                              "")
                          }, "n=", a1, ")"))[a1 > 0]
                        if (0) 
                          names(a1) %>% sapply(. %>% {
                            v1 <- hs.re(., "(?<=[\\(\\[])[^\\)\\]]+") %>% 
                              strsplit(",") %>% unlist %>% as.numeric
                            if (is.infinite(v1[1])) {
                              paste0(exp_type, switch(hs.re(., 
                                ".$"), `)` = "<", `]` = "<="), 
                                v1[2])
                            }
                            else if (is.infinite(v1[2])) {
                              paste0(exp_type, switch(hs.re(., 
                                "."), `(` = ">", `[` = ">="), 
                                v1[1])
                            }
                            else hs.pastec0(c(v1[1], switch(substr(., 
                              1, 1), `(` = "<", `[` = "<=")), 
                              exp_type, c(switch(hs.re(., ".$"), 
                                `)` = "<", `]` = "<="), v1[2]))
                          })
                        list(a2)
                      })
                      args["legend.title"] <- list(legend.title)
                      args["legend"] <- list(c(0.8, 0.8))
                      args["ggtheme"] <- list(theme_survminer(panel.spacing = unit(5, 
                        "cm")))
                      do.call(ggsurvplot, args)
                    })
                  })
                })
                .mfp.swxx("36.09.17.153858") %>% hs.wjj(go = 1)
                .mfp.swxx("36.09.17.153858") %>% hs.wjj.clean
                lapply(16:25, function(h1) {
                  wj1 <- paste0("height=", h1, ".pdf")
                  .pdf(print(P), width = 25, height = h1, wj = wj1)
                  hs.pdf.removePage1(wj1)
                  wj1
                })
                .mfp.swxx("36.09.17.153858") %>% wk.rsync.in
            }
            .pdf(quote({
                col.db <- grDevices::colors()
                c(0) %>% as.logical %>% lapply(function(conf.int) {
                  lapply(col.db[17], function(col1) {
                    args <- list(fit1, data = df, risk.table = risk.table, 
                      conf.int = conf.int, pval = p1, xlim = c(0, 
                        xlim2), xlab = xlab, break.x.by = break.x.by, 
                      ylab = ylab, surv.median.line = "hv", censor.shape = "|", 
                      censor.size = 1.5, fontsize = 3)
                    ln1 <- df[[status.cn]] %>% levels %>% length
                    if (ln1 %in% (c(2, 3))) args["palette"] <- list(switch(as.character(ln1), 
                      `2` = c("blue", "red"), `3` = c(col1, "blue", 
                        "red"))[df[[status.cn]] %>% hs.table %>% 
                      {
                        . > 0
                      }])
                    args["legend.labs"] <- local({
                      a1 <- df[, status.cn] %>% hs.table
                      if (!length(legend.labs.pre)) {
                        legend.labs.pre <- names(a1)
                      }
                      if (!length(legend.labs.detail)) legend.labs.detail <- rep("", 
                        length(a1))
                      a2 <- hs.c(names(a1), paste0(legend.labs.pre, 
                        " (", legend.labs.detail %>% {
                          ifelse(nchar(.), paste0(., "; "), "")
                        }, "n=", a1, ")"))[a1 > 0]
                      if (0) names(a1) %>% sapply(. %>% {
                        v1 <- hs.re(., "(?<=[\\(\\[])[^\\)\\]]+") %>% 
                          strsplit(",") %>% unlist %>% as.numeric
                        if (is.infinite(v1[1])) {
                          paste0(exp_type, switch(hs.re(., ".$"), 
                            `)` = "<", `]` = "<="), v1[2])
                        } else if (is.infinite(v1[2])) {
                          paste0(exp_type, switch(hs.re(., "."), 
                            `(` = ">", `[` = ">="), v1[1])
                        } else hs.pastec0(c(v1[1], switch(substr(., 
                          1, 1), `(` = "<", `[` = "<=")), exp_type, 
                          c(switch(hs.re(., ".$"), `)` = "<", 
                            `]` = "<="), v1[2]))
                      })
                      list(a2)
                    })
                    args["legend.title"] <- list(legend.title)
                    args["legend"] <- list(c(0.8, 0.8))
                    P <- do.call(ggsurvplot, args)
                    print(P)
                  })
                })
            }), wj = wj, height = height1, width = width1)
            hs.pdf.removePage1(wj)
        })
        {
            wjOut <- hs.pastec0(dirname(wj), "/", if (length(exp_type) && 
                (exp_type == "tpm")) 
                "tpm.", "fit.rd")
            hs.save(blm = c("fit1", "fit1.diff", "p1"), wj = wjOut)
        }
    }
    function(gene1, gn = 5, gn2 = 2, flag.sage1 = "sage1+-", 
        jg_wjj = .mfp.swxx("36.02.20.194147"), jg_what = {
        }, case2rm = {
        }, cancer2do = {
        }, cancer2rm = {
        }, gender2do = {
        }, individual = !length(cancer2do), break.L = c(-Inf, 
            0, 1, Inf), tbl = {
        }, redo = 0, legend.labs.pre = {
        }, ...) {
        hs.require("survminer", "survival", "clusterProfiler", 
            "org.Hs.eg.db")
        if (!length(hgnc)) 
            hgnc <<- hs.hsgly("gene.id_hgnc.35.12.06.111314")()
        if (!length(tbl)) 
            tbl <- hs.hsgly("yb,sage1_max.info.36.07.06.130805")(gene1 = gene1)
        if (length(case2rm) && (!is.null(tbl))) 
            tbl <- tbl[case_id %not.in% case2rm]
        if (length(cancer2do) && (!is.null(tbl))) 
            tbl <- tbl[project_id %in% cancer2do]
        if (length(cancer2rm) && (!is.null(tbl))) 
            tbl <- tbl[project_id %not.in% cancer2rm]
        if (length(gender2do) && (!is.null(tbl))) 
            tbl <- tbl[gender %in% gender2do]
        if (length(jg_what)) 
            return(switch(jg_what, jg_wjj = jg_wjj, tbl = tbl))
        if (is.null(tbl)) 
            return()
        gene1.name <- hs.hsgly("gene.id.converter.using_hgnc.36.01.22.160941")(gene1, 
            "ensembl", "symbol", "ensembl", "symbol")[[2]]
        if (tbl[, "TCGA-TGCT" %in% project_id]) 
            tbl <- tbl[project_id %not.in% "TCGA-TGCT"]
        df <- switch(flag.sage1, `sage1+-` = copy(tbl), `sage1-` = tbl[fpkm_ENSG00000181433 == 
            0], `sage1+` = tbl[fpkm_ENSG00000181433 > 0])
        lapply(gene1, function(g1) {
            "fpkm,tpm" %>% strsplit(",") %>% unlist %>% {
                "tpm"
            } %>% lapply(function(exp_type) {
                c("all") %>% lapply(function(cohort) {
                  if (cohort != "all") 
                    df %<>% wk.dt.subset(cohort, "project_id")
                  if (df[, .N] == 0) 
                    return()
                  if (0) 
                    c(1:5) %>% {
                      2
                    } %>% lapply(function(gn) {
                      wjOut <- hs.pastec0(jg_wjj, "/", g1, "/", 
                        cohort, ".", exp_type, ".", "gn=", gn, 
                        ".surv.pdf") %>% hs.wj
                      if (file.exists(wjOut) && (!redo)) {
                      }
                      else {
                        g1.cn <- paste0(exp_type, "_", g1)
                        if (g1.cn %not.in% names(df)) {
                          hs.browser("36.07.06.145130")
                        }
                        if (df[, NA %in% get(g1.cn)]) 
                          df <- df[!is.na(get(g1.cn))]
                        if (df[, .N] == 0) 
                          return()
                        f1 <- df[, g1.cn %>% get %>% {
                          Q <- quantile(.[. > 0], seq(1/gn, 1, 
                            by = 1/gn))
                          if (uniqueN(Q) < length(Q)) 
                            return()
                          a1 <- cut(., c(-1, 0, Q))
                          levels(a1) <- c(0, paste0(gn, "_", 
                            seq.int(gn)))
                          a1
                        }]
                        if (length(f1) == 0) 
                          return()
                        status.cn <- paste0(g1, ".status")
                        df[, `:=`(c(status.cn), f1)]
                        if (group.is.small.hs(df[, get(status.cn)], 
                          "0")) 
                          next
                        {
                          df %<>% as.data.frame
                          hs.surv_plot(df, wjOut, status.cn = status.cn, 
                            exp_type = exp_type, risk.table = TRUE, 
                            ...)
                        }
                      }
                      wjOut
                    })
                  {
                    g1.cn <- paste0(exp_type, "_", g1)
                    if (!is.list(break.L)) 
                      break.L %<>% list
                    sapply(break.L, function(break1) {
                      wjOut <- hs.pastec0(jg_wjj, "/", if (length(gene1.name) == 
                        1) 
                        gene1.name
                      else g1, "/", cohort, ".", exp_type, ".", 
                        exp_type, "_co,", break1 %>% {
                          .[!is.infinite(.)]
                        } %>% paste(collapse = ","), ".surv.pdf") %>% 
                        hs.wj
                      if (file.exists(wjOut) && (!redo)) {
                      }
                      else {
                        f1 <- df[, g1.cn %>% get %>% {
                          cut(., break1)
                        }]
                        break1 <- levels(f1) %>% hs.grepV("\\bInf\\b", 
                          inv = 1) %>% hs.re("(?<=.).+(?=.$)")
                        if ((length(levels(f1)) < 2) || (length(f1) == 
                          0)) 
                          return()
                        {
                          hist1 <- f1 %>% hs.table
                          if (sum(hist1 > 0) < 2) 
                            return()
                          if (0) {
                            i0 <- which(hist1 == 0)
                            if (length(i0)) {
                              f1 <- droplevels(f1, names(hist1)[i0])
                            }
                          }
                        }
                        status.cn <- paste0(g1, ".status")
                        df[, `:=`(c(status.cn), f1)]
                        hs.surv_plot(df, wjOut, status.cn = status.cn, 
                          exp_type = exp_type, risk.table = TRUE, 
                          legend.title = paste0(gene1.name, "\nexpression"), 
                          legend.labs.pre = legend.labs.pre, 
                          legend.labs.detail = {
                          }, ...)
                      }
                      wjOut
                    })
                  }
                })
            })
        })
    }
})
"36_10_20_020830" <- local({
    group.is.small.hs <- function(f1, l1, gs.co = 10) {
        sum(f1 == as.character(l1)) < gs.co
    }
    hs.pdf.removePage1 <- {
    }
    hs.surv_plot <- function(df, wj = "~/tmp/tmp.pdf", width1 = 25, 
        height1 = 15, group.cn, time.unit = "year", time.cn = switch(time.unit, 
            year = "survtime.year", month = "survtime.month"), 
        exp_type = {
        }, risk.table = TRUE, xlab = hs.title(time.unit), xlim2 = switch(time.unit, 
            year = 10, month = 12 * 10), break.x.by = switch(time.unit, 
            year = 1, month = 12), ylab = "Survival probability", 
        legend.title = "SAGE1\nexpression", legend.labs.pre = c("Neg", 
            "Low", "High"), legend.labs.detail = c("", "0<TPM<=1", 
            "TPM>1")) {
        if ("data.table" %in% class(df)) {
            df <- as.data.frame(df)
        }
        if (NA %in% df[, time.cn]) {
            df <- df[!is.na(df[, time.cn]), , drop = FALSE]
        }
        if ("dead" %not.in% names(df)) {
            df[, "dead"] <- 1
        }
        fit1 <- eval("survfit" %=>% bquote(.(as.symbol(x))(Surv(.(as.symbol(time.cn)), 
            dead) ~ .(as.symbol(group.cn)), data = df)))
        fit1.diff <- eval("survdiff" %=>% bquote(.(as.symbol(x))(Surv(.(as.symbol(time.cn)), 
            dead) ~ .(as.symbol(group.cn)), data = df)))
        p1 <- 1 - pchisq(fit1.diff$chisq, length(fit1.diff$n) - 
            1)
        local({
            p1 %<=>% paste("P =", format(x, scientific = 1, digits = 3))
            if (0) 
                .png({
                  par(mar = c(5, 4 * 3, 4, 2) + 0.1, mfrow = c(1, 
                    2))
                  print(ggsurvplot(fit1, data = df, risk.table = TRUE, 
                    conf.int = TRUE, pval = p1))
                  print(ggsurvplot(fit1, data = df, risk.table = TRUE, 
                    conf.int = FALSE, pval = p1))
                }, wj = wjOut, width = 480 * 6, height = 480 * 
                  3)
            if (0) {
                p1 <- "P < 1e-6 (Neg. VS Low)\nP < 1e-16 (Low VS High)"
                P <- local({
                  col.db <- grDevices::colors()
                  c(0) %=>% as.logical %=>% lapply(x, function(conf.int) {
                    lapply(col.db[17], function(col1) {
                      args <- list(fit1, data = df, risk.table = risk.table, 
                        conf.int = conf.int, pval = p1, xlim = c(0, 
                          xlim2 - 0.45), xlab = xlab, break.x.by = break.x.by, 
                        ylab = ylab, surv.median.line = "hv", 
                        censor.shape = "|", censor.size = 1.5, 
                        fontsize = 3)
                      ln1 <- df[[status.cn]] %=>% levels %=>% 
                        length
                      if (ln1 %in% (c(2, 3))) 
                        args["palette"] <- list(switch(as.character(ln1), 
                          `2` = c("blue", "red"), `3` = c(col1, 
                            "blue", "red"))[df[[status.cn]] %=>% 
                          hs.table %=>% (x > 0)])
                      args["legend.labs"] <- local({
                        a1 <- df[, status.cn] %=>% hs.table
                        if (!length(legend.labs.pre)) {
                          legend.labs.pre <- names(a1)
                        }
                        if (!length(legend.labs.detail)) 
                          legend.labs.detail <- rep("", length(a1))
                        a2 <- hs.c(names(a1), paste0(legend.labs.pre, 
                          " (", legend.labs.detail %=>% ifelse(nchar(x), 
                            paste0(x, "; "), ""), "n=", a1, ")"))[a1 > 
                          0]
                        if (0) 
                          names(a1) %=>% sapply(x, . %=>% {
                            v1 <- hs.re(x, "(?<=[\\(\\[])[^\\)\\]]+") %=>% 
                              strsplit(",") %=>% unlist %=>% 
                              as.numeric
                            if (is.infinite(v1[1])) {
                              paste0(exp_type, switch(hs.re(x, 
                                ".$"), `)` = "<", `]` = "<="), 
                                v1[2])
                            }
                            else if (is.infinite(v1[2])) {
                              paste0(exp_type, switch(hs.re(x, 
                                "."), `(` = ">", `[` = ">="), 
                                v1[1])
                            }
                            else hs.pastec0(c(v1[1], switch(substr(x, 
                              1, 1), `(` = "<", `[` = "<=")), 
                              exp_type, c(switch(hs.re(x, ".$"), 
                                `)` = "<", `]` = "<="), v1[2]))
                          })
                        list(a2)
                      })
                      args["legend.title"] <- list(legend.title)
                      args["legend"] <- list(c(0.8, 0.8))
                      args["ggtheme"] <- list(theme_survminer(panel.spacing = unit(5, 
                        "cm")))
                      do.call(ggsurvplot, args)
                    })
                  })
                })
                .mfp.swxx("36.09.17.153858") %=>% hs.wjj(go = 1)
                .mfp.swxx("36.09.17.153858") %=>% hs.wjj.clean
                lapply(16:25, function(h1) {
                  wj1 <- paste0("height=", h1, ".pdf")
                  .pdf(print(P), width = 25, height = h1, wj = wj1)
                  hs.pdf.removePage1(wj1)
                  wj1
                })
                .mfp.swxx("36.09.17.153858") %=>% wk.rsync.in
            }
            .pdf(quote({
                col.db <- grDevices::colors()
                c(0) %=>% as.logical %=>% lapply(x, function(conf.int) {
                  lapply(col.db[17], function(col1) {
                    args <- list(fit1, data = df, risk.table = risk.table, 
                      conf.int = conf.int, pval = p1, xlim = c(0, 
                        xlim2), xlab = xlab, break.x.by = break.x.by, 
                      ylab = ylab, surv.median.line = "hv", censor.shape = "|", 
                      censor.size = 1.5, fontsize = 3)
                    ln1 <- df[[group.cn]] %=>% levels %=>% length
                    if (ln1 %in% (c(2, 3))) args["palette"] <- list(switch(as.character(ln1), 
                      `2` = c("blue", "red"), `3` = c(col1, "blue", 
                        "red"))[df[[group.cn]] %=>% hs.table %=>% 
                      (x > 0)])
                    args["legend.labs"] <- local({
                      a1 <- df[, group.cn] %=>% hs.table
                      if (!length(legend.labs.pre)) {
                        legend.labs.pre <- names(a1)
                      }
                      if (!length(legend.labs.detail)) legend.labs.detail <- rep("", 
                        length(a1))
                      a2 <- hs.c(names(a1), paste0(legend.labs.pre, 
                        " (", legend.labs.detail %=>% ifelse(nchar(x), 
                          paste0(x, "; "), ""), "n=", a1, ")"))[a1 > 
                        0]
                      if (0) names(a1) %=>% sapply(x, . %>% {
                        v1 <- hs.re(., "(?<=[\\(\\[])[^\\)\\]]+") %>% 
                          strsplit(",") %>% unlist %>% as.numeric
                        if (is.infinite(v1[1])) {
                          paste0(exp_type, switch(hs.re(., ".$"), 
                            `)` = "<", `]` = "<="), v1[2])
                        } else if (is.infinite(v1[2])) {
                          paste0(exp_type, switch(hs.re(., "."), 
                            `(` = ">", `[` = ">="), v1[1])
                        } else hs.pastec0(c(v1[1], switch(substr(., 
                          1, 1), `(` = "<", `[` = "<=")), exp_type, 
                          c(switch(hs.re(., ".$"), `)` = "<", 
                            `]` = "<="), v1[2]))
                      })
                      list(a2)
                    })
                    args["legend.title"] <- list(legend.title)
                    args["legend"] <- list(c(0.8, 0.8))
                    P <- do.call(ggsurvplot, args)
                    print(P)
                  })
                })
            }), wj = wj, height = height1, width = width1)
            hs.pdf.removePage1(wj)
        })
        {
            wjOut <- hs.pastec0(dirname(wj), "/", if (length(exp_type) && 
                (exp_type == "tpm")) 
                "tpm.", "fit.rd")
            hs.save(blm = c("fit1", "fit1.diff", "p1"), wj = wjOut)
        }
    }
    function(df, redo = 0, ...) {
        hs.require("survminer", "survival")
        if (!length(hs.pdf.removePage1)) 
            hs.pdf.removePage1 <<- hs.hsgly("pdf.removePage1.36_11_04_211749")
        if (df[, .N] == 0) 
            return()
        wjOut
        if (file.exists(wjOut) && (!redo)) {
        }
        else {
            f1 <- df[, group]
            if ((length(levels(f1)) < 2) || (length(f1) == 0)) 
                return()
            {
                hist1 <- f1 %=>% hs.table
                if (sum(hist1 > 0) < 2) 
                  return()
                if (0) {
                  i0 <- which(hist1 == 0)
                  if (length(i0)) {
                    f1 <- droplevels(f1, names(hist1)[i0])
                  }
                }
            }
            status.cn <- paste0(g1, ".status")
            df[, `:=`(c(status.cn), f1)]
            hs.surv_plot(df, wjOut, group.cn = group.cn, exp_type = exp_type, 
                risk.table = TRUE, legend.title = paste0(gene1.name, 
                  "\nexpression"), legend.labs.pre = legend.labs.pre, 
                legend.labs.detail = {
                }, ...)
        }
        wjOut
    }
})
"38_11_21_074938" <- function(x, y, data, fit = {
}, x_new, ot = "fit") {
    if (!length(fit)) {
        exp1 <- bquote(lm(.(as.symbol(y)) ~ .(as.symbol(x))), 
            data = data)
        fit <- eval(exp1)
    }
    hs.switch(ot, "fit", fit1, "y", {
        predict(fit, x_new, interval = "confidence")
    })
}
"36_04_07_121916" <- local({
    scatter_plot.hs <- {
    }
    function(m1, group = {
    }, show.legend = length(group), legend.args = {
    }, show.cn = 1, wj = "~/tmp/tmp.pdf", width = 10, height = width, 
        ot = "", xy_ratio.co = 5, main.cn = "columns", main.rn = "rows", 
        plotter = .pdf, ...) {
        if (!length(scatter_plot.hs)) 
            scatter_plot.hs <<- hs.hsgly("点图.36.07.01.174228")
        m1 <- m1[rowSums(is.na(m1)) == 0, , drop = 0]
        i1 <- rowSums(is.infinite(m1)) > 0
        if (any(i1)) 
            m1 <- m1[-i1, , drop = 0]
        i1 <- hs.rowSds(m1) == 0
        if (any(i1)) 
            m1 <- m1[-i1, , drop = 0]
        pca <- prcomp(t(m1), scale = TRUE)
        pc.load <- pca$sdev %>% {
            .^2
        } %>% {
            ./sum(.)
        }
        pc.xy <- predict(pca)
        hs.switch(ot, "xy", return(list(pc.xy = pc.xy, pc.load = pc.load, 
            n1 = nrow(m1))))
        xi1 <- 1
        xi2 <- 2
        if (0) {
            xi1 <- 2
            xi2 <- 3
        }
        pch1 <- 17
        if (!length(group)) 
            group <- 1:ncol(m1)
        if (0) {
            tbl <- data.frame(x1 = pc.xy[, xi1], x2 = pc.xy[, 
                xi2])
            tbl[, "x1,x2.color"] <- sapply(1:nrow(pc.xy), ieiwk.col2rgb)
            tbl[, "yb"] <- colnames(m1)
        }
        xl1 <- paste("PC", xi1, " (", round(pc.load[xi1] * 100, 
            2), "% variance)", sep = "")
        yl1 <- paste("PC", xi2, " (", round(pc.load[xi2] * 100, 
            2), "% variance)", sep = "")
        main1 <- paste0("PCA (", ncol(m1), " ", main.cn, ", ", 
            nrow(m1), " ", main.rn, ")")
        height <- 20
        width <- pc.load %>% {
            r1 <- .[xi1]/.[xi2]
            ifelse(r1 > xy_ratio.co, 1, r1) * height
        }
        plotter({
            par(xpd = 1)
            plot(pc.xy[, xi1], pc.xy[, xi2], xlab = paste("PC", 
                xi1, " (", round(pc.load[xi1] * 100, 2), "% variance)", 
                sep = ""), ylab = paste("PC", xi2, " (", round(pc.load[xi2] * 
                100, 2), "% variance)", sep = ""), main = paste0("PCA (", 
                ncol(m1), " ", main.cn, ", ", nrow(m1), " ", 
                main.rn, ")"), col = as.integer(factor(group)), 
                pch = pch1)
            if (show.cn) 
                text(pc.xy[, xi1], pc.xy[, xi2], labels = colnames(m1), 
                  pos = 1)
            if (show.legend) {
                A1 <- list(x = "topright", legend = group %>% 
                  hs.table(order = 0) %>% {
                  paste0(names(.), " (n=", ., ")")
                }, pch = pch1, col = group %>% uniqueN %>% seq)
                if (length(legend.args)) {
                  A1[names(legend.args)] <- legend.args
                }
                do.call("legend", A1)
            }
        }, wj, width = width, height = height)
        if (0) {
            pc.xy %>% dimnames
            d1 <- as.dist(1 - cor(m1, method = "spearman"))
            d1[1]
            choose(ncol(m1), 2)
        }
        if (0) 
            .pdf({
                fig1 <- scatter_plot.hs(tbl, "x1", "x2", "x1,x2.color", 
                  xl1, yl1, main1, "yb")
                print(fig1)
            }, wj = wj, width = width, height = height)
    }
})
"36_04_16_130256" <- local({
    hs.dist <- function(mat1, side = "col") {
        as.dist(1 - cor(switch(side, row = t(mat1), col = mat1), 
            method = "spearman"))
    }
    callback <- function(hc, mat) {
        sv <- svd(t(mat))$v[, 1]
        dend <- reorder(as.dendrogram(hc), wts = sv)
        as.hclust(dend)
    }
    function(m1, wj1 = hs.wj(paste0("~/tmp/tmp.", if (capabilities()["png"]) 
        "png"
    else "pdf")), log = if (min(m2) >= 0) 
        2) {
        m1 <- m1[hs.rowSums(is.na(m1)) == 0, ]
        if (log) 
            m1 <- log2(m1 + 1)
        i1 <- hs.rowSds(m1) %>% {
            which(. > quantile(.[. > 0], 0.5))
        }
        m1 <- t(apply(m1[i1, , drop = 0], 1, function(.) (. - 
            mean(.))/sd(.)))
        column.distance <- hs.dist(m1)
        (pheatmap::pheatmap)(m1, filename = wj1, color = colorRampPalette(rev(brewer.pal(n = 7, 
            name = "RdYlBu")))(101), breaks = seq(-3, 0, length.out = 50) %or% 
            seq(0, 3, length.out = 50), cluster_cols = TRUE, 
            cluster_rows = TRUE, clustering_distance_cols = column.distance, 
            clustering_callback = callback, show_colnames = TRUE, 
            show_rownames = FALSE, cellwidth = 5, fontsize = 5, 
            height = 3)
        wj1
    }
})
"36_08_12_231620" <- local({
    hs.cb <- {
    }
    function(m1, cor.method = "spearman", cb = 1, use = "pairwise.complete.obs", 
        hc.method = "ward.D", side = "column", r2 = 0, ot = "c") {
        if (!length(hs.cb)) 
            hs.cb <<- hs.hsgly("wk.pheatmap.36.03.11.103943", 
                "callback")
        switch(side, row = {
            m1 <- t(m1)
        })
        Cor <- cor(m1, method = cor.method, use = use)
        Cor[is.na(Cor)] <- 0
        if (r2) 
            Cor <- Cor^2
        hc1 <- fastcluster::hclust(as.dist(1 - Cor), method = hc.method)
        if (cb) 
            hc1 <- local({
                i <- is.na(m1)
                if (any(i)) 
                  m1[i] <- 0
                hs.cb(hc1, m1)
            })
        switch(ot, c = hc1, o = hc1[["order"]], om = {
            m1 <- m1[, hc1[["order"]], drop = FALSE]
            switch(side, row = t(m1), column = m1)
        })
    }
})
"37_12_27_141843" <- function(df1, wj2 = "~/tmp/table0.pdf", 
    rows = {
    }, pdf.args = list(), ...) {
    hs.require("gridExtra")
    do.call(".pdf", c(quote(gridExtra::grid.table(df1, rows = rows, 
        ...)), wj2, pdf.args))
}
"38_10_02_083407" <- function(L, V) {
    paste0(L, " (", if (sum(V) == 1) 
        ""
    else paste0(V, "; "), round(V/sum(V) * 100, 2), "%)")
}
"35_07_24_102842" <- function(V, L, wj2 = {
}, width = 15, height = 10, main = "", topN = 0, plot = length(wj2), 
    plotter = .pdf) {
    hs.require("ggplot2")
    hs.browser("38.10.02.115149")
    sj1 <- if (topN) {
        hs.hsgly("饼图/topN.38_10_02_082809")(V, L, topN)
    }
    else wk.dt(V = as.vector(V), L)
    sj1[, `:=`(L, hs.factor(L, rev(L), N = 0))]
    apropos("hs.str")
    myLabel <- sj1[, hs.hsgly("饼图/加百分比.38_10_02_083407")(L = L, 
        V = V)]
    if (0) {
        sj1[L = paste0(seq_along(L) %>% {
            sprintf(paste0("%", max(nchar(.)), "d"), .)
        }, L)]
        myLabel <- sj1[, paste0(round(A/sum(A) * 100, 2), "%\n", 
            A)]
    }
    if (plot) {
        font_family <- "mono"
        P <- ggplot2::ggplot(sj1, aes(x = "V", y = V, fill = L)) + 
            ggplot2::geom_bar(stat = "identity", width = 1, position = "stack") + 
            ggplot2::coord_polar(theta = "y") + ggplot2::labs(x = "", 
            y = "", title = "") + ggplot2::scale_fill_manual(values = hs.hsgly("color.db.38_09_09_173646")(), 
            breaks = sj1[, L], labels = paste0(hs.str_pad(seq_along(myLabel)), 
                ") ", myLabel)) + ggplot2::geom_text(aes(x = 1.55, 
            y = cumsum(V) - V * 0.5, label = paste0("(", seq_along(L), 
                ")")), size = 3, family = font_family) + ggplot2::theme(text = element_text(family = font_family), 
            axis.ticks = element_blank(), panel.grid = element_blank(), 
            panel.border = element_blank(), axis.text.x = element_blank(), 
            legend.title = element_blank(), plot.title = element_text(hjust = 0.5)) + 
            ggplot2::ggtitle(main)
        plotter(print(P), wj2, width = width, height = height)
    }
}
"35_07_24_103728" <- function(V, L, wj2 = "~/swxx/tmp/tmp", width = 15, 
    height = 10, topN = 0, plot = 1, redo = 0, plotter = .png, 
    clockwise = TRUE, col = hs.hsgly("color.db.38_09_09_173646")(), 
    percent = 1, ...) {
    if (anyDuplicated(L)) {
        a1 <- tapply(V, L, sum)
        V <- unname(a1)
        L <- names(a1)
    }
    if (topN) {
        sj1 <- hs.hsgly("饼图/topN.38_10_02_082809")(V = V, 
            L = L, topN = topN)
        V <- sj1[, V]
        L <- sj1[, L]
    }
    if (percent) {
        L <- hs.hsgly("饼图/加百分比.38_10_02_083407")(L = L, 
            V = V)
        width <- width * 1.1
    }
    if (plot && (redo || (!file.exists(wj2)))) 
        plotter({
            x1 <- hs.c(L, V)
            pie(x1, clockwise = clockwise, col = col, ...)
        }, wj = wj2, width = width, height = height)
    wj2 %<>% hs.wjm(ext = "tsv.gz")
    R <- wk.dt(L, V)
    R[, `:=`("P", V/sum(V))]
}
"38_10_02_082809" <- function(V, L, topN) {
    sj1 <- wk.dt(V, L)
    wk.dt.setorderv(sj1, "V", -1)
    sj1 <- rbind(head(sj1, topN), tail(sj1, -topN)[, .(V = sum(V), 
        L = "Others")])
    sj1[, `:=`("proportion", V/sum(V))]
}
"38_04_17_165151" <- function(df) {
    hs.require("ggplot2")
    ggplot(df)
}
"2023_08_19_113708" <- function(text) {
    text %=>% hs.gsub(">=", "≥") %=>% textutils::HTMLencode() %=>% 
        gridtext::richtext_grob()
}
"36_08_06_163847" <- function(L, wj = "~/swxx/tmp.pdf", redo = 0, 
    width1 = height1 + 5, height1 = 10, N = 1, ...) {
    if (N) 
        names(L) %<>% paste0("\n(n=", sapply(L, uniqueN), ")")
    .pdf(grid.draw(VennDiagram::venn.diagram(L, {
    }, units = "cm", width = width1, height = height1, ...)), 
        wj = wj, width = width1 + 1, height = height1 + 1)
}
"2023_08_18_180631" <- function(png_fp, keep = 5, wj2 = hs.wjm(png_fp, 
    append = ".切掉白边")) {
    png <- png::readPNG(png_fp)
    if (length(dim(png)) != 3) 
        hs.browser("2023.08.18.181859", debug = 0)
    col_n <- min(dim(png)[3], 3)
    bg <- rep(1, col_n)
    co1 <- dim(png)[1]/2
    for (i in seq(ceiling(co1 + 1))) if (png[i, , ] %=>% t %=>% 
        hs.colAlls(x == bg) %=>% !all(x)) 
        break
    if (i >= co1) 
        hs.browser("2023.08.18.204612", debug = 0)
    i1 <- max(i - keep, 1)
    for (i in seq(dim(png)[1], floor(co1 - 1))) if (png[i, , 
        ] %=>% t %=>% hs.colAlls(x == bg) %=>% !all(x)) 
        break
    if (i <= co1) 
        hs.browser("2023.08.18.230457", debug = 0)
    i2 <- min(i + keep, dim(png)[1])
    co1 <- dim(png)[2]/2
    for (i in seq(ceiling(co1 + 1))) if (png[, i, ] %=>% t %=>% 
        hs.colAlls(x == bg) %=>% !all(x)) 
        break
    if (i >= co1) 
        hs.browser("2023.08.18.230539", debug = 0)
    j1 <- max(i - keep, 1)
    for (i in seq(dim(png)[2], floor(co1 - 1))) if (png[, i, 
        ] %=>% t %=>% hs.colAlls(x == bg) %=>% !all(x)) 
        break
    if (i <= co1) 
        hs.browser("2023.08.18.230539", debug = 0)
    j2 <- min(i + keep, dim(png)[2])
    if (hs.any(i1 > 1, i2 < dim(png)[1], j1 > 1, j2 < dim(png)[2])) 
        png %<=>% x[seq(i1, i2), seq(j1, j2), ]
    if (length(wj2)) {
        png::writePNG(png, wj2)
        wj2
    }
    else png
}
"36_08_12_233720" <- function(m1, color.breaks.boundary = m1 %=>% 
    {
        if (m1.is.like.cor.matrix) 
            x[upper.tri(x)]
        else x
    } %=>% x[!is.infinite(x)] %=>% range(x, na.rm = 1) %=>% abs %=>% 
    max %=>% (round(x * 2)/2) %=>% min(x, color.breaks.boundary.max), 
    color.breaks.boundary.max = 6, colors = {
    }, color.breaks.n = local({
        r1 <- range(m1, na.rm = TRUE)
        if (r1[1] >= center) {
            c(center, color.breaks.boundary)
        }
        else if (r1[2] <= center) {
            c(-color.breaks.boundary, center)
        }
        else color.breaks.boundary %>% {
            c(-., center, .)
        }
    }), center = switch(legend.name.type, or = 1, 0), color.breaks = local({
        r1 <- range(m1, na.rm = TRUE)
        if (r1[1] >= center) {
            c("white", wk.color["red"])
        }
        else if (r1[2] <= center) {
            c(wk.color["blue"], "white")
        }
        else c("purple", "black", "yellow")
    }), m1.is.like.cor.matrix = 0, name = "Value", legend.show = 1, 
    legend.name = switch(legend.name.type, lfc = expression(Log[2] * 
        "(fold change)"), fdr = expression(-Log[10] * "FDR"), 
        or = "odds ratio", evalbq(expression(Log[2] * .(legend.name.type)))), 
    legend.name.type = "lfc", legend.args = list(title = if (length(legend.name)) legend.name else name, 
        title_gp = gpar(fontfamily = fontfamily), labels_gp = gpar(fontsize = 10, 
            fontfamily = fontfamily)), legend.args_extra = {
    }, cell_width = 0.4, cell_height = cell_width * hw_ratio, 
    hw_ratio = 1, show.name.co = 200, width = if (width.full) {
        cell_width * ncol(m1)
    } else 10, width.full = ncol(m1) <= width.full.co, width.full.co = show.name.co, 
    width.extra = 10, height = if (height.full) {
        cell_height * nrow(m1)
    } else 10, height.full = (nrow(m1) <= show.name.co) || length(rn2mark), 
    height.full.co = show.name.co, height.extra = 5, rn.show = height.full, 
    rn.side = "right", row_dend_side = "left", rn.size = cell_height * 
        72.27/2.54, cn.size = cell_width * 72.27/2.54 * cn.size.factor, 
    cn.size.factor = 1, cn.show = width.full, cn.side = "top", 
    column_dend_side = "top", cluster_rows = if (any(is.na(m1))) {
        FALSE
    } else function(m1) hs.hsgly("层次聚类/fastcluster.36_08_12_231620")(m1, 
        side = "row"), cluster_columns = FALSE, use_raster = prod(dim(m1)) > 
        10000, wj = hs.fp(hs.fig_wj_wjj(), "tmp"), heatmap_legend_side = "left", 
    lgd.at = local({
        a1 <- ceiling(color.breaks.boundary * 10)/10
        r1 <- range(m1, na.rm = TRUE)
        if (r1[1] >= center) {
            c(center, a1)
        }
        else if (r1[2] <= center) {
            c(-a1, center)
        }
        else a1 %>% {
            c(-., center, .)
        }
    }), na_col = "gray", plot = 1, column_names_rot = 60, row_order = 0, 
    column_order = 0, cell.border.col = if (height.full && width.full) "gray", 
    cell.border.width = 0.1, cell_fun = if (all(c(length(hs.clean(cell.border.col)), 
        length(cell_fun.L)))) {
        function(j, i, x, y, width, height, fill) {
            here <- environment()
            if (0) 
                if (length(cell.border.col)) {
                  grid.rect(x = x, y = y, width = width, height = height, 
                    gp = gpar(col = cell.border.col, fill = NA))
                }
            cell_fun.env.import.flag <- length(cell_fun.env.L) > 
                0
            if (length(cell_fun.env.L) == 1) {
                hs.import(cell_fun.env.L[[1]])
                cell_fun.env.import.flag <- 0
            }
            for (f1i in seq_along(cell_fun.L)) {
                f1 <- cell_fun.L[[f1i]]
                if (length(f1) == 0) 
                  next
                if (cell_fun.env.import.flag) {
                  hs.import(cell_fun.env.L[[f1i]])
                }
                environment(f1) <- here
                f1()
            }
        }
    }, cell_fun.L = list(), cell_fun.env.L = list(local(function() {
    })), anno.yesno.list = {
    }, anno.yesno.col = hs.c(anno.yesno.name1, anno.yesno.col1), 
    anno.yesno.name1 = c("Yes", "No"), anno.yesno.col1 = c("red", 
        "gray"), anno.df.list = {
    }, anno.col.list = {
    }, anno.list = {
    }, rn2mark = {
    }, rn2mark.top = 0, row.anno.name.side = "top", column.anno.name.side = "left", 
    column_title = "", column_title_gp = gpar(fontsize = 10, 
        fontfamily = fontfamily), row_title_rot = 0, color.space = "RGB", 
    rect_gp = gpar(col = cell.border.col, lwd = cell.border.width), 
    fontfamily = "sans", ...) {
    hs.require(c("ComplexHeatmap", "circlize"))
    args <- list(...)
    if (any(is.infinite(m1))) {
        a1.37_05_12_142014 <- hs.deInf(as.vector(m1))
        dim(a1.37_05_12_142014) <- dim(m1)
        dimnames(a1.37_05_12_142014) <- dimnames(m1)
        m1 <- a1.37_05_12_142014
    }
    if ("column_title_rot" %not.in% names(args)) {
        args["column_title_rot"] <- list(if (("column_split" %in% 
            names(args)) && (max(nchar(rownames(args[["column_split"]]))) > 
            3)) 90 else 0)
    }
    if (!length(height)) 
        height <- if (height.full || (nrow(m1) <= show.name.co)) {
            cell_height * nrow(m1)
        }
        else 10
    if (identical(colors, NA)) {
    }
    else if (length(colors) == 0) 
        colors <- circlize::colorRamp2(color.breaks.n, color.breaks, 
            space = color.space)
    legend.args[["at"]] <- lgd.at
    lgd.labels <- as.character(lgd.at)
    if (cluster_rows %>% {
        is.numeric(.) && (length(.) == 1)
    }) 
        cluster_rows %<=>% as.logical
    if (cluster_columns %>% {
        is.numeric(.) && (length(.) == 1)
    }) 
        cluster_columns %<>% as.logical
    if (length(legend.args_extra)) 
        legend.args[names(legend.args_extra)] <- legend.args_extra
    hs.for(side1, c("bottom", "left", "top", "right"), assign(paste0(side1, 
        "_annotation"), local({
        A1 <- df1 <- col1 <- list()
        name.V <- switch(side1, right = , left = rownames(m1), 
            colnames(m1))
        {
            side1.sj <- anno.yesno.list[[side1]]
            if (length(side1.sj)) {
                for (l1.name in names(side1.sj)) {
                  jg <- hs.c(name.V, rep(anno.yesno.name1[2], 
                    length(name.V)))
                  l1 <- side1.sj[[l1.name]] %and% name.V
                  if (length(l1)) 
                    jg[l1] <- anno.yesno.name1[1]
                  df1[l1.name] <- list(jg)
                }
                col1[names(side1.sj)] <- list(anno.yesno.col)
            }
        }
        {
            side1.sj <- anno.df.list[[side1]]
            if (length(side1.sj)) {
                hs.browser("38.10.07.232628")
                for (l1.name in names(side1.sj)) {
                  l1 <- side1.sj[[l1.name]]
                  jg <- hs.c(name.V, rep(NA, length(name.V)))
                  for (l2i in seq_along(l1)) {
                    l2 <- l1[[l2i]] %and% name.V
                    if (length(l2)) 
                      jg[l2] <- names(l1)[l2i]
                  }
                  df1[l1.name] <- list(hs.factor(jg, N = 0))
                }
                side1.col <- anno.col.list[[side1]]
                hs.browser("38.10.12.171027")
                if (length(side1.col)) {
                  col1[names(side1.col)] <- side1.col
                }
                else {
                  hs.browser("38.10.12.170959")
                }
            }
        }
        {
            side1.sj <- anno.list[[side1]]
            if (length(side1.sj)) {
                A1[names(side1.sj)] <- side1.sj
                col1[names(side1.sj)] <- lapply(side1.sj, . %>% 
                  {
                    hs.cond(is.character(.), {
                      a <- hs.clean(., U = 1)
                      hs.c(a, hs.hsgly("color.db.38_09_09_173646")(n1 = length(a)))
                    }, is.factor(.), {
                      a <- levels(.)
                      hs.c(a, hs.hsgly("color.db.38_09_09_173646")(n1 = length(a)))
                    })
                  })
            }
        }
        df1 %<>% hs.nonempty %>% {
            .[sapply(., . %>% {
                any(hs.clean(., ot = "l"))
            })]
        }
        if (length(df1)) {
            A1["df"] <- list(setDF(df1, name.V))
            if (length(col1)) {
                A1["col"] <- list(col1)
                hs.hsgly("color.db.38_09_09_173646")() %>% {
                  .pdf(plot(seq_along(.), col = ., pch = 16))
                }
            }
        }
        if (length(A1)) {
            A1["which"] <- switch(side1, right = , left = "row", 
                "column")
            if (0) {
                A1["annotation_label"] <- list(gt_render(c("TPM.range", 
                  "**Annotation**_<span style='color:red'>foo</span>", 
                  colnames(A1[["df"]]))))
            }
            A1["annotation_name_rot"] <- list(switch(side1, right = , 
                left = 90, 0))
            A1["annotation_name_side"] <- list(switch(side1, 
                right = , left = row.anno.name.side, column.anno.name.side))
            A1[["col"]] <- col1
            do.call("HeatmapAnnotation", A1)
        }
    })))
    if (length(rn2mark)) {
        if (hs.grepl(rn2mark, "\n")) 
            rn2mark %<>% strsplit("\n") %>% unlist %>% hs.gsub("−", 
                "-")
        if (rn2mark.top) 
            rn2mark %<>% head(rn2mark.top)
        right_annotation <- rowAnnotation(foo = anno_mark(at = match(rn2mark, 
            rownames(m1)), labels = rn2mark))
    }
    anno.text.width <- list(right_annotation, left_annotation) %>% 
        hs.nonempty %>% {
        if (length(.)) {
            sapply(., names) %>% text.width.hs
        }
        else 0
    }
    if (!anno.text.width) {
        anno.text.width <- text.width.hs(rownames(m1)) * 1.5
    }
    anno.text.height <- list(top_annotation, bottom_annotation) %>% 
        hs.nonempty %>% {
        if (length(.)) {
            sapply(., names) %>% text.width.hs
        }
        else 0
    }
    width.extra <- sum(if (length(right_annotation)) {
        as.numeric(sum(attr(right_annotation, "width"), attr(right_annotation, 
            "gap"))/10)
    } else 0, if (length(top_annotation)) {
        get_color_mapping_list(top_annotation)[[1]] %>% attr("levels") %>% 
            text.width.hs
    } else 0, anno.text.width, width.extra)
    args <- merge_list(list(m1, col = colors, show_heatmap_legend = FALSE, 
        width = grid::unit(width, "cm"), height = grid::unit(height, 
            "cm"), show_row_names = as.logical(rn.show), show_column_names = as.logical(cn.show), 
        row_names_side = rn.side, column_names_side = cn.side, 
        use_raster = use_raster, cluster_rows = cluster_rows, 
        cluster_columns = cluster_columns, row_dend_side = row_dend_side, 
        column_dend_side = column_dend_side, na_col = na_col, 
        column_names_rot = column_names_rot, cell_fun = cell_fun, 
        rect_gp = rect_gp, right_annotation = right_annotation, 
        left_annotation = left_annotation, top_annotation = top_annotation, 
        bottom_annotation = bottom_annotation, row_title_rot = row_title_rot, 
        row_names_gp = gpar(fontsize = rn.size, fontfamily = fontfamily), 
        column_names_gp = gpar(fontsize = cn.size, fontfamily = fontfamily), 
        row_dend_gp = gpar(fontfamily = fontfamily), column_dend_gp = gpar(fontfamily = fontfamily), 
        column_title = column_title, column_title_gp = column_title_gp), 
        args)
    P <- do.call("Heatmap", args)
    if (0) {
        hs.require("magick")
        args[["col"]](-4)
    }
    if (plot) {
        height.extra <- height.extra + 5
        if (hs.grepl(wj, "(?i)\\.png$")) {
            wj <- .png(ComplexHeatmap::draw(P, heatmap_legend_side = heatmap_legend_side), 
                wj, width = width + width.extra, height = height + 
                  height.extra, family = "sans")
        }
        else wj <- .pdf({
            if (legend.show) {
                lgd1.args <- list(title = legend.name, col_fun = colors, 
                  at = lgd.at, labels = lgd.labels, direction = "horizontal")
                if (length(legend.args)) 
                  lgd1.args[names(legend.args)] <- legend.args
                lgd1 <- do.call(ComplexHeatmap::Legend, lgd1.args)
                draw(P, heatmap_legend_list = list(lgd1), heatmap_legend_side = heatmap_legend_side)
            }
            else draw(P)
        }, wj, width = width + width.extra, height = height + 
            height.extra)
        if (any(c(row_order, column_order))) {
            list(if (row_order) row_order(P) %>% sapply(. %>% 
                {
                  rownames(m1)[.]
                }), if (column_order) column_order(P) %>% sapply(. %>% 
                {
                  colnames(m1)[.]
                }))
        }
        else wj
    }
    else P
}
"36_03_11_103943" <- local({
    mat1_ordered <- {
    }
    callback <- function(hc, mat) {
        sv <- svd(t(mat))$v[, 1]
        dend <- reorder(as.dendrogram(hc), wts = sv)
        as.hclust(dend)
    }
    function(mat1, wj1.wjm, N = 51, color = colorRampPalette(c("navy", 
        "white", "firebrick3"))(N), cluster = 0, cluster_cols = as.logical(cluster), 
        cluster_rows = as.logical(cluster), show_names = 0, show_rownames = as.logical(show_names), 
        show_colnames = as.logical(show_names), pheatmap.args = list(), 
        ...) {
        hs.require(c("pheatmap", "viridis", "RColorBrewer", "dendsort"), 
            1)
        if (0) {
            test = matrix(rnorm(200), 20, 10)
            test[1:10, seq(1, 10, 2)] = test[1:10, seq(1, 10, 
                2)] + 3
            test[11:20, seq(2, 10, 2)] = test[11:20, seq(2, 10, 
                2)] + 2
            test[15:20, seq(2, 10, 2)] = test[15:20, seq(2, 10, 
                2)] + 4
            colnames(test) = paste("Test", 1:10, sep = "")
            rownames(test) = paste("Gene", 1:20, sep = "")
            hs.find.r("mat.*order")
            hs.matrix.order <- function(mat1, column = 1, row = 1) {
                d1 <- if (row) {
                  d1 <- dist(mat1)
                  c1 <- hclust(d1)
                  mat1 <- mat1[c1[["order"]], ]
                }
                d2 <- if (column) {
                  d2 <- dist(t(mat1))
                  c2 <- hclust(d2)
                  mat1 <- mat1[, c2[["order"]]]
                }
                expR1 <- quote(list(matrix = mat1, row.distance = d1, 
                  column.distance = d2, row.hclust = c1, column.hclust = c2))
                if (!column) 
                  expR1[c("column.distance", "column.hclust")] <- {
                  }
                if (!row) 
                  expR1[c("row.distance", "row.hclust")] <- {
                  }
                eval(expR1)
            }
            hs.matrix.order.spearman <- function(mat1) {
                d1 <- dist(mat1)
                c1 <- hclust(d1)
                mat1 <- mat1[c1[["order"]], ]
                d2 <- dist(t(mat1))
                c2 <- hclust(d2)
                mat1 <- mat1[, c2[["order"]]]
                list(matrix = mat1, row.distance = d1, column.distance = d2, 
                  row.hclust = c1, column.hclust = c2)
            }
            mat1.cluster.info <- hs.matrix.order(mat1, column = 0)
            {
                similar_to_sage1 = 1
                pervasive = 0
                pervasive.co = 0.8
                hs.heatmap <- function(mat1, gene1 = "SAGE1", 
                  cancer = {
                  }, normalize = 0, pervasive = 0, pervasive.co = 0.8, 
                  similar_to_gene1 = 300) {
                  if (length(cancer)) 
                    mat1 %<>% {
                      .[, colnames(.) %and% (annotation_col %>% 
                        {
                          rownames(.)[.[, "cancer"] %in% cancer]
                        })]
                    }
                  mat1 %<>% {
                    .[rowSums(. > 0) >= 10, ]
                  }
                  if (pervasive) {
                    mat1 %<>% {
                      .[rowMeans(. > 0) >= pervasive.co, ]
                    }
                  }
                  else {
                    mat1 %<>% {
                      .[rowMeans(. > 0) < pervasive.co, ]
                    }
                  }
                  if (1) {
                    distance <- mat1 %>% apply(1, . %>% {
                      i1 <- which(mat1[gene1, ] > 0)
                      p1 <- cor.test(.[i1], mat1[gene1, i1], 
                        method = "spearman") %>% {
                        c(.[["p.value"]], .[["estimate"]])
                      }
                      s1 <- which(. > 0)
                      s2 <- which(mat1[gene1, ] > 0)
                      p2 <- c(1, `odds ratio` = 1)
                      c(p1, p2)
                    }) %>% t %>% as.data.table
                    distance[, `:=`("qv1", p.adjust(V1, "fdr"))]
                    distance[, `:=`("qv2", p.adjust(V3, "fdr"))]
                    distance[, `:=`("d1", log10(qv1) * sign(rho))]
                    distance[, `:=`("d2", log10(qv2) * sign(`odds ratio`))]
                    distance[, `:=`("min", apply(data.table(d1, 
                      d2), 1, min))]
                    distance[, `:=`("gene", rownames(mat1))]
                    setorderv(distance, "min", na.last = TRUE)
                    mat1 %<>% {
                      .[distance[, gene], ]
                    }
                    if (1) {
                      mat1 %<>% {
                        rows <- if (similar_to_gene1) {
                          if (is.logical(similar_to_gene1)) {
                            distance[(min < log10(0.05)) & (rho > 
                              0), gene]
                          }
                          else distance[, head(gene, similar_to_gene1)]
                        }
                        else distance[min >= log10(0.05), gene]
                        .[rows, , drop = 0]
                      }
                    }
                  }
                  if (0) {
                    fig.wj <- paste0("~/tmp/tmp.", gene1, ",", 
                      paste(dim(mat1), collapse = "x"), ".jpeg")
                    if (file.exists(fig.wj)) 
                      return()
                  }
                  {
                    mat1 %<>% {
                      .[, order.hs(.[gene1, ])]
                    }
                    i1 <- match(gene1, rownames(mat1))
                    if (i1 != 1) {
                      mat1[c(1, i1), ] <- mat1[c(i1, 1), ]
                      rownames(mat1)[c(1, i1)] %<>% rev
                    }
                  }
                  if (0) {
                    annotation_col[, "cancer"] %>% hs.table %>% 
                      sort %>% .sh
                    annotation_col[colnames(mat1), "cancer"] %>% 
                      hs.table %>% sort %>% .sh
                    mat1 %<>% {
                      .[, order.hs(.[gene1, ])]
                    }
                    mat1 %<>% {
                      .[, .[gene1, ] == 0]
                    }
                    mat1 %<>% {
                      a1 <- hclust(as.dist(1 - cor(., method = "spearman")))
                      . <- .[, a1[["order"]]]
                      .
                    }
                    mat1 %<>% {
                      i1 <- which(.[gene1, ] == 0)
                      .[, i1] <- .[, sample(i1)]
                      .
                    }
                    mat1 %>% {
                      i1 <- which(.[gene1, ] == 0)
                    }
                  }
                  rownames(mat1) %not% cta.info[, gene.id_symbol]
                  wk.dt.subset(cta.info, rownames(mat1), "gene.id_symbol")
                  annotation_row <- local({
                    a1 <- wk.dt.subset(cta.info, rownames(mat1), 
                      "gene.id_symbol")[, specificity %>% hs.deInf]
                    rownames(mat1)[which.min(a1)]
                    wk.dt.subset(cta.info, "PRDM13", "gene.id_symbol")
                    a1 <- data.frame(log2specificity = log2(a1), 
                      is_selected = ifelse(toupper(rownames(mat1)) %in% 
                        toupper(genes_oi), "Yes", "No") %>% v.pct.hs)
                    rownames(a1) <- rownames(mat1)
                    a1
                  })
                  annotation_row[["-log10(cor_spearman.pv)"]] <- wk.dt.subset(distance, 
                    rownames(mat1), "gene")[, -min] %>% hs.deInf
                  annotation_colors[["is_selected"]] <- hs.c(levels(annotation_row[, 
                    "is_selected"]), c("black", "red"))
                  if (0) {
                    mat1 %<>% {
                      i1 <- which(.[gene1, ] > 0)
                      .[, i1] %<>% {
                        .[, hclust(as.dist(1 - cor(., method = "spearman")))[["order"]]]
                      }
                      .
                    }
                    mat1 %<>% {
                      i1 <- which(.[gene1, ] == 0)
                      .[, i1] %<>% {
                        .[, hclust(as.dist(1 - cor(., method = "spearman")))[["order"]]]
                      }
                      .
                    }
                  }
                  if (0) 
                    mat1 %<>% {
                      i1 <- which(.[gene1, ] == 0)
                      colnames(.)[i1] %<>% sample
                      .
                    }
                  if (normalize) 
                    mat1 %<>% apply(1, . %>% {
                      .[. > 0] %<>% {
                        cut(., quantile(., seq(0, 1, 0.1)), include.lowest = TRUE)
                      } %>% as.integer
                      .
                    }) %>% t
                  hs1 <- function(mat1, pos = 0, neg = 0, cluster = "sorted", 
                    cluster.row = 0) {
                    if (pos) 
                      mat1 %<>% {
                        .[, .[gene1, ] > 0]
                      }
                    if (neg) 
                      mat1 %<>% {
                        .[, .[gene1, ] == 0]
                      }
                    mat1 %<>% {
                      .[, apply(., 2, sd) > 0]
                    }
                    mat1 %<>% {
                      .[apply(., 1, sd) > 0, ]
                    }
                    mat1 %>% {
                      fig.wj <- paste0("~/tmp/tmp.", if (length(cancer)) 
                        paste0(paste(sort(cancer), collapse = ","), 
                          ".")
                      else "", gene1, ",", ifelse(pos, "pos", 
                        ""), ifelse(neg, "neg", ""), ".", paste(dim(.), 
                        collapse = "x"), ".column,", switch(cluster, 
                        sorted = cluster, paste0("clustered,", 
                          cluster)), if (cluster.row) 
                        paste0(".row,", cluster)
                      else "", ".pdf")
                      if (file.exists(fig.wj)) 
                        return()
                      hs.dist <- function(cluster, side = "col") switch(cluster, 
                        cor = as.dist(1 - cor(switch(side, row = t(.), 
                          col = .), method = "spearman")), eucli = "euclidean")
                      pheatmap(., cluster_rows = as.logical(cluster.row), 
                        cluster_cols = switch(cluster, sorted = FALSE, 
                          TRUE), clustering_distance_cols = hs.dist(cluster), 
                        clustering_distance_rows = if (cluster.row) 
                          hs.dist(cluster, side = "row"), clustering_callback = callback, 
                        annotation_col = annotation_col, annotation_row = annotation_row, 
                        annotation_colors = annotation_colors, 
                        annotation_legend = TRUE, show_colnames = FALSE, 
                        show_rownames = TRUE, cellheight = 5, 
                        fontsize = 5, filename = fig.wj)
                    }
                  }
                  hs1(mat1, pos = 1)
                  hs1(mat1, pos = 1, cluster = "cor")
                  hs1(mat1, pos = 1, cluster = "eucli")
                  hs1(mat1, neg = 1, cluster = "cor")
                  hs1(mat1, neg = 1, cluster = "eucli")
                  hs1(mat1, pos = 1, cluster = "eucli", cluster.row = 1)
                  hs1(mat1, pos = 1, cluster = "cor", cluster.row = 1)
                  hs1(mat1, neg = 1, cluster = "eucli", cluster.row = 1)
                  hs1(mat1, neg = 1, cluster = "cor", cluster.row = 1)
                }
                mat1 %>% {
                  i1 <- 1:ncol(.)
                  hs.heatmap(.[, i1], "SAGE1")
                }
                for (c1 in unique(annotation_col[, "cancer"])) .try(hs.heatmap(mat1, 
                  "SAGE1", cancer = c1))
                hs.heatmap(mat1, "MAGEA6")
                hs.heatmap(mat1 %>% {
                  .[, .["SAGE1", ] == 0]
                }, "MAGEA6")
                "MAGEA6" %in% rownames(mat1)
                hs.heatmap(mat1 %>% {
                  .[, .["SAGE1", ] > 0]
                }, "CT83")
                hs.heatmap(mat1 %>% {
                  .[, .["SAGE1", ] == 0]
                }, "CT83")
                mat1 %>% {
                  .[, .["SAGE1", ] > 0]
                } %>% dim
                for (gene1 in genes_oi) {
                  hs.say(gene1)
                  hs.heatmap(mat1, gene1)
                }
                hs.heatmap(mat2.1, "~/tmp/lihc.exp,log2,0_to_10.according_to_sage1.pdf", 
                  normalize = 1)
                hs.heatmap(mat2.1, "~/tmp/lihc.exp,log2.similar_to_sage1.pdf")
                pheatmap(mat2.1, cluster_rows = FALSE, cluster_cols = FALSE, 
                  annotation_col = annotation_col, annotation_row = annotation_row, 
                  annotation_colors = annotation_colors, annotation_legend = TRUE, 
                  show_colnames = FALSE, show_rownames = TRUE, 
                  cellheight = 5, fontsize = 5, filename = "~/tmp/lihc.exp,log2.similar_to_sage1.pdf")
            }
            mat2 <- t(scale(t(mat1)))
            mat2 %<>% {
                t(scale(t(.)))
            }
            mat2.cluster.info <- hs.matrix.order(mat2)
            {
                tcga.tbl %<>% {
                  .[, {
                    .SD[1]
                  }, by = sample_id]
                }
                annotation_col <- tcga.tbl %>% {
                  a1 <- .[, survtime] %>% {
                    .[!is.na(.)] %<>% {
                      cut(., quantile(., seq(0, 1, by = 0.1)))
                    }
                    data.frame(survtime = as.character(.))
                  }
                  rownames(a1) <- .[, sample_id]
                  a1[["cancer"]] <- .[, hs.sub(project_id, "TCGA-")]
                  a1
                }
                annotation_colors <- list(survtime = hs.c(1:10, 
                  rev(inferno(10))), cancer = unique(annotation_col[, 
                  "cancer"]) %>% {
                  hs.c(., head(col.db, length(.)))
                }, specificity = c("white", "firebrick3"))
                quantile_breaks <- function(xs, n = 10) {
                  breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
                  breaks[!duplicated(breaks)]
                }
                mat_breaks <- quantile_breaks(set.clean.hs(mat1), 
                  10)
            }
            local({
                mat1 <- mat2
                mat1.cluster.info <- mat2.cluster.info
                pheatmap(mat1, cluster_rows = mat1.cluster.info[["row.hclust"]], 
                  cluster_cols = mat1.cluster.info[["column.hclust"]], 
                  annotation_col = annotation_col, annotation_row = annotation_row, 
                  annotation_colors = annotation_colors, annotation_legend = TRUE, 
                  show_colnames = FALSE, show_rownames = TRUE, 
                  cellheight = 5, fontsize = 5, filename = "~/tmp/lihc.exp,+1,log2,std.pdf")
            })
            mat1["CFHR5", ] %>% {
                . > 0
            } %>% mean
            {
                mat2 <- mat1 %>% t %>% scale %>% t
                mat2.cluster.info <- hs.matrix.order(mat2)
                pheatmap(mat2, kmeans_k = 100, cluster_rows = mat2.cluster.info[["row.hclust"]], 
                  cluster_cols = mat2.cluster.info[["column.hclust"]], 
                  cellheight = 5, fontsize = 5, filename = "~/tmp/lihc.exp,+1,log2,std.pdf")
            }
            mat1["SAGE1", ] %>% {
                . > 0
            } %>% mean
        }
        i1 <- which(is.infinite(mat1))
        if (length(i1)) {
            max1 <- mat1 %>% {
                .[!is.infinite(.)]
            } %>% abs %>% max
            mat1[i1] %<>% {
                ifelse(sign(.) == 1, max1, -max1)
            }
        }
        breaks <- max(abs(na.omit(as.vector(mat1)))) %>% {
            seq(-., ., length.out = N)
        }
        if (0) {
            A <- list(...)
            if ("width" %not.in% names(A)) 
                mat1 %>% dim
        }
        pheatmap.args[["mat"]] <- mat1
        pheatmap.args[["show_rownames"]] <- show_rownames
        pheatmap.args[["show_colnames"]] <- show_colnames
        pheatmap.args[["cluster_cols"]] <- cluster_cols
        pheatmap.args[["cluster_rows"]] <- cluster_rows
        pheatmap.args[["color"]] <- color
        pheatmap.args[["breaks"]] <- breaks
        hs.browser("36.03.04.152113")
        {
            pheatmap.args %>% names
            pheatmap.args[["show_rownames"]]
            pheatmap.args[["show_colnames"]]
            pheatmap.args[["annotation_row"]]
        }
        if (capabilities()["png"]) {
            wj1 <- hs.wjm(wj1.wjm, add = "png")
            pheatmap.args[["fontsize"]] <- 8
            height <- 40
            height <- annotation_col %>% {
                5 + (length(.) + sum(sapply(., . %>% levels %>% 
                  length)))/2
            }
            width <- height
            png(wj1, res = 350, units = "cm", width = width, 
                height = height)
            if (0) {
                if (max(dim(mat1)) > 999) {
                  wj1 <- hs.wjm(wj1.wjm, add = "tiff")
                  hs.browser("36.02.09.211341")
                  if (0) {
                    height <- 5
                    height <- 15
                    width <- 10
                    width <- 35
                    height <- 40
                    pheatmap.args[["show_rownames"]] <- FALSE
                    wj1 %>% .ls
                  }
                  height <- annotation_col %>% {
                    5 + (length(.) + sum(sapply(., . %>% levels %>% 
                      length)))/2
                  }
                  width <- height
                  png(wj1, res = 350, units = "cm", width = width, 
                    height = height)
                }
            }
        }
        else {
            wj1 <- hs.wjm(wj1.wjm, add = "pdf")
            pdf(wj1, ...)
        }
        print(c1 <- do.call("pheatmap", pheatmap.args))
        c1 <- do.call("pheatmap", pheatmap.args)
        c1[["tree_row"]]
        c1[["tree_col"]][["order"]]
        colnames(mat1)[1019]
        mat1[, 1019]
        `?`(pheatmap)
        a2[, YT_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE]
        dev.off()
        wj1 %>% .ls
        "~/tmp/tmp.png" %>% wk.rsync.in.same_place
        wj1
    }
})
"36_06_23_173012" <- function(hc, rotate = 1, wj = "~/tmp/hclust.pdf", 
    width = 5, height = 5, ...) {
    .pdf({
        P <- ggdendro::ggdendrogram(hc, rotate = rotate, ...)
        show(P)
    }, wj = wj, width = width, height = height)
}
"2023_07_10_183513" <- function() {
    wjj2 <- hs.sjgly("2023_07_10_183513")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
    }
}
"2023_07_04_140953" <- function(png.wj, by_order = "col", plot = 0) {
    reticulate::use_python(normalizePath("~/.local/bin/python"))
    reticulate::py_run_string("\nimport cv2\nimport numpy")
    reticulate::py_run_string(paste0("src = cv2.imread( \"", 
        normalizePath(png.wj), "\")"))
    reticulate::py_run_string("gray = cv2.cvtColor( src, cv2.COLOR_BGR2GRAY)")
    reticulate::py_run_string("gray = cv2.medianBlur( gray, 5)")
    reticulate::py_run_string("rows = gray.shape[ 0]")
    reticulate::py_run_string("circles = cv2.HoughCircles( image = gray, method = cv2.HOUGH_GRADIENT_ALT, dp = 1, minDist = 7, param1 = 100, param2 = 0.9, minRadius = 2, maxRadius = int( numpy.min( src.shape[ 0:2])))")
    if (0) {
        reticulate::py_run_string("help( cv2.HoughCircles)")
        reticulate::py_run_string("\nif circles is not None:\n    circles = numpy.uint16( numpy.around( circles))\nfor i in circles[ 0, :]:\n    center = ( i[ 0], i[ 1])\n    # circle center\n    cv2.circle( src, center, 1, ( 0, 100, 100), 3)\n    # circle outline\n    radius = i[ 2]\n    cv2.circle( src, center, radius, (255, 0, 255), 3)\ncv2.imshow( \"detected circles\", src)\ncv2.waitKey( 0)\n# cv.destroyAllWindows()\n")
    }
    circles <- reticulate::py[["circles"]][1, , ]
    png <- png::readPNG(png.wj)
    bg <- hs.hsgly("2.数据学/图像处理/背景色/2023_07_26_213428")(png)
    d <- dim(png)[2] %=>% seq %=>% sapply(x, . %=>% {
        Rfast::dista(bg, png[, x, ])[1, ]
    }) > 0.1
    dm <- apply(circles, 1, . %=>% {
        c1 <- hs.hsgly("2.数据学/坐标/圆/2023_07_26_194158")(x[1], 
            x[2], x[3], x.max = dim(png)[2], y.max = dim(png)[1])
        if (nrow(c1) > 1000) 
            c1 %<=>% hs.ij(x, sample(nrow(x), 1000))
        mean(d[c1[, c(2, 1)]])
    })
    a <- circles[dm > 0.8, ] %=>% wk.dt.as
    r1 <- a[, mean(V3)]
    col_n <- a[, V1 %=>% sort %=>% unique %=>% diff %=>% sum(x > 
        (r1 * 2))] + 1
    row_n <- a[, V2 %=>% sort %=>% unique %=>% diff %=>% sum(x > 
        (r1 * 2))] + 1
    a[, `:=`("V1", V1 %=>% {
        cl <- dist(x) %=>% hclust %=>% cutree(k = col_n)
        m <- tapply(x, cl, median)
        m[cl]
    })]
    a[, `:=`("V2", V2 %=>% {
        cl <- dist(x) %=>% hclust %=>% cutree(k = row_n)
        m <- tapply(x, cl, median)
        m[cl]
    })]
    a <- hs.sort(a, use = hs.switch(by_order, "row", c("V2", 
        "V1"), "col", c("V1", "V2")), fx = 1)
    col <- apply(a, 1, . %=>% {
        c1 <- hs.hsgly("2.数据学/坐标/圆/2023_07_26_194158")(x[1], 
            x[2], x[3], x.max = dim(png)[2], y.max = dim(png)[1])
        if (nrow(c1) > 1000) 
            c1 %<=>% hs.ij(x, sample(nrow(x), 1000))
        col <- apply(png[c1[, 2], c1[, 1], ], 3, median)
        c(x[1:2], head(col, 3))
    })
    col <- apply(col, 2, function(x) {
        do.call(rgb, as.list(x[-c(1, 2)]))
    })
    if (plot) {
        col %=>% plot(seq_along(x), col = x, pch = 16, cex = 3)
        abline(v = row_n %=>% seq(x, length(col), by = x) + 0.5)
    }
    col
}
"38_09_09_173646" <- function(n1 = Inf) {
    a <- c("#DC143C", "#0000FF", "#20B2AA", "#FFA500", "#9370DB", 
        "#98FB98", "#F08080", "#1E90FF", "#7CFC00", "#FFFF00", 
        "#808000", "#FF00FF", "#FA8072", "#7B68EE", "#9400D3", 
        "#800080", "#A0522D", "#D2B48C", "#D2691E", "#87CEEB", 
        "#40E0D0", "#5F9EA0", "#FF1493", "#0000CD", "#008B8B", 
        "#FFE4B5", "#8A2BE2", "#228B22", "#E9967A", "#4682B4", 
        "#32CD32", "#F0E68C", "#FFFFE0", "#EE82EE", "#FF6347", 
        "#6A5ACD", "#9932CC", "#8B008B", "#8B4513", "#DEB887")
    head(a, n1)
}
"36_07_16_235705" <- local({
    db1 <- "red,antiquewhite4,cyan,aquamarine,aquamarine3,cyan3,darkcyan,aquamarine4,darkorchid1,blueviolet,darkmagenta,lavender,azure3,darkkhaki,black,blue,darkslateblue,cornflowerblue,cadetblue3,darkolivegreen1,chartreuse,chartreuse3,chartreuse4,darkgoldenrod1,chocolate1,darksalmon,brown,hotpink,antiquewhite,antiquewhite3" %=>% 
        hs.gre("\\w+") %=>% unlist %=>% unique
    color.db.36_09_15_094652 <- c("red", c("antiquewhite4", "snow3", 
        "aquamarine3", "cyan3", "thistle1", "aquamarine4", "thistle3", 
        "darkmagenta", "lavender", "aquamarine", "azure3", "slategray4", 
        "darkkhaki", "cornflowerblue", "cadetblue3", "darkolivegreen1", 
        "chartreuse", "chartreuse3", "chartreuse4", "darkgoldenrod1", 
        "chocolate1", "darksalmon", "hotpink", "darkcyan", "antiquewhite", 
        "antiquewhite3", "darkslateblue", "darkorchid1", "black", 
        "cyan", "blue", "blueviolet", "brown"), c("gray"))
    function(n1 = {
    }) {
        if (length(n1)) {
            head(db1, n1)
        }
        else db1
    }
})
"2023_08_18_163104" <- local({
    point1_cm <- 2.54/72.272
    cm1_point <- 1/2.54 * 72.272
    function(size = 0.5, unit = "cm") {
        size * cm1_point
    }
})
"36_07_29_230421" <- local({
    col.db <- c("red", "blue", "green", "black")
    function(x_list, xlab, main = "", wj = "~/tmp/tmp") {
        hs.require("prettyB")
        if (length(x_list) > length(col.db)) 
            stop("【36.06.29.154428】缺颜色")
        xy <- lapply(x_list, . %>% {
            data.table(x = sort(.), y = seq.int(length(.))/length(.))
        })
        legend1 <- local({
            a1 <- names(x_list)
            paste0(a1, " (n=", sapply(x_list, length), ")")
        })
        xlim <- sapply(xy, . %>% {
            .[, range(x)]
        }) %>% as.vector %>% range
        ylim <- c(0, 1)
        x0 <- 0.25
        x1 <- 5
        x2 <- xlim[2]
        xlim2 <- xlim[2]
        xlim[2] %<>% +(x1 - x0)
        lapply(xy, . %>% {
            i1 <- .[, which(x <= x0)]
            .[i1, `:=`(x, x * x1/x0)]
            .[-i1, `:=`(x, x + x1 - x0)]
        })
        hs.browser("36.07.29.215844")
        .pdf({
            xx1 <- par(xaxt = "n")
            plot_p(0, 1, xlim = xlim, ylim = ylim, type = "n", 
                xlab = xlab, ylab = "Cumulative probability", 
                main = main)
            col.db %<>% head(length(x_list))
            col1.i <- 1
            lapply(xy, . %>% {
                col1 <- col.db[col1.i]
                points(.[, x], .[, y], type = "l", col = col1)
                col1.i <<- col1.i + 1
            })
            legend("bottomright", pch = "-", legend = legend1, 
                col = col.db)
            par(xaxt = "s")
            {
                a1 <- c(0, 0.1, 0.2, x0)
                a2 <- seq(1, floor(xlim2))
                axis(1, c(a1 * x1/x0, a2 + x1 - x0), c(a1, a2), 
                  cex.axis = 0.75)
            }
        }, wj = wj)
    }
})
"36_06_29_154254" <- local({
    col.db <- c("red", "blue", "green", "black")
    function(x_list, xlab, main = "", wj = "~/tmp/tmp") {
        hs.require("prettyB")
        if (length(x_list) > length(col.db)) 
            stop("【36.06.29.154428】缺颜色")
        xy <- lapply(x_list, . %>% {
            data.table(x = sort(.), y = seq.int(length(.))/length(.))
        })
        legend1 <- local({
            a1 <- names(x_list)
            paste0(a1, " (n=", sapply(x_list, length), ")")
        })
        xlim <- sapply(xy, . %>% {
            .[, range(x)]
        }) %>% as.vector %>% range
        ylim <- c(0, 1)
        .pdf({
            plot_p(0, 1, xlim = xlim, ylim = ylim, type = "n", 
                xlab = xlab, ylab = "Cumulative probability", 
                main = main)
            col.db %<>% head(length(x_list))
            col1.i <- 1
            lapply(xy, . %>% {
                col1 <- col.db[col1.i]
                points(.[, x], .[, y], type = "l", col = col1)
                col1.i <<- col1.i + 1
            })
            legend("bottomright", pch = "-", legend = legend1, 
                col = col.db)
        }, wj = wj)
    }
})
"36_06_18_150541" <- function(D, group.cn = colnames(D)[1], value.cn = colnames(D)[2], 
    facet.cn = {
    }, facet.args = {
    }, xlab = group.cn, ylab = value.cn, main = paste0(hs.title(geom), 
        " plot"), N = !length(facet.cols.cn), scale = "count", 
    facet.rows.cn = {
    }, facet.cols.cn = {
    }, fill.color = {
    }, color = group, geom = "violin", jitter.col = ieiwk.col2rgb("black", 
        0.1), outlier.shape = {
    }, ylim = {
    }, ylim1 = {
    }, ylim.expand = ggplot2::waiver(), percentage = all(tapply(D[, 
        value.cn], D[, group.cn], sum) == 1), dot = 1, wj2 = {
    }, Q1 = 0.5) {
    if (!identical(class(D), "data.frame")) 
        D %<=>% as.data.frame
    if (!identical(class(D), "data.frame")) 
        stop("【36.06.18.155302】需要一个 data frame")
    if (any(c(group.cn, value.cn) %not.in% colnames(D))) {
        hs.browser("36.11.24.092246")
        stop("【36.06.18.155628】缺指定的列名")
    }
    if (N) 
        D[[group.cn]] %<=>% {
            v.pct.hs(x, nl = 1, P = percentage) %=>% hs.factor(x, 
                N = 0)
        }
    i1 <- match(c(group.cn, value.cn), colnames(D))
    colnames(D)[i1] <- c("group", "value")
    geom.arg <- list()
    if (geom == "violin") 
        geom.arg[["scale"]] <- scale
    if (!missing(outlier.shape)) 
        geom.arg[["outlier.shape"]] <- outlier.shape
    geom1 <- hs.semicolon2("ggplot2", paste0("geom_", switch(geom, 
        box = "boxplot", geom)))
    p <- evalbq(ggplot2::ggplot(D, ggplot2::aes(x = group, y = value, 
        color = .(substitute(color) %=>% {
            if (is.name(x)) 
                x
            else color
        })))) + do.call(geom1, geom.arg)
    if (!missing(fill.color)) 
        p <- p + ggplot2::scale_fill_manual(values = unname(fill.color))
    if (length(facet.cn)) 
        p <- eval(bquote(p + do.call(ggplot2::facet_wrap, .(c(list(bquote(~.(as.symbol(facet.cn)))), 
            facet.args)))))
    if (length(outlier.shape) && is.na(outlier.shape) && (!length(ylim))) {
        if (0) 
            ylim <- local({
                a <- tapply(D[, "value"], D[, "group"], . %>% 
                  {
                    a <- diff(quantile(., c(0.25, 0.75))) * 1.5
                    m <- median(.)
                    c(m - a, m + a)
                  })
                c(if (length(ylim1)) ylim1 else min(sapply(a, 
                  "[", 1)), max(sapply(a, "[", 2)))
            })
        ylim <- ggplot2::ggplot_build(p)[["data"]][[1]] %=>% 
            c(if (length(ylim1)) ylim1 else min(x[, "ymin"]), 
                max(x[, "ymax"]))
    }
    if (length(ylim)) 
        p %<=>% {
            x + ggplot2::scale_y_continuous(limits = ylim, expand = ylim.expand)
        }
    p <- p + ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 50, 
        hjust = 0.5, vjust = 0.5)) + ggplot2::theme(legend.position = "none") + 
        ggplot2::xlab(xlab) + ggplot2::ylab(ylab) + ggplot2::ggtitle(main)
    if (geom != "box") 
        p <- p + ggplot2::geom_boxplot(width = 0.1, fill = NA, 
            outlier.colour = NA) + ggplot2::stat_summary(fun = . %>% 
            {
                quantile(., Q1)
            }, geom = "point", fill = "white", shape = 21, size = 2)
    if (dot) 
        p <- p + ggplot2::geom_jitter(size = 0.5, width = 0.2, 
            col = jitter.col)
    if (length(c(facet.rows.cn, facet.cols.cn))) 
        p <- eval(bquote(p + ggplot2::facet_grid(.(as.symbol(ifelse(length(facet.rows.cn), 
            facet.rows.cn, "."))) ~ .(as.symbol(ifelse(length(facet.cols.cn), 
            facet.cols.cn, "."))), scales = "free")))
    p %<=>% bg_white.hs
    if (length(wj2)) {
        .fig(print(p), wj2)
    }
    else p
}
"36_09_16_194613" <- function(go1, bg.white = 0, xlim = {
}, ylim = {
}, flip.xy = 0, reverse.x = 0, reverse.y = 0, log10.x = 0, log10.y = 0, 
    title = {
    }, xlab = {
    }, ylab = {
    }, legend.show = 1, legend.text.size = {
    }, xy_ratio = 1) {
    if (bg.white) {
        go1 <- bg_white.hs(go1)
    }
    if (length(xlim)) {
        go1 <- go1 + ggplot2::scale_x_continuous(limits = xlim)
    }
    if (length(ylim)) {
        go1 <- go1 + ggplot2::scale_y_continuous(limits = ylim)
    }
    if (reverse.x) 
        go1 <- go1 + ggplot2::scale_x_reverse()
    if (reverse.y) 
        go1 <- go1 + ggplot2::scale_y_reverse()
    if (log10.x) 
        go1 <- go1 + ggplot2::scale_x_log10()
    if (log10.y) 
        go1 <- go1 + ggplot2::scale_y_log10()
    if (length(title)) 
        go1 <- go1 + ggplot2::labs(title = title)
    if (length(xlab)) 
        go1 <- go1 + ggplot2::xlab(xlab)
    if (length(ylab)) 
        go1 <- go1 + ggplot2::ylab(ylab)
    if (!legend.show) 
        go1 <- go1 + ggplot2::theme(legend.position = "none")
    if (length(legend.text.size)) 
        go1 <- go1 + ggplot2::theme(legend.text = ggplot2::element_text(size = legend.text.size))
    if (length(xy_ratio)) 
        go1 <- go1 + ggplot2::coord_fixed(ratio = xy_ratio)
    if (0) {
        theme(axis.text.x = element_text(angle = 90, vjust = 0.05, 
            hjust = 0.75))
        scale_x_discrete(labels = c("Normal\ntissue", "Primary\ntumor"))
    }
    go1
}
"39_05_28_223622" <- function(go_list, ncol = {
}, nrow = {
}, ...) {
    if (!inherits(go_list, "list")) 
        stop("[39_05_28_224305]", substitute(go_list), "不是list，而是", 
            hs.with(class(go_list), paste(x, collapse = "、")))
    do.call("+", c(go_list, list(patchwork::plot_layout(ncol = ncol, 
        nrow = nrow, ...))))
}
"38_03_08_101528" <- function(V, f1 = 3, ot = "i", lo = 1, hi = 1) {
    med1 <- median(V)
    mad1 <- mad(V[V <= med1], med1)
    mad2 <- mad(V[V >= med1], med1)
    if (mad2 == med1) 
        mad2 <- mad(V[V > med1], med1)
    if (ot %in% "mad") 
        return(c(mad1, mad2))
    co1 <- if (lo) 
        med1 - mad1 * f1
    else -Inf
    co2 <- if (hi) 
        med1 + mad2 * f1
    else Inf
    switch(ot, l = is.na(hs.cut(V, c(co1, co2))), i = which(is.na(hs.cut(V, 
        c(co1, co2)))), co = c(co1, co2))
}
"37_05_27_182956" <- function(m1) {
    order_dt <- data.table(rownames(m1), hs.rowSums(sign(m1)), 
        hs.rowSums(m1))
    setorderv(order_dt, c("V2", "V3"), -1)
    m1[order_dt[, V1], , drop = 0]
}
"38_10_07_225455" <- function(dt, ot = "name", max1 = 13) {
    hs.switch(ot, "name", dt[, sapply(.SD, . %>% {
        n1 <- hs.clean(., U = 1, ot = "n")
        (n1 %in% c(1, if (is.character(.) || is.factor(.)) length(.))) | 
            (n1 > max1)
    }) %>% {
        names(.)[.]
    }], hs.browser("38.10.07.232027"))
}
"39_02_28_135034" <- function(size, unit = "auto") {
    class(size) <- "object_size"
    format(size, unit = "auto")
}
"36_09_09_162929" <- function(v1, n1 = 100) {
    if (n1 > length(v1)) 
        stop("36.09.09.163237")
    sapply(hs.makeBatch(seq_along(v1), n1), function(i1) {
        median(v1[i1], na.rm = 1)
    })
}
"2023_08_09_153336" <- local({
    hs1 <- function(m1) {
        (m1 - hs.rowMeans(m1))/hs.rowSds(m1)
    }
    function(m1, side = 2) {
        hs.switch(side, 1, hs1(m1), 2, t(hs1(t(m1))))
    }
})
"36_08_12_224912" <- local({
    hs1 <- function(v1, n1) {
        r1 <- range(v1[v1 > 0])
        if (r1[1] == r1[2]) {
            v1[v1 > 0] <- 1
        }
        else {
            f1 <- cut(v1[v1 > 0], seq(r1[1], r1[2], length.out = n1 + 
                1), include.lowest = 1)
            levels(f1) <- seq(1/n1, 1, length.out = n1)
            v1[v1 > 0] <- as.numeric(as.vector(f1))
        }
        v1
    }
    function(v1, n1 = 20, min2 = 0) {
        r1 <- range(v1, na.rm = TRUE)
        if (r1[2] > 0) 
            v1 <- hs1(v1, n1 = n1)
        if (r1[1] < 0) 
            v1 <- -hs1(-v1, n1 = n1)
        v1
    }
})
"38_02_17_090808" <- function(v1, min2 = 1, max2 = 10) {
    v1/max(v1) * (max2 - min2) + min2
}
"38_10_18_210035" <- function(m1, target = "auto") {
    S <- hs.switch(length(dim(m1)), 2, hs.colSums(m1), 0, sapply(m1, 
        sum))
    if (target == "auto") {
        i <- which.min(abs(S - median(S)))
        target <- S[i]
    }
    hs.switch(length(dim(m1)), 2, {
        for (j in 1:ncol(m1)) {
            f1 <- target/S[j]
            m1[, j] <- m1[, j] * f1
        }
    }, 0, {
        for (j in seq_along(m1)) {
            f1 <- target/S[j]
            m1[[j]] <- m1[[j]] * f1
        }
    })
    m1
}
"38_08_18_164953" <- function(l1, ot = "m", ...) {
    jg <- l1 %>% {
        a <- names(.)
        if (anyDuplicated(a)) 
            hs.browser("38.10.09.053752")
        b <- unlist(.) %>% unique
        m1 <- matrix(0L, length(b), length(a), dimnames = list(b, 
            a))
        hs.for(cn1, a, {
            rn1 <- .[[cn1]]
            m1[rn1, cn1] <- 1L
        })
        m1
    }
    hs.switch(ot, "m", jg, "dt", wk.dt(jg, ...))
}
"36_04_07_120420" <- function(x) {
    x <- x[!is.infinite(x)]
    shapiro.test(sample(x, min(5000, length(x))))
}
"2023_07_26_213428" <- function(fig) {
    a <- evalbq(dim(fig)[2] %=>% seq %=>% sapply(x, . %=>% {
        uniqueN(.(if (length(dim(fig)) == 2) 
            quote(fig[, x])
        else if (length(dim(fig)) == 3) 
            quote(fig[, x, ])))
    }))
    if (min(a) != 1) 
        hs.browser("2023.07.03.185023", debug = 0)
    n1 <- evalbq(which(a == min(a)) %=>% lapply(x, . %=>% {
        unique(.(hs.switch(length(dim(fig)), 2, quote(hs.ij(fig, 
            , x)), 3, quote(fig[, x, ]))))
    }) %=>% do.call("rbind", x) %=>% uniqueN)
    if (n1 != 1) 
        hs.browser("2023.07.03.185255", debug = 0)
    hs.switch(length(dim(fig)), 2, fig[1, match(1, a)], 3, {
        t(fig[1, match(1, a), ])
    })
}
"38_11_24_222417" <- function(V) {
    V <- V/max(V)
    sum(1 - V)/(length(V) - 1)
}
"2023_07_26_194158" <- function(x, y, radius, x.max = {
}, y.max = {
}) {
    r2 <- radius^2
    r <- radius %=>% seq(-x, x) %=>% expand.grid(x, x, KEEP.OUT.ATTRS = FALSE) %=>% 
        as.matrix
    r %<=>% hs.ij(x, hs.rowSums(x^2) <= r2)
    r[, 1] <- r[, 1] + x
    r[, 2] <- r[, 2] + y
    i <- r < 1
    if (any(i)) 
        r[i] <- 1
    if (length(x.max)) {
        i <- which(r[, 1] > x.max)
        if (length(i)) 
            r <- hs.ij(r, -i)
    }
    if (length(y.max)) {
        i <- which(r[, 2] > y.max)
        if (length(i)) 
            r <- hs.ij(r, -i)
    }
    r
}
"38_12_09_153556" <- function(f1, N = 0) {
    f1 %=>% {
        x <- paste(x)
        x <- factor(x, names(sort(hs.table(x))))
        b <- hs.table(x)[levels(x)]
        if (N) 
            levels(x) <- paste0(names(b), " (n=", b, "; ", round(b/sum(b) * 
                100, 2), "%)")
        x
    }
}
"39_05_11_100447" <- function() {
    wjj2 <- hs.sjgly("39_05_11_100447")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        h5_wj <- hs.hsgly("表达量/39_05_02_090920")()
        sce_list <- hs.hsgly("10x2seurat.38_08_12_145141")(h5_wj)
        {
            sce_list[["ONB_311_20210824NA_on_human+mouse"]]@meta.data["AAACCCAAGGTAAGAG-1", 
                ]
            sce_list[["ONB_311_20210824NA_on_human+mouse"]] %=>% 
                rownames %=>% hs.re("[^-]*") %=>% unique
            a <- wk.fread("~/swxx/xm/db/39/4/24/pdx单细胞/表达量/ONB_311_20210824NA/outs/analysis/gem_classification.csv")
            a <- wk.fread("~/swxx/xm/db/39/4/24/pdx单细胞/表达量/ONB_311_PDX_Control_20220107NA/outs/analysis/gem_classification.csv")
            a[, call %=>% hs.table(type = "b")]
            if (0) {
                a <- h5_wj[sample_name_2do] %=>% sapply(x, . %=>% 
                  {
                    dirname(x) %=>% hs.fp("analysis/gem_classification.csv")
                  }) %=>% lapply(x, wk.fread)
                lapply(a, . %=>% {
                  x[, call %=>% hs.table(type = "b", percentage = 1)]
                })
            }
            wk.fread("~/swxx/xm/db/39/4/24/pdx单细胞/表达量/ONB_311_20210824NA/outs/metrics_summary.csv") %=>% 
                x[, .SD, .SDcols = patterns("(?i)cell")]
            sample_name_2do <- sce_list %=>% names %=>% (`%not%`("ONB_311_20210824NA"))
            for (sample_name_1 in sample_name_2do) {
                p1 <- local({
                  if (0) {
                    p1 <- sce_list[[sample_name_1]] %=>% Seurat::PercentageFeatureSet("^GRCh38-")
                    {
                      if (0) {
                        names(p1) %=>% unique %=>% sapply(x, 
                          . %=>% {
                            i1 <- which(names(p1) %in% x)
                            (length(i1) == 2) && (p1[i1[1]] == 
                              p1[i1[2]])
                          }) %=>% all
                      }
                      p1 %<=>% x[match(colnames(sce_list[[sample_name_1]]), 
                        names(x))]
                    }
                    return(p1)
                  }
                  {
                    ri_human <- sce_list[[sample_name_1]]@assays[["RNA"]]@counts %=>% 
                      rownames %=>% hs.grep("^GRCh38-")
                    ri_mouse <- sce_list[[sample_name_1]] %=>% 
                      nrow %=>% seq %=>% (`%not%`(ri_human))
                    a <- sce_list[[sample_name_1]]@assays[["RNA"]]@counts[ri_human, 
                      ] %=>% hs.colSums
                    b <- sce_list[[sample_name_1]]@assays[["RNA"]]@counts[ri_mouse, 
                      ] %=>% hs.colSums
                    return(a/(a + b) * 100)
                  }
                })
                sce_list[[sample_name_1]]@meta.data[, "human%"] <- p1
            }
            wj2 <- wjj2 %=>% hs.fp("人含量分布.pdf")
            a <- lapply(sample_name_2do, . %=>% {
                a <- h5_wj[x] %=>% dirname %=>% hs.fp("analysis/gem_classification.csv") %=>% 
                  wk.fread
                list(sample = x, `human%` = sce_list[[x]]@meta.data[, 
                  "human%"], org = a[match(colnames(sce_list[[x]]), 
                  barcode), call]) %=>% wk.dt.as
            })
            a <- rbindlist(a)
            a[, `:=`("mouse%", 100 - `human%`)]
            if (0) {
                p1 <- ggplot2::ggplot(a, ggplot2::aes(x = `human%`, 
                  y = `mouse%`, color = org)) + ggplot2::geom_point() + 
                  facet_grid(. ~ sample)
                height <- 7
                .pdf(print(p1), wj2, width = a[, sample %=>% 
                  uniqueN] * height, height = height)
                "~/swxx/xm/db/39/4/24/pdx单细胞/表达量/质控/人含量分布.pdf" %=>% 
                  wk.rsync.in(ssh = "r820")
            }
            a[, org %=>% hs.table]
            hs.hsgly("小提琴图/36.06.18.150541")(a, wj2 = wj2, 
                jitter.col = a[, c(GRCh38 = ieiwk.col2rgb("red", 
                  0.2), mm10 = ieiwk.col2rgb("blue", 0.2), Multiplet = ieiwk.col2rgb("gray", 
                  0.5))[org]])
            a[org == "mouse", hs.table(sample, type = "b")]
            if (0) {
                a <- c(0.5, 1, 2, 5, 50) %=>% hs.lapply(x, function(p1) {
                  a <- hs.lapply(sample_name_2do, function(sample_name_1) {
                    x <- sce_list[[sample_name_1]]@meta.data[, 
                      "human%"]
                    i <- sce_list[[sample_name_1]]@meta.data[, 
                      "human%"] %=>% {
                      (100 - x) > p1
                    }
                    c(sum(i), mean(i) * 100)
                  })
                  wk.dt.fromList(a, by_row = 1, nm = c("sample", 
                    "#cell", "%cell"))
                })
                a <- wk.dt.fromList(a, nm = "mouse%")
                .s(a)
            }
        }
        if (0) {
            seurat_data <- hs.hsgly("合并数据.38_12_09_141645")(sce_list)
            seurat_data %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
                normalization.method = "log", nfeatures = 2000)
            seurat_data %<=>% hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")(x, 
                reduction = "pca", dim_n = 10, wjj2 = wjj2)
            p <- Seurat::DimPlot(seurat_data, reduction = "umap", 
                shuffle = TRUE, group.by = "orig.ident", raster = TRUE)
            .pdf(print(p), width = 20, wj = hs.fp(wjj2, "umap_样本_不去批次.pdf"))
            "~/swxx/xm/db/39/4/24/pdx单细胞/表达量/质控/全貌/umap_样本_不去批次.pdf" %=>% 
                wk.rsync.in(ssh = "r820")
            seurat_data %<=>% hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")(x, 
                sample_cn = "orig.ident", dim_n = 10, wjj2 = wjj2)
            p <- Seurat::DimPlot(seurat_data, reduction = "umap", 
                shuffle = TRUE, group.by = "orig.ident", raster = TRUE)
            .pdf(print(p), width = 20, wj = hs.fp(wjj2, "harmony_umap_样本.pdf"))
            p <- Seurat::DimPlot(seurat_data, reduction = "umap", 
                shuffle = TRUE, group.by = "orig.ident", raster = TRUE)
            `?`(Seurat::DimPlot)
            seurat_data@meta.data
            .pdf(print(p), width = 20)
            "~/swxx/xm/db/39/4/24/pdx单细胞/表达量/质控/全貌/harmony_umap_样本.pdf" %=>% 
                wk.rsync.in(ssh = "r820")
            {
                seurat_data@meta.data[1, ]
                seurat_data
            }
            seurat_data %=>% str
            .ls(wjj2)
            hs.hsgly("转录组/标准流程/模块/预处理/整合、降维.38_10_09_113753")(sce_list)
        }
        {
            wjj2_here <- hs.fp(wjj2, "人")
            {
                a <- sce_list %=>% lapply(x, . %=>% {
                  if (any(hs.grepl(rownames(x), "^(?i)GRCh38-"))) {
                    x <- hs.ij(x, rownames(x) %=>% hs.grep("^(?i)GRCh38-"))
                    old_new <- rownames(x) %=>% wk.dt(x, hs.grep_sub(x, 
                      "^(?i)GRCh38-"))
                    hs.hsgly("seurat对象改基因名.39_05_16_095300")(x, 
                      old_new = old_new)
                  }
                  else x
                })
                seurat_data_human <- hs.hsgly("合并数据.38_12_09_141645")(a)
                seurat_data_human %=>% dim
                seurat_data_human@meta.data[1, ]
                {
                  cn1 <- "cell_ranger_call"
                  seurat_data_human@meta.data[, cn1] <- "NA"
                  a <- hs.hsgly("表达量/39_05_02_090920")() %=>% 
                    {
                      hs.c(names(x), hs.fp(dirname(x), "analysis/gem_classification.csv")) %=>% 
                        x[file.exists(x)]
                    }
                  names(a) %all.in% seurat_data_human@meta.data[, 
                    "orig.ident"]
                  for (sample1 in names(a)) {
                    b <- a[sample1] %=>% wk.fread
                    paste0(sample1, "_", b[, barcode])
                    seurat_data_human@meta.data[paste0(sample1, 
                      "_", b[, barcode]), cn1] <- b[, call]
                  }
                }
                seurat_data_human %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
                  normalization.method = "log", nfeatures = 2000, 
                  wjj2 = wjj2_here)
                seurat_data_human %=>% hs.hsgly("seurat/函数/idents，获取.38_12_15_114046")(x, 
                  id = "orig.ident")
                {
                  pt.size <- 0.1
                  reduction1 <- "pca"
                  reduction1 <- "harmony"
                  for (reduction1 in c("pca", "harmony")) {
                    wj2_all <- c("重叠", "分开") %=>% hs.c(x, 
                      hs.fp(wjj2_here, paste0("umap_", switch(reduction1, 
                        pca = "保留", harmony = "去除"), 
                        "批次.样本.", x, ".pdf")))
                    if (!all(file.exists(wj2_all))) 
                      seurat_data_human %<=>% hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")(x, 
                        sample_cn = "orig.ident", reduction = reduction1, 
                        dim_n = 10, wjj2 = wjj2_here)
                    wj2 <- wj2_all["重叠"]
                    if (!file.exists(wj2)) {
                      p1 <- seurat_data_human %=>% hs.hsgly("单细胞/转录组/seurat/函数/图/低维.38_12_11_113923")(x, 
                        shuffle = 1, first_is_gray = 0)
                      p1 %<=>% hs.hsgly("ggplot2.post.36_09_16_194613")(x)
                      .pdf(p1, wj2, height = 25)
                    }
                    wj2 <- wj2_all["分开"]
                    if (!file.exists(wj2)) {
                      p1 <- seurat_data_human %=>% hs.hsgly("单细胞/转录组/seurat/函数/图/低维.38_12_11_113923")(x, 
                        shuffle = 1, first_is_gray = 0, split.by = "orig.ident", 
                        combine = FALSE)
                      class(p1[[1]])
                      p1 %<=>% hs.hsgly("list.patchwork.plot_layout.39_05_28_223622")(x, 
                        ncol = 3)
                      p1 %<=>% hs.hsgly("ggplot2.post.36_09_16_194613")(x)
                      h1 <- 30
                      .fig(p1, wj2, height = h1 * 2, width = h1 * 
                        3)
                    }
                  }
                }
                {
                  h1 <- 30
                  p1 <- Seurat::DimPlot(seurat_data_human, shuffle = TRUE, 
                    raster = TRUE, cols = "BrBG", pt.size = pt.size)
                  .pdf(print(p1), width = h1 * 1.5, height = h1)
                  p1 <- Seurat::DimPlot(seurat_data_human, shuffle = TRUE, 
                    raster = TRUE, cols = "BrBG", split.by = "orig.ident")
                  h1 <- 10
                  .pdf(print(p1), width = h1 * uniqueN(seurat_data@meta.data[, 
                    "orig.ident"]), height = h1)
                  `?`(Seurat::DimPlot)
                  help(package = "Seurat")
                  seurat_data_human@meta.data[, "mouse."] %<=>% 
                    {
                      if (anyNA(x)) {
                        x[is.na(x)] <- 0
                        x
                      }
                      else x
                    }
                  hs.require("patchwork")
                  {
                    cn1 <- "mouse_pct"
                    seurat_data_human@meta.data[, cn1] <- seurat_data_human@meta.data[, 
                      "mouse."]
                    h1 <- 15
                    p1 <- Seurat::FeaturePlot(seurat_data_human, 
                      features = cn1, pt.size = pt.size, raster = TRUE, 
                      split.by = "orig.ident", keep.scale = "all", 
                      order = TRUE, combine = FALSE)
                    seurat_data_human@meta.data[, cn1] <- {
                    }
                    p1 %<=>% {
                      cmd1 <- paste0("x[[ ", seq_along(x), "]]") %=>% 
                        c("patchwork::plot_layout( ncol = 3)") %=>% 
                        paste(collapse = " + ") %=>% parse(text = x)
                      eval(cmd1)
                    }
                    .pdf(print(p1), width = h1 * 3 * 1.2, height = h1 * 
                      2)
                    cell2do <- seurat_data_human@meta.data %=>% 
                      rownames(x)[x[, "cell_ranger_call"] %not.in% 
                        c("mm10", "Multiplet")]
                    p1 <- Seurat::FeaturePlot(seurat_data_human, 
                      features = cn1, cells = cell2do, pt.size = pt.size, 
                      raster = TRUE, split.by = "orig.ident", 
                      keep.scale = "all", order = TRUE, combine = FALSE)
                    p1 %<=>% {
                      cmd1 <- paste0("x[[ ", seq_along(x), "]]") %=>% 
                        c("patchwork::plot_layout( ncol = 3)") %=>% 
                        paste(collapse = " + ") %=>% parse(text = x)
                      eval(cmd1)
                    }
                    .pdf(print(p1), width = h1 * 3 * 1.2, height = h1 * 
                      2)
                  }
                }
                h1 <- 10
                p1 <- Seurat::DimPlot(seurat_data_human, split.by = "orig.ident")
                .pdf(print(p1), width = h1 * uniqueN(seurat_data_human@meta.data[, 
                  "orig.ident"]), height = h1, wj = wj2)
                seurat_data_human
                "~/swxx/xm/db/39/4/24/pdx单细胞/表达量/质控/全貌/umap_样本_不去批次.pdf" %=>% 
                  wk.rsync.in(ssh = "r820")
                "~/swxx/xm/db/39/4/24/pdx单细胞/表达量/质控/全貌/umap_样本_去批次.pdf" %=>% 
                  wk.rsync.in(ssh = "r820")
            }
            seurat_data <- hs.hsgly("合并数据.38_12_09_141645")(sce_list)
            {
                seurat_data@meta.data[1, ]
                help(package = "Seurat")
                p1 <- Seurat::DimPlot(seurat_data_human, split.by = "orig.ident")
                seurat_data_human %=>% str
                h1 <- 10
                .pdf(print(p1), width = h1 * uniqueN(seurat_data@meta.data[, 
                  "orig.ident"]), height = h1)
                "/dev/shm/wk/tmp.pdf" %=>% wk.rsync.in(ssh = "r820")
                `?`(Seurat::DimPlot)
            }
            seurat_data %=>% dimnames
            sce_list[[1]] %=>% dimnames
            sce_list %=>% names
            sce_list[["ONB_311_20210824NA"]] %=>% dimnames
            sce_list[["ONB_311_20210824NA_on_human+mouse"]] %=>% 
                rownames %=>% hs.grep_sub("^(?i)GRCh38-") %=>% 
                hs.xor(sce_list[["ONB_311_20210824NA"]] %=>% 
                  rownames)
        }
        sce_list[[1]] %=>% .str
        sce_list <- hs.hsgly("标准流程/模块/预处理/质控/过滤细胞、基因.38_10_19_144214")(seurat_data_list = sce_list, 
            wjj2 = hs.fp(wjj2, "mid", "qc_gene,cell,sample"))
        hs.hsgly("标准流程/模块/预处理/质控/过滤细胞、基因.图表.38_10_19_130420")(seurat_data_list, 
            wjj2 = wjj2, org1 = org1)
        hs.hsgly("1.流程/单细胞/转录组/标准流程/主函数.37_12_09_143244")
    }
}
"37_07_25_182333" <- function(real = if (length(real.root)) hs.wjj(real.root, 
    "db/swxx"), real.root = "/Volumes/tf500" %=>% x[file.exists(x)][1] %=>% 
    hs.clean) {
    target <- "~/swxx"
    if (file.exists(target)) {
        if (0) {
            if (file.is_symlink(target)) {
                if (length(real)) {
                  if (Sys.readlink(target) == real) {
                  }
                  else {
                    hs.browser("37.07.25.211922")
                  }
                }
                else {
                  hs.browser("37.07.25.211859")
                }
            }
            else {
                hs.browser("37.07.25.211855")
            }
        }
    }
    else {
        if (length(real)) {
            file.symlink(real, target)
        }
        else {
            hs.browser("37.07.25.211945")
        }
    }
    target
}
"36_08_27_181922" <- function(..mem = 1) {
    wjj.36_08_27_182012 <- .mfp.swxx("36.08.27.182007")
    wj <- hs.wj(wjj.36_08_27_182012, "gene_coding.tsv.gz")
    if (file.exists(wj)) {
    }
    else {
        a1 <- fread(cmd = .sys("cat", "~/in2/swxx/sj/db/ensembl/human/Homo_sapiens.GRCh38.99.gtf", 
            "|", .sys.perl.print("print $F[ 8] if $F[ 2] eq \"gene\" and $F[ 8] =~ /gene_biotype \"protein_coding\"/"), 
            "| less -Nir"), header = FALSE, sep = ";")
        a1[, `:=`("V6", {
        })]
        setnames(a1, hs.re(a1[1], "\\w+"))
        a1[, `:=`(c("gene_version", "gene_source", "gene_biotype"), 
            {
            })]
        for (cn1 in names(a1)) {
            a1[, `:=`(c(cn1), hs.re(get(cn1), "(?<=\")[^\"]+"))]
        }
        setnames(a1, paste0("gene.id_", c("ensembl", "symbol")))
        wk.dt.fwrite(a1, wj)
    }
    fread(wj)
}
"38_12_05_100436" <- function(org1 = "human", gene_id_type = "ensembl", 
    ..mem = 1) {
    if (org1 %not.in% c("human", "macaque")) 
        hs.browser("38.12.05.100510", debug = 0)
    gtf_wj <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1 = org1, 
        what = "gtf")
    a <- hs.hsgly("1.信息学/文件/读写/提取/表格的子集.38_09_10_214239")(v1 = "gene", 
        db.fp = gtf_wj, ci1 = 3)
    a[, `:=`(gene_name, hs.re(V9, "(?<=gene_name \")[^\"]+", 
        no_match = ""))]
    a1 <- a[(V1 %in% "Y") | (gene_name %in% "XIST")]
    hs.switch(gene_id_type, "ensembl", a1[, V9 %=>% hs.re("(?<=gene_id \")[^\"]*")], 
        "symbol", a1[, hs.clean(gene_name, U = 1)])
}
"36_04_04_115908" <- function(..mem = 1) {
    sj <- hs.hsgly("tf,AnimalTFDB.wj.36.04.02.122143")()
    a1 <- seq_along(sj) %>% lapply(. %>% {
        a1 <- sj[[.]]
        a1[, `:=`("type", names(sj)[.] %>% hs.sub("Homo_sapiens_"))]
        setnames(a1, "Ensembl", "gene.id_ensembl")
        hs.hsgly("gene.id.converter.using_hgnc.36.01.22.160941")(a1, 
            "gene.id_ensembl", "gene.id_symbol", "ensembl", "symbol")
        a1[, .(gene.id_ensembl, gene.id_symbol, Family, type)]
    })
    rbindlist(a1)
}
"36_04_03_232353" <- function(wj) {
    wjj.36_03_04_003344 <- getwd()
    on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
    setwd(dirname(wj))
    nr <- readLines(wj)
    dz <- hs.re(nr, "http[^\\\"]+") %>% {
        .[hs.grep(basename(.), "cosine|processing_scripts\\.tar\\.gz", 
            inv = 1)]
    }
    if (any(file.exists(basename(dz)))) 
        return()
    sapply(dz, . %>% {
        .wk("wget -c", .)
    })
}
"36_04_04_100227" <- local({
    wjj <- "~/in2/swxx/sj/db/Harmonizome"
    sj.wj <- {
    }
    function(what = {
    }, ..mem = 1) {
        if (!length(sj.wj)) {
            sj.wj <<- hs.wjj.ls(wjj, "gene_set_library_crisp.gmt.gz", 
                R = 1)
            names(sj.wj) <<- sj.wj %>% hs.sub(paste0(wjj, "/")) %>% 
                dirname %>% hs.gsub("/", ".")
        }
        if (!length(what)) 
            return(sj.wj)
        sj.wj %<>% {
            switch(what, ..all = ., .[names(.) %and% what])
        }
        lapply(sj.wj, function(wj1) {
            a1 <- readLines(wj1)
            a1 %>% strsplit("\t") %>% {
                hs.c(sapply(., "[", 1), sapply(., tail, -2))
            }
        })
    }
})
"36_10_28_201631" <- function(..mem = 1) {
    a1 <- wk.xls.read("~/in2/swxx/sj/db/HumanTFs/DatabaseExtract_v_1.01.xlsx")
    a1[, `:=`("gene.id_symbol", hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937")(`HGNC symbol`))]
    a1[, gene.id_symbol %>% hs.grepV(";")]
    a1[, `Is TF?` %>% hs.table]
    return(a1[`Is TF?` %in% "Yes", gene.id_symbol])
    jg <- readLines("~/in2/swxx/sj/db/HumanTFs/TF_names_v_1.01.txt")
    jg <- hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937")(jg)
    if (0) {
        jg %>% {
            .[names(.) != .]
        }
        hs.grepV(jg, ";")
    }
    unname(jg)
}
"36_07_10_093656" <- function(gft_wj = "~/in2/swxx/sj/db/ensembl/human/Homo_sapiens.GRCh38.99.gtf") {
    jg_wj <- hs.wjm(gft_wj, ext = "gene.tsv.gz")
    if (file.exists(jg_wj)) {
        fread(jg_wj)
    }
    else {
        sj <- fread(cmd = .sys("cat", gft_wj, "|", .sys.perl.print("print $F[ 8] if $F[ 2] eq \"gene\"")), 
            select = c(1, 3:5))
        setnames(sj, c("gene.id_ensembl", "gene.id_symbol", "gene.source", 
            "gene.biotype"))
        lapply(names(sj), function(cn1) {
            sj[, `:=`(c(cn1), hs.re(get(cn1), "(?<=\")[^\"]+"))]
        })
        sj[1] %>% .s
        sj[, hs.table(gene_biotype) %>% sort %>% tail]
        wk.dt.fwrite(sj, jg_wj)
        sj
    }
}
"39_01_09_175801" <- function(org1 = "human", ..mem = 1) {
    gtf_wj <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1 = org1, 
        what = "gtf")
    a <- hs.hsgly("1.信息学/文件/读写/提取/表格的子集.38_09_10_214239")(v1 = "gene", 
        db.fp = gtf_wj, ci1 = 3)
    a[, `:=`(gene_id_ens, V9 %=>% hs.re.capture(x, "gene_id \"([^\"]*)\""))]
    a[, `:=`(chr, paste0("chr", V1))]
    wk.dt.2df(a, rn.cn = "gene_id_ens", cn = c("chr", "V4", "V5"))
}
"39_01_09_184034" <- function(sub = {
}) {
    jg <- .mfp.swxx("39_01_09_183726")
    if (length(sub)) 
        jg <- hs.fp(jg, sub)
    jg
}
"39_01_09_185255" <- function() {
    hs.require("infercnv")
    a <- system.file("extdata", "oligodendroglioma_annotations_downsampled.txt", 
        package = "infercnv") %=>% wk.fread(x, cn = 0)
    a[, V2 %=>% hs.table]
}
"37_12_09_214836" <- function() {
    wjj.37_12_09_223410 <- .mfp.swxx("37_12_09_223406")
    wj_in <- hs.with_wd(hs.fp(wjj.37_12_09_223410, "in"), {
        dz <- c("https://cf.10xgenomics.com/samples/cell-exp/2.1.0/pbmc4k/pbmc4k_filtered_gene_bc_matrices.tar.gz", 
            "https://cf.10xgenomics.com/samples/cell-exp/2.1.0/pbmc8k/pbmc8k_filtered_gene_bc_matrices.tar.gz", 
            "https://cf.10xgenomics.com/samples/cell-exp/6.1.0/10k_PBMC_3p_nextgem_Chromium_X/10k_PBMC_3p_nextgem_Chromium_X_molecule_info.h5", 
            "https://cf.10xgenomics.com/samples/cell-exp/6.1.0/10k_PBMC_3p_nextgem_Chromium_X/10k_PBMC_3p_nextgem_Chromium_X_filtered_feature_bc_matrix.h5", 
            "https://cf.10xgenomics.com/samples/cell-exp/6.1.0/10k_PBMC_3p_nextgem_Chromium_X/10k_PBMC_3p_nextgem_Chromium_X_filtered_feature_bc_matrix.tar.gz", 
            "https://cf.10xgenomics.com/samples/cell-exp/6.1.0/10k_PBMC_3p_nextgem_Chromium_X/10k_PBMC_3p_nextgem_Chromium_X_analysis.tar.gz", 
            "https://cf.10xgenomics.com/samples/cell-exp/6.1.0/10k_PBMC_3p_nextgem_Chromium_X/10k_PBMC_3p_nextgem_Chromium_X_metrics_summary.csv", 
            "https://cf.10xgenomics.com/samples/cell-exp/6.1.0/10k_PBMC_3p_nextgem_Chromium_X/10k_PBMC_3p_nextgem_Chromium_X_web_summary.html", 
            "https://cf.10xgenomics.com/samples/cell-exp/6.1.0/10k_PBMC_3p_nextgem_Chromium_X/10k_PBMC_3p_nextgem_Chromium_X_cloupe.cloupe")
        for (dz1 in dz) {
            if (!file.exists(basename(dz1))) 
                .wk("wget -c", dz1)
        }
        normalizePath(basename(dz))
    })
    if (0) {
        filtered_feature_bc_matrix.tar.gz <- hs.grepV(wj_in, 
            "(?i)filtered_(gene|feature)_bc_matri(x|ces).tar.gz")
        sapply(filtered_feature_bc_matrix.tar.gz, wk.untar)
    }
    hs.wjj.ls(dirname(wj_in[1]), "(?i)filtered_(gene|feature)_bc_matri(x|ces)", 
        wjj = 1) %>% sapply(. %>% hs.wjj.ls(R = 1) %>% {
        dirname(.[1])
    }) %>% unname
}
"38_09_14_114502" <- function() "~/swxx/rj/10x/cellranger"
"38_04_13_090722" <- function(org1 = "rat", ot = "") {
    if (org1 %not.in% c("human", "mouse", "rat", "sheep", "sheep (texel)", 
        "goat (black bengal)", "goat")) 
        hs.browser("38.04.13.145756")
    gtf.gz <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1 = org1, 
        what = "gtf")
    .ls(gtf.gz)
    hs.switch(ot, "gtf", return(gtf.gz))
    genome <- hs.wjm(gtf.gz, ext = 0, extN = 2, dn = 0)
    wjj2 <- "~/swxx/rj/10x/cellranger"
    genome.wjj <- hs.fp(wjj2, genome)
    if (file.exists(genome.wjj)) 
        return(genome.wjj)
    fa.gz <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1 = org1, 
        what = "genome_sequence")
    .ls(fa.gz)
    if (0) {
        fa.gz %>% .ls
        fa.gz %>% hs.wj.rm
        wk.rsync.out("/home/wk/wk/tmp/Ovis_aries.Oar_v3.1.dna.toplevel.fa.gz", 
            "/cluster/home/kweng/swxx/sj/db/ensembl/ftp.ensembl.org/pub/release-107/fasta/ovis_aries/dna", 
            ssh = "n9")
    }
    cr <- hs.hsgly("1.流程/单细胞/转录组/10x/cr/cr.38_09_14_113545")()
    if (0) {
        "~/swxx/rj/10x/cellranger" %>% .ls
        .wk(cr, "mkref -h")
    }
    if (0) {
        a <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1 = "human", 
            what = "gtf")
        a <- gtf.gz
        .less(a)
        .wk("cat_wk", a, "| perl -e", .q("while( <>) { @F = split( \"\t\"); if( ( $F[ 2] eq \"gene\") || ( $F[ 2] eq \"transcript\")) { print}}"), 
            "| less -Ni")
        gene_biotype <- .wk("cat_wk", a, "| perl -e", .q("\nwhile( <>) {\n  @F = split( \"\t\");\n  if( $F[ 2] eq \"gene\") {\n    if( /gene_biotype \"([^\"]+)/) {\n      print( $1, \"\n\");\n    }\n  }\n}\n"), 
            intern = 1) %>% hs.table %>% {
            .[order(tolower(names(.)))]
        }
        gene_biotype %>% sort
        transcript_biotype <- .wk("cat_wk", gtf, "| perl -e", 
            .q("\nwhile( <>) {\n  @F = split( \"\t\");\n  if( $F[ 2] eq \"transcript\") {\n    if( /transcript_biotype \"([^\"]+)/) {\n      print( $1, \"\n\");\n    }\n  }\n}\n"), 
            "|", .sys.uniq(), intern = 1) %>% {
            .[order(tolower(.))]
        }
        tag <- .wk("cat_wk", a, "| perl -e", .q("\nwhile( <>) {\n  @F = split( \"\t\");\n  if( $F[ 2] eq \"transcript\") {\n    if( /tag \"([^\"]+)/) {\n      print( $1, \"\n\");\n    }\n  }\n}\n"), 
            intern = 1) %>% hs.table %>% {
            .[order(tolower(names(.)))]
        }
    }
    wjj_tmp <- if (cluster.mem()["Available"] < 64) {
        "/share/38_04_29_192825"
    }
    else "/dev/shm/38_04_13_093435"
    on.exit(hs.wj.rm(wjj_tmp, 1), add = TRUE)
    hs.with_wd(hs.wjj(wjj_tmp), {
        hs.browser("2023.06.17.172740", debug = 0)
        gtf <- hs.wjm(gtf.gz, dn = 0, ext = 0)
        if (!file.exists(gtf)) 
            .wk("cat_wk", .sys.fpGood(gtf.gz), ">", .sys.fpGood(gtf))
        gtf_filtered <- hs.wjm(gtf, append = ".filtered", dn = 0)
        .head(gtf)
        gene_biotype <- .sys.cat.text(gtf) %=>% .wk("| perl -e", 
            .q("\nwhile( <>) {\n  @F = split( \"\t\");\n  if( $F[ 2] eq \"gene\") {\n    if( /gene_biotype \"([^\"]+)/) {\n      print( $1, \"\n\");\n    }\n  }\n}\n"), 
            intern = 1) %=>% hs.table %=>% x[order(tolower(names(x)))]
        if (0) {
            gene_biotype %>% sort
            tolower(names(gene_biotype)) %and% tolower(c("protein_coding", 
                "lincRNA", "lncRNA", "antisense", "IG_LV_gene", 
                "IG_V_gene", "IG_V_pseudogene", "IG_D_gene", 
                "IG_J_gene", "IG_J_pseudogene", "IG_C_gene", 
                "IG_C_pseudogene", "TR_V_gene", "TR_V_pseudogene", 
                "TR_D_gene", "TR_J_gene", "TR_J_pseudogene", 
                "TR_C_gene"))
            tolower(names(gene_biotype)) %not% tolower(c("protein_coding", 
                "lincRNA", "lncRNA", "antisense", "IG_LV_gene", 
                "IG_V_gene", "IG_V_pseudogene", "IG_D_gene", 
                "IG_J_gene", "IG_J_pseudogene", "IG_C_gene", 
                "IG_C_pseudogene", "TR_V_gene", "TR_V_pseudogene", 
                "TR_D_gene", "TR_J_gene", "TR_J_pseudogene", 
                "TR_C_gene"))
            .wk(cr, "mkgtf -h")
        }
        hs.msg("过滤gtf")
        .wk(cr, "mkgtf -h")
        .wk(cr, "mkgtf", .sys.fpGood(gtf), .sys.fpGood(gtf_filtered))
        .wk(cr, "mkgtf", .sys.fpGood(gtf), .sys.fpGood(gtf_filtered), 
            paste0("--attribute=gene_biotype:", names(gene_biotype)))
        hs.msg()
        {
            fa <- hs.wjm(fa.gz, dn = 0, ext = 0)
            if (!file.exists(fa)) {
                hs.msg(paste0("解压", fa.gz))
                .wk("cat_wk", .sys.fpGood(fa.gz), ">", .sys.fpGood(fa))
                hs.msg()
            }
        }
        hs.msg("建索引")
        .wk(cr, "mkref", "--nthreads=10", "--memgb=16", paste0("--genome=", 
            .q(genome)), paste0("--fasta=", .q(fa)), paste0("--genes=", 
            .q(gtf)), "--ref-version=1.0.0")
        hs.msg("复制索引")
        hs.wj.copy(genome, wjj2)
        hs.msg()
    })
    genome.wjj
}
"37_12_04_134421" <- function(fq = {
}, id, cmd = "count", ec = 3000, nc = 10, mem = {
}, help = 0, jg_wjj = ".", ..., org1 = "human") {
    wjj2 <- "~/swxx/rj/10x/cellranger"
    ref_wjj <- hs.with_wd(wjj2, {
        ref_all <- c(human = "https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2020-A.tar.gz", 
            mouse = "https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-mm10-2020-A.tar.gz", 
            `human&mouse` = "https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-and-mm10-2020-A.tar.gz")
        switch(org1, human = , mouse = , `human&mouse` = ref_all[org1], 
            all = ref_all) %>% sapply(. %>% {
            wj2 <- basename(.)
            wjj2 <- hs.wjm(wj2, ext = 0, extN = 2)
            if (!dir.exists(wjj2)) {
                wj.updater(dz1 = ., co1 = Inf)
                .wk("tar -xzvf", wj2)
            }
            hs.wj.portable(wjj2)
        })
    })
    if (org1 == "all") 
        return(wjj2)
    cr <- hs.with_wd(wjj2, {
        wj2 <- "cellranger-6.1.2.tar.gz"
        wjj2 <- hs.wjm(wj2, ext = 0, extN = 2)
        if (!dir.exists(wjj2)) {
            .wk("axel", "-o", .q(wj2), "-c", .q("https://cf.10xgenomics.com/releases/cell-exp/cellranger-6.1.2.tar.gz?Expires=1638822337&Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cHM6Ly9jZi4xMHhnZW5vbWljcy5jb20vcmVsZWFzZXMvY2VsbC1leHAvY2VsbHJhbmdlci02LjEuMi50YXIuZ3oiLCJDb25kaXRpb24iOnsiRGF0ZUxlc3NUaGFuIjp7IkFXUzpFcG9jaFRpbWUiOjE2Mzg4MjIzMzd9fX1dfQ__&Signature=iHq0TGe2CsS28nDtePaYc-ud1SUoL9ErxcxQbDHRg6aTzn2XgCN-5ApQXzRPv8~Sbe24icxbSX15sRqHL94kK3sIL87nB1JveF0w24ucG0G5AjqGSBBY2HVTi4uW1y4G6ab-0ihstCCGRgx5pwL5FoLMiCbhdzXcTBE5AKcwBPoIfjHJhwc7uAJl9DwTWEWiJJ0JP3shI7xSVPou9A12jrG-lnFmW3nDIXd3Ai1u8bivmnsBFupl3NYImDxFtQDIY8NadvZyqtp6XlGLIwGPkN5sEUdpYxavbIz~v9RYCtAeUkyuVRopqCocl1G3j43NiWjDJuf4dI4fLFCxe6~4nw__&Key-Pair-Id=APKAI7S6A5RYOXBWRPDA"))
            .wk("tar -xzvf", wj2)
        }
        hs.fp(hs.wj.portable(wjj2), "bin", "cellranger")
    })
    .wk(cr, "count --help")
    hs.with_wd(wjj2, {
        wj2 <- "sitecheck.txt"
        if (!file.exists(wj2)) {
            .wk(cr, "sitecheck", ">", wj2)
        }
        wjj2 <- "tiny"
        if (!dir.exists(wjj2)) {
            .wk(cr, "testrun", paste0("--id=", wjj2))
            .ls(hs.fp(wjj2, "outs/filtered_feature_bc_matrix/"))
        }
    })
    if (0) {
        .wk(cr, "--help")
        .wk(cr, "count", "--help")
    }
    if (help) {
        .wk(cr, cmd, "--help | less -Nir")
        return()
    }
    hs.with_wd(jg_wjj, {
        if (!file.exists(id)) 
            .wk(cr, cmd, "--id", id, "--transcriptome", .sys.fpGood(ref_wjj), 
                "--disable-ui", "--nosecondary", "--localcores", 
                nc, if (length(mem)) 
                  c("--localmem", mem), "--expect-cells", ec, 
                if (length(fq)) 
                  c("--fastqs", paste(.sys.fpGood(fq), collapse = ",")), 
                ...)
        normalizePath(id)
    })
}
"38_09_14_113545" <- function(version = "newest") {
    wjj0 <- hs.hsgly("1.流程/单细胞/转录组/10x/cr/根目录.38_09_14_114502")()
    v1 <- hs.switch(version, "newest", hs.wjj.ls(wjj0, "^cellranger-", 
        wjj = 1) %=>% basename %=>% hs.sub("cellranger-") %=>% 
        as.package_version %=>% max %=>% as.character, version)
    hs.fp(wjj0, paste0("cellranger-", v1), "bin/cellranger")
}
"38_08_20_224729" <- function(fq_fp) {
    all(hs.grepl(fq_fp, "(?i)_s\\d+_l\\d+_r[12]_\\d+\\.fastq\\.gz$"))
}
"38_01_25_115910" <- function(barcode.loc, features.loc, matrix.loc, 
    gene.column = 2, cell.column = 1, unique.features = TRUE, 
    strip.suffix = FALSE, cn.pre = "", cn.pre.sep = "-") {
    if (0) {
        full.data <- list()
        barcode.loc <- file.path(run, "barcodes.tsv")
        gene.loc <- file.path(run, "genes.tsv")
        features.loc <- file.path(run, "features.tsv.gz")
        matrix.loc <- file.path(run, "matrix.mtx")
    }
    if (0) {
        pre_ver_3 <- file.exists(gene.loc)
        if (!pre_ver_3) {
            addgz <- function(s) {
                return(paste0(s, ".gz"))
            }
            barcode.loc <- addgz(s = barcode.loc)
            matrix.loc <- addgz(s = matrix.loc)
        }
        if (!file.exists(barcode.loc)) {
            stop("Barcode file missing. Expecting ", basename(path = barcode.loc))
        }
        if (!pre_ver_3 && !file.exists(features.loc)) {
            stop("Gene name or features file missing. Expecting ", 
                basename(path = features.loc))
        }
        if (!file.exists(matrix.loc)) {
            stop("Expression matrix file missing. Expecting ", 
                basename(path = matrix.loc))
        }
    }
    data <- Matrix::readMM(matrix.loc)
    if (0) {
        cell.barcodes <- read.table(barcode.loc, header = FALSE, 
            sep = "\t", row.names = NULL)
        if (ncol(cell.barcodes) > 1) {
            cell.names <- cell.barcodes[, cell.column]
        }
        else {
            cell.names <- readLines(barcode.loc)
        }
    }
    cell.names <- readLines(barcode.loc)
    if (nchar(cn.pre)) 
        cell.names <- paste0(cn.pre, cn.pre.sep, cell.names)
    if (0) {
        if (all(grepl(pattern = "\\-1$", x = cell.names)) & strip.suffix) {
            cell.names <- as.vector(x = as.character(x = sapply(X = cell.names, 
                FUN = ExtractField, field = 1, delim = "-")))
        }
        if (is.null(x = names(data.dir))) {
            if (length(x = data.dir) < 2) {
                colnames(data) <- cell.names
            }
            else {
                colnames(x = data) <- paste0(i, "_", cell.names)
            }
        }
        else {
            colnames(x = data) <- paste0(names(x = data.dir)[i], 
                "_", cell.names)
        }
    }
    colnames(data) <- cell.names
    if (0) {
        feature.names <- read.delim(file = ifelse(pre_ver_3, 
            gene.loc, features.loc), header = FALSE, stringsAsFactors = FALSE)
    }
    feature.names <- wk.fread(features.loc, cn = 0)
    if (any(is.na(feature.names[[gene.column]]))) {
        warning("Some features names are NA. Replacing NA names with ID from the opposite column requested", 
            call. = FALSE, immediate. = TRUE)
        hs.browser("38.01.25.092048")
        na.features <- which(x = is.na(x = feature.names[, gene.column]))
        replacement.column <- ifelse(test = gene.column == 2, 
            yes = 1, no = 2)
        feature.names[na.features, gene.column] <- feature.names[na.features, 
            replacement.column]
    }
    if (unique.features) {
        fcols <- ncol(feature.names)
        if (fcols < gene.column) {
            stop(paste0("gene.column was set to ", gene.column, 
                " but feature.tsv.gz (or genes.tsv) only has ", 
                fcols, " columns.", " Try setting the gene.column argument to a value <= to ", 
                fcols, "."))
        }
        rownames(data) <- make.unique(feature.names[[gene.column]])
    }
    if (ncol(feature.names) > 2) {
        hs.browser("38.01.25.092213")
        data_types <- factor(x = feature.names$V3)
        lvls <- levels(x = data_types)
        if (length(x = lvls) > 1 && length(x = full.data) == 
            0) {
            message("10X data contains more than one type and is being returned as a list containing matrices of each type.")
        }
        expr_name <- "Gene Expression"
        if (expr_name %in% lvls) {
            lvls <- c(expr_name, lvls[-which(x = lvls == expr_name)])
        }
        data <- lapply(X = lvls, FUN = function(l) {
            return(data[data_types == l, , drop = FALSE])
        })
        names(x = data) <- lvls
    }
    else {
        return(data)
        data <- list(data)
    }
    if (0) {
        full.data[[length(x = full.data) + 1]] <- data
        list_of_data <- list()
        for (j in 1:length(full.data[[1]])) {
            list_of_data[[j]] <- do.call(cbind, lapply(full.data, 
                `[[`, j))
            list_of_data[[j]] <- as(list_of_data[[j]], "dgCMatrix")
        }
        names(list_of_data) <- names(full.data[[1]])
        if (length(list_of_data) == 1) {
            list_of_data[[1]]
        }
        else {
            list_of_data
        }
    }
}
"38_08_12_145141" <- function(fp, name_smart = 1, sn = names(fp)) {
    on.exit(hs.gc_here(), add = TRUE)
    if (fp %=>% {
        is.character(x) && (length(x) == 3) && (names(x) %all.in% 
            c("matrix.mtx.gz", "barcodes.tsv.gz", "features.tsv.gz"))
    }) 
        stop("[2023_06_30_120307]请提供带名称的列表")
    if (length(sn)) 
        names(fp) <- sn
    if (is.list(fp)) {
        a <- seq_along(fp) %=>% wk.mclapply(x, . %=>% {
            wj <- fp[[x]] %=>% hs.hsgly("1.流程/单细胞/转录组/标准流程/模块/10x/表达量文件夹.3个文件.39_05_01_232351")(x)
            sn <- names(fp)[x]
            wj[1] %=>% dirname %=>% hs.with_wd(x, temp = 1, {
                hs.wj.symlink(wj, ".", name = names(wj))
                hs.hsgly("10x2seurat.38_08_12_145141")(".", sn = sn)
            })
        })
        sce_list <- hs.unlist(a, recursive = 0)
    }
    else {
        fp <- hs.hsgly("10x.fp.exp.38_09_19_155735")(fp)
        if ((!length(names(fp))) && name_smart) 
            names(fp) <- fp %=>% hs.dirname(2) %=>% basename
        sce_list <- wk.mclapply(seq_along(fp), function(i1) {
            a <- hs.hsgly("read_10x.38_07_12_111111")(fp[i1])
            hs.hsgly("seurat/函数/table2seurat.2023_06_25_220926")(m1 = a, 
                xm = names(fp)[i1])
        }, mc.cores = 4)
    }
    names(sce_list) <- names(fp)
    sce_list
}
"38_09_19_155735" <- function(fp) {
    for (i1 in seq_along(fp)) {
        fp1 <- fp[i1]
        if (hs.isDir(fp1, a = 1)) {
            fp2 <- hs.fp(fp1, "outs/filtered_feature_bc_matrix.h5")
            if (file.exists(fp2)) 
                fp[i1] <- fp2
            next
            fp2 <- hs.fp(fp1, "filtered_feature_bc_matrix.h5")
            if (file.exists(fp2)) 
                fp[i1] <- fp2
        }
    }
    fp
}
"39_05_01_232351" <- function(wjj1 = ".") {
    wj <- hs.switch(length(wjj1), 1, {
        if (!dir.exists(wjj1)) 
            hs.browser("39.05.01.232955", debug = 0)
        hs.wjj.ls(wjj1, "\\.gz$")
    }, 3, wjj1, hs.browser("39.05.01.232910", debug = 0))
    if (length(wj) != 3) 
        hs.browser("39.05.01.232846", debug = 0)
    wj %<=>% normalizePath
    i1 <- sapply(wj, readLines, 1) %=>% hs.grepl("^%%MatrixMarket") %=>% 
        which
    if (length(i1) != 1) 
        hs.browser("39.05.01.223216", debug = 0)
    {
        ln <- wj[i1] %=>% readLines(10) %=>% hs.grepV("^%", inv = 1)[1] %=>% 
            hs.strsplit.v(" +")[2] %=>% as.numeric
    }
    wj2 <- wj[-i1] %=>% {
        i2 <- which(sapply(x, .wc) == ln)
        if (length(i2) != 1) 
            hs.browser("39.05.01.223721", debug = 0)
        x[i2]
    }
    hs.c(c("matrix.mtx.gz", "barcodes.tsv.gz", "features.tsv.gz"), 
        c(wj[i1], wj2) %=>% c(x, wj %not% x))
}
"38_08_19_130133" <- function(fq_fp_list, org1, is_nuclear = {
}, include_intron = if (length(is_nuclear)) is_nuclear else 1, 
    cell_n_expected = 0, force_cells = 0, fn.sn = {
    }, aggr = length(fq_fp_list) > 1, prj_fp_root = {
    }, wjj2 = ".", wjj2_here_post = "mid/cell_ranger", wjj2_here = if (length(hs.clean(wjj2_here_post))) hs.fp(wjj2, 
        wjj2_here_post) else wjj2, star_index = switch(org1, 
        human = "refdata-gex-GRCh38-2020-A", mouse = "refdata-gex-mm10-2020-A", 
        human_mouse = "refdata-gex-GRCh38-and-mm10-2020-A", rat = "Rattus_norvegicus.mRatBN7.2.106", 
        sheep = "Ovis_aries.Oar_v3.1.107", 黑山羊 = , 山羊 = "Capra_hircus.ARS1.107", 
        hs.browser("38.04.13.203433")), cr = hs.hsgly("1.流程/单细胞/转录组/10x/cr/cr.38_09_14_113545")(), 
    nc = 4, copy1 = 1, bam = 1, ot = "") {
    hs.browser("39.05.10.194252", debug = 0)
    if (!file.exists(star_index)) 
        star_index %<=>% hs.fp(hs.hsgly("1.流程/单细胞/转录组/10x/cr/根目录.38_09_14_114502")(), 
            x)
    pattern1 <- "(?i)_r[12](_\\d+)?\\.f(ast)?q\\.gz$"
    cmd <- lapply(seq_along(fq_fp_list), function(i1) {
        id1 <- names(fq_fp_list)[i1]
        {
            fq_wj <- fq_fp_list[[i1]]
            fq_wj_original <- if (hs.isDir(fq_wj, a = 1)) {
                hs.hsgly("wjj2fq.37_12_25_144226")(fq_wj, sample_wise = 0)
            }
            else fq_wj
            is_local <- prj_fp_root %=>% if (length(x)) {
                normalizePath(x) %=>% {
                  n1 <- nchar(x)
                  wjj0 <- fq_wj_original %=>% normalizePath %=>% 
                    substr(1, n1)
                  all(wjj0 == x)
                }
            }
            else 1
            fq_wj_2use <- if (is_local) {
                fq_wj_original
            }
            else hs.sub(fq_wj_original, "(?i).*/p\\d+/", hs.fp(wjj2, 
                "in/"))
        }
        cmd1 <- hs.with_wd(hs.wjj(wjj2_here), {
            wj2 <- hs.fp(id1, "outs", c("filtered_feature_bc_matrix.h5", 
                "web_summary.html")) %=>% normalizePath
            if (!all(file.exists(wj2))) {
                if (file.exists(id1)) 
                  hs.wj.bf(id1)
                hs.wj.copy(fq_wj_original, fq_wj_2use, overwrite = 0)
                hs.grepV(fq_wj_2use, pattern1, hs0 = basename) %=>% 
                  {
                    post <- wk.dt("_S1_L00", rep(seq(length(x)/2), 
                      each = 2), "_R", 1:2, "_001.fastq.gz") %>% 
                      apply(1, paste, collapse = "")
                    if (0) {
                      post <- if (all(hs.grepl(., "(?i)_s\\d+_l\\d+_r[12]_\\d+\\.f(ast)?q\\.gz$"))) {
                        hs.re(basename(.), pattern1)
                      }
                      else {
                        wk.dt("_S1_L00", rep(seq(length(.)/2), 
                          each = 2), "_R", 1:2, "_001.fastq.gz") %>% 
                          apply(1, paste, collapse = "")
                      }
                    }
                    hs.wj.symlink(x, ".", name = paste0(id1, 
                      post))
                  }
                cmd1 <- .sys(cr, "count", "--fastqs", ".", "--id", 
                  .q(id1), "--sample", .q(id1), "--transcriptome", 
                  .sys.fpGood(star_index), "--disable-ui", "--nopreflight", 
                  "--include-introns", ifelse(include_intron, 
                    "true", "false"), if (!bam) 
                    "--no-bam", hs.switch(Sys.info()["nodename"], 
                    "APT-SH-DT-716", c("--localcores 4", "--localmem 20"), 
                    c("--localcores 20", "--localmem 200")), 
                  if (cell_n_expected) 
                    c("--expect-cells", cell_n_expected), if (force_cells) 
                    c("--force-cells", force_cells))
                hs.msg(cmd1, 0)
                return(cmd1)
                .wk(cmd1)
            }
        })
        return(cmd1)
        if (!is_local) 
            hs.wj.rm(fq_wj_2use)
    })
    hs.wjj(wjj2_here) %=>% hs.with_wd(x, wk.mclapply(cmd %=>% 
        hs.nonempty, .wk, mc.cores = nc))
    sample_name <- if (hs.is_empty(fn.sn)) {
        names(fq_fp_list)
    }
    else wk.dt.subset(fn.sn, names(fq_fp_list))[[2]]
    hs.browser("39.05.02.235314", debug = 0)
    if (aggr && (length(fq_fp_list) > 1)) {
        hs.browser("39.05.02.235302", debug = 0)
        wj.csv <- hs.fp(wjj2_here, "mid/cell_ranger.aggr/sample_info.csv")
        if (!file.exists(wj.csv)) {
            a <- wk.dt(sample_id = names(fq_fp_list), molecule_h5 = normalizePath(hs.fp(wjj2, 
                "mid/cell_ranger", names(fq_fp_list), "outs/molecule_info.h5")))
            wk.dt.fwrite(a, wj.csv, sep = ",")
        }
        id1 <- "aggregated"
        if (!file.exists(hs.fp(dirname(wj.csv), id1))) 
            hs.with_wd(wj.csv, .wk(cr, "aggr", "--id", id1, "--csv", 
                wj.csv, "--localcores", 10, "--localmem", 20))
        if (0) {
            .wk(cr, "aggr --help")
            a <- wk.fread("/share/192.168.130.100/外包数据/NH/P20220401708/mid/cell_ranger.aggr/test/outs/count/filtered_feature_bc_matrix/barcodes.tsv.gz", 
                cn = 0)
            a[, `:=`(V2, hs.re(V1, "\\d+$"))]
            a[, V2 %>% hs.table]
            hs.fp(wjj2, "mid/cell_ranger", basename(fq_fp_list), 
                "outs/filtered_feature_bc_matrix/barcodes.tsv.gz") %>% 
                sapply(.wc) %>% unname
        }
    }
    exp_fp_list <- hs.fp(wjj2_here, names(fq_fp_list), "outs/filtered_feature_bc_matrix.h5") %>% 
        hs.c(names(fq_fp_list), .)
    hs.switch(ot, "seurat_data_list", hs.hsgly("10x2seurat.38_08_12_145141")(exp_fp_list), 
        exp_fp_list)
}
"38_07_12_111111" <- function(fp1, feature.column = "auto") {
    fp1 <- hs.hsgly("10x.fp.exp.38_09_19_155735")(fp1)
    hs.cond(hs.grepl(fp1, "(?i)\\.h5$"), Seurat::Read10X_h5(fp1), 
        hs.isDir(fp1, a = 1), {
            if (feature.column == "auto") {
                cn <- hs.hsgly("1.流程/单细胞/转录组/标准流程/模块/10x/表达量文件夹.3个文件.39_05_01_232351")(fp1)["features.tsv.gz"] %=>% 
                  readLines(1) %=>% hs.strsplit.v("\\s+") %=>% 
                  length
                feature.column <- min(2, cn)
            }
            Seurat::Read10X(data.dir = fp1, gene.column = feature.column)
        }, hs.browser("38.07.12.111342"))
}
"38_10_18_104617" <- function(cell_cluster_de, org1, wjj2) {
    hs.require(c("GGally", "network", "sna", "RColorBrewer"))
    sapply(seq_along(cell_cluster_de), function(comparison_i1) {
        hs.msg(paste("comparison_i1 =", comparison_i1))
        comparison1 <- names(cell_cluster_de)[comparison_i1]
        cell_cluster_de <- cell_cluster_de[[comparison_i1]]
        nn <- cell_cluster_de[, max(nchar(unique(cell_cluster)))]
        wj <- cell_cluster_de[, unique(cell_cluster) %>% as.numeric %>% 
            sort] %>% wk.mclapply(function(cgi) {
            hs.msg(paste("cgi =", cgi), 2)
            wj2 <- hs.fp(wjj2, "Group_DEG", comparison1, paste0("cluster_", 
                hs.str_pad(cgi, nn, 0), ".png"))
            if (!file.exists(wj2)) {
                if (cell_cluster_de[cell_cluster %in% cgi, !.N]) 
                  return()
                dt1 <- wk.dt.subset(cell_cluster_de, exp_l = c(cell_cluster %in% 
                  cgi, p_val_adj < 0.05, abs(avg_log2FC) > 0.25))
                if (hs.is_empty(dt1, co1 = 10)) 
                  return()
                dt1[, `:=`(gene_name_updated, hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937")(gene_id_symbol, 
                  org1 = org1))]
                dt1 <- dt1[hs.clean(gene_name_updated, ot = "i")]
                if (0) {
                  dt1[match(c("RPS27A", "RPS3A"), gene_id_symbol)]
                }
                if (hs.is_empty(dt1, co1 = 10)) 
                  return()
                link.fp <- paste0(wj2, ".link.tsv.gz")
                if (file.exists(link.fp)) {
                  link <- wk.fread(link.fp)
                }
                else {
                  link <- hs.hsgly("生物学/数据库/string/基因符号网络数据.38_01_05_175427")(org1 = org1, 
                    extract = dt1[, gene_name_updated], co1 = 900)
                  if (hs.is_empty(link)) 
                    return()
                  wk.dt.setnames(link, c("symbol1", "symbol2", 
                    "combined_score"), c("from", "to", "weight"))
                  link <- link[(tolower(from) %in% dt1[, tolower(gene_name_updated)]) & 
                    (tolower(to) %in% dt1[, tolower(gene_name_updated)])]
                  wk.dt.col.reorder(link, c("from", "to"), before = 1)
                  wk.dt.fwrite(link, link.fp)
                }
                net <- network::network(link)
                node_name <- network::get.vertex.attribute(net, 
                  "vertex.names")
                net %v% "combined_scores" <- sapply(node_name, 
                  . %>% {
                    link[(from %in% .) | (to %in% .), sum(weight)]
                  })
                tag <- local({
                  node.weight <- hs.c(node_name, network::`%v%`(net, 
                    "combined_scores"))
                  jg <- rep("", length(node.weight))
                  i1 <- tail(order(node.weight), 10)
                  jg[i1] <- rev(head(LETTERS, length(i1)))
                  hs.c(node_name, jg)
                })
                p1 <- ggnet2(net, node.color = dt1[match(tolower(node_name), 
                  tolower(gene_name_updated)), ifelse(avg_log2FC > 
                  0, "tomato", "steelblue")], node.size = "combined_scores", 
                  max_size = 5, label = tag, label.color = "black", 
                  label.size = 3) + scale_size_discrete(guide = "none") + 
                  geom_text(aes(x, y, label = tag), data.frame(x = 0.05, 
                    y = 0.99, tag = paste(c("Top 10 most connected:", 
                      hs.clean(tag) %>% {
                        paste0("  ", ., ": ", names(.))
                      } %>% sort), collapse = "\n")), vjust = 1) + 
                  labs(title = paste0("Cell cluster ", cgi, " genes changed between samples")) + 
                  theme(legend.position = c(0.85, 0.95))
                .figs(print(p1), wj2, width = 40, height = 30)
            }
        })
        wj
    })
}
"38_10_18_102752" <- function(cell_cluster_markers, seurat_data, 
    org1, wjj2, ft = "symbol") {
    hs_a <- function(dt = cell_cluster_markers, seurat_data, 
        nc = 50, ft = "symbol") {
        cluster_cn <- hs.grepV(names(dt), "(?i)cluster")
        q_cn <- c("p_val_adj", "median_qval") %and% names(dt)
        gene_cn <- c("gene_name", "gene") %and% names(dt)
        cc2do <- dt[, get(cluster_cn) %>% unique %>% paste]
        nn <- max(nchar(cc2do))
        wk.mclapply(seq_along(cc2do), function(i) {
            cell_cluster_1 <- cc2do[i]
            bc <- SeuratObject::Idents(seurat_data) %=>% names(x)[x %in% 
                cell_cluster_1]
            feature.bc_pos_n <- hs.rowSums(seurat_data@assays[["RNA"]]@counts[, 
                bc] > 0)
            u1 <- rownames(seurat_data@assays[["RNA"]]@counts)[(feature.bc_pos_n >= 
                5) | ((feature.bc_pos_n/length(bc)) > 0.05)]
            repeated <- 0
            repeat {
                a <- .try(hs.lapply(c(1), function(direction_1) {
                  s1 <- wk.dt.subset(dt, exp_l = list(get(cluster_cn) %in% 
                    cell_cluster_1, get(q_cn) < 0.05, .SD[, hs.rowAlls(sign(.SD) == 
                    direction_1), .SDcols = patterns("(?i)avg_log2FC$")]), 
                    exp_j = get(gene_cn))
                  wj_pre <- hs.fp1(wjj2, paste0("cluster_", hs.str_pad(cell_cluster_1, 
                    nn, 0)), switch(paste(direction_1), `1` = "high", 
                    `-1` = "low"))
                  hs.msg(wj_pre)
                  if (0) {
                    hs.save(s1, u1, org1, wj_pre, wj = "~/a.rd")
                    load("~/a.rd")
                  }
                  hs.hsgly("genes.understand.35.11.12.182333")(s1, 
                    u1, ft = ft, org1 = org1, anno.type.2do = c("kegg", 
                      "go"), wj_out.pre = wj_pre, ncpu = 1)
                }))
                if (!hs.is_try_error(a)) 
                  break
                repeated <- repeated + 1
                if (repeated > 10) 
                  break
            }
        }, mc.cores = nc)
    }
    repeat {
        a <- system.time(hs.hsgly("genes.understand.35.11.12.182333")(org1 = org1), 
            gcFirst = FALSE)
        if (a["elapsed"] < 1) 
            break
    }
    null <- hs_a(cell_cluster_markers, seurat_data = seurat_data, 
        ft = ft)
    {
        hs.hsgly("转录组/标准流程/模块/差异表达基因/基因集合富集分析/top_go.38_04_10_232515")(wjj2, 
            fig_n = 1)
        hs.hsgly("转录组/标准流程/模块/差异表达基因/基因集合富集分析/top_kegg.38_04_10_233503")(wjj2, 
            fig_n = 1)
    }
}
"38_10_18_103042" <- function(cell_cluster_de, seurat_data, org1, 
    wjj2) {
    u1 <- rownames(seurat_data@assays[["RNA"]]@counts)
    wjj.38_01_04_140634 <- wjj2
    hs_a <- function(nc = 50) {
        lapply(seq_along(cell_cluster_de), function(comparison_i1) {
            comparison1 <- names(cell_cluster_de)[comparison_i1]
            if (nc < 2) 
                hs.msg(comparison_i1)
            wk.mclapply(cell_cluster_de[[comparison_i1]][, unique(cell_cluster)], 
                function(cell_cluster_1) {
                  if (nc < 2) 
                    hs.msg(cell_cluster_1, 2)
                  de1 <- cell_cluster_de[[comparison_i1]][cell_cluster %in% 
                    cell_cluster_1]
                  if (hs.is_empty(de1)) 
                    return()
                  a <- hs.lapply(c(1, -1), function(direction_1) {
                    if (nc < 2) 
                      hs.msg(direction_1, 3)
                    s1 <- wk.dt.subset(de1, exp_l = list(p_val_adj < 
                      0.05, sign(avg_log2FC) == direction_1), 
                      exp_j = gene_id_symbol)
                    if (length(s1) < 10) 
                      return()
                    repeated <- 0
                    pre1 <- hs.fp(wjj.38_01_04_140634, comparison1, 
                      sprintf("cluster_%02d", as.numeric(cell_cluster_1)), 
                      switch(as.character(direction_1), `1` = "high", 
                        `-1` = "low"))
                    if (0) {
                      hs.save(s1, u1, org1, pre1, wj = "~/a.rd")
                      load("~/a.rd")
                      hs.hsgly("genes.understand.35.11.12.182333")(s1, 
                        u1, ft = "symbol", org1 = org1, anno.type.2do = c("kegg", 
                          "go"), wj_out.pre = pre1, ncpu = 1)
                    }
                    repeat {
                      a <- .try(hs.hsgly("genes.understand.35.11.12.182333")(s1, 
                        u1, ft = "symbol", org1 = org1, anno.type.2do = c("kegg", 
                          "go"), wj_out.pre = pre1, ncpu = 1))
                      if (hs.is_try_error(a)) {
                        message(a)
                      }
                      else break
                      repeated <- repeated + 1
                      if (repeated > 10) 
                        break
                    }
                  })
                }, mc.cores = nc)
        })
    }
    repeat {
        a <- system.time(hs.hsgly("genes.understand.35.11.12.182333")(org1 = org1), 
            gcFirst = FALSE)
        if (a["elapsed"] < 1) 
            break
    }
    null <- .try(hs_a())
    repeat {
        a <- system.time(hs.hsgly("genes.understand.35.11.12.182333")(org1 = org1), 
            gcFirst = FALSE)
        if (a["elapsed"] < 1) 
            break
    }
    null <- hs_a()
    {
        null <- wk.mclapply(seq_along(cell_cluster_de), function(comparison_i1) {
            message(comparison_i1)
            wjj1 <- hs.fp(wjj2, names(cell_cluster_de)[comparison_i1])
            top_go.38_04_10_232515(wjj1, fig_n = 2)
            top_kegg.38_04_10_233503(wjj1, fig_n = 2)
        })
        if (0) {
            wj2 <- hs.fp(wjj2, "report", "genes_changed_between_samples.en.kegg.top.pdf")
            if (!file.exists(wj2)) 
                local({
                  wj_in <- hs.wjj.ls(wjj.38_01_04_140634, "\\.kegg\\.tsv\\.gz$", 
                    R = 1)
                  d_v <- hs.re(basename(wj_in), "high|low")
                  dt_l <- hs.lapply(c("high", "low"), function(fx1) {
                    . <- wj_in[d_v %in% fx1]
                    dt_l <- lapply(., wk.fread)
                    names(dt_l) <- basename(dirname(.))
                    dt_l <- lapply(dt_l, . %>% {
                      head(.[qvalue < 0.05], 10)
                    }) %>% hs.nonempty(nrow = 1)
                    a <- lapply(dt_l, . %>% {
                      .[, .(Description, `-log10(qvalue)` = -log10(qvalue), 
                        GeneRatio = sapply(GeneRatio, evalStr))]
                    })
                    dt1 <- wk.dt.fromList(a, nm = "cell_cluster")
                    dt1[, `:=`(cell_cluster, cell_cluster %>% 
                      {
                        hs.factor(., as.character(sort(as.numeric(unique(.)))), 
                          N = 0)
                      })]
                    a <- dt1[, max(GeneRatio), by = Description]
                    dt1[, `:=`(Description, hs.factor(Description, 
                      a[order(V1), Description], N = 0))]
                  })
                  p_l <- hs.lapply(names(dt_l), function(fx1) {
                    dt1 <- dt_l[[fx1]]
                    ggplot(dt1, aes(cell_cluster, Description)) + 
                      geom_point(aes(size = GeneRatio, color = `-log10(qvalue)`)) + 
                      xlab("Cell cluster") + ylab("") + labs(title = paste0(switch(fx1, 
                      high = "Positive", low = "Negative"), " markers, KEGG")) + 
                      theme(plot.title = element_text(size = plot.title.size)) + 
                      theme(panel.background = {
                      }, panel.grid = element_line(color = "gray"))
                  })
                  .pdf(print(p_l[[1]] + p_l[[2]]), wj2, width = sum(sapply(dt_l, 
                    . %>% {
                      .[, uniqueN(cell_cluster) + 5] * 2
                    })), height = max(sapply(dt_l, . %>% {
                    .[, uniqueN(Description) + 2]
                  })))
                })
        }
    }
}
"38_04_10_232515" <- function(wjj1, fig_n = 1, is_cluster = 1) {
    wj_in <- hs.wjj.ls(wjj1, "\\bgo\\..*\\.tsv\\.gz$", R = 1)
    hs.browser("39.01.11.163706", debug = 0)
    if (!length(wj_in)) 
        return()
    cg_v <- basename(dirname(wj_in))
    if (is_cluster) 
        cg_v %<=>% hs.re(x, "\\d+$")
    plot.title.size <- 15
    description.size <- 10
    lapply(unique(cg_v), function(cg1) {
        wj2 <- hs.fp(wjj1, paste0(if (is_cluster) 
            "cluster_"
        else "", cg1), "top_go")
        if (file.exists(paste0(wj2, ".png"))) 
            return()
        . <- wj_in[cg_v %in% cg1]
        f1 <- hs.re(basename(.), "high|low")
        p1 <- local({
            . <- .[f1 %in% "high"]
            if (!length(.)) 
                return()
            names(.) <- hs.re(basename(.), "(?<=go\\.)[^\\.]+") %=>% 
                toupper
            dt_list <- hs.lapply(., wk.fread)
            dt_list[1]
            a1 <- lapply(dt_list, . %=>% x[head(which(qvalue < 
                0.05), 10), .(Description = Description, `-log10(qvalue)` = -log10(qvalue))])
            a1 <- a1[sapply(a1, nrow) > 0]
            if (!length(a1)) 
                return()
            dt1 <- wk.dt.fromList(a1, nm = "domain")
            dt1[, `:=`(Description, hs.factor(Description, N = 0))]
            ggplot(dt1, aes(Description, `-log10(qvalue)`, fill = domain)) + 
                geom_bar(stat = "identity") + xlab("") + ylab(quote(paste("-Log"[paste(10)], 
                "(Q-value)"))) + labs(title = paste0("Cell cluster ", 
                cg1, ", positive markers")) + theme(plot.title = element_text(size = plot.title.size)) + 
                theme(legend.position = "none") + theme(panel.background = {
            }, panel.grid = element_blank()) + theme(axis.text.x = element_text(angle = 90, 
                hjust = 1, size = description.size))
        })
        p2 <- local({
            . <- .[f1 %in% "low"]
            if (!length(.)) 
                return()
            names(.) <- hs.re(basename(.), "(?<=go\\.)[^\\.]+") %=>% 
                toupper
            dt_list <- hs.lapply(., wk.fread)
            a1 <- lapply(dt_list, . %=>% x[head(which(qvalue < 
                0.05), 10), .(Description = Description, `-log10(qvalue)` = -log10(qvalue))])
            a1 <- a1[sapply(a1, nrow) > 0]
            if (!length(a1)) 
                return()
            dt1 <- wk.dt.fromList(a1, nm = "domain")
            dt1[, `:=`(Description, hs.factor(Description, N = 0))]
            ggplot(dt1, aes(Description, `-log10(qvalue)`, fill = domain)) + 
                geom_bar(stat = "identity") + xlab("") + ylab(quote(paste("-Log"[paste(10)], 
                "(Q-value)"))) + labs(title = paste0("Cell cluster ", 
                cg1, ", negative markers")) + theme(plot.title = element_text(size = plot.title.size)) + 
                theme(panel.background = {
                }, panel.grid = element_blank()) + theme(legend.position = "right", 
                legend.direction = "vertical", legend.title = element_blank()) + 
                theme(axis.text.x = element_text(angle = 90, 
                  hjust = 1, size = description.size))
        })
        P <- p1 + p2
        .figs(print(P), wj2, width = 25 * fig_n, height = 30)
        wj2
    })
}
"38_04_10_233503" <- function(wjj1, fig_n = 1) {
    plot.title.size <- 15
    description.size <- 10
    wj_in <- hs.wjj.ls(wjj1, "\\bkegg\\.tsv\\.gz$", R = 1)
    wj_in.type <- hs.re(basename(wj_in), "\\w+")
    wj_in.type.db <- unique(wj_in.type)
    lapply(seq_along(wj_in.type.db), function(i1) {
        type_1 <- wj_in.type.db[i1]
        wj2 <- hs.fp(wjj1, paste0("top_kegg.", type_1))
        if (file.exists(paste0(wj2, ".png"))) 
            return()
        wj_in <- wj_in[wj_in.type %in% type_1]
        {
            en <- lapply(wj_in, wk.fread)
            names(en) <- wj_in %>% dirname %>% hs.re("\\d+$")
            top_term <- lapply(en, . %>% {
                .[p.adjust < 0.05, head(Description, 5)]
            }) %>% unlist %>% unique
            if (!length(top_term)) 
                return()
            dt <- wk.dt(term = rep(top_term, length(en)))
            dt[, `:=`("cluster", rep(names(en), each = length(top_term)))]
            a <- sapply(en, . %>% {
                wk.dt.subset.na(., top_term, "Description")[, 
                  p.adjust]
            })
            a <- -log10(a)
            dt[, `:=`("-Log10FDR", a)]
            rownames(a) <- top_term
            ci <- if (length(en) > 1) {
                hs.hsgly("层次聚类/fastcluster.36_08_12_231620")(a, 
                  ot = "o")
            }
            else seq(ncol(a))
            dt[, `:=`(cluster, hs.factor(cluster, colnames(a)[ci], 
                N = 0))]
            ri <- hs.hsgly("层次聚类/fastcluster.36_08_12_231620")(a, 
                ot = "o", side = "row")
            dt[, `:=`(term, hs.factor(term, rownames(a)[ri], 
                N = 0))]
            a <- sapply(en, . %>% {
                wk.dt.subset.na(., top_term, "Description")[, 
                  GeneRatio]
            })
            dt[, `:=`("GeneRatio", sapply(a, evalStr))]
            dt[, `:=`(cluster, cluster %>% {
                levels(.) %<>% {
                  paste0("Cluster ", .)
                }
                .
            })]
            p1 <- ggplot2::ggplot(dt, ggplot2::aes(cluster, term, 
                size = GeneRatio, color = `-Log10FDR`)) + ggplot2::geom_point() + 
                ggplot2::labs(x = NULL, y = NULL) + ggplot2::theme_bw() + 
                ggplot2::theme(axis.text.x = element_text(angle = 45, 
                  hjust = 1), panel.grid = element_blank())
            .figs(print(p1), wj2, width = length(en) * 0.25 + 
                10 + 5, height = length(top_term) * 0.4 + 5)
        }
        0
    })
    return()
    cg_v <- basename(dirname(wj_in)) %>% hs.re("\\d+$")
    lapply(unique(cg_v), function(cg1) {
        wj2 <- hs.fp(wjj1, paste0("cluster_", cg1), "top_kegg.pdf")
        if (!file.exists(wj2)) {
            . <- wj_in[cg_v %in% cg1]
            f1 <- hs.re(basename(.), "high|low")
            p1 <- local({
                . <- .[f1 %in% "high"]
                if (!length(.)) 
                  return()
                dt_list <- hs.lapply(., wk.fread)
                a1 <- lapply(dt_list, . %>% {
                  .[head(which(qvalue < 0.05), 10), .(Description = Description, 
                    `-log10(qvalue)` = -log10(qvalue))]
                })
                dt1 <- a1[[1]]
                dt1[, `:=`(Description, hs.factor(Description, 
                  N = 0))]
                ggplot(dt1, aes(Description, `-log10(qvalue)`)) + 
                  geom_bar(stat = "identity") + xlab("") + ylab(quote(paste("-Log"[paste(10)], 
                  "(Q-value)"))) + labs(title = paste0("Cell cluster ", 
                  cg1, ",\npositive markers")) + theme(plot.title = element_text(size = plot.title.size)) + 
                  theme(legend.position = "none") + theme(panel.background = {
                }, panel.grid = element_blank()) + theme(axis.text.x = element_text(angle = 90, 
                  hjust = 1, size = description.size))
            })
            p2 <- local({
                . <- .[f1 %in% "low"]
                if (!length(.)) 
                  return()
                dt_list <- hs.lapply(., wk.fread)
                a1 <- lapply(dt_list, . %>% {
                  .[head(which(qvalue < 0.05), 10), .(Description = Description, 
                    `-log10(qvalue)` = -log10(qvalue))]
                })
                dt1 <- a1[[1]]
                dt1[, `:=`(Description, hs.factor(Description, 
                  N = 0))]
                ggplot(dt1, aes(Description, `-log10(qvalue)`)) + 
                  geom_bar(stat = "identity") + xlab("") + ylab(quote(paste("-Log"[paste(10)], 
                  "(Q-value)"))) + labs(title = paste0("Cell cluster ", 
                  cg1, ",\ndown-regulated markers")) + theme(plot.title = element_text(size = plot.title.size)) + 
                  theme(panel.background = {
                  }, panel.grid = element_blank()) + theme(legend.position = "right", 
                  legend.direction = "vertical", legend.title = element_blank()) + 
                  theme(axis.text.x = element_text(angle = 90, 
                    hjust = 1, size = description.size))
            })
            wk.ggsave(p1 + p2, wj2, width = 25 * fig_n, height = 20)
        }
        wj2
    })
}
"38_10_18_140257" <- function(seurat_data, sample_n, only.pos = TRUE, 
    only.de = 0, wjj2) {
    wj2 <- hs.fp(wjj2, "conserved.tsv.gz")
    if (sample_n > 1) {
        hs.browser("39.01.16.150858", debug = 0)
        1
        {
            hs.msg("找群高表达基因")
            cell_cluster_markers <- hs.evalLazy({
                a <- levels(SeuratObject::Idents(seurat_data)) %>% 
                  wk.mclapply(function(cluster1) {
                    hs.msg(cluster1, 0)
                    a <- (Seurat::FindConservedMarkers)(seurat_data, 
                      ident.1 = cluster1, grouping.var = "sample_name.38_01_19_171144", 
                      only.pos = TRUE, assay = "RNA", slot = "data", 
                      verbose = 0)
                    wk.dt.setDT(a, "gene_name")
                  }, mc.cores = hs.switch(Sys.info()[["nodename"]], 
                    "APT-SH-DT-716", 2, 50))
                cn_full.38_03_11_093915 <- lapply(a, names) %>% 
                  {
                    .[[which.max(sapply(., length))]]
                  }
                lapply(seq_along(a), function(i1) {
                  cn_missed <- cn_full.38_03_11_093915 %not% 
                    names(a[[i1]])
                  if (length(cn_missed)) {
                    a[[i1]][, `:=`(c(cn_missed), NA)]
                    data.table::setcolorder(a[[i1]], cn_full.38_03_11_093915)
                  }
                  0
                })
                b <- wk.dt.fromList(a, nm = "cell_cluster")
                b[, `:=`(cell_cluster, as.integer(cell_cluster))]
                wk.dt.setorderv(b, c("cell_cluster", "max_pval"))
                b[, `:=`(max_qval, hs.rowMaxs(.SD[, .SD, .SDcols = patterns("_p_val_adj$")]))]
                b[, `:=`(median_qval, hs.rowMedians(.SD[, .SD, 
                  .SDcols = patterns("_p_val_adj$")]))]
            }, fp = wj2)
            hs.msg("完成", 0)
        }
    }
    else {
        cell_cluster_markers <- hs.evalLazy({
            a <- Seurat::FindAllMarkers(seurat_data, only.pos = only.pos, 
                max.cells.per.ident = 5000)
            wk.dt.setDT(a)
            wk.dt.col.reorder(a, "gene", before = 1)
        }, fp = wj2)
        wk.dt.setnames(cell_cluster_markers, "cluster", "cell_cluster")
        wk.dt.setnames(cell_cluster_markers, "gene", "gene_name")
        if (0) {
            a1 <- cell_cluster_markers[p_val_adj < 0.05]
            if (a1[, -1 %in% sign(avg_log2FC)]) 
                hs.browser("38.05.31.105820")
            wk.dt.setorderv(a1, c("cluster", "p_val", "avg_log2FC"), 
                c(1, 1, -1))
            {
                gene2do <- a1[, head(gene, 10), by = cluster][, 
                  unique(V1)]
                {
                  wj2 <- hs.fp(wjj2, ".top.heatmap")
                  if (!file.exists(paste0(wj2, ".png"))) 
                    local({
                      if (0) {
                        p1 <- Seurat::DoHeatmap(seurat_data, 
                          assay = "RNA", slot = "scale.data", 
                          features = gene2do, angle = 30, size = 3)
                        print(p1)
                      }
                      m1 <- seurat_data@assays[["RNA"]]@data[gene2do, 
                        , drop = FALSE] %>% apply(1, . %>% {
                        (. - mean(.))/sd(.)
                      }) %>% t
                      m1 <- m1[, Seurat::Idents(seurat_data)[colnames(m1)] %>% 
                        sort %>% names]
                      annotation_col <- hs.df(`Cell cluster` = Seurat::Idents(seurat_data)[colnames(m1)])
                      col1 <- c("darkolivegreen3", "lightsteelblue1", 
                        "dodgerblue4")
                      col1 <- c("navy", "white", "firebrick3")
                      col1 <- list(c(87, 87, 156), c(102, 186, 
                        159), c(251, 244, 185), c(231, 131, 84), 
                        c(162, 46, 74)) %>% sapply(ieiwk.col2rgb)
                      col1 <- list(c(87, 87, 156), c(102, 186, 
                        159), "white", c(231, 131, 84), c(162, 
                        46, 74)) %>% sapply(ieiwk.col2rgb)
                      col1 <- list(c(224, 224, 205), c(186, 214, 
                        232), c(65, 68, 129)) %>% sapply(ieiwk.col2rgb)
                      col1 <- (grDevices::colorRampPalette(c("purple", 
                        "black", "gold")))(100)
                      r2do <- seq(min(50, nrow(m1)))
                      p1 <- pheatmap::pheatmap(m1[r2do, ], color = (grDevices::colorRampPalette(col1, 
                        space = "Lab"))(100), breaks = seq(-2, 
                        2, length.out = 100), annotation_col = annotation_col, 
                        cluster_rows = 0, cluster_cols = 0, show_colnames = 0)
                      p1 <- ggplotify::as.ggplot(p1)
                      .png(print(p1), hs.wj(hs.wjm(wj2, append = ".50", 
                        extN = 2)), width = 30, height = length(r2do) * 
                        0.3 + 5)
                      p1 <- pheatmap::pheatmap(m1, color = (grDevices::colorRampPalette(col1, 
                        space = "Lab"))(100), breaks = seq(-2, 
                        2, length.out = 100), annotation_col = annotation_col, 
                        cluster_rows = 0, cluster_cols = 0, show_colnames = 0)
                      P <- ggplotify::as.ggplot(p1)
                      .figs(print(P), wj2, width = 30, height = length(gene2do) * 
                        0.3 + 5)
                    })
                }
                {
                  gene2do <- local({
                    if (0) {
                      cell_marker %>% hs.table
                      cell_cluster_markers[gene_name %in% "GPR171"] %>% 
                        .sh
                      cell_cluster_markers[cell_cluster %in% 
                        10, "CSTB" %in% gene_name]
                    }
                    if (0) {
                      hs.save(cell_cluster_markers, wj = "~/a.rd")
                      "~/a.rd" %>% .ls
                      load("~/a.rd")
                    }
                    markers_specific <- a1 %>% {
                      a <- tapply(.[, gene], .[, cluster], c)
                      a <- inverseList(a)
                      names(a)[sapply(a, length) < 2L]
                    }
                    return(a1[, (gene %and% markers_specific) %>% 
                      {
                        .[tolower(.) %in% tolower(names(cell_marker))]
                      } %>% head(3), by = cluster][, V1])
                    if (0) 
                      local({
                        i1 <- hs.rowAlls(cell_cluster_markers[, 
                          .SD, .SDcols = patterns("(?i)log2fc")] > 
                          0)
                        cell_cluster_markers <- cell_cluster_markers[i1 & 
                          (median_qval < 0.05), head(.SD, 100), 
                          by = cell_cluster]
                        if (0) {
                          cell_cluster_markers[, hs.table(cell_cluster)]
                          cell_cluster_markers
                        }
                        cell_cluster.db <- wk.dt.subset(cell_cluster_markers, 
                          exp_l = c(toupper(gene_name) %in% unlist(lapply(cell_marker, 
                            names))), exp_j = unique(cell_cluster) %>% 
                            sort)
                        cell_n1 <- 1
                        repeat {
                          gene.db <- (cell_marker %>% {
                            names(.)[. <= cell_n1]
                          }) %and% toupper(cell_cluster_markers_specific)
                          a <- cell_cluster_markers[toupper(gene_name) %in% 
                            gene.db, toupper(head(gene_name, 
                            3)), by = cell_cluster]
                          if (cell_cluster.db %all.in% a[, cell_cluster]) 
                            break
                          cell_n1 <- cell_n1 + 1
                          if (cell_n1 > max(cell_marker)) 
                            break
                        }
                        hs.c(a[, cell_cluster], cell_cluster_markers[match(a[, 
                          V1], tolower(gene_name)), gene_name])
                      })
                  })
                  if (0) {
                    gene2do %>% {
                      .[. %in% "GP9"]
                    }
                    gene1 <- "Rpl37a"
                  }
                  gene2do %<>% unique
                  wj2 <- hs.fp(wjj2, "specific/markers.specific.2dim.png")
                  hs.lapply("umap,tsne", function(reduction_method_1) {
                    wj2 <- hs.wjm(wj2, append = reduction_method_1)
                    co1 <- 9
                    local({
                      wj2 <- hs.wjm(wj2, append = if (length(gene2do) > 
                        co1) 
                        ".first_9")
                      if (file.exists(wj2)) 
                        return()
                      i2do <- seq(min(co1, length(gene2do)))
                      .png({
                        p1 <- Seurat::FeaturePlot(seurat_data, 
                          features = gene2do[i2do], min.cutoff = "q5", 
                          slot = "data", reduction = reduction_method_1)
                        print(p1)
                      }, wj2, width = 30, height = 10 * ceiling(length(i2do)/3))
                    })
                    if (length(gene2do) > co1) 
                      local({
                        wj2 <- hs.wjm(wj2, append = ".remaining")
                        if (file.exists(wj2)) 
                          return()
                        i2do <- seq(co1 + 1, length(gene2do))
                        .png({
                          p1 <- Seurat::FeaturePlot(seurat_data, 
                            features = gene2do, min.cutoff = "q5", 
                            slot = "data", reduction = reduction_method_1)
                          print(p1)
                        }, wj2, width = 40, height = 10 * ceiling(length(i2do)/3))
                      })
                  })
                  wj2 <- hs.fp(wjj2, "specific/markers.specific.bubble.png")
                  if (!file.exists(wj2)) 
                    local({
                      {
                        df <- local({
                          a <- seurat_data@meta.data
                          a <- expand.grid(unique(a[, "sample_name.38_01_19_171144"]), 
                            levels(a[, stable_most_weighted_cn]), 
                            gene2do)
                          colnames(a) <- c("sample", "cluster", 
                            "gene")
                          a <- wk.dt.as(a, "row_name")
                          a[, `:=`("cluster_sample", paste0(cluster, 
                            "_", sample) %>% hs.factor(unique(.), 
                            N = 0))]
                        })
                        if (0) {
                          1
                          {
                            ge <- seurat_data@assays[["RNA"]]@data[gene2do, 
                              , drop = FALSE]
                            a <- apply(ge > 0, 1, tapply, seurat_data@meta.data[, 
                              c("sample_name.38_01_19_171144", 
                                stable_most_weighted_cn)], mean)
                            df[, `:=`("percent", a * 100)]
                          }
                          {
                            {
                              a <- apply(ge, 1, tapply, seurat_data@meta.data[, 
                                c("sample_name.38_01_19_171144", 
                                  stable_most_weighted_cn)], 
                                function(x) median(x[x > 0]))
                              a <- apply(a, 2, . %>% {
                                (. - median(., na.rm = TRUE))/mad(., 
                                  na.rm = TRUE)
                              })
                            }
                            {
                              a <- apply(ge, 1, tapply, seurat_data@meta.data[, 
                                c("sample_name.38_01_19_171144", 
                                  stable_most_weighted_cn)], 
                                function(x1) log1p(mean(expm1(x1) + 
                                  1)))
                              x1 <- c(1.2, 1.4)
                              log(mean(exp(x1)))
                              log(mean(expm1(x1) + 1))
                              log1p
                              mean(x1)
                              hs.log_mean(x1)
                              a <- apply(a, 2, . %>% {
                                (. - mean(., na.rm = TRUE))/sd(., 
                                  na.rm = TRUE)
                              })
                            }
                            a[a > 2] <- 2
                            a[a < -2] <- -2
                            df[, `:=`("Expression", a)]
                          }
                        }
                        {
                          a <- seurat_data@meta.data[, c("sample_name.38_01_19_171144", 
                            stable_most_weighted_cn)]
                          a <- unique(a)
                          a <- as.matrix(a)
                          null <- apply(a, 1, . %>% {
                            sample1 <- .[1]
                            cluster1 <- .[2]
                            sample.bc <- seurat_data@meta.data %>% 
                              {
                                rownames(.)[.[, "sample_name.38_01_19_171144"] %in% 
                                  sample1]
                              }
                            cluster.bc <- seurat_data@meta.data[sample.bc, 
                              ] %>% {
                              rownames(.)[.[, stable_most_weighted_cn] %in% 
                                cluster1]
                            }
                            other.bc <- sample.bc %not% cluster.bc
                            a <- Seurat::FoldChange(seurat_data@assays[["RNA"]], 
                              cluster.bc, other.bc, features = gene2do)
                            rn1 <- wk.dt.subset(df, exp_l = c(sample %in% 
                              sample1, cluster %in% cluster1), 
                              exp_j = row_name[match(gene2do, 
                                gene)])
                            i1 <- df[, match(rn1, row_name)]
                            wk.dt.set(df, i1, "Percent", a[, 
                              "pct.1"])
                            wk.dt.set(df, i1, "L2FC", a[, "avg_log2FC"])
                          })
                          df[L2FC > 2, `:=`(L2FC, 2)]
                          df[L2FC < -2, `:=`(L2FC, -2)]
                          df[, `:=`(Percent, Percent * 100)]
                        }
                        p1 <- ggplot2::ggplot(df, ggplot2::aes(gene, 
                          cluster_sample, size = Percent, color = L2FC)) + 
                          ggplot2::geom_point() + ggplot2::labs(x = NULL, 
                          y = NULL) + ggplot2::theme_bw() + ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, 
                          hjust = 1), panel.grid = ggplot2::element_blank())
                      }
                      if (0) {
                        p1 <- Seurat::DotPlot(seurat_data, features = gene2do, 
                          cols = c("blue", "red", "purple")[1:length(sce_list)], 
                          dot.scale = 8, split.by = "sample_name.38_01_19_171144") + 
                          RotatedAxis()
                      }
                      .png(print(p1), wj2, width = 10 + length(gene2do) * 
                        0.8, height = 7 + length(levels(Idents(seurat_data))) * 
                        2 * 0.9)
                    })
                  wj2 <- hs.fp(wjj2, "specific/markers.specific.heatmap.png")
                  if (!file.exists(wj2)) 
                    local({
                      m1 <- seurat_data@assays[["RNA"]]@data[gene2do, 
                        , drop = FALSE] %>% apply(1, . %>% {
                        (. - mean(.))/sd(.)
                      }) %>% t
                      if (0) {
                        m1 <- seurat_data@assays[["RNA"]]@scale.data[gene2do, 
                          , drop = 0]
                      }
                      m1 <- m1[, Seurat::Idents(seurat_data)[colnames(m1)] %>% 
                        sort %>% names]
                      annotation_col <- hs.df(`Cell cluster` = Seurat::Idents(seurat_data)[colnames(m1)])
                      p1 <- (pheatmap::pheatmap)(m1, color = colorRampPalette(c("purple", 
                        "black", "gold"))(100), breaks = seq(-2, 
                        2, length.out = 100), annotation_col = annotation_col, 
                        cluster_rows = 0, cluster_cols = 0, show_colnames = 0)
                      p1 <- ggplotify::as.ggplot(p1)
                      .png(print(p1), wj2, width = 30, height = 15)
                      return()
                      if (0) {
                        p1 <- Seurat::DoHeatmap(seurat_data, 
                          assay = "RNA", slot = "scale.data", 
                          features = gene2do, angle = 30, size = 3)
                        print(p1)
                      }
                    })
                  null <- local({
                    wj2 <- hs.fp(wjj2, "specific/markers.specific.violin.png")
                    if (file.exists(wj2)) 
                      return()
                    p1 <- Seurat::VlnPlot(seurat_data, features = gene2do, 
                      ncol = 4, pt.size = 0)
                    .png(print(p1), wj2, width = 5 + 8 * 4, height = 3 + 
                      6 * ceiling(length(gene2do)/4))
                  })
                }
            }
            {
                markers.specific.generality <- local({
                  wj2 <- hs.fp(wjj2, "specificity.xlsx")
                  if (file.exists(wj2)) 
                    return(wk.xls.read(wj2))
                  a <- tapply(a1[, gene], a1[, cluster], c)
                  ai <- inverseList(a)
                  b <- ai[sapply(ai, length) < 2] %>% wk.dt.fromList(nm = c("gene_name", 
                    "cell_cluster"))
                  b[, `:=`(generality, 1)]
                  b[, `:=`(sharing_cluster, "")]
                  n.co <- 2
                  local(repeat {
                    cc2do <- levels(Idents(seurat_data)) %not% 
                      b[, cell_cluster]
                    if (!length(cc2do)) 
                      break
                    if (n.co > 3) 
                      break
                    ai1 <- ai[sapply(ai, length) == n.co]
                    a1 <- inverseList(ai1)
                    for (cc2do.1 in cc2do %and% names(a1)) {
                      gene1 <- a1[[cc2do.1]]
                      sharing <- sapply(ai1[gene1], function(x) paste(sort(as.numeric(x %not% 
                        cc2do.1)), collapse = ";"))
                      dt1 <- wk.dt.as(list(gene1, cc2do.1, n.co, 
                        sharing))
                      b <<- wk.dt.rbind(b, dt1)
                    }
                    n.co <- n.co + 1
                  })
                  b <- b[order(as.numeric(cell_cluster))]
                  wk.xls.write(b, wj2)
                  b
                })
                null <- local({
                  cell_cluster.v <- markers.specific.generality[, 
                    unique(cell_cluster)]
                  nn <- max(nchar(cell_cluster.v))
                  for (i1 in seq_along(cell_cluster.v)) {
                    c1 <- cell_cluster.v[i1]
                    wj2 <- hs.fp(wjj2, "specific", paste0("cluster_", 
                      hs.str_pad(c1, nn, "0"), ".xlsx"))
                    if (file.exists(wj2)) 
                      next
                    a <- markers.specific.generality[cell_cluster %in% 
                      c1]
                    b <- wk.dt.subset(cell_cluster_markers, exp_l = c(cluster %in% 
                      c1, gene %in% a[, gene_name]))
                    wk.xls.write(b, wj2)
                  }
                })
                null <- local({
                  cell_cluster.v <- markers.specific.generality[, 
                    unique(cell_cluster)]
                  nn <- max(nchar(cell_cluster.v))
                  SeuratObject::DefaultAssay(seurat_data) <- "RNA"
                  lapply(seq_along(cell_cluster.v), function(i1) {
                    c1 <- cell_cluster.v[i1]
                    wj2 <- hs.fp(wjj2, "specific", paste0("cluster_", 
                      hs.str_pad(c1, nn, "0"), ".png"))
                    if (file.exists(wj2)) 
                      return()
                    hs.msg(paste("[38_04_24_232014]", wj2))
                    a <- markers.specific.generality[cell_cluster %in% 
                      c1]
                    b <- wk.dt.subset(cell_cluster_markers, exp_l = c(cluster %in% 
                      c1, gene %in% a[, gene_name]))
                    gene2do <- b[, head(gene, 8)]
                    P <- hs.catch(lapply(seq_along(gene2do), 
                      function(i2) {
                        gene1 <- gene2do[i2]
                        p1 <- Seurat::FeaturePlot(seurat_data, 
                          gene1, reduction = "umap") + ggplot2::labs(title = paste(gene1, 
                          "expression"))
                        hs.throw(p1)
                      }))
                    P <- append(P, list(Seurat::DimPlot(seurat_data, 
                      reduction = "umap", group.by = stable_most_weighted_cn, 
                      label = 1) + ggplot2::labs(title = "Cell clusters") + 
                      ggplot2::theme(legend.position = "none")), 
                      2)
                    P <- paste0("P[[", seq_along(P), "]]") %>% 
                      paste(collapse = " + ") %>% paste0("+ patchwork::plot_layout( ncol = 3)") %>% 
                      evalStr
                    .png(print(P), wj2, width = 8 * min(3, length(gene2do) + 
                      1) + 2, height = ceiling((length(gene2do) + 
                      1)/3) * 8)
                  })
                })
            }
        }
    }
    if (!only.de) {
        {
            SeuratObject::DefaultAssay(seurat_data) <- "RNA"
            qv.cn <- if (sample_n > 1) 
                "median_qval"
            else "p_val_adj"
            {
                a1 <- cell_cluster_markers[get(qv.cn) < 0.05]
                a <- a1[, hs.rowMins(abs(.SD)), .SDcols = c(hs.grepV(names(a1), 
                  "avg_log2FC"))]
                a1[, `:=`(lfc_abs_min, a)]
                wk.dt.setorderv(a1, c("cell_cluster", qv.cn, 
                  "lfc_abs_min"), c(1, 1, -1))
                gene2do <- a1[, head(gene_name, 10), by = cell_cluster][, 
                  unique(V1)]
                if (0) {
                  wj2 <- hs.fp(wjj2, "cell_cluster_markers.top.2dim.pdf")
                  hs.cat(hs.embed_pdf(wj2, wjj2), sep = "\n", 
                    file = report_wj)
                  if (!file.exists(wj2)) 
                    .pdf({
                      p1 <- Seurat::FeaturePlot(seurat_data, 
                        features = gene2do, min.cutoff = "q5", 
                        slot = "data")
                      print(p1)
                    }, wj2, width = 30, height = ceiling(length(gene2do)/3) * 
                      10)
                  wj2 <- hs.fp(wjj2, "cell_cluster_markers.top.bubble.pdf")
                  hs.cat(hs.embed_pdf(wj2, wjj2), sep = "\n", 
                    file = report_wj)
                  if (!file.exists(wj2)) 
                    .pdf({
                      p1 <- Seurat::DotPlot(seurat_data, features = gene2do, 
                        cols = c("blue", "red"), dot.scale = 8, 
                        split.by = "sample_name.38_01_19_171144") + 
                        Seurat::RotatedAxis()
                      print(p1)
                    }, wj2, 25, 30)
                }
                {
                  wj2 <- hs.fp(wjj2, "cell_cluster_markers.top.heatmap.png")
                  local({
                    if (file.exists(wj2)) 
                      return()
                    if (0) {
                      p1 <- Seurat::DoHeatmap(seurat_data, assay = "RNA", 
                        slot = "scale.data", features = gene2do, 
                        angle = 30, size = 3)
                      print(p1)
                    }
                    m1 <- seurat_data@assays[["RNA"]]@data[gene2do, 
                      , drop = FALSE] %>% apply(1, . %>% {
                      (. - mean(.))/sd(.)
                    }) %>% t
                    m1 <- m1[, Seurat::Idents(seurat_data)[colnames(m1)] %>% 
                      sort %>% names]
                    annotation_col <- hs.df(`Cell cluster` = Seurat::Idents(seurat_data)[colnames(m1)])
                    col1 <- c("darkolivegreen3", "lightsteelblue1", 
                      "dodgerblue4")
                    col1 <- c("navy", "white", "firebrick3")
                    col1 <- list(c(87, 87, 156), c(102, 186, 
                      159), c(251, 244, 185), c(231, 131, 84), 
                      c(162, 46, 74)) %>% sapply(ieiwk.col2rgb)
                    col1 <- list(c(87, 87, 156), c(102, 186, 
                      159), "white", c(231, 131, 84), c(162, 
                      46, 74)) %>% sapply(ieiwk.col2rgb)
                    col1 <- list(c(224, 224, 205), c(186, 214, 
                      232), c(65, 68, 129)) %>% sapply(ieiwk.col2rgb)
                    col1 <- (grDevices::colorRampPalette(c("purple", 
                      "black", "gold")))(100)
                    r2do <- seq(min(50, nrow(m1)))
                    p1 <- pheatmap::pheatmap(m1[r2do, ], color = (grDevices::colorRampPalette(col1, 
                      space = "Lab"))(100), breaks = seq(-2, 
                      2, length.out = 100), annotation_col = annotation_col, 
                      cluster_rows = 0, cluster_cols = 0, show_colnames = 0)
                    p1 <- ggplotify::as.ggplot(p1)
                    .figs(print(p1), hs.wjm(wj2, append = ".50"), 
                      width = 30, height = length(r2do) * 0.3 + 
                        5)
                    p1 <- pheatmap::pheatmap(m1, color = (grDevices::colorRampPalette(col1, 
                      space = "Lab"))(100), breaks = seq(-2, 
                      2, length.out = 100), annotation_col = annotation_col, 
                      cluster_rows = 0, cluster_cols = 0, show_colnames = 0)
                    p1 <- ggplotify::as.ggplot(p1)
                    .figs(print(p1), wj2, width = 30, height = length(gene2do) * 
                      0.3 + 5)
                  })
                }
            }
            {
                gene2do <- local({
                  if (0) {
                    cell_marker %>% hs.table
                    cell_cluster_markers[gene_name %in% "GPR171"] %>% 
                      .sh
                    cell_cluster_markers[cell_cluster %in% 10, 
                      "CSTB" %in% gene_name]
                  }
                  if (0) {
                    hs.save(cell_cluster_markers, wj = "~/a.rd")
                    "~/a.rd" %>% .ls
                    load("~/a.rd")
                  }
                  cell_cluster_markers_specific <- cell_cluster_markers[get(qv.cn) < 
                    0.05][, .SD[hs.rowAlls(.SD > 0), .SDcols = patterns("(?i)log2fc")]] %>% 
                    {
                      a <- tapply(.[, gene_name], .[, cell_cluster], 
                        c)
                      a <- inverseList(a)
                      names(a)[sapply(a, length) < 2L]
                    }
                  return(cell_cluster_markers[(get(qv.cn) < 0.05), 
                    head(gene_name %and% cell_cluster_markers_specific, 
                      3), by = cell_cluster][, V1])
                  if (0) {
                    wk.dt.subset(cell_cluster_markers, exp_l = c(median_pval < 
                      0.05, .SD[, hs.rowAlls(.SD > 0), .SDcols = patterns("(?i)log2fc")], 
                      cell_cluster %in% 5), exp_j = gene_name %and% 
                      cell_cluster_markers_specific)
                  }
                  if (0) 
                    local({
                      i1 <- hs.rowAlls(cell_cluster_markers[, 
                        .SD, .SDcols = patterns("(?i)log2fc")] > 
                        0)
                      cell_cluster_markers <- cell_cluster_markers[i1 & 
                        (median_qval < 0.05), head(.SD, 100), 
                        by = cell_cluster]
                      if (0) {
                        cell_cluster_markers[, hs.table(cell_cluster)]
                        cell_cluster_markers
                      }
                      cell_cluster.db <- wk.dt.subset(cell_cluster_markers, 
                        exp_l = c(toupper(gene_name) %in% toupper(names(cell_marker))), 
                        exp_j = unique(cell_cluster) %>% sort)
                      cell_n1 <- 1
                      repeat {
                        gene.db <- (cell_marker %>% {
                          toupper(names(.))[. <= cell_n1]
                        }) %and% toupper(cell_cluster_markers_specific)
                        a <- cell_cluster_markers[toupper(gene_name) %in% 
                          gene.db, toupper(head(gene_name, 3)), 
                          by = cell_cluster]
                        if (cell_cluster.db %all.in% a[, cell_cluster]) 
                          break
                        cell_n1 <- cell_n1 + 1
                        if (cell_n1 > max(cell_marker)) 
                          break
                      }
                      hs.c(a[, cell_cluster], cell_cluster_markers[match(a[, 
                        V1], toupper(gene_name)), gene_name])
                    })
                })
                if (0) {
                  gene2do %>% {
                    .[. %in% "GP9"]
                  }
                  gene1 <- "Rpl37a"
                }
                gene2do %<>% unique
                wj2 <- hs.fp(wjj2, "specific/cell_cluster_markers.specific.2dim")
                wk.mclapply(c("umap", "tsne"), function(reduction_method_1) {
                  wj2 <- hs.wjm(wj2, add = reduction_method_1)
                  co1 <- 9
                  local({
                    wj2 <- hs.wjm(wj2, append = if (length(gene2do) > 
                      co1) 
                      ".first_9")
                    if (file.exists(paste0(wj2, ".png"))) 
                      return()
                    i2do <- seq(min(co1, length(gene2do)))
                    p1 <- Seurat::FeaturePlot(seurat_data, features = gene2do[i2do], 
                      min.cutoff = "q5", slot = "data", reduction = reduction_method_1)
                    .figs(print(p1), wj2, width = 30, height = 10 * 
                      ceiling(length(i2do)/3))
                  })
                  if (length(gene2do) > co1) 
                    local({
                      wj2 <- hs.wjm(wj2, add = ".remaining.png")
                      if (file.exists(wj2)) 
                        return()
                      i2do <- seq(co1 + 1, length(gene2do))
                      p1 <- Seurat::FeaturePlot(seurat_data, 
                        features = gene2do, min.cutoff = "q5", 
                        slot = "data", reduction = reduction_method_1)
                      .figs(print(p1), wj2, width = 40, height = 10 * 
                        ceiling(length(i2do)/3))
                    })
                })
                wj2 <- hs.fp(wjj2, "specific/cell_cluster_markers.specific.bubble.png")
                local({
                  if (file.exists(wj2)) 
                    return()
                  {
                    seurat_data@meta.data %=>% {
                      a <- expand.grid(unique(x[, "sample_name.38_01_19_171144"]), 
                        unique(x[, "cell_type"]), gene2do)
                      colnames(a) <- c("sample", "cluster", "gene")
                      a <- wk.dt.as(a, "row_name")
                      a[, `:=`("cluster_sample", paste0(cluster, 
                        "_", sample) %>% hs.factor(unique(.), 
                        N = 0))]
                    }
                    df <- local({
                      a <- seurat_data@meta.data
                      a <- expand.grid(unique(a[, "sample_name.38_01_19_171144"]), 
                        unique(a[, "cell_type"]), gene2do)
                      colnames(a) <- c("sample", "cluster", "gene")
                      a <- wk.dt.as(a, "row_name")
                      a[, `:=`("cluster_sample", paste0(cluster, 
                        "_", sample) %>% hs.factor(unique(.), 
                        N = 0))]
                    })
                    if (0) {
                      1
                      {
                        ge <- seurat_data@assays[["RNA"]]@data[gene2do, 
                          , drop = FALSE]
                        a <- apply(ge > 0, 1, tapply, seurat_data@meta.data[, 
                          c("sample_name.38_01_19_171144", "clustering_most_cell_type_cn.cell_type")], 
                          mean)
                        df[, `:=`("percent", a * 100)]
                      }
                      {
                        {
                          a <- apply(ge, 1, tapply, seurat_data@meta.data[, 
                            c("sample_name.38_01_19_171144", 
                              "clustering_most_cell_type_cn.cell_type")], 
                            function(x) median(x[x > 0]))
                          a <- apply(a, 2, . %>% {
                            (. - median(., na.rm = TRUE))/mad(., 
                              na.rm = TRUE)
                          })
                        }
                        {
                          a <- apply(ge, 1, tapply, seurat_data@meta.data[, 
                            c("sample_name.38_01_19_171144", 
                              "clustering_most_cell_type_cn.cell_type")], 
                            function(x1) log1p(mean(expm1(x1) + 
                              1)))
                          x1 <- c(1.2, 1.4)
                          log(mean(exp(x1)))
                          log(mean(expm1(x1) + 1))
                          log1p
                          mean(x1)
                          hs.log_mean(x1)
                          a <- apply(a, 2, . %>% {
                            (. - mean(., na.rm = TRUE))/sd(., 
                              na.rm = TRUE)
                          })
                        }
                        a[a > 2] <- 2
                        a[a < -2] <- -2
                        df[, `:=`("Expression", a)]
                      }
                    }
                    {
                      a <- seurat_data@meta.data[, c("sample_name.38_01_19_171144", 
                        "cell_type")]
                      a <- unique(a)
                      a <- as.matrix(a)
                      null <- apply(a, 1, . %>% {
                        sample1 <- .[1]
                        cluster1 <- .[2]
                        sample.bc <- seurat_data@meta.data %>% 
                          {
                            rownames(.)[.[, "sample_name.38_01_19_171144"] %in% 
                              sample1]
                          }
                        cluster.bc <- seurat_data@meta.data[sample.bc, 
                          ] %>% {
                          rownames(.)[.[, "cell_type"] %in% cluster1]
                        }
                        other.bc <- sample.bc %not% cluster.bc
                        a <- Seurat::FoldChange(seurat_data@assays[["RNA"]], 
                          cluster.bc, other.bc, features = gene2do)
                        rn1 <- df[sample %in% sample1][cluster %in% 
                          cluster1][, row_name[match(gene2do, 
                          gene)]]
                        i1 <- df[, match(rn1, row_name)]
                        wk.dt.set(df, i1, "Percent", a[, "pct.1"])
                        wk.dt.set(df, i1, "L2FC", a[, "avg_log2FC"])
                      })
                      df[L2FC > 2, `:=`(L2FC, 2)]
                      df[L2FC < -2, `:=`(L2FC, -2)]
                      df[, `:=`(Percent, Percent * 100)]
                    }
                    p1 <- ggplot2::ggplot(df, ggplot2::aes(gene, 
                      cluster_sample, size = Percent, color = L2FC)) + 
                      ggplot2::geom_point() + ggplot2::labs(x = NULL, 
                      y = NULL) + ggplot2::theme_bw() + ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, 
                      hjust = 1), panel.grid = ggplot2::element_blank())
                  }
                  if (0) {
                    p1 <- Seurat::DotPlot(seurat_data, features = gene2do, 
                      cols = c("blue", "red", "purple")[1:length(sce_list)], 
                      dot.scale = 8, split.by = "sample_name.38_01_19_171144") + 
                      RotatedAxis()
                  }
                  .figs(print(p1), wj2, width = 10 + length(gene2do) * 
                    0.8, height = 7 + length(levels(Idents(seurat_data))) * 
                    2 * 0.9)
                })
                wj2 <- hs.fp(wjj2, "specific/cell_cluster_markers.specific.heatmap.png")
                local({
                  if (file.exists(wj2)) 
                    return()
                  m1 <- seurat_data@assays[["RNA"]]@data[gene2do, 
                    , drop = FALSE] %>% apply(1, . %>% {
                    (. - mean(.))/sd(.)
                  }) %>% t
                  if (0) {
                    m1 <- seurat_data@assays[["RNA"]]@scale.data[gene2do, 
                      , drop = 0]
                  }
                  m1 <- m1[, Seurat::Idents(seurat_data)[colnames(m1)] %>% 
                    sort %>% names]
                  annotation_col <- hs.df(`Cell cluster` = Seurat::Idents(seurat_data)[colnames(m1)])
                  p1 <- (pheatmap::pheatmap)(m1, color = colorRampPalette(c("purple", 
                    "black", "gold"))(100), breaks = seq(-2, 
                    2, length.out = 100), annotation_col = annotation_col, 
                    cluster_rows = 0, cluster_cols = 0, show_colnames = 0)
                  p1 <- ggplotify::as.ggplot(p1)
                  .figs(print(p1), wj2, width = 30, height = 20)
                  if (0) {
                    p1 <- Seurat::DoHeatmap(seurat_data, assay = "RNA", 
                      slot = "scale.data", features = gene2do, 
                      angle = 30, size = 3)
                    print(p1)
                  }
                })
                wj2 <- hs.fp(wjj2, "specific/cell_cluster_markers.specific.violin.png")
                local({
                  if (file.exists(wj2)) 
                    return()
                  p1 <- Seurat::VlnPlot(seurat_data, features = gene2do, 
                    ncol = 4, pt.size = 0)
                  .figs(print(p1), wj2, width = 5 + 8 * 4, height = 3 + 
                    6 * ceiling(length(gene2do)/4))
                })
            }
        }
        {
            cell_cluster_markers.specific.generality <- local({
                wj2 <- hs.fp(wjj2, "specificity.tsv.gz")
                if (file.exists(wj2)) {
                  wk.fread(wj2)
                }
                else {
                  . <- wk.dt.subset(cell_cluster_markers, exp_l = c(get(qv.cn) < 
                    0.05, .SD[, hs.rowAlls(.SD > 0), .SDcols = patterns("(?i)log2fc")]))
                  a <- tapply(.[, gene_name], .[, cell_cluster], 
                    c)
                  ai <- inverseList(a)
                  b <- ai[sapply(ai, length) < 2] %>% wk.dt.fromList(nm = c("gene_name", 
                    "cell_cluster"))
                  b[, `:=`(generality, 1)]
                  b[, `:=`(sharing_cluster, "")]
                  n.co <- 2
                  local(repeat {
                    cc2do <- levels(Idents(seurat_data)) %not% 
                      b[, cell_cluster]
                    if (!length(cc2do)) 
                      break
                    if (n.co > 3) 
                      break
                    ai1 <- ai[sapply(ai, length) == n.co]
                    a1 <- inverseList(ai1)
                    for (cc2do.1 in cc2do %and% names(a1)) {
                      gene1 <- a1[[cc2do.1]]
                      sharing <- sapply(ai1[gene1], function(x) paste(sort(as.numeric(x %not% 
                        cc2do.1)), collapse = ";"))
                      dt1 <- wk.dt.as(list(gene1, cc2do.1, n.co, 
                        sharing))
                      b <<- wk.dt.rbind(b, dt1)
                    }
                    n.co <- n.co + 1
                  })
                  b <- b[order(as.numeric(cell_cluster))]
                  wk.dt.fwrite(b, wj2)
                  b
                }
            })
            null <- local({
                cell_cluster.v <- cell_cluster_markers.specific.generality[, 
                  unique(cell_cluster)]
                nn <- max(nchar(cell_cluster.v))
                for (i1 in seq_along(cell_cluster.v)) {
                  c1 <- cell_cluster.v[i1]
                  wj2 <- hs.fp(wjj2, "specific", paste0("cluster_", 
                    hs.str_pad(c1, nn, "0"), ".tsv.gz"))
                  if (file.exists(wj2)) 
                    next
                  a <- cell_cluster_markers.specific.generality[cell_cluster %in% 
                    c1]
                  b <- wk.dt.subset(cell_cluster_markers, exp_l = c(cell_cluster %in% 
                    c1, gene_name %in% a[, gene_name]))
                  wk.dt.fwrite(b, wj2)
                }
            })
            null <- local({
                cell_cluster.v <- cell_cluster_markers.specific.generality[, 
                  unique(cell_cluster)]
                nn <- max(nchar(cell_cluster.v))
                SeuratObject::DefaultAssay(seurat_data) <- "RNA"
                wk.mclapply(seq_along(cell_cluster.v), function(i1) {
                  c1 <- cell_cluster.v[i1]
                  wj2 <- hs.fp(wjj2, "specific", paste0("cluster_", 
                    hs.str_pad(c1, nn, "0"), ".png"))
                  if (file.exists(wj2)) 
                    return()
                  hs.msg(paste("[38_04_24_232014]", wj2))
                  a <- cell_cluster_markers.specific.generality[cell_cluster %in% 
                    c1]
                  b <- wk.dt.subset(cell_cluster_markers, exp_l = c(cell_cluster %in% 
                    c1, gene_name %in% a[, gene_name]))
                  gene2do <- b[, head(gene_name, 8)]
                  P <- hs.catch(lapply(seq_along(gene2do), function(i2) {
                    gene1 <- gene2do[i2]
                    p1 <- Seurat::FeaturePlot(seurat_data, gene1, 
                      reduction = "umap") + ggplot2::labs(title = paste(gene1, 
                      "expression"))
                    hs.throw(p1)
                  }))
                  P <- append(P, list(Seurat::DimPlot(seurat_data, 
                    reduction = "umap", label = 1) + ggplot2::labs(title = "Cell clusters") + 
                    ggplot2::theme(legend.position = "none")), 
                    after = 2)
                  P <- paste0("P[[", seq_along(P), "]]") %>% 
                    paste(collapse = " + ") %>% paste0("+ patchwork::plot_layout( ncol = 3)") %>% 
                    evalStr
                  .figs(print(P), wj2, width = 8 * min(3, length(gene2do) + 
                    1) + 2, height = ceiling((length(gene2do) + 
                    1)/3) * 8)
                })
            })
        }
    }
    cell_cluster_markers
}
"38_10_18_154053" <- function(seurat_data, comparison, wjj2) {
    comparison_name <- hs.hsgly("1.流程/做项目/样本组比较/比较的名称.38_09_16_095937")(comparison)
    if ("cell_cluster.sample_name" %not.in% colnames(seurat_data@meta.data)) 
        seurat_data[["cell_cluster.sample_name"]] <- paste0(Idents(seurat_data), 
            "_", seurat_data$sample_name.38_01_19_171144)
    nn <- nchar(length(comparison))
    cluster <- SeuratObject::Idents(seurat_data) %=>% {
        if (is.factor(x)) 
            levels(x)
        else unique(x)
    }
    cluster %<=>% x[order(as.numeric(x))]
    if (max(sapply(comparison, length)) > 2) {
        hs.browser("38.10.18.154649")
        hs.require("limma")
        null <- seq_along(comparison) %=>% lapply(x, function(comparison_i1) {
            comparison1 <- comparison[[comparison_i1]]
            a <- if (is.null(names(comparison))) {
                paste0("方案", hs.str_pad(comparison_i1, nn, 
                  "0"))
            }
            else hs.browser("38.06.20.171523")
            hs.msg(a)
            wjj2.38_06_20_172049 <- hs.fp(wjj2, a)
            nn <- max(nchar(cluster))
            lapply(cluster, function(cell_cluster_1) {
                hs.msg(paste0("细胞群", cell_cluster_1), 2)
                if (debug) 
                  hs.browser("38.08.25.170748")
                wj2 <- wjj2.38_06_20_172049 %=>% hs.fp(paste0("cluster_", 
                  hs.str_pad(cell_cluster_1, nn, "0")), "dif.tsv.gz")
                a <- hs.evalLazy({
                  groups <- lapply(comparison1, . %=>% paste0(cell_cluster_1, 
                    "_", x))
                  if (lapply(groups, "%and%", levels(SeuratObject::Idents(seurat_data))) %=>% 
                    sapply(length) %=>% (x > 0) %=>% (sum(x) < 
                    2)) 
                    return()
                  if (max(sapply(groups, length)) > 1) 
                    hs.browser("38.03.13.215434")
                  m1 <- seurat_data@assays[["RNA"]]@data[, Idents(seurat_data) %in% 
                    groups]
                  a <- hs.catch(hs.for(ref_i, c(length(groups), 
                    seq(length(groups) - 2)), si <- wk.dt(cluster.sample = Idents(seurat_data)[colnames(m1)] %=>% 
                    hs.factor(x, unlist(groups) %=>% c(x[ref_i], 
                      x[-ref_i]), N = 0)), {
                    design <- si[, model.matrix(~cluster.sample)]
                    fit <- limma::lmFit(m1, design)
                    efit <- limma::eBayes(fit, trend = TRUE, 
                      robust = TRUE)
                    coef <- if (ref_i == length(groups)) 
                      seq(2, ref_i)
                    else seq(2 + ref_i - 1, length(groups) - 
                      1)
                  }, coef1, coef, {
                    de <- limma::topTreat(efit, coef1, number = nrow(m1)) %=>% 
                      wk.dt.as("gene_id_symbol")
                    hs.throw(paste0(substring(colnames(design)[coef1], 
                      nchar(colnames(si)) + 1), " - ", groups[[ref_i]]), 
                      de[order(adj.P.Val, -abs(logFC))])
                  }), named = 1)
                  b <- wk.dt.fromList(a, nm = "comparison")
                  b[, `:=`(comparison, comparison %=>% hs.sub("^\\d+_") %=>% 
                    hs.sub(" - \\d+_", " - "))]
                }, fp = wj2)
                for (comparison1 in a[, unique(comparison)]) {
                  fig.wj <- wjj2.38_06_20_172049 %=>% hs.fp(paste0("cluster_", 
                    hs.str_pad(cell_cluster_1, nn, "0")), "火山图", 
                    paste0(comparison1, ".png"))
                  if (wj.update.reporter(fig.wj, wj2)) {
                    a1 <- a[comparison %in% comparison1]
                    p1 <- hs.hsgly("工作/生物学/生信/可视化/火山图.36.05.11.151845")(a1, 
                      subtitle = paste0(comparison1, ", Cell cluster ", 
                        cell_cluster_1), plot = 0, fc.co = 2^0.25)
                    .png(print(p1), fig.wj, width = 15, height = 18)
                  }
                }
            })
        })
    }
    else {
        cell_cluster_de <- seq_along(comparison) %=>% lapply(x, 
            function(comparison_i1) {
                comparison1 <- comparison[[comparison_i1]]
                wj2 <- hs.fp(wjj2, comparison_name[comparison_i1], 
                  "dif.tsv.gz")
                hs.msg(comparison_name[comparison_i1])
                hs.evalLazy({
                  a1 <- wk.mclapply(cluster, function(cell_cluster_1) {
                    hs.msg(paste0("细胞群", cell_cluster_1), 
                      2)
                    groups <- lapply(comparison1, . %=>% {
                      i <- seurat_data@meta.data[, "sample_name.38_01_19_171144"] %in% 
                        x
                      seurat_data@meta.data[, "sample_name.38_01_19_171144"] %=>% 
                        hs.table
                      a <- SeuratObject::Idents(seurat_data)[i]
                      names(a)[a %in% cell_cluster_1]
                    })
                    if (any(sapply(groups, length) < 3)) 
                      return()
                    a <- (Seurat::FindMarkers)(seurat_data@assays[["RNA"]], 
                      cells.1 = groups[[1]], cells.2 = groups[[2]], 
                      verbose = 0)
                    wk.dt.as(a, "gene_id_symbol")
                  })
                  names(a1) <- cluster
                  wk.dt.fromList(hs.nonempty(a1), nm = "cell_cluster")
                }, fp = wj2)
            })
    }
    names(cell_cluster_de) <- comparison_name
    if (0) {
        g1 <- "HES4"
        g1 <- "HS3ST1"
        v1 <- seurat_data@assays[["RNA"]]@data[g1, colnames(seurat_data)[seurat_data$cell_cluster.sample_name %in% 
            groups[2]]]
        v2 <- seurat_data@assays[["RNA"]]@data[g1, colnames(seurat_data)[seurat_data$cell_cluster.sample_name %in% 
            groups[1]]]
        mean(v1) - mean(v2)
        wilcox.test(v1, v2)
        a <- FindMarkers(seurat_data, cells.1 = colnames(seurat_data)[seurat_data$cell_cluster.sample_name %in% 
            groups[2]], cells.2 = colnames(seurat_data)[seurat_data$cell_cluster.sample_name %in% 
            groups[1]], verbose = 0)
        a <- FindAllMarkers(seurat_data, only.pos = 1, min.pct = 0.1, 
            logfc.threshold = 0.25)
    }
    {
        {
            null <- seq_along(comparison) %=>% lapply(x, function(comparison_i1) {
                comparison1 <- comparison[[comparison_i1]]
                i2do <- cell_cluster_de[[comparison_i1]][, unique(cell_cluster) %=>% 
                  as.numeric %=>% sort]
                nn <- max(nchar(i2do))
                wk.mclapply(i2do, function(ci1) {
                  a <- cell_cluster_de[[comparison_i1]][cell_cluster %in% 
                    ci1]
                  if (hs.is_empty(a)) 
                    return()
                  wj2 <- hs.fp(wjj2, comparison_name[comparison_i1], 
                    "volcano", paste0("cell_cluster", hs.str_pad(ci1, 
                      nn, 0), ".png"))
                  if (file.exists(wj2)) 
                    return()
                  p1 <- hs.hsgly("工作/生物学/生信/可视化/火山图.36.05.11.151845")(a, 
                    subtitle = paste("Cell cluster", ci1), plot = 0, 
                    fc.co = 2^0.25)
                  .figs(print(p1), wj2, width = 15)
                })
            })
        }
        if (0) {
            gene2do <- lapply(cell_cluster_de, . %>% {
                rownames(.)[.[, "p_val_adj"] < 0.05]
            } %>% head(5)) %>% hs.unlist %>% hs.table %>% sort %>% 
                head(10)
            FeaturePlot(seurat_data, features = c("HES4"), split.by = "sample_name.38_01_19_171144", 
                max.cutoff = 3, cols = c("grey", "red"))
            FeaturePlot(seurat_data, features = c("HES4"), cells = colnames(seurat_data)[seurat_data$cell_cluster.sample_name %in% 
                groups], split.by = "sample_name.38_01_19_171144", 
                max.cutoff = 3, cols = c("grey", "red"))
        }
    }
    cell_cluster_de
}
"38_10_08_204152" <- function(seurat_data, org1, wjj2) {
    ccg <- hs.hsgly("seurat/细胞周期基因.38_01_16_235954")()
    if (org1 != "human") {
        if (0) {
            hs.save(ccg, org1, wj = "~/a.rd")
            load("~/a.rd")
            . <- ccg[[1]]
            . <- hs.hsgly("同源基因.37_05_26_124410")(., 
                org1 = "human", org2 = org1, ft = "symbol", one_to_one = 0)[[2]]
        }
        ccg <- lapply(ccg, . %=>% {
            hs.hsgly("同源基因.37_05_26_124410")(x, org1 = "human", 
                org2 = org1, ft = "symbol", one_to_one = 0)[[2]] %=>% 
                {
                  a <- rownames(seurat_data@assays[["RNA"]])
                  a[tolower(a) %in% tolower(x)]
                }
        })
    }
    assay.38_01_17_001034 <- SeuratObject::DefaultAssay(seurat_data)
    SeuratObject::DefaultAssay(seurat_data) <- "RNA"
    seurat_data <- Seurat::CellCycleScoring(seurat_data, s.features = ccg[["s.genes"]], 
        g2m.features = ccg[["g2m.genes"]], set.ident = FALSE)
    if (0) {
        colnames(seurat_data@meta.data) %>% .sh
        seurat_data@meta.data[, "S.Score"] %>% summary
        seurat_data@meta.data[, "Phase"] %>% hs.table
        seurat_data@meta.data %>% {
            .[, "Phase"] %>% hs.table
            list(tapply(.[, "S.Score"], .[, "Phase"], summary), 
                tapply(.[, "G2M.Score"], .[, "Phase"], summary))
        }
    }
    hs.with_wd(hs.wjj(wjj2), {
        wj2 <- "barcode.cell_phase.xlsx"
        if (file.exists(wj2)) 
            return()
        a <- wk.dt.as(seurat_data@meta.data[, c("S.Score", "G2M.Score", 
            "Phase")], "barcode")
        wk.xls.write(a, wj2)
    })
    {
        wj2 <- hs.fp(wjj2, "cell_cycle.scores.png")
        if (!file.exists(wj2)) 
            local({
                sj1 <- seurat_data@meta.data
                df1 <- hs.df(sj1[, c("S.Score", "G2M.Score", 
                  "Phase")])
                P <- ggplot2::ggplot(df1, ggplot2::aes(S.Score, 
                  G2M.Score)) + ggplot2::geom_point(ggplot2::aes(color = Phase)) + 
                  ggplot2::theme(aspect.ratio = 1) + ggplot2::theme(legend.title = ggplot2::element_blank(), 
                  legend.key = ggplot2::element_rect(fill = NA)) + 
                  ggplot2::theme(panel.background = {
                  }, panel.grid = ggplot2::element_blank())
                .png(print(P), wj2)
            })
        wj2 <- hs.fp(wjj2, "cell_cycle.dim2.png")
        if (!file.exists(wj2)) 
            local({
                P <- hs.lapply("umap,tsne", . %>% {
                  dt1 <- data.table::as.data.table(seurat_data@reductions[[.]] %>% 
                    attr("cell.embeddings"))
                  dt1[, `:=`(phase, seurat_data@meta.data[, "Phase"])]
                  e <- bquote(ggplot2::ggplot(dt1, ggplot2::aes(.(as.symbol(names(dt1)[1])), 
                    .(as.symbol(names(dt1)[2])), color = phase)) + 
                    ggplot2::geom_point(size = rep(0.2, ncol(seurat_data))) + 
                    ggplot2::theme(aspect.ratio = 1) + ggplot2::theme(legend.title = ggplot2::element_blank(), 
                    legend.key = ggplot2::element_rect(fill = NA)) + 
                    ggplot2::theme(panel.background = {
                    }, panel.grid = ggplot2::element_blank()))
                  eval(e)
                })
                P <- P[[1]] + P[[2]]
                .png(print(P), wj2, width = 25, height = 10)
            })
        wj2 <- hs.fp(wjj2, "cell_cycle.ratio_within_cluster.png")
        if (!file.exists(wj2)) 
            local({
                dt1 <- tapply(seurat_data@meta.data[, "Phase"], 
                  SeuratObject::Idents(seurat_data), . %>% hs.table %>% 
                    wk.dt.as) %>% wk.dt.fromList
                wk.dt.setnames(dt1, 2, "phase")
                dt1[, `:=`(group, group %=>% hs.factor(N = 0))]
                P <- ggplot2::ggplot(dt1, ggplot2::aes(group, 
                  N)) + ggplot2::geom_bar(ggplot2::aes(fill = phase), 
                  stat = "identity", position = "stack") + ggplot2::xlab("Cell cluster") + 
                  ggplot2::ylab("#cell") + ggplot2::theme(axis.text.x = element_text(angle = 45, 
                  vjust = 1, hjust = 1)) + ggplot2::theme(legend.position = "top", 
                  legend.direction = "horizontal", legend.title = ggplot2::element_blank()) + 
                  ggplot2::theme(panel.background = {
                  }, panel.grid = ggplot2::element_blank())
                .png(print(P), wj2, width = 10, height = 10)
            })
    }
    SeuratObject::DefaultAssay(seurat_data) <- assay.38_01_17_001034
    seurat_data
}
"38_10_08_230829" <- function(seurat_data, wjj2, org1, sample_name.cn, 
    comparison = {
    }, comparison_name = hs.hsgly("1.流程/做项目/样本组比较/比较的名称.38_09_16_095937")(comparison)) {
    wj2 <- hs.fp(wjj2, "gsva")
    gsva <- hs.evalLazy(hs.hsgly("gsva.38_01_06_171637")(seurat_data@assays[["RNA"]]@data, 
        sample_group = Idents(seurat_data), org1 = org1, wjj2 = wjj2, 
        subset_size = 2000), fp = hs.wjm(wj2, add = "rd"))
    topN <- 20
    wk.mclapply(seq_along(gsva), function(i1) {
        message(i1)
        an1 <- names(gsva)[i1]
        wj2 <- hs.fp(wjj2, paste0(an1, ".top_", topN, ".png"))
        if (file.exists(wj2)) 
            return()
        gsva.an1 <- gsva[[an1]]
        cg <- Idents(seurat_data)[colnames(gsva.an1)]
        statistic <- apply(gsva.an1, 1, function(x1) {
            kruskal.test(x1, cg)[["statistic"]]
        })
        gs2show <- statistic %>% {
            names(.)[hs.order(.) %>% head(topN)]
        }
        m1 <- gsva.an1[gs2show, ]
        cn_ordered <- colnames(m1) %>% {
            tapply(., cg[.], c)
        } %>% lapply(. %>% {
            a <- hclust(as.dist(1 - cor(m1[, .], method = "spearman")))
            .[a[["order"]]]
        }) %>% hs.unlist
        ann_col <- hs.df(`Cell cluster` = cg)
        p1 <- pheatmap(m1[, cn_ordered], silent = TRUE, main = paste0("GSVA; ", 
            toupper(an1), ", top ", topN, "; randomly picked cells"), 
            legend = TRUE, annotation_legend = FALSE, legend_labels = "xxx", 
            annotation_col = ann_col, cluster_cols = FALSE, color = colorRampPalette(c("steelblue", 
                "white", "tomato"))(100), breaks = local({
                r1 <- range(as.vector(m1))
                uniqueN(sign(r1)) == 2
                n1 <- round((-r1[1])/diff(r1) * 100)
                unique(c(seq(r1[1], 0, length.out = n1), seq(0, 
                  r1[2], length.out = 100 - n1 + 1)))
            }), show_colnames = FALSE, show_rownames = TRUE, 
            border_color = NA)
        p1 <- ggplotify::as.ggplot(p1)
        a1 <- tapply(names(cg), cg, . %>% {
            hs.rowMedians(gsva.an1[gs2show, .])
        })
        i1 <- order(as.numeric(names(a1)))
        if (is.unsorted(i1)) 
            a1 <- a1[i1]
        m1 <- do.call("cbind", unname(a1))
        rownames(m1) <- gs2show
        colnames(m1) <- names(a1)
        ann_col <- hs.df(`Cell cluster` = colnames(m1))
        p2 <- pheatmap(m1, silent = TRUE, main = paste0("GSVA; ", 
            toupper(an1), ", top ", topN, "; randomly picked cells, median"), 
            annotation_col = ann_col, cluster_cols = FALSE, color = colorRampPalette(c("steelblue", 
                "white", "tomato"))(100), breaks = local({
                r1 <- range(as.vector(m1))
                uniqueN(sign(r1)) == 2
                n1 <- round((-r1[1])/diff(r1) * 100)
                unique(c(seq(r1[1], 0, length.out = n1), seq(0, 
                  r1[2], length.out = 100 - n1 + 1)))
            }), show_colnames = FALSE, show_rownames = TRUE, 
            border_color = NA)
        p2 <- ggplotify::as.ggplot(p2)
        P <- p1 + p2
        .figs(print(P), wj2, 100, 10)
    })
    sample_n <- seurat_data@meta.data[, sample_name.cn] %>% uniqueN
    if ((sample_n > 1) && length(comparison)) 
        local({
            wjj2 <- hs.fp(wjj2, "mid", "GSVA", "组间差异")
            wj2 <- hs.fp(wjj2, "test.tsv.gz")
            if (!file.exists(wj2)) {
                hs.require("limma")
                bc <- colnames(gsva[[1]])
                dt1 <- wk.dt(bc)
                dt1[, `:=`("sample", seurat_data@meta.data[bc, 
                  sample_name.cn])]
                dt1[, `:=`("cluster", SeuratObject::Idents(seurat_data)[bc] %>% 
                  {
                    hs.factor(., unique(sort(.)), N = 0)
                  })]
                cluster2do <- dt1[, paste(sort(as.numeric(unique(cluster))))]
                a <- lapply(seq_along(cluster2do), function(i1) {
                  cluster1 <- cluster2do[i1]
                  a <- lapply(seq_along(comparison), function(i1) {
                    comparison1 <- comparison[[i1]]
                    comparison1.name <- hs.hsgly("1.流程/做项目/样本组比较/比较的名称.38_09_16_095937")(comparison[[i1]], 
                      collapse = "")
                    bc1 <- wk.dt.subset(dt1, exp_l = c(cluster %in% 
                      cluster1, sample %in% comparison1[[1]]), 
                      exp_j = bc)
                    bc2 <- wk.dt.subset(dt1, exp_l = c(cluster %in% 
                      cluster1, sample %in% comparison1[[2]]), 
                      exp_j = bc)
                    if (!(length(bc1) && length(bc2))) 
                      return()
                    a <- lapply(seq_along(gsva), function(i1) {
                      m1 <- gsva[[i1]][, c(bc1, bc2)]
                      group <- ifelse(colnames(m1) %in% bc1, 
                        comparison1.name[1], comparison1.name[2]) %>% 
                        hs.factor(comparison1.name, N = 0)
                      design <- stats::model.matrix(~0 + group)
                      colnames(design) <- make.names(levels(group))
                      rownames(design) <- colnames(m1)
                      compare <- evalbq(limma::makeContrasts(.(as.symbol(colnames(design)[1])) - 
                        .(as.symbol(colnames(design)[2])), levels = design))
                      fit <- lmFit(m1, design)
                      fit2 <- limma::contrasts.fit(fit, compare)
                      fit3 <- limma::eBayes(fit2, trend = FALSE, 
                        robust = FALSE)
                      Diff <- limma::topTable(fit3, coef = 1, 
                        number = nrow(m1))
                      wk.dt.setDT(Diff, "gene_set")
                      wk.dt.setorderv(Diff, "P.Value")
                    })
                    names(a) <- names(gsva)
                    wk.dt.fromList(a, nm = "db")
                  })
                  names(a) <- comparison_name
                  wk.dt.fromList(a, nm = "comparison")
                })
                names(a) <- cluster2do
                a <- wk.dt.fromList(a, nm = "cell_cluster")
                a <- a[order(as.numeric(cell_cluster), comparison, 
                  db, P.Value)]
                wk.dt.fwrite(a, wj2)
            }
            {
                dt1 <- wk.fread(wj2)
                dt2 <- dt1[, 1, by = .(comparison, cell_cluster, 
                  db)]
                dt2[, `:=`(V1, {
                })]
                if (0) {
                  i1 <- wk.dt.subset(dt2, exp_l = c(comparison %in% 
                    "LT-X - LF-X", cell_cluster %in% 2, db %in% 
                    "go.bp"), ot = "i")
                  dt2[i1]
                  hs.save(dt2, wj = "~/a.rd")
                  wk.rsync.in.same_place("~/a.rd", ssh = "r820")
                  load("~/a.rd")
                  wk.dt.subset(dt2, dt2[comparison %in% "LF-X - S"][cell_cluster %in% 
                    4][db %in% "go.cc"], ot = "i")
                  wk.dt.index(dt2)
                  data.table::setkeyv
                  apropos("indi")
                }
                null <- wk.mclapply(dt2[, seq(.N)], function(i1) {
                  message(i1, " of ", dt2[, .N])
                  nn <- wk.dt.subset(dt2, dt2[i1, .(comparison, 
                    db)], exp_j = max(nchar(cell_cluster)))
                  wj2 <- dt2[i1, hs.fp(wjj2, comparison, paste0("cell_cluster_", 
                    hs.str_pad(cell_cluster, nn, 0)), paste0(db))]
                  if (all(file.exists(paste0(wj2, c(".pdf", ".png"))))) 
                    return()
                  a <- wk.dt.subset(dt1, dt2[i1])
                  if (0) {
                    file.exists(wj2)
                    hs.wj.rm(wj2)
                    hs.save(a, wj = "~/a.rd")
                    wk.rsync.in.same_place("~/a.rd", ssh = "r820")
                    load("~/a.rd")
                    a
                    wj2 <- "~/tmp.pdf"
                    hs.wj.rm(wj2)
                    hs.hsgly("生信/可视化/基因富集分析/头部条形图.38_04_26_214925")(a, 
                      x = "t", y = "gene_set", p = "P.Value", 
                      q = "adj.P.Val", wj2 = wj2, width = 20, 
                      height = 16)
                  }
                  hs.hsgly("生信/可视化/基因富集分析/头部条形图.38_04_26_214925")(a, 
                    x = "t", y = "gene_set", p = "P.Value", q = "adj.P.Val", 
                    xlab = paste0("t statistic of GSVA score, ", 
                      dt2[i1, comparison]), wj2 = wj2)
                })
            }
        })
}
"39_01_17_143300" <- local({
    monocle2.hs <- local({
        orderCells <- function(cds, root_state = NULL, num_paths = NULL, 
            reverse = NULL) {
            if (class(cds)[1] != "CellDataSet") {
                stop("Error cds is not of type 'CellDataSet'")
            }
            if (is.null(cds@dim_reduce_type)) {
                stop("Error: dimensionality not yet reduced. Please call reduceDimension() before calling this function.")
            }
            if (any(c(length(cds@reducedDimS) == 0, length(cds@reducedDimK) == 
                0))) {
                stop("Error: dimension reduction didn't prodvide correct results. Please check your reduceDimension() step and ensure correct dimension reduction are performed before calling this function.")
            }
            root_cell <- monocle:::select_root_cell(cds, root_state, 
                reverse)
            cds@auxOrderingData <- new.env(hash = TRUE)
            if (cds@dim_reduce_type == "ICA") {
                if (is.null(num_paths)) {
                  num_paths = 1
                }
                adjusted_S <- t(cds@reducedDimS)
                dp <- as.matrix(dist(adjusted_S))
                cellPairwiseDistances(cds) <- as.matrix(dist(adjusted_S))
                gp <- graph.adjacency(dp, mode = "undirected", 
                  weighted = TRUE)
                dp_mst <- minimum.spanning.tree(gp)
                minSpanningTree(cds) <- dp_mst
                next_node <<- 0
                res <- pq_helper(dp_mst, use_weights = FALSE, 
                  root_node = root_cell)
                cds@auxOrderingData[[cds@dim_reduce_type]]$root_cell <- root_cell
                order_list <- extract_good_branched_ordering(res$subtree, 
                  res$root, cellPairwiseDistances(cds), num_paths, 
                  FALSE)
                cc_ordering <- order_list$ordering_df
                row.names(cc_ordering) <- cc_ordering$sample_name
                minSpanningTree(cds) <- as.undirected(order_list$cell_ordering_tree)
                pData(cds)$Pseudotime <- cc_ordering[row.names(pData(cds)), 
                  ]$pseudo_time
                pData(cds)$State <- cc_ordering[row.names(pData(cds)), 
                  ]$cell_state
                mst_branch_nodes <- V(minSpanningTree(cds))[which(degree(minSpanningTree(cds)) > 
                  2)]$name
                minSpanningTree(cds) <- dp_mst
                cds@auxOrderingData[[cds@dim_reduce_type]]$cell_ordering_tree <- as.undirected(order_list$cell_ordering_tree)
            }
            else if (cds@dim_reduce_type == "DDRTree") {
                if (is.null(num_paths) == FALSE) {
                  message("Warning: num_paths only valid for method 'ICA' in reduceDimension()")
                }
                cc_ordering <- monocle:::extract_ddrtree_ordering(cds, 
                  root_cell)
                pData(cds)$Pseudotime <- cc_ordering[row.names(pData(cds)), 
                  ]$pseudo_time
                K_old <- reducedDimK(cds)
                old_dp <- cellPairwiseDistances(cds)
                old_mst <- minSpanningTree(cds)
                old_A <- reducedDimA(cds)
                old_W <- reducedDimW(cds)
                cds <- project2MST(cds, monocle:::project_point_to_line_segment)
                minSpanningTree(cds) <- cds@auxOrderingData[[cds@dim_reduce_type]]$pr_graph_cell_proj_tree
                root_cell_idx <- which(igraph::V(old_mst)$name == 
                  root_cell, arr.ind = T)
                cells_mapped_to_graph_root <- which(cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_closest_vertex == 
                  root_cell_idx)
                if (length(cells_mapped_to_graph_root) == 0) {
                  cells_mapped_to_graph_root <- root_cell_idx
                }
                cells_mapped_to_graph_root <- igraph::V(minSpanningTree(cds))[cells_mapped_to_graph_root]$name
                tip_leaves <- names(which(igraph::degree(minSpanningTree(cds)) == 
                  1))
                root_cell <- cells_mapped_to_graph_root[cells_mapped_to_graph_root %in% 
                  tip_leaves][1]
                if (is.na(root_cell)) {
                  root_cell <- monocle:::select_root_cell(cds, 
                    root_state, reverse)
                }
                cds@auxOrderingData[[cds@dim_reduce_type]]$root_cell <- root_cell
                cc_ordering_new_pseudotime <- monocle:::extract_ddrtree_ordering(cds, 
                  root_cell)
                pData(cds)$Pseudotime <- cc_ordering_new_pseudotime[row.names(pData(cds)), 
                  ]$pseudo_time
                if (is.null(root_state) == TRUE) {
                  closest_vertex <- cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_closest_vertex
                  pData(cds)$State <- cc_ordering[closest_vertex[, 
                    1], ]$cell_state
                }
                monocle:::reducedDimK(cds) <- K_old
                cellPairwiseDistances(cds) <- old_dp
                minSpanningTree(cds) <- old_mst
                reducedDimA(cds) <- old_A
                monocle:::reducedDimW(cds) <- old_W
                mst_branch_nodes <- igraph::V(minSpanningTree(cds))[which(igraph::degree(minSpanningTree(cds)) > 
                  2)]$name
            }
            else if (cds@dim_reduce_type == "SimplePPT") {
                if (is.null(num_paths) == FALSE) {
                  message("Warning: num_paths only valid for method 'ICA' in reduceDimension()")
                }
                cc_ordering <- extract_ddrtree_ordering(cds, 
                  root_cell)
                pData(cds)$Pseudotime <- cc_ordering[row.names(pData(cds)), 
                  ]$pseudo_time
                pData(cds)$State <- cc_ordering[row.names(pData(cds)), 
                  ]$cell_state
                mst_branch_nodes <- V(minSpanningTree(cds))[which(degree(minSpanningTree(cds)) > 
                  2)]$name
            }
            cds@auxOrderingData[[cds@dim_reduce_type]]$branch_points <- mst_branch_nodes
            cds
        }
        project2MST <- function(cds, Projection_Method) {
            dp_mst <- minSpanningTree(cds)
            Z <- reducedDimS(cds)
            Y <- reducedDimK(cds)
            cds <- monocle:::findNearestPointOnMST(cds)
            closest_vertex <- cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_closest_vertex
            closest_vertex_names <- colnames(Y)[closest_vertex]
            closest_vertex_df <- as.matrix(closest_vertex)
            row.names(closest_vertex_df) <- row.names(closest_vertex)
            tip_leaves <- names(which(igraph::degree(dp_mst) == 
                1))
            if (!is.function(Projection_Method)) {
                P <- Y[, closest_vertex]
            }
            else {
                P <- matrix(rep(0, length(Z)), nrow = nrow(Z))
                for (i in 1:length(closest_vertex)) {
                  neighbors <- names(igraph::V(dp_mst)[suppressWarnings(nei(closest_vertex_names[i], 
                    mode = "all"))])
                  projection <- NULL
                  distance <- NULL
                  Z_i <- Z[, i]
                  for (neighbor in neighbors) {
                    if (closest_vertex_names[i] %in% tip_leaves) {
                      tmp <- monocle:::projPointOnLine(Z_i, Y[, 
                        c(closest_vertex_names[i], neighbor)])
                    }
                    else {
                      tmp <- Projection_Method(Z_i, Y[, c(closest_vertex_names[i], 
                        neighbor)])
                    }
                    projection <- rbind(projection, tmp)
                    distance <- c(distance, dist(rbind(Z_i, tmp)))
                  }
                  if (!inherits(projection, "matrix")) 
                    projection <- as.matrix(projection)
                  P[, i] <- projection[which(distance == min(distance))[1], 
                    ]
                }
            }
            colnames(P) <- colnames(Z)
            dp <- as.matrix(dist(t(P)))
            min_dist = min(dp[dp != 0])
            dp <- dp + min_dist
            diag(dp) <- 0
            cellPairwiseDistances(cds) <- dp
            gp <- igraph::graph.adjacency(dp, mode = "undirected", 
                weighted = TRUE)
            dp_mst <- igraph::minimum.spanning.tree(gp)
            cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_tree <- dp_mst
            cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_dist <- P
            cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_closest_vertex <- closest_vertex_df
            cds
        }
        BEAM <- function(cds, fullModelFormulaStr = "~sm.ns(Pseudotime, df = 3)*Branch", 
            reducedModelFormulaStr = "~sm.ns(Pseudotime, df = 3)", 
            branch_states = NULL, branch_point = 1, relative_expr = TRUE, 
            branch_labels = NULL, verbose = FALSE, cores = 1, 
            ...) {
            branchTest_res <- branchTest(cds, fullModelFormulaStr = fullModelFormulaStr, 
                reducedModelFormulaStr = reducedModelFormulaStr, 
                branch_states = branch_states, branch_point = branch_point, 
                relative_expr = relative_expr, cores = cores, 
                branch_labels = branch_labels, verbose = verbose, 
                ...)
            cmbn_df <- branchTest_res[, 1:4]
            if (verbose) 
                message("pass branchTest")
            fd <- fData(cds)[row.names(cmbn_df), ]
            cmbn_df <- cbind(cmbn_df, fd)
            if (verbose) 
                message("return results")
            return(cmbn_df)
        }
        branchTest <- function(cds, fullModelFormulaStr = "~sm.ns(Pseudotime, df = 3)*Branch", 
            reducedModelFormulaStr = "~sm.ns(Pseudotime, df = 3)", 
            branch_states = NULL, branch_point = 1, relative_expr = TRUE, 
            cores = 1, branch_labels = NULL, verbose = FALSE, 
            ...) {
            if ("Branch" %in% all.vars(terms(as.formula(fullModelFormulaStr)))) {
                cds_subset <- buildBranchCellDataSet(cds = cds, 
                  branch_states = branch_states, branch_point = branch_point, 
                  branch_labels = branch_labels, ...)
            }
            else cds_subset <- cds
            branchTest_res <- differentialGeneTest(cds_subset, 
                fullModelFormulaStr = fullModelFormulaStr, reducedModelFormulaStr = reducedModelFormulaStr, 
                cores = cores, relative_expr = relative_expr, 
                verbose = verbose)
            return(branchTest_res)
        }
        buildBranchCellDataSet <- function(cds, progenitor_method = c("sequential_split", 
            "duplicate"), branch_states = NULL, branch_point = 1, 
            branch_labels = NULL, stretch = TRUE) {
            progenitor_method <- match.arg(progenitor_method)
            if (is.null(pData(cds)$State) | is.null(pData(cds)$Pseudotime)) 
                stop("Please first order the cells in pseudotime using orderCells()")
            if (is.null(branch_point) & is.null(branch_states)) 
                stop("Please either specify the branch_point or branch_states to select subset of cells")
            if (!is.null(branch_labels) & !is.null(branch_states)) {
                if (length(branch_labels) != length(branch_states)) 
                  stop("length of branch_labels doesn't match with that of branch_states")
                branch_map <- setNames(branch_labels, as.character(branch_states))
            }
            if (cds@dim_reduce_type == "DDRTree") {
                pr_graph_cell_proj_mst <- minSpanningTree(cds)
            }
            else {
                pr_graph_cell_proj_mst <- cds@auxOrderingData[[cds@dim_reduce_type]]$cell_ordering_tree
            }
            root_cell <- cds@auxOrderingData[[cds@dim_reduce_type]]$root_cell
            root_state <- pData(cds)[root_cell, ]$State
            pr_graph_root <- subset(pData(cds), State == root_state)
            if (cds@dim_reduce_type == "DDRTree") {
                closest_vertex <- cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_closest_vertex
                root_cell_point_in_Y <- closest_vertex[row.names(pr_graph_root), 
                  ]
            }
            else {
                root_cell_point_in_Y <- row.names(pr_graph_root)
            }
            root_cell <- names(which(igraph::degree(pr_graph_cell_proj_mst, 
                v = root_cell_point_in_Y, mode = "all") == 1, 
                useNames = T))[1]
            paths_to_root <- list()
            if (is.null(branch_states) == FALSE) {
                for (leaf_state in branch_states) {
                  curr_cell <- subset(pData(cds), State == leaf_state)
                  if (cds@dim_reduce_type == "DDRTree") {
                    closest_vertex <- cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_closest_vertex
                    curr_cell_point_in_Y <- closest_vertex[row.names(curr_cell), 
                      ]
                  }
                  else {
                    curr_cell_point_in_Y <- row.names(curr_cell)
                  }
                  curr_cell <- names(which(degree(pr_graph_cell_proj_mst, 
                    v = curr_cell_point_in_Y, mode = "all") == 
                    1, useNames = T))[1]
                  path_to_ancestor <- shortest_paths(pr_graph_cell_proj_mst, 
                    curr_cell, root_cell)
                  path_to_ancestor <- names(unlist(path_to_ancestor$vpath))
                  if (cds@dim_reduce_type == "DDRTree") {
                    closest_vertex <- cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_closest_vertex
                    ancestor_cells_for_branch <- row.names(closest_vertex)[which(V(pr_graph_cell_proj_mst)[closest_vertex]$name %in% 
                      path_to_ancestor)]
                  }
                  else if (cds@dim_reduce_type == "ICA") {
                    ancestor_cells_for_branch <- path_to_ancestor
                  }
                  ancestor_cells_for_branch <- intersect(ancestor_cells_for_branch, 
                    colnames(cds))
                  paths_to_root[[as.character(leaf_state)]] <- ancestor_cells_for_branch
                }
            }
            else {
                if (cds@dim_reduce_type == "DDRTree") {
                  pr_graph_cell_proj_mst <- minSpanningTree(cds)
                }
                else pr_graph_cell_proj_mst <- cds@auxOrderingData$ICA$cell_ordering_tree
                mst_branch_nodes <- cds@auxOrderingData[[cds@dim_reduce_type]]$branch_points
                branch_cell <- mst_branch_nodes[branch_point]
                mst_no_branch_point <- pr_graph_cell_proj_mst - 
                  igraph::V(pr_graph_cell_proj_mst)[branch_cell]
                path_to_ancestor <- igraph::shortest_paths(pr_graph_cell_proj_mst, 
                  branch_cell, root_cell)
                path_to_ancestor <- names(unlist(path_to_ancestor$vpath))
                for (backbone_nei in igraph::V(pr_graph_cell_proj_mst)[suppressWarnings(nei(branch_cell))]$name) {
                  descendents <- igraph::bfs(mst_no_branch_point, 
                    igraph::V(mst_no_branch_point)[backbone_nei], 
                    unreachable = FALSE)
                  descendents <- descendents$order[!is.na(descendents$order)]
                  descendents <- igraph::V(mst_no_branch_point)[descendents]$name
                  if (root_cell %in% descendents == FALSE) {
                    path_to_root <- unique(c(path_to_ancestor, 
                      branch_cell, descendents))
                    if (cds@dim_reduce_type == "DDRTree") {
                      closest_vertex <- cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_closest_vertex
                      path_to_root <- row.names(closest_vertex)[which(igraph::V(pr_graph_cell_proj_mst)[closest_vertex]$name %in% 
                        path_to_root)]
                    }
                    else {
                      path_to_root <- path_to_root
                    }
                    closest_vertex <- cds@auxOrderingData[["DDRTree"]]$pr_graph_cell_proj_closest_vertex
                    path_to_root <- intersect(path_to_root, colnames(cds))
                    paths_to_root[[backbone_nei]] <- path_to_root
                  }
                }
            }
            all_cells_in_subset <- c()
            if (is.null(branch_labels) == FALSE) {
                if (length(branch_labels) != 2) 
                  stop("Error: branch_labels must have exactly two entries")
                names(paths_to_root) <- branch_labels
            }
            for (path_to_ancestor in paths_to_root) {
                if (length(path_to_ancestor) == 0) {
                  stop("Error: common ancestors between selected State values on path to root State")
                }
                all_cells_in_subset <- c(all_cells_in_subset, 
                  path_to_ancestor)
            }
            all_cells_in_subset <- unique(all_cells_in_subset)
            common_ancestor_cells <- intersect(paths_to_root[[1]], 
                paths_to_root[[2]])
            cds <- cds[, row.names(pData(cds[, all_cells_in_subset]))]
            Pseudotime <- pData(cds)$Pseudotime
            pData <- pData(cds)
            if (stretch) {
                max_pseudotime <- -1
                for (path_to_ancestor in paths_to_root) {
                  max_pseudotime_on_path <- max(pData[path_to_ancestor, 
                    ]$Pseudotime)
                  if (max_pseudotime < max_pseudotime_on_path) {
                    max_pseudotime <- max_pseudotime_on_path
                  }
                }
                branch_pseudotime <- max(pData[common_ancestor_cells, 
                  ]$Pseudotime)
                for (path_to_ancestor in paths_to_root) {
                  max_pseudotime_on_path <- max(pData[path_to_ancestor, 
                    ]$Pseudotime)
                  path_scaling_factor <- (max_pseudotime - branch_pseudotime)/(max_pseudotime_on_path - 
                    branch_pseudotime)
                  if (is.finite(path_scaling_factor)) {
                    branch_cells <- setdiff(path_to_ancestor, 
                      common_ancestor_cells)
                    pData[branch_cells, ]$Pseudotime <- ((pData[branch_cells, 
                      ]$Pseudotime - branch_pseudotime) * path_scaling_factor + 
                      branch_pseudotime)
                  }
                }
                pData$Pseudotime <- 100 * pData$Pseudotime/max_pseudotime
            }
            pData$original_cell_id <- row.names(pData)
            if (length(paths_to_root) != 2) 
                stop("more than 2 branch states are used!")
            pData[common_ancestor_cells, "Branch"] <- names(paths_to_root)[1]
            progenitor_pseudotime_order <- order(pData[common_ancestor_cells, 
                "Pseudotime"])
            if (progenitor_method == "duplicate") {
                ancestor_exprs <- exprs(cds)[, common_ancestor_cells]
                expr_blocks <- list()
                for (i in 1:length(paths_to_root)) {
                  if (nrow(ancestor_exprs) == 1) 
                    exprs_data <- t(as.matrix(ancestor_exprs))
                  else exprs_data <- ancestor_exprs
                  colnames(exprs_data) <- paste("duplicate", 
                    i, 1:length(common_ancestor_cells), sep = "_")
                  expr_lineage_data <- exprs(cds)[, setdiff(paths_to_root[[i]], 
                    common_ancestor_cells)]
                  exprs_data <- cbind(exprs_data, expr_lineage_data)
                  expr_blocks[[i]] <- exprs_data
                }
                ancestor_pData_block <- pData[common_ancestor_cells, 
                  ]
                pData_blocks <- list()
                weight_vec <- c()
                for (i in 1:length(paths_to_root)) {
                  weight_vec <- c(weight_vec, rep(1, length(common_ancestor_cells)))
                  weight_vec_block <- rep(1, length(common_ancestor_cells))
                  new_pData_block <- ancestor_pData_block
                  row.names(new_pData_block) <- paste("duplicate", 
                    i, 1:length(common_ancestor_cells), sep = "_")
                  pData_lineage_cells <- pData[setdiff(paths_to_root[[i]], 
                    common_ancestor_cells), ]
                  weight_vec_block <- c(weight_vec_block, rep(1, 
                    nrow(pData_lineage_cells)))
                  weight_vec <- c(weight_vec, weight_vec_block)
                  new_pData_block <- rbind(new_pData_block, pData_lineage_cells)
                  new_pData_block$Branch <- names(paths_to_root)[i]
                  pData_blocks[[i]] <- new_pData_block
                }
                pData <- do.call(rbind, pData_blocks)
                exprs_data <- do.call(cbind, expr_blocks)
            }
            else if (progenitor_method == "sequential_split") {
                pData$Branch <- names(paths_to_root)[1]
                branchA <- progenitor_pseudotime_order[seq(1, 
                  length(common_ancestor_cells), by = 2)]
                pData[common_ancestor_cells[branchA], "Branch"] <- names(paths_to_root)[1]
                branchB <- progenitor_pseudotime_order[seq(2, 
                  length(common_ancestor_cells), by = 2)]
                pData[common_ancestor_cells[branchB], "Branch"] <- names(paths_to_root)[2]
                zero_pseudotime_root_cell <- common_ancestor_cells[progenitor_pseudotime_order[1]]
                exprs_data <- cbind(exprs(cds), duplicate_root = exprs(cds)[, 
                  zero_pseudotime_root_cell])
                pData <- rbind(pData, pData[zero_pseudotime_root_cell, 
                  ])
                row.names(pData)[nrow(pData)] <- "duplicate_root"
                pData[nrow(pData), "Branch"] <- names(paths_to_root)[2]
                weight_vec <- rep(1, nrow(pData))
                for (i in 1:length(paths_to_root)) {
                  path_to_ancestor <- paths_to_root[[i]]
                  branch_cells <- setdiff(path_to_ancestor, common_ancestor_cells)
                  pData[branch_cells, ]$Branch <- names(paths_to_root)[i]
                }
            }
            pData$Branch <- as.factor(pData$Branch)
            pData$State <- factor(pData$State)
            Size_Factor <- pData$Size_Factor
            fData <- fData(cds)
            colnames(exprs_data) <- row.names(pData)
            cds_subset <- newCellDataSet(as.matrix(exprs_data), 
                phenoData = new("AnnotatedDataFrame", data = pData), 
                featureData = new("AnnotatedDataFrame", data = fData), 
                expressionFamily = cds@expressionFamily, lowerDetectionLimit = cds@lowerDetectionLimit)
            pData(cds_subset)$State <- as.factor(pData(cds_subset)$State)
            pData(cds_subset)$Size_Factor <- Size_Factor
            cds_subset@dispFitInfo <- cds@dispFitInfo
            return(cds_subset)
        }
        plot_genes_branched_heatmap <- function(cds_subset, branch_point = 1, 
            branch_states = NULL, branch_labels = c("Cell fate 1", 
                "Cell fate 2"), cluster_rows = TRUE, hclust_method = "ward.D2", 
            num_clusters = 6, hmcols = NULL, branch_colors = c("#979797", 
                "#F05662", "#7990C8"), add_annotation_row = NULL, 
            add_annotation_col = NULL, show_rownames = FALSE, 
            use_gene_short_name = TRUE, scale_max = 3, scale_min = -3, 
            norm_method = c("log", "vstExprs"), trend_formula = "~sm.ns(Pseudotime, df=3) * Branch", 
            return_heatmap = FALSE, cores = 1, ...) {
            cds <- NA
            new_cds <- buildBranchCellDataSet(cds_subset, branch_states = branch_states, 
                branch_point = branch_point, progenitor_method = "duplicate", 
                ...)
            new_cds@dispFitInfo <- cds_subset@dispFitInfo
            if (is.null(branch_states)) {
                progenitor_state <- subset(pData(cds_subset), 
                  Pseudotime == 0)[, "State"]
                branch_states <- setdiff(pData(cds_subset)$State, 
                  progenitor_state)
            }
            col_gap_ind <- 101
            newdataA <- data.frame(Pseudotime = seq(0, 100, length.out = 100), 
                Branch = as.factor(unique(as.character(pData(new_cds)$Branch))[1]))
            newdataB <- data.frame(Pseudotime = seq(0, 100, length.out = 100), 
                Branch = as.factor(unique(as.character(pData(new_cds)$Branch))[2]))
            BranchAB_exprs <- genSmoothCurves(new_cds[, ], cores = cores, 
                trend_formula = trend_formula, relative_expr = T, 
                new_data = rbind(newdataA, newdataB))
            BranchA_exprs <- BranchAB_exprs[, 1:100]
            BranchB_exprs <- BranchAB_exprs[, 101:200]
            common_ancestor_cells <- row.names(pData(new_cds)[pData(new_cds)$State == 
                setdiff(pData(new_cds)$State, branch_states), 
                ])
            BranchP_num <- (100 - floor(max(pData(new_cds)[common_ancestor_cells, 
                "Pseudotime"])))
            BranchA_num <- floor(max(pData(new_cds)[common_ancestor_cells, 
                "Pseudotime"]))
            BranchB_num <- BranchA_num
            norm_method <- match.arg(norm_method)
            if (norm_method == "vstExprs") {
                BranchA_exprs <- vstExprs(new_cds, expr_matrix = BranchA_exprs)
                BranchB_exprs <- vstExprs(new_cds, expr_matrix = BranchB_exprs)
            }
            else if (norm_method == "log") {
                BranchA_exprs <- log10(BranchA_exprs + 1)
                BranchB_exprs <- log10(BranchB_exprs + 1)
            }
            heatmap_matrix <- cbind(BranchA_exprs[, (col_gap_ind - 
                1):1], BranchB_exprs)
            heatmap_matrix = heatmap_matrix[!apply(heatmap_matrix, 
                1, sd) == 0, ]
            heatmap_matrix = Matrix::t(scale(Matrix::t(heatmap_matrix), 
                center = TRUE))
            heatmap_matrix = heatmap_matrix[is.na(row.names(heatmap_matrix)) == 
                FALSE, ]
            heatmap_matrix[is.nan(heatmap_matrix)] = 0
            heatmap_matrix[heatmap_matrix > scale_max] = scale_max
            heatmap_matrix[heatmap_matrix < scale_min] = scale_min
            heatmap_matrix_ori <- heatmap_matrix
            heatmap_matrix <- heatmap_matrix[is.finite(heatmap_matrix[, 
                1]) & is.finite(heatmap_matrix[, col_gap_ind]), 
                ]
            row_dist <- as.dist((1 - cor(Matrix::t(heatmap_matrix)))/2)
            row_dist[is.na(row_dist)] <- 1
            exp_rng <- range(heatmap_matrix)
            bks <- seq(exp_rng[1] - 0.1, exp_rng[2] + 0.1, by = 0.1)
            if (is.null(hmcols)) {
                hmcols <- monocle:::blue2green2red(length(bks) - 
                  1)
            }
            ph <- (pheatmap::pheatmap)(heatmap_matrix, useRaster = T, 
                cluster_cols = FALSE, cluster_rows = TRUE, show_rownames = F, 
                show_colnames = F, clustering_distance_rows = row_dist, 
                clustering_method = hclust_method, cutree_rows = num_clusters, 
                silent = TRUE, filename = NA, breaks = bks, color = hmcols)
            annotation_row <- data.frame(Cluster = factor(cutree(ph$tree_row, 
                num_clusters)))
            if (!is.null(add_annotation_row)) {
                annotation_row <- cbind(annotation_row, add_annotation_row[row.names(annotation_row), 
                  ])
            }
            colnames(heatmap_matrix) <- c(1:ncol(heatmap_matrix))
            annotation_col <- data.frame(row.names = c(1:ncol(heatmap_matrix)), 
                `Cell Type` = c(rep(branch_labels[1], BranchA_num), 
                  rep("Pre-branch", 2 * BranchP_num), rep(branch_labels[2], 
                    BranchB_num)))
            colnames(annotation_col) <- "Cell Type"
            if (!is.null(add_annotation_col)) {
                annotation_col <- cbind(annotation_col, add_annotation_col[fData(cds[row.names(annotation_col), 
                  ])$gene_short_name, 1])
            }
            names(branch_colors) <- c("Pre-branch", branch_labels[1], 
                branch_labels[2])
            annotation_colors = list(`Cell Type` = branch_colors)
            names(annotation_colors$`Cell Type`) = c("Pre-branch", 
                branch_labels)
            if (use_gene_short_name == TRUE) {
                if (is.null(fData(cds_subset)$gene_short_name) == 
                  FALSE) {
                  feature_label <- as.character(fData(cds_subset)[row.names(heatmap_matrix), 
                    "gene_short_name"])
                  feature_label[is.na(feature_label)] <- row.names(heatmap_matrix)
                  row_ann_labels <- as.character(fData(cds_subset)[row.names(annotation_row), 
                    "gene_short_name"])
                  row_ann_labels[is.na(row_ann_labels)] <- row.names(annotation_row)
                }
                else {
                  feature_label <- row.names(heatmap_matrix)
                  row_ann_labels <- row.names(annotation_row)
                }
            }
            else {
                feature_label <- row.names(heatmap_matrix)
                row_ann_labels <- row.names(annotation_row)
            }
            row.names(heatmap_matrix) <- feature_label
            row.names(annotation_row) <- row_ann_labels
            ph_res <- (pheatmap::pheatmap)(heatmap_matrix[, ], 
                useRaster = T, cluster_cols = FALSE, cluster_rows = TRUE, 
                show_rownames = show_rownames, show_colnames = F, 
                clustering_distance_rows = row_dist, clustering_method = hclust_method, 
                cutree_rows = num_clusters, annotation_row = annotation_row, 
                annotation_col = annotation_col, annotation_colors = annotation_colors, 
                gaps_col = col_gap_ind, treeheight_row = 20, 
                breaks = bks, fontsize = 6, color = hmcols, border_color = NA, 
                silent = TRUE)
            grid::grid.rect(gp = grid::gpar("fill", col = NA))
            grid::grid.draw(ph_res$gtable)
            if (return_heatmap) {
                return(list(BranchA_exprs = BranchA_exprs, BranchB_exprs = BranchB_exprs, 
                  heatmap_matrix = heatmap_matrix, heatmap_matrix_ori = heatmap_matrix_ori, 
                  ph = ph, col_gap_ind = col_gap_ind, row_dist = row_dist, 
                  hmcols = hmcols, annotation_colors = annotation_colors, 
                  annotation_row = annotation_row, annotation_col = annotation_col, 
                  ph_res = ph_res))
            }
        }
        genSmoothCurves <- function(cds, new_data, trend_formula = "~sm.ns(Pseudotime, df = 3)", 
            relative_expr = T, response_type = "response", cores = 1) {
            expressionFamily <- cds@expressionFamily
            if (cores > 1) {
                expression_curve_matrix <- mcesApply(cds, 1, 
                  function(x, trend_formula, expressionFamily, 
                    relative_expr, new_data, fit_model_helper, 
                    responseMatrix, calculate_NB_dispersion_hint, 
                    calculate_QP_dispersion_hint) {
                    environment(fit_model_helper) <- environment()
                    environment(responseMatrix) <- environment()
                    model_fits <- fit_model_helper(x, modelFormulaStr = trend_formula, 
                      expressionFamily = expressionFamily, relative_expr = relative_expr, 
                      disp_func = cds@dispFitInfo[["blind"]]$disp_func)
                    if (is.null(model_fits)) 
                      expression_curve <- as.data.frame(matrix(rep(NA, 
                        nrow(new_data)), nrow = 1))
                    else expression_curve <- as.data.frame(responseMatrix(list(model_fits), 
                      newdata = new_data, response_type = response_type))
                    colnames(expression_curve) <- row.names(new_data)
                    expression_curve
                  }, required_packages = c("BiocGenerics", "Biobase", 
                    "VGAM", "plyr"), cores = cores, trend_formula = trend_formula, 
                  expressionFamily = expressionFamily, relative_expr = relative_expr, 
                  new_data = new_data, fit_model_helper = fit_model_helper, 
                  responseMatrix = responseMatrix, calculate_NB_dispersion_hint = calculate_NB_dispersion_hint, 
                  calculate_QP_dispersion_hint = calculate_QP_dispersion_hint)
                expression_curve_matrix <- as.matrix(do.call(rbind, 
                  expression_curve_matrix))
                return(expression_curve_matrix)
            }
            else {
                expression_curve_matrix <- smartEsApply(cds, 
                  1, function(x, trend_formula, expressionFamily, 
                    relative_expr, new_data) {
                    environment(fit_model_helper) <- environment()
                    environment(responseMatrix) <- environment()
                    model_fits <- fit_model_helper(x, modelFormulaStr = trend_formula, 
                      expressionFamily = expressionFamily, relative_expr = relative_expr, 
                      disp_func = cds@dispFitInfo[["blind"]]$disp_func)
                    if (is.null(model_fits)) {
                      expression_curve <- as.data.frame(matrix(rep(NA, 
                        nrow(new_data)), nrow = 1))
                    }
                    else expression_curve <- as.data.frame(responseMatrix(list(model_fits), 
                      new_data, response_type = response_type))
                    colnames(expression_curve) <- row.names(new_data)
                    expression_curve
                  }, convert_to_dense = TRUE, trend_formula = trend_formula, 
                  expressionFamily = expressionFamily, relative_expr = relative_expr, 
                  new_data = new_data)
                expression_curve_matrix <- as.matrix(do.call(rbind, 
                  expression_curve_matrix))
                row.names(expression_curve_matrix) <- row.names(fData(cds))
                return(expression_curve_matrix)
            }
        }
        smartEsApply <- function(X, MARGIN, FUN, convert_to_dense, 
            ...) {
            parent <- environment(FUN)
            if (is.null(parent)) 
                parent <- emptyenv()
            e1 <- new.env(parent = parent)
            multiassign(names(pData(X)), pData(X), envir = e1)
            environment(FUN) <- e1
            if (inherits(exprs(cds), c("dgCMatrix", "dgTMatrix"))) {
                res <- monocle:::sparseApply(exprs(X), MARGIN, 
                  FUN, convert_to_dense, ...)
            }
            else {
                res <- apply(exprs(X), MARGIN, FUN, ...)
            }
            if (MARGIN == 1) {
                names(res) <- row.names(X)
            }
            else {
                names(res) <- colnames(X)
            }
            res
        }
        fit_model_helper <- function(x, modelFormulaStr, expressionFamily, 
            relative_expr, disp_func = NULL, verbose = FALSE, 
            ...) {
            modelFormulaStr <- paste("f_expression", modelFormulaStr, 
                sep = "")
            orig_x <- x
            if (expressionFamily@vfamily %in% c("negbinomial", 
                "negbinomial.size")) {
                if (relative_expr) {
                  x <- x/Size_Factor
                }
                f_expression <- round(x)
                if (is.null(disp_func) == FALSE) {
                  disp_guess <- monocle:::calculate_NB_dispersion_hint(disp_func, 
                    round(orig_x))
                  if (is.null(disp_guess) == FALSE && disp_guess > 
                    0 && is.na(disp_guess) == FALSE) {
                    size_guess <- 1/disp_guess
                    if (expressionFamily@vfamily == "negbinomial") 
                      expressionFamily <- negbinomial(isize = 1/disp_guess, 
                        ...)
                    else expressionFamily <- negbinomial.size(size = 1/disp_guess, 
                      ...)
                  }
                }
            }
            else if (expressionFamily@vfamily %in% c("uninormal", 
                "binomialff")) {
                f_expression <- x
            }
            else {
                f_expression <- log10(x)
            }
            tryCatch({
                if (verbose) {
                  FM_fit <- VGAM::vglm(as.formula(modelFormulaStr), 
                    family = expressionFamily, epsilon = 0.1)
                }
                else {
                  FM_fit <- suppressWarnings(VGAM::vglm(as.formula(modelFormulaStr), 
                    family = expressionFamily, epsilon = 0.1))
                }
                FM_fit
            }, error = function(e) {
                print(e)
                backup_expression_family <- NULL
                if (expressionFamily@vfamily %in% c("negbinomial", 
                  "negbinomial.size")) {
                  disp_guess <- calculate_NB_dispersion_hint(disp_func, 
                    round(orig_x), expr_selection_func = max)
                  backup_expression_family <- negbinomial()
                }
                else if (expressionFamily@vfamily %in% c("uninormal")) {
                  backup_expression_family <- NULL
                }
                else if (expressionFamily@vfamily %in% c("binomialff")) {
                  backup_expression_family <- NULL
                }
                else {
                  backup_expression_family <- NULL
                }
                if (is.null(backup_expression_family) == FALSE) {
                  test_res <- tryCatch({
                    if (verbose) {
                      FM_fit <- VGAM::vglm(as.formula(modelFormulaStr), 
                        family = backup_expression_family, epsilon = 0.1, 
                        checkwz = TRUE)
                    }
                    else {
                      FM_fit <- suppressWarnings(VGAM::vglm(as.formula(modelFormulaStr), 
                        family = backup_expression_family, epsilon = 0.1, 
                        checkwz = TRUE))
                    }
                    FM_fit
                  }, error = function(e) {
                    NULL
                  })
                  test_res
                }
                else {
                  NULL
                }
            })
        }
        function() {
        }
    })
    function(seurat_data, monocle = 2, genes = {
    }, wjj2, trajectory.color_by = "Pseudotime", trajectory.color_by.extra = {
    }) {
        data <- seurat_data@assays[["RNA"]]@counts
        cell_metadata <- seurat_data@meta.data
        gene_annotation <- rownames(data) %=>% hs.df(gene_short_name = x, 
            row.names = x)
        hs.switch(monocle, 2, {
            hs.require("monocle")
            cds.rd <- hs.fp(wjj2, "cds.rd")
            if (file.exists(cds.rd)) {
                cds <- hs.load(cds.rd)
            }
            else {
                cds <- (monocle::newCellDataSet)(data, phenoData = new("AnnotatedDataFrame", 
                  data = cell_metadata), featureData = new("AnnotatedDataFrame", 
                  data = gene_annotation), lowerDetectionLimit = 0.5, 
                  expressionFamily = VGAM::negbinomial.size())
                cds <- estimateSizeFactors(cds)
                cds <- estimateDispersions(cds)
                if (0) {
                  disp_table <- dispersionTable(cds) %=>% wk.dt.as
                  disp_table[mean_expression %=>% {
                    x >= quantile(x, 0.25)
                  }, ]
                  disp_table %>% {
                    .[, "mean_expression"] %>% quantile(0.25)
                  }
                  disp_table %>% {
                    .[.[, "mean_expression"] %>% {
                      . > quantile(., 0.25)
                    }, "gene_id"]
                  }
                  p1 <- monocle::plot_pc_variance_explained(cds, 
                    return_all = FALSE)
                }
                if (!length(genes)) {
                  if (0) {
                    hvf <- .try(SeuratObject::VariableFeatures(seurat_data, 
                      assay = "integrated"))
                    if (hs.is_try_error(hvf)) 
                      hvf <- SeuratObject::VariableFeatures(seurat_data, 
                        assay = "RNA")
                  }
                  genes <- SeuratObject::VariableFeatures(seurat_data, 
                    assay = "RNA")
                  if (!length(genes)) {
                    seurat_data <- hs.hsgly("转录组/标准流程/模块/预处理/seurat.normalize.38_09_23_134026")(seurat_data)
                    genes <- SeuratObject::VariableFeatures(seurat_data, 
                      assay = "RNA")
                  }
                }
                cds %<=>% monocle::setOrderingFilter(x, genes)
                cds %<=>% monocle::reduceDimension(x, reduction_method = "DDRTree", 
                  max_components = 2)
                cds %<=>% hs.getter(monocle2.hs, "orderCells")(x)
                hs.save(cds, wj = cds.rd)
            }
            if (0) {
                .pdf({
                  print(plot_ordering_genes(cds))
                })
            }
            hs.browser("39.01.19.134608", debug = 0)
            for (f1 in hs.clean(trajectory.color_by %or% trajectory.color_by.extra)) {
                wj2 <- wjj2 %=>% hs.fp(x, paste0("trajectory.", 
                  f1, ".png"))
                if (file.exists(wj2)) 
                  next
                p1 <- plot_cell_trajectory(cds, color_by = f1, 
                  cell_size = 0.1, show_backbone = TRUE)
                .figs(print(p1), wj2)
            }
            {
                wj2 <- wjj2 %=>% hs.fp(x, "pseudotime.gene_exp.test.tsv.gz")
                if (file.exists(wj2)) {
                  test <- wk.fread(wj2)
                }
                else {
                  test <- seq(nrow(cds)) %=>% hs.makeBatch(x, 
                    N = 10) %=>% wk.mclapply(x, . %=>% {
                    monocle::differentialGeneTest(cds[x, ], fullModelFormulaStr = "~sm.ns(Pseudotime)")
                  })
                  test %<=>% wk.dt.fromList
                  test[, `:=`(group, {
                  })]
                  wk.dt.setorderv(test, "pval")
                  test %<=>% wk.dt.subset(x, exp_l = c(status %in% 
                    "OK", pval < 1), exp_j = .(gene_short_name, 
                    pval, qval))
                  wk.dt.fwrite(test, wj2)
                }
                local({
                  wj2 <- wjj2 %=>% hs.fp(x, "pseudotime.gene_exp.png")
                  if (file.exists(wj2)) 
                    return()
                  N <- 9
                  g <- test[qval < 0.05, gene_short_name[pval %in% 
                    head(unique(pval), N)]]
                  if (!length(g)) 
                    return()
                  if (length(g) > N) 
                    g <- g[hs.rowMedians(seurat_data@assays[["RNA"]]@data[g, 
                      ]) %=>% hs.top(x, N, end = "left", ot = "i")]
                  p1 <- monocle::plot_genes_in_pseudotime(cds[g, 
                    ], color_by = "Pseudotime", ncol = 3)
                  .figs(print(p1), wj2, height = 10 * ceiling(length(g)/3))
                })
                local({
                  wj2 <- wjj2 %=>% hs.fp(x, "gene_clusters.heatmap.png")
                  if (file.exists(wj2)) 
                    return()
                  test[, `:=`(a, seurat_data@assays[["RNA"]]@meta.features[gene_short_name, 
                    "vst.variance.standardized"])]
                  wk.dt.setorderv(test, c("pval", "a"), c(1, 
                    -1))
                  g <- test[qval < 0.05, head(gene_short_name, 
                    50)]
                  p1 <- monocle::plot_pseudotime_heatmap(cds[g, 
                    ], num_clusters = ceiling(log(length(g))), 
                    show_rownames = length(g) <= 20, return_heatmap = 1)
                  p1 <- ggplotify::as.ggplot(p1)
                  .figs(print(p1), wj2)
                })
            }
            {
                branch_points_n <- length(cds@auxOrderingData[[cds@dim_reduce_type]][["branch_points"]])
                for (branch_point_1_i in seq(branch_points_n)) {
                  wj2 <- hs.fp(wjj2, "BEAM", paste0("branch_point_", 
                    branch_point_1_i, ".test.tsv.gz"))
                  if (file.exists(wj2)) {
                    BEAM_res <- wk.fread(wj2, cn = 1)
                  }
                  else {
                    BEAM_res <- hs.getter(monocle2.hs, "BEAM")(cds, 
                      branch_point = branch_point_1_i, cores = 10)
                    BEAM_res %<=>% {
                      wk.dt.as(x) %=>% wk.dt.setorderv(x, "pval")
                    }
                    BEAM_res[, `:=`(c("family", "use_for_ordering"), 
                      {
                      })]
                    if (BEAM_res[, "FAIL" %in% status]) 
                      BEAM_res <- BEAM_res[status %in% "OK"]
                    BEAM_res[, `:=`("status", {
                    })]
                    wk.dt.col.reorder(BEAM_res, "gene_short_name", 
                      1)
                    wk.dt.fwrite(BEAM_res, wj2)
                  }
                  g <- BEAM_res[qval < 0.05, gene_short_name]
                  g %<=>% head(x, 1000)
                  if (length(g) < 2) 
                    next
                  wj3 <- wj2 %=>% hs.wjm(x, ext = "heatmap.png", 
                    extN = 3)
                  if (!file.exists(wj3)) 
                    .figs(hs.getter(monocle2.hs, "plot_genes_branched_heatmap")(cds[g, 
                      ], branch_point = branch_point_1_i, num_clusters = ceiling(log(length(g))), 
                      cores = 1, use_gene_short_name = TRUE, 
                      show_rownames = length(g) <= 20), wj3, 
                      nc = 1)
                }
            }
        }, 3, {
            hs.browser("38.09.23.090528")
            cds <- monocle3::new_cell_data_set(data, cell_metadata = cell_metadata, 
                gene_metadata = gene_annotation)
            if (0) {
                a <- seurat_data@assays[["integrated"]]@scale.data
                irlba_res <- irlba::prcomp_irlba(Matrix::t(a), 
                  n = min(get.dimensionality(seurat_data), min(dim(a)) - 
                    1), center = TRUE, scale. = TRUE)
                preproc_res <- irlba_res[["x"]]
                row.names(preproc_res) <- colnames(a)
                irlba_rotation <- irlba_res[["rotation"]]
                row.names(irlba_rotation) <- rownames(a)
                cds@preprocess_aux$gene_loadings <- irlba_rotation %*% 
                  diag(irlba_res$sdev)
                cds@preprocess_aux$prop_var_expl <- irlba_res$sdev^2/sum(irlba_res$sdev^2)
                reducedDims(cds)[["PCA"]] <- as.matrix(preproc_res)
                cds@preprocess_aux$beta <- NULL
            }
            {
                a <- SeuratObject::Embeddings(seurat_data, reduction = "umap")
                cds@int_colData$reducedDims$UMAP <- a[colnames(cds), 
                  ]
            }
            if (0) {
                cds <- monocle3.cluster_cells(cds, reduction_method = "UMAP", 
                  random_seed = 203163417)
                p1 <- plot_cells(cds, reduction_method = "UMAP", 
                  color_cells_by = "cluster_stable.cell_type_SingleR") + 
                  ggtitle("cds.umap")
            }
            if (0) {
                p1 <- plot_cells(cds, show_trajectory_graph = FALSE) + 
                  ggtitle("label by clusterID")
                p2 <- plot_cells(cds, color_cells_by = "partition", 
                  show_trajectory_graph = FALSE) + ggtitle("label by partitionID")
                p1 + p2
            }
        })
    }
})
"38_09_23_085346" <- function(seurat_data, monocle = 2, cells = {
}, genes = {
}, wjj2, cell_type_cn = {
}, trajectory.color_by = "Pseudotime", trajectory.color_by.extra = {
}) {
    if (is.character(seurat_data) && (length(seurat_data) == 
        1) && file.exists(seurat_data)) 
        seurat_data <- hs.load(seurat_data)
    hs.browser("39.01.17.142647", debug = 0)
    hs.update(cells, list(all = colnames(seurat_data)), !length(x))
    hs.for(cell_group_type_i1, seq_along(cells), cell_group_type_1 <- names(cells)[cell_group_type_i1], 
        cell_group_i1, seq_along(cells[[cell_group_type_i1]]), 
        {
            if (debug4browse.38_09_01_174554) 
                hs.browser("38.10.11.104144")
            cell_group <- cells[[cell_group_type_i1]][[cell_group_i1]]
            cell_group_name <- names(cells[[cell_group_type_i1]][cell_group_i1])
            if (!length(cell_group_name)) 
                cell_group_name <- cell_group %=>% as.numeric %=>% 
                  {
                    if (all(!is.na(x))) 
                      wk.integerRange.4cut(x)
                    else paste(x, collapse = ";")
                  }
            wjj2_here <- hs.fp(wjj2, cell_group_type_1, cell_group_name)
            cds.rd <- hs.fp(wjj2_here, "cds.rd")
            cells.ci <- switch(cell_group_type_1, cell_cluster = SeuratObject::Idents(seurat_data), 
                cell_type = seurat_data@meta.data[, cell_type_cn]) %in% 
                cell_group
            cn <- colnames(seurat_data)[cells.ci]
            {
                sce1 <- seurat_data[, cn]
                data <- GetAssayData(sce1, assay = "RNA", slot = "counts")
                cell_metadata <- sce1@meta.data
                gene_annotation <- rownames(data) %=>% hs.df(gene_short_name = x, 
                  row.names = x)
            }
            hs.switch(monocle, 2, {
                hs.require("monocle", 1)
                if (file.exists(cds.rd)) {
                  cds <- hs.load(cds.rd)
                }
                else {
                  cds <- (monocle::newCellDataSet)(data, phenoData = new("AnnotatedDataFrame", 
                    data = cell_metadata), featureData = new("AnnotatedDataFrame", 
                    data = gene_annotation), lowerDetectionLimit = 0.5, 
                    expressionFamily = VGAM::negbinomial.size())
                  cds <- estimateSizeFactors(cds)
                  cds <- estimateDispersions(cds)
                  if (0) {
                    disp_table <- dispersionTable(cds) %=>% wk.dt.as
                    disp_table[mean_expression %=>% {
                      x >= quantile(x, 0.25)
                    }, ]
                    disp_table %>% {
                      .[, "mean_expression"] %>% quantile(0.25)
                    }
                    disp_table %>% {
                      .[.[, "mean_expression"] %>% {
                        . > quantile(., 0.25)
                      }, "gene_id"]
                    }
                    p1 <- monocle::plot_pc_variance_explained(cds, 
                      return_all = FALSE)
                  }
                  if (!length(genes)) {
                    if (0) {
                      hvf <- .try(SeuratObject::VariableFeatures(sce1, 
                        assay = "integrated"))
                      if (hs.is_try_error(hvf)) 
                        hvf <- SeuratObject::VariableFeatures(sce1, 
                          assay = "RNA")
                    }
                    sce1 <- hs.hsgly("转录组/标准流程/模块/预处理/seurat.normalize.38_09_23_134026")(sce1)
                    genes <- SeuratObject::VariableFeatures(sce1, 
                      assay = "RNA")
                  }
                  cds <- monocle::setOrderingFilter(cds, genes)
                  cds %<=>% monocle::reduceDimension(x, reduction_method = "DDRTree", 
                    max_components = 2)
                  cds %<=>% monocle::orderCells(x)
                  hs.save(cds, wj = cds.rd)
                }
                if (0) {
                  .pdf({
                    print(plot_ordering_genes(cds))
                  })
                }
                for (f1 in hs.clean(trajectory.color_by %or% 
                  trajectory.color_by.extra)) {
                  wj2 <- wjj2_here %=>% hs.fp(x, paste0("trajectory.", 
                    f1, ".png"))
                  if (file.exists(wj2)) 
                    next
                  p1 <- plot_cell_trajectory(cds, color_by = f1, 
                    size = 1, show_backbone = TRUE)
                  .figs(print(p1), wj2)
                }
                {
                  wj2 <- wjj2_here %=>% hs.fp(x, "pseudotime.gene_exp.test.tsv.gz")
                  if (file.exists(wj2)) {
                    test <- wk.fread(wj2)
                  }
                  else {
                    test <- seq(nrow(cds)) %=>% hs.makeBatch(x, 
                      N = 10) %=>% wk.mclapply(x, . %=>% {
                      monocle::differentialGeneTest(cds[x, ], 
                        fullModelFormulaStr = "~sm.ns(Pseudotime)")
                    })
                    test %<>% wk.dt.fromList
                    test[, `:=`(group, {
                    })]
                    wk.dt.setorderv(test, "pval")
                    test[, status %=>% hs.table]
                    test <- wk.dt.subset(test, exp_l = c(status %in% 
                      "OK", pval < 1), exp_j = .(gene_short_name, 
                      pval, qval))
                    wk.dt.fwrite(test, wj2)
                  }
                  local({
                    wj2 <- wjj2_here %=>% hs.fp(x, "pseudotime.gene_exp.png")
                    if (file.exists(wj2)) 
                      return()
                    N <- 9
                    g <- test[qval < 0.05, gene_short_name[pval %in% 
                      head(unique(pval), N)]]
                    if (!length(g)) 
                      return()
                    if (length(g) > N) 
                      g <- g[hs.rowMedians(sce1@assays[["RNA"]]@data[g, 
                        ]) %=>% hs.top(x, N, end = "left", ot = "i")]
                    p1 <- monocle::plot_genes_in_pseudotime(cds[g, 
                      ], color_by = "Pseudotime", ncol = 3)
                    height <- 10 * ceiling(length(g)/3)
                    .figs(print(p1), wj2, height = height)
                  })
                  local({
                    wj2 <- wjj2_here %=>% hs.fp(x, "gene_clusters.heatmap.png")
                    if (file.exists(wj2)) 
                      return()
                    g <- test[qval < 0.05, gene_short_name]
                    p1 <- monocle::plot_pseudotime_heatmap(cds[g, 
                      ], num_clusters = ceiling(log(length(g))), 
                      show_rownames = length(g) <= 20, return_heatmap = 1)
                    p1 <- ggplotify::as.ggplot(p1)
                    .figs(print(p1), wj2)
                  })
                }
                {
                  branch_points_n <- length(cds@auxOrderingData[[cds@dim_reduce_type]][["branch_points"]])
                  for (branch_point_1_i in seq(branch_points_n)) {
                    wj2 <- hs.fp(wjj2_here, "BEAM", paste0("branch_point_", 
                      branch_point_1_i, ".test.tsv.gz"))
                    if (file.exists(wj2)) {
                      BEAM_res <- wk.fread(wj2, cn = 1)
                    }
                    else {
                      BEAM_res <- monocle::BEAM(cds, branch_point = branch_point_1_i, 
                        cores = 10)
                      BEAM_res %<>% wk.dt.as %=>% wk.dt.setorderv(x, 
                        "pval")
                      BEAM_res[, `:=`(c("family", "use_for_ordering"), 
                        {
                        })]
                      if (BEAM_res[, "FAIL" %in% status]) 
                        BEAM_res <- BEAM_res[status %in% "OK"]
                      BEAM_res[, `:=`("status", {
                      })]
                      wk.dt.col.reorder(BEAM_res, "gene_short_name", 
                        1)
                      wk.dt.fwrite(BEAM_res, wj2)
                    }
                    g <- BEAM_res[qval < 0.05, gene_short_name]
                    g %<>% head(1000)
                    if (length(g) < 2) 
                      next
                    wj3 <- wj2 %=>% hs.wjm(x, ext = "heatmap.png", 
                      extN = 3)
                    log(seq(1000, 50000, length.out = 10), seq(4, 
                      2, length.out = 10))
                    x <- seq(1000, 10000, by = 1000)
                    y <- seq(4, 2, length.out = length(x))
                    m1 <- lm(y ~ x)
                    predict(m1, list(x = c(20000, 1000, 2000)))
                    predict(m1, hs.df(x = c(1000, 2000)))
                    log(50000, 1.5)
                    if (!file.exists(wj3)) 
                      .figs((monocle::plot_genes_branched_heatmap)(cds[g, 
                        ], branch_point = branch_point_1_i, num_clusters = ceiling(log(length(g))), 
                        cores = 1, use_gene_short_name = TRUE, 
                        show_rownames = length(g) <= 20), wj3, 
                        nc = 1)
                  }
                }
            }, 3, {
                hs.browser("38.09.23.090528")
                cds <- monocle3::new_cell_data_set(data, cell_metadata = cell_metadata, 
                  gene_metadata = gene_annotation)
                if (0) {
                  a <- seurat_data@assays[["integrated"]]@scale.data
                  irlba_res <- irlba::prcomp_irlba(Matrix::t(a), 
                    n = min(get.dimensionality(seurat_data), 
                      min(dim(a)) - 1), center = TRUE, scale. = TRUE)
                  preproc_res <- irlba_res[["x"]]
                  row.names(preproc_res) <- colnames(a)
                  irlba_rotation <- irlba_res[["rotation"]]
                  row.names(irlba_rotation) <- rownames(a)
                  cds@preprocess_aux$gene_loadings <- irlba_rotation %*% 
                    diag(irlba_res$sdev)
                  cds@preprocess_aux$prop_var_expl <- irlba_res$sdev^2/sum(irlba_res$sdev^2)
                  reducedDims(cds)[["PCA"]] <- as.matrix(preproc_res)
                  cds@preprocess_aux$beta <- NULL
                }
                {
                  a <- SeuratObject::Embeddings(seurat_data, 
                    reduction = "umap")
                  cds@int_colData$reducedDims$UMAP <- a[colnames(cds), 
                    ]
                }
                if (0) {
                  cds <- monocle3.cluster_cells(cds, reduction_method = "UMAP", 
                    random_seed = 203163417)
                  p1 <- plot_cells(cds, reduction_method = "UMAP", 
                    color_cells_by = "cluster_stable.cell_type_SingleR") + 
                    ggtitle("cds.umap")
                }
                if (0) {
                  p1 <- plot_cells(cds, show_trajectory_graph = FALSE) + 
                    ggtitle("label by clusterID")
                  p2 <- plot_cells(cds, color_cells_by = "partition", 
                    show_trajectory_graph = FALSE) + ggtitle("label by partitionID")
                  p1 + p2
                }
            })
        })
}
"38_10_08_171232" <- function(seurat_data, org1, wjj2, nc = hs.hsgly("cpuN.38_10_08_172545")(max = 16)) {
    if (org1 %not.in% c("human", "mouse")) {
        message("Only human & mouse are supported now.")
        hs.browser("38.04.25.221451")
    }
    if (0) {
        Sys.setenv(NOT_CRAN = TRUE)
        install.packages("arrow")
    }
    hs.require(c("AUCell", "RcisTarget", "GENIE3"), 1)
    hs.require(c("zoo", "mixtools", "rbokeh"), 1)
    hs.require(c("DT", "NMF", "ComplexHeatmap", "R2HTML", "Rtsne"), 
        1)
    hs.require(c("doMC", "doRNG"), 1)
    hs.require("aertslab/SCopeLoomR", 1)
    hs.require(c("dynamicTreeCut", "aertslab/SCENIC"), 1)
    if (0) {
        wk.rsync.out("/home/wk/tmp/hs_hgnc_tfs.txt", "/home/wk/swxx/sj/db/cistarget/github.com/aertslab/pySCENIC/blob/master/resources", 
            ssh = "r820")
    }
    wj_in <- "tf,motif.tf,ranking_genome" %>% hs.lapply(. %>% 
        {
            hs.hsgly("cisTarget/数据.38_03_30_074330")(what = ., 
                org1 = org1)
        }, simplify = 1)
    wj_loom <- local({
        wj2 <- hs.fp(wjj2, "the.loom")
        if (!file.exists(wj2)) {
            hs.require(c("hhoeflin/hdf5r", "mojaveazure/loomR"), 
                1)
            seurat_data@assays[["RNA"]]@counts %=>% loomR::create(hs.wj(wj2), 
                data = x, do.transpose = TRUE, gene.attrs = list(Gene = rownames(x)), 
                cell.attrs = list(CellID = colnames(x), nGene = hs.colSums(x > 
                  0), nUMI = hs.colSums(x)))
        }
        wj2
    })
    {
        pyscenic <- hs.home.expand("~/miniconda3/envs/pyscenic/bin/pyscenic")
        {
            wj_adj <- hs.fp(wjj2, "adj.tsv")
            {
                loom1 <- loomR::connect(wj_loom)
                loom1$close_all()
            }
            if (0) {
                wj_loom <- "~/swxx/scenic_test/PBMC10k_filtered.loom"
                wj_adj <- hs.wjm(wj_loom, ext = "scenic.adj.tsv")
                wj_loom %>% .ls
                loom1 <- loomR::connect("~/swxx/scenic_test/PBMC10k_filtered.loom")
                loom1[["matrix"]]
                loom1[["col_attrs/cell_names"]]
                loom1[["gene.attrs/gene_names"]]
                loom1$get.attribute.df(1)
                loom1$col_attrs
                loom1$row_attrs
                loom1$matrix
                loom1[["col_attrs/CellID"]]
                loom1[["row_attrs/Gene"]]
                loom1[["col_attrs"]]
                loom1[["row_attrs"]]
                tf <- readLines(wj_in["tf"])
                tf %and% loom1[["row_attrs/Gene"]][]
                wj_adj %>% hs.wj.rm
                wj_adj %>% .ls
                wj_loom
            }
            if (wj.update.reporter(wj_adj, wj_loom)) {
                hs.msg("开始运行pyscenic", 0)
                cmd1 <- .sys(.sys.fpGood(pyscenic), "grn", "--num_workers", 
                  nc, "--output", .sys.fpGood(hs.wj(wj_adj)), 
                  "--method grnboost2", .sys.fpGood(wj_loom), 
                  .sys.fpGood(wj_in["tf"]))
                if (0) {
                  hs.save(cmd1, wj = "~/a.rd")
                  "~/a.rd" %>% wk.rsync.out.same_place(ssh = "hw1.n9")
                  load("~/a.rd")
                  .wk(cmd1)
                  wk.rsync.in.same_place(wj_adj, ssh = "hw1.n9")
                }
                .wk(cmd1)
                hs.msg("pyscenic运行结束", 0)
            }
            hs.fp.make_read_only(wj_adj)
        }
        {
            wj_reg <- hs.fp(hs.dirname(wj_adj), "reg.csv")
            if (file.exists(wj_reg) && (file.size(wj_reg) == 
                0)) 
                hs.wj.rm(wj_reg)
            if (wj.update.reporter(wj_reg, wj_adj)) {
                cmd1 <- .sys(.sys.fpGood(pyscenic), "ctx", .sys.fpGood(wj_adj), 
                  .sys.fpGood(wj_in["ranking_genome"]), "--annotations_fname", 
                  .sys.fpGood(wj_in["motif.tf"]), "--expression_mtx_fname", 
                  .sys.fpGood(wj_loom), "--mode \"dask_multiprocessing\"", 
                  "--min_genes 20", "-o", .sys.fpGood(wj_reg), 
                  "--num_workers", nc, "--mask_dropouts")
                message(cmd1)
                .wk(cmd1)
            }
            hs.fp.make_read_only(wj_reg)
        }
        {
            wj_auc <- hs.fp(dirname(wj_reg), "auc.csv")
            if (wj.update.reporter(wj_auc, wj_reg)) 
                .wk(pyscenic, "aucell", .sys.fpGood(wj_loom), 
                  .sys.fpGood(wj_reg), "-o", .sys.fpGood(wj_auc), 
                  "--auc_threshold 0.05", "--seed 135318", "--num_workers", 
                  nc)
            local({
                wj2 <- hs.wjm(wj_auc, ext = "tsv.gz")
                if (file.exists(wj2)) 
                  return()
                a <- wk.fread(wj_auc, ot = "dt")
                wk.dt.fwrite(a, wj2)
            })
            wj_loom_final <- hs.wjm(wj_loom, append = ".2")
            if (wj.update.reporter(wj_loom_final, wj_reg)) {
                .wk(pyscenic, "aucell", .sys.fpGood(wj_loom), 
                  .sys.fpGood(wj_reg), "-o", .sys.fpGood(wj_loom_final), 
                  "--auc_threshold 0.05", "--seed 135318", "--num_workers", 
                  nc)
            }
            {
            }
        }
    }
    {
        wj2 <- c("umap", "tsne") %>% {
            hs.c(., hs.fp(dirname(wj_loom), paste0(., ".tsv.gz")))
        }
        if (wj.update.reporter(wj2, wj_auc)) {
            pyscenic.bin <- "/dev/shm/wk_conda/envs/pyscenic/bin/pyscenic"
            cmd1 <- .sys(pyscenic.bin %>% dirname %>% hs.fp("python3"), 
                "-c", .q(paste0("\nimport os\nimport pandas as pd\nimport umap.umap_ as umap\nfrom MulticoreTSNE import MulticoreTSNE as TSNE\nauc_mtx = pd.read_csv( \"", 
                  wj_auc, "\", index_col = 0)\nrunUmap = umap.UMAP( n_neighbors = 10, min_dist = 0.4, metric = \"correlation\").fit_transform\ndr_umap = runUmap( auc_mtx)\ntsne = TSNE( n_jobs = ", 
                  nc, ")\ndr_tsne = tsne.fit_transform( auc_mtx)\nimport gzip", 
                  lapply(names(wj2), function(task1) {
                    if (wj.update.reporter(wj2[task1], wj_auc)) 
                      paste0("\nwj2 = gzip.open( \"", wj2[task1], 
                        "\", \"wb\")\na = pd.DataFrame( ", paste0("dr_", 
                          task1), ", index = auc_mtx.index.to_numpy())\na.to_csv( wj2, sep = \"\t\", header = False)\nwj2.close()\n")
                  }) %>% unlist %>% paste(collapse = ""))))
            message(cmd1)
            .wk(cmd1)
        }
    }
    {
        wj2 <- hs.fp(wjj2, "auc_dif_top.heatmap")
        if (!file.exists(wj2)) 
            local({
                auc <- local({
                  a <- wk.fread(wj_auc)
                  a <- wk.dt.2matrix(a, rn.cn = 1)
                  t(a)
                })
                topN <- 20
                cg <- Idents(seurat_data)[colnames(auc)]
                if (0) {
                  x1 <- auc[1, ]
                  t1 <- kruskal.test(x1, cg)
                  kruskal.test(auc[2, ], cg)
                  names(t1)
                  t1[["statistic"]]
                }
                statistic <- apply(auc, 1, function(x1) kruskal.test(x1, 
                  cg)[["statistic"]])
                if (0) {
                  pv <- apply(auc, 1, function(x1) kruskal.test(x1, 
                    cg)[["p.value"]])
                  qv <- p.adjust(pv, "fdr")
                  test <- wk.dt(tf = rownames(auc), statistic, 
                    pv, qv)
                  wk.dt.setorderv(test, "statistic", -1)
                  rn2show <- test[head(which(qv < 0.05), topN), 
                    tf]
                }
                rn2show <- statistic %>% {
                  names(.)[hs.order(.) %>% head(topN)]
                }
                m1 <- auc[rn2show, ]
                m1 <- apply(m1, 1, . %>% {
                  (. - mean(.))/sd(.)
                }) %>% t
                cn_ordered <- colnames(m1) %>% {
                  tapply(., cg[.], c)
                } %>% lapply(. %>% {
                  a <- hclust(as.dist(1 - cor(m1[, .], method = "spearman")))
                  .[a[["order"]]]
                }) %>% hs.unlist
                ann_col <- hs.df(`Cell cluster` = cg)
                p1 <- pheatmap(m1[, cn_ordered], silent = TRUE, 
                  main = paste0("[SCENIC] scaled regulon activity scores\nof top ", 
                    topN, " regulons"), legend = TRUE, annotation_legend = FALSE, 
                  legend_labels = "xxx", annotation_col = ann_col, 
                  cluster_cols = FALSE, color = colorRampPalette(c("steelblue", 
                    "white", "tomato"))(100), breaks = seq(-2, 
                    2, length.out = 100), show_colnames = FALSE, 
                  show_rownames = TRUE, border_color = NA)
                p1 <- ggplotify::as.ggplot(p1)
                .figs(print(p1), wj2)
            })
        if (0) {
            `?`(SCENIC::calcRSS)
            a <- wk.fread(wj_reg)
            a[1:4]
            a[4, hs.sub(V9, "[\\[\\]]")]
            module_size <- hs.gre(a[-c(1:3), V9], "(?<=\\()[^\\(\\)]+") %>% 
                sapply(. %>% hs.re("\\w+") %>% uniqueN)
            tapply(module_size, a[-c(1:3), V1], max) %>% summary
        }
    }
    if (0) {
        loomPath <- file.path(system.file("extdata", package = "SCopeLoomR"), 
            "example.loom")
        loom <- open_loom(loomPath)
        loomPath %>% .ls
        loom <- loomR::connect(loomPath, "r+")
        regulons <- get_regulons(loom)
        head(regulons)
        loom1 <- SCopeLoomR::open_loom(wj_loom_final, mode = "r+")
        regulons_incidMat <- get_regulons(loom1, column.attr.name = "Regulons")
        regulons <- regulonsToGeneLists(regulons_incidMat)
        regulonAUC <- get_regulons_AUC(loom, column.attr.name = "RegulonsAUC")
        sce = readRDS("./fibo_1000.rds")
        sce
        library(pheatmap)
        n = t(scale(t(getAUC(regulonAUC[, ]))))
        n[n > 2] = 2
        n[n < -2] = -2
        n[1:4, 1:4]
        dim(n)
        ac = data.frame(group = as.character(Idents(sce)))
        n[1:4, 1:4]
        n = n[, colnames(n) %in% colnames(sce)]
        rownames(ac) = colnames(n)
        cg = read.table("choose_tf.txt")[, 1]
        cg
        cg_n = n[rownames(n) %in% cg, ]
        pheatmap(cg_n, show_colnames = F, show_rownames = T, 
            annotation_col = ac)
        table(ac$group)
        ac$group = ifelse(ac$group %in% c(2:5, 7, 9), "mCAF", 
            "iCAF")
        pheatmap(cg_n, show_colnames = F, show_rownames = T, 
            annotation_col = ac)
        pheatmap(cg_n, show_colnames = F, show_rownames = T, 
            annotation_col = ac, filename = "heatmap_choose_regulon.png")
        dev.off()
        reg <- wk.fread(wj_reg)
        reg[2:4] %>% .sh
    }
}
"38_10_17_091000" <- function(seurat_data, prj_fp, clustering.fp = hs.fp(prj_fp, 
    "mid/clustering.gz"), clustering_chosen_cn = formals(hs.hsgly("1.流程/单细胞/转录组/标准流程/主函数.37_12_09_143244"))[["clustering_chosen_cn"]], 
    cell_type.fp = hs.fp(prj_fp, "mid/cell_type/barcode.cell_type.tsv.gz"), 
    cell_type_cn = formals(hs.hsgly("1.流程/单细胞/转录组/标准流程/主函数.37_12_09_143244"))[["cell_type_cn"]]) {
    a <- wk.fread(clustering.fp)
    if (a[, .N] != ncol(seurat_data)) 
        hs.browser("38.10.14.170826")
    seurat_data@meta.data[, clustering_chosen_cn] <- factor(a[[1]])
    a <- wk.fread(cell_type.fp)
    if (!all(a[, barcode] == colnames(seurat_data))) 
        hs.browser("38.10.14.170650")
    seurat_data@meta.data[, cell_type_cn] <- a[[2]]
    seurat_data
}
"38_10_15_214316" <- function(seurat_data, seurat_data.fp = hs.fp(prj_fp, 
    "mid/sce_ready.rd"), seurat_data.subset.fp = hs.fp(prj_fp, 
    wjj2.sub, "mid/sce_ready.rd"), wjj2.sub = "subset", cells, 
    prj_fp = if (length(prj_id)) hs.fp(hs.hsgly("中科新生命/业务/单细胞/项目/根目录.38_08_29_131836")(), 
        prj_id), prj_id = {
    }, clustering.fp = hs.fp(prj_fp, "mid/clustering.gz"), cell_type.fp = hs.fp(prj_fp, 
        "mid/cell_type/barcode.cell_type.tsv.gz"), clustering_chosen_cn = formals(hs.hsgly("1.流程/单细胞/转录组/标准流程/主函数.37_12_09_143244"))[["clustering_chosen_cn"]], 
    sample.cn = formals(hs.hsgly("1.流程/单细胞/转录组/标准流程/主函数.37_12_09_143244"))[["sample.cn"]], 
    cell_type_cn = formals(hs.hsgly("1.流程/单细胞/转录组/标准流程/主函数.37_12_09_143244"))[["cell_type_cn"]], 
    ...) {
    if (length(seurat_data.fp) && file.exists(seurat_data.fp)) {
        seurat_data <- hs.load(seurat_data.fp)
    }
    seurat_data <- hs.hsgly("单细胞/转录组/标准流程/模块/加载元信息.38_10_17_091000")(seurat_data = seurat_data, 
        prj_fp = prj_fp)
    hs.for(i1, seq_along(cells), i2, seq_along(cells[[i1]]), 
        {
            cn1 <- hs.switch(names(cells)[i1], "cell_cluster", 
                clustering_chosen_cn, "cell_type", cell_type_cn)
            i <- which(seurat_data@meta.data[, cn1] %in% cells[[i1]][[i2]])
            if (!length(i)) 
                stop("[2022_09_27_135518] 没有细胞属于列", 
                  .q(cn1), "的", paste(.q(cells[[i1]][[i2]]), 
                    collapse = "、"))
            seurat_data_subset <- seurat_data[, i]
            seurat_data_subset_list <- Seurat::SplitObject(seurat_data_subset, 
                sample.cn)
            wjj21 <- names(cells[[i1]])[i2]
            if (!length(wjj21)) 
                wjj21 <- paste(cells[[i1]][[i2]], collapse = ";")
            hs.hsgly("1.流程/单细胞/转录组/标准流程/主函数.37_12_09_143244")(wjj2.sub = wjj2.sub, 
                sce_list = seurat_data_subset_list, sce_combined.wj = seurat_data.subset.fp, 
                wjj2load = prj_fp, prj_fp = prj_fp, subset = 1, 
                ...)
        })
}
"38_09_09_172309" <- function(seurat_data, wj2 = {
}, cell_type_column_name = "cell_type") {
    reduction <- c("umap", "tsne") %and% names(seurat_data@reductions)
    cn1 <- "ct.38_09_09_164156"
    seurat_data@meta.data[, cn1] <- seurat_data@meta.data[, cell_type_column_name] %=>% 
        {
            a <- unique(x) %=>% sort
            nn <- nchar(length(a))
            b <- seq_along(a) %=>% sapply(x, . %=>% hs.str_pad(paste0("(", 
                x), nn + 1)) %=>% paste0(") ", a)
            v <- hs.c(a, b)[x]
            v[is.na(v)] <- "Unknown"
            hs.factor(v, b %or% v, N = 0)
        }
    P <- hs.lapply(reduction, function(reduction_method_1) {
        p1 <- Seurat::DimPlot(seurat_data, reduction = reduction_method_1, 
            group.by = cn1, label = FALSE, cols = hs.hsgly("color.db.38_09_09_173646")()) + 
            ggplot2::labs(title = switch(reduction_method_1, 
                umap = "UMAP", tsne = "tSNE", pca = "PCA"))
        p1$data[, "cn2"] <- seurat_data@meta.data[, cn1] %=>% 
            {
                i <- hs.grep(levels(x), "^ *\\(\\d+\\)")
                if (length(i)) {
                  levels(x)[i] %<=>% hs.re(x, "\\(\\d+\\)")
                }
                x
            }
        Seurat::LabelClusters(p1, "cn2")
    })
    P <- evalStr(paste0("P[[", seq_along(P), "]]") %=>% paste(collapse = "+"))
    if (length(wj2)) {
        if (!file.exists(wj2)) {
            text_n <- seurat_data@meta.data[, cell_type_column_name] %=>% 
                hs.clean(U = 1) %=>% nchar %=>% max
            width <- 10 + 15 * length(reduction) + text_n * 0.2
            .figs(print(P), wj2, width = width, height = 15)
        }
    }
    else {
        P
    }
}
"38_10_19_114235" <- local({
    res2cn <- function(res, tag, rev = 0) if (rev) {
        hs.sub(res, tag) %=>% hs.sub(x, "_", "-") %=>% as.numeric
    }
    else hs.sub(res, "-", "_") %=>% paste0(tag, x)
    res2do <- function(res, seurat_data = {
    }, clusterings_wj = {
    }, cn_tag = {
    }) {
        if (length(seurat_data)) {
            res %<=>% x[res2cn(x, cn_tag) %not.in% colnames(seurat_data@meta.data)]
        }
        res_done <- clusterings_wj %=>% if (length(x) && file.exists(x)) {
            readLines(x, 1) %=>% hs.strsplit.v(x, sep = "\t")
        }
        res %=>% x[res2cn(x, tag) %not.in% res_done]
    }
    get.clusterings <- local({
        hs1 <- function(sce, res = 0.1/(2^(0:20)), red = "umap", 
            dim_n = Inf, clusterings_done = {
            }, clusterings_cn_tag = "") {
            get.a.cluster <- function(res1, red = "umap") {
                message("分辨率：", res1)
                cluster_result <- monocle3:::leiden_clustering(data = reduced_dim_res, 
                  pd = sce@meta.data, k = 20, weight = FALSE, 
                  num_iter = 2, resolution_parameter = res1, 
                  random_seed = 160109, verbose = FALSE)
                factor(igraph::membership(cluster_result$optim_res))
            }
            res_done <- colnames(clusterings_done)
            reduced_dim_res <- SeuratObject::Embeddings(sce, 
                reduction = red)
            ci_max <- min(ncol(reduced_dim_res), dim_n)
            if (dim_n < ncol(reduced_dim_res)) 
                reduced_dim_res %<=>% x[, 1:dim_n, drop = 0]
            hs.switch(res, "auto", {
                res_low <- 0.1/(2^30) * as.numeric(paste0("1e-", 
                  seq(0, 9) * 3))
                res_low %<=>% x[res2cn(x, tag = clusterings_cn_tag) %not.in% 
                  colnames(clusterings_done)]
                a <- wk.mclapply(res_low, get.a.cluster, red = red)
                a <- sapply(a, function(x) levels(x) %=>% length)
                a <- a[a > 1]
                res <- if (length(a)) {
                  0.1/(2^seq(0, ceiling(log2(0.1/res_low[which.min(a)]))))
                }
                else 0.1/(2^seq(0, ceiling(log2(0.1/max(res_low)))))
                res %<=>% x[res2cn(x, tag = clusterings_cn_tag) %not.in% 
                  colnames(clusterings_done)]
            })
            if (length(res)) {
                a <- wk.mclapply(res, get.a.cluster, mc.cores = 10)
                m1 <- sapply(a, c) %=>% t %=>% unique %=>% t
                if (nzchar(clusterings_cn_tag)) 
                  colnames(m1) %<=>% paste0(clusterings_cn_tag, 
                    x)
                if (length(clusterings_done)) {
                  cn2do <- names(clusterings_done)[-1] %not% 
                    colnames(m1)
                  if (length(cn2do)) {
                    m1 %<=>% cbind(x, clusterings_done[match(rownames(m1), 
                      barcode), as.matrix(.SD), .SDcols = c(cn2do)])
                    i <- colnames(m1) %=>% res2cn(res = x, tag = clusterings_cn_tag, 
                      rev = 1) %=>% hs.order
                    m1 <- hs.matrix_ij(m1, j = i) %=>% t %=>% 
                      unique %=>% t
                  }
                }
                m1
            }
        }
        function(sce, res = "auto", red = "umap", dim_n = Inf, 
            clusterings_done = {
            }, clusterings_cn_tag = "") {
            clusterings <- hs1(sce, res, red = red, dim_n = dim_n, 
                clusterings_done = clusterings_done, clusterings_cn_tag = clusterings_cn_tag)
            cluster_n <- apply(clusterings, 2, uniqueN)
            if ((res == "auto") && (length(cluster_n) > 1)) {
                hs.for(right1, seq(ceiling(min(cluster_n)/10) * 
                  10, (log10(ncol(sce)) - 1) * 10, by = 10), 
                  {
                    left1 <- right1 - 9
                    clustering_n_co <- ifelse(left1 == 1, 3, 
                      5)
                    repeated_n <- 0
                    repeat {
                      cluster_n <- apply(clusterings, 2, uniqueN)
                      clustering_n <- cluster_n %=>% unique %=>% 
                        sum((x <= right1) & (x > left1))
                      if (clustering_n < clustering_n_co) {
                        if (left1 == 1) {
                          co1 <- res2cn(names(cluster_n)[which.min(cluster_n)], 
                            tag = clusterings_cn_tag, rev = 1)/100
                          co2 <- cluster_n[cluster_n >= right1] %=>% 
                            res2cn(names(x)[which.min(x)], tag = clusterings_cn_tag, 
                              rev = 1)
                          res2do <- .try(log10(c(co1, co2)) %=>% 
                            seq(x[1], x[2], length.out = 22)[2:21] %=>% 
                            {
                              10^x
                            })
                        }
                        else {
                          co1 <- max(res2cn(res = names(cluster_n), 
                            tag = clusterings_cn_tag, rev = 1)[which(cluster_n <= 
                            left1)])
                          if (is.infinite(co1)) 
                            co1 <- as.numeric(names(cluster_n)[which.min(cluster_n)])
                          co2 <- min(res2cn(res = names(cluster_n), 
                            tag = clusterings_cn_tag, rev = 1)[which(cluster_n > 
                            right1)])
                          res2do <- .try(log10(c(co1, co2)) %=>% 
                            seq(x[1], x[2], length.out = 22)[2:21] %=>% 
                            {
                              10^x
                            })
                        }
                        if (inherits(res2do, "try-error")) 
                          hs.browser("38.08.16.050702")
                        a <- hs1(sce = sce, res = res2do, red = red, 
                          clusterings_done = wk.dt.as(clusterings, 
                            "barcode"), clusterings_cn_tag = clusterings_cn_tag)
                        clusterings %<=>% {
                          cbind(x, a) %=>% x[, hs.order(res2cn(res = colnames(x), 
                            tag = clusterings_cn_tag, rev = 1))] %=>% 
                            t %=>% unique %=>% t
                        }
                      }
                      else break
                      repeated_n <- repeated_n + 1
                      if (repeated_n > 3) 
                        break
                    }
                  })
                cluster_n <- clusterings %=>% apply(x, 2, uniqueN) %=>% 
                  rev
                i <- cluster_n %=>% tapply(names(x), x, . %=>% 
                  x[which.max(res2cn(res = x, tag = clusterings_cn_tag, 
                    rev = 1))]) %=>% rev
                hs.matrix_ij(clusterings, j = i)
            }
            else clusterings
        }
    })
    function(seurat_data, res = "auto", red = "umap", dim_n = switch(red, 
        umap = 10, Inf), wjj2 = if (length(pca_wjj)) 
        hs.fp(pca_wjj, paste0("pc_n=", ncol(seurat_data@reductions[["pca"]]))), 
        pca_wjj = {
        }, wjj2load = {
        }, redo = 0) {
        if (red != "umap") 
            hs.browser("38.12.21.162942", debug = 0)
        {
            if (length(wjj2load)) 
                hs.browser("38.12.21.170235", debug = 0)
        }
        wj_umap <- wjj2 %=>% hs.wjj.ls("reduction_umap_c=\\d+.rd") %=>% 
            hs.c(basename(x) %=>% hs.re("\\d+"), x) %=>% x[order(as.numeric(names(x)))][2]
        wj2 <- hs.wjm(wj_umap, ext = "clusterings.tsv.gz")
        if (redo) 
            hs.wj.rm(wj2)
        {
            hs.msg("对细胞分群")
            alg <- "leiden"
            pkg <- "monocle3"
            tag <- paste0(red, "_", alg, "_", pkg, "_res.")
            hs.switch(res, "auto", {
            }, {
                res %<=>% res2do(x, seurat_data = seurat_data, 
                  cn_tag = tag)
                if (!length(res)) 
                  return(seurat_data)
                res %<=>% res2do(x, clusterings_wj = wj2, cn_tag = tag)
            })
            if (length(res)) {
                umap2 <- seurat_data@reductions[["umap"]]
                seurat_data@reductions[["umap"]] <- hs.load(wj_umap)
                clusterings <- get.clusterings(sce = seurat_data, 
                  red = red, res = res, dim_n = dim_n, clusterings_cn_tag = tag, 
                  clusterings_done = if (file.exists(wj2)) 
                    wk.fread(wj2))
                hs.msg("完成", 0)
                if (0) {
                  a <- wk.fread(wj2, ot = "m", rn.cn = 1, cn = 1, 
                    colClasses = "character")
                  all(colnames(a) == colnames(clusterings))
                  all(apply(a, 2, uniqueN) == apply(clusterings, 
                    2, uniqueN))
                  clusterings_1 <- hs.hsgly("转录组/标准流程/模块/r函数_自制.38_04_07_003149", 
                    "get.clusterings")(sce = seurat_data)
                  clusterings_2 <- hs.hsgly("转录组/标准流程/模块/r函数_自制.38_04_07_003149", 
                    "get.clusterings")(sce = seurat_data)
                  all(colnames(clusterings_1) == colnames(clusterings_2))
                  all(clusterings_1[, 1] == clusterings_2[, 1])
                  ncol(clusterings_1) %>% seq %>% sapply(. %>% 
                    {
                      all(clusterings_1[, .] == clusterings_2[, 
                        .])
                    }) %>% all
                  colnames(clusterings) == colnames(clusterings_1)
                }
                if (file.exists(wj2)) {
                  if (0) {
                    a <- wk.fread(wj2)
                    wk.dt.set(a, a[, match(colnames(seurat_data), 
                      barcode)], colnames(clusterings), wk.dt.as(clusterings))
                    wk.dt.fwrite(a, wj2)
                  }
                }
                clusterings %=>% wk.dt.as(x, keep.rownames = "barcode") %=>% 
                  wk.dt.fwrite(x, wj2)
                seurat_data@reductions[["umap"]] <- umap2
            }
        }
        if (readLines(wj2, 1) %=>% hs.strsplit.v(sep = "\t")[-1] %=>% 
            any(x %not.in% colnames(seurat_data@meta.data))) {
            clusterings <- wk.fread(wj2, ot = "m", rn.cn = 1, 
                cn = 1, colClasses = "character")
            seurat_data@meta.data[, colnames(clusterings)] <- clusterings
        }
        seurat_data
    }
})
"38_10_18_164508" <- function(seurat_data, org1, tissue_type, 
    cell_type_marker_list, cell_type_marker_list.org, cell_type_cn, 
    clustering_chosen_cn, clustering_cn2use, wjj2) {
    hs.browser("38.12.07.121032", debug = 0)
    clustering_chosen.fp <- hs.fp(wjj2, "clustering.gz")
    wj2 <- hs.fp(wjj2, "cell_type/barcode.cell_type.tsv.gz")
    if (all(file.exists(c(wj2, clustering_chosen.fp)))) {
        a <- wk.fread(wj2)
        seurat_data@meta.data[, cell_type_cn] <- a[[cell_type_cn]]
        clustering <- readLines(clustering_chosen.fp)
        clustering %<=>% hs.factor(x, unique(x) %=>% as.numeric %=>% 
            sort %=>% paste, N = 0)
        seurat_data@meta.data[, clustering_chosen_cn] <- clustering
        SeuratObject::Idents(seurat_data) <- clustering
        if (0) {
            clusterings <- seurat_data@meta.data %>% {
                .[, names(.) %>% hs.grep("^(?i)monocle3_res\\."), 
                  drop = 0]
            }
            ci <- which(apply(clusterings, 2, . %>% {
                a <- hs.table(a[[2]], .)
                all(hs.colSums(a > 0) == 1)
            }))
            if (length(ci) != 1) {
                hs.browser("38.09.09.152110")
                hs.table(clusterings[, tail(ci, 1)], a[[2]]) %>% 
                  .s
                clustering <- clusterings[, "monocle3_res.6.103515625e_06"]
            }
            clustering <- clusterings[, ci] %>% {
                hs.factor(., sort(as.numeric(unique(.))), N = 0)
            }
            hs.table(a[, cell_type], clustering) %>% .s
        }
    }
    else {
        hs.msg("预测细胞类型")
        if ((org1 %in% c("human", "mouse")) && (tissue_type == 
            "血")) {
            cell_type_ref <- hs.switch(org1, "human", hs.hsgly("转录组/标准流程/模块/r函数_自制.38_04_07_003149", 
                "get.cell_type_ref")(org1 = "human"), "mouse", 
                hs.hsgly("R/包/celldex/数据.38_03_03_143112")("ImmGenData"))
            seurat_data <- hs.hsgly("生信/1.流程/单细胞/转录组/预测细胞类型/用SingleR.38_09_06_141838")(seurat_data = seurat_data, 
                ref = cell_type_ref, cell_type_column_name = cell_type_cn, 
                clustering_chosen_cn = clustering_chosen_cn)
        }
        else {
            if (length(cell_type_marker_list)) {
                if (length(cell_type_marker_list) == 1) {
                  cell_type_marker_list <- hs.switch(paste0(org1, 
                    ".", cell_type_marker_list), "mouse.肾", 
                    hs.c(c("Endo", "Podo", "PT", "LOH", "DCT", 
                      "CD-PC", "CD-IC", "CD-Trans", "Novel1", 
                      "Fib", "Macro", "Neutro", "B", "T", "Nk", 
                      "Novel2"), list(`1` = c("Nrp1", "Kdr"), 
                      `2` = c("Nphs1", "Nphs2"), `3` = c("Slc27a2", 
                        "Lrp2"), `4` = c("Slc12a1", "Umod"), 
                      `5` = c("Umod", "Slc12a3", "Pvalb"), `6` = c("Aqp2", 
                        "Hsd11b2"), `7` = c("Atp6v1g3", "Atp6v0d2", 
                        "Insrr", "Rhbg"), `8` = c("Umod", "Aqp2", 
                        "Hsd11b2", "Atp6v1g3", "Atp6v0d2", "Insrr", 
                        "Rhbg"), `9` = c("Slc27a2", "Lrp2", "Mki67", 
                        "Cdca3", "Stmn1"), `10` = c("Plac8", 
                        "S100a4"), `11` = c("C1qa", "C1qb"), 
                      `12` = c("S100a8", "S100a9", "Ltb"), `13` = c("Cd79a", 
                        "Cd79b", "Ltb"), `14` = c("Ltb", "Cxcr6", 
                        "Nkg7"), `15` = c("Gzma", "Nkg7"), `16` = c("Mki67", 
                        "Cdca3", "Plac8", "S100a4", "Ltb", "Gzma", 
                        "Nkg7", "Stmn1"))), "human.血", list(`Naive T` = c("cd27", 
                      "ccr7", "cd8a", "cd8b"), `CD4+ memory` = c("il7r", 
                      "cd27", "ccr7"), NKT = c("znf683", "cd8a", 
                      "cd8b"), B = c("cd79a", "cd37"), `CD8+ T` = c("gzmk", 
                      "cd8a", "cd8b"), NK = c("cd160", "nkg7", 
                      "gnly", "cd247", "CCL3", "GZMB"), `CD16+ and CD14+ monocytes` = c("CD68", 
                      "FCGR3A", "CD14", "S100A12"), `Regulatory T` = c("CCR10", 
                      "IL2RA", "CD52", "CMTM7", "FOXP3"), `Monocyte derived dendritic` = c("CST3", 
                      "CD1C", "FCER1A"), `Megakaryocyte progenitors` = c("PF4", 
                      "PPBP", "PLA2G12A"), `Progenitor-NK` = c("ID2"), 
                      `Plasmacytoid dendritic cells` = c("GZMB", 
                        "CD123")))
                }
                if (cell_type_marker_list.org != org1) {
                  org2 <- hs.switch(cell_type_marker_list.org, 
                    "human", , "mouse", cell_type_marker_list.org, 
                    hs.browser("38.09.06.100751"))
                  a <- cell_type_marker_list %>% unlist %>% tolower %>% 
                    hs.clean(U = 1) %>% wk.dt
                  wk.dt.reset_name(a)
                  b <- hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937")(a[, 
                    V1])
                  a[, `:=`("symbol_updated", b)]
                  hs.hsgly("同源基因之表格.37_05_27_153942")(a, 
                    org1 = "human", org2 = "goat", gene.cn = "symbol_updated")
                  cell_type_marker_list %<>% lapply(hs.str_and, 
                    a[, symbol_updated])
                }
                if (0) {
                  a <- hs.hsgly("单细胞/转录组/预测细胞类型/用特征基因.38_08_26_122205")(seurat_data, 
                    cell_type_marker_list %>% {
                      .[names(.) %not.in% "Mono"]
                    }, cluster_n_max = 100, ot = "list")
                  clusterings %>% apply(2, uniqueN) %>% {
                    sum(. < 100)
                  }
                  a[[1]] %>% .s
                  a[[2]] %>% lapply(. %>% {
                    names(cell_type_marker_list) %not% .
                  })
                  a[[2]] %>% lapply(. %>% {
                    (names(cell_type_marker_list) %not% "Mono") %not% 
                      .
                  })
                  names(a[[2]]) %>% {
                    wk.dt(分辨率 = ., 细胞类型数 = a[[2]][.] %>% 
                      sapply(uniqueN), 群数 = clusterings[, 
                      .] %>% apply(2, uniqueN))
                  }
                  hs.str_and(cell_type_marker_list[["Neutrophil"]], 
                    rownames(seurat_data))
                  tolower(cell_type_marker_list[["Neutrophil"]]) %and% 
                    tolower(rownames(seurat_data))
                }
                cluster.cell_type <- hs.hsgly("单细胞/转录组/预测细胞类型/用特征基因.38_08_26_122205")(seurat_data, 
                  cell_type_marker_list, ot = "most_cell_types")
                clusterings <- seurat_data@meta.data %>% {
                  .[, names(.) %>% hs.grep("^(?i)monocle3_res\\."), 
                    drop = 0]
                }
                clustering <- clusterings[, names(cluster.cell_type)] %>% 
                  as.integer %>% {
                  hs.factor(., sort(unique(.)), N = 0)
                }
                SeuratObject::Idents(seurat_data) <- clustering
                seurat_data@meta.data[, clustering_chosen_cn] <- clustering
                seurat_data@meta.data[, cell_type_cn] <- wk.dt.subset.na(cluster.cell_type[[1]], 
                  clustering %>% paste)[, cell_type]
            }
            else {
                if (length(tissue_type)) 
                  seurat_data <- hs.hsgly("单细胞/转录组/预测细胞类型/用seurat.38_09_14_164012")(seurat_data = seurat_data, 
                    org1 = org1, tissue_type = tissue_type, cell_type_column_name = cell_type_cn, 
                    clustering_cn2use = clustering_cn2use, clustering_chosen_cn = clustering_chosen_cn)
            }
        }
        hs.msg("完成", 0)
        seurat_data@meta.data %>% {
            hs.c(c("barcode", cell_type_cn), list(rownames(.), 
                .[, cell_type_cn]))
        } %>% wk.dt.fwrite(wj2)
        seurat_data@meta.data %=>% list(rownames(x), x[, clustering_chosen_cn]) %=>% 
            wk.dt.fwrite(clustering_chosen.fp)
        seurat_data@meta.data %=>% {
            hs.table(x[, "cell_cluster"], x[, "cell_type"])
        } %=>% .s
    }
    wj2 <- hs.fp(wjj2, "cell_cluster/2d.png")
    local({
        if (file.exists(wj2)) 
            return()
        P <- hs.lapply("umap,tsne", function(reduction_method_1) {
            Seurat::DimPlot(seurat_data, reduction = reduction_method_1, 
                label = 1, label.size = 2, repel = 0, cols = hs.hsgly("color.db.38_09_09_173646")()) + 
                ggplot2::labs(title = switch(reduction_method_1, 
                  umap = "UMAP", tsne = "tSNE"))
        })
        cmd1 <- paste0("P[[ ", seq_along(P), "]]") %>% paste(collapse = " + ")
        p1 <- evalStr(cmd1)
        .figs(print(p1), wj2, width = 10 + 10 * 2, height = 10)
    })
    hs.hsgly("标准流程/模块/细胞分群、类型预测/2d.cell_type.38_09_09_172309")(seurat_data = seurat_data, 
        wj2 = hs.fp(wjj2, "cell_type/2d.png"), cell_type_column_name = cell_type_cn)
    null <- lapply(seq(2), function(i1) {
        rn1 <- c("umap", "tsne")[i1]
        if (rn1 %not.in% names(seurat_data@reductions)) 
            return()
        wj2 <- hs.fp(wjj2, "cell_cluster", paste0("2d.", rn1, 
            ".xy.tsv.gz"))
        if (!file.exists(wj2)) {
            a <- seurat_data@reductions[[rn1]]@cell.embeddings %>% 
                wk.dt.as(keep.rownames = "barcode")
            wk.dt.fwrite(a, wj2)
        }
    })
    hs.with_wd(hs.wjj(wjj2, "cell_type"), {
        wj2 <- "barcode.cell_type.tsv.gz"
        if (file.exists(wj2)) 
            return()
        a <- wk.dt.as(seurat_data@meta.data[, cell_type_cn, drop = 0], 
            1)
        wk.dt.setnames(a, c("barcode", "cell_type"))
        wk.dt.fwrite(a, wj2) %>% normalizePath
    })
    seurat_data
}
"38_10_09_062553" <- function(seurat_data, wjj2, sample.cn, cell_type_cn) {
    wj2 <- hs.fp(wjj2, "cell.cluster.tsv.gz")
    if (!file.exists(wj2)) {
        a <- wk.dt(colnames(seurat_data), SeuratObject::Idents(seurat_data))
        wk.dt.setnames(a, c("barcode", "cluster"))
        wk.dt.fwrite(a, wj2)
    }
    a <- wk.dt(Cell_cluster = SeuratObject::Idents(seurat_data), 
        Sample = seurat_data@meta.data[, sample.cn])
    wj2 <- hs.fp(wjj2, "sample_x_cluster.n.tsv.gz")
    if (!file.exists(wj2)) 
        local({
            b <- a[, hs.table(Cell_cluster, Sample)]
            jg <- apply(b, 2, . %>% {
                paste0(., "(", ./sum(.), ")")
            })
            rownames(jg) <- rownames(b)
            jg <- wk.dt.as(jg, keep.rownames = "cluster")
            x <- unique(wk.dt(as.numeric(Idents(seurat_data)), 
                seurat_data@meta.data[, cell_type_cn])) %>% wk.dt.setorderv("V1")
            jg[, `:=`("major_cell_type", x[, V2])]
            wk.dt.col.reorder(jg, "major_cell_type", after = 1)
            wk.dt.fwrite(jg, wj2)
        })
    wj2 <- hs.fp(wjj2, "prop.sample.png")
    if (!file.exists(wj2)) 
        local({
            p1 <- ggplot2::ggplot(a, ggplot2::aes(Sample)) + 
                ggplot2::geom_bar(ggplot2::aes(fill = Cell_cluster), 
                  stat = "count", position = "fill") + ggplot2::scale_fill_manual(values = hs.hsgly("color.db.38_09_09_173646")()) + 
                ggplot2::theme(panel.background = {
                }, panel.grid = ggplot2::element_blank()) + ggplot2::xlab("") + 
                ggplot2::ylab("Prop. of cells") + ggplot2::theme(legend.position = "right", 
                legend.title = ggplot2::element_blank(), legend.direction = "vertical", 
                axis.text.x = element_text(angle = 45, hjust = 1, 
                  vjust = 1))
            .figs(print(p1), wj2, width = 30, height = 20)
        })
    if ("sample_group_name.38_08_16_114013" %in% colnames(seurat_data@meta.data)) {
        hs.browser("38.08.31.164805")
        a <- seurat_data@meta.data %>% {
            a <- tapply(SeuratObject::Idents(seurat_data), .[, 
                "sample_group_name.38_08_16_114013"], hs.table)
            b <- tapply(.[, "sample_name.38_01_19_171144"], .[, 
                "sample_group_name.38_08_16_114013"], uniqueN)
            for (i in seq_along(a)) {
                a[[i]] <- a[[i]]/b[i]
            }
            b <- lapply(a, names) %>% wk.dt.fromList(nm = c("Group", 
                "Cell"))
            b[, `:=`("Avg. prop.", unlist(a))]
        }
        a[, `:=`(Cell, Cell %>% hs.factor(N = 0))]
        wj2 <- hs.fp(wjj2, "mid/cell_cluster/prop.group")
        if (!file.exists(wj2 %>% paste0(".png"))) 
            local({
                p1 <- ggplot2::ggplot(a, ggplot2::aes(Group, 
                  `Avg. prop.`)) + ggplot2::geom_bar(ggplot2::aes(fill = Cell), 
                  stat = "identity", position = "fill") + ggplot2::scale_fill_manual(values = allcolour) + 
                  ggplot2::theme(panel.background = {
                  }, panel.grid = ggplot2::element_blank()) + 
                  ggplot2::xlab("") + ggplot2::theme(legend.position = "right", 
                  legend.title = ggplot2::element_blank(), legend.direction = "vertical", 
                  axis.text.x = element_text(angle = 45, hjust = 1, 
                    vjust = 1))
                .figs(print(p1), wj2, width = 30, height = 20)
            })
    }
    return()
    if (0) {
        wj2 <- hs.fp(wjj2, "mid/cell_type/prop.png")
        if (!file.exists(wj2)) 
            local({
                seurat_data@meta.data[1, ]
                a <- seurat_data@meta.data %>% {
                  a <- wk.dt(Cell = .[, "cell_type"], Sample = .[, 
                    "sample_name.38_01_19_171144"] %>% hs.sub("-\\d+$"))
                  if ("sample_group_name.38_08_16_114013" %in% 
                    colnames(.)) 
                    a[, `:=`("Group", .[, "sample_group_name.38_08_16_114013"])]
                  a
                }
                {
                  p1 <- ggplot2::ggplot(a, ggplot2::aes(Cell)) + 
                    ggplot2::geom_bar(ggplot2::aes(fill = Sample), 
                      stat = "count", position = "fill") + ggplot2::theme(panel.background = {
                  }, panel.grid = ggplot2::element_blank()) + 
                    ggplot2::xlab("") + ggplot2::ylab("Prop. of cells") + 
                    ggplot2::labs(title = "Sample") + ggplot2::coord_flip() + 
                    ggplot2::theme(legend.position = "bottom", 
                      legend.title = ggplot2::element_blank(), 
                      legend.direction = "vertical")
                  if ("Group" %in% names(a)) {
                    p2 <- ggplot2::ggplot(a, ggplot2::aes(Cell)) + 
                      ggplot2::geom_bar(ggplot2::aes(fill = Group), 
                        stat = "count", position = "fill") + 
                      ggplot2::theme(panel.background = {
                      }, panel.grid = ggplot2::element_blank()) + 
                      ggplot2::xlab("") + ggplot2::ylab("Prop. of cells") + 
                      ggplot2::labs(title = "Group") + ggplot2::coord_flip() + 
                      ggplot2::theme(legend.position = "bottom", 
                        legend.title = ggplot2::element_blank(), 
                        legend.direction = "vertical") + ggplot2::theme(axis.text.y = element_blank(), 
                      axis.ticks.y = element_blank())
                    p1 <- p1 + p2
                  }
                  p2 <- ggplot2::ggplot(a[, hs.table(Cell, prop = 1) %>% 
                    {
                      .(names(.), .)
                    }], ggplot2::aes(Cell, N)) + ggplot2::geom_bar(stat = "identity") + 
                    ggplot2::theme(panel.background = {
                    }, panel.grid = ggplot2::element_blank()) + 
                    ggplot2::xlab("") + ggplot2::ylab("Prop. of cells") + 
                    ggplot2::labs(title = "Cell type") + ggplot2::coord_flip() + 
                    ggplot2::theme(legend.position = "bottom", 
                      legend.title = ggplot2::element_blank()) + 
                    ggplot2::theme(axis.text.y = element_blank(), 
                      axis.ticks.y = element_blank())
                  p1 <- p1 + p2
                  .png(print(p1), wj2, width = 30, height = 20)
                }
            })
    }
    if (0) {
        wj2 <- hs.fp(wjj2, "mid/cell_type/stat.pdf")
        if (file.exists(wj2)) {
            return()
        }
        fig_n <- seurat_data@meta.data[, cell_type_cn] %>% uniqueN
        .pdf({
            cell_cluster.cell_type_general <- seurat_data@meta.data[, 
                cell_type_cn]
            cell_cluster_2use <- cell_cluster.cell_type_general %>% 
                hs.table %>% hs.sort %>% {
                paste(names(.))
            }
            fig <- hs.lapply(seq_along(cell_cluster_2use), function(i1) {
                message(i1)
                cell_type_1 <- paste(cell_cluster_2use[i1])
                cell_type_1.cell_cluster <- SeuratObject::Idents(seurat_data)[paste(cell_cluster.cell_type_general) %in% 
                  cell_type_1] %>% hs.table %>% {
                  names(.)[. > 0]
                }
                a <- wk.dt(cell = colnames(seurat_data)[hs.factor(SeuratObject::Idents(seurat_data), 
                  ot = "str") %in% cell_type_1.cell_cluster])
                a[, `:=`(c(cell_type_1), SeuratObject::Idents(seurat_data)[cell] %>% 
                  {
                    hs.factor(., rev(sort(unique(.))), N = 0)
                  })]
                sample_name_ordered <- local({
                  a <- wk.dt(colnames(seurat_data) %>% hs.re("\\d+$"), 
                    seurat_data@meta.data[, "sample_name.38_01_19_171144"]) %>% 
                    unique
                  a[, `:=`(V1, as.numeric(V1))]
                  wk.dt.setorderv(a, "V1")
                  a[[2]]
                })
                a[, `:=`("sample", hs.factor(seurat_data@meta.data[cell, 
                  "sample_name.38_01_19_171144"], rev(sample_name_ordered), 
                  N = 0))]
                a[, `:=`("umi_n", hs.colSums(seurat_data@assays[["RNA"]]@counts[, 
                  cell]))]
                a[, `:=`(cell, {
                })]
                if (length(levels(a[[match(cell_type_1, names(a))]])) > 
                  1) {
                  b <- c("Merged", a[, -1])
                  a <- wk.dt.rbind(a, b)
                }
                if (i1%%2) 
                  wk.dt.setnames(a, cell_type_1, paste0("\n\n", 
                    cell_type_1))
                if (0) {
                  tapply(a[, sample], a[, T_cells], . %>% hs.table %>% 
                    {
                      ./sum(.) * 100
                    })
                }
                p1 <- ggplot2::ggplot(a, do.call(ggplot2::aes, 
                  list(as.symbol(paste0(ifelse(i1%%2, "\n\n", 
                    ""), cell_type_1))))) + ggplot2::geom_bar(ggplot2::aes(fill = sample), 
                  stat = "count", position = "fill") + ggplot2::theme(panel.background = {
                }, panel.grid = ggplot2::element_blank()) + ggplot2::coord_flip()
                if (i1 == 1) {
                  p1 <- p1 + ggplot2::labs(title = "Samples") + 
                    ggplot2::theme(legend.position = "none")
                }
                else if (i1 == length(cell_cluster_2use)) {
                  p1 <- p1 + ggplot2::theme(legend.position = "bottom", 
                    legend.direction = "vertical", legend.justification = c(0, 
                      1), legend.title = ggplot2::element_blank()) + 
                    ggplot2::guides(fill = ggplot2::guide_legend(reverse = 1))
                }
                else p1 <- p1 + ggplot2::theme(legend.position = "none")
                p1 <- p1 + if (i1 == length(cell_cluster_2use)) {
                  ggplot2::ylab("Fraction of cells") + ggplot2::scale_y_continuous(labels = function(x1) paste0(x1 * 
                    100, "%"))
                }
                else ggplot2::scale_y_continuous(labels = function(x1) paste0(x1 * 
                  100, "%"), name = {
                })
                p2 <- local({
                  a <- SeuratObject::Idents(seurat_data) %>% 
                    {
                      .[. %in% cell_type_1.cell_cluster]
                    } %>% hs.factor(N = 0) %>% hs.table %>% wk.dt %>% 
                    wk.dt.setnames(c(cell_type_1, "cell_n"))
                  if (a[, .N > 1]) 
                    a[, `:=`(c(cell_type_1), get(cell_type_1) %>% 
                      {
                        hs.factor(., rev(sort(as.numeric(unique(.)))), 
                          N = 0)
                      })]
                  if (length(levels(a[[paste(cell_type_1)]])) > 
                    1) 
                    a <- wk.dt.rbind(a, c("Merged", a[, -1]))
                  ggplot2::ggplot(a, do.call(ggplot2::aes, list(as.symbol(cell_type_1)))) + 
                    ggplot2::geom_bar(ggplot2::aes(y = cell_n), 
                      stat = "identity") + ggplot2::xlab("") + 
                    ggplot2::theme(panel.background = {
                    }, panel.grid = ggplot2::element_blank()) + 
                    ggplot2::coord_flip()
                })
                if (i1 == 1) 
                  p2 <- p2 + ggplot2::labs(title = "Cells")
                p2 <- p2 + if (i1 == length(cell_cluster_2use)) {
                  ggplot2::ylab("Number of cells")
                }
                else ggplot2::scale_y_continuous(name = {
                })
                p3 <- ggplot2::ggplot(a, do.call(ggplot2::aes, 
                  list(as.symbol(paste0(ifelse(i1%%2, "\n\n", 
                    ""), cell_type_1)), as.symbol("umi_n")))) + 
                  ggplot2::geom_boxplot(fill = "red") + ggplot2::stat_summary(ggplot2::aes(y = umi_n), 
                  fun = . %>% {
                    quantile(., 0.5)
                  }, geom = "point", fill = "white", shape = 21, 
                  size = 2) + ggplot2::xlab("") + ggplot2::theme(panel.background = {
                }, panel.grid = ggplot2::element_blank()) + ggplot2::coord_flip()
                if (i1 == 1) 
                  p3 <- p3 + ggplot2::labs(title = "UMI counts")
                p3 <- p3 + if (i1 == length(cell_cluster_2use)) {
                  ggplot2::ylab("Number of transcripts")
                }
                else ggplot2::scale_y_continuous(name = {
                })
                list(p1, p2, p3)
            })
            fig_flat <- hs.unlist(fig, 0)
            fig_sub_row_n <- sapply(seq_along(cell_cluster_2use), 
                function(i1) {
                  cell_type_1 <- cell_cluster_2use[i1]
                  cell_type_1.cell_cluster <- SeuratObject::Idents(seurat_data)[paste(cell_cluster.cell_type_general) %in% 
                    cell_type_1] %>% hs.table %>% {
                    names(.)[. > 0]
                  }
                  length(cell_type_1.cell_cluster)
                })
            print(evalStr(paste0("fig_flat[[ ", seq_along(fig_flat), 
                "]]") %>% paste(collapse = " + ")) + patchwork::plot_layout(ncol = 3, 
                heights = fig_sub_row_n + 0.1))
        }, wj2, width = c(10 * 3), height = 3 * fig_n)
    }
}
"38_10_18_160507" <- function(seurat_data, cell_type_cn, sample.cn, 
    sample_group.cn, cell_type_list, wjj2) {
    wj2 <- hs.fp(wjj2, "prop.sample")
    a <- seurat_data@meta.data %=>% {
        wk.dt(Cell = x[, cell_type_cn], Sample = x[, sample.cn] %=>% 
            hs.sub("-\\d+$"))
    }
    hs.fp(dirname(wj2), "sample_x_type.n.tsv.gz") %=>% {
        if (!file.exists(x)) {
            b <- a[, hs.table(Cell, Sample)]
            jg <- apply(b, 2, . %=>% paste0(x, "(", x/sum(x), 
                ")"))
            rownames(jg) <- rownames(b)
            jg <- wk.dt.as(jg, keep.rownames = "cell_type")
            if (length(cell_type_list)) {
                a1 <- a[, Cell %=>% unique %=>% list]
                a1[, `:=`("V2", hs.sub(V1, "_\\d+$"))]
                a1[, `:=`("detail", ifelse(V2 %in% names(cell_type_list), 
                  cell_type_list[V2], V1))]
                c1 <- a1[match(jg[, cell_type], V1), detail]
                jg[, `:=`("cell_type_detail", c1)]
                wk.dt.col.reorder(jg, "cell_type_detail", after = 1)
            }
            wk.dt.fwrite(jg, x)
        }
    }
    if (!file.exists(paste0(wj2, ".png"))) {
        p1 <- ggplot2::ggplot(a, ggplot2::aes(Sample)) + ggplot2::geom_bar(ggplot2::aes(fill = Cell), 
            stat = "count", position = "fill") + ggplot2::scale_fill_manual(values = hs.hsgly("color.db.38_09_09_173646")()) + 
            ggplot2::theme(panel.background = {
            }, panel.grid = ggplot2::element_blank()) + ggplot2::xlab("") + 
            ggplot2::ylab("Prop. of cells") + ggplot2::theme(legend.position = "right", 
            legend.title = ggplot2::element_blank(), legend.direction = "vertical", 
            axis.text.x = element_text(angle = 45, hjust = 1, 
                vjust = 1))
        .figs(print(p1), wj2, width = 30, height = 20)
    }
    if (sample_group.cn %in% colnames(seurat_data@meta.data)) {
        wj2 <- hs.fp(wjj2, "prop.group.png")
        if (!file.exists(wj2)) 
            local({
                a <- seurat_data@meta.data %>% {
                  a <- tapply(.[, cell_type_cn], .[, sample_group.cn], 
                    hs.table)
                  b <- tapply(.[, sample.cn], .[, sample_group.cn], 
                    uniqueN)
                  for (i in seq_along(a)) {
                    a[[i]] <- a[[i]]/b[i]
                  }
                  b <- lapply(a, names) %>% wk.dt.fromList(nm = c("Group", 
                    "Cell"))
                  b[, `:=`("Avg. prop.", unlist(a))]
                }
                p1 <- ggplot2::ggplot(a, ggplot2::aes(Group, 
                  `Avg. prop.`)) + ggplot2::geom_bar(ggplot2::aes(fill = Cell), 
                  stat = "identity", position = "fill") + ggplot2::scale_fill_manual(values = hs.hsgly("color.db.38_09_09_173646")()) + 
                  ggplot2::theme(panel.background = {
                  }, panel.grid = ggplot2::element_blank()) + 
                  ggplot2::xlab("") + ggplot2::theme(legend.position = "right", 
                  legend.title = ggplot2::element_blank(), legend.direction = "vertical", 
                  axis.text.x = element_text(angle = 45, hjust = 1, 
                    vjust = 1))
                .figs(print(p1), wj2, width = 30, height = 20)
            })
    }
}
"38_10_18_094624" <- function(seurat_data, wjj2) {
    wj2 <- hs.fp(wjj2, paste0("cluster_similarities.", c("corr_heat", 
        "pca"), ".png"))
    if (!all(file.exists(wj2))) {
        a1 <- SeuratObject::Idents(seurat_data)
        if (uniqueN(a1) < 3) 
            return()
        m1 <- sapply(seq(uniqueN(a1)), function(i1) {
            l1 <- levels(a1)[i1]
            bc <- colnames(seurat_data)[a1 %in% l1]
            hs.rowMedians(seurat_data@assays$RNA@data[, bc])
        })
        colnames(m1) <- paste0("cluster_", unique(a1))
        m1 %<=>% hs.ij(x, hs.rowAnys(x > 0))
        if (!file.exists(wj2[1])) {
            cor1 <- cor(m1, method = "spearman")
            cor1 %=>% wk.dt(keep.rownames = "cluster") %=>% wk.dt.fwrite(wj2[1] %=>% 
                hs.sub("_heat$", ".tsv.gz"))
            p1 <- pheatmap::pheatmap(cor1)
            P <- ggplotify::as.ggplot(p1)
            height1 <- width1 <- 5 + ncol(m1)
            .figs(print(P), wj2[1], width = width1, height = height1)
        }
        if (!file.exists(wj2[2])) 
            hs.hsgly("pca.36.04.07.121916")(m1, wj = wj2[2], 
                plotter = .figs)
    }
}
"38_12_16_154003" <- local({
    hs.38_07_19_104409 <- function(f1, f2, score, classify = 0) {
        f2.db <- sort(unique(f2))
        if (0) {
            m1 %=>% wk.dt.fromList(x, by_row = 1)
            f1 %=>% tapply(seq_along(x), x, function(x) head(x))
            f1 %=>% tapply(seq_along(x), x, . %=>% head)
        }
        m1 <- tapply(seq_along(f1), f1, c) %=>% sapply(x, function(x) {
            jg <- hs.c(f2.db, rep(0, length(f2.db)))
            a <- tapply(score[x], f2[x], sum)
            jg[names(a)] <- a
            jg
        })
        if (0) {
            .s(m1)
            .s(t(m1))
            m1[1, ]
            dim(m1)
            chk(m1)
            round(m1, 2) %>% .s
        }
        jg <- if (classify) {
            apply(m1, 2, . %=>% {
                x <- hs.sort(x)
                names(x[1])
            })
        }
        else {
            apply(m1, 2, . %=>% {
                x <- hs.sort(x)
                x <- x/sum(x)
                (x[1] - x[2])/2
            })
        }
        return(jg)
        if (0) {
            .s(m2)
            m1[, 1] %=>% hs.sort %=>% ((x[1] - x[2])/2)
            round(m2, 2) %>% .s
        }
        if (classify) {
            apply(m2, 2, . %>% {
                names(.)[which.max(.)]
            })
        }
        else apply(m2, 2, function(x) {
            a <- hs.sort(x)
            1 - a[2]/a[1]
        })
    }
    function(clusterings = {
    }, cell_anno, cluster_n_max = 40, clustering = {
    }, cell_type_hierarchy = {
    }) {
        if (0) {
            hs.update(clusterings, x[, !i, drop = 0], hs.if = {
                i <- apply(x, 2, uniqueN) > cluster_n_max
                any(i)
            })
        }
        local({
            if (!length(clusterings)) 
                return()
            i <- apply(clusterings, 2, uniqueN) > cluster_n_max
            if (any(i)) 
                clusterings %<<=>% x[, !i, drop = 0]
        })
        certainty_score_relative <- cell_anno %=>% {
            hs.browser("38.12.19.102555", debug = 0)
            if (inherits(x, c("list"))) {
                if (length(x) > 3) 
                  hs.browser("38.12.19.111205", debug = 0)
                i <- 1
                bc1 <- "Th431-AACACACAGCAGGCTA-1"
                bc1 <- "Tp695_516446-TTTGTCATCGAACGGA-1"
                i <- match(bc1, rownames(x[[1]]))
                ct1.ct2 <- cell_type_hierarchy %=>% tapply(x[[2]], 
                  x[[1]], c)
                x1_from_2 <- lapply(ct1.ct2[names(x[[1]])], function(ct2) {
                  hs.rowMaxs(x[[2]][, ct2])
                }) %=>% wk.dt.as
                confidence1 <- apply(x1_from_2, 1, . %=>% {
                  x <- hs.sort(x)
                  x[1] - x[2]
                })
                ct2.ct2 <- cell_type_hierarchy %=>% hs.lapply(x[[2]], 
                  function(ct2) {
                    i <- match(ct2, x[[2]])
                    x1 <- x[[1]]
                    i <- x1 %in% x1[i]
                    x[[2]][i]
                  })
                a <- wk.mclapply(seq(nrow(x[[2]])), function(i) {
                  x <- hs.sort(unlist(x[[2]][i, ]))
                  ct2 <- names(x[1])
                  ct2 <- ct2.ct2[[ct2]]
                  if (length(ct2) == 1) 
                    return(list(names(x[1]), x[1]))
                  x <- x[names(x) %in% ct2]
                  list(names(x[1]), x[1] - x[2])
                })
                dt <- rbindlist(unname(a))
                wk.dt.setnames(dt, c("ct", "confidence"))
                dt[, `:=`(confidence, confidence1 + confidence * 
                  confidence1)]
                {
                  certainty_score_relative <- dt[, confidence]/2
                  cell_anno_best <- dt[, ct]
                  cell_anno_best %=>% uniqueN
                }
                x1_from_2[i] %=>% hs.sort
                x[[2]][i, ] %=>% unlist %=>% tapply(x, wk.dt.subset(cell_type_hierarchy, 
                  names(x), 2, exp_j = .SD[[1]]), max) %=>% hs.sort %=>% 
                  (x[1] - x[2])
                x[[1]][i, ] %=>% unlist %=>% hs.sort %=>% (x[1] - 
                  x[2])
                wk.dt.subset(cell_type_hierarchy, names(x), 2, 
                  exp_j = .SD[[1]])
                x[[1]][i, ]
                dt <- wk.dt()
                wk.dt.set(dt, , "bc", rownames(x[[1]]))
                wk.dt.set(dt, , "ct1", local({
                  i <- apply(x[[1]], 1, which.max)
                  names(x[[1]])[i]
                }))
                wk.dt.set(dt, , "confidence1", apply(x[[1]], 
                  1, function(x) {
                    x <- hs.sort(x)
                    x[1] - x[2]
                  }))
                a <- wk.mclapply(seq(dt[, .N]), function(i) {
                  bc1 <- dt[i, bc]
                  ct1 <- dt[i, ct1]
                  ct2 <- wk.dt.subset(cell_type_hierarchy, ct1, 
                    exp_j = .SD[[2]])
                  x <- unlist(x[[2]][bc1, ct2])
                  jg <- if (length(ct2) == 1) {
                    list(ct2, x)
                  }
                  else {
                    x <- hs.sort(x)
                    list(names(x[1]), x[1] - x[2])
                  }
                  wk.dt.as(jg)
                }) %=>% unname %=>% rbindlist
                wk.dt.set(dt, , c("ct2", "confidence2"), a)
                a <- cell_type_hierarchy %=>% tapply(x[[2]], 
                  x[[1]], length)
                wk.dt.set(dt, , "ct2_n", dt[, a[ct1]])
                dt[, ifelse(ct2_n > 1, )]
                bc1 <- "Tp695_516446-TTTGTCATCGAACGGA-1"
                x[[1]][bc1, ] %=>% sort
                x[[2]][bc1, ] %=>% sort
                ct1 <- "ILC"
                ct1 <- "Fibroblasts"
                wk.dt.subset(cell_type_hierarchy, ct1)
                wk.dt.subset(dt, bc1, "bc")
                bc1 <- "Th431-AAACCCACAGACAATA-1"
                dt[, match(bc1, bc)]
                ct1 <- wk.dt.subset(dt, bc1, "bc", exp_j = ct1)
                dt[1, sapply(seq(.N), function(i) {
                  ct2 <- wk.dt.subset(cell_type_hierarchy, ct1, 
                    exp_j = .SD[[2]])
                  x <- x[[2]][bc1, ct2] %=>% unlist
                  if (length(x) == 1) 
                    return(hs.c(ct2, 0))
                  hs.sort(x) %=>% hs.c(names(x[1]), x[1] - x[2])
                })]
                a <- wk.mclapply(dt[, seq(.N)], function(i) {
                  bc1 <- names(clearness[i])
                  ct1 <- ct1_best[i]
                  ct2 <- wk.dt.subset(cell_type_hierarchy, ct1, 
                    exp_j = .SD[[2]])
                  x <- x[[2]][bc1, ct2] %=>% unlist
                  if (length(x) == 1) 
                    return(hs.c(ct2, 0))
                  hs.sort(x) %=>% hs.c(names(x[1]), x[1] - x[2])
                })
                x[[1]][1, ]
                clearness <- apply(x[[1]], 1, function(x) {
                  a <- hs.sort(x)
                  a[1] - a[2]
                })
                bc1 <- "Th431-AACACACAGCAGGCTA-1"
                x[[1]][bc1, ] %=>% hs.sort %=>% {
                  x[1] - x[2]
                }
                x[[1]][bc1, ] %=>% hs.sort
                clearness[bc1]
                ct1 <- x[[1]][bc1, ] %=>% hs.sort %=>% names(x)[1]
                ct2 <- wk.dt.subset(cell_type_hierarchy, ct1, 
                  exp_j = .SD[[2]])
                x[[2]][bc1, ct2] %=>% hs.sort %=>% (x[1] - x[2])
                ct1_best <- local({
                  i <- apply(x[[1]], 1, which.max)
                  names(x[[1]])[i]
                })
                a <- wk.mclapply(seq_along(clearness), function(i) {
                  bc1 <- names(clearness[i])
                  ct1 <- ct1_best[i]
                  ct2 <- wk.dt.subset(cell_type_hierarchy, ct1, 
                    exp_j = .SD[[2]])
                  x <- x[[2]][bc1, ct2] %=>% unlist
                  if (length(x) == 1) 
                    return(hs.c(ct2, 0))
                  hs.sort(x) %=>% hs.c(names(x[1]), x[1] - x[2])
                })
                clearness2 <- unlist(unname(a))
                clearness %=>% (x + x * clearness2) %=>% {
                  .pdf(plot(ieiwk.density(x)))
                }
                "/dev/shm/wengkai/tmp.pdf" %=>% wk.rsync.in
            }
            else {
                if (max(hs.rowSums(x)) > 1) {
                  1 - hs.rowMaxs(x)
                }
                else apply(x, 1, function(x) {
                  a <- hs.sort(x)
                  1 - a[2]/a[1]
                })
            }
        }
        cell_anno_best <- cell_anno %=>% {
            colnames(x)[apply(x, 1, which.max)]
        }
        if (length(clustering)) {
            return(hs.38_07_19_104409(clustering, cell_anno_best, 
                certainty_score_relative, classify = 1))
        }
        clustering_i <- ncol(clusterings)
        clustering_i <- 19
        a <- sapply(1:ncol(clusterings), function(clustering_i) {
            message(clustering_i)
            clustering <- clusterings[, clustering_i]
            a <- hs.38_07_19_104409(clustering, cell_anno_best, 
                certainty_score_relative)
            b <- hs.38_07_19_104409(cell_anno_best, clustering, 
                certainty_score_relative)
            c(mean(a, na.rm = 1), mean(b))
        }) %=>% t
        a <- sapply(1:ncol(clusterings), function(clustering_i) {
            message(clustering_i)
            clustering <- clusterings[, clustering_i]
            {
                m1 <- list(cell_anno_best, clustering, seq_along(clustering)) %=>% 
                  wk.dt.as %=>% wk.dt.setnames(x, c("ct", "cl", 
                  "i"))[, sum(certainty_score_relative[i]), by = .(ct, 
                  cl)] %=>% wk.dt.widen(x, rn = "ct", cn = "cl", 
                  vn = "V1") %=>% x[hs.order(hs.rowSums(x)), 
                  order(as.numeric(colnames(x))), drop = 0]
                hs1 <- function(x) apply(x, 2, function(x) hs.sort(x) %=>% 
                  (x/sum(x)) %=>% (x[1] - x[2])) %=>% median
                c1 <- m1 %=>% apply(x, 2, function(x) {
                  x <- hs.sort(x) %=>% {
                    x/sum(x)
                  }
                  ct2 <- names(x[1])
                  p1 <- hs.sort(unlist(hs.nonempty(lapply(ct1.ct2, 
                    function(ct2) hs.clean(max(x[ct2])))))) %=>% 
                    {
                      x[1] - x[2]
                    }
                  ct2 <- ct2.ct2[[ct2]]
                  if (length(ct2) == 1) 
                    return(x[1])
                  p2 <- x[1] - x[2]
                  p1 * (1 + p2)/2
                }) %=>% median
                c(c1, hs1(m1 %=>% t(x[hs.rowSums(x) >= 30, , 
                  drop = 0])))
            }
        }) %=>% t
        .pdf({
            plot(c(1, nrow(a)), c(0, 1), type = "n", xlab = "resolution index (high -> low)", 
                ylab = "confidence")
            points(a[, 1], type = "l", col = "black")
            points(a[, 2], type = "l", col = "blue")
            points(hs.rowMeans(a), type = "l", col = "red")
            abline(v = which.max(hs.rowMeans(a)), col = "red", 
                lty = 2)
            legend("bottomleft", legend = c("cluster", "cell type", 
                "combined"), pch = "-", col = c("black", "blue", 
                "red"), bty = "n")
        })
        hs.browser("38.12.16.174047", debug = 0)
        .s(a)
        hs.rowMeans(a) %=>% which(x > median(x)) %=>% lapply(x, 
            . %=>% {
                clustering <- clusterings[, x]
                id <- hs.38_07_19_104409(clustering, cell_anno_best, 
                  certainty_score_relative, classify = 1)
                uniqueN(id)
            })
        i <- which.max(hs.rowMeans(a))
        clustering <- clusterings[, i]
        id <- hs.38_07_19_104409(clustering, cell_anno_best, 
            certainty_score_relative, classify = 1)
        clustering %=>% hs.table %=>% hs.c(id[names(x)], x) %=>% 
            hs.sort
        list(cluster_cn = colnames(clusterings)[i], cell_type = a %=>% 
            hs.str_diversify)
    }
})
"38_11_30_144958" <- function(seurat_data, wjj2, pca_stdev_mad_n = 1, 
    pc_n = {
    }, sample_cn = {
    }, org1 = "human", redo = {
    }) {
    if (.wk("type fast_tsne")) 
        stop("[2023_07_04_173431] 用conda安装fit-sne，并软链接fast_tsne到~/.local/bin")
    on.exit(hs.gc_here(), add = TRUE)
    redo %<=>% (x %and% names(seurat_data@reductions))
    "pca" %=>% for (r1 in x) {
        if (r1 %in% redo) 
            hs.fp(wjj2, paste0("reduction_", r1, ".rd")) %=>% 
                hs.wj.rm
    }
    seurat_data %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
        normalization.method = "log", pca_stdev_mad_n = pca_stdev_mad_n, 
        pc_n = pc_n, wjj2 = wjj2, org1 = org1)
    wjj2_non_linear <- hs.fp(wjj2, paste0("pc_n=", ncol(seurat_data@reductions[["pca"]])))
    seurat_data %<=>% hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")(x, 
        sample_cn = sample_cn, redo = redo, dim_n = 10, wjj2 = wjj2_non_linear, 
        pca_wj = hs.fp(wjj2, "reduction_pca.rd"))
    return(seurat_data)
    if (length(sample_cn)) {
        wj1 <- hs.fp(wjj2_non_linear, "reduction_harmony.rd")
        if (file.exists(wj1)) {
            a <- hs.load(wj1)
            if (ncol(a) != ncol(seurat_data@reductions[["pca"]])) {
                redo <- c("harmony", "umap", "tsne")
            }
            else seurat_data@reductions[["harmony"]] <- a
        }
        else redo <- c("harmony", "umap", "tsne")
        if (file.exists(wj1)) {
            redo <- c("umap", "tsne") %=>% {
                wj <- hs.fp(wjj2_non_linear, paste0("reduction_", 
                  x, ".rd"))
                i <- file.mtime(wj) > file.mtime(wj1)
                if (any(i)) 
                  x <- x[-which(i)]
                redo %or% x
            }
        }
    }
    for (r1 in c("umap", "tsne")) {
        wj1 <- hs.fp(wjj2_non_linear, paste0("reduction_", r1, 
            ".rd"))
        if (r1 %in% redo) 
            hs.wj.rm(wj1)
        if (file.exists(wj1)) {
            seurat_data@reductions[[r1]] <- hs.load(wj1)
        }
        else {
            seurat_data %<=>% hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")(x, 
                sample_cn = sample_cn, redo = r1, dim_n = 10, 
                wjj2 = wjj2)
            seurat_data@reductions[[r1]] %=>% save(x, file = hs.wj(wj1))
        }
    }
    return(seurat_data)
    wj2 <- hs.strsplit.v("meta.features", ",") %=>% {
        hs.c(x, hs.fp(wjj2, paste0(x, ".", ifelse(x %in% "reductions", 
            "rd", "tsv.gz"))))
    }
    if (all(file.exists(wj2))) {
        if (!sum(dim(seurat_data@assays[["RNA"]]@data))) {
            hs.browser("38.11.30.145726", debug = 0)
        }
        reduction_wj <- hs.wjj.ls(wjj2, "(?i)^reduction_.*\\.rd$", 
            nm = 1)
        reduction_wj %=>% {
            for (n1 in names(x)) {
                if (n1 %in% names(seurat_data@reductions)) 
                  next
                seurat_data@reductions[[n1]] <<- hs.load(x[n1])
            }
        }
        seurat_data@reductions <- hs.load(wj2["reductions"])
        {
            a <- wk.fread(wj2["meta.features"])
            if (nrow(a) < nrow(seurat_data)) 
                hs.browser("38.12.15.150832", debug = 0)
            if (nrow(a) > nrow(seurat_data)) {
                hs.browser("39_01_17_114627", debug = 0)
                match(rownames(a))
                seurat_data %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
                  normalization.method = "log")
                seurat_data %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
                  normalization.method = "log")
            }
            seurat_data@assays[["RNA"]]@meta.features <- wk.dt.2df(a, 
                rn.cn = 1)
            seurat_data@assays[["RNA"]]@var.features <- a[, row_name[vst.variable]]
        }
        if (0) {
            seurat_data@meta.data[1, ]
            local({
                SeuratObject::Idents(seurat_data) <- "chain_pairing"
            })
        }
    }
    else {
        if (0) {
            seurat_data <- hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")(seurat_data, 
                sample_cn = sample_cn, redo = "umap")
            p1 <- hs.hsgly("单细胞/转录组/seurat/函数/图/低维.38_12_11_113923")(seurat_data, 
                id = "chain_pairing")
            .png(print(p1), width = 30)
            "/dev/shm/wengkai/tmp.png" %>% wk.rsync.in
        }
        seurat_data %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
            normalization.method = "log", redo = 1)
        seurat_data@assays[["RNA"]]@meta.features %=>% wk.dt %=>% 
            wk.dt.fwrite(x, wj2["meta.features"])
        seurat_data@reductions[["pca"]] %=>% save(x, file = hs.fp(wjj2, 
            "reduction_pca.rd"))
        seurat_data %<=>% hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")(x, 
            sample_cn = sample_cn, redo = 1)
        seurat_data@reductions %=>% {
            for (n1 in names(x)) {
                wj2 <- hs.fp(wjj2, paste0("reduction_", n1, ".rd"))
                if (file.exists(wj2)) 
                  next
                x[[n1]] %=>% save(x, file = wj2)
            }
        }
    }
    seurat_data
}
"38_11_22_105337" <- function(seurat_data, nfeatures = "auto", 
    org1 = "human", gene_id_type = "ensembl", gene_to_exclude_type = "sex") {
    seurat_data %<=>% Seurat::FindVariableFeatures(x, selection.method = "vst", 
        nfeatures = 2000)
    gene2exclude <- hs.hsgly("不考虑的.38_12_05_100105")(org1 = org1, 
        gene_id_type = gene_id_type, gene_to_exclude_type = gene_to_exclude_type)
    if (nfeatures == "auto") 
        nfeatures <- seurat_data@assays[["RNA"]]@meta.features %=>% 
            x[rownames(x) %not.in% gene2exclude, "vst.variance.standardized"] %=>% 
            {
                co <- stats::median(x) + stats::mad(x) * 2
                base::sum(x >= co)
            }
    Seurat::FindVariableFeatures(seurat_data, selection.method = "vst", 
        nfeatures = nfeatures)
}
"38_12_05_100105" <- function(org1 = "human", gene_to_exclude_type = "sex", 
    gene_id_type = "ensembl") {
    c(if ("sex" %in% gene_to_exclude_type) hs.hsgly("性别相关基因/38_12_05_100436")(org1 = org1, 
        gene_id_type = gene_id_type))
}
"38_10_09_121231" <- function(seurat_data, pc_n = hs.hsgly("转录组/标准流程/模块/预处理/降维，线性.维度.38_10_09_121529")(seurat_data), 
    reduction = (c("harmony", "pca") %and% names(seurat_data@reductions))[1], 
    dim_n = 10, sample_cn = {
    }, tsne = 1, redo = 0, pca_wj = hs.fp(pca_wjj, "reduction_pca.rd"), 
    pca_wjj = wjj2, wjj2 = hs.fp(pca_wjj, paste0("pc_n=", ncol(seurat_data@reductions[["pca"]]))), 
    ot = "seu") {
    if (ot == "wjj") 
        return(wjj2)
    if (tsne && .wk("type fast_tsne")) 
        stop("[2023_07_04_173122] 用conda安装fit-sne，并软链接fast_tsne到~/.local/bin")
    wj_up <- pca_wj
    if (reduction == "harmony") {
        wj2 <- hs.fp(wjj2, "reduction_harmony.rd")
        if (wj.update.reporter(wj2, pca_wj)) {
            hs.require("harmony")
            seurat_data %<=>% hs.hsgly("seurat.scale.38_12_20_164405")(x)
            hs.msg("harmony")
            for (a in sample_cn) seurat_data@meta.data[, a] %<=>% 
                hs.hsgly("seurat/函数/idents，获取.38_12_15_114046")(id = x)
            seurat_data %<=>% harmony::RunHarmony(x, group.by.vars = sample_cn, 
                plot_convergence = TRUE, assay.use = "RNA")
            seurat_data@reductions[["harmony"]] %=>% hs.save(x, 
                wj = wj2)
            hs.msg()
        }
        else seurat_data@reductions[["harmony"]] <- hs.load(wj2)
        wj_up <- wj2
    }
    for (c1 in c(dim_n, 2)) {
        wj2 <- hs.fp(wjj2, paste0(reduction, "_umap_c=", c1, 
            ".rd"))
        if (wj.update.reporter(wj2, wj_up)) {
            hs.msg(paste0(c1, "维umap..."))
            seurat_data %<=>% (Seurat::RunUMAP)(x, reduction = reduction, 
                dims = 1:pc_n, n.components = c1, seed.use = 2031)
            seurat_data@reductions[["umap"]] %=>% hs.save(wj = wj2)
            hs.msg()
        }
        else if (c1 == 2) 
            seurat_data@reductions[["umap"]] <- hs.load(wj2)
    }
    if (tsne) {
        wj2 <- hs.fp(wjj2, paste0(reduction, "_tsne.rd"))
        if (wj.update.reporter(wj2, wj_up)) {
            hs.msg("tsne")
            seurat_data %<=>% Seurat::RunTSNE(x, reduction = reduction, 
                dims = 1:pc_n, seed.use = 203163417, tsne.method = "FIt-SNE")
            seurat_data@reductions[["tsne"]] %=>% hs.save(wj = wj2)
            hs.msg()
        }
        else seurat_data@reductions[["tsne"]] <- hs.load(wj2)
    }
    seurat_data
}
"38_10_09_120747" <- local({
    hs.select_pc <- function(seurat_data, mad_n = 1, pc_n = {
    }, wjj2 = {
    }) {
        if (!length(pc_n)) 
            pc_n <- hs.hsgly("转录组/标准流程/模块/预处理/降维，线性.维度.38_10_09_121529")(seurat_data = seurat_data, 
                mad_n = mad_n, wjj2 = wjj2)
        {
            seurat_data@reductions[["pca"]]@cell.embeddings %<=>% 
                {
                  x[, 1:pc_n, drop = 0]
                }
            seurat_data@reductions[["pca"]]@feature.loadings %<=>% 
                {
                  x[, 1:pc_n, drop = 0]
                }
            seurat_data@reductions[["pca"]]@stdev %<=>% {
                x[1:pc_n]
            }
        }
        seurat_data
    }
    function(seurat_data, normalization.method = "log", vars.to.regress = {
    }, features = {
    }, nfeatures = 2000, redo = 0, npcs = 200, pca_stdev_mad_n = 1, 
        pc_n = {
        }, org1 = "human", gene_id_type = "ensembl", var_gene_to_exclude_type = c("sex"), 
        wjj2 = {
        }, to = "") {
        flag_save <- 1
        if (length(wjj2)) {
            rd_wj <- hs.fp(wjj2, "reduction_pca.rd")
            if (redo) 
                hs.wj.rm(rd_wj)
            if (file.exists(rd_wj)) {
                seurat_data@reductions[["pca"]] <- hs.load(rd_wj)
                flag_save <- 0
            }
        }
        if (redo) 
            seurat_data@reductions[["pca"]] <- {
            }
        if (redo || (seurat_data[["RNA"]]@data[, 2] %=>% sum %=>% 
            (x == 0)) || (seurat_data %=>% {
            !all(dim(x@assays[["RNA"]]@data) == dim(x))
        }) || (seurat_data %=>% (max(hs.colMaxs(x@assays[["RNA"]]@data[, 
            head(hs.order(x@meta.data[, "nCount_RNA"]), min(100, 
                ncol(x)))])) > 20))) 
            seurat_data %<=>% hs.hsgly("转录组/标准流程/模块/预处理/seurat.normalize.38_09_23_134026")(x, 
                method = normalization.method, nfeatures = nfeatures, 
                vars.to.regress = vars.to.regress)
        if (!length(features)) {
            wj1 <- if (length(wjj2)) 
                hs.fp(wjj2, "meta.features.tsv.gz")
            if (wj1 %=>% {
                !(length(x) && file.exists(x))
            }) 
                seurat_data %<=>% hs.hsgly("标准流程/模块/预处理/高变基因/38_11_22_105337")(x, 
                  org1 = org1, gene_id_type = gene_id_type, gene_to_exclude_type = var_gene_to_exclude_type)
            if (length(wj1)) {
                if (file.exists(wj1)) {
                  a <- wk.fread(wj1)
                  if (nrow(a) > nrow(seurat_data)) 
                    hs.browser("38.12.15.150832", debug = 0)
                  if (nrow(a) < nrow(seurat_data)) {
                    hs.browser("38.12.15.150538", debug = 0)
                    match(rownames(a))
                    seurat_data %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
                      normalization.method = "log")
                  }
                  seurat_data@assays[["RNA"]]@meta.features <- wk.dt.2df(a, 
                    rn.cn = 1)
                  seurat_data@assays[["RNA"]]@var.features <- a[, 
                    feature[vst.variable]]
                }
                else {
                  seurat_data@assays[["RNA"]]@meta.features %=>% 
                    wk.dt.as("feature") %=>% wk.dt.fwrite(wj1)
                }
            }
            features <- SeuratObject::VariableFeatures(seurat_data)
        }
        if (to == "hvg") 
            return(seurat_data)
        if (!length(seurat_data@reductions[["pca"]])) {
            if (!length(features)) {
                if (redo || (!length(SeuratObject::VariableFeatures(seurat_data)))) {
                  wj1 <- hs.fp(wjj2, "meta.features.tsv.gz")
                  if (file.exists(wj1)) {
                    a <- wk.fread(wj1)
                    if (nrow(a) > nrow(seurat_data)) 
                      hs.browser("38.12.15.150832", debug = 0)
                    if (nrow(a) < nrow(seurat_data)) {
                      hs.browser("39_01_17_114617", debug = 0)
                      match(rownames(a))
                      seurat_data %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
                        normalization.method = "log")
                    }
                    seurat_data@assays[["RNA"]]@meta.features <- wk.dt.2df(a, 
                      rn.cn = 1)
                    seurat_data@assays[["RNA"]]@var.features <- a[, 
                      row_name[vst.variable]]
                  }
                  else {
                    seurat_data %<=>% hs.hsgly("标准流程/模块/预处理/高变基因/38_11_22_105337")(x, 
                      org1 = org1, gene_id_type = gene_id_type, 
                      gene_to_exclude_type = var_gene_to_exclude_type)
                  }
                  features <- SeuratObject::VariableFeatures(seurat_data)
                }
            }
            seurat_data %<=>% hs.hsgly("seurat.scale.38_12_20_164405")(x, 
                vars.to.regress = vars.to.regress, features = features)
            seurat_data %<=>% (Seurat::RunPCA)(x, npcs = npcs, 
                features = features, seed.use = 3417, verbose = 0)
        }
        if (length(wjj2) && flag_save) 
            seurat_data@reductions[["pca"]] %=>% hs.save(wj = hs.wj(rd_wj))
        return(hs.select_pc(seurat_data, mad_n = pca_stdev_mad_n, 
            pc_n = pc_n, wjj2 = wjj2))
    }
})
"38_10_09_121529" <- function(seurat_data, n2try = 200, mad_n = 1, 
    wjj2 = {
    }) {
    if ("pca" %not.in% names(seurat_data@reductions)) 
        stop("[2022_03_01_145151]先做线性降维")
    a <- ncol(seurat_data@reductions[["pca"]])
    if (a < n2try) 
        return(a)
    l1 <- SeuratObject::Stdev(seurat_data)
    {
        d <- ieiwk.density(l1)
        mode <- d[["x"]][which.max(d[["y"]])]
        co1 <- mode %=>% (mad(l1[l1 >= x], x) * mad_n + x)
        i <- l1 >= co1
    }
    jg <- max(which(i))
    if (length(wjj2)) {
        hs.fp(wjj2, "pc_n_decision.pdf") %=>% {
            if (!file.exists(x)) 
                .pdf({
                  co1 <- mode + mad(a[a >= mode], mode)
                  co1 <- mode %=>% (mad(a[a >= x], x) + x)
                  which(a >= co1)
                  plot(d)
                  abline(v = c(median(l1), co1), col = "red", 
                    lty = 2)
                  plot(l1, type = "b")
                  abline(v = jg, col = "red", lty = 2)
                }, x)
        }
    }
    jg
}
"38_10_09_113753" <- function(seurat_data_list, wj2, wjj2, FindIntegrationAnchors.reduction = "harmony", 
    nfeatures = 2000, vars.to.regress = {
    }, sample.cn = "sample_name.38_01_19_171144", add.cell.ids = {
    }, redo = 0) {
    if (file.exists(wj2)) {
        seurat_data <- hs.load(wj2)
        sample_n <- seurat_data@meta.data[, sample.cn] %=>% uniqueN
    }
    else {
        sample_n <- length(seurat_data_list)
        if (sample_n > 1) {
            if (FindIntegrationAnchors.reduction == "harmony") {
                hs.require("harmony")
                hs.msg("merge", 0)
                seurat_data <- hs.hsgly("合并数据.38_12_09_141645")(seurat_data_list)
                hs.msg("单位/业务/单细胞/模块/降维，线性.38_10_09_120747", 
                  0)
                seurat_data %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
                  normalization.method = "log", nfeatures = nfeatures, 
                  vars.to.regress = vars.to.regress, redo = 1)
                seurat_data@meta.data[, sample.cn] <- seurat_data_list %=>% 
                  rep(names(x), sapply(x, ncol))
                hs.msg("harmony::RunHarmony", 0)
                seurat_data <- harmony::RunHarmony(seurat_data, 
                  group.by.vars = sample.cn, plot_convergence = TRUE, 
                  assay.use = "RNA")
            }
            else {
                hs.browser("38.09.02.112002")
                features <- (Seurat::SelectIntegrationFeatures)(object.list = seurat_data_list, 
                  nfeatures = Args[["nfeatures"]], assay = rep(Args[["assay2use"]], 
                    length(seurat_data_list)))
                if (FindIntegrationAnchors.reduction == "rpca") 
                  for (i in seq_along(seurat_data_list)) seurat_data_list[[i]] %<>% 
                    {
                      if (normalization.method == "log") 
                        . <- Seurat::ScaleData(., features = features, 
                          verbose = 0)
                      Seurat::RunPCA(., assay = Args[["assay2use"]], 
                        features = features, verbose = 0)
                    }
                if (normalization.method == "sct") 
                  seurat_data_list <- Seurat::PrepSCTIntegration(object.list = seurat_data_list, 
                    assay = "SCT", anchor.features = features)
                sce_anchors <- (Seurat::FindIntegrationAnchors)(object.list = seurat_data_list, 
                  assay = rep(Args[["assay2use"]], length(seurat_data_list)), 
                  anchor.features = features, reduction = FindIntegrationAnchors.reduction, 
                  k.anchor = k.anchor, normalization.method = switch(normalization.method, 
                    log = "LogNormalize", sct = "SCT"))
                seurat_data <- (Seurat::IntegrateData)(anchorset = sce_anchors, 
                  normalization.method = switch(normalization.method, 
                    log = "LogNormalize", sct = "SCT"))
                seurat_data <- seurat.normalize(seurat_data, 
                  nfeatures = Args[["nfeatures"]], vars.to.regress = vars.to.regress)
            }
        }
        else {
            seurat_data <- seurat_data_list[[1]]
            seurat_data@meta.data[, sample.cn] <- names(seurat_data_list)
            seurat_data <- Seurat::FindVariableFeatures(seurat_data, 
                nfeatures = nfeatures, selection.method = "vst")
            seurat_data %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
                normalization.method = "log", nfeatures = nfeatures, 
                vars.to.regress = vars.to.regress, redo = 1)
        }
        hs.msg("单位/业务/单细胞/模块/降维，非线性.38_10_09_121231", 
            0)
        seurat_data %<=>% hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")(x, 
            reduction = (c("harmony", "pca") %and% names(x@reductions))[1], 
            redo = redo)
        hs.msg("hs.save( seurat_data, wj = wj2)", 0)
        hs.save(seurat_data, wj = wj2)
        hs.msg()
        seurat_data
    }
    1
    {
        wj2 <- if (sample_n > 1) 
            "all_samples"
        else seurat_data@meta.data[1, sample.cn]
        wj2 <- wjj2 %=>% hs.fp("umi_gene.n", wj2)
        seurat_data %<=>% hs.hsgly("转录组/标准流程/模块/r函数_自制.38_04_07_003149", 
            "do.log10_umi_gene_calculation")(x, wj2 = wj2)
        if (sample_n > 1) {
            seurat_data@meta.data[, sample.cn] %=>% unique %=>% 
                sort %=>% wk.mclapply(x, . %=>% {
                wj2 <- hs.fp(wjj2, "umi_gene.n", x)
                hs.msg(wj2)
                cell <- colnames(seurat_data)[seurat_data@meta.data[, 
                  sample.cn] %in% x]
                hs.hsgly("转录组/标准流程/模块/r函数_自制.38_04_07_003149", 
                  "plot.umi_gene_n")(seurat_data, wj2, cell = cell)
            })
        }
    }
    if (sample_n > 1) 
        local({
            hs1 <- function(sce, wj2, cell = {
            }) {
                if (all(file.exists(paste0(wj2, c(".png", ".pdf"))))) 
                  return()
                reduction_method <- c("umap", "tsne")
                P <- lapply(seq_along(reduction_method), function(i1) {
                  reduction_method_1 <- reduction_method[i1]
                  p1 <- DimPlot(sce, reduction = reduction_method_1, 
                    group.by = sample.cn, cells = cell, raster = TRUE) + 
                    labs(title = switch(reduction_method_1, pca = "PCA", 
                      umap = "UMAP", tsne = "tSNE"))
                  if (i1 < length(reduction_method)) 
                    p1 <- p1 + ggplot2::theme(legend.position = "none")
                  p1
                })
                P <- do.call("+", P)
                .figs(print(P), wj2, width = 10 * length(reduction_method) + 
                  5, height = 10)
            }
            wj2 <- hs.fp(wjj2, "all_samples.png")
            if (!file.exists(wj2)) 
                hs1(seurat_data, wj2)
            if (sample_n > 2) {
                sn <- seurat_data@meta.data[, sample.cn] %=>% 
                  unique
                combn(sn %=>% sort, 2) %=>% {
                  wk.mclapply(seq(ncol(x)), function(ci1) {
                    s2 <- x[, ci1]
                    wj2 <- hs.fp(wjj2, paste(s2, collapse = "_and_")) %=>% 
                      paste0(".png")
                    if (!file.exists(wj2)) {
                      hs.msg(wj2)
                      cell <- seurat_data@meta.data %=>% rownames(x)[x[, 
                        sample.cn] %in% s2]
                      hs1(seurat_data, wj2, cell = cell)
                    }
                  })
                }
            }
        })
    seurat_data
}
"38_10_19_144214" <- function(seurat_data_list, feature_min_cell_n = 3, 
    org1 = "human", wjj2) {
    hs.hsgly("标准流程/模块/预处理/质控/过滤细胞、基因.图表.38_10_19_130420")(seurat_data_list, 
        wjj2 = wjj2, org1 = org1)
    seurat_data_list <- local({
        a <- wk.mclapply(seq_along(seurat_data_list), function(i1) {
            prj1 <- names(seurat_data_list)[i1]
            sce <- seurat_data_list[[prj1]]
            {
                sce <- hs.hsgly("标准流程/模块/预处理/质控/计算基因表达占比.38_10_19_134504")(sce, 
                  ot = "s", org1 = org1)
            }
            {
                a1 <- wk.dt()
                local({
                  wj2 <- hs.fp(wjj2, "before_QC", "gene_n_cutoff", 
                    paste0(prj1, ".pdf"))
                  V <- hs.colSums(sce@assays[["RNA"]]@counts > 
                    0)
                  if (0) {
                    summary(V)
                    colnames(sce_combined)[sce_combined@meta.data[, 
                      "sample_name.38_01_19_171144"] == "AEpl-2"]
                    hs.colSums(sce_combined@assays[["RNA"]]@counts[, 
                      colnames(sce_combined)[sce_combined@meta.data[, 
                        "sample_name.38_01_19_171144"] == "AEpl-2"]] > 
                      0) %>% summary
                    . <- ieiwk.density(V)
                    Vlog <- log10(V)
                    . <- ieiwk.density(Vlog)
                    .pdf({
                      i1 <- .[["y"]] %>% diff %>% sign %>% {
                        match(-1, .)
                      }
                      plot(.[["x"]], .[["y"]], type = "l")
                      x1 <- .[["x"]][i1]
                      abline(v = c(x1, x1 * 3), col = ieiwk.col2rgb("red", 
                        0.5))
                    })
                    .pdf({
                      i1 <- which.max(.[["y"]])
                      plot(.[["x"]], .[["y"]], type = "l")
                      x1 <- .[["x"]][i1]
                      a <- V[V > x1]
                      co2 <- min(a) + mad(a, min(a)) * 3
                      a <- V[V <= x1]
                      co1 <- max(a) - mad(a, max(a)) * 3
                      abline(v = c(co1, co2), col = ieiwk.col2rgb("red", 
                        0.5))
                    })
                    .pdf({
                      i1 <- which.max(.[["y"]])
                      plot(.[["x"]], .[["y"]], type = "l")
                      x1 <- .[["x"]][i1]
                      a <- Vlog[Vlog > x1]
                      co2 <- min(a) + mad(a, min(a)) * 3
                      a <- Vlog[Vlog <= x1]
                      co1 <- max(a) - mad(a, max(a)) * 3
                      abline(v = c(co1, co2), col = ieiwk.col2rgb("red", 
                        0.5))
                    })
                    "~/swxx/tmp.pdf" %>% wk.rsync.in(ssh = "r820")
                  }
                  if (min(V) > 1000) 
                    return()
                  x1 <- ieiwk.density(V) %>% {
                    if (0) 
                      .[["x"]][which.max(.[["y"]][.[["x"]] < 
                        500])]
                    i1 <- .[["y"]] %>% (diff) %>% (sign) %>% 
                      {
                        match(-1, .)
                      }
                    .[["x"]][i1]
                  }
                  co1 <- V[V <= x1] %>% {
                    mad(., max(.)) * 3 + x1
                  }
                  done <- 0
                  if ((co1/median(V)) > 0.5) {
                    Vlog <- log2(V)
                    . <- ieiwk.density(Vlog)
                    i1 <- which.max(.[["y"]])
                    x1 <- .[["x"]][i1]
                    a <- Vlog[Vlog > x1]
                    co2 <- min(a) + mad(a, min(a)) * 2
                    a <- Vlog[Vlog <= x1]
                    co1 <- max(a) - mad(a, max(a)) * 2
                    if (!file.exists(wj2)) 
                      .pdf({
                        plot(.[["x"]], .[["y"]], type = "l", 
                          xlab = expression(Log[2] * "(#gene)"), 
                          ylab = "Freq.", main = paste("Sample:", 
                            prj1))
                        abline(v = c(co1), col = "blue")
                        y1 <- .[["y"]] %>% {
                          max(.) * 0.95
                        }
                        text(co1, y1, pos = 4, labels = ceiling(2^co1), 
                          col = "blue")
                        abline(v = co2, col = "red")
                        text(co2, y1, pos = 2, labels = floor(2^co2), 
                          col = "red")
                      }, wj2)
                    co1 <- 2^co1
                    co2 <- 2^co2
                    i <- V < co1
                    a1[, `:=`("too few\ngenes", i)]
                    i <- V > co2
                    a1[, `:=`("too\nmany\ngenes", i)]
                    done <- 1
                  }
                  if (co1 > 500) 
                    warning("[2022_03_08_104957]样本", prj1, 
                      "的阳性基因数下限（", ceiling(co1), 
                      "）超过了500")
                  if (done) 
                    return()
                  i <- V < co1
                  a1[, `:=`("too few\ngenes", i)]
                  if (0) {
                    .pdf(plot(ieiwk.density(V)))
                    .pdf(hist(V, breaks = "FD"))
                  }
                  co2 <- hs.hsgly("离群者/mad.38_03_08_101528")(V, 
                    ot = "co")[2]
                  i <- V > co2
                  a1[, `:=`("too\nmany\ngenes", i)]
                  if (!file.exists(wj2)) 
                    .pdf({
                      . <- ieiwk.density(V)
                      plot(.[["x"]], .[["y"]], type = "l", xlab = "#gene", 
                        ylab = "Freq.", main = paste("Sample:", 
                          prj1))
                      abline(v = c(x1, co1), col = ieiwk.col2rgb("blue", 
                        c(0.5, 1)))
                      y1 <- .[["y"]] %>% {
                        max(.) * 0.95
                      }
                      text(co1, y1, pos = 4, labels = ceiling(co1), 
                        col = "blue")
                      abline(v = co2, col = "red")
                      text(co2, y1, pos = 2, labels = floor(co2), 
                        col = "red")
                    }, wj = wj2)
                })
                {
                  if ("percent.mt" %in% colnames(sce@meta.data)) {
                    i <- sce@meta.data[, "percent.mt"] %=>% {
                      if (median(x) == 0) {
                        co1 <- mad(x[x > 0], 0) * 3
                        x > co1
                      }
                      else hs.hsgly("离群者/mad.38_03_08_101528")(x, 
                        ot = "l")
                    }
                    if (any(i)) 
                      message("\t去掉线粒体基因含量过高（>", 
                        hs.hsgly("离群者/mad.38_03_08_101528")(sce@meta.data[, 
                          "percent.mt"], ot = "co")[2] %>% (round)(2), 
                        "%）的", sum(i), "个细胞。")
                    a1[, `:=`("too much\nmt%", i)]
                  }
                  else message("[2022_05_30_212150]竟然没有线粒体基因！")
                }
                if ("percent.hemo" %in% colnames(sce@meta.data)) {
                  i <- sce@meta.data[, "percent.hemo"] %=>% {
                    if (median(x) == 0) {
                      co1 <- mad(x[x > 0], 0) * 3
                      x > co1
                    }
                    else hs.hsgly("离群者/mad.38_03_08_101528")(x, 
                      ot = "l")
                  }
                  if (any(i)) 
                    message("\t去掉红细胞基因含量过高（>", 
                      hs.hsgly("离群者/mad.38_03_08_101528")(sce@meta.data[, 
                        "percent.hemo"], lo = 0, ot = "co")[2] %>% 
                        (round)(2), "%）的", sum(i), "个细胞。")
                  a1[, `:=`("too much\nhemo%", i)]
                }
                i <- which(!hs.colAnys(a1))
                a1[, `:=`(c(i), {
                })]
                if (ncol(a1) < 2) 
                  hs.browser("38.05.16.234550")
                wj2 <- hs.fp(wjj2, "after_QC/how_excluded", hs.p0(prj1))
                {
                  L <- apply(a1, 2, which)
                  if (!file.exists(paste0(wj2, ".pdf"))) 
                    .figs(grid.draw(VennDiagram::venn.diagram(L, 
                      {
                      }, margin = 0.2, cex = 0.5, units = "cm", 
                      width = 10, height = 5)), wj = wj2, width = 10 + 
                      1, height = 10 + 1)
                }
                i <- hs.rowAnys(a1)
                if (any(i)) {
                  message("\t共", sum(i), "个(", round(sum(i)/ncol(sce) * 
                    100, 2), "%）细胞被认为不合格。")
                  sce <- sce[, !i]
                }
            }
            {
                i <- hs.rowSums(sce@assays[["RNA"]]@counts > 
                  0) < feature_min_cell_n
                if (any(i)) {
                  message("\t样本", prj1, "：删除了", sum(i), 
                    "个阳性细胞不足", feature_min_cell_n, 
                    "个的基因。")
                  sce <- sce[!i, ]
                }
            }
            sce
        }, mc.cores = 4)
        names(a) <- names(seurat_data_list)
        a
    })
    hs.hsgly("标准流程/模块/预处理/质控/过滤细胞、基因.图表.38_10_19_130420")(seurat_data_list, 
        wjj2 = wjj2, org1 = org1, stage = "after")
    local({
        wj2 <- wjj2 %=>% hs.fp("metrics.tsv.gz")
        a <- hs.fp(wjj2, "before_QC/metrics/all_samples.tsv.gz") %=>% 
            wk.fread
        hs1 <- function(dt, post = "") {
            cn2do <- tail(names(dt), -2)
            a <- dt[, c(.N, as.list(hs.colMeans(.SD))), by = sample, 
                .SDcols = c(cn2do)]
            dt[[4]]
            wk.dt.setnames(a, seq(2, length(a)), c("#cell", paste0("Mean_", 
                cn2do, post)))
        }
        qc_before <- hs1(a)
        hs.xor(a[, barcode], seurat_data_list %=>% lapply(colnames) %=>% 
            unlist)
        if (seurat_data_list %=>% lapply(colnames) %=>% unlist %=>% 
            {
                x %all.in% a[, barcode]
            }) {
            a1 <- a[, .SD[barcode %in% colnames(seurat_data_list[[sample]])], 
                by = sample]
        }
        else {
            hs.browser("2023.06.30.151838", debug = 0)
            sample.i <- sce_combined@meta.data %=>% {
                a <- x[, "sample_name.38_01_19_171144"]
                b <- unique(a)
                i <- match(b, a)
                hs.c(b, rownames(x)[i] %>% hs.re("\\d+$"))
            }
            a1 <- a[paste0(barcode, "_", sample.i[sample]) %in% 
                colnames(sce_combined)]
        }
        qc_after <- hs1(a1)
        jg <- merge(qc_before, qc_after, by = "sample", suffixes = paste0("_", 
            c("before", "after"), "_QC"))
        wk.dt.fwrite(jg, wj2)
    })
    seurat_data_list
}
"38_10_19_130420" <- function(sce_list, wjj2, org1, stage = "before") {
    hs.require("patchwork")
    wjj2 <- hs.fp(wjj2, switch(stage, before = "before_QC", after = "after_QC"), 
        "metrics")
    wj2 <- hs.fp(wjj2, c(paste0(names(sce_list), ".png"), "all_samples.tsv.gz"))
    if (all(file.exists(wj2))) 
        return(wj2)
    a1 <- wk.mclapply(sce_list, . %=>% {
        m1 <- x@assays[["RNA"]]@counts
        c(list(umi_n = hs.colSums(m1), gene_n = hs.colSums(m1 > 
            0)), hs.hsgly("标准流程/模块/预处理/质控/计算基因表达占比.38_10_19_134504")(x, 
            org1 = org1))
    }, mc.cores = 1)
    if (uniqueN(sapply(a1, length)) != 1) 
        hs.browser("38.07.12.130009")
    {
        fig_n_max <- max(sapply(a1, length))
        sapply(seq_along(a1), function(sample_i) {
            wj2 <- hs.fp(wjj2, paste0(names(a1)[sample_i], ".png"))
            if (!file.exists(wj2)) {
                P <- hs.lapply(names(a1[[sample_i]]), function(n1) {
                  dt1 <- wk.dt.fromList(a1[[sample_i]][n1])
                  p <- hs.hsgly("小提琴图/36.06.18.150541")(dt1, 
                    main = "", xlab = "", ylab = switch(n1, gene_n = "#gene", 
                      umi_n = "#UMI", percent.mt = "UMI percentage of\nmitochondrial genes", 
                      percent.ribo = "UMI percentage of\nribosomal genes", 
                      percent.hemo = "UMI percentage of\nhemoglobin subunits"), 
                    N = 0) + ggplot2::scale_x_discrete(breaks = NULL) + 
                    ggplot2::labs(title = switch(n1, gene_n = "Gene", 
                      umi_n = "UMI", percent.mt = "Mitochondrial\ngenes", 
                      percent.ribo = "Ribosomal\ngenes", percent.hemo = "Hemoglobin\nsubunits"))
                })
                P <- eval(bquote(.(parse(text = paste0("P[[ ", 
                  1:length(P), "]]") %=>% paste(collapse = " + "))[[1]]) + 
                  patchwork::plot_layout(nrow = 1)))
                .figs(print(P), wj2, width = 5 * fig_n_max, height = 5)
            }
        })
        if (0) {
            p1 <- paste0("P[[ ", seq_along(P), "]]") %>% paste(collapse = " | ") %>% 
                paste0(" + plot_layout( ncol = 1)") %>% evalStr
            .pdf(print(p1), hs.wjm(wj2, ext = "pdf"), width = 10 * 
                fig_n_max)
            .pdf(print(P[[1]] + P[[2]] + patchwork::plot_layout(ncol = 1)), 
                hs.wjm(wj2, ext = "pdf"), width = 10 * fig_n_max, 
                height = 10 * 2)
            wk.ggsave(P[[1]] + P[[2]] + plot_layout(ncol = 1), 
                wj2, width = 10 * fig_n_max, height = 10 * length(P), 
                limitsize = FALSE)
            wk.ggsave(p1, wj2, width = 50 * fig_n_max, height = 50 * 
                length(P), limitsize = FALSE)
        }
        {
        }
    }
    {
        sapply(c("percent.mt", "percent.hemo"), function(nm1) {
            wj2 <- hs.fp(wjj2, paste0(nm1, ".pdf"))
            if (file.exists(wj2)) 
                return()
            if (0) {
                a1[[1]][["percent.mt"]] %>% summary
                a1[[1]][["percent.hemo"]] %>% summary
            }
            sj1 <- lapply(a1, function(a1) {
                a1 <- a1[[nm1]]
                if (length(a1)) {
                  sapply(seq(5, 20, 5), function(co1) {
                    mean(a1 > co1) * 100
                  })
                }
            }) %=>% hs.nonempty
            if (length(sj1)) {
                m1 <- sapply(sj1, round, digits = 2) %=>% t
                colnames(m1) <- sprintf(">%s%%", seq(0.05, 0.2, 
                  0.05) * 100)
                hs.hsgly("table2pdf.37_12_27_141843")(m1, wj2, 
                  rows = rownames(m1), pdf.args = list(width = 15))
            }
        })
    }
    {
        wj2 <- hs.fp(wjj2, "all_samples.tsv.gz")
        dt <- lapply(sce_list, colnames) %=>% wk.dt.fromList(nm = c("sample", 
            "barcode"))
        wk.dt.col.reorder(dt, "barcode", after = 0)
        hs.for(cn1, names(a1[[1]]), wk.dt.set(dt, , cn1, lapply(a1, 
            "[[", cn1) %=>% unlist))
        wk.dt.fwrite(dt, wj2)
    }
}
"38_10_19_134504" <- function(sce, percent2do = c("mt", "ribo", 
    "hemo"), percent_list = {
}, org1 = "human", ot = "p") {
    percent2do %<=>% (`%not%`(names(percent_list)))
    if (length(percent2do)) {
        total <- hs.colSums(sce@assays[["RNA"]]@counts)
        if (org1 %not.in% c("human", "mouse", "rat", "sheep", 
            "goat")) 
            stop("[2022_02_22_144727]目前只能处理人、小鼠，不能处理", 
                org1, "。")
        db1 <- hs.hsgly("基因/gene_info.37_12_31_133053")(org1) %=>% 
            wk.fread
        if (org1 %not.in% c("human", "mouse")) {
            gtf_wj <- hs.hsgly("cr/建参考基因组索引.38_04_13_090722")(org1 = org1, 
                ot = "gtf")
            gtf <- hs.hsgly("gtf/基因信息.38_09_08_204013")(gtf_wj)
            ens.des <- hs.hsgly("ensembl/基因/基于biomart/ens.描述.38_08_20_025839")(org1 = org1) %>% 
                wk.fread
            gtf.id_db <- hs.switch(org1, "rat", , "sheep", , 
                "goat", {
                  hs.hsgly("ensembl_gtf基因名缩写到ncbi基因名缩写.38_05_28_133333")(gtf_wj)
                })
            hs.38_05_30_174434 <- function(gene) {
                gene <- gtf.id_db[name_ncbi %in% gene, name %or% 
                  gene]
                hs.str_and(gene, rownames(sce))
            }
        }
        p1 <- "hemo"
        jg <- hs.lapply(percent2do, function(p1) {
            gene <- local(switch(p1, mt = local({
                hs.switch(org1, "human", , "mouse", hs.grepV(rownames(sce), 
                  "(?i)^mt-"), "rat", , "sheep", , "goat", local({
                  if (0) {
                    g <- db1[chromosome %in% "MT", ifelse(hs.clean(Symbol_from_nomenclature_authority, 
                      dirty.extra = "-", ot = "l"), Symbol_from_nomenclature_authority, 
                      Symbol)]
                    hs.38_05_30_174434(g)
                  }
                  hs.str_and(gtf[V1 %in% "MT", gene], rownames(sce))
                }))
            }), ribo = {
                hs.switch(org1, "human", , "mouse", hs.grepV(rownames(sce), 
                  "(?i)^rp[sl]"), "rat", , "sheep", , "goat", 
                  {
                    g <- wk.dt.subset(db1, exp_l = c(hs.grep(description, 
                      "^(?i)ribosomal protein")), exp_j = ifelse(hs.clean(Symbol_from_nomenclature_authority, 
                      dirty.extra = "-", ot = "l"), Symbol_from_nomenclature_authority, 
                      Symbol))
                    g <- hs.38_05_30_174434(g)
                    a <- wk.dt.subset(gtf, ens.des[hs.grep(`Gene description`, 
                      "^(?i)ribosomal protein"), `Gene stable ID`], 
                      "gene_id")[, gene] %and% rownames(sce)
                    g %or% a
                  })
            }, hemo = {
                hs.switch(org1, "human", c("HBA1", "HBA2", "HBB", 
                  "HBD", "HBE1", "HBG1", "HBG2", "HBM", "HBQ1", 
                  "HBZ"), "mouse", db1[hs.grep(description, "^(?i)hemoglobin")][type_of_gene %in% 
                  c("protein-coding"), ifelse(hs.clean(Symbol_from_nomenclature_authority, 
                  dirty.extra = "-", ot = "l"), Symbol_from_nomenclature_authority, 
                  Symbol)] %and% rownames(sce), "rat", , "sheep", 
                  , "goat", local({
                    g <- wk.dt.subset(db1, exp_l = c(hs.grep(description, 
                      "^(?i)hemoglobin"), type_of_gene %in% c("protein-coding")), 
                      exp_j = ifelse(hs.clean(Symbol_from_nomenclature_authority, 
                        dirty.extra = "-", ot = "l"), Symbol_from_nomenclature_authority, 
                        Symbol))
                    if (0) {
                      gtf.id_db[name_ncbi %in% g, unique(name)]
                      db1[Symbol %in% "Hbq1"]
                      gtf.id_db[id %in% "ENSRNOG00000070249"]
                      ens_id.ncbi_id[`gene_id,ensembl` %in% "ENSRNOG00000070249"]
                      gtf[id %in% "ENSRNOG00000070249"]
                      gtf.id_db[name_ncbi %in% "LOC120093065"]
                    }
                    g <- hs.38_05_30_174434(g)
                    a <- wk.dt.subset(gtf, ens.des[hs.grep(`Gene description`, 
                      "^(?i)hemoglobin"), `Gene stable ID`], 
                      "gene_id")[, gene] %>% hs.str_and(rownames(sce))
                    g %or% a
                  }))
            }))
            gene <- rownames(sce) %=>% x[tolower(x) %in% tolower(gene)]
            if (length(gene)) 
                hs.colSums(sce[["RNA"]]@counts[gene, ])/total * 
                  100
        })
    }
    jg <- c(percent_list, jg) %>% hs.nonempty
    names(jg) %<>% {
        paste0("percent.", .)
    }
    switch(ot, p = jg, s = {
        for (n1 in names(jg)) sce[[n1]] <- jg[[n1]]
        sce
    })
}
"38_12_23_115524" <- function(seurat_data, co_cell_n = 5, co_gene_n = 500) {
    ri <- seq(nrow(seurat_data))
    ci <- seq(ncol(seurat_data))
    flag_row <- 0
    flag_column <- 0
    repeat {
        ri1 <- hs.hsgly("去掉阳性细胞太少的基因.38_12_01_134746")(seurat_data, 
            co = co_cell_n, ot = "i", ri = ri, ci = ci)
        if (length(ri1) < length(ri)) {
            ri <- ri1
            flag_column <- 0
        }
        flag_row <- 1
        if (flag_row && flag_column) 
            break
        ci1 <- hs.hsgly("去掉阳性基因太少的细胞.38_12_23_005032")(seurat_data, 
            co = co_gene_n, ot = "i", ri = ri, ci = ci)
        if (length(ci1) < length(ci)) {
            ci <- ci1
            flag_row <- 0
        }
        flag_column <- 1
        if (flag_row && flag_column) 
            break
    }
    hs.matrix_ij(seurat_data, ri, ci)
}
"38_12_23_005032" <- function(seurat_data, co = 500, ri = {
}, ci = {
}, ot = "seu") {
    hs.hsgly("去掉阳性细胞太少的基因.38_12_01_134746")(seurat_data = seurat_data, 
        side = 2, co = co, ri = ri, ci = ci, ot = ot)
}
"38_12_01_134746" <- function(seurat_data, co = hs.switch(side, 
    1, 5, 2, 500), K = 1000, side = 1, ri = {
}, ci = {
}, ot = "seu") {
    m1 <- seurat_data@assays[["RNA"]]@counts
    if (!length(ri)) 
        ri <- seq(nrow(m1))
    if (!length(ci)) 
        ci <- seq(ncol(m1))
    hs.switch(side, 1, {
        a <- hs.makeBatch(ri, K = K) %=>% wk.mclapply(x, . %=>% 
            {
                hs.rowSums(m1[x, ci, drop = FALSE] > 0)
            }, mc.cores = 10)
        i <- ri[unlist(a) >= co]
        if (length(i) < length(ri)) {
            ri <- i
        }
    }, 2, {
        a <- hs.makeBatch(ci, K = K) %=>% wk.mclapply(x, . %=>% 
            {
                hs.colSums(m1[ri, x, drop = FALSE] > 0)
            }, mc.cores = 10)
        i <- ci[unlist(a) >= co]
        if (length(i) < length(ci)) {
            ci <- i
        }
    })
    hs.switch(ot, "seu", hs.switch(c(length(ri) < nrow(seurat_data), 
        length(ci) < ncol(seurat_data)), c(0, 0), seurat_data, 
        c(0, 1), seurat_data[, ci], c(1, 0), seurat_data[ri, 
            ], c(1, 1), seurat_data[ri, ci]), "i", i)
}
"38_12_06_175404" <- function(score, cell_group = {
}, homotypic.prop = if (length(cell_group)) DoubletFinder::modelHomotypic(cell_group) else 0.2, 
    cell = {
    }) {
    N <- length(score)
    doublet_rate <- hs.hsgly("细胞数到双胞率.38_12_05_180121")(N)
    nExp_poi <- round(doublet_rate * N * (1 - homotypic.prop))
    i <- hs.top(score, nExp_poi, ot = "i", end = "right")
    if (length(cell)) 
        cell[i]
    else i
}
"38_12_05_180121" <- function(cell_n) cell_n * 8e-06
"38_10_08_091811" <- function(fq_fp_list, wjj2, fn.sn = {
}, sample2exclude = {
}, sample2use = {
}) {
    if (!length(names(fq_fp_list))) {
        names(fq_fp_list) <- sapply(fq_fp_list, . %=>% {
            if ((length(x) == 1) && dir.exists(x)) {
                basename(x)
            }
            else basename(dirname(x[1]))
        })
    }
    if (!hs.is_empty(fn.sn)) {
        if (any(names(fq_fp_list) %not.in% fn.sn[[1]])) 
            fq_fp_list <- fq_fp_list[fn.sn[[1]]]
        if (!length(fq_fp_list)) 
            stop("[38_09_16_113938] 没有样本了")
        names(fq_fp_list) %<=>% wk.dt.subset(fn.sn, x)[[2]]
    }
    if (length(sample2exclude)) 
        fq_fp_list %<=>% x[names(x) %not.in% sample2exclude]
    if (length(sample2use)) 
        fq_fp_list %<=>% x[names(x) %in% sample2use]
    separate_sample <- hs.all(length(fq_fp_list) > 1, lapply(fq_fp_list, 
        . %=>% dirname %=>% unique) %=>% unlist %=>% uniqueN %=>% 
        (x < 2))
    wk.mclapply(seq_along(fq_fp_list), function(i1) {
        sample_name1 <- names(fq_fp_list)[i1]
        {
            if (separate_sample) 
                wjj2 <- hs.fp(wjj2, sample_name1)
            fq_wj <- fq_fp_list[[i1]]
            fq_wj_original <- if (hs.isDir(fq_wj, a = 1)) {
                hs.hsgly("wjj2fq.37_12_25_144226")(fq_wj, sample_wise = 0)
            }
            else fq_wj
            if (0) {
                is_local <- local({
                  a <- prj_fp_root %=>% normalizePath
                  n1 <- nchar(a)
                  fq_wj_original %=>% path.expand %=>% substr(1, 
                    n1) %=>% all(x == a)
                })
            }
            is_local <- 1
            fq_wj_2use <- if (is_local) {
                fq_wj_original
            }
            else hs.sub(fq_wj_original, "(?i).*/p\\d+/", hs.fp(wjj2, 
                "in/"))
            wj2 <- hs.wjm(fq_wj_2use, append = "fastqc", ext = "zip", 
                extN = 2, dn = wjj2)
            if (!all(file.exists(wj2))) {
                for (i2 in seq_along(fq_wj_original)) hs.wj.copy(fq_wj_original[i2], 
                  hs.wj(fq_wj_2use[i2]), overwrite = 0)
                hs.hsgly("fastqc.37_12_07_144326")(fq_wj_2use, 
                  wjj2 = wjj2)
            }
        }
        if (!is_local) 
            hs.wj.rm(fq_wj_2use)
    })
    fq_fp_list
}
"38_10_08_093432" <- function(wjj1, wjj2 = hs.fp(wjj1, "multiQC"), 
    wj = {
    }, redo = 0) {
    hs.with_wd(hs.wjj(wjj2), {
        wj2 <- "multiqc_report.html"
        if (redo) 
            hs.wj.rm(wj2)
        if (!file.exists(wj2)) {
            if (!length(wj)) 
                wj <- hs.wjj.ls(wjj1, "(?i)_r2(_\\d+)?.*\\.zip$")
            hs.hsgly("multiqc.37_12_08_165040")(wj = wj, args_extra = "-dd 1 --no-data-dir")
            a <- readLines("multiqc_report.html")
            i <- hs.grep(a, "mqc_analysis_path")
            m <- regexpr("(?<=\\>)[^<]+", a[i], perl = TRUE)
            regmatches(a[i], m) %<>% hs.basename(2)
            wk.dt.fwrite(a, "multiqc_report.html")
        }
        normalizePath(wj2)
    })
}
"38_09_23_134026" <- function(seurat_data, method = "log", nfeatures = 2000, 
    vars.to.regress = {
    }) {
    switch(method, log = {
        if (0) {
            getS3method("NormalizeData", "Seurat")
            getS3method("NormalizeData", "Assay")
            sce_list[[1]]@assays[["RNA"]]@counts[gene1, "AAAGAACCACAGTATC-1"]
        }
        Seurat::NormalizeData(seurat_data) %=>% Seurat::FindVariableFeatures(x, 
            selection.method = "vst", nfeatures = nfeatures)
    }, sct = {
        Seurat::SCTransform(seurat_data, variable.features.n = nfeatures, 
            method = "glmGamPoi", vars.to.regress = vars.to.regress, 
            verbose = 0)
    })
}
"38_12_20_164405" <- function(seurat_data, features = {
}, redo = 0, ...) {
    if (!length(features)) 
        features <- SeuratObject::VariableFeatures(seurat_data)
    if (redo || {
        seurat_data %=>% x@assays[[SeuratObject::DefaultAssay(x)]]@scale.data %=>% 
            {
                (!length(x)) || any(sort(rownames(x)) != sort(features))
            }
    }) {
        Seurat::ScaleData(seurat_data, features = features, verbose = TRUE, 
            assay = "RNA", ...)
    }
    else seurat_data
}
"38_10_12_230454" <- function(org1) {
    assign("wj.update.reporter.co1", 0, envir = .GlobalEnv)
    on.exit(assign("wj.update.reporter.co1", Inf, envir = .GlobalEnv), 
        add = TRUE)
    hs.hsgly("基因/gene_info.37_12_31_133053")(org1 = org1)
    repeat {
        a <- system.time(hs.hsgly("genes.understand.35.11.12.182333")(org1 = org1), 
            gcFirst = FALSE)
        if (a["elapsed"] < 1) 
            break
    }
}
"38_09_19_173807" <- function(wj2, env1 = parent.frame()) {
    if (file.exists(wj2)) 
        return()
    exp1 <- quote({
        a <- c("测序数据质控", "FastQC", .wk("fastqc --version", 
            intern = 1) %>% hs.re("[^\\s]*$") %>% trim("v"), 
            "参考基因组比对、获得表达量矩阵", 
            "Cell Ranger", hs.dirname(cr, 2) %>% basename %>% 
                hs.sub("cellranger-"), "细胞过滤、表达量标准化、降维、筛选标记基因、细胞周期分析", 
            "Seurat", paste(packageVersion("Seurat")), "细胞类型预测", 
            if (paste0(org1, ".", tissue_type) == "human.血") {
                c("SingleR", paste(packageVersion("SingleR")))
            } else {
            }, "细胞聚类、拟时序分析", "monocle3", 
            paste(packageVersion("monocle3")), "GO、KEGG功能分析", 
            "clusterProfiler", paste(packageVersion("clusterProfiler")), 
            "蛋白互作分析", "STRING", "11.5", if (cellphonedb) c("细胞通讯分析", 
                "CellPhoneDB", "3.1.0"), if (gsva) c("GSVA分析", 
                "GSVA", paste(packageVersion("GSVA"))))
        a <- matrix(a, byrow = 1, ncol = 3)
        colnames(a) <- c("任务", "工具/数据库", "版本")
        wk.dt.as(a)
    })
    a <- eval(exp1, envir = env1)
    wk.xls.write(a, wj2)
}
"39_05_16_095300" <- function(seurat_data, old_new) {
    i <- which(rownames(seurat_data@assays[["RNA"]]@counts) %in% 
        old_new[[1]])
    if (length(i)) {
        rownames(seurat_data@assays[["RNA"]]@counts)[i] %<=>% 
            {
                i <- match(x, old_new[[1]])
                old_new[[2]][i]
            }
    }
    i <- which(rownames(seurat_data@assays[["RNA"]]@data) %in% 
        old_new[[1]])
    if (length(i)) {
        rownames(seurat_data@assays[["RNA"]]@data)[i] %<=>% {
            i <- match(x, old_new[[1]])
            old_new[[2]][i]
        }
    }
    if (length(seurat_data@assays[["RNA"]]@scale.data)) {
        hs.browser("39.05.16.103404", debug = 0)
    }
    if (length(seurat_data@assays[["RNA"]]@var.features)) {
        hs.browser("39.05.16.104013", debug = 0)
    }
    if (seurat_data@assays[["RNA"]] %=>% attributes %=>% names %=>% 
        length(x %not% c("counts", "data", "scale.data", "assay.orig", 
            "var.features", "meta.features", "misc", "key", "class"))) {
        hs.browser("39.05.16.103935", debug = 0)
    }
    if (length(names(seurat_data@assays) %not% c("RNA"))) {
        hs.browser("39.05.16.104052", debug = 0)
    }
    seurat_data
}
"38_09_19_172814" <- function(wjj) {
    if (!file.exists(wjj)) 
        return()
    hs.with_wd(wjj, {
        wjj <- hs.wjj.ls(wjj, wjj = 1, F = 0) %>% hs.grepV("cell.ranger", 
            inv = 1)
        hs.msg("转tsv.gz为xlsx")
        message(wjj)
        wk.mclapply(seq_along(wjj), function(i) {
            message(i)
            wjj1 <- wjj[i]
            wj <- hs.wjj.ls(wjj1, "\\.tsv\\.gz$", R = 1)
            if (!length(wj)) 
                return()
            null <- wk.mclapply(seq_along(wj), function(j) {
                wj1 <- wj[j]
                wj2 <- hs.wjm(wj1, ext = "xlsx", extN = 2)
                if (file.exists(wj2)) 
                  return()
                message(wj1)
                a <- wk.fread(wj1, cn = 0)
                wk.xls.write(a, wj2, cn = 0)
            })
        }, mc.cores = 1)
        hs.msg("完成", 0)
    })
}
"38_04_07_003149" <- local({
    {
        hs.qc_vis <- function(sce_list, wjj2, org1, stage = "before") {
            wjj2 <- hs.fp(wjj2, switch(stage, before = "before_QC", 
                after = "after_QC"), "metrics")
            wj2 <- hs.fp(wjj2, c(paste0(names(sce_list), ".png"), 
                "all_samples.tsv.gz"))
            if (all(file.exists(wj2))) 
                return(wj2)
            a1 <- wk.mclapply(sce_list, . %>% {
                m1 <- .@assays[["RNA"]]@counts
                c(list(umi_n = hs.colSums(m1), gene_n = hs.colSums(m1 > 
                  0)), hs.percent(., org1 = org1))
            }, mc.cores = 1)
            if (uniqueN(sapply(a1, length)) != 1) 
                hs.browser("38.07.12.130009")
            {
                fig_n_max <- max(sapply(a1, length))
                sapply(seq_along(a1), function(sample_i) {
                  wj2 <- hs.fp(wjj2, paste0(names(a1)[sample_i], 
                    ".png"))
                  if (!file.exists(wj2)) {
                    P <- hs.lapply(names(a1[[sample_i]]), function(n1) {
                      dt1 <- wk.dt.fromList(a1[[sample_i]][n1])
                      hs.hsgly("小提琴图/36.06.18.150541")(dt1, 
                        main = "", xlab = "", ylab = switch(n1, 
                          gene_n = "#gene", umi_n = "#UMI", percent.mt = "UMI percentage of\nmitochondrial genes", 
                          percent.ribo = "UMI percentage of\nribosomal genes", 
                          percent.hemo = "UMI percentage of\nhemoglobin subunits"), 
                        N = 0) + ggplot2::scale_x_discrete(breaks = NULL) + 
                        ggplot2::labs(title = switch(n1, gene_n = "Gene", 
                          umi_n = "UMI", percent.mt = "Mitochondrial\ngenes", 
                          percent.ribo = "Ribosomal\ngenes", 
                          percent.hemo = "Hemoglobin\nsubunits"))
                    })
                    P <- eval(bquote(.(parse(text = paste0("P[[ ", 
                      1:length(P), "]]") %>% paste(collapse = " + "))[[1]]) + 
                      patchwork::plot_layout(nrow = 1)))
                    .figs(print(P), wj2, width = 5 * fig_n_max, 
                      height = 5)
                  }
                })
                if (0) {
                  p1 <- paste0("P[[ ", seq_along(P), "]]") %>% 
                    paste(collapse = " | ") %>% paste0(" + plot_layout( ncol = 1)") %>% 
                    evalStr
                  .pdf(print(p1), hs.wjm(wj2, ext = "pdf"), width = 10 * 
                    fig_n_max)
                  .pdf(print(P[[1]] + P[[2]] + patchwork::plot_layout(ncol = 1)), 
                    hs.wjm(wj2, ext = "pdf"), width = 10 * fig_n_max, 
                    height = 10 * 2)
                  wk.ggsave(P[[1]] + P[[2]] + plot_layout(ncol = 1), 
                    wj2, width = 10 * fig_n_max, height = 10 * 
                      length(P), limitsize = FALSE)
                  wk.ggsave(p1, wj2, width = 50 * fig_n_max, 
                    height = 50 * length(P), limitsize = FALSE)
                }
                {
                }
            }
            {
                sapply(c("percent.mt", "percent.hemo"), function(nm1) {
                  wj2 <- hs.fp(wjj2, paste0(nm1, ".pdf"))
                  if (file.exists(wj2)) 
                    return()
                  if (0) {
                    a1[[1]][["percent.mt"]] %>% summary
                    a1[[1]][["percent.hemo"]] %>% summary
                  }
                  sj1 <- lapply(a1, function(a1) {
                    a1 <- a1[[nm1]]
                    if (length(a1)) {
                      sapply(seq(5, 20, 5), function(co1) {
                        mean(a1 > co1) * 100
                      })
                    }
                  }) %>% hs.nonempty
                  if (length(sj1)) {
                    m1 <- sapply(sj1, round, digits = 2) %>% 
                      t
                    colnames(m1) <- sprintf(">%s%%", seq(0.05, 
                      0.2, 0.05) * 100)
                    hs.hsgly("table2pdf.37_12_27_141843")(m1, 
                      wj2, rows = rownames(m1), pdf.args = list(width = 15))
                  }
                })
            }
            {
                wj2 <- hs.fp(wjj2, "all_samples.tsv.gz")
                dt <- lapply(sce_list, colnames) %>% wk.dt.fromList(nm = c("sample", 
                  "barcode"))
                wk.dt.col.reorder(dt, "barcode", after = 0)
                hs.for(cn1, names(a1[[1]]), wk.dt.set(dt, , cn1, 
                  lapply(a1, . %>% {
                    .[[cn1]]
                  }) %>% unlist))
                wk.dt.fwrite(dt, wj2)
            }
        }
        plot.umi_gene_n <- function(sce, wj2, cell = {
        }) {
            if (all(file.exists(paste0(wj2, ".png")))) 
                return()
            p1 <- local({
                pl1 <- FeaturePlot(sce, features = c("umi_N_log10", 
                  "gene_N_log10"), cells = cell, reduction = "umap", 
                  combine = 0)
                p1 <- pl1[[1]] + labs(title = "Cells colored by #UMI", 
                  color = quote(Log[10]("#UMI+1")))
                p2 <- pl1[[2]] + labs(title = "Cells colored by #gene", 
                  color = quote(Log[10]("#gene")))
                pl2 <- FeaturePlot(sce, features = c("umi_N_log10", 
                  "gene_N_log10"), cells = cell, reduction = "tsne", 
                  combine = 0)
                p3 <- pl2[[1]] + labs(title = "", color = quote(Log[10]("#UMI+1")))
                p4 <- pl2[[2]] + labs(title = "", color = quote(Log[10]("#gene")))
                p1 + p2 + p3 + p4 + plot_layout(ncol = 2)
            })
            .figs(print(p1), wj2, width = 25, height = 20)
        }
        do.log10_umi_gene_calculation <- function(sce, wj2) {
            if (!(c("umi_N_log10", "gene_N_log10") %all.in% colnames(sce@meta.data))) {
                sce <- sce@assays[["RNA"]]@counts %>% {
                  a1 <- list(umi_N_log10 = log10(colSums(.) + 
                    1), gene_N_log10 = log10(colSums(. > 0)))
                  for (n1 in names(a1)) sce[[n1]] <- a1[[n1]]
                  sce
                }
            }
            if (!missing(wj2)) 
                plot.umi_gene_n(sce, wj2)
            sce
        }
    }
    {
        do.linear_dimensional_reduction <- do.nonlinear_dimensional_reduction <- {
        }
        wk.cluster_cell <- function(sce, FindClusters.algorithm, 
            sample_name, normalization.method = "log", FindClusters.resolution.db = seq(0.1, 
                2, 0.2)) {
            FindClusters.resolution.db %<>% {
                i1 <- paste0(SeuratObject::DefaultAssay(sce), 
                  "_snn_res.", .) %in% colnames(sce@meta.data)
                if (any(i1)) 
                  . <- .[!i1]
                .
            }
            if (length(FindClusters.resolution.db)) {
                if (normalization.method != "log") 
                  hs.browser("37.12.23.111954")
                pc_n <- get.dimensionality(sce)
                sce <- Seurat::FindNeighbors(sce, reduction = "pca", 
                  dims = 1:pc_n)
                mem1 <- cluster.mem()["Available"]
                nc <- 1
                a1 <- wk.mclapply(FindClusters.resolution.db, 
                  function(n1) {
                    hs.msg(paste0("对细胞分群（resolution=", 
                      n1, "）"))
                    a1 <- (Seurat::FindClusters)(sce, algorithm = FindClusters.algorithm, 
                      resolution = n1, random.seed = 203163659)
                    a1@meta.data[, paste0(SeuratObject::DefaultAssay(sce), 
                      "_snn_res.", n1)]
                  }, mc.cores = nc)
                names(a1) <- paste0(SeuratObject::DefaultAssay(sce), 
                  "_snn_res.", FindClusters.resolution.db)
                for (cn1 in names(a1)) sce@meta.data[, cn1] <- a1[[cn1]]
            }
            sce
        }
        get.clusterings <- local({
            hs1 <- function(sce, res = 0.1/(2^(0:20))) {
                get.a.cluster <- function(res1) {
                  message("分辨率：", res1)
                  cluster_result <- monocle3:::leiden_clustering(data = reduced_dim_res, 
                    pd = sce@meta.data, k = 20, weight = FALSE, 
                    num_iter = 2, resolution_parameter = res1, 
                    random_seed = 160109, verbose = FALSE)
                  factor(igraph::membership(cluster_result$optim_res))
                }
                reduced_dim_res <- SeuratObject::Embeddings(sce, 
                  reduction = "umap")
                cluster_n_min <- 5
                if ((length(res) == 1) && (res == "auto")) {
                  res_low <- 0.1/(2^30) * as.numeric(paste0("1e-", 
                    seq(0, 9) * 3))
                  a <- wk.mclapply(res_low, get.a.cluster)
                  a <- sapply(a, . %>% levels %>% length)
                  a <- a[a > 1]
                  res <- if (length(a)) {
                    0.1/(2^seq(0, ceiling(log2(0.1/res_low[which.min(a)]))))
                  }
                  else min(res_low)
                }
                a <- wk.mclapply(res, get.a.cluster, mc.cores = 10)
                sapply(a, c) %>% t %>% unique %>% t
            }
            function(sce, res = "auto") {
                clusterings <- hs1(sce, res)
                cluster_n <- apply(clusterings, 2, uniqueN)
                if (length(cluster_n) > 1) {
                  hs.for(right1, seq(ceiling(min(cluster_n)/10) * 
                    10, (log10(ncol(sce)) - 1) * 10, by = 10), 
                    {
                      left1 <- right1 - 9
                      clustering_n_co <- ifelse(left1 == 1, 3, 
                        5)
                      repeated_n <- 0
                      repeat {
                        cluster_n <- apply(clusterings, 2, uniqueN)
                        clustering_n <- cluster_n %>% unique %>% 
                          {
                            sum((. <= right1) & (. > left1))
                          }
                        if (clustering_n < clustering_n_co) {
                          if (left1 == 1) {
                            co1 <- as.numeric(names(cluster_n)[which.min(cluster_n)])/100
                            co2 <- cluster_n[cluster_n >= right1] %>% 
                              {
                                names(.)[which.min(.)]
                              } %>% as.numeric
                            res2do <- .try(log10(c(co1, co2)) %>% 
                              {
                                seq(.[1], .[2], length.out = 22)[2:21]
                              } %>% {
                              10^.
                            })
                          }
                          else {
                            co1 <- max(as.numeric(names(cluster_n)[which(cluster_n <= 
                              left1)]))
                            if (is.infinite(co1)) 
                              co1 <- as.numeric(names(cluster_n)[which.min(cluster_n)])
                            co2 <- min(as.numeric(names(cluster_n)[which(cluster_n > 
                              right1)]))
                            res2do <- .try(log10(c(co1, co2)) %>% 
                              {
                                seq(.[1], .[2], length.out = 22)[2:21]
                              } %>% {
                              10^.
                            })
                          }
                          if (is(res2do, "try-error")) 
                            hs.browser("38.08.16.050702")
                          a <- hs1(sce = sce, res = res2do)
                          apply(a, 2, uniqueN)
                          dim(a)
                          clusterings %<>% cbind(a) %>% {
                            .[, hs.order(as.numeric(colnames(.)))]
                          }
                        }
                        else break
                        repeated_n <- repeated_n + 1
                        if (repeated_n > 3) 
                          break
                      }
                    })
                  cluster_n <- clusterings %>% apply(2, uniqueN) %>% 
                    rev
                  i <- cluster_n %>% {
                    names(.)[match(unique(.), .)]
                  } %>% rev
                  clusterings[, i, drop = 0]
                }
                else clusterings
            }
        })
        do.cell_cycle_analyses <- {
        }
        plot.2d <- function(sce, wj2, what = c(if (length(hs.grep(colnames(sce@meta.data), 
            "38_01_19_171144"))) "sample", "cell cluster", "cell type")) {
            length(hs.grepV(colnames(sce@meta.data), "38_01_19_171144"))
            1
            {
                hs.lapply("pca,umap,tsne", function(reduction_method_1) {
                  wj2 <- hs.wjm(wj2, append = reduction_method_1)
                  P <- lapply(what, function(w1) {
                    switch(w1, sample = {
                      DimPlot(sce, reduction = reduction_method_1, 
                        group.by = hs.grepV(colnames(sce@meta.data), 
                          "38_01_19_171144")) + labs(title = "Samples")
                    }, `cell cluster` = {
                      DimPlot(sce, reduction = reduction_method_1, 
                        group.by = choose.clustering_resolution(sce, 
                          ot = "cn"), label = 1) + labs(title = "Cell clusters")
                    }, `cell type` = {
                      DimPlot(sce, reduction = reduction_method_1, 
                        group.by = "cluster_stable.cell_type_SingleR.general", 
                        label = 1, label.size = 2, repel = 0) + 
                        labs(title = switch(reduction_method_1, 
                          umap = "UMAP", tsne = "tSNE")) + theme(legend.text = element_text(size = 5)) + 
                        labs(title = "Cell types")
                    })
                  })
                  cmd1 <- paste0("P[[ ", seq_along(what), "]]") %>% 
                    paste(collapse = " + ")
                  p1 <- evalStr(cmd1)
                  wk.ggsave(p1, wj2, width = 5 + 10 * length(what), 
                    height = 10)
                })
            }
            if (0) {
                if (!file.exists(wj2)) 
                  .pdf({
                  }, wj2, width = 5 + 10 * length(what))
            }
        }
    }
    {
        get.cell_type_ref <- function(org1 = "human", tissue1 = {
        }) {
            hs.hsgly("R/包/celldex/数据.38_03_03_143112")(what = switch(org1, 
                human = "HumanPrimaryCellAtlasData", rat = , 
                mouse = "MouseRNAseqData", hs.browser("38.02.23.090710")))
        }
        do.SingleR <- function(sce, org1 = "human", grouping_name = choose.clustering_resolution(sce, 
            ot = "cn"), nc = 4, ref = get.cell_type_ref(org1 = org1)) {
            hs.require(c("SingleR", "celldex"))
            cell_type_ref <- ref
            if (0) 
                switch(org1, human = celldex::HumanPrimaryCellAtlasData(), 
                  mouse = celldex::MouseRNAseqData(), hs.browser("38.02.23.090710"))
            if (0) {
                attributes(cell_type_ref) %>% names
                colData(cell_type_ref)[, "label.main"] %>% hs.grepV(":")
                colData(cell_type_ref)[, "label.fine"] %>% hs.grepV(":")
                colData(cell_type_ref) %>% {
                  .[.[, "label.main"] != .[, "label.fine"], ]
                } %>% {
                  tapply(.[, "label.fine"], .[, "label.main"], 
                    unique)
                }
                attr(cell_type_ref, "colData")
                attr(cell_type_ref, "colData")[, "label.main"] %>% 
                  hs.table %>% sort
            }
            {
            }
            if (0) {
                cell_type_ref %>% slotNames
                cell_type_ref@colData[, "label.main"] %>% uniqueN
                cell_type_ref@colData[, "label.fine"] %>% uniqueN
                sce_combined@assays[["integrated"]] %>% slotNames
                sce_combined@assays[["integrated"]]@data %>% 
                  colMins %>% summary
                sce_combined@assays[["RNA"]]@data %>% colMaxs %>% 
                  summary
            }
            if (!length(grouping_name)) 
                grouping_name <- hs.grepV(colnames(sce@meta.data), 
                  paste0("^(?i)", DefaultAssay(sce), "_snn_res\\."))
            a1 <- wk.mclapply(grouping_name, function(grouping_1_name) {
                grouping_1 <- sce@meta.data[, grouping_1_name] %>% 
                  hs.factor(ot = "str")
                m1 <- sce@assays[["RNA"]]@data
                if (org1 == "rat") {
                  a <- hs.hsgly("同源基因.37_05_26_124410")(rownames(m1), 
                    org1 = org1, org2 = "mouse", ft = "symbol", 
                    one_to_one = 1)
                  a <- a[tolower(V2) %in% tolower(rownames(cell_type_ref))]
                  i <- a[, V1 %>% tolower %>% match(tolower(rownames(m1)))]
                  wk.dt.set(a, , "V1", rownames(m1)[i])
                  i <- a[, V2 %>% tolower %>% match(tolower(rownames(cell_type_ref)))]
                  wk.dt.set(a, , "V2", rownames(cell_type_ref)[i])
                  rownames(cell_type_ref)
                  m1 <- m1[a[, V1], ]
                  rownames(m1) <- a[, V2]
                }
                (SingleR::SingleR)(m1, cell_type_ref, labels = cell_type_ref$label.fine, 
                  clusters = grouping_1)
            }, mc.cores = nc)
            if (0) {
                attr(sce, "SingleR_annotations") <- hs.lapply(a1, 
                  . %>% {
                    hs.c(rownames(.), .[, "pruned.labels"])
                  })
            }
            if (0) {
                a <- SingleR(sce@assays[["RNA"]]@data, cell_type_ref, 
                  labels = cell_type_ref$label.fine)
                attributes(a) %>% names
                attr(a, "listData") %>% .str
                help(package = "SingleR")
                dd <- getDeltaFromMedian(a)
                summary(dd)
            }
            if (0) {
                a1 <- SingleR(sce@assays[["RNA"]]@data, cell_type_ref, 
                  labels = cell_type_ref$label.fine)
                sce@meta.data[, cluster_name_1] %>% hs.table
                hs.table(a1[["integrated_snn_res.0.5"]][colnames(sce), 
                  "pruned.labels"], sce@meta.data[, "seurat_annotations"]) %>% 
                  .s
            }
            if (length(grouping_name) > 1) {
                a1
            }
            else {
                if (0) {
                  a <- local({
                    a <- a1[[1]] %>% {
                      hs.c(rownames(.), .[, "pruned.labels"])
                    }
                    a[hs.factor(sce@meta.data[, grouping_name], 
                      ot = "str")]
                  })
                  a %all.in% colData(cell_type_ref)[, "label.fine"]
                  a %not% colData(cell_type_ref)[, "label.main"]
                }
                sce[["cluster_stable.cell_type_SingleR"]] <- local({
                  a <- a1[[1]] %>% {
                    hs.c(rownames(.), .[, "pruned.labels"])
                  }
                  a[hs.factor(sce@meta.data[, grouping_name], 
                    ot = "str")]
                })
                sce[["cluster_stable.cell_type_SingleR.general"]] <- sce@meta.data[, 
                  "cluster_stable.cell_type_SingleR"] %>% {
                  if (0) {
                    jg <- hs.c(., rep(NA, length(.)))
                    i1 <- !is.na(.)
                    a1 <- . %not% NA
                    a2 <- hs.re(a1, "(?i)(t_cell:cd\\d.|[^:]+)")
                    jg[i1] <- hs.c(a1, a2)[.[i1]]
                  }
                  jg <- colData(cell_type_ref)[match(., colData(cell_type_ref)[, 
                    "label.fine"]), "label.main"]
                  jg
                }
                sce
            }
        }
    }
    {
        monocle3.cluster_cells <- function(cds, reduction_method = c("UMAP", 
            "tSNE", "PCA", "LSI", "Aligned"), k = 20, cluster_method = c("leiden", 
            "louvain"), num_iter = 2, partition_qval = 0.05, 
            weight = FALSE, resolution = NULL, random_seed = NULL, 
            verbose = FALSE, ...) {
            reduction_method <- match.arg(reduction_method)
            cluster_method <- match.arg(cluster_method)
            assertthat::assert_that(methods::is(cds, "cell_data_set"))
            assertthat::assert_that(is.character(reduction_method))
            assertthat::assert_that(assertthat::is.count(k))
            assertthat::assert_that(is.logical(weight))
            assertthat::assert_that(assertthat::is.count(num_iter))
            if (!is.null(resolution) & (cluster_method == "louvain")) {
                message(paste("Resolution can only be used when cluster_method is", 
                  "'leiden'. Switching to leiden clustering."))
                cluster_method <- "leiden"
            }
            if (!is.null(resolution)) {
                assertthat::assert_that(is.numeric(resolution))
            }
            assertthat::assert_that(is.numeric(partition_qval))
            assertthat::assert_that(is.logical(verbose))
            assertthat::assert_that(!is.null(reducedDims(cds)[[reduction_method]]), 
                msg = paste("No dimensionality reduction for", 
                  reduction_method, "calculated.", "Please run reduce_dimension with", 
                  "reduction_method =", reduction_method, "before running cluster_cells"))
            reduced_dim_res <- reducedDims(cds)[[reduction_method]]
            if (is.null(random_seed)) {
                random_seed <- sample.int(.Machine$integer.max, 
                  1)
            }
            if (verbose) 
                message("Running ", cluster_method, " clustering algorithm ...")
            switch(cluster_method, leiden = {
                cds <- monocle3:::add_citation(cds, "leiden")
                cluster_result <- leiden_clustering(data = reduced_dim_res, 
                  pd = colData(cds), k = k, weight = weight, 
                  num_iter = num_iter, resolution_parameter = resolution, 
                  random_seed = random_seed, verbose = verbose, 
                  ...)
                if (length(unique(cluster_result$optim_res$membership)) > 
                  1) {
                  cluster_graph_res <- monocle3:::compute_partitions(cluster_result$g, 
                    cluster_result$optim_res, partition_qval, 
                    verbose)
                  partitions <- igraph::components(cluster_graph_res$cluster_g)$membership[cluster_result$optim_res$membership]
                  partitions <- as.factor(partitions)
                } else {
                  partitions <- rep(1, nrow(colData(cds)))
                }
                names(partitions) <- row.names(reduced_dim_res)
                clusters <- factor(igraph::membership(cluster_result$optim_res))
                cds@clusters[[reduction_method]] <- list(cluster_result = cluster_result, 
                  partitions = partitions, clusters = clusters)
            }, louvain = {
                cluster_result <- louvain_clustering(data = reduced_dim_res, 
                  pd = colData(cds), k = k, weight = weight, 
                  num_iter = num_iter, random_seed = random_seed, 
                  verbose = verbose, ...)
                if (length(unique(cluster_result$optim_res$membership)) > 
                  1) {
                  cluster_graph_res <- monocle3:::compute_partitions(cluster_result$g, 
                    cluster_result$optim_res, partition_qval, 
                    verbose)
                  partitions <- igraph::components(cluster_graph_res$cluster_g)$membership[cluster_result$optim_res$membership]
                  partitions <- as.factor(partitions)
                } else {
                  partitions <- rep(1, nrow(colData(cds)))
                }
                names(partitions) <- row.names(reduced_dim_res)
                clusters <- factor(igraph::membership(cluster_result$optim_res))
                cds@clusters[[reduction_method]] <- list(cluster_result = cluster_result, 
                  partitions = partitions, clusters = clusters)
            })
            cds <- monocle3:::add_citation(cds, "clusters")
            cds <- monocle3:::add_citation(cds, "partitions")
            return(cds)
        }
        leiden_clustering <- function(data, pd, k = 20, weight = NULL, 
            num_iter = 2, resolution_parameter = 1e-04, random_seed = NULL, 
            verbose = FALSE, ...) {
            extra_arguments <- list(...)
            if ("partition_type" %in% names(extra_arguments)) {
                partition_type <- extra_arguments[["partition_type"]]
            }
            else partition_type <- "CPMVertexPartition"
            if ("initial_membership" %in% names(extra_arguments)) {
                initial_membership <- extra_arguments[["initial_membership"]]
            }
            else initial_membership <- NULL
            if ("weights" %in% names(extra_arguments)) {
                edge_weights <- extra_arguments[["weights"]]
            }
            else edge_weights <- NULL
            if ("node_sizes" %in% names(extra_arguments)) {
                node_sizes <- extra_arguments[["node_sizes"]]
            }
            else node_sizes <- NULL
            if (partition_type %in% c("ModularityVertexPartition", 
                "SignificanceVertexPartition", "SurpriseVertexPartition")) {
                resolution_parameter = NA
            }
            else if (is.null(resolution_parameter)) {
                resolution_parameter = 1e-04
            }
            if (is.null(num_iter)) 
                num_iter = 2
            if (random_seed == 0L) 
                random_seed = NULL
            cell_names <- row.names(pd)
            if (!identical(cell_names, row.names(pd))) 
                stop("Phenotype and row name from the data don't match")
            if (is.data.frame(data)) 
                data <- as.matrix(data)
            if (!is.matrix(data)) 
                stop("Wrong input data, should be a data frame of matrix!")
            if (k < 1) {
                stop("k must be a positive integer!")
            }
            else if (k > nrow(data) - 2) {
                k <- nrow(data) - 2
                warning(paste("RANN counts the point itself, k must be smaller than\nthe", 
                  "total number of points - 1 (all other points) - 1", 
                  "(itself)!"))
            }
            if (verbose) {
                message("Run kNN based graph clustering starts:", 
                  "\n", "  -Input data of ", nrow(data), " rows and ", 
                  ncol(data), " columns", "\n", "  -k is set to ", 
                  k)
                message("  Finding nearest neighbors...")
            }
            t1 <- system.time(tmp <- RANN::nn2(data, data, k + 
                1, searchtype = "standard"))
            neighborMatrix <- tmp[[1]][, -1]
            distMatrix <- tmp[[2]][, -1]
            if (verbose) 
                message("    DONE. Run time: ", t1[3], "s")
            if (verbose) 
                message("  Compute jaccard coefficient between nearest-neighbor sets ...")
            t2 <- system.time(links <- monocle3:::jaccard_coeff(neighborMatrix, 
                weight))
            if (verbose) 
                message("    DONE. Run time: ", t2[3], "s")
            if (verbose) 
                message("  Build undirected graph from the weighted links ...")
            links <- links[links[, 1] > 0, ]
            relations <- as.data.frame(links)
            colnames(relations) <- c("from", "to", "weight")
            relations$from <- cell_names[relations$from]
            relations$to <- cell_names[relations$to]
            t3 <- system.time(g <- igraph::graph.data.frame(relations, 
                directed = FALSE))
            if (verbose) 
                message("    DONE. Run time: ", t3[3], "s")
            if (verbose) 
                message("  Run leiden clustering ...")
            t_start <- Sys.time()
            if (verbose) {
                table_results <- data.frame(resolution_parameter = double(), 
                  quality = double(), modularity = double(), 
                  significance = double(), number_clusters = integer())
            }
            best_modularity <- -1
            best_result <- NULL
            best_resolution_parameter <- "No resolution"
            for (i in 1:length(resolution_parameter)) {
                cur_resolution_parameter <- resolution_parameter[i]
                cluster_result <- leidenbase::leiden_find_partition(g, 
                  partition_type = partition_type, initial_membership = initial_membership, 
                  edge_weights = edge_weights, node_sizes = node_sizes, 
                  seed = random_seed, resolution_parameter = cur_resolution_parameter, 
                  num_iter = num_iter, verbose = verbose)
                quality <- cluster_result[["quality"]]
                modularity <- cluster_result[["modularity"]]
                significance <- cluster_result[["significance"]]
                if (verbose) 
                  table_results <- rbind(table_results, data.frame(resolution_parameter = cur_resolution_parameter, 
                    quality = quality, modularity = modularity, 
                    significance = significance, cluster_count = max(cluster_result[["membership"]])))
                if (verbose) 
                  message("    Current resolution is ", cur_resolution_parameter, 
                    "; Modularity is ", modularity, "; Quality is ", 
                    quality, "; Significance is ", significance, 
                    "; Number of clusters is ", max(cluster_result[["membership"]]))
                if (modularity > best_modularity) {
                  best_result <- cluster_result
                  best_resolution_parameter <- cur_resolution_parameter
                  best_modularity <- modularity
                }
                if (is.null(best_result)) {
                  best_result <- cluster_result
                  best_resolution_parameter <- NULL
                  best_modularity <- cluster_result[["modularity"]]
                }
            }
            t_end <- Sys.time()
            if (verbose) {
                message("    Done. Run time: ", t_end - t_start, 
                  "s\n")
                message("  Clustering statistics")
                selected <- vector(mode = "character", length = length(resolution_parameter))
                for (irespar in 1:length(resolution_parameter)) {
                  if (identical(table_results[["resolution_parameter"]][irespar], 
                    best_resolution_parameter)) 
                    selected[irespar] <- "*"
                  else selected[irespar] <- " "
                }
                print(cbind(` ` = " ", table_results, selected), 
                  row.names = FALSE)
                message()
                message("  Cell counts by cluster")
                membership <- best_result[["membership"]]
                membership_frequency <- stats::aggregate(data.frame(cell_count = membership), 
                  list(cluster = membership), length)
                membership_frequency <- cbind(` ` = " ", membership_frequency, 
                  cell_fraction = sprintf("%.3f", membership_frequency[["cell_count"]]/sum(membership_frequency[["cell_count"]])))
                print(membership_frequency, row.names = FALSE)
                message()
                message("  Maximal modularity is ", best_modularity, 
                  " for resolution parameter ", best_resolution_parameter)
                message("\n  Run kNN based graph clustering DONE.\n  -Number of clusters: ", 
                  max(best_result[["membership"]]))
            }
            if (igraph::vcount(g) < 3000) {
                coord <- NULL
                edge_links <- NULL
            }
            else {
                coord <- NULL
                edge_links <- NULL
            }
            igraph::V(g)$names <- as.character(igraph::V(g))
            out_result <- list(membership = best_result[["membership"]], 
                modularity = best_result[["modularity"]])
            names(out_result$membership) = cell_names
            return(list(g = g, relations = relations, distMatrix = distMatrix, 
                coord = coord, edge_links = edge_links, optim_res = out_result))
        }
        learn_graph <- function(cds, use_partition = FALSE, close_loop = TRUE, 
            learn_graph_control = NULL, verbose = FALSE) {
            reduction_method <- "UMAP"
            if (!is.null(learn_graph_control)) {
                assertthat::assert_that(methods::is(learn_graph_control, 
                  "list"))
                (assertthat::assert_that)(all(names(learn_graph_control) %in% 
                  c("euclidean_distance_ratio", "geodesic_distance_ratio", 
                    "minimal_branch_len", "orthogonal_proj_tip", 
                    "prune_graph", "scale", "ncenter", "rann.k", 
                    "maxiter", "eps", "L1.gamma", "L1.sigma")), 
                  msg = "Unknown variable in learn_graph_control")
            }
            euclidean_distance_ratio <- ifelse(is.null(learn_graph_control$euclidean_distance_ratio), 
                1, learn_graph_control$euclidean_distance_ratio)
            geodesic_distance_ratio <- ifelse(is.null(learn_graph_control$geodesic_distance_ratio), 
                1/3, learn_graph_control$geodesic_distance_ratio)
            minimal_branch_len <- ifelse(is.null(learn_graph_control$minimal_branch_len), 
                10, learn_graph_control$minimal_branch_len)
            orthogonal_proj_tip <- ifelse(is.null(learn_graph_control$orthogonal_proj_tip), 
                FALSE, learn_graph_control$orthogonal_proj_tip)
            prune_graph <- ifelse(is.null(learn_graph_control$prune_graph), 
                TRUE, learn_graph_control$prune_graph)
            ncenter <- learn_graph_control$ncenter
            scale <- ifelse(is.null(learn_graph_control$scale), 
                FALSE, learn_graph_control$scale)
            rann.k <- ifelse(is.null(learn_graph_control$rann.k), 
                25, learn_graph_control$rann.k)
            maxiter <- ifelse(is.null(learn_graph_control$maxiter), 
                10, learn_graph_control$maxiter)
            eps <- ifelse(is.null(learn_graph_control$eps), 1e-05, 
                learn_graph_control$eps)
            L1.gamma <- ifelse(is.null(learn_graph_control$L1.gamma), 
                0.5, learn_graph_control$L1.gamma)
            L1.sigma <- ifelse(is.null(learn_graph_control$L1.sigma), 
                0.01, learn_graph_control$L1.sigma)
            assertthat::assert_that(methods::is(cds, "cell_data_set"))
            (assertthat::assert_that)(reduction_method %in% c("UMAP"), 
                msg = paste0("unsupported or invalid reduction method '", 
                  reduction_method, "'"))
            assertthat::assert_that(is.logical(use_partition))
            assertthat::assert_that(is.logical(close_loop))
            assertthat::assert_that(is.logical(verbose))
            assertthat::assert_that(is.logical(orthogonal_proj_tip))
            assertthat::assert_that(is.logical(prune_graph))
            assertthat::assert_that(is.logical(scale))
            assertthat::assert_that(is.numeric(euclidean_distance_ratio))
            assertthat::assert_that(is.numeric(geodesic_distance_ratio))
            assertthat::assert_that(is.numeric(minimal_branch_len))
            if (!is.null(ncenter)) {
                assertthat::assert_that(assertthat::is.count(ncenter))
            }
            assertthat::assert_that(assertthat::is.count(maxiter))
            assertthat::assert_that(assertthat::is.count(rann.k))
            assertthat::assert_that(is.numeric(eps))
            assertthat::assert_that(is.numeric(L1.sigma))
            assertthat::assert_that(is.numeric(L1.sigma))
            (assertthat::assert_that)(!is.null(reducedDims(cds)[[reduction_method]]), 
                msg = paste("No dimensionality reduction for", 
                  reduction_method, "calculated.", "Please run reduce_dimension with", 
                  "reduction_method =", reduction_method, "and cluster_cells before running", 
                  "learn_graph."))
            (assertthat::assert_that)(!is.null(cds@clusters[[reduction_method]]), 
                msg = paste("No cell clusters for", reduction_method, 
                  "calculated.", "Please run cluster_cells with", 
                  "reduction_method =", reduction_method, "before running learn_graph."))
            if (use_partition) {
                partition_list <- cds@clusters[[reduction_method]]$partitions
            }
            else {
                partition_list <- rep(1, nrow(colData(cds)))
            }
            multi_tree_DDRTree_res <- (monocle3:::multi_component_RGE)(cds, 
                scale = scale, reduction_method = reduction_method, 
                partition_list = partition_list, irlba_pca_res = reducedDims(cds)[[reduction_method]], 
                max_components = max_components, ncenter = ncenter, 
                rann.k = rann.k, maxiter = maxiter, eps = eps, 
                L1.gamma = L1.gamma, L1.sigma = L1.sigma, close_loop = close_loop, 
                euclidean_distance_ratio = euclidean_distance_ratio, 
                geodesic_distance_ratio = geodesic_distance_ratio, 
                prune_graph = prune_graph, minimal_branch_len = minimal_branch_len, 
                verbose = verbose)
            rge_res_W <- multi_tree_DDRTree_res$ddrtree_res_W
            rge_res_Z <- multi_tree_DDRTree_res$ddrtree_res_Z
            rge_res_Y <- multi_tree_DDRTree_res$ddrtree_res_Y
            cds <- multi_tree_DDRTree_res$cds
            dp_mst <- multi_tree_DDRTree_res$dp_mst
            principal_graph(cds)[[reduction_method]] <- dp_mst
            cds@principal_graph_aux[[reduction_method]]$dp_mst <- rge_res_Y
            cds <- project2MST(cds, monocle3:::project_point_to_line_segment, 
                orthogonal_proj_tip, verbose, reduction_method, 
                rge_res_Y)
            cds
        }
        project2MST <- function(cds, Projection_Method, orthogonal_proj_tip = FALSE, 
            verbose, reduction_method, rge_res_Y) {
            dp_mst <- principal_graph(cds)[[reduction_method]]
            Z <- t(reducedDims(cds)[[reduction_method]])
            Y <- rge_res_Y
            cds <- monocle3:::findNearestPointOnMST(cds, reduction_method, 
                rge_res_Y)
            closest_vertex <- cds@principal_graph_aux[[reduction_method]]$pr_graph_cell_proj_closest_vertex
            closest_vertex_names <- colnames(Y)[closest_vertex[, 
                1]]
            tip_leaves <- names(which(igraph::degree(dp_mst) == 
                1))
            if (!is.function(Projection_Method)) {
                P <- Y[, closest_vertex]
            }
            else {
                P <- matrix(rep(0, length(Z)), nrow = nrow(Z))
                nearest_edges <- matrix(rep(0, length(Z[1:2, 
                  ])), ncol = 2)
                row.names(nearest_edges) <- colnames(cds)
                for (i in 1:length(closest_vertex)) {
                  neighbors <- names(igraph::neighborhood(dp_mst, 
                    nodes = closest_vertex_names[i], mode = "all")[[1]])[-1]
                  projection <- NULL
                  distance <- NULL
                  Z_i <- Z[, i]
                  for (neighbor in neighbors) {
                    if (closest_vertex_names[i] %in% tip_leaves) {
                      if (orthogonal_proj_tip) {
                        tmp <- projPointOnLine(Z_i, Y[, c(closest_vertex_names[i], 
                          neighbor)])
                      }
                      else {
                        tmp <- Projection_Method(Z_i, Y[, c(closest_vertex_names[i], 
                          neighbor)])
                      }
                    }
                    else {
                      tmp <- Projection_Method(Z_i, Y[, c(closest_vertex_names[i], 
                        neighbor)])
                    }
                    if (any(is.na(tmp))) {
                      tmp <- Y[, neighbor]
                    }
                    projection <- rbind(projection, tmp)
                    distance <- c(distance, stats::dist(rbind(Z_i, 
                      tmp)))
                  }
                  if (class(projection)[1] != "matrix") {
                    projection <- as.matrix(projection)
                  }
                  which_min <- which.min(distance)
                  if (length(which_min) == 0) 
                    browser()
                  P[, i] <- projection[which_min, ]
                  nearest_edges[i, ] <- c(closest_vertex_names[i], 
                    neighbors[which_min])
                }
            }
            colnames(P) <- colnames(Z)
            dp_mst_list <- igraph::decompose.graph(dp_mst)
            dp_mst_df <- NULL
            partitions <- cds@clusters[[reduction_method]]$partitions
            assertthat::assert_that(!is.null(names(cds@clusters[[reduction_method]]$partitions)), 
                msg = "names(cds@clusters[[reduction_method]]$partitions) == NULL")
            assertthat::assert_that(!(length(colnames(cds)) != 
                length(names(cds@clusters[[reduction_method]]$partitions))), 
                msg = "length( colnames(cds) ) != length(names(cds@clusters[[reduction_method]]$partitions))")
            assertthat::assert_that(!any(colnames(cds) != names(cds@clusters[[reduction_method]]$partitions)), 
                msg = "colnames(cds)!=names(cds@clusters[[reduction_method]]$partitions)")
            if (length(dp_mst_list) == 1 && max(as.integer(levels(partitions))) > 
                1) {
                levels(partitions)[as.numeric(levels(partitions)) != 
                  "1"] <- "1"
            }
            if (!is.null(partitions)) {
                for (cur_partition in sort(unique(partitions))) {
                  data_df <- NULL
                  if (verbose) {
                    message("\nProjecting cells to principal points for partition: ", 
                      cur_partition)
                  }
                  subset_cds_col_names <- names(partitions[partitions == 
                    cur_partition])
                  cur_z <- Z[, subset_cds_col_names]
                  cur_p <- P[, subset_cds_col_names]
                  if (ncol(cur_p) > 0 && nrow(cur_p) > 0) {
                    cur_centroid_name <- igraph::V(dp_mst_list[[1]])$name
                    cur_nearest_edges <- nearest_edges[subset_cds_col_names, 
                      ]
                    data_df <- cbind(as.data.frame(t(cur_p)), 
                      apply(cur_nearest_edges, 1, sort) %>% t())
                    row.names(data_df) <- colnames(cur_p)
                    colnames(data_df) <- c(paste0("P_", 1:nrow(cur_p)), 
                      "source", "target")
                    data_df$distance_2_source <- sqrt(colSums((cur_p - 
                      rge_res_Y[, as.character(data_df[, "source"])])^2))
                    data_df <- data_df %>% tibble::rownames_to_column() %>% 
                      dplyr::mutate(group = paste(source, target, 
                        sep = "_")) %>% dplyr::arrange(group, 
                      dplyr::desc(-distance_2_source))
                    data_df <- data_df %>% dplyr::group_by(group) %>% 
                      dplyr::mutate(new_source = dplyr::lag(rowname), 
                        new_target = rowname)
                    data_df[is.na(data_df$new_source), "new_source"] <- as.character(as.matrix(data_df[is.na(data_df$new_source), 
                      "source"]))
                    added_rows <- which(is.na(data_df$new_source) & 
                      is.na(data_df$new_target))
                    data_df <- as.data.frame(data_df, stringsAsFactors = F)
                    data_df <- as.data.frame(as.matrix(data_df), 
                      stringsAsFactors = F)
                    data_df[added_rows, c("new_source", "new_target")] <- data_df[added_rows - 
                      1, c("rowname", "target")]
                    aug_P = cbind(cur_p, rge_res_Y)
                    data_df$weight <- sqrt(colSums((aug_P[, data_df$new_source] - 
                      aug_P[, data_df$new_target]))^2)
                    data_df$weight <- data_df$weight + min(data_df$weight[data_df$weight > 
                      0])
                    edge_list <- as.data.frame(igraph::get.edgelist(dp_mst_list[[as.numeric(cur_partition)]]), 
                      stringsAsFactors = FALSE)
                    dp <- as.matrix(stats::dist(t(rge_res_Y)[cur_centroid_name, 
                      ]))
                    edge_list$weight <- dp[cbind(edge_list[, 
                      1], edge_list[, 2])]
                    {
                      dimnames(dp)
                      dp[as.matrix(edge_list)]
                      edge_list[, 1] %not% rownames(dp)
                    }
                    colnames(edge_list) <- c("new_source", "new_target", 
                      "weight")
                    dp_mst_df <- Reduce(rbind, list(dp_mst_df, 
                      data_df[, c("new_source", "new_target", 
                        "weight")], edge_list))
                  }
                }
            }
            dp_mst <- igraph::graph.data.frame(dp_mst_df, directed = FALSE)
            cds@principal_graph_aux[[reduction_method]]$pr_graph_cell_proj_tree <- dp_mst
            cds@principal_graph_aux[[reduction_method]]$pr_graph_cell_proj_dist <- P
            closest_vertex_df <- as.matrix(closest_vertex)
            row.names(closest_vertex_df) <- row.names(closest_vertex)
            cds@principal_graph_aux[[reduction_method]]$pr_graph_cell_proj_closest_vertex <- closest_vertex_df
            cds
        }
        graph_test <- function(cds, neighbor_graph = c("knn", 
            "principal_graph"), reduction_method = "UMAP", k = 25, 
            method = c("Moran_I"), alternative = "greater", expression_family = "quasipoisson", 
            cores = 1, verbose = FALSE) {
            neighbor_graph <- match.arg(neighbor_graph)
            lw <- calculateLW(cds, k = k, verbose = verbose, 
                neighbor_graph = neighbor_graph, reduction_method = reduction_method)
            if (verbose) {
                message("Performing Moran's I test: ...")
            }
            exprs_mat <- SingleCellExperiment::counts(cds)[, 
                attr(lw, "region.id"), drop = FALSE]
            sz <- size_factors(cds)[attr(lw, "region.id")]
            wc <- spdep::spweights.constants(lw, zero.policy = TRUE, 
                adjust.n = TRUE)
            test_res <- pbmcapply::pbmclapply(row.names(exprs_mat), 
                FUN = function(x, sz, alternative, method, expression_family) {
                  exprs_val <- exprs_mat[x, ]
                  if (expression_family %in% c("uninormal", "binomialff")) {
                    exprs_val <- exprs_val
                  }
                  else {
                    exprs_val <- log10(exprs_val/sz + 0.1)
                  }
                  test_res <- tryCatch({
                    if (method == "Moran_I") {
                      mt <- suppressWarnings(monocle3:::my.moran.test(exprs_val, 
                        lw, wc, alternative = alternative))
                      data.frame(status = "OK", p_value = mt$p.value, 
                        morans_test_statistic = mt$statistic, 
                        morans_I = mt$estimate[["Moran I statistic"]])
                    }
                    else if (method == "Geary_C") {
                      gt <- suppressWarnings(my.geary.test(exprs_val, 
                        lw, wc, alternative = alternative))
                      data.frame(status = "OK", p_value = gt$p.value, 
                        geary_test_statistic = gt$statistic, 
                        geary_C = gt$estimate[["Geary C statistic"]])
                    }
                  }, error = function(e) {
                    data.frame(status = "FAIL", p_value = NA, 
                      morans_test_statistic = NA, morans_I = NA)
                  })
                }, sz = sz, alternative = alternative, method = method, 
                expression_family = expression_family, mc.cores = cores, 
                ignore.interactive = TRUE)
            if (verbose) {
                message("returning results: ...")
            }
            test_res <- do.call(rbind.data.frame, test_res)
            row.names(test_res) <- row.names(cds)
            test_res <- merge(test_res, rowData(cds), by = "row.names")
            row.names(test_res) <- test_res[, 1]
            test_res[, 1] <- NULL
            test_res$q_value <- 1
            test_res$q_value[which(test_res$status == "OK")] <- stats::p.adjust(subset(test_res, 
                status == "OK")[, "p_value"], method = "BH")
            test_res$status = as.character(test_res$status)
            test_res[row.names(cds), ]
        }
        calculateLW <- function(cds, k, neighbor_graph, reduction_method, 
            verbose = FALSE) {
            if (verbose) {
                message("retrieve the matrices for Moran's I test...")
            }
            knn_res <- NULL
            principal_g <- NULL
            cell_coords <- reducedDims(cds)[[reduction_method]]
            if (neighbor_graph == "knn") {
                knn_res <- RANN::nn2(cell_coords, cell_coords, 
                  min(k + 1, nrow(cell_coords)), searchtype = "standard")[[1]]
            }
            else if (neighbor_graph == "principal_graph") {
                pr_graph_node_coords <- cds@principal_graph_aux[[reduction_method]]$dp_mst
                principal_g <- igraph::get.adjacency(cds@principal_graph[[reduction_method]])[colnames(pr_graph_node_coords), 
                  colnames(pr_graph_node_coords)]
            }
            exprs_mat <- exprs(cds)
            if (neighbor_graph == "knn") {
                if (is.null(knn_res)) {
                  knn_res <- RANN::nn2(cell_coords, cell_coords, 
                    min(k + 1, nrow(cell_coords)), searchtype = "standard")[[1]]
                }
                links <- monocle3:::jaccard_coeff(knn_res[, -1], 
                  F)
                links <- links[links[, 1] > 0, ]
                relations <- as.data.frame(links)
                colnames(relations) <- c("from", "to", "weight")
                knn_res_graph <- igraph::graph.data.frame(relations, 
                  directed = T)
                knn_list <- lapply(1:nrow(knn_res), function(x) knn_res[x, 
                  -1])
                region_id_names <- colnames(cds)
                id_map <- 1:ncol(cds)
                names(id_map) <- id_map
                points_selected <- 1:nrow(knn_res)
                knn_list <- lapply(points_selected, function(x) id_map[as.character(knn_res[x, 
                  -1])])
            }
            else if (neighbor_graph == "principal_graph") {
                cell2pp_map <- cds@principal_graph_aux[[reduction_method]]$pr_graph_cell_proj_closest_vertex
                if (is.null(cell2pp_map)) {
                  stop(paste("Error: projection matrix for each cell to principal", 
                    "points doesn't exist, you may need to rerun learn_graph"))
                }
                cell2pp_map <- cell2pp_map[row.names(cell2pp_map) %in% 
                  row.names(colData(cds)), , drop = FALSE]
                cell2pp_map <- cell2pp_map[colnames(cds), ]
                if (verbose) {
                  message("Identify connecting principal point pairs ...")
                }
                knn_res <- RANN::nn2(cell_coords, cell_coords, 
                  min(k + 1, nrow(cell_coords)), searchtype = "standard")[[1]]
                principal_g_tmp <- principal_g
                diag(principal_g_tmp) <- 1
                cell_membership <- as.factor(cell2pp_map)
                uniq_member <- sort(unique(cell_membership))
                membership_matrix <- Matrix::sparse.model.matrix(~cell_membership + 
                  0)
                colnames(membership_matrix) <- levels(uniq_member)
                feasible_space <- membership_matrix %*% Matrix::tcrossprod(principal_g_tmp[as.numeric(levels(uniq_member)), 
                  as.numeric(levels(uniq_member))], membership_matrix)
                links <- monocle3:::jaccard_coeff(knn_res[, -1], 
                  F)
                links <- links[links[, 1] > 0, ]
                relations <- as.data.frame(links)
                colnames(relations) <- c("from", "to", "weight")
                knn_res_graph <- igraph::graph.data.frame(relations, 
                  directed = T)
                tmp_a <- igraph::get.adjacency(knn_res_graph)
                block_size <- 10000
                num_blocks = ceiling(nrow(tmp_a)/block_size)
                if (verbose) {
                  message("start calculating valid kNN graph ...")
                }
                tmp <- NULL
                for (j in 1:num_blocks) {
                  if (j < num_blocks) {
                    block_a <- tmp_a[((((j - 1) * block_size) + 
                      1):(j * block_size)), ]
                    block_b <- feasible_space[((((j - 1) * block_size) + 
                      1):(j * block_size)), ]
                  }
                  else {
                    block_a <- tmp_a[((((j - 1) * block_size) + 
                      1):(nrow(tmp_a))), ]
                    block_b <- feasible_space[((((j - 1) * block_size) + 
                      1):(nrow(tmp_a))), ]
                  }
                  cur_tmp <- block_a * block_b
                  if (is.null(tmp)) {
                    tmp <- cur_tmp
                  }
                  else {
                    tmp <- rbind(tmp, cur_tmp)
                  }
                }
                if (verbose) {
                  message("Calculating valid kNN graph, done ...")
                }
                region_id_names <- colnames(cds)
                id_map <- 1:ncol(cds)
                names(id_map) <- id_map
                knn_list <- slam::rowapply_simple_triplet_matrix(slam::as.simple_triplet_matrix(tmp), 
                  function(x) {
                    res <- which(as.numeric(x) > 0)
                    if (length(res) == 0) 
                      res <- 0L
                    res
                  })
            }
            else {
                stop("Error: unrecognized neighbor_graph option")
            }
            names(knn_list) <- id_map[names(knn_list)]
            class(knn_list) <- "nb"
            attr(knn_list, "region.id") <- region_id_names
            attr(knn_list, "call") <- match.call()
            lw <- spdep::nb2listw(knn_list, zero.policy = TRUE)
            lw
        }
    }
    get.dimensionality <- {
    }
    choose.clustering_resolution <- function(sce, ot = "sce", 
        cn.pattern = switch(cn.pattern.type, seurat = paste0("^", 
            DefaultAssay(sce), "_snn_res\\."), monocle = paste0("monocle3_res.*\\d$")), 
        cn.pattern.type = "monocle") {
        hs.require("clustree")
        sc3_stability <- clustree:::calc_sc3_stability(sce@meta.data %>% 
            {
                .[, hs.grep(names(.), cn.pattern), drop = FALSE]
            })
        if (0) {
            sce@meta.data[1, ]
            apply(sc3_stability, 2, unique)
            sc3_stability %>% {
                tapply(.[, "cluster"], .[, "resolution"], isSorted)
            } %>% all
            stable_most_median_i <- sc3_stability %>% {
                tapply(.[, "stability"], .[, "resolution"], median)
            } %>% which.max
        }
        {
            a <- sce@meta.data %>% {
                apply(.[, hs.grepV(colnames(.), cn.pattern), 
                  drop = FALSE], 2, . %>% hs.table %>% {
                  .[order(as.numeric(names(.)))]
                })
            }
            sc3_stability <- cbind(sc3_stability, cluster_size = as.vector(unlist(a)))
        }
        stable_most_weighted_i <- sc3_stability %>% {
            sapply(unique(.[, "resolution"]), function(resolution1) {
                a <- .[.[, "resolution"] %in% resolution1, ]
                mean(a[, "stability"] * a[, "cluster_size"])
            })
        } %>% which.max
        stable_most_weighted_cn <- sce@meta.data %>% colnames %>% 
            hs.grepV(cn.pattern) %>% stable_most_weighted_i[]
        switch(ot, sce = {
            Idents(sce) <- sce@meta.data[, stable_most_weighted_cn]
            sce
        }, cn = stable_most_weighted_cn)
    }
    {
        hs.fp_relative <- function(wj1, root1) {
            hs.sub(normalizePath(wj1), normalizePath(root1), 
                ".")
        }
        hs.embed_pdf <- function(wj1, root1, wj = 1) {
            wj1 <- hs.fp_relative(wj1, root1)
            c(if (wj) paste0("[[", wj1, "]]"), "#+BEGIN_EXPORT html", 
                paste0("  <iframe style=\"height: 400px; width: 800px\" src=\"", 
                  wj1, "#view=fit\">"), "  </iframe>", "#+END_EXPORT")
        }
        hs.embed_tbl <- function(tbl1, wj2, tbl1.xls, rn = 1) {
            if (!missing(tbl1.xls)) 
                tbl1 <- wk.xls.read(wj2)
            if (!rn) 
                rownames(tbl1) <- {
                }
            a1 <- xtable::xtable(tbl1) %>% print(type = "html") %>% 
                capture.output %>% {
                c("#+BEGIN_EXPORT html", ., "#+END_EXPORT")
            }
            if (!missing(wj2)) {
                hs.cat(a1, sep = "\n", file = wj2)
            }
            else a1
        }
    }
    hs.out_pre <- function(stage, step, stages, steps) {
        "00qc_01fastq"
        "00qc_02cell"
        stage_db <- list("qc", "cell_clustering", "cell_annotation")
        stage_i <- match(stage, stages)
        step_i <- match(step, steps)
        hs.name <- function(x1) if (length(names(x1))) 
            names(x1)
        else x1
        step_name <- hs.name(steps[step_i])
        stage_name <- hs.name(stages[stage_i])
        paste0(sprintf("%02d", stage_i), stage_name)
    }
    function() {
        {
            hs.require(c("Seurat", "immunogenomics/harmony"))
            hs.require(c("patchwork", "ggplotify", "htmltools"))
            {
                hs.require("metap")
            }
            {
                hs.require("EnhancedVolcano", 1)
            }
            {
                hs.require("cole-trapnell-lab/leidenbase", 1)
                hs.require("cole-trapnell-lab/monocle3", 1)
            }
        }
        if (!length(do.linear_dimensional_reduction)) 
            do.linear_dimensional_reduction <<- hs.hsgly("预处理/降维，线性.38_10_09_120747")
        if (!length(do.nonlinear_dimensional_reduction)) 
            do.nonlinear_dimensional_reduction <- hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")
        if (!length(do.cell_cycle_analyses)) 
            do.cell_cycle_analyses <<- hs.hsgly("转录组/标准流程/模块/高级分析/细胞周期.38_10_08_204152")
        if (!length(get.dimensionality)) 
            get.dimensionality <<- hs.hsgly("转录组/标准流程/模块/预处理/降维，线性.维度.38_10_09_121529")
    }
})
"37_12_09_143244" <- function(prj_fp = if (length(prj_id)) hs.fp(hs.hsgly("中科新生命/业务/单细胞/项目/根目录.38_08_29_131836")(), 
    prj_id), fq_fp_list = hs.hsgly("wjj2fq.37_12_25_144226")(fq_fp_wjj, 
    R = 1, wjj_wise = fq_fp_list.wjj_wise), fq_fp_list.wjj_wise = 1, 
    prj_id = {
    }, prj_fp_root = prj_fp, fq_fp_wjj = hs.fp(prj_fp, "in"), 
    prj_src_fp = {
    }, bam = 1, copy1 = 1, exp_fp_list = if (length(prj_fp) && 
        (!length(fq_fp_list))) hs.fp(prj_fp, "mid/cell_ranger") %>% 
        hs.wjj.ls(wjj = 1) %>% hs.grepV(hs0 = basename, "^[_\\.]", 
        inv = 1) %>% {
        hs.c(basename(.), hs.fp(., "outs/filtered_feature_bc_matrix.h5"))
    }, org1 = "human", tissue_type = {
    }, fn.sn = {
    }, cell_type_marker_list = {
    }, cell_type_marker_list.org = org1, cell_type_list = {
    }, sample2exclude = {
    }, sample2use = {
    }, sce_list = {
    }, sce_combined.wj = hs.fp(if (subset) wjj2 else wjj2load, 
        "mid/sce_ready.rd"), sce_combined = {
    }, comparison = {
    }, clustering_cn2use = {
    }, wjj2 = local({
        a <- if (length(prj_fp)) 
            prj_fp
        else "."
        if (length(wjj2.sub)) 
            a <- hs.fp(a, wjj2.sub)
        a
    }), wjj2.sub = {
    }, wjj2load = prj_fp, subset = 0, feature_min_cell_n = 3, 
    cell_max_mito_p = 15, normalization.method = "log", vars.to.regress = {
    }, FindIntegrationAnchors.reduction = "harmony", cr.only = 0, 
    xlsx.only = 0, update.only = 0, load.seurat_data.only = 0, 
    force_cells = 0, cell_n_expected = 0, sample.cn = "sample_name.38_01_19_171144", 
    sample_group.cn = "sample_group_name.38_08_16_114013", clustering_chosen_cn = "cell_cluster", 
    cell_type_cn = "cell_type", task_list = switch(task_type, 
        std = {
        }, advanced = {
        }), cells = {
    }, clustering_cn_4subsetting = {
    }, task_type = "std", monocle = if (length(monocle.cells)) 2 else 0, 
    monocle.cells = {
    }, gsva = task_type == "advanced", scenic = task_type == 
        "advanced", cell_cycle = task_type == "advanced", cellphonedb = task_type == 
        "advanced") {
    Args <- list(nfeatures = switch(normalization.method, log = 2, 
        sct = 3) * 1000, assay2use = switch(normalization.method, 
        log = "RNA", sct = "SCT"))
    hs.browser("38.10.21.155311", debug = 1)
    if (length(sce_list)) {
        sce_combined <- hs.hsgly("转录组/标准流程/模块/预处理/整合、降维.38_10_09_113753")(seurat_data_list = sce_list, 
            wj2 = sce_combined.wj, wjj2 = hs.fp(wjj2, "mid/sample_integration"), 
            FindIntegrationAnchors.reduction = FindIntegrationAnchors.reduction, 
            nfeatures = Args[["nfeatures"]], vars.to.regress = vars.to.regress, 
            sample.cn = sample.cn, redo = 1)
    }
    if (update.only) {
        hs.hsgly("更新基因数据库.38_10_12_230454")(org1 = org1)
        return()
    }
    else assign("wj.update.reporter.co1", Inf, envir = .GlobalEnv)
    if (xlsx.only) {
        hs.hsgly("tsv2xlsx.38_09_19_172814")(hs.fp(wjj2, "mid"))
        return()
    }
    if (length(fq_fp_list) && (!subset)) {
        wjj2_here <- hs.fp(prj_fp, "mid", "原始数据质控", 
            "fastqc")
        hs.msg("1.流程/单细胞/转录组/细胞/fastqc.38_10_08_091811")
        fq_fp_list <- hs.hsgly("标准流程/模块/预处理/质控/fastqc.38_10_08_091811")(fq_fp_list = fq_fp_list, 
            wjj2 = wjj2_here, fn.sn = fn.sn, sample2exclude = sample2exclude, 
            sample2use = sample2use)
        hs.msg("1.流程/单细胞/转录组/细胞/multiqc.38_10_08_093432")
        hs.hsgly("标准流程/模块/预处理/质控/multiqc.38_10_08_093432")(wjj1 = wjj2_here)
    }
    if ((!length(sce_combined)) && length(sce_combined.wj) && 
        (!file.exists(sce_combined.wj))) {
        hs.msg("单位/业务/单细胞/模块/cell_ranger.38_08_19_130133")
        sce_list <- hs.hsgly("标准流程/模块/10x/cell_ranger.38_08_19_130133")(fq_fp_list = fq_fp_list, 
            org1 = org1, cell_n_expected = cell_n_expected, force_cells = force_cells, 
            is_nuclear = is_nuclear, fn.sn = fn.sn, wjj2 = prj_fp, 
            prj_fp_root = prj_fp_root, aggr = 1, nc = 4, ot = "seurat_data_list", 
            copy1 = copy1, bam = bam)
        if (cr.only) 
            return()
        on.exit(hs.hsgly("tsv2xlsx.38_09_19_172814")(hs.fp(wjj2, 
            "mid")), add = TRUE)
        sce_list <- hs.hsgly("标准流程/模块/预处理/质控/过滤细胞、基因.38_10_19_144214")(seurat_data_list = sce_list, 
            wjj2 = hs.fp(wjj2, "mid", "qc_gene,cell,sample"))
    }
    if (cr.only) 
        return()
    hs.msg("单位/业务/单细胞/模块/整合、降维.38_10_09_113753")
    if (!length(sce_combined)) {
        sce_combined <- hs.hsgly("转录组/标准流程/模块/预处理/整合、降维.38_10_09_113753")(seurat_data_list = sce_list, 
            wj2 = sce_combined.wj, wjj2 = hs.fp(wjj2, "mid/sample_integration"), 
            FindIntegrationAnchors.reduction = FindIntegrationAnchors.reduction, 
            nfeatures = Args[["nfeatures"]], add.cell.ids = names(seurat_data_list), 
            vars.to.regress = vars.to.regress, sample.cn = sample.cn)
    }
    sample_n <- sce_combined@meta.data[, sample.cn] %>% uniqueN
    hs.msg("单位/业务/单细胞/模块/细胞分群.38_10_19_114235")
    sce_combined %<=>% hs.hsgly("单细胞/转录组/标准流程/模块/细胞分群、类型预测/分群.38_10_19_114235")(seurat_data = x, 
        wjj2 = hs.fp(wjj2, "mid"), wjj2load = if (!subset) 
            wjj2load)
    hs.msg("单位/业务/单细胞/模块/细胞类型预测.38_10_18_164508")
    sce_combined <- hs.hsgly("单细胞/转录组/标准流程/模块/细胞分群、类型预测/类型预测.38_10_18_164508")(seurat_data = sce_combined, 
        org1 = org1, tissue_type = tissue_type, cell_type_marker_list = cell_type_marker_list, 
        cell_type_marker_list.org = cell_type_marker_list.org, 
        cell_type_cn = cell_type_cn, clustering_chosen_cn = clustering_chosen_cn, 
        clustering_cn2use = clustering_cn2use, wjj2 = hs.fp(wjj2, 
            "mid"))
    if (load.seurat_data.only) 
        return(sce_combined)
    if ((sample_n > 1) && (!length(comparison))) {
        comparison <- list(as.list(sort(unique(sce_combined@meta.data[, 
            sample.cn]))[1:2]))
    }
    hs.msg("单位/业务/单细胞/模块/细胞群之间的表达谱关联.38_10_18_094624")
    hs.hsgly("转录组/标准流程/模块/细胞分群、类型预测/细胞群之间的表达谱关联.38_10_18_094624")(seurat_data = sce_combined, 
        wjj2 = hs.fp(wjj2, "mid/cell_cluster"))
    hs.msg("单位/业务/单细胞/模块/细胞分群比例分析.38_10_09_062553")
    hs.hsgly("转录组/标准流程/模块/细胞分群、类型预测/细胞比例分析，分群.38_10_09_062553")(seurat_data = sce_combined, 
        wjj2 = hs.fp(wjj2, "mid/cell_cluster"), sample.cn = sample.cn, 
        cell_type_cn = cell_type_cn)
    hs.msg("单位/业务/单细胞/模块/细胞比例分析，类型.38_10_18_160507")
    hs.hsgly("转录组/标准流程/模块/细胞分群、类型预测/细胞比例分析，类型.38_10_18_160507")(seurat_data = sce_combined, 
        cell_type_cn = cell_type_cn, sample.cn = sample.cn, sample_group.cn = "sample_group_name.38_08_16_114013", 
        cell_type_list = cell_type_list, wjj2 = hs.fp(wjj2, "mid/cell_type"))
    hs.msg("单位/业务/单细胞/模块/差异表达基因/群高表达基因.38_10_18_140257")
    cell_cluster_markers <- hs.hsgly("单细胞/转录组/标准流程/模块/差异表达基因/群高表达.38_10_18_140257")(seurat_data = sce_combined, 
        sample_n = sample_n, wjj2 = hs.fp(wjj2, "mid/差异表达分析/特征基因"))
    if (sample_n > 1) {
        hs.msg("单位/业务/单细胞/模块/差异表达基因/组间.38_10_18_154053")
        cell_cluster_de <- hs.hsgly("单细胞/转录组/标准流程/模块/差异表达基因/组间.38_10_18_154053")(seurat_data = sce_combined, 
            comparison = comparison, wjj2 = hs.fp(wjj2, "mid/差异表达分析/dif.between_groups"))
    }
    hs.hsgly("转录组/标准流程/模块/差异表达基因/基因集合富集分析/群高表达基因.38_10_18_102752")(cell_cluster_markers = cell_cluster_markers, 
        seurat_data = sce_combined, org1 = org1, wjj2 = hs.fp(wjj2, 
            "mid/差异基因富集分析/群高表达基因"))
    if (sample_n > 1) {
        hs.hsgly("转录组/标准流程/模块/差异表达基因/基因集合富集分析/组间差异表达基因.38_10_18_103042")(cell_cluster_de = cell_cluster_de, 
            seurat_data = sce_combined, wjj2 = hs.fp(wjj2, "mid/差异基因富集分析/Group_DEG"), 
            org1 = org1)
        hs.hsgly("单细胞/转录组/标准流程/模块/差异表达基因/蛋白互作/38_10_18_104617")(cell_cluster_de = cell_cluster_de, 
            org1 = org1, wjj2 = hs.fp(wjj2, "mid/蛋白互作"))
    }
    if (cell_cycle) 
        sce_combined %<=>% hs.hsgly("转录组/标准流程/模块/高级分析/细胞周期.38_10_08_204152")(seurat_data = x, 
            org1 = org1, wjj2 = hs.fp(wjj2, "mid/cell_cycle"))
    if (gsva) 
        hs.hsgly("转录组/标准流程/模块/高级分析/gsva.38_10_08_230829")(seurat_data = sce_combined, 
            wjj2 = hs.fp(wjj2, "mid/GSVA"), sample_name.cn = sample.cn, 
            comparison = comparison, org1 = org1)
    if (monocle) 
        hs.hsgly("转录组/标准流程/模块/高级分析/monocle.38_09_23_085346")(seurat_data = sce_combined, 
            monocle = monocle, cells = monocle.cells, wjj2 = hs.fp(wjj2, 
                "mid/轨迹分析", paste0("monocle", monocle)))
    if (scenic) 
        hs.hsgly("转录组/标准流程/模块/高级分析/scenic.38_10_08_171232")(seurat_data = sce_combined, 
            org1 = org1, wjj2 = hs.fp(wjj2, "mid/转录调控网络分析/scenic"))
    if (cellphonedb) 
        hs.hsgly("1.流程/单细胞/转录组/细胞通讯/cellphonedb/38_09_21_152841")(seurat_data = sce_combined, 
            wjj2 = wjj2 %>% hs.fp("mid/细胞通讯/CellPhoneDb"), 
            org1 = org1, cell_type_cn = cell_type_cn)
    wjj2
}
"38_08_26_122205" <- local({
    get.clustering_name.most_cell_types <- function(L) {
        L[[2]] %>% sapply(uniqueN) %>% sort %>% {
            names(.)[match(max(.), .)]
        }
        a <- wk.dt(L[[1]] %>% sapply(nrow), L[[2]] %>% sapply(uniqueN), 
            names(L[[1]]))
        wk.dt.setnames(a, c("cluster_n", "cell_type_n", "clustering_name"))
        wk.dt.setorderv(a, c("cluster_n", "cell_type_n"))
        a[match(max(cell_type_n), cell_type_n), clustering_name]
    }
    function(seurat_data, cell_type_marker_list, ot = "list", 
        cluster_n_max = 40) {
        cell_type_marker_list %<>% lapply(hs.str_and, rownames(seurat_data))
        cell_type_marker_list %<>% hs.nonempty
        markers <- cell_type_marker_list %>% unlist %>% unique
        m1 <- seurat_data@assays[["RNA"]]@data[markers, ]
        markers_rank <- apply(m1, 1, rank, ties.method = "min")/ncol(m1)
        clusterings <- seurat_data@meta.data %>% {
            .[, colnames(.) %>% hs.grep("(?i)monocle3_res\\.")]
        }
        clusterings %<>% {
            .[, order(apply(., 2, uniqueN))]
        }
        fg <- which(apply(clusterings, 2, uniqueN) <= cluster_n_max) %>% 
            wk.mclapply(function(clustering_i) {
                hs.msg(clustering_i)
                clustering <- clusterings[, clustering_i] %>% 
                  as.integer
                cell_type_marker_list %>% seq_along %>% sapply(function(marker_list_1_i) {
                  marker_list_1 <- cell_type_marker_list[[marker_list_1_i]]
                  marker_list_1_rank <- markers_rank[, marker_list_1, 
                    drop = 0]
                  tapply(seq(nrow(marker_list_1_rank)), clustering, 
                    function(i) {
                      marker_list_1_rank[i, , drop = 0] %>% hs.colMeans
                    }) %>% sapply(mean)
                })
            })
        cell_type_list <- lapply(fg, . %>% {
            . <- .[order(as.numeric(rownames(.))), , drop = 0]
            hs.c(rownames(.), names(cell_type_marker_list)[apply(., 
                1, which.max)])
        })
        L <- list(fg, cell_type_list)
        hs.switch(ot, "list", L, "most_cell_types", {
            clustering_name <- get.clustering_name.most_cell_types(L)
            if (0) {
                fg[[clustering_name]] %>% .s
                colnames(fg[[clustering_name]]) <- names(cell_type_marker_list)
            }
            jg <- L[[2]][[clustering_name]] %>% {
                a <- tapply(seq_along(.), ., c)
                b <- seq_along(a) %>% lapply(. %>% {
                  if (length(a[[.]]) > 1) {
                    paste0(names(a)[.], "_", seq_along(a[[.]]))
                  }
                  else names(a)[.]
                })
                c <- wk.dt(rownames(L[[1]][[clustering_name]]) %>% 
                  {
                    .[order(as.numeric(.))]
                  }, rep(names(a), sapply(a, length))[order(unlist(a))], 
                  unlist(b)[order(unlist(a))], cell_type_list[[clustering_name]] %>% 
                    {
                      sapply(seq_along(.), function(ri1) {
                        fg[[clustering_name]][ri1, match(.[ri1], 
                          names(cell_type_marker_list))]
                      })
                    }, fg[[clustering_name]] %>% apply(1, . %>% 
                    hs.order %>% 2[]) %>% {
                    names(cell_type_marker_list)[.]
                  }, fg[[clustering_name]] %>% apply(1, . %>% 
                    hs.sort %>% 2[]))
                wk.dt.setnames(c, c("cluster", "cell_type", "cell_subtype", 
                  "score", "alternative", "alternative.score"))
                c[, `:=`("confidence", 1 - alternative.score/score)]
            }
            jg <- list(jg)
            names(jg) <- clustering_name
            jg
        })
    }
})
"38_09_14_164012" <- function(seurat_data, org1, tissue_type, 
    cluster_n_max = 40, cell_type_column_name = "cell_type", 
    clustering_cn2use = {
    }, clustering_chosen_cn = "cell_cluster") {
    hs.browser("38.12.16.153505", debug = 0)
    ref <- hs.switch(org1, "human", local({
        fp.rd <- "~/swxx/sj/db/Tabula Sapiens" %=>% hs.fp(x, 
            hs.switch(tissue_type, "血", "Blood", "心", "Heart", 
                "膀胱", "Bladder", "肺", "Lung", tissue_type) %=>% 
                paste0(x, ".rd"))
        fp_ready.rd <- hs.wjm(fp.rd, append = "ready")
        if (file.exists(fp_ready.rd)) {
            hs.load(fp_ready.rd)
        }
        else {
            ref <- hs.load(fp.rd)
            if (0) {
                fp.rd %>% normalizePath
                ref@meta.data[1, ]
                ref@meta.data[, "cell_ontology_class"] %>% hs.factor(N = 0) %>% 
                  hs.table
            }
            m1 <- ref@assays[["RNA"]]@counts
            i <- hs.rowSums(m1 > 0)
            a <- SeuratObject::CreateSeuratObject(counts = m1[i >= 
                10, ])
            a@meta.data <- ref@meta.data[i, ]
            a <- NormalizeData(a)
            a <- FindVariableFeatures(a, selection.method = "vst", 
                nfeatures = 2000, verbose = FALSE)
            a %<>% (hs.hsgly("转录组/标准流程/模块/r函数_自制.38_04_07_003149", 
                "do.linear_dimensional_reduction"))
            hs.save(a, wj = fp_ready.rd)
            a
        }
    }), "mouse", hs.hsgly("单细胞/Tabula Muris/Single-cell RNA-seq data from Smart-seq2 sequencing of FACS sorted cells (v2).38_07_14_095423")(hs.switch(tissue_type, 
        "肾", "Kidney", "脑", c("Brain_Myeloid", "Brain_Non-Myeloid"), 
        "肠", "Large_Intestine", "心", "Heart", "肺", "Lung", 
        tissue_type)), hs.browser("38.08.24.222130"))
    cell_type_cn <- colnames(ref@meta.data) %and% c("cell_ontology_class", 
        cell_type_column_name)
    if (0) {
        ref %>% str
        ref@meta.data[, cell_type_cn] %>% hs.factor(N = 0) %>% 
            hs.table
    }
    if (seurat_data %=>% {
        is.character(x) && (length(x) == 1) && file.exists(x)
    }) 
        seurat_data %<=>% hs.load
    pc_n <- max(ncol(ref@reductions[["pca"]]), hs.hsgly("转录组/标准流程/模块/r函数_自制.38_04_07_003149", 
        "get.dimensionality")(seurat_data))
    hs.msg("anchors <- Seurat::FindTransferAnchors( reference = ref, query = seurat_data, dims = seq( pc_n))")
    anchors <- Seurat::FindTransferAnchors(reference = ref, query = seurat_data, 
        reference.assay = SeuratObject::DefaultAssay(ref), query.assay = SeuratObject::DefaultAssay(seurat_data), 
        dims = seq(pc_n))
    if (0) {
        SeuratObject::VariableFeatures(ref) %and% SeuratObject::VariableFeatures(seurat_data)
        SeuratObject::VariableFeatures(seurat_data, assay = "integrated")
    }
    hs.msg("完成", 0)
    predictions <- Seurat::TransferData(anchorset = anchors, 
        refdata = ref@meta.data[, cell_type_cn])
    i <- which(predictions[, "predicted.id"] %in% "")
    if (length(i)) {
        ci2exclude <- c(1, ncol(predictions))
        j <- predictions[i, -ci2exclude] %=>% apply(1, which.max)
        predictions[i, "predicted.id"] <- colnames(predictions)[-ci2exclude] %>% 
            hs.sub("prediction.score.") %=>% x[j]
    }
    if (0) {
        predictions %>% {
            .[.[, "predicted.id"] == "", ]
        }
        predictions[, "predicted.id"]
        predictions[1, ]
        predictions[, "predicted.id"] %>% hs.table
    }
    certainty_score_relative <- apply(predictions[, -1], 1, function(x) {
        a <- hs.sort(x)[2:3]
        1 - a[2]/a[1]
    })
    if (0) {
        clusterings %>% apply(2, uniqueN) %>% {
            which(. <= cluster_n_max)
        }
        clustering_i <- match("0.0001953125", colnames(clusterings))
    }
    clusterings <- if (length(clustering_cn2use)) {
        seurat_data@meta.data[, clustering_cn2use, drop = 0]
    }
    else seurat_data@meta.data %=>% hs.ij(x, j = names(x) %=>% 
        hs.grep("(?i)monocle3_res\\."))
    for (i in seq(ncol(clusterings))) {
        clusterings[, i] %<=>% hs.factor(x, sort(unique(x)), 
            N = 0)
    }
    clustering_i_2do <- if (length(clustering_cn2use)) {
        match(clustering_cn2use, colnames(clusterings))
    }
    else clusterings %=>% apply(2, uniqueN) %=>% which(x <= cluster_n_max)
    a <- sapply(clustering_i_2do, function(clustering_i) {
        message(clustering_i)
        clustering <- clusterings[, clustering_i]
        a <- hs.38_07_19_104409(clustering, predictions[, "predicted.id"], 
            certainty_score_relative)
        b <- hs.38_07_19_104409(predictions[, "predicted.id"], 
            clustering, certainty_score_relative)
        c(mean(a), mean(b))
    })
    if (length(clustering_i_2do) == 1) 
        colnames(a) <- clustering_cn2use
    i <- which.max(hs.colMeans(a))
    if (0) {
        colnames(a)[i]
        clusterings[, colnames(a)[i]] %>% uniqueN
        a[, i]
    }
    clustering <- clusterings[, colnames(a)[i]]
    if (!isSorted(as.numeric(levels(clustering)))) 
        clustering %<=>% hs.factor(x, sort(as.numeric(levels(x))), 
            N = 0)
    SeuratObject::Idents(seurat_data) <- clustering
    if (clustering_chosen_cn %=>% {
        (length(x) == 1) && nzchar(x)
    }) 
        seurat_data@meta.data[, clustering_chosen_cn] <- clustering
    a <- hs.38_07_19_104409(clustering, predictions[, "predicted.id"], 
        certainty_score_relative, classify = 1)
    a <- hs.str_diversify(a)
    if (0) 
        a[paste(clustering)] %=>% hs.table
    seurat_data@meta.data[, cell_type_column_name] <- a[paste(clustering)] %>% 
        unname
    seurat_data
}
"38_09_06_141838" <- function(seurat_data, org1, tissue_type, 
    ref = {
    }, cell_type_column_name = "cell_type", clustering_cn2use = {
    }, clustering_chosen_cn = "cell_cluster", cluster_n_max = 40) {
    m1 <- seurat_data@assays[["RNA"]]@data
    clusterings <- seurat_data@meta.data %>% {
        .[, names(.) %>% hs.grep("^(?i)monocle3_res\\."), drop = 0]
    }
    clustering_cn2do <- if (length(clustering_cn2use)) {
        clustering_cn2use
    }
    else clusterings %>% apply(2, uniqueN) %>% hs.sort %>% {
        names(.)[. <= cluster_n_max]
    }
    set.seed(223511)
    a <- wk.mclapply(clustering_cn2do, function(clustering_cn) {
        hs.msg(clustering_cn)
        clustering <- clusterings[, clustering_cn]
        level <- unique(clustering) %>% paste %>% {
            .[order(as.numeric(.))]
        }
        n_wanted <- 5000
        ci2do <- if (length(clustering) > n_wanted) {
            a <- hs.catch({
                hs.for(li1, seq_along(level) %>% rev, {
                  co1 <- ceiling(n_wanted/li1)
                  l1 <- level[li1]
                  i <- which(clustering %in% l1)
                  i <- if (length(i) <= co1) {
                    i
                  }
                  else sample(i, co1)
                  n_wanted <- n_wanted - length(i)
                  hs.throw(i)
                })
            })
            unlist(a) %>% sort
        }
        else seq_along(clustering)
        b <- (SingleR::SingleR)(m1[, ci2do], ref, labels = ref$label.fine, 
            clusters = clustering[ci2do])
        if (0) {
            cell <- (SingleR::SingleR)(seurat_data@assays[["RNA"]]@data[, 
                i], ref, labels = ref$label.fine)
        }
        a <- hs.table(b[, "pruned.labels"], level)
        .s(a)
        a
    })
    names(a) <- clustering_cn2do
    cell_type_n <- sapply(a, . %>% {
        sum(rownames(.) %not.in% NA)
    })
    clustering_most_cell_type_cn <- cell_type_n %>% rev %>% {
        names(.)[match(max(.), .)]
    }
    if (0) 
        clusterings[, clustering_most_cell_type_cn] %>% {
            hs.factor(., unique(.) %>% as.numeric %>% sort, N = 0)
        } %>% hs.table %>% .s
    SeuratObject::Idents(seurat_data) <- seurat_data@meta.data[, 
        clustering_most_cell_type_cn]
    if ((length(clustering_chosen_cn) == 1) && nzchar(clustering_chosen_cn)) 
        seurat_data@meta.data[, clustering_chosen_cn] <- seurat_data@meta.data[, 
            clustering_most_cell_type_cn]
    seurat_data@meta.data[, cell_type_column_name] <- a[[clustering_most_cell_type_cn]] %>% 
        {
            hs.c(colnames(.), rownames(.)[apply(. > 0, 2, which)])
        } %>% paste(SeuratObject::Idents(seurat_data))[]
    if (0) {
        a[[clustering_most_cell_type_cn]] %>% {
            hs.c(colnames(.), rownames(.)[apply(. > 0, 2, which)])
        } %>% paste(SeuratObject::Idents(sce_combined))[] %>% 
            hs.table
    }
    seurat_data
}
"38_12_19_110421" <- function(..mem = 1) {
    db_wjj <- hs.hsgly("配置/38_12_12_140723")()
    hs.fp(db_wjj, "encyclopedia_table.xlsx") %=>% wk.xls.read
}
"38_12_12_140723" <- function() {
    wjj2 <- hs.sjgly("38_12_12_140723")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        .wk(paste0("CELLTYPIST_FOLDER=", .q(normalizePath(wjj2))), 
            "celltypist --update-models")
    }
    wjj2
}
"38_12_12_140628" <- function(sub = {
}) {
    jg <- .mfp.swxx("38_12_12_140631")
    if (length(sub)) 
        jg <- hs.fp(jg, sub)
    jg
}
"38_12_12_140912" <- function(cmd = "--help") {
    db_wjj <- hs.hsgly("配置/38_12_12_140723")()
    .sys(paste0("CELLTYPIST_FOLDER=", .q(normalizePath(db_wjj))), 
        "celltypist", cmd)
}
"39_01_10_151840" <- function() {
    wjj2 <- hs.sjgly("39_01_10_151840")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        wk.rsync.out("~/org/工作/生物学/生信/1.流程/单细胞/转录组/预测细胞类型/celltypist/encyclopedia_table/encyclopedia_table.xlsx", 
            wjj2)
    }
    hs.fp(wjj2, "encyclopedia_table.xlsx") %=>% wk.xls.read
}
"38_11_24_003610" <- function(V) {
    jg <- list()
    length(V)/100
    if (min(V) > 1000) 
        return()
    V_density <- ieiwk.density(V, n = 100)
    `?`(density)
    x1 <- V_density %>% {
        if (0) 
            .[["x"]][which.max(.[["y"]][.[["x"]] < 500])]
        i1 <- .[["y"]] %>% diff %>% sign %>% {
            match(-1, .)
        }
        .[["x"]][i1]
    }
    V_density %>% "y"[[]] %>% diff
    V_density %>% {
        .[["x"]][which.max(.[["y"]])]
    }
    co1 <- V[V <= x1] %>% {
        mad(., max(.))
    } %>% {
        . * 3 + x1
    }
    hs.browser("38.11.24.004952", debug = 0)
    {
        .pdf(plot(hs.sort(V), type = "l"), width = 50)
        .pdf({
            plot(V_density)
            abline(v = V_density %>% {
                .[["x"]][which.max(.[["y"]])]
            })
        })
        "~/swxx/tmp.pdf" %>% wk.rsync.in(ssh = "fwq")
        co1
    }
    if ((co1/median(V)) > 0.5) {
        Vlog <- log2(V)
        . <- ieiwk.density(Vlog)
        i1 <- which.max(.[["y"]])
        x1 <- .[["x"]][i1]
        a <- Vlog[Vlog > x1]
        a_min <- min(a)
        co2 <- a_min + mad(a, a_min) * 2
        a <- Vlog[Vlog <= x1]
        co1 <- max(a) - mad(a, max(a)) * 2
        .pdf({
            plot(.)
            abline(v = c(co1, co2))
            text(c(co1, co2), par("usr")[c(3, 4)] %>% mean, labels = round(2^c(co1, 
                co2), 1))
        })
        co1 <- 2^co1
        co2 <- 2^co2
    }
    else {
        co2 <- hs.hsgly("离群者/mad.38_03_08_101528")(V, ot = "co")[2]
    }
    jg[["too_few"]][["cut_off"]] <- co1
    jg[["too_few"]][["i"]] <- V < co1
    jg[["too_many"]][["cut_off"]] <- co2
    jg[["too_many"]][["i"]] <- V > co2
    if (co1 > 500) 
        warning("[2022_03_08_104957]样本的阳性基因数下限（", 
            ceiling(co1), "）超过了500")
    if (0) {
        .pdf({
            plot(V_density)
            abline(v = jg %>% sapply("[[", "cut_off"))
        })
    }
    jg
}
"36_09_08_181701" <- function(sj) {
    NormalizeData(sj)
}
"36_09_08_182404" <- function(sj, rm.mt = 1) {
    hs.require(c("Seurat", "sctransform"))
    Args <- list(sj)
    if (rm.mt) {
        sj <- hs.hsgly("percent_mt.36.09.08.183156")(sj)
        Args[["vars.to.regress"]] <- "percent.mt"
    }
    do.call("SCTransform", Args)
}
"38_12_15_001757" <- local({
    FindClusters.default <- function(object, modularity.fxn = 1, 
        initial.membership = NULL, node.sizes = NULL, resolution = 0.8, 
        method = "matrix", algorithm = 1, n.start = 10, n.iter = 10, 
        random.seed = 0, group.singletons = TRUE, temp.file.location = NULL, 
        edge.file.name = NULL, verbose = TRUE, ...) {
        CheckDots(...)
        if (is.null(x = object)) {
            stop("Please provide an SNN graph")
        }
        if (tolower(x = algorithm) == "louvain") {
            algorithm <- 1
        }
        if (tolower(x = algorithm) == "leiden") {
            algorithm <- 4
        }
        if (nbrOfWorkers() > 1) {
            clustering.results <- future_lapply(X = resolution, 
                FUN = function(r) {
                  if (algorithm %in% c(1:3)) {
                    ids <- RunModularityClustering(SNN = object, 
                      modularity = modularity.fxn, resolution = r, 
                      algorithm = algorithm, n.start = n.start, 
                      n.iter = n.iter, random.seed = random.seed, 
                      print.output = verbose, temp.file.location = temp.file.location, 
                      edge.file.name = edge.file.name)
                  }
                  else if (algorithm == 4) {
                    Seurat:::RunLeiden
                    ids <- Seurat:::RunLeiden(object = object, 
                      method = method, partition.type = "RBConfigurationVertexPartition", 
                      initial.membership = initial.membership, 
                      node.sizes = node.sizes, resolution.parameter = r, 
                      random.seed = random.seed, n.iter = n.iter)
                  }
                  else {
                    stop("algorithm not recognised, please specify as an integer or string")
                  }
                  names(x = ids) <- colnames(x = object)
                  ids <- GroupSingletons(ids = ids, SNN = object, 
                    verbose = verbose)
                  results <- list(factor(x = ids))
                  names(x = results) <- paste0("res.", r)
                  return(results)
                })
            clustering.results <- as.data.frame(x = clustering.results)
        }
        else {
            clustering.results <- data.frame(row.names = colnames(x = object))
            for (r in resolution) {
                if (algorithm %in% c(1:3)) {
                  ids <- RunModularityClustering(SNN = object, 
                    modularity = modularity.fxn, resolution = r, 
                    algorithm = algorithm, n.start = n.start, 
                    n.iter = n.iter, random.seed = random.seed, 
                    print.output = verbose, temp.file.location = temp.file.location, 
                    edge.file.name = edge.file.name)
                }
                else if (algorithm == 4) {
                  ids <- RunLeiden(object = object, method = method, 
                    partition.type = "RBConfigurationVertexPartition", 
                    initial.membership = initial.membership, 
                    node.sizes = node.sizes, resolution.parameter = r, 
                    random.seed = random.seed, n.iter = n.iter)
                }
                else {
                  stop("algorithm not recognised, please specify as an integer or string")
                }
                names(x = ids) <- colnames(x = object)
                ids <- GroupSingletons(ids = ids, SNN = object, 
                  group.singletons = group.singletons, verbose = verbose)
                clustering.results[, paste0("res.", r)] <- factor(x = ids)
            }
        }
        return(clustering.results)
    }
    FindClusters.Seurat <- function(object, graph.name = NULL, 
        modularity.fxn = 1, initial.membership = NULL, node.sizes = NULL, 
        resolution = 0.8, method = "matrix", algorithm = 1, n.start = 10, 
        n.iter = 10, random.seed = 0, group.singletons = TRUE, 
        temp.file.location = NULL, edge.file.name = NULL, verbose = TRUE, 
        ...) {
        CheckDots(...)
        graph.name <- graph.name %||% paste0(DefaultAssay(object = object), 
            "_snn")
        if (!graph.name %in% names(x = object)) {
            stop("Provided graph.name not present in Seurat object")
        }
        if (!is(object = object[[graph.name]], class2 = "Graph")) {
            stop("Provided graph.name does not correspond to a graph object.")
        }
        clustering.results <- FindClusters(object = object[[graph.name]], 
            modularity.fxn = modularity.fxn, initial.membership = initial.membership, 
            node.sizes = node.sizes, resolution = resolution, 
            method = method, algorithm = algorithm, n.start = n.start, 
            n.iter = n.iter, random.seed = random.seed, group.singletons = group.singletons, 
            temp.file.location = temp.file.location, edge.file.name = edge.file.name, 
            verbose = verbose, ...)
        colnames(x = clustering.results) <- paste0(graph.name, 
            "_", colnames(x = clustering.results))
        object <- AddMetaData(object = object, metadata = clustering.results)
        Idents(object = object) <- colnames(x = clustering.results)[ncol(x = clustering.results)]
        levels <- levels(x = object)
        levels <- tryCatch(expr = as.numeric(x = levels), warning = function(...) {
            return(levels)
        }, error = function(...) {
            return(levels)
        })
        Idents(object = object) <- factor(x = Idents(object = object), 
            levels = sort(x = levels))
        object[["seurat_clusters"]] <- Idents(object = object)
        cmd <- LogSeuratCommand(object = object, return.command = TRUE)
        slot(object = cmd, name = "assay.used") <- DefaultAssay(object = object[[graph.name]])
        object[[slot(object = cmd, name = "name")]] <- cmd
        return(object)
    }
})
"2023_06_30_131109" <- function(seurat_data, id1) {
    id_temp <- "2023_06_30_131204"
    seurat_data@meta.data[, id_temp] <- hs.switch(length(id1), 
        1, seurat_data@meta.data[, id1], ncol(seurat_data), id1, 
        hs.browser("2023.06.30.131337", debug = 0))
    jg <- Seurat::SplitObject(seurat_data, id_temp)
    for (i1 in seq_along(jg)) {
        jg[[i1]]@meta.data[, id_temp] <- {
        }
    }
    jg
}
"38_12_11_113923" <- function(seurat_data, id = {
}, shuffle = 0, pt.size = 0.5, raster = 1, first_is_gray = 1, 
    ot = "p", ...) {
    id <- hs.hsgly("seurat/函数/idents，获取.38_12_15_114046")(seurat_data = seurat_data, 
        id = id)
    if (uniqueN(id) > 40) {
        a <- sort(hs.table(id))
        others <- names(head(a, -39))
        id[id %in% others] <- "others"
    }
    if (ot == "text width") 
        return(text.width.hs(levels(id)))
    Seurat::Idents(seurat_data) <- hs.hsgly("r/因子/对类排序.38_12_09_153556")(id, 
        N = 1)
    col_n <- uniqueN(Seurat::Idents(seurat_data))
    front <- if (!shuffle) 
        levels(Seurat::Idents(seurat_data))
    cols <- c(if (first_is_gray) "gray", hs.hsgly("color.db.38_09_09_173646")()) %=>% 
        head(col_n)
    Seurat::DimPlot(seurat_data, order = front, shuffle = as.logical(shuffle), 
        pt.size = pt.size, cols = cols, raster = as.logical(raster), 
        ...)
}
"38_12_15_112614" <- function(seurat_data, id, id_db = {
}, ..., width = 40, height = 30, wjj2 = ".") {
    id <- hs.hsgly("seurat/函数/idents，获取.38_12_15_114046")(seurat_data = seurat_data, 
        id = id)
    wj2 <- hs.fp(wjj2, "all_cells.png")
    if (!file.exists(wj2)) {
        p1 <- hs.hsgly("单细胞/转录组/seurat/函数/图/低维.38_12_11_113923")(seurat_data = seurat_data, 
            id = id, reduction = "umap", shuffle = 1, ...)
        .fig(print(p1), wj = wj2, width = width * 1.2, height = height)
    }
    if (!length(id_db)) {
        id_db <- hs.table(id) %=>% hs.sort(x) %=>% names
    }
    else if (is.function(id_db)) {
        id_db %<=>% x(id)
    }
    n_width <- nchar(max(seq_along(id_db)))
    for (i in seq_along(id_db)) {
        id1_name <- id_db[i]
        wj2 <- hs.fp(wjj2, paste0(hs.str_pad(i, n_width, 0), 
            "_", id1_name, ".png"))
        if (file.exists(wj2)) 
            next
        id1 <- ifelse(id %in% id1_name, id1_name, "others")
        p1 <- hs.hsgly("单细胞/转录组/seurat/函数/图/低维.38_12_11_113923")(seurat_data = seurat_data, 
            id = id1, reduction = "umap", ...) + ggplot2::labs(title = id1_name)
        .fig(print(p1), wj = wj2, width = width, height = height)
    }
}
"38_12_22_165909" <- local({
    DotPlot <- function(object, assay = NULL, features, cols = c("lightgrey", 
        "blue"), col.min = -2.5, col.max = 2.5, dot.min = 0, 
        dot.scale = 6, idents = NULL, group.by = NULL, split.by = NULL, 
        cluster.idents = FALSE, scale = TRUE, scale.by = "radius", 
        scale.min = NA, scale.max = NA) {
        assay <- assay %||% DefaultAssay(object = object)
        DefaultAssay(object = object) <- assay
        split.colors <- !is.null(x = split.by) && !any(cols %in% 
            rownames(x = brewer.pal.info))
        scale.func <- switch(EXPR = scale.by, size = scale_size, 
            radius = scale_radius, stop("'scale.by' must be either 'size' or 'radius'"))
        feature.groups <- NULL
        if (is.list(features) | any(!is.na(names(features)))) {
            feature.groups <- unlist(x = sapply(X = 1:length(features), 
                FUN = function(x) {
                  return(rep(x = names(x = features)[x], each = length(features[[x]])))
                }))
            if (any(is.na(x = feature.groups))) {
                warning("Some feature groups are unnamed.", call. = FALSE, 
                  immediate. = TRUE)
            }
            features <- unlist(x = features)
            names(x = feature.groups) <- features
        }
        cells <- unlist(x = CellsByIdentities(object = object, 
            idents = idents))
        data.features <- FetchData(object = object, vars = features, 
            cells = cells)
        data.features$id <- if (is.null(x = group.by)) {
            Idents(object = object)[cells, drop = TRUE]
        }
        else {
            object[[group.by, drop = TRUE]][cells, drop = TRUE]
        }
        if (!is.factor(x = data.features$id)) {
            data.features$id <- factor(x = data.features$id)
        }
        id.levels <- levels(x = data.features$id)
        data.features$id <- as.vector(x = data.features$id)
        if (!is.null(x = split.by)) {
            splits <- object[[split.by, drop = TRUE]][cells, 
                drop = TRUE]
            if (split.colors) {
                if (length(x = unique(x = splits)) > length(x = cols)) {
                  stop("Not enough colors for the number of groups")
                }
                cols <- cols[1:length(x = unique(x = splits))]
                names(x = cols) <- unique(x = splits)
            }
            data.features$id <- paste(data.features$id, splits, 
                sep = "_")
            unique.splits <- unique(x = splits)
            id.levels <- paste0(rep(x = id.levels, each = length(x = unique.splits)), 
                "_", rep(x = unique(x = splits), times = length(x = id.levels)))
        }
        data.plot <- lapply(X = unique(x = data.features$id), 
            FUN = function(ident) {
                data.use <- data.features[data.features$id == 
                  ident, 1:(ncol(x = data.features) - 1), drop = FALSE]
                apply(data.use, 2, mean)
                apply(data.use, 2, . %=>% log1p(mean(expm1(x))))
                avg.exp <- apply(X = data.use, MARGIN = 2, FUN = function(x) {
                  return(mean(x = expm1(x = x)))
                })
                log1p(avg.exp)
                pct.exp <- apply(X = data.use, MARGIN = 2, FUN = PercentAbove, 
                  threshold = 0)
                return(list(avg.exp = avg.exp, pct.exp = pct.exp))
            })
        names(x = data.plot) <- unique(x = data.features$id)
        if (cluster.idents) {
            mat <- do.call(what = rbind, args = lapply(X = data.plot, 
                FUN = unlist))
            mat <- scale(x = mat)
            id.levels <- id.levels[hclust(d = dist(x = mat))$order]
        }
        data.plot <- lapply(X = names(x = data.plot), FUN = function(x) {
            data.use <- as.data.frame(x = data.plot[[x]])
            data.use$features.plot <- rownames(x = data.use)
            data.use$id <- x
            return(data.use)
        })
        data.plot <- do.call(what = "rbind", args = data.plot)
        if (!is.null(x = id.levels)) {
            data.plot$id <- factor(x = data.plot$id, levels = id.levels)
        }
        ngroup <- length(x = levels(x = data.plot$id))
        if (ngroup == 1) {
            scale <- FALSE
            warning("Only one identity present, the expression values will be not scaled", 
                call. = FALSE, immediate. = TRUE)
        }
        else if (ngroup < 5 & scale) {
            warning("Scaling data with a low number of groups may produce misleading results", 
                call. = FALSE, immediate. = TRUE)
        }
        hs.browser("39.01.16.162200", debug = 0)
        data.features %=>% mean(expm1(x[x[, "id"] %in% "Th", 
            "ENSG00000011465"]))
        avg.exp.scaled <- sapply(X = unique(x = data.plot$features.plot), 
            FUN = function(x) {
                data.use <- data.plot[data.plot$features.plot == 
                  x, "avg.exp"]
                if (scale) {
                  data.use <- scale(x = data.use)
                  data.use <- MinMax(data = data.use, min = col.min, 
                    max = col.max)
                }
                else {
                  data.use <- log1p(x = data.use)
                }
                return(data.use)
            })
        avg.exp.scaled <- as.vector(x = t(x = avg.exp.scaled))
        if (split.colors) {
            avg.exp.scaled <- as.numeric(x = cut(x = avg.exp.scaled, 
                breaks = 20))
        }
        data.plot$avg.exp.scaled <- avg.exp.scaled
        data.plot$features.plot <- factor(x = data.plot$features.plot, 
            levels = features)
        data.plot$pct.exp[data.plot$pct.exp < dot.min] <- NA
        data.plot$pct.exp <- data.plot$pct.exp * 100
        if (split.colors) {
            splits.use <- vapply(X = as.character(x = data.plot$id), 
                FUN = gsub, FUN.VALUE = character(length = 1L), 
                pattern = paste0("^((", paste(sort(x = levels(x = object), 
                  decreasing = TRUE), collapse = "|"), ")_)"), 
                replacement = "", USE.NAMES = FALSE)
            data.plot$colors <- mapply(FUN = function(color, 
                value) {
                return(colorRampPalette(colors = c("grey", color))(20)[value])
            }, color = cols[splits.use], value = avg.exp.scaled)
        }
        color.by <- ifelse(test = split.colors, yes = "colors", 
            no = "avg.exp.scaled")
        if (!is.na(x = scale.min)) {
            data.plot[data.plot$pct.exp < scale.min, "pct.exp"] <- scale.min
        }
        if (!is.na(x = scale.max)) {
            data.plot[data.plot$pct.exp > scale.max, "pct.exp"] <- scale.max
        }
        if (!is.null(x = feature.groups)) {
            data.plot$feature.groups <- factor(x = feature.groups[data.plot$features.plot], 
                levels = unique(x = feature.groups))
        }
        hs.browser("39.01.16.161401", debug = 0)
        plot <- ggplot(data = data.plot, mapping = aes_string(x = "features.plot", 
            y = "id")) + geom_point(mapping = aes_string(size = "pct.exp", 
            color = color.by)) + scale.func(range = c(0, dot.scale), 
            limits = c(scale.min, scale.max)) + theme(axis.title.x = element_blank(), 
            axis.title.y = element_blank()) + guides(size = guide_legend(title = "Percent Expressed")) + 
            labs(x = "Features", y = ifelse(test = is.null(x = split.by), 
                yes = "Identity", no = "Split Identity")) + cowplot::theme_cowplot()
        if (!is.null(x = feature.groups)) {
            plot <- plot + facet_grid(facets = ~feature.groups, 
                scales = "free_x", space = "free_x", switch = "y") + 
                theme(panel.spacing = unit(x = 1, units = "lines"), 
                  strip.background = element_blank())
        }
        if (split.colors) {
            plot <- plot + scale_color_identity()
        }
        else if (length(x = cols) == 1) {
            plot <- plot + scale_color_distiller(palette = cols)
        }
        else {
            plot <- plot + scale_color_gradient(low = cols[1], 
                high = cols[2])
        }
        if (!split.colors) {
            plot <- plot + guides(color = guide_colorbar(title = "Average Expression"))
        }
        return(plot)
    }
    function(seurat_data, feature, id = {
    }, id_db = unique(id) %=>% sort) {
        if (!length(id)) 
            id <- Seurat::Idents(seurat_data)
        Seurat::Idents(seurat_data) <- hs.hsgly("seurat/函数/idents，获取.38_12_15_114046")(seurat_data = seurat_data, 
            id = id)
        DotPlot(object = seurat_data, assay = "RNA", features = feature) + 
            theme(axis.text.x = element_text(angle = 90, vjust = 0.05, 
                hjust = 0.75))
    }
})
"36_09_08_183156" <- local({
    db <- {
    }
    function(sj) {
        if (!length(db)) 
            db <<- hs.hsgly("线粒体基因.36.09.08.184137")()
        if ("percent.mt" %in% colnames(attr(sj, "meta.data"))) {
        }
        else {
            gene_id_type <- ifelse(all(hs.grepl(rownames(sj), 
                "^ENS")), "ensembl", "symbol")
            gene_mt <- rownames(sj) %and% db[, switch(gene_id_type, 
                symbol = gene.id_symbol, ensembl = gene.id_ensmebl)]
            sj %<>% PercentageFeatureSet(features = gene_mt, 
                col.name = "percent.mt")
        }
        sj
    }
})
"36_09_08_184137" <- local({
    db <- {
    }
    function() {
        if (!length(db)) 
            db <<- hs.wj(.mfp.swxx("36.09.08.184556"), "gene_mt.tsv.gz") %>% 
                {
                  if (file.exists(.)) {
                    wk.fread(.)
                  }
                  else {
                    gtf_wj <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1 = "human", 
                      what = "gtf")
                    a1 <- wk.fread(cmd = .sys("cat_wk", .sys.fpGood(gtf_wj), 
                      "|", .sys.perl.print("print if $F[ 2] eq \"gene\" and $F[ 0] eq \"MT\"")), 
                      header = FALSE)
                    a1[, `:=`("gene.id_ensmebl", hs.re(V9, "(?<=gene_id \")[^\"]+"))]
                    a1[, `:=`("gene.id_symbol", hs.re(V9, "(?<=gene_name \")[^\"]+"))]
                    jg <- a1[, .SD, .SDcols = patterns("^gene\\.")]
                    wk.dt.fwrite(jg, .)
                    jg
                  }
                }
        db
    }
})
"36_09_08_170130" <- function(sj) {
    cn <- hs.hsgly("filter.rna_n.36.09.08.164956")(sj, ot = "s")
    sj <- hs.hsgly("percent_mt.36.09.08.183156")(sj)
    attr(sj, "meta.data")[cn, ] %>% {
        x1 <- .[, "nCount_RNA"]
        x2 <- .[, "percent.mt"]
        x1 %<>% log1p
        cl <- dbscan::hdbscan(data.frame(x1, x2), minPts = 25)
        fit1 <- lowess(x1, x2, f = 0.05)
        i2 <- (cl[["cluster"]] %>% {
            . %in% names(sort(hs.table(.), decreasing = TRUE)[1])
        }) | (match(x1, fit1[["x"]]) %>% {
            x2 < fit1[["y"]][.]
        })
        rownames(.)[i2]
    }
}
"36_09_07_233753" <- function(sj) {
    cn <- hs.hsgly("filter.rna_n.36.09.08.164956")(sj, ot = "s")
    attr(sj, "meta.data")[cn, ] %>% {
        x1 <- .[, "nCount_RNA"]
        x2 <- .[, "nFeature_RNA"]
        x1 %<>% log1p
        x2 %<>% log1p
        if (0) {
            cl <- dbscan::hdbscan(data.frame(x1, x2), minPts = 25)
            cl[["cluster"]] == 2
            .[cl[["cluster"]] == 2, "nFeature_RNA"] %>% summary
        }
        fit1 <- lowess(x1, x2, f = 0.05)
        d1 <- match(x1, fit1[["x"]]) %>% {
            max(x2 - fit1[["y"]][.]) * 1.1
        }
        i2 <- match(x1, fit1[["x"]]) %>% {
            x2 > (fit1[["y"]][.] - d1)
        }
        return(rownames(.)[i2])
        sj__seurat.L[[sj1.name]] %>% dim
        sj__seurat.L[[sj1.name]][, i2] %>% dim
        if (0) 
            .pdf(expression({
                if (1) {
                  plot(x1, x2, type = "n")
                  i1 <- match(x1, fit1[["x"]]) %>% {
                    x2 > (fit1[["y"]][.] - d1)
                  }
                  points(x1[i1], x2[i1], col = "red", pch = ".")
                  i1 <- match(x1, fit1[["x"]]) %>% {
                    x2 <= (fit1[["y"]][.] - d1)
                  }
                  points(x1[i1], x2[i1], col = "gray", pch = ".")
                }
                if (0) {
                  plot(x1, x2)
                }
                lines(fit1, col = "red")
                lines(fit1[["x"]], fit1[["y"]] - d1, col = "pink")
            }), width = 20)
    }
}
"36_09_08_164956" <- function(sj, ot = "i") {
    attr(sj, "meta.data") %>% {
        i1 <- which(.[, "nFeature_RNA"] %>% {
            . > min(quantile(., 0.1), 200)
        } & .[, "nCount_RNA"] %>% {
            . > min(quantile(., 0.1), 1000)
        })
        switch(ot, i = i1, s = rownames(.)[i1])
    }
}
"38_12_15_114046" <- function(seurat_data, id = {
}, id_db_hs = function(x) {
    L <- levels(x) %and% paste(unique(x))
    if (!length(L)) 
        L <- unique(x)
    a <- as.numeric(L)
    if (anyNA(a)) {
        hs.str_sort(L)
    }
    else {
        if (0) {
            x <- hs.sample(seq(10), 100) %=>% paste
            x <- hs.sample(rnorm(10), 100) %=>% paste
            hs.factor(x, sort(unique(as.numeric(x))), N = 0) %=>% 
                hs.table
            hs.table(x) %=>% x[order(as.numeric(names(x)))]
        }
        L[order(a)]
    }
}) {
    id <- hs.switch(length(id), 0, SeuratObject::Idents(seurat_data), 
        1, seurat_data@meta.data[, id], id)
    hs.factor(id, id_db_hs(id), N = 0)
}
"2023_06_25_220926" <- function(fp = {
}, m1, cn.hs = {
}, xm = "proj") {
    on.exit(hs.gc_here(), add = TRUE)
    if (length(fp)) 
        m1 <- wk.fread(fp, ot = "m", rn.cn = 1) %=>% Matrix::Matrix(sparse = TRUE)
    if (length(cn.hs)) 
        colnames(m1) %<=>% cn.hs
    if (sum(m1[, 1]) < 1) 
        hs.browser("2023.06.27.203909", debug = 0)
    tpm <- hs.cond(hs.hsgly("is_tpm.2023_06_29_194530")(m1, data_type = "sc"), 
        m1, m1 %=>% hs.colMaxs %=>% {
            max(x) < 20
        }, {
            hs.anti_log <- m1[, 1] %=>% {
                hs2 <- function(x) hs.num_is_equal(sum(hs1(x)), 
                  10000)
                mean(x == 0)
                max(x)
                sum(exp(x) - 1)
                hs1 <- expm1
                if (hs2(x)) 
                  return(hs1)
                hs1 <- function(x) 2^x - 1
                if (hs2(x)) 
                  return(hs1)
                if (.wk("gzip -t", hs.fpGood(fp))) {
                  stop("[2023_06_30_105440 gzip文件损坏]", 
                    fp)
                }
                hs.browser("2023.06.27.203247", debug = 0)
            }
            hs.anti_log(m1) %=>% Matrix::Matrix(sparse = 1)
        }, m1[, 1] %=>% (sum(abs(x - round(x))) == 0), m1, hs.browser("2023.06.27.205942", 
            debug = 0))
    jg <- SeuratObject::CreateSeuratObject(counts = tpm, project = xm)
    jg[["RNA"]]@data %<=>% {
        Matrix::sparseMatrix(1, 1, dims = dim(x), dimnames = dimnames(x)) %=>% 
            as("dMatrix")
    }
    if (0) {
        x <- matrix(0, 3, 4)
        Matrix::sparseMatrix(1, 1, dims = dim(x), dimnames = dimnames(x)) %=>% 
            as("dMatrix")
        jg %<=>% Seurat::NormalizeData(x)
        jg[["RNA"]]@counts <- as(matrix(0), "dgCMatrix")
        jg[["RNA"]]@data <- as(matrix(0), "dgCMatrix")
        help(package = "Matrix")
        `?`(Matrix::sparseMatrix)
        a <- Matrix::sparseMatrix(1, 1, dims = dim(jg[["RNA"]]@counts)) %=>% 
            as("dgCMatrix")
        dim(jg[["RNA"]]@counts)
        jg[["RNA"]]@counts %=>% dim
        jg[["RNA"]]@counts <- jg[["RNA"]]@data
        os("jg")
        abs(jg[["RNA"]]@data[, 1] - log1p(tpm[, 1])) %=>% mas
    }
    jg
}
"37_12_17_110130" <- function() {
}
"37_12_09_125141" <- function() {
    hs.require(c("Seurat", "dplyr", "patchwork"))
    wk.install_github("umap-learn", conda = "~/miniconda3/bin/conda")
    hs.require("reticulate")
    reticulate::py_install("umap-learn", method = "conda", conda = "~/miniconda3/bin/conda")
}
"38_04_07_003035" <- function() {
}
"2023_06_09_170446" <- function(si_fp, fq_is_clean = 0, prj_root = ".", 
    ...) {
    fp.r <- hs.wjj.ls(prj_root, paste0("^文件夹\\.", wk.muid.pattern, 
        ".*\\.r(\\.gz)?$"))
    if (!length(fp.r)) 
        fp.r <- hs.fp(prj_root, "文件夹.", paste0(.muid4machine(), 
            ".r"))
    hs1 <- bquote(function(sub = {
    }) {
        jg <- .(prj_root)
        if (length(sub)) 
            jg <- hs.fp(jg, sub)
        jg
    })
    c("##", deparse(hs1)) %=>% cat(sep = "\n", file = fp.r %=>% 
        hs.wj)
    si <- wk.fread(si_fp)
    if (!fq_is_clean) {
        dirname(hs.hsgly("启动.2023_06_09_170446", PO = 1)) %=>% 
            .ls
        hs_uid <- "fq/clean/2023_06_09_164805"
        hs.sjgly(hs_uid)
        hs.with_wd(prj_root, {
            hs.hsgly("fq/clean/2023_06_09_164805")(fq_fp)
        })
        hs.fp(prj_root, hs.sjgly("2023_06_09_164805"))
        hs.sjgly("2023_06_09_164805")
        wjj2_here <- xxx
        hs.fp(prj_root, )
        si
        hs.hsgly(hs_uid)
    }
}
"2023_06_09_165618" <- function() {
    "."
}
"2023_06_09_164805" <- function(fq_fp, fq_is_clean = 0) {
    wjj2 <- hs.sjgly("2023_06_09_164805")
    if (!fq_is_clean) {
    }
}
"35_08_06_153838" <- function(wjIn, wjjOut = dirname(wjIn[1]), 
    yb1 = hs.hsgly("wjm2ybm.35.08.20.200110")(wjIn[1]), nc = 9, 
    Q = 20) {
    sj <- hs.hsgly("fq.clean.wj.35.08.20.191126")(wjIn = wjIn, 
        yb1 = yb1, wjj2 = wjjOut)
    trimmomatic.wjj <- hs.hsgly("rj.find.35.08.20.070052")("trimmomatic")
    trimmomatic.wj <- hs.listWj(trimmomatic.wjj, "trimmomatic-[0-9\\.]+.jar", 
        R = 1)
    adapter.wjj <- trimmomatic.wj %>% dirname %>% hs.fp("adapters")
    rl1 <- hs.hsgly("fq.readLength.35.07.15.162917")(wjIn[1])
    if (0) 
        headcrop <- c(mRNA = 0, DNA = 0, miRNA = 0, KC = 12)
    headcrop <- 0
    paired <- length(wjIn) > 1
    if (!file.exists(sj[["clean"]][1])) {
        .wk("java", "-jar", trimmomatic.wj, ifelse(sj[["paired"]], 
            "PE", "SE"), "-threads", nc, .sys.fpGood(sj[["raw"]]), 
            .sys.fpGood(sj[["clean"]]), paste0("ILLUMINACLIP:", 
                normalizePath(adapter.wjj), "/", ifelse(paired, 
                  "TruSeq3-PE-2.fa:2:30:10:1:true", "TruSeq3-SE.fa:2:30:10")), 
            "LEADING:3", "TRAILING:3", paste0("SLIDINGWINDOW:4:", 
                15), if (headcrop) 
                paste0("HEADCROP:", headcrop), paste0("AVGQUAL:", 
                Q), paste0("MINLEN:", max(floor(rl1/2), 25)))
        if (0) {
            .sys.cat.skip("clean/SRR2097343.R1.paired.fq.gz")
            a1 <- "clean/SRR2099756.R2.paired.fq.gz" %>% .sys.cat.skip %>% 
                pipe %>% readLines(40000)
            a1[seq(2, length(a1), by = 4)] %>% nchar %>% summary
            a1 <- .wk("samtools", "view", "mapping/bam/SRR2097343.bam", 
                "| cut -f9", "| head -n", 10000, intern = 1)
            summary(abs(as.integer(a1)))
        }
    }
    sj
}
"37_01_08_164853" <- function(wjj1 = ".", search_clean_wjj = 1) {
    if (search_clean_wjj) {
        wjj1 <- hs.wjj.ls(wjj1, "(?i)clean", wjj = 1)
        if (length(wjj1) != 1) 
            hs.browser("37.01.08.165135")
    }
    wj <- fq_wj.hs.wjj(wjj1)
    hs.hsgly("sample.fq_wj.hs.fq_wj.37_01_11_003552")(wj)
}
"37_01_11_003552" <- function(fq_wj) {
    fq_wj %>% {
        tapply(., hs.sub(basename(.), "(?i)([_\\.]r[12])?(\\.(paired|clean))?\\.fq\\.gz$") %>% 
            basename, c)
    }
}
"37_05_26_164900" <- function(cfg.xlsx) {
    cfg.xlsx
}
"36_04_05_182133" <- local({
    wjj1 <- {
    }
    gtf.hs <- function(wj) {
        gtf.wj <- hs.wjm(wj, ext = "gtf", extN = 2, dn = wjj1)
        if (!file.exists(gtf.wj)) {
            ci2drop <- match(c("#bin", "milliDiv", "milliDel", 
                "milliIns", "genoLeft", "repStart", "repEnd", 
                "repLeft", "id"), unlist(strsplit(readLines(wj, 
                1), "\t")))
            repeats_ucsc <- fread(wj, drop = ci2drop)
            {
                repeat.class.wj <- hs.wjm(wj, append = ".class", 
                  extN = 2)
                if (!file.exists(repeat.class.wj)) 
                  repeats_ucsc[, unique(.SD), .SDcols = c("repName", 
                    "repClass", "repFamily")] %>% wk.dt.fwrite(repeat.class.wj)
            }
            setnames(repeats_ucsc, "genoName", "V1")
            repeats_ucsc[, `:=`("V1", hs.sub(V1, "^chr"))]
            repeats_ucsc[, `:=`("V2", paste0(hs.re(basename(wj), 
                "\\w+"), "_rmsk"))]
            repeats_ucsc[, `:=`("V3", "exon")]
            repeats_ucsc[, `:=`("genoStart", genoStart + 1)]
            setnames(repeats_ucsc, "genoStart", "V4")
            setnames(repeats_ucsc, "genoEnd", "V5")
            setnames(repeats_ucsc, "swScore", "V6")
            setnames(repeats_ucsc, "strand", "V7")
            repeats_ucsc[, `:=`("V8", ".")]
            repeats_ucsc[, `:=`("V9", sprintf("gene_id \"%s\"; transcript_id \"%s____%s_%s_%s\"; family_id \"%s\"; class_id \"%s\";", 
                repName, repName, V1, V4, V5, repFamily, repClass))]
            repeats_ucsc[, `:=`(c("repName", "repClass", "repFamily"), 
                {
                })]
            setcolorder(repeats_ucsc, names(repeats_ucsc) %>% 
                hs.re("\\d+") %>% as.integer %>% sort %>% paste0("V", 
                .))
            wk.dt.fwrite(repeats_ucsc, gtf.wj, col.names = FALSE, 
                quote = FALSE)
        }
        gtf.wj
    }
    function(org1) {
        if (!length(wjj1)) 
            wjj1 <<- .mfp.swxx("36.04.06.104457") %>% hs.wjj
        switch(org1, grcm38 = {
            gtf.hs("~/in2/swxx/sj/db/repeats/ucsc_gonomeBrowser/mouse/mm10.repeats.tsv.gz")
        }, grch38 = {
            gtf.hs("~/in2/swxx/sj/db/repeats/ucsc_gonomeBrowser/human/hg38.repeats.tsv.gz")
        }, hs.browser("36.04.06.111548"))
    }
})
"36_04_21_132815" <- function(bam.wj, gv = "grch38") {
    n1 <- "RPKM"
    ignoreDuplicates <- 0
    bs <- 50
    rl <- hs.hsgly("bam.read_length.36.05.05.223906")(bam.wj)
    egs <- switch(paste0(gv, ".", rl), grch38.150 = 2862010578, 
        hs.browser("36.05.05.215126"))
    sapply(bam.wj, function(wj1) {
        bai.wj <- hs.wjm(wj1, add = "bai")
        if (file.mtime(bai.wj) < file.mtime(wj1)) {
            .cat.time()
            message("【samtools index】", wj1)
            .wk("samtools index -@ 4", wj1)
            .cat.time()
        }
        wj2 <- hs.wjm(wj1, append = paste0(".normalizeUsing=", 
            n1, "_bs=", bs, "_ignoreDuplicates=", ifelse(ignoreDuplicates, 
                1, 0), "_effectiveGenomeSize=", egs), ext = "bw")
        if (!file.exists(wj2)) 
            .wk("bamCoverage", "-b", .sys.fpGood(wj1), "-o", 
                .sys.fpGood(wj2), "-p max/2", "--binSize", bs, 
                if (ignoreDuplicates) 
                  "--ignoreDuplicates", "--effectiveGenomeSize", 
                egs, "--normalizeUsing", n1)
        wj2
    })
}
"36_01_31_215610" <- function(de = if (length(de.wj)) wk.fread(de.wj), 
    de.wj = {
    }, gene.cn = gene_id_name.37_04_29_145914(names(de)), log2fc.cn = hs.grepV(names(de), 
        "(?i)l(og2|og)?F(old)?C(hange)?"), fc.cn = {
    }, ft = guess_gene_id_type.37_04_28_165736(de[[gene.cn]]), 
    q.cn.db = {
    }, p.cn = hs.grepV(names(de), "(?i)p[_\\-\\.]?v(al)?")[1], 
    stat.cn = {
    }, wjj1 = if (length(de.wj)) {
        hs.wjj(hs.wjm(de.wj, ext = 0, extN = if (hs.grepl(de.wj, 
            "\\.(xls(x)?|tsv|csv|txt)$")) {
            1
        }
        else if (hs.grepl(de.wj, "\\.(tsv|)\\.gz$")) {
            2
        }
        else hs.browser("36.12.22.195425")))
    } else "~/tmp", q.co = 0.05, deg = 1, gs = 1, u1 = {
    }, u1.name = {
    }, org1 = "human", tpm1 = 0, top = 0, nc = hs.NC(), RcisTarget = 1) {
    hs.hsgly("genes.understand.35.11.12.182333")
    if (!length(gene.cn)) 
        gene.cn <- hs.grepV(names(de), "gene([\\._]?id)?")
    1
    {
        id <- de[[gene.cn]] %>% hs.grepV(";")
        de[[gene.cn]]
        a <- de[, get(gene.cn) %>% {
            hs.hsgly("update.gene_id.37_04_26_114304")(.)
        }]
        de[, `:=`(c(gene.cn), get(gene.cn) %>% {
            hs.hsgly("update.gene_id.37_04_26_114304")(.)
        })]
    }
    pv.cn <- if (length(q.cn.db)) 
        q.cn.db[1]
    else hs.clean(pv.hs.str(names(de))[1])
    if (length(pv.cn) && de[, NA %in% get(pv.cn)]) {
        de <- de[!is.na(get(pv.cn))]
    }
    if (length(pv.cn)) 
        de <- de[order(get(pv.cn))]
    if (tpm1) {
        hs.browser("36.06.12.192405")
    }
    en_wjj <- paste(c(wjj1, paste(c("en", if (length(u1.name)) u1.name), 
        collapse = "_"), if (tpm1) paste0("tpm,", tpm1)), collapse = "/")
    log2fc <- if (length(fc.cn)) 
        de[, log2(get(fc.cn))]
    else de[[log2fc.cn]]
    exp.L <- expandingList()
    if (!length(q.cn.db)) 
        q.cn.db <- hs.grepV(names(de), "(?i)padj|qv|fdr")
    if (gene.cn %not.in% names(de)) 
        gene.cn <- hs.grepV(names(de), "(?i)gene(.?id)?")[1]
    de <- de[, .SD, .SDcols = c(gene.cn, log2fc.cn, fc.cn, q.cn.db, 
        stat.cn) %and% names(de)]
    if (ft != "ncbi") {
        gene.cn2 <- "gene_id,ncbi"
        switch(org1, human = hs.hsgly("gene.id.converter.using_hgnc.36.01.22.160941")(de, 
            gene.cn, gene.cn2, ft, "ncbi"), {
            id1.id2 <- hs.hsgly("编号种类转换库.37_03_29_103455")(ft, 
                "ncbi", org1 = org1)
            cn1 <- names(id1.id2)[1]
            id1.id2[, `:=`(c(cn1), get(cn1) %>% toupper)]
            id1.id2 %<>% {
                hs.c(.[[2]], as.list(.[[1]]))
            } %>% inverseList %>% sapply(paste, collapse = ";")
            de[get(gene.cn) %in% names(id1.id2), `:=`("gene_id,ncbi", 
                id1.id2[get(gene.cn)])]
            de[get(gene.cn) %>% hs.grep(";"), `:=`("gene_id,ncbi", 
                strsplit(get(gene.cn), ";") %>% sapply(. %>% 
                  {
                    jg <- hs.clean(id1.id2[.]) %>% paste(collapse = ";")
                    ifelse(jg == "", NA, jg)
                  }))]
        })
        ft <- "ncbi"
        gene.cn <- gene.cn2
    }
    if (de[, !is.character(get(gene.cn))]) 
        de[, `:=`(c(gene.cn), as.character(get(gene.cn)))]
    if (gs) {
        1
        {
            rnk <- de[hs.clean(get(gene.cn), ot = "l"), .(get(gene.cn), 
                hs.deInf(get(log2fc.cn)))]
            {
                i1 <- rnk[[2]] == 0
                if (any(i1)) 
                  rnk <- rnk[!i1]
            }
            {
                wjj2 <- hs.wjj(hs.wjm(de.wj, ext = 0, extN = 2), 
                  "en/gsea_ucsd,broad/log2fc")
                local({
                  wjj.36_03_04_003344 <- getwd()
                  on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
                  setwd(wjj2)
                  hs.hsgly("gene.enrich_rank_ucsd_broad.37_04_01_164138")(rnk, 
                    ".", org1 = org1)
                })
            }
            if (0) {
                exp.L[["add"]](bquote(hs.hsgly("genes.understand.35.11.12.182333")(rnk = .(rnk), 
                  wj_out.pre = .(hs.wjj(wjj1, "en", "gsea", "log2fc/")), 
                  org1 = .(org1), ft = "ncbi")))
            }
        }
        if (0) {
            if (length(p.cn)) {
                rnk <- de[, .(get(gene.cn), hs.deInf(log10(get(p.cn)) * 
                  ifelse(log2fc > 0, -1, 1)))]
                exp.L[["add"]](bquote(hs.hsgly("genes.understand.35.11.12.182333")(rnk = .(rnk), 
                  wj_out.pre = .(hs.wjj(wjj1, "en", "gsea", "log10pv/")), 
                  org1 = .(org1), ft = .(ft))))
            }
            if (length(stat.cn)) {
                rnk <- de[, .(get(gene.cn), get(stat.cn))]
                exp.L[["add"]](bquote(hs.hsgly("genes.understand.35.11.12.182333")(rnk = .(rnk), 
                  wj_out.pre = .(hs.wjj(wjj1, "en", "gsea", "stat/")), 
                  org1 = .(org1), ft = .(ft))))
            }
        }
    }
    if (deg) {
        if (length(u1)) {
            if (length(u1.name) == 0) 
                stop("【36.05.19.162659】需要文件名")
        }
        else u1 <- hs.clean(de[, get(gene.cn) %>% as.character] %>% 
            strsplit(";+") %>% unlist)
        debug <- 0
        i1 <- is.na(de[[q.cn.db[1]]])
        if (any(i1)) {
            hs.browser("36.06.04.161205")
        }
        if (!length(q.cn.db)) 
            hs.stop("@36.06.04.120919@ 没有找到qv列")
        if (0) {
            q.cn.2do <- paste0("padj_fc", c(1.2, 1.5, 2)) %and% 
                q.cn.db
            if (length(q.cn.2do)) 
                for (q1.cn in q.cn.2do) {
                  i1 <- de[, get(q1.cn) < q.co]
                  for (fx1 in c("dif", "up", "down")) {
                    message("\t", "【", q1.cn, ";", fx1, "】")
                    s1 <- de[i1 & switch(fx1, dif = 1, up = get(log2fc.cn) > 
                      0, down = get(log2fc.cn) < 0), get(gene.cn)]
                    hs.hsgly("genes.understand.35.11.12.182333")(s1, 
                      u1, ft = "ncbi", wj_out.pre = hs.wj(en_wjj, 
                        q1.cn, q.co, fx1), org1 = org1)
                  }
                }
        }
        {
            if (length(q.cn.db) != 1) {
                message("[36.06.04.122200]qv不止1列: ", paste(q.cn.db, 
                  collapse = "; "), "。用", q.cn.db[1], "。")
            }
            q.cn <- q.cn.db[1]
            s1.db <- list(up = de[get(log2fc.cn) > 0, hs.clean(get(gene.cn) %>% 
                strsplit(";+") %>% unlist, U = 1)], down = de[get(log2fc.cn) < 
                0, hs.clean(get(gene.cn) %>% strsplit(";+") %>% 
                unlist, U = 1)])
            fc.co.db <- c(1.2, 1.5, 2, 1)
            if (RcisTarget && (org1 %in% c("human", "mouse", 
                "fruitfly"))) {
                if (org1 %in% "fruitfly") {
                  message("待完善")
                  hs.browser("37.04.01.150246")
                }
                jg.wj.37_04_01_160518 <- hs.fp(en_wjj, "tf_target.xlsx")
                if (file.exists(jg.wj.37_04_01_160518)) {
                }
                else {
                  hs.require("RcisTarget")
                  RcisTarget.motifRankings.file <- local({
                    wj <- "~/in2/swxx/sj/db/cistarget/databases" %>% 
                      hs.wjj.ls("\\.feather$", R = 1)
                    names(wj) <- hs.re(basename(wj), "[^_]+")
                    wj[switch(org1, human = "hg38", mouse = "mm10")]
                  })
                  motifRankings <- importRankings(RcisTarget.motifRankings.file)
                  motifAnnotations.name <- paste0("motifAnnotations_", 
                    switch(org1, human = "hgnc", mouse = "mgi"))
                  evalbq(data(.(as.symbol(motifAnnotations.name))))
                  gt <- switch(org1, human = "symbol", mouse = "mgi_symbol", 
                    hs.browser("37.04.01.101302"))
                  gene_list <- hs.hsgly("差异表达的.37_04_02_092139")(de, 
                    gene.cn = gene.cn, ft = "ncbi", tt = gt, 
                    org1 = org1)
                  hs.hsgly("tf.using_RcisTarget.37_04_01_170603")(gene_list, 
                    org1 = org1, wjj2 = en_wjj, ft = gt)
                  if (0) {
                    jg <- cisTarget(gene_list, motifRankings, 
                      motifAnnot = get(motifAnnotations.name))
                    jg %>% wk.xls.write(hs.fp(en_wjj, "tf_target.xlsx"))
                  }
                }
            }
            for (fc.co in fc.co.db) {
                i1 <- de[, get(q.cn) < q.co] & (abs(log2fc) > 
                  log2(fc.co))
                for (fx1 in c("dif", "up", "down")) {
                  message("\t", "【fc.co,", fc.co, ";", fx1, 
                    "】")
                  s1 <- de[i1 & switch(fx1, dif = 1, up = get(log2fc.cn) > 
                    0, down = get(log2fc.cn) < 0), get(gene.cn) %>% 
                    strsplit(";+") %>% unlist %>% hs.clean(U = 1)]
                  s1_all <- switch(fx1, dif = s1.db, s1.db[[fx1]])
                  wj_out.pre <- hs.wj(en_wjj, "padj_fc1", q.co, 
                    "fc", fc.co, fx1)
                  exp.L[["add"]](bquote(hs.hsgly("genes.understand.35.11.12.182333")(.(s1), 
                    .(u1), s1_all = .(s1_all), ft = "ncbi", wj_out.pre = .(wj_out.pre), 
                    org1 = .(org1))))
                  if (fc.co == 1) {
                    exp.L[["add"]](bquote(hs.hsgly("genes.understand.35.11.12.182333")(.(s1), 
                      .(u1), s1_all = .(s1_all), wj_out.pre = .(wj_out.pre), 
                      org1 = .(org1), top = .(top), wj_out.pre.4top = .(hs.wj(en_wjj, 
                        fx1, "top")), ft = .(ft))))
                  }
                }
            }
        }
    }
    if (1) {
        jg.wjj <- exp.L[["get"]]() %>% {
            lapply(seq_along(.), function(i1) {
                message("[37_04_01_160712]", i1)
                eval(.[[i1]])
            })
        }
    }
    if (0) 
        jg.wjj <- wk.mclapply(exp.L[["get"]](), eval, mc.cores = nc)
    if (1 && deg) {
        message("准备 kegg pathview 的数据")
        wjj1 <- jg.wjj[[1]]
        rnk <- de[hs.clean(get(gene.cn), ot = "l", dirty.extra = ""), 
            hs.c(get(gene.cn), hs.deInf(log2fc))]
        wk.mclapply(jg.wjj, function(wjj1) {
            wj2do <- hs.wjj.ls(wjj1, "(?i)kegg\\.tsv\\.gz$") %>% 
                hs.grepV("(?i)msigdb", inv = 1)
            for (wj1 in wj2do) {
                message(wj1)
                en1 <- fread(wj1)
                qv.cn <- hs.grepV(names(en1), "(?i)qv")
                pw_i <- en1[, which(get(qv.cn) < 0.1)]
                if (length(pw_i)) 
                  hs.wjj(hs.wjm(wj1, ext = 0, extN = 2), go = 1)
                en.gene.cn <- c("geneID", "core_enrichment") %and% 
                  names(en1)
                for (pw1i in pw_i) {
                  pw1 <- en1[pw1i, ID]
                  sj.wj <- paste0(pw1, ".pathview.rd")
                  if (file.exists(sj.wj)) 
                    next
                  gene <- en1[pw1i, hs.re(strsplit(get(en.gene.cn), 
                    "/")[[1]], "\\d+")]
                  geneList <- rnk[gene]
                  hs.save(geneList, pw1, org1, wj = sj.wj)
                }
            }
        })
        if (0) 
            lapply(jg.wjj, function(wjj1) {
                wj2do <- hs.wjj.ls(wjj1, "(?i)kegg\\.tsv\\.gz$")
                lapply(wj2do, function(wj1) {
                  hs.wjj(hs.wjm(wj1, ext = 0, extN = 2), go = 1)
                  hs.wjj.ls(".", "pathview\\.rd$") %>% normalizePath
                })
            }) %>% unlist %>% hs.wj.rm
    }
}
"36_01_30_215755" <- local({
    hs.add.group.mean <- function(de) {
        pd1 <- length(hs.grep(names(de), yb.info[, levels(condition)[2]])) == 
            0
    }
    hs.add.fc <- function(de) {
        cn <- "foldChange"
        if (cn %not.in% names(de)) {
            de[, `:=`(c(cn), 2^log2FoldChange)]
            wk.dt.col.reorder(de, cn, after = "log2FoldChange")
            1
        }
        else 0
    }
    function(X, gn, control, wjj1, condition = {
    }, condition.extra = {
    }, shrink.log2fc = 1, nc = 8, output.exp = 1, fc.co.v = head(fc.co.db, 
        fc.co.n), fc.co.db = c(1, 1.2, 1.5, 2, 4, 8), fc.co.n = 1, 
        de.wj = {
        }, dds.wj = {
        }, dds.redo = 0, add.symbol = 0, debug = 0, focus = 1) {
        hs.require(c("DESeq2", "IHW", "BiocParallel", "apeglm"))
        if (nc > 1) {
            register(MulticoreParam(nc))
        }
        if (length(control) > 1) {
            if (length(condition) && (length(condition) != length(condition))) {
                stop("36.02.29.231406")
            }
        }
        if (!length(condition)) 
            condition <- vector("list", length(control))
        comparison.seen <- c()
        for (i1 in seq_along(control)) {
            control_non <- gn %not% control[i1]
            comparison <- cbind(control[i1], control_non) %>% 
                apply(1, . %>% sort %>% paste(collapse = "."))
            if (!length(condition[[i1]])) 
                condition[i1] <- list(control_non[comparison %not.in% 
                  comparison.seen])
            if (length(condition[[i1]])) 
                comparison.seen %<>% (`%or%`(condition[[i1]]))
        }
        de.wj <- sapply(seq_along(control), function(control1.i) {
            control1 <- control[control1.i]
            condition <- condition[[control1.i]]
            if (!length(condition)) 
                return()
            if (0) 
                if (focus) {
                  hs.browser("36.06.13.110650")
                  i1 <- gn %in% c(control1, condition)
                  if (!all(i1)) {
                    gn %<>% {
                      .[i1]
                    }
                    X <- X[, i1, drop = 0]
                  }
                }
            yb.info <- data.table(id = colnames(X), condition = gn %>% 
                factor %>% relevel(control1))
            if (length(condition.extra)) {
                hs.browser("36.10.06.213807")
                yb.info[, `:=`(names(condition.extra), condition.extra)]
            }
            if (length(dds.wj) == 0) 
                dds.wj <- hs.wj(wjj1, "dds.rd")
            de.wj <- condition %>% {
                hs.c(., hs.wj(wjj1, paste0(., "_vs_", control1), 
                  "de.tsv.gz"))
            }
            load.dds <- !all(file.exists(de.wj))
            exp.wj <- c("exp", "counts_normalized") %>% {
                hs.c(., hs.wj(wjj1, paste0(., ".tsv.gz")))
            }
            if (output.exp) {
                if (all(file.exists(exp.wj))) {
                  output.exp <- 0
                }
                else load.dds <- 1
            }
            if ((!load.dds) && (!debug)) 
                return(de.wj)
            if (exists("dds", inherits = FALSE)) {
            }
            else if (file.exists(dds.wj) && (!dds.redo)) {
                dds <- dds.wj %>% hs.load
            }
            if ((!exists("dds", inherits = FALSE)) || (colData(dds)[, 
                "condition"] %>% levels %>% {
                .[1] != control1
            })) {
                dds <- eval(parse(text = paste0("DESeqDataSetFromMatrix( X, yb.info, design = ~ ", 
                  paste(names(yb.info)[-1], collapse = " + "), 
                  ")")))
                dds %<>% DESeq(parallel = nc > 1)
            }
            if (!file.exists(dds.wj)) 
                hs.save(dds, wj = dds.wj)
            if (length(rownames(dds))) {
                gene.cn <- ifelse(rownames(dds) %>% sample(min(9, 
                  nrow(dds))) %>% hs.grepl("^(?i)ens") %>% all, 
                  "gene.id_ensembl", "gene.id_symbol")
                if (gene.cn == "gene.id_ensembl") 
                  add.symbol <- 1
                if (output.exp) {
                  exp1.wj <- exp.wj["exp"]
                  if (!file.exists(exp1.wj)) {
                    a1 <- assay(do.call(ifelse(nrow(dds) < 2000, 
                      "rlog", "vst"), list(dds, blind = FALSE))) %>% 
                      as.data.table(gene.cn)
                    wk.dt.fwrite(a1, exp1.wj)
                  }
                  if (0) {
                    exp1.wj <- exp.wj["exp_blind"]
                    if (!file.exists(exp1.wj)) {
                      a1 <- assay(do.call(ifelse(nrow(dds) < 
                        2000, "rlog", "vst"), list(dds, blind = !0))) %>% 
                        as.data.table(1)
                      setnames(a1, "rn", gene.cn)
                      wk.dt.fwrite(a1, exp1.wj)
                    }
                  }
                  exp1.wj <- exp.wj["counts_normalized"]
                  if (!file.exists(exp1.wj)) {
                    a1 <- counts(dds, normalized = TRUE)
                    a2 <- as.data.table(a1, keep.rownames = gene.cn)
                    wk.dt.fwrite(a2, exp1.wj)
                  }
                }
            }
            for (de.i in seq_along(de.wj)) {
                de1.wj <- de.wj[de.i]
                if (file.exists(de1.wj)) {
                  de <- fread(de1.wj, nrow = 1)
                  if (!(paste0("padj_fc", fc.co.v) %all.in% names(de))) {
                    hs.rm.wj(de1.wj)
                  }
                }
                if (!file.exists(de1.wj)) {
                  condition1 <- names(de.wj)[de.i]
                  comparison1.i <- match(condition1, yb.info[, 
                    levels(condition)])
                  comparison1 <- resultsNames(dds)[comparison1.i]
                  hs.cat("\n【", comparison1, "】")
                  deg <- fc.co.v %>% lapply(. %>% {
                    fc.co <- .
                    hs.cat("\t【fold change cutoff】", fc.co)
                    a1 <- results(dds, contrast = c("condition", 
                      condition1, control1), lfcThreshold = log(fc.co, 
                      2), filterFun = ihw, parallel = nc > 1)
                  })
                  names(deg) <- fc.co.v
                  r1 <- as.data.table(deg[[1]])
                  if (length(rownames(dds))) 
                    r1[, `:=`(c(gene.cn), rownames(deg[[1]]))]
                  pq.cn <- c("pvalue", "padj")
                  pq.cn %>% {
                    setnames(r1, ., paste0(., "_fc1"))
                  }
                  r1[, `:=`(c("stat", "weight"), {
                  })]
                  if (length(deg) > 1) 
                    for (cn1 in pq.cn) r1[, `:=`(paste0(cn1, 
                      "_fc", names(deg)[-1]), deg[-1] %>% lapply("[[", 
                      cn1))]
                  wk.dt.col.reorder(r1, cn.pattern = c("^pvalue", 
                    "^padj"))
                  if (shrink.log2fc) {
                    deg2 <- lfcShrink(dds, coef = comparison1.i, 
                      type = "apeglm", parallel = nc > 1)
                    r1[, `:=`("log2FoldChange_shrunken_apeglm", 
                      deg2[, "log2FoldChange"])]
                  }
                  p.cn <- "pvalue_fc1"
                  if (length(rownames(dds))) 
                    setorderv(r1, p.cn, na.last = TRUE)
                  if (add.symbol) 
                    hs.hsgly("de_tbl.add.symbol.36.01.31.231208")(r1)
                  hs.add.fc(r1)
                  if (0) {
                    r1[gene.id_symbol %in% c("SAGE1")]
                    r1[(padj_fc1 < 0.05) & (abs(log2FoldChange) > 
                      log2(1.2)), sign(log2FoldChange) %>% hs.table]
                    u1 <- r1[(padj_fc1 < 0.05) & (abs(log2FoldChange) > 
                      log2(1.2)), gene.id_ensembl]
                    s1 <- r1[(padj_fc1 < 0.05) & (log2FoldChange > 
                      log2(1.2)), gene.id_ensembl]
                    fisher.test.ele(u1, s1, cta)
                    s1 <- r1[(padj_fc1 < 0.05) & (log2FoldChange < 
                      -log2(1.2)), gene.id_ensembl]
                    fisher.test.ele(u1, s1, cta)
                    u1 <- r1[hs.clean(padj_fc1, ot = "i"), gene.id_ensembl]
                    s1 <- r1[(padj_fc1 < 0.05) & (log2FoldChange > 
                      log2(1.2)), gene.id_ensembl]
                    fisher.test.ele(u1, s1, cta)
                    s1 <- r1[(padj_fc1 < 0.05) & (log2FoldChange < 
                      -log2(1.2)), gene.id_ensembl]
                    a1 <- hs.hsgly("gene.id.converter.using_hgnc.36.01.22.160941")(r1[(padj_fc1 < 
                      0.05) & (log2FoldChange > log2(1.2)), .(gene.id_ensembl %and% 
                      cta)], "V1", "V2", "ensembl", "symbol")
                    as.data.frame(a1) %>% .s
                  }
                  {
                    counts_normalized <- counts(dds, normalized = TRUE) %>% 
                      as.data.table
                    if (length(rownames(dds))) 
                      counts_normalized[, `:=`(c(gene.cn), rownames(dds))]
                    for (c1 in c(control1, names(de.wj)[de.i])) {
                      yb <- yb.info[which(as.matrix(condition) == 
                        c1), id]
                      a1 <- counts_normalized[, rowMeans(.SD), 
                        .SDcols = c(yb)]
                      if (length(rownames(dds))) {
                        names(a1) <- counts_normalized[, get(gene.cn)]
                        a1 <- r1[, a1[get(gene.cn)]]
                      }
                      r1[, `:=`(c(paste0("mean.", c1)), a1)]
                      a1 <- counts_normalized[, rowMedians(as.matrix(.SD)), 
                        .SDcols = c(yb)]
                      if (length(rownames(dds))) {
                        names(a1) <- counts_normalized[, get(gene.cn)]
                        a1 <- r1[, a1[get(gene.cn)]]
                      }
                      r1[, `:=`(c(paste0("median.", c1)), a1)]
                    }
                  }
                  wk.dt.col.reorder(r1, cn.pattern = c(if (length(rownames(dds))) "(?i)^gene", 
                    "(?i)mean", "(?i)median", "(?i)foldchange"), 
                    before = "lfcSE")
                  i1 <- if (length(rownames(dds))) 
                    r1[, which(is.na(get(p.cn)))]
                  wk.dt.fwrite(r1, de1.wj, i1)
                }
            }
            de.wj
        })
        names(de.wj) <- control
        de.wj
    }
})
"36_03_23_144833" <- function(wjj0 = ".", group.hs = NA, control = {
}, condition = {
}, outliers = list({
}), bam.wj = hs.wjj.ls(file.path(wjj0, "mapping"), "\\.bam$", 
    R = 1), pca = !de, de = length(control), en = de, org1 = "human", 
    gv = switch(org1, human = "grch38", mouse = "grcm38"), gtf.type = "gene", 
    rpt = 0, sj.name = if (rpt) "gene+repeat,non_overlapping" else "gene", 
    gtf.wj = switch(gv, grch38 = "~/swxx/sj/db/ensembl/human/Homo_sapiens.GRCh38.100.gtf", 
        grcm38 = "~/swxx/sj/db/ensembl/mouse/release/102_38/Mus_musculus.GRCm38.102.gtf", 
        grcm39 = "~/swxx/sj/db/ensembl/ftp.ensembl.org/pub/release-105/gtf/mus_musculus/Mus_musculus.GRCm39.105.gtf", 
        hs.browser("36.12.15.154726")), pca.wj.name = {
    }, group.drop = {
    }, yb.drop = {
    }, group.2do = {
    }, yb.2do = {
    }, gene_merge_to = {
    }, gene_to_be_merged = {
    }, only = {
    }, dds.redo = 0, gc.redo = 0, counts.wj = NULL, tpm = 1, 
    focus = 1, featureCounts.A = {
    }, ...) {
    hs.with_wd(hs.wjj(wjj0), {
        if (!length(counts.wj)) {
            counts.wj <- c("gene", if (rpt) "repeat") %>% {
                hs.c(., lapply(., . %>% {
                  gtf.type <- .
                  hs.lapply("om,OM,Om", . %>% {
                    hs.hsgly("xm.bam.to_expression.35.08.13.211511")(wj_bam = bam.wj, 
                      gv = gv, sj.name = sj.name, gtf.type = gtf.type, 
                      wj_gtf = gtf.wj, O = switch(substr(., 1, 
                        1), o = 0, O = 1), M = switch(substr(., 
                        2, 2), m = 0, M = 1), featureCounts.A = featureCounts.A, 
                      redo = gc.redo)
                  }) %>% unlist
                }))
            }
        }
        if (identical(only, "counts.wj")) {
            return(counts.wj)
        }
    })
    hs.browser("38.04.06.003751")
    hs.require("DESeq2")
    if (tpm) {
        if (length(counts.wj) != 1) 
            hs.browser("37.03.24.182255")
        dlff1 <- names(counts.wj[["gene"]])[1]
        message("37_03_24_175908")
        lapply(names(counts.wj[["gene"]]), function(dlff1) {
            counts.wj <- counts.wj[["gene"]][dlff1]
            tpm.wj <- counts.wj %>% hs.wjm(ext = 0, extN = 2) %>% 
                hs.wj("tpm.tsv.gz")
            message(tpm.wj)
            if (!file.exists(tpm.wj)) {
                gc <- counts.wj %>% fread(drop = 2:5)
                if (0) {
                  gc[1]
                  gc[[3]]
                  gc[1, .(strsplit(Start, ";")[[1]], strsplit(End, 
                    ";")[[1]])] %>% lapply(as.numeric) %>% {
                    .[[2]] - .[[1]] + 1
                  } %>% sum
                  gc[1, .(strsplit(Start, ";")[[1]], strsplit(End, 
                    ";")[[1]])] %>% lapply(as.numeric) %>% {
                    IRanges(.[[1]], .[[2]])
                  } %>% reduce %>% width %>% sum
                }
                X <- wk.dt.2matrix(gc, rn.cn = 1, cn2rm = 1:2)
                X <- X[rowSums(X == 0) < ncol(X), ]
                {
                  tpm <- apply(X, 2, hs.hsgly("counts2tpm.36.05.06.170737"), 
                    gv = gv, len = gc[match(rownames(X), Geneid), 
                      Length])
                  tpm %<>% as.data.table("gene.id_ensembl")
                  if (0) {
                    wk.dt.fwrite(tpm, "~/tmp/tmp.tsv.gz")
                    tpm <- fread("~/tmp/tmp.tsv.gz")
                    hs.hsgly("gene.id+.converter.using_ensembl.36.05.26.232037")(tpm, 
                      gv = "grcm38")
                  }
                  switch(gv, grch38 = hs.hsgly("gene_ids.adder.using_hgnc.36.03.24.173032")(tpm), 
                    grcm38 = hs.hsgly("gene.id+.converter.using_ensembl.36.05.26.232037")(tpm, 
                      gv = gv), hs.browser("36.05.26.231319"))
                  wk.dt.fwrite(tpm, tpm.wj)
                }
                if (0) 
                  if (!file.exists(fpkm.wj)) {
                    dds <- DESeq2::DESeqDataSetFromMatrix(X, 
                      data.frame(colnames(X)), design = ~1)
                    gene.length <- hs.hsgly("gene.length_unstranded.unique_reads.36.04.21.091051")(unique = 0)
                    mcols(dds) <- data.frame(basepairs = wk.dt.subset.na(gene.length, 
                      rownames(X), "gene.id_ensembl")[, length])
                    fpkm <- fpkm(dds)
                    fpkm %<>% as.data.table("gene.id_ensembl")
                    hs.hsgly("gene_ids.adder.using_hgnc.36.03.24.173032")(fpkm)
                    wk.dt.fwrite(fpkm, fpkm.wj)
                  }
            }
        })
    }
    gc.wj <- counts.wj[[gtf.type]]
    message("37_03_24_175823")
    if (identical(NA, group.hs)) {
        ybm <- if (length(wjj0)) {
            readLines(gc.wj[1], 1) %>% strsplit("\t") %>% unlist %>% 
                tail(-6)
        }
        else gc.wj[1] %>% readLines(1) %>% strsplit("\t") %>% 
            unlist %>% hs.grepV("(?i)^gene", inv = 1)
        return(ybm)
    }
    dlff <- names(gc.wj)
    if (!length(dlff)) {
        dlff <- "unknown"
        if (length(gc.wj) > 1) {
            hs.browser("36.06.04.170445")
        }
    }
    lapply(dlff, function(dlff1) {
        if (dlff1 != "unknown") {
            gc.wj %<>% {
                .[dlff1]
            }
        }
        jg.wjj <- do.call(hs.wjj, as.list(c("supp/de", sj.name, 
            gv, gtf.type, ifelse(dlff1 == "unknown", paste0("quantitator_", 
                dlff1), paste0("featureCounts_", dlff1)), "DESeq2")))
        gc <- if (length(yb.2do)) {
            fread(gc.wj, select = c(hs.re(readLines(gc.wj, 1), 
                "[^\t]+"), yb.2do))
        }
        else fread(gc.wj, drop = switch(dlff1, unknown = {
        }, 2:6))
        X <- as.matrix(gc, 1)
        if (c(gene_merge_to, gene_to_be_merged) %>% {
            (length(.) > 1) && all(. %in% rownames(X))
        }) {
            X[gene_merge_to, ] <- colSums(X[c(gene_merge_to, 
                gene_to_be_merged), ])
            X <- X[which(rowSums(X > 0) > 0) %not% match(gene_to_be_merged, 
                rownames(X)), ]
        }
        if (0) {
            a1 <- X[c(gene_merge_to, gene_to_be_merged), ]
            colnames(a1) %<>% group.hs
        }
        group <- group.hs(colnames(X))
        if (length(group.drop)) {
            i1 <- which(group %not.in% group.drop)
            X <- X[, i1, drop = 0]
            group %<>% {
                .[i1]
            }
        }
        if (length(yb.drop)) {
            i1 <- which(colnames(X) %not.in% yb.drop)
            X <- X[, i1, drop = 0]
            group %<>% {
                .[i1]
            }
        }
        if (length(group.2do)) {
            i1 <- which(group %in% group.2do)
            X <- X[, i1, drop = 0]
            group %<>% {
                .[i1]
            }
        }
        if ((!length(pca.wj.name)) && (length(group.drop) || 
            length(yb.drop))) {
            pca.wj.name <- paste0("remove,", paste(c(sort(group.drop), 
                sort(yb.drop)), collapse = ","))
        }
        if (pca) {
            if (0) {
                dds <- DESeqDataSetFromMatrix(X, data.frame(colnames(X)), 
                  design = ~1)
                dds %<>% {
                  DESeq(.)
                }
                nc <- counts(dds, 1)
                hs.hsgly("pca.36.04.07.121916")(log2(nc), wj = "~/tmp/pca")
                hs.hsgly("clustering,replacing_PCA.36.04.16.130256")(log2(nc + 
                  1))
            }
            pca.wj <- hs.hsgly("deseq2.pca.36.02.19.153251")(X, 
                group, jg.wjj, legend.x = "bottomright", redo = 1, 
                pca.wj.name = pca.wj.name)
        }
        if (de) {
            de.wj <- hs.hsgly("de.wj.hs.X.36.03.10.130448")(X = X, 
                gn = group, control = control, condition = condition, 
                outliers = outliers, wjj0 = jg.wjj, dds.redo = 1, 
                focus = focus) %>% hs.wj.portable
            if (en) {
                for (wj1 in de.wj) {
                  hs.hsgly("differential_gene_expression.enrichment.36.01.31.215610")(de.wj = wj1, 
                    org1 = switch(gv, grch38 = "human", grcm38 = "mouse", 
                      hs.browser("36.10.25.185442")), ...)
                }
            }
        }
        c(if (pca) c(pca = pca.wj), if (de) c(de = de.wj)) %>% 
            hs.wj.portable
    })
}
"36_03_10_130448" <- function(X, gn, control, condition = {
}, outliers = list({
}), wjj0, dds.redo = 0, focus = 1, ...) {
    de.wj <- lapply(seq_along(outliers), function(outliers.i) {
        outlier <- outliers[[outliers.i]]
        if (length(outlier)) {
            i1 <- colnames(X) %in% outlier
            if (any(i1)) {
                X %<>% {
                  .[, !i1, drop = 0]
                }
                gn <- gn[!i1]
            }
        }
        yb <- gn %>% unique %>% sort %>% sapply(. %>% {
            paste0(., "\t", sort(colnames(X)[gn == .]) %>% paste(collapse = ";;;;"))
        }) %>% unname
        de.wjj <- hs.wj(wjj0, if (focus) 
            "2groups"
        else "allSamples")
        if (0) 
            de.wjj <- local({
                yb.wj <- hs.wj(hs.wjj.ls(wjj0, wj = 0, wjj = 1), 
                  "yb.tsv.gz")
                yb.wj %<>% {
                  .[file.exists(.)]
                }
                hs.browser("36.06.10.134938")
                if (length(yb.wj)) {
                  yb.wj %<>% {
                    .[sapply(., . %>% readLines %>% identical(yb))]
                  }
                  if (length(yb.wj)) {
                    return(hs.wj.portable(dirname(yb.wj)))
                  }
                }
                wjj1 <- hs.wjj(wjj0, gn %>% unique %>% sort %>% 
                  paste(collapse = ",") %>% paste0(".", hs.gsub(.muid4machine(), 
                  "\\.", "_")))
                yb.wj <- hs.wj(wjj1, "yb.tsv.gz")
                con1 <- gzfile(yb.wj)
                on.exit(close(con1), add = TRUE)
                cat(yb, sep = "\n", file = con1)
                wjj1
            })
        hs.hsgly("deseq2.2+_groups.36.01.30.215755")(X, gn, control = control, 
            condition = condition, wjj1 = de.wjj, dds.redo = dds.redo, 
            focus = focus, ...)
    })
    hs.unlist(de.wj)
}
"36_02_19_153251" <- local({
    cols <- c("#999999", "#FF0099", "#E69F00", "#56B4E9", "#009E73", 
        "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#990000", 
        "#9900cc", "#66FF66", "#663300", "#0000FF", "#CC0033", 
        "#FF0000", "#000099", "#660066", "#333333", "#FFCCCC", 
        "#993366", "#33CC33", "#000099", "#CC9900")
    exp_better.hs <- function(exp1, id.type = {
    }) {
        if (!length(id.type)) 
            1
        gene.cn <- ifelse(rownames(exp1) %>% sample(min(9, length(.))) %>% 
            hs.grepl("^(?i)ens") %>% all, "gene.id_ensembl", 
            "gene.id_symbol")
        exp1 %<>% as.data.table(keep.rownames = gene.cn)
        if (gene.cn == "gene.id_symbol") 
            hs.hsgly("de_tbl.add.symbol.36.01.31.231208")(exp1, 
                cn1 = gene.cn, ft = "symbol")
        exp1
    }
    function(X, gn, wjj1, legend.x = "right", redo = 0, pca.wj.name = {
    }, sample_name.show = 0, ...) {
        hs.require(c("DESeq2", "RColorBrewer"))
        plot.wj <- hs.wj(wjj1, hs.pastec.("pca", pca.wj.name, 
            "pdf"))
        if (0) 
            if (file.exists(plot.wj) && (!redo)) 
                return(plot.wj %>% hs.wj.portable)
        yb.info <- data.table(id = colnames(X), condition = factor(gn))
        dds <- DESeqDataSetFromMatrix(X, yb.info, design = ~1)
        hs1.name <- ifelse(nrow(dds) < 2000, "rlog", "vst")
        a1 <- do.call(hs1.name, list(dds, blind = TRUE))
        {
            exp.wj <- hs.wj(wjj1, paste0("exp.", hs1.name, ",blind=T.tsv.gz"))
            if (length(rownames(X)) && (!file.exists(exp.wj))) {
                exp.wj %>% .ls
                exp1 <- assay(a1)
                if (length(rownames(X))) 
                  exp1 %<>% exp_better.hs
                wk.dt.fwrite(exp1, exp.wj)
            }
            exp.wj <- hs.wj(wjj1, "exp.counts,normalized.tsv.gz")
            if (length(rownames(X)) && (!file.exists(exp.wj))) {
                dds %<>% DESeq
                exp1 <- counts(dds, normalized = TRUE)
                if (length(rownames(X))) 
                  exp1 %<>% exp_better.hs
                wk.dt.fwrite(exp1, exp.wj)
            }
        }
        if (0) {
            pcaData <- plotPCA(a1, intgroup = c("condition"), 
                returnData = TRUE)
            percentVar <- round(100 * attr(pcaData, "percentVar"))
            fig <- ggplot(pcaData, aes(PC1, PC2, color = condition)) + 
                geom_point(size = 3) + xlab(paste0("PC1: ", percentVar[1], 
                "% variance")) + ylab(paste0("PC2: ", percentVar[2], 
                "% variance")) + coord_fixed()
            .pdf({
                show(fig)
            }, wj = plot.wj, width = 20, height = 20, ...)
            "~/in2/swxx/xm/db/36/09/18/36_09_18_190952/pca.pdf" %>% 
                wk.rsync.in
        }
        if (redo || (!file.exists(plot.wj))) {
            pc.xy <- plotPCA(a1, intgroup = c("condition"), returnData = TRUE)
            varPct <- round(attr(pc.xy, "percentVar") * 100, 
                1)
            .pdf({
                pch1 <- 16
                if (0) {
                  if (yb.info[, condition] %=>% levels %=>% (length(x) > 
                    11)) 
                    hs.browser("36.03.01.134723")
                  col1 <- yb.info[, condition] %=>% levels %=>% 
                    hs.c(x, brewer.pal(length(x), "Set3"))
                }
                col1 <- yb.info[, condition] %=>% levels %=>% 
                  hs.c(x, head(cols, length(x)))
                for (xi in 1) {
                  plot(pc.xy[, 1], pc.xy[, 2], xlab = paste0("PC", 
                    xi, " (", varPct[1], "% variance)"), ylab = paste0("PC", 
                    xi + 1, " (", varPct[2], "% variance)"), 
                    col = col1[gn], main = "", pch = pch1)
                  x.ss <- par("usr")[1:2] %=>% (diff(x)/15)
                  if (sample_name.show) 
                    text(pc.xy[, 1] - x.ss, pc.xy[, 2], pc.xy[, 
                      "name"], pos = 3, xpd = 1)
                  legend(legend.x, legend = names(col1), col = col1, 
                    pch = pch1)
                }
            }, wj = plot.wj, width = 20, height = 20, ...)
        }
        plot.wj %>% hs.wj.portable
    }
})
"37_05_27_192408" <- function(root, org1 = "human") {
    de_fp <- hs.hsgly("de_fp.37_05_26_182426")(root)
    hs.hsgly("en.37_05_12_000258")(de_fp, org1 = org1)
    hs.hsgly("en_rank_热图.37_05_13_221522")(root)
}
"37_05_26_182426" <- function(wjj1, wanted = {
}) {
    wj <- hs.wjj.ls(wjj1, "\\.tsv\\.gz$")
    names(wj) <- hs.wjm(wj, dn = 0, ext = 0, extN = 2)
    if (length(wanted)) {
        wj[wanted]
    }
    else wj
}
"37_05_12_000258" <- function(de_wj, org1, ...) {
    for (wj1 in de_wj) {
        hs.hsgly("differential_gene_expression.enrichment.36.01.31.215610")(wj1, 
            org1 = org1, ...)
    }
}
"37_05_13_221522" <- function(root, wjj2, q_co = 0.1) {
    wjj <- hs.hsgly("en_rank_folder.37_05_13_224153")(root)
    significantly_enriched <- hs.hsgly("folders_to_enrichemnt.37_05_13_230726")(wjj, 
        q_co = q_co)
    hs.hsgly("排序富集热图.37_05_12_154907")(significantly_enriched, 
        wjj2)
}
"37_05_13_224153" <- function(root) {
    wj <- hs.wjj.ls(root, "\\.tsv\\.gz$")
    nm <- hs.wjm(wj, dn = 0, ext = 0, extN = 2)
    wjj <- hs.wjm(wj, ext = 0, extN = 2) %>% hs.fp("en/gsea_ucsd,broad/log2fc")
    names(wjj) <- nm
    wjj
}
"37_05_27_113458" <- function(pw, de_fp, org1 = "human", org_ref = org1[1], 
    de_wanted = {
    }, wjj2 = hs.fp(dirname(de_fp[1]), "图表"), bn2 = hs.pastec0(if (length(wj2_pre)) c(wj2_pre, 
        "_"), "lfc_heatmap.pdf"), wj2_pre = {
    }, wj2 = hs.fp(wjj2, bn2), root = {
    }) {
    if (length(pw) > 1) 
        hs.browser("37.05.27.184356")
    gene <- hs.hsgly("基因集合的成员.37_05_27_100106")(pw, 
        org1 = org_ref)
    if (length(de_wanted)) 
        de_fp <- de_fp[de_wanted]
    de <- lapply(de_fp, wk.fread)
    for (i1 in seq_along(de_fp)) {
        i1.org <- org1[hs.recycle(i1, length(org1))]
        if (i1.org != org_ref) {
            de[[i1]] %<>% {
                hs.hsgly("同源基因之表格.37_05_27_153942")(., 
                  i1.org, org_ref, gt = "symbol")
            }
        }
    }
    gene_used <- lapply(de, . %>% {
        gn_cn <- gene_id_name.37_04_29_145914(names(.))
        qv_cn <- qvalue_name.37_04_29_150328(names(.))
        .[get(qv_cn) < 0.05, get(gn_cn)]
    }) %>% unlist %>% unique
    gene %<>% (`%and%`(gene_used))
    m1 <- sapply(de, . %>% {
        lfc_cn <- log_fold_change_name.37_04_29_150019(names(.))
        gn_cn <- gene_id_name.37_04_29_145914(names(.))
        .[match(gene, get(gn_cn)), get(lfc_cn)]
    })
    rownames(m1) <- gene
    m1[is.na(m1)] <- 0
    m1 <- hs.hsgly("从正到负_从大到小.37_05_27_182956")(m1)
    pw.name <- pw
    hs.hsgly("ComplexHeatmap/36.08.12.233720")(m1, legend.name.type = "lfc", 
        column_title = pw.name, wj = wj2, color.breaks = c("green", 
            "black", "red"), color.breaks.boundary.max = 3, cluster_rows = 0)
}
"36_05_06_170737" <- local({
    gene.length <- list()
    function(counts, gv = "grch38", len = {
    }, release = "105") {
        if (gv %not.in% names(gene.length)) {
            gene.length[gv] <<- list(hs.hsgly("gene.length_unstranded.unique_reads.36.04.21.091051")(gv = gv, 
                release = release, unique = 0))
        }
        effLen <- if (length(len)) 
            len
        else wk.dt.subset.na(gene.length[[gv]], names(counts), 
            "gene.id_ensembl")[, length]
        jg <- hs.c(names(counts), rep(NA, length(counts)))
        i1 <- which(effLen > 0)
        rate <- log(counts[i1]) - log(effLen[i1])
        denom <- log(sum(exp(rate)))
        jg[i1] <- exp(rate - denom + log(1e+06))
        names(jg)[is.na(jg)]
        jg
    }
})
"36_05_06_170652" <- local({
    c1 <- log(1e+06)
    function(fpkm) {
        exp(log(fpkm) - log(sum(fpkm)) + c1)
    }
})
"2023_06_29_194530" <- function(m1, sum1 = switch(data_type, 
    bulk = 1e+06, sc = 10000), data_type = "bulk") {
    s1 <- sum(m1[, 1])
    if (!hs.num_is_equal(s1, sum1)) 
        return(FALSE)
    s1 <- sum(m1[, 2])
    if (!hs.num_is_equal(s1, sum1)) 
        return(FALSE)
    d1 <- m1 %=>% hs.colSums %=>% abs(x - sum1) %=>% max
    hs.num_is_equal(d1)
}
"39_03_20_154050" <- function(de, cn_wanted = "gene") {
    hs.switch(cn_wanted, "gene", {
        a <- hs.grepV(names(de), "(?i)gene.*symbol")
        if (length(a) == 1) 
            return(a)
        a <- hs.grepV(names(de), "(?i)symbol")
        if (length(a) == 1) 
            return(a)
        hs.msg("39_03_20_154614")
    }, "q", {
        a <- hs.grepV(names(de), "(?i)adj.*p.*val.*")
        if (length(a) == 1) 
            return(a)
        hs.msg("39_03_20_154608")
    }, "fc", {
        a <- hs.grepV(names(de), "(?i)log.*fc")
        if (length(a) == 1) 
            return(a)
        hs.msg("39_03_20_154639")
    })
}
"37_09_26_175722" <- function(exp1, sample_group = {
}, co_group = 0.5, co_exp = 0, gene_side = 2 - (ncol(exp1) == 
    length(sample_group))) {
    m1 <- if (co_exp > 0) {
        exp1 >= co_exp
    }
    else exp1 > 0
    i1 <- apply(m1, gene_side, . %>% {
        all(tapply(., sample_group, mean) < co_group)
    })
    if (any(i1)) 
        exp1 %<>% {
            .[!i1, , drop = 0]
        }
    exp1
}
"38_09_30_162626" <- function(exp_wj, gtf_wj, wj2 = sapply(exp_wj, 
    function(wj1) hs.fp(dirname(wj1), hs.sub(basename(wj1), "\\.gene_id\\.", 
        ".symbol.")))) {
    if (wj.update.reporter(wj2, exp_wj)) {
        gtf <- .sys.cat.text(gtf_wj) %>% .sys("| perl -e", .q("while( <>) { @F = split( \"\t\"); if( $F[ 2] eq \"gene\") { print}}")) %>% 
            {
                wk.fread(cmd = .)
            }
        gtf[, `:=`(ensg, hs.re(V9, "(?<=gene_id \")[^\"]+"))]
        gtf[, `:=`(symbol, hs.re(V9, "(?<=gene_name \")[^\"]+", 
            no_match = ""))]
        gtf[hs.clean(symbol, ot = "l"), if (.N > 1) 
            .I, by = symbol]
        gtf[, `:=`(i, .I)]
        gtf[, {
            a <- if (.N == 1) {
                symbol
            }
            else {
                message(symbol)
                if (nchar(substr(symbol, 1, 1))) {
                  paste0(symbol, "____", ensg)
                }
                else {
                  ensg
                }
            }
            wk.dt.set(gtf, i, "gene_id", a)
            TRUE
        }, by = symbol]
        sapply(seq_along(exp_wj), function(i1) {
            wj1 <- exp_wj[i1]
            wj2 <- wj2[i1]
            if (wj.update.reporter(wj2, wj1)) {
                hs.msg(wj1)
                a1 <- wk.fread(wj1)
                a <- gtf[match(a1[, Geneid], ensg), gene_id]
                a1[, `:=`(Geneid, a)]
                wk.dt.fwrite(a1, wj2)
            }
            0
        })
    }
    wj2
}
"38_09_16_095937" <- local({
    hs1 <- function(comparison1, rename = 0, collapse = " - ") {
        jg <- names(comparison1)
        if (hs.any(rename, !length(jg))) {
            jg <- sapply(comparison1, . %>% sort %>% paste(collapse = "____"))
        }
        if ((length(collapse) == 1) && nzchar(collapse)) 
            jg <- paste(jg, collapse = collapse)
        jg
    }
    function(comparison, ...) {
        if (is.list(comparison[[1]])) {
            seq_along(comparison) %>% sapply(. %>% {
                jg <- names(comparison)[.] %>% hs.clean
                if (hs.any(!length(jg))) 
                  jg <- hs1(comparison[[.]], ...)
                jg
            })
        }
        else {
            hs1(comparison, ...)
        }
    }
})
"38_01_14_124719" <- function(snv, ref.cn = names(snv)[1], alt.cn = names(snv)[2]) {
    tbl <- snv[, hs.table(get(ref.cn), get(alt.cn))]
    tbl[c("A", "G", "C", "T"), c("G", "A", "T", "C")] %>% {
        sum(diag(.))/sum(.) * 100
    }
}
"36_05_04_140727" <- local({
    pic.bin.wjj <- "~/in2/swxx/rj/PIC-master"
    function(gv = "grch38", source = "ensembl") {
        pic.db.wjj <- .mfp.swxx(switch(source, ensembl = "36.05.04.141138", 
            ncbi = "36.06.17.221554")) %>% hs.wjj(gv)
        wj <- hs.wjj.ls(pic.db.wjj)
        names(wj) <- tolower(hs.re(hs.wjm(wj, dn = 0, ext = 0), 
            "[^_]+$"))
        return(wj)
        setwd(pic.db.wjj)
        {
            bin <- hs.wj(pic.bin.wjj, "extract_genomic_locations.py")
            {
                gtf.wj <- "~/in2/swxx/sj/db/ensembl/human/Homo_sapiens.GRCh38.99.gtf"
                gsize.wj <- "~/in2/swxx/sj/jyz/human/ensembl/Homo_sapiens.GRCh38.dna.primary_assembly.fa.fai"
            }
            {
                gtf.wj <- "~/in2/swxx/sj/db/ensembl/human/Homo_sapiens.GRCh37.87.gtf"
                gsize.wj <- "~/in2/swxx/sj/jyz/human/ensembl/Homo_sapiens.GRCh37.dna.primary_assembly.fa.fai"
            }
            bin <- hs.wj(pic.bin.wjj, "chr", "extract_genomic_locations.py")
            {
                gtf.wj <- "~/in2/swxx/sj/db/ensembl/human/Homo_sapiens.GRCh38.99.gtf"
                gsize.wj <- "~/in2/swxx/sj/jyz/human/ncbi/GCF_000001405.39_GRCh38.p13_genomic.fna.fai"
            }
        }
        .wk("less -Ni", hs.wj(pic.bin.wjj, "README.md"))
        gsize.wj %>% .ls1
        gsize.wj %>% .less
        hs.wj(pic.bin.wjj, "extract_genomic_locations.py") %>% 
            .less
        gtf.wj %>% .less
        .sys("python", bin, "--gtf", gtf.wj, "--TSSup 50 --TSSdown 300 --GBdown 3000 --gsize", 
            gsize.wj)
        pic.db.wjj %>% .ls
        wj1 <- hs.wj(pic.db.wjj, "gene_GB.bed")
        wj1 <- hs.wj(pic.db.wjj, "gene_tss.bed")
        a1 <- fread(wj1)
        a1[, V1 %>% hs.table] %>% .sh
        wk.dt.fwrite(a1[V1 %in% c(1:22, "X", "Y", "MT")], wj1, 
            col.names = !1)
    }
})
"35_11_12_182333" <- local({
    go.loaded <- list()
    kegg.loaded <- list()
    list.hs.gmt <- function(wj_gmt) {
        nr <- readLines(wj_gmt) %=>% strsplit("\t")
        term2name <- lapply(nr, . %=>% {
            if (length(x) > 2) {
                head(x, 2)
            }
        }) %=>% do.call("rbind", x) %=>% wk.dt.as
        term2gene <- lapply(nr, . %=>% {
            if (length(x) > 2) {
                data.table(x[1], tail(x, -2) %=>% unique)
            }
        }) %=>% rbindlist %=>% unique
        list(term2gene = term2gene, term2name = term2name)
    }
    cellMarkers <- {
    }
    msigdb.info <- {
    }
    if (0) {
        msigdb <- list()
        msigdb.wjj <- local({
            wjj <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("msigdb") %=>% 
                hs.wjj.ls(wjj = 1)
            if (length(wjj) > 1) {
                names(wjj) <- basename(wjj) %=>% hs.re("(?<=v)\\d+(\\.\\d+)?")
                wjj <- wjj[which.max(as.numeric(names(wjj)))]
            }
            wjj
        })
        msigdb.info.hs <- function() {
            hs.hsgly("term.description.36_12_14_134354")(msigdb.wjj)
        }
        msigdb.hs <- function(org1 = "human") {
            if (is.null(msigdb[[org1]])) {
                if (0) {
                  wj <- hs.wjj.ls(msigdb.wjj, paste0("\\.entrez", 
                    switch(org1, human = "", mouse = "\\.mouse"), 
                    "\\.gmt$"), R = 1)
                  wj <- hs.grepV(wj, "(c1\\..*|c2\\.(cgp|cp\\..*)|c3\\.(mir|tft)\\..*|c4\\.c(gn|m)|c5\\.go\\.(bp|cc|mf).*|c[678]\\..*|h\\..*)\\.v\\d+")
                  names(wj) <- wj %=>% basename %=>% hs.sub("\\.v\\d+\\.\\d+\\.entrez.*\\.gmt$")
                  hs.browser("37.09.30.102132")
                  msigdb[org1] <<- lapply(wj, read.gmt)
                }
                {
                  msigdb[[org1]] <<- hs.hsgly("数据.38_01_08_211229")(org1 = org1)
                }
            }
        }
        for (on1 in paste0("msigdb.", c("human", "mouse"))) {
            assign(on1, list())
        }
    }
    gProfiler.db <- {
    }
    hs.output <- function(r1, wj1, gsea = 0, org1 = "human") {
        .try(wk.dt.setDT(r1))
        if (org1 %not.in% c("human", "mouse")) {
            wk.dt.fwrite(r1, wj1)
            return(wj1)
        }
        if (is.list(r1) && nrow(r1)) {
            if (gsea) {
                cn1 <- "Description"
                i1 <- hs.match(cn1, names(r1))
                if (length(msigdb.info) && length(i1)) {
                  i2 <- hs.match(r1[[i1]], msigdb.info[, STANDARD_NAME])
                  i3 <- which(!is.na(i2))
                  if (length(i3)) {
                    r1[[i1]][i3] <- msigdb.info[i2[i3], paste0("【BRIEF】", 
                      DESCRIPTION_BRIEF, DESCRIPTION_FULL %=>% 
                        ifelse(nchar(x), paste0("【FULL】", 
                          x), ""))]
                  }
                }
            }
            a1 <- hs.hsgly("clusterProfiler.add.geneSymbol.35.11.20.170355")(r1, 
                org1 = org1)
            wk.dt.fwrite(a1, wj1)
        }
        else cat("", file = wj1)
    }
    hs.rnk <- function(genes, scores, ft = "ENSEMBL", tt = switch(org1, 
        yeast = "ENSEMBL", "ENTREZID"), org1 = "human") {
        hs.require("clusterProfiler")
        hs.c(genes, scores) %=>% {
            if (ft != tt) {
                a1 <- names(x) %=>% bitr(fromType = ft, toType = tt, 
                  OrgDb = hs.hsgly("org.db.36.06.08.172512")(org1 = org1, 
                    ot = "nam"))
                x <- x[a1[, ft]]
                names(x) <- a1[, "ENTREZID"]
            }
            sort(x, decreasing = TRUE)
        }
    }
    hs.wj_out <- function(wj_out.pre, post = "", ext = "tsv.gz") {
        wj_out.pre %=>% paste0(x, if (hs.grepl(x, "/$")) 
            ""
        else ".", post, ".", ext)
    }
    hs.barplot <- function(wj1, wj2 = hs.wjm(wj1, ext = if (top) 
        paste0("top_", top)
    else 0, extN = ifelse(hs.grepl(wj1, "\\.gz$"), 2, 1)), co1 = 0.05, 
        top = 0, topN = 500, bar.width = 1, pdf.height = {
        }, text.cex = 1, sj1, qv.c = {
        }, main1 = {
        }, path.cn = "Description", redo = 0) {
        if (redo) 
            hs.wj.rm(wj2)
        if (file.exists(wj2)) 
            return(wj2)
        if (missing(sj1)) 
            sj1 <- wk.fread(wj1)
        if (hs.is_empty(sj1)) 
            return()
        if (!length(qv.c)) 
            qv.c <- hs.grepV(names(sj1), "(?i)^q[\\.-_]?v(al)?")[1]
        if (!length(qv.c)) 
            qv.c <- hs.grepV(names(sj1), "^p[\\._-]?adj")[1]
        if (!length(qv.c)) 
            hs.browser("【35.12.26.092732】need one and only one column for q-values")
        i1 <- which(sj1[[qv.c]] < co1)
        if (!length(i1)) 
            return()
        sj1 %<=>% hs.ij(x, i1)
        sj1[, `:=`(Description, Description %=>% hs.factor(x, 
            rev(x), N = 0))]
        if (top > 0) 
            sj1 <- head(sj1, top)
        sj1[, `:=`("-log10q", -log(get(qv.c), 10))]
        if (!length(main1)) 
            main1 <- hs.wjm(wj1, dn = 0, ext = 0, extN = 1 + 
                hs.grepl(wj1, "(?i)\\.gz$")) %=>% hs.sub("^aaa\\.")
        xlab1 <- bquote("Enrichment Scores (-Log"[10] * "FDR)")
        text.width <- text.width.hs(sj1[, get(path.cn)])
        xlab1.width <- text.width.hs(xlab1)
        plot.height <- length(sj1[, get(path.cn)])
        fig.height <- sum(plot.height, 5)
        plot.width <- sj1[, ceiling(max(`-log10q`)) + 0.5]
        mai.4 <- max(1, (xlab1.width - plot.width)/2 + 1)
        fig.width <- sum(plot.width, mai.4, text.width)
        p1 <- ggplot2::ggplot(sj1, ggplot2::aes(Description, 
            `-log10q`)) + ggplot2::geom_bar(stat = "identity") + 
            ggplot2::geom_hline(yintercept = -log10(0.05), col = "pink", 
                linetype = 2) + ggplot2::coord_flip() + ggplot2::xlab("") + 
            ggplot2::ylab(bquote("Enrichment Scores (-Log"[10] * 
                "FDR)")) + ggplot2::theme_bw() + ggplot2::theme(panel.grid = ggplot2::element_blank())
        .try(.figs(print(p1), wj2, width = fig.width, height = fig.height))
        return()
        hs.38_08_29_222326 <- function() {
            y1 <- sj1[, `-log10q` %=>% rev]
            op1 <- par(pin = c(plot.width, plot.height)/2.54, 
                mai = c(strheight("W", "inches") * 20, text.width + 
                  1, 2, mai.4)/2.54, usr = c(0, max(y1), 0, sj1[, 
                  .N]), xaxs = "i", yaxs = "i", xpd = TRUE)
            on.exit(par(op1), add = TRUE)
            col1 <- if ("NES" %in% names(sj1)) {
                sj1[, ifelse(NES > 0, "#A20056E5", "#3B4992E5")]
            }
            else {
                "#A20056E5"
            }
            col1 %<=>% rev
            y2 <- barplot(y1, border = NA, horiz = TRUE, col = col1, 
                axes = TRUE, xlab = xlab1, main = main1, width = bar.width, 
                usr = c(0, max(y1), 0, sj1[, .N]))
            text(0, y2, sj1[, rev(get(path.cn))], pos = 2, cex = text.cex)
            leg1 <- {
                if ("NES" %in% names(sj1)) {
                  c("Pos", "Neg")
                }
                else {
                }
            }
            if (length(leg1)) 
                legend(par("usr")[1] - diff(par("usr")[1:2])/diff(par("plt")[1:2]) * 
                  par("plt")[1]/2, par("usr")[4] + 2, legend = leg1, 
                  pch = 15, col = c("#A20056E5", "#3B4992E5"), 
                  bty = "n", xpd = TRUE)
            lines(rep(-log(co1, 10), 2), c(0, max(y2) + diff(y2[1:2])), 
                lty = 2, lwd = 2, col = "pink", xpd = FALSE)
        }
        hs.browser("38.08.29.215211")
        .pdf(hs.38_08_29_222326(), wj2, width = fig.width, height = fig.height)
        evalbq(.pdf(.(exp1), wj2, width = fig.width/2.54, height = fig.height/2.54))
        bquote(.pdf(.(exp1), wj2, width = fig.width/2.54, height = fig.height/2.54, 
            local = 1))
        .png(exp1, wj2, width = fig.width/2.54, height = fig.height/2.54)
        wj2
    }
    tbl.hs <- function(tbl.wj) {
        a1 <- fread(tbl.wj, data.table = 0)
        rownames(a1) <- a1[[1]]
        a1
    }
    bubble_plot.wj.hs <- function(rd, rd.wj = {
    }, tbl.wj, tbl = {
    }, co1 = 0.05, top = 0, bubble_plot.wj = hs.wjm(tbl.wj, ext = hs.pastec0(if (top) 
        c("top_", top, ".")
    else "", "bubble_plot"), extN = 2), pdf.width = {
    }, pdf.height = {
    }, redo = 0, ...) {
        if (redo) 
            hs.wj.rm(bubble_plot.wj)
        if (file.exists(paste0(bubble_plot.wj, ".png"))) 
            return()
        if (length(rd.wj)) {
            rd <- hs.load(rd.wj)
        }
        if (identical(tbl, {
        })) 
            tbl <- tbl.hs(tbl.wj)
        if (top > 0) 
            tbl %<=>% head(x, top)
        bubble_plot.wj %=>% {
            q.cn <- hs.grepV(names(tbl), "qval")
            if (!length(q.cn)) 
                q.cn <- hs.grepV(names(tbl), "p\\.?adj")
            n1 <- sum(tbl[[q.cn]] < co1, na.rm = 1)
            if (n1 > 1) {
                text.width <- text.width.hs(head(tbl, n1)[, Description])
                p1 <- enrichplot::dotplot(rd, showCategory = n1, 
                  ...)
                width <- if (length(pdf.width)) 
                  pdf.width
                else 15 + text.width
                height <- if (length(pdf.height)) 
                  pdf.height
                else 5 + n1
                .try(.figs(print(p1), width = width, height = height, 
                  wj = x))
            }
            x
        }
    }
    hs.en <- function(tbl.wj, s1, u1, et, ot = {
    }, org1, OrgDb, TERM2GENE, seed = {
    }, redo = {
    }) {
        tried_wj <- hs.wjm(tbl.wj, ext = "tried.txt", extN = 2)
        if (length(s1) < 5) 
            cat("", file = tried_wj)
        if (file.exists(tried_wj)) 
            return()
        if (!all(file.exists(hs.wjm(tbl.wj, ext = c("rd", "pdf", 
            "bubble_plot.pdf"), extN = 2)))) {
            hs.msg(hs.pastec0("【", et, if (length(ot)) 
                c(", ", ot), "】", tbl.wj))
            if (0) {
                et == "do"
                "padj_fc2 down"
            }
            rd.wj <- tbl.wj %=>% hs.wjm(ext = "rd", extN = 2)
            if (c(et, ot) %=>% paste(collapse = ".") %=>% {
                x %in% redo
            }) 
                hs.wj.rm(rd.wj)
            r1 <- rd.wj %=>% {
                {
                  hs1 <- switch(et, reactome = , reaction = , 
                    go = , kegg = , cell_markers = enricher, 
                    do = enrichDO)
                  if (0) {
                    ReactomePA:::get_Reactome_DATA
                    ReactomePA:::reactomeEXTID2PATHID
                    ReactomePA:::getALLEG
                    reactome.db::reactomeEXTID2PATHID
                    function(organism = "human") {
                      ReactomePA_Env <- get_Reactome_Env()
                      if (exists("organism", envir = ReactomePA_Env, 
                        inherits = FALSE)) {
                        org <- get("organism", envir = ReactomePA_Env)
                        if (org == organism && exists("PATHID2EXTID", 
                          envir = ReactomePA_Env) && exists("EXTID2PATHID", 
                          envir = ReactomePA_Env) && exists("PATHID2NAME", 
                          envir = ReactomePA_Env)) {
                          return(ReactomePA_Env)
                        }
                      }
                      ALLEG <- getALLEG(organism)
                      EXTID2PATHID <- as.list(reactomeEXTID2PATHID)
                      EXTID2PATHID <- EXTID2PATHID[names(EXTID2PATHID) %in% 
                        ALLEG]
                      PATHID2EXTID <- as.list(reactomePATHID2EXTID)
                      PATHID2NAME <- as.list(reactomePATHID2NAME)
                      PI <- names(PATHID2NAME)
                      PATHID2NAME <- lapply(PATHID2NAME, function(x) x[1])
                      names(PATHID2NAME) <- PI
                      PATHID2EXTID <- PATHID2EXTID[names(PATHID2EXTID) %in% 
                        names(PATHID2NAME)]
                      PATHID2EXTID <- PATHID2EXTID[names(PATHID2EXTID) %in% 
                        unique(unlist(EXTID2PATHID))]
                      PATHID2EXTID <- lapply(PATHID2EXTID, function(x) intersect(x, 
                        ALLEG))
                      PATHID2NAME <- PATHID2NAME[names(PATHID2NAME) %in% 
                        names(PATHID2EXTID)]
                      PATHID2NAME <- unlist(PATHID2NAME)
                      PATHID2NAME <- gsub("^\\w+\\s\\w+:\\s+", 
                        "", PATHID2NAME)
                      assign("PATHID2EXTID", PATHID2EXTID, envir = ReactomePA_Env)
                      assign("EXTID2PATHID", EXTID2PATHID, envir = ReactomePA_Env)
                      assign("PATHID2NAME", PATHID2NAME, envir = ReactomePA_Env)
                      return(ReactomePA_Env)
                    }
                    enrichGO
                    env1 <- clusterProfiler:::get_GO_data("org.Hs.eg.db", 
                      "MF", keytype = "ENTREZID")
                    goterms <- AnnotationDbi::Ontology(GO.db::GOTERM)
                    go2gene <- suppressMessages(AnnotationDbi::mapIds(org.Hs.eg.db, 
                      keys = "CC", column = "ENTREZID", keytype = "GOCC", 
                      multiVals = "list"))
                    ls("package:org.Hs.eg.db") %=>% .sh
                    `?`(org.Hs.egGO)
                    `?`(org.Hs.egGO2EG)
                    go.ncbi <- as.list(org.Hs.egGO2EG)
                    env1 %=>% ls
                    env1[["EXTID2PATHID"]]
                    env1[["PATHID2EXTID"]]
                    clusterProfiler:::enricher_internal
                    DOSE:::EXTID2TERMID
                    egGO <- as.list(org.Hs.egGO)
                    id1 <- "GO:0000287"
                    env1[["PATHID2EXTID"]][[id1]]
                    go.ncbi[[id1]]
                    go.term.hs.id("GO:0000287")
                    go.term.hs.id("GO:0000287")
                  }
                  args <- list(gene = as.character(s1), universe = as.character(u1), 
                    pAdjustMethod = "fdr", pvalueCutoff = 1.1, 
                    qvalueCutoff = 1.1)
                  switch(et, go = {
                    if (0) {
                      args["OrgDb"] <- list(OrgDb)
                      args["ont"] <- list(toupper(ot))
                      if (org1 %in% "yeast") args["keyType"] <- list("ENSEMBL")
                    }
                    {
                      a <- go.loaded[[org1]][[ot]]
                      args["TERM2GENE"] <- a["term2gene"]
                      args["TERM2NAME"] <- a["term2name"]
                    }
                  }, reactome = {
                    if (0) {
                      args["organism"] <- org1
                    }
                    {
                      a <- hs.hsgly("reactome.gene.4clusterProfiler.36_12_31_171255")(org1)
                      args["TERM2GENE"] <- a["term2gene"]
                      args["TERM2NAME"] <- a["term2name"]
                    }
                  }, reaction = {
                    a <- hs.hsgly("reactome.gene.4clusterProfiler.36_12_31_171255")(org1, 
                      type = "reaction")
                    args["TERM2GENE"] <- a["term2gene"]
                    args["TERM2NAME"] <- a["term2name"]
                  }, kegg = {
                    args["TERM2GENE"] <- list(TERM2GENE[, .(set_id, 
                      gene)] %=>% unique)
                    args["TERM2NAME"] <- list(TERM2GENE[, .(set_id, 
                      set)] %=>% unique)
                  }, cell_markers = {
                    args["TERM2GENE"] <- list(TERM2GENE)
                    args["minGSSize"] <- list(3)
                  })
                  a1 <- do.call(hs1, args)
                  if (et == "go") 
                    a1@ontology <- toupper(ot)
                  hs.save(a1, wj = hs.wj(x))
                  a1
                }
            }
            cat("", file = tried_wj)
            if (!length(r1)) 
                return()
            if (et == "go") {
                fig.wj <- rd.wj %=>% hs.wjm(ext = "net")
                if (!file.exists(fig.wj)) {
                  .figs(clusterProfiler::plotGOgraph(r1), wj = fig.wj, 
                    width = 15, height = 15)
                }
            }
            tbl <- r1 %=>% wk.dt.as
            bubble_plot.wj.hs(r1, top = 10, tbl.wj = tbl.wj, 
                tbl = tbl)
            hs.output(tbl, tbl.wj, org1 = org1)
            hs.barplot(tbl.wj)
        }
    }
    function(s1 = {
    }, u1 = {
    }, rnk = {
    }, s1_all = {
    }, org1 = "human", ft = {
    }, tt = switch(org1, yeast = "ENSEMBL", "ENTREZID"), wj_out.pre, 
        wj_out.pre.4top = {
        }, co1 = 1.1, seed = {
        }, top = 0, anno.type.2do = {
        }, ncpu = 2, db_extra = {
        }, db_extra_only = 1) {
        if ("preloaded.39_03_23_105643" %=>% {
            hs.exists(x, .GlobalEnv) && (!get(x, envir = .GlobalEnv))
        }) 
            stop("[38_01_08_175317]请先预加载")
        org1 <- hs.hsgly("物种/中文名.common_name.38_09_06_110337")(org1)
        BiocParallel::register(BiocParallel::MulticoreParam(workers = ncpu))
        if (!length(ft)) 
            ft <- hs.hsgly("编号种类猜测.37_04_28_165736")(if (length(s1)) 
                s1
            else rnk[[1]])
        if (org1 %not.in% names(kegg.loaded)) 
            kegg.loaded[[org1]] <<- hs.hsgly("kegg/通路/基因.36.03.18.133126")(org1 = org1)
        if (org1 %not.in% names(go.loaded)) 
            go.loaded[[org1]] <<- "bp,cc,mf" %=>% hs.strsplit.v(",") %=>% 
                hs.lapply(x, . %=>% {
                  hs.hsgly("go.gene.4clusterProfiler.36_12_31_143101")(org1, 
                    ot = x)
                })
        if (length(s1) && length(u1)) {
            switch(org1, human = {
                u1 %<=>% hs.clean(hs.hsgly("gene.id.converter.using_hgnc.36.01.22.160941")(x, 
                  "ensg", "ncbi", tolower(ft), "ncbi")[[2]], 
                  U = 1)
                s1 %<=>% hs.clean(hs.hsgly("gene.id.converter.using_hgnc.36.01.22.160941")(x, 
                  "ensg", "ncbi", tolower(ft), "ncbi")[[2]], 
                  U = 1)
                s1_all %<=>% {
                  if (is.list(x)) {
                    lapply(x, . %=>% hs.hsgly("gene.id.converter.using_hgnc.36.01.22.160941")(x, 
                      "ensg", "ncbi", tolower(ft), "ncbi")[[2]] %=>% 
                      hs.clean(U = 1))
                  } else if (length(x)) {
                    hs.clean(hs.hsgly("gene.id.converter.using_hgnc.36.01.22.160941")(x, 
                      "ensg", "ncbi", tolower(ft), "ncbi")[[2]], 
                      U = 1)
                  }
                }
                s1_ncbi <- s1
                u1_ncbi <- u1
                s1_all_ncbi <- s1_all
            }, mouse = {
                switch(ft, ensembl = {
                  u1 <- wk.dt.subset(hs.hsgly("gene.info.36_09_21_151142")(), 
                    u1, "gene.id_ensmebl")[, as.character(unique(gene.id_ncbi))]
                  s1 <- wk.dt.subset(hs.hsgly("gene.info.36_09_21_151142")(), 
                    s1, "gene.id_ensmebl")[, as.character(unique(gene.id_ncbi))]
                }, symbol = {
                  u1 <- wk.dt.subset(hs.hsgly("gene.info.36_09_21_151142")(), 
                    exp_l = c(tolower(gene.id_symbol) %in% tolower(u1)), 
                    exp_j = paste(hs.clean(gene.id_ncbi, U = 1)))
                  s1 <- wk.dt.subset(hs.hsgly("gene.info.36_09_21_151142")(), 
                    exp_l = c(tolower(gene.id_symbol) %in% tolower(s1)), 
                    exp_j = paste(hs.clean(gene.id_ncbi, U = 1)))
                })
                u1_ncbi <- u1
                s1_ncbi <- s1
            }, {
                u1_ncbi <- hs.hsgly("geneid1-geneid2.37_04_28_165424")(u1, 
                  ft = "symbol", tt = "ncbi", org1 = org1) %=>% 
                  hs.clean(U = 1)
                s1_ncbi <- hs.hsgly("geneid1-geneid2.37_04_28_165424")(s1, 
                  ft = "symbol", tt = "ncbi", org1 = org1) %=>% 
                  hs.clean(U = 1)
                if (is.list(s1_all)) {
                  s1_all_ncbi <- s1_all %=>% lapply(x, . %=>% 
                    {
                      do.call("bitr", list(x, ft, "ENTREZID", 
                        org_db.name))[, "ENTREZID"]
                    })
                } else {
                  if (length(s1_all)) s1_all_ncbi <- s1_all %=>% 
                    do.call("bitr", list(x, toupper(ft), "ENTREZID", 
                      org_db.name))[, "ENTREZID"]
                }
            })
            s1 %<=>% {
                x %and% u1
            }
            s1_ncbi %<=>% {
                x %and% u1_ncbi
            }
            if (top) {
                if (length(s1) >= 100) {
                  return()
                }
                if (org1 != "human") {
                  hs.browser("36.06.09.135731")
                }
                if (length(wj_out.pre.4top)) 
                  wj_out.pre <- hs.fp(wj_out.pre.4top, topN, 
                    "")
            }
            if (hs.any(!length(anno.type.2do), "kegg" %in% anno.type.2do)) {
                tbl.wj <- hs.wj_out(wj_out.pre, "kegg")
                hs.en(tbl.wj, s1 = s1_ncbi, u1 = u1_ncbi, et = "kegg", 
                  TERM2GENE = kegg.loaded[[org1]], org1 = org1, 
                  seed = seed)
            }
            if (hs.any(!length(anno.type.2do), "reactome" %in% 
                anno.type.2do)) {
                if ((length(s1_ncbi) < 100) && top) {
                  s1_ncbi <- if (is.list(s1_all_ncbi)) {
                    unlist(lapply(s1_all_ncbi, head, topN))
                  }
                  else head(s1_all_ncbi, topN)
                }
                tbl.wj <- hs.wj_out(wj_out.pre, "reactome")
                hs.en(tbl.wj, s1 = s1_ncbi, u1 = u1_ncbi, et = "reactome", 
                  org1 = org1, seed = seed)
            }
            if (hs.any(!length(anno.type.2do), "reaction" %in% 
                anno.type.2do)) {
                if ((length(s1_ncbi) < 100) && top) {
                  s1_ncbi <- if (is.list(s1_all_ncbi)) {
                    unlist(lapply(s1_all_ncbi, head, topN))
                  }
                  else head(s1_all_ncbi, topN)
                }
                tbl.wj <- hs.wj_out(wj_out.pre, "reaction")
                hs.en(tbl.wj, s1 = s1_ncbi, u1 = u1_ncbi, et = "reaction", 
                  org1 = org1, seed = seed)
            }
            if (hs.any(!length(anno.type.2do), any(hs.grepl(anno.type.2do, 
                "^(?i)go\\b")))) {
                if (length(anno.type.2do)) {
                  a1 <- hs.grepV(anno.type.2do, "^(?i)go\\b")
                  domain2do <- if (identical(a1, "go")) {
                    c("bp", "cc", "mf")
                  }
                  else hs.sub(a1, "^go[\\._]")
                }
                else domain2do <- c("bp", "cc", "mf")
                lapply(domain2do, function(od1) {
                  hs.hsgly("go.gene.4clusterProfiler.36_12_31_143101")(org1, 
                    ot = od1)
                  tbl.wj <- hs.wj_out(wj_out.pre, paste0("go.", 
                    od1))
                  hs.en(tbl.wj, s1 = s1_ncbi, u1 = u1_ncbi, et = "go", 
                    ot = od1, OrgDb = org_db, org1 = org1, seed = seed)
                })
            }
            if (0) {
                wj_out <- hs.wj_out(wj_out.pre, "keggModule")
                if (!file.exists(wj_out)) {
                  r1 <- enrichMKEGG(gene = s1, universe = u1, 
                    pAdjustMethod = "fdr", qvalueCutoff = 1.1, 
                    pvalueCutoff = 1.1) %=>% summary
                  hs.output(r1, wj_out)
                }
                hs.barplot(wj_out)
            }
            if (hs.all(org1 %in% c("human"), hs.any(!length(anno.type.2do), 
                "cell_markers" %in% anno.type.2do))) {
                if (!length(cellMarkers)) 
                  cellMarkers <<- local({
                    a1 <- hs.fp(hs.hsgly("CellMarker.35.11.12.120554")(), 
                      "Human_cell_markers.txt.gz") %=>% wk.fread
                    cn <- c("tissueType", "cancerType", "cellName")
                    if (0) {
                      a1[, {
                        .SD
                        hs.browser("35.11.12.205130")
                        .(cell = paste(unname(.BY), collapse = ", "), 
                          marker = list(hs.gre(geneID, "\\d+") %=>% 
                            unlist))
                      }, by = c(cn)][, `:=`(c(cn), {
                      })]
                    }
                    a1[, .(cell = paste(unname(.BY), collapse = ", "), 
                      marker = hs.gre(geneID, "\\d+") %=>% unlist), 
                      by = c(cn)][, `:=`(c(cn), {
                    })]
                  })
                tbl.wj <- hs.wj_out(wj_out.pre, "cell_markers")
                hs.en(tbl.wj, s1 = s1, u1 = u1, et = "cell_markers", 
                  TERM2GENE = cellMarkers, org1 = org1, seed = seed)
            }
            if (org1 %in% c("human") && (!length(anno.type.2do) || 
                "ncg" %in% anno.type.2do)) {
                wj_out <- hs.wj_out(wj_out.pre, "ncg")
                if (!file.exists(wj_out)) {
                  r1 <- enrichNCG(gene = s1, universe = u1, pAdjustMethod = "fdr", 
                    pvalueCutoff = 1.1, qvalueCutoff = 1.1) %=>% 
                    wk.dt.as
                  hs.output(r1, wj_out)
                }
                hs.barplot(wj_out)
            }
            if (org1 %in% c("human") && (!length(anno.type.2do) || 
                "dgn" %in% anno.type.2do)) {
                wj_out <- hs.wj_out(wj_out.pre, "dgn")
                if (!file.exists(wj_out)) {
                  r1 <- enrichDGN(gene = s1, universe = u1, pAdjustMethod = "fdr", 
                    pvalueCutoff = 1.1, qvalueCutoff = 1.1) %=>% 
                    wk.dt.as
                  hs.output(r1, wj_out)
                }
                hs.barplot(wj_out)
            }
            if (org1 %in% c("human") && (!length(anno.type.2do) || 
                "do" %in% anno.type.2do)) {
                tbl.wj <- hs.wj_out(wj_out.pre, "do")
                hs.en(tbl.wj, s1 = s1, u1 = u1, et = "do", org1 = org1, 
                  seed = seed)
            }
            if (!length(anno.type.2do) || "msigdb" %in% anno.type.2do) {
                if (org1 %in% c("human", "mouse")) {
                  msigdb <- hs.hsgly("数据.38_01_08_211229")(org1 = org1)[["gene"]]
                  for (i1 in seq_along(msigdb)) {
                    db1 <- msigdb[[i1]]
                    db1.name <- names(msigdb)[i1]
                    wj_out <- hs.wj_out(wj_out.pre, paste0("msigdb.", 
                      db1.name))
                    if (!file.exists(wj_out)) {
                      cat("doing ", db1.name, "\n", sep = "")
                      r1 <- enricher(gene = as.character(s1), 
                        universe = as.character(u1), TERM2GENE = db1, 
                        pAdjustMethod = "fdr", pvalueCutoff = 1.1, 
                        qvalueCutoff = 1.1) %=>% wk.dt.as
                      hs.output(r1, wj_out, org1 = org1)
                    }
                    hs.barplot(wj_out)
                  }
                }
            }
            if (0) {
                if (org1 %in% c("human")) {
                  db <- gProfiler.db
                  for (i1 in seq_along(db)) {
                    db1 <- db[[i1]]
                    db1.name <- names(db)[i1]
                    wj_out <- hs.wj_out(wj_out.pre, paste0("db.", 
                      db1.name))
                    if (!file.exists(wj_out)) {
                      cat("doing ", db1.name, "\n", sep = "")
                      A1 <- list(gene = as.character(s1), universe = as.character(u1), 
                        TERM2GENE = db1[["term2gene"]], pAdjustMethod = "fdr", 
                        pvalueCutoff = 1.1, qvalueCutoff = 1.1)
                      if ("term2name" %in% names(db1)) 
                        A1["TERM2NAME"] <- db1["term2name"]
                      r1 <- do.call(enricher, A1) %=>% wk.dt.as
                      hs.output(r1, wj_out, org1 = org1)
                    }
                    hs.barplot(wj_out)
                  }
                }
            }
        }
        else if (length(rnk)) {
            wk.dt.reset_name(rnk)
            if (anyDuplicated(rnk[[1]])) {
                rnk <- rnk[hs.clean(rnk[[1]], ot = "l"), V2[which.max(abs(V2))], 
                  by = V1]
                setnames(rnk, paste0("V", 1:ncol(rnk)))
            }
            pv.co <- 1
            if (ft == "ensembl") {
                switch(org1, human = {
                  rnk_ncbi <- rnk %=>% hs.rnk(x[[1]], x[[2]], 
                    ft = toupper(ft), tt = "ENTREZID", org1 = org1) %=>% 
                    wk.dt(names(x), x)
                  rnk %<=>% {
                    hs.rnk(x[[1]], x[[2]], ft = toupper(ft), 
                      org1 = org1) %=>% wk.dt(names(x), x)
                  }
                }, mouse = {
                  id.tbl <- wk.dt.subset(hs.hsgly("gene.info.36_09_21_151142")(), 
                    rnk[[1]], "gene.id_ensmebl")
                  rnk <- wk.dt.subset(rnk, id.tbl[, gene.id_ensmebl], 
                    1)
                  rnk[[1]] <- id.tbl[, as.character(gene.id_ncbi)]
                })
                setnames(rnk, paste0("V", 1:ncol(rnk)))
                if (anyDuplicated(rnk[[1]])) {
                  rnk <- rnk[hs.clean(rnk[[1]], ot = "l"), V2[which.max(abs(V2))], 
                    by = V1]
                }
                setnames(rnk, paste0("V", 1:ncol(rnk)))
            }
            else if (ft %in% "symbol") {
                rnk[, `:=`(V1, hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937")(V1))]
                rnk[, `:=`(V1, toupper(V1))]
                hs.hsgly("编号转换之表格.37_03_22_145251")(rnk, 
                  ft, "ncbi")
                if (0) {
                  rnk[hs.clean(gene_id_ncbi, ot = "i", inv = 1)]
                  hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937")("FLJ31104")
                }
                rnk[, `:=`(V1, {
                })]
                data.table::setcolorder(rnk, c("gene_id_ncbi", 
                  "V2"))
                wk.dt.reset_name(rnk)
                a1 <- rnk[, V1] %=>% strsplit(";")
                n1 <- sapply(a1, length)
                rnk <- rnk[, .(unlist(a1), rep(V2, n1))]
            }
            wk.dt.setorderv(rnk, "V2", -1)
            rnk_ncbi <- rnk <- hs.c(rnk[[1]], rnk[[2]])
            if (length(db_extra)) {
                if (ft != tt) {
                  db_extra %<=>% lapply(x, . %=>% {
                    a1 <- data.table::copy(x)
                    cn.37_10_28_173723 <- data.table::copy(names(a1))
                    a1 <- hs.hsgly("dt.id1_to_id2.37_03_22_145251")(a1, 
                      "symbol", "ncbi", id1.cn = cn.37_10_28_173723[2])
                    cn2rm.37_10_28_173445 <- names(a1)[c(-1, 
                      -ncol(a1))]
                    a1[, `:=`(c(cn2rm.37_10_28_173445), {
                    })]
                    setnames(a1, cn.37_10_28_173723)
                  })
                }
                for (i1 in seq_along(db_extra)) {
                  db1_name <- names(db_extra)[i1]
                  message("\t【", db1_name, "】")
                  wj_out.pre1 <- hs.fp(wj_out.pre, db1_name)
                  wj_out <- hs.wj_out(wj_out.pre, db1_name)
                  if (!hs.wj.exists(wj_out, size = 1)) {
                    db1 <- data.table::copy(db_extra[[i1]])
                    db1 %<=>% wk.dt.expand(x, names(x)[2])
                    db1[, `:=`(member, as.character(member))]
                    gsea.jg <- GSEA(rnk, TERM2GENE = db1, minGSSize = 5, 
                      pvalueCutoff = pv.co)
                    save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd", 
                      extN = 2) %=>% hs.wj)
                    r1 <- gsea.jg %=>% wk.dt.as
                    hs.output(r1, wj_out, gsea = 1)
                  }
                  hs.barplot(wj_out)
                }
                if (db_extra_only) 
                  return()
            }
            if (1) {
                tbl.wj <- hs.wj_out(wj_out.pre, "gsea.kegg")
                rd.wj <- hs.wjm(tbl.wj, ext = "rd", extN = 2)
                if (!hs.wj.exists(tbl.wj, size = 1)) {
                  message("\t【kegg】")
                  if (0) {
                    set.seed(36042921)
                    gsea.jg <- gseKEGG(rnk, pvalueCutoff = pv.co, 
                      use_internal_data = FALSE, seed = !0)
                  }
                  gsea.jg <- GSEA(rnk, TERM2GENE = kegg.loaded[[org1]][, 
                    .(set_id, gene)], TERM2NAME = unique(kegg.loaded[[org1]][, 
                    .(set_id, set)]), pAdjustMethod = "fdr", 
                    pvalueCutoff = pv.co)
                  hs.save(gsea.jg, wj = rd.wj %=>% hs.wj)
                  tbl <- gsea.jg %=>% as.data.table
                  hs.output(tbl, tbl.wj, gsea = 1, org1 = org1)
                }
                {
                  barplot.wj <- tbl.wj %=>% hs.wjm(x, ext = "pdf", 
                    extN = ifelse(hs.grepl(x, "\\.gz$"), 2, 1))
                  if (!file.exists(barplot.wj)) {
                    rd <- rd.wj %=>% hs.load
                    hs.barplot(tbl.wj, barplot.wj)
                  }
                }
                if (0) 
                  local({
                    gse1 <- hs.load(rd.wj)
                    tbl <- gse1 %=>% wk.dt.as
                    co1 <- 0.2
                    if (tbl[qvalues < co1, .N < 3]) 
                      co1 <- 0.2
                    pdf.wjj <- hs.wjm(tbl.wj, ext = 0, extN = 2)
                    gs2do <- tbl[qvalues < co1, ID]
                    if (!length(gs2do)) 
                      return()
                    gs2do %<=>% {
                      x[!file.exists(hs.wj(pdf.wjj, paste0(x, 
                        ".pdf")))]
                    }
                    lapply(gs2do, function(gs1) {
                      pdf.wj <- hs.wj(pdf.wjj, paste0(gs1, ".pdf"))
                      if (!file.exists(pdf.wj)) {
                        .pdf({
                          print(wk.gseaplot2(gse1, gs1, title = tbl[match(gs1, 
                            ID), paste0(Description, "\n", paste("Q:", 
                            format(qvalues, scientific = 1, digits = 2)))]))
                        }, wj = pdf.wj, width = 10, height = 10)
                      }
                    })
                  })
            }
            if (0) {
                wj_out <- hs.wj_out(wj_out.pre, "gsea.kegg_module")
                if (!file.exists(wj_out)) {
                  cat("\t【kegg module】\n")
                  gsea.jg <- gseMKEGG(rnk, pvalueCutoff = pv.co)
                  save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd"))
                  r1 <- gsea.jg %=>% summary
                  r1 %<=>% wk.dt.as
                  hs.output(r1, wj_out, gsea = 1)
                }
                hs.barplot(wj_out)
            }
            if (1) {
                c("BP", "CC", "MF") %=>% lapply(x, function(od1) {
                  wj_out <- hs.wj_out(wj_out.pre, paste0("gsea.go.", 
                    od1))
                  if (!hs.wj.exists(wj_out, size = 1)) {
                    hs.cat("\t【go，", od1, "】")
                    a <- hs.hsgly("go.gene.4clusterProfiler.36_12_31_143101")(org1, 
                      ot = od1)
                    gsea.jg <- GSEA(rnk, TERM2GENE = a[["term2gene"]], 
                      TERM2NAME = a[["term2name"]], minGSSize = 5, 
                      pvalueCutoff = pv.co)
                    if (0) {
                      clusterProfiler:::get_GO_data
                      clusterProfiler:::build_Anno
                      OrgDb <- org.Hs.eg.db
                      ont <- "BP"
                      keyType <- "ENTREZID"
                      GO_DATA <- clusterProfiler:::get_GO_data(OrgDb, 
                        ont, keyType)
                      hs.getter(GO_DATA, "EXTID2PATHID")
                      enrichGO
                      enricher
                      USER_DATA <- clusterProfiler:::build_Anno(TERM2GENE, 
                        TERM2NAME)
                      function(OrgDb, ont, keytype) {
                        GO_Env <- clusterProfiler:::get_GO_Env()
                        use_cached <- FALSE
                        ont2 <- NULL
                        if (exists("ont", envir = GO_Env, inherits = FALSE)) 
                          ont2 <- get("ont", envir = GO_Env)
                        if (exists("organism", envir = GO_Env, 
                          inherits = FALSE) && exists("keytype", 
                          envir = GO_Env, inherits = FALSE) && 
                          !is.null(ont2)) {
                          org <- get("organism", envir = GO_Env)
                          kt <- get("keytype", envir = GO_Env)
                          if (org == get_organism(OrgDb) && keytype == 
                            kt && (ont == ont2 || ont2 == "ALL") && 
                            exists("goAnno", envir = GO_Env, 
                              inherits = FALSE)) {
                            use_cached <- TRUE
                          }
                        }
                        if (use_cached) {
                          goAnno <- get("goAnno", envir = GO_Env)
                          if (!is.null(ont2) && ont2 != ont) {
                            goAnno <- goAnno[goAnno$ONTOLOGYALL == 
                              ont, ]
                          }
                        }
                        else {
                          OrgDb <- load_OrgDb(OrgDb)
                          kt <- keytypes(OrgDb)
                          if (!keytype %in% kt) {
                            stop("keytype is not supported...")
                          }
                          kk <- keys(OrgDb, keytype = keytype)
                          goterms <- AnnotationDbi::Ontology(GO.db::GOTERM)
                          if (ont != "ALL") {
                            goterms <- goterms[goterms == ont]
                          }
                          go2gene <- suppressMessages(AnnotationDbi::mapIds(OrgDb, 
                            keys = names(goterms), column = keytype, 
                            keytype = "GOALL", multiVals = "list"))
                          goAnno <- stack(go2gene)
                          colnames(goAnno) <- c(keytype, "GOALL")
                          goAnno <- unique(goAnno[!is.na(goAnno[, 
                            1]), ])
                          goAnno$ONTOLOGYALL <- goterms[goAnno$GOALL]
                          assign("goAnno", goAnno, envir = GO_Env)
                          assign("keytype", keytype, envir = GO_Env)
                          assign("ont", ont, envir = GO_Env)
                          assign("organism", get_organism(OrgDb), 
                            envir = GO_Env)
                        }
                        GO2GENE <- unique(goAnno[, c(2, 1)])
                        GO_DATA <- build_Anno(GO2GENE, get_GO2TERM_table())
                        goOnt.df <- goAnno[, c("GOALL", "ONTOLOGYALL")] %=>% 
                          unique
                        if (!is.null(ont2) && ont2 == "ALL") {
                          return(GO_DATA)
                        }
                        goOnt <- goOnt.df[, 2]
                        names(goOnt) <- goOnt.df[, 1]
                        assign("GO2ONT", goOnt, envir = GO_DATA)
                        return(GO_DATA)
                      }
                    }
                    save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd"))
                    r1 <- gsea.jg %=>% wk.dt.as
                    hs.output(r1, wj_out, gsea = 1, org1 = org1)
                  }
                  wj_out %=>% hs.wjm(x, ext = "pdf", extN = 1 + 
                    hs.grepl(x, "\\.gz$")) %=>% if (!file.exists(x)) {
                    hs.barplot(wj_out)
                  }
                })
            }
            if (1) {
                wj_out <- hs.wj_out(wj_out.pre, "gsea.reactome")
                if (!hs.wj.exists(wj_out, size = 1)) {
                  hs.cat("\t【Reactome】")
                  gsea.jg <- gsePathway(rnk_ncbi, pAdjustMethod = "fdr", 
                    pvalueCutoff = pv.co, organism = org1)
                  save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd", 
                    extN = 2))
                  r1 <- gsea.jg %=>% wk.dt.as
                  hs.output(r1, wj_out, gsea = 1, org1 = org1)
                }
                wj_out %=>% hs.wjm(x, ext = "pdf", extN = 1 + 
                  hs.grepl(x, "\\.gz$")) %=>% if (!file.exists(x)) {
                  hs.barplot(wj_out)
                }
            }
            if (0) {
                wj_out <- hs.wj_out(wj_out.pre, "gsea.reaction")
                if (!hs.wj.exists(wj_out, size = 1)) {
                  hs.cat("\t【Reaction】")
                  a <- hs.hsgly("reactome.gene.4clusterProfiler.36_12_31_171255")(org1, 
                    type = "reaction")
                  gsea.jg <- GSEA(rnk, TERM2GENE = a[["term2gene"]], 
                    TERM2NAME = a[["term2name"]], minGSSize = 5, 
                    pvalueCutoff = pv.co)
                  save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd", 
                    extN = 2))
                  r1 <- gsea.jg %=>% as.data.table
                  hs.output(r1, wj_out, gsea = 1, org1 = org1)
                }
                a <- wk.fread(wj_out)
                if (!a[, Description %=>% anyDuplicated]) 
                  hs.barplot(wj_out)
            }
            if (org1 %in% "human") 
                local({
                  return()
                  wj_out <- hs.wj_out(wj_out.pre, "gsea.cell_markers")
                  if (!hs.wj.exists(wj_out, size = 1)) {
                    hs.say("\t【cell markers】")
                    gsea.jg <- GSEA(rnk, TERM2GENE = cellMarkers, 
                      minGSSize = 5, pvalueCutoff = pv.co)
                    save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd", 
                      extN = 2))
                    r1 <- gsea.jg %=>% wk.dt.as
                    hs.output(r1, wj_out, gsea = 1)
                  }
                  hs.barplot(wj_out)
                })
            if (0) {
                wj_out <- hs.wj_out(wj_out.pre, "gsea.ncg")
                if (!file.exists(wj_out)) {
                  cat("\t【network of cancer gene】\n")
                  gsea.jg <- gseNCG(rnk, pAdjustMethod = "fdr", 
                    pvalueCutoff = pv.co)
                  save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd"))
                  r1 <- gsea.jg %=>% summary
                  r1 %<=>% wk.dt.as
                  hs.output(r1, wj_out, gsea = 1)
                }
                hs.barplot(wj_out)
            }
            if (0) {
                wj_out <- hs.wj_out(wj_out.pre, "gsea.dgn")
                if (!file.exists(wj_out)) {
                  gsea.jg <- gseDGN(rnk, pAdjustMethod = "fdr", 
                    pvalueCutoff = pv.co)
                  save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd"))
                  r1 <- gsea.jg %=>% summary
                  r1 %<=>% wk.dt.as
                  hs.output(r1, wj_out, gsea = 1)
                }
                hs.barplot(wj_out)
            }
            if (0) {
                wj_out <- hs.wj_out(wj_out.pre, "gsea.do")
                if (!file.exists(wj_out)) {
                  gsea.jg <- gseDO(rnk, pAdjustMethod = "fdr", 
                    pvalueCutoff = pv.co)
                  save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd"))
                  r1 <- gsea.jg %=>% summary
                  r1 %<=>% wk.dt.as
                  hs.output(r1, wj_out, gsea = 1)
                }
                hs.barplot(wj_out)
            }
            if (org1 %in% c("human", "mouse")) 
                local({
                  return()
                  i.db <- seq_along(msigdb)
                  for (i1 in i.db) {
                    db1 <- msigdb[[i1]]
                    db1.name <- names(msigdb)[i1]
                    wj_out <- hs.wj_out(wj_out.pre, paste0("gsea.msigdb.", 
                      db1.name))
                    if (!hs.wj.exists(wj_out)) {
                      message("\t【", db1.name, "】")
                      hs.browser("37.10.06.143459")
                      gsea.jg <- GSEA(rnk, TERM2GENE = db1, pAdjustMethod = "fdr", 
                        pvalueCutoff = pv.co)
                      save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd"))
                      r1 <- gsea.jg %=>% wk.dt.as
                      hs.output(r1, wj_out, gsea = 1, org1 = org1)
                    }
                    hs.barplot(wj_out)
                  }
                })
            if (org1 %in% c("human")) {
                db <- gProfiler.db
                for (i1 in seq_along(db)) {
                  db1.name <- names(db)[i1]
                  db1 <- db[[i1]]
                  wj_out <- hs.wj_out(wj_out.pre, paste0("gsea.", 
                    db1.name))
                  if (!hs.wj.exists(wj_out)) {
                    message("\t【", db1.name, "】")
                    A1 <- list(rnk, TERM2GENE = db1[["term2gene"]], 
                      pAdjustMethod = "fdr", pvalueCutoff = pv.co)
                    if ("term2name" %in% names(db1)) 
                      A1["TERM2NAME"] <- db1["term2name"]
                    if (0) {
                      A1[["BPPARAM"]] <- BiocParallel::registered()[["MulticoreParam"]]
                    }
                    gsea.jg <- do.call(GSEA, A1)
                    save(gsea.jg, file = wj_out %=>% hs.wjm(ext = "rd"))
                    r1 <- gsea.jg %=>% wk.dt.as
                    hs.output(r1, wj_out, gsea = 1, org1 = org1)
                  }
                  hs.barplot(wj_out)
                }
            }
        }
        dirname(paste0(wj_out.pre, "a1"))
    }
})
"39_01_11_164432" <- function(wj, wj2 = "/dev/shm/wk/tmp", topN = 5) {
    en <- lapply(wj, wk.fread)
    top_term <- lapply(en, . %>% {
        .[p.adjust < 0.05, head(Description, topN)]
    }) %>% unlist %>% unique
    if (!length(top_term)) 
        return()
    dt <- wk.dt(term = rep(top_term, length(en)))
    dt[, `:=`("group", rep(names(en), each = length(top_term)))]
    a <- sapply(en, . %>% {
        wk.dt.subset.na(., top_term, "Description")[, p.adjust]
    })
    a <- -log10(a)
    dt[, `:=`("-Log10FDR", a)]
    rownames(a) <- top_term
    ci <- if (length(en) > 1) {
        hs.hsgly("层次聚类/fastcluster.36_08_12_231620")(a, 
            ot = "o")
    }
    else seq(ncol(a))
    dt[, `:=`(group, hs.factor(group, colnames(a)[ci], N = 0))]
    ri <- hs.hsgly("层次聚类/fastcluster.36_08_12_231620")(a, 
        ot = "o", side = "row")
    dt[, `:=`(term, hs.factor(term, rownames(a)[ri], N = 0))]
    a <- sapply(en, . %>% {
        wk.dt.subset.na(., top_term, "Description")[, GeneRatio]
    })
    dt[, `:=`("GeneRatio", sapply(a, evalStr))]
    dt[, `:=`(group, group %=>% {
        levels(x) %<=>% paste0("Group ", x)
        x
    })]
    p1 <- ggplot2::ggplot(dt, ggplot2::aes(group, term, size = GeneRatio, 
        color = `-Log10FDR`)) + ggplot2::geom_point() + ggplot2::labs(x = NULL, 
        y = NULL) + ggplot2::theme_bw() + ggplot2::theme(axis.text.x = element_text(angle = 45, 
        hjust = 1), panel.grid = element_blank())
    dt[, term %=>% levels %=>% nchar %=>% max]
    .figs(print(p1), wj2, width = length(en) * 0.25 + 10 + 5 + 
        dt[, term %=>% levels %=>% nchar %=>% max %=>% {
            x * 0.1
        }], height = length(top_term) * 0.5 + 5)
}
"38_08_29_221614" <- function(wjj1) {
    wj <- hs.wjj.ls(wjj1, "\\.tsv\\.gz$")
    for (i in seq_along(wj)) {
        wj1 <- wj[i]
        hs.hsgly("genes.understand.35.11.12.182333", "hs.barplot")(wj1)
    }
}
"37_10_28_162549" <- function(gs, rnk, wjj2 = ".") {
    hs.hsgly("genes.understand.35.11.12.182333")(rnk = rnk, ft = "symbol", 
        wj_out.pre = hs.fp(wjj2, "en/"), db_extra = gs)
}
"36_09_21_161440" <- local({
    hs.deg <- function(gene, qv, log2fc, fx, qv.co = 0.05, log2fc.co, 
        fc.co) {
        i1 <- qv < qv.co
        if (!missing(fx)) 
            i1 <- i1 & (sign(log2fc) %in% sign(fx))
        if (!missing(log2fc.co)) 
            i1 <- i1 & (abs(log2fc) > log2fc.co)
        if (!missing(fc.co)) 
            i1 <- i1 & (abs(log2fc) > log2(fc.co))
        gene[i1]
    }
    function(de, jg.wjj = ".", gene.cn = hs.hsgly("获取列名.39_03_20_154050")(de, 
        cn_wanted = "gene"), log2fc.cn = hs.hsgly("获取列名.39_03_20_154050")(de, 
        cn_wanted = "fc"), q.cn = hs.hsgly("获取列名.39_03_20_154050")(de, 
        cn_wanted = "q"), org1 = "human", fc.co.v = c(1.1, 1.2, 
        1.5, 2, 3, 4, 8), q.co.v = c(0.05, 0.1), n_core = Ncpu()[1]) {
        hs.hsgly("基因/富集分析/用clusterProfiler/预载.39_03_23_095724")(org1 = org1)
        if (!inherits(de, "data.table")) 
            de %<=>% wk.dt.as
        u1 <- de[!is.na(get(q.cn)), get(gene.cn)]
        cmd.L <- expandingSequence()
        rnk <- de[hs.clean(get(log2fc.cn), ot = "l"), .(get(gene.cn), 
            get(log2fc.cn))]
        cmd.L[["add"]](bquote(hs.hsgly("基因/富集分析/用clusterProfiler/基于.35.11.12.182333")(rnk = .(rnk), 
            wj_out.pre = .(hs.wjj(jg.wjj, "gene_rank", "")), 
            org1 = .(org1))))
        hs.for(fx1, c("up", "down", "dif"), fc.co, fc.co.v, q.co, 
            q.co.v, {
                s1 <- hs.deg(de[[gene.cn]], de[[q.cn]], de[[log2fc.cn]], 
                  switch(fx1, up = 1, down = -1, dif = c(-1, 
                    1)), fc.co = fc.co, qv.co = q.co) %=>% hs.clean
                cmd.L[["add"]](bquote(hs.hsgly("基因/富集分析/用clusterProfiler/基于.35.11.12.182333")(.(s1), 
                  .(u1), wj_out.pre = .(hs.wj(jg.wjj, paste0("qv_", 
                    q.co, "__fc_", fc.co), fx1)), org1 = .(org1))))
            })
        wk.mclapply(cmd.L[["get"]](), function(cmd1) .try(eval(cmd1)), 
            mc.cores = n_core)
    }
})
"38_06_23_094846" <- local({
    hs.oddsRatio.38_06_23_094118 <- function(r1, r2, no_inf = 0) {
        v <- hs.strsplit.v(c(r1, r2), split = "/") %>% as.numeric
        v[2] <- v[2] - v[1]
        v[3] <- v[3] - v[1]
        v[4] <- v[4] - sum(v[1:3])
        if (no_inf && any(v[2:3] == 0)) {
            if (v[1] == 0) {
                hs.browser("38.06.23.145040")
            }
            if (v[2] == 0) {
                v[2] <- 1
                if (v[1] < 1) 
                  hs.browser("38.06.23.150015")
                v[1] <- v[1] - 1
                v[3] <- v[3] + 1
                if (v[4] < 1) 
                  hs.browser("38.06.23.150025")
                v[4] <- v[4] - 1
            }
            if (v[3] == 0) {
                v[3] <- 1
                if (v[1] < 1) 
                  hs.browser("38.06.23.145055")
                v[1] <- v[1] - 1
                v[2] <- v[2] + 1
                if (v[4] < 1) 
                  hs.browser("38.06.23.150035")
                v[4] <- v[4] - 1
            }
        }
        m1 <- matrix(v, 2)
        fisher.test(m1)[["estimate"]]
    }
    function(en.wj) {
        wk.mclapply(en.wj, function(wj1) {
            message(wj1)
            cn <- cn.hs.wj(wj1)
            if ("oddsRatio" %in% cn) 
                return()
            if (!(c("GeneRatio", "BgRatio") %all.in% cn)) 
                hs.browser("38.06.23.131913")
            wj1.bf <- paste0(wj1, ".bf")
            if (file.exists(wj1.bf)) 
                hs.browser("38.06.23.130736")
            en <- wk.fread(wj1)
            file.rename(wj1, wj1.bf)
            or <- en[, .(GeneRatio, BgRatio)] %>% apply(1, . %>% 
                {
                  hs.oddsRatio.38_06_23_094118(.[1], .[2])
                })
            if (length(or) != en[, .N]) 
                hs.browser("38.06.23.132222")
            en[, `:=`(oddsRatio, or)]
            wk.dt.col.reorder(en, cn.pattern = "(?i)ratio")
            wk.dt.fwrite(en, wj1)
            hs.wj.rm(wj1.bf)
        })
    }
})
"39_03_23_095724" <- local({
    kegg.loaded <- list()
    go.loaded <- list()
    function(org1 = "human", db = c("go", "kegg")) {
        hs.require(c("clusterProfiler", "topGO", "org.Hs.eg.db", 
            "org.Mm.eg.db", "DOSE", "ReactomePA", "RColorBrewer", 
            "ontologyIndex", "Rgraphviz"))
        if (!exists("preloaded.39_03_23_105643", envir = .GlobalEnv)) 
            assign("preloaded.39_03_23_105643", 0, envir = .GlobalEnv)
        if ("kegg" %in% db) {
            if (org1 %not.in% names(kegg.loaded)) {
                hs.msg("预加载kegg", 0)
                kegg.loaded[[org1]] <<- hs.hsgly("kegg/通路/基因.36.03.18.133126")(org1 = org1)
            }
        }
        if ("go" %in% db) {
            if (c("bp", "cc", "mf") %not.all.in% names(go.loaded[[org1]])) {
                hs.msg("预加载go", 0)
                local({
                  domain2do <- c("bp", "cc", "mf") %not% names(go.loaded[[org1]])
                  null <- lapply(domain2do, . %=>% {
                    message(x)
                    a <- hs.hsgly("go.gene.4clusterProfiler.36_12_31_143101")(org1, 
                      ot = x)
                    go.loaded[[org1]][[x]] <<- a
                    {
                    }
                  })
                })
            }
        }
        if (0) {
            if (hs.all(org1 %in% c("human"), !length(gProfiler.db))) {
                if (preload) {
                  hs.msg("预加载 gProfiler")
                  gProfiler.db <<- local({
                    gmt_wj <- hs.hsgly("gProfiler.data.37_01_06_195308")()
                    wj2 <- hs.fp(dirname(gmt_wj[1]), "a.rd")
                    if (wj.update.reporter(wj2, gmt_wj)) {
                      gProfiler.db <- hs.hsgly("gProfiler.data.37_01_06_195308")() %>% 
                        {
                          hs.c(paste0("gProfiler.", hs.wjm(., 
                            dn = 0, ext = 0, extN = 2) %>% hs.gsub(":", 
                            ".")), .)
                        } %>% lapply(list.hs.gmt)
                      hs.save(gProfiler.db, wj = wj2)
                    }
                    hs.load(wj2)
                  })
                }
                else stop("[38_03_27_120413]请先预加载gProfiler ( org1 = ", 
                  .q("xxx"), ")")
            }
        }
        hs.hsgly("gene.id.converter.using_hgnc.36.01.22.160941")
        assign("preloaded.39_03_23_105643", 1, envir = .GlobalEnv)
    }
})
"38_01_06_171637" <- function(m1, sample_group = {
}, gs_db = c("go", "kegg"), gs_l = {
}, org1 = "human", gt = "symbol", wjj2 = {
}, inclusive = length(gs_l), subset_size = 2000) {
    hs.require("GSVA")
    if (org1 %not.in% c("human", "mouse")) 
        hs.browser("38.01.06.171734")
    rownames(m1) %<>% {
        hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937")(.)
    }
    if (!length(gs_l)) 
        gs_l <- hs.catch({
            {
                kegg <- hs.hsgly("kegg/通路/基因.36.03.18.133126")(org1)
                hs.throw("kegg", wk.dt.tapply(kegg, "gene.id_symbol", 
                  "set"))
            }
            hs.lapply("bp,cc,mf", function(ot) {
                go <- hs.hsgly("go.gene.4clusterProfiler.36_12_31_143101")(org1, 
                  ot = ot, gt = gt)
                term.gene <- go[["term2gene"]] %>% {
                  tapply(.[, "member"], .[, "group"], c)
                }
                names(term.gene) <- go[["term2name"]] %>% {
                  i1 <- match(names(term.gene), .[, "term"])
                  .[i1, "name"]
                }
                hs.throw(paste0("go.", ot), term.gene)
            })
        }, named = 1)
    if (length(sample_group)) {
        total <- subset_size
        if (total < length(sample_group)) {
            howmany <- local({
                tbl <- hs.table(sample_group) %>% sort
                howmany <- rep(0, length(tbl))
                for (i1 in seq_along(tbl)) {
                  n1 <- tbl[i1]/sum(tbl[seq(i1, length(tbl))]) * 
                    (total - sum(head(howmany, i1)))
                  howmany[i1] <- ifelse(n1 < 30, min(tbl[i1], 
                    30), n1)
                }
                hs.c(names(tbl), round(howmany))
            })
            if (0) {
                names(howmany) %>% {
                  wk.dt(howmany[.], hs.table(sample_group)[.])
                } %>% .s
            }
            ci2do <- tapply(seq_len(ncol(m1)), sample_group, 
                c) %>% {
                sapply(names(.), function(gn1) {
                  set.seed(380109.233703)
                  sort(hs.sample(.[[gn1]], howmany[gn1]))
                })
            } %>% hs.unlist %>% sort
            m1 <- m1[, ci2do]
            sample_group <- sample_group[ci2do]
            if (length(wjj2)) 
                wk.dt.fwrite(colnames(m1), hs.fp(wjj2, "barcode_seleted.tsv.gz"))
        }
    }
    if (0) {
        gs <- gs_l[[1]]
        gs <- gs_l[[2]]
        length(gs)
        gsva <- list(kegg = a)
        .str(gsva)
        hs.save(m1, gs_l, wj = "~/a.rd")
        "~/a.rd" %>% .ls
        "~/a.rd" %>% wk.rsync.out.same_place(ssh = "hw1.n9")
        load("~/a.rd")
        hs.save(gsva, wj = "~/swxx/xm/db/192.168.130.100/外包数据/JN晶能/P20220100715/报告/12.GSVA分析/gsva.rd")
        "~/swxx/xm/db/192.168.130.100/外包数据/JN晶能/P20220100715/报告/12.GSVA分析/gsva.rd" %>% 
            wk.rsync.in.same_place(ssh = "hw1.n9")
    }
    nc <- hs.switch(Sys.info()["nodename"], "APT-SH-DT-716", 
        2, {
            a <- Ncpu()[1]/2
            if (a < 3) 
                a <- Ncpu()[2] - 1
            a
        })
    rn <- if (inclusive) {
        toupper(rownames(m1)) %>% {
            hs.c(., strsplit(., "[;\\|]"))
        } %>% wk.dt.fromList
    }
    else wk.dt(member = toupper(rownames(m1)))
    rn <- rownames(m1) %>% {
        a <- strsplit(., "[;\\|]")
        a <- wk.dt(rep(., sapply(a, length)), unlist(a))
        wk.dt.setnames(a, c("original", "single"))
    }
    hs.lapply(gs_l, function(gs) {
        if (inclusive) {
            gs %<>% lapply(. %>% {
                rn[hs.str_match(., single), unique(original)]
            })
        }
        else {
            hs.browser("38.09.15.222055")
            member <- hs.unlist(gs)
            i <- member %in% rn[, member]
            member %not% rn[, member]
            gs <- tapply(member[i], rep(names(gs), sapply(gs, 
                length))[i], list)
        }
        l <- sapply(gs, length)
        i <- (l >= 5) & (l <= 500)
        gs <- gs[i]
        if (0) 
            if (!missing(sample_group)) {
                all(colnames(gsva) == names(sample_group2))
                hs.require("limma")
                design <- model.matrix(~sample_group2)
                fit1 <- lmFit(gsva, design)
                fit1 <- eBayes(fit1)
                result_limma <- topTable(fit1, n = Inf) %>% as.data.table("gene_set")
                result_limma[adj.P.Val < 0.05, .N]
                fit1 <- apply(gsva, 1, . %>% {
                  aov(. ~ sample_group2)
                })
                pv <- sapply(fit1, function(x1) anova(x1)[1, 
                  "Pr(>F)"])
                pv <- apply(gsva, 1, function(x1) kruskal.test(x1, 
                  sample_group2)[["p.value"]])
                qv <- p.adjust(pv, "fdr")
                hs.table(qv < 0.05)
            }
        if (length(gs)) {
            (GSVA::gsva)(m1, gs, parallel.sz = nc)
        }
    })
}
"37_04_29_160710" <- function(geneid, ft = hs.hsgly("编号种类猜测.37_04_28_165736")(geneid)) {
}
"37_04_28_165736" <- function(gene_id, sep = ";", to_sep = 1) {
    id_all <- as.character(gene_id)
    if (to_sep) 
        id_all <- strsplit(id_all, sep) %>% unlist
    id_all <- unique(id_all) %>% tolower
    if (mean(hs.grepl(id_all, "^ens.*\\d{8,}$")) > 0.5) {
        "ensembl"
    }
    else if (mean(hs.grepl(id_all, "^\\d+$")) > 0.5) {
        "ncbi"
    }
    else "symbol"
}
"37_04_28_165424" <- function(gene_id, ft, tt, org1 = "human", 
    sep = ";") {
    gene_id <- as.character(gene_id)
    db <- hs.hsgly("编号种类转换库.37_03_29_103455")(ft, 
        tt, org1 = org1)
    db.c1 <- db[[1]] %>% toupper
    gene_id_list <- as.character(gene_id) %>% toupper %>% strsplit(sep)
    a1 <- data.table(rep(gene_id, sapply(gene_id_list, length)), 
        unlist(gene_id_list))
    i1 <- a1[, match(V2, db.c1)]
    a1[, `:=`(V3, db[[2]][i1])]
    jg <- a1[hs.clean(V3, ot = "l"), {
        if (uniqueN(V3) == 1) 
            V3[1]
        else paste(V3, collapse = ";")
    }, by = V1]
    hs.c(gene_id, hs.c(jg[[1]], jg[[2]])[gene_id])
}
"37_03_29_103455" <- local({
    org1.id <- {
    }
    hs.tt_join <- function(dt1) {
        cn1 <- names(dt1)[1]
        cn2 <- names(dt1)[2]
        if (anyDuplicated(dt1[[cn1]])) {
            dt1 <- dt1[, paste(get(cn2), collapse = ";"), by = c(cn1)]
            setnames(dt1, c(cn1, cn2))
        }
        dt1
    }
    ncbi.symbol <- function(org1, out = "dt", update_db = 0) {
        jg.wj <- .mfp.swxx(paste0("37_03_31_164408/", org1, ".tsv.gz"))
        if (out == "wj") 
            return(jg.wj)
        if (update_db) 
            jg.wj %=>% hs.wj.rm
        if (file.exists(jg.wj)) {
            wk.fread(jg.wj)
        }
        else {
            jg <- local({
                gene_info <- hs.hsgly("species.genes.36_12_28_140507")(org1, 
                  what = "all", symbol2upper_case = 0)
                a1 <- gene_info[Symbol_from_nomenclature_authority %not.in% 
                  "-", .(GeneID, Symbol = Symbol_from_nomenclature_authority)]
                switch(org1, human = {
                  aa <- hs.hsgly("hgnc/用总表更新symbol.38_01_03_204741")(a1[, 
                    Symbol], recent = 1)
                  i1 <- aa != a1[, Symbol]
                  if (any(i1)) a1[i1, `:=`(Symbol, aa[i1])]
                }, mouse = message("[37_04_20_133639|@待办@]小鼠可以考虑用mgi的信息来更新号"))
                setnames(a1, paste0("gene_id_", c("ncbi", "symbol")))
                a2 <- gene_info[Symbol_from_nomenclature_authority %in% 
                  "-", .(GeneID, Symbol)]
                if (0) {
                  gene_info[GeneID %in% 3337213]
                  aa <- hs.hsgly("hgnc/用总表更新symbol.38_01_03_204741")(a2[, 
                    Symbol], recent = 1)
                  i1 <- aa != a2[, Symbol]
                  if (any(i1)) 
                    a2[i1, `:=`(Symbol, aa[i1])]
                  a2[i1]
                  aa[i1]
                }
                wk.dt.rbind(a1, a2, U = 1)
            })
            if (0) {
                a1[, "3337213" %in% gene_id_ncbi]
                jg[, "3337213" %in% gene_id_ncbi]
                jg1 <- local({
                  wj1 <- hs.hsgly("ftp.ncbi.nih.gov/gene/DATA.37_04_29_161915")("gene2accession.gz")
                  cmd1 <- .sys("cat_wk", wj1, "| perl -e", .q(paste0("\nwhile( <>) {\n  $flag = 0;\n  if( /^", 
                    org1.id, "\\t/) {\n    chomp;\n    @F = split( \"\t\");\n    print( $F[ 1], \"\t\", $F[ 15], \"\n\");\n    $flag = 1;\n  } else {\n    if( $flag) {\n      exit;\n    }\n  }\n}\n")), 
                    "| uniq | sort -n | uniq")
                  a1 <- wk.fread(cmd = cmd1, cn = 0)
                  setnames(a1, cn.hs.wj(wj1)[c(2, 16)])
                  a1[GeneID %not.in% jg[, gene_id_ncbi], .(GeneID, 
                    Symbol)]
                })
                if (jg1[, .N]) {
                  jg1[, `:=`(Symbol, hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937")(Symbol, 
                    org1 = org1, recent = 1))]
                  jg <- wk.dt.rbind(jg1, jg2, U = 1)
                }
            }
            switch(org1, human = {
                if (0) {
                  a <- jg[, hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937")(gene_id_symbol, 
                    org1 = org1, recent = 1)]
                  a[names(a) != a]
                  jg[, "ARNT" %in% gene_id_symbol]
                  hs.getter(hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937"), 
                    "db")[[org1]]["ACP1"]
                }
                if (0) {
                  hgnc <- hs.hsgly("gene.id_hgnc.35.12.06.111314")()
                  a1 <- hgnc[hs.clean(`NCBI Gene ID`, ot = "l"), 
                    .(`NCBI Gene ID`, `Approved symbol`)]
                  setnames(a1, names(jg))
                  jg %<>% rbind(a1) %>% unique
                  wk.dt.rbind(jg, a1, U = 1)
                  a1[`gene_id,ncbi` %not.in% jg[, `gene_id,ncbi`]]
                }
            })
            wk.dt.fwrite(jg, jg.wj)
            jg
        }
    }
    ens.symbol <- function(org1) {
        jg1 <- local({
            a1 <- hs.hsgly("ens,symbol.37_09_14_164105")(org1)
            if (org1 %in% "human") {
                cn2 <- names(a1)[2]
                a2 <- a1[, hs.hsgly("hgnc/用总表更新symbol.38_01_03_204741")(get(cn2), 
                  recent = 1)]
                i1 <- a2 != names(a2)
                if (0) {
                  wk.dt.subset.na(a1, "ENSG00000102125")
                  a2["TAZ"]
                  a2["U3"]
                }
                if (any(i1)) {
                  a1[i1, `:=`(c(cn2), a2[i1])]
                }
            }
            a1
        })
        jg2 <- local({
            ens_ncbi <- hs.hsgly("编号种类转换库.37_03_29_103455")("ens", 
                "ncbi", org1)
            ncbi_symbol <- hs.hsgly("编号种类转换库.37_03_29_103455")("ncbi", 
                "symbol", org1)
            if (0) {
                a1 <- hs.hsgly("dt.id1_to_id2.37_03_22_145251")(ens_ncbi, 
                  dt2 = ncbi_symbol, id1.cn = "gene_id_ncbi")
            }
            a1 <- merge(ens_ncbi, ncbi_symbol)
            a1[, unique(gene_id_symbol) %>% sort %>% paste(collapse = ";"), 
                by = gene_id_ensembl]
        })
        setnames(jg2, names(jg1))
        jg <- rbind(jg1, jg2) %>% unique
        if (0) {
            wk.dt.subset.na(jg1, "ENSG00000238297")
            wk.dt.subset.na(jg2, "ENSG00000238297")
            wk.dt.subset.na(a1, "ENSG00000238297")
        }
        if (0) {
            a <- jg[, {
                if (.N > 1) 
                  .(gene_id_symbol, gene_id_symbol[which.min(nchar(gene_id_symbol))])
            }, by = gene_id_ensembl]
        }
        a <- jg[, {
            if (.N > 1) 
                gene_id_symbol[which.min(nchar(gene_id_symbol))]
            else gene_id_symbol
        }, by = gene_id_ensembl]
        setnames(a, names(jg))
        if (0) {
            "/Volumes/tf500/db/swxx/sj/db/ncbi/gene/gene2ensembl.gz" %>% 
                .head
            a <- wk.fread(cmd = .sys("cat_wk", "/Volumes/tf500/db/swxx/sj/db/ncbi/gene/gene2ensembl.gz", 
                "| perl -e", .q("\nwhile( <>) {\n  @F = split( \"\t\");\n  if( $F[ 2] eq \"ENSG00000102125\") {\n    print;\n  }\n}\n")), 
                cn = 0)
            .sh(a)
            lapply(a, unique)
            wk.dt.subset.na(a, "ENSG00000102125")
            a1[external_gene_name %in% "U3", .N]
            a1[external_gene_name %in% "LAP3"]
        }
        a
    }
    ncbi.ens <- function(org1, out = "dt", update_db = 0) {
        wj_in <- hs.hsgly("gene_id.ncbi,ensembl.36.06.08.154157")(org1 = org1, 
            update_db = update_db)
        if (out == "wj") 
            return(wj_in)
        sj <- wk.fread(wj_in)
        wk.dt.reset_name(sj)
        sj <- sj[hs.clean(V1, ot = "i")]
        if (0) {
            sj[hs.grepl(V2, "ENSG00000237402")]
        }
        if (org1 == "human") {
            a1 <- hs.hsgly("gene.id_hgnc.35.12.06.111314")(update_db = update_db)
            sj_hgnc <- lapply(c("NCBI Gene ID", "NCBI Gene ID(supplied by NCBI)"), 
                function(id1) {
                  a <- lapply(c("Ensembl gene ID", "Ensembl ID(supplied by Ensembl)"), 
                    function(id2) {
                      a1[hs.clean(get(id1), ot = "l") & hs.clean(get(id2), 
                        ot = "l"), .(get(id1), get(id2))]
                    })
                  rbindlist(a) %>% unique
                }) %>% rbindlist %>% unique
            setnames(sj_hgnc, names(sj))
            sj %=>% rbind(x, sj_hgnc) %=>% unique
        }
        if (0) {
            sj_ens <- hs.hsgly("ensembl.ncbi.35.06.18.103558")(org1 = org1)
            sj <- unique(rbind(sj, unique(sj_ens[, .(V1 = entrezgene_id, 
                V2 = ensembl_gene_id)])))
        }
        setnames(sj, paste0("gene_id_", c("ncbi", "ensembl")))
    }
    ncbi.mgi_symbol <- function(...) {
        wj_in <- "~/swxx/sj/db/ncbi/gene/mouse/Mus_musculus.gene_info.gz"
        dz1 <- "ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/GENE_INFO/Mammalia/Mus_musculus.gene_info.gz"
        if (wj.update.reporter(wj_in)) {
            wj.updater(wj_in, dz1)
        }
        {
            ci <- match(c("GeneID", "Symbol_from_nomenclature_authority"), 
                cn.hs.wj(wj_in))
            cmd1 <- .sys("~/org/dm/sh/rj/cat_wk", wj_in, "| perl -e", 
                .q(paste0("{\nwhile( <>) {\n  chomp;\n  @F = split( \"\\t\");\n  if( $F[ ", 
                  ci[2], "] ne \"-\") {\n    print( $F[ ", ci[1] - 
                    1, "], \"\\t\", $F[ ", ci[2] - 1, "], \"\\n\");\n  }\n}\n}")))
            jg <- wk.fread(cmd = cmd1, cn = 1)
            setnames(jg, paste0("gene_id,", c("ncbi", "symbol_mgi")))
            return(jg)
        }
        if (0) {
            wj_in.content <- fread(wj_in)
            hs.require("RcisTarget")
            motifRankings <- importRankings("~/swxx/sj/db/cistarget/databases/mus_musculus/mm10/refseq_r80/mc9nr/gene_based/mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather")
            motifRankings %>% attr("rankings") %>% colnames %>% 
                (`%in%`(wj_in.content[, Symbol])) %>% mean
            motifRankings %>% attr("rankings") %>% colnames %>% 
                (`%in%`(wj_in.content[, Symbol_from_nomenclature_authority])) %>% 
                mean
            motifRankings %>% attr("rankings") %>% colnames %>% 
                hs.grepV("(?i)ATP6")
            a[, "-" %in% Symbol_from_nomenclature_authority]
            b <- a[(Symbol_from_nomenclature_authority %not.in% 
                "-") & (Symbol %not.in% "-") & (Symbol != Symbol_from_nomenclature_authority), 
                .(GeneID, Symbol, Symbol_from_nomenclature_authority, 
                  dbXrefs)]
            b[, Symbol_from_nomenclature_authority %and% (motifRankings %>% 
                attr("rankings") %>% colnames)]
            motifRankings %>% attr("rankings") %>% colnames
        }
    }
    ncbi.uniprot <- function(org1) {
        wjj.37_03_26_175130 <- .mfp.swxx("37_03_26_175130") %>% 
            hs.wjj
        wj2 <- hs.fp(wjj.37_03_26_175130, paste0(org1, ".tsv.gz"))
        if (file.exists(wj2)) {
            fread(wj2)
        }
        else {
            wj1 <- "~/swxx/sj/db/ncbi/gene/gene2accession.gz"
            if (0) {
                .wk("cat_wk", "~/in2/swxx/sj/db/ncbi/gene/gene2accession.gz", 
                  "| perl -e", .q("while( <>) { if( /\\bP09544\\b/) { print;}}"))
                .wk("cat_wk", "/data/swxx/sj/db/uniprot/idmapping/HUMAN_9606_idmapping_selected.tab.gz", 
                  "| perl -e", .q("while( <>) { if( /\\bA4D0V1\\b/) { print; exit;}}"))
                .wk("cat_wk", "~/in2/swxx/sj/db/uniprot/UniProtKB/Reviewed/uniprot_sprot.dat.gz", 
                  "| less -Ni")
                .wk("cat_wk", "~/in2/swxx/sj/db/uniprot/UniProtKB/Reviewed/uniprot_sprot.dat.gz", 
                  "| perl -e", .q("while( <>) { if( /^(ID|AC)\\s/) { print;}}"), 
                  "| less -Ni")
            }
            cn_wanted <- c("GeneID", "protein_accession.version", 
                "protein_gi", "mature_peptide_accession.version", 
                "mature_peptide_gi", "Symbol")
            ci_wanted <- cn_wanted %>% {
                hs.c(., match(., wj1.cn))
            }
            cmd1 <- .sys("cat_wk", wj1, "| perl -e", .q(paste0("\nwhile( <>) {\n  $flag = 0;\n  if( /^", 
                org1.id, "\\t/) {\n    chomp;\n    @F = split( /\\t/);\n    if( $F[ ", 
                ci_wanted["protein_accession.version"] - 1, "] ne \"-\") {\n      print( ", 
                paste0("$F[ ", ci_wanted - 1, "]") %>% paste(collapse = ", \"\\t\", "), 
                ", \"\\n\");\n    }\n    $flag = 1;\n  } else {\n    if( $flag) {\n      exit;\n    }\n  }\n}\n")))
            jg <- wk.fread(cmd = cmd1, cn = 0)
            setnames(jg, cn_wanted)
            if (jg[hs.grep(protein_accession.version, "^(NP|XP|WP|YP)_", 
                inv = 1), length(hs.grepV(protein_accession.version, 
                "^[^_]_"))]) 
                hs.browser("37.03.26.170031")
            jg <- jg[hs.grep(protein_accession.version, "^(NP|XP|WP|YP)_", 
                inv = 1)]
            jg[, `:=`(protein_accession, hs.sub(protein_accession.version, 
                "\\.\\d+$"))]
            ncbi.uniprot <- unique(jg[, .(GeneID, protein_accession)])
            wj1_uniprot <- paste0("~/swxx/sj/db/uniprot/idmapping/", 
                switch(org1, human = "HUMAN_9606", mouse = "MOUSE_10090"), 
                "_idmapping.dat.gz")
            uniprot <- wk.fread(wj1_uniprot, cn = 0)
            if (0) {
                ncbi.uniprot[, protein_accession] %and% uniprot[, 
                  V1]
                uniprot[hs.filter(V1, function(x1) x1 %in% ncbi.uniprot[, 
                  protein_accession], V2, function(x1) x1 %in% 
                  "GeneID")]
            }
            jg1 <- uniprot[V2 %in% "GeneID", .(V3, V1)] %>% unique
            setnames(jg1, paste0("gene_id,", c("ncbi", "uniprot")))
            jg2 <- local({
                a1 <- uniprot[, if ("GeneID" %not.in% V2) 
                  1, by = V1][[1]]
                a1 %<>% {
                  . %or% hs.sub(., "-\\d+$")
                } %>% (`%not%`(jg1[, `gene_id,uniprot`]))
                ncbi.uniprot[protein_accession %in% a1] %>% setnames(names(jg1))
            })
            jg_wanted <- rbind(jg1, jg2)
            wk.dt.fwrite(jg_wanted, wj2)
            jg_wanted
        }
    }
    function(ft, tt = "ncbi", org1 = "human", tt_join = 0, update_db = 0, 
        .mem = 1) {
        if (any(c(ft, tt) %not.in% c("ncbi", "ens", "mgi_symbol", 
            "uniprot", "symbol"))) 
            stop("[37.03.29.104751]需整合")
        if (0) 
            if (c("ens", "symbol") %all.in% c(ft, tt)) {
                dt1 <- ens.symbol(org1)
                if (0) {
                  wk.dt.subset.na(dt1, "ENSG00000102125")
                }
                if (tt == "ens") 
                  wk.dt.col.reorder(dt1, 2, before = 1)
                if (tt_join) 
                  dt1 <- hs.tt_join(dt1)
                return(dt1)
            }
        if ("mgi_symbol" %in% c(ft, tt)) 
            org1 <- "mouse"
        org1.id <<- hs.hsgly("ncbi/物种/名称.编号.36_12_28_140636")(org1)
        if ("ncbi" %in% c(ft, tt)) {
            dt1 <- get(paste0("ncbi.", c(ft, tt) %not% "ncbi"))(org1 = org1)
            if (tt == "ncbi") 
                wk.dt.col.reorder(dt1, 2, before = 1)
            jg <- dt1
        }
        else {
            jg <- local({
                m1 <- get(paste0("ncbi.", ft))(org1 = org1, update_db = update_db)
                m2 <- get(paste0("ncbi.", tt))(org1 = org1, update_db = update_db)
                if (0) 
                  for (on1 in c("m1", "m2")) {
                    evalbq({
                      if (.(as.symbol(on1))[, !is.character(`gene_id,ncbi`)]) {
                        .(as.symbol(on1))[, `:=`("gene_id,ncbi", 
                          as.character(`gene_id,ncbi`))]
                      }
                    })
                  }
                jg <- merge(m1, m2, by.x = names(m1)[1], by.y = names(m2)[1])
                jg[, `:=`("gene_id_ncbi", {
                })]
            })
            hs.switch(c(ft, tt), list("ens", "symbol"), {
                ft.cn <- "gene_id_ensembl"
                tt.cn <- "gene_id_symbol"
                gtf.fp <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1 = org1)
                gtf <- hs.hsgly("gtf/基因信息.38_09_08_204013")(gtf.fp)
                a <- gtf[gene_id %not.in% jg[hs.clean(get(tt.cn), 
                  ot = "i"), get(ft.cn)]][hs.clean(gene_name, 
                  ot = "i"), .(gene_id, gene_name)]
                jg <- wk.dt.rbind(jg, a)
            })
            jg <- unique(jg)
        }
        if (tt_join) {
            jg <- hs.tt_join(jg)
            if (0) {
                jg[, gene_id_symbol] %not% dt1[, gene_id_symbol]
                dt1[, gene_id_symbol] %not% jg[, gene_id_symbol]
                jg[gene_id_symbol %in% c("TTTY9A")]
                dt1[gene_id_symbol %in% c("SNORD29")]
                jg[, "FOSL2-AS1" %in% gene_id_symbol]
                jg[gene_id_ensembl %in% "ENSG00000223004"]
            }
        }
        jg
    }
})
"37_03_22_145251" <- function(dt1, id1 = {
}, id2 = {
}, id1.cn = colnames(dt1)[1], id2.cn = colnames(dt2)[2], dt2 = hs.hsgly("编号种类转换库.37_03_29_103455")(id1, 
    id2, org1 = org1, tt_join = 1), update = 1, org1 = "human", 
    dup.data = {
    }, dup.uniq.fun1 = identity, dup.uniq.fun2 = which.max, replace = 0, 
    ...) {
    if (0) {
        dt2[hs.grepl(gene_id_symbol, ";")]
        dt2[hs.grepl(gene_id_ensembl, "ENSG00000237402")]
    }
    id1.cn1_0 <- id1.cn[1]
    if (update) {
        local({
            hs1 <- hs.hsgly(switch(id1, ens = "编号更新之ensembl.36_12_20_235632", 
                ncbi = "编号更新之ncbi.36_10_15_163409", 
                symbol = "gene.id_symbol.updater.36_09_12_235937"))
            id1.2 <- paste0(id1.cn, "_updated")
            id <- dt1[[id1.cn[1]]]
            if (id1 %in% "ens") 
                id <- hs.sub(id, "\\.[^\\.]+$")
            a1.37_10_25_111347 <- hs1(id)
            dt1[, `:=`(c(id1.2), a1.37_10_25_111347)]
            if (0) {
                dt1 %=>% hs.ij(x, x[[id1]] != x[[id1.2]])
            }
            id1.cn[1] <<- id1.2
        })
    }
    else {
        if ((id1 %in% "ens") && (dt1[, any(hs.grepl(unique(get(id1.cn[1])), 
            "\\.\\d+$"))])) {
            id1.cn[1] <- "id.37_10_25_135926"
            dt1[, `:=`(c(id1.cn[1]), hs.sub(get(id1.cn1_0), "\\.\\d+$"))]
        }
    }
    id1.cn %<=>% c(x, names(dt2)[1])
    if (missing(id2.cn)) 
        local({
            id2.cn <<- names(dt2) %not% id1.cn
        })
    if (id2.cn %in% names(dt1)) 
        return(dt1)
    if (0) {
        if (identical(id1, "ens")) 
            hs.browser("37.09.17.234411")
        dt2[hs.grepl(`gene_id,ncbi`, ";")]
        dt2[, anyDuplicated(gene_id_ncbi)]
        dt2[hs.grepl(gene_id_symbol, ";")]
    }
    1
    {
        local({
            dt1.id <- dt1[, unique(.SD), .SDcols = rep(id1.cn[1], 
                2)]
            wk.dt.reset_name(dt1.id)
            dt1.id %<>% wk.dt.expand(2)
            id2.v <- wk.dt.subset.na(dt2, dt1.id[[2]], id1.cn[2])[[2]]
            dt1.id[, `:=`(V3, id2.v)]
            if (0) {
                dt1.id[hs.grepl(V1, ";")]
            }
            a1 <- dt1.id[, {
                jg <- hs.c(strsplit(as.character(V1), ";")[[1]], 
                  rep("", .N))
                i1 <- hs.clean(V3, ot = "l")
                if (any(i1)) 
                  jg[i1] <- V3[i1]
                paste(jg, collapse = ";")
            }, by = V1]
            a1 <- wk.dt.subset.na(a1, dt1[[id1.cn[1]]])[[2]]
            dt1[, `:=`(c(id2.cn), a1)]
        })
        if (0) {
            dt1[[id1.cn[1]]] %>% as.character %>% strsplit(";") %>% 
                {
                  k1 <- unlist(.)
                  if (class(k1) != class(dt2[[id1.cn[2]]])) 
                    class(k1) <- class(dt2[[id1.cn[2]]])
                  id2 <- wk.dt.subset.na(dt2, k1, id1.cn[2])[[id2.cn]]
                  i1 <- rep.int(seq_along(.), sapply(., length))
                  id2 <- wk.dt.subset.na(dt2, k1, id1.cn[2])[[id2.cn]]
                  id2 %<>% tapply(i1, . %>% unique %>% sort %>% 
                    paste(collapse = ";"))
                  dt1[unique(i1), `:=`(c(id2.cn), id2)]
                }
        }
    }
    if (0) {
        dt1[hs.clean(`gene_id,ncbi`, inv = 1, ot = "i")]
        dt1[symbol %in% "IL6ST-DT"]
        dt1[symbol %in% "FLJ31104"]
        dt1[symbol %in% "TBCE"]
        dt1[, if (!length(hs.clean(`gene_id,ncbi`))) 
            .SD, by = symbol][symbol %not.in% c("U6", "Y_RNA", 
            "Metazoa_SRP", "U1", "U2", "U3", "U8", "U4", "SNORD116", 
            "SNORD81")][!hs.grepl(symbol, "(?i)^sno")]
        wk.dt.subset.na(dt1, "ENSG00000238297")
        if (any(hs.grepl(dt1[[id1.cn[1]]], ";"))) 
            dt1 <- local({
                a1 <- strsplit(dt1[[id1.cn[1]]], ";")
                i1 <- rep(dt1[, seq.int(.N)], sapply(a1, length))
                jg <- dt1[i1]
                jg[, `:=`(c(id1.cn[1]), unlist(a1))]
            })
        jg <- merge(dt1, dt2, all.x = TRUE, by.x = id1.cn[1], 
            by.y = id1.cn[2], ...)
        if (anyDuplicated(jg[[id2.cn]])) {
            if (length(dup.data)) {
                i1 <- local({
                  if (id1.cn %in% colnames(dup.data)) {
                    v1 <- id1.cn
                    v2 <- colnames(dup.data) %not% v1
                  }
                  else {
                    v1 <- colnames(dup.data)[1]
                    v2 <- colnames(dup.data)[2]
                  }
                  if (!is.numeric(dup.data[[v2]])) {
                    hs.browser("37.09.15.065302")
                  }
                  dup.data <- merge(dup.data, dt2, all.x = TRUE, 
                    by.x = "V1", by.y = id1.cn)
                  dup.data[, `:=`(i1, seq.int(.N))]
                  dup.data[, i1[dup.uniq.fun2(get(v2))], by = c(id2.cn)][, 
                    V1]
                })
                jg <- jg[i1]
            }
            else {
                hs.browser("37.09.15.065231")
            }
        }
    }
    if ("id.37_10_25_135926" %in% colnames(dt1)) {
        dt1[, `:=`(id.37_10_25_135926, {
        })]
        id1.cn[1] <- id1.cn1_0
    }
    if (replace) {
        dt1[, `:=`(c(id1.cn[1]), get(id2.cn))]
        dt1[, `:=`(c(id2.cn), {
        })]
    }
    else if (missing(id2.cn) && length(id1) && length(id2)) {
        setnames(dt1, id2.cn, paste0("gene_id_", id2))
    }
    wk.dt.col.reorder(dt1, c(id1.cn1_0, id1.cn, id2.cn) %and% 
        colnames(dt1))
    dt1
}
"36_12_20_235632" <- local({
    org.sj <- list()
    org.sj.hs.org1 <- function(org1) {
        if (org1 %not.in% names(org.sj)) {
            org.sj[[org1]] <<- wk.fread(hs.hsgly("gene.id_ensembl.history.36_12_21_002840")(org1))
            hs.hsgly("gene.id_ensembl.history.36_12_21_002840")(org1) %=>% 
                .wc
        }
        org.sj[[org1]]
    }
    function(v1, org1 = "human") {
        db <- org.sj.hs.org1(org1)
        v1 <- trim(toupper(v1))
        i1 <- v1 %in% db[, old]
        if (any(i1)) 
            v1[i1] %<=>% db[old %in% x, paste(new, collapse = ";"), 
                by = old][, hs.c(old, V1)][x]
        v1
    }
})
"36_10_15_163409" <- local({
    db <- list()
    function(v1, org1 = "human", as.char = 0) {
        wjj_ncbi <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("ncbi")
        wj1 <- hs.hsgly("ftp.ncbi.nih.gov/gene/DATA.37_04_29_161915")("gene_history.gz")
        wj2 <- hs.wjm(wj1, ext = "replaced.tsv.gz")
        wj3 <- hs.wjm(wj1, ext = "withdrawn.tsv.gz")
        if (wj.update.reporter(c(wj2, wj3), wj1)) {
            a1 <- wk.fread(wj1)
            if (wj.update.reporter(wj2, wj1)) {
                message("[2022_02_18_091423]生成", wj2)
                if (0) {
                  db <- a1[, {
                    id <- Discontinued_GeneID[GeneID == "-"]
                    repeat {
                      id1 <- Discontinued_GeneID[GeneID %in% 
                        id] %not% id
                      if (length(id1)) {
                        .SD
                        hs.browser("36.12.23.003301")
                        id <- c(id, id1)
                      }
                      else break
                      .SD[Discontinued_GeneID %not.in% id]
                    }
                    i1 <- which((GeneID %not.in% id) & (Discontinued_GeneID %not.in% 
                      id))
                    .SD[i1]
                  }, by = "#tax_id"]
                  db_final <- db[, {
                    i1 <- which(GeneID %in% Discontinued_GeneID)
                    if (length(i1)) {
                      for (i1.1 in i1) {
                        id1 <- GeneID[i1.1]
                        repeat {
                          i2 <- match(id1, Discontinued_GeneID, 
                            0)
                          if (identical(i2, 0L)) {
                            break
                          }
                          else {
                            id1 <- GeneID[i2]
                            GeneID[i1.1] <- id1
                          }
                        }
                      }
                    }
                    .(GeneID, Discontinued_GeneID)
                  }, by = "#tax_id"]
                  wk.dt.fwrite(db_final, "~/swxx/sj/db/ncbi/gene/gene_history.updatable.tsv.gz")
                }
                jg <- a1[, {
                  id.db <- id <- GeneID %not% Discontinued_GeneID %not% 
                    "-"
                  repeat {
                    id1 <- Discontinued_GeneID[GeneID %in% id.db] %not% 
                      id.db
                    if (length(id1)) {
                      id.db %<>% c(id1)
                    }
                    else break
                  }
                  if (length(id.db)) {
                    id2do <- id.db %not% id
                    if (0) {
                      a1 <- if (length(id2do) > 1000) {
                        unlist(wk.mclapply(id2do, id2.hs.id1))
                      }
                      else sapply(id2do, id2.hs.id1)
                      wk.dt(old = id2do, new = a1)
                    }
                    jg <- .SD[match(id2do, Discontinued_GeneID), 
                      .(Discontinued_GeneID, GeneID)]
                    repeat {
                      i <- jg[, which(GeneID %not.in% id)]
                      if (length(i)) {
                        old <- jg[i, GeneID]
                        new <- jg[match(old, Discontinued_GeneID), 
                          GeneID]
                        wk.dt.set(jg, i, "GeneID", new)
                      }
                      else break
                    }
                    wk.dt.setnames(jg, c("old", "new"))
                  }
                }]
                if (0) {
                  jg <- a1[, {
                    message("tax_id: ", `#tax_id`)
                    id2.hs.id1 <- function(id1) {
                      repeat {
                        i1 <- match(id1, Discontinued_GeneID)
                        id2 <- GeneID[i1]
                        if (id2 %in% id) {
                          return(id2)
                        }
                        else id1 <- id2
                      }
                    }
                    id.db <- id <- GeneID %not% Discontinued_GeneID %not% 
                      "-"
                    repeat {
                      id1 <- Discontinued_GeneID[GeneID %in% 
                        id.db] %not% id.db
                      if (length(id1)) {
                        id.db %<>% c(id1)
                      }
                      else break
                    }
                    if (length(id.db)) {
                      id2do <- id.db %not% id
                      if (0) {
                        a1 <- if (length(id2do) > 1000) {
                          unlist(wk.mclapply(id2do, id2.hs.id1))
                        }
                        else sapply(id2do, id2.hs.id1)
                        wk.dt(old = id2do, new = a1)
                      }
                      jg <- .SD[match(id2do, Discontinued_GeneID), 
                        .(Discontinued_GeneID, GeneID)]
                      repeat {
                        i <- jg[, which(GeneID %not.in% id)]
                        if (length(i)) {
                          old <- jg[i, GeneID]
                          new <- jg[match(old, Discontinued_GeneID), 
                            GeneID]
                          wk.dt.set(jg, i, "GeneID", new)
                        }
                        else break
                      }
                      wk.dt.setnames(jg, c("old", "new"))
                    }
                  }, by = "#tax_id"]
                }
                wk.dt.fwrite(jg, wj2)
            }
            if (wj.update.reporter(wj3, wj1)) {
                message("[2022_02_18_091510]生成", wj3)
                id_replaced <- wk.fread(wj2)
                jg <- a1[GeneID %not.in% id_replaced[, old %or% 
                  new], {
                  id2do <- GeneID %or% Discontinued_GeneID %not% 
                    "-"
                  unique(.SD[Discontinued_GeneID %in% id2do, 
                    .(Discontinued_GeneID, Discontinued_Symbol %>% 
                      toupper)])
                }]
                wk.dt.setnames(jg, "V2", "gene_id_symbol")
                wk.dt.fwrite(jg, wj3)
            }
        }
        if (!length(db)) {
            db <<- wk.fread(wj2)
        }
        i1 <- v1 %in% db[, old]
        if (any(i1)) {
            a1 <- v1[i1]
            db1 <- db[old %in% a1]
            n1 <- uniqueN(a1)
            message(n1, " gene", ifelse(n1 > 1, "s", ""), " updated.")
            v1[i1] <- db[match(a1, old), new]
        }
        if (as.char) 
            as.character(v1)
        else v1
    }
})
"36_09_12_235937" <- local({
    hs.recent <- {
    }
    db <- list()
    db.hs_f <- function(org1, recent = 1) {
        wjj.36_11_16_133818 <- .mfp.swxx("36_11_16_133816")
        jg.wj <- hs.wjm(org1, add = "tsv.gz", dn = wjj.36_11_16_133818)
        wj1 <- hs.hsgly("ftp.ncbi.nih.gov/gene/DATA.37_04_29_161915")("gene_history.gz")
        switch(org1, mouse = {
        }, human = {
            wj1 %<=>% {
                x %or% formals(hs.hsgly("gene.id_hgnc.35.12.06.111314"))[["wj1"]]
            }
            wj1 %<=>% (x %or% formals(hs.hsgly("gene.id_hgnc.35.12.06.111314"))[["wj1"]])
        })
        if (wj.update.reporter(jg.wj, wj1)) {
            db_ncbi <- hs.hsgly("species.genes.36_12_28_140507")(org1, 
                what = "all")
            if (0) {
                db_ncbi[GeneID %in% 441072]
            }
            if (0) {
                if (org1 != "human") {
                  hs.browser("38.01.03.234310")
                  db_ens <- hs.hsgly("ens,symbol.37_09_14_164105")(org1)
                  hs.hsgly("dt.id1_to_id2.37_03_22_145251")(db_ens, 
                    "ens", "ncbi", org1 = org1, update = 0)
                  hs.hsgly("dt.id1_to_id2.37_03_22_145251")(db_ens, 
                    "ncbi", "symbol", id1.cn = "gene_id_ncbi", 
                    update = 1, org1 = org1)
                  hs.browser("38.01.01.125125")
                  if (0) 
                    db_ens[gene_id_ensembl != gene_id_ensembl_updated]
                  db_ens <- db_ens[gene_id_symbol != gene_id_symbol][{
                    a1 <- strsplit(gene_id_symbol, ";")
                    sapply(seq.int(.N), function(i1) {
                      gene_id_symbol[i1] %not.in% a1[[i1]]
                    })
                  }]
                }
            }
            sep1 <- "|"
            jg <- switch(org1, human = local({
                jg_hgnc <- uid.obj("用总表更新symbol.38_01_03_204741", 
                  "old.new")
                jg_ncbi <- local({
                  {
                    if (0) {
                      db_ncbi[Symbol_from_nomenclature_authority %in% 
                        (names(jg_hgnc) %not% db_hgnc[, `Approved symbol`]), 
                        .N]
                      db_ncbi[Symbol_from_nomenclature_authority %in% 
                        (names(jg_hgnc) %not% db_hgnc[, `Approved symbol`])]
                      db_ncbi[Symbol_from_nomenclature_authority %in% 
                        (names(jg_hgnc) %not% db_hgnc[, `Approved symbol`])][2]
                      jg_hgnc["SEA"]
                      jg_hgnc["ZBED6CL"]
                      db_ncbi[Symbol_from_nomenclature_authority %in% 
                        (names(jg_hgnc) %not% db_hgnc[, `Approved symbol`]), 
                        jg_hgnc[Symbol_from_nomenclature_authority]]
                    }
                    db_hgnc <- hs.hsgly("gene.id_hgnc.35.12.06.111314")()
                    db_ncbi[Symbol_from_nomenclature_authority %in% 
                      (names(jg_hgnc) %not% db_hgnc[, `Approved symbol`]), 
                      `:=`(Symbol_from_nomenclature_authority, 
                        hs.recent(jg_hgnc[Symbol_from_nomenclature_authority]))]
                  }
                  if (0) {
                    db_hgnc[, `HGNC ID`] %not% db_ncbi[, dbXrefs %>% 
                      hs.re("(?i)(?<=hgnc:)\\d+")]
                    db_ncbi[, dbXrefs %>% hs.re("(?i)(?<=hgnc:)\\d+")] %not% 
                      db_hgnc[, `HGNC ID`]
                    db_ncbi[, dbXrefs %>% hs.re("(?i)hgnc:\\d+")] %not% 
                      db_hgnc[, `HGNC ID`]
                    db_ncbi[, dbXrefs %>% hs.gre("(?i)hgnc:\\d+") %>% 
                      sapply(length) %>% hs.table]
                  }
                  if (0) {
                    local({
                      i1 <- db_ncbi[, hs.grep(dbXrefs, "(?i)(?<=hgnc:)\\d+")]
                      db_ncbi[i1, `:=`(hgnc_id, hs.re(dbXrefs, 
                        "(?i)(?<=hgnc:)\\d+"))]
                      if (db_ncbi[hs.clean(hgnc_id, ot = "i"), 
                        anyDuplicated(hgnc_id)]) {
                        hs.browser("37.10.05.234120")
                      }
                      id1 <- db_hgnc[, `HGNC ID`] %and% db_ncbi[, 
                        hgnc_id]
                      if (0) {
                        if (db_ncbi[hgnc_id %in% id1, anyDuplicated(hgnc_id)] | 
                          db_hgnc[`HGNC ID` %in% id1, anyDuplicated(`HGNC ID`)]) {
                          hs.browser("37.03.31.124848")
                        }
                        db_hgnc[, uniqueN(`Approved symbol`), 
                          by = `HGNC ID`][, V1 %>% uniqueN]
                      }
                      i2 <- db_ncbi[, match(id1, hgnc_id)]
                      i3 <- db_hgnc[, match(id1, `HGNC ID`)]
                      if (db_ncbi[i2, any(Symbol_from_nomenclature_authority != 
                        db_hgnc[i3, `Approved symbol`])]) {
                        db_ncbi[i2, `:=`(Symbol_from_nomenclature_authority, 
                          db_hgnc[i3, `Approved symbol`])]
                      }
                    })
                    db_ncbi[Symbol_from_nomenclature_authority %in% 
                      (names(jg_hgnc) %not% db_hgnc[, `Approved symbol`]), 
                      `:=`(Symbol_from_nomenclature_authority, 
                        hs.recent(jg_hgnc[Symbol_from_nomenclature_authority]))]
                  }
                  ncbi.jg1 <- db_ncbi[Symbol_from_nomenclature_authority != 
                    "-", {
                    i1 <- Symbol_from_nomenclature_authority == 
                      Symbol
                    s1 <- paste0(Symbol_from_nomenclature_authority, 
                      "|", Symbol)
                    s2 <- paste0(s1, "|", Synonyms)
                    s3 <- paste0(Symbol, "|", Synonyms)
                    a1 <- ifelse(Synonyms == "-", ifelse(i1, 
                      Symbol, s1), ifelse(i1, s3, s2)) %=>% strsplit(x, 
                      "\\|") %=>% lapply(x, "%not%", "-")
                    a2 <- ifelse(i1, Symbol, s1) %=>% strsplit(x, 
                      "\\|")
                    a3 <- data.table(new = rep(ifelse(i1, Symbol, 
                      s1), sapply(a1, length)) %=>% strsplit(x, 
                      "\\|") %=>% unlist, old = unlist(rep(a1, 
                      sapply(a2, length))))
                    a3[, paste(new, collapse = sep1), by = old][, 
                      hs.c(old, V1)]
                  }]
                  if (0) {
                    ncbi.jg1["-"]
                  }
                  ncbi.jg2 <- db_ncbi[Symbol_from_nomenclature_authority == 
                    "-", {
                    a1 <- ifelse(Synonyms == "-", Symbol, paste0(Symbol, 
                      sep1, Synonyms)) %=>% strsplit(x, "\\|")
                    a2 <- wk.dt(new = rep(Symbol, sapply(a1, 
                      length)), old = unlist(a1))
                    jg <- a2[, paste(new, collapse = sep1), by = old][, 
                      hs.c(old, V1)]
                    a <- names(jg) %and% names(ncbi.jg1)
                    jg[a] %<=>% {
                      paste0(x, sep1, ncbi.jg1[a]) %=>% strsplit(x, 
                        sep1, fixed = TRUE) %=>% sapply(x, . %=>% 
                        {
                          paste(unique(x), collapse = sep1)
                        })
                    }
                    jg
                  }]
                  i1 <- names(ncbi.jg1) %not.in% names(ncbi.jg2)
                  jg <- c(ncbi.jg1[i1], ncbi.jg2)
                  if (0) {
                    jg <- db_ncbi[Synonyms != "-", {
                      a1 <- hs.c(ifelse(Symbol_from_nomenclature_authority == 
                        "-", Symbol, Symbol_from_nomenclature_authority), 
                        strsplit(Synonyms, "\\|"))
                      a2 <- inverseList(a1)
                      sapply(a2, paste, collapse = ";")
                    }]
                    jg %>% {
                      .[names(.) != .]
                    } %>% {
                      a1 <- strsplit(., ";")
                      .[sapply(seq_along(.), function(i1) {
                        names(.)[i1] %in% a1[[i1]]
                      })]
                    }
                  }
                  jg %=>% x[names(x) != x]
                })
                a1 <- names(jg_hgnc) %and% names(jg_ncbi)
                jg1 <- paste0(jg_hgnc[a1], sep1, jg_ncbi[a1]) %=>% 
                  strsplit(sep1, fixed = TRUE) %=>% sapply(x, 
                  . %=>% {
                    unique(x) %=>% paste(collapse = sep1)
                  }) %=>% hs.c(a1, x)
                return(list(jg_hgnc, jg_ncbi) %=>% lapply(x, 
                  . %=>% {
                    x[names(x) %not.in% a1]
                  }) %>% unlist %>% c(jg1))
                if (0) {
                  a <- list(jg_hgnc, jg_ncbi) %>% lapply(. %>% 
                    {
                      .[names(.) %not.in% a1]
                    }) %>% unlist %>% c(jg1)
                  b <- fread(jg.wj)[, hs.c(old, new)] %>% {
                    .[. != names(.)]
                  }
                  names(a) %not% names(b)
                  names(b) %not% names(a)
                  g1 <- "ZSCAN5DP"
                  g1 <- "ADAM23"
                  g1 <- "MASCRNA"
                  a[g1]
                  b[g1]
                  ncbi.jg2[g1]
                  jg_hgnc[g1]
                  jg_ncbi[g1]
                  hgnc.jg1[g1]
                  hgnc.jg2[g1]
                  db_ncbi[(Symbol %in% g1) | (Symbol_from_nomenclature_authority %in% 
                    g1) | hs.grepl(Synonyms, paste0("\\b", g1, 
                    "\\b"))]
                  db_hgnc[(`Approved symbol` %in% g1) | hs.grepl(`Previous symbols`, 
                    paste0("\\b", g1, "\\b")) | hs.grepl(`Alias symbols`, 
                    paste0("\\b", g1, "\\b"))]
                }
            }), mouse = {
                if (0) {
                  local({
                    a <- "~/in2/swxx/sj/db/ncbi/gene/gene2accession.gz" %>% 
                      readLines(9) %>% {
                      wk.fread(text = .)
                    }
                    a[1]
                    b <- .wk("cat_wk", "~/in2/swxx/sj/db/ncbi/gene/gene2accession.gz | perl -e", 
                      .q("\nwhile( <>) {\n  if( /BC048546/) {\n    print;\n    exit;\n  }\n}\n"), 
                      intern = 1)
                    hs.c(names(a), strsplit(b, "\t")[[1]])
                  })
                }
                {
                  if (0) {
                    db_ens <- fread(wj1["ensembl"])
                    db_ens <- local({
                      i1 <- hs.clean(db_ens[[2]], ot = "l")
                      hs.c(toupper(db_ens[[1]][i1]), as.list(toupper(db_ens[[2]][i1])))
                    })
                  }
                  db_ens <- list()
                }
                synonyms.cn <- "Synonyms"
                if (recent) {
                  current <- db_ncbi[, toupper(Symbol) %>% unique]
                  synonyms.cn <- "Synonyms_no_current"
                  db_ncbi[Synonyms != "-", `:=`(c(synonyms.cn), 
                    strsplit(toupper(Synonyms), "\\|") %>% sapply(. %>% 
                      {
                        paste(. %not% current, collapse = "|")
                      }))]
                }
                db_ncbi <- wk.dt.subset(db_ncbi, exp_l = c(hs.clean(get(synonyms.cn), 
                  dirty.extra = "-", ot = "l")), exp_j = hs.c(toupper(Symbol), 
                  strsplit(get(synonyms.cn), "\\|")), ot = "r")
                c(db_ens, db_ncbi) %>% inverseList %>% sapply(. %>% 
                  unique %>% paste(collapse = "|"))
            }, {
                db_ncbi <- local({
                  db_ncbi[Synonyms != "-"] %>% {
                    hs.c(.[, ifelse(Symbol_from_nomenclature_authority != 
                      "-", Symbol_from_nomenclature_authority, 
                      Symbol)], strsplit(.[, Synonyms], "\\|"))
                  }
                })
                c(db_ncbi) %>% inverseList %>% sapply(. %>% unique %>% 
                  paste(collapse = sep1))
            })
            if (0) {
                jg["LAP3"]
                jg["IFNA"]
            }
            if (0) {
                jg["A1BG"]
                b <- jg %>% {
                  data.table(old = names(.), new = .)
                } %>% setorderv("old")
                a <- jg.wj %>% wk.fread
                a[, old] %not% b[, old]
                b[old %not.in% a[, old]]
            }
            jg %>% {
                data.table(old = names(.), new = .)
            } %>% setorderv("old") %>% wk.dt.fwrite(jg.wj)
            db[org1] <<- list(jg)
        }
        else {
            if (org1 %not.in% names(db)) {
                jg <- wk.fread(jg.wj)
                db[org1] <<- list(jg[, hs.c(old, new)])
            }
        }
    }
    if (0) {
        db_ncbi[, hs.grepV(dbXrefs, "AC008736")]
        "~/swxx/sj/db/ncbi/gene/gene_info.gz" %>% wk.bash
    }
    function(v1, keep.old = 1, flat = 0, org1 = "human", recent = 1) {
        if (!length(hs.recent)) 
            hs.recent <<- hs.hsgly("新旧并存取新.38_01_03_221659")
        db.hs_f(org1, recent = recent)
        db <- db[[org1]]
        jg <- hs.c(v1, rep("", length(v1)))
        v2 <- trim(v1) %>% toupper
        i1 <- v2 %not.in% names(db)
        if (any(i1)) 
            jg[i1] <- v1[i1]
        i1 <- !i1
        if (any(i1)) 
            jg[i1] <- db[v2[i1]]
        if (recent) {
            if (0) {
                a <- hs.gre(jg, ";")
                sapply(a, length) %>% hs.table
                db["LAP3"]
                jg["LAP3"]
            }
            jg <- hs.recent(jg)
        }
        if (flat && any(hs.grepl(jg, "\\|"))) {
            a1 <- strsplit(jg, "\\|")
            wk.dt(input = rep(names(jg), sapply(a1, length)), 
                updated = unlist(a1))
        }
        else jg
    }
})
"37_05_27_100106" <- local({
    get.pw_gene <- function(pw1, kegg.db, combine_method = "intersect") {
        if (combine_method != "intersect") 
            stop("37_05_06_155358")
        hs.lapply(pw1, function(pw1) kegg.db[set %in% pw1, gene]) %>% 
            inverseList %>% {
            names(.)[sapply(., length) == length(pw1)]
        }
    }
    function(pw, db1 = "kegg", org1 = "human", tt = "symbol") {
        kegg.db <- hs.hsgly("kegg/通路/基因.36.03.18.133126")(org1 = org1)
        get.pw_gene(pw, kegg.db) %>% {
            hs.hsgly("geneid1-geneid2.37_04_28_165424")(., "ncbi", 
                tt, org1 = org1)
        }
    }
})
"37_05_26_124410" <- function(gene1 = {
}, org1 = "human", org2 = "mouse", ft = "ncbi", tt = ft, one_to_one = 1) {
    if (ft %not.in% c("ncbi", "symbol")) 
        hs.browser("37.05.26.125613")
    homologue_ncbi <- hs.hsgly("ncbi_org1_org2_homologous_gene.37_04_30_151333")(org1 = org1, 
        org2 = org2, one_to_one = one_to_one)
    if (0) {
        homologue_ncbi[human_gene_id %in% 721]
    }
    jg <- if (ft == "symbol") {
        org1.db <- hs.hsgly("编号种类转换库.37_03_29_103455")(ft, 
            "ncbi", org1 = org1)
        org2.db <- hs.hsgly("编号种类转换库.37_03_29_103455")(ft, 
            "ncbi", org1 = org2)
        if (org1.db[, is.numeric(gene_id_ncbi)]) 
            org1.db[, `:=`(gene_id_ncbi, paste(gene_id_ncbi))]
        if (org2.db[, is.numeric(gene_id_ncbi)]) 
            org2.db[, `:=`(gene_id_ncbi, paste(gene_id_ncbi))]
        wk.dt.replace(homologue_ncbi, paste, . %>% {
            !is.character(.)
        })
        x.cn1 <- "gene_id_ncbi"
        y.cn1 <- names(homologue_ncbi)[1]
        a1 <- merge(org1.db, homologue_ncbi, by.x = x.cn1, by.y = y.cn1)
        a2 <- merge(org2.db, homologue_ncbi, by.x = x.cn1, by.y = names(homologue_ncbi)[2])
        merge(a1[, .SD, .SDcol = c(1, 2)], a2[, .SD, .SDcol = -1], 
            by.x = "gene_id_ncbi", by.y = names(homologue_ncbi)[1])[, 
            .SD, .SDcol = -1]
    }
    else homologue_ncbi
    if (0) {
        jg[gene_id_symbol.x != gene_id_symbol.y]
        jg[gene_id_symbol.x %in% "CD33"]
    }
    wk.dt.reset_name(jg)
    if (0) {
        jg[tolower(get(names(jg)[1])) != tolower(get(names(jg)[2]))]
        jg[, V1 %>% anyDuplicated]
        jg[V1 %in% "MBL2"]
        sum(gene1 %in% "MBL2")
        jg[hs.clean(V1, ot = "i", inv = 1)]
        jg[, mean(V1 != V2)]
    }
    if (length(gene1)) {
        i <- tolower(jg[[1]]) %in% tolower(gene1)
        a <- jg[i]
        return(a)
    }
    jg
}
"37_05_27_153942" <- function(sj1, org1 = "mouse", org2 = "human", 
    gt = "symbol", gene.cn = {
    }, cn2 = {
    }) {
    if (is(sj1, "data.table")) {
        cn1 <- if (length(gene.cn)) 
            gene.cn
        else gene_id_name.37_04_29_145914(names(sj1))
        if (!length(cn1)) 
            cn1 <- names(sj1)[1]
        db1 <- hs.hsgly("同源基因.37_05_26_124410")(sj1[[cn1]], 
            org1 = org1, org2 = org2, ft = gt, one_to_one = 0)
        if (0) {
            db1[V1 != V2]
            hs.save(sj1, wj = "~/a.rd")
            load("~/a.rd")
            org1 <- "human"
            org2 <- "rat"
            gt <- "symbol"
            gene.cn <- "gene_updated"
            cn2 <- "gene_rat"
        }
        db1 <- db1[, hs.clean(V2, U = 1) %>% paste(collapse = ";"), 
            by = V1]
        wk.dt.reset_name(db1)
        i <- hs.str_match(sj1[[cn1]], db1[, V1])
        c.38_09_08_143112 <- db1[["V2"]][i]
        if (!length(cn2)) 
            cn2 <- cn1
        sj1[, `:=`(c(cn2), c.38_09_08_143112)]
    }
    else hs.browser("37.05.27.154738")
    sj1
}
"36_12_28_140507" <- function(org1 = "human", what = "gene_id", 
    symbol2upper_case = 0, ..mem = 1) {
    wj2 <- hs.switch(org1, "human", {
        hs.hsgly("ftp.ncbi.nih.gov/gene/DATA.37_04_29_161915")("GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz")
    }, "mouse", {
        hs.hsgly("ftp.ncbi.nih.gov/gene/DATA.37_04_29_161915")("GENE_INFO/Mammalia/Mus_musculus.gene_info.gz")
    }, "rat", {
        hs.hsgly("ftp.ncbi.nih.gov/gene/DATA.37_04_29_161915")("GENE_INFO/Mammalia/Rattus_norvegicus.gene_info.gz")
    }, {
        wj1 <- hs.hsgly("ftp.ncbi.nih.gov/gene/DATA.37_04_29_161915")("gene_info.gz")
        if (0) {
            .ls(wj1)
            .less(wj1)
            .wk("zcat", wj1, "| cut -f3 | uniq | sort | uniq")
        }
        wj2 <- hs.wjm(org1, add = "gene_id.tsv.gz", dn = dirname(wj1))
        if (wj.update.reporter(wj2, wj1)) {
            org1.id <- hs.hsgly("ncbi/物种/名称.编号.36_12_28_140636")(org1)
            readLines(wj1, 1) %>% cat(sep = "\n", file = .sys("| gzip >", 
                .sys.fpGood(wj2)))
            .wk(.sys.cat.text(wj1), "| perl -e", .q(paste0("\nwhile( <>) {\n  $flag = 0;\n  if( /^", 
                org1.id, "\\t/) {\n    print;\n    $flag = 1;\n  } else {\n    if( $flag) {\n      exit;\n    }\n  }\n}\n")), 
                "| uniq | sort -n | uniq | gzip >>", .sys.fpGood(wj2))
        }
        wj2
    })
    switch(what, wj = wj2, gene_id = {
        cmd1 <- .sys(.sys.cat.text(wj2), "| perl -e", .q("while( <>) { /\t([^\t]*)/; print( $1, \"\n\")}"))
        wk.fread(cmd = cmd1)
    }, all = {
        jg <- wk.fread(wj2)
        if (symbol2upper_case) {
            hs.lapply("Symbol,Symbol_from_nomenclature_authority,Synonyms", 
                function(cn1) {
                  jg[, `:=`(c(cn1), toupper(get(cn1)))]
                  {
                  }
                })
        }
        {
            i1 <- wk.dt.subset(jg, exp_l = c(tolower(Symbol_from_nomenclature_authority) != 
                tolower(Symbol), Symbol_from_nomenclature_authority != 
                "-"), ot = "i")
            if (length(i1)) {
                jg[i1, `:=`(Synonyms, hs.ifelse(Synonyms == "-", 
                  function(i1) Symbol[i1], function(i1) paste0(Symbol[i1], 
                    "|", Synonyms[i1])))]
                jg[i1, `:=`(Symbol, Symbol_from_nomenclature_authority)]
            }
        }
        jg
    })
}
"36_05_11_151845" <- local({
    color.V <- c("royalblue", "red2")
    function(tbl = {
    }, x1.cn = hs.grepV(colnames(tbl), "(?i)l(og(2)?)?f(old)?c(hange)?")[1], 
        y1.cn = hs.grepV(colnames(tbl), "(?i)(p[\\._-]?(val[\\._-]?)?adj|adj[\\._-]p)")[1], 
        rn.cn = hs.grepV(names(tbl), "gene([\\._]?id)?")[1], 
        fc.co = 1.2, y1.co = 0.05, x1 = {
        }, y1 = {
        }, tbl.wj = {
        }, selectLab = {
        }, wj = if (length(tbl.wj)) 
            hs.wjm(tbl.wj, ext = "tmp.pdf", extN = ifelse(hs.grepl(tbl.wj, 
                "\\.tsv\\.gz$"), 2, stop("36.05.24.103854")))
        else "~/tmp/tmp.pdf", hs = 20, title = "Volcano plot", 
        subtitle = "", labSize = 3, pointSize = 2, xlim = range(tbl[[x1.cn]] %=>% 
            x[!is.infinite(x)]) * xlim.factor, xlim.factor = 1, 
        ylim = c(0, max(-log10(tbl[, y1.cn]), na.rm = !0) + 0.1), 
        drawConnectors = TRUE, caption = paste0("Total = ", nrow(tbl), 
            " ", caption.right), caption.right = "gene", right.name = "Up", 
        left.name = "Down", color.alpha = 0.2, w1 = 20, h1 = 20, 
        plot = 1, x.axis = 1, xlab = bquote(~Log[2] ~ .(xlab.right)), 
        xlab.right = "Fold change", border = "partial", ...) {
        hs.require("EnhancedVolcano")
        if (length(tbl.wj) == 1) {
            tbl <- fread(tbl.wj, data.table = 0)
        }
        tbl <- tbl[hs.clean(get(y1.cn), ot = "l")]
        if (length(x1) && length(y1)) {
            tbl <- data.frame(x1 = x1, y1 = y1)
            rownames(tbl) <- names(x1)
            if (length(x1.cn) && length(y1.cn)) {
                colnames(tbl) <- c(x1.cn, y1.cn)
            }
        }
        wk.dt.setorderv(tbl, y1.cn)
        zero.i1 <- which(tbl[[y1.cn]] == 0)
        if (length(zero.i1)) {
            extreme1 <- min(tbl[-zero.i1, get(y1.cn)]) * 0.1
            tbl[zero.i1, `:=`(c(y1.cn), extreme1)]
        }
        rn <- tbl[[rn.cn]]
        i1 <- tbl[, which((get(y1.cn) < y1.co) & (abs(get(x1.cn)) > 
            log2(fc.co)))]
        up_i1 <- tbl[i1, i1[which(get(x1.cn) > 0)]]
        down_i1 <- tbl[i1, i1[which(get(x1.cn) < 0)]]
        if (!identical(class(tbl), "data.frame")) {
            tbl <- as.data.frame(tbl)
        }
        colCustom <- rep("grey", nrow(tbl))
        names(colCustom) <- rep("Insig", length(colCustom))
        colCustom[up_i1] <- "red2"
        names(colCustom)[up_i1] <- length(up_i1) %=>% rep(paste0(right.name, 
            " (n=", x, ")"), x)
        colCustom[down_i1] <- "royalblue"
        names(colCustom)[down_i1] <- length(down_i1) %=>% rep(paste0(left.name, 
            " (n=", x, ")"), x)
        if (0) {
            tbl %>% {
                i1 <- .[, x1.cn] > 0
                .[i1, rn.cn][hs.top((.[i1, x1.cn] %>% {
                  ./max(.)
                }) * ((-log10(.[i1, y1.cn])) %>% {
                  ./max(.)
                }), 20, "right", ot = "i")]
            }
        }
        p1 <- EnhancedVolcano(tbl, lab = rn, labSize = labSize, 
            caption = caption, selectLab = if (length(selectLab)) {
                selectLab %and% rn
            }
            else {
                tbl %=>% {
                  i1 <- x[, x1.cn] > 0
                  a1 <- x[i1, rn.cn][hs.top((x[i1, x1.cn] %=>% 
                    {
                      x/max(x)
                    }) * ((-log10(x[i1, y1.cn])) %=>% {
                    x/max(x)
                  }), hs, "right", ot = "i")]
                  i1 <- x[, x1.cn] < 0
                  a2 <- x[i1, rn.cn][hs.top((abs(x[i1, x1.cn]) %=>% 
                    {
                      x/max(x)
                    }) * ((-log10(x[i1, y1.cn])) %=>% {
                    x/max(x)
                  }), hs, "right", ot = "i")]
                  a1 %or% a2
                }
            }, x = x1.cn, y = y1.cn, ylab = quote(~-Log[10] ~ 
                italic(Q)), xlab = xlab, title = title, subtitle = subtitle, 
            , FCcutoff = log2(fc.co), pCutoff = y1.co, colCustom = colCustom, 
            colAlpha = color.alpha, pointSize = pointSize, xlim = xlim, 
            ylim = ylim, drawConnectors = as.logical(drawConnectors), 
            endsConnectors = "last", border = border, borderWidth = 1, 
            ...)
        p1 <- p1 + theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(), panel.background = element_blank(), 
            axis.line = element_line(colour = "black"))
        if (!x.axis) 
            p1 <- p1 + theme(axis.title.x = element_blank(), 
                axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
                axis.line.x = element_blank())
        if (plot) {
            .pdf({
                print(p1)
            }, wj = wj, width = w1, height = h1)
        }
        else p1
    }
})
"38_04_26_214925" <- function(de, p, q, x, y, top = 20, xlab = x, 
    wj2 = "~/swxx/tmp.pdf", ...) {
    if (file.exists(wj2)) 
        return(wj2)
    if (0) {
        p <- "P.Value"
        q <- "adj.P.Val"
        x <- "t"
        y <- "gene_set"
        top <- 20
    }
    de <- de[head(order(get(p)), top)]
    wk.dt.setorderv(de, x)
    de[, `:=`(c(y), hs.factor(get(y), N = 0))]
    de[, `:=`(direction.38_04_26_215714, "Insig.")]
    de[get(q) < 0.05, `:=`(direction.38_04_26_215714, ifelse(sign(get(x)) > 
        0, "High", "Low"))]
    p1 <- evalbq(ggplot2::ggplot(de, ggplot2::aes(.(as.symbol(x)), 
        .(as.symbol(y)), fill = direction.38_04_26_215714))) + 
        ggplot2::geom_col() + ggplot2::scale_fill_manual(values = c(High = "#36638a", 
        Insig. = "#cccccc", Low = "#7bcd7b") %=>% x[names(x) %in% 
        de[, direction.38_04_26_215714]]) + ggplot2::xlab(xlab) + 
        ggplot2::ylab("") + ggplot2::theme(panel.background = {
    }, legend.position = "none", axis.text.y = ggplot2::element_blank(), 
        axis.ticks.y = ggplot2::element_blank())
    co <- 0.3
    range <- range(de[[x]])
    ratio <- range %=>% hs.with(x, abs(x), x2/max(max(x2), diff(x1)))
    hs.x1.38_04_27_112049 <- function(fx1) {
        switch(fx1, `-` = {
            if ((range[2] > 0) && (ratio[2] > co)) range[2] else range[1]
        }, `+` = {
            if ((range[1] < 0) && (ratio[1] > co)) range[1] else range[2]
        })/2
    }
    x1 <- c("-", "+") %=>% sapply(hs.x1.38_04_27_112049)
    width <- if (uniqueN(sign(x1)) > 1) 
        20
    else 10
    height <- de[, .N] * 0.5
    hs.place.38_04_27_112311 <- function(x1) ifelse(x1 > 0, "left", 
        "right")
    hs.color.38_04_27_155917 <- function(fx1, x1) switch(paste(uniqueN(sign(c(fx1, 
        x1)))), `1` = "black", `2` = "gray")
    de1 <- de[direction.38_04_26_215714 %in% "High"]
    if (de1[, .N]) {
        x1.38_04_27_112049 <- unname(x1["+"])
        p1 <- p1 + evalbq(de1[, {
            ggfittext::geom_fit_text(ggplot2::aes(.(x1.38_04_27_112049), 
                .(as.symbol(y))), .SD, label = .(as.symbol(y)), 
                place = .(hs.place.38_04_27_112311(x1.38_04_27_112049)), 
                width = .(abs(x1.38_04_27_112049) * 1.9), reflow = TRUE, 
                grow = TRUE)
        }])
    }
    de1 <- de[direction.38_04_26_215714 %in% "Low"]
    if (de1[, .N]) {
        x1.38_04_27_112049 <- unname(x1["-"])
        p1 <- p1 + evalbq(de1[, {
            ggfittext::geom_fit_text(ggplot2::aes(.(x1.38_04_27_112049), 
                .(as.symbol(y))), .SD, label = .(as.symbol(y)), 
                place = .(hs.place.38_04_27_112311(x1.38_04_27_112049)), 
                width = .(abs(x1.38_04_27_112049) * 1.9), reflow = TRUE, 
                grow = TRUE)
        }])
    }
    de1 <- wk.dt.subset(de, exp_l = c(direction.38_04_26_215714 %in% 
        "Insig.", get(x) > 0))
    if (de1[, .N]) {
        x1.38_04_27_112049 <- unname(x1["+"])
        p1 <- p1 + evalbq(de1[, {
            ggfittext::geom_fit_text(ggplot2::aes(.(x1.38_04_27_112049), 
                .(as.symbol(y))), .SD, label = .(as.symbol(y)), 
                place = .(hs.place.38_04_27_112311(x1.38_04_27_112049)), 
                width = .(abs(x1.38_04_27_112049) * 1.9), color = .(hs.color.38_04_27_155917(1, 
                  x1.38_04_27_112049)), reflow = TRUE, grow = TRUE)
        }])
    }
    de1 <- wk.dt.subset(de, exp_l = c(direction.38_04_26_215714 %in% 
        "Insig.", get(x) < 0))
    if (de1[, .N]) {
        x1.38_04_27_112049 <- unname(x1["-"])
        p1 <- p1 + evalbq(de1[, {
            ggfittext::geom_fit_text(ggplot2::aes(.(x1.38_04_27_112049), 
                .(as.symbol(y))), .SD, label = .(as.symbol(y)), 
                place = .(hs.place.38_04_27_112311(x1.38_04_27_112049)), 
                width = .(abs(x1.38_04_27_112049) * 1.9), color = .(hs.color.38_04_27_155917(-1, 
                  x1.38_04_27_112049)), reflow = TRUE, grow = TRUE)
        }])
    }
    if (length(wj2)) {
        .pdf(print(p1), wj2, width = width, height = height)
        .png(print(p1), wj2, width = width, height = height)
    }
    else p1
}
"36_08_11_231030" <- function(en, en.wj = {
}, pdf.wj = "headtmap.pdf", q.co = 0.05, n.co = 1, cn.anno.df = {
}, cn.anno.name = colnames(cn.anno.df)[1]) {
    if (length(en.wj)) 
        en <- hs.lapply(en.wj, fread)
    q.cn <- hs.grepV(names(en[[1]]), "^qv")
    anno.db <- lapply(en, . %>% {
        .[get(q.cn) < q.co, Description]
    }) %>% inverseList
    anno.db %<>% {
        names(.)[sapply(., length) >= n.co]
    }
    if (length(anno.db) < 2) 
        return()
    m1 <- sapply(en, . %>% {
        wk.dt.subset.na(., anno.db, "Description")[, NES]
    })
    qv <- sapply(en, . %>% {
        wk.dt.subset.na(., anno.db, "Description")[, get(q.cn)]
    })
    colnames(m1) <- colnames(qv) <- names(en)
    rownames(m1) <- rownames(qv) <- anno.db
    m1[is.na(m1)] <- 0
    n1 <- max(m1)
    color.breaks <- c(wk.color["blue"], "white", wk.color["red"])
    nes.lgd.at <- (floor(n1 * 2)/2) %>% {
        seq(-., ., 0.5)
    }
    nes.lgd.labels <- local({
        a1 <- as.character(nes.lgd.at)
        if (tail(nes.lgd.at, 1) < n1) {
            a1[1] %<>% {
                paste0("<=", .)
            }
            a1[length(a1)] %<>% {
                paste0(">=", .)
            }
        }
        a1
    })
    q.lgd.at <- c(0, 1, -log10(0.05), seq(2, max(c(2, floor(n1)))))
    q.lgd.labels <- lgd.at %>% {
        a1 <- head(., -1)
        i1 <- as.integer(a1) != a1
        a2 <- as.character(a1)
        if (any(i1)) 
            a2[i1] <- format(a1[i1], digits = 2)
        c(a2, paste0(">=", tail(., 1)))
    }
    colors <- circlize::colorRamp2(nes.lgd.at %>% tail(1) %>% 
        {
            c(-., 0, .)
        }, color.breaks)
    Args <- list(m1, colors, show_heatmap_legend = FALSE, width = grid::unit(0.5 * 
        ncol(m1), "cm"), height = grid::unit(0.5 * nrow(m1), 
        "cm"), row_names_side = "right", use_raster = FALSE, 
        cell_fun = function(j, i, x, y, width, height, fill) {
            qv[i, j] %>% {
                ifelse(is.na(.) || (. >= 0.05), "", "*")
            }
            grid.text(qv[i, j] %>% {
                ifelse(is.na(.) || (. >= 0.05), "", "*")
            }, x, y)
        })
    if (length(cn.anno.df)) {
        cn.anno <- HeatmapAnnotation(df = cn.anno.df, annotation_name_side = "left")
        Args["top_annotation"] <- list(cn.anno)
        Args["bottom_annotation"] <- list(cn.anno)
    }
    P <- do.call(Heatmap, Args)
    lgd1 <- Legend(title = "Normalized enrichment score", col = colors, 
        at = nes.lgd.at, labels = nes.lgd.labels, direction = "horizontal")
    .pdf(draw(P, heatmap_legend_list = list(lgd1), heatmap_legend_side = "top"), 
        height = max(20, nrow(m1)/4), width = text.width.hs(rownames(m1)) * 
            2.54 + 20, wj = hs.wj(pdf.wj))
}
"36_08_07_150833" <- function(en, en.wj = {
}, pdf.wj = "headtmap.pdf", q.co = 0.05, n.co = 1) {
    if (length(en.wj)) 
        en <- hs.lapply(en.wj, fread)
    q.cn <- hs.grepV(names(en[[1]]), "^qv")
    anno.db <- lapply(en, . %>% {
        .[get(q.cn) < q.co, Description]
    }) %>% inverseList
    anno.db %<>% {
        names(.)[sapply(., length) >= n.co]
    }
    if (length(anno.db) < 2) 
        return()
    m1 <- sapply(en, . %>% {
        wk.dt.subset.na(., anno.db, "Description")[, -log10(get(q.cn))]
    })
    colnames(m1) <- names(en)
    rownames(m1) <- anno.db
    m1[is.na(m1)] <- 0
    n1 <- max(m1)
    color.breaks <- c("black", "white", "red")
    lgd.at <- c(0, 1, -log10(0.05), seq(2, max(c(2, floor(n1)))))
    lgd.labels <- lgd.at %>% {
        a1 <- head(., -1)
        i1 <- as.integer(a1) != a1
        a2 <- as.character(a1)
        if (any(i1)) 
            a2[i1] <- format(a1[i1], digits = 2)
        c(a2, paste0(">=", tail(., 1)))
    }
    colors <- circlize::colorRamp2(c(0, -log10(0.05), max(c(2, 
        floor(n1)))), color.breaks)
    P <- Heatmap(m1, colors, show_heatmap_legend = FALSE, width = grid::unit(0.5 * 
        ncol(m1), "cm"), height = grid::unit(0.5 * nrow(m1), 
        "cm"), row_names_side = "right", use_raster = FALSE)
    lgd1 <- Legend(title = as.expression(bquote(~"Log"[10] * 
        "Q")), col = colors, at = lgd.at, labels = lgd.labels, 
        direction = "horizontal")
    .pdf(draw(P, heatmap_legend_list = list(lgd1), heatmap_legend_side = "top"), 
        height = max(20, nrow(m1)/4), width = 20, wj = hs.wj(pdf.wj))
}
"36_03_10_142422" <- function(height, xlab1 = "", main1 = "", 
    col1 = c("#A20056E5", "gray"), wj = hs.wj(wjj, "bar_plot.pdf"), 
    wjj = "~/tmp", post.plot = {
    }, beside = length(dim(height)), errBar = NULL, horiz = TRUE, 
    ref1 = {
    }, names.arg = {
    }, width1 = {
    }, height1 = {
    }, par.par = {
    }, ...) {
    entries <- if (length(names.arg)) {
        names.arg
    }
    else if (length(dim(height))) 
        colnames(height)
    else names(height)
    if (horiz) {
        entries %<>% rev
        if (length(dim(height))) {
            height %<>% {
                .[, rev(1:ncol(.)), drop = FALSE]
            }
        }
        else height %<>% rev
        if (length(errBar)) 
            errBar %<>% {
                .[, rev(1:ncol(.)), drop = FALSE]
            }
    }
    max1 <- if (length(dim(height))) 
        max(colSums(height))
    else max(height)
    col1 <- if (length(dim(height))) 
        col1
    else col1[1]
    text.width <- text.width.hs(entries, cm = 1)
    xlab1.width <- text.width.hs(xlab1, cm = 1)
    mai <- c(1, if (horiz) text.width else 1, 1, 0.8)
    if (!length(width1)) 
        width1 <- text.width/2.54 + max(5, xlab1.width/2.54) + 
            1
    if (!length(height1)) 
        height1 <- 5 + (if (horiz) 
            (length(entries) %>% {
                if (. > 15) 
                  (. - 15) * 0.15
                else 0
            })
        else 0)
    pin <- c(width1, height1)
    pdf(wj, width = width1 + sum(mai[c(2, 4)]), height = height1 + 
        sum(mai[c(1, 3)]))
    on.exit(dev.off(), add = TRUE)
    op1 <- par(mai = mai, pin = pin, xpd = TRUE, family = "")
    on.exit(par(op1), add = TRUE)
    if (length(par.par)) {
        op <- do.call("par", par.par)
        on.exit(par(op), add = TRUE)
    }
    Args <- list(height, border = NA, horiz = horiz, col = col1, 
        axes = TRUE, xlab = xlab1, main = main1, beside = beside, 
        plot = TRUE, names.arg = rep("", length(entries)), ...)
    if (length(dim(height))) {
        Args["legend.text"] <- list(rownames(height))
        Args[["args.legend"]] <- list(x = 0, y = -2, xpd = TRUE)
    }
    eval(bquote(Args[.(if (horiz) 
        "xlim"
    else "ylim")] <- list(c(0, .(max(height) %>% ceiling %>% 
        {
            n1 <- nchar(.) - 1
            ceiling(./(10^n1)) * (10^n1)
        })))))
    y1 <- do.call("barplot", Args)
    if (length(dim(y1)) && (length(y1) > max(dim(y1)))) {
        lapply(seq_along(errBar), function(i1) {
            x1 <- height[i1]
            x2 <- x1 + errBar[i1]
            lines(c(x1, x2), rep(y1[i1], 2))
            y1[i1] + 0.25 * c(-1, 1)
            lines(rep(x2, 2), y1[i1] + 0.25 * c(-1, 1))
        })
        y1 <- colMeans(y1)
    }
    hs.text <- function(label, x1, y1, pos = 1, adj = 1) {
        if (is.list(label)) {
            xy <- cbind(x1, y1)
            lapply(seq_along(label), function(i1) {
                text(xy[i1, 1], xy[i1, 2], label[[i1]], pos = pos, 
                  adj = adj)
            })
        }
        else text(x1, y1, label, pos = pos, adj = adj)
    }
    1
    {
        xy <- if (horiz) 
            cbind(-max(strwidth(entries)), y1)
        else cbind(y1, 0)
        for (i1 in 1:nrow(xy)) {
            text(xy[i1, 1], xy[i1, 2], entries[i1], adj = if (horiz) 
                0
            else 0.5)
        }
    }
    {
        if (max1 > 100) {
            seq(50, max1, by = 100) %>% sapply(. %>% {
                lines(rep(., 2), range(y1), col = "gray", lty = 2)
            })
            seq(100, max1, by = 100) %>% sapply(. %>% {
                lines(rep(., 2), range(y1), col = "gray", lty = 1)
            })
        }
    }
    if (length(ref1)) {
        args <- list(col = "gray", lty = 2)
        args[[ifelse(horiz, "v", "h")]] <- ref1
        do.call(abline, args)
    }
    1
    {
        if (length(dim(height)) == 0) {
            if (horiz) {
                text(height, y1, height, pos = 4)
            }
        }
    }
    if (length(post.plot)) 
        eval(post.plot)
    wj %>% hs.wj.portable
}
"37_11_26_202754" <- local({
    hs.gs_local_path <- function(gs1, path_local_root = "~/swxx/sj/db/gatk") {
        if (hs.grepl(gs1, "(?i)\\.ba[im]$")) 
            return(gs1)
        if (hs.grepl(gs1, "^gs:/")) {
            wj2 <- hs.sub(gs1, "gs:/", path_local_root)
            if (!file.exists(wj2)) {
                .wk("gsutil", "cp -crP", gs1, dirname(wj2) %>% 
                  hs.wjj)
            }
            return(wj2)
        }
        else return(gs1)
    }
    mutect2.inputs <- {
    }
    if (0) {
        sapply(mutect2.inputs, hs.gs_local_path, path_local_root = "/Users/wk/sshfs/r820/media/wk/data/swxx/sj/db/gatk")
    }
    GetSampleName <- function(bam_wj) {
        H <- .wk("samtools view -H", .sys.fpGood(bam_wj), intern = 1)
        rg <- hs.grepV(H, "(?i)^@rg")
        if (length(rg) != 1) 
            stop("37_11_26_142003")
        rg <- strsplit(rg, "\t")[[1]]
        return(hs.grepV(rg, "(?i)^sm:") %>% hs.sub("(?i)sm:"))
    }
    get_fs_in_gb <- function(wj) {
        sum(file.size(wj))/1e+09
    }
    M2 <- function(intervals = {
    }, ref_fasta, ref_fai, ref_dict, tumor_reads, tumor_reads_index, 
        normal_reads = {
        }, normal_reads_index = {
        }, pon = {
        }, pon_idx = {
        }, gnomad = {
        }, gnomad_idx = {
        }, m2_extra_args = {
        }, getpileupsummaries_extra_args = {
        }, make_bamout = 0, run_ob_filter = 0, compress, gga_vcf = {
        }, gga_vcf_idx = {
        }, variants_for_contamination = {
        }, variants_for_contamination_idx = {
        }, gatk_override = {
        }, gcs_project_for_requester_pays = {
        }, gatk_docker, mem = {
        }, preemptible = {
        }, max_retries = {
        }, disk_space = {
        }, cpu = {
        }, use_ssd = 0) {
        output_vcf <- hs.pastec0(intervals, ".output.vcf", if (compress) 
            ".gz")
        output_vcf_idx <- hs.pastec0(output_vcf, ".", ifelse(compress, 
            "tbi", "idx"))
        output_stats <- paste0(output_vcf, ".stats")
        machine_mem <- if (length(mem)) 
            mem * 1000
        else 3500
        command_mem <- machine_mem - 500
        tumor_sample <- GetSampleName(tumor_reads)
        if (length(normal_reads)) {
            normal_sample <- GetSampleName(normal_reads)
        }
        if (wj.update.reporter(output_vcf, c(tumor_reads, normal_reads, 
            ref_fasta, gnomad, pon, intervals, gga_vcf))) {
            .wk("gatk", "--java-options", paste0("-Xmx", command_mem, 
                "m"), "Mutect2", "-R", ref_fasta, "-I", tumor_reads, 
                "-tumor", tumor_sample, if (length(normal_reads)) 
                  c("-I", normal_reads, "-normal", normal_sample), 
                if (length(gnomad)) 
                  c("--germline-resource", gnomad), if (length(pon)) 
                  c("-pon", pon), if (length(intervals)) 
                  c("-L", intervals), if (length(gga_vcf)) 
                  c("--alleles", gga_vcf), "-O", output_vcf, 
                if (make_bamout) 
                  "--bam-output bamout.bam", if (run_ob_filter) 
                  c("--f1r2-tar-gz", paste0(intervals, ".f1r2.tar.gz")), 
                m2_extra_args, {
                })
        }
        {
            if (length(variants_for_contamination)) {
                GetPileupSummaries <- function(type = "tumor") {
                  wj_reads <- get(paste0(type, "_reads"))
                  wj2 <- paste0(intervals, ".", type, "-pileups.table")
                  if (wj.update.reporter(wj2, c(wj_reads, ref_fasta, 
                    variants_for_contamination))) {
                    .wk("gatk", "--java-options", paste0("-Xmx", 
                      command_mem, "m"), "GetPileupSummaries", 
                      "-R", ref_fasta, "-I", wj_reads, if (length(intervals)) 
                        c("--interval-set-rule INTERSECTION -L", 
                          intervals), "-V", variants_for_contamination, 
                      "-L", variants_for_contamination, "-O", 
                      wj2, getpileupsummaries_extra_args, {
                      })
                  }
                }
                GetPileupSummaries()
                if (length(normal_reads)) {
                  GetPileupSummaries("normal")
                }
            }
        }
        jg <- list(unfiltered_vcf = output_vcf, unfiltered_vcf_idx = output_vcf_idx, 
            output_bamOut = paste0(intervals, ".bamout.bam"), 
            tumor_sample = tumor_sample, stats = output_stats, 
            f1r2_counts = paste0(intervals, ".f1r2.tar.gz"), 
            tumor_pileups = paste0(intervals, ".tumor-pileups.table"))
        if (length(normal_reads)) {
            jg[["normal_sample"]] <- normal_sample
            jg[["normal_pileups"]] <- hs.wjj.ls(".", "\\normal-pileups\\.table$")
        }
        return(jg)
    }
    MergeVCFs <- function(input_vcfs, input_vcf_indices, output_name, 
        compress, runtime_params) {
        output_vcf <- hs.pastec0(output_name, ".vcf", if (compress) 
            ".gz")
        output_vcf_idx <- hs.pastec0(output_vcf, ".", if (compress) 
            "tbi"
        else "idx")
        if (wj.update.reporter(c(output_vcf, output_vcf_idx), 
            input_vcfs)) {
            .wk("gatk", "--java-options", paste0("-Xmx", runtime_params[["command_mem"]], 
                "m"), "MergeVcfs", paste("-I", input_vcfs), "-O", 
                output_vcf)
        }
        return(list(merged_vcf = output_vcf, merged_vcf_idx = output_vcf_idx))
    }
    MergeBamOuts <- function(ref_fasta, ref_fai, ref_dict, bam_outs, 
        output_vcf_name, runtime_params, disk_space = {
        }) {
        mem <- paste0("-Xmx", runtime_params[["command_mem"]], 
            "m")
        .wk("gatk", "--java-options", mem, "GatherBamFiles", 
            paste("-I", bam_outs), "-O", "unsorted.out.bam", 
            "-R", ref_fasta)
        {
            wj2_bam <- paste0(output_vcf_name, ".out.bam")
            .wk("gatk", "--java-options", mem, "SortSam", "-I", 
                "unsorted.out.bam", "-O", wj2_bam, "--SORT_ORDER", 
                "coordinate", "-VALIDATION_STRINGENCY", "LENIENT")
            .wk("gatk", "--java-options", mem, "BuildBamIndex", 
                "-I", wj2_bam, "-VALIDATION_STRINGENCY", "LENIENT")
        }
        return(list(merged_bam_out = wj2_bam, merged_bam_out_index = hs.wjm(wj2_bam, 
            ext = "bai")))
    }
    MergeStats <- function(stats, runtime_params) {
        wj2 <- "merged.stats"
        if (wj.update.reporter(wj2, stats)) {
            .wk("gatk", "--java-options", paste0("-Xmx", runtime_params[["command_mem"]], 
                "m"), "MergeMutectStats", paste("-stats", stats), 
                "-O", wj2)
        }
        return(list(merged_stats = wj2))
    }
    MergePileupSummaries <- function(input_tables, output_name, 
        ref_dict, runtime_params) {
        wj2 <- paste0(output_name, ".tsv")
        if (wj.update.reporter(wj2, input_tables)) {
            .wk("gatk", "--java-options", paste0("-Xmx", runtime_params[["command_mem"]], 
                "m"), "GatherPileupSummaries", "--sequence-dictionary", 
                ref_dict, paste("-I", input_tables), "-O", wj2)
        }
        return(list(merged_table = wj2))
    }
    LearnReadOrientationModel <- function(f1r2_tar_gz, runtime_params, 
        mem = {
        }) {
        wj2 <- "artifact-priors.tar.gz"
        if (wj.update.reporter(wj2, f1r2_tar_gz)) {
            machine_mem <- c(mem, runtime_params[["machine_mem"]])[1]
            command_mem <- machine_mem - 1000
            .wk("gatk", "--java-options", paste0("-Xmx", command_mem, 
                "m"), "LearnReadOrientationModel", paste("-I", 
                f1r2_tar_gz), "-O", wj2)
        }
        return(list(artifact_prior_table = wj2))
    }
    CalculateContamination <- function(intervals = {
    }, tumor_pileups, normal_pileups = {
    }, runtime_params) {
        wj2 <- "contamination.table"
        if (wj.update.reporter(wj2, c(tumor_pileups, normal_pileups))) {
            .wk("gatk", "--java-options", paste0("-Xmx", runtime_params[["command_mem"]], 
                "m"), "CalculateContamination", "-I", tumor_pileups, 
                "-O", wj2, "--tumor-segmentation", "segments.table", 
                if (length(normal_pileups)) 
                  c("-matched", normal_pileups))
        }
        return(list(contamination_table = wj2, maf_segments = "segments.table"))
    }
    Filter <- function(intervals = {
    }, ref_fasta, ref_fai, ref_dict, unfiltered_vcf, unfiltered_vcf_idx, 
        output_name, compress, mutect_stats = {
        }, artifact_priors_tar_gz = {
        }, contamination_table = {
        }, maf_segments = {
        }, m2_extra_filtering_args = {
        }, runtime_params, disk_space = {
        }) {
        output_vcf <- hs.pastec0(output_name, ".vcf", if (compress) 
            ".gz")
        output_vcf_idx <- hs.pastec0(output_vcf, ".", if (compress) 
            "tbi"
        else "idx")
        if (wj.update.reporter(c(output_vcf, output_vcf_idx), 
            c(intervals, ref_fasta, ref_fai, ref_dict, unfiltered_vcf, 
                unfiltered_vcf_idx, mutect_stats, contamination_table, 
                maf_segments, artifact_priors_tar_gz))) {
            .wk("gatk", "--java-options", paste0("-Xmx", runtime_params[["command_mem"]], 
                "m"), "FilterMutectCalls", "-V", unfiltered_vcf, 
                "-R", ref_fasta, "-O", output_vcf, if (length(contamination_table)) 
                  c("--contamination-table", contamination_table), 
                if (length(maf_segments)) 
                  c("--tumor-segmentation", maf_segments), if (length(artifact_priors_tar_gz)) 
                  c("--ob-priors", artifact_priors_tar_gz), if (length(mutect_stats)) 
                  c("-stats", mutect_stats), "--filtering-stats", 
                "filtering.stats", m2_extra_filtering_args)
        }
        return(list(filtered_vcf = output_vcf, filtered_vcf_idx = output_vcf_idx, 
            filtering_stats = "filtering.stats"))
    }
    FilterAlignmentArtifacts <- function(ref_fasta, ref_fai, 
        ref_dict, input_vcf, input_vcf_idx, reads, reads_index, 
        output_name, compress, realignment_index_bundle, realignment_extra_args = {
        }, gcs_project_for_requester_pays = {
        }, runtime_params, mem) {
        output_vcf <- hs.pastec0(output_name, ".vcf", if (compress) 
            ".gz")
        output_vcf_idx <- hs.pastec0(output_vcf, ".", if (compress) 
            "tbi"
        else "idx")
        if (wj.update.reporter(c(output_vcf, output_vcf_idx), 
            c(ref_fasta, ref_fai, ref_dict, reads, reads_index, 
                realignment_index_bundle, input_vcf, input_vcf_idx))) {
            machine_mem <- mem
            command_mem <- machine_mem - 500
            .wk("gatk", "--java-options", paste0("-Xmx", command_mem, 
                "m"), "FilterAlignmentArtifacts", "-R", ref_fasta, 
                "-V", input_vcf, "-I", reads, "--bwa-mem-index-image", 
                realignment_index_bundle, realignment_extra_args, 
                "-O", output_vcf, if (length(gcs_project_for_requester_pays)) 
                  c("--gcs-project-for-requester-pays", gcs_project_for_requester_pays))
        }
        return(list(filtered_vcf = output_vcf, filtered_vcf_idx = output_vcf_idx))
    }
    Funcotate <- function(ref_fasta, ref_fai, ref_dict, input_vcf, 
        input_vcf_idx, reference_version, output_file_base_name, 
        output_format, compress, use_gnomad, data_sources_tar_gz = "gs://broad-public-datasets/funcotator/funcotator_dataSources.v1.7.20200521s.tar.gz", 
        control_id = {
        }, case_id = {
        }, sequencing_center = {
        }, sequence_source = {
        }, transcript_selection_mode = "CANONICAL", transcript_selection_list = {
        }, annotation_defaults = {
        }, annotation_overrides = {
        }, funcotator_excluded_fields = {
        }, filter_funcotations = 0, interval_list = {
        }, extra_args = {
        }, gcs_project_for_requester_pays = {
        }, runtime_params, disk_space = {
        }, default_ram_mb = 3000, default_disk_space_gb = 100) {
        output_maf <- paste0(output_file_base_name, ".maf")
        output_maf_index <- paste0(output_maf, ".idx")
        output_vcf <- hs.pastec0(output_file_base_name, ".vcf", 
            if (compress) 
                ".gz")
        output_vcf_idx <- hs.pastec0(output_vcf, ".", if (compress) 
            "tbi"
        else "idx")
        output_file <- if (output_format == "MAF") 
            output_maf
        else output_vcf
        output_file_index <- if (output_format == "MAF") 
            output_maf_index
        else output_vcf_idx
        message("Extracting data sources zip file...")
        hs.browser("37.11.26.210330")
        datasources_wjj <- hs.wjj("datasources_dir")
        .wk("tar -zxvf", data_sources_tar_gz, "-C", datasources_wjj, 
            "--strip-components 1")
        DATA_SOURCES_FOLDER <- datasources_wjj
        if (use_gnomad) {
            message("Enabling gnomAD...")
            for (potential_gnomad_gz in c("gnomAD_exome.tar.gz", 
                "gnomAD_genome.tar.gz")) {
                wj1 <- hs.fp(DATA_SOURCES_FOLDER, potential_gnomad_gz)
                if (file.exists(wj1)) {
                  hs.with_wd(DATA_SOURCES_FOLDER, {
                    .wk("tar -zvxf", potential_gnomad_gz)
                  })
                }
                else message("ERROR: Cannot find gnomAD folder: ", 
                  potential_gnomad_gz)
            }
        }
        1
        {
            .wk("gatk", "--java-options", paste0("-Xmx", runtime_params[["command_mem"]], 
                "m"), "Funcotator", "--data-sources-path", DATA_SOURCES_FOLDER, 
                "--ref-version", reference_version, "--output-file-format", 
                output_format, "-R", ref_fasta, "-V", input_vcf, 
                "-O", output_file, if (length(interval_list)) 
                  c("-L", interval_list), "--annotation-default", 
                paste0("normal_barcode:", if (length(control_id)) 
                  control_id
                else "Unknown"), "--annotation-default", paste0("tumor_barcode:", 
                  if (length(case_id)) 
                    case_id
                  else "Unknown"), "--annotation-default", paste0("Center:", 
                  if (length(sequencing_center)) 
                    sequencing_center
                  else "Unknown"), "--annotation-default", paste0("source:", 
                  if (length(sequence_source)) 
                    sequence_source
                  else "Unknown"), if (length(transcript_selection_mode)) 
                  c("--transcript-selection-mode", transcript_selection_mode), 
                if (length(transcript_selection_list)) 
                  paste("--transcript-list", transcript_selection_list), 
                if (length(annotation_defaults)) 
                  paste("--annotation-default", annotation_defaults), 
                if (length(annotation_overrides)) 
                  paste("--annotation-override", annotation_overrides), 
                if (length(funcotator_excluded_fields)) 
                  paste("--exclude-field", funcotator_excluded_fields), 
                if (length(filter_funcotations) && filter_funcotations) 
                  "--remove-filtered-variants", extra_args, {
                })
            if (output_format == "MAF") {
                .wk("touch", .q(output_maf_index))
            }
        }
        return(list(funcotated_output_file = output_file, funcotated_output_file_index = output_file_index))
    }
    Mutect2 <- function(intervals = {
    }, ref_fasta, ref_fai, ref_dict, tumor_reads, tumor_reads_index, 
        normal_reads = {
        }, normal_reads_index = {
        }, pon = {
        }, pon_idx = {
        }, scatter_count, gnomad = {
        }, gnomad_idx = {
        }, variants_for_contamination = {
        }, variants_for_contamination_idx = {
        }, realignment_index_bundle = hs.gs_local_path(mutect2.inputs[["Mutect2.realignment_index_bundle"]]), 
        realignment_extra_args = {
        }, run_orientation_bias_mixture_model_filter = 0, m2_extra_args = {
        }, m2_extra_filtering_args = {
        }, getpileupsummaries_extra_args = {
        }, split_intervals_extra_args = {
        }, make_bamout = 0, compress_vcfs = 0, gga_vcf = {
        }, gga_vcf_idx = {
        }, gcs_project_for_requester_pays = {
        }, run_funcotator = 0, sequencing_center = {
        }, sequence_source = {
        }, funco_reference_version = {
        }, funco_output_format = {
        }, funco_compress = 0, funco_use_gnomad_AF = 0, funco_data_sources_tar_gz = {
        }, funco_transcript_selection_mode = {
        }, funco_transcript_selection_list = hs.gs_local_path(mutect2.inputs[["Mutect2.funco_transcript_selection_list"]]), 
        funco_annotation_defaults = {
        }, funco_annotation_overrides = {
        }, funcotator_excluded_fields = {
        }, funco_filter_funcotations = 0, funcotator_extra_args = {
        }, funco_default_output_format = "MAF", gatk_docker, 
        gatk_override = {
        }, basic_bash_docker = "ubuntu:16.04", filter_funcotations = 0, 
        preemptible = {
        }, max_retries = {
        }, small_task_cpu = 2, small_task_mem = 4, small_task_disk = 100, 
        boot_disk_size = 12, learn_read_orientation_mem = 8000, 
        filter_alignment_artifacts_mem = 9000, emergency_extra_disk = {
        }, large_input_to_output_multiplier = 2.25, small_input_to_output_multiplier = 2, 
        cram_to_bam_multiplier = 6) {
        preemptible_or_default <- c(preemptible, 2)[1]
        max_retries_or_default <- c(max_retries, 2)[1]
        compress <- c(compress_vcfs, 0)[1]
        run_ob_filter <- c(run_orientation_bias_mixture_model_filter, 
            0)[1]
        make_bamout_or_default <- c(make_bamout, 0)[1]
        run_funcotator_or_default <- c(run_funcotator, 0)[1]
        filter_funcotations_or_default <- c(filter_funcotations, 
            1)[1]
        {
            ref_size <- ceiling(get_fs_in_gb(c(ref_fasta, ref_dict, 
                ref_fai)))
            tumor_reads_size <- ceiling(get_fs_in_gb(c(tumor_reads, 
                tumor_reads_index)))
            gnomad_vcf_size <- if (length(gnomad)) 
                ceiling(get_fs_in_gb(gnomad))
            else 0
            normal_reads_size <- if (length(normal_reads)) 
                ceiling(get_fs_in_gb(c(normal_reads, normal_reads_index)))
            else 0
        }
        {
            funco_tar_size <- if (length(funco_data_sources_tar_gz)) 
                ceiling(get_fs_in_gb(funco_data_sources_tar_gz) * 
                  3)
            else 100
            gatk_override_size <- if (length(gatk_override)) 
                ceiling(get_fs_in_gb(gatk_override))
            else 0
        }
        {
            disk_pad <- 10 + gatk_override_size + c(emergency_extra_disk, 
                0)[1]
        }
        {
            output_basename <- hs.wjm(tumor_reads, ext = 0, dn = 0)
            unfiltered_name <- paste0(output_basename, "-unfiltered")
            filtered_name <- paste0(output_basename, "-filtered")
            funcotated_name <- paste0(output_basename, "-funcotated")
            output_vcf_name <- paste0(output_basename, ".vcf")
        }
        tumor_cram_to_bam_disk <- ceiling(tumor_reads_size * 
            cram_to_bam_multiplier)
        normal_cram_to_bam_disk <- ceiling(normal_reads_size * 
            cram_to_bam_multiplier)
        standard_runtime <- list(gatk_docker = gatk_docker, gatk_override = gatk_override, 
            max_retries = max_retries_or_default, preemptible = preemptible_or_default, 
            cpu = small_task_cpu, machine_mem = small_task_mem * 
                1000, command_mem = small_task_mem * 1000 - 500, 
            disk = small_task_disk + disk_pad, boot_disk_size = boot_disk_size)
        m2_output_size <- tumor_reads_size/scatter_count
        m2_per_scatter_size <- tumor_reads_size + normal_reads_size + 
            ref_size + gnomad_vcf_size + m2_output_size + disk_pad
        SplitIntervals.jg <- SplitIntervals(intervals = intervals, 
            ref_fasta = ref_fasta, scatter_count = scatter_count, 
            split_intervals_extra_args = split_intervals_extra_args, 
            runtime_params = standard_runtime)
        M2.jg <- wk.mclapply(SplitIntervals.jg[["interval_files"]], 
            function(subintervals) {
                M2(intervals = subintervals, ref_fasta = ref_fasta, 
                  ref_fai = ref_fai, ref_dict = ref_dict, tumor_reads = tumor_reads, 
                  tumor_reads_index = tumor_reads_index, normal_reads = normal_reads, 
                  normal_reads_index = normal_reads_index, pon = pon, 
                  pon_idx = pon_idx, gnomad = gnomad, gnomad_idx = gnomad_idx, 
                  preemptible = preemptible, max_retries = max_retries, 
                  m2_extra_args = m2_extra_args, getpileupsummaries_extra_args = getpileupsummaries_extra_args, 
                  variants_for_contamination = variants_for_contamination, 
                  variants_for_contamination_idx = variants_for_contamination_idx, 
                  make_bamout = make_bamout_or_default, run_ob_filter = run_ob_filter, 
                  compress = compress, gga_vcf = gga_vcf, gga_vcf_idx = gga_vcf_idx, 
                  gatk_override = gatk_override, gatk_docker = gatk_docker, 
                  disk_space = m2_per_scatter_size, gcs_project_for_requester_pays = gcs_project_for_requester_pays)
            }, mc.cores = scatter_count)
        merged_vcf_size <- ceiling(get_fs_in_gb(unlist(lapply(M2.jg, 
            "[[", "unfiltered_vcf"))))
        merged_bamout_size <- ceiling(get_fs_in_gb(unlist(lapply(M2.jg, 
            "[[", "output_bamOut"))))
        if (run_ob_filter) {
            LearnReadOrientationModel.jg <- LearnReadOrientationModel(f1r2_tar_gz = unlist(lapply(M2.jg, 
                "[[", "f1r2_counts")), runtime_params = standard_runtime, 
                mem = learn_read_orientation_mem)
        }
        MergeVCFs.jg <- MergeVCFs(input_vcfs = unlist(lapply(M2.jg, 
            "[[", "unfiltered_vcf")), input_vcf_indices = unlist(lapply(M2.jg, 
            "[[", "unfiltered_vcf_idx")), output_name = unfiltered_name, 
            compress = compress, runtime_params = standard_runtime)
        MergeStats.jg <- MergeStats(stats = unlist(lapply(M2.jg, 
            "[[", "stats")), runtime_params = standard_runtime)
        if (length(variants_for_contamination)) {
            MergeTumorPileups.jg <- MergePileupSummaries(input_tables = unlist(lapply(M2.jg, 
                "[[", "tumor_pileups")), output_name = output_basename, 
                ref_dict = ref_dict, runtime_params = standard_runtime)
            if (length(normal_reads)) {
                MergeNormalPileups.jg <- MergePileupSummaries(input_tables = unlist(lapply(M2.jg, 
                  "[[", "normal_pileups")), output_name = output_basename, 
                  ref_dict = ref_dict, runtime_params = standard_runtime)
            }
            CalculateContamination.jg <- CalculateContamination(tumor_pileups = MergeTumorPileups.jg[["merged_table"]], 
                normal_pileups = if (length(normal_reads)) 
                  MergeNormalPileups.jg[["merged_table"]], runtime_params = standard_runtime)
        }
        Filter.jg <- Filter(ref_fasta = ref_fasta, ref_fai = ref_fai, 
            ref_dict = ref_dict, intervals = intervals, unfiltered_vcf = MergeVCFs.jg[["merged_vcf"]], 
            unfiltered_vcf_idx = MergeVCFs.jg[["merged_vcf_idx"]], 
            output_name = filtered_name, compress = compress, 
            mutect_stats = MergeStats.jg[["merged_stats"]], contamination_table = CalculateContamination.jg[["contamination_table"]], 
            maf_segments = CalculateContamination.jg[["maf_segments"]], 
            artifact_priors_tar_gz = LearnReadOrientationModel.jg[["artifact_prior_table"]], 
            m2_extra_filtering_args = m2_extra_filtering_args, 
            runtime_params = standard_runtime, disk_space = ceiling(get_fs_in_gb(MergeVCFs.jg[["merged_vcf"]]) * 
                small_input_to_output_multiplier) + disk_pad)
        jg <- list(filtered_vcf = Filter.jg[["filtered_vcf"]], 
            filtered_vcf_idx = Filter.jg[["filtered_vcf_idx"]], 
            filtering_stats = Filter.jg[["filtering_stats"]], 
            mutect_stats = MergeStats.jg[["merged_stats"]], contamination_table = CalculateContamination.jg[["contamination_table"]], 
            maf_segments = CalculateContamination.jg[["maf_segments"]], 
            read_orientation_model_params = LearnReadOrientationModel.jg[["artifact_prior_table"]])
        if (length(realignment_index_bundle)) {
            FilterAlignmentArtifacts.jg <- FilterAlignmentArtifacts(ref_fasta = ref_fasta, 
                ref_fai = ref_fai, ref_dict = ref_dict, reads = tumor_reads, 
                reads_index = tumor_reads_index, realignment_index_bundle = realignment_index_bundle[1], 
                realignment_extra_args = realignment_extra_args, 
                compress = compress, output_name = hs.sub(filtered_name, 
                  "[^_\\-\\.]+$", "AlignmentArtifacts_filtered"), 
                input_vcf = Filter.jg[["filtered_vcf"]], input_vcf_idx = Filter.jg[["filtered_vcf_idx"]], 
                runtime_params = standard_runtime, mem = filter_alignment_artifacts_mem, 
                gcs_project_for_requester_pays = gcs_project_for_requester_pays)
            jg[["filtered_vcf"]] %<>% c(FilterAlignmentArtifacts.jg[["filtered_vcf"]])
            jg[["filtered_vcf_idx"]] %<>% c(FilterAlignmentArtifacts.jg[["filtered_vcf_idx"]])
        }
        if (run_funcotator_or_default) {
            funcotate_vcf_input <- c(FilterAlignmentArtifacts.jg[["filtered_vcf"]], 
                Filter.jg[["filtered_vcf"]])[1]
            funcotate_vcf_input_index <- c(FilterAlignmentArtifacts.jg[["filtered_vcf_idx"]], 
                Filter.jg[["filtered_vcf_idx"]])[1]
            Funcotate.jg <- Funcotate(ref_fasta = ref_fasta, 
                ref_fai = ref_fai, ref_dict = ref_dict, input_vcf = funcotate_vcf_input, 
                input_vcf_idx = funcotate_vcf_input_index, reference_version = c(funco_reference_version, 
                  "hg19")[1], output_file_base_name = hs.wjm(funcotate_vcf_input, 
                  ext = "annotated"), output_format = if (length(funco_output_format)) 
                  funco_output_format
                else funco_default_output_format, compress = if (length(funco_compress)) 
                  funco_compress[1]
                else 0, use_gnomad = if (length(funco_use_gnomad_AF)) 
                  funco_use_gnomad_AF[1]
                else 0, data_sources_tar_gz = funco_data_sources_tar_gz, 
                case_id = M2.jg[[1]][["tumor_sample"]], control_id = M2.jg[[1]][["normal_sample"]], 
                sequencing_center = sequencing_center, sequence_source = sequence_source, 
                transcript_selection_mode = funco_transcript_selection_mode, 
                transcript_selection_list = funco_transcript_selection_list, 
                annotation_defaults = funco_annotation_defaults, 
                annotation_overrides = funco_annotation_overrides, 
                funcotator_excluded_fields = funcotator_excluded_fields, 
                filter_funcotations = filter_funcotations_or_default, 
                extra_args = funcotator_extra_args, runtime_params = standard_runtime, 
                disk_space = ceiling(get_fs_in_gb(funcotate_vcf_input) * 
                  large_input_to_output_multiplier) + funco_tar_size + 
                  disk_pad)
        }
        if (make_bamout_or_default) {
            MergeBamOuts.jg <- MergeBamOuts(ref_fasta = ref_fasta, 
                ref_fai = ref_fai, ref_dict = ref_dict, bam_outs = unlist(lapply(M2.jg, 
                  "[[", "output_bamOut")), output_vcf_name = MergeVCFs.jg[["merged_vcf"]] %>% 
                  hs.wjm(ext = 0), runtime_params = standard_runtime, 
                disk_space = ceiling(merged_bamout_size * large_input_to_output_multiplier) + 
                  disk_pad)
            jg[["bamout"]] <- MergeBamOuts.jg[["merged_bam_out"]]
            jg[["bamout_index"]] <- MergeBamOuts.jg[["merged_bam_out_index"]]
        }
        if (run_funcotator_or_default) {
            funcotate_vcf_input <- c(FilterAlignmentArtifacts.jg[["filtered_vcf"]], 
                Filter.jg[["filtered_vcf"]])[1]
            funcotate_vcf_input_index <- c(FilterAlignmentArtifacts.jg[["filtered_vcf_idx"]], 
                Filter.jg[["filtered_vcf_idx"]])[1]
            Funcotate.jg <- Funcotate(ref_fasta = ref_fasta, 
                ref_fai = ref_fai, ref_dict = ref_dict, input_vcf = funcotate_vcf_input, 
                input_vcf_idx = funcotate_vcf_input_index, reference_version = c(funco_reference_version, 
                  "hg19")[1], output_file_base_name = hs.wjm(funcotate_vcf_input, 
                  ext = "annotated"), output_format = if (length(funco_output_format)) 
                  funco_output_format
                else funco_default_output_format, compress = if (length(funco_compress)) 
                  funco_compress[1]
                else 0, use_gnomad = if (length(funco_use_gnomad_AF)) 
                  funco_use_gnomad_AF[1]
                else 0, data_sources_tar_gz = funco_data_sources_tar_gz, 
                case_id = M2.jg[[1]][["tumor_sample"]], control_id = M2.jg[[1]][["normal_sample"]], 
                sequencing_center = sequencing_center, sequence_source = sequence_source, 
                transcript_selection_mode = funco_transcript_selection_mode, 
                transcript_selection_list = funco_transcript_selection_list, 
                annotation_defaults = funco_annotation_defaults, 
                annotation_overrides = funco_annotation_overrides, 
                funcotator_excluded_fields = funcotator_excluded_fields, 
                filter_funcotations = filter_funcotations_or_default, 
                extra_args = funcotator_extra_args, runtime_params = standard_runtime, 
                disk_space = ceiling(get_fs_in_gb(funcotate_vcf_input) * 
                  large_input_to_output_multiplier) + funco_tar_size + 
                  disk_pad)
            jg[["funcotated_file_index"]] <- Funcotate.jg[["funcotated_output_file_index"]]
            jg[["funcotated_file_index"]] <- Funcotate.jg[["funcotated_output_file_index"]]
        }
        return(jg)
    }
    SplitIntervals <- function(intervals = {
    }, ref_fasta, scatter_count, split_intervals_extra_args = {
    }, runtime_params) {
        if (!length(intervals)) 
            return()
        wj2 <- hs.wjj.ls(".", "\\.interval_list$")
        if (length(wj2) == scatter_count) {
            return(list(interval_files = wj2))
        }
        else hs.wj.rm(wj2)
        wjj2 <- hs.fp(hs.wjm(intervals, add = "scattered"), paste0("n=", 
            scatter_count))
        wj2 <- hs.wjj.ls(wjj2, "\\.interval_list$")
        if (length(wj2) != scatter_count) {
            hs.wj.rm(wj2)
            hs.with_wd(wjj2, .wk("gatk", "--java-options", paste0("-Xmx", 
                runtime_params[["command_mem"]], "m"), "SplitIntervals", 
                "-R", ref_fasta, "-L", intervals, "-scatter", 
                scatter_count, "-O", ".", split_intervals_extra_args))
        }
        wj2 <- hs.wjj.ls(wjj2, "\\.interval_list$")
        file.symlink(wj2, ".")
        return(list(interval_files = basename(wj2)))
    }
    function(intervals = {
    }, ref_fasta = hs.gs_local_path(mutect2.inputs[["Mutect2.ref_fasta"]]), 
        ref_fai = hs.hsgly("fa.fai.37_04_14_182739")(ref_fasta), 
        ref_dict = hs.hsgly("fa.dict.37_12_12_102937")(ref_fasta), 
        pair_list = {
        }, pairs = if (length(pair_list)) 
            wk.fread(pair_list, cn = 0), pon = hs.gs_local_path(mutect2.inputs[["Mutect2.pon"]]), 
        pon_idx = hs.gs_local_path(mutect2.inputs[["Mutect2.pon_idx"]]), 
        gnomad = hs.gs_local_path(mutect2.inputs[["Mutect2.gnomad"]]), 
        gnomad_idx = hs.gs_local_path(mutect2.inputs[["Mutect2.gnomad_idx"]]), 
        variants_for_contamination = hs.gs_local_path(mutect2.inputs[["Mutect2.variants_for_contamination"]]), 
        variants_for_contamination_idx = hs.gs_local_path(mutect2.inputs[["Mutect2.variants_for_contamination_idx"]]), 
        filter_funcotations = 0, run_orientation_bias_mixture_model_filter = 0, 
        scatter_count, m2_extra_args = mutect2.inputs[["Mutect2.m2_extra_args"]], 
        m2_extra_filtering_args = {
        }, compress_vcfs = 0, make_bamout = 0, gcs_project_for_requester_pays = {
        }, sequencing_center = {
        }, sequence_source = {
        }, run_funcotator = 0, funco_reference_version = {
        }, funco_data_sources_tar_gz = {
        }, funco_transcript_selection_mode = {
        }, funco_transcript_selection_list = {
        }, funco_annotation_defaults = {
        }, funco_annotation_overrides = {
        }, gatk_docker = mutect2.inputs[["Mutect2.gatk_docker"]], 
        preemptible_attempts = {
        }, gatk_override = {
        }, what = {
        }) {
        if (!length(mutect2.inputs)) 
            mutect2.inputs <<- rjson::fromJSON(file = "~/org/工作/生物学/生信/宽所/gatk/somatic/资料/脚本/mutect2.inputs.json")
        if (length(what)) {
            return(switch(what, ref = ref_fasta))
        }
        pairs <- as.matrix(pairs)
        wk.mclapply(1:nrow(pairs), function(row_i) {
            row <- pairs[row_i, ]
            if (length(row) == 4) {
                normal_bam <- row[3]
                normal_bai <- row[4]
            }
            else normal_bam <- normal_bai <- {
            }
            Mutect2.jg <- Mutect2(intervals = intervals, ref_fasta = ref_fasta, 
                ref_fai = ref_fai, ref_dict = ref_dict, tumor_reads = row[1], 
                tumor_reads_index = row[2], normal_reads = normal_bam, 
                normal_reads_index = normal_bai, pon = pon, pon_idx = pon_idx, 
                scatter_count = scatter_count, gnomad = gnomad, 
                gnomad_idx = gnomad_idx, variants_for_contamination = variants_for_contamination, 
                variants_for_contamination_idx = variants_for_contamination_idx, 
                filter_funcotations = filter_funcotations, run_orientation_bias_mixture_model_filter = run_orientation_bias_mixture_model_filter, 
                m2_extra_args = m2_extra_args, m2_extra_filtering_args = m2_extra_filtering_args, 
                sequencing_center = sequencing_center, sequence_source = sequence_source, 
                run_funcotator = run_funcotator, funco_reference_version = funco_reference_version, 
                funco_data_sources_tar_gz = funco_data_sources_tar_gz, 
                funco_transcript_selection_mode = funco_transcript_selection_mode, 
                funco_transcript_selection_list = funco_transcript_selection_list, 
                funco_annotation_defaults = funco_annotation_defaults, 
                funco_annotation_overrides = funco_annotation_overrides, 
                make_bamout = make_bamout, compress_vcfs = compress_vcfs, 
                gatk_override = gatk_override, gatk_docker = gatk_docker, 
                preemptible = preemptible_attempts, gcs_project_for_requester_pays = gcs_project_for_requester_pays)
            jg <- Mutect2.jg[c("filtered_vcf", "filtered_vcf_idx", 
                "contamination_table")]
            if (make_bamout) {
                jg["m2_bamout"] <- Mutect2.jg["bamout"]
                jg["m2_bamout_index"] <- Mutect2.jg["bamout_index"]
            }
            return(hs.lapply(jg, normalizePath))
        }, mc.cores = 1)
    }
})
"39_02_21_114353" <- function(sub = {
}) {
    jg <- .mfp.swxx("39_02_21_114435")
    if (length(sub)) 
        jg <- hs.fp(jg, sub)
    jg
}
"39_05_22_180320" <- function() {
    wjj2 <- hs.sjgly("39_05_22_180320")
    wj <- hs.fp("~/d/rjk/微信/WeChat Files/sophiewk/FileStorage/File/2023-05", 
        c("ONB_311_Major_clusters.rds", "ONB_311_Malignant_clusters.rds"))
    return(hs.wjm(wj, dn = wjj2) %=>% hs.fp_add_name)
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        if (0) {
            for (wj1 in wj) wk.rsync.out(wj1, wjj2, ssh = "r820")
        }
        .ls(wjj2)
        ONB_311_Major_clusters <- hs.fp(wjj2, "ONB_311_Major_clusters.rds") %=>% 
            hs.load
        ONB_311_Major_clusters@meta.data[1, ]
        ONB_311_Major_clusters@meta.data[, "major_type"] %=>% 
            hs.table
        a <- hs.fp(wjj2, "ONB_311_Malignant_clusters.rds") %=>% 
            hs.load
        str(a)
        a@meta.data[1, ]
    }
}
"2023_06_04_222244" <- function() {
    wjj2 <- hs.sjgly("2023_06_04_222244")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        h5_wj <- hs.hsgly("h5_fp.39_05_02_090920")()
        sce_list <- hs.hsgly("10x2seurat.38_08_12_145141")(h5_wj)
        {
            if (0) {
                sce_list[["ONB_311_20210824NA_on_human+mouse"]]@meta.data["AAACCCAAGGTAAGAG-1", 
                  ]
                sce_list[["ONB_311_20210824NA_on_human+mouse"]] %=>% 
                  rownames %=>% hs.re("[^-]*") %=>% unique
                a <- wk.fread("~/swxx/xm/db/39/4/24/pdx单细胞/表达量/ONB_311_20210824NA/outs/analysis/gem_classification.csv")
                a <- wk.fread("~/swxx/xm/db/39/4/24/pdx单细胞/表达量/ONB_311_PDX_Control_20220107NA/outs/analysis/gem_classification.csv")
                a[, call %=>% hs.table(type = "b")]
                a <- h5_wj[sample_name_2do] %=>% sapply(x, . %=>% 
                  {
                    dirname(x) %=>% hs.fp("analysis/gem_classification.csv")
                  }) %=>% lapply(x, wk.fread)
                lapply(a, . %=>% {
                  x[, call %=>% hs.table(type = "b", percentage = 1)]
                })
                wk.fread("~/swxx/xm/db/39/4/24/pdx单细胞/表达量/ONB_311_20210824NA/outs/metrics_summary.csv") %=>% 
                  x[, .SD, .SDcols = patterns("(?i)cell")]
            }
            sample_name_2do <- sce_list %=>% names %=>% (`%not%`("ONB_311_20210824NA"))
            cn1 <- "human_pct"
            for (sample_name_1 in sample_name_2do) {
                p1 <- local({
                  if (0) {
                    p1 <- sce_list[[sample_name_1]] %=>% Seurat::PercentageFeatureSet("^GRCh38-")
                    {
                      if (0) {
                        names(p1) %=>% unique %=>% sapply(x, 
                          . %=>% {
                            i1 <- which(names(p1) %in% x)
                            (length(i1) == 2) && (p1[i1[1]] == 
                              p1[i1[2]])
                          }) %=>% all
                      }
                      p1 %<=>% x[match(colnames(sce_list[[sample_name_1]]), 
                        names(x))]
                    }
                    return(p1)
                  }
                  {
                    ri_human <- sce_list[[sample_name_1]]@assays[["RNA"]]@counts %=>% 
                      rownames %=>% hs.grep("^GRCh38-")
                    ri_mouse <- sce_list[[sample_name_1]] %=>% 
                      nrow %=>% seq %=>% (`%not%`(ri_human))
                    a <- sce_list[[sample_name_1]]@assays[["RNA"]]@counts[ri_human, 
                      ] %=>% hs.colSums
                    b <- sce_list[[sample_name_1]]@assays[["RNA"]]@counts[ri_mouse, 
                      ] %=>% hs.colSums
                    return(a/(a + b) * 100)
                  }
                })
                sce_list[[sample_name_1]]@meta.data[, cn1] <- p1
            }
            wj2 <- wjj2 %=>% hs.fp("人含量分布.pdf")
            if (!file.exists(wj2)) {
                a <- lapply(sample_name_2do, . %=>% {
                  a <- h5_wj[x] %=>% dirname %=>% hs.fp("analysis/gem_classification.csv") %=>% 
                    wk.fread
                  jg <- list(x, sce_list[[x]]@meta.data[, cn1], 
                    a[match(colnames(sce_list[[x]]), barcode), 
                      call])
                  names(jg) <- c("sample", cn1, "org")
                  wk.dt.as(jg)
                })
                a %<=>% rbindlist
                a[, `:=`(hs.sub(cn1, "human", "mouse"), 100 - 
                  get(cn1))]
                if (0) {
                  p1 <- ggplot2::ggplot(a, ggplot2::aes(x = `human%`, 
                    y = `mouse%`, color = org)) + ggplot2::geom_point() + 
                    facet_grid(. ~ sample)
                  height <- 7
                  .pdf(print(p1), wj2, width = a[, sample %=>% 
                    uniqueN] * height, height = height)
                  "~/swxx/xm/db/39/4/24/pdx单细胞/表达量/质控/人含量分布.pdf" %=>% 
                    wk.rsync.in(ssh = "r820")
                  a[, org %=>% hs.table]
                }
                hs.hsgly("小提琴图/36.06.18.150541")(a, wj2 = wj2, 
                  jitter.col = a[, c(GRCh38 = ieiwk.col2rgb("red", 
                    0.2), mm10 = ieiwk.col2rgb("blue", 0.2), 
                    Multiplet = ieiwk.col2rgb("gray", 0.5))[org]])
            }
            if (0) {
                a[, org %=>% unique]
                a[org == "mm10", hs.table(sample, type = "b", 
                  percentage = 1)]
                a <- c(0.5, 1, 2, 5, 50) %=>% hs.lapply(x, function(p1) {
                  a <- hs.lapply(sample_name_2do, function(sample_name_1) {
                    x <- sce_list[[sample_name_1]]@meta.data[, 
                      "human_pct"]
                    i <- sce_list[[sample_name_1]]@meta.data[, 
                      "human_pct"] %=>% {
                      (100 - x) > p1
                    }
                    c(sum(i), mean(i) * 100)
                  })
                  wk.dt.fromList(a, by_row = 1, nm = c("sample", 
                    "#cell", "%cell"))
                })
                a <- wk.dt.fromList(a, nm = "mouse_pct")
                .s(a)
            }
        }
        {
            {
                a <- sce_list %=>% lapply(x, . %=>% {
                  if (any(hs.grepl(rownames(x), "^(?i)GRCh38-"))) {
                    x <- hs.ij(x, rownames(x) %=>% hs.grep("^(?i)GRCh38-"))
                    old_new <- rownames(x) %=>% wk.dt(x, hs.grep_sub(x, 
                      "^(?i)GRCh38-"))
                    hs.hsgly("seurat对象改基因名.39_05_16_095300")(x, 
                      old_new = old_new)
                  }
                  else x
                })
                seurat_data_human <- hs.hsgly("合并数据.38_12_09_141645")(a)
                seurat_data_human %=>% dim
                seurat_data_human@meta.data[1, ]
                {
                  cn1 <- "cell_ranger_call"
                  seurat_data_human@meta.data[, cn1] <- "NA"
                  a <- hs.hsgly("表达量/39_05_02_090920")() %=>% 
                    {
                      hs.c(names(x), hs.fp(dirname(x), "analysis/gem_classification.csv")) %=>% 
                        x[file.exists(x)]
                    }
                  names(a) %all.in% seurat_data_human@meta.data[, 
                    "orig.ident"]
                  for (sample1 in names(a)) {
                    b <- a[sample1] %=>% wk.fread
                    paste0(sample1, "_", b[, barcode])
                    seurat_data_human@meta.data[paste0(sample1, 
                      "_", b[, barcode]), cn1] <- b[, call]
                  }
                }
                seurat_data_human %<=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")(x, 
                  normalization.method = "log", nfeatures = 2000, 
                  wjj2 = wjj2)
                seurat_data_human %=>% hs.hsgly("seurat/函数/idents，获取.38_12_15_114046")(x, 
                  id = "orig.ident")
                {
                  pt.size <- 0.1
                  reduction1 <- "pca"
                  reduction1 <- "harmony"
                  wjj2 %=>% .ls
                  hs.fp(wjj2, "umap_去除批次.样本.分开.pdf")
                  wjj2_here <- wjj2
                  for (reduction1 in c("pca", "harmony")) {
                    wj2_all <- c("重叠", "分开") %=>% hs.c(x, 
                      hs.fp(wjj2_here, paste0("umap_", switch(reduction1, 
                        pca = "保留", harmony = "去除"), 
                        "批次.样本.", x, ".pdf")))
                    if (!all(file.exists(wj2_all))) 
                      seurat_data_human %<=>% hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")(x, 
                        sample_cn = "orig.ident", reduction = reduction1, 
                        dim_n = 10, wjj2 = wjj2_here)
                    {
                    }
                    wj2 <- wj2_all["重叠"]
                    if (!file.exists(wj2)) {
                      p1 <- seurat_data_human %=>% hs.hsgly("单细胞/转录组/seurat/函数/图/低维.38_12_11_113923")(x, 
                        shuffle = 1, first_is_gray = 0)
                      p1 %<=>% hs.hsgly("ggplot2.post.36_09_16_194613")(x)
                      .pdf(p1, wj2, height = 25)
                    }
                    wj2 <- wj2_all["分开"]
                    if (!file.exists(wj2)) {
                      p1 <- seurat_data_human %=>% hs.hsgly("单细胞/转录组/seurat/函数/图/低维.38_12_11_113923")(x, 
                        shuffle = 1, first_is_gray = 0, split.by = "orig.ident", 
                        combine = FALSE)
                      p1 %<=>% hs.hsgly("list.patchwork.plot_layout.39_05_28_223622")(x, 
                        ncol = 3)
                      p1 %<=>% hs.hsgly("ggplot2.post.36_09_16_194613")(x)
                      .fig(p1, height = h1 * 2, width = h1 * 
                        3)
                      .fig(p1, wj2, height = h1 * 2, width = h1 * 
                        3)
                    }
                  }
                }
                {
                  h1 <- 30
                  p1 <- Seurat::DimPlot(seurat_data_human, shuffle = TRUE, 
                    raster = TRUE, cols = "BrBG", pt.size = pt.size)
                  .pdf(print(p1), width = h1 * 1.5, height = h1)
                  p1 <- Seurat::DimPlot(seurat_data_human, shuffle = TRUE, 
                    raster = TRUE, cols = "BrBG", split.by = "orig.ident")
                  h1 <- 10
                  .pdf(print(p1), width = h1 * uniqueN(seurat_data@meta.data[, 
                    "orig.ident"]), height = h1)
                  `?`(Seurat::DimPlot)
                  help(package = "Seurat")
                  seurat_data_human@meta.data[, "mouse."] %<=>% 
                    {
                      if (anyNA(x)) {
                        x[is.na(x)] <- 0
                        x
                      }
                      else x
                    }
                  hs.require("patchwork")
                  {
                    cn1 <- "mouse_pct"
                    seurat_data_human@meta.data[, cn1] <- seurat_data_human@meta.data[, 
                      "mouse."]
                    h1 <- 15
                    p1 <- Seurat::FeaturePlot(seurat_data_human, 
                      features = cn1, pt.size = pt.size, raster = TRUE, 
                      split.by = "orig.ident", keep.scale = "all", 
                      order = TRUE, combine = FALSE)
                    seurat_data_human@meta.data[, cn1] <- {
                    }
                    p1 %<=>% {
                      cmd1 <- paste0("x[[ ", seq_along(x), "]]") %=>% 
                        c("patchwork::plot_layout( ncol = 3)") %=>% 
                        paste(collapse = " + ") %=>% parse(text = x)
                      eval(cmd1)
                    }
                    .pdf(print(p1), width = h1 * 3 * 1.2, height = h1 * 
                      2)
                    cell2do <- seurat_data_human@meta.data %=>% 
                      rownames(x)[x[, "cell_ranger_call"] %not.in% 
                        c("mm10", "Multiplet")]
                    p1 <- Seurat::FeaturePlot(seurat_data_human, 
                      features = cn1, cells = cell2do, pt.size = pt.size, 
                      raster = TRUE, split.by = "orig.ident", 
                      keep.scale = "all", order = TRUE, combine = FALSE)
                    p1 %<=>% {
                      cmd1 <- paste0("x[[ ", seq_along(x), "]]") %=>% 
                        c("patchwork::plot_layout( ncol = 3)") %=>% 
                        paste(collapse = " + ") %=>% parse(text = x)
                      eval(cmd1)
                    }
                    .pdf(print(p1), width = h1 * 3 * 1.2, height = h1 * 
                      2)
                  }
                }
                h1 <- 10
                p1 <- Seurat::DimPlot(seurat_data_human, split.by = "orig.ident")
                .pdf(print(p1), width = h1 * uniqueN(seurat_data_human@meta.data[, 
                  "orig.ident"]), height = h1, wj = wj2)
                seurat_data_human
                "~/swxx/xm/db/39/4/24/pdx单细胞/表达量/质控/全貌/umap_样本_不去批次.pdf" %=>% 
                  wk.rsync.in(ssh = "r820")
                "~/swxx/xm/db/39/4/24/pdx单细胞/表达量/质控/全貌/umap_样本_去批次.pdf" %=>% 
                  wk.rsync.in(ssh = "r820")
            }
            seurat_data <- hs.hsgly("合并数据.38_12_09_141645")(sce_list)
            {
                seurat_data@meta.data[1, ]
                chk("Seurat")
                p1 <- Seurat::DimPlot(seurat_data_human, split.by = "orig.ident")
                seurat_data_human %=>% str
                h1 <- 10
                .pdf(print(p1), width = h1 * uniqueN(seurat_data@meta.data[, 
                  "orig.ident"]), height = h1)
                "/dev/shm/wk/tmp.pdf" %=>% wk.rsync.in
                `?`(Seurat::DimPlot)
            }
            seurat_data %=>% dimnames
            sce_list[[1]] %=>% dimnames
            sce_list %=>% names
            sce_list[["ONB_311_20210824NA"]] %=>% dimnames
            sce_list[["ONB_311_20210824NA_on_human+mouse"]] %=>% 
                rownames %=>% hs.grep_sub("^(?i)GRCh38-") %=>% 
                hs.xor(sce_list[["ONB_311_20210824NA"]] %=>% 
                  rownames)
        }
    }
}
"2023_06_06_034437" <- function() {
    wjj2 <- hs.sjgly("2023_06_06_034437")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        seurat_data_human <- hs.hsgly("人的/2023_06_04_222244")()
        {
            ref <- hs.hsgly("百奥智汇的结果/39_05_22_180320")()["ONB_311_Major_clusters"] %=>% 
                hs.load
            cell_type_cn <- "major_type"
            if (0) {
                str(ref)
                ref@meta.data[, cell_type_cn] %=>% hs.table %=>% 
                  sort
            }
            pc_n <- max(ncol(ref@reductions[["pca"]]), hs.hsgly("转录组/标准流程/模块/r函数_自制.38_04_07_003149", 
                "get.dimensionality")(seurat_data_human))
            anchors <- Seurat::FindTransferAnchors(reference = ref, 
                query = seurat_data_human, reference.assay = SeuratObject::DefaultAssay(ref), 
                query.assay = SeuratObject::DefaultAssay(seurat_data_human), 
                dims = seq(pc_n))
            predictions <- Seurat::TransferData(anchorset = anchors, 
                refdata = ref@meta.data[, cell_type_cn])
            seurat_data_human@meta.data[, colnames(predictions)] <- predictions
            id1 <- "predicted.id"
            for (reduction1 in c("pca", "harmony")) {
                wj2_all <- c("重叠", "分开") %=>% {
                  hs.c(x, hs.fp(wjj2, paste0("umap_", switch(reduction1, 
                    pca = "保留", harmony = "去除"), "批次.样本.", 
                    x, ".pdf")))
                }
                if (!all(file.exists(wj2_all))) 
                  seurat_data_human %<=>% hs.hsgly("转录组/标准流程/模块/预处理/降维，非线性.38_10_09_121231")(x, 
                    sample_cn = "orig.ident", reduction = reduction1, 
                    dim_n = 10, wjj2 = wjj2)
                wj2 <- wj2_all["重叠"]
                if (!file.exists(wj2)) {
                  p1 <- seurat_data_human %=>% hs.hsgly("单细胞/转录组/seurat/函数/图/低维.38_12_11_113923")(x, 
                    id = id1, shuffle = 1, first_is_gray = 0)
                  p1 %<=>% hs.hsgly("ggplot2.post.36_09_16_194613")(x)
                  .pdf(p1, wj2, height = 25)
                }
                wj2 <- wj2_all["分开"]
                if (!file.exists(wj2)) {
                  p1 <- seurat_data_human %=>% hs.hsgly("单细胞/转录组/seurat/函数/图/低维.38_12_11_113923")(x, 
                    id = id1, shuffle = 1, first_is_gray = 0, 
                    split.by = "orig.ident", combine = FALSE)
                  p1 %<=>% hs.hsgly("list.patchwork.plot_layout.39_05_28_223622")(x, 
                    ncol = 3)
                  p1 %<=>% hs.hsgly("ggplot2.post.36_09_16_194613")(x)
                  .pdf(p1, wj2, height = h1 * 2, width = h1 * 
                    3)
                  "~/swxx/xm/db/39/4/24/pdx单细胞/表达量/人的/umap_去除批次.样本.分开.pdf" %=>% 
                    wk.rsync.in(ssh = "r820")
                }
            }
            seurat_data_human@meta.data[1, ]
        }
    }
}
"39_05_02_090920" <- function() {
    wjj2 <- hs.sjgly("39_05_02_090920")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        fq_fp_list <- hs.hsgly("质控/39_05_02_085932")()
        sce_list <- hs.hsgly("转录组/标准流程/模块/10x/cell_ranger.38_08_19_130133")(fq_fp_list = fq_fp_list, 
            org1 = "human_mouse", wjj2 = wjj2, wjj2_here_post = {
            }, nc = 1, ot = "seurat_data_list")
        sce_list <- hs.hsgly("转录组/标准流程/模块/10x/cell_ranger.38_08_19_130133")(fq_fp_list = fq_fp_list["ONB_311_20210824NA"], 
            org1 = "human", wjj2 = paste0(wjj2, "_人基因组"), 
            wjj2_here_post = {
            }, nc = 1, ot = "seurat_data_list")
    }
    wjj2 %=>% c(x, paste0(x, "_人基因组")) %=>% lapply(x, 
        . %=>% hs.wjj.ls(x, wjj = 1)) %=>% unlist %=>% hs.fp("outs/filtered_feature_bc_matrix.h5", 
        e = 1) %=>% {
        a <- hs.hsgly("公私.39_05_12_131245")(x)[["private"]] %=>% 
            hs.dirname(2)
        i <- which(a == "表达量/ONB_311_20210824NA")
        a[i] %<=>% {
            basename(x) %=>% paste0("_on_human+mouse")
        }
        a[-i] %<=>% basename
        names(x) <- a
        x
    }
}
"39_05_26_072156" <- function() {
    wjj2 <- hs.sjgly("39_05_26_072156")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        seurat_data <- hs.hsgly("单细胞/转录组/预测细胞类型/用seurat.38_09_14_164012")(seurat_data = seurat_data, 
            org1 = org1, tissue_type = tissue_type, cell_type_column_name = cell_type_cn, 
            clustering_cn2use = clustering_cn2use, clustering_chosen_cn = clustering_chosen_cn)
    }
}
"39_05_02_090046" <- function() {
    "~/swxx/sj/db/2023/5/1/杨婧艺-pdx单细胞" %=>% c(hs.fp(x, 
        "ENB_311"), hs.fp(x, "ONB-311-PDX") %=>% hs.wjj.ls(wjj = 1))
}
"39_05_02_085932" <- function() {
    wjj2 <- hs.sjgly("39_05_02_085932")
    if (0) {
        wjj2 %=>% hs.fp("multiQC/multiqc_report.html") %=>% .ls
        "~/swxx/xm/db/39/4/24/pdx单细胞/原始数据/质控/multiQC/multiqc_report.html" %=>% 
            wk.rsync.in(ssh = "r820")
    }
    sample_wjj <- hs.hsgly("原始数据/39_05_02_090046")()
    if (0) {
        lapply(sample_wjj, . %=>% {
            hs.hsgly("文件/检查完整性/md5sum.38_05_13_124417")(x)
        })
    }
    {
        fq_fp_list <- sample_wjj %=>% lapply(x, hs.hsgly("wjj2fq.37_12_25_144226")) %=>% 
            hs.unlist(recursive = 0, use.names = 1)
        fq_fp_list %=>% hs.hsgly("标准流程/模块/预处理/质控/fastqc.38_10_08_091811")(fq_fp_list = x, 
            wjj2 = wjj2)
        hs.hsgly("标准流程/模块/预处理/质控/multiqc.38_10_08_093432")(wjj1 = wjj2)
        fq_fp_list
    }
}
"39_05_05_133945" <- function() {
    wjj2 <- hs.sjgly("39_05_05_133945")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        hs.with_wd(wjj2, hs.hsgly("云/阿里/oss.39_05_05_134720")(e = "https://oss-cn-guangzhou.aliyuncs.com", 
            i = "LTAI5tRJbBmyS4VqR6xm5jTU", k = "9T8S5yoFNC31oqzGOws52mkIWW4AAA") %=>% 
            .wk("sync -u", "oss://genedenovo-gdr22110307-8-c1oy7mmadojf23xr", 
                "./"))
        wjj2 %=>% hs.wjj.ls("(doc|exe)$") %=>% hs.wj.rm
        wjj2 %=>% hs.wjj.ls(wjj = 1) %=>% lapply(x, . %=>% hs.hsgly("文件/检查完整性/md5sum.38_05_13_124417")(x))
    }
    hs.wjj.ls(wjj2, "(?i)\\.h5$", R = 1)
}
"39_05_03_175507" <- function() {
    wjj2 <- hs.sjgly("39_05_03_175507")
    hs.fp(wjj2, "in") %=>% .ls
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        if (0) 
            wk.rsync.out("~/org/工作/生物学/生信/零活/39/5/3/变异瀑布图/瀑布图绘制.xlsx", 
                hs.fp(wjj2, "in"), ssh = "r820")
        sj <- hs.fp(wjj2, "in/瀑布图绘制.xlsx") %=>% wk.xls.read("")
        sj[["基因突变"]][, mutation_type %=>% hs.table %=>% 
            hs.sort]
        sj[["基因突变"]][hs.grep(mutation_type, "^(cn|gene)_", 
            inv = 1)]
        sj[["基因突变"]][nchar(ref) %=>% {
            (x > 1) & (x == nchar(alt))
        }, mutation_type %=>% hs.table]
        simple <- sj[["基因突变"]] %=>% wk.dt.subset(x, exp_l = c(hs.grep(mutation_type, 
            "^(cn|gene)_", inv = 1), pos %=>% as.numeric %=>% 
            is.na %=>% `!`, mutation_type %not.in% "splice_region_variant"))
        simple[, `:=`("pos", as.numeric(pos))]
        if (0) {
            simple %=>% dim
            db <- c("A", "C", "G", "T")
            db <- "(?i)[acgt]"
            simple %=>% wk.dt.subset(exp_l = c((ref %not.in% 
                db) | (alt %not.in% db)), exp_j = .(ref, alt)) %=>% 
                hs.df %=>% .s
        }
        maf <- simple[, .(Hugo_Symbol = gene, Chromosome = chrom, 
            Start_Position = pos, End_Position = pos + nchar(ref) - 
                1, Reference_Allele = ref, Tumor_Seq_Allele2 = alt, 
            Variant_Classification = mutation_type, Tumor_Sample_Barcode = sample_id)]
        simple[mutation_type %in% "missense_variant", all(nchar(ref) == 
            nchar(alt))]
        i <- simple[, which((mutation_type %in% "missense_variant") | 
            (nchar(ref) == nchar(alt)))]
        wk.dt.set(maf, i, "Variant_Type", "SNP")
        i <- simple[, which(nchar(ref) < nchar(alt))]
        wk.dt.set(maf, i, "Variant_Type", "INS")
        i <- simple[, which(nchar(ref) > nchar(alt))]
        wk.dt.set(maf, i, "Variant_Type", "DEL")
        if (0) {
            maf[, Variant_Type %=>% hs.table]
            simple[(nchar(ref) > 1) & (mutation_type %in% "missense_variant"), 
                .(ref, alt)]
            "Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, Variant_Classification, Variant_Type and Tumor_Sample_Barcode" %=>% 
                hs.strsplit.v("(,|and) *") %=>% trim %=>% (`%not%`(names(maf)))
        }
        {
            `?`(maftools::read.maf)
            maf[, Variant_Classification %=>% hs.table %=>% hs.sort]
            maf[Variant_Classification %in% "missense_variant", 
                `:=`("Variant_Classification", "Missense_Mutation")]
            maf[(Variant_Classification %in% "frameshift_variant") & 
                (Variant_Type %in% "INS"), `:=`("Variant_Classification", 
                "Frame_Shift_Ins")]
            maf[(Variant_Classification %in% "frameshift_variant") & 
                (Variant_Type %in% "DEL"), `:=`("Variant_Classification", 
                "Frame_Shift_Del")]
            maf[Variant_Classification %in% "stop_gained", `:=`("Variant_Classification", 
                "Nonsense_Mutation")]
            maf[Variant_Classification %in% c("splice_acceptor_variant", 
                "splice_donor_variant"), `:=`("Variant_Classification", 
                "Splice_Site")]
            maf[hs.grep(Variant_Classification, "inframe_deletion"), 
                `:=`("Variant_Classification", "In_Frame_Del")]
            maf[hs.grep(Variant_Classification, "inframe_insertion"), 
                `:=`("Variant_Classification", "In_Frame_Ins")]
            maf[Variant_Classification %in% "start_lost", `:=`("Variant_Classification", 
                "Translation_Start_Site")]
            maf[Variant_Classification %in% "stop_lost", `:=`("Variant_Classification", 
                "Nonstop_Mutation")]
            wk.dt.fwrite(maf, hs.fp(wjj2, "maf.tsv.gz"))
            "~/swxx/xm/db/39/5/3/变异瀑布图/maf.tsv.gz" %=>% 
                wk.rsync.in(ssh = "r820")
        }
        maf <- maftools::read.maf(maf)
        {
            .pdf(maftools::plotmafSummary(maf = maf, rmOutlier = TRUE, 
                addStat = "median", dashboard = TRUE, titvRaw = FALSE), 
                wj = hs.fp(wjj2, "mafSummary"))
            "~/swxx/xm/db/39/5/3/变异瀑布图/mafSummary.pdf" %=>% 
                wk.rsync.in(ssh = "r820")
            .pdf(maftools::oncoplot(maf = maf, top = 10), wj = hs.fp(wjj2, 
                "oncoplot"))
            "~/swxx/xm/db/39/5/3/变异瀑布图/oncoplot.pdf" %=>% 
                wk.rsync.in(ssh = "r820")
        }
        a <- sj[["基因突变"]][, .(Hugo_Symbol = gene, Chromosome = chrom, 
            Start_Position = pos, End_Position = pos + nchar(ref) - 
                1, Reference_Allele = ref, Tumor_Seq_Allele2 = alt, 
            Variant_Classification = mutation_type, "Variant_Type")]
        sj[["基因突变"]][, mutation_type %=>% hs.table %=>% 
            sort]
        `?`(maftools::read.maf)
        {
            laml.maf.wj <- system.file("extdata", "tcga_laml.maf.gz", 
                package = "maftools")
            laml.maf <- wk.fread(laml.maf.wj)
            laml.maf[, Tumor_Sample_Barcode %=>% unique] %=>% 
                cat(sep = "\n")
            laml.maf[, all(Tumor_Seq_Allele1 == Reference_Allele)]
            laml.maf[, Variant_Classification %=>% hs.table %=>% 
                hs.sort]
            laml.maf[, hs.table(Variant_Classification, Variant_Type)]
        }
    }
}
"39_05_21_113523" <- function() {
    wjj2 <- hs.sjgly("39_05_21_113523")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        wk.rsync.out("~/d/迅雷下载/Figure. 2c,d,e.rds", wjj2, 
            ssh = "r820")
    }
    wj1 <- hs.wjj.ls(wjj2, "(?i)\\.rds$")
    a <- hs.load(wj1)
    a@phenoData@data[1, ]
    a@phenoData@data[, "seurat_clusters"] %=>% hs.table
    str(a)
    a %=>% class
    a@phenoData@data %=>% str
}
"39_05_21_102955" <- function(sub = {
}) {
    jg <- .mfp.swxx("39_05_21_102959")
    if (length(sub)) 
        jg <- hs.fp(jg, sub)
    jg
}
"39_05_08_221225" <- function() {
    wjj2 <- hs.sjgly("39_05_08_221225")
    if (0) {
        wk.rsync.out("~/org/工作/生物学/生信/零活/39/05/07/lncrna探针定位/blast/5H26M6Y5013-Alignment.txt.gz", 
            wjj2, ssh = "r820")
        .ls(wjj2)
    }
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        wj1 <- hs.fp(wjj2, "5H26M6Y5013-Alignment.txt.gz")
        if (0) {
            .less(wj1)
        }
        a <- readLines(wj1)
        a <- hs.sepSeq(a, sep.i = hs.grep(a, "^Query #"))
        if (0) {
            a[[3]] %=>% tail(x, -match("Alignments:", x)) %=>% 
                hs.sepSeq(x, sep.i = hs.grep(x, ">"))[-1]
            lb()
            x <- 101
        }
        dt1 <- seq(2, length(a)) %=>% lapply(x, . %=>% {
            hs.msg(x)
            a1 <- a[[x]]
            if ("No significant similarity found." %in% a1) 
                return()
            alignment <- a1 %=>% tail(x, -match("Alignments:", 
                x)) %=>% hs.sepSeq(x, sep.i = hs.grep(x, ">")) %=>% 
                x[sapply(x, . %=>% {
                  substr(x[1], 1, 1) == ">"
                })]
            jg <- lapply(alignment, function(a1) {
                r1 <- hs.sepSeq(a1, sep.i = hs.grep(a1, "^Range \\d+:"))
                jg <- r1 %=>% x[sapply(x, "[", 1) %=>% hs.grep("^Range \\d+:")] %=>% 
                  lapply(x, . %=>% {
                    range <- x[1] %=>% hs.re(":.*") %=>% hs.gre(x, 
                      "\\d+")[[1]] %=>% as.numeric
                    score <- hs.grepV(x, "^Score:") %=>% hs.re("\\d+") %=>% 
                      as.numeric
                    wk.dt(score, start = range[1], end = range[2], 
                      strand = hs.grepV(x, "^Identities:") %=>% 
                        hs.re("Strand:.*") %=>% trim %=>% hs.re("[^ ]+$"))
                  }) %=>% rbindlist
                a <- a1 %=>% hs.grepV("^Sequence ID:") %=>% trim
                jg[, `:=`("seq_len", hs.re(a, "\\d+$") %=>% as.numeric)]
                jg[, `:=`("seq_id", hs.sub(a, "Length: *\\d+$") %=>% 
                  trim %=>% hs.re("[^ ]*$"))]
                jg[, `:=`("seq_title", substring(a1[1], 2))]
            })
            jg %<=>% rbindlist
            jg[, `:=`("query_id", a1[1] %=>% hs.sub("^Query #\\d+: ") %=>% 
                hs.re("[^ ]+"))]
            wk.dt.col.reorder(jg, c("query_id", "seq_title", 
                "seq_id", "seq_len"), before = 1)
        })
        dt1 %<=>% rbindlist
        dt1_best <- dt1[, .SD[score %=>% which(x == max(x))], 
            by = query_id]
        if (0) {
            dt1_best[, {
                if (!any(hs.grepl(seq_id, "(?i)^nc|nm|xm|nr"))) {
                  .SD
                }
            }, by = query_id]
            dt1_best[, any(hs.grepl(seq_id, "(?i)^nc|nm|xm|nr")), 
                by = query_id][, hs.table(V1)]
        }
        dt1_best %<=>% x[hs.grepl(seq_id, "(?i)^nc|nm|xm|nr"), 
            {
                a <- .SD[score %=>% {
                  x == max(x)
                }]
                i <- !hs.grepl(seq_id, "(?i)^nc")
                if (any(i)) {
                  a[i]
                }
                else {
                  if (!all(hs.grepl(seq_id, "(?i)^nc"))) 
                    hs.browser("39.05.08.220638", debug = 0)
                  a
                }
            }, by = query_id]
        if (0) {
            dt1_best[, any(hs.grepl(seq_id, "(?i)^nm|xm|nr")), 
                by = query_id][, hs.table(V1)]
            dt1_best[hs.grepl(seq_id, "(?i)^nc")]
            dt1_best[, all(end > start)]
        }
        {
            gtf <- "~/swxx/sj/db/ensembl/ftp.ensembl.org/pub/release-106/gtf/homo_sapiens/Homo_sapiens.GRCh38.106.gtf.gz" %=>% 
                wk.fread
            ri <- dt1_best[, hs.grep(seq_id, "(?i)^nc")]
            a <- dt1_best[ri, GenomicRanges::GRanges(seqnames = seq_title %=>% 
                hs.re("(?<=chromosome )[^,]+") %=>% trim, ranges = IRanges::IRanges(start = start, 
                end = end), query_id = query_id)]
            gtf_exon <- gtf[V3 %in% "exon"]
            gtf_exon[, `:=`("gene_id_ens", V9 %=>% hs.re("(?<=gene_id \")[^\"]+"))]
            b <- gtf_exon[, GenomicRanges::GRanges(seqnames = V1, 
                ranges = IRanges::IRanges(start = V4, end = V5), 
                gene_id_ens = gene_id_ens, V9 = V9)]
            ovlp <- GenomicRanges::findOverlaps(a, b)
            ovlp %<=>% as.data.table
            for (qi1 in ovlp[, queryHits %=>% unique]) {
                si <- ovlp[queryHits %in% qi1, subjectHits]
                wk.dt.set(dt1_best, ri[qi1], "gene_id_ens", b[si] %=>% 
                  GenomicRanges::mcols(x)[, "gene_id_ens"] %=>% 
                  paste(collapse = ","))
                b1 <- b[si] %=>% GenomicRanges::mcols(x)[, "V9"]
                gene_id_symbol <- b1 %=>% {
                  i <- hs.grepl(x, "gene_name \"")
                  ifelse(i, hs.re(x[i], "(?<=gene_name \")[^\"]+"), 
                    "") %=>% paste(collapse = ",")
                }
                wk.dt.set(dt1_best, ri[qi1], "gene_id_symbol", 
                  gene_id_symbol)
            }
            ri <- dt1_best[, hs.grep(seq_id, "(?i)^nc") %and% 
                which(is.na(gene_id_ens))]
            a <- dt1_best[ri, GenomicRanges::GRanges(seqnames = seq_title %=>% 
                hs.re("(?<=chromosome )[^,]+") %=>% trim, ranges = IRanges::IRanges(start = start, 
                end = end), query_id = query_id)]
            gtf_transcript <- gtf[V3 %in% "transcript"]
            gtf_transcript[, `:=`("gene_id_ens", V9 %=>% hs.re("(?<=gene_id \")[^\"]+"))]
            b <- gtf_transcript[, GenomicRanges::GRanges(seqnames = V1, 
                ranges = IRanges::IRanges(start = V4, end = V5), 
                gene_id_ens = gene_id_ens, V9 = V9)]
            ovlp <- GenomicRanges::findOverlaps(a, b)
            ovlp %<=>% wk.dt.as
            for (qi1 in ovlp[, queryHits %=>% unique]) {
                si <- ovlp[queryHits %in% qi1, subjectHits]
                wk.dt.set(dt1_best, ri[qi1], "gene_id_ens", b[si] %=>% 
                  GenomicRanges::mcols(x)[, "gene_id_ens"] %=>% 
                  paste(collapse = ","))
                b1 <- b[si] %=>% GenomicRanges::mcols(x)[, "V9"]
                gene_id_symbol <- b1 %=>% {
                  i <- hs.grepl(x, "gene_name \"")
                  ifelse(i, hs.re(x[i], "(?<=gene_name \")[^\"]+"), 
                    "") %=>% paste(collapse = ",")
                }
                wk.dt.set(dt1_best, ri[qi1], "gene_id_symbol", 
                  gene_id_symbol)
            }
            dt1_best[hs.grep(seq_id, "(?i)^nc"), hs.table(is.na(gene_id_ens))]
            {
                refseq <- hs.hsgly("ncbi/基因/refseq/39_05_25_122220")() %=>% 
                  wk.fread(cn = 1)
                refseq[, `:=`(RNA_nucleotide_accession, hs.sub(RNA_nucleotide_accession.version, 
                  "\\.\\d+$"))]
                a <- dt1_best[, hs.sub(seq_id, "\\.\\d+$")]
                i <- which((a %in% refseq[, RNA_nucleotide_accession]) & 
                  (dt1_best[, hs.clean(gene_id_symbol, ot = "l", 
                    inv = 1)]))
                wk.dt.set(dt1_best, i, "gene_id_symbol", refseq[match(a[i], 
                  RNA_nucleotide_accession), Symbol])
                dt1_best
            }
            wk.dt.fwrite(dt1_best, wjj2 %=>% hs.fp("结果.tsv.gz"))
        }
        {
            "~/swxx/xm/db/39/05/07/lncrna探针定位/blast/结果.tsv.gz" %=>% 
                wk.rsync.in("~/org/工作/生物学/生信/零活/39/05/07/lncrna探针定位/blast", 
                  ssh = "r820")
            wj1 <- "~/org/工作/生物学/生信/零活/39/05/07/lncrna探针定位/blast/结果.tsv.gz"
            dt1_best <- wj1 %=>% wk.fread
            ri <- dt1_best[, hs.grep(seq_id, "(?i)^nm|xm|nr")]
            a <- local({
                wj_temp <- "~/39_05_09_004820.tsv.gz"
                on.exit(hs.wj.rm(wj_temp), add = TRUE)
                dt1_best[ri, seq_id %=>% hs.sub("\\.\\d+$") %=>% 
                  unique] %=>% cat(sep = "\n", file = .sys("| perl -e", 
                  .q("\nwhile( <>) {\n  chomp;\n  $transcript{ $_} = 1;\n}\nopen( FH, \"pigz -cd ~/Downloadsm/gene2refseq.gz |\");\nwhile( <FH>) {\n  if( /^9606\t/) {\n    $do_line = 1;\n  } else {\n    if( $do_line) {\n      goto DO_EXIT;\n    } else {\n      next;\n    }\n  }\n  @F = split( \"\t\");\n  $F[ 3] =~ s/\\.\\d+\\s*$//;\n  if( exists( $transcript{ $F[ 3]})) {\n    print;\n    #delete $transcript{ $F[ 3]};\n  }\n}\nDO_EXIT: do{\n  print( \"39_05_09_003233\n\");\n  close( FH);\n}\n"), 
                  "| pigz >", hs.fpGood(wj_temp)))
                a <- wk.fread(wj_temp, cn = 0)
                cn <- "~/Downloads/gene2refseq.gz" %=>% readLines(1) %=>% 
                  hs.strsplit.v("\t")
                wk.dt.setnames(a, cn)
            })
            a[, `:=`("RNA_nucleotide_accession", RNA_nucleotide_accession.version %=>% 
                hs.sub("\\.\\d+$"))]
            t_g <- a[, .(Symbol = Symbol %=>% unique %=>% sort %=>% 
                paste(collapse = ",")), by = RNA_nucleotide_accession]
            t1 <- dt1_best[, seq_id %=>% hs.sub("\\.\\d+$")]
            a <- wk.dt.subset.na(t_g, t1)[, Symbol]
            dt1_best[!is.na(a)][seq_id %=>% hs.grep("(?i)^nc")]
            i <- which(!is.na(a))
            wk.dt.set(dt1_best, i, "gene_id_symbol", a[i])
            for (ri1 in dt1_best[, gene_id_ens %=>% hs.grep(",")]) {
                a <- dt1_best[ri1, {
                  a <- list(hs.strsplit.v(gene_id_ens, ","), 
                    hs.strsplit.v(gene_id_symbol, ","))
                  if (identical(unique(a[[2]]), "")) {
                    a[[2]] <- rep("", length(a[[1]]))
                  }
                  if (sapply(a, length) %=>% {
                    uniqueN(x) == 1
                  }) {
                    a <- wk.dt.as(a) %=>% unique
                    wk.dt.setorderv(a, "V1")
                  }
                }]
                if (length(a)) {
                  wk.dt.set(dt1_best, ri1, "gene_id_ens", a[, 
                    V1] %=>% paste(collapse = ","))
                  wk.dt.set(dt1_best, ri1, "gene_id_symbol", 
                    a[, V2] %=>% paste(collapse = ","))
                }
                else {
                  hs.browser("39.05.09.011850", debug = 0)
                  dt1_best[ri1]
                }
            }
            dt1_best %=>% wk.xls.write(hs.fp(wjj2, "结果.xlsx"))
            "/home/wk/swxx/xm/db/39/5/7/lncrna探针定位/blast/结果.xlsx" %=>% 
                wk.rsync.in(ssh = "r820")
            dt1_best[hs.clean(gene_id_symbol, inv = 1, ot = "i"), 
                query_id %=>% uniqueN]
            dt1_best[hs.clean(gene_id_ens, ot = "l") | hs.clean(gene_id_symbol, 
                ot = "l"), query_id %=>% uniqueN]
        }
    }
}
"39_05_07_233010" <- function() {
    wjj2 <- hs.sjgly("39_05_07_233010")
    seq.fa <- hs.hsgly("seq.fa/39_05_07_175106")()
    wj.psl <- hs.hsgly("blat/使用/输出到文件夹.39_05_07_194055")(sequence_wj = seq.fa, 
        wjj2 = wjj2, type_seq = "rna", extra = "-fine")
    cn <- wj.psl %=>% readLines(10)
    cn %<=>% {
        x[3:4] %=>% strsplit("\t") %=>% lapply(trim) %=>% paste0(x[[1]], 
            head(c(x[[2]], rep("", length(x[[1]]))), length(x[[1]])))
    }
    psl <- wk.fread(wj.psl, skip = 5, cn = 0)
    wk.dt.setnames(psl, cn)
    psl_best <- psl[, .SD[match == max(match)], by = Qname]
    if (0) {
        psl_best[, match %=>% hs.table]
        psl_best %=>% {
            q1 <- x[match %=>% {
                x == min(x)
            }, Qname]
            x[Qname %in% q1]
        }
        psl_best[, .N, by = Qname][, N] %=>% hs.table
    }
}
"39_05_07_232313" <- function() {
    wjj2 <- hs.sjgly("39_05_07_232313")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        wk.rsync.out("~/org/工作/生物学/生信/零活/39/05/07/lncrna探针定位/in/GSE111762.xlsx", 
            wjj2, ssh = "r820")
    }
    wjj2 %=>% hs.fp("GSE111762.xlsx")
}
"39_05_07_175106" <- function() {
    wjj2 <- hs.sjgly("39_05_07_175106")
    wj2 <- wjj2 %=>% hs.fp("seq.fa")
    if (!file.exists(wj2)) {
        sj_wj <- hs.hsgly("in/39_05_07_232313")()
        sj <- sj_wj %=>% wk.xls.read("1.5倍")
        sj[, paste0(paste0(">", ID, "\n", SEQUENCE))] %=>% cat(sep = "\n", 
            file = wj2)
    }
    wj2
}
"2023_06_17_171923" <- function() {
    wjj2 <- hs.sjgly("2023_06_17_171923")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
        if (0) {
            wk.rsync.out("/media/sf_D_DRIVE/db/软件库/微信/WeChat Files/sophiewk/FileStorage/File/2023-06/AmexT_v47.FULL.gtf", 
                hs.fp(wjj2, "in"), ssh = "r820")
            wk.rsync.out("/media/sf_D_DRIVE/db/软件库/微信/WeChat Files/sophiewk/FileStorage/File/2023-06/AmexT_v47.FULL.fa", 
                hs.fp(wjj2, "in"), ssh = "r820")
        }
        wj <- hs.fp(wjj2, "in") %=>% hs.wjj.ls %=>% hs.c(hs.fp.ext(x), 
            x)
        fa.gz <- wj["fa"]
        gtf_corrected.fp <- wj["gtf"] %=>% hs.wjm(append = ".corrected")
        wj["gtf"] %=>% .bash
        gtf <- wj["gtf"] %=>% wk.fread
        gtf[1]
        wj["fa"] %=>% .head
        wj["fa"] %=>% .sys.cat.text %=>% .wk("| perl -e", .q("while( <>) { if( /^>/) { print;}}"))
        gtf_corrected.fp <- xxx
    }
}
"2023_08_12_122149" <- function() {
    wjj2 <- hs.sjgly("2023_08_12_122149")
    bam_fp <- hs.hsgly("bam/39_08_12_002835")("bam")
    seurat_fp <- hs.hsgly("bam/39_08_12_002835")("rd")
    seurat <- seurat_fp %=>% hs.load
    seurat %=>% hs.str
    ec_AACCTAGGGTTTCTCCAT
    seurat %=>% x[sapply(x, inherits, "Seurat")]
    seurat %=>% lapply(class)
    seurat %=>% x[names(x)[sapply(x, inherits, "Seurat")] %=>% 
        sort] %=>% sapply(dim)
    seurat %=>% names
    seurat[["w"]] %=>% str
    seurat[["w"]]@meta.data[, "orig.ident"] %=>% hs.table
    bam_fp %=>% lapply(x, . %=>% .sys("samtools view", hs.fpGood(x))) %=>% 
        unlist %=>% paste(collapse = "\n") %=>% .wk("| less -Ni")
    bam_fp[1] %=>% .wk("samtools view", hs.fpGood(x), "| less -Ni")
    bam_fp[1] %=>% .wk("samtools view -H", hs.fpGood(x), "| less -Ni")
    .wk("samtools view --help")
    .wk("dropest -M -V -b -f -L eiEIBA", "-c", "/home/wk/swxx/rj/src/dropEst/configs/drop_seq_velocyto.xml", 
        hs.fpGood(bam_fp[1]))
    "/home/wk/swxx/rj/src/dropEst/configs/seq_well.xml" %=>% 
        .ls1
    hs.hsgly("bam/is_pe.36.06.30.093208")
    getwd()
    hs.wjj.ls(wjj2, wj = 1, wjj = 1)
}
"39_08_12_002835" <- function(what = "all") {
    wjj2 <- hs.sjgly("39_08_12_002835")
    if (0) {
        hs.wjj(wjj2)
        .bash(wjj2)
    }
    hs.switch(what, "bam", hs.wjj.ls(wjj2, "(?i)\\.bam$"), "rd", 
        hs.wjj.ls(wjj2, "(?i)\\.rdata$"), "all", c("bam", "rd") %=>% 
            hs.lapply(x, . %=>% {
                hs.hsgly("数据/39_08_12_002835")(what = x)
            }))
}
"39_03_02_161930" <- function() "~/swxx/xm/db"
"38_06_26_135942" <- function() {
    rj_wjj <- "~/swxx/rj/obsutil"
    dz1 <- "https://obs-community.obs.cn-north-1.myhuaweicloud.com/obsutil/current/obsutil_linux_amd64.tar.gz"
    wj.tar.gz <- hs.wjm(dz1, dn = rj_wjj)
    if (0) {
        wk.wget(dz1, rj_wjj)
        .ls(wj.tar.gz)
        .bash()
        .unzip(wj.tar.gz, l = 0)
        .bash("~/swxx/rj/obsutil/obsutil_linux_amd64/obsutil_linux_amd64_5.3.4/setup.sh")
    }
    hs.wjm(wj.tar.gz, ext = 0, extN = 2) %>% hs.wjj.ls(R = 1)
}
"39_05_07_195309" <- function() {
    hs.hsgly("生物学/根文件夹.37_07_25_182333")() %=>% 
        hs.fp("rj")
}
"36_06_27_111221" <- function(fa_wj, jg_wjj = fa_wj %=>% hs.sub("\\.gz$") %=>% 
    hs.wjm(ext = "mapper_index/bwa"), jg_prefix = basename(fa_wj), 
    bwa.bin = "bwa") {
    if (!file.exists(hs.wjm(jg_prefix, add = "bwt", dn = jg_wjj))) 
        .wk(bwa.bin, "index", "-p", hs.wj(jg_wjj, jg_prefix), 
            fa_wj)
    file.path(jg_wjj, jg_prefix)
}
"36_06_27_112023" <- function(mapper, mapper.bin = hs.switch(mapper, 
    "hisat-3n", hs.hsgly("hisat-3n/路径.38_11_15_115013")(), 
    mapper)) {
    hs.switch(mapper, "bwa", {
        .wk(mapper.bin, "2>&1", "| grep -i -m 1", .q("^version"), 
            intern = 1) %>% hs.sub("Version: ")
    }, "hisat-3n", , .wk(mapper.bin, "--version", intern = 1)[1] %>% 
        hs.re("[^ ]+$"))
}
"38_11_15_115013" <- function(what = "hisat-3n") {
    wjj0 <- "~/swxx/rj/hisat-3n"
    hs.fp(wjj0, what)
}
"38_11_15_155307" <- function(fq_wj, base_change = "C,T", seq_type = "dna", 
    stranded = 0, nt = 8) {
    bin1 <- hs.hsgly("hisat-3n/路径.38_11_15_115013")()
    .wk(.sys.fpGood(bin1), "-q", "-1", fq_wj[1], "-2", fq_wj[2], 
        "-x", hs.hsgly("生信/软件/序列比对/index.36.06.30.205408")(fa_wj, 
            mapper = "hisat-3n", additional = paste("--base-change", 
                base_change)), "--base-change", base_change, 
        "--no-repeat-index", "--repeat", if (seq_type == "dna") 
            "--no-spliced-alignment", if (stranded) 
            "--directional-mapping", "-p", nt, "-mm")
}
"36_06_30_205408" <- function(fa_wj, jg_wjj = hs.wjj(dirname(fa_wj), 
    "index", mapper, hs.hsgly("生信/软件/序列比对/version.36.06.27.112023")(mapper)), 
    jg_prefix = basename(fa_wj), mapper, mapper.bin = hs.switch(mapper, 
        "hisat-3n", {
            hs.hsgly("hisat-3n/路径.38_11_15_115013")("hisat-3n-build")
        }, mapper), additional = hs.switch(mapper, "hisat-3n", 
        "--base-change C,T --repeat-index"), nt = 8) {
    wj_out <- hs.fp(jg_wjj, jg_prefix)
    hs.with_wd(hs.wj(wj_out), {
        if (!length(hs.wjj.ls())) {
            hs.switch(mapper, "bwa", {
                .wk(mapper.bin, "index", "-p", wj_out, fa_wj)
            }, "bowtie2", {
                .wk(paste0(mapper.bin, "-build"), "--threads", 
                  nt, fa_wj, wj_out)
            }, "hisat-3n", {
                fa_wj <- hs.hsgly("基因组/解压到内存.38_11_15_145951")(fa.gz = fa_wj)
                .wk(mapper.bin, additional, "-p", nt, fa_wj, 
                  basename(wj_out))
                if (hs.grepl(fa_wj, "^/dev/shm/")) 
                  hs.wj.rm(fa_wj)
            })
        }
    })
    wj_out
}
"39_05_05_134720" <- function(i, k, e, ...) {
    A <- list(...)
    i1 <- match("path", names(A), nomatch = 0)
    if (i1) 
        hs.update(A[[i1]], paste0(x, "/"), length(x) && (!hs.grepl(x, 
            "/$")))
    bin1 <- "~/swxx/rj/ossutil-linux-amd64/ossutil"
    hs.update(e, paste0("https://", x), !hs.grepl(x, "^http"))
    hs.update(e, hs.sub(x, "(\\.[^\\.]+\\.com)?$", ".aliyuncs.com"), 
        !hs.grepl(x, "\\.com$"))
    .sys(c(bin1, "-i", i, "-k", k, "-e", e, A))
}
"2023_06_01_131342" <- function() {
    "~/rj/linuxnd/lnd"
}
"37_12_19_180840" <- function(version = "hg19") {
    wjj_annovar <- "~/swxx/rj/annovar"
    remote <- "https://www.openbioinformatics.org/annovar/download"
    local <- hs.fp(wjj_annovar, "humandb")
    avdblist_wj <- wj.updater(dz1 = hs.fp(remote, paste0(version, 
        "_avdblist.txt.gz")), wjj2 = local, co1 = 30)
    avdblist <- wk.fread(avdblist_wj)
    avdblist %>% as.data.frame %>% .s
    bn_all <- c("hg19_refGeneWithVer.txt.gz", "hg19_refGeneWithVerMrna.fa.gz", 
        "hg19_refGeneVersion.txt.gz", "hg19_refGene.txt.gz", 
        "hg19_refGeneMrna.fa.gz", "hg19_avsnp150.txt.gz", "hg19_avsnp150.txt.idx.gz", 
        "hg19_cg69.txt.gz", "hg19_cg69.txt.idx.gz", "hg19_caddgt20.txt.gz", 
        "hg19_caddgt20.txt.idx.gz", "hg19_clinvar_20210501.txt.gz", 
        "hg19_clinvar_20210501.txt.idx.gz", "hg19_kaviar_20150923.txt.gz", 
        "hg19_kaviar_20150923.txt.idx.gz", "hg19_hrcr1.txt.gz", 
        "hg19_hrcr1.txt.idx.gz", "hg19_gwava.txt.gz", "hg19_gwava.txt.idx.gz", 
        "hg19_cosmic70.txt.gz", "hg19_cosmic70.txt.idx.gz", "hg19_icgc28.txt.gz", 
        "hg19_icgc28.txt.idx.gz", "hg19_nci60.txt.gz", "hg19_nci60.txt.idx.gz", 
        "hg19_exac03nontcga.txt.gz", "hg19_exac03nontcga.txt.idx.gz", 
        "hg19_dbnsfp42c.txt.gz", "hg19_dbnsfp42c.txt.idx.gz", 
        "hg19_dbscsnv11.txt.gz", "hg19_dbscsnv11.txt.idx.gz", 
        "hg19_ensGeneMrna.fa.gz", "hg19_ensGene.txt.gz", "hg19_revel.txt.gz", 
        "hg19_revel.txt.idx.gz", "hg19_mitimpact24.txt.gz", "hg19_mitimpact24.txt.idx.gz", 
        )
    bn_versioned <- c("hg19_avsnp150.txt.gz", "hg19_avsnp150.txt.idx.gz", 
        "hg19_clinvar_20210501.txt.gz", "hg19_clinvar_20210501.txt.idx.gz", 
        "hg19_kaviar_20150923.txt.gz", "hg19_kaviar_20150923.txt.idx.gz", 
        "hg19_cosmic70.txt.gz", "hg19_cosmic70.txt.idx.gz", "hg19_icgc28.txt.gz", 
        "hg19_icgc28.txt.idx.gz", "hg19_dbnsfp42c.txt.gz", "hg19_dbnsfp42c.txt.idx.gz", 
        "hg19_dbscsnv11.txt.gz", "hg19_dbscsnv11.txt.idx.gz", 
        "hg19_mitimpact24.txt.gz", "hg19_mitimpact24.txt.idx.gz")
    bn_all %>% hs.re("[^\\.]+$") %>% hs.table
    if (0) {
        avdblist[, hs.re(V1, "[^\\.]+$") %>% hs.table]
    }
    bn1_i <- 1
    avdblist
    bn1 <- bn[6]
    bn1 <- bn_versioned[1]
    sapply(bn_versioned, function(bn1) {
        wj1.z <- hs.fp(local, bn1)
        wj1.z.time_remote <- avdblist[match(bn1, V1), V2] %>% 
            {
                as.POSIXct(sub("(.{4})(.{2})", "\\1-\\2-", .))
            }
        wj1 <- hs.wjm(wj1.z, ext = 0)
        if (!file.exists(wj1)) {
            wj.updater(dz1 = hs.fp(remote, bn1), wjj2 = local)
            if (hs.re(bn1, "[^\\.]+$") != "gz") 
                hs.browser("37.12.19.221347")
            hs.with_wd(local, .wk("pigz -d", .q(bn1)))
        }
    })
    bn1 <- (bn %not% bn_versioned)[1]
    sapply(bn_all %not% bn_versioned, function(bn1) {
        wj1.z <- hs.fp(local, bn1)
        wj1.z.time_remote <- avdblist[match(bn1, V1), V2] %>% 
            {
                as.POSIXct(sub("(.{4})(.{2})", "\\1-\\2-", .))
            }
        wj1 <- hs.wjm(wj1.z, ext = 0)
        .ls(c(wj1, wj1.z))
        wj1 %>% {
            if (file.exists(.)) {
                if (wj1.z.time_remote - file.mtime(.) > 24) {
                  hs.browser("37.12.19.231837")
                  hs.wj.rm(.)
                }
            }
        }
        wj1.z %>% {
            if (file.exists(.)) {
                if (wj1.z.time_remote - file.mtime(.) > 24) {
                  hs.wj.rm(.)
                }
            }
        }
        if (!file.exists(wj1)) {
            wj.updater(dz1 = hs.fp(remote, bn1), wjj2 = local)
            if (hs.re(bn1, "[^\\.]+$") != "gz") 
                hs.browser("37.12.19.221347")
            hs.with_wd(local, .wk("pigz -d", .q(bn1)))
        }
    })
}
"32_12_18_094220" <- function(wj.in, wjj.out = {
}, wj.out.qz = hs.wjm(wj.in, ext = 0), wj.log = hs.wjm(wj.in, 
    ext = "av.log"), do.separate = 1, all.separate = 1, rm.zjjg = 0, 
    redo1 = 0, redo2 = 0, protocol.all = "refGene,ensGene,cytoBand,genomicSuperDups,popfreq_all_20150413,gnomad211_exome,gnomad211_genome,avsnp150,dbnsfp35c,dbnsfp33a,mcap13,spidex,revel,regsnpintron,mitimpact24,intervar_20180118,kaviar_20150923,cosmic77,nci60,clinvar_20190211,icgc21", 
    combine = 0, gene.model.separate = 0, compress = 1, only.wj = 0) {
    if (0) {
        t1 <- "/gpfsDATA/home/calvin/rj/annovar/humandb/hg19_dbnsfp35c.txt" %>% 
            fread(nrows = 2)
        t1 <- "/gpfsDATA/home/calvin/rj/annovar/humandb/hg19_dbnsfp33a.txt" %>% 
            fread(nrows = 2)
        names(t1) %>% hs.grep("(?i)cadd|poly", v = 1)
    }
    "refGene,cytoBand,genomicSuperDups,popfreq_all_20150413,gnomad_genome,avsnp147,dbnsfp33a,spidex,cosmic77,nci60,clinvar_20170130,icgc21"
    wjj.36_03_04_003344 <- getwd()
    on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
    setwd(dirname(wj.in))
    wj.in %<>% basename
    wj.out <- hs.wjm(wj.in, ext = "hg19_multianno.txt")
    wj.out.real <- hs.wjm(wj.out, add = if (compress) 
        "gz", dn = if (length(wjj.out)) 
        wjj.out
    else 1)
    if (0) {
        .sys.do(.sys.pigz(), wj.out.real, "|", "wc -l")
        .sh(ieiwk.fread(.sys(.sys.pigz(), wj.out.real, "|", .sys.head(2))))
    }
    if (all.separate) {
        jg.wj <- {
            protocol <- unlist(strsplit(protocol.all, ",")) %rm% 
                ""
            protocol1 <- "gnomad211_genome"
            wk.sapply(unlist(strsplit(protocol.all, ",")) %rm% 
                "", function(protocol1) {
                operation1 <- hs.hsgly("annovar/protocol2operation.33.04.19.171210")(protocol1)
                qz <- hs.wjm(wj.in, ext = protocol1)
                wj.out <- hs.wjm(qz, add = "hg19_multianno.txt")
                wj.out.2 <- hs.wjm(wj.out, add = if (compress) 
                  "gz", dn = if (length(wjj.out)) 
                  wjj.out
                else 1)
                if (only.wj) 
                  return(normalizePath(wj.out.2))
                if (0) {
                  wj.out
                  apropos("copy")
                  .wk("zcat", wj.out.2, ">", wj.out)
                  wj.out %>% .less
                }
                if (!file.exists(wj.out.2)) {
                  .wk("perl", .sys.annovar("table_annovar.pl"), 
                    wj.in, "--outfile", qz, .sys.annovar("humandb"), 
                    "--buildver hg19", "--remove", "--protocol", 
                    protocol1, "--operation", operation1, if (protocol1 %in% 
                      c("refGene", "ensGene")) 
                      if (gene.model.separate) 
                        c("--argument", "\"--separate\""), "--nastring .")
                  skip <- 0
                  if (protocol1 %in% c("gnomad211_genome", "gnomad211_exome")) {
                    cn <- names(wk.fread(cmd = .sys("head", "-n 1", 
                      wj.out), cn = 1))
                    cn[6:length(cn)] %<>% {
                      paste0(protocol1, ".", .)
                    }
                    wk.dt.fwrite(as.list(cn), wj.out.2, append = FALSE)
                    skip <- 1
                  }
                  .wk(if (skip) 
                    c("tail", "-n", paste0("+", skip + 1))
                  else "cat", wj.out, if (compress) 
                    "| gzip", ">>", wj.out.2)
                  wj.out %>% hs.rm.wj
                }
                wj.out.2 %>% normalizePath
            }) %>% unlist
        }
        if (0) {
            jg.wj[1] %>% .ls
            jg.wj[1] %>% hs.wjm(add = "bak") %>% .wk("cat", ., 
                "| gzip -cd | wc -l")
        }
        wj1 <- jg.wj[1]
        if (combine) {
            db1 <- "clinvar_20170130"
            if (file.exists(wj.out.real)) {
                list(wj.out.real, jg.wj)
            }
            else {
                db1 <- jg.wj[1] %>% names
                a <- wk.sapply(names(jg.wj), function(db1) {
                  .ss(db1)
                  wj1 <- jg.wj[db1]
                  hs.hsgly("annovar.multianno.sep.read.33.01.21.212759")(wj1, 
                    db1)
                })
                row1 <- unique(do.call("c", unname(lapply(a, 
                  rownames))))
                r <- matrix(nrow = length(row1), ncol = sum(sapply(a, 
                  ncol)) + 1)
                r[, 1] <- row1
                ci <- 2
                null <- sapply(seq_along(a), function(ai) {
                  .ss(ai)
                  a1 <- a[[ai]]
                  r1 <- as.matrix(a1[row1, , drop = 0])
                  r[, seq(ci, length.out = ncol(a1))] <<- r1
                  ci <<- ci + ncol(a1)
                })
                if (0) {
                  a1 <- c("Chr", "Start", "End", "Ref", "Alt", 
                    unlist(sapply(a, colnames)))
                  .sh(data.frame(seq_along(a1), a1))
                  ic <- grep("(?i)eigen", a1)
                  a1[ic]
                  r[c(49, 50), ic - 4]
                  r[1, ic - 4]
                  r[, ic[2] - 4]
                  r[, ic[4] - 4]
                  which(r[, ic[1]] != r[, ic[2]])
                  grep("[^\\.]", r[seq(99), ic[1]])
                  r[1, ]
                }
                if (rm.zjjg) 
                  unlink(jg.wj)
                {
                  if (0) {
                    ieiwk.write.tsv.gz(t(c("Chr", "Start", "End", 
                      "Ref", "Alt", unlist(sapply(a, colnames)))), 
                      wj.out, col.names = 0, row.names = 0)
                    ieiwk.write.tsv.gz(r, wj.out, col.names = 0, 
                      row.names = 0, append = 1)
                  }
                  r.2 <- lapply(seq.int(ncol(r)), function(i1) r[, 
                    i1])
                  names(r.2) <- c(paste0("Chr", "Start", "End", 
                    "Ref", "Alt", sep = "\t"), unlist(sapply(a, 
                    colnames)))
                  list(ieiwk.fwrite.gz(r.2, wj.out.real, quote = 0), 
                    jg.wj)
                }
            }
        }
        else jg.wj
    }
    else if (0) {
        protocol1 <- "refGene,cytoBand,genomicSuperDups,popfreq_all_20150413,avsnp142,dbnsfp30a,eigen,spidex,mcap,cosmic77,nci60"
        operation1 <- paste(c("g", "r", "r", rep("f", length(unlist(gre(",", 
            protocol1))) - 2)), collapse = ",")
        .wk("perl", .sys.annovar("table_annovar.pl"), .sys.file.fifo(wj.in), 
            "--outfile", wj.out.qz, .sys.annovar("humandb"), 
            "--buildver hg19", "--remove", "--protocol", protocol1, 
            "--operation", operation1, "--nastring .", log = wj.log)
        if (do.separate) {
            for (protocol1 in c("clinvar_20160302", "icgc21")) {
                operation1 <- "f"
                qz <- hs.wjm(wj.in, ext = protocol1)
                .wk("perl", .sys.annovar("table_annovar.pl"), 
                  wj.in, "--outfile", qz, .sys.annovar("humandb"), 
                  "--buildver hg19", "--remove", "--protocol", 
                  protocol1, "--operation", operation1, "--nastring .", 
                  log = hs.wjm(qz, append = "av", add = "log"))
            }
        }
        hs.wjm(wj.in, ext = "hg19_multianno.txt")
    }
}
"33_04_19_171210" <- function(protocol1) {
    sapply(unlist(strsplit(protocol1, ",")), function(p1) {
        switch(p1, refGene = , ensGene = "g", cytoBand = , genomicSuperDups = "r", 
            "f")
    }) %>% paste(collapse = ",")
}
"32_12_20_232837" <- function(wj.in, ext = "avinput", shuchu.zygosity = 0, 
    wjj.out = {
    }, gzip = 0, includeinfo = 1) {
    if (!length(wjj.out)) {
        wjj.out <- dirname(wj.in)
        wj.in %<>% basename
    }
    hs.with_wd(wjj.out, {
        if (gzip) 
            ext <- hs.wjm(ext, add = "gz")
        wj2 <- hs.wjm(wj.in, ext = ext, extN = 3)
        if (wj.update.reporter(wj2, wj.in)) {
            .wk("perl", .sys.annovar("convert2annovar.pl"), "--format", 
                "vcf4old", "--includeinfo", .q(wj.in), if (!includeinfo) 
                  c("|", .sys.cut("-10")), "|", "sort", "|", 
                "uniq", if (gzip) 
                  c("|", .sys.pbgzip()), ">", .q(wj2))
            if (shuchu.zygosity) {
                wj.out.zygosity <- hs.wjm(wj.in, ext = "zygosity.tsv.gz", 
                  extN = 3)
                .wk("perl", .sys.annovar("convert2annovar.pl"), 
                  "--format", "vcf4old", .q(wj.in), "|", "sort", 
                  "|", "uniq", "|", .sys.pbgzip(), ">", .q(wj.out.zygosity))
            }
        }
        normalizePath(wj2)
    })
}
"39_05_07_195240" <- function(genome_version = switch(org, human = "hs1"), 
    org = "human") {
    wjj2 <- hs.sjgly("39_05_07_195240")
    if (genome_version %not.in% c("hs1")) 
        hs.browser("39.05.07.200339", debug = 0)
    hs.wjj(wjj2, genome_version) %=>% hs.with_wd({
        wj <- hs.fp("https://hgdownload.soe.ucsc.edu/goldenPath", 
            genome_version, "bigZips", paste0(genome_version, 
                c(".2bit", ".2bit.bpt")))
        for (wj1 in wj) {
            if (basename(wj1) %=>% file.exists) 
                next
            .wk("wget -c", wj1)
        }
        wj[1] %=>% basename %=>% normalizePath
    })
}
"39_05_07_194055" <- function(sequence, sequence_wj = {
}, wjj2 = ".", wj2 = "out.psl", org = "human", genome_version = switch(org, 
    human = "hs1"), type_db = "dna", type_seq = "dna", type_out = "psl", 
    extra = "", run = 0) {
    wj_bin <- hs.hsgly("blat/文件.39_05_07_222820")()
    if (0) {
        .wk(wj_bin, "| less -Ni")
    }
    wj_2bit <- hs.hsgly("2bit/39_05_07_195240")(genome_version = genome_version)
    hs.wjj(wjj2) %=>% hs.with_wd({
        if (!file.exists(wj2)) {
            if (!length(sequence_wj)) {
                sequence_wj <- "seq.fa"
                cat(sequence, sep = "\n", file = sequence_wj)
            }
            .wk(wj_bin, wj_2bit, sequence_wj, wj2, paste0("-t=", 
                type_db), paste0("-q=", type_seq), paste0("-out=", 
                type_out), extra)
        }
        normalizePath(wj2)
    })
}
"39_05_07_222820" <- function() {
    wj1 <- "~/rj/miniconda3/envs/r/bin/blat"
    if (!file.exists(wj1)) 
        hs.browser("39.05.07.223002", debug = 0)
    wj1
}
"38_12_09_141645" <- function(seurat_data_list, add.cell.ids = names(seurat_data_list)) {
    hs.require("Seurat")
    seurat_data <- seurat_data_list %=>% hs.switch(length(x), 
        1, x[[1]], SeuratObject:::merge.Seurat(x[[1]], x[-1], 
            add.cell.ids = add.cell.ids))
    for (cn1 in colnames(seurat_data@meta.data)) {
        a <- seurat_data@meta.data[, cn1]
        if (is(a, "character")) 
            seurat_data@meta.data[, cn1] <- factor(a, hs.str_sort(unique(a)))
    }
    seurat_data
}
"38_01_16_235954" <- function(only_one = 1, ..mem = 1) {
    a1 <- Seurat::cc.genes
    a1 <- lapply(a1, hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937"))
    if (only_one) {
        i <- match(c("JPT1;NOTCH1;MTRNR2L1", "JPT1;NOTCH1"), 
            a1[["g2m.genes"]]) %>% hs.clean
        if (length(i)) 
            a1[["g2m.genes"]][i] <- "JPT1"
        if (lapply(a1, hs.grepV, ";") %>% hs.nonempty %>% length) {
            hs.browser("38.01.17.000343")
        }
    }
    a1
}
"38_12_09_102041" <- function(seurat_data, wj2 = "a.h5ad") {
    if (!file.exists(wj2)) {
        hs.require(c("Seurat", "SeuratDisk"))
        SaveH5Seurat(pbmc3k.final, filename = "pbmc3k.h5Seurat")
    }
    wj2
}
"36_06_26_232616" <- function(bam_wj, nt = 4) {
    for (wj1 in bam_wj) {
        .cat.time()
        hs.say("【", wj1, "】")
        .wk("samtools index", "-@", nt, wj1)
        .cat.time()
    }
}
"32_12_21_000647" <- function(wj.bam, wjj.jg = hs.dirname(wj.bam[1], 
    ifelse(length(wj.bam) == 1, 1, 2)), qz = ifelse(length(wj.bam) == 
    1, hs.wjm(wj.bam, ext = 0), "allFam"), n1 = Sys.info()["nodename"], 
    n.nodes = 1, w.which = seq_along(w), q = 1, var = 1, sv = 1, 
    x = .sys.speedseq("ceph18.b37.lumpy.exclude.2014-01-15.bed"), 
    xcs = hs.NC(), wj.ref = .sys.speedseq("wj.ref"), bg = 0) {
    wjj.36_03_04_003344 <- getwd()
    on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
    same_wjj <- normalizePath(dirname(wj.bam[1])) == normalizePath(wjj.jg)
    wj.bam %<>% normalizePath
    hs.wjj(wjj.jg, go = 1)
    if (same_wjj) {
    }
    else {
        a <- c(wj.bam, hs.wjm(wj.bam, add = "bai"))
        if (any(file.exists(file.path(wjj.jg, basename(a))))) 
            stop("37_01_11_192844 已经有文件存在")
        file.symlink(a, ".")
        on.exit(hs.rm.wj(file.path(wjj.jg, basename(a))), add = TRUE)
    }
    wj.bam %<>% basename
    bam.head <- wk.fread(cmd = .sys("samtools", "view", "-H", 
        wj.bam[1], "|", .sys.perl.print("print if $F[ 0] eq \"\\@SQ\"")), 
        cn = 0)
    chr <- bam.head[V1 == "@SQ", hs.grepl(V2, "(?i)chr\\d") %>% 
        any]
    MT <- bam.head[V1 == "@SQ", any(hs.grepl(V2, "(?i):(chr)?mt$"))]
    W <- hs.hsgly("speedseq.var.-w_chrxxx.multi_nodes.32.12.26.211911")(n.nodes, 
        chr = chr, mt = MT)
    jg <- {
    }
    if (var) {
        wj2 <- paste0(qz, ".vcf.gz")
        jg %<>% c(wj2)
        if (!file.exists(wj2)) {
            if (length(W) < 2) {
                if (!nzchar(n1)) 
                  n1 <- give.me.a.node(10, "speedseq")
                if (!nzchar(xcs)) 
                  xcs <- 8
                cmd <- .sys(.sys.speedseq("wj.speedseq"), "var", 
                  "-o", qz, if (length(W)) 
                    c("-w", W), "-q", q, "-t", xcs, wj.ref, wj.bam)
                .wk(cmd)
                if (0) 
                  wk.e(bquote({
                    .wk(.(cmd), log = .(hs.wjm(qz, add = "var.log")))
                  }), n1, bg = bg)
            }
            else {
                hs.browser("35.09.25.090948")
                n1 <- give.me.a.node(10, "ceph18.b37.include.2014-01-15")
                for (i1 in w.which) {
                  w1 <- W[i1]
                  qz1 <- hs.wjm(qz, add = sprintf("part%s", i1))
                  cmd <- .sys(.sys.speedseq("wj.speedseq"), "var", 
                    "-o", qz1, "-w", w1, "-q", q, "-t", xcs, 
                    wj.ref, wj.bam)
                  .wk(cmd)
                  if (0) 
                    ieiwk.eval(bquote({
                      .wk(.(cmd), log = .(hs.wjm(qz1, add = "var.log")))
                    }), n1, bg = bg)
                }
            }
        }
    }
    if (sv) {
        cmd <- .sys(.sys.speedseq("wj.speedseq"), "sv", "-o", 
            qz, if (nchar(x)) 
                c("-x", .sys.speedseq("ceph18.b37.lumpy.exclude.2014-01-15.bed")), 
            "-B", paste(wj.bam, collapse = ","), "-D", paste(hs.wjm(wj.bam, 
                append = ".discordants"), collapse = ","), "-S", 
            paste(hs.wjm(wj.bam, append = ".splitters"), collapse = ","), 
            "-g", "-d", "-R", wj.ref, "-t", xcs)
        hs.browser("37.01.11.211606")
        .wk(cmd)
        if (0) 
            wk.e(bquote({
                .wk(.(cmd), log = hs.wjm(.(qz), add = "sv.log"))
            }), n1, bg = bg)
    }
    normalizePath(jg)
}
"36_12_30_102332" <- function(wj1, wjj2 = dirname(wj1), gz = 1) {
    wj2 <- hs.wjm(wj1, dn = wjj2, ext = hs.pastec0("fq", if (gz) 
        ".gz"))
    paired <- hs.hsgly("bam/is_pe.36.06.30.093208")(wj1)
    if (paired) 
        wj2 <- hs.wjm(wj2, append = paste0(".", c("r1", "r2")), 
            extN = 2)
    .wk("samtools fastq -@ 4", if (paired) {
        if (gz) {
            c("-1", ">( paste - - - - | sort -k1,1 -S 3G | tr", 
                .q("\t"), .q("\n"), "| gzip >", wj2[1], ")", 
                "-2", ">( paste - - - - | sort -k1,1 -S 3G | tr", 
                .q("\t"), .q("\n"), "| gzip >", wj2[2], ")")
        }
        else {
            c("-1", wj2[1], "-2", wj2[2])
        }
    }
    else {
        c("-0", wj2)
    }, wj1)
}
"36_05_05_223906" <- function(bam.wj) {
    a1 <- .wk("samtools view", bam.wj[1], "| cut -f10 | head -n 100000", 
        intern = 1)
    max(nchar(a1))
}
"38_01_12_221325" <- function(bam_wj, bed_wj, mq = 20) {
    wj2 <- hs.wjm(bam_wj, add = "depth.gz")
    if (!file.exists(wj2)) 
        hs.with_wd(dirname(wj2), {
            hs.browser("38.01.12.224305")
            .wk("samtools", "depth", "-a", "-Q", mq, c("-b", 
                .sys.fpGood(bed_wj)), .sys.fpGood(bam_wj), "| gzip >", 
                .q(basename(wj2)))
        })
    wj2
}
"37_11_28_141923" <- function(bam_wj, nt = 4, ..., with_bam = 1) {
    wj2 <- hs.wjm(bam_wj, ext = "bai")
    if (file.exists(wj2)) {
        if (wj.update.reporter(wj2, bam_wj)) {
            .wk("samtools index", "-@", nt, hs.wjm(bam_wj, ext = 0), 
                hs.wjm(.sys.fpGood(bam_wj), ext = 0))
        }
        return(wj2)
    }
    wj2 <- hs.wjm(bam_wj, add = "bai")
    if (wj.update.reporter(wj2, bam_wj)) {
        .wk("samtools index", "-@", nt, .sys.fpGood(bam_wj))
    }
    c(if (with_bam) bam_wj, wj2)
}
"36_07_16_215544" <- function(bam_wj, dedup = 0) {
    bw_wj <- switch(as.character(length(bam_wj)), `1` = hs.wjm(bam_wj[1], 
        ext = "bw"), `2` = hs.wjm(bam_wj[1], ext = "bw", append = paste0("..-", 
        hs.wjm(bam_wj[2], dn = 0, ext = 0))), {
        .r("36.07.16.220039")
        hs.browser("36.07.16.220040")
    })
    if (dedup) 
        bw_wj %<>% hs.wjm(append = "..dedup")
    log_wj <- hs.wjm(bw_wj, add = "log")
    if (file.exists(bw_wj)) {
    }
    else .wk(switch(as.character(length(bam_wj)), `1` = {
        hs.cat("【bamCoverage】", bam_wj)
        c("bamCoverage", "-b", bam_wj)
    }, `2` = {
        hs.cat("【bamCompare】", bam_wj)
        hs.cat("\t", bam_wj[1])
        hs.cat("\t", bam_wj[2])
        c("bamCompare", "-b1", bam_wj[1], "-b2", bam_wj[2], "--scaleFactorsMethod None")
    }), "-o", bw_wj, "--binSize 20", "--smoothLength 60", "--normalizeUsing BPM", 
        "--centerReads", if (dedup) 
            "--ignoreDuplicates", if (hs.hsgly("bam/is_pe.36.06.30.093208")(bam_wj[1])) {
            "--samFlagInclude 64"
        }
        else {
            .r("36.07.16.213519")
            hs.browser("36.07.16.213508")
        }, "-p", hs.NC(), "2>", log_wj)
    bw_wj
}
"36_07_16_212356" <- function(bam_wj) {
    a1 <- .wk("bamPEFragmentSize", "-b", wj1, "-p", nc2, intern = 1)
    hs.re(hs.grepV(a1, "^Median.*\\d+")[1], "\\d+")
}
"36_06_30_093208" <- function(bam_wj) {
    if (0) {
        .wk("samtools", "view", .sys.fpGood(wj_bam[1]), "| head -n 1000", 
            intern = 1) %>% strsplit("\t") %>% sapply("[", 8) %>% 
            as.numeric %>% any
    }
    !identical(.wk("samtools view", .sys.fpGood(bam_wj), .pipe(), 
        "head -n 100000", "| cut -f7", "| uniq", intern = 1), 
        "*")
}
"37_12_12_082631" <- function(bed_fp) {
    wj2 <- hs.wjm(bed_fp, append = ".4gatk")
    if (wj.update.reporter(wj2, bed_fp)) {
        cmd_convert <- .sys("cat_wk", .sys.fpGood(bed_fp), "| perl -e", 
            .q("\nwhile( <>) {\n  if( /^chrmt?\t/i) {\n    print( \"MT\t\", $')\n  } elsif( /^chr/i) {\n     print( $')\n  }\n}"))
        if (0) {
            cmd_1st_column_uniq <- .sys("perl -e", .q("while( <>) { /[^\\s]*/; print( $&, \"\n\")}"), 
                "| uniq -c | less -Ni")
            .wk(cmd_convert, "|", cmd_1st_column_uniq)
        }
        .wk(cmd_convert, ">", .sys.fpGood(wj2))
    }
    wj2
}
"37_12_12_102937" <- function(fa.wj) {
    sapply(fa.wj, function(fa.wj) {
        Z <- hs.grepl(fa.wj, "(?i)z$")
        wj2 <- hs.wjm(fa.wj, ext = "dict", extN = Z + 1)
        if (!file.exists(wj2)) {
            wj2 <- if (Z) 
                hs.wjm(fa.wj, ext = "dict")
            else hs.wjm(fa.wj, add = "dict")
        }
        if (wj.update.reporter(wj2, fa.wj)) 
            hs.with_wd(dirname(fa.wj), .wk("samtools dict", .q(basename(fa.wj)), 
                ">", .q(basename(wj2))))
        wj2
    })
}
"37_04_14_182739" <- function(fa.wj) {
    sapply(fa.wj, function(fa.wj) {
        Z <- hs.grepl(fa.wj, "(?i)z$")
        fai.wj <- paste0(fa.wj, ".", if (Z) 
            "gzi"
        else "fai")
        if (wj.update.reporter(fai.wj, fa.wj)) 
            hs.with_wd(dirname(fa.wj), .wk("samtools faidx", 
                .q(basename(fa.wj))))
        fai.wj
    })
}
"39_04_10_100139" <- function(fp, pattern1 = "(?i)([_\\.]r?[12])?\\.f(ast)?q(\\.gz)?$") {
    mid <- basename(fp) %=>% hs.sub(pattern1) %=>% hs.factor(x, 
        unique(x), N = 0) %=>% tapply(x, x, length) %=>% rep(paste0("SMPL_", 
        hs.str_pad(seq_along(x), pad = "0")), x)
    wjm <- paste0(mid, basename(fp) %=>% hs.re(pattern1))
    hs.fp(dirname(fp), wjm)
}
"38_01_12_213550" <- function(fq_wj) {
    paired <- hs.hsgly("isPaired.35.08.18.232544")(fq_wj)
    paired && {
        read2 <- readLines(fq_wj[1], 5)
        hs.browser("38.01.12.214013")
        read2[1] == read2[2]
    }
}
"38_01_12_212153" <- function(fq_wj) {
    interleaved <- hs.hsgly("是否交织.38_01_12_213550")(fq_wj)
    bs <- ifelse(interleaved, 8, 4)
    paired <- hs.hsgly("isPaired.35.08.18.232544")(fq_wj)
    if (paired) {
        fq_wj <- fq_wj[seq(from = 1, to = length(fq_wj), by = 2)]
    }
    ln <- wk.mclapply(fq_wj, .wc)
    hs.unlist(ln)/bs
}
"37_12_07_144326" <- local({
    hs1 <- function(fq, wjj2 = hs.fp(dirname(fq[1]), "fastqc")) {
        wj_st <- paste0(fq, ".st")
        if (file.exists(wj_st)) 
            stop("[38_04_21_173421]文件不完整")
        wj_html <- hs.wjm(fq, append = "_fastqc", ext = "html", 
            extN = 1 + hs.grepl(fq, "\\.gz$"), dn = wjj2)
        if (!file.exists(wj_html)) 
            .wk("fastqc", "-o", .sys.fpGood(hs.wjj(wjj2)), "-t", 
                1, .sys.fpGood(fq))
        normalizePath(wjj2)
    }
    function(fq, nc = Ncpu()[2]/4, ...) {
        if (!length(.wk("type fastqc", intern = 1))) 
            stop("[38_04_21_172933] 缺fastqc。sudo apt install fastqc")
        a <- wk.mclapply(fq, hs1, ..., mc.cores = nc)
        hs.c(fq, a)
    }
})
"37_12_25_144530" <- function(fq_wj) {
    tapply(fq_wj, hs.hsgly("wjm2ybm.35.08.20.200110")(fq_wj), 
        c)
}
"35_08_18_232544" <- function(wj) {
    (length(wj) > 1) || local({
        read <- readLines(wj[1], 5)[c(1, 5)]
        uniqueN(hs.re(read, ".*\\d+:\\d+:\\d+\\s")) == 1
    })
}
"37_12_08_165040" <- local({
    hs1 <- function(fq, wjj2 = hs.fp(dirname(fq[1]), "fastqc"), 
        nc = 10) {
        wj_st <- paste0(fq, ".st")
        if (file.exists(wj_st)) 
            return("")
        wj_html <- hs.wjm(fq, append = "_fastqc", ext = "html", 
            extN = 1 + hs.grepl(fq, "\\.gz$"), dn = wjj2)
        if (!file.exists(wj_html)) {
            .wk("fastqc", "-o", .sys.fpGood(hs.wjj(wjj2)), "-t", 
                nc, .sys.fpGood(fq))
        }
        normalizePath(wjj2)
    }
    function(wj, wjj2 = ".", args_extra = "", help = 0) {
        if (help) {
            .wk("multiqc --help | less -Nir")
            return()
        }
        hs.with_wd(wjj2, {
            if (!file.exists("multiqc_report.html")) 
                .wk("multiqc", args_extra, .sys.fpGood(wj))
        })
    }
})
"38_04_04_215853" <- function(fq.wj) {
    wk.mclapply(seq_along(fq.wj), function(i) {
        wj1 <- fq.wj[i]
        wj2 <- hs.wjm(wj1, append = paste0(c(1, 2)), extN = 2)
        if (wj.update.reporter(wj2, wj1)) {
            read_size <- as.integer(hs.re(readLines(wj1, 1), 
                "\\d+$"))/2
            con1 <- pipe(.sys("cat_wk", wj1), "r")
            hs.msg(wj1)
            lines_read <- 0
            repeat {
                chunk <- readLines(con1, 4e+06)
                if (!length(chunk)) 
                  break
                lines_read <- lines_read + length(chunk)
                i1 <- seq(1, length(chunk), by = 4)
                {
                  dt <- wk.dt(V1 = hs.sub(chunk[i1], "(\\.\\d+) \\d+.*", 
                    paste0("\\1.1 1")), V2 = substr(chunk[i1 + 
                    1], 1, read_size))
                  dt[, `:=`(V3, "+")]
                  dt[, `:=`(V4, substr(chunk[i1 + 3], 1, read_size))]
                  wk.dt.fwrite(t(dt) %>% as.vector, wj2[1])
                }
                {
                  dt[, `:=`(V1, hs.sub(V1, "\\.1 ", ".2 "))]
                  dt[, `:=`(V2, substring(chunk[i1 + 1], read_size + 
                    1))]
                  dt[, `:=`(V4, substring(chunk[i1 + 3], read_size + 
                    1))]
                  wk.dt.fwrite(t(dt) %>% as.vector, wj2[2])
                }
                hs.msg(paste0("处理了", lines_read, "行"), 
                  0)
            }
            close(con1)
        }
    })
}
"37_12_25_144226" <- function(wjj1, F = 1, R = 0, sample_wise = !wjj_wise, 
    wjj_wise = 0, ...) {
    wj <- hs.wjj.ls(wjj1, "\\.f(ast)?q(\\.gz)?$", F = F, R = R)
    if (length(wj)) {
        if (sample_wise) {
            wj <- tapply(wj, hs.hsgly("wjm2ybm.35.08.20.200110")(wj), 
                c)
        }
        else if (wjj_wise) {
            wj <- tapply(wj, basename(dirname(wj)), c)
        }
    }
    wj
}
"35_08_20_200110" <- function(wj) {
    hs.sub(basename(wj), "\\.f(ast)?q(\\.gz)?$") %>% sapply(revStr.int) %>% 
        hs.re("^\\d+[_\\.\\-]*", after = 1) %>% hs.re("^(?i)\\d+[ri][_\\.\\-]*", 
        after = 1) %>% hs.re("^(?i)\\d+l[_\\.\\-]*", after = 1) %>% 
        hs.re("^(?i)\\d+s[_\\.\\-]*", after = 1) %>% sapply(revStr.int)
}
"38_09_08_204013" <- function(gtf.fp, ..mem = 1) {
    if (0) {
        cmd1 <- .sys.cat.text(gtf.fp) %=>% .sys("| perl -e", 
            .q("\nwhile( <>) {\n  @F = split( \"\t\");\n  if( $F[ 2] eq \"gene\") {\n    print;\n  }\n}\n"))
        gtf <- wk.fread(cmd = cmd1)
    }
    gtf <- hs.hsgly("1.信息学/文件/读写/提取/表格的子集.38_09_10_214239")("gene", 
        gtf.fp, ci1 = 3)
    for (cn1 in paste0("gene_", c("id", "name", "biotype"))) {
        message("[38_09_08_204542] 提取", cn1)
        gtf[, `:=`(c(cn1), hs.re(V9, paste0("(?<=", cn1, " \")[^\"]*"), 
            no_match = ""))]
    }
    gtf[, `:=`("gene", ifelse(hs.clean(gene_name, ot = "l"), 
        gene_name, gene_id))]
    gtf[, `:=`("i.38_08_20_024125", seq(.N))]
    a <- gtf[, if (.N > 1) 
        .(i.38_08_20_024125, paste0(gene, ".", seq(.N))), by = gene]
    if (!hs.is_empty(a)) 
        wk.dt.set(gtf, a[, i.38_08_20_024125], "gene", a[, V2])
    gtf[, `:=`("i.38_08_20_024125", {
    })]
    gtf
}
"37_12_14_222827" <- function(vcf_wj, wjj2 = ".", rm = 0) {
    sapply(vcf_wj, function(vcf_wj1) {
        if (hs.grepl(vcf_wj1, "z$")) {
            hs.browser("37.12.14.230529")
            if (normalizePath(wjj2) == getwd()) {
                return(vcf_wj1)
            }
            else {
                (if (rm) 
                  hs.wj.move
                else hs.wj.copy)(vcf_wj1, wjj2)
                return(hs.wjm(vcf_wj1, dn = wjj2))
            }
        }
        else {
            wj2 <- hs.wjm(vcf_wj1, dn = wjj2, add = "gz")
            if (wj.update.reporter(wj2, vcf_wj1)) {
                hs.browser("37.12.14.225650")
                .wk("bcftools convert", .sys.fpGood(vcf_wj1), 
                  "-Oz", "-o", .sys.fpGood(wj2))
                .wk("bcftools index", .sys.fpGood(wj2), "-t")
                if (rm) 
                  hs.wj.rm(vcf_wj1)
            }
            return(wj2)
        }
    })
}
"2023_06_14_140936" <- function(sub = {
}) {
    jg <- hs.hsgly("根文件夹.37_07_25_182333")() %=>% hs.fp("rj")
    if (length(sub)) 
        jg <- hs.fp(jg, sub)
    jg
}
"2023_06_14_140548" <- function() {
    wjj2 <- hs.sjgly("2023_06_14_140548")
    arg.c <- hs.fp(wjj2, "config.yaml")
    if (!file.exists(arg.c)) {
        .wk("bulker", "init", "-c", hs.fpGood(arg.c))
        a <- arg.c %=>% readLines
        map <- list(c("default_crate_folder", wjj2 %=>% hs.sub("~", 
            "$HOME") %=>% hs.fp("crate")), c("singularity_image_folder", 
            wjj2 %=>% hs.sub("~", "$HOME") %=>% hs.fp("singularity_image")))
        for (map1_i in seq_along(map)) {
            map1 <- map[[map1_i]]
            i <- hs.grep(a, paste0("^\\s*", map1[1], ":"))
            if (length(i) != 1) 
                hs.browser("2023_06_14_143844", debug = 0)
            a[i] %<=>% hs.sub(paste0("(?<=", map1[1], ": ).*"), 
                map1[2])
        }
        cat(a, sep = "\n", file = arg.c)
    }
    crate1 <- "databio/pepatac:1.0.9"
    .wk("bulker load", "-c", hs.fpGood(arg.c), crate1)
    .wk("bulker activate", "-c", hs.fpGood(arg.c), crate1)
    arg.c %=>% .less
}
"2023_06_01_155232" <- function(what = "singularity") {
    hs.switch(what, "singularity", "~/swxx/rj/lib/singularity_for_nf-core")
}
"2023_06_05_111615" <- function() {
    .wk("nf-core")
    a <- hs.hsgly("文件.2023_05_31_135733")(what = "atac")
    b <- lapply(seq_along(a), function(i1) {
        a1 <- a[i1]
        sn1 <- basename(a1)
        fp <- hs.hsgly("wjj2fq.37_12_25_144226")(a1)
        c(hs.sub(sn1, "-\\d*$"), normalizePath(fp), hs.re(sn1, 
            "\\d*$"))
    })
    b <- wk.dt.fromList(b, by_row = 1)
    wk.dt.setnames(b, c("sample", "fastq_1", "fastq_2", "replicate"))
    wd1 <- "~/swxx/2023_06_05_142906"
    si_fp <- hs.fp(wd1, "si.csv")
    wk.dt.fwrite(b, si_fp, sep = ",")
    t2t_wjj <- "~/swxx/t2t"
    if (0) {
        t2t_wjj %=>% .ls
        t2t_wjj %=>% .bash
        hs.fp(t2t_wjj, "chm13v2.0_RefSeq_Liftoff_v5.gff3") %=>% 
            .head
        hs.fp(t2t_wjj, "Homo_sapiens-GCA_009914755.4-softmasked.fa.gz") %=>% 
            .head
    }
    {
        {
            t2t_fa_wj <- hs.fp(t2t_wjj, "chm13v2.0_maskedY_rCRS.fa")
            t2t_gtf_wj <- hs.fp(t2t_wjj, "GCF_009914755.1_T2T-CHM13v2.0_genomic.gtf.gz")
            t2t_gff_wj <- hs.fp(t2t_wjj, "chm13v2.0_RefSeq_Liftoff_v5.gff.gz")
            t2t_fa_wj %=>% .head
            t2t_gtf_wj %=>% .head
        }
        if (0) {
            t2t_fa_wj <- hs.fp(t2t_wjj, "Homo_sapiens-GCA_009914755.4-softmasked.fa.gz")
            t2t_gtf_wj <- hs.fp(t2t_wjj, "Homo_sapiens-GCA_009914755.4-2022_07-genes.gtf.gz")
        }
    }
    if (0) {
        .wk("nextflow", "run", .q(normalizePath("~/swxx/rj/nf-core/pipeline/atacseq/2.0/workflow")), 
            "--show_hidden_params true", "--help")
        a <- hs.wjj.ls("/home/wk/swxx/2023_06_05_142906", "\\.bam$", 
            R = 1)
        a <- a[order(basename(a))]
    }
    wd1 %=>% .ls
    hs.with_wd(wd1, .wk("export", paste0("NXF_OFFLINE=", .q("TRUE")), 
        "&&", "export", paste0("NXF_SINGULARITY_CACHEDIR=", .q(normalizePath("~/swxx/rj/nf-core/singularity_cache"))), 
        "&&", "export", paste0("NXF_OPTS=", .q("-Xms1g -Xmx4g")), 
        "&&", "nextflow run", .q(normalizePath("~/swxx/rj/nf-core/pipeline/atacseq/2.0/workflow")), 
        "--input", normalizePath(si_fp), "--max_memory", "50.GB", 
        "--skip_trimming true", "--read_length", "150", "--macs_gsize", 
        "2929823009", "--mito_name", "chrM", "--fasta", .q(normalizePath(t2t_fa_wj)), 
        "--bwa_index", t2t_fa_wj %=>% hs.wjm(ext = "index/bwa") %=>% 
            normalizePath %=>% .q, "--gff", .q(normalizePath(t2t_gff_wj)), 
        "-profile singularity", "-resume", "--outdir ."))
    if (0) {
        .ls(wd1)
        run_info <- hs.with_wd(wd1, .wk("nextflow -log .nextflow.log.2 log", 
            intern = 1)) %=>% wk.fread(text = x)
        run_info[, `SESSION ID` %=>% unique]
        wjj <- hs.with_wd(wd1, .wk("nextflow log voluminous_mcclintock", 
            intern = 1))
        wjj <- hs.with_wd(wd1, .wk("nextflow log voluminous_mcclintock", 
            intern = 1))
        .wk("du -sh", wjj[1])
        hs.hsgly("统计大小/2023_06_09_121941")(wjj)
        a <- hs.with_wd(wd1, .wk("nextflow clean", "-n", "voluminous_mcclintock", 
            intern = 1))
        a <- hs.with_wd(wd1, .wk("nextflow clean", "-n", "8561a1f8-73e7-4428-a112-375e6dbb019b", 
            intern = 1))
        hs.sub(a, "Would remove ") %=>% hs.hsgly("统计大小/2023_06_09_121941")(x)
        length(wjj)
    }
    {
        a <- wk.fread("~/swxx/t2t/GCF_009914755.1_T2T-CHM13v2.0_feature_count.txt.gz")
        wk.dt.set(a, , "Unique Ids", as.integer(a[, `Unique Ids`]))
        wk.dt.setorderv(a, "Unique Ids", -1)
        .s(a)
    }
    .chk("~/.nextflow")
    .less("~/.nextflow.log")
    .ls("~/.singularity")
    gtf_hg38 <- wk.fread("~/swxx/t2t/GCF_000001405.40_GRCh38.p14_genomic.gtf.gz")
    gtf <- wk.fread("~/swxx/t2t/GCF_009914755.1_T2T-CHM13v2.0_genomic.gtf.gz")
    {
        gff_fp <- "~/swxx/t2t/Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3.gz"
        gff_fp <- "~/swxx/t2t/GCF_009914755.1_T2T-CHM13v2.0_genomic.gff.gz"
        gff <- ape::read.gff(gff_fp)
        gff %<=>% wk.dt.as
        gff[, type %=>% hs.table %=>% sort]
        gff_type_nongene <- c("exon", "CDS", "three_prime_UTR", 
            "five_prime_UTR", "pseudogenic_transcript", "unconfirmed_transcript", 
            "V_gene_segment", "J_gene_segment", "D_gene_segment", 
            "C_gene_segment", "region")
        gff_type_gene <- gff[, type %=>% levels %=>% hs.grepV("(?i)gene$")]
        gff[type %in% gff_type_gene, attributes %=>% hs.re("(?<=biotype\\=)[^;]*") %=>% 
            hs.table %=>% sort]
        gff[type %in% gff_type_gene, .N]
        gff[type %in% gff_type_gene][1]
        gff[type %not.in% gff_type_nongene, type %=>% hs.table %=>% 
            sort]
        gff[type %in% "lnc_RNA"][1]
        gff[type %in% "mRNA"][1:2]
        gff[type %in% "unconfirmed_transcript"]
        gff[, which(type %in% "region")]
        gff[, V2 %=>% hs.table]
    }
    gtf[2]
    gtf[V3 == "gene", V1 %=>% hs.table]
    gtf[V3 == "gene"]
    gene_hg38 <- gtf_hg38[V3 == "gene", V9 %=>% hs.re("(?<=gene_id \")[^\"]*")]
    gene_t2t <- gtf[V3 == "gene", V9 %=>% hs.re("(?<=gene_id \")[^\"]*")]
    gene_hg38 %=>% anyDuplicated
    hs.xor(gene_t2t, gene_hg38)
    gtf_hg38[V3 == "gene"][gene_hg38 == "DDX11L1"]
    gtf_hg38[V3 == "gene"][gene_hg38 %not.in% gene_t2t, V9 %=>% 
        hs.re("(?<=pseudo \")[^\"]*") %=>% hs.table]
    gtf_hg38[V3 == "gene"][, V9 %=>% hs.re("(?<=pseudo \")[^\"]*") %=>% 
        hs.table]
    gtf_hg38[V3 == "gene"][, V9 %=>% hs.re("(?<=pseudo \")[^\"]*", 
        no_match = "")]
    gtf_hg38[V3 == "gene"][3]
    gtf_hg38[V3 == "gene", .N]
    gtf_hg38[1]
}
"38_02_25_222216" <- function(what) {
    wjj2 <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("10x") %>% 
        hs.fp("dataset", what)
    dz1 <- switch(what, pbmc3k = "http://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz", 
        "neuron_10k_v3")
    wj2 <- hs.wjm(dz1, dn = wjj2)
    if (!file.exists(wj2)) 
        wk.wget(dz1, wjj2)
    wk.untar(wj2) %>% hs.wjj.ls(R = 1) %>% {
        dirname(.[1])
    }
}
"36_04_25_230138" <- wjj0 <- "~/in2/swxx/sj/db/mclp"
"37_12_28_151351" <- function() {
}
"38_07_18_103101" <- function() evalq(wjj.38_07_18_103250 <- "~/swxx/sj/db/mca", 
    envir = .GlobalEnv)
"38_07_18_112023" <- function(wj1 = {
}, ..mem = 1) {
    "http://bis.zju.edu.cn/MCA/data/dge/One-Year-Brain.zip"
    hs.hsgly("单细胞/mca/数据.38_07_18_103101")()
    if (!length(wj1)) {
        .chk(hs.fp(wjj.38_07_18_103250, "in/exp"))
        return()
    }
    hs.require("Seurat")
    wj1 <- hs.fp(wjj.38_07_18_103250, "in/exp", wj1)
    if (!file.exists(wj1)) 
        hs.with_wd(hs.wjj(wjj.38_07_18_103250, "in/exp"), {
            dz1 <- hs.wjm(wj1, dn = "http://bis.zju.edu.cn/MCA/data/dge")
            .wk("wget -c", dz1)
        })
    m1 <- local({
        a <- hs.hsgly("列出文件名.38_07_14_100544")(wj1) %>% 
            hs.grepV("dge\\.csv")
        if (!length(a)) 
            hs.browser("38.07.18.112742")
        a <- wk.fread(cmd = hs.hsgly("文件/解压/指定文件.38_07_14_102931")(wj1, 
            a))
        a <- wk.dt.2matrix(a, rn.cn = 1)
        b <- hs.hsgly("列出文件名.38_07_14_100544")(wj1) %>% 
            hs.grepV("gene\\.csv")
        b <- wk.fread(cmd = hs.hsgly("文件/解压/指定文件.38_07_14_102931")(wj1, 
            b), cn = 1)
        rownames(a) <- b[, x]
        a
    })
    meta <- local({
        a <- hs.hsgly("列出文件名.38_07_14_100544")(wj1) %>% 
            hs.grepV("barcodes_anno\\.csv")
        a <- wk.fread(cmd = hs.hsgly("文件/解压/指定文件.38_07_14_102931")(wj1, 
            a))
        wk.dt.2matrix(a, rn.cn = 1)
    })
    ref <- SeuratObject::CreateSeuratObject(counts = m1)
    if (!all(rownames(meta) == rownames(ref@meta.data))) 
        hs.browser("38.07.18.113654")
    for (cn1 in colnames(meta)) ref@meta.data[, cn1] <- meta[, 
        cn1]
    ref <- NormalizeData(ref)
    ref <- FindVariableFeatures(ref, selection.method = "vst", 
        nfeatures = 2000, verbose = FALSE)
    ref %<>% (hs.hsgly("转录组/标准流程/模块/r函数_自制.38_04_07_003149", 
        "do.linear_dimensional_reduction"))
    ref
}
"38_07_14_095423" <- local({
    remove_unannotated_cells <- function(m1, meta) {
        colnames(m1) <- tolower(colnames(m1))
        ci <- colnames(m1) %in% meta[, cell]
        if (!all(ci)) 
            m1 <- m1[, ci, drop = 0]
        m1
    }
    function(tissue = {
    }, ..mem = 0) {
        wjj1 <- "~/swxx/sj/db/Tabula Muris"
        if (0) {
            "/Volumes/tf500/Tabula Muris/Single-cell RNA-seq data from Smart-seq2 sequencing of FACS sorted cells (v2)" %>% 
                wk.rsync.out(wjj1, ssh = "hp1")
        }
        if (0) {
            dim(meta)
            meta[, cell_ontology_class %=>% {
                uniqueN(x) == uniqueN(tolower(x))
            }]
            meta[1]
        }
        hs.with_wd(hs.fp(wjj1, "Single-cell RNA-seq data from Smart-seq2 sequencing of FACS sorted cells (v2)"), 
            {
                wj.zip <- "10700143.zip"
                wj <- hs.hsgly("列出文件名.38_07_14_100544")(wj.zip, 
                  ot = "v")
                wj <- hs.grepV(wj, "^(?i)FACS/.*\\.csv$")
                names(wj) <- hs.wjm(wj, dn = 0, ext = 0) %>% 
                  hs.sub("-counts$")
                if (!length(tissue)) {
                  message("请选择组织：")
                  wk.dt(组织 = names(wj), tar路径 = wj) %>% 
                    .s
                  return()
                }
                meta <- wk.fread(hs.fp(wjj1, "Single-cell RNA-seq data from Smart-seq2 sequencing of FACS sorted cells (v2)", 
                  "13088129.csv.gz"))
                meta[, `:=`(cell, tolower(cell))]
                wj1 <- "FACS/Brain_Non-Myeloid-counts.csv"
                if (0) {
                  i1 <- 1
                  wj1 <- wj[i1]
                }
                wj <- wj[tissue]
                m1_list <- lapply(wj, function(wj1) {
                  m1 <- wk.fread(cmd = hs.hsgly("文件/解压/指定文件.38_07_14_102931")(wj.zip, 
                    wj1), rn.cn = 1)
                  remove_unannotated_cells(m1, meta)
                })
                if (length(wj) > 1) {
                  a <- lapply(m1_list, rownames)
                  if ((sapply(a, length) %=>% (uniqueN(x) == 
                    1)) && (wk.dt.as(a) %=>% t %=>% unique %=>% 
                    (nrow(x) == 1))) {
                  }
                  else hs.browser("38.08.22.140949")
                }
                {
                  a <- m1_list %=>% sapply(x, function(m1) {
                    hs.rowSums(m1 > 0)
                  })
                  i <- hs.rowAnys(a >= 3)
                  m1_list %<=>% lapply(x, function(m1) m1[i, 
                    , drop = 0])
                  m1 <- do.call("cbind", unname(m1_list))
                }
                ref <- SeuratObject::CreateSeuratObject(counts = m1)
                if (0) {
                  rownames(m1) %=>% hs.grepV("-") %=>% .sh
                }
                ref %<=>% (Seurat::NormalizeData)(x, verbose = FALSE) %=>% 
                  (Seurat::FindVariableFeatures)(x, selection.method = "vst", 
                    nfeatures = 2000) %=>% (Seurat::ScaleData)(x, 
                  verbose = 0) %=>% hs.hsgly("预处理/降维，线性.38_10_09_120747")
                ref@meta.data[, "cell_type"] <- meta[match(colnames(ref), 
                  cell), cell_ontology_class]
                if (0) {
                  ref@meta.data[, "cell_type"] %=>% hs.table
                  ref@meta.data[1, ]
                }
                ref
            })
    }
})
"38_08_22_142212" <- function() {
    1
    {
        wjj1 <- "~/swxx/sj/db/Tabula Sapiens"
        if (0) {
        }
        hs.wjj(wjj1)
    }
}
"36_06_17_071658" <- {
    list(ensembl = c("~/in2/swxx/sj/jyz/human/ensembl/Homo_sapiens.GRCh38.dna.primary_assembly.fa", 
        "~/in2/swxx/sj/jyz/human/ensembl/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa"), 
        ncbi = "~/in2/swxx/sj/jyz/human/ncbi/GCF_000001405.39_GRCh38.p13_genomic.fna")
}
"36_09_21_151142" <- function(..mem = 1) {
    jg.wj <- .mfp.swxx("36_09_21_151246.tsv.gz")
    if (file.exists(jg.wj)) {
        wk.fread(jg.wj)
    }
    else {
        a1 <- hs.hsgly("MGI_Gene_Model_Coord.36_09_21_130734")()
        if (0) {
            a2 <- hs.hsgly("MGI.gff3.36_09_21_143042")()
            b1 <- a1[hs.clean(gene.id_ensmebl, ot = "l") & hs.clean(gene.id_ncbi, 
                ot = "l"), .(gene.id_ensmebl, gene.id_ncbi)]
            b2 <- a2[hs.clean(gene.id_ensmebl, ot = "l") & hs.clean(gene.id_ncbi, 
                ot = "l"), .(gene.id_ensmebl, gene.id_ncbi)]
        }
        jg <- wk.dt.subset(a1, exp_l = c(hs.clean(gene.id_ensembl, 
            ot = "l"), hs.clean(gene.id_ncbi, ot = "l")), exp_j = .(gene.id_ensembl, 
            gene.id_ncbi, gene.id_symbol))
        wk.dt.fwrite(jg, jg.wj)
        jg
    }
}
"36_09_21_130734" <- function(..mem = 1) {
    jg.wj <- .mfp.swxx("36_09_21_133649.tsv.gz")
    if (file.exists(jg.wj)) {
        wk.fread(jg.wj)
    }
    else {
        wj1 <- hs.with_wd("~/swxx/sj/db/mgi/gene", {
            dz1 <- "http://www.informatics.jax.org/downloads/reports/MGI_Gene_Model_Coord.rpt"
            wj2 <- basename(dz1)
            if (wj.update.reporter(wj2)) 
                hs.wj.rm(wj2)
            wk.wget(dz1, dirname(wj2))
        })
        a1 <- wk.fread(cmd = .sys("cat_wk", .sys.fpGood(wj1), 
            "| perl -e", .q("while( <>) { s/\t$//; print}")), 
            cn = 1)
        wk.dt.setnames(a1, "1. MGI accession id", "gene.id_mgi")
        wk.dt.setnames(a1, "3. marker symbol", "gene.id_symbol")
        wk.dt.setnames(a1, "4. marker name", "gene.name")
        wk.dt.setnames(a1, "6. Entrez gene id", "gene.id_ncbi")
        wk.dt.setnames(a1, "11. Ensembl gene id", "gene.id_ensembl")
        jg <- a1[, .SD, .SDcols = patterns("^gene\\.")]
        wk.dt.replace(jg, . %>% {
            .[. %in% "null"] <- NA
            .
        }, . %>% {
            "null" %in% .
        })
        wk.dt.fwrite(jg, jg.wj)
        jg
    }
}
"36_09_21_143042" <- function(..mem = 1) {
    jg.wj <- .mfp.swxx("36_09_21_143112.tsv.gz")
    if (file.exists(jg.wj)) {
        fread(jg.wj)
    }
    else {
        a1 <- fread(cmd = .sys(.sys.cat.skip("~/in2/swxx/sj/db/mgi/gene/MGI.gff3.gz", 
            header = 0), "|", .sys.perl.print("print if $F[ 1] eq \"MGI\"")))
        a1[, .N]
        jg <- a1
        jg[, `:=`("gene.id_symbol", hs.re(V9, "(?<=Name=)[^;,]+", 
            no_match = ""))]
        jg[, `:=`("gene.id_mgi", hs.re(V9, "(?<=gene_id=MGI:)\\d+", 
            no_match = ""))]
        jg[, `:=`("gene.id_name", hs.re(V9, "(?<=description=)[^;]+", 
            no_match = ""))]
        jg[, `:=`("gene.id_name", hs.gsub(gene.id_name, "%2c", 
            ","))]
        jg[, `:=`("gene.id_ncbi", hs.re(V9, "(?<=NCBI_Gene:)\\d+", 
            no_match = ""))]
        jg[, `:=`("gene.id_ensmebl", hs.re(V9, "(?<=ENSEMBL:)[^,;]+", 
            no_match = ""))]
        jg[, `:=`("gene.biotype", hs.re(V9, "(?<=so_term_name=)[^,;]+", 
            no_match = ""))]
        jg[, hs.re(V9, "(?<=mgi_type=)[^,;]+", no_match = "") %>% 
            hs.table]
        jg[, hs.table(hs.re(V9, "(?<=mgi_type=)[^,;]+", no_match = ""), 
            V3)]
        jg[, hs.table(gene.biotype, V3)]
        jg[, hs.table(gene.biotype, hs.re(V9, "(?<=mgi_type=)[^,;]+", 
            no_match = ""))] %>% .sh
        jg[, gene.biotype %>% hs.table %>% sort]
        a1[, V3 %>% hs.table]
        setnames(jg, "V1", "gene.chr_id")
        setnames(jg, "V4", "gene.chr_start")
        setnames(jg, "V5", "gene.chr_end")
        setnames(jg, "V7", "gene.chr_strand")
        jg[, `:=`(c("V2", "V3", "V6", "V8", "V9"), {
        })]
        wk.dt.fwrite(jg, jg.wj)
        jg
    }
}
"35_11_15_144010" <- local({
    hs1 <- function(t1) switch(t1, ens = "ENSEMBL", ncbi = "ENTREZID", 
        sym = "SYMBOL", .)
    function(id1, from = "ens", to = "sym", org1 = "human") {
        org_db.name <- switch(org1, human = "org.Hs.eg.db", mouse = "org.Mm.eg.db", 
            yeast = "org.Sc.sgd.db", fly = "org.Dm.eg.db", worm = "org.Ce.eg.db", 
            zebrafish = "org.Dr.eg.db")
        hs.require(c("clusterProfiler", org_db.name))
        from %<>% {
            switch(., ens = "ENSEMBL", ncbi = "ENTREZID", sym = "SYMBOL", 
                .)
        }
        to %<>% {
            switch(., ens = "ENSEMBL", ncbi = "ENTREZID", sym = switch(org1, 
                yeast = "GENENAME", "SYMBOL"), .)
        }
        a1 <- unique(id1) %>% bitr(fromType = from, toType = to, 
            OrgDb = org_db.name) %>% as.data.table
        a1 <- a1[, paste(sort(unique(get(to))), collapse = ";"), 
            by = c(from)]
        wk.dt.subset.na(a1, id1)
    }
})
"36_03_24_173032" <- function(tbl, cn1 = names(tbl)[1], ft = if (hs.grepl(tbl[[cn1]][1], 
    "^(?i)ensg")) "ensembl", tt = c("symbol", "name", "alias", 
    "ensembl", "ncbi") %not% ft, put.to.front = 1) {
    for (t1 in tt) {
        id1.id2 <- hs.hsgly("id1+.id2+.36.04.16.111737")(tbl[[cn1]], 
            ft, t1)
        cn2 <- switch(t1, symbol = "gene.id_symbol", name = "gene.name", 
            alias = "gene.id_alias", ensembl = "gene.id_ensembl", 
            ncbi = "gene.id_ncbi")
        tbl[, `:=`(c(cn2), id1.id2)]
    }
    if (put.to.front) 
        setcolorder(tbl, wk.dt.get.j(tbl, grep_or = "^gene\\."))
    tbl
}
"36_01_22_160941" <- local({
    hgnc <- {
    }
    if (0) {
        hgnc[(`NCBI Gene ID` %in% 414245) | (`NCBI Gene ID(supplied by NCBI)` %in% 
            414245)]
    }
    hs1 <- function(t1) {
        switch(t1, symbol = "Approved symbol", ensembl = "Ensembl gene ID", 
            name = "Approved name", uniprot = "gene.id_uniprot", 
            hgnc = "HGNC ID", ncbi = "NCBI Gene ID", hs.browser("36.01.17.221755"))
    }
    function(dt1 = {
    }, cn1 = "a", cn2 = "b", ft, tt, sep = ";") {
        if (!length(hgnc)) {
            hgnc <<- hs.hsgly("gene.id_hgnc.35.12.06.111314")()
            hgnc[, `:=`("NCBI Gene ID", as.character(`NCBI Gene ID`))]
            {
                i1 <- hgnc[, which(`Ensembl gene ID` == "")]
                hgnc[, which(hs.clean(`Ensembl gene ID`, inv = 1, 
                  ot = "l") & hs.clean(`Ensembl ID(supplied by Ensembl)`, 
                  ot = "l"))]
                if (length(i1)) 
                  hgnc[i1, `:=`("Ensembl gene ID", `Ensembl ID(supplied by Ensembl)`)]
            }
        }
        if (length(dt1) == 0) 
            return(hgnc)
        if (is.vector(dt1)) {
            dt1 %<>% as.character %>% data.table
            setnames(dt1, cn1)
        }
        cn1_hgnc <- hs1(ft)
        cn2_hgnc <- hs1(tt)
        a1 <- wk.dt.subset(hgnc, dt1[[cn1]], cn1_hgnc)[, .SD, 
            .SDcols = c(cn1_hgnc, cn2_hgnc)] %>% unique
        if (a1[[cn1_hgnc]] %>% {
            length(.) > uniqueN(.)
        }) 
            a1 <- a1[, paste(sort(unique(get(cn2_hgnc))), collapse = sep), 
                by = c(cn1_hgnc)]
        setnames(a1, c(cn1, cn2))
        dt1[, `:=`(c(cn2), wk.dt.subset.na(a1, get(cn1), cn1)[[cn2]])]
    }
})
"36_10_15_204106" <- function() {
    hgnc <- hs.hsgly("gene.id_hgnc.35.12.06.111314")()
    Homo_sapiens.gene_info <- fread("~/in2/swxx/sj/db/ncbi/gene/Homo_sapiens.gene_info.2020-10-15.gz")
    jg <- Homo_sapiens.gene_info[, .(GeneID, Symbol)]
    hgnc[1, .(`Approved symbol`, `NCBI Gene ID`, )]
    hgnc[hs.clean(`NCBI Gene ID`, ot = "l", inv = 1) & hs.clean(`NCBI Gene ID(supplied by NCBI)`, 
        ot = "l"), `:=`(`NCBI Gene ID`, `NCBI Gene ID(supplied by NCBI)`)]
    hgnc[hs.clean(`NCBI Gene ID`, ot = "l", inv = 1)]
    hgnc[Status == "Approved"][hs.clean(`NCBI Gene ID`, ot = "l", 
        inv = 1)]
    hgnc[!hs.clean(`NCBI Gene ID`, ot = "l"), .(`NCBI Gene ID`, 
        `NCBI Gene ID(supplied by NCBI)`)]
    hgnc[`NCBI Gene ID(supplied by NCBI)` == 100874108]
    jg[hs.clean(Symbol, ot = "l")]
    jg[, Symbol %>% nchar %>% hs.table]
    id_common <- hgnc[, `NCBI Gene ID`] %and% jg[, GeneID]
    jg[GeneID %not.in% id_common]
    jg[Symbol %>% {
        . %in% .[duplicated(.)]
    }] %>% setorderv("Symbol") %>% .s
    jg[, GeneID %>% anyDuplicated]
}
"36_06_08_172512" <- function(org1, ot = "obj", load = 1, install = 1) {
    org_db.name <- switch(org1, human = "org.Hs.eg.db", mouse = "org.Mm.eg.db", 
        rat = "org.Rn.eg.db", yeast = "org.Sc.sgd.db", fly = "org.Dm.eg.db", 
        worm = "org.Ce.eg.db", zebrafish = "org.Dr.eg.db", hs.browser("36.06.08.183348"))
    if (load) {
        hs.require(org_db.name, install = install)
    }
    switch(ot, obj = get(org_db.name), nam = org_db.name, col = columns(get(org_db.name)), 
        ls = ls(paste0("package:", org_db.name)))
}
"36_06_02_145930" <- local({
    wjj0 <- .mfp.swxx("36.06.02.152620")
    id1.id2.hs.orgDb <- function(db, cn1, cn2) {
        a1 <- as.list(db)
        a2 <- data.table(gene.id_ncbi = names(a1))
        a3 <- a2[, {
            as.character(a1[[gene.id_ncbi]])
        }, by = "gene.id_ncbi"]
        setnames(a3, c(cn1, cn2))
    }
    org.wj <- hs.fp(wjj0, "1.org")
    if (0) {
        cat("* org.Hs.eg.db", sep = "\n", file = org.wj)
        cat("** column name", sep = "\n", file = org.wj, append = 1)
        cat(names(orgDb), sep = "\n", file = org.wj, append = 1)
    }
    pn <- "org.Hs.eg.db"
    function(..mem = 1) {
        if (!hs.havePackages(pn)) 
            stop(rep("\n", 2), "@36.06.02.154606@ 先安装", 
                pn, rep("\n", 4))
        V <- installed.packages(noCache = 1)[pn, "Version"]
        wj_out <- hs.wjm(V, add = "tsv.gz", dn = wjj0)
        if (file.exists(wj_out)) {
            note.update_time(wj_out, paste(pn, V))
            jg <- fread(wj_out)
            jg[, `:=`("gene.id_ncbi", as.character(gene.id_ncbi))]
        }
        ..meta.wj_out
        hs.require("org.Hs.eg.db")
        ls("package:org.Hs.eg.db") %>% .sh
        orgDb <- id1.id2.hs.orgDb(org.Hs.egENSEMBL, "gene.id_ncbi", 
            "gene.id_ensembl")
        a1 <- as.list(org.Hs.egSYMBOL)
        orgDb %<>% {
            .[, data.table(.SD, gene.id_symbol = a1[[gene.id_ncbi]]), 
                by = "gene.id_ncbi"]
        }
        a1 <- as.list(org.Hs.egUNIPROT)
        orgDb %<>% {
            .[, data.table(.SD, gene.id_uniprot = as.character(a1[[gene.id_ncbi]])), 
                by = "gene.id_ncbi"]
        }
        a1 <- as.list(org.Hs.egGENENAME)
        orgDb %<>% {
            .[, data.table(.SD, gene.name = as.character(a1[[gene.id_ncbi]])), 
                by = "gene.id_ncbi"]
        }
        a1 <- as.list(org.Hs.egALIAS2EG)
        a1 %<>% inverseList
        orgDb %<>% {
            .[, data.table(.SD, gene.id_alias = as.character(a1[[gene.id_ncbi]])), 
                by = "gene.id_ncbi"]
        }
        wk.dt.fwrite(orgDb, wj_out)
        orgDb %>% uniqueN
        orgDb %>% dim
        wj_out %>% .ls
    }
})
"37_04_06_191822" <- local({
    hs1 <- function(wj_in) {
        wjj.36_03_04_003344 <- getwd()
        on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
        setwd(dirname(wj_in))
        wj_in %<>% basename
        wj_out <- hs.wjm(wj_in, append = ".id.tax_id")
        if (wj.update.reporter(wj_out, wj_in)) {
            .wk(.sys.cat.text(wj_in), "| perl -e", .q("\nwhile( <>) {\n  if( /^ID\\s/) {\n    $' =~ /\\s([^\\s]+)/;\n    $id = $1;\n    next;\n  }\n  if( /^OX\\s/) {\n    $' =~ /NCBI_TaxID=(\\d+)/;\n    print( $id, \"\\t\", $1, \"\\n\");\n  }\n}\n"), 
                "| gzip >", wj_out)
        }
        wj_out <- hs.wjm(wj_in, append = ".id.ac")
        if (wj.update.reporter(wj_out, wj_in)) {
            .wk(.sys.cat.text(wj_in), "| perl -e", .q("\nwhile( <>) {\n  if( /^ID\\s/) {\n    /\\s([^\\s]+)/;\n    $id = $1;\n    next;\n  }\n  if( /^AC\\s/) {\n    $' =~ /\\s+(.+)/;\n    $ac = $1;\n    foreach $ac1 ( split( \"; *\", $1)) {\n      print( $id, \"\\t\", $ac1, \"\\n\");\n    }\n  }\n}\n"), 
                "| gzip >", wj_out)
        }
        wj_out <- hs.wjm(wj_in, append = ".ac.ncbi")
        if (wj.update.reporter(wj_out, wj_in)) {
            .wk(.sys.cat.text(wj_in), "| perl -e", .q("\nwhile( <>) {\n  if( /^ID\\s/) {\n    ## 输出上一个id的内容\n    foreach $ac1 ( split( \"; *\", $ac)) {\n      foreach $v1 ( values %geneid) {\n        print( $ac1, \"\\t\", $v1, \"\\n\");\n      }\n    }\n\n    $geneid_i1 = 0;\n    undef %geneid;\n    next;\n  }\n  if( /^AC\\s/) {\n    /^AC\\s+(.+)/;\n    $ac = $1;\n    next;\n  }\n  if( /^DR\\s+GeneID;/) {\n    /\\d+/;\n    $geneid_i1 = $geneid_i1 + 1;\n    $geneid{ $geneid_i1} = $&;\n  }\n}\n"), 
                "| gzip >", wj_out)
        }
        if (0) 
            if (wj.update.reporter(wj_out, wj_in)) {
                .wk(.sys.cat.text(wj_in), "| perl -e", .q("\nwhile( <>) {\n  if( /^ID\\s/) {\n    ## 输出上一个id的内容\n    foreach $ac1 ( split( \"; *\", $ac)) {\n      print( $id, \"\\t\", $ac1, \"\\tAC\", \"\\t\", $org, \"\\n\");\n    }\n    foreach $v1 ( values %geneid) {\n      print( $id, \"\\t\", $v1, \"\\tGeneID\", \"\\t\", $org, \"\\n\");\n    }\n\n    /\\s([^\\s]+)/;\n    $id = $1;\n    $geneid_i1 = 0;\n    undef %geneid;\n  }\n  if( /^AC\\s/) {\n    /^AC\\s+(.+)/;\n    $ac = $1;\n  }\n  if( /^DR\\s+GeneID;/) {\n    /\\d+/;\n    $geneid_i1 = $geneid_i1 + 1;\n    $geneid{ $geneid_i1} = $&;\n  }\n  if( /^OX\\s+NCBI_TaxID=(\\d+)/) {\n    $org = $1;\n  }\n}\n"), 
                  "| gzip >", wj_out)
                if (0) {
                  jg <- wk.fread.chunk.do.lines(gzfile(wj_in), 
                    hs = . %>% {
                      sj <- hs.sepSeq(., sep.i = hs.grep(., "^ID\\s"))
                      a1 <- sj %>% lapply(. %>% {
                        id <- hs.grepV(., "^ID") %>% hs.re("(?<= )[^ ]+")
                        ac <- hs.grepV(., "^AC") %>% hs.re("(?<= )[^ ]+") %>% 
                          strsplit(";") %>% unlist %>% trim
                        ncbi <- unlist(hs.gre(hs.grepV(., "DR\\s+GeneID"), 
                          "\\d+"))
                        if (!length(ncbi)) 
                          ncbi <- ""
                        expand.grid(id, ac, ncbi, stringsAsFactors = FALSE) %>% 
                          as.data.table
                      })
                      rbindlist(a1)
                    }, hs.sep.i = . %>% hs.grep("^ID\\s"), nrows = 1e+06)
                  jg <- rbindlist(jg)
                  setnames(jg, c("ID", "AC", "gene.id_ncbi"))
                  wk.dt.fwrite(jg, wj_out)
                }
                if (0) {
                  ens.uniprot <- local({
                    a1 <- fread("/Volumes/in2/mouse/ens.uniprot.txt.gz")
                    a1 <- a1[, hs.clean(`UniProtKB/Swiss-Prot ID` %or% 
                      `UniProtKB/TrEMBL ID`), by = "Gene stable ID"]
                    setnames(a1, paste0("gene.id_", c("ensembl", 
                      "uniprot")))
                  })
                  ens.ncbi <- local({
                    a1 <- fread("/Volumes/in2/mouse/ens.ncbi.txt.gz")
                    setnames(a1, paste0("gene.id_", c("ensembl", 
                      "ncbi")))
                    a1[hs.clean(gene.id_ncbi, ot = "l")]
                  })
                  uniprot.ncbi
                  a1 <- wk.dt.absorb1(ens.ncbi, ens.uniprot)
                  a1[gene.id_ensembl %in% "ENSMUSG00000034853"]
                  .mfp.swxx("36_12_11_193512")
                  ens.ncbi[`NCBI gene (formerly Entrezgene) ID` %in% 
                    "329910"]
                  ens.uniprot[`Gene stable ID` %in% "ENSMUSG00000034853"]
                }
            }
        jg <- hs.wjj.ls(".", hs.wjm(wj_in, append = "\\..+"))
        jg %>% {
            hs.c(hs.sub(basename(.), hs.sub(wj_in, "[^\\.]+$")) %>% 
                hs.sub("\\.gz$"), hs.wj.portable(.))
        }
    }
    hs2 <- function(org1, reviewed = 1) {
        wjj0 <- hs.fp("~/in2/swxx/sj/db/uniprot/UniProtKB", hs.title(paste0(ifelse(reviewed, 
            "", "un"), "reviewed")))
        org1.fp.str <- switch(org1, human = "human", mouse = "rodents", 
            hs.browser("37.04.07.150926"))
        wj_in <- hs.fp(wjj0, paste0("uniprot_", ifelse(reviewed, 
            "sprot", "trembl"), "_", org1.fp.str, ".dat.gz"))
        dz1 <- hs.fp("https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions", 
            basename(wj_in))
        wj.updater(wj_in, dz1, co1 = 56)
        hs1(wj_in)
    }
    function(org1, reviewed = 1) {
        switch(as.character(reviewed), `0` = , `1` = hs2(org1, 
            reviewed), `2` = sapply(c(0, 1), function(reviewed) hs2(org1 = org1, 
            reviewed = reviewed))) %>% t %>% as.data.table
    }
})
"37_04_07_161443" <- function(id1, ft = "ac", tt = "id") {
}
"36_07_24_091107" <- local({
    db_wj <- "~/in2/swxx/sj/db/ncbi/gene/generifs_basic.fst"
    gene_id <- {
    }
    function(id, merge = 1) {
        if (!length(gene_id)) 
            gene_id <<- local({
                if (0) {
                  wk.wget("ftp://ftp.ncbi.nih.gov/gene/GeneRIF/generifs_basic.gz", 
                    dirname(db_wj))
                }
                wj1 <- hs.wjm(db_wj, ext = "gz")
                if ((!file.exists(db_wj)) || (file.mtime(db_wj) < 
                  file.mtime(wj1))) {
                  sj <- fread(wj1, quote = "")
                  wk.fst.write(sj, db_wj)
                }
                wk.fst.read(db_wj, "Gene ID")
            })
        ri <- gene_id[, which(`Gene ID` %in% id)]
        jg <- wk.fst.read(db_wj, ri = ri, ci = -1)
        if (merge) {
            jg[, {
                .(gene.description_ncbiRIF = paste(paste0("N=", 
                  .N, ". ", `GeneRIF text`, " (", `last update timestamp`, 
                  "; pmid:", `PubMed ID (PMID) list`, ")")[order(`last update timestamp`, 
                  decreasing = TRUE)], collapse = " "))
            }, by = .(gene.id_ncbi = `Gene ID`)]
        }
        else jg
    }
})
"36_07_24_114431" <- local({
    db <- list()
    function(id, org1 = "human") {
        if (org1 %not.in% names(db)) {
            jg_wj <- .mfp.swxx(paste0(switch(org1, human = "36.08.05.214638", 
                mouse = "36_12_06_231347"), ".tsv.gz"))
            db[org1] <<- list(fread(jg_wj))
        }
        db[[org1]][match(id, gene.id_ncbi), gene.description_ncbi]
    }
})
"36_07_24_113535" <- function(id) {
    hs.require("rentrez")
    id.L <- hs.makeBatch(id, K = 100)
    hs.c(id, unlist(lapply(id.L, . %>% {
        a1 <- entrez_summary("gene", ., always_return_list = 1)
        sapply(a1, "[[", "summary")
    })))
}
"36_06_20_224138" <- function(org1 = "human", info1 = "FUNCTION", 
    ..mem = 1) {
    wj1 <- "~/swxx/sj/db/uniprot/UniProtKB/Reviewed/uniprot_sprot.dat.gz"
    wj1 <- "~/swxx/sj/db/uniprot/ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/uniprot_sprot.dat.gz"
    wj_out <- hs.wjm(wj1, append = paste0(".", org1, ".", info1))
    if (!file.exists(wj_out)) {
        a1 <- wk.fread.chunk.do.lines(gzfile(wj1), hs = . %>% 
            {
                sj <- hs.sepSeq(., sep.i = hs.grep(., "^ID\\s"))
                names(sj) <- sapply(sj, . %>% {
                  hs.re(.[1], "(?<= )[^ ]+")
                })
                i1 <- hs.grep(names(sj), paste0("_", switch(org1, 
                  human = "HUMAN", mouse = "MOUSE", hs.browser("36.12.06.214633")), 
                  "$"))
                if (length(i1)) {
                  if (0) {
                    hs.browser("36.06.21.075828")
                    sj[[i1[1]]] %>% hs.grepV("(?i)ensg")
                    sj[[i1[1]]] %>% hs.grepV("^DR\\s") %>% .ss
                    a1 <- sj[i1] %>% sapply(. %>% hs.grepV("^DR\\s+GeneID;"))
                    sj[i1[which(sapply(a1, length) > 1)][1]]
                    sapply(a1, length) %>% hs.table
                  }
                  sj[i1[1]]
                  jg <- lapply(i1, function(i1) {
                    jg <- rep("", 3)
                    names(jg) <- c("FUNCTION", "SUBCELLULAR LOCATION", 
                      "KEY WORDS")
                    . <- sj[[i1]] %>% hs.grepV("^CC\\s")
                    if (0) {
                      "3BP5_HUMAN"
                      if (names(sj)[i1] == "2ABB_HUMAN") 
                        hs.browser("37.03.02.123257")
                    }
                    sep.i <- hs.grep(., "^CC\\s+-!- ")
                    if (length(sep.i)) {
                      CC <- hs.sepSeq(., sep.i = sep.i)
                      names(CC) <- sapply(CC, "[", 1) %>% hs.sub("CC\\s+-!-\\s+") %>% 
                        hs.re("[^:]+")
                      jg[1:2] <- c(if ("FUNCTION" %in% names(CC)) {
                        a1 <- CC[["FUNCTION"]] %>% hs.sub("CC\\s+")
                        a1[1] %<>% hs.sub("[^:]+:\\s+")
                        paste(a1, collapse = " ")
                      } else "", if ("SUBCELLULAR LOCATION" %in% 
                        names(CC)) {
                        a1 <- CC[names(CC) %in% "SUBCELLULAR LOCATION"]
                        a1 %<>% lapply(. %>% hs.sub("CC\\s+"))
                        a1 %>% lapply(. %>% {
                          .[1] %<>% hs.sub("[^:]+:\\s+")
                          paste(., collapse = " ") %>% hs.sub(" Note=.*") %>% 
                            hs.gsub("{[^{}]+}") %>% hs.gsub("\\[[^\\[\\]]+\\]: ") %>% 
                            strsplit(" *[\\.;] *")
                        }) %>% unlist %>% unique %>% {
                          paste(c("", ., ""), collapse = "|")
                        }
                      } else "")
                    }
                    . <- sj[[i1]] %>% hs.grepV("^KW\\s")
                    jg[3] <- hs.sub(., "^KW\\s+") %>% paste(collapse = " ") %>% 
                      hs.gsub("\\s*[\\.;]\\s*", "|") %>% {
                      paste0("|", ., "|")
                    } %>% hs.gsub("\\|{2,}", "|")
                    jg
                  })
                  names(jg) <- names(sj)[i1]
                  jg
                }
            }, hs.sep.i = . %>% hs.grep("^ID\\s"))
        jg <- unlist(a1, 0)
        tbl <- do.call("rbind", unname(jg)) %>% as.data.table %>% 
            setnames(names(jg[[1]]))
        tbl[, `:=`(uniprot_id, names(jg))]
        setcolorder(tbl, c("uniprot_id", names(jg[[1]])))
        if (0) {
            tbl[17]
            sj["2ABB_HUMAN"]
            tbl[, `SUBCELLULAR LOCATION`]
            tbl[hs.grep(uniprot_id, "^(?i)eda_")]
        }
        wk.dt.fwrite(tbl, wj_out)
    }
    wj_out
}
"38_11_15_145951" <- function(fa.gz) {
    if (!hs.grepl(fa.gz, "(?i)\\.gz$")) 
        return(fa.gz)
    wjj0 <- "/dev/shm"
    if (dir.exists(wjj0)) {
        wj2 <- hs.fp(wjj0, "38_11_15_150418", Sys.info()["user"]) %>% 
            paste0(normalizePath(fa.gz)) %>% hs.wjm(ext = 0)
        if (!file.exists(wj2)) {
            .sys.cat.text(fa.gz) %>% .wk(">", .sys.fpGood(hs.wj(wj2)))
        }
    }
    else wj2 <- fa.gz
    wj2
}
"38_11_11_162045" <- function(chr) {
    a <- chr %=>% tapply(x, hs.grepl(x, "\\d") %=>% as.numeric, 
        c)
    a[["1"]] %<=>% x[hs.re(x, "\\d+") %=>% as.numeric %=>% order]
    a[["0"]] %<=>% x[order(x)]
    a[c("1", "0")] %=>% hs.unlist
}
"38_11_15_114746" <- function() "~/swxx/sj/基因组/人/t2t/chm13v2.0.fa.gz"
"2023_06_14_181536" <- function() {
    hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")() %=>% 
        hs.fp("genome")
}
"2023_06_14_201613" <- function(org1 = "human") {
    hs.switch(org1, "human", "hg38", hs.browser("2023.06.14.194258", 
        debug = 0))
}
"2023_06_14_122113" <- function(...) {
    wjj2 <- hs.sjgly("2023_06_14_122113")
    arg.c <- hs.wj(wjj2, "config.yaml")
    if ((!file.exists(arg.c)) || (configr::read.config(arg.c)[["genome_folder"]] != 
        normalizePath(wjj2))) {
        hs.wj.rm(arg.c)
        .wk("refgenie init -c", hs.fpGood(arg.c))
    }
    if (0) {
        .ls(wjj2)
        .less(arg.c)
    }
    .wk("refgenie", ..., "-c", arg.c %=>% hs.fpGood)
}
"39_02_25_102641" <- function(org1 = "human") {
    hs.hsgly("编号种类转换库.37_03_29_103455")()
}
"37_09_13_095732" <- local({
    wjj0 <- {
    }
    function(what = {
    }) {
        if (!length(wjj0)) 
            wjj0 <<- hs.hsgly("生物学/根文件夹.37_07_25_182333")() %=>% 
                hs.fp("sj/db")
        if (length(what)) {
            wjj1 <- hs.wjj.ls(wjj0, paste0("^(?i)", tolower(what), 
                "$"), wjj = 1)
            if (length(wjj1) == 1L) 
                return(wjj1)
            {
                return(hs.fp(wjj0, what))
            }
        }
        else wjj0
    }
})
"36_04_13_155130" <- function() {
    a1 <- readLines("/home/wk/org/工作/生物学/生信/组学数据库/药物/myCancerGenome/通路/pathway.txt")
    i1 <- which(a1 == "")
    a2 <- hs.sepSeq(a1, sep.i = i1)
    a2 <- hs.c(sapply(a2, . %>% {
        hs.grep(., "^#", v = 1)[1]
    }) %>% trim("[ #]"), lapply(a2, hs.grep, "^[^#]", v = 1))
    jg <- data.table(gene.id_symbol = unlist(a2), pathway = rep(names(a2), 
        sapply(a2, length)))
    if (0) {
        jg[, if (.N > 1) 
            .SD, by = "gene.id_symbol"]
        jg[, gene.id_symbol %>% uniqueN]
    }
    jg
}
"38_02_18_161229" <- function() {
    wjj.38_02_18_162222 <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("cellphonedb")
    if (0) {
        wk.rsync.out(wjj.38_02_18_162222, hs.dirname(wjj.38_02_18_162222))
        wjj.38_02_18_162222 %>% .ls
    }
}
"38_03_30_074330" <- local({
    feather.v1_to_v2 <- function(fp1, fp2 = hs.wjm(fp1, append = ".v2")) {
        .wk("~/swxx/rj/create_cisTarget_databases-master/convert_cistarget_databases_v1_to_v2.py", 
            "-i", .sys.fpGood(fp1), "-o", .sys.fpGood(fp2))
        fp2
    }
    if (0) {
        feather.v1_to_v2("~/swxx/sj/db/cistarget/resources.aertslab.org/cistarget/databases/mm10__refseq-r80__10kb_up_and_down_tss.mc9nr.feather")
        feather.v1_to_v2("~/swxx/sj/db/cistarget/resources.aertslab.org/cistarget/databases/hg38__refseq-r80__10kb_up_and_down_tss.mc9nr.feather")
    }
    function(what = "tf", org1 = "human") {
        wjj.38_03_30_074450 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")(), 
            "sj/db/cistarget")
        switch(what, tf = local({
            hs.fp(wjj.38_03_30_074450, "github.com/aertslab/pySCENIC/blob/master/resources", 
                switch(org1, human = "hs_hgnc_tfs.txt", mouse = "mm_mgi_tfs.txt"))
        }), motif.tf = local({
            a1 <- "resources.aertslab.org/cistarget/motif2tf"
            a2 <- switch(org1, human = "motifs-v9nr_clust-nr.hgnc-m0.001-o0.0.tbl", 
                mouse = "motifs-v9nr_clust-nr.mgi-m0.001-o0.0.tbl")
            wk.wget(hs.fp(paste0("https://", a1), a2), hs.fp(wjj.38_03_30_074450, 
                a1), gz = 0)
        }), ranking_genome = local({
            a1 <- "resources.aertslab.org/cistarget/databases"
            a2 <- switch(org1, human = hs.fp("homo_sapiens/hg38/refseq_r80/mc9nr/gene_based", 
                "hg38__refseq-r80__10kb_up_and_down_tss.mc9nr.genes_vs_motifs.rankings.feather"), 
                mouse = hs.fp("mus_musculus/mm10/refseq_r80/mc9nr/gene_based", 
                  "mm10__refseq-r80__10kb_up_and_down_tss.mc9nr.genes_vs_motifs.rankings.feather"))
            fp1 <- wk.wget(hs.fp(paste0("https://", a1), a2), 
                hs.fp(wjj.38_03_30_074450, a1))
            feather.v1_to_v2(fp1)
        }))
    }
})
"36_04_14_142241" <- function() {
}
"34_01_29_185206" <- function(id = {
}) {
    sj <- local({
        a1 <- "~/Downloads/WRckFaM3WXtbNnro-4960312.idmapper.txt" %>% 
            readLines
        a2 <- hs.sepSeq(a1)
        sapply(a2, . %>% {
            . %not% ""
        }) %>% hs.nonempty %>% sapply(. %>% {
            c(hs.re(.[2], "[^\\.]+"), hs.grep(., "retired", v = 1)[1] %>% 
                hs.re("(?<=retired).+") %>% hs.re("\\d+"))
        }) %>% t %>% as.data.table
    })
    hs.require("biomaRt", 1)
    db.ea <- listEnsemblArchives() %>% as.data.table
    r1 <- sj[, {
        v1 <- as.integer(V2) - 1
        url1 <- wk.dt.subset(db.ea, paste(v1), "version")[, url]
        mart1 <- useMart("ENSEMBL_MART_ENSEMBL", host = url1)
        ds1 <- useDataset("mmusculus_gene_ensembl", mart = mart1)
        if (0) {
            `?`(listMarts)
            m1 <- listMarts()
            `?`(useMart)
            `?`(listDatasets)
            db.ds <- listDatasets(mart1) %>% as.data.table
            db.ds[description %>% hs.grep("(?i)mouse|human")]
            ma1 <- listAttributes(ds1) %>% as.data.table
            mf1 <- listFilters(ds1) %>% as.data.table
            ma1[, name]
            ma1[, page %>% table]
            ma1[page == "homologs"][description %>% hs.grep("(?i)human")]
            ma1[, description %>% hs.grep("(?i)name|symbol", 
                v = 1) %>% hs.grep("(?i)chro", v = 1, inv = 1)]
            ma1[page == "feature_page"][description %>% hs.grep("(?i)entrez")]
            ma1[description == "MGI symbol"]
            ma1[description == "Associated Gene Name"]
            ma1[description == "UniProt Gene Name"]
            ma1[description == "WikiGene Name"]
            ma1[name == "protein_id"]
            mf1[name %>% hs.grep("(?i)peptide")]
            `?`(getBM)
            getBM(attributes = c("ensembl_peptide_id", ma1[page == 
                "feature_page", name[126]]), filters = "ensembl_peptide_id", 
                values = V1, mart = ds1)
        }
        hs.browser("35.01.29.203511")
        a1 <- getBM(attributes = ma1[page == "feature_page", 
            name], filters = "ensembl_peptide_id", values = V1, 
            mart = ds1)
        getBM(attributes = c("ensembl_peptide_id", "external_gene_name", 
            "embl", "entrezgene", "mgi_id"), filters = "ensembl_peptide_id", 
            values = V1, mart = ds1) %>% .sh
    }, by = .(V2)]
}
"36_12_21_003842" <- function(org1 = "human", version = "in_use") {
    ver1 <- hs.hsgly("数据库版本.37_09_14_140925")(version)
    wj1 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")(), 
        "sj/db/ensembl", org1, "release", ver1) %>% hs.wjj.ls("\\.gtf\\.gz$")
    if (!length(wj1)) {
        dz0 <- paste0("http://ftp.ensembl.org/pub/release-", 
            hs.re(ver1, "\\d+"), "/gtf/", hs.hsgly("ensembl_species_names.scientific_name.36_12_25_190120")(org1))
        a1 <- readLines(dz0)
        wjm1 <- hs.re(a1, "[^>]+\\.gtf\\.gz(?=</a>)") %>% hs.grepV("(chr|abinitio)\\.gtf\\.gz$", 
            inv = 1)
        dz1 <- hs.fp(hs.sub(dz0, "http", "ftp"), wjm1)
        wj1 <- hs.fp("~/swxx/sj/db/ensembl", org1, "release", 
            ver1, wjm1)
        wj.updater(wj1, dz1)
    }
    wj1
}
"36_05_26_233803" <- local({
    cn.hs <- function(name) switch(name, name = "gene.name", 
        paste0("gene.id_", name))
    function(id1, id2, ..mem = 1) {
        id1.ci <- ci.list[[id1]]
        id2.ci <- ci.list[[id2]]
        a1 <- hgnc[, unique(.SD), .SDcols = c(id1.ci, id2.ci)]
        a1 <- a1[, unlist(.SD) %not% c("", NA), by = c(id2.ci)]
        if ("V1" %not.in% names(a1)) 
            hs.browser("36.07.09.120746")
        setnames(a1, "V1", cn.hs(id1))
        setcolorder(a1, c(length(a1), 1:(length(a1) - 1)))
        switch(as.character(length(id2.ci)), `1` = {
            setnames(a1, 2, cn.hs(id2))
        }, `2` = {
            switch(id2, alias = {
                a1[[2]] <- paste0(a1[[2]], ", ", a1[[3]]) %>% 
                  trim(", ")
                a1[, `:=`(tail(names(a1), -2), {
                })]
            }, {
                a1 <- a1[, unique(unlist(.SD)), by = c(cn.hs(id1))]
            })
            setnames(a1, sapply(c(id1, id2), cn.hs))
        }, stop("36.04.13.024226 id2不能超过两列"))
    }
})
"37_09_13_135206" <- function(org1 = "human", gv = 38, what = "gtf") {
    species <- hs.hsgly("列表.37_01_25_194601")()
    if (0) {
        species[hs.grep(`Common name`, "sheep")]
        species[hs.grep(`Common name`, "macaque")]
    }
    release_n <- hs.hsgly("生物学/数据库/ensembl/最新版本号.37_12_30_165505")()
    if (hs.all(org1 == "human", gv == 37)) {
        hs.browser("37.09.30.161350")
        release_n <- 75
    }
    release_path <- paste0("http://ftp.ensembl.org/pub/release-", 
        release_n)
    org_path <- species[`Common name` %in% org1, ftp_fp]
    if (0) {
        species[`Common name` %in% org1]
        species[hs.grep(`Common name`, org1)]
        species[hs.grep(`Common name`, "goat")]
    }
    url1 <- hs.fp1(release_path, switch(what, genome_sequence = c("fasta", 
        org_path, "dna"), ncbi_gene_id = c("tsv", org_path), 
        gtf = c("gtf", org_path), regulation = c("regulation", 
            org_path), hs.browser("38.04.09.134023")))
    html <- hs.evalLazy(xml2::read_html(url1), hs4save = xml2::write_html, 
        hs4load = xml2::read_html, fp = hs.fp(hs.hsgly("ensembl/url2fp.38_04_09_140039")(url1), 
            "a.html.gz"))
    wjm <- xml2::xml_find_all(html, "//a/@href") %=>% xml2::xml_text(x) %=>% 
        hs.grepV(switch(what, genome_sequence = "(?i)[^\"]*\\.dna.(primary_assembly|toplevel).fa.gz", 
            gtf = paste0("(?i)", release_n, "\\.gtf\\.gz$"), 
            ncbi_gene_id = "(?i)\\.entrez\\.tsv\\.gz$", regulation = "(?i)\\.gff\\.gz$"))
    if ((length(wjm) > 1) && any(hs.grepl(wjm, "primary_assembly"))) 
        wjm <- hs.grepV(wjm, "primary_assembly")
    if (length(wjm) != 1) 
        hs.browser("37.09.13.140913")
    dz1 <- hs.fp(url1, wjm)
    wj2 <- hs.hsgly("ensembl/url2fp.38_04_09_140039")(dz1)
    if (!length(hs.fp(wj2, robust = 1))) {
        hs.hsgly("1.信息学/文件/获取/用aria2c.2023_06_26_120607")(dz1, 
            wjj2 = dirname(wj2))
        if (0) {
            .ls(wj2)
            .head(wj2)
            a <- wk.fread(wj2)
            a[gene_stable_id %in% "ENSCHIG00010010790"]
            wk.rsync.out("/Volumes/tf500/Ovis_aries_rambouillet.Oar_rambouillet_v1.0.107.gtf.gz", 
                "/home/wk/swxx/sj/db/ensembl/ftp.ensembl.org/pub/release-107/gtf/ovis_aries_rambouillet", 
                ssh = "hp1")
            wk.rsync.out("/Volumes/tf500/Ovis_aries_rambouillet.Oar_rambouillet_v1.0.dna.toplevel.fa.gz", 
                "~/swxx/sj/db/ensembl/ftp.ensembl.org/pub/release-107/fasta/ovis_aries_rambouillet/dna", 
                ssh = "hp1")
        }
    }
    if (0) {
        fa.gz <- hs.fp(wjj2, wjm)
        fa <- hs.wjm(fa.gz, ext = 0)
        if (file.exists(fa)) {
        }
        else {
            wk.wget(hs.fp(url1, wjm), wjj2)
            .wk("gzip -cd", .sys.fpGood(fa.gz), ">", .sys.fpGood(fa))
        }
        hs.hsgly("fa.fai.37_04_14_182739")(fa)
        fa
    }
    hs.fp(wj2, robust = 1)
}
"38_01_05_114445" <- function(org1 = "human") {
    wj_in <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1, 
        what = "gtf")
    wj_out <- hs.wjm(wj_in, append = ".ensg,enst,ensp")
    if (!file.exists(wj_out)) {
        cmd1 <- .sys("cat_wk", .sys.fpGood(wj_in), "| perl -e", 
            .q("\nwhile( <>) {\n  @F = split( \"\t\");\n  if( $F[ 2] eq \"CDS\") {\n    $F[ 8] =~ /\\bgene_id \"([^\"]+)/;\n    $g = $1;\n    $F[ 8] =~ /\\btranscript_id \"([^\"]+)/;\n    $t = $1;\n    $F[ 8] =~ /\\bprotein_id \"([^\"]+)/;\n    $p = $1;\n    print( $g, \"\t\", $t, \"\t\", $p, \"\n\");\n  }\n}\n"), 
            "| uniq")
        .wk("cat", "<( echo", .q(paste(c("gene", "transcript", 
            "protein"), collapse = "\t")), ")", "<(", cmd1, ")", 
            "| gzip >", .sys.fpGood(wj_out))
    }
    wj_out
}
"36_05_27_060241" <- local({
    db <- list(symbol = "external_gene_name", name = "description", 
        alias = "external_synonym", uniprot = "uniprotswissprot", 
        uniprot2 = "uniprotsptrembl", ncbi = "entrezgene_id", 
        ncbi_symbol = "entrezgene_accession", enst = "ensembl_transcript_id", 
        ensp = "ensembl_peptide_id", go = c("namespace_1003", 
            "go_id", "go_linkage_type"))
    function(id1 = {
    }, org1 = "human", version = hs.hsgly("biomaRt.datatset.version_wanted.35.04.30.214633")(org1 = org1), 
        redo = 0) {
        if (!length(id1)) 
            return(db)
        if (id1 == "go") 
            warning("36.05.27.210356 seems go annotations from ensembl contain only direct associations")
        id2 <- db[[id1]]
        wj_out <- .mfp("36.05.27.060411.tsv.gz", hs.hsgly("wjj.ensembl.35.10.09.102213")())
        wj_out <- hs.wjm(wj_out, append = paste0(".v=", version, 
            ".", paste(id2, collapse = ",")), extN = 2)
        if (redo) 
            hs.wj.rm(wj_out)
        ..wj_old <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1, 
            what = "gtf")
        ..meta.wj_out
        message("【ensembl】", id1)
        R <- switch(id2, external_gene_name = local({
            a1 <- hs.hsgly("ens,symbol.37_09_14_164105")(org1, 
                ..mem = 0)
            if (0) {
                wk.dt.subset.na(a1, "ENSG00000102125")
            }
            setnames(a1, c("ensembl_gene_id", id2))
        }), hs.browser("37.09.14.183816"))
        if (0) {
            R <- hs.hsgly("ensembl.getBM.35.05.01.101614")(version = version, 
                attributes = c("ensembl_gene_id", id2))
        }
        wk.dt.fwrite(R, hs.wj(wj_out))
        R
    }
})
"36_01_27_001159" <- local({
    hs1 <- function(t1) switch(t1, symbol = "external_gene_name", 
        ensembl = "ensembl_gene_id", name = "description", uniprot = "uniprotswissprot", 
        hgnc = "hgnc_id", alias = "external_synonym", ncbi = "entrezgene_id", 
        hs.browser("36.06.08.151659"))
    db <- list()
    function(dt1 = {
    }, cn1, cn2, ft, tt, gv = if (length(org1)) 
        hs.hsgly("biomaRt.datatset.version_wanted.35.04.30.214633")(org1)
    else "grch38", org1 = {
    }, sep = ";;;;") {
        hs.browser("37.03.24.190613")
        cn2_regular <- switch(cn2, name = "gene.name", symbol = "gene.id_symbol", 
            ensembl = "gene.id_ensembl", cn2)
        if (length(dt1) == 0) 
            return()
        if (gv %not.in% names(db)) 
            db[gv] <<- list(hs.hsgly("ensg.symbol.name.alias.ncbi.35.05.01.094450")(gv))
        if (is.vector(dt1)) {
            dt1 <- data.table(dt1)
            setnames(dt1, cn1)
        }
        cn1_db <- hs1(ft)
        cn2_db <- hs1(tt)
        if ("uniprotswissprot" %in% c(cn1_db, cn2_db)) {
            db1 <- hs.hsgly("ensembl.id.36.05.27.060241")("uniprot", 
                version = gv)
            db[[gv]] <<- merge(db[[gv]], db1, by = "ensembl_gene_id")
        }
        if (0) 
            if ("external_synonym" %in% c(cn1_db, cn2_db)) {
                db1 <- hs.hsgly("ensembl.id.36.05.27.060241")("alias", 
                  version = gv)
                db[[gv]] <<- merge(db[[gv]], db1, by = "ensembl_gene_id")
            }
        a1 <- wk.dt.subset(db[[gv]], dt1[[cn1]], cn1_db)[, unique(.SD), 
            .SDcols = c(cn1_db, cn2_db)]
        if (anyDuplicated(a1[[cn1_db]])) 
            a1 <- a1[, paste(sort(unique(get(cn2_db))), collapse = sep), 
                by = c(cn1_db)]
        setnames(a1, c(cn1, cn2_regular))
        dt1[, `:=`(c(cn2), wk.dt.subset.na(a1, get(cn1), cn1)[[cn2_regular]])]
        if (cn2 == "name") {
            hs.browser("36.05.27.053700")
            dt1[, `:=`(c(cn2), hs.sub(get(cn2), " *\\[[^\\[\\]]+\\]$"))]
        }
        dt1
    }
})
"36_05_26_232037" <- function(tbl, cn1 = names(tbl)[1], ft = if (hs.grepl(tbl[[cn1]][1], 
    "^(?i)ens[a-z]*g")) "ensembl", tt = c("symbol", "name", "alias", 
    "ensembl", "ncbi") %not% ft, put.to.front = 1, gv = "grch38") {
    t1 <- tt[3]
    for (t1 in tt) {
        cn2 <- switch(t1, symbol = "gene.id_symbol", name = "gene.name", 
            alias = "gene.id_alias", ensembl = "gene.id_ensembl", 
            ncbi = "gene.id_ncbi")
        if (cn2 %in% names(tbl)) 
            next
        hs.hsgly("gene.id.converter.using_ensembl.36.01.27.001159")(tbl, 
            cn1, cn2, ft, t1, gv = gv)
    }
    if (put.to.front) 
        setcolorder(tbl, wk.dt.get.j(tbl, grep_or = "^gene\\."))
    tbl
}
"35_06_11_201251" <- function(version = hs.hsgly("biomaRt.datatset.version_wanted.35.04.30.214633")()) {
    wj_out <- .mfp("35.06.11.201347.tsv.gz", "~/in2/swxx/sj/db/ensembl")
    if (file.exists(wj_out)) {
        fread(wj_out)
    }
    else {
        R <- hs.hsgly("ensembl.getBM.35.05.01.101614")(version = version, 
            attributes = c("ensembl_gene_id", "gene_biotype"))
        fwrite(R, wj_out, sep = "\t")
        R
    }
}
"38_08_20_025839" <- function(org1 = "sheep") {
    wjj0.38_08_20_030335 <- .mfp.swxx("38_08_20_030342")
    if (!length(org1)) {
        wjj0.38_08_20_030335 %>% .ls
        return()
    }
    wjj2 <- wjj0.38_08_20_030335 %>% hs.fp(org1)
    wj <- wjj2 %>% hs.wjj.ls
    if (hs.is_empty(wj)) {
        message("[2022_08_20_112621] 因为网络慢，所以选择手动从biomart获取2列的表然后传到", 
            wjj2)
        wjj2
    }
    else wj
}
"36_09_12_162116" <- function(version = hs.hsgly("biomaRt.datatset.version_wanted.35.04.30.214633")()) {
    wj_out <- .mfp("36.09.12.162136.tsv.gz", "~/swxx/sj/db/ensembl")
    return(fread(wj_out))
    if (0) {
        R <- hs.hsgly("ensembl.getBM.35.05.01.101614")(version = version, 
            attributes = c("ensembl_gene_id", "go_id"))
        wk.dt.fwrite(R, wj_out)
    }
}
"36_09_12_203510" <- local({
    ens.go <- {
    }
    go <- {
    }
    function(ensg, go.domain = {
    }) {
        if (!length(ens.go)) 
            ens.go <<- hs.hsgly("ensg.go.36.09.12.162116")()
        if (!length(go)) 
            go <<- hs.hsgly("go.36.09.12.155114")()
        if (!length(go.domain)) 
            go.domain <- c("bp", "cc", "mf")
        id.desc <- if (uniqueN(go.domain) < 3) {
            go.domain <- c(bp = "biological_process", cc = "cellular_component", 
                mf = "molecular_function")[go.domain]
            go[(go_id != "") & (namespace_1003 %in% go.domain), 
                hs.c(go_id, name_1006)]
        }
        else {
            go[go_id != "", hs.c(go_id, name_1006)]
        }
        jg <- ens.go[(ensembl_gene_id %in% ensg) & (go_id %in% 
            names(id.desc))]
        a1 <- wk.dt.subset(go, jg[, go_id], "go_id")
        jg[, `:=`(names(a1), a1)]
        setnames(jg, "ensembl_gene_id", "gene.id_ensembl")
        setnames(jg, "namespace_1003", "go_domain")
        setnames(jg, "name_1006", "go_description")
    }
})
"35_05_01_094450" <- function(version = hs.hsgly("biomaRt.datatset.version_wanted.35.04.30.214633")(), 
    redo = 0, ..mem = 1) {
    wj_out <- .mfp(hs.wjm("35.05.01.094633.tsv.gz", append = paste0(".v=", 
        version), extN = 2), hs.hsgly("wjj.ensembl.35.10.09.102213")())
    hs.browser("37.03.24.190709")
    ..wj_old <- hs.hsgly("ensembl.gene.35.04.30.160024")(version, 
        ..only.wj_out = 1)
    if (redo) 
        hs.wj.rm(wj_out)
    ..meta.wj_out
    ensg.symbol <- hs.hsgly("ensembl.id.36.05.27.060241")("symbol", 
        version = version, redo = 0)
    ensg.name <- hs.hsgly("ensembl.id.36.05.27.060241")("name", 
        version = version, redo = 0)
    ensg.ncbi <- hs.hsgly("ensembl.id.36.05.27.060241")("ncbi", 
        version = version, redo = 0)
    ensg.alias <- hs.hsgly("ensembl.id.36.05.27.060241")("alias", 
        version = version, redo = 0)
    R <- merge(ensg.symbol, ensg.name, by = "ensembl_gene_id")
    R %<>% merge(ensg.ncbi, by = "ensembl_gene_id")
    R %<>% merge(ensg.alias, by = "ensembl_gene_id")
    if (0) 
        R <- hs.hsgly("ensembl.getBM.35.05.01.101614")(version = version, 
            attributes = c("ensembl_gene_id", "external_gene_name", 
                "description"))
    wk.dt.fwrite(R, hs.wj(wj_out))
    R
}
"38_09_05_104722" <- function(org1 = "human") {
    wjj2.38_09_05_104918 <- .mfp.swxx("38_09_05_104926")
    wj2 <- hs.fp(wjj2.38_09_05_104918, org1) %>% hs.wjj.ls
    if (length(wj2)) {
    }
    else {
        message("须从ensembl biomart获取\"ncbi,ens基因号\"到", 
            hs.fp(wjj2.38_09_05_104918, org1, ""))
        return()
        if (0) {
            wk.rsync.out("~/tmp/mart_export.txt.gz", "~/swxx/xm/db/38/09/05/38_09_05_104926/sheep", 
                ssh = "n9")
        }
    }
    wj2
}
"36_09_12_215754" <- function(version = hs.hsgly("biomaRt.datatset.version_wanted.35.04.30.214633")()) {
    wj_out <- .mfp("36.09.12.215814.tsv.gz", "~/swxx/sj/db/ensembl")
    return(fread(wj_out))
    if (0) {
        R <- hs.hsgly("ensembl.getBM.35.05.01.101614")(version = version, 
            attributes = c("ensembl_gene_id", "reactome"))
        wk.dt.fwrite(R, wj_out)
    }
}
"36_09_12_220739" <- local({
    ens.reactome <- reactome <- {
    }
    function(ensg) {
        if (!length(ens.reactome)) 
            ens.reactome <<- hs.hsgly("ensg.reactome.36.09.12.215754")()
        if (!length(reactome)) 
            reactome <- hs.hsgly("id.描述.36.09.12.221334")()
        id.desc <- reactome[, hs.c(reactome_id, reactome_description)]
        jg <- ens.reactome[(ensembl_gene_id %in% ensg) & (reactome %in% 
            names(id.desc))]
        a1 <- wk.dt.subset(reactome, jg[, reactome], "reactome_id")[, 
            .(reactome_description)]
        jg[, `:=`(names(a1), a1)]
        jg[, `:=`("reactome", {
        })]
    }
})
"38_09_02_160626" <- function(org1 = "sheep") {
    wjj0.38_09_02_160925 <- .mfp.swxx("38_09_02_160921")
    wjj2 <- wjj0.38_09_02_160925 %>% hs.fp(org1)
    wj <- wjj2 %>% hs.wjj.ls
    if (hs.is_empty(wj)) {
        message("[38_09_02_160845] 因为网络慢，所以选择手动从biomart获取2列的表然后传到", 
            wjj2)
        wjj2
    }
    else wj
}
"36_09_12_155114" <- function(version = hs.hsgly("biomaRt.datatset.version_wanted.35.04.30.214633")()) {
    wj_out <- .mfp("36.09.12.155745.tsv.gz", "~/swxx/sj/db/ensembl")
    return(fread(wj_out))
    if (0) {
        R <- hs.hsgly("ensembl.getBM.35.05.01.101614")(ens_gene = {
        }, version = version, attributes = c("namespace_1003", 
            "go_id", "name_1006"))
        wk.dt.fwrite(R, wj_out)
    }
}
"36_10_23_133502" <- function(org1 = "human", wj1 = hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1 = org1), 
    ot = "gene") {
    wj2 <- hs.wjm(wj1, ext = "gene.tsv.gz")
    if (wj.update.reporter(wj2, wj1)) {
        tbl <- wk.fread(hs.hsgly("gtf_gene.37_01_18_121321")(org1))
        tbl[, `:=`("gene.id_ensembl", hs.re(V9, "(?<=gene_id \")[^\"]+"))]
        tbl[, `:=`("gene.id_symbol", hs.re(V9, "(?<=gene_name \")[^\"]+", 
            no_match = ""))]
        tbl[, `:=`("gene.biotype", hs.re(V9, "(?<=gene_biotype \")[^\"]+"))]
        tbl[, `:=`(c("V3", "V6", "V8", "V9"), {
        })]
        setnames(tbl, paste0("V", c(1, 2, 4, 5, 7)), c("chr", 
            "source", "start", "end", "strand"))
        wk.dt.fwrite(tbl, wj2)
    }
    wj3 <- hs.wjm(wj1, ext = "gene_stat.tsv.gz")
    if (wj.update.reporter(wj3, wj2)) {
        tbl <- wk.fread(wj2)
        a1 <- tbl[, .N, by = gene.biotype]
        setorderv(a1, "N", -1)
        wk.dt.fwrite(a1, wj3)
    }
    switch(ot, gene = wj2, stat = wj3)
}
"36_03_25_094459" <- function() {
    gene.gtf.wj <- "~/in2/swxx/sj/db/ensembl/human/Homo_sapiens.GRCh38.99.gtf"
    gene_repeat.gtf.wj <- hs.wjm(gene.gtf.wj, append = "+repeat,non_overlapping")
    return(gene_repeat.gtf.wj)
    gene.gtf <- fread(gene.gtf.wj, header = FALSE)
    repeat.gtf <- "~/in2/swxx/sj/db/repeats/ucsc_gonomeBrowser/human/hg38.repeats.gtf" %>% 
        fread
    repeat.gtf[, `:=`("V1", hs.sub(V1, "chr"))]
    repeat.gtf[V1 == "M", `:=`("V1", "MT")]
    exon.gr <- gene.gtf[V3 == "exon", ieiwk.GRanges(V1, start = V4, 
        end = V5)]
    repeat.gr <- repeat.gtf[, ieiwk.GRanges(V1, start = V4, end = V5, 
        I = repeat.gtf[, seq.int(.N)])]
    {
        repeat.1.gr <- setdiff(repeat.gr, exon.gr)
        repeat.1.gr %>% width %>% sum
        exon.gr %>% reduce %>% width %>% sum
        repeat.1.gtf <- repeat.1.gr %>% {
            data.table(V1 = as.vector(seqnames(.)), V2 = "hg38_rmsk", 
                V3 = "exon", V4 = start(.), V5 = end(.), V6 = ".", 
                V7 = "+", V8 = ".", V9 = "gene_id \"repeat_nonoverlapping\"; transcript_id \"repeat_nonoverlapping\"; exon_id \"repeat_nonoverlapping\";")
        }
        wk.dt.fwrite(gene.gtf, gene_repeat.gtf.wj, col.names = FALSE, 
            append = FALSE)
        wk.dt.fwrite(repeat.1.gtf, gene_repeat.gtf.wj, col.names = FALSE, 
            append = TRUE)
        gene_repeat.gtf.wj %>% .ls1
        gene_repeat.gtf.wj %>% .tail
    }
}
"36_04_21_121618" <- function() {
    gtf.wj <- hs.hsgly("gene+repeat,non_overlapping.gtf.wj.36.03.25.094459")()
    jg.wj <- hs.wjm(gtf.wj, add = "gene_length.tsv.gz")
    return(fread(jg.wj))
    nr <- fread(text = .wk("cat", gtf.wj, "| gawk -v FS=\"\t\" -v OFS=\"\t\"", 
        .q("{ if( $3 == \"exon\") { split( $9, A, \"; \"); print( $1, $4, $5, A[ 1], A[ 6])}}"), 
        intern = 1), sep = "\t", header = !1)
    ensg.symbol <- local({
        a1 <- unique(nr[, .(V4, V5)])
        a1[V5 == "", `:=`("V5", V4)]
        a1[, hs.c(V4 %>% hs.re("(?<=\")[^\"]+"), V5 %>% hs.re("(?<=\")[^\"]+"))]
    })
    nr %>% tail
    nr[, `:=`("V4", V4 %>% hs.re("(?<=\")[^\"]+"))]
    nr[V5 == "", `:=`("V5", V4)]
    nr[V4 != "repeat_nonoverlapping", `:=`("V5", V5 %>% hs.re("(?<=\")[^\"]+"))]
    L <- nr[, {
        sum(width(reduce(ieiwk.GRanges(V1, start = V2, end = V3))))
    }, by = V4]
    L[, V4 %>% duplicated %>% any]
    L[, V1 %>% min]
    setnames(L, c("gene.id_ensembl", "length"))
    L[gene.id_ensembl %in% names(ensg.symbol), `:=`("gene.id_symbol", 
        ensg.symbol[gene.id_ensembl])]
    wk.dt.fwrite(L, jg.wj)
    jg.wj %>% .ls1
}
"36_04_21_091051" <- function(unique = FALSE, gv = "grch38", 
    release = "110", ..mem = 1) {
    hs.browser("2023.08.04.190952", debug = 0)
    gtf.wj <- switch(gv, grch38 = paste0("~/swxx/sj/db/ensembl/ftp.ensembl.org/pub/release-", 
        release, "/gtf/homo_sapiens/Homo_sapiens.GRCh38.", release, 
        ".gtf.gz"), grch37 = "~/in2/swxx/sj/db/ensembl/human/Homo_sapiens.GRCh37.87.gtf", 
        grcm38 = "~/in2/swxx/sj/db/ensembl/mouse/Mus_musculus.GRCm38.101.gtf", 
        hs.browser("36.05.26.204957"))
    jg.wj <- local({
        Arg <- list(gtf.wj)
        a <- if (unique) 
            "gene_length,unique.tsv.gz"
        else "gene_length.tsv.gz"
        if (hs.wj.is_compressed(gtf.wj)) {
            Arg[["ext"]] <- a
        }
        else Arg[["add"]] <- a
        do.call("hs.wjm", Arg)
    })
    if (file.exists(jg.wj)) {
        wk.fread(jg.wj)
    }
    else {
        if (0) {
            nr1 <- fread(text = readLines(gtf.wj, 9))
            nr1 <- fread(text = .wk("cat", gtf.wj, "| head", 
                "| gawk -v FS=\"\t\" -v OFS=\"\t\"", .q("{ if( $3 == \"exon\") { split( $9, A, \"; \"); print( $1, $4, $5, A[ 1], A[ 6])}}"), 
                intern = 1), sep = "\t", header = !1)
            nr1[1:3]
        }
        nr <- wk.fread(cmd = .sys("cat_wk", gtf.wj, "| gawk -v FS=\"\t\" -v OFS=\"\t\"", 
            .q("{ if( $3 == \"exon\") { split( $9, A, \"; \"); print( $1, $4, $5, A[ 1], A[ 6])}}")), 
            sep = "\t", header = FALSE)
        nr <- .sys.cat.text(gtf.wj) %=>% .sys("| gawk -v FS=\"\t\" -v OFS=\"\t\"", 
            .q("{ if( $3 == \"exon\") { split( $9, A, \"; \"); print( $1, $4, $5, A[ 1], A[ 6])}}")) %=>% 
            wk.fread(cmd = x, sep = "\t", header = FALSE)
        .sys.cat.skip(gtf.wj) %=>% .wk("| perl -e", .q("\nwhile( <>) {\n  print;\n}\n"), 
            "| head")
        nr <- .sys.cat.text(gtf.wj) %=>% .sys("| perl -e", .q("\n$, = \"\t\";\nwhile( <>) {\n  @F = split( \"\t\");\n  if( $F[ 2] eq \"exon\") {\n    @A = split( \"; \", $F[ 8]);\n    print( $F[ 0], $F[ 3], $F[ 4], $A[ 0], \"$A[ 5]\n\");\n  }\n}\n")) %=>% 
            wk.fread(cmd = x)
        if (0) {
            nr[, .N]
            .wk("cat", gtf.wj, "| gawk", .q("{ if( $3 == \"exon\") print}"), 
                "| wc -l")
            nr[, V5 %>% tail]
        }
        if (0) {
            ensg.symbol <- local({
                a1 <- nr[, unique(V4)]
                a1 %<=>% nr[match(x, V4), .(V4, V5)]
                a1[V5 == "", `:=`("V5", V4)]
                a1[, hs.c(V4 %>% hs.re("(?<=\")[^\"]+"), V5 %>% 
                  hs.re("(?<=\")[^\"]+"))]
            })
        }
        gtf <- hs.hsgly("gtf/基因信息.38_09_08_204013")(gtf.wj)
        ensg.symbol <- gtf[, hs.c(gene_id, gene)]
        L <- nr[, {
            gr <- ieiwk.GRanges(V1, start = V2, end = V3)
            L <- wk.mclapply(unique(V4), . %=>% {
                i1 <- which(V4 == x)
                a1 <- gr[i1]
                if (unique) 
                  a1 <- setdiff(a1, gr[-i1])
                a1 <- IRanges::reduce(a1)
                sum(width(a1))
            })
            hs.c(names(L) %>% hs.re("(?<=\")[^\"]+"), unlist(L))
        }, by = V1]
        if (0) {
            a1 <- fread("~/in2/swxx/xm/jz/dw/sage1/pdx,cells/邓玮X101SC19113319-Z03-J001-B1-39_20191210/supp/3.Expression/Read_counts.293t.gene+repeat,non_overlapping.grch38.gene.gene_id.tsv.gz", 
                select = 2)
            a1[[1]] %>% unique %>% strsplit(";") %>% unlist %>% 
                unique %>% .sh
        }
        genes <- nr[, {
            unique(V4) %>% hs.re("(?<=\")[^\"]+")
        }, by = V1][[2]]
        jg <- tapply(L[[2]], genes, sum)
        jg <- data.table(gene.id_ensembl = names(jg), gene.id_symbol = ensg.symbol[names(jg)], 
            length = jg)
        wk.dt.fwrite(jg, jg.wj, add = 0)
        jg
    }
}
"37_01_18_121321" <- function(gtf.wj = hs.hsgly("ensembl/获取数据.37_09_13_135206")()) {
    if (!file.exists(gtf.wj)) 
        stop("37_01_18_121908")
    wj2 <- hs.wjm(gtf.wj, append = ".gene", extN = 2)
    if (!file.exists(wj2)) {
        .wk(.sys.cat.skip(gtf.wj, cn = 0), "|", .sys.perl.print("print if $F[ 2] eq \"gene\""), 
            "| gzip >", .sys.fpGood(wj2))
    }
    wj2
}
"36_12_21_002840" <- function(org1 = "human") {
    wj <- hs.hsgly("ftp.ensembl.org/pub/current_mysql.37_04_26_164820")(org1, 
        what = "stable_id_event")
    jg.wj <- hs.wjm(wj, ext = "old_new.tsv.gz", extN = 2)
    if (wj.update.reporter(jg.wj, wj)) 
        local({
            cmd1 <- .sys("perl -e", .q(paste0("\n$i1 = 0;\nopen( fp1, ", 
                .q(.sys.cat.text(hs.hsgly("ftp.ensembl.org/pub/current_mysql.37_04_26_164820")(org1, 
                  what = "mapping_session"), pipe = 1)), ");\n$i1 = 0;\nwhile( <fp1>) {\n  $i1 = $i1 + 1;\n  /\\d+/;\n  $mapping_session_id_order{ $&} = $i1;\n}\nclose( fp1);\nopen( fp1, ", 
                .q(.sys.cat.text(hs.hsgly("ftp.ensembl.org/pub/current_mysql.37_04_26_164820")(org1, 
                  what = "gene"), pipe = 1)), ");\nwhile( <fp1>) {\n  @F = split/\\t/;\n  $gene_current{ $F[ 12]} = 1;\n}\nclose( fp1);\nopen( fp1, ", 
                .q(.sys.cat.text(wj, pipe = 1)), ");\nwhile( <fp1>) {\n  @F = split( /\\t/);\n  if( $F[ 5] eq \"gene\") {\n    if( $F[ 0] ne \"\\\\N\") {\n      if( ( ! exists( $gene_history{ $F[ 0]})) || ( $mapping_session_id_order{ $F[ 4]} > $gene_mapping_session_id_order{ $F[ 0]})) {\n        $gene_history{ $F[ 0]} = $F[ 2];\n        $gene_mapping_session_id_order{ $F[ 0]} = $mapping_session_id_order{ $F[ 4]};\n      } elsif ( $mapping_session_id_order{ $F[ 4]} == $gene_mapping_session_id_order{ $F[ 0]}) {\n        $gene_history{ $F[ 0]} .= \";$F[ 2]\";\n      }\n    }\n  }\n}\nclose( fp1);\nsub sub1 {\n  my @gene = split( /;/, $_[ 0]);\n  my $i1 = 0;\n  my $vl = @gene;\n  while( $i1 < $vl) {\n    if( $gene[ $i1] ne \"\\\\N\") {\n      if( ! exists( $gene_current{ $gene[ $i1]})) {\n        $gene[ $i1] = $gene_history{ $gene[ $i1]};\n      }\n    }\n    $i1++;\n  }\n  join( \";\", @gene);\n}\nforeach $k1 ( keys %gene_history) {\n  $gene2 = $gene_history{ $k1};\n  LOOP1: while( 1) {\n    $jg = sub1( $gene2);\n    if( $gene2 eq $jg) {\n      $gene_history{ $k1} = $gene2;\n      last LOOP1;\n    } else {\n      $gene2 = $jg;\n    }\n  }\n}\nforeach $k1 ( keys %gene_history) {\n  foreach $v1 ( split( /;/, $gene_history{ $k1})) {\n    if( ( $v1 ne \"\\\\N\") && ( $v1 ne $k1) && exists( $gene_current{ $v1})) {\n      print( $k1, \"\\t\", $v1, \"\\n\");\n    }\n  }\n}\n")), 
                "| sort | uniq | perl -e", .q("print( \"old\\tnew\\n\"); while( <>) { print;}"), 
                "| gzip >", jg.wj)
            .wk(cmd1)
            return()
            jg.wj %>% .head
            jg.wj %>% .wc
            jg.wj %>% .file
            jg.wj %>% .ls
            jg <- wk.fread(cmd = cmd1, cn = 0)
            if (0) {
                jg[, hs.grepl(unique(V1), "^ENS") %>% mean]
                jg_old <- wk.fread(jg.wj)
                jg[, V1 %>% hs.grepV("^ENS")] %xor% jg_old[, 
                  old]
                {
                  jg_old[old %in% "ENSG00000124732"]
                  jg[V1 %in% "ENSG00000124732"]
                  .wk(.sys.cat.text(wj), "| perl -e", .q("while( <>) {\nif( /ENSG00000124732/) {\n  print;\n}\n}\n"))
                }
            }
            jg_old[, old %>% uniqueN] - length(jg_old[, old] %and% 
                jg[, V1])
            if (0) {
                gtf.wj <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1)
                gtf9 <- fread(cmd = .sys("zcat", gtf.wj, "|", 
                  .sys.perl.print("print $F[ 8] if $F[ 2] eq \"gene\"")))
                ens_id2do <- gtf9[, hs.re(V1, "ENSG\\d+")]
            }
            ens_id2do <- hs.hsgly("ftp.ensembl.org/pub/current_mysql.37_04_26_164820")(org1, 
                what = "gene", out = "data")
            ens_id.db <- ens_id2do
            a1 <- wk.fread(cmd = .sys(.sys.cat.text(wj), "| perl -e", 
                .q("while( <>) {\n@F = split/\\t/;\nif( $F[ 5] eq \"gene\") {\n  print;\n}\n}")))
            it1 <- 1
            repeat {
                id_extra <- a1[(V1 %in% ens_id.db) | (V3 %in% 
                  ens_id.db), V1 %not% ens_id.db]
                if (length(id_extra)) {
                  message(it1)
                  it1 <- it1 + 1
                  ens_id.db %<>% c(id_extra)
                }
                else break
            }
            a1 <- a1[(V1 %in% ens_id.db) | (V3 %in% ens_id.db)]
            ens_id.db <- a1[, V1 %or% V3]
            mapping_session <- hs.hsgly("ftp.ensembl.org/pub/current_mysql.37_04_26_164820")(org1, 
                what = "mapping_session", out = "content")
            if (!a1[, V5 %all.in% mapping_session[, V1]]) 
                hs.browser("37.04.27.103032")
            record1 <- ens_id.db[-1] %>% {
                hs.c(., rep(NA, length(.)))
            }
            id2.hs.id1 <- function(id1) {
                counter1[id1] <<- counter1[id1] + 1
                a2 <- a1[V1 %in% id1]
                if (a2[, .N]) {
                  v5.latest <- tail(mapping_session[, V1] %and% 
                    a2[, V5], 1)
                  a2 <- a2[(V5 %in% v5.latest) & (V3 %in% ens_id.db)]
                  if (a2[, .N == 1]) {
                    if (a2[, V3 == id1]) 
                      return(id1)
                    if (a2[, V3 == "\\N"]) 
                      return("")
                    id2.hs.id1(a2[, V3])
                  }
                  else {
                    if (a2[, all(V3 == "\\N")]) 
                      return("")
                    sapply(a2[, V3 %not% "\\N"], id2.hs.id1) %>% 
                      hs.clean(U = 1) %>% sort %>% paste(collapse = ";")
                  }
                }
                else {
                  a2 <- a1[(V3 %in% id1)]
                  v5.latest <- tail(mapping_session[, V1] %and% 
                    a2[, V5], 1)
                  a2 <- a2[V5 %in% v5.latest]
                  if (a2[, all(V3 == id1)]) 
                    return(id1)
                  hs.say(id1)
                  stop("36.12.20.152053")
                }
            }
            jg <- lapply(2:length(ens_id.db), function(id1.i) {
                id1 <- ens_id.db[id1.i]
                id2.hs.id1(id1)
                hs.table(counter1)
                i1 <- which.max(counter1)
                if (counter1[i1] > 1) {
                  hs.browser("37.04.27.105600")
                  id2.hs.id1("ENSMUSG00000070003")
                }
            })
            a2 <- unlist(jg)
            names(a2) <- ens_id.db[-1]
            a3 <- strsplit(a2, ";") %>% lapply("%and%", ens_id2do) %>% 
                hs.nonempty %>% sapply(paste, collapse = ";")
            jg <- a3[names(a3) != a3] %>% strsplit(";") %>% wk.dt.fromList
            wk.dt.setnames(jg, c("old", "new"))
            wk.dt.fwrite(jg, jg.wj)
        })
    jg.wj
}
"37_09_14_150617" <- function(org1 = "human") {
    wj_in <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1, 
        what = "ncbi_gene_id")
    a1 <- wk.fread(cmd = .sys(.sys.cat.skip(wj_in), "| cut -f1,4", 
        "| uniq", "| tail -n +2", "| sort | uniq"), cn = 0)
    setnames(a1, paste0("gene_id,", c("ensembl", "ncbi")))
}
"37_09_14_164105" <- function(org1 = "human") {
    wj_in <- hs.hsgly("ensembl/获取数据.37_09_13_135206")(org1, 
        what = "gtf")
    wj_out <- hs.wjm(wj_in, append = ".ens,symbol")
    if (file.exists(wj_out)) {
        wk.fread(wj_out)
    }
    else {
        if (0) {
            wj_in %>% .ls1
            wj_in %>% .less
            .wk("zcat", wj_in, "|", .perl.e("\nif( $F[ 2] eq \"gene\") {\n  print;\n}\n", 
                sys = 1, a = 1), "| less -Ni")
        }
        cmd1 <- .sys(.sys.cat.text(wj_in), "| perl -e", .q("\nwhile( <>) {\n  @F = split( \"\t\");\n  if( $F[ 2] eq \"gene\") {\n    if( $F[ 8] =~ /gene_name \"([^\"]*)/) {\n      $c1 = $1;\n      $F[ 8] =~ /gene_id \"([^\"]*)/;\n      print( $1, \"\t\", $c1, \"\n\");\n    }\n  }\n}"))
        a1 <- wk.fread(cmd = cmd1, cn = 0)
        a1[, `:=`(V2, toupper(V2))]
        setnames(a1, paste0("gene_id_", c("ensembl", "symbol")))
        wk.dt.fwrite(a1, wj_out)
        a1
    }
}
"37_09_14_140925" <- function(ver1 = "in_use", ..mem = 1) switch(ver1, 
    in_use = 104, current = local({
        a1 <- readLines("http://ftp.ensembl.org/pub/current_README", 
            3)
        hs.re(a1, "\\d+")
    }))
"37_01_25_194601" <- function(..mem = 1) {
    wjj <- .mfp.swxx("37_01_25_194716")
    dz1 <- "http://www.ensembl.org/info/about/species.html"
    wj2 <- hs.wjm(dz1, add = "gz", dn = wjj)
    html <- wj2 %=>% {
        if (file.exists(x)) {
            readLines(x)
        }
        else {
            a <- readLines(dz1)
            cat(a, sep = "\n", file = .sys("| gzip >", .sys.fpGood(hs.wj(x))))
            a
        }
    }
    i <- hs.grep(html, "^ +<table\\b")
    a <- html[i + 2] %=>% strsplit("</tr>", fixed = 1) %=>% unlist
    a[1] %<=>% hs.sub(x, "^\\s*<tbody>")
    a <- hs.sub(a, "</tbody>")
    a %<=>% trim
    a %<=>% hs.clean
    a %<=>% paste0(x, "</tr>")
    x <- lapply(a, . %=>% {
        xml2::read_xml(x) %=>% xml2::as_list(x)[[1]]
    })
    if (0) {
        sapply(x, length) %=>% hs.table
    }
    ftp_fp <- sapply(x, . %=>% attr(x[[1]][[2]], "href")) %=>% 
        trim(x, "/")
    ftp_fp %=>% anyDuplicated
    c1 <- sapply(a, hs.re, "(?<=>)[^<]+") %=>% unname
    c2 <- sapply(x, . %=>% unlist(x[[2]]))
    c3 <- sapply(x, . %=>% unlist(x[[3]]))
    c4 <- sapply(x, . %=>% unlist(x[[4]]))
    c5 <- sapply(x, . %=>% unlist(x[[5]]))
    c6 <- sapply(x, . %=>% {
        x <- x[[6]]
        if ("span" %in% names(x)) {
            x[[1]][[2]]
        }
        else unlist(x)
    })
    c7 <- sapply(x, . %=>% unlist(x[[7]]))
    c8 <- sapply(x, . %=>% unlist(x[[8]]))
    jg <- mget(paste0("c", 1:8)) %=>% wk.dt.as
    wk.dt.setnames(jg, html[i + 1] %=>% hs.gre("(?<=>)[^<]+") %=>% 
        unlist)
    jg[, `:=`("ftp_fp", ftp_fp)]
    if (0) {
        wk.rsync.out("/Users/wk/Downloads/Species.csv", wjj)
        hs.wj.move("/Users/wk/Downloads/Species.csv", hs.wjj(wjj))
        .ls(wjj)
    }
    if (0) {
        wj1 <- hs.fp(wjj, "Species.csv")
        note.update_time(wj2)
        jg <- wk.fread(wj1)
    }
    for (cn1 in c("Common name", "Scientific name", "ftp_fp")) jg[, 
        `:=`(c(cn1), tolower(get(cn1)))]
    jg
}
"38_04_09_142327" <- function(gv1) {
    hs.browser("38.04.09.152736")
    release_n <- hs.hsgly("生物学/数据库/ensembl/最新版本号.37_12_30_165505")()
    dz1 <- paste0("http://ftp.ensembl.org/pub/release-", release_n, 
        "/species_EnsemblVertebrates.txt")
    dt <- wk.fread(hs.hsgly("ensembl/url2fp.38_04_09_140039")(dz1, 
        save = 1))
    cn <- names(dt)[-1]
    cn[1] %<>% hs.sub("^#")
    dt <- dt[, .SD, .SDcols = -ncol(dt)]
    wk.dt.setnames(dt, cn)
    gv1
    dt[hs.grep(assembly, "(?i)grc[hm]\\d+")]
}
"36_03_20_155420" <- local({
    org.gv <- {
    }
    function(org1 = "human", gv1 = {
    }) {
        if (!length(org.gv)) 
            org.gv <<- list(human = c("grch38", "grch37"), mouse = c("grcm39", 
                "grcm38"), yeast = "r64-1-1") %=>% wk.dt.fromList(nm = c("org", 
                "gv"))
        if (length(gv1)) {
            org.gv[match(tolower(gv1), gv), org]
        }
        else org.gv[match(tolower(org1), org), gv]
    }
})
"38_09_06_110337" <- function(org1 = "黑山羊") {
    hs.switch(org1, "黑山羊", "goat (black bengal)", "山羊", 
        "goat", org1)
}
"38_09_08_100005" <- function(x) {
    v1 <- hs.hsgly("物种/中文名.common_name.38_09_06_110337")(x)
    hs.hsgly("ensembl/物种/common_name.ncbi_tax_id.38_09_08_095424")(v1)
}
"38_09_08_095424" <- function(common_name) {
    common_name <- hs.switch(common_name, "山羊", "goat", "黑山羊", 
        "goat (black bengal)", "yeast", "s. cerevisiae", common_name)
    db <- hs.hsgly("列表.37_01_25_194601")()
    db[hs.str_in(`Common name`, common_name), `Taxon ID`]
}
"37_12_30_165505" <- function() {
    dz1 <- "http://ftp.ensembl.org/pub/current_gtf/homo_sapiens/CHECKSUMS"
    (hs.hsgly("ensembl/url2fp.38_04_09_140039")(dz1)) %=>% dirname %=>% 
        hs.wjj %=>% hs.with_wd({
        wj1 <- wj.updater(dz1 = dz1, co1 = 30)
        nr <- wk.fread(wj1)
        nr[, hs.re(V3, "\\.(\\d+).gtf.gz")[1]] %=>% hs.re("\\d+")
    })
}
"38_04_09_140039" <- function(url1, save = 0) {
    wjj_ensembl <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("ensembl")
    wj2 <- hs.fp(wjj_ensembl, hs.sub(url1, "(ftp|https?)://"))
    if (save) {
        if (hs.grepl(url1, "/$")) {
            url1 %<>% hs.sub("/+$")
            wj2 <- hs.fp(wjj_ensembl, hs.sub(url1, "(ftp|https?)://"), 
                "a.html.gz")
            if (wj.update.reporter(wj2, co1 = 30)) 
                wj.updater(wj2, url1, co1 = 30)
            return(wj2)
        }
        else {
            wj2 <- wj.updater(dz1 = url1, wjj2 = dirname(wj2), 
                co1 = Inf)
            co1 <- max(30, file.size(wj2)/(1000^2)/10)
            if (wj.update.reporter(wj2, co1 = co1)) 
                wj.updater(dz1 = url1, wjj2 = dirname(wj2), co1 = 30)
            return(wj2)
        }
    }
    else return(wj2)
}
"39_02_21_143729" <- function() {
    "~/rj/gdc-client_v1.6.1_Ubuntu_x64/gdc-client"
}
"38_01_17_160209" <- function(org1 = "human", genome_version = switch(org1, 
    human = "hg38")) {
    wjj.38_01_17_160952 <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("gencode")
    dz1 <- switch(genome_version, hg38 = "https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/gencode.v39.annotation.gtf.gz", 
        hg37 = "https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/GRCh37_mapping/gencode.v39lift37.annotation.gtf.gz", 
        hs.browser("38.01.17.160821"))
    wj2 <- hs.fp(wjj.38_01_17_160952, hs.re(dz1, "ftp\\..*"))
    hs.with_wd(dirname(wj2), {
        hs.browser("38.01.17.173220")
        .wk("axel -ac", dz1)
    })
}
"39_01_11_145019" <- function(sub = {
}) {
    jg <- "~/swxx/sj/db/go"
    if (length(sub)) 
        jg <- hs.fp(jg, sub)
    jg
}
"39_01_11_145257" <- function() {
    wjj2 <- hs.sjgly("39_01_11_145257")
    dz1 <- "http://purl.obolibrary.org/obo/go/go-basic.obo"
    wj2 <- hs.wjm(dz1, dn = wjj2)
    if (!file.exists(wj2)) {
        wk.wget(dz1, wjj2, gz = 0)
    }
    wj2
}
"39_01_11_150504" <- function(sub = {
}) {
    jg <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("gprofiler")
    if (length(sub)) 
        jg <- hs.fp(jg, sub)
    jg
}
"37_01_06_195308" <- function() {
    wjj2 <- hs.sjgly("37_01_06_195308")
    {
        hs.with_wd(wjj2, {
            lapply(c("https://biit.cs.ut.ee/gprofiler/static/gprofiler_hsapiens.name.zip", 
                "https://biit.cs.ut.ee/gprofiler/static/gprofiler_hsapiens.ENSG.zip"), 
                function(dz1) {
                  wk.wget(dz1)
                })
        })
        wk.decompress(wjj2)
        hs.fp(wjj2, "gprofiler_hsapiens.name") %=>% hs.wjj.ls(x, 
            "\\.name\\.gmt$") %=>% wk.mclapply(x, . %=>% {
            message(x)
            wj2 <- hs.wjm(x, ext = "ncbi.gmt", extN = 2)
            if (wj.update.reporter(wj2, x)) {
                nr <- readLines(x) %=>% strsplit(x, "\t")
                if (nr[[1]][2] %in% c("CORUM root", "HP root", 
                  "HPA root", "MIRNA root", "WIKIPATHWAYS", "REACTOME root term")) {
                  nr[[1]] <- {
                  }
                }
                id.db <- lapply(nr, tail, -2) %=>% unlist %=>% 
                  hs.clean(x, U = 1) %=>% wk.dt(x, symbol = x)
                if (0) {
                  id.db[old != new]
                  id.db[hs.clean(ncbi, ot = "i", inv = 1)] %=>% 
                    .s
                }
                hs.hsgly("编号转换之表格.37_03_22_145251")(id.db, 
                  "symbol", "ncbi")
                wk.dt.setnames(id.db, c("old", "new", "ncbi"))
                if (0) {
                  id.db[new != old]
                  {
                    id.db[hs.grepl(ncbi, ";")]
                    a <- hs.lapply(nr, . %=>% {
                      a1 <- id.db[match(tail(x, -2), old)]
                      if (a1[, any(hs.grepl(ncbi, ";"))]) 
                        a1
                    }) %=>% hs.nonempty
                    sapply(a, nrow) %=>% max
                  }
                  id.db[hs.clean(ncbi, inv = 1, ot = "i")]
                  wj2 %=>% .ls1
                  wj2 %=>% .head
                }
                hs.makeBatch(seq_along(nr), K = 10) %=>% {
                  sink(file = wj2)
                  on.exit(sink())
                  for (i in x) {
                    a <- sapply(i, function(i) {
                      x <- nr[[i]]
                      ncbi <- tail(x, -2) %=>% wk.dt.subset(id.db, 
                        x, "old", exp_j = unique(ncbi) %=>% hs.strsplit.v(x, 
                          sep = ";") %=>% unique %=>% as.numeric %=>% 
                          sort)
                      paste(c(x[1:2], ncbi), collapse = "\t")
                    })
                    wk.dt.fwrite(a, "")
                  }
                }
            }
        }, mc.cores = 1)
    }
    hs.fp(wjj2, "gprofiler_hsapiens.name") %=>% hs.wjj.ls(x, 
        "\\.ncbi\\.gmt$")
}
"39_05_24_135710" <- function() {
    wjj2 <- hs.sjgly("39_05_24_135710")
    if (length(hs.wjj.ls(wjj2, wj = 1, wjj = 1)) < 2) {
    }
}
"38_01_03_221659" <- function(v1, sep1 = "[;\\|]", sep2 = sep1) {
    jg <- hs.c(names(v1), rep("", length(v1)))
    a1 <- strsplit(v1, sep1)
    a2 <- data.table(new = unlist(a1), i1 = rep(seq_along(v1), 
        sapply(a1, length)))
    if (0) {
        v1["LAP3"]
    }
    a3 <- a2[, {
        old <- names(v1)[i1]
        if (old %in% new) 
            old
        else v1[i1]
    }, by = i1]
    if (a3[, i1 %>% anyDuplicated]) 
        hs.browser("37.09.13.184246")
    jg[a3[, i1]] <- a3[, V1]
    if (0) {
        jg %>% {
            .[names(.) != .]
        }
        jg["LAP3"]
    }
    return(jg)
    jg <- data.table(old = rep(names(v1), sapply(a1, length)), 
        new = unlist(a1))
    jg <- jg[, if (old %in% new) 
        old
    else paste(new, collapse = sep2), by = old]
    jg[, hs.c(old, V1)[names(v1)]]
}
"38_01_03_204741" <- local({
    old.new <- {
    }
    function(v1, recent = 0) {
        if (!length(old.new)) 
            old.new <<- local({
                hgnc <- hs.hsgly("gene.id_hgnc.35.12.06.111314")()
                if (0) {
                  db_hgnc[`HGNC ID` %in% "55804"]
                  db_hgnc[, `Locus type` %>% unique %>% hs.grepV("(?i)site|region|unknown", 
                    inv = 1)]
                  hgnc[, `Locus type` %>% hs.table %>% hs.sort]
                  hgnc[, hs.table(`Locus type`, `Locus group`)] %>% 
                    .s
                  hgnc[, hs.table(`Locus group`)] %>% .s
                  hgnc[, .N, by = `Approved symbol`][, N %>% 
                    hs.table]
                  hgnc[`Approved symbol` %>% {
                    . %in% .[duplicated(.)]
                  }, .(`Approved symbol`, `Previous symbols`, 
                    `Alias symbols`)][, uniqueN(.SD), by = `Approved symbol`][, 
                    V1 %>% uniqueN]
                }
                hgnc.jg1 <- hgnc[, {
                  a1 <- strsplit(`Previous symbols`, ", ")
                  a1 <- lapply(seq.int(.N), function(i1) {
                    c(`Approved symbol`[i1], a1[[i1]])
                  })
                  a2 <- data.table(new = rep(`Approved symbol`, 
                    sapply(a1, length)), old = unlist(a1))
                  {
                    a2[, `:=`(type, rep(`Locus group`, sapply(a1, 
                      length)) %>% hs.factor(c("protein-coding gene", 
                      "non-coding RNA", "pseudogene", "other"), 
                      N = 0))]
                    setorderv(a2, "type")
                  }
                  a2[, paste(unique(new), collapse = "|"), by = old][, 
                    hs.c(old, V1)] %>% {
                    .[. != names(.)]
                  }
                }]
                if (0) {
                  hgnc.jg1 %>% {
                    .[sapply(seq_along(.), function(i1) {
                      names(.)[i1] %in% strsplit(.[i1], ";")[[1]]
                    })]
                  }
                }
                hgnc.jg2 <- hgnc[, {
                  a1 <- strsplit(`Alias symbols`, ", ")
                  a1 <- lapply(seq.int(.N), function(i1) {
                    c(`Approved symbol`[i1], a1[[i1]])
                  })
                  a2 <- data.table(new = rep(`Approved symbol`, 
                    sapply(a1, length)), old = unlist(a1))
                  {
                    a2[, `:=`(type, rep(`Locus group`, sapply(a1, 
                      length)) %>% hs.factor(c("protein-coding gene", 
                      "non-coding RNA", "pseudogene", "other"), 
                      N = 0))]
                    setorderv(a2, "type")
                  }
                  jg <- a2[, paste(unique(new), collapse = "|"), 
                    by = old][old != V1, hs.c(old, V1)]
                  jg[names(jg) %not.in% names(hgnc.jg1)]
                }]
                c(hgnc.jg1, hgnc.jg2) %>% {
                  .[. != names(.)]
                }
            })
        jg <- rep(NA, length(v1))
        i1 <- v1 %in% names(old.new)
        if (any(i1)) 
            jg[i1] <- old.new[v1[i1]]
        if (wk.false %in% i1) {
            i1 <- !i1
            jg[i1] <- v1[i1]
        }
        jg <- hs.c(v1, jg)
        if (recent) 
            jg <- hs.hsgly("新旧并存取新.38_01_03_221659")(jg)
        jg
    }
})
"36_04_12_233708" <- function(..mem = 1) {
    hgnc <- hs.hsgly("gene.id_hgnc.35.12.06.111314")()
    ci1 <- "Ensembl gene ID"
    ci2 <- "Ensembl ID(supplied by Ensembl)"
    a1 <- hgnc[, .(symbol = paste(sort(unique(`Approved symbol`)), 
        collapse = ";")), by = c(ci1)]
    a2 <- hgnc[, .(symbol = paste(sort(unique(`Approved symbol`)), 
        collapse = ";")), by = c(ci2)]
    if (0) {
        common <- a1[, get(ci1) %not% c("", NA)] %and% a2[, get(ci2)]
        i1 <- which(wk.dt.subset(a1, common, ci1)[, symbol] != 
            wk.dt.subset(a2, common, ci2)[, symbol])
        wk.dt.subset(a1, common[i1], ci1)
        wk.dt.subset(a2, common[i1], ci2)
    }
    a2.specific <- a2[, get(ci2) %not% c("", NA)] %not% a1[, 
        get(ci1)]
    a2.specific %<>% {
        wk.dt.subset(a2, ., ci2)
    }
    setnames(a2.specific, ci2, ci1)
    jg <- rbind(a1, a2.specific)[get(ci1) %not.in% c("", NA)]
    setnames(jg, paste0(c("gene.id_"), c("ensembl", "symbol")))
}
"35_12_06_111314" <- function(wj1 = "~/swxx/sj/db/hgnc/combined.tsv.gz", 
    update_db = 0) {
    if (update_db) 
        hs.wj.rm(wj1)
    if (wj.update.reporter(wj1, co1 = Inf)) {
        hgnc <- wk.fread("https://www.genenames.org/cgi-bin/download/custom?col=gd_hgnc_id&col=gd_app_sym&col=gd_app_name&col=gd_status&col=gd_prev_sym&col=gd_aliases&col=gd_pub_chrom_map&col=gd_pub_acc_ids&col=gd_pub_refseq_ids&col=gd_name_aliases&col=md_eg_id&col=md_refseq_id&col=md_prot_id&col=md_ensembl_id&col=gd_pub_eg_id&col=gd_date_mod&col=gd_date2app_or_res&col=gd_prev_name&col=gd_locus_type&col=gd_locus_group&col=gd_pub_ensembl_id&col=family.id&col=gd_enz_ids&status=Approved&status=Entry%20Withdrawn&hgnc_dbtag=on&order_by=gd_app_sym_sort&format=text&submit=submit", 
            quote = "")
        if (0) 
            hgnc %>% dim
        hgnc[, `:=`("HGNC ID", as.integer(hs.sub(`HGNC ID`, "HGNC:")))]
        hs.lapply("Approved symbol,Previous symbols,Alias symbols", 
            function(cn1) {
                set(hgnc, , cn1, toupper(hgnc[[cn1]]))
                1
            })
        hgnc %<>% wk.dt.expand("UniProt ID(supplied by UniProt)", 
            ", ")
        if (0) {
            hgnc[, c("UBA1", "H3C2", "HIST2H3B") %in% `Approved symbol`]
            hgnc[`Approved symbol` %in% c("UBA1", "H3C2", "HIST2H3B")] %>% 
                .sh
            dim(hgnc)
            hgnc[1]
            hgnc[, Status %>% hs.table]
            hgnc[Status %in% "Symbol Withdrawn"]
        }
        {
            {
                cn2merge.list <- list(ncbi = c("NCBI Gene ID", 
                  "NCBI Gene ID(supplied by NCBI)"), ensembl = c("Ensembl gene ID", 
                  "Ensembl ID(supplied by Ensembl)"))
                if (0) {
                  lapply(names(cn2merge.list), function(cn2) {
                    cn1 <- cn2merge.list[[cn2]]
                    cn2 %<>% {
                      paste0("gene.id_", .)
                    }
                    hgnc[, `:=`(c(cn2), get(cn1[1]))]
                    hgnc[hs.clean(get(cn2), inv = 1, ot = "i"), 
                      `:=`(c(cn2), get(cn1[2]))]
                    1
                  })
                }
                {
                  for (cn2do in cn2merge.list) {
                    if (hgnc[, any(hs.grepl(get(cn2do[2]), ","))]) 
                      hs.browser("37.09.13.061559")
                    i1 <- hgnc[, {
                      a1 <- get(cn2do[1])
                      a2 <- get(cn2do[2])
                      which(hs.clean(a2, ot = "l") & (a1 != a2))
                    }]
                    if (length(i1)) {
                      a1 <- hgnc[i1]
                      a1[, `:=`(c(cn2do[1]), get(cn2do[2]))]
                      hgnc %<>% rbind(a1)
                    }
                  }
                }
            }
            cn2merge.list <- list(alias = c("Previous symbols", 
                "Alias symbols"))
            lapply(names(cn2merge.list), function(cn2) {
                cn1 <- cn2merge.list[[cn2]]
                cn2 %<>% {
                  paste0("gene.id_", .)
                }
                a2 <- paste0(hgnc[[cn1[1]]], ", ", hgnc[[cn1[2]]]) %>% 
                  hs.sub("^, ") %>% hs.sub(", $")
                hgnc[, `:=`(c(cn2), a2)]
                1
            })
            if (0) {
                c("NCBI Gene ID", "NCBI Gene ID(supplied by NCBI)", 
                  "Ensembl gene ID", "Ensembl ID(supplied by Ensembl)", 
                  "UniProt ID(supplied by UniProt)", "Approved symbol", 
                  "Previous symbols", "Alias symbols", "Approved name") %not% 
                  names(hgnc)
            }
        }
        wk.dt.fwrite(hgnc[Status %in% "Approved"], wj1)
    }
    wk.fread(wj1, quote = "")
}
"36_04_13_010453" <- local({
    hgnc <- {
    }
    cn.hs.id1 <- function(id1) {
        switch(id1, uniprot = "UniProt ID(supplied by UniProt)", 
            symbol = "Approved symbol", name = "Approved name", 
            ncbi = , ensembl = , alias = paste0("gene.id_", id1))
    }
    cn.hs <- function(name) switch(name, name = "gene.name", 
        paste0("gene.id_", name))
    function(id1, id2, ..mem = 1) {
        if (!length(hgnc)) {
            hgnc <<- hs.hsgly("gene.id_hgnc.35.12.06.111314")()
            for (cn1 in hs.grepV(names(hgnc), "(?i)symbol")) hgnc[[cn1]] %<<=>% 
                toupper
        }
        id1.cn <- cn.hs.id1(id1)
        id2.cn <- cn.hs.id1(id2)
        a1 <- hgnc[, unique(.SD), .SDcols = c(id1.cn, id2.cn)]
        setnames(a1, sapply(c(id1, id2), cn.hs))
    }
})
"36_04_13_094913" <- function(id1, id2, sep1 = ";", ..mem = 1) {
    a1 <- hs.hsgly("id1,id2.36.04.13.010453")(id1, id2)
    cn <- names(a1)
    if (length(cn) > 2) 
        hs.browser("36.04.13.095554")
    jg <- a1[hs.clean(get(cn[1])), {
        a1 <- as.character(get(cn[2]))
        if (.N > 1) 
            paste(unique(sort(a1)), collapse = sep1)
        else a1
    }, by = c(cn[1])]
    jg %>% {
        hs.c(.[[1]], .[[2]])
    }
}
"36_04_16_111737" <- function(id1, ft = if (hs.grep(id1[1], "^(?i)ensg")) "ensembl", 
    tt = "symbol", sep1 = ";;;;") {
    if (length(ft) == 0) 
        stop("36.04.16.112009")
    id1.list <- strsplit(as.character(id1), "[;,]+\\s*")
    id1.id2.db <- hs.hsgly("id1.id2+.36.04.13.094913")(ft, tt)
    id1 %>% as.character %>% {
        hs.c(., strsplit(., "[;,]+\\s*") %>% sapply(. %>% {
            a1 <- id1.id2.db[.]
            if (length(a1) > 1) 
                a1 %<>% sort %>% paste(collapse = sep1)
            a1
        }))
    }
}
"36_04_12_233405" <- function(..mem = 1) {
    a1 <- hs.hsgly("ncbi,symbol.36.04.13.005401")()
    if (0) {
        a1[, anyDuplicated(gene.id_symbol)]
        a1[, anyDuplicated(gene.id_ncbi)]
    }
    a1[, .(gene.id_symbol = if (.N > 1) 
        paste(gene.id_symbol, collapse = ";")
    else gene.id_symbol), by = "gene.id_ncbi"]
}
"36_04_13_005401" <- function(..mem = 1) {
    hgnc <- hs.hsgly("gene.id_hgnc.35.12.06.111314")()
    ci1 <- "NCBI Gene ID"
    ci2 <- "NCBI Gene ID(supplied by NCBI)"
    a1 <- hgnc[, .(symbol = unique(`Approved symbol`)), by = c(ci1)]
    a2 <- hgnc[, .(symbol = unique(`Approved symbol`)), by = c(ci2)]
    a2.specific <- a2[, get(ci2) %not% c("", NA)] %not% a1[, 
        get(ci1)]
    a2.specific %<>% {
        wk.dt.subset(a2, ., ci2)
    }
    setnames(a2.specific, ci2, ci1)
    jg <- rbind(a1, a2.specific)[get(ci1) %not.in% c("", NA)]
    setnames(jg, c("gene.id_ncbi", "gene.id_symbol"))
    jg[, `:=`("gene.id_ncbi", as.character(gene.id_ncbi))]
}
"36_04_12_234007" <- function(..mem = 1) {
    hgnc <- hs.hsgly("gene.id_hgnc.35.12.06.111314")() %>% fread
    a1 <- hgnc[`Approved symbol` %not.in% c("", NA), .(symbol = `Approved symbol`, 
        alias = paste0(`Previous symbols`, ", ", `Alias symbols`))]
    a1 <- a1[, `:=`("alias", alias %>% trim(", "))]
    jg <- a1[, .(alias = if (.N > 1) 
        alias %not% c("", NA)
    else alias), by = "symbol"]
    setnames(jg, names(jg), paste0("gene.id_", names(jg)))
}
"36_04_12_234512" <- function(..mem = 1) {
    hgnc <- hs.hsgly("gene.id_hgnc.35.12.06.111314")()
    a1 <- hgnc[Status == "Approved", .(symbol = `Approved symbol`, 
        name = `Approved name`)]
    jg <- a1[, .(name = if (.N > 1) 
        name %not% c("", NA)
    else name), by = "symbol"]
    setnames(jg, c("gene.name", "gene.id_symbol"))
}
"38_02_27_114727" <- local({
    wk.xml.kv <- function(xml, k1, no = "") {
        jg <- if (k1 %in% names(xml)) 
            xmlToList(xml[[k1]])
        if (!length(jg)) 
            jg <- no
        jg
    }
    wk.xml.ks_v <- function(xml, k, no = if (length(k) > 1) 
        "") {
        jg <- hs.lapply(k, function(k1) wk.xml.kv(xml, k1, no = no))
        if (length(k) > 1) 
            wk.dt.as(jg)
        else hs.clean(jg)
    }
    hs.disposition <- function(x) {
        i <- hs.grep(names(x), "(?i)descendant")
        paste0(if ("term" %in% names(x)) 
            paste(c(x[["term"]], if (length(i)) ":"), collapse = "")
        else "", if (length(i)) 
            unlist(sapply(x[i], hs.disposition))
        else "")
    }
    function(what = "name", only_wj = 0, redo = 0) {
        wjj.38_03_20_133117 <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("hmdb")
        dz1 <- "https://hmdb.ca/system/downloads/current/hmdb_metabolites.zip"
        wj1 <- wk.wget(dz1, wjj.38_03_20_133117)
        if (0) {
            .less(wj1)
            {
                .wk("unzip -p", wj1, "| perl -e", .q("\nwhile( <>) {\n  #chomp;\n  if( /<\\/metabolite>/) { $n++;}\n}\nprint( $n, \"\n\");\n"))
                .wk("unzip -p", wj1, "| perl -e", .q("\nwhile( <>) {\n  if( /hmdb\\d+/i) {\n    if( not( exists( $id{ $&}))) {\n      $id{ $&} = 1;\n    }\n  }\n}\nprint( scalar keys( %id), \"\n\");\n"))
                .wk("unzip -p", wj1, "| perl -e", .q("\nwhile( <>) {\n  chomp;\n  if( /<\\/metabolite>\\s/) { $n++;}\n}\nprint( $n, \"\n\");\n"))
            }
            .wk("unzip -p", wj1, "| perl -e", .q("\nwhile( <>) {\n  if( /<kegg_id> *([^< ]+) *</i) {\n    if( not( exists( $id{ $1}))) {\n      $id{ $1} = 1;\n    }\n  }\n}\nforeach $k ( keys( %id)) {\n  print( $k, \"\n\");\n}\n"), 
                "| gzip > ~/a.txt.gz")
            .wk("unzip -p", wj1, "| head -n 10000000 | perl -e", 
                .q("\nwhile( <>) {\n  if( /<kegg_id>C06390/i) {\n    print;\n    exit;\n  }\n}\n"), 
                intern = 1)
            .wc("~/a.txt.gz")
        }
        wj2 <- hs.wjm(wj1, ext = paste0("id_", what, ".tsv.gz"))
        if (only_wj) 
            return(wj2)
        if (redo) 
            hs.wj.rm(wj2)
        if (wj.update.reporter(wj2, wj1)) {
            hs.wj.rm(wj2)
            hs.require("XML")
            nc <- hs.switch(what, "name", , "disposition", 10, 
                1)
            if (0) {
                .wk("unzip -p", wj1, "| wc -l")
                .wk("gzip -cd", .sys.fpGood(wj1), "| wc -l")
                xml <- readLines(wj1, 2435)
                xml <- xmlParse(tail(xml, -2))
                xml.root <- xmlRoot(xml)
                xml.root %>% hs.methods
                xml.root[["name"]] %>% xmlToList
                xml.root[["synonyms"]] %>% xmlToList %>% unlist
                xml.root[["iupac_name"]] %>% xmlToList
                xml.root[["traditional_iupac"]]
                xml.root %>% wk.less
                names(xml.root)
            }
            {
                hs.msg("开始处理xml")
                con1 <- .sys("unzip -p", .sys.fpGood(wj1), pipe = "r")
                on.exit(close(con1), add = TRUE)
                readLines(con1, 2)
                chunk.size <- 1e+06
                lines_read <- 0
                line_start <- "<metabolite>"
                line_end <- "</metabolite>"
                remained <- NULL
                repeat {
                  hs.msg(paste("开始处理下一批", lines_read), 
                    2)
                  Lines <- readLines(con1, chunk.size)
                  if (length(Lines)) {
                    lines_read <- lines_read + length(Lines)
                    hs.msg(paste("Lines read so far:", lines_read), 
                      0)
                  }
                  else {
                    if (!match(line_start, remained, nomatch = 0)) 
                      break
                    Lines <- remained
                    remained <- NULL
                  }
                  sep.i <- which(Lines %in% line_start)
                  if (!length(sep.i)) {
                    remained <- c(remained, Lines)
                    next
                  }
                  a1 <- hs.sepSeq(Lines, sep.i = sep.i)
                  if (length(remained)) {
                    a1[[1]] <- c(remained, a1[[1]])
                    remained <- NULL
                  }
                  if (a1[[1]][1] != line_start) 
                    a1[[1]] <- NULL
                  if (!any(length(a1), length(remained))) 
                    break
                  i2do.max <- length(a1)
                  if (i <- match(line_end, a1[[i2do.max]], nomatch = 0)) {
                    if (i < length(a1[[i2do.max]])) {
                      remained <- tail(a1[[i2do.max]], -i)
                      a1[[i2do.max]] %<>% head(i)
                    }
                  }
                  else {
                    remained <- a1[[i2do.max]]
                    i2do.max <- i2do.max - 1
                  }
                  if (i2do.max) {
                    if (0) 
                      what <- "inchi_key"
                    a <- wk.mclapply(1:i2do.max, function(i1) {
                      xml <- xmlRoot(xmlParse(a1[[i1]]))
                      if (0) {
                        wk.less(xml)
                        names(xml) %>% hs.grepV("(?i)smile")
                      }
                      id1 <- xml[["accession"]] %>% xmlToList
                      v1 <- switch(what, title = {
                        name <- hs.clean(hs.unlist(lapply(c("name"), 
                          function(k1) wk.xml.kv(xml, k1))), 
                          U = 1)
                        tapply(name, tolower(name), function(x1) {
                          if (length(x1) > 1) {
                            x1 <- hs.sort(x1)
                            reg <- gregexpr("[A-Z]", x1, perl = TRUE)
                            x1[which.max(sapply(reg, function(reg1) sum(reg1 > 
                              0L)))]
                          } else x1
                        }) %>% unname
                      }, name = {
                        name <- hs.clean(hs.unlist(lapply(c("name", 
                          "synonyms", "iupac_name", "traditional_iupac"), 
                          function(k1) wk.xml.kv(xml, k1))), 
                          U = 1)
                        tapply(name, tolower(name), function(x1) {
                          if (length(x1) > 1) {
                            x1 <- hs.sort(x1)
                            reg <- gregexpr("[A-Z]", x1, perl = TRUE)
                            x1[which.max(sapply(reg, function(reg1) sum(reg1 > 
                              0L)))]
                          } else x1
                        }) %>% unname
                      }, common_name = wk.xml.kv(xml, "name"), 
                        disposition = {
                          a <- wk.xml.kv(xml, "ontology")
                          jg <- if (!hs.all(is.character(a), 
                            length(a) == 1)) {
                            i2 <- match("Disposition", sapply(a, 
                              function(x) x[["term"]]), nomatch = 0)
                            if (i2) {
                              b <- a[[i2]][hs.grep(names(a[[i2]]), 
                                "(?i)descendant")]
                              hs.disposition(b)
                            }
                          }
                          {
                            a <- wk.xml.kv(xml, "biological_properties")
                            if (!all(hs.grepl(names(a), "(?i)location|pathway"))) {
                              hs.browser("38.07.04.192431")
                              wk.less(a)
                              length(a)
                            }
                            c(jg, a[hs.grep(names(a), "(?i)location")] %>% 
                              lapply(. %>% unlist %>% trim %>% 
                                hs.clean) %>% hs.nonempty %>% 
                              {
                                paste0(names(.), ":", sapply(., 
                                  paste, collapse = ";"))
                              })
                          }
                        }, keggID = wk.xml.kv(xml, "kegg_id"), 
                        pubchemCompoundID = wk.xml.kv(xml, "pubchem_compound_id"), 
                        smiles = wk.xml.kv(xml, "smiles"), cas_id = wk.xml.kv(xml, 
                          "cas_registry_number"), formula = wk.xml.kv(xml, 
                          "chemical_formula"), mass = wk.xml.ks_v(xml, 
                          c("average_molecular_weight", "monisotopic_molecular_weight")), 
                        inchi = wk.xml.kv(xml, "inchi"), inchi_key = wk.xml.kv(xml, 
                          "inchikey"), class = wk.xml.ks_v(xml[["taxonomy"]], 
                          c("super_class", "class", "sub_class")), 
                        direct_parent = local({
                          a <- wk.xml.kv(xml, "taxonomy")
                          if ("direct_parent" %in% names(a)) {
                            a[["direct_parent"]]
                          } else ""
                        }))
                      if (length(v1)) 
                        hs.c(id1, list(unname(v1)))
                    }, mc.cores = nc)
                    hs.msg("extracted", 0)
                    a2 <- hs.unlist(unname(a), 0, 1)
                    if (!length(a2)) 
                      next
                    dt <- wk.dt.fromList(a2)
                    wk.dt.replace(dt, trim)
                    wk.dt.setnames(dt, c("id", switch(what, class = c("super_class", 
                      "class", "sub_class"), mass = c("average_molecular_weight", 
                      "monisotopic_molecular_weight"), what)))
                    if (0) {
                      dt[keggID %>% hs.clean(ot = "l")]
                      if (dt[, "HMDB0014568" %in% id]) 
                        hs.browser("38.07.09.211357")
                      dt[id %in% "HMDB0014568"]
                      dt[id %in% c("HMDB0010330", "HMDB0010331"), 
                        keggID]
                      .less(wj1)
                      a1[["HMDB0014568"]]
                      a1[c("HMDB0010446", "HMDB0014568")] %>% 
                        wk.dt.fromList
                      list(a = NA, b = "b") %>% wk.dt.fromList %>% 
                        .s
                    }
                    wk.dt.fwrite(dt, wj2, smart = 1)
                    hs.msg("wrote", 0)
                  }
                }
                if (0) {
                  dt[, .N]
                  dt[, uniqueN(group)]
                  dt[, uniqueN(member)]
                  {
                    a <- dt[, {
                      if (.N > 1) {
                        .SD
                      }
                    }, by = member]
                    sapply(a, uniqueN)
                    a[, hs.table(member) %>% hs.table] %>% .sh
                    a[, hs.table(member) %>% {
                      .[which.max(.)]
                    }]
                    a[member %in% "triacylglycerol"]
                  }
                }
            }
        }
        wj2
    }
})
"38_07_04_212712" <- function() {
}
"37_09_23_120223" <- function() {
    wjj1 <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("icgc")
    return(wjj1)
    {
        hs.wjj(wjj1)
        "/Users/wk/Downloads/simple_somatic_mutation.open.BLCA-CN.tsv.gz"
    }
}
"38_10_01_002406" <- function() {
    wj2 <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg") %>% 
        hs.fp("compound/class.tsv.gz")
    db_wj <- c(class.pathway = hs.hsgly("kegg/通路/分类_网页.38_10_01_074028")(), 
        pathway.compound = hs.hsgly("kegg/通路/化合物.38_10_01_121811")())
    if (wj.update.reporter(wj2, db_wj)) {
        db <- db_wj %>% lapply(wk.fread, colClasses = "character")
        dt <- wk.dt.copy(db[["class.pathway"]])
        dt[, `:=`(pathway, paste0("path:map", pathway))]
        wk.dt.absorb1(dt, db[["pathway.compound"]], by1 = "pathway", 
            by2 = "pathway")
        jg <- dt[, .(class1, class2, compound)]
        jg %<>% wk.dt.expand("compound")
        jg <- jg[hs.clean(compound, ot = "i")]
        jg <- unique(jg)
        if (0) {
            jg[, uniqueN(compound), by = class2] %>% wk.dt.setorderv("class2") %>% 
                .s
        }
        wk.dt.fwrite(jg, wj2)
    }
    if (0) {
        db_wj <- c(pathway.class = hs.hsgly("kegg/通路/分类.38_09_30_222336")(), 
            pathway.module = hs.hsgly("kegg/通路/模块.38_10_01_001959")(), 
            module.compound = hs.hsgly("kegg/模块/id.compound.38_10_01_002654")())
        if (wj.update.reporter(wj2, db_wj)) {
            db <- db_wj %>% lapply(wk.fread)
            dt <- wk.dt.copy(db[["pathway.class"]])
            wk.dt.absorb1(dt, db[["pathway.module"]], by1 = "id", 
                by2 = "pathway")
            wk.dt.absorb1(dt, db[["module.compound"]], by1 = "module", 
                by2 = "id")
            jg <- dt[, .(class1, class2, compound)]
            jg %<>% wk.dt.expand("compound")
            jg <- jg[hs.clean(compound, ot = "i")]
            if (0) {
                jg[, uniqueN(compound), by = class2] %>% wk.dt.setorderv("class2") %>% 
                  .s
            }
            wk.dt.fwrite(jg, wj2)
        }
    }
    wj2
}
"38_05_21_103033" <- function() {
    wj2 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
        "compound", "id.drug.tsv.gz")
    if (!file.exists(wj2)) {
        sj.wjj <- hs.hsgly("kegg/化合物/id.info_flat.38_05_20_212307")()
        sj.wj <- hs.wjj.ls(sj.wjj)
        REMARK <- lapply(seq_along(sj.wj), function(i) {
            wj1 <- sj.wj[i]
            hs.load(wj1)[[1]][["REMARK"]]
        })
        names(REMARK) <- hs.wjm(sj.wj, dn = 0, ext = 0)
        REMARK %<>% hs.nonempty
        a <- REMARK %>% lapply(. %>% {
            hs.grepV(., "(?i)same as:")
        }) %>% hs.nonempty
        jg <- sapply(a, . %>% {
            hs.gre(., "(?i)\\bd\\d+")
        }) %>% hs.nonempty
        jg <- wk.dt.fromList(jg, nm = c("compound", "drug"))
        if (0) {
            jg[, if (.N > 1) 
                .SD, by = drug]
        }
        wk.dt.fwrite(jg, wj2)
    }
    wj2
}
"38_08_31_025713" <- function(id = {
}) {
    if (length(id) == 0) {
        wj2 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
            "compound", "id.cas_id.tsv.gz")
        if (!file.exists(wj2)) {
            id <- hs.hsgly("kegg/化合物/id.name.38_02_24_131718")() %>% 
                wk.xls.read %>% {
                .[[1]]
            } %>% unique
            jg <- wk.dt(id)
            jg[, `:=`("cas_id", hs.na(""))]
            for (i in seq_along(id)) {
                message(i)
                if (jg[i, !is.na(cas_id)]) 
                  next
                id1 <- id[i]
                wj1 <- hs.hsgly("kegg/化合物/id.info_flat.38_05_20_212307")(id1)
                if (!file.exists(wj1)) 
                  next
                a <- hs.load(wj1)
                a1 <- a[[1]][["DBLINKS"]] %>% hs.grepV("(?i)cas")
                if (length(a1)) {
                  id2 <- hs.gre(a1, "\\d+-\\d+-\\d+")[[1]]
                  if (length(id2) > 1) 
                    id2 %<>% paste(collapse = ";")
                  if (!length(id2)) 
                    hs.browser("38.08.30.162127")
                  wk.dt.set(jg, i, "cas_id", id2)
                }
            }
            wk.dt.fwrite(jg, wj2)
        }
        return(wj2)
    }
    hs.browser("38_08_31_030145")
}
"38_05_20_212307" <- local({
    hs1 <- function(id) {
        wj2 <- hs.fp(wjj2, paste0(id, ".rd"))
        local({
            j.db <- which(!file.exists(wj2))
            i <- length(j.db)
            repeat {
                hs.switch(i, 0, break, 1, {
                }, .sleep(1))
                j <- j.db[i]
                hs.msg(i, 0)
                id1 <- id[j]
                pre1 <- hs.cond(hs.grepl(id1, "^(?i)c\\d+$"), 
                  "cpd", hs.grepl(id1, "^(?i)d\\d+$"), "dr", 
                  hs.browser("38.08.27.113616"))
                a <- .try(KEGGREST::keggGet(paste0(pre1, ":", 
                  id[j])))
                if (!hs.is_try_error(a)) 
                  hs.save(a, wj = wj2[j])
                i <- i - 1
            }
        })
        wj2
    }
    function(id = {
    }) {
        wjj2 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
            "compound", "id.info_flat")
        hs.switch(length(id), 0, {
            id.name <- hs.hsgly("kegg/化合物/id.name.38_02_24_131718")() %>% 
                wk.xls.read
            id <- id.name[, kegg_id %>% unique %>% hs.grepV("^C")]
            hs1(id)
            return(wjj2)
        }, return(hs1(id)))
    }
})
"38_09_01_072617" <- function(id = {
}) {
    if (length(id) == 0) {
        wj2 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
            "compound", "id.mass.tsv.gz")
        if (!file.exists(wj2)) {
            id <- hs.hsgly("kegg/化合物/id.name.38_02_24_131718")() %>% 
                wk.xls.read %>% {
                .[[1]]
            } %>% unique
            jg <- wk.dt(id)
            cn <- c("FORMULA", "EXACT_MASS", "MOL_WEIGHT")
            for (cn1 in cn) jg[, `:=`(c(cn1), hs.na(""))]
            for (i in seq_along(id)) {
                message(i)
                if (jg[i, !is.na(FORMULA)]) 
                  next
                id1 <- id[i]
                wj1 <- hs.hsgly("kegg/化合物/id.info_flat.38_05_20_212307")(id1)
                if (!file.exists(wj1)) 
                  next
                a <- hs.load(wj1)[[1]]
                a1 <- a[cn]
                a1 %<>% lapply(. %>% {
                  if (length(.)) 
                    .
                  else ""
                })
                wk.dt.set(jg, i, cn, a1)
            }
            wk.dt.fwrite(jg, wj2)
        }
        return(wj2)
    }
    hs.browser("38.08.30.162642")
}
"38_02_24_131718" <- function() {
    wj2 <- hs.wj(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
        "compound", "name.id.xlsx")
    if (wj.update.reporter(wj2, co1 = Inf)) {
        hs.require("KEGGREST")
        kegg.cpd <- KEGGREST::keggList("compound")
        kegg.drug <- KEGGREST::keggList("drug")
        if (0) {
            kegg.drug[1] %>% strsplit(";\\s*")
        }
        L <- c(kegg.cpd, kegg.drug) %>% strsplit("; *") %>% lapply(hs.clean) %>% 
            hs.nonempty
        names(L) %<>% hs.sub(".*:")
        dt <- wk.dt.fromList(L)
        dt[, `:=`(member, trim(member))]
        if (0) {
            dt[member != trim(member), c(member, trim(member))]
            dt[hs.clean(V1, inv = 1, ot = "l")]
        }
        wk.dt.setnames(dt, c("kegg_id", "name"))
        wk.xls.write(dt, wj2)
    }
    wj2
}
"38_10_25_113343" <- local({
    db <- {
    }
    function(id = {
    }) {
        if (!length(db)) {
            db <<- hs.hsgly("kegg/化合物/id.name.38_02_24_131718")() %=>% 
                wk.xls.read
            db <<- db[, name[which.min(nchar(name))], by = kegg_id]
        }
        if (length(id)) {
            wk.dt.subset.na(db, id, "kegg_id")
        }
        else db
    }
})
"38_08_30_161536" <- function(id = {
}) {
    if (length(id) == 0) {
        wj2 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
            "compound", "id.pubchem_id.tsv.gz")
        if (!file.exists(wj2)) {
            id <- hs.hsgly("kegg/化合物/id.name.38_02_24_131718")() %>% 
                wk.xls.read %>% {
                .[[1]]
            } %>% unique
            jg <- wk.dt(id)
            jg[, `:=`("pubchem_sid", hs.na(""))]
            i <- match("D02012", id)
            for (i in seq_along(id)) {
                message(i)
                if (jg[i, !is.na(pubchem_sid)]) 
                  next
                id1 <- id[i]
                wj1 <- hs.hsgly("kegg/化合物/id.info_flat.38_05_20_212307")(id1)
                if (!file.exists(wj1)) 
                  next
                a <- hs.load(wj1)
                a1 <- a[[1]][["DBLINKS"]] %>% hs.grepV("(?i)pubchem")
                if (length(a1)) {
                  id2 <- hs.gre(a1, "\\d+")[[1]]
                  if (length(id2) > 1) 
                    hs.browser("38.08.30.162114")
                  if (!length(id2)) 
                    hs.browser("38.08.30.162127")
                  wk.dt.set(jg, i, "pubchem_sid", id2)
                }
            }
            {
                id <- jg[, hs.clean(pubchem_sid, U = 1) %>% list]
                wk.dt.setnames(id, "sid")
                wj_db <- "~/swxx/sj/db/pubchem/ftp.ncbi.nlm.nih.gov/pubchem/Compound/Extras/CID-SID.gz"
                wj_db %>% .head
                a <- hs.hsgly("1.信息学/文件/读写/提取/表格的子集.38_09_10_214239")(v1 = id[, 
                  sid], db.fp = wj_db, ci1 = 2)
                wk.dt.setnames(a, c("cid", "sid", "type"))
                if (0) {
                  a[, type %>% hs.table]
                  id1 <- a[match(2, type), sid]
                  a[sid %in% id1]
                  a[sid %in% sid[type %in% 2], if (1 %not.in% 
                    type) 
                    .SD, by = sid]
                  a[, anyDuplicated(sid)]
                  a[, anyDuplicated(cid)]
                  jg[hs.clean(pubchem_sid, ot = "i"), anyDuplicated(pubchem_sid)]
                  a_bf <- a
                }
                a <- a[type %in% 1] %>% wk.dt.collapse(by = "sid", 
                  U = 1)
                if (0) {
                  jg[id %in% "D02012"]
                  a_bf[sid %in% "7849074"]
                  a[sid %in% "7849074"]
                }
                wk.dt.absorb1(jg, a, by1 = "pubchem_sid", by2 = "sid", 
                  what = "cid", as = "pubchem_cid", keep_old = 0)
            }
            wk.dt.fwrite(jg, wj2, cn = 1)
        }
        return(wj2)
    }
    hs.browser("38.08.30.162642")
}
"38_10_01_002654" <- function() {
    wjj_in <- hs.hsgly("kegg/模块/id.info_flat.38_09_30_165723")()
    wj2 <- hs.wjm(wjj_in, ext = "compound.tsv.gz")
    if (!file.exists(wj2)) {
        wj <- hs.wjj.ls(wjj_in, "(?i)\\.rd$")
        names(wj) <- hs.wjm(wj, dn = 0, ext = 0)
        sj <- lapply(wj, function(wj1) {
            sj1 <- hs.load(wj1)
            names(sj1[[1]][["COMPOUND"]])
        })
        jg <- wk.dt.fromList(sj, nm = c("id", "compound"))
        wk.dt.fwrite(jg, wj2)
        jg
    }
    wj2
}
"38_09_30_165723" <- local({
    hs1 <- function(id) {
        wj2 <- hs.fp(wjj2, paste0(id, ".rd"))
        local({
            j.db <- which(!file.exists(wj2))
            i <- length(j.db)
            repeat {
                hs.switch(i, 0, break, 1, {
                }, .sleep(1))
                j <- j.db[i]
                hs.msg(i, 0)
                id1 <- id[j]
                a <- .try(KEGGREST::keggGet(paste0("md:", id[j])))
                if (!hs.is_try_error(a)) 
                  hs.save(a, wj = wj2[j])
                i <- i - 1
            }
        })
        wj2
    }
    function(id = {
    }) {
        wjj2 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
            "module", "id.info_flat")
        hs.switch(length(id), 0, {
            id.name <- hs.hsgly("kegg/模块/id.name.38_09_30_165917")() %>% 
                wk.xls.read
            hs1(id.name[, hs.sub(kegg_id, "md:")])
            return(wjj2)
        }, return(hs1(id)))
    }
})
"38_09_30_165917" <- function() {
    wj2 <- hs.wj(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
        "module", "name.id.xlsx")
    if (wj.update.reporter(wj2, co1 = Inf)) {
        hs.require("KEGGREST")
        kegg.module <- KEGGREST::keggList("module")
        dt <- kegg.module %>% {
            wk.dt(names(.), .)
        }
        wk.dt.setnames(dt, c("kegg_id", "name"))
        wk.xls.write(dt, wj2)
        if (0) {
            "/home/wk/swxx/sj/db/kegg/module/name.id.xlsx" %>% 
                wk.rsync.in(ssh = "hp1")
        }
    }
    wj2
}
"38_09_30_222336" <- function(ot = "wj") {
    wj2 <- hs.wj(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
        "通路", "分类.tsv.gz")
    if (!file.exists(wj2)) {
        wj <- hs.hsgly("kegg/通路/信息.38_09_30_222601")() %>% 
            hs.wjj.ls("(?i)\\.rd$")
        names(wj) <- hs.wjm(wj, dn = 0, ext = 0)
        a <- lapply(wj, function(wj1) {
            sj1 <- hs.load(wj1)
            hs.browser("38.10.01.001429")
            sj1[[1]][["CLASS"]]
        })
        sapply(a, length) %>% hs.table
        a[which(sapply(a, length) == 0)]
        jg <- wk.dt.fromList(a, nm = c("id", "class"))
        if (0) {
            jg[, class %>% hs.re("[^;]+", no_match = "") %>% 
                hs.table %>% wk.dt]
        }
        jg[, `:=`(class1, hs.re(class, "[^;]*"))]
        jg[, `:=`(class2, hs.re(class, "(?<=;).*") %>% trim)]
        jg[, `:=`(class, {
        })]
        wk.dt.fwrite(jg, wj2)
    }
    hs.switch(ot, "wj", wj2, "dt", wk.fread(wj2))
}
"38_10_01_074028" <- function() {
    wj1 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
        "通路", "pathway.html.gz")
    if (!file.exists(wj1)) 
        .wk("wget", "-O -", "https://www.genome.jp/kegg/pathway.html", 
            "| gzip >", .sys.fpGood(hs.wj(wj1)))
    wj2 <- hs.wjm(wj1, ext = "tsv.gz", extN = 2)
    if (wj.update.reporter(wj2, wj1)) {
        a <- readLines(wj1)
        a <- hs.sepSeq(a, sep.p = "^\\<h4 *id=")
        a[[3]] %>% head(20)
        jg <- a[-1] %>% lapply(. %>% {
            message(.[1])
            a <- hs.sepSeq(., sep.p = "^\\<b")
            jg <- a[-1] %>% lapply(. %>% {
                hs.grepV(., "^ *<dt>") %>% hs.re("\\d+")
            })
            names(jg) <- a[-1] %>% sapply("[", 1) %>% hs.re(">.*") %>% 
                hs.sub(">\\d+\\.\\d+ *") %>% hs.sub("</b>")
            wk.dt.fromList(jg, nm = c("class2", "pathway"))
        })
        names(jg) <- a[-1] %>% sapply("[", 1) %>% hs.re("(?<=\\>).*") %>% 
            hs.sub("\\d\\. *") %>% hs.sub("</h4>")
        jg <- wk.dt.fromList(jg, nm = "class1")
        wk.dt.fwrite(jg, wj2)
    }
    wj2
}
"38_09_30_215837" <- function() {
    wj2 <- hs.wj(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
        "通路", "号.名称.tsv.gz")
    if (wj.update.reporter(wj2, co1 = Inf)) {
        hs.require("KEGGREST")
        db <- KEGGREST::keggList("pathway")
        jg <- db %>% {
            wk.dt(names(.), .)
        }
        wk.dt.setnames(jg, c("kegg_id", "name"))
        wk.dt.fwrite(jg, wj2)
    }
    wj2
}
"38_10_01_121811" <- function() {
    wj2 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
        "通路", "化合物.tsv.gz")
    if (!file.exists(wj2)) {
        pathway_id.name <- hs.hsgly("kegg/通路/号.名称.38_09_30_215837")() %>% 
            wk.fread
        L <- pathway_id.name[, kegg_id] %>% hs.lapply(function(pw1) {
            message(pw1)
            a <- readLines(paste0("https://www.genome.jp/dbget-bin/get_linkdb?-t+compound+", 
                pw1))
            a1 <- hs.grepV(a, "<a href=\"/entry/")
            hs.re(a1, "(?<=\\<a href=\"/entry/)[^\"]+")
        })
        jg <- wk.dt.fromList(L, nm = c("pathway", "compound"))
        wk.dt.fwrite(jg, wj2)
    }
    wj2
}
"36_03_18_133126" <- function(org1 = "human", redo = 0, out = "obj", 
    ..mem = 1) {
    jg.wj <- .mfp.swxx("36.01.14.162629.tsv.gz") %>% hs.wj
    if (org1 != "human") 
        jg.wj <- hs.wjm(jg.wj, append = paste0(".", org1), extN = 2)
    if (out == "fp") 
        return(jg.wj)
    if (redo) 
        hs.wj.rm(jg.wj)
    if (wj.update.reporter(jg.wj, co1 = 30)) {
        jg <- local({
            org1.kegg_id <- hs.hsgly("org.kegg_id.36.06.08.185551")(org1)
            R.utils::setOption("clusterProfiler.download.method", 
                "auto")
            db1 <- clusterProfiler:::prepare_KEGG(org1.kegg_id, 
                "KEGG", "kegg")
            if (0) {
                ls(db1)
                db1[["EXTID2PATHID"]]
                db1[["PATHID2EXTID"]]
                db1[["PATHID2NAME"]]
                clusterProfiler:::prepare_KEGG
                clusterProfiler:::download_KEGG
                a <- clusterProfiler:::download.KEGG.Path("hsa")
                names(a)
                clusterProfiler:::download.KEGG.Path
                keggpathid2extid.df <- clusterProfiler:::kegg_link("hsa", 
                  "pathway")
                getOption("clusterProfiler.download.method")
                dl <- tryCatch(downloader::download(url, quiet = TRUE, 
                  ...), error = function(e) NULL)
                `?`(utils::download.file)
            }
            kegg.id_name <- db1[["PATHID2NAME"]]
            kegg.name_id <- kegg.id_name %>% {
                hs.c(., names(.))
            }
            kegg.gene_id <- db1[["EXTID2PATHID"]]
            kegg.gene_id %<>% inverseList
            names(kegg.gene_id) %<>% {
                kegg.id_name[.]
            }
            if (0) {
                hs.hsgly("基因/gene_info.37_12_31_133053")(org1 = org1)
                species <- clusterProfiler:::organismMapper("hsa")
                keyType <- "kegg"
                KEGG_DATA <- clusterProfiler:::prepare_KEGG(org1.kegg_id, 
                  "KEGG", keyType)
                ls(KEGG_DATA)
                KEGG_DATA[["PATHID2EXTID"]] %>% head
            }
            a1 <- kegg.gene_id %>% {
                db1 <- .
                lapply(names(.), . %>% {
                  data.table(., db1[[.]])
                })
            }
            a1 <- rbindlist(a1)
            wk.dt.setnames(a1, c("set", "gene"))
            a1[, `:=`(gene, as.numeric(gene))]
            a1[, `:=`("set_id", kegg.name_id[set])]
            kegg.name_id %>% head
            if (org1 %in% "human") {
                a1[, `:=`("gene.id_symbol", hs.hsgly("gene.id.converter.using_hgnc.36.01.22.160941")(gene, 
                  "gene.id_ncbi", "gene.id_symbol", "ncbi", "symbol")[, 
                  gene.id_symbol])]
            }
            else {
                hs.hsgly("dt.id1_to_id2.37_03_22_145251")(a1, 
                  id1 = "ncbi", id2 = "symbol", id1.cn = "gene", 
                  id2.cn = "gene.id_symbol", org1 = org1)
                if (a1[, any(gene != gene_updated)]) 
                  a1[, `:=`(gene, gene_updated)]
                a1[, `:=`("gene_updated", {
                })]
                wk.dt.col.reorder(a1, c("set", "gene", "set_id"), 
                  before = 1)
            }
            a1
        })
        wk.dt.fwrite(jg, jg.wj)
    }
    note.update_time(jg.wj)
    wk.fread(jg.wj)
}
"38_10_01_001959" <- function(ot = "wj") {
    wj2 <- hs.wj(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
        "通路", "模块.tsv.gz")
    if (!file.exists(wj2)) {
        wj <- hs.hsgly("kegg/通路/信息.38_09_30_222601")() %>% 
            hs.wjj.ls("(?i)\\.rd$")
        names(wj) <- hs.wjm(wj, dn = 0, ext = 0)
        a <- lapply(wj, function(wj1) {
            sj1 <- hs.load(wj1)
            names(sj1[[1]][["MODULE"]])
        })
        jg <- wk.dt.fromList(a, nm = c("pathway", "module"))
        wk.dt.fwrite(jg, wj2)
    }
    hs.switch(ot, "wj", wj2, "dt", wk.fread(wj2))
}
"38_09_30_222601" <- local({
    hs1 <- function(id) {
        wj2 <- hs.fp(wjj2, paste0(id, ".rd"))
        local({
            j.db <- which(!file.exists(wj2))
            i <- length(j.db)
            repeat {
                hs.switch(i, 0, break, 1, {
                }, .sleep(1))
                j <- j.db[i]
                hs.msg(i, 0)
                id1 <- id[j]
                a <- .try(KEGGREST::keggGet(paste0("path:", id[j])))
                if (!hs.is_try_error(a)) 
                  hs.save(a, wj = wj2[j])
                i <- i - 1
            }
        })
        wj2
    }
    function(id = {
    }) {
        wjj2 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("kegg"), 
            "通路", "号.信息")
        hs.switch(length(id), 0, {
            id.name <- hs.hsgly("kegg/通路/号.名称.38_09_30_215837")() %>% 
                wk.fread
            hs1(id.name[, hs.sub(kegg_id, "path:")])
            return(wjj2)
        }, return(hs1(id)))
    }
})
"38_09_14_150155" <- function() {
    wjj.38_09_14_150301 <- .mfp.swxx("38_09_14_150305")
    dz1 <- "http://rest.kegg.jp/list/organism"
    wj2 <- dz1 %>% hs.wjm(dn = wjj.38_09_14_150301, add = ".tsv.gz")
    wk.wget2wj(dz1, wj2)
}
"38_09_14_154201" <- local({
    db1 <- {
    }
    function(name, search = 0) {
        if (!length(db1)) {
            db1 <<- hs.hsgly("kegg/物种/列表.38_09_14_150155")() %>% 
                wk.fread(cn = 0, sep = "\t")
            db1[, `:=`("org_name", V3 %>% trim %>% hs.re("\\([^\\(\\)]*\\)$", 
                no_match = "") %>% trim("[\\(\\)]"))]
            if (0) {
                db1[, hs.clean(org_name, ot = "t")]
            }
        }
        if (search) {
            db1[hs.grep(org_name, name)]
        }
        else {
            jg <- db1[hs.str_match(name, org_name), V2]
            if ((length(jg) == 1) && is.na(jg)) {
                .s(db1[hs.grep(paste0("(?i)", org_name), name)])
            }
            jg
        }
    }
})
"38_01_08_211229" <- local({
    msigdb <- list()
    function(org1) {
        hs.require("msigdbr")
        if (org1 %in% c("sheep")) {
            warning("[38_09_02_164956]", org1, ": 没有msigdb数据，需要准备")
            return()
        }
        if (!length(msigdb[[org1]])) {
            message("[2022_01_08_214144] 准备msigdb数据")
            if (0) {
                msigdbr_species()
                babelgene::orthologs
                babelgene::species
            }
            db1 <- msigdbr(species = switch(org1, human = "Homo sapiens", 
                mouse = "Mus musculus", rat = "Rattus norvegicus", 
                hs.browser("37.09.30.104253")))
            setDT(db1)
            db1[, `:=`(entrez_gene, hs.hsgly("基因对应/更新/ncbi.36_10_15_163409")(entrez_gene))]
            msigdb[[org1]][["gene"]] <<- hs.catch(db1[, {
                name1 <- unlist(.BY) %>% hs.clean %>% paste(collapse = "/")
                hs.throw(name1, .SD[, unique(.SD), .SDcols = c("gs_name", 
                  "entrez_gene")])
            }, by = .(gs_cat, gs_subcat)], named = 1)
            if (0) {
                db1[, unique(.SD)[, hs.table(gs_subcat, gs_cat)], 
                  .SDcols = c("gs_cat", "gs_subcat", "gs_name")] %>% 
                  .s
            }
            msigdb[[org1]][["description"]] <<- db1[, {
                unique(.SD)
            }, .SDcols = c("gs_name", "gs_description")]
        }
        msigdb[[org1]]
    }
})
"38_03_20_000049" <- function() {
    jg <- hs.hsgly("获取文件.39_05_25_150955")(c("gene/DATA/gene_orthologs.gz", 
        "pub/HomoloGene/current/homologene.data"))
    if (0) {
        dz.V <- hs.hsgly("获取文件.39_05_25_150955")(c("gene/DATA/gene_orthologs.gz", 
            "pub/HomoloGene/current/homologene.data"), only = "dz")
        .wk("curl -I", dz.V["homologene"])
    }
    jg
}
"36_06_08_154157" <- function(org1 = "human", update_db = 0) {
    if (org1 %in% c("sheep")) 
        return(hs.hsgly("ensembl/基因/基于biomart/ens.ncbi.38_09_05_104722")(org1 = org1))
    wj1 <- hs.hsgly("ftp.ncbi.nih.gov/gene/DATA.37_04_29_161915")("gene2ensembl.gz")
    if (0) {
        dz1 <- "https://ftp.ncbi.nlm.nih.gov/gene/DATA/gene2ensembl.gz"
        wj1 <- hs.wjm(dz1, dn = "~/swxx/sj/db/ncbi/gene")
        wj.updater(wj1, dz1)
        .ls(wj1)
        .head(wj1)
    }
    org1.code <- hs.hsgly("ncbi/物种/名称.编号.36_12_28_140636")(org1 = org1, 
        ot = "id")
    wj2 <- hs.wjm(wj1, append = paste0(".", org1))
    if (update_db) 
        hs.wj.rm(wj2)
    if (wj.update.reporter(wj2, wj1)) {
        cmd1 <- .sys(.sys.cat.text(wj1), "| perl -e", .q(paste0("\nwhile( <>) {\n  $flag = 0;\n  if( /^", 
            org1.code, "\\t/) {\n    @F = split( /\\t/);\n    print( $F[ 1], \"\\t\", $F[ 2], \"\\n\");\n    $flag = 1;\n  } else {\n    if( $flag) {\n      exit;\n    }\n  }\n}\n")), 
            "| uniq | sort -k 1n | uniq | gzip >", wj2)
        message(cmd1)
        .wk(cmd1)
    }
    wj2
}
"37_12_31_133053" <- function(org1 = "human") {
    org1 <- hs.hsgly("ensembl/物种/common_name.ncbi_tax_id.38_09_08_095424")(org1)
    org1 <- hs.hsgly("ncbi/物种/名称.编号.36_12_28_140636")(org1, 
        ot = "name")
    wjj_ncbi <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("ncbi")
    if (org1 %not.in% c("human", "mouse", "rat", "sheep", "goat", 
        "yeast")) 
        stop("[37_12_31_134129]")
    if (0) {
        a <- readLines("http://ftp.ncbi.nlm.nih.gov/gene/GENE_INFO/Mammalia")
        a <- readLines("https://ftp.ncbi.nlm.nih.gov/gene/DATA/GENE_INFO/Mammalia")
        "/Rattus_norvegicus.gene_info.gz"
        dz1 <- "ftp.ncbi.nih.gov/gene/GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz"
        hs.with_wd(hs.fp(wjj_ncbi, dirname(dz1)), {
            wj.updater(dz1 = dz1, co1 = 10)
        })
    }
    hs.cond(org1 %in% c("human", "mouse", "rat"), hs.hsgly("文件/获取/从ftp.37_04_29_185838")("ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia", 
        what = paste0(switch(org1, human = "Homo_sapiens", mouse = "Mus_musculus", 
            rat = "Rattus_norvegicus"), ".gene_info.gz"), wjj0 = wjj_ncbi), 
        org1 %in% c("yeast"), hs.hsgly("文件/获取/从ftp.37_04_29_185838")("ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Fungi", 
            what = paste0(switch(org1, yeast = "Saccharomyces_cerevisiae"), 
                ".gene_info.gz"), wjj0 = wjj_ncbi), org1 %in% 
            c("sheep", "goat"), {
            wj <- hs.hsgly("文件/获取/从ftp.37_04_29_185838")("ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia", 
                what = "All_Mammalia.gene_info.gz", wjj0 = wjj_ncbi)
            wj2 <- hs.fp(dirname(wj), paste0(org1, ".tsv.gz"))
            if (wj.update.reporter(wj2, wj)) {
                org1_id <- hs.hsgly("ncbi/物种/名称.编号.36_12_28_140636")(org1, 
                  ot = "id")
                cmd1 <- .sys("perl -e", .q(paste0("\nopen( wj, \"cat_wk ", 
                  wj, "| head |\");\n$a = readline( wj);\nprint( $a);\nclose( wj);\nopen( wj, \"cat_wk ", 
                  .sys.fpGood(wj), " |\");\n$flag1 = 0;\nwhile( <wj>) {\n  @F = split( \"\t\");\n  if( $F[ 0] eq \"", 
                  org1_id, "\") {\n    print;\n    $flag1 = 1;\n  } else {\n    if( $flag1) {\n      exit;\n    }\n  }\n}\nclose( wj);\n")), 
                  "| gzip >", .sys.fpGood(wj2))
                message(cmd1)
                .wk(cmd1)
            }
            wj2
        })
}
"39_05_25_122220" <- function(org1 = "human") {
    wjj2 <- hs.sjgly("39_05_25_122220")
    if (!hs.grepl(org1, "^\\d+$")) 
        org1 %<=>% hs.hsgly("ncbi/物种/名称.编号.36_12_28_140636")(x)
    wj2 <- hs.fp(wjj2, paste0(org1, ".tsv.gz"))
    wj1 <- hs.hsgly("获取文件.39_05_25_150955")("gene/DATA/gene2refseq.gz")
    if (wj.update.reporter(wj2, wj1)) {
        .head(wj1)
        hs.hsgly("1.信息学/文件/读写/提取/表格的子集.38_09_10_214239")(org1, 
            db.fp = wj1, wj2 = wj2, ci1 = 1, head = 1, consecutive = 1)
    }
    wj2
}
"39_01_20_130947" <- function(gene) {
    wjj2 <- hs.sjgly("39_01_20_130947")
    hs.lapply(gene, function(g1) {
        hs.msg(g1)
        wj2 <- hs.fp(wjj2, "html", paste0(g1, ".gz"))
        if (file.exists(wj2)) {
            a <- readLines(wj2)
        }
        else {
            dz1 <- paste0("https://www.ncbi.nlm.nih.gov/gene?db=gene&report=generif&term=", 
                g1)
            repeat {
                a <- readLines(dz1)
                a1 <- hs.grep(a, "^\\s*<div class=\"rprt-section\">")
                if (length(a1)) {
                  wk.dt.fwrite(a, wj2)
                  break
                }
            }
        }
        a1 <- hs.grepV(a, "^\\s*<div class=\"rprt-section\">")
        if (hs.grepl(a1, "(?i)No GeneRIF available for this Gene ID")) 
            return()
        wk.dt(XML::readHTMLTable(a1[1])[[1]])
    })
}
"39_05_25_143854" <- function(sub = {
}) {
    jg <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("ncbi")
    if (length(sub)) 
        jg <- hs.fp(jg, sub)
    jg
}
"36_12_28_140636" <- local({
    db <- {
    }
    function(org1 = "human", ot = "") {
        if (!length(db)) 
            local({
                db <- c(9606, "human", 10090, "mouse", 10116, 
                  "rat", 9940, "sheep", 9925, "goat", 559292, 
                  "yeast")
                db %<=>% matrix(2) %=>% t %=>% wk.dt
                wk.dt.setnames(db, c("id", "name"))
                db[, `:=`(id, as.numeric(id))]
                db <<- db
            })
        jg <- hs.switch(ot, "name", {
            if (hs.grepl(org1, "^\\d+$")) {
                db[match(org1, id), name]
            }
            else org1
        }, "id", {
            if (!hs.grepl(org1, "^\\d+$")) {
                db[match(org1, name), id]
            }
            else org1
        }, {
            if (hs.grepl(org1, "^\\d+$")) {
                db[match(org1, id), name]
            }
            else db[match(org1, name), id]
        })
        if (!length(jg)) 
            hs.browser("36.12.28.140800")
        jg
    }
})
"38_04_03_164704" <- function(prj1) {
    hs.require("rentrez")
    a <- entrez_search("sra", prj1)
    b <- entrez_summary("sra", a[["ids"]])
    sapply(b, function(x) hs.re(x[["runs"]], "(?i)srr\\d+"))
}
"39_05_25_150544" <- function(sub = {
}) {
    jg <- "ftp://ftp.ncbi.nih.gov"
    if (length(sub)) 
        jg <- hs.fp(jg, sub)
    jg
}
"39_05_25_150955" <- function(sub, only = {
}) {
    wjj2 <- hs.sjgly("39_05_25_150955")
    dz.V <- hs.hsgly("根地址.39_05_25_150544")(sub)
    dz.V %<=>% hs.fp_add_name
    if (identical(only, "dz")) 
        return(dz.V)
    wj_out.V <- dz.V %=>% hs.sub(".*://ftp[^/]*") %=>% paste0(wjj2, 
        x) %=>% hs.c(names(dz.V), x) %=>% {
        i1 <- !hs.grepl(x, "\\.gz$")
        if (any(i1)) 
            x[i1] %<=>% paste0(".gz")
        x
    }
    for (i1 in seq_along(dz.V)) {
        fn1 <- names(dz.V[i1])
        wj1 <- wj_out.V[fn1]
        if (!file.exists(wj1)) {
            dz1 <- dz.V[fn1]
            wj.updater(wj1, dz1, co1 = 30)
        }
    }
    wj_out.V
}
"2023_06_25_181210" <- function(dz) {
    {
        html <- xml2::read_html(dz)
        jg <- html %=>% xml2::xml_find_all("//pre/a/@href") %=>% 
            xml2::xml_text(x)[-1]
        return(jg)
    }
    if (0) {
        a <- readLines(dz)
        a %=>% tail(x, -hs.grep(x, "<pre>")[1]) %=>% head(x, 
            hs.grep(x, "</pre>")[1] - 1) %=>% hs.re("(?<=>).*") %=>% 
            hs.re(".*(?=<)")
    }
}
"37_02_06_192853" <- function(id1) {
    hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("ncbi"), 
        "geo/gse", id1)
}
"36_08_13_095708" <- function(id1, wjj1 = {
}) {
    wjj0 <- hs.fp("ftp.ncbi.nlm.nih.gov/geo/series", hs.sub(id1, 
        "\\d{1,3}$", "nnn"), id1)
    if (!length(wjj1)) 
        wjj1 <- hs.fp(hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("ncbi"), 
            wjj0)
    wj <- hs.wjj.ls(wjj1)
    if (length(wj)) {
        wj_type <- hs.wjm(wj, dn = 0, ext = 0) %=>% hs.re("[^_\\.]*$")
        info <- list(type = wj_type, wj = wj) %=>% wk.dt.as
    }
    else {
        url <- c("matrix", "soft") %=>% hs.c(x, paste0("https://", 
            wjj0, "/", x))
        url %<=>% lapply(x, . %=>% hs.fp(x, hs.hsgly("工作/生物学/数据库/ncbi/geo/ftp.wj.2023_06_25_181210")(x)))
        info <- url %=>% wk.dt.fromList(nm = c("type", "url"))
        info %<=>% hs.ij(x, x[, wj %=>% basename %=>% order])
        info[, `:=`("wj", hs.wjm(url, dn = hs.wj(wjj1)))]
        info[, for (i in .I) {
            if (!file.exists(wj[i])) 
                hs.hsgly("1.信息学/文件/获取/用aria2c.2023_06_26_120607")(url[i], 
                  wjj1)
        }]
        if (0) {
            info["soft", wj[1], on = "type"] %=>% .less
            info["matrix", wj[1], on = "type"] %=>% .less
        }
    }
    if (info[type == "matrix", .N > 1]) {
        hs.msg("[2023_08_07_171853|考虑用]hs.hsgly( 'series_matrix到表格.39_03_02_202432')")
        return(info)
    }
    si <- info["matrix", hs.hsgly("series_matrix到表格.39_03_02_202432")(series_matrix_wj = wj), 
        on = "group"]
    if (0) {
        si %=>% dim
        si[1]
        lapply(si, unique) %=>% {
            .[sapply(., length) == 1]
        }
        sapply(si, uniqueN) %=>% hs.table
        a1 <- lapply(si, hs.table) %=>% x[sapply(, length) == 
            7]
    }
    list(file = info[, wj], si = si)
}
"36_09_22_105025" <- function(gse_id1, wjj1 = hs.hsgly("gse号1.文件夹.37_02_06_192853")(gse_id1)) {
    gse.info <- hs.hsgly("gse号.信息.36.08.13.095708")(gse_id1, 
        hs.wjj(wjj1))
    if (hs.is_empty(gse.info[["info"]])) {
        wj1 <- gse.info[["file"]]["soft"]
        a1 <- readLines(wj1)
        hs.grepV(a1, "^!Series_sample_id") %>% hs.re("(?i)gsm\\d+")
    }
    else gse.info[["info"]][, Sample_geo_accession]
}
"36_08_13_103848" <- function(id.V, wjj1) {
    if (!missing(wjj1)) {
        wj <- hs.fp(wjj1, "gsm.srr.tsv.gz")
        if (file.exists(wj)) {
            dt1 <- wk.fread(wj)
            if (dt1[, id.V %all.in% gsm]) 
                return(dt1[match(id.V, gsm)])
        }
    }
    hs.require("rentrez")
    if (0) {
        help(package = "rentrez")
        db <- entrez_dbs()
        entrez_info("sra")
        a <- entrez_search("bioproject", "PRJNA773222")
        a <- entrez_search("sra", "PRJNA773222")
        a <- entrez_search("sra", "PRJNA737757", retmax = 56)
        a <- entrez_search("sra", "PRJNA737757")
        names(a)
        XML::xmlToList(a[["file"]])[["IdList"]] %=>% length
        a[["QueryTranslation"]]
        a[["ids"]]
        b <- entrez_fetch(db = "sra", id = a[["ids"]], rettype = "xml", 
            parsed = TRUE, retmax = a[["count"]])
        str(b)
        .l(b)
        "<EXPERIMENT alias="
        apropos("method")
        showMethods(class = class(b))
        utils::.S3methods(class = class(b))
        help(package = "XML")
        xml_list <- XML::xmlToList(b)
        xml_list %=>% names
        a
        methods::getMethods
        utils::methods
        apropos("function")
        `?`(methods::getMethods)
        b <- entrez_summary("sra", a[["ids"]])
        b <- entrez_summary("sra", "SRX11146001")
        b <- entrez_summary("bioproject", a[["ids"]])
        b <- as.data.table(b)
        names(b)
        entrez_summary("sra", a[["ids"]])
        a <- entrez_search("sra", "SRX12713559")
        b <- entrez_summary("sra", a[["ids"]])
        as.data.table(b)
    }
    jg <- local({
        a1 <- lapply(id.V, . %>% {
            hs.msg(., 0)
            entrez_search("sra", .)
        })
        a2 <- entrez_summary("sra", sapply(a1, "[[", "ids"))
        if (length(id.V) > 1) {
            rbindlist(lapply(a2, as.data.table))
        }
        else as.data.table(a2)
    })
    hs.msg("done", 0)
    jg[, `:=`("gsm", expxml %>% hs.re("(?i)gsm\\d+", no_match = ""))]
    jg[, `:=`("srr", runs %>% hs.re("(?i)srr\\d+", no_match = ""))]
    if (!missing(wjj1)) 
        wk.dt.fwrite(jg, wj)
    return(jg[match(id.V, gsm)])
    if (0) {
        jg[1]
        a1 <- jg[1, hs.gre(expxml, "<[^<>]+>")][[1]]
        a2 <- jg[1, hs.gre(expxml, "(?<=>)[^<>]*(?=<)")][[1]]
        v1 <- head(as.vector(rbind(a1, a2)), -1)
        start.i <- hs.grep(v1, "<[^/]")
        end.i <- hs.grep(v1, "</.*>")
        v1[start.i]
        v1[end.i]
        hs.grepV(v1, "/>$")
        hs.grepV(v1, "/.+>$")
        end.i <- hs.grep(v1, "^</")
        i2 <- end.i[1]
        sapply(end.i, function(i2) {
        })
        function(tree = list(), v1) {
            if (hs.grepl(v1[1], "/>$")) {
            }
        }
    }
}
"39_03_02_202432" <- function(series_matrix_wj, clean = 1) {
    a1 <- readLines(series_matrix_wj)
    jg <- hs.grepV(a1, "^!") %=>% hs.grepV("\t[^\t]*\t") %=>% 
        strsplit("\t") %=>% hs.c(sapply(x, "[", 1) %=>% hs.sub("."), 
        lapply(x, . %=>% trim(x[-1], "\""))) %=>% wk.dt.setDT
    if (clean) 
        wk.dt.clean(jg)
    else jg
}
"37_12_04_120750" <- function(bin = {
}, help = 0) {
    wjj2 <- hs.wjj(hs.hsgly("生物学/根文件夹.37_07_25_182333")(), 
        "rj")
    if (!hs.file_system_is_linkable(wjj2)) 
        wjj2 <- "~/rj"
    sratoolkit.wjj <- hs.wjj.ls(wjj2, "^sratoolkit", wjj = 1)
    if (!length(sratoolkit.wjj)) {
        dz1 <- "ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-ubuntu64.tar.gz"
        wj_tar <- wk.axel(dz1, wjj2)
        hs.with_wd(wjj2, {
            .wk("tar -xzvf", basename(dz1))
        })
        sratoolkit.wjj <- hs.wjj.ls(wjj2, "^sratoolkit", wjj = 1)
        hs.browser("37.12.04.132445")
        .wk(hs.fp(sratoolkit.wjj, "bin", "vdb-config"), "--interactive")
    }
    wj1 <- do.call(hs.fp, as.list(c(sratoolkit.wjj, "bin", bin)))
    if (help) {
        .wk(wj1, "--help | less -Ni")
    }
    else wj1
}
"37_12_08_124805" <- function(id, wjj2 = ".", chk = 1) {
    hs.lapply(id, function(id1) {
        hs.msg(id1)
        hs.with_wd(hs.wjj(wjj2, id1), {
            repeat {
                wj2 <- hs.wjj.ls(".", paste0("^", id1, "(_\\d+)\\.fq(\\.gz)?$"))
                if (!length(wj2)) 
                  wj2 <- hs.wjj.ls(".", paste0("^", id1, "(_\\d+)\\.fastq(\\.gz)?$"))
                if (!length(wj2)) 
                  wj2 <- hs.wjj.ls(".", paste0("^", id1, "\\.f(ast)?q(\\.gz)?$"))
                if (!length(wj2)) {
                  wj2 <- hs.wjj.ls(".", paste0("^", id1, "(?i)\\.sra$"))
                  if (length(wj2)) {
                    .wk(hs.hsgly("sratoolkit.37_12_04_120750")("fastq-dump"), 
                      "-v", "-F", "--defline-qual", .q("+"), 
                      "--readids", "--split-3", "--gzip", "-O .", 
                      wj2)
                    wj2 <- hs.wjj.ls(".", paste0("^", id1, "\\.f(ast)?q(\\.gz)?$"))
                  }
                  else {
                    hs.msg("下载")
                    .wk(hs.hsgly("sratoolkit.37_12_04_120750")("fastq-dump"), 
                      "-v", "-F", "--defline-qual", .q("+"), 
                      "--readids", "--split-3", "--gzip", "-O .", 
                      id1)
                    hs.msg("结束", 0)
                    next
                    if (0) {
                      .wk(hs.hsgly("sratoolkit.37_12_04_120750")("fastq-dump"), 
                        "-v", "-Z", "--readids", "-O .", id1, 
                        "| gzip >", wj2)
                    }
                  }
                }
                if (chk) {
                  hs.msg(hs.pastec("[gzip -t]", .q(wj2)), 0)
                  if (all(hs.hsgly("文件/检查完整性/gz.37_12_08_111112")(wj2))) {
                    hs.msg("结束", 0)
                    break
                  }
                  else {
                    message("\t\t", normalizePath(wj2), "获取不完整；现删除后开始重新获取")
                    hs.wj.rm(wj2)
                  }
                }
                else break
            }
            normalizePath(wj2)
        })
    })
}
"37_02_09_134736" <- function(id1, wjj2 = ".") {
    hs.with_wd(wjj2, {
        sapply(id1, function(id1) {
            .wk(hs.hsgly("sratoolkit.37_12_04_120750")("prefetch"), 
                "--max-size 100G", "-v", "-O .", id1)
        })
    })
}
"36_08_13_123902" <- function(sra_wj, fq_wjj = {
}) {
    wjj1 <- "/dev/shm/36.07.21.071129"
    wjj2 <- "/dev/shm/36.07.21.071143"
    on.exit(hs.wj.rm(c(wjj1, wjj2), 1), add = TRUE)
    for (i1 in sra_wj %>% seq_along) {
        .cat.time()
        hs.say(i1, "of", length(sra_wj))
        fq.wjj1 <- if (length(fq_wjj)) 
            fq_wjj
        else dirname(sra_wj[i1])
        if (length(fq_wj.hs.wjj(fq.wjj1, str2include = paste0(basename(sra_wj[i1]), 
            "(_1)?\\.")))) 
            next
        hs.browser("37.12.04.211024")
        file.symlink()
        sra_wj[i1]
        fq.wjj1
        .wk(hs.hsgly("sratoolkit.37_12_04_120750")("fasterq-dump"), 
            "-t", hs.wjj(wjj1), "-O", hs.wjj(wjj2), "-e", hs.NC(), 
            "-p", "-S", "-v", sra_wj[i1])
        hs.wjj.ls(wjj2) %>% lapply(. %>% {
            .wk("pigz", .)
        })
        hs.wjj.ls(wjj2) %>% {
            hs.wj.move(., hs.wjm(., dn = fq.wjj1))
        }
        .cat.time()
    }
}
"38_08_27_094838" <- function(name, wjj_mid = {
}, wjj_mid.rm = 1) {
    name.db <- name %>% tolower %>% unique
    sj.wjj <- "~/swxx/sj/db/pubchem/ftp.ncbi.nlm.nih.gov/pubchem/Compound/Extras"
    if (!length(wjj_mid)) 
        wjj_mid <- hs.wj.temp()
    if (wjj_mid.rm) 
        on.exit(hs.wj.rm(wjj_mid, 1))
    anno_wj <- hs.with_wd(hs.wjj(wjj_mid), {
        wj1 <- "name.txt.gz"
        if (!file.exists(wj1)) 
            cat(name.db, sep = "\n", file = .sys("| gzip >", 
                .q(wj1)))
        wj <- c("CID-Title.gz", "CID-IUPAC.gz", "CID-MeSH.gz", 
            "CID-Synonym-filtered.gz", "CID-Synonym-unfiltered.gz")
        hs.fp(sj.wjj, wj) %>% wk.mclapply(function(wj_db) {
            wj2 <- hs.wjm(wj1, append = hs.wjm(wj_db, dn = 0, 
                ext = 0), append_sep = ".", extN = 2, ext = "tsv.gz")
            if (!file.exists(wj2)) {
                message(wj2)
                .wk("perl -e", .q(paste0("\nopen( WJ1, ", .q(.sys.cat.text(wj1, 
                  pipe = 1)), ");\nwhile( <WJ1>) {\n  chomp;\n  $name{ $_} = 1;\n}\nclose( WJ1);\n\nopen( wj1, ", 
                  .q(.sys.cat.text(wj_db, pipe = 1)), ");\nwhile( <wj1>) {\n  chomp;\n  @F = split( \"\t\");\n  @G = split( /; */, lc( $F[ 1]));\n  for( $i = 0; $i <= $#G; $i++) {\n    if( exists( $name{ $G[ $i]})) {\n      print( $F[ 0], \"\t\", $G[ $i], \"\n\");\n    }\n  }\n}\nclose( wj1);\n")), 
                  "| gzip >", .sys.fpGood(wj2))
            }
            normalizePath(wj2)
        })
    })
    {
        jg <- wk.dt(name.db, "")
        wk.dt.setnames(jg, c("name", "pubchem.cid"))
        hs.for(wj1, anno_wj, {
            message(wj1)
            a <- .try(wk.fread(wj1, cn = 0))
            if (is(a, "try-error")) 
                next
            wk.dt.setnames(a, c("id", "name"))
            a <- a[, paste(sort(unique(id)), collapse = ";"), 
                by = name]
            wk.dt.setnames(a, 2, "id")
            ri <- jg[, hs.clean(pubchem.cid, ot = "i", inv = 1)]
            c1 <- wk.dt.subset.na(a, jg[ri, name], "name")[, 
                id]
            wk.dt.set(jg, ri, "pubchem.cid", c1)
        })
    }
    jg
}
"38_08_27_104536" <- function(name) {
    name.db <- name %>% tolower %>% unique
    jg <- wk.dt(name.db, "")
    wk.dt.setnames(jg, c("name", "id"))
    hmdb.wj <- c("title", "name") %>% sapply(. %>% {
        hs.hsgly("生物学/数据库/hmdb/id.info.38_02_27_114727")(what = .)
    })
    hmdb.title <- hs.hsgly("生物学/数据库/hmdb/id.info.38_02_27_114727")(what = "title") %>% 
        wk.fread
    hmdb.title[, `:=`(title, tolower(title))]
    hs.for(wj1i, seq_along(hmdb.wj), {
        ri <- jg[, hs.clean(id, ot = "i", inv = 1)]
        db1 <- hmdb.wj[wj1i] %>% wk.fread
        cn2 <- names(db1)[2]
        db1[, `:=`(c(cn2), get(cn2) %>% tolower)]
        a <- wk.dt.subset(db1, jg[ri, name], 2)
        a <- a[, paste(names(a)[1] %>% get %>% sort, collapse = ";"), 
            by = c(names(a)[2])]
        c1 <- wk.dt.subset.na(a, jg[ri, name])[, V1]
        wk.dt.set(jg, ri, 2L, c1)
    })
    wk.dt.setnames(jg, "id", "hmdb.id")
}
"38_08_27_102154" <- function(name) {
    name.db <- name %>% tolower %>% unique
    kegg <- hs.hsgly("kegg/化合物/id.name.38_02_24_131718")() %>% 
        wk.xls.read
    kegg[, `:=`("name", tolower(name))]
    kegg[hs.grep(kegg_id, "(?i)d"), `:=`("name2", hs.sub(name, 
        " *\\([^\\(\\)]*\\)$"))]
    jg <- wk.dt(name.db, "")
    wk.dt.setnames(jg, c("name", "kegg.id"))
    hs.for(ci1, 2:3, {
        ri <- jg[, hs.clean(kegg.id, ot = "i", inv = 1)]
        a <- wk.dt.subset(kegg, jg[ri, name], ci1)[, .SD, .SDcols = c(1, 
            ci1)]
        a <- a[, paste(kegg_id %>% sort, collapse = ";"), by = c(names(a)[2])]
        c1 <- wk.dt.subset.na(a, jg[ri, name])[, V1]
        wk.dt.set(jg, ri, 2L, c1)
    })
    jg
}
"38_09_04_100304" <- function() {
    wjj.38_09_04_191404 <- hs.hsgly("pubchem/cid.page/page.38_09_02_073647")()
    wj2 <- hs.fp(wjj.38_09_04_191404, "cid.cas_id.tsv.gz")
    wj <- hs.wjj.ls(wjj.38_09_04_191404, "^(?i)\\d+\\.json\\.gz$")
    if (wj.update.reporter(wj2, wj)) {
        names(wj) <- hs.wjm(wj, dn = 0, ext = 0, extN = 2)
        if (0) {
            wj["11615849"]
            i <- match("11615849", names(wj))
        }
        a <- wk.mclapply(seq(length(wj), 1), function(i) {
            message(i)
            wj1 <- wj[i]
            json <- .wk("cat_wk", .sys.fpGood(wj1), intern = 1)
            a <- jsonlite::fromJSON(json)
            j <- match("Names and Identifiers", a[[1]][["Section"]][["TOCHeading"]])
            j1 <- match("Other Identifiers", a[[1]][["Section"]][["Section"]][[j]][, 
                "TOCHeading"], nomatch = 0)
            if (!j1) 
                return()
            a1 <- a[[1]][["Section"]][["Section"]][[j]][[j1, 
                "Section"]]
            j2 <- match("CAS", a1[, "TOCHeading"], nomatch = 0)
            if (!j2) 
                return()
            cas_id <- a1[[j2, "Information"]][, "Value"] %>% 
                unlist %>% hs.clean(U = 1)
            if (length(cas_id) > 1) 
                cas_id %<>% paste(collapse = ";")
            if (length(cas_id) != 1) 
                hs.browser("38.09.04.190558")
            c(names(wj1), cas_id)
        })
        jg <- do.call("rbind", a)
        jg %<>% {
            .[order(as.numeric(.[, 1])), , drop = 0]
        }
        wk.dt.fwrite(jg, wj2, cn = 0)
    }
    wj2
}
"38_09_19_234900" <- function() {
    wjj.38_09_04_191404 <- hs.hsgly("pubchem/cid.page/page.38_09_02_073647")()
    wj2 <- hs.fp(wjj.38_09_04_191404, "cid.name.tsv.gz")
    wj <- hs.wjj.ls(wjj.38_09_04_191404, "^(?i)\\d+\\.json\\.gz$")
    hs.browser("38.09.20.001307")
    if (wj.update.reporter(wj2, wj)) {
        names(wj) <- hs.wjm(wj, dn = 0, ext = 0, extN = 2)
        if (0) {
            wj["11615849"]
            i <- match("11615849", names(wj))
        }
        a <- wk.mclapply(seq(length(wj), 1), function(i) {
            message(i)
            wj1 <- wj[i]
            json <- .wk("cat_wk", .sys.fpGood(wj1), intern = 1)
            a <- jsonlite::fromJSON(json)
            j <- match("Names and Identifiers", a[[1]][["Section"]][["TOCHeading"]])
            j1 <- match("Other Identifiers", a[[1]][["Section"]][["Section"]][[j]][, 
                "TOCHeading"], nomatch = 0)
            if (!j1) 
                return()
            a1 <- a[[1]][["Section"]][["Section"]][[j]][[j1, 
                "Section"]]
            j2 <- match("CAS", a1[, "TOCHeading"], nomatch = 0)
            if (!j2) 
                return()
            cas_id <- a1[[j2, "Information"]][, "Value"] %>% 
                unlist %>% hs.clean(U = 1)
            if (length(cas_id) > 1) 
                cas_id %<>% paste(collapse = ";")
            if (length(cas_id) != 1) 
                hs.browser("38.09.04.190558")
            c(names(wj1), cas_id)
        })
        jg <- do.call("rbind", a)
        jg %<>% {
            .[order(as.numeric(.[, 1])), , drop = 0]
        }
        wk.dt.fwrite(jg, wj2, cn = 0)
    }
    wj2
}
"38_09_02_073647" <- function(id = {
}) {
    wjj2.38_09_02_073916 <- "~/swxx/sj/db/pubchem/Compound/cid_page" %>% 
        hs.wjj
    wj_is_ok <- function(wj1, id1) {
        file.exists(wj1) && {
            a <- .wk("cat_wk", .sys.fpGood(wj2), intern = 1)
            length(a) && (tail(a, 1) == "}") && {
                a[4] %=>% hs.re("\\d+") %=>% (x == id1)
            }
        }
    }
    for (i in seq_along(id)) {
        hs.msg(i)
        id1 <- id[i]
        wj2 <- hs.fp(wjj2.38_09_02_073916, paste0(id1, ".json.gz"))
        dz1 <- paste0("https://pubchem.ncbi.nlm.nih.gov/rest/pug_view/data/compound/", 
            id1, "/JSON/?response_type=save&response_basename=compound_CID_", 
            id1)
        j <- 0
        repeat {
            if (wj_is_ok(wj2, id1)) 
                break
            if (0) {
                a <- .try(readLines(dz1))
                if (!hs.is_try_error(a)) 
                  break
                cat(a, sep = "\n", file = .sys("| gzip >", .sys.fpGood(wj2)))
            }
            if (0) {
                .wk("axel -o", .sys.fpGood(hs.wjm(wj2, ext = 0)), 
                  .q(dz1))
                .wk("gzip -f", .sys.fpGood(hs.wjm(wj2, ext = 0)))
            }
            .wk("wget -O -", .q(dz1), "| gzip >", .sys.fpGood(wj2))
            j <- j + 1
            if (j > 10) 
                break
        }
    }
    wjj2.38_09_02_073916
}
"38_03_06_205116" <- function(what = "name", ..mem = 1) {
    if (0) {
        db1.type <- "compound"
        hs.lapply("compound,substance", function(db1.type) {
            hs.with_wd(hs.fp("~/swxx/sj/db/pubchem", db1.type), 
                {
                  dz1 <- hs.fp("https://ftp.ncbi.nlm.nih.gov/pubchem", 
                    hs.title(db1.type), "CURRENT-Full/XML")
                  html.fp <- "the.html.gz"
                  if (wj.update.reporter(html.fp, co1 = 30)) 
                    hs.wj.rm(html.fp)
                  html <- hs.evalLazy({
                    readLines(dz1)
                  }, fp = html.fp, hs4save = function(x1, fp1) cat(x1, 
                    sep = "\n", file = fp1), hs4load = function(fp1) readLines(fp1))
                  {
                    {
                      gz.fp <- hs.re(html, "[^\"]*\\.xml\\.gz(?=\")")
                      gz.fp %>% length
                      gz.fp %>% {
                        .[!file.exists(.)]
                      }
                      i1 <- 1
                      gz.fp1 <- "Compound_155000001_155500000.xml.gz"
                      hs.browser("38.03.06.075107")
                      wk.mclapply(seq_along(gz.fp), function(i1) {
                        gz.fp1 <- gz.fp[i1]
                        gz.ok.fp1 <- paste0(gz.fp1, ".ok")
                        if (wj.update.reporter(gz.ok.fp1, gz.fp1)) {
                          a <- .wk("gzip -t", .q(gz.fp1), intern = 1)
                          if (is.null(attr(a, "status"))) {
                            cat("", file = gz.ok.fp1)
                          }
                        }
                      }, mc.cores = 4)
                      seq_along(gz.fp) %>% vapply(. %>% {
                        gz.fp1 <- gz.fp[.]
                        gz.ok.fp1 <- paste0(gz.fp1, ".ok")
                        if (file.exists(gz.ok.fp1)) 
                          ""
                        else gz.fp1
                      }, "") %>% hs.clean %T>% cat(sep = "\n") %>% 
                        {
                          hs.fp(dz1, .)
                        } %>% cat(sep = "\n")
                    }
                    if (0) {
                      md5.dz <- hs.fp(dz1, hs.re(html, "[^\"]*\\.xml\\.gz.md5(?=\")"))
                      md5.fp <- basename(md5.dz)
                      md5.dz %>% {
                        .[!file.exists(md5.fp)]
                      } %>% vapply(. %>% {
                        .wk("wget -c", .)
                        TRUE
                      }, TRUE)
                      md5.fp[1]
                      .wk("md5sum -c", .q(md5.fp[1]))
                      .wk("gzip -cd", "Compound_000000001_000500000.xml.gz", 
                        "| wc -l")
                    }
                  }
                  if (0) {
                    hs.fp(dz1, hs.re(a, "[^\"]*\\.xml\\.gz(?=\")"))[-c(1:51)] %>% 
                      head(50) %>% cat(sep = "\n", file = "~/a.txt")
                    head(hs.re(a, "[^\"]*\\.xml\\.gz(?=\")") %not% 
                      hs.wjj.ls("/Volumes/tf500/pubchem", F = 0), 
                      Inf) %>% {
                      hs.fp(dz1, .)
                    } %>% cat(sep = "\n", file = "~/tmp/a.txt")
                    {
                      dz1 <- "https://ftp.ncbi.nlm.nih.gov/pubchem/Substance/CURRENT-Full/XML"
                      a <- readLines(dz1)
                      hs.re(a, "[^\"]*\\.xml\\.gz(?=\")") %>% 
                        {
                          hs.fp(dz1, .)[1:50]
                        } %>% cat(sep = "\n", file = "~/tmp/a.txt")
                    }
                  }
                })
        })
    }
    {
        wj <- wk.dt(wj = hs.wjj.ls("~/swxx/sj/db/pubchem/compound", 
            "\\.xml\\.gz$"))
        wj[, `:=`("size", file.size(wj))]
        wj[, `:=`("ok", file.exists(paste0(wj, ".ok")))]
        if (wj[c(!ok), .N]) 
            hs.browser("38.03.06.205644")
        setorderv(wj, "size", -1)
        if (0) {
            what <- "name"
            wj1 <- wj[1, wj]
            .wc(wj1)
            .less(wj1)
            wj1 <- "/Volumes/tf500/pubchem/Compound_000000001_000500000.xml.gz"
            wj1 <- "~/swxx/sj/db/pubchem/compound/Compound_000000001_000500000.xml.gz"
            wj[, hs.wjm(wj, append = ".id_name.tsv") %>% {
                .[file.exists(.)]
            }] %>% .ls
            wj[, hs.wjm(wj, append = ".id_name.tsv") %>% {
                .[file.exists(.)]
            } %>% file.size %>% sum %>% `/`(1e+09)]
            a <- wj[, hs.wjm(wj, append = ".id_name.tsv") %>% 
                {
                  .[file.exists(.)][1]
                }] %>% wk.fread
            a[, name %>% nchar %>% hs.table %>% {
                max(.)/sum(.)
            }]
        }
        id_name.wj <- wj[, seq(.N)] %>% wk.mclapply(function(i1) {
            wj1 <- wj[i1, wj]
            wj2 <- hs.wjm(wj1, append = ".id_name.tsv")
            if (wj.update.reporter(wj2, wj1)) {
                hs.require("XML")
                line_start <- "  <PC-Compound>"
                line_end <- "  </PC-Compound>"
                chunk.size <- 1e+06
                con1 <- .sys("pigz -cd", .sys.fpGood(wj1), pipe = "r")
                remained <- readLines(con1, 1000)
                remained <- tail(remained, -match(line_start, 
                  remained) + 1)
                lines_read <- 0
                acc <- expandingList()
                repeat {
                  Lines <- readLines(con1, chunk.size)
                  if (!length(Lines) && !length(remained)) 
                    break
                  if (length(Lines)) {
                    lines_read <- lines_read + length(Lines)
                    message("\tLines read: ", lines_read, ".")
                  }
                  else {
                    if (length(remained)) {
                      Lines <- remained
                      remained <- NULL
                    }
                  }
                  if (length(remained)) {
                    Lines <- c(remained, Lines)
                    remained <- NULL
                  }
                  sep.i <- which(Lines %in% line_start)
                  if (!length(sep.i)) {
                    remained <- c(remained, Lines)
                    next
                  }
                  a1 <- hs.sepSeq(Lines, sep.i = sep.i)
                  i2do.max <- length(a1)
                  if (a1[[i2do.max]] %=>% tail(1) %=>% (x != 
                    line_end)) {
                    remained <- a1[[i2do.max]]
                    i2do.max <- i2do.max - 1
                  }
                  if (i2do.max > 0) {
                    a2 <- wk.mclapply(1:i2do.max, function(i1) {
                      xml <- xmlRoot(xmlParse(a1[[i1]]))
                      if (0) {
                        a1[[i1]] %>% cat(sep = "\n", file = "~/tmp/a.txt")
                        names(xml) %>% hs.grepV("(?i)pubchem")
                        xml %>% names
                        xml[["PC-Compound_props"]][["PC-InfoData"]]
                        apropos("xml") %>% .sh
                        names(xml)[9]
                        df1 <- xmlToDataFrame(xml[[9]])
                        xml[10]
                        names(xml) %>% length
                        names(xml) %>% hs.table
                        df1 <- xmlToDataFrame(xml[["PC-Compound_props"]])
                        getNodeSet(xml)
                        df1[6, ]
                        .s(df1)
                        xmlToList(xml[["PC-Compound_props"]])[[7]]
                        xmlToList(xml[["PC-Compound_props"]]) %>% 
                          lapply(. %>% {
                            .[["PC-InfoData_value"]][["PC-InfoData_value_sval"]]
                          })
                      }
                      id1 <- xmlToDataFrame(xml[[1]])[, 1]
                      if (id1 %in% 439918) {
                        hs.browser("38.03.12.114135")
                        wk.less(Lines)
                      }
                      v1 <- switch(what, name = {
                        dt1 <- xmlToDataFrame(xml[["PC-Compound_props"]]) %>% 
                          setDT
                        dt1[hs.grep(`PC-InfoData_urn`, "(?i)IUPAC Name"), 
                          `PC-InfoData_value` %>% {
                            .[!duplicated(tolower(.))]
                          }]
                      })
                      if (0) {
                        if (length(v1)) 
                          acc[["add"]](id1, unname(v1))
                        TRUE
                      }
                      if (length(v1)) 
                        data.table(id1, v1)
                    }, mc.cores = 1)
                    jg <- rbindlist(unname(a2))
                    setnames(jg, c("id", "name"))
                    acc[["add"]](jg)
                  }
                }
                close(con1)
                jg <- acc[["get"]]()
                jg[[1]] %>% wk.dt.fwrite(wj2)
                if (length(jg) > 1) 
                  vapply(2:length(jg), function(i1) {
                    wk.dt.fwrite(jg[[i1]], wj2, add = 1)
                    TRUE
                  }, TRUE)
            }
            wj2
        }, mc.cores = 40)
        {
            wjj2.38_03_08_220212 <- wj[1, dirname(wj) %>% hs.fp("name_id")]
            wj2 <- hs.fp(wjj2.38_03_08_220212, "breaks.gz")
            wj_in <- wj[1, dirname(wj) %>% hs.wjj.ls("\\.xml\\.id_name\\.tsv\\.gz$")]
            if (wj.update.reporter(wj2, wj_in)) 
                local({
                  a <- .sys("find", wj[1, dirname(wj)], "-iname", 
                    "*.xml.id_name.tsv.gz", "| xargs -I {} cat_wk {}", 
                    "| perl -e", .q("while( <>) {\n  $i1 = index( $_, \"\t\");\n  $ht{ length( $_) - $i1 - 2} += 1;\n}\nforeach $k ( keys %ht) {\n  print( $k, \"\t\", $ht{ $k}, \"\n\");\n}\n"), 
                    "| sort -t \"\t\" -k 1n")
                  if (0) {
                    .wk("echo a | tee >&2", intern = 1)
                    .wk("echo a | tee >&1", intern = 1)
                    a <- wk.fread(cmd = .sys("find", wj[1, dirname(wj)], 
                      "-iname", "*.xml.id_name.tsv.gz", "| head -n 1", 
                      "| xargs -I {} cat_wk {}", "| perl -e", 
                      .q("while( <>) {\n  $i1 = index( $_, \"\t\");\n  $ht{ length( $_) - $i1 - 2} += 1;\n}\nforeach $k ( keys %ht) {\n  print( $k, \"\t\", $ht{ $k}, \"\n\");\n}\n")))
                    .wk("find", wj[1, dirname(wj)], "-iname", 
                      "*.xml.id_name.tsv.gz", "| head -n 1")
                    "/cluster/home/kweng/swxx/sj/db/pubchem/compound/Compound_050000001_050500000.xml.id_name.tsv.gz" %>% 
                      .ls
                    "/cluster/home/kweng/swxx/sj/db/pubchem/compound/Compound_050000001_050500000.xml.gz" %>% 
                      .ls
                    "/cluster/home/kweng/swxx/sj/db/pubchem/compound" %>% 
                      hs.wjj.ls("html") %>% {
                      .wk("less", .)
                    }
                    a[, sum(V1 * V2)]
                  }
                  dt1 <- wk.fread(cmd = a)
                  dt1[V1 %in% 4, `:=`(V2, V2 - wj[1, dirname(wj) %>% 
                    hs.wjj.ls("\\.xml\\.id_name\\.tsv\\.gz$") %>% 
                    length])]
                  if (0) {
                    dt1[, sum(V1 * V2)/1e+09]
                  }
                  breaks <- dt1[, V2 %>% {
                    if (0) {
                      co1 <- sum(.) %=>% (x/100)
                      cs <- cumsum(.)
                    }
                    cs <- cumsum(as.numeric(V1 * V2))
                    co1 <- sum(V1 * V2)/100
                    if (0) {
                      I <- sapply(co1 * seq(100), . %>% {
                        max(which(cs < .))
                      }) %>% unique
                    }
                    I <- local({
                      acc <- rep(0, 100 * 1.5)
                      acc.i <- 1
                      i1 <- i0 <- 0
                      repeat {
                        i1 <- i1 + 1
                        if (i1 > length(cs)) 
                          break
                        sum1 <- cs[i1] - {
                          if (i0 > 0) 
                            cs[i0]
                          else 0
                        }
                        if (sum1 > co1) {
                          acc[acc.i] <- i1
                          acc.i <- acc.i + 1
                          i0 <- i1
                          next
                        }
                        if (i1 < length(cs)) {
                          sum2 <- cs[i1 + 1] - {
                            if (i0 > 0) 
                              cs[i0]
                            else 0
                          }
                          if (sum2 > co1) {
                            if ((sum1/co1) > 0.5) {
                              acc[acc.i] <- i1
                              acc.i <- acc.i + 1
                              i0 <- i1
                              next
                            }
                          }
                        }
                      }
                      acc[acc > 0] %or% length(cs)
                    })
                    if (tail(I, 2) %=>% {
                      cs[x[2]] - cs[x[1]]
                    } %=>% ((x/co1) < 0.5)) 
                      I <- I[1 - length(I)]
                    I <- head(I, -1)
                    if (0) {
                      diff(cs[I[1:2]])/co1
                      cs[I] %>% diff %>% (`/`(co1)) %>% summary
                      length(I)
                    }
                    V1[I]
                  }]
                  wk.dt.fwrite(list(breaks), wj2)
                })
            breaks <- c(1, wk.fread(wj2)[[1]], Inf)
            wj2 <- local({
                f1 <- hs.cut(breaks, breaks, hi = 1)
                sapply(seq(length(breaks) - 1), function(i2) {
                  co1 <- breaks[i2] + switch(substr(levels(f1)[i2], 
                    1, 1), `[` = 0, `(` = 1)
                  co2 <- breaks[i2 + 1] + switch(hs.re(levels(f1)[i2], 
                    ".$"), `]` = 0, `)` = -1)
                  hs.fp(wjj2.38_03_08_220212, hs.pastec0(if (is.infinite(co2)) 
                    c(co1, "+")
                  else paste(co1 %or% co2, collapse = "-"), ".tsv.gz"))
                })
            })
            if (0) {
                basename(wj2) %>% .ss
            }
            if (wj.update.reporter(wj2, unlist(id_name.wj))) 
                null <- vapply(seq_along(id_name.wj), function(i1) {
                  wj1 <- id_name.wj[[i1]]
                  message("\t", wj1)
                  a1 <- wk.fread(wj1)
                  if (a1[hs.grep(name, ", "), .N]) 
                    hs.browser("38.03.08.215646")
                  f1 <- a1[, nchar(name) %>% hs.cut(breaks, hi = 1)]
                  f1.tab <- hs.table(f1)
                  null <- vapply(seq(length(breaks) - 1), function(i2) {
                    message("\t\t", i2)
                    if (f1.tab[i2]) {
                      a1[f1 %in% levels(f1)[i2], .(name, id)] %>% 
                        wk.dt.fwrite(wj2[i2], add = 1)
                    }
                    TRUE
                  }, TRUE)
                  TRUE
                }, TRUE)
            wj2
        }
    }
}
"38_09_20_001808" <- function(name) {
}
"38_08_07_183149" <- function(related = 0) {
    for (related in 0:1) {
        wjj0.38_08_07_184747 <- "/Volumes/tf500/pubchem/cid_cas"
        if (related) 
            wjj0.38_08_07_184747 %<>% paste0("_related")
        hs.with_wd(hs.wjj(wjj0.38_08_07_184747), {
            ready.fp <- "ready_json.txt"
            repeat {
                if (file.exists(ready.fp)) 
                  break
                get.json <- function(page_i, related = 0) {
                  wj2 <- paste0("page_", page_i, ".json.gz")
                  repeat {
                    if (file.exists(wj2)) {
                      message(wj2)
                      a <- readLines(wj2)
                      json <- .try(jsonlite::fromJSON(a))
                      if ((!is(json, "try-error")) && ("Page" %in% 
                        names(json[["Annotations"]]))) {
                        break
                      }
                      else hs.wj.rm(wj2)
                    }
                    if (!file.exists(wj2)) {
                      dz1 <- if (related) {
                        paste0("https://pubchem.ncbi.nlm.nih.gov/rest/pug_view/annotations/heading/JSON/?source=CAS%20Common%20Chemistry&heading_type=Compound&heading=Related%20CAS&page=", 
                          page_i, "&response_type=save&response_basename=PubChemAnnotations_CAS%20Common%20Chemistry_heading%3DRelated%20CAS")
                      }
                      else paste0("https://pubchem.ncbi.nlm.nih.gov/rest/pug_view/annotations/heading/JSON/?source=CAS%20Common%20Chemistry&heading_type=Compound&heading=CAS&page=", 
                        page_i, "&response_type=save&response_basename=PubChemAnnotations_CAS%20Common%20Chemistry_heading%3DCAS")
                      .wk("wget -O -", .q(dz1), "| gzip >", .sys.fpGood(wj2))
                    }
                  }
                  wj2
                }
                json <- readLines(get.json(1))
                hs.for(page_i, tail(json) %>% hs.grepV("^\\s*\"TotalPages\":") %>% 
                  hs.re("\\d+") %>% as.numeric %>% seq, get.json(page_i))
                cat("", file = ready.fp)
            }
            wj <- hs.wjj.ls(".", "^page_\\d+\\.json\\.gz$", F = 0)
            wj2 <- "extracted.tsv.gz"
            if (wj.update.reporter(wj2, wj)) {
                wj1 <- wj[2]
                hs.for(wj1, wj, {
                  hs.msg(wj1)
                  a <- readLines(wj1)
                  json <- jsonlite::fromJSON(a)
                  dt <- json[["Annotations"]][["Annotation"]]
                  dt <- wk.dt.as(dt)
                  jg <- dt[, LinkedRecords %>% {
                    wk.dt(unlist(.), rep(SourceID, sapply(., 
                      length)))
                  }]
                  jg <- wk.dt.setnames(jg, c("pubchemCompoundID", 
                    "CAS_ID"))
                  wk.dt.fwrite(jg, wj2, smart = 1)
                })
            }
        })
    }
    wj <- paste0("/Volumes/tf500/pubchem/cid_cas", c("", "_related")) %>% 
        hs.fp("extracted.tsv.gz")
    if (0) {
        sj <- lapply(wj, wk.fread)
        identical(sj[[1]], sj[[2]])
    }
    dt <- wj[1] %>% wk.fread
    if (0) {
        apply(dt, 2, uniqueN)
        json <- "/Volumes/tf500/pubchem/cid_cas/page_1.json.gz" %>% 
            readLines %>% (jsonlite::fromJSON)
        a <- json[["Annotations"]][["Annotation"]] %>% wk.dt.as
        a[LinkedRecords %=>% (sapply(x, length) == 0)][2]
        x <- readLines("https://pubchem.ncbi.nlm.nih.gov/compound/57376607#section=CAS&fullscreen=true")
        dt[pubchemCompoundID %in% "57376607"]
        dt[seq(.N) %>% sample(1)]
        json <- jsonlite::fromJSON("https://pubchem.ncbi.nlm.nih.gov/rest/pug_view/data/compound/57376607/JSON/?response_type=save&response_basename=compound_CID_57376607")
        json <- jsonlite::fromJSON("https://pubchem.ncbi.nlm.nih.gov/rest/pug_view//data/compound/57376607#section=Other-Identifiers&fullscreen=true")
        os("json")
        .l(json)
        json[["Record"]][["Section"]][[2, "Section"]][[3, "Section"]][[1, 
            "Information"]][, "Value"] %>% unlist %>% unique
    }
}
"38_09_09_072128" <- function(smiles) {
    wj2 <- "~/swxx/sj/db/pubchem/smiles.cid.tsv.gz"
    hs.smiles2do <- function(smiles, wj2) {
        smiles_done <- if (file.exists(wj2)) {
            wk.fread(wj2, cn = 0, colClasses = "character")[, 
                V1]
        }
        hs.str_not(smiles, smiles_done) %>% hs.clean
    }
    smiles <- smiles2do
    smiles %<>% hs.smiles2do(wj2)
    . <- "CCCCCCCC\\C=C/CCCC(O)CC(O)=O"
    . <- "CCCCCCCC/C=C\\CCCC(CC(=O)O)O"
    jg <- smiles %>% lapply(. %>% {
        dz1 <- hs.fp("https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles", 
            curl::curl_escape(.), "cids/txt")
        dz1 <- hs.fp("https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles", 
            curl::curl_escape(hs.gsub(., "/", ".")), "cids/txt")
        a <- .wk("wget -O -", .q(dz1), intern = 1)
        if (length(a) > 1) 
            a <- paste(a, collapse = ";")
        a
    })
    names(jg) <- smiles
    names(jg)[which(sapply(jg, length) == 0)]
    "https://pubchem.ncbi.nlm.nih.gov/sdq/sdqagent.cgi?infmt=json&outfmt=csv&query={%22download%22:%22*%22,%22collection%22:%22compound%22,%22where%22:{%22ands%22:[{%22input%22:{%22type%22:%22netcachekey%22,%22idtype%22:%22cid%22,%22key%22:%22puEBjiBhRd1y883qT5KEx68jwUPW4lfKLe9Mhjb-Xoc252I%22}}]},%22order%22:[%22relevancescore,desc%22],%22start%22:1,%22limit%22:10000000,%22downloadfilename%22:%22PubChem_compound_smiles_CCCCCCCCCCCCCCCCCCOCC(COP(O)(%3DO)OCCN)OC(%3DO)CC%5C%5CC%3DC%2FC%5C%5CC%3DC%2FCCCCCCCCCC%22}"
    "https://pubchem.ncbi.nlm.nih.gov/sdq/sdqagent.cgi?infmt=json&outfmt=csv&query={%22download%22:%22*%22,%22collection%22:%22compound%22,%22where%22:{%22ands%22:[{%22input%22:{%22type%22:%22netcachekey%22,%22idtype%22:%22cid%22,%22key%22:%22byjIQagSza76hM-dTeWGsK1TTjPYwsKWuLPZ2qOiy9uju_c%22}}]},%22order%22:[%22relevancescore,desc%22],%22start%22:1,%22limit%22:10000000,%22downloadfilename%22:%22PubChem_compound_smiles_CCCCCCCCCCCCCCCCCCOCC(COP(O)(%3DO)OCCN)OC(%3DO)CC%5C%5CC%3DC%2FC%5C%5CC%3DC%2FCCCCCCCCCC%22}"
    a <- readLines("https://pubchem.ncbi.nlm.nih.gov/#query=CCCCCCCCCCCCCCCCCCOCC(COP(O)(%3DO)OCCN)OC(%3DO)CC%5C%5CC%3DC%2FC%5C%5CC%3DC%2FCCCCCCCCCC")
    smiles %>% cat(sep = "\n", file = "~/a.txt")
    "~/a.txt" %>% wk.rsync.in(ssh = "hp1")
    wj2
}
"36_12_31_171255" <- function(org1 = "human", type = "pathway", 
    out = "obj", ..mem = 1) {
    wjj1 <- hs.hsgly("生物学/数据库/文件夹.37_09_13_095732")("reactome")
    wj1 <- file.path(wjj1, switch(type, pathway = "NCBI2Reactome_All_Levels.txt.gz", 
        reaction = "NCBI2ReactomeReactions.txt.gz"))
    wj.updater(wj1, hs.wjm(wj1, dn = "https://reactome.org/download/current", 
        ext = 0), co1 = 30, gz = 1)
    wjj2 <- hs.fp(wjj1, type)
    wj2 <- hs.fp(wjj2, paste0(org1, ".rd"))
    if (out == "fp") 
        return(wj2)
    if (wj.update.reporter(wj2, wj1)) {
        sj <- wk.fread(cmd = .sys(.sys.cat.skip(wj1), "|", .sys.perl.print(paste("print if $F[ 5] eq", 
            .q(switch(org1, human = "Homo sapiens", mouse = "Mus musculus", 
                hs.browser("36.12.31.175947"))), "and $F[ 0] =~ /^[0-9]+$/")), 
            "| cut -f1,2,4"), cn = 0, quote = "")
        sj[, `:=`(V1, hs.hsgly("基因对应/更新/ncbi.36_10_15_163409")(V1, 
            org1 = org1))]
        term2gene <- sj[, .(V2, V1)] %=>% unique
        term2name <- sj[, .(V2, V3)] %=>% unique
        hs.save(term2gene, term2name, wj = wj2 %=>% hs.wj)
    }
    hs.load(wj2)
}
"38_01_05_175427" <- function(org1 = "human", extract, update = 1, 
    co1 = 500) {
    if (org1 %not.in% c("human", "mouse", "rat", "sheep", "goat")) 
        hs.browser("38.01.05.235907")
    wj2 <- hs.fp(.mfp.swxx("38_01_05_175705"), paste0(org1, ".tsv.gz"))
    ens_db_wj <- hs.hsgly("ensembl/基因/基因-蛋白.38_01_05_114445")(org1 = org1)
    wj_in <- c(hs.hsgly("string/数据.37_12_30_111108")(out = "wj"), 
        ens_db_wj, environment(hs.hsgly("编号种类转换库.37_03_29_103455"))[["ncbi.symbol"]](org1 = org1, 
            out = "wj"), environment(hs.hsgly("编号种类转换库.37_03_29_103455"))[["ncbi.ens"]](org1 = org1, 
            out = "wj"))
    if (wj.update.reporter(wj2, wj_in)) {
        string_db <- hs.hsgly("string/数据.37_12_30_111108")(org1 = org1, 
            co1 = 0)
        ens_db <- wk.fread(ens_db_wj)
        ensg.symbol <- hs.hsgly("编号种类转换库.37_03_29_103455")("ens", 
            "symbol", org1 = org1, tt_join = 0)
        ensg.symbol[gene_id_ensembl %in% "ENSCHIG00000000514"]
        ensg.symbol %<>% wk.dt.collapse("gene_id_symbol")
        c1 <- wk.dt.subset.na(ensg.symbol, ens_db[, gene], exp_j = gene_id_symbol)
        ens_db[, `:=`("gene_id_symbol", c1)]
        ensp.ensg <- ens_db[, unique(.SD)[, hs.c(protein, gene)], 
            .SDcols = c("protein", "gene")]
        if (0) {
            hs.save(ens_db, org1, wj = "~/a.rd")
            load("~/a.rd")
            ens_db[hs.clean(gene_id_symbol, inv = 1, ot = "l"), 
                .N]
        }
        a <- wk.dt.subset.na(ens_db, string_db[, protein1], "protein")[, 
            gene_id_symbol]
        string_db[, `:=`(symbol1, a)]
        a <- wk.dt.subset.na(ens_db, string_db[, protein2], "protein")[, 
            gene_id_symbol]
        string_db[, `:=`(symbol2, a)]
        string_db[, `:=`(c("protein1", "protein2"), {
        })]
        hs1 <- uid.obj("37_12_30_111108", "hs.dedup")
        a <- wk.dt.subset(string_db, exp_l = c(hs.clean(symbol1, 
            ot = "l"), hs.clean(symbol2, ot = "l"), symbol1 != 
            symbol2), exp_j = hs1(unique(.SD)))
        b <- hs1(a, cn1 = "symbol1", cn2 = "symbol2", cn3 = "combined_score")
        wk.dt.fwrite(b, wj2)
        if (0) {
            .wc(wj2)
            wk.dt.subset(a, a[, .(symbol1 = symbol2, symbol2 = symbol1)], 
                ot = "i")
        }
    }
    if (!missing(extract)) {
        if (update) 
            extract <- hs.hsgly("更新/gene.id_symbol.updater.36_09_12_235937")(extract)
        a <- if (co1 > 0) {
            cmd1 <- .sys(.sys.cat.text(wj2), "| perl -e", .q(paste0("while( <>) {\n/\\d+/;\nif( $& >= ", 
                co1, ") { print}\n}")))
            a <- wk.fread(cmd = cmd1, cn = 0)
            wk.dt.setnames(a, fp = wj2)
        }
        else wk.fread(wj2, cn = 1)
        jg <- wk.dt.subset(a, exp_l = c(tolower(symbol1) %in% 
            tolower(extract), tolower(symbol2) %in% tolower(extract), 
            if (co1 > 0) combined_score >= co1))
        return(jg)
    }
    wj2
}
"38_01_05_130005" <- function(symbol, org1 = "human", co1 = 0) {
    string_db <- hs.hsgly("string/数据.37_12_30_111108")(org1 = org1)
    ens_db_wj <- hs.hsgly("ensembl/基因/基因-蛋白.38_01_05_114445")(org1 = org1)
    ens_db <- wk.fread(ens_db_wj)
    a <- data.table(symbol = symbol)
    symbol_ensg <- hs.hsgly("dt.id1_to_id2.37_03_22_145251")(a, 
        "symbol", "ens")
    ensg <- symbol_ensg[, gene_id_ens]
    a <- merge(symbol_ensg, ens_db, by.x = "gene_id_ens", by.y = "gene")
    ensp <- a[, protein]
    wk.dt.subset(string_db, exp_l = c(protein1 %in% ensp, protein2 %in% 
        ensp, if (co1 > 0) combined_score >= co1))
}
"37_09_27_111846" <- function() {
    clinical <- "~/swxx/sj/db/tcga/rna-seq/fpkm/meta" %>% hs.wjj.ls("^clinical\\.cart\\..*\\.tar\\.gz") %>% 
        lapply(. %>% wk.untar("clinical.tsv", AO = 1) %>% {
            wk.fread(cmd = .)
        }) %>% rbindlist
    if (0) {
        clinical[1] %>% .sh
        wj1 <- hs.wjj.ls("~/swxx/sj/db/tcga/rna-seq/fpkm/meta", 
            "^clinical\\.cases_selection\\..*\\.tar\\.gz")
        clinical <- wk.fread(cmd = wk.untar(wj1, "clinical.tsv", 
            O = 1))
        a1 <- clinical[, {
            un <- sapply(.SD, uniqueN)
            if (0) {
                a1 <- hs.table(un)
                hs.cat("**** ", case_id, " ****")
                .sh(.SD[, .SD, .SDcols = c(names(un)[un > 1])])
            }
            names(un)[un > 1]
        }, by = case_id]
        a1[, V1 %>% hs.table]
    }
    clinical[, `:=`(c("treatment_or_therapy", "treatment_type"), 
        {
        })]
    clinical %<>% unique
    if (0) {
        wjm_case <- local({
            cmd1 <- .sys(.sys.cat.text(hs.wjj.ls("~/swxx/sj/db/tcga/rna-seq/fpkm/meta", 
                "files\\..*\\.json")), "| jq -r", .q(".[] | [ .file_name, ( .cases | .[] | .case_id)] | @tsv"))
            a1 <- wk.fread(cmd = cmd1, cn = 0)
            setnames(a1, c("file_name", "case_id"))
        })
        clinical <- clinical[case_id %in% wjm_case[, case_id]]
    }
    wk.dt.clean(clinical)
}
"39_02_21_115648" <- function(sub = {
}) {
    jg <- "~/swxx/sj/db/gdc/tcga"
    if (length(sub)) 
        jg <- hs.fp(jg, sub)
    jg
}
"36_07_24_165033" <- function(type1 = "median") {
    wjj0 <- .mfp.swxx("36.07.31.200909")
    wj <- hs.wjj.ls(wjj0)
    names(wj) <- hs.wjm(wj, dn = 0, ext = 0, extN = 2)
    return(fread(wj[type1]))
    hs.require("jsonlite")
    json <- paste0("~/in2/swxx/sj/db/tcga/rna-seq/htseq-counts/meta/tcga.500", 
        c("+", "-")) %>% lapply(. %>% {
        jsonlite::fromJSON(hs.wjj.ls(., "^(?i)biospecimen"))
    })
    hs.wjm(c("median", "max", "min"), add = "tsv.gz", dn = wjj0) %>% 
        .ls
    lapply(c("median", "max", "min"), function(cell_percentage.type) {
        jg_wj <- hs.wjm(cell_percentage.type, add = "tsv.gz", 
            dn = wjj0)
        if (file.exists(jg_wj)) {
        }
        else {
            tbl <- hs.hsgly("yb,sage1_max.info.36.07.06.130805")()
            hs1 <- switch(cell_percentage.type, median = colMedians, 
                max = colMaxs, min = colMins)
            lapply(json, function(json) {
                slides.cn.db <- local({
                  a1 <- json[, "samples"] %>% lapply(. %>% {
                    unique(do.call("c", .[, "portions"] %>% lapply(. %>% 
                      {
                        if ("slides" %in% colnames(.)) {
                          unique(do.call("c", lapply(.[, "slides"], 
                            colnames)))
                        }
                      })))
                  })
                  unique(unlist(a1))
                })
                slides.cn.db %<>% sort
                i1 <- which(json[, "case_id"] %in% tbl[, case_id])
                a1 <- json[i1, "samples"] %>% lapply(. %>% {
                  a1 <- .[, "portions"] %>% lapply(. %>% {
                    if ("slides" %in% colnames(.)) {
                      a1 <- lapply(.[, "slides"], . %>% {
                        if (length(.)) {
                          cn1 <- slides.cn.db %and% names(.)
                          cn2 <- slides.cn.db %not% names(.)
                          jg <- .[, cn1]
                          if (length(cn2)) {
                            a1 <- as.data.frame(matrix(NA, nrow(jg), 
                              length(cn2), dimnames = list(NULL, 
                                cn2)))
                            cbind(jg, a1)[, slides.cn.db]
                          }
                          else jg
                        }
                      })
                      rbindlist(a1)
                    }
                  })
                  rbindlist(a1)
                })
                names(a1) <- tbl[match(json[i1, "case_id"], case_id), 
                  sample_id]
                a2 <- hs.lapply(names(a1), . %>% {
                  a2 <- a1[[.]]
                  if (length(a2)) {
                    a2 <- a2[hs.grepl(submitter_id, paste0("^", 
                      .))]
                    a2[, `:=`("sample_id", .)]
                  }
                })
                a3 <- rbindlist(a2)
                a4 <- a3[, {
                  if (.N > 1) {
                    a1 <- hs1(as.matrix(.SD), na.rm = TRUE)
                    a1[a1 %in% c(NaN, Inf, -Inf)] <- NA
                    as.data.table(t(hs.c(cn_numeric, a1)))
                  }
                  else {
                    .SD
                  }
                }, by = sample_id, .SDcols = c(cn_numeric)]
                tbl[match(a4[, sample_id], sample_id), `:=`(names(a4), 
                  a4)]
            })
            wk.dt.fwrite(tbl, jg_wj)
        }
    })
    if (0) {
        a3[, {
            if (.N > 1) {
                .N
            }
        }, by = sample_id][, .N]
        cn_numeric <- hs.grepV(names(a3), "(?i)(percent|number)")
        1 - colMeans(a3[, is.na(.SD), .SDcols = c(cn_numeric)])
        cn_numeric %<>% {
            .[!apply(a3[, is.na(.SD), .SDcols = c(.)], 2, all)]
        }
        {
            tbl[1]
            cn1 <- "percent_lymphocyte_infiltration"
            i1 <- tbl[, which((!is.na(percent_lymphocyte_infiltration)) & 
                (tpm_ENSG00000181433 > 0))]
            a1 <- tbl[i1, percent_lymphocyte_infiltration]
            sj <- tapply(a1, f1, list)
            cn <- hs.grepV(names(tbl), "percent|number")
            jg <- sapply(cn, . %>% {
                i1 <- tbl[, which((!is.na(get(.))) & (tpm_ENSG00000181433 > 
                  0))]
                tbl[i1, c(.N, cor.test(get(.), tpm_ENSG00000181433, 
                  method = "spearman")[c("p.value", "estimate")] %>% 
                  unlist)]
            }) %>% t %>% as.data.table("cell_percentage")
            setnames(jg, "V1", "sample_size")
            setorderv(jg, "estimate.rho", -1)
            jg[, `:=`(qv, p.adjust(p.value, "fdr"))]
            .s(jg)
            tbl[!is.na(percent_lymphocyte_infiltration), cor.test(percent_lymphocyte_infiltration, 
                tpm_ENSG00000181433, method = "spearman")]
        }
    }
    if (0) {
        a1 <- json[, "samples"] %>% lapply(. %>% {
            .[, "portions"] %>% lapply(. %>% {
                if ("slides" %in% colnames(.)) 
                  length(.[, "slides"])
            })
        })
        a1[[4]]
        sapply(a1, length) %>% hs.table
        a2 <- sapply(a1, . %>% {
            max(unlist(.))
        })
        hs.table(a2)
        length(a2)
        a1[a2 == 30]
        json[a2 == 30, "samples"][[1]][2, "portions"][[1]][, 
            "slides"]
        a1 <- json[1, "samples"][[1]]
        a1[, "submitter_id"]
        a1[2, "portions"][[1]][, "slides"]
    }
    local({
        i1 <- which(json[, "case_id"] %in% tbl[, case_id])
        a1 <- json[i1, "samples"] %>% lapply(. %>% {
            a1 <- .[, "portions"] %>% lapply(. %>% {
                if ("slides" %in% colnames(.)) {
                  a1 <- lapply(.[, "slides"], . %>% {
                    if (length(.)) {
                      cn1 <- slides.cn.db %and% names(.)
                      cn2 <- slides.cn.db %not% names(.)
                      jg <- .[, cn1]
                      if (length(cn2)) {
                        a1 <- as.data.frame(matrix(NA, nrow(jg), 
                          length(cn2), dimnames = list(NULL, 
                            cn2)))
                        cbind(jg, a1)[, slides.cn.db]
                      }
                      else jg
                    }
                  })
                  rbindlist(a1)
                }
            })
            rbindlist(a1)
        })
        names(a1) <- tbl[match(json[i1, "case_id"], case_id), 
            sample_id]
        a2 <- hs.lapply(names(a1), . %>% {
            a2 <- a1[[.]]
            if (length(a2)) {
                a2 <- a2[hs.grepl(submitter_id, paste0("^", .))]
                a2[, `:=`("sample_id", .)]
            }
        })
        a3 <- rbindlist(a2)
        a3[, {
            if (.N > 1) {
                .N
            }
        }, by = sample_id]
        cn_numeric <- hs.grepV(names(a3), "(?i)(percent|number)")
        1 - colMeans(a3[, is.na(.SD), .SDcols = c(cn_numeric)])
        cn_numeric %<>% {
            .[!apply(a3[, is.na(.SD), .SDcols = c(.)], 2, all)]
        }
        a4 <- a3[, {
            if (.N > 1) {
                a1 <- colMedians(as.matrix(.SD), na.rm = TRUE)
                a1[is.nan(a1)] <- NA
                as.data.table(t(hs.c(cn_numeric, a1)))
            }
            else {
                .SD
            }
        }, by = sample_id, .SDcols = c(cn_numeric)]
        1 - colMeans(a3[, is.na(.SD), .SDcols = c(cn_numeric)])
        a3[sample_id %in% "TCGA-CH-5769-01A"]
        a3[, `:=`("sample_id", names(a2))]
        tbl[1]
        sapply(a1, length) %>% hs.table
        a1[1]
        a1 <- hs.nonempty(a1)
        a1[1]
        unique(unlist(a1))
    })
    json[1, "samples"][[1]][, "submitter_id"]
    json[1, "samples"][[1]][, "portions"] %>% .sh
    json[1, "samples"][[1]][2, "portions"][[1]][, "slides"]
    json[1, "samples"][[1]][1, "portions"][[1]][, "slides"]
    json[1, "samples"][[1]][3, "portions"][[1]][, "slides"][[1]]
    {
        for (case_id1 in json[, "case_id"] %and% tbl[, case_id]) {
            case_id1.sample_id1 <- tbl[case_id %in% case_id1, 
                sample_id]
            match(case_id1, json[, "case_id"])
            a1 <- json[match(case_id1, json[, "case_id"]), "samples"][[1]] %>% 
                {
                  .[.[, "submitter_id"] == case_id1.sample_id1, 
                    ]
                }
            hs.cat(nrow(a1))
            if (nrow(a1) != 1) 
                hs.browser("36.07.27.115320")
        }
        wjm.yb[case_id %in% case_id1]
        tbl <- hs.hsgly("yb,sage1_max.info.36.07.06.130805")()
        a1[, "portions"][[1]] %>% .sh
    }
    case.slides <- local({
        a1 <- json[, "samples"] %>% lapply(. %>% {
            .[, "portions"] %>% lapply(. %>% {
                if ("slides" %in% colnames(.)) {
                  hs.browser("36.07.24.181931")
                  .[, "slides"]
                  dim(.)
                  .[2, ]
                  .sh(.)
                  unique(do.call("c", lapply(.[, "slides"], colnames)))
                }
            })
        })
        unique(unlist(a1))
    })
    "~/in2/swxx/sj/db/tcga/rna-seq/htseq-counts/meta/tcga.500+" %>% 
        .ls
}
"39_02_21_120004" <- function(ct = {
}, ot = "tpm", log = 1, legacy = FALSE, gene_id_type = "symbol") {
    wjj2 <- hs.sjgly("39_02_21_120004")
    hs.require("TCGAbiolinks")
    hs.switch(length(ct), 0, return(TCGAbiolinks::getGDCprojects()[["project_id"]] %=>% 
        hs.grepV(x, "^(?i)tcga") %=>% hs.sub(x, "(?i)tcga-") %=>% 
        sort), {
        ct %<=>% toupper
        if (0) {
            query <- (TCGAbiolinks::GDCquery)(project = paste0("TCGA-", 
                ct), data.category = hs.switch(legacy, TRUE, 
                "Gene expression", FALSE, "Transcriptome Profiling"), 
                data.type = hs.switch(legacy, TRUE, "Gene expression quantification", 
                  FALSE, "Gene Expression Quantification"), platform = "Illumina HiSeq", 
                file.type = "normalized_results", experimental.strategy = "RNA-Seq", 
                barcode = c("TCGA-14-0736-02A-01R-2005-01", "TCGA-06-0211-02A-02R-2005-01"), 
                legacy = FALSE)
        }
        wjj2_here <- hs.fp(wjj2, paste0("TCGA-", ct), "harmonized", 
            "Transcriptome_Profiling", "Gene_Expression_Quantification")
        if (!dir.exists(wjj2_here)) {
            query <- (TCGAbiolinks::GDCquery)(project = paste0("TCGA-", 
                ct), data.category = hs.switch(legacy, TRUE, 
                "Gene expression", FALSE, "Transcriptome Profiling"), 
                data.type = hs.switch(legacy, TRUE, "Gene expression quantification", 
                  FALSE, "Gene Expression Quantification"), workflow.type = "STAR - Counts")
            si.fp <- hs.fp(wjj2_here, "si.tsv.gz")
            si <- query[, "results"][[1]]
            wk.dt.fwrite(si, si.fp)
            (TCGAbiolinks::GDCdownload)(query = query, method = "api", 
                files.per.chunk = 10, directory = wjj2)
            wjj2_here %=>% hs.wjj.ls(x, "(?i)\\.tsv$", R = 1) %=>% 
                for (wj1 in x) .wk("gzip", .sys.fpGood(wj1))
        }
        si_hs <- function() hs.fp(wjj2_here, "si.tsv.gz") %=>% 
            wk.fread
        hs.switch(ot, "wjj", wjj2_here, "si", si_hs(), "tpm", 
            {
                si <- si_hs()
                wj <- si[, hs.fp(wjj2_here, file_id, paste0(file_name, 
                  ".gz"))]
                fp2cn <- function(fp1) fp1 %=>% basename %=>% 
                  hs.sub(x, "\\.gz$") %=>% si[match(x, file_name), 
                  cases]
                jg <- local({
                  a <- wj[1] %=>% readLines %=>% x[-c(1, 3:6)] %=>% 
                    wk.fread(text = x, cn = 1)
                  if (0) {
                    wk.dt.setorderv(a, "tpm_unstranded", -1)
                    a[gene_id %=>% hs.grep(x, "\\.\\d+$", inv = 1), 
                      .(gene_id, gene_name, gene_type)] %=>% 
                      .s
                    a[gene_name %=>% hs.grep(x, "\\.\\d+$")][gene_type %in% 
                      "protein_coding"]
                  }
                  a1 <- a[, .(gene_id)]
                  wk.dt.set(a1, , fp2cn(wj[1]), a[, tpm_unstranded %=>% 
                    if (log) 
                      log1p(x)
                    else x])
                })
                if (length(wj) > 1) 
                  for (i in 2:length(wj)) {
                    a <- wj[i] %=>% readLines %=>% x[-c(1, 3:6)] %=>% 
                      wk.fread(text = x, cn = 1)
                    if (!all(a[, gene_id] == jg[, gene_id])) 
                      hs.browser("39.02.22.084210", debug = 0)
                    wk.dt.set(jg, , fp2cn(wj[i]), a[, tpm_unstranded %=>% 
                      if (log) 
                        log1p(x)
                      else x])
                  }
                hs.switch(gene_id_type, "symbol", {
                  hs.hsgly("dt.id1_to_id2.37_03_22_145251")(jg, 
                    id1 = "ens", id2 = "symbol", id1.cn = "gene_id")
                  jg[hs.clean(gene_id_symbol, inv = 1, ot = "i"), 
                    `:=`(gene_id_symbol, gene_id_updated)]
                  jg[, `:=`(c("gene_id", "gene_id_updated"), 
                    {
                    })]
                }, "ens", {
                  jg[, `:=`(gene_id, hs.hsgly("基因对应/更新/ensembl.36_12_20_235632")(gene_id))]
                  if (0) {
                    jg[hs.sub(gene_id, "\\.[^\\.]+$") != gene_id_updated] %=>% 
                      {
                        cbind(x[, gene_id_symbol], a[match(x[, 
                          gene_id], gene_id), gene_name]) %=>% 
                          .sh
                      }
                  }
                }, hs.browser("39.02.22.200125", debug = 0))
                return(jg)
            }, hs.browser("39.03.06.180419", debug = 0))
    })
}
"39_02_22_135805" <- function(ct = "acc") {
    wjj2 <- hs.sjgly("39_02_22_135805")
    ct %<=>% toupper
    wj2 <- hs.fp(wjj2, paste0(ct, ".tsv.gz"))
    if (!file.exists(wj2)) {
        query <- TCGAbiolinks::GDCquery_clinic(project = paste0("TCGA-", 
            ct), type = "clinical")
        wk.dt.fwrite(query, wj2)
    }
    wj2
}
"36_01_11_201948" <- function(genes, exp.type = "counts_normalized") {
    wjj.36_03_04_003344 <- getwd()
    on.exit(setwd(wjj.36_03_04_003344), add = TRUE)
    setwd("~/in2/swxx/sj/db/tcga/rna-seq/htseq-counts")
    projects <- hs.hsgly("tcga..project.sample_szie.36.03.03.234542")()
    exp.type <- switch(exp.type, counts_raw = "counts", counts_normalized = "counts_normalized", 
        vsd = "vsd")
    hs.c(projects[, prj], lapply(projects[, prj], function(prj1) {
        hs.say(prj1)
        exp.wj <- hs.wj("DESeq2", "exp", paste0(prj1, ".", exp.type, 
            ".tsv.gz"))
        exp <- exp.wj %>% fread
        wk.dt.subset(exp, genes, "gene.id_ensembl")
    }))
}
"36_06_28_143030" <- function(..mem = 1) {
    jg_wjj <- .mfp.swxx("36.06.28.143122") %=>% hs.wjj
    wk.fread(hs.fp(jg_wjj, "sample_id.info.tsv.gz"))
}
"36_07_17_153324" <- function(org1 = "human", ..mem = 1) {
    wj_in <- paste0("~/swxx/sj/db/uniprot/idmapping/", switch(org1, 
        human = "HUMAN_9606", mouse = "MOUSE_10090"), "_idmapping.dat.gz")
    wj_jg <- .mfp.swxx(paste0(switch(org1, human = "36.08.03.225621", 
        mouse = "36_12_06_212216"), ".tsv.gz"))
    if (file.exists(wj_jg) && (file.mtime(wj_jg) >= file.mtime(wj_in))) {
        fread(wj_jg)
    }
    else {
        db.name.V <- c("Ensembl", "GeneID", "Gene_Name", "UniProtKB-ID")
        sj <- fread(wj_in, header = FALSE)
        if (0) {
            wj_in %>% .less
            "O88942"
            sj[(V1 %in% "Q62230") & (V2 %in% "Ensembl")]
        }
        jg <- sj[V2 %in% db.name.V, {
            a1 <- as.list(tapply(V3[i1], V2[i1], unique))
            a1 <- hs.c(db.name.V, a1[db.name.V])
            a1 <- lapply(a1, function(a1) if (length(a1)) 
                a1
            else "")
            do.call("expand.grid", c(a1, stringsAsFactors = FALSE))
        }, by = V1]
        setnames(jg, c("gene.id_uniprot", "gene.id_ensembl", 
            "gene.id_ncbi", "gene.id_symbol", "gene.id_uniprot_2"))
        id.description <- fread(hs.hsgly("uniprot.gene.anno.36.06.20.224138")(org1 = org1))
        wk.dt.absorb1(jg, id.description, by1 = "gene.id_uniprot_2", 
            by2 = "uniprot_id")
        wk.dt.fwrite(jg, wj_jg %>% hs.wj)
        jg
    }
}
